// vue-data-ui/dist/BaseZoomControls-CoFievy9.js
import { openBlock as c, createElementBlock as a, normalizeStyle as e, unref as d, normalizeClass as m, createElementVNode as l, createVNode as s, toDisplayString as f, createCommentVNode as z } from "vue";
import { l as y } from "./exposedLib-H4lmBl1t.js";
import i from "./BaseIcon-CNx8Tzl5.js";
import { _ as b } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const g = ["disabled"], C = {
  __name: "BaseZoomControls",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    scale: {
      type: Number,
      default: 0
    },
    withDirection: {
      type: Boolean,
      default: !1
    },
    isFullscreen: {
      type: Boolean,
      default: !1
    },
    isCursorPointer: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["zoomIn", "zoomOut", "resetZoom", "switchDirection"],
  setup(o, { emit: u }) {
    const n = u;
    return (h, t) => (c(), a("div", {
      class: m({ "vue-data-ui-zoom-controls": !0, "vue-data-ui-zoom-controls-fullscreen": o.isFullscreen }),
      "data-dom-to-png-ignore": "",
      style: e({
        border: o.config.style.chart.controls.border,
        backgroundColor: o.config.style.chart.controls.backgroundColor,
        padding: o.config.style.chart.controls.padding,
        borderRadius: o.config.style.chart.controls.borderRadius,
        "--vue-data-ui-zoom-control-button-color": o.config.style.chart.controls.buttonColor,
        "--vue-data-ui-zoom-control-button-color-hover": d(y)(o.config.style.chart.controls.buttonColor, 0.2)
      })
    }, [
      l("button", {
        onClick: t[0] || (t[0] = (r) => n("zoomOut")),
        class: "vue-data-ui-zoom-controls-button",
        "data-cy-zoom-out": "",
        style: e({ cursor: o.isCursorPointer ? "pointer" : "default" })
      }, [
        s(i, {
          name: "zoomMinus",
          stroke: o.config.style.chart.controls.color,
          size: o.config.style.chart.controls.fontSize * 1.2
        }, null, 8, ["stroke", "size"])
      ], 4),
      l("button", {
        class: "vue-data-ui-zoom-controls-button-zoom",
        onClick: t[1] || (t[1] = (r) => n("resetZoom")),
        "data-cy-zoom-reset": "",
        style: e({
          color: o.config.style.chart.controls.color,
          width: o.config.style.chart.controls.fontSize * 4 + "px",
          borderRadius: o.config.style.chart.controls.borderRadius,
          fontSize: o.config.style.chart.controls.fontSize + "px",
          cursor: o.isCursorPointer ? "pointer" : "default"
        })
      }, f(Math.round(o.scale * 100)) + "% ", 5),
      l("button", {
        onClick: t[2] || (t[2] = (r) => n("zoomIn")),
        class: "vue-data-ui-zoom-controls-button",
        "data-cy-zoom-in": "",
        style: e({
          cursor: o.isCursorPointer ? "pointer" : "default"
        })
      }, [
        s(i, {
          name: "zoomPlus",
          stroke: o.config.style.chart.controls.color,
          size: o.config.style.chart.controls.fontSize * 1.2
        }, null, 8, ["stroke", "size"])
      ], 4),
      l("button", {
        disabled: o.scale === 1,
        onClick: t[3] || (t[3] = (r) => n("resetZoom")),
        class: "vue-data-ui-zoom-controls-button",
        style: e({
          cursor: o.isCursorPointer ? "pointer" : "default"
        })
      }, [
        s(i, {
          name: "revert",
          stroke: o.config.style.chart.controls.color,
          size: o.config.style.chart.controls.fontSize * 1.2
        }, null, 8, ["stroke", "size"])
      ], 12, g),
      o.withDirection ? (c(), a("button", {
        key: 0,
        onClick: t[4] || (t[4] = (r) => n("switchDirection")),
        class: "vue-data-ui-zoom-controls-button",
        style: e({
          cursor: o.isCursorPointer ? "pointer" : "default"
        })
      }, [
        s(i, {
          name: "direction",
          stroke: o.config.style.chart.controls.color,
          size: o.config.style.chart.controls.fontSize * 1.2
        }, null, 8, ["stroke", "size"])
      ], 4)) : z("", !0)
    ], 6));
  }
}, P = /* @__PURE__ */ b(C, [["__scopeId", "data-v-f6b8b4d1"]]);
export {
  P as B
};
