// vue-data-ui/dist/BaseLegendToggle-Y61xahI4.js
import { openBlock as i, createElementBlock as l, normalizeClass as a, createElementVNode as s, normalizeStyle as d, withKeys as g, createVNode as c } from "vue";
import m from "./BaseIcon-CNx8Tzl5.js";
import { _ as u } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const C = {
  __name: "BaseLegendToggle",
  props: {
    backgroundColor: { type: String, default: "#CCCCCC" },
    color: { type: String, default: "#2D353C" },
    fontSize: { type: Number, default: 14 },
    checked: { type: Boolean },
    isCursorPointer: { type: Boolean }
  },
  emits: ["toggle"],
  setup(e, { emit: n }) {
    const o = n;
    return (f, t) => (i(), l("div", {
      class: a({ "vue-ui-legend-toggle-wrapper": e.isCursorPointer }),
      "data-dom-to-png-ignore": ""
    }, [
      s("div", {
        class: "vue-ui-legend-toggle",
        role: "button",
        tabindex: "0",
        onClick: t[0] || (t[0] = (r) => o("toggle")),
        onKeydown: t[1] || (t[1] = g((r) => o("toggle"), ["enter"])),
        style: d({
          position: "relative",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          backgroundColor: e.backgroundColor,
          padding: e.fontSize / 4 + "px"
        })
      }, [
        c(m, {
          name: e.checked ? "minus" : "check",
          stroke: e.color,
          size: e.fontSize * 0.6,
          "stroke-width": e.fontSize / 4
        }, null, 8, ["name", "stroke", "size", "stroke-width"])
      ], 36)
    ], 2));
  }
}, v = /* @__PURE__ */ u(C, [["__scopeId", "data-v-7d3908a2"]]);
export {
  v as B
};
