// vue-data-ui/dist/ColorPicker-j0ySewhI.js
import { ref as u, computed as T, watchEffect as U, onMounted as W, onBeforeUnmount as Y, watch as j, withDirectives as q, openBlock as s, createElementBlock as v, withKeys as A, createElementVNode as i, normalizeStyle as a, renderSlot as G, normalizeProps as H, guardReactiveProps as J, createVNode as b, withModifiers as n, Fragment as R, renderList as M, createCommentVNode as N, createBlock as Q, Teleport as X, unref as Z, nextTick as _ } from "vue";
import g from "./BaseIcon-CNx8Tzl5.js";
import { W as ee } from "./lib-GwPnJSrY.js";
import { v as oe } from "./vClickOutside-C6WiFswA.js";
import { _ as te } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ne = ["onClick"], re = { style: { position: "absolute", top: "50%", left: "50%", transform: "translate(-50%,-46%)" } }, le = ["value"], ie = ["onClick"], ue = { style: { position: "absolute", top: "50%", left: "50%", transform: "translate(-50%,-46%)" } }, se = ["value"], ae = {
  __name: "ColorPicker",
  props: {
    value: {
      type: String,
      default: "#ffffff"
    },
    size: {
      type: String,
      default: "50px"
    },
    backgroundColor: {
      type: String,
      default: "#FFFFFF"
    },
    buttonBorderColor: {
      type: String,
      default: "#FFFFFF"
    },
    teleported: {
      type: Boolean,
      default: !1
    },
    isCursorPointer: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["update:value"],
  setup(o, { emit: $ }) {
    const c = o, F = $, O = u(null), C = u(null), d = u(null), l = u(!1), m = u(!1), y = u({ top: 0, left: 0 }), V = T(() => ({
      backgroundColor: c.value,
      width: "100%",
      height: "100%",
      cursor: c.isCursorPointer ? "pointer" : "default"
    })), f = T(() => ee(c.value));
    function w(t) {
      F("update:value", t), l.value = !1;
    }
    function E(t) {
      F("update:value", t.target.value);
    }
    function P(t) {
      t?.stopPropagation?.(), m.value = !0, d.value?.click();
    }
    function D() {
      l.value = !1;
    }
    function L() {
      !m.value && l.value && D();
    }
    async function K() {
      l.value = !l.value, l.value && c.teleported && (await _(), x());
    }
    function x() {
      if (!C.value) return;
      const t = C.value.getBoundingClientRect();
      y.value = { top: t.top + 36, left: t.right - 48 };
    }
    function p() {
      l.value && c.teleported && x();
    }
    function k() {
      setTimeout(() => m.value = !1, 0);
    }
    U((t) => {
      const e = d.value;
      if (!e) return;
      const r = () => k(), I = () => k(), S = () => {
      };
      e.addEventListener("blur", r), e.addEventListener("change", I), e.addEventListener("input", S), t(() => {
        e.removeEventListener("blur", r), e.removeEventListener("change", I), e.removeEventListener("input", S);
      });
    });
    function B() {
      k();
    }
    function h() {
      document.visibilityState === "visible" && k();
    }
    W(() => {
      window.addEventListener("scroll", p, { passive: !0 }), window.addEventListener("resize", p, { passive: !0 }), window.addEventListener("focus", B), document.addEventListener("visibilitychange", h);
    }), Y(() => {
      window.removeEventListener("scroll", p), window.removeEventListener("resize", p), window.removeEventListener("focus", B), document.removeEventListener("visibilitychange", h);
    }), j(
      () => c.value,
      (t) => {
        d.value && (d.value.value = t);
      }
    );
    const z = u([
      "#000000",
      "#FFFFFF",
      "#FF5733",
      "#33FF57",
      "#3357FF",
      "#FFC300",
      "#800080",
      "#FF1493",
      "#00CED1"
    ]);
    return (t, e) => q((s(), v("div", {
      ref_key: "wrapperRef",
      ref: O,
      onKeydown: A(L, ["esc"]),
      style: { height: "100%", width: "100%", position: "relative" }
    }, [
      i("button", {
        ref_key: "buttonRef",
        ref: C,
        class: "icon",
        onClick: K,
        style: a(V.value),
        type: "button"
      }, [
        G(t.$slots, "annotator-action-color", H(J({ color: f.value })), () => [
          b(g, {
            name: "palette",
            stroke: f.value,
            size: 22
          }, null, 8, ["stroke"])
        ], !0)
      ], 4),
      l.value && !o.teleported ? (s(), v("div", {
        key: 0,
        tabindex: "0",
        class: "vue-ui-color-picker",
        style: a({
          backgroundColor: o.backgroundColor,
          position: "absolute",
          left: "calc(100% + 30px)",
          top: "50%",
          transform: "translateY(-18%)"
        }),
        onMousedown: e[2] || (e[2] = n(() => {
        }, ["stop"])),
        onClick: e[3] || (e[3] = n(() => {
        }, ["stop"])),
        onTouchstart: e[4] || (e[4] = n(() => {
        }, ["stop"]))
      }, [
        (s(!0), v(R, null, M(z.value, (r) => (s(), v("button", {
          key: r,
          class: "vue-ui-color-picker-option",
          type: "button",
          style: a({ backgroundColor: r, outline: `1px solid ${o.buttonBorderColor}`, cursor: o.isCursorPointer ? "pointer" : "default" }),
          onClick: () => w(r)
        }, null, 12, ne))), 128)),
        i("button", {
          class: "vue-ui-color-picker-option",
          type: "button",
          style: a({ backgroundColor: o.value, outline: `1px solid ${o.buttonBorderColor}`, cursor: o.isCursorPointer ? "pointer" : "default" }),
          onClick: n(P, ["stop"]),
          onMousedown: e[0] || (e[0] = n(() => {
          }, ["stop"])),
          onTouchstart: e[1] || (e[1] = n(() => {
          }, ["stop"]))
        }, [
          i("div", re, [
            b(g, {
              name: "colorPicker",
              stroke: f.value,
              size: 22
            }, null, 8, ["stroke"])
          ]),
          i("input", {
            ref_key: "colorInput",
            ref: d,
            type: "color",
            value: o.value,
            class: "hidden-input",
            onInput: E
          }, null, 40, le)
        ], 36)
      ], 36)) : N("", !0),
      l.value && o.teleported ? (s(), Q(X, {
        key: 1,
        to: "body"
      }, [
        i("div", {
          tabindex: "0",
          class: "vue-ui-color-picker",
          style: a({
            backgroundColor: o.backgroundColor,
            position: "fixed",
            top: y.value.top + "px",
            left: y.value.left + "px",
            zIndex: 2147483647
          }),
          onMousedown: e[7] || (e[7] = n(() => {
          }, ["stop"])),
          onClick: e[8] || (e[8] = n(() => {
          }, ["stop"])),
          onTouchstart: e[9] || (e[9] = n(() => {
          }, ["stop"]))
        }, [
          (s(!0), v(R, null, M(z.value, (r) => (s(), v("button", {
            key: r,
            class: "vue-ui-color-picker-option",
            type: "button",
            style: a({ backgroundColor: r, outline: `1px solid ${o.buttonBorderColor}`, cursor: o.isCursorPointer ? "pointer" : "default" }),
            onClick: () => w(r)
          }, null, 12, ie))), 128)),
          i("button", {
            class: "vue-ui-color-picker-option",
            type: "button",
            style: a({ backgroundColor: o.value, outline: `1px solid ${o.buttonBorderColor}`, cursor: o.isCursorPointer ? "cursor" : "default" }),
            onClick: n(P, ["stop"]),
            onMousedown: e[5] || (e[5] = n(() => {
            }, ["stop"])),
            onTouchstart: e[6] || (e[6] = n(() => {
            }, ["stop"]))
          }, [
            i("div", ue, [
              b(g, {
                name: "colorPicker",
                stroke: f.value,
                size: 22
              }, null, 8, ["stroke"])
            ]),
            i("input", {
              ref_key: "colorInput",
              ref: d,
              type: "color",
              value: o.value,
              class: "hidden-input",
              onInput: E
            }, null, 40, se)
          ], 36)
        ], 36)
      ])) : N("", !0)
    ], 32)), [
      [Z(oe), L]
    ]);
  }
}, ke = /* @__PURE__ */ te(ae, [["__scopeId", "data-v-5fdacedc"]]);
export {
  ke as C
};
