// vue-data-ui/dist/DataTable-Bo1pmAqG.js
import { useCssVars as A, unref as t, computed as D, ref as k, onMounted as E, openBlock as o, createElementBlock as l, normalizeClass as C, normalizeStyle as r, withKeys as R, createVNode as B, createCommentVNode as m, createElementVNode as a, toDisplayString as I, Fragment as g, renderList as v, renderSlot as $ } from "vue";
import { _ as K } from "./Shape-CSyWYgXc.js";
import O from "./BaseIcon-CNx8Tzl5.js";
import { _ as P } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const T = { class: "vue-ui-data-table" }, j = { style: { display: "flex", "align-items": "center", "justify-content": "flex-end", "padding-right": "3px", gap: "3px" } }, F = { style: { width: "12px", height: "12px" } }, L = {
  key: 0,
  height: "12",
  width: "12",
  viewBox: "0 0 20 20",
  style: { background: "none" }
}, M = ["fill"], q = ["data-cell"], G = {
  dir: "auto",
  style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" }
}, H = {
  key: 0,
  height: "12",
  width: "12",
  viewBox: "0 0 20 20",
  style: { background: "none", overflow: "visible" }
}, J = {
  __name: "DataTable",
  props: {
    colNames: {
      type: Array,
      default() {
        return [];
      }
    },
    head: Array,
    body: Array,
    title: String,
    config: Object,
    withCloseButton: { type: Boolean, default: !0 },
    isCursorPointer: { type: Boolean, default: !1 }
  },
  emits: ["close"],
  setup(e, { emit: N }) {
    A((s) => ({
      v4faaafda: t(_)
    }));
    const f = e, { backgroundColor: h, color: y, outline: p } = f.config.th, { backgroundColor: z, color: S, outline: _ } = f.config.td, V = D(() => f.config.breakpoint), b = k(null), w = k(!1);
    E(() => {
      const s = new ResizeObserver((i) => {
        i.forEach((n) => {
          w.value = n.contentRect.width < V.value;
        });
      });
      b.value && s.observe(b.value);
    });
    const x = N;
    return (s, i) => (o(), l("div", {
      ref_key: "tableContainer",
      ref: b,
      style: r(`width: 100%; container-type: inline-size; position:relative;${e.withCloseButton ? "padding-top: 36px;" : ""}overflow:auto`),
      class: C({
        "atom-data-table": !0,
        "vue-ui-responsive": w.value
      })
    }, [
      e.withCloseButton ? (o(), l("div", {
        key: 0,
        "data-dom-to-png-ignore": "",
        role: "button",
        tabindex: "0",
        style: r(`width:32px; position: absolute; top: 0; right:4px; padding: 0 0px; display: flex; align-items:center;justify-content:center;height: 36px; width: 32px; cursor:${e.isCursorPointer ? "pointer" : "default"}; background:${t(h)};`),
        onClick: i[0] || (i[0] = (n) => x("close")),
        onKeypress: i[1] || (i[1] = R((n) => x("close"), ["enter"]))
      }, [
        B(O, {
          name: "close",
          stroke: t(y),
          "stroke-width": 2
        }, null, 8, ["stroke"])
      ], 36)) : m("", !0),
      a("table", T, [
        a("caption", {
          style: r({ backgroundColor: t(h), color: t(y), outline: t(p) }),
          class: "vue-ui-data-table__caption"
        }, I(e.title), 5),
        a("thead", null, [
          a("tr", {
            role: "row",
            style: r([{ "font-variant-numeric": "tabular-nums" }, { backgroundColor: t(h), color: t(y) }]),
            class: "vue-ui-data-table__thead-row"
          }, [
            (o(!0), l(g, null, v(e.head, (n, d) => (o(), l("th", {
              role: "cell",
              style: r({
                outline: t(p)
              }),
              key: `th_${d}`
            }, [
              a("div", j, [
                a("div", F, [
                  n.color ? (o(), l("svg", L, [
                    a("circle", {
                      cx: "10",
                      cy: "10",
                      r: "10",
                      fill: n.color
                    }, null, 8, M)
                  ])) : m("", !0)
                ]),
                $(s.$slots, "th", { th: n }, void 0, !0)
              ])
            ], 4))), 128))
          ], 4)
        ]),
        a("tbody", null, [
          (o(!0), l(g, null, v(e.body, (n, d) => (o(), l("tr", {
            role: "row",
            style: r([{ "font-variant-numeric": "tabular-nums" }, { backgroundColor: t(z), color: t(S) }]),
            class: C({ "vue-ui-data-table__tbody__row": !0, "vue-ui-data-table__tbody__row-even": d % 2 === 0, "vue-ui-data-table__tbody__row-odd": d % 2 !== 0 })
          }, [
            (o(!0), l(g, null, v(n, (c, u) => (o(), l("td", {
              role: "cell",
              "data-cell": (e.colNames[u] && e.colNames[u].name ? e.colNames[u].name : "") || e.colNames[u] || "",
              style: r({ outline: t(_) }),
              class: "vue-ui-data-table__tbody__td"
            }, [
              a("div", G, [
                c.color ? (o(), l("svg", H, [
                  B(K, {
                    plot: { x: 10, y: 10 },
                    color: c.color,
                    radius: 9,
                    shape: e.config.shape || c.shape || "circle"
                  }, null, 8, ["color", "shape"])
                ])) : m("", !0),
                $(s.$slots, "td", { td: c }, void 0, !0)
              ])
            ], 12, q))), 256))
          ], 6))), 256))
        ])
      ])
    ], 6));
  }
}, Y = /* @__PURE__ */ P(J, [["__scopeId", "data-v-a9965cf9"]]);
export {
  Y as default
};
