// vue-data-ui/dist/Legend-1LMMsPxL.js
import { openBlock as n, createElementBlock as a, normalizeStyle as d, renderSlot as r, Fragment as f, renderList as m, normalizeClass as g, createVNode as p, mergeProps as h, unref as k, createCommentVNode as y } from "vue";
import { a8 as v } from "./lib-GwPnJSrY.js";
import { _ as b } from "./Shape-CSyWYgXc.js";
import { _ as x } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const S = ["id"], $ = ["onClick", "viewBox"], B = {
  __name: "Legend",
  props: {
    legendSet: {
      type: Array,
      default() {
        return [];
      }
    },
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    id: {
      type: String,
      default: ""
    },
    clickable: {
      type: Boolean,
      default: !0
    },
    isCursorPointer: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["clickMarker"],
  setup(t, { emit: s }) {
    const c = s;
    function u(o, l) {
      c("clickMarker", { legend: o, i: l });
    }
    return (o, l) => (n(), a("div", {
      id: t.id,
      class: "vue-data-ui-legend",
      style: d({
        background: t.config.backgroundColor,
        color: t.config.color,
        paddingBottom: (t.config.paddingBottom ?? 0) + "px",
        paddingTop: (t.config.paddingTop ?? 12) + "px",
        fontWeight: t.config.fontWeight,
        fontSize: `var(--legend-font-size, ${t.config.fontSize ?? 14}px)`
      })
    }, [
      r(o.$slots, "legendTitle", { titleSet: t.legendSet }, void 0, !0),
      r(o.$slots, "legendToggle", {}, void 0, !0),
      (n(!0), a(f, null, m(t.legendSet, (e, i) => (n(), a("div", {
        key: `legend_${i}`,
        class: g({ "vue-data-ui-legend-item": !0, active: t.clickable && t.isCursorPointer })
      }, [
        e.shape ? (n(), a("svg", {
          key: 0,
          onClick: (C) => u(e, i),
          height: "1em",
          width: "1em",
          viewBox: e.shape && e.shape === "star" ? "-10 -10 80 80" : "0 0 60 60",
          style: d(`overflow: visible; opacity:${e.opacity}`)
        }, [
          p(b, {
            stroke: "none",
            shape: e.shape,
            radius: 30,
            plot: {
              x: 30,
              y: e.shape === "triangle" ? 36 : 30
            },
            fill: e.color
          }, null, 8, ["shape", "plot", "fill"]),
          r(o.$slots, "legend-pattern", h({ ref_for: !0 }, {
            legend: e,
            index: k(v)(e.absoluteIndex) ? e.absoluteIndex : i
          }), void 0, !0)
        ], 12, $)) : y("", !0),
        r(o.$slots, "item", {
          legend: e,
          index: i
        }, void 0, !0)
      ], 2))), 128))
    ], 12, S));
  }
}, I = /* @__PURE__ */ x(B, [["__scopeId", "data-v-ca899be3"]]);
export {
  I as default
};
