// vue-data-ui/dist/NonSvgPenAndPaper-CGk_xThe.js
import { ref as u, computed as ge, onMounted as ze, nextTick as Xe, onBeforeUnmount as Ye, watch as ee, openBlock as g, createElementBlock as y, Fragment as J, normalizeStyle as C, normalizeClass as M, createElementVNode as $, createVNode as P, toDisplayString as ye, unref as xe, withDirectives as we, vModelText as ke, createCommentVNode as R, withModifiers as be, renderList as me } from "vue";
import { j as Ce, x as We, l as Re, X as U } from "./lib-GwPnJSrY.js";
import q from "./BaseIcon-CNx8Tzl5.js";
import { C as Ue } from "./ColorPicker-j0ySewhI.js";
import { _ as qe } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Ve = ["disabled"], Ie = ["data-mode", "xmlns", "viewBox"], Ke = ["cx", "cy", "r", "fill"], Ge = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], He = ["x1", "y1", "x2", "y2", "stroke", "stroke-width", "marker-end"], Oe = ["d", "stroke", "stroke-width"], Qe = ["x", "y", "fill", "font-size"], je = ["x", "dy"], Je = ["d", "stroke", "stroke-width"], Ze = {
  __name: "NonSvgPenAndPaper",
  props: {
    parent: {
      type: HTMLElement
    },
    backgroundColor: {
      type: String,
      default: "#FFFFFF"
    },
    color: {
      type: String,
      default: "#2D353C"
    },
    active: {
      type: Boolean,
      default: !1
    },
    isCursorPointer: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["close"],
  setup(s, { emit: Ae }) {
    const S = s, Se = Ae, m = u([]), E = u([]), te = u("0 0 0 0"), A = u(S.color), T = u(1), v = u(null), V = u(!1), x = u(null), F = u({ x: 0, y: 0 }), D = u([""]), z = u({ row: 0, col: 0 }), L = u(16);
    let $e = 1;
    const ne = ["arrow", "text", "line", "draw"], I = u(0), a = ge(() => ne[I.value]), c = u(null);
    u(null), u(null);
    const K = u(`arrow-${Ce()}`), le = u(`arrow-def-${Ce()}`), Ee = {
      arrow: "plotArrow",
      text: "text",
      line: "plotLine",
      draw: "annotator"
    };
    function Te() {
      I.value + 1 >= ne.length ? I.value = 0 : I.value += 1;
    }
    function Z() {
      return $e++;
    }
    function De(n) {
      const t = p.value;
      if (!t) return { x: 0, y: 0 };
      const e = t.createSVGPoint(), r = !!n.touches?.length || !!n.changedTouches?.length ? n.touches?.[0] || n.changedTouches?.[0] : null, o = r ? r.clientX : n.clientX, i = r ? r.clientY : n.clientY;
      e.x = o, e.y = i;
      const d = t.getScreenCTM()?.inverse();
      return d ? e.matrixTransform(d) : { x: 0, y: 0 };
    }
    function G(n) {
      if (!v.value || a.value !== "text" || V.value) return;
      const { x: t, y: e } = De(n);
      F.value = { x: t, y: e }, D.value = [""], z.value = { row: 0, col: 0 };
      const l = document.createElementNS("http://www.w3.org/2000/svg", "text");
      l.setAttribute("x", t), l.setAttribute("y", e), l.setAttribute("fill", A.value), l.setAttribute("font-size", L.value), l.setAttribute("font-family", "sans-serif"), l.setAttribute("class", "vue-data-ui-doodle"), l.setAttribute("dominant-baseline", "hanging"), l.setAttribute("pointer-events", "all");
      const r = document.createElementNS("http://www.w3.org/2000/svg", "tspan");
      r.setAttribute("x", t), r.setAttribute("dy", "0"), r.textContent = "", l.appendChild(r), l.style.pointerEvents = "none", l.style.userSelect = "none", v.value.appendChild(l), x.value = l, V.value = !0, window.addEventListener("keydown", re), window.addEventListener("mousedown", ue, !0), ae(), oe();
    }
    function re(n) {
      if (!V.value) return;
      let { row: t, col: e } = z.value, l = D.value.slice(), r = !1;
      if (n.key === "Enter") {
        const o = l[t], i = o.slice(0, e), d = o.slice(e);
        l.splice(t, 1, i, d), t += 1, e = 0, r = !0, n.preventDefault();
      } else if (n.key === "Backspace") {
        if (e > 0)
          l[t] = l[t].slice(0, e - 1) + l[t].slice(e), e -= 1, r = !0;
        else if (t > 0) {
          const o = l[t - 1].length;
          l[t - 1] += l[t], l.splice(t, 1), t -= 1, e = o, r = !0;
        }
        n.preventDefault();
      } else if (n.key === "Delete")
        e < l[t].length ? (l[t] = l[t].slice(0, e) + l[t].slice(e + 1), r = !0) : t < l.length - 1 && (l[t] += l[t + 1], l.splice(t + 1, 1), r = !0), n.preventDefault();
      else if (n.key === "ArrowLeft")
        e > 0 ? e -= 1 : t > 0 && (t -= 1, e = l[t].length), r = !0, n.preventDefault();
      else if (n.key === "ArrowRight")
        e < l[t].length ? e += 1 : t < l.length - 1 && (t += 1, e = 0), r = !0, n.preventDefault();
      else if (n.key === "ArrowUp")
        t > 0 && (t -= 1, e = Math.min(e, l[t].length), r = !0), n.preventDefault();
      else if (n.key === "ArrowDown")
        t < l.length - 1 && (t += 1, e = Math.min(e, l[t].length), r = !0), n.preventDefault();
      else if (n.key.length === 1 && !n.ctrlKey && !n.metaKey && !n.altKey)
        l[t] = l[t].slice(0, e) + n.key + l[t].slice(e), e += 1, r = !0, n.preventDefault();
      else if (n.key === "Escape") {
        se(!0);
        return;
      } else n.key === "Tab" && n.preventDefault();
      r && (D.value = l, z.value = { row: t, col: e }, ae(), oe());
    }
    function ae() {
      const n = x.value, { x: t, y: e } = F.value;
      for (; n.firstChild; ) n.removeChild(n.firstChild);
      D.value.forEach((l, r) => {
        const o = document.createElementNS("http://www.w3.org/2000/svg", "tspan");
        o.setAttribute("x", t), o.setAttribute("dy", r === 0 ? "0" : `${L.value * 1.2}`), o.textContent = l.length ? l : "​", n.appendChild(o);
      });
    }
    function oe() {
      const n = v.value.querySelector(".vue-data-ui-svg-caret");
      n && v.value.removeChild(n);
      const t = x.value;
      if (!t) return;
      const { x: e, y: l } = F.value, { row: r, col: o } = z.value, i = L.value, d = t.childNodes[r];
      if (!d) return;
      let h = d.textContent.slice(0, o);
      h.endsWith(" ") && (h += " ");
      const f = document.createElementNS("http://www.w3.org/2000/svg", "text");
      f.setAttribute("x", e), f.setAttribute("y", l), f.setAttribute("font-size", i), f.setAttribute("font-family", "sans-serif"), f.textContent = h || "", v.value.appendChild(f);
      const w = f.getBBox();
      v.value.removeChild(f);
      let k = l + r * i * 1.2, N = e + w.width;
      const b = document.createElementNS("http://www.w3.org/2000/svg", "rect");
      b.setAttribute("x", N), b.setAttribute("y", k), b.setAttribute("width", 1), b.setAttribute("height", i), b.setAttribute("fill", A.value), b.setAttribute("class", "vue-data-ui-svg-caret"), v.value.appendChild(b);
    }
    function ue(n) {
      if (x.value && !x.value.contains(n.target)) {
        const t = x.value.children;
        t.length === 1 && (t[0].textContent === "" || t[0].textContent === "​") && x.value.remove(), se(!1);
      }
    }
    function se(n = !1) {
      window.removeEventListener("keydown", re), window.removeEventListener("mousedown", ue, !0);
      const t = v.value?.querySelector(".vue-data-ui-svg-caret");
      t && v.value.removeChild(t);
      const e = D.value.every(
        (l) => !l || l === "​"
      );
      n || e || m.value.push({
        id: Z(),
        type: "text",
        x: F.value.x,
        y: F.value.y,
        color: A.value,
        fontSize: L.value,
        lines: D.value.map((l) => l)
      }), x.value && v.value && v.value.contains(x.value) && v.value.removeChild(x.value), V.value = !1, x.value = null, D.value = [""], z.value = { row: 0, col: 0 };
    }
    const B = ge(() => We(S.color, 0.6));
    function _({ width: n, height: t }) {
      te.value = `0 0 ${n} ${t}`;
    }
    const H = u(null);
    ze(() => {
      Xe(() => {
        if (S.parent) {
          H.value = new ResizeObserver((e) => {
            for (const l of e) {
              const { width: r, height: o } = l.contentRect;
              _({ width: r, height: o });
            }
          }), H.value.observe(S.parent);
          const { width: n, height: t } = S.parent.getBoundingClientRect();
          _({ width: n, height: t });
        }
      }), v.value = p.value.querySelector("g");
    }), Ye(() => {
      H.value && H.value.disconnect();
    }), ee(
      () => S.parent,
      (n) => {
        if (!n) return;
        const { width: t, height: e } = S.parent.getBoundingClientRect();
        _({ width: t, height: e });
      },
      { immediate: !0 }
    ), ee(a, (n) => {
      p.value && (p.value.removeEventListener("mousedown", G), p.value.removeEventListener("touchstart", G), n === "text" && (p.value.addEventListener("mousedown", G), p.value.addEventListener("touchstart", G, { passive: !1 })));
    });
    const X = u(!1), Y = u(""), p = u(null), O = u(!1), ie = u(null);
    ee(
      () => S.active,
      (n) => {
        p.value && (p.value.style.touchAction = n ? "none" : "");
      },
      { immediate: !0 }
    );
    function ce(n) {
      if (n.cancelable && n.preventDefault(), a.value !== "draw" || !p.value) return;
      X.value = !0;
      const { x: t, y: e } = W(n);
      Y.value = `M ${t} ${e}`;
    }
    function ve(n) {
      if (n.cancelable && n.preventDefault(), a.value !== "draw" || !X.value || !p.value) return;
      const { x: t, y: e } = W(n);
      Y.value += ` ${t} ${e}`;
    }
    function Le() {
      const n = p.value;
      if (!n) return;
      let t = n.querySelector(`defs#${le.value}`);
      if (t || (t = document.createElementNS(U, "defs"), t.setAttribute("id", le.value), n.appendChild(t)), t.querySelector(`#${K.value}`)) return;
      const e = 6, l = document.createElementNS(U, "marker");
      l.setAttribute("id", K.value), l.setAttribute("markerUnits", "strokeWidth"), l.setAttribute("markerWidth", String(e)), l.setAttribute("markerHeight", String(e)), l.setAttribute("refX", String(e - 2)), l.setAttribute("refY", String(e / 2)), l.setAttribute("orient", "auto"), l.setAttribute("viewBox", `0 0 ${e} ${e}`);
      const r = document.createElementNS(U, "path");
      r.setAttribute("d", `M 0 0 L ${e} ${e / 2} L 0 ${e} z`), r.setAttribute("fill", "context-stroke"), r.setAttribute("stroke", "none"), l.appendChild(r), t.appendChild(l);
    }
    function de(n) {
      if (n.cancelable && n.preventDefault(), !["line", "arrow"].includes(a.value) || !p.value || !v.value) return;
      a.value === "arrow" && Le(), O.value = !0;
      const { x: t, y: e } = W(n);
      ie.value = { x: t, y: e }, c.value = document.createElementNS(U, "line"), c.value.setAttribute("x1", t), c.value.setAttribute("y1", e), c.value.setAttribute("x2", t), c.value.setAttribute("y2", e), c.value.setAttribute("stroke", A.value), c.value.setAttribute("stroke-width", T.value), c.value.setAttribute("stroke-linecap", "round"), c.value.setAttribute("class", "vue-data-ui-doodle"), a.value === "arrow" && c.value.setAttribute("marker-end", `url(#${K.value})`), v.value.appendChild(c.value);
    }
    function fe(n) {
      if (n.cancelable && n.preventDefault(), !["line", "arrow"].includes(a.value) || !O.value || !c.value) return;
      const { x: t, y: e } = W(n);
      c.value.setAttribute("x2", t), c.value.setAttribute("y2", e);
    }
    function Q(n) {
      if (n?.cancelable && n.preventDefault(), !O.value || !c.value) return;
      const { x: t, y: e } = W(n);
      c.value.setAttribute("x2", t), c.value.setAttribute("y2", e), m.value.push({
        id: Z(),
        type: a.value,
        // 'line' or 'arrow'
        x1: parseFloat(c.value.getAttribute("x1")),
        y1: parseFloat(c.value.getAttribute("y1")),
        x2: t,
        y2: e,
        strokeWidth: T.value,
        color: A.value
      }), E.value = [], v.value.contains(c.value) && v.value.removeChild(c.value), c.value = null, ie.value = null, O.value = !1;
    }
    function pe(n) {
      const t = n.trim().split(/\s+/);
      if (t.length < 4)
        return n;
      const e = t.slice(1).map(Number);
      if (e.length % 2 !== 0)
        return n;
      const l = Be(e), r = [`M ${l[0]} ${l[1]}`];
      for (let d = 2; d < l.length - 2; d += 2) {
        const h = l[d - 2], f = l[d - 1], w = l[d], k = l[d + 1], N = (h + w) / 2, b = (f + k) / 2;
        r.push(`Q ${h} ${f} ${N} ${b}`);
      }
      const o = l[l.length - 2], i = l[l.length - 1];
      return r.push(`L ${o} ${i}`), r.join(" ");
    }
    function Be(n, t = 1) {
      const e = [...n];
      for (let l = 2; l < n.length - 2; l += 2) {
        const r = n[l], o = n[l + 1], i = n[l - 2], d = n[l - 1], h = n[l + 2], f = n[l + 3];
        e[l] = r + t * ((i + h) / 2 - r), e[l + 1] = o + t * ((d + f) / 2 - o);
      }
      return e;
    }
    function Ne(n) {
      const t = n.trim().split(/\s+/);
      let e = "", l = "", r = null, o = null;
      for (let i = 0; i < t.length; i += 1) {
        const d = t[i];
        if (isNaN(d)) {
          if (l = d, l === "M" || l === "L")
            r = parseFloat(t[++i]), o = parseFloat(t[++i]), e += `${l}${r} ${o}`;
          else if (l === "Q") {
            const h = parseFloat(t[++i]), f = parseFloat(t[++i]), w = parseFloat(t[++i]), k = parseFloat(t[++i]);
            h === r && f === o ? e += `t${w - r} ${k - o}` : e += `q${h - r} ${f - o} ${w - r} ${k - o}`, r = w, o = k;
          }
        } else {
          const h = parseFloat(d), f = parseFloat(t[++i]);
          if (l === "L") {
            const w = h - r, k = f - o;
            w === 0 ? e += `v${k}` : k === 0 ? e += `h${w}` : e += `l${w} ${k}`, r = h, o = f;
          } else if (l === "Q") {
            const w = h, k = f, N = parseFloat(t[++i]), b = parseFloat(t[++i]);
            w === r && k === o ? e += `t${N - r} ${b - o}` : e += `q${w - r} ${k - o} ${N - r} ${b - o}`, r = N, o = b;
          }
        }
      }
      return e;
    }
    function j(n) {
      n?.cancelable && n.preventDefault(), a.value === "draw" && (X.value && (m.value.push({
        id: Z(),
        strokeWidth: T.value,
        path: Ne(pe(Y.value)),
        color: A.value
      }), E.value = [], Y.value = ""), X.value = !1);
    }
    function W(n) {
      if (!p.value) return { x: 0, y: 0 };
      const t = p.value.getBoundingClientRect(), l = !!n.touches?.length || !!n.changedTouches?.length ? n.touches?.[0] || n.changedTouches?.[0] : null, r = l ? l.clientX : n.clientX, o = l ? l.clientY : n.clientY;
      return {
        x: r - t.left,
        y: o - t.top
      };
    }
    function Me() {
      if (m.value.length > 0) {
        const n = m.value.pop();
        E.value.push(n);
      }
    }
    function Pe() {
      if (E.value.length > 0) {
        const n = E.value.pop();
        m.value.push(n);
      }
    }
    function Fe() {
      m.value = [], E.value = [];
    }
    const he = u(null);
    return (n, t) => (g(), y(J, null, [
      s.active ? (g(), y("div", {
        key: 0,
        "data-dom-to-png-ignore": "",
        class: M({
          "vue-ui-pen-and-paper-actions": !0,
          visible: s.active
        }),
        style: C({ backgroundColor: s.backgroundColor })
      }, [
        $("button", {
          class: "vue-ui-pen-and-paper-action",
          style: C({
            backgroundColor: s.backgroundColor,
            border: `1px solid ${B.value}`,
            cursor: s.isCursorPointer ? "pointer" : "default"
          }),
          onClick: t[0] || (t[0] = (e) => Se("close"))
        }, [
          P(q, {
            name: "close",
            stroke: s.color
          }, null, 8, ["stroke"])
        ], 4),
        $("button", {
          class: M({
            "vue-ui-pen-and-paper-action": !0
          }),
          style: C({
            padding: "0 !important",
            cursor: s.isCursorPointer ? "pointer" : "default"
          })
        }, [
          P(Ue, {
            value: A.value,
            "onUpdate:value": t[1] || (t[1] = (e) => A.value = e),
            backgroundColor: s.backgroundColor,
            buttonBorderColor: B.value
          }, null, 8, ["value", "backgroundColor", "buttonBorderColor"])
        ], 4),
        $("button", {
          class: M(["vue-ui-pen-and-paper-action", { "vue-ui-pen-and-paper-action-active": a.value === "text" }]),
          onClick: t[2] || (t[2] = (e) => Te()),
          style: C({
            backgroundColor: s.backgroundColor,
            border: `1px solid ${B.value}`,
            cursor: s.isCursorPointer ? "pointer" : "default"
          })
        }, [
          P(q, {
            name: Ee[a.value],
            stroke: s.color
          }, null, 8, ["name", "stroke"]),
          $("div", {
            style: C({
              position: "absolute",
              bottom: "-20px",
              color: B.value,
              width: "100%",
              textAlign: "center",
              fontSize: "12px"
            })
          }, ye(xe(Re)({
            v: a.value === "text" ? L.value : T.value,
            s: "px",
            r: 1
          })), 5)
        ], 6),
        $("button", {
          class: M({
            "vue-ui-pen-and-paper-action": !0,
            "vue-ui-pen-and-paper-action-disabled": !m.value.length
          }),
          disabled: !m.value.length,
          style: C({
            backgroundColor: s.backgroundColor,
            border: `1px solid ${B.value}`,
            marginTop: "20px",
            cursor: s.isCursorPointer ? "pointer" : "default"
          }),
          onClick: Me
        }, [
          P(q, {
            name: "refresh",
            stroke: s.color
          }, null, 8, ["stroke"])
        ], 14, Ve),
        $("button", {
          class: M({
            "vue-ui-pen-and-paper-action": !0,
            "vue-ui-pen-and-paper-action-disabled": !E.value.length
          }),
          style: C({
            backgroundColor: s.backgroundColor,
            border: `1px solid ${B.value}`,
            cursor: s.isCursorPointer ? "pointer" : "default"
          }),
          onClick: Pe
        }, [
          P(q, {
            name: "refresh",
            stroke: s.color,
            style: { transform: "scaleX(-1)" }
          }, null, 8, ["stroke"])
        ], 6),
        $("button", {
          class: M([{
            "vue-ui-pen-and-paper-action": !0,
            "vue-ui-pen-and-paper-action-disabled": !m.value.length
          }, "vue-ui-pen-and-paper-action"]),
          style: C({
            backgroundColor: s.backgroundColor,
            border: `1px solid ${B.value}`,
            cursor: s.isCursorPointer ? "pointer" : "default"
          }),
          onClick: Fe
        }, [
          P(q, {
            name: "trash",
            stroke: s.color
          }, null, 8, ["stroke"])
        ], 6),
        a.value === "draw" ? we((g(), y("input", {
          key: 0,
          ref_key: "range",
          ref: he,
          type: "range",
          class: "vertical-range",
          min: 0.5,
          max: 12,
          step: 0.1,
          "onUpdate:modelValue": t[3] || (t[3] = (e) => T.value = e),
          style: C({ accentColor: s.color })
        }, null, 4)), [
          [ke, T.value]
        ]) : R("", !0),
        a.value === "text" ? we((g(), y("input", {
          key: 1,
          ref_key: "range",
          ref: he,
          type: "range",
          class: "vertical-range",
          min: 3,
          max: 48,
          step: 0.1,
          "onUpdate:modelValue": t[4] || (t[4] = (e) => L.value = e),
          style: C({ accentColor: s.color })
        }, null, 4)), [
          [ke, L.value]
        ]) : R("", !0)
      ], 6)) : R("", !0),
      (g(), y("svg", {
        "data-mode": a.value,
        ref_key: "svgElement",
        ref: p,
        xmlns: xe(U),
        viewBox: te.value,
        class: M({
          "vue-ui-pen-and-paper": !0,
          inactive: !s.active
        }),
        onMousedown: t[5] || (t[5] = (e) => a.value === "draw" ? ce(e) : a.value === "line" || a.value === "arrow" ? de(e) : null),
        onMousemove: t[6] || (t[6] = (e) => a.value === "draw" ? ve(e) : a.value === "line" || a.value === "arrow" ? fe(e) : null),
        onMouseup: t[7] || (t[7] = (e) => a.value === "draw" ? j(e) : a.value === "line" || a.value === "arrow" ? Q(e) : null),
        onMouseleave: t[8] || (t[8] = (e) => a.value === "draw" ? j(e) : a.value === "line" || a.value === "arrow" ? Q(e) : null),
        onTouchstart: t[9] || (t[9] = be((e) => a.value === "draw" ? ce(e) : a.value === "line" || a.value === "arrow" ? de(e) : null, ["prevent"])),
        onTouchmove: t[10] || (t[10] = be((e) => a.value === "draw" ? ve(e) : a.value === "line" || a.value === "arrow" ? fe(e) : null, ["prevent"])),
        onTouchend: t[11] || (t[11] = (e) => a.value === "draw" ? j(e) : a.value === "line" || a.value === "arrow" ? Q(e) : null),
        onTouchcancel: t[12] || (t[12] = (e) => a.value === "draw" ? j(e) : a.value === "line" || a.value === "arrow" ? Q(e) : null)
      }, [
        $("g", {
          ref_key: "G",
          ref: v
        }, [
          (g(!0), y(J, null, me(m.value, (e) => (g(), y(J, {
            key: e.id
          }, [
            e.path && e.path.replace("M", "").split(" ").length === 2 ? (g(), y("circle", {
              key: 0,
              cx: e.path.replace("M", "").split(" ")[0],
              cy: e.path.replace("M", "").split(" ")[1],
              r: e.strokeWidth / 2,
              fill: e.color
            }, null, 8, Ke)) : e.type === "line" ? (g(), y("line", {
              key: 1,
              class: "vue-ui-pen-and-paper-path",
              x1: e.x1,
              y1: e.y1,
              x2: e.x2,
              y2: e.y2,
              stroke: e.color,
              "stroke-width": e.strokeWidth
            }, null, 8, Ge)) : e.type === "arrow" ? (g(), y("line", {
              key: 2,
              class: "vue-ui-pen-and-paper-path",
              x1: e.x1,
              y1: e.y1,
              x2: e.x2,
              y2: e.y2,
              stroke: e.color,
              "stroke-width": e.strokeWidth,
              "marker-end": `url(#${K.value})`
            }, null, 8, He)) : e.path ? (g(), y("path", {
              key: 3,
              class: "vue-ui-pen-and-paper-path",
              d: e.path,
              stroke: e.color,
              "stroke-width": e.strokeWidth,
              fill: "none"
            }, null, 8, Oe)) : e.type === "text" ? (g(), y("text", {
              key: 4,
              x: e.x,
              y: e.y,
              fill: e.color,
              "font-size": e.fontSize,
              "font-family": "sans-serif",
              "dominant-baseline": "hanging",
              class: "vue-ui-pen-and-paper-text"
            }, [
              (g(!0), y(J, null, me(e.lines, (l, r) => (g(), y("tspan", {
                key: r,
                x: e.x,
                dy: r === 0 ? "0" : e.fontSize * 1.2
              }, ye(l.length ? l : "​"), 9, je))), 128))
            ], 8, Qe)) : R("", !0)
          ], 64))), 128))
        ], 512),
        X.value ? (g(), y("path", {
          key: 0,
          class: "vue-ui-pen-and-paper-path vue-ui-pen-and-paper-path-drawing",
          d: pe(Y.value),
          stroke: A.value,
          "stroke-width": T.value * 1.1,
          fill: "none"
        }, null, 8, Je)) : R("", !0)
      ], 42, Ie))
    ], 64));
  }
}, rt = /* @__PURE__ */ qe(Ze, [["__scopeId", "data-v-ea4aacae"]]);
export {
  rt as default
};
