// vue-data-ui/dist/Slicer-DZJA9IG-.js
import { useCssVars as Ot, ref as m, computed as s, onMounted as lt, onBeforeUnmount as Re, watch as U, nextTick as ve, onUpdated as jt, openBlock as i, createElementBlock as o, normalizeStyle as L, createElementVNode as P, createVNode as Ut, renderSlot as Zt, createCommentVNode as c, unref as de, Fragment as b, renderList as R, normalizeProps as Yt, mergeProps as Jt, normalizeClass as Z, withDirectives as at, vModelText as nt, toDisplayString as $e } from "vue";
import { j as qt, X as Kt, W as Ae, L as it, c as ot, M as ut, b as rt, au as st } from "./lib-GwPnJSrY.js";
import { t as ct, u as Qt, d as _t } from "./useResponsive-DfdjqQps.js";
import el from "./BaseIcon-CNx8Tzl5.js";
import { _ as tl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ll = ["data-minimap"], al = {
  class: "vue-data-ui-slicer-labels",
  style: { position: "relative", "z-index": "1", "pointer-events": "none" }
}, nl = {
  key: 0,
  style: { width: "100%", position: "relative" }
}, il = {
  key: 0,
  class: "minimap",
  style: { width: "100%" }
}, ol = ["xmlns", "viewBox"], ul = ["id"], rl = ["stop-color"], sl = ["width", "height", "stroke"], cl = ["d", "stroke"], vl = ["x", "width", "y", "height", "fill"], dl = ["d", "stroke"], hl = ["cx", "cy", "fill", "stroke"], ml = ["x", "width", "height", "fill", "rx"], fl = ["x", "width", "height", "rx", "fill"], gl = ["x", "width", "height", "fill", "rx"], yl = ["x2", "y1", "y2", "stroke"], xl = ["d", "fill"], Ml = ["d", "stroke"], kl = { key: 6 }, bl = ["x", "width", "y", "height", "fill"], wl = ["d", "stroke"], Cl = ["x", "width", "height", "fill", "stroke", "stroke-width"], Sl = ["x", "width", "height"], pl = ["fill", "stroke"], Ll = ["stroke"], Pl = ["fill"], Nl = ["x", "width", "height", "fill", "stroke", "stroke-width"], Rl = ["x", "width", "height"], $l = ["fill", "stroke"], Al = ["stroke"], Tl = ["fill"], El = ["cx", "cy", "stroke", "fill"], Bl = ["cx", "cy", "fill"], Wl = ["cx", "cy", "stroke", "fill"], Il = ["cx", "cy", "fill"], Fl = { key: 12 }, Vl = ["cx", "cy", "fill", "stroke"], Dl = ["cx", "cy", "stroke", "fill"], zl = ["cx", "cy", "fill"], Hl = ["cx", "cy", "stroke", "fill"], Xl = ["cx", "cy", "fill"], Gl = ["height", "x", "width", "onMouseenter"], Ol = ["min", "max"], jl = ["min", "max"], Te = 48, Ul = {
  __name: "Slicer",
  props: {
    background: {
      type: String,
      default: "#FFFFFF"
    },
    borderColor: {
      type: String,
      default: "#FFFFFF"
    },
    fontSize: {
      type: Number,
      default: 14
    },
    labelLeft: {
      type: [String, Number],
      default: ""
    },
    labelRight: {
      type: [String, Number],
      default: ""
    },
    textColor: {
      type: String,
      default: "#1A1A1A"
    },
    inputColor: {
      type: String,
      default: "#1A1A1A"
    },
    max: {
      type: Number,
      default: 0
    },
    min: {
      type: Number,
      default: 0
    },
    selectColor: {
      type: String,
      default: "#4A4A4A"
    },
    useResetSlot: {
      type: Boolean,
      default: !1
    },
    valueStart: {
      type: [Number, String],
      default: 0
    },
    valueEnd: {
      type: [Number, String],
      default: 0
    },
    minimap: {
      type: Array,
      default: []
    },
    smoothMinimap: {
      type: Boolean,
      default: !1
    },
    minimapSelectedColor: {
      type: String,
      default: "#1f77b4"
    },
    minimapSelectionRadius: {
      type: Number,
      default: 12
    },
    minimapLineColor: {
      type: String,
      default: "#2D353C"
    },
    minimapSelectedColorOpacity: {
      type: Number,
      default: 0.2
    },
    minimapSelectedIndex: {
      type: Number,
      default: null
    },
    minimapIndicatorColor: {
      type: String,
      default: "#2D353C"
    },
    refreshStartPoint: {
      type: Number,
      default: 0
    },
    refreshEndPoint: {
      type: Number,
      default: null
    },
    enableRangeHandles: {
      type: Boolean,
      default: !1
    },
    enableSelectionDrag: {
      type: Boolean,
      default: !0
    },
    verticalHandles: {
      type: Boolean,
      default: !1
    },
    minimapCompact: {
      type: Boolean,
      default: !1
    },
    allMinimaps: {
      type: Array,
      default() {
        return [];
      }
    },
    minimapMerged: {
      type: Boolean,
      default: !1
    },
    minimapFrameColor: {
      type: String,
      default: "#e1e5e8"
    },
    cutNullValues: {
      type: Boolean,
      default: !1
    },
    focusOnDrag: {
      type: Boolean,
      default: !1
    },
    focusRangeRatio: {
      type: Number,
      default: 0.1
    },
    maxWidth: {
      type: Number,
      default: null
    },
    isCursorPointer: {
      type: Boolean,
      default: !1
    },
    additionalMinimapHeight: {
      type: Number,
      default: 0
    },
    handleType: {
      type: String,
      default: ""
    },
    handleWidth: {
      type: Number,
      default: 20
    },
    handleBorderWidth: {
      type: Number,
      default: 1
    },
    handleIconColor: {
      type: String,
      default: null
    },
    handleBorderColor: {
      type: String,
      default: null
    },
    handleFill: {
      type: String,
      default: null
    }
  },
  emits: ["update:start", "update:end", "reset", "trapMouse"],
  setup(l, { expose: vt, emit: dt }) {
    Ot((e) => ({
      b3608678: gt.value,
      v3e454538: g.value,
      v68ad08a9: xt.value,
      v1e7a8668: l.selectColor,
      a2950614: yt.value
    }));
    const n = l, C = m(null), S = m(n.min), p = m(n.max), M = s(() => !!n.allMinimaps.length), Ee = m(qt()), $ = s(() => M.value && n.minimapCompact), he = m(0);
    lt(() => {
      const e = () => {
        C.value && (he.value = C.value.getBoundingClientRect().width);
      };
      e();
      const a = ct(e, 50);
      window.addEventListener("resize", a), Re(() => {
        window.removeEventListener("resize", a);
      });
    });
    const ht = s(() => n.refreshEndPoint === null ? n.max : n.refreshEndPoint), Y = dt, J = s(() => {
      if ($.value) {
        const a = Math.max(1, oe.value - 1);
        return xe.value / a * 100;
      }
      const e = Math.max(1, n.max - n.min);
      return (S.value - n.min) / e * 100;
    }), re = s(() => {
      if ($.value) {
        const a = Math.max(1, oe.value - 1);
        return Me.value / a * 100;
      }
      const e = Math.max(1, n.max - n.min);
      return (p.value - n.min) / e * 100;
    }), me = s(() => (J.value + re.value) / 2), fe = s(() => {
      if (!C.value) return !1;
      const e = C.value.getBoundingClientRect().width;
      return J.value / 100 * e - Pe.value / 2 < 0;
    }), ge = s(() => {
      if (!C.value) return !1;
      const e = C.value.getBoundingClientRect().width;
      return re.value / 100 * e + ce.value / 2 > e;
    }), N = s(() => {
      const e = ge.value ? `calc(${me.value}% - ${ie.value.width}px)` : fe.value ? `calc(${me.value}% - 8px)` : `calc(${me.value}% - ${ie.value.width / 2}px)`;
      return {
        left: `${J.value}%`,
        width: `${Math.max(0, re.value - J.value)}%`,
        background: n.selectColor,
        tooltipLeft: `calc(${J.value}% - ${fe.value ? 9 : Pe.value / 2}px)`,
        tooltipRight: `calc(${re.value}% - ${ge.value ? ce.value - 9 : ce.value / 2}px)`,
        tooltipCenter: e,
        arrowLeft: !fe.value,
        arrowRight: !ge.value
      };
    }), q = s(() => Math.max(0, (K.value || 1) - 1)), ye = s(() => Math.max(0, (oe.value || 1) - 1));
    function se(e) {
      return ye.value === 0 ? 0 : Math.round(e / ye.value * q.value);
    }
    function Be(e) {
      return q.value === 0 ? 0 : Math.round(e / q.value * ye.value);
    }
    const mt = (e) => Math.floor(n.min + e), ft = (e) => Math.ceil(n.min + e), xe = s({
      get() {
        return $.value ? se(F.value) : Number(S.value);
      },
      set(e) {
        if ($.value) {
          const a = Math.round(+e || 0), t = Math.max(0, se(Math.max(0, H.value - 1))), u = Math.min(Math.max(0, a), t), r = Be(u);
          _(mt(r));
        } else {
          let a = Math.round(+e || 0);
          const t = Number(p.value) - 1, u = Math.min(Math.max(n.min, a), t);
          _(u);
        }
      }
    }), Me = s({
      get() {
        return $.value ? se(Math.max(F.value, H.value - 1)) : Number(p.value);
      },
      set(e) {
        if ($.value) {
          const a = Math.round(+e || 0), t = se(F.value), u = Math.max(a, t), r = Be(u);
          ee(ft(r + 1));
        } else {
          let a = Math.round(+e || 0);
          const t = Number(S.value) + 1, u = Math.max(t, Math.min(a, n.max));
          ee(u);
        }
      }
    }), gt = s(() => n.inputColor), yt = s(() => n.background), xt = s(() => `${n.selectColor}33`), g = s(() => n.borderColor), We = s(() => {
      const e = [];
      for (let a = 0; a < n.max; a += 1)
        e.push(a);
      return e;
    });
    function Ie() {
      Y("reset");
    }
    U(
      () => n.min,
      (e) => {
        Number(S.value) < Number(e) && (S.value = Number(e)), Number(p.value) < Number(e) && (p.value = Number(e));
      }
    ), U(
      () => n.max,
      (e) => {
        Number(S.value) > Number(e) && (S.value = Number(e)), Number(p.value) > Number(e) && (p.value = Number(e));
      }
    );
    const ke = m(null), f = m({
      width: 1,
      height: 1
    }), O = m(null);
    lt(() => {
      if (M.value) {
        const e = ct(() => {
          const { width: a, height: t } = Qt({
            chart: ke.value
          });
          f.value.width = a, f.value.height = t - 47;
        });
        O.value = new ResizeObserver(e), O.value.observe(ke.value);
      }
    }), Re(() => {
      O.value && O.value.disconnect();
    });
    const K = s(() => Math.max(...n.allMinimaps.map((e) => e.series.length), n.minimap.length || 0)), k = s(() => {
      if (!K.value) return 0;
      const e = Math.max(1, K.value - (n.minimapCompact ? 1 : 0));
      return f.value.width / e;
    }), Fe = s(() => {
      const e = [];
      if (Array.isArray(n.minimap) && n.minimap.length && n.minimapMerged && e.push(...n.minimap.filter(Number.isFinite)), Array.isArray(n.allMinimaps) && n.allMinimaps.length)
        for (const a of n.allMinimaps)
          a?.isVisible && Array.isArray(a?.series) && e.push(...a.series.filter(Number.isFinite));
      return e.length ? { min: Math.min(...e), max: Math.max(...e) } : { min: 0, max: 1 };
    }), D = s(() => Fe.value.min), j = s(() => Fe.value.max), Mt = s(() => D.value < 0 && j.value > 0 || j.value <= 0 ? D.value : 0), kt = s(() => Math.max(1e-9, Math.max(Math.abs(D.value), Math.abs(j.value)))), bt = s(() => D.value < 0 && j.value > 0), wt = (e) => {
      const a = Math.max(1, f.value.height);
      if (bt.value) {
        const t = kt.value;
        return (1 - (Math.max(-1, Math.min(1, e / t)) + 1) / 2) * a;
      } else {
        const t = Math.min(0, D.value), u = Math.max(0, j.value);
        return a - (e - t) / Math.max(1e-9, u - t) * a;
      }
    }, Ve = s(() => wt(0)), De = s(() => n.allMinimaps.length ? n.allMinimaps.filter((e) => e.type === "bar").length : 0), be = s(() => k.value / (De.value || 1) * 0.8);
    function ze(e, a, t) {
      const u = be.value, r = Math.max(1, De.value), h = K.value - 1;
      return t === 0 ? e + u / 2 * a : t === h ? e - u / 2 * (r - a) : e - r * u / 2 + a * u;
    }
    function He(e, a) {
      return [0, K.value - 1].includes(a) ? be.value / 2 : be.value;
    }
    function Ct(e, a, t) {
      const u = (x, d, V) => Math.max(d, Math.min(V, x));
      if (a <= 0) {
        const x = Math.max(1e-9, 0 - e);
        return (d) => t - (d - e) / x * t;
      }
      if (e >= 0) {
        const x = Math.max(1e-9, a - 0);
        return (d) => t - (d - 0) / x * t;
      }
      const h = Math.max(1e-9, Math.max(Math.abs(e), Math.abs(a)));
      return (x) => (1 - (u(x / h, -1, 1) + 1) / 2) * t;
    }
    function Xe(e, a = !1) {
      if (!e || !e.length)
        return {
          fullSet: "",
          points: [],
          selectionSet: "",
          sliced: [],
          firstPlot: null,
          lastPlot: null,
          hasFull: !1,
          hasSelection: !1,
          fullMarkers: [],
          selectionMarkers: []
        };
      const t = Math.max(1, f.value.height), u = Ct(D.value, j.value, t), r = e.length, h = Math.min(Math.max(0, F.value), Math.max(0, r - 1)), x = Math.min(r, Math.max(h + 1, H.value)), d = e.map((y, et) => {
        const ue = y, tt = Number.isFinite(ue), Xt = k.value * et + (n.minimapCompact ? 0 : k.value / 2), Gt = u(0);
        return {
          x: Xt,
          y: tt ? u(ue) : NaN,
          v: ue,
          value: ue == null ? null : tt ? ue : null,
          y0: Gt,
          i: et
        };
      }), V = (y) => y >= 0 && y < d.length && Number.isFinite(d[y]?.value), X = d.filter((y) => Number.isFinite(y.value) && !V(y.i - 1) && !V(y.i + 1)), Ne = X.filter((y) => y.i >= h && y.i < x), G = d.slice(h, x), zt = d.length >= 2 ? n.smoothMinimap || a ? n.cutNullValues ? it(d) : ot(d.filter((y) => y.value != null)) : n.cutNullValues ? ut(d) : rt(d.filter((y) => y.value != null)) : "", Ht = G.length >= 2 ? n.smoothMinimap || a ? n.cutNullValues ? it(G) : ot(G.filter((y) => y.value != null)) : n.cutNullValues ? ut(G) : rt(G.filter((y) => y.value != null)) : "";
      return {
        fullSet: zt,
        points: d,
        selectionSet: Ht,
        sliced: G,
        firstPlot: d[h] || null,
        lastPlot: d[Math.max(0, x - 1)] || null,
        hasFull: d.length >= 2,
        hasSelection: G.length >= 2,
        fullMarkers: X,
        selectionMarkers: Ne
      };
    }
    const v = s(() => n.minimap.length ? Xe(n.minimap) : {
      fullSet: "",
      selectionSet: "",
      sliced: [],
      firstPlot: null,
      lastPlot: null,
      hasFull: !1,
      hasSelection: !1
    }), Q = s(() => n.allMinimaps.length ? n.allMinimaps.map((e, a) => {
      const t = Xe(e?.series || [], !!e.smooth), u = e?.id ?? e?.name ?? a;
      return {
        key: typeof u == "object" ? JSON.stringify(u) : String(u),
        color: e?.color,
        ...t,
        isVisible: e.isVisible,
        type: e.type || "line"
      };
    }) : []), w = s(() => {
      const e = F.value, a = Math.max(e + 1, H.value);
      return {
        x: k.value * e + (n.minimapCompact ? 0 : k.value / 2),
        width: k.value * (a - e) - k.value
      };
    }), A = m(n.minimapSelectedIndex);
    function St(e) {
      A.value = _e(n.valueStart) + e;
    }
    const pt = _t(St, 60);
    U(() => n.minimapSelectedIndex, (e, a) => {
      if ([null, void 0].includes(e)) {
        A.value = null;
        return;
      }
      e !== a && pt(e);
    }, { immediate: !0 });
    function Lt(e) {
      A.value = e;
      const a = F.value, t = H.value;
      e >= a && e < t && !I.value && Y("trapMouse", e - a);
    }
    const we = m(null), Ce = m(null);
    function _(e) {
      S.value = e, we.value && (we.value.value = e), Y("update:start", Number(S.value));
    }
    function ee(e) {
      p.value = e, Ce.value && (Ce.value.value = e), Y("update:end", Number(p.value));
    }
    const Se = s(() => n.valueEnd - n.valueStart), Pt = s(() => Se.value < n.max - n.min), te = m(!1);
    let Nt = m(null);
    const Rt = s(() => (he.value - 48) / (n.max - n.min) * Se.value), $t = s(() => Math.max(1, he.value - Te - Rt.value)), At = s(() => Math.max(1, n.max - n.min - Se.value)), Tt = s(() => At.value / $t.value), Ge = m(0), pe = m(0), Oe = m(0), Et = m(0);
    let T = null, E = null, B = null, W = null;
    const je = m(n.min);
    function Le(e) {
      if (!C.value) return n.min;
      const a = C.value.getBoundingClientRect(), t = a.left + Te / 2, u = a.right - Te / 2, r = Math.max(1, u - t), x = (Math.max(t, Math.min(e, u)) - t) / r, d = Math.max(1, n.max - n.min);
      return Math.round(n.min + x * d);
    }
    const Ue = async (e) => {
      if (z.value = !0, !n.enableSelectionDrag) return;
      const a = e.type === "touchstart";
      a || e.stopPropagation();
      const t = a && e.targetTouches && e.targetTouches[0] ? e.targetTouches[0] : null, u = a ? t ? t.target : null : e.target;
      if (!u || !(u instanceof Element) || u.classList && u.classList.contains("range-handle")) return;
      te.value = !0;
      const r = a ? t ? t.clientX : 0 : e.clientX;
      if (Nt.value = r, Ge.value = r, n.focusOnDrag && !Pt.value && C.value) {
        je.value = Le(r);
        const h = Math.min(0.95, Math.max(0.05, n.focusRangeRatio)), x = n.max - n.min, d = Math.max(1, Math.round(x * h)), V = Math.floor(d / 2);
        let X = je.value - V;
        X = Math.max(n.min, Math.min(X, n.max - d));
        const Ne = Math.min(n.max, X + d);
        _(X), ee(Ne), st(C.value, "mouseup"), await ve(), st(C.value, "mousedown", { clientX: r });
      }
      pe.value = Number(S.value), Oe.value = Number(p.value), Et.value = Tt.value, T = a ? "touchmove" : "mousemove", E = a ? "touchend" : "mouseup", B = a ? Wt : Bt, W = a ? Ft : It, window.addEventListener(T, B, { passive: !1 }), window.addEventListener(E, W);
    };
    function Bt(e) {
      te.value && Ze(e.clientX);
    }
    function Wt(e) {
      if (!te.value || !C.value) return;
      const a = e.target;
      if (!(a instanceof Element) || !C.value.contains(a) || a.classList && a.classList.contains("range-handle")) return;
      e.preventDefault();
      const t = e.targetTouches && e.targetTouches[0] ? e.targetTouches[0] : null;
      t && Ze(t.clientX);
    }
    function Ze(e) {
      if (!te.value) return;
      const a = Le(Ge.value), u = Le(e) - a, r = Math.max(1, Oe.value - pe.value);
      let h = Math.round(pe.value + u);
      const x = n.min, d = n.max - r;
      h < x && (h = x), h > d && (h = d);
      const V = h + r;
      _(h), ee(V);
    }
    function It() {
      Ye();
    }
    function Ft() {
      Ye();
    }
    function Ye() {
      te.value = !1, T && B && window.removeEventListener(T, B), E && W && window.removeEventListener(E, W), T = E = null, B = W = null;
    }
    const I = m(!1), le = m(null), ae = m(null), Pe = m(1), ce = m(1), z = m(!1);
    function Je() {
      le.value && (Pe.value = le.value.getBoundingClientRect().width);
    }
    function qe() {
      ae.value && (ce.value = ae.value.getBoundingClientRect().width);
    }
    const Ke = m(0);
    function Qe(e) {
      Ke.value = e === "start" ? 1 : 0;
    }
    const ne = m(!1), ie = m({
      width: 0,
      left: 0
    });
    U([S, p], async () => {
      if (await ve(), !le.value || !ae.value) {
        ne.value = !1, ie.value = { width: 0, left: 0 };
        return;
      }
      const e = le.value.getBoundingClientRect(), a = ae.value.getBoundingClientRect();
      ne.value = e.x + e.width > a.x;
      const t = e.x + e.width / 2, u = a.x + a.width / 2, r = e.width + a.width, h = (t + u) / 2;
      ie.value = {
        width: r,
        left: h - r / 2
      };
    }), jt(() => {
      Je(), qe();
    }), U(
      () => n.labelLeft,
      () => {
        ve(Je);
      },
      { deep: !0 }
    ), U(
      () => n.labelRight,
      () => {
        ve(qe);
      },
      { deep: !0 }
    ), vt({
      setStartValue: _,
      setEndValue: ee
    }), Re(() => {
      O.value && O.value.disconnect(), T && B && window.removeEventListener(T, B), E && W && window.removeEventListener(E, W), T = E = null, B = W = null;
    });
    const oe = s(() => Math.max(1, n.max - n.min));
    function _e(e) {
      const a = Math.floor(e - n.min);
      return Math.min(Math.max(0, a), oe.value);
    }
    function Vt(e) {
      const a = Math.ceil(e - n.min);
      return Math.min(Math.max(0, a), oe.value);
    }
    const F = s(() => _e(S.value)), H = s(() => Vt(p.value)), Dt = s(() => {
      if (!We.value.length) return null;
      if (A.value >= F.value && A.value < H.value) {
        const e = A.value;
        return {
          x1: k.value * e + (n.minimapCompact ? 0 : k.value / 2),
          x2: k.value * e + (n.minimapCompact ? 0 : k.value / 2),
          y1: 0,
          y2: Math.max(f.value.height, 0),
          stroke: n.minimapIndicatorColor,
          "stroke-linecap": "round",
          "stroke-dasharray": 2,
          "stroke-width": 1
        };
      } else
        return null;
    });
    return (e, a) => (i(), o("div", {
      "data-minimap": M.value,
      "data-dom-to-png-ignore": "",
      style: L([{ padding: "0 48px" }, {
        maxWidth: l.maxWidth ? l.maxWidth + "px" : void 0,
        margin: l.maxWidth ? "0 auto" : void 0
      }]),
      class: "vue-data-ui-zoom",
      ref_key: "zoomWrapper",
      ref: C,
      onMousedown: Ue,
      onTouchstart: Ue,
      onTouchend: a[11] || (a[11] = (t) => z.value = !1)
    }, [
      P("div", al, [
        l.valueStart !== l.refreshStartPoint || l.valueEnd !== ht.value ? (i(), o("div", nl, [
          l.useResetSlot ? Zt(e.$slots, "reset-action", {
            key: 1,
            reset: Ie
          }, void 0, !0) : (i(), o("button", {
            key: 0,
            tabindex: "0",
            role: "button",
            class: "vue-data-ui-refresh-button",
            style: L({
              top: M.value ? "36px" : "-16px",
              pointerEvents: "all !important",
              cursor: l.isCursorPointer ? "pointer" : "default"
            }),
            onClick: Ie
          }, [
            Ut(el, {
              name: "refresh",
              stroke: l.textColor
            }, null, 8, ["stroke"])
          ], 4))
        ])) : c("", !0)
      ]),
      P("div", {
        class: "double-range-slider",
        ref_key: "minimapWrapper",
        ref: ke,
        style: L([{ "z-index": "0" }, M.value ? {
          "--minimap-unit-px": k.value + "px",
          "--minimap-offset-px": (l.minimapCompact ? 0 : k.value / 2) + "px"
        } : void 0]),
        onMouseenter: a[9] || (a[9] = (t) => z.value = !0),
        onMouseleave: a[10] || (a[10] = (t) => z.value = !1)
      }, [
        M.value ? (i(), o("div", il, [
          (i(), o("svg", {
            key: `mm-${l.minimapMerged ? "merged" : "split"}-${l.minimapCompact ? "compact" : "normal"}`,
            xmlns: de(Kt),
            viewBox: `0 0 ${Math.max(0, f.value.width)} ${Math.max(0, f.value.height)}`
          }, [
            P("defs", null, [
              P("linearGradient", {
                id: Ee.value,
                x1: "0%",
                y1: "0%",
                x2: "0%",
                y2: "100%"
              }, [
                P("stop", {
                  offset: "0%",
                  "stop-color": `${l.minimapLineColor}50`
                }, null, 8, rl),
                a[12] || (a[12] = P("stop", {
                  offset: "100%",
                  "stop-color": "transparent"
                }, null, -1))
              ], 8, ul)
            ]),
            l.minimapCompact ? (i(), o("rect", {
              key: 0,
              class: "vue-ui-zoom-minimap-frame",
              x: 0,
              y: 0,
              width: f.value.width,
              height: f.value.height,
              fill: "none",
              stroke: l.minimapFrameColor,
              rx: 3
            }, null, 8, sl)) : c("", !0),
            l.minimapMerged ? (i(), o("path", {
              key: 1,
              d: `M${v.value.fullSet}`,
              stroke: `${l.minimapLineColor}`,
              fill: "none",
              "stroke-width": "1",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: { opacity: "0.6" }
            }, null, 8, cl)) : (i(), o(b, { key: 2 }, [
              (i(!0), o(b, null, R(Q.value.filter((t) => t.type === "bar"), (t, u) => (i(), o("g", null, [
                (i(!0), o(b, null, R(t.points, (r, h) => (i(), o(b, null, [
                  t.isVisible && !isNaN(r.y) ? (i(), o("rect", {
                    key: 0,
                    x: ze(r.x, u, h),
                    width: He(u, h),
                    y: r.v >= 0 ? r.y : r.y0,
                    height: r.v >= 0 ? r.y0 - r.y : r.y - r.y0,
                    fill: t.color,
                    style: { opacity: "0.6" }
                  }, null, 8, vl)) : c("", !0)
                ], 64))), 256))
              ]))), 256)),
              (i(!0), o(b, null, R(Q.value.filter((t) => t.type === "line"), (t) => (i(), o("g", {
                key: String(t.key)
              }, [
                t.isVisible ? (i(), o("path", {
                  key: 0,
                  d: `M ${t.fullSet}`,
                  fill: "none",
                  stroke: t.color,
                  style: { opacity: "0.6" }
                }, null, 8, dl)) : c("", !0),
                t.isVisible ? (i(!0), o(b, { key: 1 }, R(t.fullMarkers, (u) => (i(), o("circle", {
                  key: `sel-dot-under-${t.key}-${u.i}`,
                  cx: u.x,
                  cy: u.y,
                  r: "2",
                  fill: t.color,
                  stroke: g.value,
                  "stroke-width": "0.5",
                  style: { opacity: "0.6" }
                }, null, 8, hl))), 128)) : c("", !0)
              ]))), 128))
            ], 64)),
            P("rect", {
              x: w.value.x,
              width: w.value.width < 0 ? 0 : w.value.width,
              height: Math.max(f.value.height, 0),
              y: 0,
              fill: g.value,
              rx: l.minimapSelectionRadius,
              stroke: "none"
            }, null, 8, ml),
            P("rect", {
              x: w.value.x,
              width: w.value.width < 0 ? 0 : w.value.width,
              height: Math.max(f.value.height, 0),
              y: 0,
              rx: l.minimapSelectionRadius,
              fill: g.value
            }, null, 8, fl),
            P("rect", {
              x: w.value.x,
              width: w.value.width < 0 ? 0 : w.value.width,
              height: Math.max(f.value.height, 0),
              y: 0,
              fill: l.minimapSelectedColor,
              rx: l.minimapSelectionRadius,
              style: L({
                opacity: l.minimapSelectedColorOpacity
              })
            }, null, 12, gl),
            A.value !== null && !I.value ? (i(), o("line", Yt(Jt({ key: 3 }, Dt.value)), null, 16)) : c("", !0),
            !l.minimapMerged && Mt.value < 0 ? (i(), o("line", {
              key: 4,
              class: "slicer-minimap-zero-line",
              x1: 0,
              x2: f.value.width,
              y1: Ve.value,
              y2: Ve.value,
              stroke: l.minimapFrameColor,
              "stroke-width": "0.5"
            }, null, 8, yl)) : c("", !0),
            l.minimapMerged ? (i(), o(b, { key: 5 }, [
              v.value && v.value.sliced && v.value.sliced.length ? (i(), o(b, { key: 0 }, [
                v.value.selectionSet ? (i(), o("path", {
                  key: 0,
                  d: `M${v.value.sliced[0].x},${Math.max(f.value.height, 0)} ${v.value.selectionSet} L${v.value.sliced[v.value.sliced.length - 1].x},${Math.max(f.value.height, 0)}Z`,
                  fill: `url(#${Ee.value})`,
                  stroke: "none",
                  style: { opacity: "1" }
                }, null, 8, xl)) : c("", !0),
                v.value.selectionSet ? (i(), o("path", {
                  key: 1,
                  d: `M ${v.value.selectionSet}`,
                  stroke: `${l.minimapLineColor}`,
                  fill: "transparent",
                  "stroke-width": "2",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round"
                }, null, 8, Ml)) : c("", !0)
              ], 64)) : c("", !0)
            ], 64)) : (i(), o("g", kl, [
              (i(!0), o(b, null, R(Q.value.filter((t) => t.type === "bar"), (t, u) => (i(), o("g", null, [
                (i(!0), o(b, null, R(t.points, (r, h) => (i(), o(b, null, [
                  t.isVisible && !isNaN(r.y) ? (i(), o("rect", {
                    key: 0,
                    x: ze(r.x, u, h),
                    width: He(u, h),
                    y: r.v >= 0 ? r.y : r.y0,
                    height: r.v >= 0 ? r.y0 - r.y : r.y - r.y0,
                    fill: t.color,
                    style: L({
                      opacity: h >= l.valueStart && h <= l.valueEnd - 1 ? 1 : 0
                    })
                  }, null, 12, bl)) : c("", !0)
                ], 64))), 256))
              ]))), 256)),
              (i(!0), o(b, null, R(Q.value.filter((t) => t.type === "line"), (t) => (i(), o("g", {
                key: String(t.key)
              }, [
                t && t.hasSelection && t.selectionSet && t.isVisible ? (i(), o("path", {
                  key: 0,
                  d: `M ${t.selectionSet}`,
                  stroke: t.color,
                  fill: "transparent",
                  "stroke-width": "2",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round"
                }, null, 8, wl)) : c("", !0)
              ]))), 128))
            ])),
            M.value && l.minimapCompact ? (i(), o("rect", {
              key: 7,
              class: "vue-ui-zoom-compact-minimap-handle",
              x: w.value.x - Math.min(40, Math.max(20, l.handleWidth)),
              y: 0,
              width: Math.min(40, Math.max(20, l.handleWidth)),
              height: f.value.height,
              fill: l.handleFill || g.value,
              stroke: l.handleBorderColor || l.textColor,
              "stroke-width": l.handleBorderWidth,
              rx: 3
            }, null, 8, Cl)) : c("", !0),
            l.handleType && l.handleType !== "empty" ? (i(), o("svg", {
              key: 8,
              x: w.value.x - Math.min(40, Math.max(20, l.handleWidth)),
              y: 0,
              width: Math.min(40, Math.max(20, l.handleWidth)),
              height: f.value.height,
              viewBox: "0 0 20 20"
            }, [
              l.handleType === "arrow" ? (i(), o("path", {
                key: 0,
                d: "M 7 7 L 3 10 L 7 13 L 7 7 M 13 7 L 17 10 L 13 13 L 13 7",
                fill: g.value,
                stroke: l.handleIconColor || l.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, pl)) : l.handleType === "chevron" ? (i(), o("path", {
                key: 1,
                d: "M 6 7 L 4 10 L 6 13 M 14 7 L 16 10 L 14 13",
                fill: "none",
                stroke: l.handleIconColor || l.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, Ll)) : l.handleType === "grab" ? (i(), o("path", {
                key: 2,
                d: "M 8 5 A 1 1 0 0 0 8 7 A 1 1 0 0 0 8 5 M 8 9 A 1 1 0 0 0 8 11 A 1 1 0 0 0 8 9 M 8 13 A 1 1 0 0 0 8 15 A 1 1 0 0 0 8 13 M 12 5 A 1 1 0 0 0 12 7 A 1 1 0 0 0 12 5 M 12 9 A 1 1 0 0 0 12 11 A 1 1 0 0 0 12 9 M 12 13 A 1 1 0 0 0 12 15 A 1 1 0 0 0 12 13",
                fill: l.handleIconColor || l.textColor,
                stroke: "none",
                opacity: "0.6"
              }, null, 8, Pl)) : c("", !0)
            ], 8, Sl)) : c("", !0),
            M.value && l.minimapCompact ? (i(), o("rect", {
              key: 9,
              class: "vue-ui-zoom-compact-minimap-handle",
              x: w.value.x + w.value.width,
              y: 0,
              width: Math.min(40, Math.max(20, l.handleWidth)),
              height: f.value.height,
              fill: l.handleFill || g.value,
              stroke: l.handleBorderColor || l.textColor,
              "stroke-width": l.handleBorderWidth,
              rx: 3
            }, null, 8, Nl)) : c("", !0),
            l.handleType && l.handleType !== "empty" ? (i(), o("svg", {
              key: 10,
              x: w.value.x + w.value.width,
              y: 0,
              width: Math.min(40, Math.max(20, l.handleWidth)),
              height: f.value.height,
              viewBox: "0 0 20 20"
            }, [
              l.handleType === "arrow" ? (i(), o("path", {
                key: 0,
                d: "M 7 7 L 3 10 L 7 13 L 7 7 M 13 7 L 17 10 L 13 13 L 13 7",
                fill: g.value,
                stroke: l.handleIconColor || l.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, $l)) : l.handleType === "chevron" ? (i(), o("path", {
                key: 1,
                d: "M 6 7 L 4 10 L 6 13 M 14 7 L 16 10 L 14 13",
                fill: "none",
                stroke: l.handleIconColor || l.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, Al)) : l.handleType === "grab" ? (i(), o("path", {
                key: 2,
                d: "M 8 5 A 1 1 0 0 0 8 7 A 1 1 0 0 0 8 5 M 8 9 A 1 1 0 0 0 8 11 A 1 1 0 0 0 8 9 M 8 13 A 1 1 0 0 0 8 15 A 1 1 0 0 0 8 13 M 12 5 A 1 1 0 0 0 12 7 A 1 1 0 0 0 12 5 M 12 9 A 1 1 0 0 0 12 11 A 1 1 0 0 0 12 9 M 12 13 A 1 1 0 0 0 12 15 A 1 1 0 0 0 12 13",
                fill: l.handleIconColor || l.textColor,
                stroke: "none",
                opacity: "0.6"
              }, null, 8, Tl)) : c("", !0)
            ], 8, Rl)) : c("", !0),
            l.minimapMerged ? (i(), o(b, { key: 11 }, [
              v.value && v.value.firstPlot && v.value.firstPlot.value !== null ? (i(), o("circle", {
                key: 0,
                cx: v.value.firstPlot.x,
                cy: v.value.firstPlot.y,
                "stroke-width": "0.5",
                stroke: g.value,
                r: "4",
                fill: l.minimapLineColor
              }, null, 8, El)) : c("", !0),
              v.value && v.value.firstPlot && v.value.firstPlot.value !== null ? (i(), o("circle", {
                key: 1,
                cx: v.value.firstPlot.x,
                cy: v.value.firstPlot.y,
                r: "2",
                fill: g.value
              }, null, 8, Bl)) : c("", !0),
              v.value && v.value.lastPlot && v.value.lastPlot.value !== null ? (i(), o("circle", {
                key: 2,
                cx: v.value.lastPlot.x,
                cy: v.value.lastPlot.y,
                "stroke-width": "0.5",
                stroke: g.value,
                r: "4",
                fill: l.minimapLineColor
              }, null, 8, Wl)) : c("", !0),
              v.value && v.value.lastPlot && v.value.lastPlot.value !== null ? (i(), o("circle", {
                key: 3,
                cx: v.value.lastPlot.x,
                cy: v.value.lastPlot.y,
                r: "2",
                fill: g.value
              }, null, 8, Il)) : c("", !0)
            ], 64)) : (i(), o("g", Fl, [
              (i(!0), o(b, null, R(Q.value.filter((t) => t.type === "line"), (t) => (i(), o("g", {
                key: String(t.key)
              }, [
                t.isVisible ? (i(!0), o(b, { key: 0 }, R(t.selectionMarkers, (u) => (i(), o("circle", {
                  key: `sel-dot-${t.key}-${u.i}`,
                  cx: u.x,
                  cy: u.y,
                  r: "2.5",
                  fill: t.color,
                  stroke: g.value,
                  "stroke-width": "0.5"
                }, null, 8, Vl))), 128)) : c("", !0),
                t && t.firstPlot && t.isVisible && t.firstPlot.value !== null ? (i(), o("circle", {
                  key: 1,
                  cx: t.firstPlot.x,
                  cy: t.firstPlot.y,
                  "stroke-width": "0.5",
                  stroke: g.value,
                  r: "4",
                  fill: t.color
                }, null, 8, Dl)) : c("", !0),
                t && t.firstPlot && t.isVisible && t.firstPlot.value !== null ? (i(), o("circle", {
                  key: 2,
                  cx: t.firstPlot.x,
                  cy: t.firstPlot.y,
                  r: "2",
                  fill: g.value
                }, null, 8, zl)) : c("", !0),
                t && t.lastPlot && t.isVisible && t.lastPlot.value !== null ? (i(), o("circle", {
                  key: 3,
                  cx: t.lastPlot.x,
                  cy: t.lastPlot.y,
                  "stroke-width": "0.5",
                  stroke: g.value,
                  r: "4",
                  fill: t.color
                }, null, 8, Hl)) : c("", !0),
                t && t.lastPlot && t.isVisible && t.lastPlot.value !== null ? (i(), o("circle", {
                  key: 4,
                  cx: t.lastPlot.x,
                  cy: t.lastPlot.y,
                  r: "2",
                  fill: g.value
                }, null, 8, Xl)) : c("", !0)
              ]))), 128))
            ])),
            (i(!0), o(b, null, R(We.value, (t, u) => (i(), o("rect", {
              key: `trap-${u}`,
              y: 0,
              height: Math.max(f.value.height, 0),
              fill: "transparent",
              style: L([{ "pointer-events": "all !important" }, {
                cursor: t >= F.value && t < H.value && l.enableSelectionDrag ? I.value ? "grabbing" : "grab" : "default"
              }]),
              x: k.value * u - (l.minimapCompact ? k.value / 2 : 0),
              width: k.value < 0 ? 0 : k.value,
              onMousedown: a[0] || (a[0] = (r) => I.value = !0),
              onMouseup: a[1] || (a[1] = (r) => I.value = !1),
              onMouseenter: (r) => Lt(t),
              onMouseleave: a[2] || (a[2] = (r) => {
                A.value = null, Y("trapMouse", null);
              })
            }, null, 44, Gl))), 128))
          ], 8, ol))
        ])) : c("", !0),
        P("div", {
          class: "slider-track",
          style: L({ visibility: M.value && l.minimapCompact ? "hidden" : "visible" })
        }, null, 4),
        P("div", {
          class: Z({
            "range-highlight": !0,
            move: l.enableSelectionDrag
          }),
          onMousedown: a[3] || (a[3] = (t) => I.value = !0),
          onMouseup: a[4] || (a[4] = (t) => I.value = !1),
          style: L({
            ...N.value,
            cursor: I.value ? "grabbing" : "grab",
            visibility: M.value && l.minimapCompact ? "hidden" : "visible"
          })
        }, null, 38),
        l.enableRangeHandles ? at((i(), o("input", {
          key: 1,
          ref_key: "rangeStart",
          ref: we,
          type: "range",
          class: Z({
            "range-left": !0,
            "range-handle": !0,
            "range-minimap": M.value && l.verticalHandles,
            "range-invisible": M.value && l.minimapCompact
          }),
          min: l.min,
          max: $.value && l.minimapCompact ? q.value : l.max,
          step: 1,
          "onUpdate:modelValue": a[5] || (a[5] = (t) => xe.value = t),
          onMouseenter: a[6] || (a[6] = (t) => Qe("start"))
        }, null, 42, Ol)), [
          [
            nt,
            xe.value,
            void 0,
            { number: !0 }
          ]
        ]) : c("", !0),
        l.enableRangeHandles ? at((i(), o("input", {
          key: 2,
          ref_key: "rangeEnd",
          ref: Ce,
          type: "range",
          class: Z({
            "range-right": !0,
            "range-handle": !0,
            "range-minimap": M.value && l.verticalHandles,
            "range-invisible": M.value && l.minimapCompact
          }),
          min: l.min,
          max: $.value && l.minimapCompact ? q.value : l.max,
          step: 1,
          "onUpdate:modelValue": a[7] || (a[7] = (t) => Me.value = t),
          onMouseenter: a[8] || (a[8] = (t) => Qe("end"))
        }, null, 42, jl)), [
          [
            nt,
            Me.value,
            void 0,
            { number: !0 }
          ]
        ]) : c("", !0),
        l.labelLeft ? (i(), o("div", {
          key: 3,
          ref_key: "tooltipLeft",
          ref: le,
          class: Z({
            "range-tooltip": !0,
            "range-tooltip-visible": z.value,
            "range-tooltip-arrow": N.value.arrowLeft && !l.verticalHandles,
            "range-tooltip-arrow-left": !N.value.arrowLeft && !l.verticalHandles
          }),
          style: L({
            left: N.value.tooltipLeft,
            color: de(Ae)(l.selectColor),
            backgroundColor: l.selectColor,
            border: `1px solid ${g.value}`,
            zIndex: `${Ke.value + 4}`,
            visibility: ne.value || l.labelLeft === l.labelRight ? "hidden" : "visible",
            top: M.value && l.minimapCompact ? "calc(-100% - 18px)" : "-100%"
          })
        }, $e(l.labelLeft), 7)) : c("", !0),
        (ne.value || l.labelLeft === l.labelRight) && (l.labelLeft || l.labelRight) ? (i(), o("div", {
          key: 4,
          ref: "tooltipMerge",
          class: Z({
            "range-tooltip": !0,
            "range-tooltip-visible": z.value,
            "range-tooltip-arrow": !0,
            "range-tooltip-arrow-left": !N.value.arrowLeft && !l.verticalHandles,
            "range-tooltip-arrow-right": !N.value.arrowRight && !l.verticalHandles
          }),
          style: L({
            left: N.value.tooltipCenter,
            width: ie.value.width + "px",
            color: de(Ae)(l.selectColor),
            backgroundColor: l.selectColor,
            border: `1px solid ${g.value}`,
            zIndex: "4",
            top: M.value && l.minimapCompact ? "calc(-100% - 18px)" : "-100%"
          })
        }, $e(l.labelLeft === l.labelRight ? l.labelLeft : `${l.labelLeft} - ${l.labelRight}`), 7)) : c("", !0),
        l.labelRight ? (i(), o("div", {
          key: 5,
          ref_key: "tooltipRight",
          ref: ae,
          class: Z({
            "range-tooltip": !0,
            "range-tooltip-visible": z.value,
            "range-tooltip-arrow": N.value.arrowRight && !l.verticalHandles,
            "range-tooltip-arrow-right": !N.value.arrowRight && !l.verticalHandles
          }),
          style: L({
            left: N.value.tooltipRight,
            color: de(Ae)(l.selectColor),
            backgroundColor: l.selectColor,
            border: `1px solid ${g.value}`,
            zIndex: "4",
            visibility: ne.value || l.labelLeft === l.labelRight ? "hidden" : "visible",
            top: M.value && l.minimapCompact ? "calc(-100% - 18px)" : "-100%"
          })
        }, $e(l.labelRight), 7)) : c("", !0)
      ], 36)
    ], 44, ll));
  }
}, Ql = /* @__PURE__ */ tl(Ul, [["__scopeId", "data-v-c31208c3"]]);
export {
  Ql as S
};
