// vue-data-ui/dist/SlicerPreview-I2J0UUF_.js
import { useCssVars as wl, ref as f, computed as s, watch as ee, onMounted as Sl, onBeforeUnmount as He, onUpdated as Ct, nextTick as ve, openBlock as u, createElementBlock as r, normalizeStyle as L, createElementVNode as F, createVNode as Cl, renderSlot as Nt, createCommentVNode as c, unref as ke, Fragment as M, renderList as S, normalizeProps as Pt, mergeProps as pe, normalizeClass as te, withDirectives as Lt, withKeys as At, vModelText as $t, toDisplayString as Xe } from "vue";
import { j as Nl, T as Pl, X as Ll, W as Ke, L as Ft, c as Rt, M as Tt, b as Et, au as Bt } from "./lib-GwPnJSrY.js";
import { t as Ze, u as Al, d as $l } from "./useResponsive-DfdjqQps.js";
import Fl from "./BaseIcon-CNx8Tzl5.js";
import { _ as Rl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Tl = ["data-minimap"], El = {
  class: "vue-data-ui-slicer-labels",
  style: { position: "relative", "z-index": "1", "pointer-events": "none" }
}, Bl = {
  key: 0,
  style: { width: "100%", position: "relative" }
}, Il = {
  key: 0,
  class: "minimap",
  style: { width: "100%" }
}, Wl = ["xmlns", "viewBox"], Vl = ["id"], Dl = ["stop-color"], zl = ["width", "height", "stroke"], Hl = ["d", "stroke"], Xl = ["d", "fill"], Kl = ["x", "y", "width", "height", "fill"], Zl = ["d", "stroke", "stroke-dasharray"], jl = ["cx", "cy", "fill", "stroke"], Ul = ["cx", "cy", "fill", "stroke"], Ol = ["x", "width", "height", "fill", "rx"], ql = ["x", "width", "height", "rx", "fill"], Yl = ["x", "width", "height", "fill", "rx", "aria-valuemin", "aria-valuemax", "aria-valuenow", "aria-valuetext"], Gl = ["x2", "y1", "y2", "stroke"], Jl = { key: "merged-tree" }, Ql = ["d", "fill"], _l = ["d", "stroke"], ea = ["cx", "cy", "stroke", "fill"], ta = ["cx", "cy", "stroke", "fill"], la = { key: "split-tree" }, aa = ["x", "y", "width", "height", "fill"], na = ["d", "stroke", "stroke-dasharray"], ia = ["cx", "cy", "fill", "stroke"], ua = ["x", "width", "height", "fill", "stroke", "stroke-width"], ra = ["x", "width", "height"], oa = ["fill", "stroke"], sa = ["stroke"], ca = ["fill"], ma = ["x", "width", "height", "fill", "stroke", "stroke-width"], va = ["x", "width", "height"], da = ["fill", "stroke"], ha = ["stroke"], fa = ["fill"], ga = { key: "merged-tree" }, ya = ["cx", "cy", "stroke", "fill"], xa = ["cx", "cy", "fill"], ba = ["cx", "cy", "stroke", "fill"], Ma = ["cx", "cy", "fill"], ka = { key: 1 }, pa = ["d", "stroke", "stroke-dasharray"], wa = ["cx", "cy", "fill", "stroke"], Sa = ["cx", "cy", "stroke", "fill"], Ca = ["cx", "cy", "fill"], Na = ["cx", "cy", "stroke", "fill"], Pa = ["cx", "cy", "fill"], La = ["cx", "cy", "fill", "stroke"], Aa = ["x", "height", "width", "onMouseenter"], $a = ["min", "max", "tabindex"], Fa = ["min", "max", "tabindex"], je = 48, Ra = {
  __name: "SlicerPreview",
  props: {
    uuid: {
      type: String,
      default: ""
    },
    immediate: {
      type: Boolean,
      default: !0
    },
    background: {
      type: String,
      default: "#FFFFFF"
    },
    borderColor: {
      type: String,
      default: "#FFFFFF"
    },
    fontSize: {
      type: Number,
      default: 14
    },
    labelLeft: {
      type: [String, Number],
      default: ""
    },
    labelRight: {
      type: [String, Number],
      default: ""
    },
    textColor: {
      type: String,
      default: "#1A1A1A"
    },
    inputColor: {
      type: String,
      default: "#1A1A1A"
    },
    max: {
      type: Number,
      default: 0
    },
    min: {
      type: Number,
      default: 0
    },
    selectColor: {
      type: String,
      default: "#4A4A4A"
    },
    useResetSlot: {
      type: Boolean,
      default: !1
    },
    valueStart: {
      type: [Number, String],
      default: 0
    },
    valueEnd: {
      type: [Number, String],
      default: 0
    },
    minimap: {
      type: Array,
      default: []
    },
    smoothMinimap: {
      type: Boolean,
      default: !1
    },
    minimapSelectedColor: {
      type: String,
      default: "#1f77b4"
    },
    minimapSelectionRadius: {
      type: Number,
      default: 12
    },
    minimapLineColor: {
      type: String,
      default: "#2D353C"
    },
    minimapSelectedColorOpacity: {
      type: Number,
      default: 0.2
    },
    minimapSelectedIndex: {
      type: Number,
      default: null
    },
    minimapIndicatorColor: {
      type: String,
      default: "#2D353C"
    },
    refreshStartPoint: {
      type: Number,
      default: 0
    },
    refreshEndPoint: {
      type: Number,
      default: null
    },
    enableRangeHandles: {
      type: Boolean,
      default: !1
    },
    enableSelectionDrag: {
      type: Boolean,
      default: !0
    },
    verticalHandles: {
      type: Boolean,
      default: !1
    },
    timeLabels: {
      type: Array
    },
    isPreview: {
      type: Boolean,
      default: !1
    },
    preciseLabels: {
      type: Array,
      default() {
        return [];
      }
    },
    usePreciseLabels: {
      type: Boolean,
      default: !1
    },
    selectedSeries: {
      type: Object
    },
    customFormat: {
      type: [Function, null]
    },
    minimapCompact: {
      type: Boolean,
      default: !1
    },
    allMinimaps: {
      type: Array,
      default() {
        return [];
      }
    },
    minimapMerged: {
      type: Boolean,
      default: !1
    },
    minimapFrameColor: {
      type: String,
      default: "#e1e5e8"
    },
    cutNullValues: {
      type: Boolean,
      default: !1
    },
    focusOnDrag: {
      type: Boolean,
      default: !1
    },
    focusRangeRatio: {
      type: Number,
      default: 0.1
    },
    minScale: {
      type: Number,
      default: null
    },
    maxScale: {
      type: Number,
      default: null
    },
    // Used to center zero line in VueUiStackBar (non distributed mode)
    forceZeroCenter: {
      type: Boolean,
      default: !1
    },
    maxWidth: {
      type: Number,
      default: null
    },
    isCursorPointer: {
      type: Boolean,
      default: !1
    },
    additionalMinimapHeight: {
      type: Number,
      default: 0
    },
    handleType: {
      type: String,
      default: ""
    },
    handleWidth: {
      type: Number,
      default: 20
    },
    handleBorderWidth: {
      type: Number,
      default: 1
    },
    handleIconColor: {
      type: String,
      default: null
    },
    handleBorderColor: {
      type: String,
      default: null
    },
    handleFill: {
      type: String,
      default: null
    }
  },
  emits: ["futureStart", "futureEnd", "update:start", "update:end", "reset", "trapMouse"],
  setup(n, { expose: It, emit: Wt }) {
    wl((e) => ({
      v739151cd: Dt.value,
      v0b86da41: b.value,
      v5dc2f2f2: Ht.value,
      v2887c91e: n.selectColor,
      v27a3657f: zt.value
    }));
    const a = n, P = f(null), g = f(a.min), y = f(a.max), x = s(() => !!a.allMinimaps.length), we = f(Nl()), T = f(!1), j = s(() => x.value && a.minimapCompact), Ue = f(0), Se = Ze((e) => E("futureStart", e), 0), Ce = Ze((e) => E("futureEnd", e), 0), U = s({
      get: () => g.value,
      set(e) {
        const l = Math.min(e, y.value - 1);
        l !== g.value && (g.value = l, ae.value && (ae.value.value = String(l)), a.immediate ? E("update:start", Number(l)) : T.value && Se(l));
      }
    }), O = s({
      get: () => y.value,
      set(e) {
        const l = Math.max(e, g.value + 1);
        l !== y.value && (y.value = l, ne.value && (ne.value.value = String(l)), a.immediate ? E("update:end", Number(l)) : T.value && Ce(l));
      }
    });
    let Oe = null;
    function A() {
      clearTimeout(Oe), a.immediate || (E("update:start", Number(g.value)), E("update:end", Number(y.value))), T.value = !1;
    }
    const Vt = s(() => a.refreshEndPoint === null ? a.max : a.refreshEndPoint), E = Wt, de = s(() => {
      if (j.value) {
        const l = Math.max(1, Y.value - 1);
        return fe.value / l * 100;
      }
      const e = Math.max(1, a.max - a.min);
      return (g.value - a.min) / e * 100;
    }), Ne = s(() => {
      if (j.value) {
        const l = Math.max(1, Y.value - 1);
        return ge.value / l * 100;
      }
      const e = Math.max(1, a.max - a.min);
      return (y.value - a.min) / e * 100;
    }), Pe = s(() => (de.value + Ne.value) / 2), Le = s(() => {
      if (!P.value) return !1;
      const e = Math.max(1, a.max - a.min);
      return P.value.getBoundingClientRect().width * ((g.value - a.min) / e) - ye.value / 2 < 0;
    }), Ae = s(() => {
      if (!P.value) return !1;
      const e = Math.max(1, a.max - a.min), l = P.value.getBoundingClientRect().width;
      return l * ((y.value - a.min) / e) + oe.value / 2 > l;
    }), R = s(() => {
      const e = Ae.value ? `calc(${Pe.value}% - ${ce.value.width}px)` : Le.value ? `calc(${Pe.value}% - 8px)` : `calc(${Pe.value}% - ${ce.value.width / 2}px)`;
      return {
        left: `${de.value}%`,
        width: `${Math.max(0, Ne.value - de.value)}%`,
        background: a.selectColor,
        tooltipLeft: `calc(${de.value}% - ${Le.value ? 9 : ye.value / 2}px)`,
        tooltipRight: `calc(${Ne.value}% - ${Ae.value ? oe.value - 9 : oe.value / 2}px)`,
        tooltipCenter: e,
        arrowLeft: !Le.value,
        arrowRight: !Ae.value
      };
    }), Dt = s(() => a.inputColor), zt = s(() => a.background), Ht = s(() => `${a.selectColor}33`), b = s(() => a.borderColor), qe = s(() => {
      let e = [];
      for (let l = 0; l < a.max; l += 1)
        e.push(l);
      return e;
    });
    function Ye() {
      E("reset");
    }
    ee(
      () => a.min,
      (e) => {
        Number(g.value) < Number(e) && (g.value = Number(e)), Number(y.value) < Number(e) && (y.value = Number(e));
      }
    ), ee(
      () => a.max,
      (e) => {
        Number(g.value) > Number(e) && (g.value = Number(e)), Number(y.value) > Number(e) && (y.value = Number(e));
      }
    );
    const he = f(null), d = f({
      width: 1,
      height: 1
    }), q = f(null);
    Sl(() => {
      if (x.value) {
        const e = Ze(() => {
          if (!he.value) return;
          const { width: l, height: t } = Al({ chart: he.value }), i = Math.max(0, Math.round(l)), o = Math.max(0, Math.round(t - 47));
          (i !== d.value.width || o !== d.value.height) && (d.value.width = i, d.value.height = o + a.additionalMinimapHeight);
        }, 0);
        q.value = new ResizeObserver(e), q.value.observe(he.value);
      }
    }), He(() => {
      q.value && q.value.disconnect();
    });
    const Y = s(() => Math.max(1, a.max - a.min));
    function Ge(e) {
      const l = Math.floor(e - a.min);
      return Math.min(Math.max(0, l), Y.value);
    }
    function Xt(e) {
      const l = Math.ceil(e - a.min);
      return Math.min(Math.max(0, l), Y.value);
    }
    const K = s(() => Ge(g.value)), G = s(() => Xt(y.value)), $e = s(() => Math.max(...a.allMinimaps.map((e) => e.series.length))), Kt = s(() => x.value && a.minimapCompact ? 40 : 0), Zt = s(() => Kt.value / 2);
    s(() => Math.max(1, d.value.width - Zt.value * 2));
    const w = s(() => {
      const e = Math.max(1, $e.value - (a.minimapCompact ? 1 : 0));
      return d.value.width / e;
    }), Je = s(() => a.allMinimaps.length ? a.allMinimaps.filter((e) => e.type === "bar" && e.isVisible).length : 0), Fe = s(() => w.value / (Je.value || 1) * 0.8);
    function Qe(e, l, t) {
      const i = Fe.value, o = Math.max(1, Je.value), m = $e.value - 1;
      return t === 0 ? e + i / 2 * l : t === m ? e - i / 2 * (o - l) : e - o * i / 2 + l * i;
    }
    function _e(e, l) {
      return [0, $e.value - 1].includes(l) ? Fe.value / 2 : Fe.value;
    }
    const jt = s(() => {
      const e = [];
      if (Array.isArray(a.minimap) && a.minimap.length && a.minimapMerged && e.push(...a.minimap.filter(Number.isFinite)), Array.isArray(a.allMinimaps) && a.allMinimaps.length)
        for (const l of a.allMinimaps)
          l?.isVisible && Array.isArray(l?.series) && e.push(...l.series.filter(Number.isFinite));
      return e.length ? { min: Math.min(...e), max: Math.max(...e) } : { min: 0, max: 1 };
    }), et = s(() => {
      if (a.minScale == null && a.forceZeroCenter) return null;
      const e = Number(a.minScale);
      return Number.isFinite(e) ? e : null;
    }), tt = s(() => {
      if (a.maxScale == null && a.forceZeroCenter) return null;
      const e = Number(a.maxScale);
      return Number.isFinite(e) ? e : null;
    }), Re = s(() => et.value !== null && tt.value !== null), lt = s(() => {
      const { min: e, max: l } = jt.value, t = et.value, i = tt.value;
      let o, m;
      return t !== null && i !== null ? (o = Math.min(t, e), m = Math.max(i, l)) : (o = t !== null ? t : e, m = i !== null ? i : l), Number.isFinite(o) || (o = 0), Number.isFinite(m) || (m = 1), o === m ? m = o + 1 : o > m && ([o, m] = [m, o]), { min: o, max: m };
    }), J = s(() => lt.value.min), le = s(() => lt.value.max), Ut = s(() => J.value < 0 && le.value > 0 || le.value <= 0 ? J.value : 0), Ot = (e) => {
      const l = Math.max(1, d.value.height);
      return at(
        J.value,
        le.value,
        l,
        Re.value
      )(e);
    }, Te = s(() => Ot(0));
    function at(e, l, t, i = !1) {
      const o = (h, N, $) => Math.max(N, Math.min($, h)), B = Math.max(1e-9, l - e);
      if (i)
        return (h) => {
          const N = (h - e) / B;
          return t - N * t;
        };
      if (l <= 0) {
        const h = Math.max(1e-9, 0 - e);
        return (N) => t - (N - e) / h * t;
      } else if (e >= 0) {
        const h = Math.max(1e-9, l - 0);
        return (N) => t - (N - 0) / h * t;
      } else {
        const h = Math.max(1e-9, Math.max(Math.abs(e), Math.abs(l)));
        return (N) => (1 - (o(N / h, -1, 1) + 1) / 2) * t;
      }
    }
    function nt(e, l = !1) {
      if (!e || !e.length)
        return {
          fullSet: "",
          points: [],
          selectionSet: "",
          sliced: [],
          firstPlot: null,
          lastPlot: null,
          hasFull: !1,
          hasSelection: !1,
          fullMarkers: [],
          selectionMarkers: [],
          dashed: !1
        };
      const t = Math.max(1, d.value.height), i = at(J.value, le.value, t, Re.value), o = e.length, m = Math.min(Math.max(0, K.value), Math.max(0, o - 1)), B = Math.min(o, Math.max(m + 1, G.value)), h = e.map((p, wt) => {
        const me = p, St = Number.isFinite(me), gl = w.value * wt + (a.minimapCompact ? 0 : w.value / 2), yl = (Ml, kl, pl) => Math.max(kl, Math.min(pl, Ml)), xl = Re.value ? yl(0, J.value, le.value) : 0, bl = i(xl);
        return {
          x: gl,
          y: St ? i(me) : NaN,
          v: me,
          value: me == null ? null : St ? me : null,
          y0: bl,
          i: wt
        };
      }), N = (p) => p >= 0 && p < h.length && Number.isFinite(h[p]?.value), $ = h.filter((p) => Number.isFinite(p.value) && !N(p.i - 1) && !N(p.i + 1)), Me = $.filter((p) => p.i >= m && p.i < B), Z = h.slice(m, B), hl = h.length >= 2 ? a.smoothMinimap || l ? a.cutNullValues ? Ft(h) : Rt(h.filter((p) => p.value != null)) : a.cutNullValues ? Tt(h) : Et(h.filter((p) => p.value != null)) : "", fl = Z.length >= 2 ? a.smoothMinimap || l ? a.cutNullValues ? Ft(Z) : Rt(Z.filter((p) => p.value != null)) : a.cutNullValues ? Tt(Z) : Et(Z.filter((p) => p.value != null)) : "";
      return {
        fullSet: hl,
        points: h,
        selectionSet: fl,
        sliced: Z,
        firstPlot: h[m] || null,
        lastPlot: h[Math.max(0, B - 1)] || null,
        hasFull: h.length >= 2,
        hasSelection: Z.length >= 2,
        fullMarkers: $,
        selectionMarkers: Me
      };
    }
    const v = s(() => a.minimap.length ? nt(a.minimap) : []), I = s(() => a.allMinimaps.length ? a.allMinimaps.map((e, l) => {
      const t = nt(e?.series || [], !!e.smooth), i = e?.id ?? e?.name ?? l;
      return {
        key: typeof i == "object" ? JSON.stringify(i) : String(i),
        color: e?.color,
        ...t,
        temperatureColors: e?.temperatureColors ?? null,
        isVisible: e.isVisible,
        type: e.type || "line",
        dashed: e.dashed ?? !1
      };
    }) : []), C = s(() => {
      const e = K.value, l = Math.max(e + 1, G.value);
      return {
        x: w.value * e + (a.minimapCompact ? 0 : w.value / 2),
        width: w.value * (l - e) - w.value
      };
    }), W = f(a.minimapSelectedIndex), it = (e) => Math.round(a.min + e), fe = s({
      get() {
        return j.value ? K.value : Number(U.value);
      },
      set(e) {
        if (j.value) {
          const l = Math.round(+e || 0);
          Be(it(l));
        } else {
          let l = Math.round(+e || 0);
          const t = Number(y.value) - 1, i = Math.min(Math.max(a.min, l), t);
          ae.value && (ae.value.valueAsNumber = i), Be(i);
        }
      }
    }), ge = s({
      get() {
        return j.value ? Math.max(K.value, G.value - 1) : Number(O.value);
      },
      set(e) {
        if (j.value) {
          const l = Math.round(+e || 0);
          We(it(l + 1));
        } else {
          let l = Math.round(+e || 0);
          const t = Number(g.value) + 1, i = Math.max(t, Math.min(l, a.max));
          ne.value && (ne.value.valueAsNumber = i), We(i);
        }
      }
    });
    function qt(e) {
      W.value = Ge(a.valueStart) + e;
    }
    const Yt = $l(qt, 60);
    ee(() => a.minimapSelectedIndex, (e, l) => {
      if ([null, void 0].includes(e)) {
        W.value = null;
        return;
      }
      e !== l && Yt(e);
    }, { immediate: !0 });
    function Gt(e) {
      W.value = e;
      const l = K.value, t = G.value;
      e >= l && e < t && !X.value && E("trapMouse", e - l);
    }
    const ae = f(null), ne = f(null);
    function ut(e) {
      if (typeof e == "object" && e && "target" in e) {
        const t = e.target, i = "valueAsNumber" in t ? t.valueAsNumber : +t.value;
        return Number.isFinite(i) ? i : NaN;
      }
      const l = +e;
      return Number.isFinite(l) ? l : NaN;
    }
    let Ee = 0;
    function Be(e) {
      T.value = !0;
      const l = ut(e);
      Number.isFinite(l) && (cancelAnimationFrame(Ee), Ee = requestAnimationFrame(() => {
        U.value = l;
      }));
    }
    let Ie = 0;
    function We(e) {
      T.value = !0;
      const l = ut(e);
      Number.isFinite(l) && (cancelAnimationFrame(Ie), Ie = requestAnimationFrame(() => {
        O.value = l;
      }));
    }
    He(() => {
      cancelAnimationFrame(Ee), cancelAnimationFrame(Ie);
    });
    const ie = s(() => a.valueEnd - a.valueStart), Jt = s(() => ie.value < a.max - a.min), Q = f(!1);
    let Qt = f(null);
    const _t = s(() => (Ue.value - 48) / (a.max - a.min) * ie.value), el = s(() => Math.max(1, Ue.value - je - _t.value)), tl = s(() => Math.max(1, a.max - a.min - ie.value)), ll = s(() => tl.value / el.value), rt = f(0), ot = f(0), al = f(0), nl = f(0);
    let V = null, D = null, z = null, H = null;
    const st = f(a.min);
    function Ve(e) {
      if (!P.value) return a.min;
      const l = P.value.getBoundingClientRect(), t = l.left + je / 2, i = l.right - je / 2, o = Math.max(1, i - t), B = (Math.max(t, Math.min(e, i)) - t) / o, h = Math.max(1, a.max - a.min);
      return Math.round(a.min + B * h);
    }
    const ct = async (e) => {
      if (T.value = !0, _.value = !0, !a.enableSelectionDrag) return;
      const l = e.type === "touchstart";
      l || e.stopPropagation();
      const t = l && e.targetTouches && e.targetTouches[0] ? e.targetTouches[0] : null, i = l ? t ? t.target : null : e.target;
      if (!i || !(i instanceof Element) || i.classList && i.classList.contains("range-handle")) return;
      Q.value = !0;
      const o = l ? t ? t.clientX : 0 : e.clientX;
      if (Qt.value = o, rt.value = o, a.focusOnDrag && !Jt.value && P.value) {
        st.value = Ve(o);
        const m = Math.min(0.95, Math.max(0.05, a.focusRangeRatio)), B = a.max - a.min, h = Math.max(1, Math.round(B * m)), N = Math.floor(h / 2);
        let $ = st.value - N;
        $ = Math.max(a.min, Math.min($, a.max - h));
        const Me = Math.min(a.max, $ + h);
        U.value = $, O.value = Me, Se($), Ce(Me), Bt(P.value, "mouseup"), await ve(), Bt(P.value, "mousedown", { clientX: o });
      }
      ot.value = g.value, al.value = y.value, nl.value = ll.value, V = l ? "touchmove" : "mousemove", D = l ? "touchend" : "mouseup", z = l ? ul : il, H = l ? ol : rl, window.addEventListener(V, z, { passive: !1 }), window.addEventListener(D, H);
    };
    function il(e) {
      Q.value && mt(e.clientX);
    }
    function ul(e) {
      if (!Q.value || !P.value) return;
      const l = e.target;
      if (!(l instanceof Element) || !P.value.contains(l) || l.classList && l.classList.contains("range-handle")) return;
      e.preventDefault();
      const t = e.targetTouches && e.targetTouches[0] ? e.targetTouches[0] : null;
      t && mt(t.clientX);
    }
    function mt(e) {
      if (!Q.value) return;
      const l = Ve(rt.value), i = Ve(e) - l;
      let o = Math.round(ot.value + i);
      o = Math.max(a.min, Math.min(o, a.max - ie.value));
      const m = o + ie.value;
      U.value = o, O.value = m, Se(o), Ce(m);
    }
    function rl() {
      vt();
    }
    function ol() {
      vt();
    }
    function vt() {
      Q.value = !1, V && z && window.removeEventListener(V, z), D && H && window.removeEventListener(D, H), V = D = null, z = H = null, A();
    }
    const X = f(!1), ue = f(null), re = f(null), ye = f(1), oe = f(1), _ = f(!1);
    function De() {
      if (ue.value) {
        const e = Math.round(ue.value.getBoundingClientRect().width);
        e !== ye.value && (ye.value = e);
      }
    }
    function ze() {
      if (re.value) {
        const e = Math.round(re.value.getBoundingClientRect().width);
        e !== oe.value && (oe.value = e);
      }
    }
    Ct(() => {
      De(), ze();
    });
    const dt = f(0);
    function ht(e) {
      dt.value = e === "start" ? 1 : 0;
    }
    const se = f(!1), ce = f({
      width: 0,
      left: 0
    });
    ee([g, y], async () => {
      if (await ve(), !ue.value || !re.value) {
        se.value = !1, ce.value = { width: 0, left: 0 };
        return;
      }
      const e = ue.value.getBoundingClientRect(), l = re.value.getBoundingClientRect();
      se.value = e.x + e.width > l.x;
      const t = e.x + e.width / 2, i = l.x + l.width / 2, o = e.width + l.width, m = (t + i) / 2;
      ce.value = {
        width: o,
        left: m - o / 2
      };
    }), Ct(() => {
      De(), ze();
    }), ee(() => a.labelLeft, () => {
      ve(De);
    }, { deep: !0 }), ee(() => a.labelRight, () => {
      ve(ze);
    }, { deep: !0 });
    const k = s(() => {
      let e = "", l = "", t = !1;
      if (Pl(a.customFormat))
        try {
          const i = a.customFormat({
            absoluteIndex: g.value,
            seriesIndex: g.value,
            datapoint: a.selectedSeries,
            timeLabel: a.preciseLabels[g.value],
            side: "left"
          }), o = a.customFormat({
            absoluteIndex: y.value - 1,
            seriesIndex: -1,
            datapoint: a.selectedSeries,
            timeLabel: a.preciseLabels[y.value - 1],
            side: "right"
          });
          typeof i == "string" && typeof o == "string" && (e = i, l = o, t = !0);
        } catch {
          t = !1;
        }
      if (!t) {
        const i = a.usePreciseLabels ? a.preciseLabels.find((m) => m.absoluteIndex === g.value) : a.timeLabels.find((m) => m.absoluteIndex === g.value), o = a.usePreciseLabels ? a.preciseLabels.find((m) => m.absoluteIndex === y.value - 1) : a.timeLabels.find((m) => m.absoluteIndex === y.value - 1);
        e = i ? i.text : "", l = o ? o.text : "";
      }
      return { left: e, right: l };
    });
    He(() => {
      q.value && q.value.disconnect(), V && z && window.removeEventListener(V, z), D && H && window.removeEventListener(D, H), V = D = null, z = H = null, clearTimeout(Oe);
    });
    const sl = s(() => {
      if (!qe.value.length) return null;
      if (W.value >= K.value && W.value < G.value) {
        const e = W.value, l = w.value * e + (a.minimapCompact ? 0 : w.value / 2);
        return {
          x1: l,
          x2: l,
          y1: 0,
          y2: Math.max(d.value.height, 0),
          stroke: a.minimapIndicatorColor,
          "stroke-linecap": "round",
          "stroke-dasharray": 2,
          "stroke-width": 1
        };
      }
      return null;
    }), ft = s(() => Number(g.value) < Number(y.value) - 1), gt = s(() => Number(g.value) > Number(a.min)), yt = s(() => Number(y.value) < Number(a.max)), xt = s(() => Number(y.value) > Number(g.value) + 1);
    function bt(e) {
      const l = Number(g.value) + e;
      e > 0 && !ft.value || e < 0 && !gt.value || (T.value = !0, U.value = l, A());
    }
    function Mt(e) {
      const l = Number(y.value) + e;
      e > 0 && !yt.value || e < 0 && !xt.value || (T.value = !0, O.value = l, A());
    }
    const xe = {
      plus: () => bt(1),
      minus: () => bt(-1),
      canPlus: ft,
      canMinus: gt
    }, be = {
      plus: () => Mt(1),
      minus: () => Mt(-1),
      canPlus: yt,
      canMinus: xt
    };
    function cl(e) {
      if (!e || kt(e.target)) return;
      const l = e.key, t = l === "ArrowLeft" || l === "ArrowDown" || l === "-" || l === "Subtract", i = l === "ArrowRight" || l === "ArrowUp" || l === "+" || l === "Add";
      !t && !i || (e.preventDefault(), e.stopPropagation(), i ? (xe.plus(), be.plus()) : (xe.minus(), be.minus()));
    }
    function kt(e) {
      return e instanceof HTMLElement && (e.isContentEditable || e.tagName === "INPUT" || e.tagName === "TEXTAREA" || e.tagName === "SELECT");
    }
    function ml(e) {
      if (!x.value || !a.minimapCompact) return;
      const l = P.value;
      if (!l) return;
      const t = e === "start" ? '[data-cy="slicer-compact-handle-left"]' : '[data-cy="slicer-compact-handle-right"]', i = l.querySelector(t);
      i && i instanceof SVGElement && typeof i.focus == "function" && i.focus();
    }
    function pt(e, l) {
      if (!l || kt(l.target)) return;
      const t = l.key, i = t === "ArrowLeft" || t === "ArrowDown" || t === "-" || t === "Subtract", o = t === "ArrowRight" || t === "ArrowUp" || t === "+" || t === "Add";
      !i && !o || (l.preventDefault(), l.stopPropagation(), e === "start" ? o ? xe.plus() : xe.minus() : o ? be.plus() : be.minus(), ve(() => ml(e)));
    }
    const vl = s(() => ({
      tabindex: 0,
      role: "slider",
      "aria-label": "Range start",
      "aria-valuemin": Number(a.min),
      "aria-valuemax": Math.max(Number(a.min), Number(y.value) - 1),
      "aria-valuenow": Number(g.value)
    })), dl = s(() => ({
      tabindex: 0,
      role: "slider",
      "aria-label": "Range end",
      "aria-valuemin": Math.min(Number(a.max), Number(g.value) + 1),
      "aria-valuemax": Number(a.max),
      "aria-valuenow": Number(y.value)
    }));
    return It({
      setStartValue: Be,
      setEndValue: We
    }), (e, l) => (u(), r("div", {
      "data-minimap": x.value,
      "data-dom-to-png-ignore": "",
      style: L([{ padding: "0 48px" }, {
        maxWidth: n.maxWidth ? n.maxWidth + "px" : void 0,
        margin: n.maxWidth ? "0 auto" : void 0
      }]),
      class: "vue-data-ui-zoom",
      ref_key: "zoomWrapper",
      ref: P,
      onMousedown: ct,
      onTouchstart: ct
    }, [
      F("div", El, [
        n.valueStart !== n.refreshStartPoint || n.valueEnd !== Vt.value ? (u(), r("div", Bl, [
          n.useResetSlot ? Nt(e.$slots, "reset-action", {
            key: 1,
            reset: Ye
          }, void 0, !0) : (u(), r("button", {
            key: 0,
            tabindex: "0",
            role: "button",
            class: "vue-data-ui-refresh-button",
            style: L({
              top: x.value ? "36px" : "-16px",
              pointerEvents: "all !important",
              cursor: n.isCursorPointer ? "pointer" : "default"
            }),
            onClick: Ye
          }, [
            Cl(Fl, {
              name: "refresh",
              stroke: n.textColor
            }, null, 8, ["stroke"])
          ], 4))
        ])) : c("", !0)
      ]),
      F("div", {
        class: "double-range-slider",
        ref_key: "minimapWrapper",
        ref: he,
        style: L([{ "z-index": "0" }, x.value ? {
          "--minimap-unit-px": w.value + "px",
          "--minimap-offset-px": (n.minimapCompact ? 0 : w.value / 2) + "px"
        } : void 0]),
        onMouseenter: l[15] || (l[15] = (t) => _.value = !0),
        onMouseleave: l[16] || (l[16] = (t) => _.value = !1)
      }, [
        x.value ? (u(), r("div", Il, [
          (u(), r("svg", {
            key: `mm-${n.minimapMerged ? "merged" : "split"}-${n.minimapCompact ? "compact" : "normal"}`,
            xmlns: ke(Ll),
            viewBox: `0 0 ${Math.max(0, d.value.width)} ${Math.max(0, d.value.height)}`
          }, [
            F("defs", null, [
              F("linearGradient", {
                id: we.value,
                x1: "0%",
                y1: "0%",
                x2: "0%",
                y2: "100%"
              }, [
                F("stop", {
                  offset: "0%",
                  "stop-color": `${n.minimapLineColor}50`
                }, null, 8, Dl),
                l[17] || (l[17] = F("stop", {
                  offset: "100%",
                  "stop-color": "transparent"
                }, null, -1))
              ], 8, Vl)
            ]),
            n.minimapCompact ? (u(), r("rect", {
              key: 0,
              class: "vue-ui-zoom-minimap-frame",
              x: 0,
              y: 0,
              width: d.value.width,
              height: d.value.height,
              fill: "none",
              stroke: n.minimapFrameColor,
              rx: 3
            }, null, 8, zl)) : c("", !0),
            e.$slots.slotMap ? c("", !0) : (u(), r(M, { key: 1 }, [
              n.minimapMerged ? (u(), r("path", {
                key: 0,
                d: `M${v.value.fullSet}`,
                stroke: `${n.minimapLineColor}`,
                fill: "none",
                "stroke-width": "1",
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                style: { opacity: "0.6" }
              }, null, 8, Hl)) : c("", !0),
              n.minimapMerged && !n.minimapCompact ? (u(), r("path", {
                key: 1,
                d: `M${w.value / 2},${Math.max(d.value.height, 0)} ${v.value.fullSet} L${d.value.width - w.value / 2},${Math.max(d.value.height, 0)}Z`,
                fill: `url(#${we.value})`,
                stroke: "none",
                style: { opacity: "0.6" }
              }, null, 8, Xl)) : n.minimapMerged ? c("", !0) : (u(), r(M, { key: 2 }, [
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "bar" && t.isVisible), (t, i) => (u(), r("g", null, [
                  (u(!0), r(M, null, S(t.points, (o, m) => (u(), r(M, null, [
                    t && !isNaN(o.y) ? (u(), r("rect", {
                      key: 0,
                      x: Qe(o.x, i, m),
                      y: o.v >= 0 ? o.y : o.y0,
                      width: _e(i, m),
                      height: o.v >= 0 ? o.y0 - o.y : o.y - o.y0,
                      fill: t.color,
                      style: { opacity: 0.6 }
                    }, null, 8, Kl)) : c("", !0)
                  ], 64))), 256))
                ]))), 256)),
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "line"), (t) => (u(), r("g", null, [
                  t.isVisible ? (u(), r("path", {
                    key: 0,
                    d: `M ${t.fullSet}`,
                    fill: "none",
                    stroke: t.color,
                    style: { opacity: "0.6" },
                    "stroke-dasharray": t.dashed ? "2 4" : 0
                  }, null, 8, Zl)) : c("", !0),
                  t.isVisible && n.cutNullValues ? (u(!0), r(M, { key: 1 }, S(t.fullMarkers, (i) => (u(), r("circle", {
                    key: `sel-dot-under-${t.key}-${i.i}`,
                    cx: i.x,
                    cy: i.y,
                    r: "2",
                    fill: t.color,
                    stroke: b.value,
                    "stroke-width": "0.5",
                    style: { opacity: "0.6" }
                  }, null, 8, jl))), 128)) : c("", !0)
                ]))), 256)),
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "plot"), (t) => (u(), r("g", null, [
                  (u(!0), r(M, null, S(t.points, (i) => (u(), r("g", null, [
                    t.isVisible && i.value !== null ? (u(), r("circle", {
                      key: `sel-plot-under-${t.key}-${i.i}`,
                      cx: i.x,
                      cy: i.y,
                      r: "2",
                      fill: t.color,
                      stroke: b.value,
                      "stroke-width": "0.5",
                      style: { opacity: "0.6" }
                    }, null, 8, Ul)) : c("", !0)
                  ]))), 256))
                ]))), 256))
              ], 64))
            ], 64)),
            F("rect", {
              x: C.value.x,
              y: 0,
              width: Math.max(0, C.value.width),
              height: Math.max(d.value.height, 0),
              fill: b.value,
              rx: n.minimapSelectionRadius,
              stroke: "none"
            }, null, 8, Ol),
            F("rect", {
              x: C.value.x,
              y: 0,
              width: C.value.width < 0 ? 0 : C.value.width,
              height: Math.max(d.value.height, 0),
              rx: n.minimapSelectionRadius,
              fill: b.value,
              style: L({ opacity: Q.value || T.value ? 0 : 1 })
            }, null, 12, ql),
            F("rect", {
              x: C.value.x,
              y: 0,
              width: C.value.width < 0 ? 0 : C.value.width,
              height: Math.max(d.value.height, 0),
              fill: n.minimapSelectedColor,
              rx: n.minimapSelectionRadius,
              style: L({ opacity: n.minimapSelectedColorOpacity }),
              tabindex: "0",
              role: "slider",
              "aria-label": "Selected range",
              "aria-valuemin": Number(a.min),
              "aria-valuemax": Number(a.max),
              "aria-valuenow": Number(g.value),
              "aria-valuetext": k.value.left && k.value.right ? `${k.value.left} – ${k.value.right}` : void 0,
              onKeydown: cl
            }, null, 44, Yl),
            !n.minimapMerged && Ut.value < 0 ? (u(), r("line", {
              key: 2,
              class: "slicer-minimap-zero-line",
              x1: 0,
              x2: d.value.width,
              y1: Te.value,
              y2: Te.value,
              stroke: n.minimapFrameColor,
              "stroke-width": "0.5"
            }, null, 8, Gl)) : c("", !0),
            e.$slots.slotMap ? c("", !0) : (u(), r(M, { key: 3 }, [
              n.minimapMerged ? (u(), r("g", Jl, [
                v.value && v.value.sliced && v.value.sliced.length ? (u(), r(M, { key: 0 }, [
                  v.value.selectionSet ? (u(), r("path", {
                    key: 0,
                    d: `M${v.value.sliced[0].x},${Math.max(d.value.height, 0)} ${v.value.selectionSet} L${v.value.sliced[v.value.sliced.length - 1].x},${Math.max(d.value.height, 0)}Z`,
                    fill: `url(#${we.value})`,
                    stroke: "none",
                    style: { opacity: "1" }
                  }, null, 8, Ql)) : c("", !0),
                  v.value.selectionSet ? (u(), r("path", {
                    key: 1,
                    d: `M ${v.value.selectionSet}`,
                    stroke: `${n.minimapLineColor}`,
                    fill: "transparent",
                    "stroke-width": "2",
                    "stroke-linecap": "round",
                    "stroke-linejoin": "round"
                  }, null, 8, _l)) : c("", !0)
                ], 64)) : c("", !0),
                v.value && v.value.firstPlot ? (u(), r("circle", {
                  key: 1,
                  cx: v.value.firstPlot.x,
                  cy: v.value.firstPlot.y,
                  "stroke-width": "0.5",
                  stroke: b.value,
                  r: "3",
                  fill: n.minimapLineColor
                }, null, 8, ea)) : c("", !0),
                v.value && v.value.lastPlot ? (u(), r("circle", {
                  key: 2,
                  cx: v.value.lastPlot.x,
                  cy: v.value.lastPlot.y,
                  "stroke-width": "0.5",
                  stroke: b.value,
                  r: "3",
                  fill: n.minimapLineColor
                }, null, 8, ta)) : c("", !0)
              ])) : (u(), r("g", la, [
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "bar" && t.isVisible), (t, i) => (u(), r("g", null, [
                  (u(!0), r(M, null, S(t.points, (o, m) => (u(), r(M, null, [
                    t && !isNaN(o.y) ? (u(), r("rect", {
                      key: 0,
                      x: Qe(o.x, i, m),
                      y: o.v >= 0 ? o.y : o.y0,
                      width: _e(i, m),
                      height: o.v >= 0 ? o.y0 - o.y : o.y - o.y0,
                      fill: t.color,
                      style: L({ opacity: m >= U.value && m < O.value ? 1 : 0 })
                    }, null, 12, aa)) : c("", !0)
                  ], 64))), 256))
                ]))), 256)),
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "line"), (t, i) => (u(), r("g", {
                  key: String(t.key)
                }, [
                  t && t.hasSelection && t.selectionSet && t.isVisible ? (u(), r("path", {
                    key: 0,
                    d: `M ${t.selectionSet}`,
                    stroke: t.temperatureColors ? `url(#temperature_grad_line_${i}_${n.uuid})` : t.color,
                    fill: "transparent",
                    "stroke-width": "2",
                    "stroke-linecap": "round",
                    "stroke-linejoin": "round",
                    "stroke-dasharray": t.dashed ? "2 4" : 0
                  }, null, 8, na)) : c("", !0)
                ]))), 128)),
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "plot"), (t) => (u(), r("g", {
                  key: String(t.key)
                }, [
                  (u(!0), r(M, null, S(t.points, (i) => (u(), r("g", null, [
                    t.isVisible && i.value !== null ? (u(), r("circle", {
                      key: 0,
                      cx: i.x,
                      cy: i.y,
                      r: "2",
                      fill: t.color,
                      stroke: b.value,
                      "stroke-width": "0.5",
                      style: { opacity: "0.6" }
                    }, null, 8, ia)) : c("", !0)
                  ]))), 256))
                ]))), 128))
              ]))
            ], 64)),
            e.$slots.slotMap ? Nt(e.$slots, "slotMap", Pt(pe({ key: 4 }, {
              width: Math.max(0, d.value.width),
              height: Math.max(0, d.value.height),
              zeroY: Te.value,
              unitW: Math.max(0, w.value)
            })), void 0, !0) : c("", !0),
            x.value && n.minimapCompact ? (u(), r("rect", pe({
              key: 5,
              class: "vue-ui-zoom-compact-minimap-handle",
              x: C.value.x - Math.min(40, Math.max(20, n.handleWidth)),
              y: 0,
              width: Math.min(40, Math.max(20, n.handleWidth)),
              height: d.value.height,
              fill: n.handleFill || b.value,
              stroke: n.handleBorderColor || n.textColor,
              "stroke-width": n.handleBorderWidth,
              rx: 3
            }, vl.value, {
              onKeydown: l[0] || (l[0] = (t) => pt("start", t))
            }), null, 16, ua)) : c("", !0),
            n.handleType && n.handleType !== "empty" ? (u(), r("svg", {
              key: 6,
              x: C.value.x - Math.min(40, Math.max(20, n.handleWidth)),
              y: 0,
              width: Math.min(40, Math.max(20, n.handleWidth)),
              height: d.value.height,
              viewBox: "0 0 20 20"
            }, [
              n.handleType === "arrow" ? (u(), r("path", {
                key: 0,
                d: "M 7 7 L 3 10 L 7 13 L 7 7 M 13 7 L 17 10 L 13 13 L 13 7",
                fill: b.value,
                stroke: n.handleIconColor || n.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, oa)) : n.handleType === "chevron" ? (u(), r("path", {
                key: 1,
                d: "M 6 7 L 4 10 L 6 13 M 14 7 L 16 10 L 14 13",
                fill: "none",
                stroke: n.handleIconColor || n.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, sa)) : n.handleType === "grab" ? (u(), r("path", {
                key: 2,
                d: "M 8 5 A 1 1 0 0 0 8 7 A 1 1 0 0 0 8 5 M 8 9 A 1 1 0 0 0 8 11 A 1 1 0 0 0 8 9 M 8 13 A 1 1 0 0 0 8 15 A 1 1 0 0 0 8 13 M 12 5 A 1 1 0 0 0 12 7 A 1 1 0 0 0 12 5 M 12 9 A 1 1 0 0 0 12 11 A 1 1 0 0 0 12 9 M 12 13 A 1 1 0 0 0 12 15 A 1 1 0 0 0 12 13",
                fill: n.handleIconColor || n.textColor,
                stroke: "none",
                opacity: "0.6"
              }, null, 8, ca)) : c("", !0)
            ], 8, ra)) : c("", !0),
            x.value && n.minimapCompact ? (u(), r("rect", pe({
              key: 7,
              class: "vue-ui-zoom-compact-minimap-handle",
              x: C.value.x + C.value.width,
              y: 0,
              width: Math.min(40, Math.max(20, n.handleWidth)),
              height: d.value.height,
              fill: n.handleFill || b.value,
              stroke: n.handleBorderColor || n.textColor,
              "stroke-width": n.handleBorderWidth,
              rx: 3
            }, dl.value, {
              onKeydown: l[1] || (l[1] = (t) => pt("end", t))
            }), null, 16, ma)) : c("", !0),
            n.handleType && n.handleType !== "empty" ? (u(), r("svg", {
              key: 8,
              x: C.value.x + C.value.width,
              y: 0,
              width: Math.min(40, Math.max(20, n.handleWidth)),
              height: d.value.height,
              viewBox: "0 0 20 20"
            }, [
              n.handleType === "arrow" ? (u(), r("path", {
                key: 0,
                d: "M 7 7 L 3 10 L 7 13 L 7 7 M 13 7 L 17 10 L 13 13 L 13 7",
                fill: b.value,
                stroke: n.handleIconColor || n.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, da)) : n.handleType === "chevron" ? (u(), r("path", {
                key: 1,
                d: "M 6 7 L 4 10 L 6 13 M 14 7 L 16 10 L 14 13",
                fill: "none",
                stroke: n.handleIconColor || n.textColor,
                "stroke-width": 0.618,
                "stroke-linejoin": "round",
                "stroke-linecap": "round"
              }, null, 8, ha)) : n.handleType === "grab" ? (u(), r("path", {
                key: 2,
                d: "M 8 5 A 1 1 0 0 0 8 7 A 1 1 0 0 0 8 5 M 8 9 A 1 1 0 0 0 8 11 A 1 1 0 0 0 8 9 M 8 13 A 1 1 0 0 0 8 15 A 1 1 0 0 0 8 13 M 12 5 A 1 1 0 0 0 12 7 A 1 1 0 0 0 12 5 M 12 9 A 1 1 0 0 0 12 11 A 1 1 0 0 0 12 9 M 12 13 A 1 1 0 0 0 12 15 A 1 1 0 0 0 12 13",
                fill: n.handleIconColor || n.textColor,
                stroke: "none",
                opacity: "0.6"
              }, null, 8, fa)) : c("", !0)
            ], 8, va)) : c("", !0),
            W.value !== null && !X.value ? (u(), r("line", Pt(pe({ key: 9 }, sl.value)), null, 16)) : c("", !0),
            e.$slots.slotMap ? c("", !0) : (u(), r(M, { key: 10 }, [
              n.minimapMerged ? (u(), r("g", ga, [
                v.value && v.value.firstPlot && v.value.firstPlot.value !== null ? (u(), r("circle", {
                  key: 0,
                  cx: v.value.firstPlot.x,
                  cy: v.value.firstPlot.y,
                  "stroke-width": "0.5",
                  stroke: b.value,
                  r: "4",
                  fill: n.minimapLineColor
                }, null, 8, ya)) : c("", !0),
                v.value && v.value.firstPlot && v.value.firstPlot.value !== null ? (u(), r("circle", {
                  key: 1,
                  cx: v.value.firstPlot.x,
                  cy: v.value.firstPlot.y,
                  r: 2,
                  fill: b.value
                }, null, 8, xa)) : c("", !0),
                v.value && v.value.lastPlot && v.value.lastPlot.value !== null ? (u(), r("circle", {
                  key: 2,
                  cx: v.value.lastPlot.x,
                  cy: v.value.lastPlot.y,
                  "stroke-width": "0.5",
                  stroke: b.value,
                  r: "4",
                  fill: n.minimapLineColor
                }, null, 8, ba)) : c("", !0),
                v.value && v.value.lastPlot && v.value.lastPlot.value !== null ? (u(), r("circle", {
                  key: 3,
                  cx: v.value.lastPlot.x,
                  cy: v.value.lastPlot.y,
                  r: "2",
                  fill: b.value
                }, null, 8, Ma)) : c("", !0)
              ])) : (u(), r("g", ka, [
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "line"), (t, i) => (u(), r("g", {
                  key: String(t.key)
                }, [
                  t && t.hasSelection && t.selectionSet && t.isVisible ? (u(), r("path", {
                    key: 0,
                    d: `M ${t.selectionSet}`,
                    stroke: t.temperatureColors ? `url(#temperature_grad_line_${i}_${n.uuid})` : t.color,
                    fill: "transparent",
                    "stroke-width": "2",
                    "stroke-linecap": "round",
                    "stroke-linejoin": "round",
                    "stroke-dasharray": t.dashed ? "2 4" : 0
                  }, null, 8, pa)) : c("", !0),
                  t.isVisible && n.cutNullValues ? (u(!0), r(M, { key: 1 }, S(t.selectionMarkers, (o) => (u(), r("circle", {
                    key: `sel-dot-${t.key}-${o.i}`,
                    cx: o.x,
                    cy: o.y,
                    r: "2.5",
                    fill: t.color,
                    stroke: b.value,
                    "stroke-width": "0.5"
                  }, null, 8, wa))), 128)) : c("", !0),
                  t && t.firstPlot && t.isVisible && t.firstPlot.value !== null ? (u(), r("circle", {
                    key: 2,
                    cx: t.firstPlot.x,
                    cy: t.firstPlot.y,
                    "stroke-width": "0.5",
                    stroke: b.value,
                    r: "4",
                    fill: t.color
                  }, null, 8, Sa)) : c("", !0),
                  t && t.firstPlot && t.isVisible && t.firstPlot.value !== null ? (u(), r("circle", {
                    key: 3,
                    cx: t.firstPlot.x,
                    cy: t.firstPlot.y,
                    r: "2",
                    fill: b.value
                  }, null, 8, Ca)) : c("", !0),
                  t && t.lastPlot && t.isVisible && t.lastPlot.value !== null ? (u(), r("circle", {
                    key: 4,
                    cx: t.lastPlot.x,
                    cy: t.lastPlot.y,
                    "stroke-width": "0.5",
                    stroke: b.value,
                    r: "4",
                    fill: t.color
                  }, null, 8, Na)) : c("", !0),
                  t && t.lastPlot && t.isVisible && t.lastPlot.value !== null ? (u(), r("circle", {
                    key: 5,
                    cx: t.lastPlot.x,
                    cy: t.lastPlot.y,
                    r: "2",
                    fill: b.value
                  }, null, 8, Pa)) : c("", !0)
                ]))), 128)),
                (u(!0), r(M, null, S(I.value.filter((t) => t.type === "plot"), (t) => (u(), r("g", {
                  key: String(t.key)
                }, [
                  (u(!0), r(M, null, S(t.points, (i) => (u(), r("g", null, [
                    t.isVisible && n.cutNullValues && i.value !== null ? (u(), r("circle", {
                      key: `sel-plot-${t.key}-${i.i}`,
                      cx: i.x,
                      cy: i.y,
                      r: "2.5",
                      fill: t.color,
                      stroke: b.value,
                      "stroke-width": "0.5"
                    }, null, 8, La)) : c("", !0)
                  ]))), 256))
                ]))), 128))
              ]))
            ], 64)),
            (u(!0), r(M, null, S(qe.value, (t, i) => (u(), r("rect", {
              x: w.value * i - (n.minimapCompact ? w.value / 2 : 0),
              y: 0,
              height: Math.max(d.value.height, 0),
              width: w.value < 0 ? 0 : w.value,
              fill: "transparent",
              style: L([{ "pointer-events": "all !important" }, { cursor: t >= K.value && t < G.value && n.enableSelectionDrag ? X.value ? "grabbing" : "grab" : "default" }]),
              onMousedown: l[2] || (l[2] = (o) => X.value = !0),
              onMouseup: l[3] || (l[3] = (o) => X.value = !1),
              onMouseenter: (o) => Gt(t),
              onMouseleave: l[4] || (l[4] = (o) => {
                W.value = null, E("trapMouse", null);
              })
            }, null, 44, Aa))), 256))
          ], 8, Wl))
        ])) : c("", !0),
        F("div", {
          class: "slider-track",
          style: L({ visibility: x.value && n.minimapCompact ? "hidden" : "visible" })
        }, null, 4),
        F("div", {
          class: te({ "range-highlight": !0, move: n.enableSelectionDrag }),
          onMousedown: l[5] || (l[5] = (t) => X.value = !0),
          onMouseup: l[6] || (l[6] = (t) => X.value = !1),
          style: L({
            ...R.value,
            cursor: X.value ? "grabbing" : "grab",
            visibility: x.value && n.minimapCompact ? "hidden" : "visible"
          })
        }, null, 38),
        n.enableRangeHandles ? Lt((u(), r("input", {
          key: 1,
          "aria-label": "range-handle-left",
          ref_key: "rangeStart",
          ref: ae,
          type: "range",
          class: te({
            "range-left": !0,
            "range-handle": !0,
            "range-minimap": x.value && n.verticalHandles,
            "range-invisible": x.value && n.minimapCompact
          }),
          min: n.min,
          max: n.minimapCompact && x.value ? Math.max(0, Y.value - 1) : n.max,
          tabindex: x.value ? -1 : 0,
          "onUpdate:modelValue": l[7] || (l[7] = (t) => fe.value = t),
          onFocus: l[8] || (l[8] = (t) => x.value && t.target.blur()),
          onInput: l[9] || (l[9] = (t) => fe.value = t.target.valueAsNumber),
          onChange: A,
          onKeyup: At(A, ["enter"]),
          onBlur: A,
          onMouseenter: l[10] || (l[10] = (t) => ht("start")),
          onPointerup: A
        }, null, 42, $a)), [
          [
            $t,
            fe.value,
            void 0,
            { number: !0 }
          ]
        ]) : c("", !0),
        n.enableRangeHandles ? Lt((u(), r("input", {
          key: 2,
          "aria-label": "range-handle-right",
          ref_key: "rangeEnd",
          ref: ne,
          type: "range",
          class: te({
            "range-right": !0,
            "range-handle": !0,
            "range-minimap": x.value && n.verticalHandles,
            "range-invisible": x.value && n.minimapCompact
          }),
          min: n.min,
          max: n.minimapCompact && x.value ? Math.max(0, Y.value - 1) : n.max,
          tabindex: x.value ? -1 : 0,
          onFocus: l[11] || (l[11] = (t) => x.value && t.target.blur()),
          "onUpdate:modelValue": l[12] || (l[12] = (t) => ge.value = t),
          onInput: l[13] || (l[13] = (t) => ge.value = t.target.valueAsNumber),
          onChange: A,
          onKeyup: At(A, ["enter"]),
          onBlur: A,
          onMouseenter: l[14] || (l[14] = (t) => ht("end")),
          onPointerup: A
        }, null, 42, Fa)), [
          [
            $t,
            ge.value,
            void 0,
            { number: !0 }
          ]
        ]) : c("", !0),
        k.value.left ? (u(), r("div", {
          key: 3,
          ref_key: "tooltipLeft",
          ref: ue,
          class: te({
            "range-tooltip": !0,
            "range-tooltip-visible": _.value,
            "range-tooltip-arrow": R.value.arrowLeft && !n.verticalHandles,
            "range-tooltip-arrow-left": !R.value.arrowLeft && !n.verticalHandles
          }),
          style: L({
            left: R.value.tooltipLeft,
            color: ke(Ke)(n.selectColor),
            backgroundColor: n.selectColor,
            border: `1px solid ${b.value}`,
            zIndex: `${dt.value + 4}`,
            visibility: se.value || k.value.left === k.value.right ? "hidden" : "visible",
            top: x.value && n.minimapCompact ? "calc(-100% - 12px)" : "-100%"
          })
        }, Xe(k.value.left), 7)) : c("", !0),
        (se.value || k.value.left === k.value.right) && (k.value.left || k.value.right) ? (u(), r("div", {
          key: 4,
          ref: "tooltipMerge",
          class: te({
            "range-tooltip": !0,
            "range-tooltip-visible": _.value,
            "range-tooltip-arrow": !0,
            "range-tooltip-arrow-left": !R.value.arrowLeft && !n.verticalHandles,
            "range-tooltip-arrow-right": !R.value.arrowRight && !n.verticalHandles
          }),
          style: L({
            left: R.value.tooltipCenter,
            width: ce.value.width + "px",
            color: ke(Ke)(n.selectColor),
            backgroundColor: n.selectColor,
            border: `1px solid ${b.value}`,
            zIndex: "4",
            top: x.value && n.minimapCompact ? "calc(-100% - 12px)" : "-100%"
          })
        }, Xe(k.value.left === k.value.right ? k.value.left : `${k.value.left} - ${k.value.right}`), 7)) : c("", !0),
        k.value.right ? (u(), r("div", {
          key: 5,
          ref_key: "tooltipRight",
          ref: re,
          class: te({
            "range-tooltip": !0,
            "range-tooltip-visible": _.value,
            "range-tooltip-arrow": R.value.arrowRight && !n.verticalHandles,
            "range-tooltip-arrow-right": !R.value.arrowRight && !n.verticalHandles
          }),
          style: L({
            left: R.value.tooltipRight,
            color: ke(Ke)(n.selectColor),
            backgroundColor: n.selectColor,
            border: `1px solid ${b.value}`,
            zIndex: "4",
            visibility: se.value || k.value.left === k.value.right ? "hidden" : "visible",
            top: x.value && n.minimapCompact ? "calc(-100% - 12px)" : "-100%"
          })
        }, Xe(k.value.right), 7)) : c("", !0)
      ], 36)
    ], 44, Tl));
  }
}, Va = /* @__PURE__ */ Rl(Ra, [["__scopeId", "data-v-060b5d70"]]);
export {
  Va as S
};
