// vue-data-ui/dist/Tooltip-DL3QgjZf.js
import { ref as i, onMounted as L, onUnmounted as R, watch as w, nextTick as W, computed as g, openBlock as Y, createBlock as j, Teleport as q, createElementBlock as H, normalizeStyle as I, normalizeClass as V, renderSlot as C, createElementVNode as D, createCommentVNode as G } from "vue";
import { u as U } from "./lib-GwPnJSrY.js";
function J({ tooltip: n, chart: c, clientPosition: t, positionPreference: v = "center", defaultOffsetY: h = 24, blockShiftY: x = !1 }) {
  const l = i(0), a = i(h);
  if (n && c) {
    const { width: u, height: m } = n.getBoundingClientRect(), { right: f, left: p, bottom: S } = c.getBoundingClientRect();
    v === "center" && (t.x + u / 2 > f ? l.value = -u + (f - t.x) : t.x - u / 2 < p ? l.value = -u + (u - (t.x - p)) : l.value = -u / 2), v === "right" && (t.x + u > f ? l.value = -u + (f - t.x) : l.value = 0), v === "left" && (t.x < p + u ? l.value = -u + (u - (t.x - p)) : l.value = -u), t.y + m > S && !x && (a.value = -m - h);
  }
  return {
    top: t.y + a.value,
    left: t.x + l.value
  };
}
function K(n, c, t) {
  L(() => n.addEventListener(c, t)), R(() => n.removeEventListener(c, t));
}
function Q() {
  const n = i(0), c = i(0);
  return K(window, "mousemove", (t) => {
    n.value = t.clientX, c.value = t.clientY;
  }), { x: n, y: c };
}
const Z = ["aria-hidden"], P = ["innerHTML"], te = {
  __name: "Tooltip",
  props: {
    teleportTo: {
      type: String,
      default: "body"
    },
    backgroundColor: {
      type: String,
      default: "#FFFFFF"
    },
    color: {
      type: String,
      default: "#000000"
    },
    content: String,
    maxWidth: {
      type: String,
      default: "300px"
    },
    parent: {
      type: Object
    },
    show: {
      type: Boolean,
      default: !1
    },
    isCustom: {
      type: Boolean,
      default: !1
    },
    fontSize: {
      type: [Number, String],
      default: 14
    },
    borderRadius: {
      type: Number,
      default: 4
    },
    borderColor: {
      type: String,
      default: "#e1e5e8"
    },
    borderWidth: {
      type: Number,
      default: 1
    },
    backgroundOpacity: {
      type: Number,
      default: 100
    },
    position: {
      type: String,
      default: "center"
    },
    offsetY: {
      type: Number,
      default: 24
    },
    blockShiftY: {
      type: Boolean,
      default: !1
    },
    isFullscreen: {
      type: Boolean,
      default: !1
    },
    smooth: {
      type: Boolean,
      default: !0
    },
    backdropFilter: {
      type: Boolean,
      default: !0
    },
    smoothForce: {
      type: Number,
      default: 0.18
    },
    smoothSnapThreshold: {
      type: Number,
      default: 0.25
    }
  },
  setup(n, { expose: c }) {
    const t = n, v = i(null), { x: h, y: x } = Q(t.parent), l = i({ x: 0, y: 0 }), a = i({ x: 0, y: 0 }), u = i(null), m = i(null), f = i({ scaleX: 1, scaleY: 1 }), p = i(null);
    function S() {
      const e = M(v.value);
      if (u.value = e, !e) {
        m.value = null, f.value = { scaleX: 1, scaleY: 1 }, p.value = null;
        return;
      }
      m.value = e.getBoundingClientRect(), f.value = X(e), p.value = t.parent?.getBoundingClientRect?.() || null;
    }
    function T(e) {
      return e ? !!(e.transform && e.transform !== "none" || e.perspective && e.perspective !== "none" || e.filter && e.filter !== "none" || e.backdropFilter && e.backdropFilter !== "none" || e.contain && e.contain.includes("paint") || e.willChange && (e.willChange.includes("transform") || e.willChange.includes("filter"))) : !1;
    }
    function M(e) {
      let o = e?.parentElement || null;
      for (; o && o !== document.documentElement; ) {
        const r = getComputedStyle(o);
        if (T(r)) return o;
        o = o.parentElement;
      }
      return null;
    }
    function X(e) {
      if (!e) return { scaleX: 1, scaleY: 1 };
      const r = getComputedStyle(e).transform;
      if (!r || r === "none") return { scaleX: 1, scaleY: 1 };
      try {
        const d = new DOMMatrixReadOnly(r), s = Math.hypot(d.a, d.b), A = Math.hypot(d.c, d.d);
        return {
          scaleX: s || 1,
          scaleY: A || 1
        };
      } catch {
        return { scaleX: 1, scaleY: 1 };
      }
    }
    function E({ x: e, y: o }) {
      h.value = e, x.value = o;
    }
    let y = null;
    function k() {
      if (!t.show) {
        b();
        return;
      }
      if (!t.smooth) {
        a.value.x = l.value.x, a.value.y = l.value.y, b();
        return;
      }
      const e = l.value.x - a.value.x, o = l.value.y - a.value.y;
      if (Math.abs(e) <= t.smoothSnapThreshold && Math.abs(o) <= t.smoothSnapThreshold) {
        a.value.x = l.value.x, a.value.y = l.value.y, b();
        return;
      }
      a.value.x += e * t.smoothForce, a.value.y += o * t.smoothForce, y = requestAnimationFrame(k);
    }
    function F() {
      y == null && t.show && t.smooth && (y = requestAnimationFrame(k));
    }
    function b() {
      y != null && (cancelAnimationFrame(y), y = null);
    }
    w([h, x], ([e, o]) => {
      l.value.x = e, l.value.y = o, t.smooth ? F() : (a.value.x = e, a.value.y = o);
    }), w(
      () => t.show,
      async (e) => {
        if (!e) {
          b();
          return;
        }
        const o = h.value, r = x.value;
        l.value.x = o, l.value.y = r, a.value.x = o, a.value.y = r, await W(), S(), F();
      }
    ), R(() => {
      b();
    });
    const N = g(() => {
      const e = m.value;
      if (!e) return { x: a.value.x, y: a.value.y };
      const { scaleX: o, scaleY: r } = f.value;
      return {
        x: (a.value.x - e.left) / o,
        y: (a.value.y - e.top) / r
      };
    }), $ = g(() => {
      const e = m.value, o = p.value;
      if (!e || !o) return t.parent;
      const { scaleX: r, scaleY: d } = f.value;
      return {
        ...t.parent,
        getBoundingClientRect() {
          const s = o;
          return {
            left: (s.left - e.left) / r,
            top: (s.top - e.top) / d,
            right: (s.right - e.left) / r,
            bottom: (s.bottom - e.top) / d,
            width: s.width / r,
            height: s.height / d,
            x: (s.left - e.left) / r,
            y: (s.top - e.top) / d
          };
        }
      };
    }), B = g(() => {
      const e = J({
        tooltip: v.value,
        chart: $.value,
        clientPosition: N.value,
        positionPreference: t.position,
        defaultOffsetY: t.offsetY,
        blockShiftY: t.blockShiftY
      });
      return {
        top: Math.round(e.top),
        left: Math.round(e.left)
      };
    }), z = g(() => U(t.backgroundColor, t.backgroundOpacity)), O = g(() => {
      const e = {
        pointerEvents: "none",
        position: "fixed",
        top: "0px",
        left: "0px",
        transform: `translate3d(${B.value.left}px, ${B.value.top}px, 0)`,
        borderRadius: `${t.borderRadius}px`,
        border: `${t.borderWidth}px solid ${t.borderColor}`,
        zIndex: 2147483647
      };
      return t.isCustom || Object.assign(e, {
        background: z.value,
        color: t.color,
        maxWidth: t.maxWidth,
        fontSize: `${t.fontSize}px`
      }), e;
    });
    return c({
      placeTooltip: E
    }), (e, o) => (Y(), j(q, {
      to: n.isFullscreen ? n.parent : n.teleportTo
    }, [
      n.show ? (Y(), H("div", {
        key: 0,
        ref_key: "tooltip",
        ref: v,
        role: "tooltip",
        "aria-hidden": !n.show,
        "aria-live": "polite",
        class: V({
          "vue-data-ui-custom-tooltip": n.isCustom,
          "vue-data-ui-tooltip": !n.isCustom,
          "vue-data-ui-tooltip-backdrop": n.backdropFilter
        }),
        style: I(O.value)
      }, [
        C(e.$slots, "tooltip-before"),
        C(e.$slots, "default"),
        D("div", { innerHTML: n.content }, null, 8, P),
        C(e.$slots, "tooltip-after")
      ], 14, Z)) : G("", !0)
    ], 8, ["to"]));
  }
};
export {
  te as default
};
