// vue-data-ui/dist/useObjectBindings-BlXBVtRa.js
import { u as p } from "./useConfig-CIgyKZ6R.js";
import { watchEffect as j, markRaw as g, computed as h, isRef as b } from "vue";
const k = [
  "vue_ui_xy",
  "vue_ui_donut",
  "vue_ui_treemap",
  "vue_ui_waffle",
  "vue_ui_radar",
  "vue_ui_quadrant",
  "vue_ui_gauge",
  "vue_ui_wheel",
  "vue_ui_tiremarks",
  "vue_ui_chestnut",
  "vue_ui_onion",
  "vue_ui_vertical_bar",
  "vue_ui_horizontal_bar",
  "vue_ui_heatmap",
  "vue_ui_scatter",
  "vue_ui_candlestick",
  "vue_ui_sparkline",
  "vue_ui_sparkbar",
  "vue_ui_sparkstackbar",
  "vue_ui_sparkhistogram",
  "vue_ui_sparkgauge",
  "vue_ui_spark_trend",
  "vue_ui_quick_chart",
  "vue_ui_age_pyramid",
  "vue_ui_relation_circle",
  "vue_ui_thermometer",
  "vue_ui_rings",
  "vue_ui_donut_evolution",
  "vue_ui_mood_radar",
  "vue_ui_molecule",
  "vue_ui_nested_donuts",
  "vue_ui_galaxy",
  "vue_ui_strip_plot",
  "vue_ui_dumbbell",
  "vue_ui_3d_bar",
  "vue_ui_table_sparkline",
  "vue_ui_table_heatmap",
  "vue_ui_word_cloud",
  "vue_ui_xy_canvas",
  "vue_ui_flow",
  "vue_ui_parallel_coordinate_plot",
  "vue_ui_timer",
  "vue_ui_carousel_table",
  "vue_ui_gizmo",
  "vue_ui_stackbar",
  "vue_ui_bullet",
  "vue_ui_funnel",
  "vue_ui_history_plot",
  "vue_ui_world",
  "vue_ui_ridgeline",
  "vue_ui_chord",
  "vue_ui_stackline",
  "vue_ui_dag",
  "vue_ui_geo",
  "vue_ui_bump",
  "vue_ui_cursor",
  "vue_ui_accordion",
  "vue_ui_kpi",
  "vue_ui_mini_loader",
  "vue_ui_smiley",
  "vue_ui_rating",
  "vue_ui_annotator",
  "vue_ui_dashboard",
  "vue_ui_skeleton",
  "vue_ui_table",
  "vue_ui_digits",
  "vue_ui_circle_pack"
];
function D(e, n = {}) {
  return k.includes(e) ? p(n)[e] : (console.error(`VueDataUi - getVueDataUiConfig : ${e} is not a valid component name for this utility.
Use snake case names, for example 'vue_ui_xy', 'vue_ui_donut', etc.`), {});
}
const y = /* @__PURE__ */ Object.assign({ "./themes/vue_ui_3d_bar.json": () => import("./vue_ui_3d_bar-CL0iX8Qi.js").then((e) => e.default), "./themes/vue_ui_age_pyramid.json": () => import("./vue_ui_age_pyramid-8H8WSL7J.js").then((e) => e.default), "./themes/vue_ui_bullet.json": () => import("./vue_ui_bullet-Ckqieba5.js").then((e) => e.default), "./themes/vue_ui_bump.json": () => import("./vue_ui_bump-CIht647H.js").then((e) => e.default), "./themes/vue_ui_candlestick.json": () => import("./vue_ui_candlestick-BAMn-VGw.js").then((e) => e.default), "./themes/vue_ui_chestnut.json": () => import("./vue_ui_chestnut-BloFvDuV.js").then((e) => e.default), "./themes/vue_ui_chord.json": () => import("./vue_ui_chord-Bc1_ocwK.js").then((e) => e.default), "./themes/vue_ui_circle_pack.json": () => import("./vue_ui_circle_pack-CLmw8nHe.js").then((e) => e.default), "./themes/vue_ui_dag.json": () => import("./vue_ui_dag-ZOwuDBZ8.js").then((e) => e.default), "./themes/vue_ui_donut.json": () => import("./vue_ui_donut-CA4dgV8J.js").then((e) => e.default), "./themes/vue_ui_donut_evolution.json": () => import("./vue_ui_donut_evolution-DFQU-rqE.js").then((e) => e.default), "./themes/vue_ui_dumbbell.json": () => import("./vue_ui_dumbbell-BCOHDiPk.js").then((e) => e.default), "./themes/vue_ui_flow.json": () => import("./vue_ui_flow-BErINgox.js").then((e) => e.default), "./themes/vue_ui_funnel.json": () => import("./vue_ui_funnel-yuaMZ-yb.js").then((e) => e.default), "./themes/vue_ui_galaxy.json": () => import("./vue_ui_galaxy-D95AN4FT.js").then((e) => e.default), "./themes/vue_ui_gauge.json": () => import("./vue_ui_gauge-BMEuRC4J.js").then((e) => e.default), "./themes/vue_ui_geo.json": () => import("./vue_ui_geo--kg4VoVX.js").then((e) => e.default), "./themes/vue_ui_heatmap.json": () => import("./vue_ui_heatmap-C9u993qE.js").then((e) => e.default), "./themes/vue_ui_history_plot.json": () => import("./vue_ui_history_plot-BCKk1CrQ.js").then((e) => e.default), "./themes/vue_ui_horizontal_bar.json": () => import("./vue_ui_horizontal_bar-BEagrV0T.js").then((e) => e.default), "./themes/vue_ui_molecule.json": () => import("./vue_ui_molecule-BtvlINUh.js").then((e) => e.default), "./themes/vue_ui_mood_radar.json": () => import("./vue_ui_mood_radar-BUs6Fdn8.js").then((e) => e.default), "./themes/vue_ui_nested_donuts.json": () => import("./vue_ui_nested_donuts-DvTrWT_C.js").then((e) => e.default), "./themes/vue_ui_onion.json": () => import("./vue_ui_onion-CxWH9_OZ.js").then((e) => e.default), "./themes/vue_ui_parallel_coordinate_plot.json": () => import("./vue_ui_parallel_coordinate_plot-DH3cFJ3n.js").then((e) => e.default), "./themes/vue_ui_quadrant.json": () => import("./vue_ui_quadrant-DjRs--Uz.js").then((e) => e.default), "./themes/vue_ui_quick_chart.json": () => import("./vue_ui_quick_chart-DMCjEjPe.js").then((e) => e.default), "./themes/vue_ui_radar.json": () => import("./vue_ui_radar-DryGBOrz.js").then((e) => e.default), "./themes/vue_ui_relation_circle.json": () => import("./vue_ui_relation_circle-CCroen7h.js").then((e) => e.default), "./themes/vue_ui_ridgeline.json": () => import("./vue_ui_ridgeline-j4rVmpEp.js").then((e) => e.default), "./themes/vue_ui_rings.json": () => import("./vue_ui_rings-CxOrekqA.js").then((e) => e.default), "./themes/vue_ui_scatter.json": () => import("./vue_ui_scatter-CO5Iq9ai.js").then((e) => e.default), "./themes/vue_ui_spark_trend.json": () => import("./vue_ui_spark_trend-3BjrSprO.js").then((e) => e.default), "./themes/vue_ui_sparkbar.json": () => import("./vue_ui_sparkbar-BsMzz5Um.js").then((e) => e.default), "./themes/vue_ui_sparkgauge.json": () => import("./vue_ui_sparkgauge-KnjASOZs.js").then((e) => e.default), "./themes/vue_ui_sparkhistogram.json": () => import("./vue_ui_sparkhistogram-GBdzIVvS.js").then((e) => e.default), "./themes/vue_ui_sparkline.json": () => import("./vue_ui_sparkline-CQ8DJVJx.js").then((e) => e.default), "./themes/vue_ui_sparkstackbar.json": () => import("./vue_ui_sparkstackbar-BSexvOuM.js").then((e) => e.default), "./themes/vue_ui_stackbar.json": () => import("./vue_ui_stackbar-CuDC4O4b.js").then((e) => e.default), "./themes/vue_ui_stackline.json": () => import("./vue_ui_stackline-QTh-nuxE.js").then((e) => e.default), "./themes/vue_ui_strip_plot.json": () => import("./vue_ui_strip_plot-B9ypi_F3.js").then((e) => e.default), "./themes/vue_ui_table_heatmap.json": () => import("./vue_ui_table_heatmap-CkvipPVy.js").then((e) => e.default), "./themes/vue_ui_table_sparkline.json": () => import("./vue_ui_table_sparkline-yQdijcyJ.js").then((e) => e.default), "./themes/vue_ui_thermometer.json": () => import("./vue_ui_thermometer-BTmjRBH8.js").then((e) => e.default), "./themes/vue_ui_tiremarks.json": () => import("./vue_ui_tiremarks-DBotWfi8.js").then((e) => e.default), "./themes/vue_ui_treemap.json": () => import("./vue_ui_treemap-D9-AIkT-.js").then((e) => e.default), "./themes/vue_ui_vertical_bar.json": () => import("./vue_ui_vertical_bar-CELc39b8.js").then((e) => e.default), "./themes/vue_ui_waffle.json": () => import("./vue_ui_waffle-M2h-uJQ_.js").then((e) => e.default), "./themes/vue_ui_wheel.json": () => import("./vue_ui_wheel-LqBz3lv4.js").then((e) => e.default), "./themes/vue_ui_word_cloud.json": () => import("./vue_ui_word_cloud-mKAtBxHR.js").then((e) => e.default), "./themes/vue_ui_xy.json": () => import("./vue_ui_xy-esktmpg-.js").then((e) => e.default), "./themes/vue_ui_xy_canvas.json": () => import("./vue_ui_xy_canvas-6BBDSowT.js").then((e) => e.default) }), w = Object.fromEntries(
  Object.entries(y).map(([e, n]) => {
    const u = e.match(/\/themes\/(.+)\.json$/);
    return [u ? u[1] : e, n];
  })
), m = /* @__PURE__ */ new Map();
let x = async (e) => {
  const n = w[e];
  if (!n) {
    const u = new Error("ENOENT");
    throw u.code = "ENOENT", u;
  }
  return await n();
};
async function V(e) {
  if (!e) return null;
  if (m.has(e)) return m.get(e);
  try {
    const n = await x(e);
    return m.set(e, n), n;
  } catch (n) {
    return console.warn(`[getThemeConfig] Missing theme file: ${e}.json`, n), null;
  }
}
function O(e, n) {
  if (n <= 0 || e.length < 3) return e;
  const u = e.length, o = Array.from({ length: u });
  for (let _ = 0; _ < u; _ += 1) {
    const t = Math.max(0, _ - n);
    let i = 0;
    for (let a = t; a <= _; a += 1) i += e[a].value;
    o[_] = i / (_ - t + 1);
  }
  const r = Array.from({ length: u });
  for (let _ = 0; _ < u; _ += 1) {
    const t = Math.min(u - 1, _ + n);
    let i = 0;
    for (let a = _; a <= t; a += 1) i += e[a].value;
    r[_] = i / (t - _ + 1);
  }
  const s = e.map((_) => ({ ..._ }));
  for (let _ = 1; _ < u - 1; _ += 1) {
    const t = _ / (u - 1);
    s[_].value = (1 - t) * o[_] + t * r[_];
  }
  return s;
}
function E(e, n) {
  if (n <= 0 || e.length < 3) return e;
  const u = 1 / (1 + n), o = e.length, r = Array.from({ length: o });
  r[0] = e[0].value;
  for (let t = 1; t < o; t += 1)
    r[t] = u * e[t].value + (1 - u) * r[t - 1];
  const s = Array.from({ length: o });
  s[o - 1] = e[o - 1].value;
  for (let t = o - 2; t >= 0; t -= 1)
    s[t] = u * e[t].value + (1 - u) * s[t + 1];
  const _ = e.map((t) => ({ ...t }));
  for (let t = 1; t < o - 1; t += 1) {
    const i = t / (o - 1);
    _[t].value = (1 - i) * r[t] + i * s[t];
  }
  return _;
}
function P(e, n) {
  let u = e;
  return u = O(u, n.averageWindow), u = E(u, n.smoothingTau), u;
}
function d(e, n = [], u = !0) {
  const o = [];
  if (e && typeof e == "object") {
    if (Array.isArray(e) && u)
      return [];
    for (const r of Object.keys(e)) {
      const s = (
        /** @type {any} */
        e[r]
      );
      if (Array.isArray(s) && u)
        continue;
      const _ = n.concat(r);
      s && typeof s == "object" ? o.push(...d(s, _, u)) : o.push(_);
    }
  }
  return o;
}
function v(e, n) {
  return n.reduce((u, o) => u?.[o], e);
}
function c(e, n, u) {
  let o = b(e) ? e.value : e;
  for (let r = 0; r < n.length - 1; r += 1) {
    const s = n[r];
    (!Object.prototype.hasOwnProperty.call(o, s) || typeof o[s] != "object") && (o[s] = {}), o = o[s];
  }
  o[n[n.length - 1]] = u;
}
function f(e, n, u, o) {
  const r = n.split(o);
  let s = e;
  for (let _ = 0; _ < r.length - 1; _ += 1) {
    const t = r[_];
    s[t] || (s[t] = {}), s = s[t];
  }
  s[r[r.length - 1]] = u;
}
function T(e, n) {
  const { delimiter: u = ".", skipArrays: o = !0 } = n || {}, r = {};
  function s() {
    Object.keys(r).forEach((i) => delete r[i]);
    const t = d(e.value, [], o);
    for (const i of t) {
      const a = i.join(u);
      r[a] = h({
        get: () => v(e.value, i),
        set: (l) => c(e.value, i, l)
      });
    }
  }
  j(s), s();
  const _ = {
    get(t, i) {
      return typeof i == "string" || i.startsWith("__v_") ? i in t ? Reflect.get(t, i) : (f(e.value, i, void 0, u), r[i] = h({
        get: () => v(e.value, i),
        set: (a) => c(e.value, i, a)
      }), i.startsWith("__v_") || console.warn(`Vue Data UI - useObjectBindings: no binding found for key "${i}". Please verify you are binding to a property path which exists on the object.`), "") : !0;
    },
    set(t, i, a) {
      return typeof i == "string" || i.startsWith("__v_") ? i in t ? Reflect.set(t, i, a) : (f(e.value, i, a, u), r[i] = h({
        get: () => v(e.value, i),
        set: (l) => c(e.value, i, l)
      }), i.startsWith("__v_") || console.warn(`Vue Data UI - useObjectBindings: cannot set unknown binding "${i}".`), !0) : !0;
    }
  };
  return g(new Proxy(r, _));
}
export {
  P as a,
  D as b,
  V as g,
  T as u
};
