// vue-data-ui/dist/useStableElementSize-DCNs0Tx7.js
import { ref as h, onBeforeUnmount as q, nextTick as y } from "vue";
function N(i, r) {
  return Number.isFinite(i) && Number.isFinite(r);
}
function x(i, r) {
  return i.width === r.width && i.height === r.height;
}
function I({
  elementRef: i,
  minimumWidth: r = 2,
  minimumHeight: d = 2,
  stableFramesRequired: C = 2,
  once: f = !1,
  onSizeAccepted: v
}) {
  const w = h(0), F = h(0), a = h(!1);
  let n = null, u = 0, o = { width: 0, height: 0 }, l = 0, c = !1;
  function b() {
    u && (cancelAnimationFrame(u), u = 0);
  }
  function p(e, t) {
    return !(!N(e, t) || e < r || t < d);
  }
  function m() {
    b(), n && (n.disconnect(), n = null);
  }
  function z(e, t) {
    w.value = e, F.value = t, a.value = !0, typeof v == "function" && v({ width: e, height: t }), c = !0, f && m();
  }
  function A(e, t) {
    if (f && c) return;
    if (!p(e, t)) {
      a.value = !1, l = 0, o = { width: 0, height: 0 };
      return;
    }
    const s = { width: e, height: t };
    x(s, o) ? l += 1 : (l = 1, o = s), l >= C && z(e, t);
  }
  function g() {
    const e = i.value;
    if (!e) {
      a.value = !1;
      return;
    }
    const t = e.getBoundingClientRect();
    A(Math.round(t.width), Math.round(t.height));
  }
  function S(e = 2) {
    b();
    let t = e;
    const s = () => {
      g(), t -= 1, t > 0 ? u = requestAnimationFrame(s) : u = 0;
    };
    u = requestAnimationFrame(s);
  }
  async function O() {
    await y(), S(2);
    const e = i.value;
    e && (n && (n.disconnect(), n = null), n = new ResizeObserver(() => {
      f && c || (a.value = !1, l = 0, S(2));
    }), n.observe(e));
  }
  return q(m), {
    width: w,
    height: F,
    isStable: a,
    start: O,
    stop: m,
    measureOnce: g
  };
}
export {
  I as u
};
