// vue-data-ui/dist/utils.js
import { u as a, g as e, b as t, a as s } from "./useObjectBindings-BlXBVtRa.js";
import { i as o, h as r, g as i, f as l, e as g, d as m, b as n, c as u, a as f } from "./lib-GwPnJSrY.js";
import { m as C, s as c, d, l as h } from "./exposedLib-H4lmBl1t.js";
const D = {
  applyDataCorrection: s,
  getVueDataUiConfig: t,
  getThemeConfig: e,
  abbreviate: f,
  createSmoothPath: u,
  createStraightPath: n,
  createTSpans: m,
  createWordCloudDatasetFromPlainText: g,
  formatSmallValue: l,
  getCumulativeAverage: i,
  getCumulativeMedian: r,
  getPalette: o,
  lightenColor: h,
  darkenColor: d,
  shiftColorHue: c,
  mergeConfigs: C,
  useObjectBindings: a
};
export {
  f as abbreviate,
  s as applyDataCorrection,
  u as createSmoothPath,
  n as createStraightPath,
  m as createTSpans,
  g as createWordCloudDatasetFromPlainText,
  d as darkenColor,
  D as default,
  l as formatSmallValue,
  i as getCumulativeAverage,
  r as getCumulativeMedian,
  o as getPalette,
  e as getThemeConfig,
  t as getVueDataUiConfig,
  h as lightenColor,
  C as mergeConfigs,
  c as shiftColorHue,
  a as useObjectBindings
};
