// vue-data-ui/dist/vue-ui-xy-f67V1xjd.js
import { defineAsyncComponent as Ee, useSlots as zu, getCurrentInstance as Iu, ref as m, shallowRef as Ou, computed as g, onMounted as Pt, toRefs as Nu, watch as ht, watchEffect as dl, nextTick as Ft, onBeforeUnmount as Nl, openBlock as r, createElementBlock as i, normalizeStyle as w, normalizeClass as tt, createBlock as $e, unref as v, withCtx as X, renderSlot as R, normalizeProps as ge, guardReactiveProps as be, createCommentVNode as h, createElementVNode as p, createSlots as uo, createVNode as ft, Fragment as d, renderList as y, mergeProps as lt, toDisplayString as oe, Teleport as Pu, resolveDynamicComponent as Fu, withDirectives as Vu, vModelCheckbox as Ru, createTextVNode as so } from "vue";
import { j as Pl, q as Fl, t as no, n as Bu, D as Vt, E as at, s as Vl, p as Rl, F as hl, G as E, H as Bl, I as Hu, J as Du, K as V, L as ro, c as io, M as co, b as vo, N as Wu, O as Eu, P as Yu, Q as ho, R as fo, k as ie, S as Xu, T as mo, U as Gu, z as go, A as Fe, l as Ve, v as ot, u as ze, V as qu, a as Uu, W as Ye, d as fl, Y as Le, Z as Zu, _ as Xe, $ as Rt, m as Ju, o as bo, a0 as Ku, a1 as Qu, a2 as mt } from "./lib-GwPnJSrY.js";
import { u as po, a as xo, b as ju } from "./useTimeLabels-C-A2uZKi.js";
import { u as es } from "./useConfig-CIgyKZ6R.js";
import { u as ts } from "./usePrinter-DiK5R6k-.js";
import { u as ls, B as as } from "./BaseScanner-BhhkKVoj.js";
import { u as os } from "./useSvgExport-ByUukOZt.js";
import { u as Bt } from "./useNestedProp-BBT0mbHh.js";
import { u as us } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ss } from "./useStableElementSize-DCNs0Tx7.js";
import { u as ns } from "./useTimeLabelCollider-CIsgDrl9.js";
import rs from "./img-BYYzY_QU.js";
import is from "./Title-BfyrP-jp.js";
import cs from "./vue_ui_xy-esktmpg-.js";
import { _ as ml } from "./Shape-CSyWYgXc.js";
import { S as vs } from "./SlicerPreview-I2J0UUF_.js";
import ds from "./vue-ui-accordion-BGHTqIBt.js";
import { B as hs } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as fs } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ms = ["id"], gs = ["id"], bs = ["viewBox", "aria-label"], ps = ["x", "y", "width", "height"], xs = { key: 1 }, ys = { class: "vue-ui-xy-grid" }, ks = ["stroke", "x1", "x2", "y1", "y2"], _s = ["stroke", "x1", "x2", "y1", "y2"], ws = { key: 1 }, Ss = ["x1", "x2", "y1", "y2", "stroke"], $s = ["x1", "x2", "y1", "y2", "stroke"], Ls = ["x1", "x2", "y1", "y2", "stroke"], Ms = { key: 3 }, Cs = ["d", "stroke"], As = { key: 4 }, Ts = ["d", "stroke"], zs = ["id"], Is = ["stop-color"], Os = ["stop-color"], Ns = ["stop-color"], Ps = ["id"], Fs = ["stop-color"], Vs = ["stop-color"], Rs = ["stop-color"], Bs = ["id"], Hs = ["stop-color"], Ds = ["stop-color"], Ws = ["id"], Es = ["stop-color"], Ys = ["stop-color"], Xs = ["id"], Gs = ["stop-color"], qs = ["stop-color"], Us = { key: 0 }, Zs = ["id"], Js = ["stop-color", "offset"], Ks = ["x", "y", "height", "width", "fill"], Qs = ["x", "y", "width"], js = { key: 0 }, en = ["x", "y", "height", "width", "fill"], tn = ["x", "y", "height", "width", "rx", "fill", "stroke", "stroke-width"], ln = ["x", "y", "height", "width", "rx", "fill", "stroke", "stroke-width"], an = ["width", "x", "y"], on = ["stroke", "x1", "x2", "y1", "y2"], un = { key: 3 }, sn = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], nn = ["x", "y", "width", "height", "stroke", "stroke-width", "stroke-linecap", "stroke-linejoin", "stroke-dasharray"], rn = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], cn = ["fill", "font-size", "transform"], vn = ["x1", "x2", "y1", "y2", "stroke"], dn = ["x", "y", "font-size", "fill"], hn = ["x1", "x2", "y1", "y2", "stroke"], fn = ["x", "y", "font-size", "fill"], mn = ["width", "x", "y"], gn = { style: { width: "100%" } }, bn = ["d", "stroke", "stroke-width", "stroke-dasharray"], pn = ["d", "stroke", "stroke-width", "stroke-dasharray"], xn = { key: 6 }, yn = { key: 7 }, kn = ["d", "fill", "fill-opacity"], _n = { key: 0 }, wn = ["d", "fill"], Sn = ["d", "fill"], $n = ["d", "fill"], Ln = ["d", "fill"], Mn = ["d", "stroke", "stroke-width", "stroke-dasharray"], Cn = ["d", "stroke", "stroke-width", "stroke-dasharray"], An = ["d", "stroke", "stroke-width", "stroke-dasharray"], Tn = ["d", "stroke", "stroke-width", "stroke-dasharray"], zn = ["width", "x", "y"], In = { style: { width: "100%" } }, On = { key: 8 }, Nn = ["x", "y", "font-size", "fill"], Pn = ["x", "y", "font-size", "fill", "font-weight"], Fn = { key: 9 }, Vn = ["x", "y", "font-size", "fill"], Rn = { key: 10 }, Bn = ["x", "y"], Hn = ["innerHTML"], Dn = ["x", "y"], Wn = ["innerHTML"], En = ["x1", "x2", "y1", "y2", "stroke"], Yn = { key: 11 }, Xn = ["x", "y", "font-size", "fill"], Gn = { key: 12 }, qn = ["x", "y"], Un = ["innerHTML"], Zn = ["x", "y"], Jn = ["innerHTML"], Kn = ["x1", "x2", "y1", "y2", "stroke"], Qn = ["x", "y", "font-size", "fill", "innerHTML"], jn = ["x", "y", "font-size", "fill", "innerHTML"], er = ["x", "y", "font-size", "fill", "innerHTML"], tr = ["x", "y", "font-size", "fill", "innerHTML"], lr = { key: 0 }, ar = ["id"], or = ["fill", "stroke"], ur = ["x1", "x2", "y1", "y2", "stroke", "marker-end"], sr = ["x1", "x2", "y1", "y2", "stroke", "marker-end"], nr = ["x", "y", "font-size", "fill", "stroke"], rr = ["id"], ir = ["stop-color"], cr = ["stop-color"], vr = ["x", "y", "width", "height", "fill", "onMouseenter"], dr = ["font-size", "fill", "transform"], hr = ["x", "y", "font-size", "fill"], fr = ["text-anchor", "font-size", "fill", "transform", "onClick"], mr = ["text-anchor", "font-size", "fill", "transform", "innerHTML", "onClick"], gr = { key: 15 }, br = ["x1", "y1", "x2", "y2", "stroke", "stroke-width", "stroke-dasharray"], pr = ["x1", "y1", "x2", "y2", "stroke", "stroke-width", "stroke-dasharray"], xr = ["y", "x", "width", "height", "fill"], yr = ["id", "x", "y", "font-size", "fill", "text-anchor"], kr = {
  key: 16,
  style: { "pointer-events": "none" }
}, _r = ["x", "y"], wr = ["innerHTML"], Sr = ["cx", "cy", "r", "fill"], $r = ["data-start", "data-end"], Lr = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Mr = {
  class: "vue-ui-xy-tag-arrow",
  height: "20",
  viewBox: "0 0 10 20",
  style: { position: "absolute", right: "100%", top: "50%", transform: "translateY(-50%)" }
}, Cr = ["fill"], Ar = ["innerHTML"], Tr = {
  class: "vue-ui-xy-tag-arrow",
  height: "100%",
  viewBox: "0 0 10 20",
  style: { position: "absolute", left: "100%", top: "50%", transform: "translateY(-50%)" }
}, zr = ["fill"], Ir = ["innerHTML"], Or = {
  class: "vue-ui-xy-tag-arrow",
  height: "20",
  viewBox: "0 0 10 20",
  style: { position: "absolute", right: "100%", top: "50%", transform: "translateY(-50%)" }
}, Nr = ["fill"], Pr = ["innerHTML"], Fr = {
  class: "vue-ui-xy-tag-arrow",
  height: "100%",
  viewBox: "0 0 10 20",
  style: { position: "absolute", left: "100%", top: "50%", transform: "translateY(-50%)" }
}, Vr = ["fill"], Rr = ["innerHTML"], Br = ["id"], Hr = ["onClick"], Dr = {
  key: 0,
  viewBox: "0 0 20 12",
  height: "1em",
  width: "1.43em"
}, Wr = ["stroke", "fill"], Er = {
  key: 1,
  viewBox: "0 0 40 40",
  height: "1em",
  width: "1em"
}, Yr = ["fill"], Xr = ["fill"], Gr = {
  key: 2,
  viewBox: "0 0 12 12",
  height: "1em",
  width: "1em"
}, qr = {
  style: { display: "flex", "flex-direction": "row", gap: "6px", "align-items": "center", "padding-left": "6px" },
  "data-dom-to-png-ignore": ""
}, Ur = ["innerHTML"], Zr = {
  __name: "vue-ui-xy",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    selectedXIndex: {
      type: Number,
      default: void 0
    }
  },
  emits: ["selectTimeLabel", "selectX", "selectLegend", "zoomStart", "zoomEnd", "zoomReset", "copyAlt"],
  setup(gl, { expose: yo, emit: ko }) {
    const O = gl, _o = Ee(() => import("./DataTable-Bo1pmAqG.js")), wo = Ee(() => import("./Tooltip-DL3QgjZf.js")), So = Ee(() => import("./UserOptions-CrLqEimB.js")), Hl = Ee(() => import("./BaseIcon-CNx8Tzl5.js")), $o = Ee(() => import("./vue-ui-table-sparkline-ClpLLKi2.js")), Lo = Ee(() => import("./PackageVersion-DN9d0ME5.js")), Mo = Ee(() => import("./PenAndPaper-DWV1FMIi.js")), Co = Ee(() => import("./BaseDraggableDialog-DSpKTn3R.js")), Ge = ko, Dl = zu(), Ao = Iu(), { vue_ui_xy: gt } = es(), { isThemeValid: To, warnInvalidTheme: zo } = us(), Me = m(null), bt = m(null), Ie = m(null), ut = m(null), pt = m(null), xt = m(null), Io = m(null), bl = m(null), pl = m(null), Ht = m(null), xl = m(null), yl = m(null), Be = m(null), kl = m(null), Oo = m(0), I = m(null), Wl = m(!0), ce = m(600), K = m(1e3), Dt = m("0 0 1000 600"), No = m({ x: 0, y: 0 }), yt = m({ line: "line", bar: "bar", plot: "plot" }), kt = m(!1), qe = m(!1), El = m(!1), Ue = m(null), q = m([]), P = m(Pl()), Yl = m(0), Xl = m(0), Gl = m(0), _t = m(!0), _l = m(0), F = m(null), ql = m(!1), wl = m(!0), Wt = m(!0), ve = m(null), Re = m({}), Et = m(null), Ul = m(!1), st = m(null), Yt = m(!1), Xt = m(!1), _ = m(null), Zl = Ou(null), Jl = m(!1), Kl = m(0), Ql = m(0), jl = ss({
      elementRef: Zl,
      minimumWidth: 2,
      minimumHeight: 2,
      stableFramesRequired: 2,
      once: !1,
      onSizeAccepted: () => {
        nt();
      }
    });
    function wt() {
      Zl.value = Me.value?.parentNode ?? null;
    }
    function ea() {
      return new Promise((l) => {
        requestAnimationFrame(() => {
          requestAnimationFrame(l);
        });
      });
    }
    async function nt() {
      const l = ++Ql.value;
      Jl.value = !1, await Ft(), await ea(), await ea(), l === Ql.value && (Kl.value += 1, Jl.value = !0);
    }
    const Po = g(() => ({
      height: ce.value,
      width: K.value
    }));
    function ta(l) {
      return Number.isFinite(l) ? Math.max(0, Math.floor(l)) : 0;
    }
    function Sl(l, u, e = 0) {
      return Number.isFinite(l) && Number.isFinite(u) && Math.abs(u) > 1e-9 ? l / u : e;
    }
    const la = m(!1), x = m({
      xAxis: 18,
      yAxis: 12,
      dataLabels: 20,
      plotLabels: 10
    }), Ce = m({ plot: 3, line: 3 });
    Pt(() => {
      Ul.value = !0, O.dataset.length && O.dataset.forEach((l, u) => {
        [null, void 0].includes(l.series) && Fl({
          componentName: "VueUiXy",
          type: "datasetSerieAttribute",
          property: "series (number[])",
          index: u,
          debug: St.value
        });
      }), wt(), jl.start(), nt();
    });
    function $l() {
      if (!Object.keys(O.config || {}).length)
        return gt;
      const l = Bt({
        userConfig: O.config,
        defaultConfig: gt
      });
      O.config && Rt(O.config, "chart.highlightArea") && (Array.isArray(O.config.chart.highlightArea) ? l.chart.highlightArea = O.config.chart.highlightArea : l.chart.highlightArea = [O.config.chart.highlightArea]), O.config && Rt(O.config, "chart.annotations") && Array.isArray(O.config.chart.annotations) && O.config.chart.annotations.length ? l.chart.annotations = O.config.chart.annotations.map((a) => Bt({
        defaultConfig: gt.chart.annotations[0],
        userConfig: a
      })) : l.chart.annotations = [], O.config && Rt(O.config, "chart.grid.position") && O.config.chart.grid.position === "start" && O.dataset.length && O.dataset.some((a) => a.type === "bar") && (l.chart.grid.position = "middle", Rt(O.config, "debug") && console.warn("Vue Data UI - VueUiXy - config.chart.grid.position was overriden to `middle` because your dataset contains a bar")), O.config && Rt(O.config, "chart.highlightArea") && (Array.isArray(O.config.chart.highlightArea) ? l.chart.highlightArea = O.config.chart.highlightArea.map((a) => aa({
        defaultConfig: gt.chart.highlightArea,
        userConfig: a
      })) : l.chart.highlightArea = aa({
        defaultConfig: gt.chart.highlightArea,
        userConfig: O.config.chart.highlightArea
      }));
      const u = l.theme;
      if (!u) return l;
      if (!To.value(l))
        return zo(l), l;
      const e = Bt({
        userConfig: cs[u] || O.config,
        defaultConfig: l
      }), s = Bt({
        userConfig: O.config,
        defaultConfig: e
      });
      return {
        ...s,
        customPalette: s.customPalette.length ? s.customPalette : Ju[u] || Rl
      };
    }
    function aa({ defaultConfig: l, userConfig: u }) {
      return Bt({
        defaultConfig: l,
        userConfig: u
      });
    }
    const Ll = g({
      get() {
        return !!O.dataset && O.dataset.length;
      },
      set(l) {
        return l;
      }
    }), t = m($l()), Ze = g(() => t.value.chart.userOptions.useCursorPointer), f = m({
      dataLabels: { show: !0 },
      showTooltip: !0,
      showTable: !1,
      isStacked: !1,
      useIndividualScale: !1
    });
    function Ml() {
      la.value ? (f.value.isStacked = t.value.chart.grid.labels.yAxis.stacked, f.value.useIndividualScale == null && (f.value.useIndividualScale = t.value.chart.grid.labels.yAxis.useIndividualScale)) : (f.value = {
        dataLabels: { show: !0 },
        showTooltip: t.value.chart.tooltip.show === !0,
        showTable: t.value.showTable === !0,
        isStacked: t.value.chart.grid.labels.yAxis.stacked,
        useIndividualScale: t.value.chart.grid.labels.yAxis.useIndividualScale
      }, la.value = !0);
    }
    const St = g(() => !!t.value.debug), Fo = g(() => no({
      defaultConfig: {
        useCssAnimation: !1,
        showTable: !1,
        chart: {
          annotations: [],
          highlightArea: [],
          backgroundColor: "#99999930",
          grid: {
            stroke: "#6A6A6A",
            labels: {
              show: !1,
              axis: {
                yLabel: "",
                xLabel: ""
              },
              xAxisLabels: { show: !1 },
              yAxis: {
                commonScaleSteps: 10,
                useNiceScale: !0,
                scaleMin: 0,
                scaleMax: 134
              },
              zeroLine: { show: !0 }
            }
          },
          padding: {
            top: 12,
            bottom: 24,
            left: 24,
            right: 24
          },
          userOptions: { show: !1 },
          zoom: {
            show: !1,
            startIndex: null,
            endIndex: null
          }
        },
        bar: {
          serieName: { show: !1 },
          labels: { show: !1 },
          border: {
            useSerieColor: !1,
            stroke: "#999999"
          }
        },
        line: {
          dot: {
            useSerieColor: !1,
            fill: "#8A8A8A"
          },
          labels: { show: !1 }
        }
      },
      userConfig: t.value.skeletonConfig ?? {}
    })), Vo = g(() => O.config?.skeletonDataset ?? [
      {
        name: "",
        series: [0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 134],
        type: "line",
        smooth: !0,
        color: "#BABABA"
      },
      {
        name: "",
        series: [0, 0.5, 1, 1.5, 2.5, 4, 6.5, 10.5, 17, 27.5, 44.5, 67],
        type: "bar",
        color: "#CACACA"
      }
    ]), { loading: ue, FINAL_DATASET: pe, manualLoading: Gt } = ls({
      ...Nu(O),
      FINAL_CONFIG: t,
      prepareConfig: $l,
      callback: () => {
        Promise.resolve().then(async () => {
          await el(), f.value.showTable = t.value.showTable;
        });
      },
      skeletonDataset: Vo.value,
      skeletonConfig: no({
        defaultConfig: t.value,
        userConfig: Fo.value
      })
    });
    function Ro(l) {
      const u = /* @__PURE__ */ new WeakMap();
      return (e, ...s) => {
        let a = u.get(e);
        const n = JSON.stringify(s);
        if (a && a.has(n)) return a.get(n);
        const o = l(e, ...s);
        return a || (a = /* @__PURE__ */ new Map(), u.set(e, a)), a.set(n, o), o;
      };
    }
    const Bo = Ro((l, u) => Bl({ data: l, threshold: u })), He = (l) => Bo(l, t.value.downsample.threshold), Oe = g({
      get: () => Math.max(...pe.value.map((l) => He(l.series).length)),
      set: (l) => l
    }), b = m({ start: 0, end: Oe.value }), Ae = m({ start: 0, end: Oe.value }), oa = g(() => t.value.chart.zoom.preview.enable && (Ae.value.start !== b.value.start || Ae.value.end !== b.value.end));
    function ua(l, u) {
      Ae.value[l] = u;
    }
    function rt() {
      const l = Math.max(
        1,
        ...pe.value.map((s) => He(s.series).length)
      );
      let u = Math.max(0, Math.min(b.value.start ?? 0, l - 1)), e = Math.max(u + 1, Math.min(b.value.end ?? l, l));
      (!Number.isFinite(u) || !Number.isFinite(e) || e <= u) && (u = 0, e = l), b.value = { start: u, end: e }, Ae.value.start = u, Ae.value.end = e, Ie.value && (Ie.value.setStartValue(u), Ie.value.setEndValue(e));
    }
    const Ho = g(() => {
      const { left: l, top: u, width: e, height: s } = c.value, a = b.value.start, o = b.value.end - a, C = e / o, A = Ae.value.start - a, N = Ae.value.end - a, S = Math.max(0, Math.min(o, A)), T = Math.max(0, Math.min(o, N));
      return {
        x: l + S * C,
        y: u,
        width: (T - S) * C,
        height: s,
        fill: t.value.chart.zoom.preview.fill,
        stroke: t.value.chart.zoom.preview.stroke,
        "stroke-width": t.value.chart.zoom.preview.strokeWidth,
        "stroke-dasharray": t.value.chart.zoom.preview.strokeDasharray,
        "stroke-linecap": "round",
        "stroke-linejoin": "round",
        style: {
          pointerEvents: "none",
          transition: "none !important",
          animation: "none !important"
        }
      };
    });
    ht(() => O.selectedXIndex, (l) => {
      if ([null, void 0].includes(O.selectedXIndex)) {
        _.value = null;
        return;
      }
      const u = l - b.value.start;
      u < 0 || l >= b.value.end ? _.value = null : _.value = u ?? null;
    }, { immediate: !0 });
    const { isPrinting: qt, isImaging: Cl, generatePdf: sa, generateImage: na } = ts({
      elementId: `vue-ui-xy_${P.value}`,
      fileName: t.value.chart.title.text || "vue-ui-xy",
      options: t.value.chart.userOptions.print
    }), Do = m(!1), Ut = g(() => Bu(t.value.customPalette)), it = g(() => {
      const l = t.value.chart.grid.labels.yAxis.scaleMin;
      if (l == null) return null;
      const u = Number(l);
      return Number.isFinite(u) ? u : null;
    }), Wo = g(() => it.value !== null ? c.value.bottom : de.value), $t = g(() => {
      const l = t.value.chart.grid.labels.yAxis.scaleMax;
      if (l == null) return null;
      const u = Number(l);
      return Number.isFinite(u) ? u : null;
    }), ra = g(() => it.value !== null || $t.value !== null), ia = g(() => {
      const l = De.value.filter((u) => !q.value.includes(u.id)).flatMap((u) => Array.isArray(u.series) ? u.series : []).filter(Number.isFinite);
      return l.length ? {
        min: Math.min(...l),
        max: Math.max(...l)
      } : { min: 0, max: 1 };
    });
    function ca(l, u) {
      let e = Number.isFinite(l) ? l : 0, s = Number.isFinite(u) ? u : 1;
      return e === s ? s = e + 1 : e > s && ([e, s] = [s, e]), { min: e, max: s };
    }
    const Zt = g(() => {
      const { min: l, max: u } = ia.value;
      if (!ra.value) {
        const o = l;
        return o > 0 ? 0 : o;
      }
      const e = it.value !== null ? it.value : l > 0 ? 0 : l, s = $t.value !== null ? $t.value : u, a = l < e ? l : e, n = u > s ? u : s;
      return ca(a, n).min;
    }), Jt = g(() => {
      const { min: l, max: u } = ia.value;
      if (!ra.value) {
        const o = u;
        return Zt.value === o ? o + 1 : o;
      }
      const e = it.value !== null ? it.value : l > 0 ? 0 : l, s = $t.value !== null ? $t.value : u, a = l < e ? l : e, n = u > s ? u : s;
      return ca(a, n).max;
    }), Te = g(() => t.value.chart.grid.labels.yAxis.useNiceScale ? Vt(Zt.value, Jt.value < 0 ? 0 : Jt.value, t.value.chart.grid.labels.yAxis.commonScaleSteps) : at(Zt.value, Jt.value < 0 ? 0 : Jt.value, t.value.chart.grid.labels.yAxis.commonScaleSteps)), Lt = g(() => [null, void 0].includes(t.value.chart.grid.labels.yAxis.scaleMin) ? Te.value.min >= 0 ? 0 : Math.abs(Te.value.min) : -Te.value.min), De = g(() => Wl.value ? pe.value.map((l, u) => {
      const e = He(l.series), s = `uniqueId_${u}`;
      return {
        ...l,
        slotAbsoluteIndex: u,
        series: e.map((a) => hl(a) ? a : null).slice(b.value.start, b.value.end),
        color: Vl(l.color ? l.color : Ut.value[u] ? Ut.value[u] : Rl[u]),
        id: s,
        scaleLabel: l.scaleLabel || s
      };
    }) : pe.value), se = g(() => De.value.map((l, u) => ({
      absoluteIndex: u,
      ...l,
      series: l.series.map((e) => e + Lt.value),
      absoluteValues: l.series,
      segregate: () => ol(l),
      isSegregated: q.value.includes(l.id)
    }))), we = g(() => De.value.map((l, u) => ({
      ...l,
      series: l.series.map((e) => e + Lt.value),
      absoluteValues: l.series
    })).filter((l) => !q.value.includes(l.id))), Eo = g(() => q.value.length === se.value.length);
    function Yo() {
      let l = 0;
      xl.value && (l = Array.from(xl.value.querySelectorAll("text")).reduce((a, n) => {
        const o = n.getComputedTextLength();
        return (o > a ? o : a) + t.value.chart.grid.labels.yAxis.scaleValueOffsetX;
      }, 0));
      const u = pl.value ? pl.value.getBoundingClientRect().width + t.value.chart.grid.labels.axis.yLabelOffsetX + x.value.yAxis : 0, e = t.value.chart.grid.labels.yAxis.crosshairSize;
      return l + u + e;
    }
    const Mt = m(0), ct = m(0);
    function va() {
      const l = Ht.value;
      if (!l) {
        Mt.value = 0, ct.value = 0;
        return;
      }
      try {
        const u = l.getBBox();
        Mt.value = u?.height ?? 0, ct.value = u?.x ?? 0;
      } catch {
        Mt.value = 0, ct.value = 0;
      }
    }
    const da = g(() => {
      let l = 0;
      if (bl.value)
        try {
          l = bl.value.getBBox().height || 0;
        } catch {
          l = 0;
        }
      return l + Mt.value + x.value.xAxis;
    }), Xo = g(() => pe.value.some((l) => l.useProgression)), c = g(() => {
      Kl.value;
      let l = 0;
      const u = 36;
      t.value.chart.grid.labels.show && (f.value.useIndividualScale && !f.value.isStacked ? l = (pe.value.length - q.value.length) * (t.value.chart.grid.labels.yAxis.labelWidth + u) : f.value.useIndividualScale && f.value.isStacked ? l = t.value.chart.grid.labels.yAxis.labelWidth + u : l = Yo());
      const e = t.value.chart.labels.fontSize * 1.1, s = Xo.value ? 24 : 6;
      Ht.value && ct.value < 0 && (l += Math.abs(ct.value));
      const a = K.value - l - t.value.chart.grid.labels.yAxis.crosshairSize - s - t.value.chart.padding?.left - t.value.chart.padding?.right, n = t.value.chart.grid.position === "middle" ? 0 : a / $.value / 2;
      return {
        top: t.value.chart.padding?.top + e,
        right: K.value - s - t.value.chart.padding?.right,
        bottom: ce.value - da.value - t.value.chart.padding?.bottom - t.value.chart.grid.labels.axis.xLabelOffsetY,
        left: l + t.value.chart.grid.labels.yAxis.crosshairSize - n + t.value.chart.padding?.left,
        height: ce.value - da.value - t.value.chart.padding?.top - t.value.chart.padding?.bottom - e - t.value.chart.grid.labels.axis.xLabelOffsetY,
        width: a,
        scaleLabelX: l,
        individualOffsetX: u
      };
    }), Go = g(() => {
      const l = t.value.chart.grid.position === "middle" ? 1 : 0, u = $.value + l, e = E(c.value?.top), s = E(c.value?.bottom);
      return Array.from({ length: u }).map((a, n) => {
        const o = c.value.width / $.value * n + c.value?.left + U.value;
        return `M${o},${e} L${o},${s}`;
      }).join(" ");
    }), qo = g(() => {
      if (!t.value.chart.grid.labels.xAxis.showCrosshairs) return "";
      const l = c.value.width / $.value, u = t.value.chart.grid.labels.xAxis.crosshairSize, e = t.value.chart.grid.labels.xAxis.crosshairsAlwaysAtZero;
      return jt.value.map((s, a) => {
        if (!s || !s.text) return null;
        const n = c.value?.left + l * a + l / 2, o = e ? de.value - (de.value === c.value?.bottom ? 0 : u / 2) : c.value?.bottom, C = e ? de.value + u / (de.value === c.value?.bottom ? 1 : 2) : c.value?.bottom + u;
        return `M${n},${o} L${n},${C}`;
      }).filter(Boolean).join(" ");
    });
    function ha() {
      return !!Ao?.vnode.props?.onSelectTimeLabel;
    }
    function Uo(l, u, e) {
      if (!Et.value) {
        const s = document.createElement("canvas");
        Et.value = s.getContext("2d");
      }
      return Et.value.font = `normal ${l}px sans-serif`, Et.value;
    }
    function fa() {
      const l = Me.value.querySelectorAll(".vue-ui-xy-tag");
      l.length && Array.from(l).forEach((u) => u.style.opacity = "0");
    }
    function Kt(l, u, e, s, a) {
      e && (Re.value[`${l}_${u}_${s}_${a}`] = e);
    }
    const ma = m(!1);
    async function ga(l = !1) {
      await Ft(), ma.value = l, ql.value && (Wt.value = l);
    }
    function Al() {
      kt.value = !kt.value;
    }
    const le = m([]), Qt = m([]);
    let ba = 0;
    dl(() => {
      const l = ++ba;
      (async () => {
        const u = Math.max(
          ...pe.value.map(
            (s) => Bl({
              data: s.series,
              threshold: t.value.downsample.threshold
            }).length
          )
        ), e = await po({
          values: t.value.chart.grid.labels.xAxisLabels.values,
          maxDatapoints: u,
          formatter: t.value.chart.grid.labels.xAxisLabels.datetimeFormatter,
          start: b.value.start,
          end: b.value.end
        });
        l === ba && (le.value = e);
      })();
    });
    let pa = 0;
    dl(() => {
      const l = ++pa;
      (async () => {
        const u = Math.max(
          ...pe.value.map(
            (s) => Bl({
              data: s.series,
              threshold: t.value.downsample.threshold
            }).length
          )
        ), e = await po({
          values: t.value.chart.grid.labels.xAxisLabels.values,
          maxDatapoints: u,
          formatter: t.value.chart.grid.labels.xAxisLabels.datetimeFormatter,
          start: 0,
          end: Oe.value
        });
        l === pa && (Qt.value = e);
      })();
    });
    const Zo = g(() => {
      const l = t.value.chart.grid.labels.xAxisLabels.modulo;
      return le.value.length ? Math.min(l, [...new Set(le.value.map((u) => u.text))].length) : l;
    }), jt = g(() => {
      const l = t.value.chart.grid.labels.xAxisLabels, u = le.value || [], e = Qt.value || [], s = b.value.start ?? 0, a = _.value, n = $.value, o = u.map((A) => A?.text ?? ""), C = e.map((A) => A?.text ?? "");
      return Hu(
        !!l.showOnlyFirstAndLast,
        !!l.showOnlyAtModulo,
        Math.max(1, Zo.value || 1),
        o,
        C,
        s,
        a,
        n
      );
    }), Jo = g(() => (jt.value || []).map((l) => l?.text ?? "").join("|"));
    Pt(() => {
      requestAnimationFrame(() => {
        va();
      }), ht(
        [
          () => Jo.value,
          () => t.value.chart.grid.labels.xAxisLabels.rotation,
          () => x.value.xAxis,
          () => K.value,
          () => ce.value
        ],
        async () => {
          await Ft(), requestAnimationFrame(() => {
            va();
          });
        },
        { flush: "post" }
      );
    }), Nl(() => {
      Mt.value = 0, ct.value = 0, jl.stop();
    });
    function xa(l, u) {
      const e = we.value.map((s) => ({
        shape: s.shape || null,
        name: s.name,
        color: s.color,
        type: s.type,
        value: s.absoluteValues.find((a, n) => n === u),
        comments: s.comments || [],
        prefix: s.prefix || t.value.chart.labels.prefix,
        suffix: s.suffix || t.value.chart.labels.suffix
      }));
      Ge("selectTimeLabel", {
        datapoint: e,
        absoluteIndex: l.absoluteIndex,
        label: l.text
      });
    }
    const $ = g(() => {
      const l = ta((b.value.end ?? 0) - (b.value.start ?? 0));
      return Math.max(1, l);
    });
    function Ko(l) {
      F.value = l;
    }
    function ya() {
      f.value.isStacked = !f.value.isStacked, f.value.isStacked ? f.value.useIndividualScale = !0 : f.value.useIndividualScale = t.value.chart.grid.labels.yAxis.useIndividualScale;
    }
    function Tl(l) {
      St.value && l.autoScaling && (t.value.chart.grid.labels.yAxis.useIndividualScale || console.warn(`VueUiXy (datapoint: ${l.name}) : autoScaling only works when config.chart.grid.labels.yAxis.useIndividualScale is set to true`), t.value.chart.grid.labels.yAxis.stacked || console.warn(`VueUiXy (datapoint: ${l.name}) : autoScaling only works when config.chart.grid.labels.yAxis.stacked is set to true`));
    }
    function Qo(l, u) {
      const e = ta(l), s = Array(e).fill(0);
      for (let a = 0; a < u.length && a < e; a += 1) s[a] = u[a] ?? 0;
      return s;
    }
    function ka(l) {
      const u = Math.max(...pe.value.map((e) => He(e.series).length));
      return l > u ? u : l < 0 || t.value.chart.zoom.startIndex !== null && l < t.value.chart.zoom.startIndex ? t.value.chart.zoom.startIndex !== null ? t.value.chart.zoom.startIndex + 1 : 1 : l;
    }
    const Ct = m(!1), _a = m(!1), wa = m(0), Sa = m(0);
    function $a(l, u) {
      const e = Number(l), s = Number(u);
      return !Number.isFinite(e) || !Number.isFinite(s) ? !1 : !Object.is(e, s);
    }
    function el() {
      if (!Ct.value) {
        Ct.value = !0;
        try {
          const { startIndex: l, endIndex: u } = t.value.chart.zoom, e = Math.max(...pe.value.map((n) => He(n.series).length));
          wa.value = 0, Sa.value = e;
          const s = l ?? 0, a = u != null ? Math.min(ka(u + 1), e) : e;
          tl.value = !0, b.value.start = s, b.value.end = a, Ae.value.start = s, Ae.value.end = a, rt(), _a.value = !0;
        } finally {
          queueMicrotask(() => {
            tl.value = !1;
          }), Ct.value = !1;
        }
      }
    }
    const tl = m(!1);
    function jo(l) {
      if (Ct.value || tl.value) return;
      const u = Number(l);
      Ge("zoomStart", {
        index: u,
        isZoom: $a(u, wa.value)
      }), Number.isFinite(u) && u !== b.value.start && (b.value.start = u, Ae.value.start = u, rt());
    }
    function eu(l) {
      if (Ct.value || tl.value) return;
      const u = ka(l);
      Ge("zoomEnd", {
        index: u,
        isZoom: $a(u, Sa.value)
      }), u !== b.value.end && (b.value.end = u, Ae.value.end = u, rt());
    }
    async function tu() {
      await el(), Ge("zoomReset");
    }
    function xe(l) {
      return ![null, void 0, NaN, 1 / 0, -1 / 0].includes(l);
    }
    const La = g(() => Te.value.max + Lt.value);
    function Je(l) {
      return l / (xe(La.value) ? La.value : 1);
    }
    const de = g(() => isNaN(Je(Lt.value)) ? c.value?.bottom : c.value?.bottom - c.value.height * Je(Lt.value));
    function Ma(l) {
      const u = ![null, void 0].includes(t.value.chart.grid.labels.yAxis.scaleMin) && t.value.chart.grid.labels.yAxis.scaleMin > 0 && Zt.value >= 0 ? c.value?.bottom : de.value;
      return l.value >= 0 ? V(u - l.y <= 0 ? 1e-5 : u - l.y) : V(l.y - de.value <= 0 ? 1e-5 : l.y - de.value);
    }
    function Ca(l) {
      return l.value >= 0 ? V(l.zeroPosition - l.y <= 0 ? 1e-5 : l.zeroPosition - l.y) : V(l.y - l.zeroPosition <= 0 ? 1e-5 : l.zeroPosition - l.y);
    }
    const J = g(() => {
      const l = Math.max(1, $.value), u = Math.max(1, c.value.width), e = Math.max(1, De.value.filter((s) => s.type === "bar" && !q.value.includes(s.id)).length);
      return {
        bar: Sl(u, l * e, 1),
        plot: Sl(u, l, 1),
        line: Sl(u, l, 1)
      };
    });
    function We() {
      return f.value.useIndividualScale && f.value.isStacked ? J.value.line - c.value.width / $.value * 0.1 : J.value.bar;
    }
    function ll(l) {
      return f.value.useIndividualScale && f.value.isStacked ? l.x + c.value.width / $.value * 0.05 : l.x + J.value.bar / 2;
    }
    function Aa(l) {
      return l.value >= 0 ? l.y : [null, void 0, NaN, 1 / 0, -1 / 0].includes(de.value) ? c?.bottom.value : de.value;
    }
    function Ta(l) {
      return l.value >= 0 ? l.y : [null, void 0, NaN, 1 / 0, -1 / 0].includes(l.zeroPosition) ? 0 : l.zeroPosition;
    }
    const At = m(null);
    function za(l) {
      const u = ve.value;
      if (!u) return null;
      if (u.createSVGPoint && u.getScreenCTM) {
        const T = u.createSVGPoint();
        T.x = l.clientX, T.y = l.clientY;
        const Y = u.getScreenCTM();
        if (Y) {
          const k = T.matrixTransform(Y.inverse());
          return { x: k.x, y: k.y, ok: !0 };
        }
      }
      const e = u.getBoundingClientRect(), s = u.viewBox?.baseVal || { x: 0, y: 0, width: e.width, height: e.height }, a = Math.min(e.width / s.width, e.height / s.height), n = s.width * a, o = s.height * a, C = (e.width - n) / 2, A = (e.height - o) / 2, N = (l.clientX - e?.left - C) / a + s.x, S = (l.clientY - e?.top - A) / a + s.y;
      return { x: N, y: S, ok: !0 };
    }
    let Ke = 0;
    function lu(l) {
      kt.value || (Ke && cancelAnimationFrame(Ke), Ke = requestAnimationFrame(() => {
        Ke = 0;
        const u = za(l);
        if (!u || !ve.value) {
          al();
          return;
        }
        const { left: e, right: s, top: a, bottom: n, width: o } = c.value;
        if (u.x < e || u.x > s || u.y < a || u.y > n) {
          al();
          return;
        }
        const C = u.x - e, A = o / $.value, N = Math.floor(C / A);
        N >= 0 && N < $.value ? At.value !== N && (At.value = N, Ga(!0, N)) : al();
      }));
    }
    function al() {
      Ke && (cancelAnimationFrame(Ke), Ke = 0), At.value = null, Ga(!1, null);
    }
    function Ia(l) {
      const u = za(l);
      if (u && ve.value) {
        const { left: e, right: s, top: a, bottom: n, width: o } = c.value;
        if (u.x >= e && u.x <= s && u.y >= a && u.y <= n) {
          const C = o / Math.max(1, $.value), A = Math.floor((u.x - e) / C);
          if (A >= 0 && A < $.value) {
            Oa(A);
            return;
          }
        }
      }
      At.value != null && Oa(At.value);
    }
    function Oa(l) {
      const u = we.value.map((e) => ({
        name: e.name,
        value: [null, void 0, NaN].includes(e.absoluteValues[l]) ? null : e.absoluteValues[l],
        color: e.color,
        type: e.type
      }));
      Ge(
        "selectX",
        {
          dataset: u,
          index: l,
          indexLabel: t.value.chart.grid.labels.xAxisLabels.values[l]
        }
      ), t.value.events.datapointClick && t.value.events.datapointClick({ datapoint: u, seriesIndex: l + b.value.start });
    }
    function au() {
      return se.value.map((l) => ({
        values: l.absoluteValues,
        color: l.color,
        name: l.name,
        type: l.type
      }));
    }
    async function ou({ scale: l = 2 } = {}) {
      if (!Me.value) return;
      const { width: u, height: e } = Me.value.getBoundingClientRect(), s = u / e, { imageUri: a, base64: n } = await rs({ domElement: Me.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: a,
        base64: n,
        title: t.value.chart.title.text,
        width: u,
        height: e,
        aspectRatio: s
      };
    }
    function uu() {
      q.value.length ? q.value = [] : se.value.forEach((l) => {
        q.value.push(l.id);
      });
    }
    function ol(l) {
      if (q.value.includes(l.id))
        q.value = q.value.filter((u) => u !== l.id);
      else {
        if (q.value.length + 1 === De.value.length) return;
        q.value.push(l.id);
      }
      Ge("selectLegend", we.value.map((u) => ({
        name: u.name,
        values: u.absoluteValues,
        color: u.color,
        type: u.type
      }))), _l.value += 1;
    }
    function Na(l) {
      if (!se.value.length)
        return t.value.debug && console.warn("VueUiXy - There are no series to show."), null;
      const u = se.value.find((e) => e.name === l);
      return u || (t.value.debug && console.warn(`VueUiXy - Series name not found "${l}"`), null);
    }
    function su(l) {
      const u = Na(l);
      u !== null && q.value.includes(u.id) && ol({ id: u.id });
    }
    function nu(l) {
      const u = Na(l);
      u !== null && (q.value.includes(u.id) || ol({ id: u.id }));
    }
    const ru = g(() => {
      const l = t.value.chart.title.text || "Chart visualization", u = t.value.chart.title.subtitle.text || "";
      return `${l}. ${u}`;
    }), Pa = g(() => ({
      linePlot: $.value > t.value.line.dot.hideAboveMaxSerieLength
    })), iu = g(() => t.value.chart.userOptions.show && (!t.value.chart.title.show || !t.value.chart.title.text)), cu = g(() => {
      if (Array.isArray(t.value.chart.highlightArea))
        return t.value.chart.highlightArea.map((u) => {
          const e = Math.min(u.to, Oe.value - 1);
          return {
            ...u,
            span: u.from === e ? 1 : e < u.from ? 0 : e - u.from + 1
          };
        });
      const l = {
        ...t.value.chart.highlightArea,
        to: Math.min(t.value.chart.highlightArea.to, Oe.value - 1)
      };
      return [{ ...l, span: l.from === l.to ? 1 : l.to < l.from ? 0 : l.to - l.from + 1 }];
    }), Fa = g(() => Wl.value ? pe.value.map((l, u) => ({
      ...l,
      series: He(l.series),
      id: `uniqueId_${u}`,
      color: Vl(l.color ? l.color : Ut.value[u] ? Ut.value[u] : Rl[u])
    })) : pe.value), vu = g(() => we.value.map((l) => {
      const u = l.absoluteValues.map((e) => [void 0, null].includes(e) ? 0 : e);
      return {
        id: l.id,
        name: l.name,
        color: l.color,
        values: Qo($.value, u)
      };
    })), du = g(() => ({
      responsiveBreakpoint: t.value.table.responsiveBreakpoint,
      roundingValues: t.value.table.rounding,
      showAverage: !1,
      showMedian: !1,
      showTotal: !1,
      fontFamily: t.value.chart.fontFamily,
      prefix: t.value.chart.labels.prefix,
      suffix: t.value.chart.labels.suffix,
      colNames: le.value.map((l, u) => t.value.table.useDefaultTimeFormat ? l.text : vt.value(u + b.value.start, t.value.table.timeFormat)),
      thead: {
        backgroundColor: t.value.table.th.backgroundColor,
        color: t.value.table.th.color,
        outline: t.value.table.th.outline
      },
      tbody: {
        backgroundColor: t.value.table.td.backgroundColor,
        color: t.value.table.td.color,
        outline: t.value.table.td.outline
      },
      userOptions: {
        show: !1
      },
      sparkline: {
        animation: { show: !1 }
      }
    })), U = g(() => t.value.chart.grid.position === "middle" ? 0 : c.value.width / $.value / 2), Qe = g(() => Du(se.value.filter((l) => !q.value.includes(l.id))));
    function hu(l, u) {
      const e = /* @__PURE__ */ Object.create(null);
      for (let s = 0; s < l.length; s += 1) {
        const a = l[s], n = String(u(a));
        e[n] || (e[n] = []), e[n].push(a);
      }
      return e;
    }
    const L = g(() => {
      const l = hu(Qe.value, (e) => e.scaleLabel), u = {};
      for (const [e, s] of Object.entries(l)) {
        const a = s.flatMap((n) => n.absoluteValues);
        u[e] = {
          min: Math.min(...a) || 0,
          max: Math.max(...a) || 1,
          groupId: `scale_group_${Pl()}`
        };
      }
      return u;
    }), fu = g(() => {
      const l = De.value.filter((u) => u.type === "bar").filter((u) => !q.value.includes(u.id)).length;
      return c.value.width / $.value / l - ul.value * l;
    }), ul = g(() => J.value.line * t.value.bar.periodGap), sl = g(() => Math.max(1e-5, We() - (f.value.useIndividualScale && f.value.isStacked ? 0 : ul.value))), nl = g(() => sl.value * Math.min(Math.abs(t.value.bar.innerGap), 0.95)), mu = g(() => {
      if (!t.value.chart.zoom.minimap.show) return [];
      const l = Fa.value.filter((a) => !q.value.includes(a.id)), u = Math.max(...l.map((a) => a.series.length)), e = [];
      for (let a = 0; a < u; a += 1)
        e.push(l.map((n) => n.series[a] || 0).reduce((n, o) => (n || 0) + (o || 0), 0));
      const s = Math.min(...e);
      return e.map((a) => a + (s < 0 ? Math.abs(s) : 0));
    }), gu = g(() => t.value.chart.zoom.minimap.show ? Fa.value.map((u) => ({
      ...u,
      isVisible: !q.value.includes(u.id)
    })) : []), je = g(() => we.value.map((l) => ({
      slotAbsoluteIndex: l.slotAbsoluteIndex,
      shape: l.shape || null,
      name: l.name,
      color: l.color,
      type: l.type,
      value: l.absoluteValues.find((u, e) => e === _.value),
      comments: l.comments || [],
      prefix: l.prefix || t.value.chart.labels.prefix,
      suffix: l.suffix || t.value.chart.labels.suffix
    }))), Va = g(() => Te.value.ticks.map((l) => ({
      y: l >= 0 ? de.value - c.value.height * Je(l) : de.value + c.value.height * Je(Math.abs(l)),
      value: l,
      prefix: t.value.chart.labels.prefix,
      suffix: t.value.chart.labels.suffix
    }))), Ra = g(() => {
      const l = t.value.chart.annotations;
      if (!l || !Array.isArray(l) || l.every((S) => !S.show)) return [];
      const u = l.filter(
        (S) => S.show && (S.yAxis.yTop != null || S.yAxis.yBottom != null)
      );
      if (!u.length) return [];
      const { left: e, right: s } = c.value, a = de.value, n = c.value.height, o = Te.value.min, A = Te.value.max - o, N = (S) => {
        const T = (S - 0) / A;
        return a - T * n;
      };
      return u.map((S) => {
        const { yAxis: { yTop: T, yBottom: Y, label: k } } = S, B = T != null && Y != null && T !== Y, z = T == null ? null : N(T), ae = Y == null ? null : N(Y), fe = Uo(k.fontSize);
        fe.font = `${k.fontSize}px sans-serif`;
        const Z = fe.measureText(k.text).width, Se = k.fontSize, Ne = (k.position === "start" ? e + k.padding?.left : s - k.padding?.right) + k.offsetX, ke = (z != null && ae != null ? Math.min(z, ae) : z ?? ae) - k.fontSize / 3 + k.offsetY - k.padding?.top;
        let D;
        k.textAnchor === "middle" ? D = Ne - Z / 2 - k.padding?.left : k.textAnchor === "end" ? D = Ne - Z - k.padding?.right : D = Ne - k.padding?.left;
        const W = ke - Se * 0.75 - k.padding?.top;
        return {
          show: ![z, ae, W].includes(NaN),
          id: `annotation_y_${Pl()}`,
          hasArea: B,
          areaHeight: B ? Math.abs(z - ae) : 0,
          yTop: z,
          yBottom: ae,
          config: S.yAxis,
          x1: e,
          x2: s,
          _text: { x: Ne, y: ke },
          _box: {
            x: D,
            y: W,
            width: Z + k.padding?.left + k.padding?.right,
            height: Se + k.padding?.top + k.padding?.bottom,
            fill: k.backgroundColor,
            stroke: k.border.stroke,
            rx: k.border.rx,
            ry: k.border.ry,
            strokeWidth: k.border.strokeWidth
          }
        };
      });
    });
    function Ba(l, u) {
      const e = l[u - 1], s = l[u + 1];
      let a = !!e && !!s && e.value == null && s.value == null || !e && !!s && s.value == null || !!e && !s && e.value == null;
      return xe(l[u].value) && a && t.value.line.cutNullValues;
    }
    const ye = g(() => {
      const l = Qe.value.filter((o) => ["bar", "line", "plot"].includes(o.type)), u = l.length, e = t.value.chart.grid.labels.yAxis.gap, s = f.value.isStacked, a = s ? e * (u - 1) : 0, n = c.value.height - a;
      return l.filter((o) => o.type === "bar").map((o, C) => {
        Tl(o);
        const A = L.value[o.scaleLabel].min, N = L.value[o.scaleLabel].max;
        o.absoluteValues.filter((H) => ![null, void 0].includes(H)).map((H) => (H - A) / (N - A));
        const S = {
          valueMin: A,
          valueMax: N < 0 ? 0 : N
        }, T = {
          max: o.scaleMax || Math.max(...o.absoluteValues) || 1,
          min: o.scaleMin || Math.min(...o.absoluteValues.filter((H) => ![void 0, null].includes(H))) > 0 ? 0 : Math.min(...o.absoluteValues.filter((H) => ![null, void 0].includes(H)))
        }, Y = o.scaleSteps || t.value.chart.grid.labels.yAxis.commonScaleSteps, k = 1.0000001, B = t.value.chart.grid.labels.yAxis.useNiceScale ? Vt(T.min, T.max === T.min ? T.max * k : T.max, Y) : at(T.min, T.max === T.min ? T.max * k : T.max, Y), z = t.value.chart.grid.labels.yAxis.useNiceScale ? Vt(S.valueMin, S.valueMax === S.valueMin ? S.valueMax * k : S.valueMax, Y) : at(S.valueMin, S.valueMax === S.valueMin ? S.valueMax * k : S.valueMax, Y), ae = B.min >= 0 ? 0 : Math.abs(B.min), fe = 0, Z = B.max + ae, Se = z.max + Math.abs(fe), Ne = o.stackIndex, dt = u - 1 - Ne, ke = s ? 1 - o.cumulatedStackRatio : 0, D = s ? n * ke + e * dt : 0, W = s ? n * o.stackRatio : c.value.height, Q = c.value?.bottom - D - W * ae / Z, _e = c.value?.bottom - D - W * fe / Se, j = se.value.filter((H) => H.type === "bar").filter((H) => !q.value.includes(H.id)).length, It = o.series.map((H, ne) => {
          const Ot = f.value.useIndividualScale ? (o.absoluteValues[ne] + ae) / Z : Je(H), et = f.value.useIndividualScale && f.value.isStacked ? c.value?.left + c.value.width / $.value * ne : c.value?.left + J.value.bar * C + J.value.bar * ne * j - fu.value / 2 - C * ul.value;
          return {
            yOffset: V(D),
            individualHeight: V(W),
            x: V(et),
            y: c.value?.bottom - D - W * Ot,
            value: o.absoluteValues[ne],
            zeroPosition: V(Q),
            individualMax: V(Z),
            comment: o.comments && o.comments.slice(b.value.start, b.value.end)[ne] || ""
          };
        }), me = o.absoluteValues.map((H) => z.min >= 0 ? (H - Math.abs(z.min)) / (z.max - Math.abs(z.min)) : (H + Math.abs(z.min)) / (z.max + Math.abs(z.min))), Pe = o.series.map((H, ne) => {
          const Ot = f.value.useIndividualScale && f.value.isStacked ? c.value?.left + c.value.width / $.value * ne : c.value?.left - J.value.bar / 2 + J.value.bar * C + J.value.bar * ne * se.value.filter((et) => et.type === "bar").filter((et) => !q.value.includes(et.id)).length;
          return {
            yOffset: V(D),
            individualHeight: V(W),
            x: V(Ot),
            y: V(c.value?.bottom - V(D) - (V(W) * me[ne] || 0)),
            value: o.absoluteValues[ne],
            zeroPosition: V(Q),
            individualMax: V(Z),
            comment: o.comments && o.comments.slice(b.value.start, b.value.end)[ne] || ""
          };
        }), G = B.ticks.map((H) => ({
          y: H >= 0 ? Q - W * (H / Z) : Q + W * Math.abs(H) / Z,
          value: H,
          prefix: o.prefix || t.value.chart.labels.prefix,
          suffix: o.suffix || t.value.chart.labels.suffix,
          datapoint: o
        })), ee = z.ticks.map((H) => {
          const ne = (H - z.min) / (z.max - z.min);
          return {
            y: H >= 0 ? _e - W * ne : _e + W * ne,
            value: H,
            prefix: o.prefix || t.value.chart.labels.prefix,
            suffix: o.suffix || t.value.chart.labels.suffix,
            datapoint: o
          };
        });
        return L.value[o.scaleLabel].name = o.name, L.value[o.scaleLabel].groupName = o.scaleLabel, L.value[o.scaleLabel].groupColor = t.value.chart.grid.labels.yAxis.groupColor || o.color, L.value[o.scaleLabel].color = o.color, L.value[o.scaleLabel].scaleYLabels = o.autoScaling ? ee : G, L.value[o.scaleLabel].zeroPosition = o.autoScaling ? _e : Q, L.value[o.scaleLabel].individualMax = o.autoScaling ? Se : Z, L.value[o.scaleLabel].scaleLabel = o.scaleLabel, L.value[o.scaleLabel].id = o.id, L.value[o.scaleLabel].yOffset = D, L.value[o.scaleLabel].individualHeight = W, L.value[o.scaleLabel].autoScaleYLabels = ee, L.value[o.scaleLabel].unique = Qe.value.filter((H) => H.scaleLabel === o.scaleLabel).length === 1, {
          ...o,
          yOffset: D,
          autoScaleYLabels: ee,
          individualHeight: W,
          scaleYLabels: o.autoScaling ? ee : G,
          individualScale: o.autoScaling ? z : B,
          individualMax: o.autoScaling ? Se : Z,
          zeroPosition: o.autoScaling ? _e : Q,
          plots: o.autoScaling ? Pe : It,
          groupId: L.value[o.scaleLabel].groupId
        };
      });
    }), te = g(() => {
      const l = Qe.value.filter((o) => ["bar", "line", "plot"].includes(o.type)), u = l.length, e = t.value.chart.grid.labels.yAxis.gap, s = f.value.isStacked, a = s ? e * (u - 1) : 0, n = c.value.height - a;
      return l.filter((o) => o.type === "line").map((o, C) => {
        Tl(o);
        const A = L.value[o.scaleLabel].min, N = L.value[o.scaleLabel].max;
        o.absoluteValues.filter((M) => ![null, void 0].includes(M)).map((M) => (M - A) / (N - A));
        const S = {
          valueMin: A,
          valueMax: N
        }, T = {
          max: o.scaleMax || Math.max(...o.absoluteValues) || 1,
          min: o.scaleMin || (Math.min(...o.absoluteValues) > 0 ? 0 : Math.min(...o.absoluteValues))
        }, Y = o.scaleSteps || t.value.chart.grid.labels.yAxis.commonScaleSteps, k = 1.0000001, B = t.value.chart.grid.labels.yAxis.useNiceScale ? Vt(T.min, T.max === T.min ? T.max * k : T.max, Y) : at(T.min, T.max === T.min ? T.max * k : T.max, Y), z = t.value.chart.grid.labels.yAxis.useNiceScale ? Vt(S.valueMin, S.valueMax === S.valueMin ? S.valueMax * k : S.valueMax, Y) : at(S.valueMin, S.valueMax === S.valueMin ? S.valueMax * k : S.valueMax, Y), ae = B.min >= 0 ? 0 : Math.abs(B.min), fe = 0, Z = B.max + Math.abs(ae), Se = z.max + Math.abs(fe), Ne = o.stackIndex, dt = u - 1 - Ne, ke = s ? 1 - o.cumulatedStackRatio : 0, D = s ? n * ke + e * dt : 0, W = s ? n * o.stackRatio : c.value.height, Q = c.value?.bottom - D - W * ae / Z, _e = c.value?.bottom - D - W * fe / Se, j = o.series.map((M, re) => {
          const Tu = f.value.useIndividualScale ? (o.absoluteValues[re] + Math.abs(ae)) / Z : Je(M);
          return {
            x: V(c.value?.left + J.value.line / 2 + J.value.line * re),
            y: V(c.value?.bottom - D - W * Tu),
            value: o.absoluteValues[re],
            comment: o.comments && o.comments.slice(b.value.start, b.value.end)[re] || ""
          };
        }), It = o.absoluteValues.map((M) => z.min >= 0 ? (M - Math.abs(z.min)) / (z.max - Math.abs(z.min)) : (M + Math.abs(z.min)) / (z.max + Math.abs(z.min))), me = o.series.map((M, re) => [void 0, null].includes(o.absoluteValues[re]) ? {
          x: V(c.value?.left + J.value.line / 2 + J.value.line * re),
          y: Q,
          value: o.absoluteValues[re],
          comment: o.comments && o.comments.slice(b.value.start, b.value.end)[re] || ""
        } : {
          x: V(c.value?.left + J.value.line / 2 + J.value.line * re),
          y: V(c.value?.bottom - D - (W * It[re] || 0)),
          value: o.absoluteValues[re],
          comment: o.comments && o.comments.slice(b.value.start, b.value.end)[re] || ""
        }), Pe = o.dashIndices && Array.isArray(o.dashIndices) && o?.dashIndices?.length > 0, G = t.value.line.cutNullValues ? ro(j) : io(j.filter((M) => M.value !== null)), ee = t.value.line.cutNullValues ? ro(me) : io(me.filter((M) => M.value !== null)), H = t.value.line.cutNullValues ? co(j) : vo(j.filter((M) => M.value !== null)), ne = t.value.line.cutNullValues ? co(me) : vo(me.filter((M) => M.value !== null)), Ot = Pe ? Wu(t.value.line.cutNullValues ? j : j.filter((M) => M.value !== null), o.dashIndices.map((M) => M - b.value.start)) : [], et = Pe ? Eu(t.value.line.cutNullValues ? j : j.filter((M) => M.value !== null), o.dashIndices.map((M) => M - b.value.start)) : [], cl = B.ticks.map((M) => ({
          y: M >= 0 ? Q - W * (M / Z) : Q + W * Math.abs(M) / Z,
          value: M,
          prefix: o.prefix || t.value.chart.labels.prefix,
          suffix: o.suffix || t.value.chart.labels.suffix,
          datapoint: o
        })), vl = z.ticks.map((M) => {
          const re = (M - z.min) / (z.max - z.min);
          return {
            y: M >= 0 ? _e - W * re : _e + W * re,
            value: M,
            prefix: o.prefix || t.value.chart.labels.prefix,
            suffix: o.suffix || t.value.chart.labels.suffix,
            datapoint: o
          };
        });
        L.value[o.scaleLabel].name = o.name, L.value[o.scaleLabel].groupName = o.scaleLabel, L.value[o.scaleLabel].groupColor = t.value.chart.grid.labels.yAxis.groupColor || o.color, L.value[o.scaleLabel].color = o.color, L.value[o.scaleLabel].scaleYLabels = o.autoScaling ? vl : cl, L.value[o.scaleLabel].zeroPosition = o.autoScaling ? _e : Q, L.value[o.scaleLabel].individualMax = o.autoScaling ? Se : Z, L.value[o.scaleLabel].scaleLabel = o.scaleLabel, L.value[o.scaleLabel].id = o.id, L.value[o.scaleLabel].yOffset = D, L.value[o.scaleLabel].individualHeight = W, L.value[o.scaleLabel].autoScaleYLabels = vl, L.value[o.scaleLabel].unique = Qe.value.filter((M) => M.scaleLabel === o.scaleLabel).length === 1;
        const Au = f.value.useIndividualScale ? o.autoScaling ? _e : Q : Wo.value, Nt = Math.max(
          Math.max(
            o.autoScaling ? _e : cl.at(-1) ? cl.at(-1).y : 0,
            c.value?.top
          ),
          Au
        );
        return {
          ...o,
          temperatureColors: o.temperatureColors ? o.temperatureColors.map((M) => Vl(M)) : null,
          yOffset: D,
          autoScaleYLabels: vl,
          individualHeight: W,
          scaleYLabels: o.autoScaling ? vl : cl,
          individualScale: o.autoScaling ? z : B,
          individualMax: o.autoScaling ? Se : Z,
          zeroPosition: o.autoScaling ? _e : Q,
          curve: o.autoScaling ? ee : G,
          plots: o.autoScaling ? me : j,
          dashedStraight: Ot,
          dashedSmooth: et,
          hasDashedSegments: Pe,
          area: o.useArea ? f.value.useIndividualScale ? t.value.line.cutNullValues ? ho(
            o.autoScaling ? me : j,
            Nt
          ) : fo(
            o.autoScaling ? me.filter((M) => M.value !== null) : j.filter((M) => M.value !== null),
            Nt
          ) : t.value.line.cutNullValues ? ho(j, Nt) : fo(j.filter((M) => M.value !== null), Nt) : "",
          curveAreas: o.useArea ? Yu(
            o.autoScaling ? t.value.line.cutNullValues ? me : me.filter((M) => M.value !== null) : t.value.line.cutNullValues ? j : j.filter((M) => M.value !== null),
            Nt,
            t.value.line.cutNullValues
          ) : [],
          straight: o.autoScaling ? ne : H,
          groupId: L.value[o.scaleLabel].groupId
        };
      });
    }), he = g(() => {
      const l = Qe.value.filter((o) => ["bar", "line", "plot"].includes(o.type)), u = l.length, e = t.value.chart.grid.labels.yAxis.gap, s = f.value.isStacked, a = s ? e * (u - 1) : 0, n = c.value.height - a;
      return l.filter((o) => o.type === "plot").map((o) => {
        Tl(o);
        const C = L.value[o.scaleLabel].min, A = L.value[o.scaleLabel].max;
        o.absoluteValues.filter((G) => ![null, void 0].includes(G)).map((G) => (G - C) / (A - C));
        const N = {
          valueMin: C,
          valueMax: A
        }, S = {
          max: o.scaleMax || Math.max(...o.absoluteValues) || 1,
          min: o.scaleMin || Math.min(...o.absoluteValues) > 0 ? 0 : Math.min(...o.absoluteValues)
        }, T = o.scaleSteps || t.value.chart.grid.labels.yAxis.commonScaleSteps, Y = 1.0000001, k = at(S.min, S.max === S.min ? S.max * Y : S.max, T), B = at(N.valueMin, N.valueMax === N.valueMin ? N.valueMax * Y : N.valueMax, T), z = k.min >= 0 ? 0 : Math.abs(k.min), ae = 0, fe = k.max + z, Z = B.max + Math.abs(ae), Se = o.stackIndex, Ne = u - 1 - Se, dt = s ? 1 - o.cumulatedStackRatio : 0, ke = s ? n * dt + e * Ne : 0, D = s ? n * o.stackRatio : c.value.height, W = c.value?.bottom - ke - D * z / fe, Q = c.value?.bottom - ke - D * ae / Z, _e = o.series.map((G, ee) => {
          const H = f.value.useIndividualScale ? (o.absoluteValues[ee] + Math.abs(z)) / fe : Je(G);
          return {
            x: V(c.value?.left + J.value.plot / 2 + J.value.plot * ee),
            y: V(c.value?.bottom - ke - D * H),
            value: o.absoluteValues[ee],
            comment: o.comments && o.comments.slice(b.value.start, b.value.end)[ee] || ""
          };
        }), j = o.absoluteValues.map((G) => B.min >= 0 ? (G - Math.abs(B.min)) / (B.max - Math.abs(B.min)) : (G + Math.abs(B.min)) / (B.max + Math.abs(B.min))), It = o.series.map((G, ee) => ({
          x: V(c.value?.left + J.value.plot / 2 + J.value.plot * ee),
          y: V(c.value?.bottom - ke - (D * j[ee] || 0)),
          value: o.absoluteValues[ee],
          comment: o.comments && o.comments.slice(b.value.start, b.value.end)[ee] || ""
        })), me = k.ticks.map((G) => ({
          y: G >= 0 ? W - D * (G / fe) : W + D * Math.abs(G) / fe,
          value: G,
          prefix: o.prefix || t.value.chart.labels.prefix,
          suffix: o.suffix || t.value.chart.labels.suffix,
          datapoint: o
        })), Pe = B.ticks.map((G) => {
          const ee = (G - B.min) / (B.max - B.min);
          return {
            y: G >= 0 ? Q - D * ee : Q + D * ee,
            value: G,
            prefix: o.prefix || t.value.chart.labels.prefix,
            suffix: o.suffix || t.value.chart.labels.suffix,
            datapoint: o
          };
        });
        return L.value[o.scaleLabel].name = o.name, L.value[o.scaleLabel].groupName = o.scaleLabel, L.value[o.scaleLabel].groupColor = t.value.chart.grid.labels.yAxis.groupColor || o.color, L.value[o.scaleLabel].color = o.color, L.value[o.scaleLabel].scaleYLabels = o.autoScaling ? Pe : me, L.value[o.scaleLabel].zeroPosition = o.autoScaling ? Q : W, L.value[o.scaleLabel].individualMax = o.autoScaling ? Z : fe, L.value[o.scaleLabel].scaleLabel = o.scaleLabel, L.value[o.scaleLabel].id = o.id, L.value[o.scaleLabel].yOffset = ke, L.value[o.scaleLabel].individualHeight = D, L.value[o.scaleLabel].autoScaleYLabels = Pe, L.value[o.scaleLabel].unique = Qe.value.filter((G) => G.scaleLabel === o.scaleLabel).length === 1, {
          ...o,
          yOffset: ke,
          autoScaleYLabels: Pe,
          individualHeight: D,
          scaleYLabels: o.autoScaling ? Pe : me,
          individualScale: o.autoScaling ? B : k,
          individualMax: o.autoScaling ? Z : fe,
          zeroPosition: o.autoScaling ? Q : W,
          plots: o.autoScaling ? It : _e,
          groupId: L.value[o.scaleLabel].groupId
        };
      });
    }), Tt = g(() => {
      const l = te.value.map((n) => ({
        name: n.name,
        color: n.color,
        scale: n.individualScale,
        scaleYLabels: n.scaleYLabels,
        zero: n.zeroPosition,
        max: n.individualMax,
        scaleLabel: n.scaleLabel || "",
        id: n.id,
        yOffset: n.yOffset || 0,
        individualHeight: n.individualHeight || c.value.height,
        autoScaleYLabels: n.autoScaleYLabels
      })), u = ye.value.map((n) => ({
        name: n.name,
        color: n.color,
        scale: n.individualScale,
        scaleYLabels: n.scaleYLabels,
        zero: n.zeroPosition,
        max: n.individualMax,
        scaleLabel: n.scaleLabel || "",
        id: n.id,
        yOffset: n.yOffset || 0,
        individualHeight: n.individualHeight || c.value.height
      })), e = he.value.map((n) => ({
        name: n.name,
        color: n.color,
        scale: n.individualScale,
        scaleYLabels: n.scaleYLabels,
        // FIX
        zero: n.zeroPosition,
        max: n.individualMax,
        scaleLabel: n.scaleLabel || "",
        id: n.id,
        yOffset: n.yOffset || 0,
        individualHeight: n.individualHeight || c.value.height
      })), s = f.value.useIndividualScale && !f.value.isStacked ? Object.values(L.value) : [...l, ...u, ...e], a = s.flatMap((n) => n).length;
      return s.flatMap((n, o) => {
        let C = 0;
        C = f.value.isStacked ? c.value?.left : c.value?.left / a * (o + 1);
        const A = f.value.useIndividualScale && !f.value.isStacked ? n.unique ? n.name : n.groupName : n.name;
        return {
          unique: n.unique,
          id: n.id,
          groupId: n.groupId,
          scaleLabel: n.scaleLabel,
          name: ie(
            t.value.chart.grid.labels.yAxis.serieNameFormatter,
            A,
            A,
            n
          ),
          color: f.value.useIndividualScale && !f.value.isStacked ? n.unique ? n.color : n.groupColor : n.color,
          scale: n.scale,
          yOffset: n.yOffset,
          individualHeight: n.individualHeight,
          x: C,
          yLabels: n.scaleYLabels || n.scale.ticks.map((N) => ({
            y: N >= 0 ? n.zero - n.individualHeight * (N / n.max) : n.zero + n.individualHeight * Math.abs(N) / n.max,
            value: N
          }))
        };
      });
    }), Ha = g(() => {
      const l = t.value.line.interLine || {}, u = l.pairs || [], e = l.colors || [];
      if (!u.length) return [];
      const s = [];
      return u.forEach((a, n) => {
        const [o, C] = Array.isArray(a) ? a : [a.a, a.b];
        if (!o || !C) return;
        const A = te.value.find((k) => k.name === o), N = te.value.find((k) => k.name === C);
        if (!A || !N || A.type !== "line" || N.type !== "line") return;
        const S = e?.[n]?.[0] ?? A.color, T = e?.[n]?.[1] ?? N.color;
        Xu({
          lineA: A.plots,
          lineB: N.plots,
          smoothA: !!A.smooth,
          smoothB: !!N.smooth,
          colorLineA: S,
          colorLineB: T,
          sampleStepPx: 2,
          cutNullValues: t.value.line.cutNullValues
        }).forEach((k, B) => {
          s.push({ ...k, key: `inter_${o}_${C}_${n}_${B}` });
        });
      }), s;
    }), Da = g(() => ({
      datapoint: je.value,
      seriesIndex: _.value,
      series: se.value,
      bars: ye.value,
      lines: te.value,
      plots: he.value,
      config: t.value
    })), Wa = m({ months: [], shortMonths: [], days: [], shortDays: [] });
    let Ea = 0;
    dl(() => {
      const l = ++Ea, u = t.value.chart.grid.labels.xAxisLabels.datetimeFormatter;
      (async () => {
        const e = await xo(u.locale).catch(() => xo("en"));
        l === Ea && (Wa.value = e.data);
      })();
    });
    const vt = g(() => {
      const l = t.value.chart.grid.labels.xAxisLabels.datetimeFormatter, u = ju({
        useUTC: l.useUTC,
        locale: Wa.value,
        januaryAsYear: l.januaryAsYear
      });
      return (e, s) => {
        const n = t.value.chart.grid.labels.xAxisLabels.values?.[e];
        return n == null ? "" : u.formatDate(new Date(n), s);
      };
    }), Ya = g(() => (t.value.chart.grid.labels.xAxisLabels.values || []).map((u, e) => ({
      text: vt.value(e, t.value.chart.zoom.timeFormat),
      absoluteIndex: e
    }))), bu = g(() => {
      let l = "", u = je.value.map((a) => a.value).filter((a) => hl(a) && a !== null).reduce((a, n) => Math.abs(a) + Math.abs(n), 0);
      const e = le.value[_.value], s = t.value.chart.tooltip.customFormat;
      if (mo(s) && Gu(() => s({
        absoluteIndex: _.value + b.value.start,
        seriesIndex: _.value,
        datapoint: je.value,
        series: se.value,
        bars: ye.value,
        lines: te.value,
        plots: he.value,
        config: t.value,
        dateLabel: e
      })))
        return s({
          absoluteIndex: _.value + b.value.start,
          seriesIndex: _.value,
          datapoint: je.value,
          series: se.value,
          bars: ye.value,
          lines: te.value,
          plots: he.value,
          config: t.value,
          dateLabel: e
        });
      if (e && e.text && t.value.chart.tooltip.showTimeLabel) {
        const a = vt.value(_.value + b.value.start, t.value.chart.tooltip.timeFormat);
        l += `<div style="padding-bottom: 6px; margin-bottom: 4px; border-bottom: 1px solid ${t.value.chart.tooltip.borderColor}; width:100%">${t.value.chart.grid.labels.xAxisLabels.datetimeFormatter.enable && !t.value.chart.tooltip.useDefaultTimeFormat ? a : e.text}</div>`;
      }
      return je.value.forEach((a) => {
        if (hl(a.value)) {
          let n = "", o = "";
          switch (yt.value[a.type]) {
            case "bar":
              n = `<svg viewBox="0 0 40 40" height="14" width="14">${Dl.pattern ? `<rect x="0" y="0" rx="1" stroke="none" height="40" width="40" fill="${a.color}" />` : ""}<rect x="0" y="0" rx="1" stroke="none" height="40" width="40" fill="${Dl.pattern ? `url(#pattern_${P.value}_${a.slotAbsoluteIndex}` : a.color}" /></svg>`;
              break;
            case "line":
              !a.shape || !["star", "triangle", "square", "diamond", "pentagon", "hexagon"].includes(a.shape) ? o = `<circle cx="10" cy="8" r="4" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" fill="${a.color}" />` : a.shape === "triangle" ? o = `<path d="${Fe({ plot: { x: 10, y: 8 }, radius: 4, sides: 3, rotation: 0.52 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" />` : a.shape === "square" ? o = `<path d="${Fe({ plot: { x: 10, y: 8 }, radius: 4, sides: 4, rotation: 0.8 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" />` : a.shape === "diamond" ? o = `<path d="${Fe({ plot: { x: 10, y: 8 }, radius: 4, sides: 4, rotation: 0 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" />` : a.shape === "pentagon" ? o = `<path d="${Fe({ plot: { x: 10, y: 8 }, radius: 4, sides: 5, rotation: 0.95 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" />` : a.shape === "hexagon" ? o = `<path d="${Fe({ plot: { x: 10, y: 8 }, radius: 4, sides: 6, rotation: 0 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" />` : a.shape === "star" && (o = `<polygon stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" fill="${a.color}" points="${go({ plot: { x: 10, y: 8 }, radius: 4 })}" />`), n = `<svg viewBox="0 0 20 12" height="14" width="20"><rect rx="1.5" x="0" y="6.5" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="0.5" height="3" width="20" fill="${a.color}" />${o}</svg>`;
              break;
            case "plot":
              if (!a.shape || !["star", "triangle", "square", "diamond", "pentagon", "hexagon"].includes(a.shape)) {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" fill="${a.color}" /></svg>`;
                break;
              }
              if (a.shape === "star") {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><polygon stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" fill="${a.color}" points="${go({ plot: { x: 6, y: 6 }, radius: 5 })}" /></svg>`;
                break;
              }
              if (a.shape === "triangle") {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Fe({ plot: { x: 6, y: 6 }, radius: 6, sides: 3, rotation: 0.52 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>`;
                break;
              }
              if (a.shape === "square") {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Fe({ plot: { x: 6, y: 6 }, radius: 6, sides: 4, rotation: 0.8 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>`;
                break;
              }
              if (a.shape === "diamond") {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Fe({ plot: { x: 6, y: 6 }, radius: 5, sides: 4, rotation: 0 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>`;
                break;
              }
              if (a.shape === "pentagon") {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Fe({ plot: { x: 6, y: 6 }, radius: 5, sides: 5, rotation: 0.95 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>`;
                break;
              }
              if (a.shape === "hexagon") {
                n = `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Fe({ plot: { x: 6, y: 6 }, radius: 5, sides: 6, rotation: 0 }).path}" fill="${a.color}" stroke="${t.value.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>`;
                break;
              }
          }
          l += `<div style="display:flex;flex-direction:row; align-items:center;gap:3px;"><div style="width:20px">${n}</div> ${a.name}: <b>${t.value.chart.tooltip.showValue ? ie(
            a.type === "line" ? t.value.line.labels.formatter : a.type === "bar" ? t.value.bar.labels.formatter : t.value.plot.labels.formatter,
            a.value,
            Ve({
              p: a.prefix,
              v: a.value,
              s: a.suffix,
              r: t.value.chart.tooltip.roundingValue
            }),
            { datapoint: a }
          ) : ""}</b> ${t.value.chart.tooltip.showPercentage ? `(${Ve({
            v: V(Math.abs(a.value) / u * 100),
            s: "%",
            r: t.value.chart.tooltip.roundingPercentage
          })})` : ""}</div>`, t.value.chart.comments.showInTooltip && a.comments.length && a.comments.slice(b.value.start, b.value.end)[_.value] && (l += `<div class="vue-data-ui-tooltip-comment" style="background:${a.color}20; padding: 6px; margin-bottom: 6px; border-left: 1px solid ${a.color}">${a.comments.slice(b.value.start, b.value.end)[_.value]}</div>`);
        }
      }), `<div style="border-radius:4px;padding:12px;font-variant-numeric: tabular-nums;color:${t.value.chart.tooltip.color}">${l}</div>`;
    }), Xa = g(() => {
      if (De.value.length === 0) return { head: [], body: [], config: {}, columnNames: [] };
      const l = we.value.map((e) => ({
        label: e.name,
        color: e.color,
        type: e.type
      })), u = [];
      return le.value.forEach((e, s) => {
        const a = t.value.table.useDefaultTimeFormat ? [e.text] : [vt.value(s + b.value.start, t.value.table.timeFormat)];
        we.value.forEach((n) => {
          a.push(xe(n.absoluteValues[s]) ? Number(n.absoluteValues[s].toFixed(t.value.table.rounding)) : "");
        }), u.push(a);
      }), { head: l, body: u };
    }), rl = g(() => {
      const l = t.value.table.showSum;
      let u = [""].concat(we.value.map((n) => n.name));
      l && (u = u.concat(' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>'));
      let e = [];
      for (let n = 0; n < $.value; n += 1) {
        const o = we.value.map((C) => C.absoluteValues[n] ?? 0).reduce((C, A) => C + A, 0);
        e.push(
          [
            t.value.table.useDefaultTimeFormat ? le.value[n].text ?? "-" : vt.value(n + b.value.start, t.value.table.timeFormat)
          ].concat(we.value.map(
            (C) => ie(
              C.type === "line" ? t.value.line.labels.formatter : C.type === "bar" ? t.value.bar.labels.formatter : t.value.plot.labels.formatter,
              C.absoluteValues[n] ?? 0,
              Ve({
                p: C.prefix || t.value.chart.labels.prefix,
                v: C.absoluteValues[n] ?? 0,
                s: C.suffix || t.value.chart.labels.suffix,
                r: t.value.table.rounding
              })
            )
          )).concat(l ? (o ?? 0).toFixed(t.value.table.rounding) : [])
        );
      }
      const s = {
        th: {
          backgroundColor: t.value.table.th.backgroundColor,
          color: t.value.table.th.color,
          outline: t.value.table.th.outline
        },
        td: {
          backgroundColor: t.value.table.td.backgroundColor,
          color: t.value.table.td.color,
          outline: t.value.table.td.outline
        },
        breakpoint: t.value.table.responsiveBreakpoint
      }, a = [t.value.table.columnNames.period].concat(we.value.map((n) => n.name)).concat(t.value.table.columnNames.total);
      return { head: u, body: e, config: s, colNames: a };
    });
    function zl(l = null) {
      const u = [[t.value.chart.title.text], [t.value.chart.title.subtitle.text], [""]], e = ["", ...Xa.value.head.map((o) => o.label)], s = Xa.value.body, a = u.concat([e]).concat(s), n = Ku(a);
      l ? l(n) : Qu({ csvContent: n, title: t.value.chart.title.text || "vue-ui-xy" });
    }
    function Ga(l, u = null) {
      if (Eo.value) return;
      El.value = l;
      const e = we.value.map((s) => ({
        name: s.name,
        value: [null, void 0, NaN].includes(s.absoluteValues[u]) ? null : s.absoluteValues[u],
        color: s.color,
        type: s.type
      }));
      l ? (_.value = u, t.value.events.datapointEnter && t.value.events.datapointEnter({ datapoint: e, seriesIndex: u + b.value.start })) : (_.value = null, t.value.events.datapointLeave && t.value.events.datapointLeave({ datapoint: e, seriesIndex: u + b.value.start }));
    }
    function qa() {
      f.value.showTable = !f.value.showTable;
    }
    function Ua() {
      f.value.dataLabels.show = !f.value.dataLabels.show;
    }
    function Za() {
      f.value.showTooltip = !f.value.showTooltip;
    }
    function Ja(l) {
      qe.value = l, Yl.value += 1;
    }
    function Ka() {
      if (!t.value.responsiveProportionalSizing) {
        x.value.dataLabels = t.value.chart.grid.labels.fontSize, x.value.yAxis = t.value.chart.grid.labels.axis.fontSize, x.value.xAxis = t.value.chart.grid.labels.xAxisLabels.fontSize, x.value.plotLabels = t.value.chart.labels.fontSize, Ce.value.plot = t.value.plot.radius, Ce.value.line = t.value.line.radius;
        return;
      }
      x.value.dataLabels = mt({
        relator: ce.value,
        adjuster: 400,
        source: t.value.chart.grid.labels.fontSize,
        threshold: 10,
        fallback: 10
      }), x.value.yAxis = mt({
        relator: K.value,
        adjuster: 1e3,
        source: t.value.chart.grid.labels.axis.fontSize,
        threshold: 10,
        fallback: 10
      }), x.value.xAxis = mt({
        relator: K.value,
        adjuster: 1e3,
        source: t.value.chart.grid.labels.xAxisLabels.fontSize,
        threshold: 10,
        fallback: 10
      }), x.value.plotLabels = mt({
        relator: K.value,
        adjuster: 800,
        source: t.value.chart.labels.fontSize,
        threshold: 10,
        fallback: 10
      }), Ce.value.plot = mt({
        relator: K.value,
        adjuster: 800,
        source: t.value.plot.radius,
        threshold: 1,
        fallback: 1
      }), Ce.value.line = mt({
        relator: K.value,
        adjuster: 800,
        source: t.value.line.radius,
        threshold: 1,
        fallback: 1
      });
    }
    function Il() {
      bo(O.dataset) ? (Fl({
        componentName: "VueUiXy",
        type: "dataset",
        debug: St.value
      }), Gt.value = !0) : O.dataset.forEach((u, e) => {
        [null, void 0].includes(u.name) && (Fl({
          componentName: "VueUiXy",
          type: "datasetSerieAttribute",
          property: "name (string)",
          index: e,
          debug: St.value
        }), Gt.value = !0);
      }), St.value && O.dataset.forEach((u) => {
        u.series.forEach((e, s) => {
          hl(e) || console.warn(`VueUiXy has detected an unsafe value in your dataset:
-----> The serie '${u.name}' contains the value '${e}' at index ${s}.
'${e}' was converted to null to allow the chart to display.`);
        });
      }), bo(O.dataset) || (Gt.value = t.value.loading), ql.value = t.value.chart.userOptions.showOnChartHover, wl.value = t.value.chart.userOptions.keepStateOnChartLeave, Wt.value = !t.value.chart.userOptions.showOnChartHover, Ml();
      const l = 12;
      if (t.value.responsive) {
        const e = Me.value.parentNode;
        Be.value && (Be.value.unobserve(kl.value), Be.value.disconnect());
        const { height: s, width: a } = e.getBoundingClientRect();
        let n = null, o = 0;
        t.value.chart.title.show && bt.value && (n = bt.value, o = n.getBoundingClientRect().height);
        let C = null, A = 0;
        t.value.chart.zoom.show && Oe.value > 6 && Ll.value && Ie.value && Ie.value.$el && (C = Ie.value.$el, A = C.getBoundingClientRect().height);
        let N = null, S = 0;
        t.value.chart.legend.show && ut.value && (N = ut.value, S = N.getBoundingClientRect().height);
        let T = 0;
        pt.value && (T = pt.value.getBoundingClientRect().height);
        let Y = 0;
        xt.value && (Y = xt.value.getBoundingClientRect().height), ce.value = s - o - S - A - T - Y - l, K.value = a, Dt.value = `0 0 ${K.value < 0 ? 10 : K.value} ${ce.value < 0 ? 10 : ce.value}`, Ka();
        const k = new ResizeObserver((B) => {
          for (const z of B)
            t.value.chart.title.show && bt.value ? o = bt.value.getBoundingClientRect().height : o = 0, Ie.value && Ie.value.$el ? A = Ie.value.$el.getBoundingClientRect().height : A = 0, ut.value ? S = ut.value.getBoundingClientRect().height : S = 0, pt.value ? T = pt.value.getBoundingClientRect().height : T = 0, xt.value ? Y = xt.value.getBoundingClientRect().height : Y = 0, requestAnimationFrame(() => {
              ce.value = z.contentRect.height - o - S - A - T - Y - l, K.value = z.contentBoxSize[0].inlineSize ?? z.contentRect.width, Dt.value = `0 0 ${K.value < 0 ? 10 : K.value} ${ce.value < 0 ? 10 : ce.value}`, Ka(), wt(), nt();
            });
        });
        Be.value = k, kl.value = e, k.observe(e);
      } else
        ce.value = t.value.chart.height, K.value = t.value.chart.width, x.value.dataLabels = t.value.chart.grid.labels.fontSize, x.value.yAxis = t.value.chart.grid.labels.axis.fontSize, x.value.xAxis = t.value.chart.grid.labels.xAxisLabels.fontSize, x.value.plotLabels = t.value.chart.labels.fontSize, Ce.value.plot = t.value.plot.radius, Ce.value.line = t.value.line.radius, Dt.value = `0 0 ${K.value} ${ce.value}`;
      wt(), nt();
    }
    function Qa(l) {
      No.value = {
        x: l.clientX,
        y: l.clientY
      };
    }
    Pt(() => {
      Il(), el(), document.addEventListener("mousemove", Qa, { passive: !0 }), document.addEventListener("scroll", fa, { passive: !0 });
    }), Nl(() => {
      document.removeEventListener("scroll", fa, { passive: !0 }), document.removeEventListener("mousemove", Qa, { passive: !0 }), Be.value && (Be.value.unobserve(kl.value), Be.value.disconnect(), Be.value = null);
    }), ns({
      timeLabelsEls: Ht,
      timeLabels: le,
      slicer: b,
      configRef: t,
      rotationPath: ["chart", "grid", "labels", "xAxisLabels", "rotation"],
      autoRotatePath: ["chart", "grid", "labels", "xAxisLabels", "autoRotate", "enable"],
      isAutoSize: Do,
      height: ce,
      width: K,
      rotation: t.value.chart.grid.labels.xAxisLabels.autoRotate.angle
    });
    const il = m(!1), ja = m(null), eo = m(200), pu = g(
      () => _.value ?? F.value ?? 0
    );
    function xu() {
      const l = Math.ceil(eo.value || 200);
      return Math.min(Math.max(l, 1), 200);
    }
    function yu() {
      const l = xu(), u = 200, e = Math.max(1, $.value), s = c.value.width / e, n = c.value?.left + pu.value * s + s / 2 - l / 2 - (u - l) / 2, o = c.value?.left - (u - l) / 2, C = c.value?.right - (u + l) / 2, A = Math.max(o, Math.min(n, C));
      return V(A);
    }
    Pt(() => {
      let l = null, u = null;
      const e = (n) => {
        cancelAnimationFrame(u), u = requestAnimationFrame(() => {
          eo.value = Math.min(Math.max(Math.ceil(n || 0), 1), 200);
        });
      }, s = new ResizeObserver((n) => {
        let o = n.find((C) => C.target === l) || n[0];
        o && e(o.contentRect.width || 200);
      }), a = dl((n) => {
        const o = ja.value;
        l && l !== o && (s.unobserve(l), l = null), o && o !== l && (Ft(() => {
          o.offsetParent !== null && e(o.offsetWidth || o.getBoundingClientRect().width || 200);
        }), s.observe(o), l = o), n(() => {
          l && (s.unobserve(l), l = null);
        });
      });
      Nl(() => {
        try {
          l && s.unobserve(l), s.disconnect(), a();
        } catch {
        }
      });
    });
    const ku = g(() => {
      if ([null, void 0].includes(_.value) && [null, void 0].includes(F.value)) return "";
      const l = (_.value != null ? _.value : 0) || (F.value != null ? F.value : 0), u = t.value.chart.timeTag.customFormat;
      if (il.value = !1, mo(u))
        try {
          const e = u({
            absoluteIndex: l + b.value.start,
            seriesIndex: l,
            datapoint: je.value,
            bars: ye.value,
            lines: te.value,
            plots: he.value,
            config: t.value
          });
          if (typeof e == "string")
            return il.value = !0, e;
        } catch {
          console.warn("Custom format cannot be applied on timeTag."), il.value = !1;
        }
      if (!il.value)
        return [null, void 0].includes(le.value[l]) ? "" : t.value.chart.grid.labels.xAxisLabels.datetimeFormatter.enable && !t.value.chart.timeTag.useDefaultFormat ? vt.value(l + b.value.start, t.value.chart.timeTag.timeFormat) : le.value[l].text;
    });
    ht(() => O.dataset, (l) => {
      Array.isArray(l) && l.length > 0 && (Gt.value = !1), Oe.value = Math.max(...pe.value.map((u) => He(u.series).length)), b.value = {
        start: 0,
        end: Oe.value
      }, Oo.value += 1, _l.value += 1, wt(), nt(), rt();
    }, { deep: !0 }), ht(() => O.config, (l) => {
      ue.value || (t.value = $l()), Il(), Gl.value += 1, Xl.value += 1, Ml(), wt(), nt(), rt();
    }, { deep: !0 });
    const Ol = m(!1);
    function to() {
      const l = Me.value?.parentNode;
      if (!l) {
        Ol.value = !1;
        return;
      }
      const u = l.getBoundingClientRect();
      Ol.value = u.width > 2 && u.height > 2;
    }
    Pt(() => {
      to();
      const l = new ResizeObserver(() => {
        to(), Ol.value && (Il(), rt(), el());
      });
      Me.value?.parentNode && l.observe(Me.value.parentNode);
    }), ht(t, () => {
      Ml();
    }, { immediate: !0 });
    const zt = g(() => {
      const l = t.value.table.useDialog && !t.value.showTable, u = f.value.showTable;
      return {
        component: l ? Co : ds,
        title: `${t.value.chart.title.text}${t.value.chart.title.subtitle.text ? `: ${t.value.chart.title.subtitle.text}` : ""}`,
        props: l ? {
          backgroundColor: t.value.table.th.backgroundColor,
          color: t.value.table.th.color,
          headerColor: t.value.table.th.color,
          headerBg: t.value.table.th.backgroundColor,
          isFullscreen: qe.value,
          fullscreenParent: Me.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8)
        } : {
          hideDetails: !0,
          config: {
            open: u,
            maxHeight: 1e4,
            body: {
              backgroundColor: t.value.chart.backgroundColor,
              color: t.value.chart.color
            },
            head: {
              backgroundColor: t.value.chart.backgroundColor,
              color: t.value.chart.color
            }
          }
        }
      };
    });
    ht(() => f.value.showTable, (l) => {
      t.value.showTable || (l && t.value.table.useDialog && st.value ? st.value.open() : st.value && st.value.close && st.value.close());
    });
    function lo() {
      f.value.showTable = !1, yl.value && yl.value.setTableIconState(!1);
    }
    const _u = g(() => se.value.map((l) => ({
      shape: l.type === "bar" ? "square" : l.shape ?? "circle",
      color: l.color,
      name: l.name
    }))), wu = g(() => t.value.chart.backgroundColor), Su = g(() => t.value.chart.legend), $u = g(() => t.value.chart.title), { exportSvg: Lu, getSvg: Mu } = os({
      svg: ve,
      title: $u,
      legend: Su,
      legendItems: _u,
      backgroundColor: wu
    });
    async function ao({ isCb: l }) {
      Xt.value = !0, await Ft();
      try {
        if (l) {
          const { blob: u, url: e, text: s, dataUrl: a } = await Mu();
          await Promise.resolve(t.value.chart.userOptions.callbacks.svg({ blob: u, url: e, text: s, dataUrl: a }));
        } else
          await Promise.resolve(Lu());
      } finally {
        Xt.value = !1;
      }
    }
    function Cu(l) {
      if (l?.stage === "start") {
        Yt.value = !0;
        return;
      }
      if (l?.stage === "end") {
        Yt.value = !1;
        return;
      }
      na();
    }
    async function oo() {
      if (Ge("copyAlt", {
        config: {
          ...t.value,
          formattedDates: le.value
        },
        dataset: {
          lines: te.value,
          bars: ye.value,
          plots: he.value
        }
      }), !t.value.chart.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(t.value.chart.userOptions.callbacks.altCopy({
        config: {
          ...t.value,
          formattedDates: le.value
        },
        dataset: {
          lines: te.value,
          bars: ye.value,
          plots: he.value
        }
      }));
    }
    return yo({
      getData: au,
      getImage: ou,
      generatePdf: sa,
      generateImage: na,
      generateSvg: ao,
      generateCsv: zl,
      hideSeries: nu,
      showSeries: su,
      toggleStack: ya,
      toggleTable: qa,
      toggleLabels: Ua,
      toggleTooltip: Za,
      toggleAnnotator: Al,
      toggleFullscreen: Ja,
      copyAlt: oo
    }), (l, u) => (r(), i("div", {
      id: `vue-ui-xy_${P.value}`,
      class: tt(`vue-data-ui-component vue-ui-xy ${qe.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${t.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      ref_key: "chart",
      ref: Me,
      style: w(`background:${t.value.chart.backgroundColor}; color:${t.value.chart.color};width:100%;font-family:${t.value.chart.fontFamily};${t.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: u[6] || (u[6] = () => ga(!0)),
      onMouseleave: u[7] || (u[7] = () => ga(!1)),
      onClick: Ia
    }, [
      t.value.chart.userOptions.buttons.annotator && ve.value ? (r(), $e(v(Mo), {
        key: 0,
        svgRef: ve.value,
        backgroundColor: t.value.chart.backgroundColor,
        color: t.value.chart.color,
        active: kt.value,
        isCursorPointer: Ze.value,
        onClose: Al
      }, {
        "annotator-action-close": X(() => [
          R(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": X(({ color: e }) => [
          R(l.$slots, "annotator-action-color", ge(be({ color: e })), void 0, !0)
        ]),
        "annotator-action-draw": X(({ mode: e }) => [
          R(l.$slots, "annotator-action-draw", ge(be({ mode: e })), void 0, !0)
        ]),
        "annotator-action-undo": X(({ disabled: e }) => [
          R(l.$slots, "annotator-action-undo", ge(be({ disabled: e })), void 0, !0)
        ]),
        "annotator-action-redo": X(({ disabled: e }) => [
          R(l.$slots, "annotator-action-redo", ge(be({ disabled: e })), void 0, !0)
        ]),
        "annotator-action-delete": X(({ disabled: e }) => [
          R(l.$slots, "annotator-action-delete", ge(be({ disabled: e })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : h("", !0),
      iu.value ? (r(), i("div", {
        key: 1,
        ref_key: "noTitle",
        ref: xt,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%; background:transparent"
      }, null, 512)) : h("", !0),
      t.value.chart.title.show ? (r(), i("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: bt,
        class: "vue-ui-xy-title",
        style: w(`font-family:${t.value.chart.fontFamily}`)
      }, [
        (r(), $e(is, {
          key: `title_${Gl.value}`,
          config: {
            title: {
              cy: "xy-div-title",
              ...t.value.chart.title
            },
            subtitle: {
              cy: "xy-div-subtitle",
              ...t.value.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 4)) : h("", !0),
      p("div", {
        id: `legend-top-${P.value}`
      }, null, 8, gs),
      t.value.chart.userOptions.show && (wl.value || Wt.value) ? (r(), $e(v(So), {
        ref_key: "userOptionsRef",
        ref: yl,
        key: `user_options_${Yl.value}`,
        backgroundColor: t.value.chart.backgroundColor,
        color: t.value.chart.color,
        isPrinting: v(qt),
        isImaging: v(Cl),
        uid: P.value,
        hasTooltip: t.value.chart.userOptions.buttons.tooltip && t.value.chart.tooltip.show,
        hasPdf: t.value.chart.userOptions.buttons.pdf,
        hasXls: t.value.chart.userOptions.buttons.csv,
        hasImg: t.value.chart.userOptions.buttons.img,
        hasSvg: t.value.chart.userOptions.buttons.svg,
        hasLabel: t.value.chart.userOptions.buttons.labels,
        hasTable: t.value.chart.userOptions.buttons.table,
        hasStack: gl.dataset.length > 1 && t.value.chart.userOptions.buttons.stack,
        hasFullscreen: t.value.chart.userOptions.buttons.fullscreen,
        hasAltCopy: t.value.chart.userOptions.buttons.altCopy,
        isStacked: f.value.isStacked,
        isFullscreen: qe.value,
        chartElement: l.$refs.chart,
        position: t.value.chart.userOptions.position,
        isTooltip: f.value.showTooltip,
        titles: { ...t.value.chart.userOptions.buttonTitles },
        hasAnnotator: t.value.chart.userOptions.buttons.annotator,
        isAnnotation: kt.value,
        callbacks: t.value.chart.userOptions.callbacks,
        tableDialog: t.value.table.useDialog,
        printScale: t.value.chart.userOptions.print.scale,
        isCursorPointer: Ze.value,
        onToggleFullscreen: Ja,
        onGeneratePdf: v(sa),
        onGenerateCsv: zl,
        onGenerateImage: Cu,
        onGenerateSvg: ao,
        onToggleTable: qa,
        onToggleLabels: Ua,
        onToggleStack: ya,
        onToggleTooltip: Za,
        onToggleAnnotator: Al,
        onCopyAlt: oo,
        style: w({
          visibility: wl.value ? Wt.value ? "visible" : "hidden" : "visible"
        })
      }, uo({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: X(({ isOpen: e, color: s }) => [
            R(l.$slots, "menuIcon", ge(be({ isOpen: e, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: X(() => [
            R(l.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: X(() => [
            R(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionCsv ? {
          name: "optionCsv",
          fn: X(() => [
            R(l.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: X(() => [
            R(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: X(() => [
            R(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionTable ? {
          name: "optionTable",
          fn: X(() => [
            R(l.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionLabels ? {
          name: "optionLabels",
          fn: X(() => [
            R(l.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        l.$slots.optionStack ? {
          name: "optionStack",
          fn: X(({ isStack: e }) => [
            R(l.$slots, "optionStack", ge(be({ isStack: e })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: X(({ toggleFullscreen: e, isFullscreen: s }) => [
            R(l.$slots, "optionFullscreen", ge(be({ toggleFullscreen: e, isFullscreen: s })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: X(({ toggleAnnotator: e, isAnnotator: s }) => [
            R(l.$slots, "optionAnnotator", ge(be({ toggleAnnotator: e, isAnnotator: s })), void 0, !0)
          ]),
          key: "10"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: X(({ altCopy: e }) => [
            R(l.$slots, "optionAltCopy", ge(be({ altCopy: e })), void 0, !0)
          ]),
          key: "11"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasLabel", "hasTable", "hasStack", "hasFullscreen", "hasAltCopy", "isStacked", "isFullscreen", "chartElement", "position", "isTooltip", "titles", "hasAnnotator", "isAnnotation", "callbacks", "tableDialog", "printScale", "isCursorPointer", "onGeneratePdf", "style"])) : h("", !0),
      (r(), i("svg", {
        ref_key: "svgRef",
        ref: ve,
        xmlns: "http://www.w3.org/2000/svg",
        class: tt([{ "vue-data-ui-fullscreen--on": qe.value, "vue-data-ui-fulscreen--off": !qe.value }, "vue-ui-xy-svg vue-data-ui-svg"]),
        width: "100%",
        viewBox: Dt.value,
        style: w({
          background: "transparent",
          color: t.value.chart.color,
          fontFamily: t.value.chart.fontFamily
        }),
        "aria-label": ru.value,
        role: "img",
        "aria-live": "polite",
        preserveAspectRatio: "xMidYMid",
        onMousemove: lu,
        onMouseleave: al,
        onClick: Ia
      }, [
        p("g", {
          ref_key: "G",
          ref: Io,
          class: "vue-data-ui-g"
        }, [
          ft(v(Lo)),
          l.$slots["chart-background"] ? (r(), i("foreignObject", {
            key: 0,
            x: c.value?.left + U.value < 0 ? 0 : c.value?.left + U.value,
            y: c.value?.top,
            width: c.value.width - (t.value.chart.grid.position === "middle" ? 0 : c.value.width / $.value) < 0 ? 0 : c.value.width - (t.value.chart.grid.position === "middle" ? 0 : c.value.width / $.value),
            height: c.value.height < 0 ? 0 : c.value.height,
            style: {
              pointerEvents: "none"
            }
          }, [
            R(l.$slots, "chart-background", {}, void 0, !0)
          ], 8, ps)) : h("", !0),
          $.value > 0 ? (r(), i("g", xs, [
            p("g", ys, [
              t.value.chart.grid.labels.xAxis.showBaseline ? (r(), i("line", {
                key: 0,
                stroke: t.value.chart.grid.stroke,
                "stroke-width": "1",
                x1: c.value?.left + U.value,
                x2: c.value?.right - U.value,
                y1: v(E)(c.value?.bottom),
                y2: v(E)(c.value?.bottom),
                "stroke-linecap": "round",
                style: { animation: "none !important" }
              }, null, 8, ks)) : h("", !0),
              f.value.useIndividualScale ? t.value.chart.grid.showHorizontalLines ? (r(!0), i(d, { key: 2 }, y(Tt.value, (e) => (r(), i("g", null, [
                e.id === I.value && e.yLabels.length ? (r(!0), i(d, { key: 0 }, y(e.yLabels, (s) => (r(), i("line", {
                  x1: c.value?.left + U.value,
                  x2: c.value?.right - U.value,
                  y1: v(E)(s.y),
                  y2: v(E)(s.y),
                  stroke: e.color,
                  "stroke-width": 0.5,
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, $s))), 256)) : e.yLabels.length ? (r(!0), i(d, { key: 1 }, y(e.yLabels, (s) => (r(), i("line", {
                  x1: c.value?.left + U.value,
                  x2: c.value?.right - U.value,
                  y1: v(E)(s.y),
                  y2: v(E)(s.y),
                  stroke: t.value.chart.grid.stroke,
                  "stroke-width": 0.5,
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, Ls))), 256)) : h("", !0)
              ]))), 256)) : h("", !0) : (r(), i(d, { key: 1 }, [
                t.value.chart.grid.labels.yAxis.showBaseline ? (r(), i("line", {
                  key: 0,
                  stroke: t.value.chart.grid.stroke,
                  "stroke-width": "1",
                  x1: c.value?.left + U.value,
                  x2: c.value?.left + U.value,
                  y1: v(E)(c.value?.top),
                  y2: v(E)(c.value?.bottom),
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, _s)) : h("", !0),
                t.value.chart.grid.showHorizontalLines ? (r(), i("g", ws, [
                  (r(!0), i(d, null, y(Va.value, (e) => (r(), i("line", {
                    x1: c.value?.left + U.value,
                    x2: c.value?.right,
                    y1: v(E)(e.y),
                    y2: v(E)(e.y),
                    stroke: t.value.chart.grid.stroke,
                    "stroke-width": 0.5,
                    "stroke-linecap": "round",
                    style: { animation: "none !important" }
                  }, null, 8, Ss))), 256))
                ])) : h("", !0)
              ], 64)),
              t.value.chart.grid.showVerticalLines ? (r(), i("g", Ms, [
                p("path", {
                  d: Go.value,
                  "stroke-width": 0.5,
                  stroke: t.value.chart.grid.stroke,
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, Cs)
              ])) : h("", !0),
              t.value.chart.grid.labels.xAxisLabels.show ? (r(), i("g", As, [
                p("path", {
                  d: qo.value,
                  stroke: t.value.chart.grid.stroke,
                  "stroke-width": 1,
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, Ts)
              ])) : h("", !0)
            ]),
            (r(!0), i(d, null, y(ye.value, (e, s) => (r(), i("defs", {
              key: `def_rect_${s}`
            }, [
              l.$slots["bar-gradient"] ? R(l.$slots, "bar-gradient", lt({
                key: 0,
                ref_for: !0
              }, { series: e, positiveId: `rectGradient_pos_${s}_${P.value}`, negativeId: `rectGradient_neg_${s}_${P.value}` }), void 0, !0) : (r(), i(d, { key: 1 }, [
                p("linearGradient", {
                  id: `rectGradient_pos_${s}_${P.value}`,
                  x2: "0%",
                  y2: "100%"
                }, [
                  p("stop", {
                    offset: "0%",
                    "stop-color": e.color
                  }, null, 8, Is),
                  p("stop", {
                    offset: "62%",
                    "stop-color": `${v(ot)(e.color, 0.02)}`
                  }, null, 8, Os),
                  p("stop", {
                    offset: "100%",
                    "stop-color": `${v(ot)(e.color, 0.05)}`
                  }, null, 8, Ns)
                ], 8, zs),
                p("linearGradient", {
                  id: `rectGradient_neg_${s}_${P.value}`,
                  x2: "0%",
                  y2: "100%"
                }, [
                  p("stop", {
                    offset: "0%",
                    "stop-color": `${v(ot)(e.color, 0.05)}`
                  }, null, 8, Fs),
                  p("stop", {
                    offset: "38%",
                    "stop-color": `${v(ot)(e.color, 0.02)}`
                  }, null, 8, Vs),
                  p("stop", {
                    offset: "100%",
                    "stop-color": e.color
                  }, null, 8, Rs)
                ], 8, Ps)
              ], 64))
            ]))), 128)),
            (r(!0), i(d, null, y(he.value, (e, s) => (r(), i("defs", {
              key: `def_plot_${s}`
            }, [
              p("radialGradient", {
                id: `plotGradient_${s}_${P.value}`,
                cx: "50%",
                cy: "50%",
                r: "50%",
                fx: "50%",
                fy: "50%"
              }, [
                p("stop", {
                  offset: "0%",
                  "stop-color": `${v(ot)(e.color, 0.05)}`
                }, null, 8, Hs),
                p("stop", {
                  offset: "100%",
                  "stop-color": e.color
                }, null, 8, Ds)
              ], 8, Bs)
            ]))), 128)),
            (r(!0), i(d, null, y(te.value, (e, s) => (r(), i(d, {
              key: `def_line_${e.id}`
            }, [
              p("defs", null, [
                p("radialGradient", {
                  id: `lineGradient_${s}_${P.value}`,
                  cx: "50%",
                  cy: "50%",
                  r: "50%",
                  fx: "50%",
                  fy: "50%"
                }, [
                  p("stop", {
                    offset: "0%",
                    "stop-color": `${v(ot)(e.color, 0.05)}`
                  }, null, 8, Es),
                  p("stop", {
                    offset: "100%",
                    "stop-color": e.color
                  }, null, 8, Ys)
                ], 8, Ws),
                l.$slots["area-gradient"] ? R(l.$slots, "area-gradient", lt({
                  key: 0,
                  ref_for: !0
                }, { series: e, id: `areaGradient_${s}_${P.value}` }), void 0, !0) : (r(), i("linearGradient", {
                  key: 1,
                  id: `areaGradient_${s}_${P.value}`,
                  x1: "0%",
                  x2: "100%",
                  y1: "0%",
                  y2: "0%"
                }, [
                  p("stop", {
                    offset: "0%",
                    "stop-color": `${v(ze)(v(ot)(e.color, 0.03), t.value.line.area.opacity)}`
                  }, null, 8, Gs),
                  p("stop", {
                    offset: "100%",
                    "stop-color": `${v(ze)(e.color, t.value.line.area.opacity)}`
                  }, null, 8, qs)
                ], 8, Xs))
              ]),
              e.temperatureColors ? (r(), i("defs", Us, [
                p("linearGradient", {
                  id: `temperature_grad_line_${s}_${P.value}`,
                  gradientTransform: "rotate(90)"
                }, [
                  (r(!0), i(d, null, y(e.temperatureColors, (a, n) => (r(), i("stop", {
                    key: `temperature_grad_stop_${s}_${n}_${P.value}`,
                    "stop-color": a,
                    offset: v(qu)(n, e.temperatureColors.length)
                  }, null, 8, Js))), 128))
                ], 8, Zs)
              ])) : h("", !0)
            ], 64))), 128)),
            (r(!0), i(d, null, y(cu.value, (e) => (r(), i("g", null, [
              e.show ? (r(), i(d, { key: 0 }, [
                (r(!0), i(d, null, y(e.span, (s, a) => (r(), i("g", null, [
                  p("rect", {
                    style: w({
                      transition: "none",
                      opacity: e.from + a >= b.value.start && e.from + a <= b.value.end - 1 ? 1 : 0
                    }),
                    x: c.value?.left + c.value.width / $.value * (e.from + a - b.value.start),
                    y: c.value?.top,
                    height: c.value.height < 0 ? 10 : c.value.height,
                    width: c.value.width / $.value < 0 ? 1e-5 : c.value.width / $.value,
                    fill: v(ze)(e.color, e.opacity)
                  }, null, 12, Ks)
                ]))), 256)),
                (r(!0), i(d, null, y(e.span, (s, a) => (r(), i("g", null, [
                  e.caption.text && a === 0 ? (r(), i("foreignObject", {
                    key: 0,
                    x: c.value?.left + c.value.width / $.value * (e.from + a - b.value.start) - (e.caption.width === "auto" ? 0 : e.caption.width / 2 - c.value.width / $.value * e.span / 2),
                    y: c.value?.top + e.caption.offsetY,
                    style: w({
                      overflow: "visible",
                      opacity: e.to >= b.value.start && e.from < b.value.end ? 1 : 0
                    }),
                    height: "1",
                    width: e.caption.width === "auto" ? c.value.width / $.value * e.span : e.caption.width
                  }, [
                    p("div", {
                      style: w(`padding:${e.caption.padding}px;text-align:${e.caption.textAlign};font-size:${e.caption.fontSize}px;color:${e.caption.color};font-weight:${e.caption.bold ? "bold" : "normal"}`)
                    }, oe(e.caption.text), 5)
                  ], 12, Qs)) : h("", !0)
                ]))), 256))
              ], 64)) : h("", !0)
            ]))), 256)),
            ma.value ? (r(), i("g", js, [
              (r(!0), i(d, null, y($.value, (e, s) => (r(), i("g", {
                key: `tooltip_trap_highlighter_${s}`
              }, [
                p("rect", {
                  x: c.value?.left + c.value.width / $.value * s,
                  y: c.value?.top,
                  height: c.value.height < 0 ? 10 : c.value.height,
                  width: c.value.width / $.value < 0 ? 1e-5 : c.value.width / $.value,
                  fill: [F.value, _.value, Ue.value].includes(s) ? v(ze)(t.value.chart.highlighter.color, t.value.chart.highlighter.opacity) : "transparent",
                  style: { transition: "none !important", animation: "none !important" }
                }, null, 8, en)
              ]))), 128))
            ])) : h("", !0),
            ye.value.length ? (r(!0), i(d, { key: 1 }, y(ye.value, (e, s) => (r(), i("g", {
              key: `serie_bar_${e.id}`,
              class: tt(`serie_bar_${s}`),
              style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
            }, [
              (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i("g", {
                key: `bar_plot_${s}_${n}`
              }, [
                xe(a.value) ? (r(), i("rect", {
                  key: 0,
                  x: ll(a) + nl.value / 2,
                  y: f.value.useIndividualScale ? Ta(a) : Aa(a),
                  height: f.value.useIndividualScale ? Math.abs(Ca(a)) : Math.abs(Ma(a)),
                  width: sl.value - nl.value,
                  rx: t.value.bar.borderRadius,
                  fill: t.value.bar.useGradient ? a.value >= 0 ? `url(#rectGradient_pos_${s}_${P.value})` : `url(#rectGradient_neg_${s}_${P.value})` : e.color,
                  stroke: t.value.bar.border.useSerieColor ? e.color : t.value.bar.border.stroke,
                  "stroke-width": t.value.bar.border.strokeWidth,
                  style: w({
                    transition: v(ue) || !t.value.bar.showTransition ? void 0 : `all ${t.value.bar.transitionDurationMs}ms ease-in-out`
                  })
                }, null, 12, tn)) : h("", !0),
                xe(a.value) && l.$slots.pattern ? (r(), i("rect", {
                  key: 1,
                  x: ll(a) - nl.value / 2,
                  y: f.value.useIndividualScale ? Ta(a) : Aa(a),
                  height: f.value.useIndividualScale ? Math.abs(Ca(a)) : Math.abs(Ma(a)),
                  width: sl.value - nl.value,
                  rx: t.value.bar.borderRadius,
                  fill: `url(#pattern_${P.value}_${e.slotAbsoluteIndex})`,
                  stroke: t.value.bar.border.useSerieColor ? e.color : t.value.bar.border.stroke,
                  "stroke-width": t.value.bar.border.strokeWidth,
                  style: w({ transition: v(ue) || !t.value.bar.showTransition ? void 0 : `all ${t.value.bar.transitionDurationMs}ms ease-in-out` })
                }, null, 12, ln)) : h("", !0),
                a.comment && t.value.chart.comments.show ? (r(), i("foreignObject", {
                  key: 2,
                  style: { overflow: "visible" },
                  height: "12",
                  width: sl.value + t.value.chart.comments.width,
                  x: ll(a) - t.value.chart.comments.width / 2 + t.value.chart.comments.offsetX,
                  y: v(V)(a.y) + t.value.chart.comments.offsetY + 6
                }, [
                  R(l.$slots, "plot-comment", {
                    plot: { ...a, color: e.color, seriesIndex: s, datapointIndex: n }
                  }, void 0, !0)
                ], 8, an)) : h("", !0)
              ]))), 128))
            ], 6))), 128)) : h("", !0),
            !f.value.useIndividualScale && t.value.chart.grid.labels.zeroLine.show ? (r(), i("line", {
              key: 2,
              stroke: t.value.chart.grid.stroke,
              "stroke-width": "1",
              x1: c.value?.left + U.value,
              x2: c.value?.right,
              y1: v(E)(de.value),
              y2: v(E)(de.value),
              "stroke-linecap": "round",
              style: { animation: "none !important" }
            }, null, 8, on)) : h("", !0),
            t.value.chart.highlighter.useLine && (![null, void 0].includes(_.value) || ![null, void 0].includes(F.value)) ? (r(), i("g", un, [
              p("line", {
                x1: c.value?.left + c.value.width / $.value * ((_.value !== null ? _.value : 0) || (F.value !== null ? F.value : 0)) + c.value.width / $.value / 2,
                x2: c.value?.left + c.value.width / $.value * ((_.value !== null ? _.value : 0) || (F.value !== null ? F.value : 0)) + c.value.width / $.value / 2,
                y1: v(E)(c.value?.top),
                y2: v(E)(c.value?.bottom),
                stroke: t.value.chart.highlighter.color,
                "stroke-width": t.value.chart.highlighter.lineWidth,
                "stroke-dasharray": t.value.chart.highlighter.lineDasharray,
                "stroke-linecap": "round",
                style: { transition: "none !important", animation: "none !important", "pointer-events": "none" }
              }, null, 8, sn)
            ])) : h("", !0),
            t.value.chart.grid.frame.show ? (r(), i("rect", {
              key: 4,
              style: { pointerEvents: "none", transition: "none", animation: "none !important" },
              x: c.value?.left + U.value < 0 ? 0 : c.value?.left + U.value,
              y: c.value?.top,
              width: c.value.width - (t.value.chart.grid.position === "middle" ? 0 : c.value.width / $.value) < 0 ? 0 : c.value.width - (t.value.chart.grid.position === "middle" ? 0 : c.value.width / $.value),
              height: c.value.height < 0 ? 0 : c.value.height,
              fill: "transparent",
              stroke: t.value.chart.grid.frame.stroke,
              "stroke-width": t.value.chart.grid.frame.strokeWidth,
              "stroke-linecap": t.value.chart.grid.frame.strokeLinecap,
              "stroke-linejoin": t.value.chart.grid.frame.strokeLinejoin,
              "stroke-dasharray": t.value.chart.grid.frame.strokeDasharray
            }, null, 8, nn)) : h("", !0),
            t.value.chart.grid.labels.show ? (r(), i("g", {
              key: 5,
              ref_key: "scaleLabels",
              ref: xl
            }, [
              f.value.useIndividualScale ? (r(), i(d, { key: 0 }, [
                (r(!0), i(d, null, y(Tt.value, (e) => (r(), i("g", null, [
                  p("line", {
                    x1: e.x + U.value - c.value.individualOffsetX,
                    x2: e.x + U.value - c.value.individualOffsetX,
                    y1: f.value.isStacked ? v(E)(c.value?.bottom - e.yOffset - e.individualHeight) : v(E)(c.value?.top),
                    y2: f.value.isStacked ? v(E)(c.value?.bottom - e.yOffset) : v(E)(c.value?.bottom),
                    stroke: e.color,
                    "stroke-width": t.value.chart.grid.stroke,
                    "stroke-linecap": "round",
                    style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.3 : 1};transition:opacity 0.2s ease-in-out; animation: none !important`)
                  }, null, 12, rn)
                ]))), 256)),
                (r(!0), i(d, null, y(Tt.value, (e) => (r(), i("g", {
                  style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.3 : 1};transition:opacity 0.2s ease-in-out`)
                }, [
                  p("text", {
                    fill: e.color,
                    "font-size": x.value.dataLabels * 0.8,
                    "text-anchor": "middle",
                    transform: `translate(${e.x - x.value.dataLabels * 0.8 / 2 + U.value}, ${f.value.isStacked ? c.value?.bottom - e.yOffset - e.individualHeight / 2 : c.value?.top + c.value.height / 2}) rotate(-90)`
                  }, oe(e.name) + " " + oe(e.scaleLabel && e.unique && e.scaleLabel !== e.id ? `-
                                    ${e.scaleLabel}` : ""), 9, cn),
                  (r(!0), i(d, null, y(e.yLabels, (s, a) => (r(), i(d, null, [
                    t.value.chart.grid.labels.yAxis.showCrosshairs ? (r(), i("line", {
                      key: 0,
                      x1: e.x + 3 + U.value - t.value.chart.grid.labels.yAxis.crosshairSize - c.value.individualOffsetX,
                      x2: e.x + U.value - c.value.individualOffsetX,
                      y1: v(E)(s.y),
                      y2: v(E)(s.y),
                      stroke: e.color,
                      "stroke-width": 1,
                      "stroke-linecap": "round",
                      style: { animation: "none !important" }
                    }, null, 8, vn)) : h("", !0)
                  ], 64))), 256)),
                  (r(!0), i(d, null, y(e.yLabels, (s, a) => (r(), i("text", {
                    x: e.x - 5 + U.value - c.value.individualOffsetX,
                    y: v(E)(s.y) + x.value.dataLabels / 3,
                    "font-size": x.value.dataLabels,
                    "text-anchor": "end",
                    fill: e.color
                  }, oe(v(ie)(
                    t.value.chart.grid.labels.yAxis.formatter,
                    s.value,
                    v(Ve)({
                      p: s.prefix,
                      v: s.value,
                      s: s.suffix,
                      r: t.value.chart.grid.labels.yAxis.rounding
                    }),
                    { datapoint: s.datapoint, seriesIndex: a }
                  )), 9, dn))), 256))
                ], 4))), 256))
              ], 64)) : (r(!0), i(d, { key: 1 }, y(Va.value, (e, s) => (r(), i("g", {
                key: `yLabel_${s}`
              }, [
                xe(e) && e.value >= Te.value.min && e.value <= Te.value.max && t.value.chart.grid.labels.yAxis.showCrosshairs ? (r(), i("line", {
                  key: 0,
                  x1: c.value?.left + U.value,
                  x2: c.value?.left + U.value - t.value.chart.grid.labels.yAxis.crosshairSize,
                  y1: v(E)(e.y),
                  y2: v(E)(e.y),
                  stroke: t.value.chart.grid.stroke,
                  "stroke-width": "1",
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, hn)) : h("", !0),
                e.value >= Te.value.min && e.value <= Te.value.max ? (r(), i("text", {
                  key: 1,
                  x: c.value.scaleLabelX - t.value.chart.grid.labels.yAxis.crosshairSize,
                  y: v(V)(e.y + x.value.dataLabels / 3),
                  "font-size": x.value.dataLabels,
                  "text-anchor": "end",
                  fill: t.value.chart.grid.labels.color
                }, oe(xe(e.value) ? v(ie)(
                  t.value.chart.grid.labels.yAxis.formatter,
                  e.value,
                  v(Ve)({
                    p: e.prefix,
                    v: e.value,
                    s: e.suffix,
                    r: t.value.chart.grid.labels.yAxis.rounding
                  })
                ) : ""), 9, fn)) : h("", !0)
              ]))), 128))
            ], 512)) : h("", !0),
            (r(!0), i(d, null, y(he.value, (e, s) => (r(), i("g", {
              key: `serie_plot_${e.id}`,
              class: tt(`serie_plot_${s}`),
              style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
            }, [
              (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i("g", {
                key: `circle_plot_${s}_${n}`
              }, [
                a && xe(a.value) ? (r(), $e(ml, {
                  key: 0,
                  shape: ["triangle", "square", "diamond", "pentagon", "hexagon", "star"].includes(e.shape) ? e.shape : "circle",
                  color: t.value.plot.useGradient ? `url(#plotGradient_${s}_${P.value})` : t.value.plot.dot.useSerieColor ? e.color : t.value.plot.dot.fill,
                  plot: { x: v(V)(a.x), y: v(V)(a.y) },
                  radius: _.value !== null && _.value === n || F.value !== null && F.value === n ? (Ce.value.plot || 6) * 1.5 : Ce.value.plot || 6,
                  stroke: t.value.plot.dot.useSerieColor ? t.value.chart.backgroundColor : e.color,
                  strokeWidth: t.value.plot.dot.strokeWidth,
                  transition: v(ue) || !t.value.plot.showTransition || _.value !== null && _.value === n || F.value !== null && F.value === n ? void 0 : `all ${t.value.plot.transitionDurationMs}ms ease-in-out`
                }, null, 8, ["shape", "color", "plot", "radius", "stroke", "strokeWidth", "transition"])) : h("", !0),
                a.comment && t.value.chart.comments.show ? (r(), i("foreignObject", {
                  key: 1,
                  style: { overflow: "visible" },
                  height: "12",
                  width: t.value.chart.comments.width,
                  x: a.x - t.value.chart.comments.width / 2 + t.value.chart.comments.offsetX,
                  y: a.y + t.value.chart.comments.offsetY + 6
                }, [
                  p("div", gn, [
                    R(l.$slots, "plot-comment", {
                      plot: { ...a, color: e.color, seriesIndex: s, datapointIndex: n }
                    }, void 0, !0)
                  ])
                ], 8, mn)) : h("", !0)
              ]))), 128))
            ], 6))), 128)),
            (r(!0), i(d, null, y(te.value, (e, s) => (r(), i("g", {
              key: `serie_line_${e.id}`,
              class: tt(`serie_line_${s}`),
              style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
            }, [
              e.smooth && e.plots.length > 1 && e.curve ? (r(), i("path", {
                key: 0,
                d: `M${e.curve}`,
                stroke: t.value.chart.backgroundColor,
                "stroke-width": t.value.line.strokeWidth + 1,
                "stroke-dasharray": e.dashed ? t.value.line.strokeWidth * 2 : 0,
                fill: "none",
                style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
              }, null, 12, bn)) : e.plots.length > 1 && e.straight ? (r(), i("path", {
                key: 1,
                d: `M${e.straight}`,
                stroke: t.value.chart.backgroundColor,
                "stroke-width": t.value.line.strokeWidth + 1,
                "stroke-dasharray": e.dashed ? t.value.line.strokeWidth * 2 : 0,
                fill: "none",
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
              }, null, 12, pn)) : h("", !0)
            ], 6))), 128)),
            l.$slots.pattern ? (r(), i("defs", xn, [
              (r(!0), i(d, null, y(De.value, (e, s) => R(l.$slots, "pattern", lt({
                key: `serie_pattern_slot_${e.id}`
              }, { ref_for: !0 }, { ...e, seriesIndex: e.slotAbsoluteIndex, patternId: `pattern_${P.value}_${s}` }), void 0, !0)), 128))
            ])) : h("", !0),
            Ha.value.length && !f.value.isStacked ? (r(), i("g", yn, [
              (r(!0), i(d, null, y(Ha.value, (e) => (r(), i("path", {
                key: e.key,
                d: e.d,
                fill: e.color,
                "fill-opacity": t.value.line.interLine.fillOpacity,
                stroke: "none",
                "pointer-events": "none",
                style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
              }, null, 12, kn))), 128))
            ])) : h("", !0),
            (r(!0), i(d, null, y(te.value, (e, s) => (r(), i("g", {
              key: `serie_line_above_${e.id}`,
              class: tt(`serie_line_${s}`),
              style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
            }, [
              e.useArea && e.plots.length > 1 ? (r(), i("g", _n, [
                e.smooth ? (r(!0), i(d, { key: 0 }, y(e.curveAreas, (a, n) => (r(), i(d, { key: n }, [
                  a ? (r(), i("path", {
                    key: 0,
                    d: a,
                    fill: t.value.line.area.useGradient ? `url(#areaGradient_${s}_${P.value})` : v(ze)(e.color, t.value.line.area.opacity),
                    style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
                  }, null, 12, wn)) : h("", !0),
                  l.$slots.pattern && a ? (r(), i("path", {
                    key: 1,
                    d: a,
                    fill: `url(#pattern_${P.value}_${e.slotAbsoluteIndex})`,
                    style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
                  }, null, 12, Sn)) : h("", !0)
                ], 64))), 128)) : (r(!0), i(d, { key: 1 }, y(e.area.split(";"), (a, n) => (r(), i(d, { key: n }, [
                  a ? (r(), i("path", {
                    key: 0,
                    d: `M${a}Z`,
                    fill: t.value.line.area.useGradient ? `url(#areaGradient_${s}_${P.value})` : v(ze)(e.color, t.value.line.area.opacity),
                    style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
                  }, null, 12, $n)) : h("", !0),
                  l.$slots.pattern && a ? (r(), i("path", {
                    key: 1,
                    d: `M${a}Z`,
                    fill: `url(#pattern_${P.value}_${e.slotAbsoluteIndex})`,
                    style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
                  }, null, 12, Ln)) : h("", !0)
                ], 64))), 128))
              ])) : h("", !0),
              !e.hasDashedSegments && e.smooth && e.plots.length > 1 && e.curve ? (r(), i("path", {
                key: 1,
                d: `M${e.curve}`,
                stroke: e.temperatureColors ? `url(#temperature_grad_line_${s}_${P.value})` : e.color,
                "stroke-width": t.value.line.strokeWidth,
                "stroke-dasharray": e.dashed ? t.value.line.strokeWidth * 2 : 0,
                fill: "none",
                "stroke-linecap": "round",
                style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
              }, null, 12, Mn)) : e.hasDashedSegments ? (r(), i(d, { key: 2 }, [
                e.smooth ? (r(!0), i(d, { key: 0 }, y(e.dashedSmooth, (a) => (r(), i("path", {
                  key: a.path,
                  fill: "none",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  d: `M ${a.path}`,
                  stroke: e.temperatureColors ? `url(#temperature_grad_line_${s}_${P.value})` : e.color,
                  "stroke-width": t.value.line.strokeWidth,
                  "stroke-dasharray": a.dashed ? t.value.line.strokeWidth * 2 : 0
                }, null, 8, Cn))), 128)) : (r(!0), i(d, { key: 1 }, y(e.dashedStraight, (a) => (r(), i("path", {
                  key: a.path,
                  fill: "none",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  d: `M ${a.path}`,
                  stroke: e.temperatureColors ? `url(#temperature_grad_line_${s}_${P.value})` : e.color,
                  "stroke-width": t.value.line.strokeWidth,
                  "stroke-dasharray": a.dashed ? t.value.line.strokeWidth * 2 : 0
                }, null, 8, An))), 128))
              ], 64)) : e.plots.length > 1 && e.straight ? (r(), i("path", {
                key: 3,
                d: `M${e.straight}`,
                stroke: e.temperatureColors ? `url(#temperature_grad_line_${s}_${P.value})` : e.color,
                "stroke-width": t.value.line.strokeWidth,
                "stroke-dasharray": e.dashed ? t.value.line.strokeWidth * 2 : 0,
                fill: "none",
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                style: w({ transition: v(ue) || !t.value.line.showTransition ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out` })
              }, null, 12, Tn)) : h("", !0),
              (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                key: `circle_line_${s}_${n}`
              }, [
                !Pa.value.linePlot && a && xe(a.value) || Pa.value.linePlot && a && xe(a.value) && (_.value !== null && _.value === n || F.value !== null && F.value === n) || Ba(e.plots, n) ? (r(), $e(ml, {
                  key: 0,
                  shape: ["triangle", "square", "diamond", "pentagon", "hexagon", "star"].includes(e.shape) ? e.shape : "circle",
                  color: t.value.line.useGradient ? `url(#lineGradient_${s}_${P.value})` : t.value.line.dot.useSerieColor ? e.color : t.value.line.dot.fill,
                  plot: { x: v(V)(a.x), y: v(V)(a.y) },
                  radius: _.value !== null && _.value === n || F.value !== null && F.value === n ? (Ce.value.line || 6) * 1.5 : (Ba(e.plots, n), Ce.value.line || 6),
                  stroke: t.value.line.dot.useSerieColor ? t.value.chart.backgroundColor : e.color,
                  strokeWidth: t.value.line.dot.strokeWidth,
                  transition: v(ue) || !t.value.line.showTransition || _.value !== null && _.value === n || F.value !== null && F.value === n ? void 0 : `all ${t.value.line.transitionDurationMs}ms ease-in-out`
                }, null, 8, ["shape", "color", "plot", "radius", "stroke", "strokeWidth", "transition"])) : h("", !0),
                a.comment && t.value.chart.comments.show ? (r(), i("foreignObject", {
                  key: 1,
                  style: { overflow: "visible" },
                  height: "12",
                  width: t.value.chart.comments.width,
                  x: a.x - t.value.chart.comments.width / 2 + t.value.chart.comments.offsetX,
                  y: a.y + t.value.chart.comments.offsetY + 6
                }, [
                  p("div", In, [
                    R(l.$slots, "plot-comment", {
                      plot: { ...a, color: e.color, seriesIndex: s, datapointIndex: n }
                    }, void 0, !0)
                  ])
                ], 8, zn)) : h("", !0)
              ], 64))), 128))
            ], 6))), 128)),
            (t.value.bar.labels.show || t.value.bar.serieName.show) && f.value.dataLabels.show ? (r(), i("g", On, [
              (r(!0), i(d, null, y(ye.value, (e, s) => (r(), i(d, {
                key: `xLabel_bar_${e.id}`
              }, [
                (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                  key: `xLabel_bar_${s}_${n}`
                }, [
                  a && (!Object.hasOwn(e, "dataLabels") || e.dataLabels === !0 || _.value !== null && _.value === n || F.value !== null && F.value === n) && t.value.bar.labels.show ? (r(), i("text", {
                    key: 0,
                    x: f.value.useIndividualScale && f.value.isStacked ? a.x + J.value.line / 2 : ll(a) + We() / 2 - ul.value / 2,
                    y: v(V)(a.y) + (a.value >= 0 ? t.value.bar.labels.offsetY : -t.value.bar.labels.offsetY * 3),
                    "text-anchor": "middle",
                    "font-size": x.value.plotLabels,
                    fill: t.value.bar.labels.color,
                    style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                  }, oe(xe(a.value) ? v(ie)(
                    t.value.bar.labels.formatter,
                    a.value,
                    v(Ve)({
                      p: e.prefix || t.value.chart.labels.prefix,
                      v: a.value,
                      s: e.suffix || t.value.chart.labels.suffix,
                      r: t.value.bar.labels.rounding
                    }),
                    {
                      datapoint: a,
                      serie: e
                    }
                  ) : ""), 13, Nn)) : h("", !0),
                  a && t.value.bar.serieName.show ? (r(), i("text", {
                    key: 1,
                    x: f.value.useIndividualScale && f.value.isStacked ? a.x + J.value.line / 2 : a.x + We() * 1.1,
                    y: a.y + (a.value > 0 ? t.value.bar.serieName.offsetY : -t.value.bar.serieName.offsetY * 3),
                    "text-anchor": "middle",
                    "font-size": x.value.plotLabels,
                    fill: t.value.bar.serieName.useSerieColor ? e.color : t.value.bar.serieName.color,
                    "font-weight": t.value.bar.serieName.bold ? "bold" : "normal",
                    style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                  }, oe(t.value.bar.serieName.useAbbreviation ? v(Uu)({
                    source: e.name,
                    length: t.value.bar.serieName.abbreviationSize
                  }) : e.name), 13, Pn)) : h("", !0)
                ], 64))), 128))
              ], 64))), 128))
            ])) : h("", !0),
            t.value.plot.labels.show && f.value.dataLabels.show ? (r(), i("g", Fn, [
              (r(!0), i(d, null, y(he.value, (e, s) => (r(), i(d, {
                key: `xLabel_plot_${e.id}`
              }, [
                (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                  key: `xLabel_plot_${s}_${n}`
                }, [
                  a && !Object.hasOwn(e, "dataLabels") || e.dataLabels === !0 || _.value !== null && _.value === n || F.value !== null && F.value === n ? (r(), i("text", {
                    key: 0,
                    x: a.x,
                    y: a.y + t.value.plot.labels.offsetY,
                    "text-anchor": "middle",
                    "font-size": x.value.plotLabels,
                    fill: t.value.plot.labels.color,
                    style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                  }, oe(xe(a.value) ? v(ie)(
                    t.value.plot.labels.formatter,
                    a.value,
                    v(Ve)({
                      p: e.prefix || t.value.chart.labels.prefix,
                      v: a.value,
                      s: e.suffix || t.value.chart.labels.suffix,
                      r: t.value.plot.labels.rounding
                    }),
                    {
                      datapoint: a,
                      serie: e
                    }
                  ) : ""), 13, Vn)) : h("", !0)
                ], 64))), 128))
              ], 64))), 128))
            ])) : (r(), i("g", Rn, [
              (r(!0), i(d, null, y(he.value, (e, s) => (r(), i(d, {
                key: `xLabel_plot_${e.id}`
              }, [
                (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                  key: `xLabel_plot_${s}_${n}`
                }, [
                  t.value.plot.tag.followValue ? (r(), i(d, { key: 1 }, [
                    [F.value, _.value, Ue.value].includes(n) && e.useTag ? (r(), i("line", {
                      key: 0,
                      class: "vue-ui-xy-tag-plot",
                      x1: c.value?.left,
                      x2: c.value?.right,
                      y1: a.y,
                      y2: a.y,
                      "stroke-width": 1,
                      "stroke-linecap": "round",
                      "stroke-dasharray": "2",
                      stroke: e.color
                    }, null, 8, En)) : h("", !0)
                  ], 64)) : (r(), i(d, { key: 0 }, [
                    a && n === 0 && e.useTag && e.useTag === "start" ? (r(), i("foreignObject", {
                      key: 0,
                      x: a.x,
                      y: a.y - 20,
                      height: 24,
                      width: "150",
                      style: w(`overflow: visible; opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                    }, [
                      p("div", {
                        style: w(`padding: 3px; background:${v(ze)(e.color, 80)};color:${v(Ye)(e.color)};width:fit-content;font-size:${x.value.plotLabels}px;border-radius: 2px;`),
                        innerHTML: v(ie)(
                          t.value.plot.tag.formatter,
                          a.value,
                          e.name,
                          {
                            datapoint: a,
                            seriesIndex: n,
                            serieName: e.name
                          }
                        )
                      }, null, 12, Hn)
                    ], 12, Bn)) : h("", !0),
                    a && n === e.plots.length - 1 && e.useTag && e.useTag === "end" ? (r(), i("foreignObject", {
                      key: 1,
                      x: a.x - e.name.length * (x.value.plotLabels / 2),
                      y: a.y - 20,
                      height: 24,
                      width: "150",
                      style: w(`overflow: visible; opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                    }, [
                      p("div", {
                        style: w(`padding: 3px; background:${v(ze)(e.color, 80)};color:${v(Ye)(e.color)};width:fit-content;font-size:${x.value.plotLabels}px;border-radius: 2px;`),
                        innerHTML: v(ie)(
                          t.value.plot.tag.formatter,
                          a.value,
                          e.name,
                          {
                            datapoint: a,
                            seriesIndex: n,
                            serieName: e.name
                          }
                        )
                      }, null, 12, Wn)
                    ], 12, Dn)) : h("", !0)
                  ], 64))
                ], 64))), 128))
              ], 64))), 128))
            ])),
            t.value.line.labels.show && f.value.dataLabels.show ? (r(), i("g", Yn, [
              (r(!0), i(d, null, y(te.value, (e, s) => (r(), i(d, {
                key: `xLabel_line_${e.id}`
              }, [
                (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                  key: `xLabel_line_${s}_${n}`
                }, [
                  a && !Object.hasOwn(e, "dataLabels") || e.dataLabels === !0 || _.value !== null && _.value === n || F.value !== null && F.value === n ? (r(), i("text", {
                    key: 0,
                    x: a.x,
                    y: a.y + (a.value >= 0 ? t.value.line.labels.offsetY : -t.value.line.labels.offsetY * 3),
                    "text-anchor": "middle",
                    "font-size": x.value.plotLabels,
                    fill: t.value.line.labels.color,
                    style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                  }, oe(xe(a.value) ? v(ie)(
                    t.value.line.labels.formatter,
                    a.value,
                    v(Ve)({
                      p: e.prefix || t.value.chart.labels.prefix,
                      v: a.value,
                      s: e.suffix || t.value.chart.labels.suffix,
                      r: t.value.line.labels.rounding
                    }),
                    {
                      datapoint: a,
                      serie: e
                    }
                  ) : ""), 13, Xn)) : h("", !0)
                ], 64))), 128))
              ], 64))), 128))
            ])) : (r(), i("g", Gn, [
              (r(!0), i(d, null, y(te.value, (e, s) => (r(), i(d, {
                key: `xLabel_line_${e.id}`
              }, [
                (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                  key: `xLabel_line_${s}_${n}`
                }, [
                  t.value.line.tag.followValue ? (r(), i(d, { key: 1 }, [
                    [F.value, _.value, Ue.value].includes(n) && e.useTag ? (r(), i("line", {
                      key: 0,
                      class: "vue-ui-xy-tag-line",
                      x1: c.value?.left,
                      x2: c.value?.right,
                      y1: a.y,
                      y2: a.y,
                      "stroke-width": 1,
                      "stroke-linecap": "round",
                      "stroke-dasharray": "2",
                      stroke: e.color
                    }, null, 8, Kn)) : h("", !0)
                  ], 64)) : (r(), i(d, { key: 0 }, [
                    a && n === 0 && e.useTag && e.useTag === "start" ? (r(), i("foreignObject", {
                      key: 0,
                      x: a.x,
                      y: a.y - 20,
                      height: 24,
                      width: "150",
                      style: w(`overflow: visible; opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                    }, [
                      p("div", {
                        style: w(`padding: 3px; background:${v(ze)(e.color, 80)};color:${v(Ye)(e.color)};width:fit-content;font-size:${x.value.plotLabels}px;border-radius: 2px;`),
                        innerHTML: v(ie)(
                          t.value.line.tag.formatter,
                          a.value,
                          e.name,
                          {
                            datapoint: a,
                            seriesIndex: n,
                            serieName: e.name
                          }
                        )
                      }, null, 12, Un)
                    ], 12, qn)) : h("", !0),
                    a && n === e.plots.length - 1 && e.useTag && e.useTag === "end" ? (r(), i("foreignObject", {
                      key: 1,
                      x: a.x,
                      y: a.y - 20,
                      height: 24,
                      width: "150",
                      style: w(`overflow: visible; opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                    }, [
                      p("div", {
                        style: w(`padding: 3px; background:${v(ze)(e.color, 80)};color:${v(Ye)(e.color)};width:fit-content;font-size:${x.value.plotLabels}px;border-radius: 2px;`),
                        innerHTML: v(ie)(
                          t.value.line.tag.formatter,
                          a.value,
                          e.name,
                          {
                            datapoint: a,
                            seriesIndex: n,
                            serieName: e.name
                          }
                        )
                      }, null, 12, Jn)
                    ], 12, Zn)) : h("", !0)
                  ], 64))
                ], 64))), 128))
              ], 64))), 128))
            ])),
            (r(!0), i(d, null, y(te.value, (e, s) => (r(), i(d, {
              key: `xLabel_line_${e.id}`
            }, [
              (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                key: `xLabel_line_${s}_${n}`
              }, [
                a && n === 0 && e.showSerieName && e.showSerieName === "start" ? (r(), i("text", {
                  key: 0,
                  x: a.x - x.value.plotLabels,
                  y: a.y,
                  "font-size": x.value.plotLabels,
                  "text-anchor": "end",
                  fill: e.color,
                  innerHTML: v(fl)({
                    content: e.name,
                    fontSize: x.value.plotLabels,
                    fill: e.color,
                    x: a.x - x.value.plotLabels,
                    y: a.y,
                    maxWords: 2
                  }),
                  style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                }, null, 12, Qn)) : h("", !0),
                a && n === e.plots.length - 1 && e.showSerieName && e.showSerieName === "end" ? (r(), i("text", {
                  key: 1,
                  x: a.x + x.value.plotLabels,
                  y: a.y,
                  "font-size": x.value.plotLabels,
                  "text-anchor": "start",
                  fill: e.color,
                  innerHTML: v(fl)({
                    content: e.name,
                    fontSize: x.value.plotLabels,
                    fill: e.color,
                    x: a.x + x.value.plotLabels,
                    y: a.y,
                    maxWords: 2
                  }),
                  style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                }, null, 12, jn)) : h("", !0)
              ], 64))), 128))
            ], 64))), 128)),
            (r(!0), i(d, null, y(he.value, (e, s) => (r(), i(d, {
              key: `xLabel_plot_${e.id}`
            }, [
              (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
                key: `xLabel_plot_${s}_${n}`
              }, [
                a && n === 0 && e.showSerieName && e.showSerieName === "start" ? (r(), i("text", {
                  key: 0,
                  x: a.x - x.value.plotLabels,
                  y: a.y,
                  "font-size": x.value.plotLabels,
                  "text-anchor": "end",
                  fill: e.color,
                  innerHTML: v(fl)({
                    content: e.name,
                    fontSize: x.value.plotLabels,
                    fill: e.color,
                    x: a.x - x.value.plotLabels,
                    y: a.y,
                    maxWords: 2
                  }),
                  style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                }, null, 12, er)) : h("", !0),
                a && n === e.plots.length - 1 && e.showSerieName && e.showSerieName === "end" ? (r(), i("text", {
                  key: 1,
                  x: a.x + x.value.plotLabels,
                  y: a.y,
                  "font-size": x.value.plotLabels,
                  "text-anchor": "start",
                  fill: e.color,
                  innerHTML: v(fl)({
                    content: e.name,
                    fontSize: x.value.plotLabels,
                    fill: e.color,
                    x: a.x + x.value.plotLabels,
                    y: a.y,
                    maxWords: 2
                  }),
                  style: w(`opacity:${I.value ? I.value === e.groupId ? 1 : 0.2 : 1};transition:opacity 0.2s ease-in-out`)
                }, null, 12, tr)) : h("", !0)
              ], 64))), 128))
            ], 64))), 128)),
            (r(!0), i(d, null, y([...he.value, ...te.value, ...ye.value], (e, s) => (r(), i(d, {
              key: `progression-${s}`
            }, [
              Object.hasOwn(e, "useProgression") && e.useProgression === !0 && !isNaN(v(Le)(e.plots).trend) ? (r(), i("g", lr, [
                p("defs", null, [
                  p("marker", {
                    id: `progression_arrow_${s}`,
                    markerWidth: "9",
                    markerHeight: "9",
                    viewBox: "-1 -1 9 9",
                    markerUnits: "userSpaceOnUse",
                    refX: "7",
                    refY: 7 / 2,
                    orient: "auto",
                    overflow: "visible"
                  }, [
                    p("polygon", {
                      points: "0,0 7,3.5 0,7",
                      fill: e.color,
                      stroke: t.value.chart.backgroundColor,
                      "stroke-width": "1",
                      "stroke-linejoin": "round"
                    }, null, 8, or)
                  ], 8, ar)
                ]),
                e.plots.length > 1 ? (r(), i("line", {
                  key: 0,
                  x1: v(Le)(e.plots).x1 + (e.type === "bar" ? We() : 0),
                  x2: v(Le)(e.plots).x2 + (e.type === "bar" ? We() : 0),
                  y1: v(E)(v(Le)(e.plots).y1),
                  y2: v(E)(v(Le)(e.plots).y2),
                  "stroke-width": 1,
                  stroke: t.value.chart.backgroundColor,
                  "stroke-dasharray": 2,
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  "marker-end": `url(#progression_arrow_${s})`
                }, null, 8, ur)) : h("", !0),
                e.plots.length > 1 ? (r(), i("line", {
                  key: 1,
                  x1: v(Le)(e.plots).x1 + (e.type === "bar" ? We() : 0),
                  x2: v(Le)(e.plots).x2 + (e.type === "bar" ? We() : 0),
                  y1: v(E)(v(Le)(e.plots).y1),
                  y2: v(E)(v(Le)(e.plots).y2),
                  "stroke-width": 1,
                  stroke: e.color,
                  "stroke-dasharray": 2,
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  "marker-end": `url(#progression_arrow_${s})`
                }, null, 8, sr)) : h("", !0),
                e.plots.length > 1 ? (r(), i("text", {
                  key: 2,
                  "text-anchor": "middle",
                  x: v(Le)(e.plots).x2 + (e.type === "bar" ? We() : 0),
                  y: v(Le)(e.plots).y2 - 12,
                  "font-size": x.value.plotLabels,
                  fill: e.color,
                  stroke: t.value.chart.backgroundColor,
                  "stroke-width": 4,
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  "paint-order": "stroke fill"
                }, oe(v(Ve)({
                  v: v(Le)(e.plots).trend * 100,
                  s: "%",
                  r: 2
                })), 9, nr)) : h("", !0)
              ])) : h("", !0)
            ], 64))), 128)),
            f.value.useIndividualScale && !f.value.isStacked ? (r(), i(d, { key: 13 }, [
              p("defs", null, [
                (r(!0), i(d, null, y(Tt.value, (e, s) => (r(), i("linearGradient", {
                  id: `individual_scale_gradient_${P.value}_${s}`,
                  x1: "0%",
                  x2: "100%",
                  y1: "0%",
                  y2: "0%"
                }, [
                  p("stop", {
                    offset: "0%",
                    "stop-color": t.value.chart.backgroundColor,
                    "stop-opacity": "0"
                  }, null, 8, ir),
                  p("stop", {
                    offset: "100%",
                    "stop-color": e.color,
                    "stop-opacity": "0.2"
                  }, null, 8, cr)
                ], 8, rr))), 256))
              ]),
              (r(!0), i(d, null, y(Tt.value, (e, s) => (r(), i("rect", {
                x: e.x - t.value.chart.grid.labels.yAxis.labelWidth + U.value - c.value.individualOffsetX,
                y: c.value?.top,
                width: t.value.chart.grid.labels.yAxis.labelWidth + c.value.individualOffsetX,
                height: c.value.height < 0 ? 10 : c.value.height,
                fill: I.value === e.groupId ? `url(#individual_scale_gradient_${P.value}_${s})` : "transparent",
                onMouseenter: (a) => I.value = e.groupId,
                onMouseleave: u[0] || (u[0] = (a) => I.value = null)
              }, null, 40, vr))), 256))
            ], 64)) : h("", !0),
            p("g", null, [
              t.value.chart.grid.labels.axis.yLabel && !f.value.useIndividualScale ? (r(), i("text", {
                key: 0,
                ref_key: "yAxisLabel",
                ref: pl,
                "font-size": x.value.yAxis,
                fill: t.value.chart.grid.labels.color,
                transform: `translate(${t.value.chart.grid.labels.axis.fontSize}, ${c.value?.top + c.value.height / 2}) rotate(-90)`,
                "text-anchor": "middle",
                style: { transition: "none" }
              }, oe(t.value.chart.grid.labels.axis.yLabel), 9, dr)) : h("", !0),
              t.value.chart.grid.labels.axis.xLabel ? (r(), i("text", {
                key: 1,
                ref_key: "xAxisLabel",
                ref: bl,
                "text-anchor": "middle",
                x: K.value / 2,
                y: ce.value - 3,
                "font-size": x.value.yAxis,
                fill: t.value.chart.grid.labels.color
              }, oe(t.value.chart.grid.labels.axis.xLabel), 9, hr)) : h("", !0)
            ]),
            t.value.chart.grid.labels.xAxisLabels.show ? (r(), i("g", {
              key: 14,
              ref_key: "timeLabelsEls",
              ref: Ht
            }, [
              l.$slots["time-label"] ? (r(!0), i(d, { key: 0 }, y(jt.value, (e, s) => R(l.$slots, "time-label", lt({
                key: `time_label_${e.id}`,
                ref_for: !0
              }, {
                x: c.value?.left + c.value.width / $.value * s + c.value.width / $.value / 2,
                y: c.value?.bottom,
                fontSize: x.value.xAxis,
                fill: t.value.chart.grid.labels.xAxisLabels.color,
                transform: `translate(${c.value?.left + c.value.width / $.value * s + c.value.width / $.value / 2}, ${c.value?.bottom + x.value.xAxis * 1.3 + t.value.chart.grid.labels.xAxisLabels.yOffset}), rotate(${t.value.chart.grid.labels.xAxisLabels.rotation})`,
                absoluteIndex: e.absoluteIndex,
                content: e.text,
                textAnchor: t.value.chart.grid.labels.xAxisLabels.rotation > 0 ? "start" : t.value.chart.grid.labels.xAxisLabels.rotation < 0 ? "end" : "middle",
                show: e && e.text
              }), void 0, !0)), 128)) : (r(!0), i(d, { key: 1 }, y(jt.value, (e, s) => (r(), i("g", {
                key: `time_label_${s}`
              }, [
                e && e.text ? (r(), i(d, { key: 0 }, [
                  String(e.text).includes(`
`) ? (r(), i("text", {
                    key: 1,
                    class: "vue-data-ui-time-label",
                    "text-anchor": t.value.chart.grid.labels.xAxisLabels.rotation > 0 ? "start" : t.value.chart.grid.labels.xAxisLabels.rotation < 0 ? "end" : "middle",
                    "font-size": x.value.xAxis,
                    fill: t.value.chart.grid.labels.xAxisLabels.color,
                    transform: `translate(${c.value?.left + c.value.width / $.value * s + c.value.width / $.value / 2}, ${c.value?.bottom + x.value.xAxis * 1.5}), rotate(${t.value.chart.grid.labels.xAxisLabels.rotation})`,
                    style: w({
                      cursor: ha() && Ze.value ? "pointer" : "default"
                    }),
                    innerHTML: v(Zu)({
                      content: String(e.text),
                      fontSize: x.value.xAxis,
                      fill: t.value.chart.grid.labels.xAxisLabels.color,
                      x: 0,
                      y: 0
                    }),
                    onClick: () => xa(e, s)
                  }, null, 12, mr)) : (r(), i("text", {
                    key: 0,
                    class: "vue-data-ui-time-label",
                    "text-anchor": t.value.chart.grid.labels.xAxisLabels.rotation > 0 ? "start" : t.value.chart.grid.labels.xAxisLabels.rotation < 0 ? "end" : "middle",
                    "font-size": x.value.xAxis,
                    fill: t.value.chart.grid.labels.xAxisLabels.color,
                    transform: `translate(${c.value?.left + c.value.width / $.value * s + c.value.width / $.value / 2}, ${c.value?.bottom + x.value.xAxis * 1.5}), rotate(${t.value.chart.grid.labels.xAxisLabels.rotation})`,
                    style: w({
                      cursor: ha() && Ze.value ? "pointer" : "default"
                    }),
                    onClick: () => xa(e, s)
                  }, oe(e.text || ""), 13, fr))
                ], 64)) : h("", !0)
              ]))), 128))
            ], 512)) : h("", !0),
            Ra.value.length && !f.value.isStacked ? (r(), i("g", gr, [
              (r(!0), i(d, null, y(Ra.value, (e) => (r(), i("g", {
                key: e.uid
              }, [
                e.yTop && e.show && isFinite(e.yTop) ? (r(), i("line", {
                  key: 0,
                  x1: e.x1,
                  y1: e.yTop,
                  x2: e.x2,
                  y2: e.yTop,
                  stroke: e.config.line.stroke,
                  "stroke-width": e.config.line.strokeWidth,
                  "stroke-dasharray": e.config.line.strokeDasharray,
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, br)) : h("", !0),
                e.yBottom && e.show && isFinite(e.yBottom) ? (r(), i("line", {
                  key: 1,
                  x1: e.x1,
                  y1: e.yBottom,
                  x2: e.x2,
                  y2: e.yBottom,
                  stroke: e.config.line.stroke,
                  "stroke-width": e.config.line.strokeWidth,
                  "stroke-dasharray": e.config.line.strokeDasharray,
                  "stroke-linecap": "round",
                  style: { animation: "none !important" }
                }, null, 8, pr)) : h("", !0),
                e.hasArea && e.show && isFinite(e.yTop) && isFinite(e.yBottom) ? (r(), i("rect", {
                  key: 2,
                  y: Math.min(e.yTop, e.yBottom),
                  x: e.x1,
                  width: c.value.width,
                  height: v(V)(e.areaHeight, 0),
                  fill: v(ze)(e.config.area.fill, e.config.area.opacity),
                  style: { animation: "none !important" }
                }, null, 8, xr)) : h("", !0),
                e.config.label.text && e.show && isFinite(e._box.y) ? (r(), i("rect", lt({
                  key: 3,
                  class: "vue-ui-xy-annotation-label-box"
                }, { ref_for: !0 }, e._box, { style: { animation: "none !important", transition: "none !important" } }), null, 16)) : h("", !0),
                e.config.label.text && e.show && isFinite(e._text.y) ? (r(), i("text", {
                  key: 4,
                  id: e.id,
                  class: "vue-ui-xy-annotation-label",
                  x: e._text.x,
                  y: e._text.y,
                  "font-size": e.config.label.fontSize,
                  fill: e.config.label.color,
                  "text-anchor": e.config.label.textAnchor
                }, oe(e.config.label.text), 9, yr)) : h("", !0)
              ]))), 128))
            ])) : h("", !0),
            t.value.chart.timeTag.show && (![null, void 0].includes(_.value) || ![null, void 0].includes(F.value)) ? (r(), i("g", kr, [
              (r(), i("foreignObject", {
                x: yu(),
                y: c.value?.bottom,
                width: "200",
                height: "40",
                style: { overflow: "visible !important" }
              }, [
                p("div", {
                  ref_key: "timeTagEl",
                  ref: ja,
                  class: "vue-ui-xy-time-tag",
                  style: w(`width: fit-content;margin: 0 auto;text-align:center;padding:3px 12px;background:${t.value.chart.timeTag.backgroundColor};color:${t.value.chart.timeTag.color};font-size:${t.value.chart.timeTag.fontSize}px`),
                  innerHTML: ku.value
                }, null, 12, wr)
              ], 8, _r)),
              p("circle", {
                cx: c.value?.left + c.value.width / $.value * ((_.value !== null ? _.value : 0) || (F.value !== null ? F.value : 0)) + c.value.width / $.value / 2,
                cy: c.value?.bottom,
                r: t.value.chart.timeTag.circleMarker.radius,
                fill: t.value.chart.timeTag.circleMarker.color
              }, null, 8, Sr)
            ])) : h("", !0)
          ])) : h("", !0),
          oa.value ? (r(), i("rect", lt({ key: 2 }, Ho.value, {
            "data-start": b.value.start,
            "data-end": b.value.end
          }), null, 16, $r)) : h("", !0),
          R(l.$slots, "svg", {
            svg: {
              ...Po.value,
              isPrintingImg: v(qt) | v(Cl) | Yt.value,
              isPrintingSvg: Xt.value,
              data: [...te.value, ...ye.value, ...he.value],
              drawingArea: c.value
            }
          }, void 0, !0)
        ], 512)
      ], 46, bs)),
      l.$slots.watermark ? (r(), i("div", Lr, [
        R(l.$slots, "watermark", ge(be({ isPrinting: v(qt) || v(Cl) || Yt.value || Xt.value })), void 0, !0)
      ])) : h("", !0),
      (r(!0), i(d, null, y(te.value, (e, s) => (r(), i(d, {
        key: `tag_line_${e.id}`
      }, [
        (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
          key: `tag_line_${s}_${n}`
        }, [
          [F.value, _.value, Ue.value].includes(n) && e.useTag && e.useTag === "end" && t.value.line.tag.followValue ? (r(), i("div", {
            key: 0,
            ref_for: !0,
            ref: (o) => Kt(s, n, o, "right", "line"),
            class: "vue-ui-xy-tag",
            "data-tag": "right",
            style: w({
              position: "fixed",
              top: v(Xe)({
                svgElement: ve.value,
                x: c.value?.right + t.value.line.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_right_line`],
                position: "right"
              })?.top + "px",
              left: v(Xe)({
                svgElement: ve.value,
                x: c.value?.right + t.value.line.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_right_line`],
                position: "right"
              })?.left + "px",
              height: "fit-content",
              width: "fit-content",
              background: e.color,
              color: v(Ye)(e.color),
              padding: "0 6px",
              fontSize: t.value.line.tag.fontSize + "px",
              opacity: 1
            })
          }, [
            (r(), i("svg", Mr, [
              p("path", {
                d: "M 0,10 10,0 10,20 Z",
                fill: e.color,
                stroke: "none"
              }, null, 8, Cr)
            ])),
            p("div", {
              class: "vue-ui-xy-tag-content",
              innerHTML: v(ie)(
                t.value.line.tag.formatter,
                a.value,
                e.name,
                {
                  datapoint: a,
                  seriesIndex: n,
                  serieName: e.name
                }
              )
            }, null, 8, Ar)
          ], 4)) : h("", !0),
          [F.value, _.value, Ue.value].includes(n) && e.useTag && e.useTag === "start" && t.value.line.tag.followValue ? (r(), i("div", {
            key: 1,
            ref_for: !0,
            ref: (o) => Kt(s, n, o, "left", "line"),
            class: "vue-ui-xy-tag",
            "data-tag": "left",
            style: w({
              position: "fixed",
              top: v(Xe)({
                svgElement: ve.value,
                x: c.value?.left - t.value.line.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_left_line`],
                position: "left"
              })?.top + "px",
              left: v(Xe)({
                svgElement: ve.value,
                x: c.value?.left - t.value.line.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_left_line`],
                position: "left"
              })?.left + "px",
              height: "fit-content",
              width: "fit-content",
              background: e.color,
              color: v(Ye)(e.color),
              padding: "0 6px",
              fontSize: t.value.line.tag.fontSize + "px",
              opacity: 1
            })
          }, [
            (r(), i("svg", Tr, [
              p("path", {
                d: "M 0,0 10,10 0,20 Z",
                fill: e.color,
                stroke: "none"
              }, null, 8, zr)
            ])),
            p("div", {
              class: "vue-ui-xy-tag-content",
              innerHTML: v(ie)(
                t.value.line.tag.formatter,
                a.value,
                e.name,
                {
                  datapoint: a,
                  seriesIndex: n,
                  serieName: e.name
                }
              )
            }, null, 8, Ir)
          ], 4)) : h("", !0)
        ], 64))), 128))
      ], 64))), 128)),
      (r(!0), i(d, null, y(he.value, (e, s) => (r(), i(d, {
        key: `tag_plot_${e.id}`
      }, [
        (r(!0), i(d, null, y(e.plots, (a, n) => (r(), i(d, {
          key: `tag_plot_${s}_${n}`
        }, [
          [F.value, _.value, Ue.value].includes(n) && e.useTag && e.useTag === "end" && t.value.plot.tag.followValue ? (r(), i("div", {
            key: 0,
            ref_for: !0,
            ref: (o) => Kt(s, n, o, "right", "plot"),
            class: "vue-ui-xy-tag",
            "data-tag": "right",
            style: w({
              position: "fixed",
              top: v(Xe)({
                svgElement: ve.value,
                x: c.value?.right + t.value.plot.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_right_plot`],
                position: "right"
              })?.top + "px",
              left: v(Xe)({
                svgElement: ve.value,
                x: c.value?.right + t.value.plot.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_right_plot`],
                position: "right"
              })?.left + "px",
              height: "fit-content",
              width: "fit-content",
              background: e.color,
              color: v(Ye)(e.color),
              padding: "0 6px",
              fontSize: t.value.plot.tag.fontSize + "px",
              opacity: 1
            })
          }, [
            (r(), i("svg", Or, [
              p("path", {
                d: "M 0,10 10,0 10,20 Z",
                fill: e.color,
                stroke: "none"
              }, null, 8, Nr)
            ])),
            p("div", {
              class: "vue-ui-xy-tag-content",
              innerHTML: v(ie)(
                t.value.plot.tag.formatter,
                a.value,
                e.name,
                {
                  datapoint: a,
                  seriesIndex: n,
                  serieName: e.name
                }
              )
            }, null, 8, Pr)
          ], 4)) : h("", !0),
          [F.value, _.value, Ue.value].includes(n) && e.useTag && e.useTag === "start" && t.value.plot.tag.followValue ? (r(), i("div", {
            key: 1,
            ref_for: !0,
            ref: (o) => Kt(s, n, o, "left", "plot"),
            class: "vue-ui-xy-tag",
            "data-tag": "left",
            style: w({
              position: "fixed",
              top: v(Xe)({
                svgElement: ve.value,
                x: c.value?.left - t.value.plot.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_left_plot`],
                position: "left"
              })?.top + "px",
              left: v(Xe)({
                svgElement: ve.value,
                x: c.value?.left - t.value.plot.tag.fontSize / 1.5,
                y: a.y,
                element: Re.value[`${s}_${n}_left_plot`],
                position: "left"
              })?.left + "px",
              height: "fit-content",
              width: "fit-content",
              background: e.color,
              color: v(Ye)(e.color),
              padding: "0 6px",
              fontSize: t.value.plot.tag.fontSize + "px",
              opacity: 1
            })
          }, [
            (r(), i("svg", Fr, [
              p("path", {
                d: "M 0,0 10,10 0,20 Z",
                fill: e.color,
                stroke: "none"
              }, null, 8, Vr)
            ])),
            p("div", {
              class: "vue-ui-xy-tag-content",
              innerHTML: v(ie)(
                t.value.plot.tag.formatter,
                a.value,
                e.name,
                {
                  datapoint: a,
                  seriesIndex: n,
                  serieName: e.name
                }
              )
            }, null, 8, Rr)
          ], 4)) : h("", !0)
        ], 64))), 128))
      ], 64))), 128)),
      t.value.chart.zoom.show && Oe.value > 6 && Ll.value && _a.value ? (r(), $e(vs, {
        key: 5,
        ref_key: "chartSlicer",
        ref: Ie,
        uuid: P.value,
        allMinimaps: gu.value,
        background: t.value.chart.zoom.color,
        borderColor: t.value.chart.backgroundColor,
        customFormat: t.value.chart.zoom.customFormat,
        cutNullValues: t.value.line.cutNullValues,
        enableRangeHandles: t.value.chart.zoom.enableRangeHandles,
        enableSelectionDrag: t.value.chart.zoom.enableSelectionDrag,
        end: b.value.end,
        focusOnDrag: t.value.chart.zoom.focusOnDrag,
        focusRangeRatio: t.value.chart.zoom.focusRangeRatio,
        fontSize: t.value.chart.zoom.fontSize,
        useResetSlot: t.value.chart.zoom.useResetSlot,
        immediate: !t.value.chart.zoom.preview.enable,
        inputColor: t.value.chart.zoom.color,
        isPreview: oa.value,
        labelLeft: le.value[0] ? le.value[0].text : "",
        labelRight: le.value.at(-1) ? le.value.at(-1).text : "",
        max: Oe.value,
        min: 0,
        minimap: mu.value,
        minimapCompact: t.value.chart.zoom.minimap.compact,
        minimapFrameColor: t.value.chart.zoom.minimap.frameColor,
        minimapIndicatorColor: t.value.chart.zoom.minimap.indicatorColor,
        minimapLineColor: t.value.chart.zoom.minimap.lineColor,
        minimapMerged: t.value.chart.zoom.minimap.merged,
        minimapSelectedColor: t.value.chart.zoom.minimap.selectedColor,
        minimapSelectedColorOpacity: t.value.chart.zoom.minimap.selectedColorOpacity,
        minimapSelectedIndex: _.value,
        minimapSelectionRadius: t.value.chart.zoom.minimap.selectionRadius,
        preciseLabels: Ya.value.length ? Ya.value : Qt.value,
        refreshEndPoint: t.value.chart.zoom.endIndex !== null ? t.value.chart.zoom.endIndex + 1 : Math.max(...gl.dataset.map((e) => He(e.series).length)),
        refreshStartPoint: t.value.chart.zoom.startIndex !== null ? t.value.chart.zoom.startIndex : 0,
        selectColor: t.value.chart.zoom.highlightColor,
        selectedSeries: je.value,
        smoothMinimap: t.value.chart.zoom.minimap.smooth,
        start: b.value.start,
        textColor: t.value.chart.color,
        timeLabels: Qt.value,
        usePreciseLabels: t.value.chart.grid.labels.xAxisLabels.datetimeFormatter.enable && !t.value.chart.zoom.useDefaultFormat,
        valueEnd: b.value.end,
        valueStart: b.value.start,
        verticalHandles: t.value.chart.zoom.minimap.verticalHandles,
        minScale: t.value.chart.grid.labels.yAxis.scaleMin,
        maxScale: t.value.chart.grid.labels.yAxis.scaleMax,
        maxWidth: t.value.chart.zoom.maxWidth,
        additionalMinimapHeight: t.value.chart.zoom.minimap.additionalHeight,
        handleType: t.value.chart.zoom.minimap.handleType,
        handleIconColor: t.value.chart.zoom.minimap.handleIconColor,
        handleBorderWidth: t.value.chart.zoom.minimap.handleBorderWidth,
        handleBorderColor: t.value.chart.zoom.minimap.handleBorderColor,
        handleFill: t.value.chart.zoom.minimap.handleFill,
        handleWidth: t.value.chart.zoom.minimap.handleWidth,
        isCursorPointer: Ze.value,
        onFutureEnd: u[1] || (u[1] = (e) => ua("end", e)),
        onFutureStart: u[2] || (u[2] = (e) => ua("start", e)),
        onReset: tu,
        onTrapMouse: Ko,
        "onUpdate:end": eu,
        "onUpdate:start": jo
      }, {
        "reset-action": X(({ reset: e }) => [
          R(l.$slots, "reset-action", ge(be({ reset: e })), void 0, !0)
        ]),
        _: 3
      }, 8, ["uuid", "allMinimaps", "background", "borderColor", "customFormat", "cutNullValues", "enableRangeHandles", "enableSelectionDrag", "end", "focusOnDrag", "focusRangeRatio", "fontSize", "useResetSlot", "immediate", "inputColor", "isPreview", "labelLeft", "labelRight", "max", "minimap", "minimapCompact", "minimapFrameColor", "minimapIndicatorColor", "minimapLineColor", "minimapMerged", "minimapSelectedColor", "minimapSelectedColorOpacity", "minimapSelectedIndex", "minimapSelectionRadius", "preciseLabels", "refreshEndPoint", "refreshStartPoint", "selectColor", "selectedSeries", "smoothMinimap", "start", "textColor", "timeLabels", "usePreciseLabels", "valueEnd", "valueStart", "verticalHandles", "minScale", "maxScale", "maxWidth", "additionalMinimapHeight", "handleType", "handleIconColor", "handleBorderWidth", "handleBorderColor", "handleFill", "handleWidth", "isCursorPointer"])) : h("", !0),
      p("div", {
        id: `legend-bottom-${P.value}`
      }, null, 8, Br),
      Ul.value ? (r(), $e(Pu, {
        key: 6,
        to: t.value.chart.legend.position === "top" ? `#legend-top-${P.value}` : `#legend-bottom-${P.value}`
      }, [
        t.value.chart.legend.show ? (r(), i("div", {
          key: 0,
          ref_key: "chartLegend",
          ref: ut,
          class: "vue-ui-xy-legend",
          style: w({
            fontSize: `var(--legend-font-size, ${t.value.chart.legend.fontSize ?? 14}px)`
          })
        }, [
          t.value.chart.legend.selectAllToggle.show && se.value.length > 2 && !v(ue) ? (r(), $e(hs, {
            key: 0,
            backgroundColor: t.value.chart.legend.selectAllToggle.backgroundColor,
            color: t.value.chart.legend.selectAllToggle.color,
            fontSize: t.value.chart.legend.fontSize,
            checked: q.value.length > 0,
            onToggle: uu
          }, null, 8, ["backgroundColor", "color", "fontSize", "checked"])) : h("", !0),
          (r(!0), i(d, null, y(se.value, (e, s) => (r(), i("div", {
            key: `div_legend_item_${s}`,
            onClick: (a) => ol(e),
            class: tt({
              "vue-ui-xy-legend-item-alone": se.value.length === 1,
              "vue-ui-xy-legend-item": !0,
              "vue-ui-xy-legend-item-segregated": q.value.includes(e.id)
            }),
            style: w({
              cursor: Ze.value ? "pointer" : "default"
            })
          }, [
            yt.value[e.type] === "line" ? (r(), i("svg", Dr, [
              p("rect", {
                x: "0",
                y: "7.5",
                rx: "1.5",
                stroke: t.value.chart.backgroundColor,
                "stroke-width": 0.5,
                height: "3",
                width: "20",
                fill: e.color
              }, null, 8, Wr),
              ft(ml, {
                plot: { x: 10, y: 9 },
                radius: 4,
                color: e.color,
                shape: ["triangle", "square", "diamond", "pentagon", "hexagon", "star"].includes(e.shape) ? e.shape : "circle",
                stroke: t.value.chart.backgroundColor,
                strokeWidth: 0.5
              }, null, 8, ["color", "shape", "stroke"])
            ])) : yt.value[e.type] === "bar" ? (r(), i("svg", Er, [
              yt.value[e.type] === "bar" && l.$slots.pattern ? (r(), i("rect", {
                key: 0,
                x: "4",
                y: "4",
                rx: "1",
                height: "32",
                width: "32",
                stroke: "none",
                fill: e.color
              }, null, 8, Yr)) : h("", !0),
              yt.value[e.type] === "bar" ? (r(), i("rect", {
                key: 1,
                x: "4",
                y: "4",
                rx: "1",
                height: "32",
                width: "32",
                stroke: "none",
                fill: l.$slots.pattern ? `url(#pattern_${P.value}_${e.slotAbsoluteIndex})` : e.color
              }, null, 8, Xr)) : h("", !0)
            ])) : (r(), i("svg", Gr, [
              ft(ml, {
                plot: { x: 6, y: 6 },
                radius: 5,
                color: e.color,
                shape: ["triangle", "square", "diamond", "pentagon", "hexagon", "star"].includes(e.shape) ? e.shape : "circle"
              }, null, 8, ["color", "shape"])
            ])),
            p("span", {
              style: w(`color:${t.value.chart.legend.color}`)
            }, oe(e.name), 5)
          ], 14, Hr))), 128))
        ], 4)) : (r(), i("div", {
          key: 1,
          ref_key: "chartLegend",
          ref: ut
        }, [
          R(l.$slots, "legend", { legend: se.value }, void 0, !0)
        ], 512))
      ], 8, ["to"])) : h("", !0),
      l.$slots.source ? (r(), i("div", {
        key: 7,
        ref_key: "source",
        ref: pt,
        dir: "auto"
      }, [
        R(l.$slots, "source", {}, void 0, !0)
      ], 512)) : h("", !0),
      ft(v(wo), {
        teleportTo: t.value.chart.tooltip.teleportTo,
        show: f.value.showTooltip && El.value,
        backgroundColor: t.value.chart.tooltip.backgroundColor,
        color: t.value.chart.tooltip.color,
        fontSize: t.value.chart.tooltip.fontSize,
        borderRadius: t.value.chart.tooltip.borderRadius,
        borderColor: t.value.chart.tooltip.borderColor,
        borderWidth: t.value.chart.tooltip.borderWidth,
        backgroundOpacity: t.value.chart.tooltip.backgroundOpacity,
        position: t.value.chart.tooltip.position,
        offsetY: t.value.chart.tooltip.offsetY,
        parent: l.$refs.chart,
        content: bu.value,
        isFullscreen: qe.value,
        isCustom: t.value.chart.tooltip.customFormat && typeof t.value.chart.tooltip.customFormat == "function",
        smooth: t.value.chart.tooltip.smooth,
        backdropFilter: t.value.chart.tooltip.backdropFilter,
        smoothForce: t.value.chart.tooltip.smoothForce,
        smoothSnapThreshold: t.value.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": X(() => [
          R(l.$slots, "tooltip-before", ge(be({ ...Da.value })), void 0, !0)
        ]),
        "tooltip-after": X(() => [
          R(l.$slots, "tooltip-after", ge(be({ ...Da.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      Ll.value && t.value.chart.userOptions.buttons.table ? (r(), $e(Fu(zt.value.component), lt({ key: 8 }, zt.value.props, {
        ref_key: "tableUnit",
        ref: st,
        onClose: lo
      }), uo({
        content: X(() => [
          p("div", {
            style: w(`${v(qt) || t.value.table.useDialog ? "" : "max-height:400px"};${t.value.table.useDialog ? "height: fit-content; " : ""};overflow:auto;width:100%;${t.value.table.useDialog ? "" : "margin-top:48px"}`)
          }, [
            p("div", qr, [
              Vu(p("input", {
                type: "checkbox",
                "onUpdate:modelValue": u[4] || (u[4] = (e) => _t.value = e)
              }, null, 512), [
                [Ru, _t.value]
              ]),
              p("div", {
                onClick: u[5] || (u[5] = (e) => _t.value = !_t.value),
                style: w({
                  cursor: Ze.value ? "pointer" : "default"
                })
              }, [
                ft(v(Hl), {
                  name: "chartLine",
                  size: 20,
                  stroke: t.value.chart.color
                }, null, 8, ["stroke"])
              ], 4)
            ]),
            _t.value ? (r(), $e(v($o), {
              key: `sparkline_${_l.value}`,
              dataset: vu.value,
              config: du.value
            }, null, 8, ["dataset", "config"])) : (r(), $e(v(_o), {
              key: `table_${Xl.value}`,
              colNames: rl.value.colNames,
              head: rl.value.head,
              body: rl.value.body,
              config: rl.value.config,
              title: t.value.table.useDialog ? "" : zt.value.title,
              withCloseButton: !t.value.table.useDialog,
              onClose: lo
            }, {
              th: X(({ th: e }) => [
                p("div", { innerHTML: e }, null, 8, Ur)
              ]),
              td: X(({ td: e }) => [
                so(oe(isNaN(Number(e)) ? e : v(Ve)({
                  p: t.value.chart.labels.prefix,
                  v: e,
                  s: t.value.chart.labels.suffix,
                  r: t.value.table.rounding
                })), 1)
              ]),
              _: 1
            }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton"]))
          ], 4)
        ]),
        _: 2
      }, [
        t.value.table.useDialog ? {
          name: "title",
          fn: X(() => [
            so(oe(zt.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        t.value.table.useDialog ? {
          name: "actions",
          fn: X(() => [
            p("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: u[3] || (u[3] = (e) => zl(t.value.chart.userOptions.callbacks.csv))
            }, [
              ft(v(Hl), {
                name: "fileCsv",
                stroke: zt.value.props.color
              }, null, 8, ["stroke"])
            ])
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : h("", !0),
      R(l.$slots, "skeleton", {}, () => [
        v(ue) ? (r(), $e(as, { key: 0 })) : h("", !0)
      ], !0)
    ], 46, ms));
  }
}, mi = /* @__PURE__ */ fs(Zr, [["__scopeId", "data-v-42010bee"]]);
export {
  mi as default
};
