// vue-data-ui/dist/vue-ui-parallel-coordinate-plot-UFg10bHP.js
import { defineAsyncComponent as R, computed as f, ref as d, toRefs as Xt, watch as se, shallowRef as tt, onMounted as Yt, onBeforeUnmount as qt, watchEffect as Zt, openBlock as n, createElementBlock as u, unref as c, normalizeStyle as M, normalizeClass as q, createBlock as I, withCtx as h, renderSlot as p, normalizeProps as T, guardReactiveProps as P, createCommentVNode as g, createElementVNode as L, createSlots as at, createVNode as we, Fragment as z, renderList as O, toDisplayString as H, Teleport as Jt, resolveDynamicComponent as Kt, mergeProps as Qt, createTextVNode as lt, nextTick as ot } from "vue";
import { j as Ae, t as st, m as ea, p as ne, o as nt, q as it, r as ta, n as aa, s as la, D as oa, b as sa, c as na, ai as ia, X as ua, Z as ra, T as ut, a2 as ie, a0 as ca, a1 as va, k as rt, l as ct, U as da } from "./lib-GwPnJSrY.js";
import { t as vt, u as ha } from "./useResponsive-DfdjqQps.js";
import { u as pa } from "./useConfig-CIgyKZ6R.js";
import { u as fa } from "./usePrinter-DiK5R6k-.js";
import { u as ma, B as ya } from "./BaseScanner-BhhkKVoj.js";
import { u as ga } from "./useSvgExport-ByUukOZt.js";
import { u as $e } from "./useNestedProp-BBT0mbHh.js";
import { u as ba } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ka } from "./useUserOptionState-BIvW1Kz7.js";
import { u as xa } from "./useChartAccessibility-9icAAmYg.js";
import { u as Ca } from "./useTimeLabelCollider-CIsgDrl9.js";
import wa from "./vue_ui_parallel_coordinate_plot-DH3cFJ3n.js";
import Aa from "./Title-BfyrP-jp.js";
import $a from "./Legend-1LMMsPxL.js";
import { _ as Ta } from "./Shape-CSyWYgXc.js";
import Pa from "./img-BYYzY_QU.js";
import { B as za } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Fa } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Sa = ["id"], Ma = ["id"], Ia = ["xmlns", "viewBox"], Na = ["x", "y", "width", "height"], La = { style: { "pointer-events": "none" } }, Oa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], _a = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Da = { key: 0 }, Ra = ["x", "y", "fill", "font-size", "font-weight"], Ba = ["fill", "font-size", "font-weight", "text-anchor", "transform"], Ea = ["fill", "font-size", "font-weight", "text-anchor", "transform", "innerHTML"], Ua = { key: 0 }, Ha = ["x", "y", "font-size", "fill"], Va = ["width", "x", "y"], Wa = { style: { width: "100%" } }, ja = ["x", "y", "fill", "font-weight", "font-size", "onMouseenter", "onMouseleave", "onClick"], Ga = ["d", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Xa = ["d", "onMouseenter", "onMouseleave", "onClick"], Ya = {
  key: 4,
  class: "vue-data-ui-watermark"
}, qa = ["id"], Za = ["onClick"], Ja = ["innerHTML"], Ka = {
  __name: "vue-ui-parallel-coordinate-plot",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "selectDatapoint", "copyAlt"],
  setup(dt, { expose: ht, emit: pt }) {
    const ft = R(() => import("./Tooltip-DL3QgjZf.js")), mt = R(() => import("./BaseIcon-CNx8Tzl5.js")), yt = R(() => import("./vue-ui-accordion-BGHTqIBt.js")), gt = R(() => import("./DataTable-Bo1pmAqG.js")), bt = R(() => import("./PenAndPaper-DWV1FMIi.js")), kt = R(() => import("./UserOptions-CrLqEimB.js")), xt = R(() => import("./PackageVersion-DN9d0ME5.js")), Ct = R(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_parallel_coordinate_plot: wt } = pa(), { isThemeValid: At, warnInvalidTheme: $t } = ba(), A = dt, Z = f({
      get() {
        return !!A.dataset && A.dataset.length;
      },
      set(t) {
        return t;
      }
    }), Te = d(0), N = d(null), Pe = d(null), ze = d(null), Fe = d(null), Se = d(null), Me = d(0), Ie = d(0), Ne = d(0), Le = d(!1), j = d(null), ue = d(null), J = d(!1), K = d(!1), re = d(null), B = d(Ae()), E = d(!1);
    function Oe(t) {
      E.value = t, Te.value += 1;
    }
    const e = d(he()), U = f(() => e.value.userOptions.useCursorPointer), Tt = f(() => st({
      defaultConfig: {
        useCssAnimation: !1,
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            legend: {
              show: !0,
              backgroundColor: "transparent"
            },
            yAxis: {
              stroke: "#6A6A6A",
              labels: {
                showAxisNames: !1,
                axisNames: [],
                ticks: {
                  color: "#6A6A6A"
                }
              }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: _, FINAL_DATASET: Pt, manualLoading: ce } = ma({
      ...Xt(A),
      FINAL_CONFIG: e,
      prepareConfig: he,
      callback: () => {
        Promise.resolve().then(async () => {
          await ot(), y.value.showTable = e.value.table.show;
        });
      },
      skeletonDataset: A.config?.skeletonDataset ?? [
        {
          name: "",
          shape: "circle",
          color: "transparent",
          series: [
            {
              name: "",
              values: [1, 10, 100, 1e3]
            }
          ]
        },
        {
          name: "",
          shape: "circle",
          color: "#CACACA",
          series: [
            {
              name: "",
              values: [0.2, 3, 50, 800]
            }
          ]
        },
        {
          name: "",
          shape: "circle",
          color: "transparent",
          series: [
            {
              name: "",
              values: [0, 0, 0, 0]
            }
          ]
        }
      ],
      skeletonConfig: st({
        defaultConfig: e.value,
        userConfig: Tt.value
      })
    }), { userOptionsVisible: ve, setUserOptionsVisibility: _e, keepUserOptionState: De } = ka({ config: e.value }), { svgRef: de } = xa({ config: e.value.style.chart.title });
    function he() {
      const t = $e({
        userConfig: A.config,
        defaultConfig: wt
      }), l = t.theme;
      if (!l) return t;
      if (!At.value(t))
        return $t(t), t;
      const a = $e({
        userConfig: wa[l] || A.config,
        defaultConfig: t
      }), s = $e({
        userConfig: A.config,
        defaultConfig: a
      });
      return {
        ...s,
        customPalette: s.customPalette.length ? s.customPalette : ea[l] || ne
      };
    }
    se(() => A.config, (t) => {
      _.value || (e.value = he()), ve.value = !e.value.userOptions.showOnChartHover, Be(), Me.value += 1, Ne.value += 1, Ie.value += 1, y.value.dataLabels.show = e.value.style.chart.yAxis.labels.datapoints.show, y.value.showTable = e.value.table.show, y.value.showTooltip = e.value.style.chart.tooltip.show;
    }, { deep: !0 }), se(() => A.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (ce.value = !1);
    }, { deep: !0 });
    const D = tt(null), V = tt(null);
    Yt(() => {
      Le.value = !0, Be();
    });
    const Re = f(() => !!e.value.debug);
    function Be() {
      if (nt(A.dataset) ? (it({
        componentName: "VueUiParallelCoordinatePlot",
        type: "dataset",
        debug: Re.value
      }), ce.value = !0) : A.dataset.forEach((t, l) => {
        ta({
          datasetObject: t,
          requiredAttributes: ["name", "series"]
        }).forEach((a) => {
          Z.value = !1, it({
            componentName: "VueUiParallelCoordinatePlot",
            type: "datasetSerieAttribute",
            property: a,
            index: l,
            debug: Re.value
          });
        });
      }), nt(A.dataset) || (ce.value = e.value.loading), e.value.responsive) {
        const t = vt(() => {
          const { width: l, height: a } = ha({
            chart: N.value,
            title: e.value.style.chart.title.text ? Pe.value : null,
            legend: e.value.style.chart.legend.show ? ze.value : null,
            source: Fe.value,
            noTitle: Se.value
          });
          requestAnimationFrame(() => {
            v.value.width = l, v.value.height = a - 12, e.value.responsiveProportionalSizing ? (v.value.plotSize = ie({
              relator: Math.min(l, a),
              adjuster: 600,
              source: e.value.style.chart.plots.radius,
              threshold: 2,
              fallback: 2
            }), v.value.ticksFontSize = ie({
              relator: Math.min(l, a),
              adjuster: 600,
              source: e.value.style.chart.yAxis.labels.ticks.fontSize,
              threshold: 10,
              fallback: 10
            }), v.value.datapointFontSize = ie({
              relator: Math.min(l, a),
              adjuster: 600,
              source: e.value.style.chart.yAxis.labels.datapoints.fontSize,
              threshold: 10,
              fallback: 10
            }), v.value.axisNameFontSize = ie({
              relator: Math.min(l, a),
              adjuster: 600,
              source: e.value.style.chart.yAxis.labels.axisNamesFontSize,
              threshold: 12,
              fallback: 12
            })) : (v.value.plotSize = e.value.style.chart.plots.radius, v.value.ticksFontSize = e.value.style.chart.yAxis.labels.ticks.fontSize, v.value.datapointFontSize = e.value.style.chart.yAxis.labels.datapoints.fontSize, v.value.axisNameFontSize = e.value.style.chart.yAxis.labels.axisNamesFontSize);
          });
        });
        D.value && (V.value && D.value.unobserve(V.value), D.value.disconnect()), D.value = new ResizeObserver(t), V.value = N.value.parentNode, D.value.observe(V.value);
      }
    }
    qt(() => {
      D.value && (V.value && D.value.unobserve(V.value), D.value.disconnect());
    });
    const { isPrinting: pe, isImaging: fe, generatePdf: Ee, generateImage: Ue } = fa({
      elementId: `pcp_${B.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-parallel-coordinate-plot",
      options: e.value.userOptions.print
    }), zt = f(() => e.value.userOptions.show && !e.value.style.chart.title.text), v = d({
      height: e.value.style.chart.height,
      width: e.value.style.chart.width,
      plotSize: e.value.style.chart.plots.radius,
      // ratio 100
      ticksFontSize: e.value.style.chart.yAxis.labels.ticks.fontSize,
      // ratio 42.85
      datapointFontSize: e.value.style.chart.yAxis.labels.datapoints.fontSize,
      axisNameFontSize: e.value.style.chart.yAxis.labels.axisNamesFontSize
    }), He = f(() => v.value.width), Ve = f(() => v.value.height), G = d(0), Ft = vt((t) => {
      G.value = t;
    }, 100);
    Zt((t) => {
      const l = re.value;
      if (!l) return;
      const a = new ResizeObserver((s) => {
        Ft(s[0].contentRect.height);
      });
      a.observe(l), t(() => a.disconnect());
    });
    const m = f(() => {
      const { top: t, right: l, bottom: a, left: s } = e.value.style.chart.padding, o = Ve.value, r = He.value;
      return {
        chartHeight: Math.max(1e-3, o),
        chartWidth: Math.max(1e-3, r),
        height: Math.max(1e-3, o - t - a - G.value - v.value.datapointFontSize * 2),
        width: Math.max(1e-3, r - s - l),
        top: t + G.value + v.value.datapointFontSize,
        left: s,
        right: r - l,
        bottom: o - a - v.value.datapointFontSize
      };
    }), St = f(() => aa(e.value.customPalette)), y = d({
      dataLabels: {
        show: e.value.style.chart.yAxis.labels.datapoints.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    se(e, () => {
      y.value = {
        dataLabels: {
          show: e.value.style.chart.yAxis.labels.datapoints.show
        },
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const k = d([]);
    function Mt() {
      k.value.length ? k.value = [] : ee.value.forEach((t) => {
        k.value.push(t.id);
      });
    }
    function Q(t) {
      k.value.includes(t) ? k.value = k.value.filter((l) => l !== t) : k.value.push(t);
    }
    const F = f(() => Pt.value.map((t, l) => {
      const a = la(t.color) || St.value[l] || ne[l] || ne[l % ne.length];
      return {
        ...t,
        series: t.series.map((s) => ({
          ...s,
          id: Ae(),
          color: a
        })),
        seriesIndex: l,
        color: a,
        id: Ae(),
        shape: t.shape || "circle"
      };
    }));
    function We(t) {
      if (!F.value.length)
        return e.value.debug && console.warn("VueUiParallelCoordinatePlot - There are no series to show."), null;
      const l = F.value.find((a) => a.name === t);
      return l || (e.value.debug && console.warn(`VueUiParallelCoordinatePlot - Series name not found "${t}"`), null);
    }
    function It(t) {
      const l = We(t);
      l !== null && k.value.includes(l.id) && Q(l.id);
    }
    function Nt(t) {
      const l = We(t);
      l !== null && (k.value.includes(l.id) || Q(l.id));
    }
    const ee = f(() => F.value.map((t) => ({
      ...t,
      opacity: k.value.includes(t.id) ? 0.5 : 1,
      segregate: () => Q(t.id),
      isSegregated: k.value.includes(t.id),
      shape: t.shape || "circle"
    }))), Lt = f(() => ({
      cy: "pcp-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), je = f(() => Math.max(
      ...F.value.filter((t) => !k.value.includes(t.id)).map((t) => Math.max(...t.series.flatMap((l) => l.values.length)))
    )), $ = f(() => m.value.width / je.value), me = f(() => F.value.filter((t) => !k.value.includes(t.id))), b = f(() => {
      let t = [];
      for (let l = 0; l < je.value; l += 1) {
        const a = Math.min(...me.value.flatMap((x) => x.series.map((C) => C.values[l] || 0) || 0)), s = Math.max(...me.value.flatMap((x) => x.series.map((C) => C.values[l] || 0) || 0)), o = s === a ? a / 4 : a, r = s === a ? s * 2 : s, i = oa(o, r, e.value.style.chart.yAxis.scaleTicks), S = i.ticks.map((x, C) => {
          const jt = i.min < 0 ? x + Math.abs(i.min) : x - Math.abs(i.min), Gt = i.min < 0 ? i.max + Math.abs(i.min) : i.max - Math.abs(i.min);
          return {
            y: m.value.bottom - m.value.height * (jt / Gt),
            x: m.value.left + $.value * l + $.value / 2,
            value: x
          };
        });
        t.push({
          scale: i,
          ticks: S,
          name: e.value.style.chart.yAxis.labels.axisNames[l] || `Y-${l + 1}`
        });
      }
      return t;
    }), X = f(() => me.value.map((t, l) => ({
      ...t,
      series: t.series.map((a, s) => ({
        ...a,
        datapoints: a.values.map((o, r) => {
          const i = b.value[r].scale.min < 0 ? (o || 0) + Math.abs(b.value[r].scale.min) : (o || 0) - Math.abs(b.value[r].scale.min), S = b.value[r].scale.min < 0 ? b.value[r].scale.max + Math.abs(b.value[r].scale.min) : b.value[r].scale.max - Math.abs(b.value[r].scale.min);
          return {
            name: a.name,
            seriesName: t.name,
            axisIndex: r,
            datapointIndex: s,
            seriesIndex: l,
            value: o || 0,
            x: m.value.left + $.value * r + $.value / 2,
            y: m.value.bottom - m.value.height * (i / S),
            comment: a.comments && a.comments[r] || ""
          };
        })
      }))
    })).map((t) => ({
      ...t,
      series: t.series.map((l) => {
        const a = sa(l.datapoints), s = na(l.datapoints), o = ia(e.value.style.chart.lines.smooth ? `M ${s}` : `M ${a}`);
        return {
          ...l,
          smoothPath: s,
          straightPath: a,
          pathLength: o
        };
      })
    })));
    function Ge({ value: t, index: l, datapoint: a }) {
      return rt(
        e.value.style.chart.yAxis.labels.formatters[l] || null,
        t,
        ct({
          p: e.value.style.chart.yAxis.labels.prefixes[l] || "",
          v: t,
          s: e.value.style.chart.yAxis.labels.suffixes[l] || "",
          r: e.value.style.chart.yAxis.labels.roundings[l] || 0
        }),
        { datapoint: a, seriesIndex: l }
      );
    }
    const w = d(null), ye = d(null), ge = d(!1), be = d("");
    function te({ shape: t, serie: l, S: a }) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: { ...l, shape: t }, seriesIndex: a }), w.value = null, ge.value = null;
    }
    function ae({ shape: t, serieName: l, serie: a, relativeIndex: s, seriesIndex: o, S: r }) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: { ...a, shape: t }, seriesIndex: r }), ye.value = { serie: a, relativeIndex: s, seriesIndex: o, series: F.value, scales: b.value }, ge.value = !0, w.value = a.id;
      let i = "";
      const S = e.value.style.chart.tooltip.customFormat;
      ut(S) && da(() => S({
        serie: a,
        seriesIndex: a.seriesIndex,
        series: F.value,
        config: e.value,
        scales: b.value
      })) ? be.value = S({
        serie: a,
        seriesIndex: a.seriesIndex,
        series: F.value,
        config: e.value,
        scales: b.value
      }) : (i += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${l ? l + " - " : ""}${a.name}</div>`, b.value.map((x) => x.name).forEach((x, C) => {
        i += `
                <div class="vue-ui-tooltip-item" style="text-align:left">
                    <span>${x}: </span>
                    <span>
                        ${rt(
          e.value.style.chart.yAxis.labels.formatters[C] || null,
          a.datapoints[C].value,
          ct({
            p: e.value.style.chart.yAxis.labels.prefixes[C] || "",
            v: a.datapoints[C].value,
            s: e.value.style.chart.yAxis.labels.suffixes[C] || "",
            r: e.value.style.chart.yAxis.labels.roundings[C] || ""
          }),
          { datapoint: a.datapoints[C], seriesIndex: C }
        )}    
                    </span>
                </div>
            `, e.value.style.chart.comments.showInTooltip && a.datapoints[C].comment && (i += `<div class="vue-data-ui-tooltip-comment" style="background:${a.color}20; padding: 6px; margin-bottom: 6px; border-left: 1px solid ${a.color}">${a.datapoints[C].comment}</div>`);
      }), be.value = `<div>${i}</div>`);
    }
    function Ot() {
      return F.value;
    }
    const W = f(() => {
      const t = [e.value.table.columnNames.series].concat([e.value.table.columnNames.item]).concat(b.value.map((o) => o.name)), l = X.value.flatMap((o, r) => o.series.map((i) => [o.name].concat([i.name]).concat(i.values))), a = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, s = [e.value.table.columnNames.series].concat([e.value.table.columnNames.item]).concat(b.value.map((o) => o.name));
      return {
        body: l,
        head: t,
        config: a,
        colNames: s
      };
    }), Xe = f(() => X.value.length === 0 ? { head: [], body: [], config: {}, columnNames: [] } : {
      head: W.value.head,
      body: W.value.body
    });
    function ke(t = null) {
      const l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], a = Xe.value.head, s = Xe.value.body, o = l.concat([a]).concat(s), r = ca(o);
      t ? t(r) : va({ csvContent: r, title: e.value.style.chart.title.text || "vue-ui-parallel-coordinate-plot" });
    }
    const xe = pt;
    function le({ serie: t, shape: l, S: a }) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: { ...t, shape: l }, seriesIndex: a }), xe("selectDatapoint", t);
    }
    function Ye(t) {
      xe("selectLegend", {
        ...t,
        isSegregated: !t.isSegregated,
        opacity: t.isSegregated ? 1 : 0.5
      });
    }
    function qe() {
      y.value.showTable = !y.value.showTable;
    }
    function Ze() {
      y.value.dataLabels.show = !y.value.dataLabels.show;
    }
    function Je() {
      y.value.showTooltip = !y.value.showTooltip;
    }
    const oe = d(!1);
    function Ce() {
      oe.value = !oe.value;
    }
    async function _t({ scale: t = 2 } = {}) {
      if (!N.value) return;
      const { width: l, height: a } = N.value.getBoundingClientRect(), s = l / a, { imageUri: o, base64: r } = await Pa({ domElement: N.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: o,
        base64: r,
        title: e.value.style.chart.title.text,
        width: l,
        height: a,
        aspectRatio: s
      };
    }
    const Dt = d({ start: 0, end: 1 }), Rt = f(() => b.value.map((t) => t.name));
    Ca({
      timeLabelsEls: re,
      timeLabels: Rt,
      slicer: Dt,
      configRef: e,
      rotationPath: ["style", "chart", "yAxis", "labels", "axisNamesRotation"],
      autoRotatePath: ["style", "chart", "yAxis", "labels", "axisNamesAutoRotate", "enable"],
      isAutoSize: !1,
      width: He,
      height: Ve,
      targetClass: ".vue-ui-parallel-coordinate-plot-x-label",
      rotation: e.value.style.chart.yAxis.labels.axisNamesAutoRotate.angle
    });
    const Y = f(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = y.value.showTable;
      return {
        component: t ? Ct : yt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: E.value,
          fullscreenParent: N.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: U.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    se(() => y.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && j.value ? j.value.open() : "close" in j.value && j.value.close());
    });
    function Ke() {
      y.value.showTable = !1, ue.value && ue.value.setTableIconState(!1);
    }
    const Bt = f(() => e.value.style.chart.backgroundColor), Et = f(() => e.value.style.chart.legend), Ut = f(() => e.value.style.chart.title), { exportSvg: Ht, getSvg: Vt } = ga({
      svg: de,
      title: Ut,
      legend: Et,
      legendItems: ee,
      backgroundColor: Bt
    });
    async function Qe({ isCb: t }) {
      K.value = !0, await ot();
      try {
        if (t) {
          const { blob: l, url: a, text: s, dataUrl: o } = await Vt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: a, text: s, dataUrl: o }));
        } else
          await Promise.resolve(Ht());
      } finally {
        K.value = !1;
      }
    }
    function Wt(t) {
      if (t?.stage === "start") {
        J.value = !0;
        return;
      }
      if (t?.stage === "end") {
        J.value = !1;
        return;
      }
      Ue();
    }
    async function et() {
      if (xe("copyAlt", {
        config: e.value,
        dataset: X.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: X.value
      }));
    }
    return ht({
      getData: Ot,
      getImage: _t,
      generateCsv: ke,
      generatePdf: Ee,
      generateImage: Ue,
      generateSvg: Qe,
      hideSeries: Nt,
      showSeries: It,
      toggleTable: qe,
      toggleLabels: Ze,
      toggleTooltip: Je,
      toggleAnnotator: Ce,
      toggleFullscreen: Oe,
      copyAlt: et
    }), (t, l) => (n(), u("div", {
      ref_key: "pcpChart",
      ref: N,
      class: q(`vue-data-ui-component vue-ui-pcp ${E.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: M(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height:100%" : ""}`),
      id: `pcp_${B.value}`,
      onMouseenter: l[2] || (l[2] = () => c(_e)(!0)),
      onMouseleave: l[3] || (l[3] = () => c(_e)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (n(), I(c(bt), {
        key: 0,
        svgRef: c(de),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: oe.value,
        isCursorPointer: U.value,
        onClose: Ce
      }, {
        "annotator-action-close": h(() => [
          p(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": h(({ color: a }) => [
          p(t.$slots, "annotator-action-color", T(P({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": h(({ mode: a }) => [
          p(t.$slots, "annotator-action-draw", T(P({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": h(({ disabled: a }) => [
          p(t.$slots, "annotator-action-undo", T(P({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": h(({ disabled: a }) => [
          p(t.$slots, "annotator-action-redo", T(P({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": h(({ disabled: a }) => [
          p(t.$slots, "annotator-action-delete", T(P({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : g("", !0),
      zt.value ? (n(), u("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Se,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : g("", !0),
      e.value.style.chart.title.text ? (n(), u("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Pe,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (n(), I(Aa, {
          key: `title_${Me.value}`,
          config: {
            title: {
              cy: "pcp-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "pcp-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : g("", !0),
      L("div", {
        id: `legend-top-${B.value}`
      }, null, 8, Ma),
      e.value.userOptions.show && Z.value && (c(De) || c(ve)) ? (n(), I(c(kt), {
        ref_key: "userOptionsRef",
        ref: ue,
        key: `user_options_${Te.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: c(pe),
        isImaging: c(fe),
        uid: B.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: E.value,
        isTooltip: y.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: N.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: oe.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: U.value,
        onToggleFullscreen: Oe,
        onGeneratePdf: c(Ee),
        onGenerateCsv: ke,
        onGenerateImage: Wt,
        onGenerateSvg: Qe,
        onToggleTable: qe,
        onToggleLabels: Ze,
        onToggleTooltip: Je,
        onToggleAnnotator: Ce,
        onCopyAlt: et,
        style: M({
          visibility: c(De) ? c(ve) ? "visible" : "hidden" : "visible"
        })
      }, at({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: h(({ isOpen: a, color: s }) => [
            p(t.$slots, "menuIcon", T(P({ isOpen: a, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: h(() => [
            p(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: h(() => [
            p(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: h(() => [
            p(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: h(() => [
            p(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: h(() => [
            p(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: h(() => [
            p(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: h(({ toggleFullscreen: a, isFullscreen: s }) => [
            p(t.$slots, "optionFullscreen", T(P({ toggleFullscreen: a, isFullscreen: s })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: h(({ toggleAnnotator: a, isAnnotator: s }) => [
            p(t.$slots, "optionAnnotator", T(P({ toggleAnnotator: a, isAnnotator: s })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: h(({ altCopy: a }) => [
            p(t.$slots, "optionAltCopy", T(P({ altCopy: a })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : g("", !0),
      (n(), u("svg", {
        ref_key: "svgRef",
        ref: de,
        xmlns: c(ua),
        class: q({ "vue-data-ui-fullscreen--on": E.value, "vue-data-ui-fulscreen--off": !E.value }),
        viewBox: `0 0 ${m.value.chartWidth <= 0 ? 10 : m.value.chartWidth} ${m.value.chartHeight <= 0 ? 10 : m.value.chartHeight}`,
        style: M(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color}`)
      }, [
        we(c(xt)),
        t.$slots["chart-background"] ? (n(), u("foreignObject", {
          key: 0,
          x: m.value.left,
          y: m.value.top,
          width: m.value.width,
          height: m.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          p(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Na)) : g("", !0),
        (n(!0), u(z, null, O(b.value, (a, s) => (n(), u("g", La, [
          L("line", {
            x1: m.value.left + $.value * s + $.value / 2,
            x2: m.value.left + $.value * s + $.value / 2,
            y1: m.value.top,
            y2: m.value.bottom,
            stroke: e.value.style.chart.yAxis.stroke,
            "stroke-width": e.value.style.chart.yAxis.strokeWidth
          }, null, 8, Oa),
          e.value.style.chart.yAxis.labels.ticks.show ? (n(), u(z, { key: 0 }, [
            (n(!0), u(z, null, O(a.ticks, (o) => (n(), u("line", {
              x1: o.x,
              x2: o.x - 10,
              y1: o.y,
              y2: o.y,
              stroke: e.value.style.chart.yAxis.stroke,
              "stroke-width": e.value.style.chart.yAxis.strokeWidth,
              style: M(`opacity:${w.value && !y.value.showTooltip ? 0.2 : 1}`)
            }, null, 12, _a))), 256)),
            c(_) ? g("", !0) : (n(), u("g", Da, [
              (n(!0), u(z, null, O(a.ticks, (o) => (n(), u("text", {
                x: o.x - 12 + e.value.style.chart.yAxis.labels.ticks.offsetX,
                y: o.y + e.value.style.chart.yAxis.labels.ticks.offsetY + v.value.ticksFontSize / 3,
                fill: e.value.style.chart.yAxis.labels.ticks.color,
                "text-anchor": "end",
                "font-size": v.value.ticksFontSize,
                "font-weight": e.value.style.chart.yAxis.labels.ticks.bold ? "bold" : "normal",
                style: M(`opacity:${w.value && !y.value.showTooltip ? 0.2 : 1}`)
              }, H(Ge({ value: o.value, index: s, datapoint: o })), 13, Ra))), 256))
            ]))
          ], 64)) : g("", !0)
        ]))), 256)),
        e.value.style.chart.yAxis.labels.showAxisNames ? (n(), u("g", {
          key: 1,
          ref_key: "xAxisLabels",
          ref: re
        }, [
          (n(!0), u(z, null, O(b.value, (a, s) => (n(), u(z, null, [
            String(a.name).includes(`
`) ? (n(), u("text", {
              key: 1,
              class: "vue-ui-parallel-coordinate-plot-x-label",
              fill: e.value.style.chart.yAxis.labels.axisNamesColor,
              "font-size": v.value.axisNameFontSize,
              "font-weight": e.value.style.chart.yAxis.labels.axisNamesBold ? "bold" : "",
              "text-anchor": e.value.style.chart.yAxis.labels.axisNamesRotation === 0 ? "middle" : e.value.style.chart.yAxis.labels.axisNamesRotation < 0 ? "start" : "end",
              transform: `translate(${m.value.left + $.value * s + $.value / 2}, ${G.value - v.value.axisNameFontSize}), rotate(${e.value.style.chart.yAxis.labels.axisNamesRotation})`,
              innerHTML: c(ra)({
                content: String(a.name),
                fontSize: v.value.axisNameFontSize,
                fill: e.value.style.chart.yAxis.labels.axisNamesColor,
                x: 0,
                y: 0
              })
            }, null, 8, Ea)) : (n(), u("text", {
              key: 0,
              class: "vue-ui-parallel-coordinate-plot-x-label",
              fill: e.value.style.chart.yAxis.labels.axisNamesColor,
              "font-size": v.value.axisNameFontSize,
              "font-weight": e.value.style.chart.yAxis.labels.axisNamesBold ? "bold" : "",
              "text-anchor": e.value.style.chart.yAxis.labels.axisNamesRotation === 0 ? "middle" : e.value.style.chart.yAxis.labels.axisNamesRotation < 0 ? "start" : "end",
              transform: `translate(${m.value.left + $.value * s + $.value / 2}, ${G.value - v.value.axisNameFontSize}), rotate(${e.value.style.chart.yAxis.labels.axisNamesRotation})`
            }, H(a.name), 9, Ba))
          ], 64))), 256))
        ], 512)) : g("", !0),
        (n(!0), u(z, null, O(X.value, (a, s) => (n(), u("g", null, [
          (n(!0), u(z, null, O(a.series, (o, r) => (n(), u("g", null, [
            e.value.style.chart.plots.show ? (n(), u("g", Ua, [
              (n(!0), u(z, null, O(o.datapoints, (i, S) => (n(), I(Ta, {
                plot: { x: i.x, y: i.y },
                color: a.color,
                shape: a.shape,
                radius: a.shape === "triangle" ? v.value.plotSize * 1.2 : v.value.plotSize,
                stroke: e.value.style.chart.backgroundColor,
                strokeWidth: 0.5,
                onMouseenter: (x) => ae({
                  shape: a.shape,
                  serieName: a.name,
                  serie: o,
                  relativeIndex: r,
                  seriesIndex: o.seriesIndex,
                  S: s
                }),
                onMouseleave: (x) => te({ serie: o, shape: a.shape, S: s }),
                style: M(`opacity:${w.value ? w.value === o.id ? e.value.style.chart.plots.opacity : 0.2 : e.value.style.chart.plots.opacity}`),
                onClick: () => le({ serie: o, shape: a.shape, S: s })
              }, null, 8, ["plot", "color", "shape", "radius", "stroke", "onMouseenter", "onMouseleave", "style", "onClick"]))), 256)),
              y.value.showTooltip ? g("", !0) : (n(), u(z, { key: 0 }, [
                w.value && w.value === o.id && o.datapoints.length ? (n(), u("text", {
                  key: 0,
                  x: o.datapoints[0].x - v.value.ticksFontSize,
                  y: o.datapoints[0].y + v.value.ticksFontSize / 3,
                  "text-anchor": "end",
                  "font-size": v.value.ticksFontSize,
                  fill: a.color,
                  "font-weight": "bold"
                }, H(o.name), 9, Ha)) : g("", !0)
              ], 64)),
              e.value.style.chart.comments.show ? (n(!0), u(z, { key: 1 }, O(o.datapoints, (i) => (n(), u("g", null, [
                i.comment ? (n(), u("foreignObject", {
                  key: 0,
                  style: { overflow: "visible" },
                  height: "12",
                  width: e.value.style.chart.comments.width,
                  x: i.x - e.value.style.chart.comments.width / 2 + e.value.style.chart.comments.offsetX,
                  y: i.y + e.value.style.chart.comments.offsetY + 6
                }, [
                  L("div", Wa, [
                    p(t.$slots, "plot-comment", {
                      plot: { ...i, color: a.color }
                    }, void 0, !0)
                  ])
                ], 8, Va)) : g("", !0)
              ]))), 256)) : g("", !0)
            ])) : g("", !0),
            !c(_) && (y.value.dataLabels.show || w.value && w.value === o.id) ? (n(!0), u(z, { key: 1 }, O(o.datapoints, (i, S) => (n(), u("text", {
              x: i.x + 12 + e.value.style.chart.yAxis.labels.datapoints.offsetX,
              y: i.y + e.value.style.chart.yAxis.labels.datapoints.offsetY + v.value.datapointFontSize / 3,
              fill: e.value.style.chart.yAxis.labels.datapoints.useSerieColor ? a.color : e.value.style.chart.yAxis.labels.datapoints.color,
              "text-anchor": "start",
              "font-weight": e.value.style.chart.yAxis.labels.datapoints.bold ? "bold" : "normal",
              class: q({ "vue-ui-pcp-animated": !1, "vue-ui-pcp-transition": !c(_) }),
              "font-size": v.value.datapointFontSize,
              onMouseenter: (x) => ae({
                shape: a.shape,
                serieName: a.name,
                serie: o,
                relativeIndex: r,
                seriesIndex: o.seriesIndex,
                S: s
              }),
              onMouseleave: (x) => te({ serie: o, shape: a.shape, S: s }),
              onClick: () => le({ serie: o, shape: a.shape, S: s }),
              style: M(`opacity:${w.value ? w.value === o.id ? 1 : 0.2 : 1}`)
            }, H(Ge({ value: i.value, index: S, datapoint: i })), 47, ja))), 256)) : g("", !0),
            L("path", {
              d: `M${e.value.style.chart.lines.smooth ? o.smoothPath : o.straightPath}`,
              stroke: a.color,
              "stroke-width": e.value.style.chart.lines.strokeWidth,
              fill: "none",
              class: q({ "vue-ui-pcp-animated vue-data-ui-line-animated": e.value.useCssAnimation, "vue-ui-pcp-transition": !c(_) }),
              onMouseenter: (i) => ae({
                shape: a.shape,
                serieName: a.name,
                serie: o,
                relativeIndex: r,
                seriesIndex: o.seriesIndex,
                S: s
              }),
              onMouseleave: (i) => te({ serie: o, shape: a.shape, S: s }),
              onClick: () => le({ serie: o, shape: a.shape, S: s }),
              style: M(`opacity:${w.value ? w.value === o.id ? e.value.style.chart.lines.opacity : 0.2 : e.value.style.chart.lines.opacity}; stroke-dasharray:${o.pathLength}; stroke-dashoffset: ${e.value.useCssAnimation ? o.pathLength : 0}`)
            }, null, 46, Ga),
            y.value.showTooltip ? (n(), u("path", {
              key: 2,
              d: `M${e.value.style.chart.lines.smooth ? o.smoothPath : o.straightPath}`,
              stroke: "transparent",
              "stroke-width": 12,
              fill: "none",
              class: q({ "vue-ui-pcp-animated vue-data-ui-line-animated": e.value.useCssAnimation, "vue-ui-pcp-transition": !c(_) }),
              onMouseenter: (i) => ae({
                shape: a.shape,
                serieName: a.name,
                serie: o,
                relativeIndex: r,
                seriesIndex: o.seriesIndex,
                S: s
              }),
              onMouseleave: (i) => te({ serie: o, shape: a.shape, S: s }),
              onClick: () => le({ serie: o, shape: a.shape, S: s }),
              style: { opacity: "0" }
            }, null, 42, Xa)) : g("", !0)
          ]))), 256))
        ]))), 256)),
        p(t.$slots, "svg", {
          svg: {
            ...m.value,
            isPrintingImg: c(pe) | c(fe) | J.value,
            isPrintingSvg: K.value
          }
        }, void 0, !0)
      ], 14, Ia)),
      t.$slots.watermark ? (n(), u("div", Ya, [
        p(t.$slots, "watermark", T(P({ isPrinting: c(pe) || c(fe) || J.value || K.value })), void 0, !0)
      ])) : g("", !0),
      L("div", {
        id: `legend-bottom-${B.value}`
      }, null, 8, qa),
      Le.value ? (n(), I(Jt, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${B.value}` : `#legend-bottom-${B.value}`
      }, [
        L("div", {
          ref_key: "chartLegend",
          ref: ze
        }, [
          e.value.style.chart.legend.show && Z.value ? (n(), I($a, {
            key: `legend_${Ie.value}`,
            legendSet: ee.value,
            config: Lt.value,
            isCursorPointer: U.value,
            onClickMarker: l[0] || (l[0] = ({ legend: a }) => {
              Q(a.id), Ye(a);
            })
          }, {
            item: h(({ legend: a, index: s }) => [
              L("div", {
                onClick: (o) => {
                  a.segregate(), Ye(a);
                },
                style: M(`opacity:${k.value.includes(a.id) ? 0.5 : 1}`)
              }, H(a.name), 13, Za)
            ]),
            legendToggle: h(() => [
              ee.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !c(_) ? (n(), I(za, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: k.value.length > 0,
                isCursorPointer: U.value,
                onToggle: Mt
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : g("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : p(t.$slots, "legend", {
            key: 1,
            legend: F.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : g("", !0),
      t.$slots.source ? (n(), u("div", {
        key: 6,
        ref_key: "source",
        ref: Fe,
        dir: "auto"
      }, [
        p(t.$slots, "source", {}, void 0, !0)
      ], 512)) : g("", !0),
      we(c(ft), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: y.value.showTooltip && ge.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: N.value,
        content: be.value,
        isFullscreen: E.value,
        isCustom: c(ut)(e.value.style.chart.tooltip.customFormat),
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": h(() => [
          p(t.$slots, "tooltip-before", T(P({ ...ye.value })), void 0, !0)
        ]),
        "tooltip-after": h(() => [
          p(t.$slots, "tooltip-after", T(P({ ...ye.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      Z.value && e.value.userOptions.buttons.table ? (n(), I(Kt(Y.value.component), Qt({ key: 7 }, Y.value.props, {
        ref_key: "tableUnit",
        ref: j,
        onClose: Ke
      }), at({
        content: h(() => [
          (n(), I(c(gt), {
            key: `table_${Ne.value}`,
            colNames: W.value.colNames,
            head: W.value.head,
            body: W.value.body,
            config: W.value.config,
            title: e.value.table.useDialog ? "" : Y.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: U.value,
            onClose: Ke
          }, {
            th: h(({ th: a }) => [
              L("div", { innerHTML: a }, null, 8, Ja)
            ]),
            td: h(({ td: a }) => [
              lt(H(a), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: h(() => [
            lt(H(Y.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: h(() => [
            L("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (a) => ke(e.value.userOptions.callbacks.csv)),
              style: M({ cursor: U.value ? "pointer" : "default" })
            }, [
              we(c(mt), {
                name: "fileCsv",
                stroke: Y.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : g("", !0),
      p(t.$slots, "skeleton", {}, () => [
        c(_) ? (n(), I(ya, { key: 0 })) : g("", !0)
      ], !0)
    ], 46, Sa));
  }
}, gl = /* @__PURE__ */ Fa(Ka, [["__scopeId", "data-v-5ab7113f"]]);
export {
  gl as default
};
