// vue-data-ui/dist/vue-ui-table-sparkline-ClpLLKi2.js
import { useCssVars as Ge, defineAsyncComponent as K, ref as c, useSlots as Le, onMounted as ke, computed as k, watch as Ce, shallowRef as Re, onBeforeUnmount as He, openBlock as v, createElementBlock as p, unref as u, normalizeClass as Q, normalizeStyle as s, createElementVNode as f, toDisplayString as x, createCommentVNode as b, withDirectives as qe, createVNode as N, Fragment as Y, renderList as Z, createBlock as Xe, createSlots as je, withCtx as $, renderSlot as S, mergeProps as ee, nextTick as Je } from "vue";
import { u as te } from "./useNestedProp-BBT0mbHh.js";
import { j as Ke, m as Qe, p as M, n as Ye, o as Ze, q as _e, r as et, ao as tt, s as at, k as U, l as B, a0 as lt, a1 as ot } from "./lib-GwPnJSrY.js";
import { u as nt } from "./useConfig-CIgyKZ6R.js";
import { u as st } from "./usePrinter-DiK5R6k-.js";
import { u as ut } from "./useThemeCheck-D1ZYXUwG.js";
import { u as rt } from "./useUserOptionState-BIvW1Kz7.js";
import { v as it } from "./vClickOutside-C6WiFswA.js";
import dt from "./vue_ui_table_sparkline-yQdijcyJ.js";
import { _ as vt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ct = ["id"], pt = { style: { "z-index": "1", "padding-right": "24px" } }, ft = {
  key: 0,
  style: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center"
  }
}, mt = {
  key: 0,
  style: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center"
  }
}, gt = ["onClick"], bt = ["onClick"], yt = ["data-cell"], ht = ["data-cell", "onPointerenter"], kt = ["data-cell"], Ct = ["data-cell"], _t = ["data-cell"], xt = ["data-cell"], wt = {
  key: 1,
  ref: "source",
  dir: "auto"
}, At = {
  __name: "vue-ui-table-sparkline",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["copyAlt"],
  setup(xe, { expose: we, emit: Ae }) {
    Ge((t) => ({
      v7cfb4b44: t.tdo
    }));
    const Se = K(() => import("./vue-ui-sparkline-B2RDrdiR.js")), F = K(() => import("./BaseIcon-CNx8Tzl5.js")), Ie = K(() => import("./UserOptions-CrLqEimB.js")), { vue_ui_table_sparkline: Oe } = nt(), { isThemeValid: $e, warnInvalidTheme: Pe } = ut(), C = xe, ze = Ae, E = c(Ke()), ae = c(0), G = c(0), L = c(null), De = Le(), le = c(!1);
    ke(() => {
      De["chart-background"] && console.warn("VueUiTableSparkline does not support the #chart-background slot.");
    });
    const e = k({
      get: () => se(),
      set: (t) => t
    }), P = k(() => e.value.userOptions.useCursorPointer), { userOptionsVisible: R, setUserOptionsVisibility: oe, keepUserOptionState: ne } = rt({ config: e.value });
    function se() {
      const t = te({
        userConfig: C.config,
        defaultConfig: Oe
      }), l = t.theme;
      if (!l) return t;
      if (!$e.value(t))
        return Pe(t), t;
      const a = te({
        userConfig: dt[l] || C.config,
        defaultConfig: t
      }), n = te({
        userConfig: C.config,
        defaultConfig: a
      });
      return {
        ...n,
        customPalette: n.customPalette.length ? n.customPalette : Qe[l] || M
      };
    }
    Ce(() => C.config, (t) => {
      e.value = se(), R.value = !e.value.userOptions.showOnChartHover, ve(), G.value += 1;
    }, { deep: !0 }), Ce(() => C.dataset, (t) => {
      m.value = z.value, G.value += 1;
    }, { deep: !0 });
    const { isPrinting: ue, isImaging: re, generatePdf: ie, generateImage: de } = st({
      elementId: `table_${E.value}`,
      fileName: e.value.title.text || "vue-ui-table-sparkline",
      options: e.value.userOptions.print
    }), Ne = k(() => Ye(e.value.customPalette)), T = c(null), H = c(!1), Fe = k(() => e.value.responsiveBreakpoint);
    ke(() => {
      ve();
    });
    const q = c(e.value.colNames), I = Re(null);
    function ve() {
      Ze(C.dataset) && _e({
        componentName: "VueUiTableSparkline",
        type: "dataset"
      }), I.value && I.value.disconnect(), I.value = new ResizeObserver((t) => {
        t.forEach((l) => {
          H.value = l.contentRect.width < Fe.value;
        });
      }), T.value && I.value.observe(T.value), q.value = [];
      for (let t = 0; t < pe.value; t += 1)
        q.value.push(e.value.colNames[t] || `col ${t}`);
    }
    He(() => {
      I.value && I.value.disconnect();
    });
    const V = k(() => (C.dataset.forEach((t, l) => {
      et({
        datasetObject: t,
        requiredAttributes: ["name", "values"]
      }).forEach((a) => {
        _e({
          componentName: "VueUiTableSparkline",
          type: "datasetSerieAttribute",
          property: a,
          index: l
        });
      });
    }), C.dataset.map((t, l) => {
      const a = (t.values || []).map((d) => isNaN(d) ? 0 : d ?? 0), n = a.reduce((d, i) => d + i, 0), r = n / a.length, o = tt(a);
      return {
        ...t,
        values: t.values || [],
        color: at(t.color) || Ne.value[l] || M[l] || M[l % M.length],
        sum: n,
        average: r,
        median: o,
        sparklineDataset: a.map((d, i) => ({
          period: e.value.colNames[i] || `col ${i}`,
          value: d || 0
        }))
      };
    })));
    function Te(t) {
      const a = (t[0]?.values || []).map(
        (r, o) => t.map((d) => d?.values[o] || [])
      ).map(
        (r) => r.map((o, d) => [o, d]).sort((o, d) => d[0] - o[0]).map((o) => o[1])
      );
      return t.map((r, o) => ({
        ...r,
        values: r.values || [],
        orders: a[o]
      }));
    }
    const z = k(() => Te(V.value)), m = c(z.value), Ve = k(() => Math.max(...m.value.map((t) => (t.values || []).length))), X = c(void 0), j = c(!1), g = c(void 0), ce = c(1);
    function O() {
      j.value = !1, g.value = void 0, y.value = void 0, ce.value = 1, D.value.forEach((t) => t.state = 1), w.value = {
        name: 1,
        sum: 1,
        average: 1,
        median: 1
      }, m.value = z.value;
    }
    function W(t, l, a) {
      if (["name", "sum", "average", "median"].includes(t.type)) {
        We(t.type, l, a);
        return;
      }
      if (!ge(l)) return;
      if (A.value = l, y.value = void 0, ![null, void 0].includes(g.value) && l !== g.value && (g.value = void 0, O()), D.value[l].state === a && g.value === l) {
        g.value = void 0, O();
        return;
      }
      j.value = !0, g.value = l;
      const n = z.value.map((i) => i.values[l] || []), r = a;
      D.value[l].state = r, ce.value = r, l === X.value ? X.value = void 0 : X.value = l;
      const d = n.map((i, h) => [h, i]).sort((i, h) => r * (h[1] - i[1])).map((i) => i[0]).map((i) => z.value[i]);
      m.value = d, G.value += 1;
    }
    const pe = k(() => Math.max(...C.dataset.map((t) => (t.values || []).length))), _ = k(() => {
      let t = q.value.map((a) => ({
        type: "reg",
        value: a
      }));
      if (!t.length)
        for (let a = 0; a < pe.value; a += 1)
          t.push({ type: "reg", value: `col ${a + 1}` });
      e.value.showTotal && (t = [...t, { type: "sum", value: e.value.translations.total }]);
      let l;
      return e.value.showAverage && e.value.showMedian ? l = [
        ...t,
        { type: "average", value: e.value.translations.average },
        { type: "median", value: e.value.translations.median }
      ] : e.value.showAverage && !e.value.showMedian ? l = [...t, { type: "average", value: e.value.translations.average }] : !e.value.showAverage && e.value.showMedian ? l = [...t, { type: "median", value: e.value.translations.median }] : l = t, e.value.showSparklines ? [...l, { type: "chart", value: e.value.translations.chart }] : l;
    }), w = c({
      name: 1,
      sum: 1,
      average: 1,
      median: 1
    }), y = c(void 0);
    function We(t, l, a) {
      if (!m.value || m.value.length === 0 || !be(t)) return;
      if (y.value !== t && (y.value = void 0), ![null, void 0].includes(g.value) && l !== g.value && O(), g.value = void 0, w.value[t] === a && y.value) {
        y.value = void 0, O();
        return;
      }
      y.value = t, j.value = !0, w.value[t] = a, [null, void 0].includes(l) || (D.value[l].state = w.value[t]);
      const n = w.value[t], r = [...m.value].sort((o, d) => {
        const i = o[t] ?? (typeof o[t] == "number" ? 0 : ""), h = d[t] ?? (typeof d[t] == "number" ? 0 : "");
        return typeof i == "string" && typeof h == "string" ? n === -1 ? i.localeCompare(h) : h.localeCompare(i) : typeof i == "number" && typeof h == "number" ? n === -1 ? i - h : h - i : 0;
      });
      m.value = r;
    }
    const A = c(void 0), J = c(void 0);
    function Me({ dataIndex: t, serieIndex: l }) {
      A.value = t, J.value = l, L.value[t] && !H.value && L.value[t].scrollIntoView({ behavior: "smooth", block: "nearest", inline: "center" });
    }
    const fe = c(!1);
    function Ue(t) {
      fe.value = t, ae.value += 1;
    }
    function me(t = null) {
      Je(() => {
        const l = [e.value.translations.serie].concat(_.value), a = V.value.map((o, d) => [
          [o.name],
          o.values,
          [o.sum],
          [o.average],
          [o.median]
        ]), n = [l].concat(a), r = lt(n);
        t ? t(r) : ot({
          csvContent: r,
          title: e.value.title.text || "vue-ui-table-sparkline"
        });
      });
    }
    function ge(t) {
      return e.value.sortedDataColumnIndices.includes(t);
    }
    function be(t) {
      return t.type === "name" || t === "name" ? e.value.sortedSeriesName : t.type === "sum" || t === "sum" ? e.value.sortedSum : t.type === "average" || t === "average" ? e.value.sortedAverage : t.type === "median" || t === "median" ? e.value.sortedMedian : !1;
    }
    function ye(t, l, a) {
      return ["sum", "average", "median"].includes(l.type) ? y.value === l.type && w.value[l.type] === a ? 1 : 0.3 : t === g.value && D.value[t].state === a ? 1 : 0.3;
    }
    function Be() {
      e.value.resetSortOnClickOutside && O();
    }
    const D = k({
      get: () => _.value.map((t) => ({ state: 1 })),
      set: (t) => t
    });
    function Ee(t) {
      if (t?.stage === "start") {
        le.value = !0;
        return;
      }
      if (t?.stage === "end") {
        le.value = !1;
        return;
      }
      de();
    }
    async function he() {
      if (ze("copyAlt", {
        config: e.value,
        dataset: V.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: V.value
      }));
    }
    return we({
      generatePdf: ie,
      generateImage: de,
      generateCsv: me,
      restoreOrder: O,
      copyAlt: he
    }), (t, l) => (v(), p("div", {
      ref_key: "tableContainer",
      ref: T,
      class: Q({ "vue-data-ui-component": !0, "vue-ui-table-sparkline": !0, "vue-ui-responsive": H.value }),
      style: { overflow: "hidden" },
      id: `table_${E.value}`,
      onMouseenter: l[3] || (l[3] = () => u(oe)(!0)),
      onMouseleave: l[4] || (l[4] = () => u(oe)(!1))
    }, [
      e.value.title.text ? (v(), p("div", {
        key: 0,
        class: "vue-ui-table-sparkline-caption",
        style: s({ backgroundColor: e.value.title.backgroundColor })
      }, [
        f("div", {
          class: "atom-title",
          style: s({
            fontSize: `${e.value.title.fontSize}px`,
            fontWeight: e.value.title.bold ? "bold" : "normal",
            color: e.value.title.color,
            textAlign: e.value.title.textAlign
          })
        }, x(e.value.title.text), 5),
        e.value.title.subtitle.text ? (v(), p("div", {
          key: 0,
          class: "atom-subtitle",
          style: s({
            fontSize: `${e.value.title.subtitle.fontSize}px`,
            fontWeight: e.value.title.subtitle.bold ? "bold" : "normal",
            color: e.value.title.subtitle.color,
            textAlign: e.value.title.textAlign
          })
        }, x(e.value.title.subtitle.text), 5)) : b("", !0)
      ], 4)) : b("", !0),
      f("div", {
        style: { overflow: "auto" },
        onPointerleave: l[2] || (l[2] = (a) => {
          J.value = void 0, A.value = void 0;
        })
      }, [
        f("table", {
          class: "vue-ui-data-table",
          style: s({ fontFamily: e.value.fontFamily, position: "relative" })
        }, [
          f("thead", pt, [
            qe((v(), p("tr", {
              role: "row",
              class: "vue-ui-data-table__thead-row",
              style: s({
                backgroundColor: e.value.thead.backgroundColor,
                color: e.value.thead.color
              })
            }, [
              f("th", {
                role: "cell",
                style: s({
                  backgroundColor: e.value.thead.backgroundColor,
                  border: e.value.thead.outline,
                  textAlign: e.value.thead.textAlign,
                  fontWeight: e.value.thead.bold ? "bold" : "normal"
                }),
                class: "sticky-col-first"
              }, [
                f("div", {
                  style: s({
                    display: "flex",
                    flexDirection: "row",
                    alignItems: "center",
                    gap: "3px",
                    justifyContent: e.value.thead.textAlign
                  })
                }, [
                  f("span", null, x(e.value.translations.serie), 1),
                  m.value.length > 1 && e.value.sortedSeriesName ? (v(), p("div", ft, [
                    f("button", {
                      class: "vue-ui-table-sparkline-sorting-button vue-ui-table-sparkline-sorting-button-down",
                      onClick: l[0] || (l[0] = (a) => W({ type: "name" }, null, -1)),
                      style: s({
                        cursor: P.value ? "pointer" : "default"
                      })
                    }, [
                      N(u(F), {
                        size: 12,
                        name: "arrowBottom",
                        stroke: e.value.thead.color,
                        style: s({
                          opacity: y.value === "name" && w.value.name === -1 ? 1 : 0.3
                        })
                      }, null, 8, ["stroke", "style"])
                    ], 4),
                    f("button", {
                      class: "vue-ui-table-sparkline-sorting-button vue-ui-table-sparkline-sorting-button-up",
                      onClick: l[1] || (l[1] = (a) => W({ type: "name" }, null, 1)),
                      style: s({
                        cursor: P.value ? "pointer" : "default"
                      })
                    }, [
                      N(u(F), {
                        size: 12,
                        name: "arrowTop",
                        stroke: e.value.thead.color,
                        style: s({
                          opacity: y.value === "name" && w.value.name === 1 ? 1 : 0.3
                        })
                      }, null, 8, ["stroke", "style"])
                    ], 4)
                  ])) : b("", !0)
                ], 4)
              ], 4),
              (v(!0), p(Y, null, Z(_.value, (a, n) => (v(), p("th", {
                role: "cell",
                style: s({
                  background: e.value.thead.backgroundColor,
                  border: e.value.thead.outline,
                  textAlign: e.value.thead.textAlign,
                  fontWeight: e.value.thead.bold ? "bold" : "normal",
                  minWidth: n === _.value.length - 1 ? `${e.value.sparkline.dimensions.width}px` : "48px",
                  paddingRight: n === _.value.length - 1 && e.value.userOptions.show ? "36px" : ""
                }),
                class: Q({ "sticky-col": n === _.value.length - 1 && e.value.showSparklines })
              }, [
                f("div", {
                  style: s({
                    display: "flex",
                    flexDirection: "row",
                    alignItems: "center",
                    gap: "3px",
                    justifyContent: e.value.thead.textAlign
                  })
                }, [
                  f("span", null, x(a.value), 1),
                  m.value.length > 1 && (ge(n) || be(a)) ? (v(), p("div", mt, [
                    f("button", {
                      class: "vue-ui-table-sparkline-sorting-button vue-ui-table-sparkline-sorting-button-down",
                      onClick: () => W(a, n, -1),
                      style: s({
                        cursor: P.value ? "pointer" : "default"
                      })
                    }, [
                      N(u(F), {
                        size: 12,
                        name: "arrowBottom",
                        stroke: e.value.thead.color,
                        style: s({
                          opacity: ye(n, a, -1)
                        })
                      }, null, 8, ["stroke", "style"])
                    ], 12, gt),
                    f("button", {
                      class: "vue-ui-table-sparkline-sorting-button vue-ui-table-sparkline-sorting-button-up",
                      onClick: () => W(a, n, 1),
                      style: s({
                        cursor: P.value ? "pointer" : "default"
                      })
                    }, [
                      N(u(F), {
                        size: 12,
                        name: "arrowTop",
                        stroke: e.value.thead.color,
                        style: s({
                          opacity: ye(n, a, 1)
                        })
                      }, null, 8, ["stroke", "style"])
                    ], 12, bt)
                  ])) : b("", !0)
                ], 4),
                e.value.userOptions.show && n === _.value.length - 1 && (u(ne) || u(R)) ? (v(), Xe(u(Ie), {
                  ref_for: !0,
                  ref: "details",
                  key: `user_option_${ae.value}`,
                  backgroundColor: e.value.thead.backgroundColor,
                  color: e.value.thead.color,
                  isPrinting: u(ue),
                  isImaging: u(re),
                  uid: E.value,
                  hasPdf: e.value.userOptions.buttons.pdf,
                  hasXls: e.value.userOptions.buttons.csv,
                  hasImg: e.value.userOptions.buttons.img,
                  hasFullscreen: e.value.userOptions.buttons.fullscreen,
                  hasAltCopy: e.value.userOptions.buttons.altCopy,
                  isFullscreen: fe.value,
                  titles: { ...e.value.userOptions.buttonTitles },
                  chartElement: T.value,
                  position: e.value.userOptions.position,
                  callbacks: e.value.userOptions.callbacks,
                  printScale: e.value.userOptions.print.scale,
                  isCursorPointer: P.value,
                  onToggleFullscreen: Ue,
                  onGeneratePdf: u(ie),
                  onGenerateImage: Ee,
                  onGenerateCsv: me,
                  onCopyAlt: he,
                  style: s({
                    visibility: u(ne) ? u(R) ? "visible" : "hidden" : "visible"
                  })
                }, je({ _: 2 }, [
                  t.$slots.menuIcon ? {
                    name: "menuIcon",
                    fn: $(({ isOpen: r, color: o }) => [
                      S(t.$slots, "menuIcon", ee({ ref_for: !0 }, { isOpen: r, color: o }), void 0, !0)
                    ]),
                    key: "0"
                  } : void 0,
                  t.$slots.optionPdf ? {
                    name: "optionPdf",
                    fn: $(() => [
                      S(t.$slots, "optionPdf", {}, void 0, !0)
                    ]),
                    key: "1"
                  } : void 0,
                  t.$slots.optionCsv ? {
                    name: "optionCsv",
                    fn: $(() => [
                      S(t.$slots, "optionCsv", {}, void 0, !0)
                    ]),
                    key: "2"
                  } : void 0,
                  t.$slots.optionImg ? {
                    name: "optionImg",
                    fn: $(() => [
                      S(t.$slots, "optionImg", {}, void 0, !0)
                    ]),
                    key: "3"
                  } : void 0,
                  t.$slots.optionFullscreen ? {
                    name: "optionFullscreen",
                    fn: $(({ toggleFullscreen: r, isFullscreen: o }) => [
                      S(t.$slots, "optionFullscreen", ee({ ref_for: !0 }, { toggleFullscreen: r, isFullscreen: o }), void 0, !0)
                    ]),
                    key: "4"
                  } : void 0,
                  t.$slots.optionAltCopy ? {
                    name: "optionAltCopy",
                    fn: $(({ altCopy: r }) => [
                      S(t.$slots, "optionAltCopy", ee({ ref_for: !0 }, { altCopy: r }), void 0, !0)
                    ]),
                    key: "5"
                  } : void 0
                ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasXls", "hasImg", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "callbacks", "printScale", "isCursorPointer", "onGeneratePdf", "style"])) : b("", !0)
              ], 6))), 256))
            ], 4)), [
              [u(it), Be]
            ])
          ]),
          f("tbody", null, [
            (v(!0), p(Y, null, Z(m.value, (a, n) => (v(), p("tr", {
              role: "row",
              style: s({
                backgroundColor: e.value.tbody.backgroundColor,
                color: e.value.tbody.color
              }),
              class: Q({ "vue-ui-data-table__tbody__row": !0, "vue-ui-data-table__tbody__row-even": n % 2 === 0, "vue-ui-data-table__tbody__row-odd": n % 2 !== 0 })
            }, [
              f("td", {
                role: "cell",
                style: s({
                  backgroundColor: e.value.tbody.backgroundColor,
                  border: e.value.tbody.outline,
                  fontSize: `${e.value.tbody.fontSize}px`,
                  fontWeight: e.value.tbody.bold ? "bold" : "normal",
                  textAlign: e.value.tbody.textAlign
                }),
                "data-cell": e.value.translations.serie,
                class: "vue-ui-data-table__tbody__td sticky-col-first"
              }, [
                f("div", {
                  dir: "auto",
                  style: s({
                    display: "flex",
                    flexDirection: "row",
                    alignItems: "center",
                    gap: "6px",
                    justifyContent: e.value.tbody.textAlign
                  })
                }, [
                  e.value.tbody.showColorMarker ? (v(), p("span", {
                    key: 0,
                    style: s({ color: a.color })
                  }, "⬤", 4)) : b("", !0),
                  f("span", null, x(a.name ?? "-"), 1)
                ], 4)
              ], 12, yt),
              (v(!0), p(Y, null, Z(Ve.value, (r, o) => (v(), p("td", {
                dir: "auto",
                role: "cell",
                ref_for: !0,
                ref_key: "TD",
                ref: L,
                style: s({
                  border: e.value.tbody.outline,
                  fontSize: `${e.value.tbody.fontSize}px`,
                  fontWeight: e.value.tbody.bold ? "bold" : "normal",
                  textAlign: e.value.tbody.textAlign,
                  background: A.value !== void 0 && o === A.value ? e.value.tbody.selectedColor.useSerieColor ? `${a.color.length > 7 ? a.color.slice(0, -2) : a.color}33` : e.value.tbody.selectedColor.fallback : ""
                }),
                "data-cell": _.value[o] ? _.value[o].value : "",
                class: "vue-ui-data-table__tbody__td",
                onPointerenter: (d) => {
                  J.value = n, A.value = o;
                }
              }, x([null, void 0].includes(a.values[o]) ? "-" : u(U)(
                e.value.formatter,
                Number(a.values[o]),
                u(B)({
                  p: e.value.prefix,
                  v: Number(a.values[o]),
                  s: e.value.suffix,
                  r: e.value.roundingValues
                }),
                { datapoint: a, seriesIndex: n, datapointIndex: o }
              )), 45, ht))), 256)),
              e.value.showTotal ? (v(), p("td", {
                key: 0,
                dir: "auto",
                role: "cell",
                style: s({
                  border: e.value.tbody.outline,
                  fontSize: `${e.value.tbody.fontSize}px`,
                  fontWeight: e.value.tbody.bold ? "bold" : "normal",
                  textAlign: e.value.tbody.textAlign
                }),
                "data-cell": e.value.translations.total,
                class: "vue-ui-data-table__tbody__td"
              }, x(u(U)(
                e.value.formatter,
                a.sum,
                u(B)({
                  p: e.value.prefix,
                  v: a.sum,
                  s: e.value.suffix,
                  r: e.value.roundingTotal
                }),
                { datapoint: a.sum, seriesIndex: n }
              )), 13, kt)) : b("", !0),
              e.value.showAverage ? (v(), p("td", {
                key: 1,
                dir: "auto",
                role: "cell",
                style: s({
                  border: e.value.tbody.outline,
                  fontSize: `${e.value.tbody.fontSize}px`,
                  fontWeight: e.value.tbody.bold ? "bold" : "normal",
                  textAlign: e.value.tbody.textAlign
                }),
                "data-cell": e.value.translations.average,
                class: "vue-ui-data-table__tbody__td"
              }, x(u(U)(
                e.value.formatter,
                a.average,
                u(B)({
                  p: e.value.prefix,
                  v: a.average,
                  s: e.value.suffix,
                  r: e.value.roundingAverage
                }),
                { datapoint: a.average, seriesIndex: n }
              )), 13, Ct)) : b("", !0),
              e.value.showMedian ? (v(), p("td", {
                key: 2,
                dir: "auto",
                role: "cell",
                style: s({
                  border: e.value.tbody.outline,
                  fontSize: `${e.value.tbody.fontSize}px`,
                  fontWeight: e.value.tbody.bold ? "bold" : "normal",
                  textAlign: e.value.tbody.textAlign
                }),
                "data-cell": e.value.translations.median,
                class: "vue-ui-data-table__tbody__td"
              }, x(u(U)(
                e.value.formatter,
                a.median,
                u(B)({
                  p: e.value.prefix,
                  v: a.median,
                  s: e.value.suffix,
                  r: e.value.roundingMedian
                }),
                { datapoint: a.median, seriesIndex: n }
              )), 13, _t)) : b("", !0),
              e.value.showSparklines ? (v(), p("td", {
                key: 3,
                role: "cell",
                "data-cell": e.value.translations.chart,
                style: s({
                  border: e.value.tbody.outline,
                  fontSize: `${e.value.tbody.fontSize}px`,
                  fontWeight: e.value.tbody.bold ? "bold" : "normal",
                  textAlign: e.value.tbody.textAlign,
                  backgroundColor: e.value.tbody.backgroundColor,
                  padding: "0"
                }),
                class: "vue-ui-data-table__tbody__td sticky-col"
              }, [
                N(u(Se), {
                  onHoverIndex: ({ index: r }) => Me({ dataIndex: r, serieIndex: n }),
                  "height-ratio": e.value.sparkline.dimensions.heightRatio,
                  "forced-padding": 30,
                  dataset: a.sparklineDataset,
                  showInfo: !1,
                  selectedIndex: A.value,
                  config: {
                    type: e.value.sparkline.type,
                    style: {
                      backgroundColor: "transparent",
                      animation: {
                        show: e.value.sparkline.animation.show && !u(ue) && !u(re),
                        animationFrames: e.value.sparkline.animation.animationFrames
                      },
                      padding: {
                        right: 12
                      },
                      line: {
                        color: a.color,
                        smooth: e.value.sparkline.smooth,
                        strokeWidth: e.value.sparkline.strokeWidth
                      },
                      bar: {
                        color: a.color
                      },
                      area: {
                        color: a.color,
                        opacity: e.value.sparkline.showArea ? 16 : 0,
                        useGradient: e.value.sparkline.useGradient
                      },
                      verticalIndicator: {
                        color: a.color
                      },
                      plot: {
                        radius: 9,
                        stroke: e.value.tbody.backgroundColor,
                        strokeWidth: 3
                      }
                    }
                  }
                }, null, 8, ["onHoverIndex", "height-ratio", "dataset", "selectedIndex", "config"])
              ], 12, xt)) : b("", !0)
            ], 6))), 256))
          ])
        ], 4)
      ], 32),
      t.$slots.source ? (v(), p("div", wt, [
        S(t.$slots, "source", {}, void 0, !0)
      ], 512)) : b("", !0)
    ], 42, ct));
  }
}, Vt = /* @__PURE__ */ vt(At, [["__scopeId", "data-v-0c4dbd44"]]);
export {
  Vt as default
};
