// vue-data-ui/dist/vue-ui-gizmo-DEec3-v9.js
import { defineAsyncComponent as V, ref as k, useSlots as F, onMounted as y, computed as l, toRefs as S, watch as I, openBlock as o, createElementBlock as s, normalizeStyle as M, createVNode as U, unref as i, createElementVNode as t, createCommentVNode as r, Fragment as C, toDisplayString as x, renderSlot as P, createBlock as E } from "vue";
import { u as O } from "./useConfig-CIgyKZ6R.js";
import { j as T, q as j, t as w, k as b, l as z } from "./lib-GwPnJSrY.js";
import { u as q } from "./useNestedProp-BBT0mbHh.js";
import { u as R, B as H } from "./BaseScanner-BhhkKVoj.js";
import { _ as J } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const K = { class: "vue-data-ui-component vue-ui-gizmo-wrapper" }, Q = ["viewBox", "width"], W = { key: 0 }, X = ["id"], Y = ["stop-color"], Z = ["stop-color"], ee = ["stroke"], te = ["stroke"], oe = ["width", "fill"], ae = ["fill"], se = { key: 0 }, re = ["id"], ne = ["stroke"], le = ["stroke", "stroke-dasharray", "stroke-dashoffset"], ie = ["filter"], ue = ["stroke", "stroke-dasharray", "stroke-dashoffset"], de = ["fill"], ce = {
  __name: "vue-ui-gizmo",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Number
    }
  },
  setup(A) {
    const G = V(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_gizmo: L } = O(), a = A, u = k(T()), N = F();
    y(() => {
      g();
    }), y(() => {
      N["chart-background"] && d.value && console.warn("VueUiGizmo does not support the #chart-background slot.");
    });
    const d = l(() => e.value.debug);
    function g() {
      !a.dataset && a.dataset !== 0 && j({
        componentName: "VueUiGizmo",
        type: "dataset",
        debug: d.value
      }), a.dataset < 0 && d.value && console.warn("VueUiGizmo: dataset cannot be negative."), a.dataset > 100 && d.value && console.warn(`VueUiGizmo: gauge is maxed out, as dataset exceeds 100% (props.dataset = ${a.dataset})`);
    }
    const e = k(v()), $ = l(() => w({
      defaultConfig: {
        stroke: "#6A6A6A80",
        color: "#6A6A6A",
        gradientColor: "#CACACA",
        textColor: "transparent"
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: B, FINAL_DATASET: p } = R({
      ...S(a),
      FINAL_CONFIG: e,
      prepareConfig: v,
      dsIsNumber: !0,
      skeletonDataset: a.config?.skeletonDataset ?? 50,
      skeletonConfig: w({
        defaultConfig: e.value,
        userConfig: $.value
      })
    });
    function v() {
      return q({
        userConfig: a.config,
        defaultConfig: L
      });
    }
    I(() => a.config, (h) => {
      e.value = v(), g();
    }, { deep: !0 });
    const _ = l(() => ({
      battery: { width: e.value.showPercentage ? 150 : 80, height: 50 },
      gauge: { width: 80, height: 80 }
    })[e.value.type]), m = l(() => Math.min(Math.max(0, p.value), 100)), c = l(() => Math.max(0, p.value)), f = l(() => {
      const n = 2 * Math.PI * 35, D = n - m.value / 100 * n;
      return {
        dasharray: `${n} ${n}`,
        dashoffset: D
      };
    });
    return (h, n) => (o(), s("div", K, [
      (o(), s("svg", {
        class: "vue-ui-gizmo",
        viewBox: `0 0 ${_.value.width} ${_.value.height}`,
        width: e.value.size,
        style: M({
          background: "transparent",
          fontFamily: e.value.fontFamily
        })
      }, [
        U(i(G)),
        e.value.useGradient ? (o(), s("defs", W, [
          t("linearGradient", {
            id: `gizmo_gradient_${u.value}`,
            x1: "0%",
            x2: "100%",
            y1: "0%",
            y2: "0%"
          }, [
            t("stop", {
              offset: "0%",
              "stop-color": e.value.gradientColor
            }, null, 8, Y),
            t("stop", {
              offset: "100%",
              "stop-color": e.value.color
            }, null, 8, Z)
          ], 8, X)
        ])) : r("", !0),
        e.value.type === "battery" ? (o(), s(C, { key: 1 }, [
          t("path", {
            d: "M 5 10 L 5 40 C 5 43 7 45 9 45 L 65 45 C 68 45 70 43 70 40 L 70 38 C 73 38 75 38 75 33 L 75 17 C 75 12 73 12 70 12 L 70 10 C 70 7 68 5 65 5 L 10 5 C 7 5 5 7 5 10",
            stroke: e.value.stroke,
            "stroke-width": 2,
            fill: "none"
          }, null, 8, ee),
          t("path", {
            d: "M 70 16 L 70 34",
            stroke: e.value.stroke,
            "stroke-width": 2,
            fill: "none",
            style: { opacity: "0.5" }
          }, null, 8, te),
          t("rect", {
            x: 10,
            y: 10,
            height: 30,
            width: 55 * (m.value / 100),
            fill: e.value.useGradient ? `url(#gizmo_gradient_${u.value})` : e.value.color,
            rx: 2
          }, null, 8, oe),
          e.value.showPercentage ? (o(), s("text", {
            key: 0,
            x: 85,
            y: 32,
            "text-anchor": "start",
            "font-size": "20",
            fill: e.value.textColor
          }, x(i(b)(
            e.value.formatter,
            c.value,
            i(z)({
              v: c.value,
              s: "%"
            })
          )), 9, ae)) : r("", !0)
        ], 64)) : r("", !0),
        e.value.type === "gauge" ? (o(), s(C, { key: 2 }, [
          e.value.useGradient ? (o(), s("defs", se, [
            t("filter", {
              id: `blur_${u.value}`,
              x: "-50%",
              y: "-50%",
              width: "200%",
              height: "200%"
            }, [...n[0] || (n[0] = [
              t("feGaussianBlur", {
                in: "SourceGraphic",
                stdDeviation: 1
              }, null, -1)
            ])], 8, re)
          ])) : r("", !0),
          t("circle", {
            cx: 40,
            cy: 40,
            r: 35,
            stroke: e.value.stroke,
            "stroke-width": 8,
            fill: "none"
          }, null, 8, ne),
          t("circle", {
            cx: 40,
            cy: 40,
            r: 35,
            stroke: e.value.color,
            "stroke-width": 8,
            "stroke-dasharray": f.value.dasharray,
            "stroke-dashoffset": f.value.dashoffset,
            "stroke-linecap": "round",
            fill: "none",
            style: { transform: "rotate(-90deg)", "transform-origin": "50% 50%" }
          }, null, 8, le),
          e.value.useGradient ? (o(), s("g", {
            key: 1,
            filter: `url(#blur_${u.value})`
          }, [
            t("circle", {
              cx: 40,
              cy: 40,
              r: 35,
              stroke: e.value.gradientColor,
              "stroke-width": 1,
              fill: "none",
              "stroke-dasharray": f.value.dasharray,
              "stroke-dashoffset": f.value.dashoffset,
              style: { transform: "rotate(-90deg)", "transform-origin": "50% 50%" }
            }, null, 8, ue)
          ], 8, ie)) : r("", !0),
          e.value.showPercentage ? (o(), s("text", {
            key: 2,
            x: 40,
            y: 47,
            "text-anchor": "middle",
            "font-size": "20",
            fill: e.value.textColor
          }, x(i(b)(
            e.value.formatter,
            c.value,
            i(z)({
              v: c.value,
              s: "%"
            })
          )), 9, de)) : r("", !0)
        ], 64)) : r("", !0)
      ], 12, Q)),
      P(h.$slots, "skeleton", {}, () => [
        i(B) ? (o(), E(H, { key: 0 })) : r("", !0)
      ], !0)
    ]));
  }
}, me = /* @__PURE__ */ J(ce, [["__scopeId", "data-v-c88b56eb"]]);
export {
  me as default
};
