// vue-data-ui/dist/vue-ui-ridgeline-CwvDh5pG.js
import { defineAsyncComponent as N, computed as v, ref as i, shallowRef as ft, toRefs as ka, onBeforeUnmount as gt, watch as ae, nextTick as he, watchEffect as yt, openBlock as n, createElementBlock as r, unref as u, normalizeStyle as C, normalizeClass as fe, createBlock as _, withCtx as c, renderSlot as d, normalizeProps as P, guardReactiveProps as S, createCommentVNode as h, createSlots as Ie, createVNode as Re, createElementVNode as y, Fragment as m, renderList as A, mergeProps as Ne, toDisplayString as V, resolveDynamicComponent as Ca, createTextVNode as Fe } from "vue";
import { j as Ee, t as pt, n as Aa, s as wa, p as ge, as as mt, P as $a, b as bt, c as _a, ai as Pa, X as Sa, W as Ta, Z as Oa, k as ye, l as pe, m as La, q as Be, r as xt, o as Ma, a0 as za, a1 as Da, ab as Ia } from "./lib-GwPnJSrY.js";
import { t as kt, u as Ra } from "./useResponsive-DfdjqQps.js";
import { u as Na } from "./useConfig-CIgyKZ6R.js";
import { u as Fa, B as Ea } from "./BaseScanner-BhhkKVoj.js";
import { u as Ba } from "./usePrinter-DiK5R6k-.js";
import { u as Xa } from "./useSvgExport-ByUukOZt.js";
import { u as Xe } from "./useNestedProp-BBT0mbHh.js";
import { u as Ha } from "./useTimeLabels-C-A2uZKi.js";
import { u as Ua } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Va } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Wa } from "./useChartAccessibility-9icAAmYg.js";
import { u as Ga } from "./useTimeLabelCollider-CIsgDrl9.js";
import ja from "./vue_ui_ridgeline-j4rVmpEp.js";
import Ya from "./Legend-1LMMsPxL.js";
import qa from "./Title-BfyrP-jp.js";
import { _ as Za } from "./Shape-CSyWYgXc.js";
import Ja from "./img-BYYzY_QU.js";
import { B as Ka } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Qa } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const el = ["id"], tl = ["xmlns", "viewBox"], al = ["id"], ll = ["stop-color"], ol = ["stop-color"], sl = ["stop-color"], nl = ["stop-color"], rl = ["id"], ul = ["stop-color"], il = ["stop-color"], cl = ["stop-color"], vl = ["stop-color"], dl = { key: 0 }, hl = ["fill", "d"], fl = ["stroke", "stroke-width", "d"], gl = ["fill", "d"], yl = ["stroke", "stroke-dasharray", "stroke-width", "d"], pl = ["x1", "y1", "x2", "y2", "stroke", "stroke-width", "stroke-dasharray"], ml = ["cx", "cy", "stroke", "stroke-width", "r", "fill"], bl = ["x", "y", "font-size", "font-weight", "fill", "text-decoration", "onMouseenter", "onClick"], xl = { key: 0 }, kl = ["font-size", "fill", "font-weight", "transform", "text-anchor"], Cl = ["font-size", "fill", "font-weight", "transform", "text-anchor", "innerHTML"], Al = ["x", "y", "width", "height", "onMouseenter", "onMouseleave", "onClick"], wl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], $l = ["cx", "cy", "stroke", "stroke-width", "r", "fill"], _l = ["x", "y", "text-anchor", "font-size", "fill"], Pl = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Sl = ["onClick"], Tl = ["innerHTML"], Ol = {
  __name: "vue-ui-ridgeline",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "selectDatapoint", "selectX", "copyAlt"],
  setup(Ct, { expose: At, emit: wt }) {
    const $t = N(() => import("./vue-ui-xy-f67V1xjd.js")), _t = N(() => import("./BaseIcon-CNx8Tzl5.js")), Pt = N(() => import("./vue-ui-accordion-BGHTqIBt.js")), St = N(() => import("./DataTable-Bo1pmAqG.js")), Tt = N(() => import("./PenAndPaper-DWV1FMIi.js")), Ot = N(() => import("./UserOptions-CrLqEimB.js")), Lt = N(() => import("./PackageVersion-DN9d0ME5.js")), He = N(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_ridgeline: Mt } = Na(), { isThemeValid: zt, warnInvalidTheme: Dt } = Ua(), M = Ct, q = v({
      get() {
        return Array.isArray(E.value) && E.value.length > 0;
      },
      set(t) {
        return t;
      }
    }), le = wt, T = i(null), Ue = i(null), Ve = i(null), z = ft(null), F = ft(null), We = i(null), Ge = i(null), me = i(0), It = i(0), be = i(0), w = i(Ee()), je = i(0), Rt = i(0), Ye = i(512), g = i(null), W = i(null), xe = i(null), qe = i(0), ke = i(null), Z = i(null), Ce = i(null), oe = i(!1), se = i(!1);
    function Ae() {
      const t = Xe({
        userConfig: M.config,
        defaultConfig: Mt
      }), l = t.theme;
      if (!l) return t;
      if (!zt.value(t))
        return Dt(t), t;
      const a = Xe({
        userConfig: ja[l] || M.config,
        defaultConfig: t
      }), s = Xe({
        userConfig: M.config,
        defaultConfig: a
      });
      return {
        ...s,
        customPalette: s.customPalette.length ? s.customPalette : La[l] || ge
      };
    }
    const e = i(Ae()), O = v(() => e.value.userOptions.useCursorPointer), Nt = v(() => pt({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            areas: {
              maxPoint: {
                show: !1
              },
              opacity: 0.9,
              stroke: {
                useSerieColor: !0
              }
            },
            legend: {
              backgroundColor: "transparent"
            },
            padding: {
              right: -24,
              left: 0
            },
            xAxis: {
              labels: {
                values: []
              }
            },
            yAxis: {
              labels: {
                fontSize: 0
              }
            },
            zeroLine: { show: !1 }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: G, FINAL_DATASET: E, manualLoading: we } = Fa({
      ...ka(M),
      FINAL_CONFIG: e,
      prepareConfig: Ae,
      skeletonDataset: M.config?.skeletonDataset ?? [
        {
          name: "_",
          datapoints: [
            {
              name: "__",
              color: "#999999",
              values: [28.639, 32.04, 41.134, 44.525, 21.151, 2.436, 0.218, 0.024, 2e-3, 0, 0, 0]
            },
            {
              name: "_",
              color: "#CACACA",
              values: [13.253, 15.621, 23.36, 33.698, 29.935, 10.874, 2.364, 0.561, 0.107, 0.02, 6e-3, 4e-3]
            }
          ]
        },
        {
          name: "_",
          datapoints: [
            {
              name: "_",
              color: "#999999",
              values: [10.851, 13.195, 21.617, 36.556, 42.292, 21.006, 3.398, 0.223, 0.013, 1e-3, 0, 0]
            },
            {
              name: "_",
              color: "#CACACA",
              values: [3.171, 4.115, 8.108, 18.248, 31.641, 29.063, 12.031, 2.742, 0.504, 0.102, 0.032, 0.021]
            }
          ]
        },
        {
          name: "_",
          datapoints: [
            {
              name: "_",
              color: "#999999",
              values: [1.731, 2.334, 5.125, 13.626, 29.911, 38.524, 24.168, 7.646, 1.575, 0.317, 0.097, 0.063]
            },
            {
              name: "_",
              color: "#CACACA",
              values: [0.25, 0.367, 1.026, 3.944, 13.635, 28.891, 30.149, 15.419, 4.714, 1.246, 0.442, 0.299]
            }
          ]
        },
        {
          name: "_",
          datapoints: [
            {
              name: "_",
              color: "#999999",
              values: [0.034, 0.054, 0.194, 1.065, 5.747, 20.735, 38.306, 32.899, 15.318, 5.566, 2.422, 1.76]
            },
            {
              name: "_",
              color: "#CACACA",
              values: [1e-3, 2e-3, 9e-3, 0.095, 1.124, 8.342, 27.115, 35.08, 21.449, 9.093, 4.243, 3.143]
            }
          ]
        },
        {
          name: "_",
          datapoints: [
            {
              name: "_",
              color: "#999999",
              values: [0, 1e-3, 4e-3, 0.051, 0.567, 3.322, 14.215, 44.783, 40.351, 20.377, 9.866, 7.378]
            },
            {
              name: "_",
              color: "#CACACA",
              values: [0, 0, 0, 0, 1e-3, 0.11, 4.136, 27.498, 43.24, 29.807, 17.345, 13.678]
            }
          ]
        },
        {
          name: "_",
          datapoints: [
            {
              name: "_",
              color: "#999999",
              values: [0, 0, 0, 0, 0.025, 0.598, 3.886, 10.645, 54.479, 45.953, 30.814, 24.55]
            },
            {
              name: "_",
              color: "#CACACA",
              values: [0, 0, 0, 0, 0, 0, 7e-3, 1.655, 26.63, 52.017, 45.192, 39.651]
            }
          ]
        }
      ],
      skeletonConfig: pt({
        defaultConfig: e.value,
        userConfig: Nt.value
      })
    }), $ = i(Math.min(
      e.value.style.chart.areas.height,
      e.value.style.chart.areas.rowHeight
    )), {
      userOptionsVisible: ne,
      setUserOptionsVisibility: Ze,
      keepUserOptionState: Je
    } = Va({ config: e.value }), { svgRef: re } = Wa({ config: e.value.style.chart.title }), $e = v(() => !!e.value.debug);
    function Ke() {
      const t = E.value || [];
      if (!Array.isArray(t) || t.length === 0) {
        Be({ componentName: "VueUiRidgeline", type: "dataset", debug: $e.value }), we.value = !0;
        return;
      }
      if (t.forEach((l, a) => {
        xt({
          datasetObject: l,
          requiredAttributes: ["name", "datapoints"]
        }).forEach((s) => {
          q.value = !1, Be({
            componentName: "VueUiRidgeline",
            type: "datasetSerieAttribute",
            property: s,
            index: a,
            debug: $e.value
          });
        }), Array.isArray(l.datapoints) && l.datapoints.length && l.datapoints.forEach((s, o) => {
          xt({
            datasetObject: s,
            requiredAttributes: ["name", "values"]
          }).forEach((f) => {
            q.value = !1, Be({
              componentName: "VueUiRidgeline",
              type: "datasetSerieAttribute",
              property: `datapoint.${f}`,
              index: `${a}-${o}`,
              debug: $e.value
            });
          });
        });
      }), Rt.value = t.length, $.value = Math.min(
        e.value.style.chart.areas.height,
        e.value.style.chart.areas.rowHeight
      ), Ma(M.dataset) || (we.value = e.value.loading), e.value.responsive) {
        const l = kt(() => {
          const { width: a, height: s } = Ra({
            chart: T.value,
            title: e.value.style.chart.title.text ? Ue.value : null,
            legend: e.value.style.chart.legend.show ? Ve.value : null,
            source: We.value,
            noTitle: Ge.value,
            padding: e.value.style.chart.padding
          });
          requestAnimationFrame(() => {
            Ye.value = a, $.value = t.length ? s / t.length : 0, qe.value = s - 12;
          });
        });
        z.value && (F.value && z.value.unobserve(F.value), z.value.disconnect()), z.value = new ResizeObserver(l), F.value = T.value?.parentNode || null, F.value && z.value.observe(F.value);
      }
    }
    gt(() => {
      z.value && (F.value && z.value.unobserve(F.value), z.value.disconnect());
    });
    const { isPrinting: _e, isImaging: Pe, generatePdf: Qe, generateImage: et } = Ba({
      elementId: `vue-ui-ridgeline_${w.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-ridgeline",
      options: e.value.userOptions.print
    }), Ft = v(() => e.value.userOptions.show && !e.value.style.chart.title.text), Et = v(() => Aa(e.value.customPalette)), B = i({
      showTable: e.value.table.show
    });
    ae(
      () => E.value,
      async (t) => {
        Array.isArray(t) && t.length && (ne.value = !e.value.userOptions.showOnChartHover, await he(), Ke(), me.value += 1, be.value += 1, B.value.showTable = e.value.table.show);
      },
      { deep: !0, immediate: !0 }
    ), ae(() => M.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (we.value = !1);
    }, { deep: !0 }), ae(
      () => G.value,
      async (t) => {
        t || (await he(), Ke());
      },
      { immediate: !0 }
    ), ae(
      () => M.config,
      () => {
        e.value = Ae(E.value || null), ne.value = !e.value.userOptions.showOnChartHover, $.value = Math.min(e.value.style.chart.areas.height, e.value.style.chart.areas.rowHeight), B.value.showTable = e.value.table.show, me.value += 1, be.value += 1;
      },
      { deep: !0 }
    );
    const Se = v(() => e.value.style.chart.areas.height / e.value.style.chart.areas.rowHeight);
    function Bt() {
      b.value.length ? b.value = [] : R.value.forEach((t) => {
        b.value.push(t.id);
      });
    }
    function ue(t) {
      b.value.includes(t) ? b.value = b.value.filter((l) => l !== t) : b.value.push(t), le("selectLegend", I.value);
    }
    function tt(t) {
      if (!R.value.length)
        return e.value.debug && console.warn("VueUiRidgeline - There are no series to show."), null;
      const l = R.value.find((a) => a.name === t);
      return l || (e.value.debug && console.warn(`VueUiRidgeline - Series name not found "${t}"`), null);
    }
    function Xt(t) {
      const l = tt(t);
      l !== null && b.value.includes(l.id) && ue(l.id);
    }
    function Ht(t) {
      const l = tt(t);
      l !== null && (b.value.includes(l.id) || ue(l.id));
    }
    const D = v(() => q.value ? (E.value || []).map((t) => ({
      ...t,
      labelLen: Zt(t.name, e.value.style.chart.yAxis.labels.fontSize),
      uid: Ee(),
      datapoints: t.datapoints.map((l, a) => {
        const s = l.color ? wa(l.color) : Et.value[a] || ge[a] || ge[a % ge.length], o = mt(l.name);
        return { ...l, color: s, id: o };
      })
    })) : []), at = v(() => e.value.style.chart.padding.top + $.value * (E.value || []).length + $.value * Se.value + e.value.style.chart.padding.bottom), p = v(() => {
      const t = e.value.style.chart.padding;
      return {
        width: Ye.value,
        height: at.value,
        padding: t
      };
    }), Ut = v(() => p.value.width), Vt = v(() => p.value.height), Te = i(0), Wt = kt((t) => {
      Te.value = t;
    }, 100);
    yt((t) => {
      const l = ke.value;
      if (!l) return;
      const a = new ResizeObserver((s) => {
        Wt(s[0].contentRect.height);
      });
      a.observe(l), t(() => a.disconnect());
    }), gt(() => {
      Te.value = 0;
    });
    const x = v(() => ({
      fullHeight: p.value.height + Te.value,
      top: p.value.padding.top,
      left: p.value.padding.left,
      right: p.value.width - p.value.padding.right,
      bottom: p.value.height - e.value.style.chart.padding.bottom,
      width: p.value.width - (p.value.padding.left + p.value.padding.right)
    })), J = v(() => Math.max(...D.value.flatMap((t) => t.datapoints.map((l) => l.values.length)))), j = i([]);
    let lt = 0;
    yt(() => {
      const t = ++lt;
      (async () => {
        const l = await Ha({
          values: e.value.style.chart.xAxis.labels.values,
          maxDatapoints: J.value,
          formatter: e.value.style.chart.xAxis.labels.datetimeFormatter,
          start: 0,
          end: e.value.style.chart.xAxis.labels.values.length
        });
        t === lt && (j.value = l);
      })();
    });
    const Y = v(() => {
      const t = Math.max(...D.value.map((o) => o.labelLen)), l = p.value.padding.left + t + 16 + e.value.style.chart.yAxis.labels.offsetX, a = (x.value.width - l) / J.value, s = [];
      for (let o = 0; o < J.value; o += 1)
        s.push({
          selectorX: l + a * o,
          x: l + a * o - a / 2,
          y: x.value.top,
          label: e.value.style.chart.xAxis.labels.values[o] ? j.value[o]?.text ?? "" : "",
          index: o,
          width: a,
          height: at.value
        });
      return s;
    });
    function Oe(t) {
      return D.value.map((l) => l.datapoints.map((a) => ({
        dp: a,
        selected: a.values[t.index]
      })));
    }
    function Gt(t) {
      g.value = t;
      const l = Oe(t);
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: l, seriesIndex: t.index });
    }
    function jt(t) {
      g.value = null;
      const l = Oe(t);
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: t.index });
    }
    function Yt(t) {
      const l = Oe(t);
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: t.index }), le("selectX", l);
    }
    function qt(t, l) {
      return t.length * l / 2 + l;
    }
    function ot(t, l, a) {
      const s = qt(l, a);
      return t + s > x.value.right;
    }
    function Zt(t, l, a = "sans-serif") {
      const o = document.createElement("canvas").getContext("2d"), f = typeof l == "number" ? `${l}px` : l;
      return o.font = `${f} ${a}`, o.measureText(t).width;
    }
    const I = v(() => {
      $.value;
      const t = Math.max(...D.value.map((k) => k.labelLen)), l = Math.max(...D.value.flatMap((k) => k.datapoints.flatMap((Q) => Q.values))), a = Math.min(...D.value.flatMap((k) => k.datapoints.flatMap((Q) => Q.values))), s = p.value.padding.left + t + 16 + e.value.style.chart.yAxis.labels.offsetX, o = (x.value.width - s) / J.value, f = Math.abs(Math.min(a, 0)), L = l + f;
      function ct(k) {
        return isNaN(k / L) ? 0 : k / L;
      }
      return D.value.map((k, Q) => {
        const vt = x.value.top + $.value * Q, H = x.value.top + vt + $.value * Se.value * (1 - ct(f));
        return {
          ...k,
          label: {
            x: s - e.value.style.chart.yAxis.labels.fontSize,
            y: H
          },
          datapoints: k.datapoints.map((ee) => {
            const U = ee.values.map((te, ya) => {
              const pa = isNaN(te) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(te) ? 0 : te || 0, ma = s + o * ya, ba = x.value.top + vt + $.value * Se.value * (1 - ct(pa + f)), xa = te === Math.max(...ee.values);
              return {
                x: ma,
                y: ba,
                value: te,
                isMaxPoint: xa,
                zero: H
              };
            }), da = `${$a(U, H, !1, !1)}`, ha = `M ${s},${H} ${bt(U)} ${U.at(-1).x},${H}`, fa = `M ${s},${H} ${U.at(-1).x},${H}`, dt = `M ${_a(U)}`, ht = `M ${bt(U)}`, ga = Pa(
              e.value.style.chart.areas.smooth ? dt : ht
            );
            return {
              ...ee,
              uid: Ee(),
              plots: U,
              smoothPath: da,
              straightPath: ha,
              zeroPath: fa,
              pathLength: ga,
              smoothPathRidge: dt,
              straightPathRidge: ht
            };
          }).filter((ee) => !b.value.includes(ee.id))
        };
      });
    }), b = i([]);
    function Jt(t) {
      const l = /* @__PURE__ */ new Map();
      return t.forEach((a) => {
        a.datapoints.forEach((s, o) => {
          const f = mt(s.name);
          l.has(f) || l.set(f, {
            id: f,
            name: s.name,
            color: s.color,
            shape: "circle",
            segregate: () => ue(f),
            isSegregated: b.value.includes(f),
            opacity: b.value.includes(f) ? 0.5 : 1
          });
        });
      }), Array.from(l.values());
    }
    const R = v(() => Jt(D.value)), Kt = v(() => ({
      cy: "donut-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), X = i(!1);
    function st(t) {
      X.value = t, je.value += 1;
    }
    const ie = i({}), ce = i([]);
    function Qt(t) {
      le("selectDatapoint", t), e.value.style.chart.dialog.show && (ce.value = t.datapoints.map((l) => ({
        name: l.name,
        color: l.color,
        type: "line",
        useArea: !0,
        smooth: e.value.style.chart.areas.smooth,
        series: l.values
      })), W.value = t, ie.value = Ia({
        ...e.value.style.chart.dialog.xyChart,
        responsive: !0,
        // Overriding 
        chart: {
          ...e.value.style.chart.dialog.xyChart.chart,
          grid: {
            ...e.value.style.chart.dialog.xyChart.chart.grid,
            labels: {
              ...e.value.style.chart.dialog.xyChart.chart.grid.labels,
              xAxisLabels: {
                ...e.value.style.chart.dialog.xyChart.chart.grid.labels.xAxisLabels,
                values: e.value.style.chart.xAxis.labels.values,
                // Overriding
                autoRotate: {
                  enable: !0,
                  angle: e.value.style.chart.dialog.xyChart.chart.grid.labels.xAxisLabels.autoRotate.angle
                },
                // overriding
                datetimeFormatter: e.value.style.chart.xAxis.labels.datetimeFormatter
                // Overriding
              }
            }
          },
          tooltip: {
            ...e.value.style.chart.dialog.xyChart.chart.tooltip,
            showTimeLabel: e.value.style.chart.xAxis.labels.values.length > 0
            // Overriding
          },
          userOptions: {
            ...e.value.style.chart.dialog.xyChart.chart.userOptions,
            buttons: {
              ...e.value.style.chart.dialog.xyChart.chart.userOptions.buttons,
              altCopy: e.value.userOptions.buttons.altCopy
            },
            callbacks: {
              ...e.value.style.chart.dialog.xyChart.chart.userOptions.callbacks,
              altCopy: () => {
              }
            },
            useCursorPointer: e.value.userOptions.useCursorPointer
            // Overriding
          }
        }
      }), xe.value && xe.value.open());
    }
    const Le = i(null);
    function ea(t) {
      Le.value = t;
    }
    function ta() {
      Le.value = null;
    }
    const ve = i(!1);
    function Me() {
      ve.value = !ve.value;
    }
    function nt() {
      B.value.showTable = !B.value.showTable;
    }
    const rt = v(() => ({ body: I.value.flatMap((a) => a.datapoints.flatMap((s) => ({
      ...s,
      rowName: `${a.name}: ${s.name}`
    }))).map((a) => [
      {
        name: a.rowName,
        color: a.color
      },
      ...a.values
    ]) })), de = v(() => {
      const t = [
        e.value.table.columnNames.series,
        ...j.value.map((s) => s.text)
      ], l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.series,
          ...j.value.map((s) => s.text)
        ],
        head: t,
        body: rt.value.body,
        config: l
      };
    });
    function ze(t = null) {
      he(() => {
        const l = [
          [e.value.table.columnNames.series, ...j.map((o) => [o.text])],
          ...rt.value.body.map((o, f) => [o[0].name, ...o.slice(1)])
        ], a = [
          [e.value.style.chart.title.text],
          [e.value.style.chart.title.subtitle.text]
        ].concat(l), s = za(a);
        t ? t(s) : Da({
          csvContent: s,
          title: e.value.style.chart.title.text || "vue-ui-ridgeline"
        });
      });
    }
    function aa() {
      return I.value;
    }
    async function la({ scale: t = 2 } = {}) {
      if (!T.value) return;
      const { width: l, height: a } = T.value.getBoundingClientRect(), s = l / a, { imageUri: o, base64: f } = await Ja({ domElement: T.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: o,
        base64: f,
        title: e.value.style.chart.title.text,
        width: l,
        height: a,
        aspectRatio: s
      };
    }
    const oa = v(() => ({ min: 0, max: J.value }));
    Ga({
      timeLabelsEls: ke,
      timeLabels: j,
      slicer: oa,
      configRef: e,
      rotationPath: ["style", "chart", "xAxis", "labels", "rotation"],
      autoRotatePath: ["style", "chart", "xAxis", "labels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: Ut,
      height: Vt,
      targetClass: ".vue-ui-ridgeline-x-axis-label",
      angle: e.value.style.chart.xAxis.labels.autoRotate.angle
    });
    const K = v(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = B.value.showTable;
      return {
        component: t ? He : Pt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: X.value,
          fullscreenParent: T.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: O.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ae(() => B.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && Z.value ? Z.value.open() : "close" in Z.value && Z.value.close());
    });
    function ut() {
      B.value.showTable = !1, Ce.value && Ce.value.setTableIconState(!1);
    }
    const sa = v(() => e.value.style.chart.backgroundColor), na = v(() => ({
      ...e.value.style.chart.legend,
      position: "bottom"
    })), ra = v(() => e.value.style.chart.title), { exportSvg: ua, getSvg: ia } = Xa({
      svg: re,
      title: ra,
      legend: na,
      legendItems: R,
      backgroundColor: sa
    });
    async function it({ isCb: t }) {
      se.value = !0, await he();
      try {
        if (t) {
          const { blob: l, url: a, text: s, dataUrl: o } = await ia();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: a, text: s, dataUrl: o }));
        } else
          await Promise.resolve(ua());
      } finally {
        se.value = !1;
      }
    }
    function ca(t) {
      if (t?.stage === "start") {
        oe.value = !0;
        return;
      }
      if (t?.stage === "end") {
        oe.value = !1;
        return;
      }
      et();
    }
    function va(t) {
      De("xy-zoom", t.dataset, t.config);
    }
    async function De(t = "main-chart", l = null, a = null) {
      if (le("copyAlt", {
        source: t,
        config: t === "main-chart" ? e.value : ie.value,
        dataset: t === "main-chart" ? I.value : ce.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        source: t,
        config: t === "main-chart" ? e.value : ie.value,
        dataset: t === "main-chart" ? I.value : ce.value
      }));
    }
    return At({
      getData: aa,
      getImage: la,
      generateImage: et,
      generateSvg: it,
      generatePdf: Qe,
      generateCsv: ze,
      hideSeries: Ht,
      showSeries: Xt,
      toggleAnnotator: Me,
      toggleTable: nt,
      toggleFullscreen: st,
      copyAlt: De
    }), (t, l) => (n(), r("div", {
      ref_key: "ridgelineChart",
      ref: T,
      class: fe(`vue-data-ui-component vue-ui-ridgeline ${X.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      id: `vue-ui-ridgeline_${w.value}`,
      style: C({
        fontFamily: e.value.style.fontFamily,
        width: "100%",
        textAlign: "center",
        background: e.value.style.chart.backgroundColor,
        height: e.value.responsive ? "100%" : void 0
      }),
      onMouseenter: l[3] || (l[3] = () => u(Ze)(!0)),
      onMouseleave: l[4] || (l[4] = () => u(Ze)(!1))
    }, [
      e.value.userOptions.buttons.annotator && u(re) ? (n(), _(u(Tt), {
        key: 0,
        color: e.value.style.chart.color,
        backgroundColor: e.value.style.chart.backgroundColor,
        active: ve.value,
        svgRef: u(re),
        isCursorPointer: O.value,
        onClose: Me
      }, {
        "annotator-action-close": c(() => [
          d(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": c(({ color: a }) => [
          d(t.$slots, "annotator-action-color", P(S({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": c(({ mode: a }) => [
          d(t.$slots, "annotator-action-draw", P(S({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": c(({ disabled: a }) => [
          d(t.$slots, "annotator-action-undo", P(S({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": c(({ disabled: a }) => [
          d(t.$slots, "annotator-action-redo", P(S({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": c(({ disabled: a }) => [
          d(t.$slots, "annotator-action-delete", P(S({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["color", "backgroundColor", "active", "svgRef", "isCursorPointer"])) : h("", !0),
      d(t.$slots, "userConfig", {}, void 0, !0),
      Ft.value ? (n(), r("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Ge,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : h("", !0),
      e.value.style.chart.title.text ? (n(), r("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Ue,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (n(), _(qa, {
          key: `title_${me.value}`,
          config: {
            title: {
              cy: "ridgeline-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "ridgeline-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : h("", !0),
      e.value.userOptions.show && q.value && (u(Je) || u(ne)) ? (n(), _(u(Ot), {
        ref_key: "userOptionsRef",
        ref: Ce,
        key: `user_option_${je.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: u(_e),
        isImaging: u(Pe),
        uid: w.value,
        hasTooltip: !1,
        callbacks: e.value.userOptions.callbacks,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: !1,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: X.value,
        printScale: e.value.userOptions.print.scale,
        chartElement: T.value,
        position: e.value.userOptions.position,
        isTooltip: !1,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: ve.value,
        tableDialog: e.value.table.useDialog,
        style: C({ visibility: u(Je) ? u(ne) ? "visible" : "hidden" : "visible" }),
        isCursorPointer: O.value,
        onToggleFullscreen: st,
        onGeneratePdf: u(Qe),
        onGenerateCsv: ze,
        onGenerateImage: ca,
        onGenerateSvg: it,
        onToggleTable: nt,
        onToggleAnnotator: Me,
        onCopyAlt: De
      }, Ie({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: c(({ isOpen: a, color: s }) => [
            d(t.$slots, "menuIcon", P(S({ isOpen: a, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: c(() => [
            d(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: c(() => [
            d(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: c(() => [
            d(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: c(() => [
            d(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: c(() => [
            d(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: c(() => [
            d(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: c(({ toggleFullscreen: a, isFullscreen: s }) => [
            d(t.$slots, "optionFullscreen", P(S({ toggleFullscreen: a, isFullscreen: s })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: c(({ toggleAnnotator: a, isAnnotator: s }) => [
            d(t.$slots, "optionAnnotator", P(S({ toggleAnnotator: a, isAnnotator: s })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: c(({ altCopy: a }) => [
            d(t.$slots, "optionAltCopy", P(S({ altCopy: a })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "callbacks", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "printScale", "chartElement", "position", "titles", "hasAnnotator", "isAnnotation", "tableDialog", "style", "isCursorPointer", "onGeneratePdf"])) : h("", !0),
      (n(), r("svg", {
        ref_key: "svgRef",
        ref: re,
        xmlns: u(Sa),
        class: fe({ "vue-data-ui-fullscreen--on": X.value, "vue-data-ui-fulscreen--off": !X.value }),
        viewBox: `0 0 ${p.value.width <= 0 ? 10 : p.value.width} ${x.value.fullHeight <= 0 ? 10 : x.value.fullHeight}`,
        style: C(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color};${e.value.responsive ? `height: ${qe.value}px; width: 100%;` : ""}`)
      }, [
        Re(u(Lt)),
        y("defs", null, [
          (n(!0), r(m, null, A(R.value, (a, s) => (n(), r("linearGradient", {
            id: `gradient-${a.id}-${w.value}`,
            x1: "50%",
            y1: "0%",
            x2: "50%",
            y2: "100%"
          }, [
            y("stop", {
              offset: "0%",
              "stop-color": a.color,
              "stop-opacity": "1"
            }, null, 8, ll),
            y("stop", {
              offset: "30%",
              "stop-color": a.color,
              "stop-opacity": "0.7"
            }, null, 8, ol),
            y("stop", {
              offset: "70%",
              "stop-color": a.color,
              "stop-opacity": "0.3"
            }, null, 8, sl),
            y("stop", {
              offset: "100%",
              "stop-color": a.color,
              "stop-opacity": "0.1"
            }, null, 8, nl)
          ], 8, al))), 256)),
          (n(!0), r(m, null, A(I.value, (a, s) => (n(), r(m, null, [
            (n(!0), r(m, null, A(a.datapoints, (o, f) => (n(), r("linearGradient", {
              key: `grad${o.id}`,
              id: `gradient-single-${w.value}-${o.uid}`,
              x1: "50%",
              y1: "0%",
              x2: "50%",
              y2: "100%"
            }, [
              y("stop", {
                offset: "0%",
                "stop-color": o.color,
                "stop-opacity": "1"
              }, null, 8, ul),
              y("stop", {
                offset: "30%",
                "stop-color": o.color,
                "stop-opacity": "0.7"
              }, null, 8, il),
              y("stop", {
                offset: "70%",
                "stop-color": o.color,
                "stop-opacity": "0.3"
              }, null, 8, cl),
              y("stop", {
                offset: "100%",
                "stop-color": o.color,
                "stop-opacity": "0.1"
              }, null, 8, vl)
            ], 8, rl))), 128))
          ], 64))), 256))
        ]),
        (n(!0), r(m, null, A(I.value, (a, s) => (n(), r("g", {
          key: `ds-${s}`
        }, [
          (n(!0), r(m, null, A(a.datapoints, (o, f) => (n(), r("g", {
            key: o.id
          }, [
            t.$slots.pattern ? (n(), r("g", dl, [
              y("defs", null, [
                d(t.$slots, "pattern", Ne({ ref_for: !0 }, { datapointIndex: s, seriesIndex: f, patternId: `pattern_${w.value}_${o.uid}` }), void 0, !0)
              ])
            ])) : h("", !0),
            y("path", {
              fill: t.$slots.pattern ? `url(#pattern_${w.value}_${o.uid})` : e.value.style.chart.backgroundColor,
              stroke: "none",
              "stroke-linecap": "round",
              d: e.value.style.chart.areas.smooth ? o.smoothPath : o.straightPath,
              style: C({
                opacity: e.value.style.chart.areas.opacity
              })
            }, null, 12, hl),
            y("path", {
              fill: "none",
              stroke: e.value.style.chart.areas.stroke.useSerieColor ? o.color : e.value.style.chart.areas.stroke.color,
              "stroke-width": e.value.style.chart.areas.strokeWidth,
              d: e.value.style.chart.areas.smooth ? o.smoothPathRidge : o.straightPathRidge,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: fe({ "vue-ui-ridgeline-animate": e.value.useCssAnimation && !u(G) }),
              style: C({
                strokeDasharray: o.pathLength,
                strokeDashoffset: e.value.useCssAnimation ? o.pathLength : 0
              })
            }, null, 14, fl),
            y("path", {
              fill: e.value.style.chart.areas.useGradient ? e.value.style.chart.areas.useCommonColor ? `url(#gradient-${o.id}-${w.value})` : `url(#gradient-single-${w.value}-${o.uid})` : o.color,
              stroke: "none",
              d: e.value.style.chart.areas.smooth ? o.smoothPath : o.straightPath,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: fe({ "vue-ui-ridgeline-animate": e.value.useCssAnimation && !u(G) }),
              style: C({
                strokeDasharray: o.pathLength,
                strokeDashoffset: e.value.useCssAnimation ? o.pathLength : 0
              })
            }, null, 14, gl),
            e.value.style.chart.zeroLine.show ? (n(), r("path", {
              key: 1,
              stroke: e.value.style.chart.zeroLine.useSerieColor ? o.color : e.value.style.chart.zeroLine.stroke,
              "stroke-dasharray": e.value.style.chart.zeroLine.strokeDasharray,
              "stroke-width": e.value.style.chart.zeroLine.strokeWidth,
              d: o.zeroPath,
              "stroke-linecap": "round"
            }, null, 8, yl)) : h("", !0),
            e.value.style.chart.areas.maxPoint.show && o.plots.length > 1 ? (n(!0), r(m, { key: 2 }, A(o.plots, (L) => (n(), r(m, null, [
              L.isMaxPoint ? (n(), r("line", {
                key: 0,
                x1: L.x,
                y1: L.y,
                x2: L.x,
                y2: L.zero,
                stroke: e.value.style.chart.areas.maxPoint.adaptStrokeToBackground ? u(Ta)(o.color) : e.value.style.chart.areas.maxPoint.stroke,
                "stroke-width": e.value.style.chart.areas.maxPoint.strokeWidth,
                "stroke-linecap": "round",
                "stroke-dasharray": e.value.style.chart.areas.maxPoint.strokeDasharray
              }, null, 8, pl)) : h("", !0)
            ], 64))), 256)) : h("", !0),
            o.plots.length === 1 ? (n(), r("circle", {
              key: 3,
              cx: o.plots[0].x,
              cy: o.plots[0].y,
              stroke: e.value.style.chart.selector.dot.stroke,
              "stroke-width": e.value.style.chart.selector.dot.strokeWidth,
              r: e.value.style.chart.selector.dot.radius,
              fill: e.value.style.chart.selector.dot.useDatapointColor ? o.color : e.value.style.chart.selector.dot.fill,
              style: {
                pointerEvents: "none"
              }
            }, null, 8, ml)) : h("", !0)
          ]))), 128)),
          y("text", {
            x: a.label.x,
            y: a.label.y,
            "text-anchor": "end",
            "font-size": e.value.style.chart.yAxis.labels.fontSize,
            "font-weight": e.value.style.chart.yAxis.labels.bold ? "bold" : "normal",
            fill: e.value.style.chart.yAxis.labels.color,
            style: C({
              cursor: e.value.style.chart.dialog.show && O.value ? "pointer" : "default"
            }),
            "text-decoration": e.value.style.chart.dialog.show && (Le.value === s || W.value && a.uid === W.value.uid) ? "underline" : "",
            onMouseenter: (o) => ea(s),
            onMouseleave: ta,
            onClick: (o) => Qt(a)
          }, V(a.name), 45, bl)
        ]))), 128)),
        e.value.style.chart.xAxis.labels.values.length ? (n(), r("g", {
          key: 0,
          ref_key: "timeLabelsEls",
          ref: ke
        }, [
          (n(!0), r(m, null, A(Y.value, (a, s) => d(t.$slots, "time-label", Ne({ ref_for: !0 }, {
            show: a && !e.value.style.chart.xAxis.labels.showOnlyFirstAndLast && !e.value.style.chart.xAxis.labels.showOnlyAtModulo || a && e.value.style.chart.xAxis.labels.showOnlyFirstAndLast && (s === 0 || s === Y.value.length - 1) || a && g.value && g.value.index === s || a && !e.value.style.chart.xAxis.labels.showOnlyFirstAndLast && e.value.style.chart.xAxis.labels.showOnlyAtModulo && s % Math.floor(Y.value.length / e.value.style.chart.xAxis.labels.modulo) === 0,
            fontSize: e.value.style.chart.xAxis.labels.fontSize,
            content: a.label,
            textAnchor: e.value.style.chart.xAxis.labels.rotation > 0 ? "start" : e.value.style.chart.xAxis.labels.rotation < 0 ? "end" : "middle",
            fill: e.value.style.chart.xAxis.labels.color,
            transform: `translate(${a.selectorX}, ${x.value.top + a.height + e.value.style.chart.xAxis.labels.offsetY}), rotate(${e.value.style.chart.xAxis.labels.rotation})`,
            x: a.selectorX,
            y: x.value.bottom + e.value.style.chart.xAxis.labels.offsetY
          }), () => [
            a && !e.value.style.chart.xAxis.labels.showOnlyFirstAndLast && !e.value.style.chart.xAxis.labels.showOnlyAtModulo || a && e.value.style.chart.xAxis.labels.showOnlyFirstAndLast && (s === 0 || s === Y.value.length - 1) || a && g.value && g.value.index === s || a && !e.value.style.chart.xAxis.labels.showOnlyFirstAndLast && e.value.style.chart.xAxis.labels.showOnlyAtModulo && s % Math.floor(Y.value.length / e.value.style.chart.xAxis.labels.modulo) === 0 ? (n(), r("g", xl, [
              String(a.label).includes(`
`) ? (n(), r("text", {
                key: 1,
                class: "vue-ui-ridgeline-x-axis-label",
                "font-size": e.value.style.chart.xAxis.labels.fontSize,
                fill: e.value.style.chart.xAxis.labels.color,
                "font-weight": e.value.style.chart.xAxis.labels.bold ? "bold" : "normal",
                transform: `translate(${a.selectorX}, ${x.value.bottom + e.value.style.chart.xAxis.labels.offsetY}), rotate(${e.value.style.chart.xAxis.labels.rotation})`,
                "text-anchor": e.value.style.chart.xAxis.labels.rotation > 0 ? "start" : e.value.style.chart.xAxis.labels.rotation < 0 ? "end" : "middle",
                style: C({
                  opacity: g.value ? g.value.index === s ? 1 : 0.2 : 1
                }),
                innerHTML: u(Oa)({
                  content: String(a.label),
                  fontSize: e.value.style.chart.xAxis.labels.fontSize,
                  fill: e.value.style.chart.xAxis.labels.color,
                  x: 0,
                  y: 0
                })
              }, null, 12, Cl)) : (n(), r("text", {
                key: 0,
                class: "vue-ui-ridgeline-x-axis-label",
                "font-size": e.value.style.chart.xAxis.labels.fontSize,
                fill: e.value.style.chart.xAxis.labels.color,
                "font-weight": e.value.style.chart.xAxis.labels.bold ? "bold" : "normal",
                transform: `translate(${a.selectorX}, ${x.value.bottom + e.value.style.chart.xAxis.labels.offsetY}), rotate(${e.value.style.chart.xAxis.labels.rotation})`,
                "text-anchor": e.value.style.chart.xAxis.labels.rotation > 0 ? "start" : e.value.style.chart.xAxis.labels.rotation < 0 ? "end" : "middle",
                style: C({
                  opacity: g.value ? g.value.index === s ? 1 : 0.2 : 1
                })
              }, V(a.label), 13, kl))
            ])) : h("", !0)
          ], !0)), 256))
        ], 512)) : h("", !0),
        y("g", null, [
          (n(!0), r(m, null, A(Y.value, (a, s) => (n(), r("rect", {
            x: a.x,
            y: a.y,
            width: a.width < 0 ? 0.1 : a.width,
            height: a.height < 0 ? 0.1 : a.height,
            fill: "transparent",
            onMouseenter: (o) => Gt(a),
            onMouseleave: (o) => jt(a),
            onClick: () => Yt(a)
          }, null, 40, Al))), 256)),
          e.value.style.chart.selector.show && g.value ? (n(), r("line", {
            key: 0,
            x1: g.value.selectorX,
            x2: g.value.selectorX,
            y1: g.value.y,
            y2: g.value.y + g.value.height - $.value / 2,
            stroke: e.value.style.chart.selector.stroke,
            "stroke-width": e.value.style.chart.selector.strokeWidth,
            "stroke-dasharray": e.value.style.chart.selector.strokeDasharray,
            "stroke-linecap": "round",
            style: { pointerEvents: "none" }
          }, null, 8, wl)) : h("", !0),
          g.value ? (n(!0), r(m, { key: 1 }, A(I.value, (a) => (n(), r(m, null, [
            (n(!0), r(m, null, A(a.datapoints, (s) => (n(), r(m, null, [
              (n(!0), r(m, null, A(s.plots, (o, f) => (n(), r(m, null, [
                g.value && g.value.index === f ? (n(), r("circle", {
                  key: 0,
                  cx: o.x,
                  cy: o.y,
                  stroke: e.value.style.chart.selector.dot.stroke,
                  "stroke-width": e.value.style.chart.selector.dot.strokeWidth,
                  r: e.value.style.chart.selector.dot.radius,
                  fill: e.value.style.chart.selector.dot.useDatapointColor ? s.color : e.value.style.chart.selector.dot.fill,
                  style: {
                    pointerEvents: "none"
                  }
                }, null, 8, $l)) : h("", !0),
                g.value && g.value.index === f ? (n(), r("text", {
                  key: 1,
                  x: ot(
                    o.x,
                    u(ye)(
                      e.value.style.chart.selector.labels.formatter,
                      o.value,
                      u(pe)({
                        p: e.value.style.chart.xAxis.labels.prefix,
                        v: o.value,
                        s: e.value.style.chart.xAxis.labels.suffix,
                        r: e.value.style.chart.selector.labels.rounding
                      })
                    ),
                    e.value.style.chart.selector.labels.fontSize
                  ) ? o.x - e.value.style.chart.selector.labels.fontSize / 2 : o.x + e.value.style.chart.selector.labels.fontSize / 2,
                  y: o.y + e.value.style.chart.selector.labels.fontSize / 3,
                  "text-anchor": ot(
                    o.x,
                    u(ye)(
                      e.value.style.chart.selector.labels.formatter,
                      o.value,
                      u(pe)({
                        p: e.value.style.chart.xAxis.labels.prefix,
                        v: o.value,
                        s: e.value.style.chart.xAxis.labels.suffix,
                        r: e.value.style.chart.selector.labels.rounding
                      })
                    ),
                    e.value.style.chart.selector.labels.fontSize
                  ) ? "end" : "start",
                  "font-size": e.value.style.chart.selector.labels.fontSize,
                  fill: e.value.style.chart.selector.labels.color,
                  style: {
                    pointerEvents: "none"
                  }
                }, V(u(ye)(
                  e.value.style.chart.selector.labels.formatter,
                  o.value,
                  u(pe)({
                    p: e.value.style.chart.xAxis.labels.prefix,
                    v: o.value,
                    s: e.value.style.chart.xAxis.labels.suffix,
                    r: e.value.style.chart.selector.labels.rounding
                  })
                )), 9, _l)) : h("", !0)
              ], 64))), 256))
            ], 64))), 256))
          ], 64))), 256)) : h("", !0)
        ]),
        d(t.$slots, "svg", {
          svg: {
            ...p.value,
            isPrintingImg: u(_e) | u(Pe) | oe.value,
            isPrintingSvg: se.value
          }
        }, void 0, !0)
      ], 14, tl)),
      t.$slots.watermark ? (n(), r("div", Pl, [
        d(t.$slots, "watermark", P(S({ isPrinting: u(_e) || u(Pe) || oe.value || se.value })), void 0, !0)
      ])) : h("", !0),
      y("div", {
        ref_key: "chartLegend",
        ref: Ve
      }, [
        e.value.style.chart.legend.show ? (n(), _(Ya, {
          key: `legend_${be.value}`,
          legendSet: R.value,
          config: Kt.value,
          isCursorPointer: O.value,
          onClickMarker: l[0] || (l[0] = ({ legend: a }) => ue(a.id))
        }, Ie({
          item: c(({ legend: a }) => [
            u(G) ? h("", !0) : (n(), r("div", {
              key: 0,
              style: C(`opacity:${b.value.includes(a.id) ? 0.5 : 1}`),
              onClick: (s) => a.segregate()
            }, V(a.name), 13, Sl))
          ]),
          legendToggle: c(() => [
            R.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !u(G) ? (n(), _(Ka, {
              key: 0,
              backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
              color: e.value.style.chart.legend.selectAllToggle.color,
              fontSize: e.value.style.chart.legend.fontSize,
              checked: b.value.length > 0,
              isCursorPointer: O.value,
              onToggle: Bt
            }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : h("", !0)
          ]),
          _: 2
        }, [
          t.$slots.pattern ? {
            name: "legend-pattern",
            fn: c(({ legend: a, index: s }) => [
              Re(Za, {
                shape: a.shape,
                radius: 30,
                stroke: "none",
                plot: { x: 30, y: 30 },
                fill: `url(#pattern_${w.value}_${s})`
              }, null, 8, ["shape", "fill"])
            ]),
            key: "0"
          } : void 0
        ]), 1032, ["legendSet", "config", "isCursorPointer"])) : h("", !0),
        d(t.$slots, "legend", { legend: R.value }, void 0, !0)
      ], 512),
      t.$slots.source ? (n(), r("div", {
        key: 5,
        ref_key: "source",
        ref: We,
        dir: "auto"
      }, [
        d(t.$slots, "source", {}, void 0, !0)
      ], 512)) : h("", !0),
      q.value && e.value.userOptions.buttons.table ? (n(), _(Ca(K.value.component), Ne({ key: 6 }, K.value.props, {
        ref_key: "tableUnit",
        ref: Z,
        onClose: ut
      }), Ie({
        content: c(() => [
          (n(), _(u(St), {
            key: `table_${It.value}`,
            colNames: de.value.colNames,
            head: de.value.head,
            body: de.value.body,
            config: de.value.config,
            title: e.value.table.useDialog ? "" : K.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: O.value,
            onClose: ut
          }, {
            th: c(({ th: a }) => [
              y("div", { innerHTML: a }, null, 8, Tl)
            ]),
            td: c(({ td: a }) => [
              Fe(V(a.name ? a.name : u(ye)(
                e.value.style.chart.selector.labels.formatter,
                a,
                u(pe)({
                  p: e.value.style.chart.xAxis.labels.prefix,
                  v: a,
                  s: e.value.style.chart.xAxis.labels.suffix,
                  r: e.value.table.td.roundingValue
                })
              )), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: c(() => [
            Fe(V(K.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: c(() => [
            y("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (a) => ze(e.value.userOptions.callbacks.csv)),
              style: C({ cursor: O.value ? "pointer" : "default" })
            }, [
              Re(u(_t), {
                name: "fileCsv",
                stroke: K.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : h("", !0),
      e.value.style.chart.dialog.show ? (n(), _(u(He), {
        key: 7,
        ref_key: "dialog",
        ref: xe,
        onClose: l[2] || (l[2] = (a) => W.value = null),
        backgroundColor: e.value.style.chart.dialog.backgroundColor,
        color: e.value.style.chart.dialog.color,
        headerBg: e.value.style.chart.dialog.header.backgroundColor,
        headerColor: e.value.style.chart.dialog.header.color,
        isFullscreen: X.value,
        fullscreenParent: T.value,
        isCursorPointer: O.value,
        withPadding: ""
      }, {
        title: c(() => [
          Fe(V(W.value.name), 1)
        ]),
        content: c(() => [
          W.value ? (n(), _(u($t), {
            key: 0,
            config: ie.value,
            dataset: ce.value,
            onCopyAlt: va
          }, null, 8, ["config", "dataset"])) : h("", !0)
        ]),
        _: 1
      }, 8, ["backgroundColor", "color", "headerBg", "headerColor", "isFullscreen", "fullscreenParent", "isCursorPointer"])) : h("", !0),
      d(t.$slots, "skeleton", {}, () => [
        u(G) ? (n(), _(Ea, { key: 0 })) : h("", !0)
      ], !0)
    ], 46, el));
  }
}, Jl = /* @__PURE__ */ Qa(Ol, [["__scopeId", "data-v-84fd5425"]]);
export {
  Jl as default
};
