// vue-data-ui/dist/vue-ui-scatter-FdAAYJOK.js
import { defineAsyncComponent as ie, computed as P, ref as $, toRefs as Ll, watch as Ee, shallowRef as Et, onMounted as Bl, onBeforeUnmount as Pl, openBlock as i, createElementBlock as c, unref as y, normalizeStyle as ne, normalizeClass as It, createBlock as le, withCtx as T, renderSlot as z, normalizeProps as K, guardReactiveProps as J, createCommentVNode as b, createElementVNode as w, createSlots as Ft, createVNode as Ce, Fragment as Q, renderList as ae, toDisplayString as G, Teleport as zl, resolveDynamicComponent as Ol, mergeProps as Dl, createTextVNode as Wt, nextTick as Nt } from "vue";
import { j as El, t as Xt, m as Il, p as he, o as Yt, q as ot, n as Fl, r as Rt, aj as Wl, s as st, al as Nl, c as Xl, u as Ie, X as Yl, ak as Rl, k as ve, K as q, l as oe, a0 as Ul, a1 as Vl, T as jl, U as Gl } from "./lib-GwPnJSrY.js";
import { t as ql, u as Hl } from "./useResponsive-DfdjqQps.js";
import { u as Zl } from "./useConfig-CIgyKZ6R.js";
import { u as Kl, B as Jl } from "./BaseScanner-BhhkKVoj.js";
import { u as Ql } from "./usePrinter-DiK5R6k-.js";
import { u as ea } from "./useSvgExport-ByUukOZt.js";
import { u as nt } from "./useNestedProp-BBT0mbHh.js";
import { u as ta } from "./useThemeCheck-D1ZYXUwG.js";
import { u as la } from "./useUserOptionState-BIvW1Kz7.js";
import { u as aa } from "./useChartAccessibility-9icAAmYg.js";
import oa from "./vue_ui_scatter-CO5Iq9ai.js";
import sa from "./Legend-1LMMsPxL.js";
import na from "./Title-BfyrP-jp.js";
import { _ as ut } from "./Shape-CSyWYgXc.js";
import ua from "./img-BYYzY_QU.js";
import { B as ra } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as ia } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const va = ["id"], ca = ["id"], ya = ["xmlns", "viewBox"], da = ["width", "height"], ha = { key: 1 }, fa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], ga = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], pa = { key: 2 }, ma = ["id"], ba = ["stop-color"], ka = ["stop-color"], xa = ["id"], wa = ["stop-color"], $a = ["stop-color"], _a = ["x", "y", "width", "height", "fill", "stroke", "stroke-width", "rx"], Ma = ["x", "y", "width", "height", "onMouseenter"], Ca = {
  key: 2,
  style: { "pointer-events": "none" }
}, Aa = ["x", "y", "width", "height", "fill", "fill-opacity"], Sa = ["x1", "x2", "y2", "stroke", "stroke-dasharray", "stroke-width"], Ta = ["x1", "x2", "y2", "stroke", "stroke-dasharray", "stroke-width"], La = ["x", "y", "height", "width", "fill", "stroke", "stroke-width", "rx"], Ba = ["x", "y", "width", "height", "onMouseenter"], Pa = {
  key: 2,
  style: { "pointer-events": "none" }
}, za = ["x", "y", "width", "height", "fill", "fill-opacity"], Oa = ["x1", "x2", "y1", "y2", "stroke", "stroke-dasharray", "stroke-width"], Da = ["x1", "x2", "y1", "y2", "stroke", "stroke-dasharray", "stroke-width"], Ea = {
  key: 0,
  style: { "pointer-events": "none" }
}, Ia = ["d", "stroke", "stroke-width"], Fa = ["d", "stroke", "stroke-width"], Wa = ["d", "stroke", "stroke-width"], Na = ["d", "stroke", "stroke-width"], Xa = { key: 3 }, Ya = ["points", "fill", "stroke-width", "stroke-dasharray", "stroke"], Ra = { key: 0 }, Ua = ["cx", "cy", "r", "fill", "stroke", "stroke-width", "onMouseover", "onMouseleave", "onClick"], Va = { key: 1 }, ja = ["clip-path"], Ga = ["d", "fill", "stroke", "stroke-width", "stroke-opacity"], qa = {
  key: 0,
  style: { "pointer-events": "none" }
}, Ha = ["x", "y", "width", "height"], Za = {
  key: 6,
  style: { "pointer-events": "none !important" }
}, Ka = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Ja = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Qa = ["x", "y", "font-size", "fill", "font-weight", "text-anchor"], eo = ["x", "y", "font-size", "fill", "font-weight"], to = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], lo = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], ao = ["x", "y", "font-size", "fill", "font-weight", "text-anchor"], oo = ["id", "transform", "font-size", "font-weight", "fill"], so = ["font-size", "fill", "transform"], no = ["transform", "font-size", "fill"], uo = ["x", "y", "font-size", "fill"], ro = ["x", "y", "font-size", "fill"], io = ["font-size", "font-weight", "fill", "x", "y"], vo = ["id"], co = ["x", "y", "width", "height"], yo = {
  key: 11,
  style: { pointerEvents: "none" }
}, ho = ["x1", "x2", "y1", "y2", "stroke-dasharray", "stroke", "stroke-width", "clip-path"], fo = ["x", "y", "fill", "font-size", "font-weight"], go = {
  key: 4,
  class: "vue-data-ui-watermark"
}, po = ["id"], mo = ["onClick"], bo = {
  key: 0,
  style: { width: "100%", display: "flex", "align-items": "center", "justify-content": "center" }
}, ko = {
  viewBox: "0 0 20 20",
  height: "20",
  width: "20",
  style: { overflow: "hidden", background: "transparent" }
}, xo = { key: 0 }, wo = ["innerHTML"], $o = {
  __name: "vue-ui-scatter",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["copyAlt"],
  setup(Ut, { expose: Vt, emit: jt }) {
    const Gt = ie(() => import("./Tooltip-DL3QgjZf.js")), qt = ie(() => import("./BaseIcon-CNx8Tzl5.js")), Ht = ie(() => import("./vue-ui-accordion-BGHTqIBt.js")), Zt = ie(() => import("./DataTable-Bo1pmAqG.js")), Kt = ie(() => import("./PenAndPaper-DWV1FMIi.js")), Jt = ie(() => import("./UserOptions-CrLqEimB.js")), Qt = ie(() => import("./PackageVersion-DN9d0ME5.js")), el = ie(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_scatter: tl } = Zl(), { isThemeValid: ll, warnInvalidTheme: al } = ta(), ee = Ut, ol = jt, rt = P(() => !!ee.dataset && ee.dataset.length), X = $(El()), Fe = $(!1), We = $(""), it = $(0), se = $(null), vt = $(null), ct = $(null), yt = $(null), dt = $(null), ht = $(0), ft = $(0), gt = $(0), O = $([]), pt = $(!1), be = $(null), ke = $(null), xe = $(null), Ne = $(null), Ae = $(!1), Se = $(!1), Xe = $(null), sl = $(null), nl = $(null), Ye = $(null), e = $(Ge()), ce = P(() => e.value.userOptions.useCursorPointer);
    function ul(l = 100, a = 0.8, t = {}) {
      const { meanX: o = 0, sdX: s = 1, meanY: u = 0, sdY: r = 1, seed: f } = t;
      let v = (f ?? Math.floor(Math.random() * 2 ** 31)) >>> 0;
      const p = () => (v = v * 1664525 + 1013904223 >>> 0, v / 2 ** 32), L = () => {
        let m = 0, W = 0;
        for (; m === 0; ) m = p();
        for (; W === 0; ) W = p();
        return Math.sqrt(-2 * Math.log(m)) * Math.cos(2 * Math.PI * W);
      }, d = l / 2, M = Array.from({ length: d }, L), F = Array.from({ length: d }, L), C = (m) => m.reduce((W, Me) => W + Me, 0) / m.length, D = C(M), Y = C(F);
      for (let m = 0; m < d; m += 1)
        M[m] -= D, F[m] -= Y;
      const B = (m, W) => m.reduce((Me, De, at) => Me + De * W[at], 0), _ = (m) => B(m, m), h = B(F, M) / _(M), S = F.map((m, W) => m - h * M[W]), j = _(M) / d, te = _(S) / d, re = Math.sqrt((1 - a * a) * j / te), U = M.map((m, W) => a * m + re * S[W]), de = M.concat(M.map((m) => -m)), lt = U.concat(U.map((m) => -m)), $e = (m) => Math.sqrt(_(m) / m.length), pe = (m, W, Me) => {
        const De = $e(m);
        return m.map((at) => Me + (De ? at / De * W : 0));
      }, me = pe(de, s, o), _e = pe(lt, r, u);
      for (let m = me.length - 1; m > 0; m -= 1) {
        const W = Math.floor(p() * (m + 1));
        [me[m], me[W]] = [me[W], me[m]], [_e[m], _e[W]] = [_e[W], _e[m]];
      }
      return me.map((m, W) => ({ x: m, y: _e[W] }));
    }
    const rl = P(() => Xt({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        useCssAnimation: !1,
        style: {
          backgroundColor: "#99999930",
          layout: {
            axis: {
              stroke: "#6A6A6A"
            },
            correlation: {
              label: { show: !1 }
            },
            dataLabels: {
              xAxis: { show: !1 },
              yAxis: { show: !1 }
            },
            marginalBars: {
              fill: "#99999960"
            },
            padding: { top: 12, right: 12, bottom: 12, left: 12 },
            plots: {
              stroke: "#6A6A6A"
            }
          },
          legend: {
            backgroundColor: "transparent"
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: Re, FINAL_DATASET: Ue, manualLoading: Ve } = Kl({
      ...Ll(ee),
      FINAL_CONFIG: e,
      prepareConfig: Ge,
      skeletonDataset: ee.config?.skeletonDataset ?? [
        {
          name: "",
          color: "#999999",
          values: ul(100, 0.5, { seed: 42 })
        }
      ],
      skeletonConfig: Xt({
        defaultConfig: e.value,
        userConfig: rl.value
      })
    }), { userOptionsVisible: je, setUserOptionsVisibility: mt, keepUserOptionState: bt } = la({ config: e.value }), { svgRef: Te } = aa({ config: e.value.style.title });
    function Ge() {
      const l = nt({
        userConfig: ee.config,
        defaultConfig: tl
      }), a = l.theme;
      if (!a) return l;
      if (!ll.value(l))
        return al(l), l;
      const t = nt({
        userConfig: oa[a] || ee.config,
        defaultConfig: l
      }), o = nt({
        userConfig: ee.config,
        defaultConfig: t
      });
      return {
        ...o,
        customPalette: o.customPalette.length ? o.customPalette : Il[a] || he
      };
    }
    Ee(() => ee.config, (l) => {
      Re.value || (e.value = Ge()), je.value = !e.value.userOptions.showOnChartHover, xt(), ht.value += 1, ft.value += 1, gt.value += 1, H.value.showTable = e.value.table.show, H.value.showTooltip = e.value.style.tooltip.show, kt.value && e.value.usePerformanceMode && console.warn(`VueUiScatter : You are using performance mode

- downsampling is disabled in this mode, all plots are rendered

- plot significance is not active in this mode (all plots have the same opacity) 

- Depending on plot density, shapes might not display a border (stroke) to avoid fuzziness 

ℹ️ To remove this warning, set config.debug to false.`);
    }, { deep: !0 }), Ee(() => ee.dataset, (l) => {
      Array.isArray(l) && l.length > 0 && (Ve.value = !1);
    }, { deep: !0 });
    const ue = Et(null), fe = Et(null);
    Bl(() => {
      pt.value = !0, xt();
    });
    const kt = P(() => !!e.value.debug);
    function xt() {
      if (Yt(ee.dataset) && (ot({
        componentName: "VueUiScatter",
        type: "dataset",
        debug: kt.value
      }), Ve.value = !0), Yt(ee.dataset) || (Ve.value = e.value.loading), e.value.responsive) {
        const l = ql(() => {
          const { width: a, height: t } = Hl({
            chart: se.value,
            title: e.value.style.title.text ? vt.value : null,
            legend: e.value.style.legend.show ? ct.value : null,
            source: yt.value,
            noTitle: dt.value
          });
          requestAnimationFrame(() => {
            N.value.width = a, N.value.height = t;
          });
        });
        ue.value && (fe.value && ue.value.unobserve(fe.value), ue.value.disconnect()), ue.value = new ResizeObserver(l), fe.value = se.value.parentNode, ue.value.observe(fe.value);
      }
    }
    Pl(() => {
      ue.value && (fe.value && ue.value.unobserve(fe.value), ue.value.disconnect());
    });
    const { isPrinting: qe, isImaging: He, generatePdf: wt, generateImage: $t } = Ql({
      elementId: `vue-ui-scatter_${X.value}`,
      fileName: e.value.style.title.text || "vue-ui-scatter",
      options: e.value.userOptions.print
    }), il = P(() => e.value.userOptions.show && !e.value.style.title.text), _t = P(() => Fl(e.value.customPalette)), H = $({
      showTable: e.value.table.show,
      showTooltip: e.value.style.tooltip.show
    });
    Ee(e, () => {
      H.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.tooltip.show
      };
    }, { immediate: !0 });
    const N = $({
      height: e.value.style.layout.height,
      width: e.value.style.layout.width
    }), Le = P(() => e.value.style.layout.marginalBars.show ? e.value.style.layout.marginalBars.size + e.value.style.layout.marginalBars.offset : 0), n = P(() => {
      let l = 0, a = 0;
      return Xe.value && (l = Xe.value.getBBox().width + 6), Ye.value && (a = Ye.value.getBBox().height + 6), {
        top: e.value.style.layout.padding.top + Le.value + e.value.style.layout.dataLabels.yAxis.fontSize * 2,
        right: N.value.width - e.value.style.layout.padding.right - Le.value - 6,
        bottom: N.value.height - e.value.style.layout.padding.bottom - a,
        left: e.value.style.layout.padding.left + l,
        height: N.value.height - e.value.style.layout.padding.top - e.value.style.layout.padding.bottom - Le.value - a - e.value.style.layout.dataLabels.yAxis.fontSize * 2,
        width: N.value.width - e.value.style.layout.padding.left - e.value.style.layout.padding.right - Le.value - l - 6
      };
    }), E = P(() => {
      Ue.value.forEach((s, u) => {
        Rt({
          datasetObject: s,
          requiredAttributes: ["values"]
        }).forEach((r) => {
          ot({
            componentName: "VueUiScatter",
            type: "datasetSerieAttribute",
            property: r,
            index: u
          });
        }), s.values && s.values.forEach((r, f) => {
          Rt({
            datasetObject: r,
            requiredAttributes: ["x", "y"]
          }).forEach((v) => {
            ot({
              componentName: "VueUiScatter",
              type: "datasetSerieAttribute",
              property: `values.${v}`,
              index: `${u} - ${f}`
            });
          });
        });
      });
      const l = Math.min(...Z.value.filter((s) => !O.value.includes(s.id)).flatMap((s) => s.values.map((u) => u.x))), a = Math.max(...Z.value.filter((s) => !O.value.includes(s.id)).flatMap((s) => s.values.map((u) => u.x))), t = Math.min(...Z.value.filter((s) => !O.value.includes(s.id)).flatMap((s) => s.values.map((u) => u.y))), o = Math.max(...Z.value.filter((s) => !O.value.includes(s.id)).flatMap((s) => s.values.map((u) => u.y)));
      return { xMin: l >= 0 ? 0 : l, xMax: a, yMin: t >= 0 ? 0 : t, yMax: o };
    }), I = P(() => ({
      x: n.value.left + Math.abs(E.value.xMin) / (E.value.xMax + Math.abs(E.value.xMin)) * n.value.width,
      y: n.value.bottom - Math.abs(E.value.yMin) / (E.value.yMax + Math.abs(E.value.yMin)) * n.value.height
    })), Z = P(() => Ue.value.map((l, a) => {
      const t = `cluster_${X.value}_${a}`;
      return {
        ...l,
        values: Wl({
          data: l.values,
          threshold: e.value.usePerformanceMode ? l.values.length + 1 : e.value.downsample.threshold
        }),
        id: t,
        color: l.color ? l.color : _t.value[a] || he[a] || he[a % he.length],
        opacity: O.value.includes(t) ? 0.5 : 1,
        shape: l.shape ?? "circle",
        segregate: () => Pe(t),
        isSegregated: O.value.includes(t)
      };
    })), vl = P(() => ({
      cy: "scatter-div-legend",
      backgroundColor: e.value.style.legend.backgroundColor,
      color: e.value.style.legend.color,
      fontSize: e.value.style.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.legend.bold ? "bold" : ""
    })), ge = P(() => Z.value.map((l, a) => ({
      ...l,
      plots: l.values.map((t) => ({
        x: n.value.left + (t.x + Math.abs(E.value.xMin)) / (E.value.xMax + Math.abs(E.value.xMin)) * n.value.width,
        y: n.value.bottom - (t.y + Math.abs(E.value.yMin)) / (E.value.yMax + Math.abs(E.value.yMin)) * n.value.height,
        v: {
          ...t,
          name: t.name || ""
        },
        clusterName: l.name,
        color: l.color ? l.color : _t.value[a] || he[a] || he[a % he.length],
        id: `plot_${X.value}_${Math.random()}`,
        weight: t.weight ?? e.value.style.layout.plots.radius
      }))
    })).filter((l) => !O.value.includes(l.id))), V = P(() => {
      const a = ({ m: t, b: o, rect: s, verticalX: u = null }) => {
        const { left: r, right: f, top: v, bottom: p } = s, L = [], d = (_, h) => {
          Number.isFinite(_) && Number.isFinite(h) && L.push({ x: _, y: h });
        }, M = ({ x: _, y: h }) => _ >= r - 1e-9 && _ <= f + 1e-9 && h >= v - 1e-9 && h <= p + 1e-9;
        u !== null ? u >= r - 1e-9 && u <= f + 1e-9 && (d(u, v), d(u, p)) : Number.isFinite(t) && (d(r, t * r + o), d(f, t * f + o), Math.abs(t) > 1e-9 ? (d((v - o) / t, v), d((p - o) / t, p)) : o >= v - 1e-9 && o <= p + 1e-9 && (d(r, o), d(f, o)));
        const F = L.filter(M), C = [];
        for (const _ of F)
          C.some((h) => Math.abs(h.x - _.x) < 1e-6 && Math.abs(h.y - _.y) < 1e-6) || C.push(_);
        if (C.length < 2) return null;
        let D = C[0], Y = C[1], B = 0;
        for (let _ = 0; _ < C.length; _ += 1)
          for (let h = _ + 1; h < C.length; h += 1) {
            const S = C[_].x - C[h].x, j = C[_].y - C[h].y, te = S * S + j * j;
            te > B && (B = te, D = C[_], Y = C[h]);
          }
        return { x1: D.x, y1: D.y, x2: Y.x, y2: Y.y };
      };
      return ge.value.map((t) => {
        const o = t.plots.length, s = t.plots.reduce((h, S) => h + S.x, 0) / o, u = t.plots.reduce((h, S) => h + S.y, 0) / o;
        let r = 0, f = 0;
        for (const h of t.plots) {
          const S = h.x - s, j = h.y - u;
          r += S * j, f += S * S;
        }
        let v, p, L = null;
        f < 1e-9 ? (L = s, v = 1 / 0, p = null) : (v = r / f, p = u - v * s);
        let d, M;
        L !== null ? (d = 1 / 0, M = null) : (d = v, M = p);
        const F = t.plots.every((h) => h.v && typeof h.v.x == "number" && typeof h.v.y == "number");
        let C = NaN;
        if (o >= 2) {
          let h = 0, S = 0;
          F ? (h = t.plots.reduce((U, de) => U + de.v.x, 0) / o, S = t.plots.reduce((U, de) => U + de.v.y, 0) / o) : (h = s, S = -u);
          let j = 0, te = 0, re = 0;
          for (const U of t.plots) {
            const de = F ? U.v.x : U.x, lt = F ? U.v.y : -U.y, $e = de - h, pe = lt - S;
            j += $e * pe, te += $e * $e, re += pe * pe;
          }
          if (te >= 1e-9 && re >= 1e-9) {
            const U = j / Math.sqrt(te * re);
            C = Math.max(-1, Math.min(1, U));
          }
        }
        const D = a({ m: d, b: M, rect: n.value, verticalX: L });
        if (!D)
          return {
            ...t,
            correlation: null,
            label: null,
            plots: t.plots.map((h) => ({
              ...h,
              deviation: 0,
              shape: t.shape,
              color: st(t.color)
            }))
          };
        const Y = (D.x1 + D.x2) / 2, B = (D.y1 + D.y2) / 2, _ = { x: Y, y: B };
        return {
          ...t,
          color: st(t.color),
          correlation: {
            ...D,
            coefficient: C
          },
          label: _,
          plots: t.plots.map((h) => {
            let S, j;
            L !== null ? (S = L, j = h.y) : Math.abs(d) < 1e-9 ? (S = h.x, j = M) : (S = (h.x + d * h.y - d * M) / (1 + d * d), j = (d * h.x + d * d * h.y + M) / (1 + d * d));
            const te = h.x - S, re = h.y - j, U = Math.sqrt(te * te + re * re);
            return {
              ...h,
              deviation: U,
              shape: t.shape,
              color: st(t.color)
            };
          })
        };
      });
    }), Mt = P(() => Math.max(...V.value.flatMap((l) => l.plots.map((a) => Math.abs(a.deviation)))));
    function cl() {
      return V.value;
    }
    function Ct(l, a) {
      const t = Array.isArray(l) ? l.flatMap((B) => B.plots.map((_) => ({
        x: _.x,
        y: _.y
      }))) : l.plots.map((B) => ({
        x: B.x,
        y: B.y
      }));
      let o = 1 / 0, s = -1 / 0, u = 1 / 0, r = -1 / 0;
      t.forEach(({ x: B, y: _ }) => {
        o = Math.min(o, B), s = Math.max(s, B), u = Math.min(u, _), r = Math.max(r, _);
      });
      const f = s - o, v = r - u, p = f / a, L = v / a, d = Array(a).fill(0), M = Array(a).fill(0);
      t.forEach(({ x: B, y: _ }) => {
        const h = Math.floor((B - o) / p), S = Math.floor((_ - u) / L);
        d[h] || (d[h] = 0), M[S] || (M[S] = 0), d[h] += 1, M[S] += 1;
      });
      const F = [], C = [];
      for (let B = 0; B < a; B += 1)
        F.push(o + (B + 0.5) * p), C.push(u + (B + 0.5) * L);
      const D = Math.max(...d), Y = Math.max(...M);
      return { x: d, y: M, avgX: F, avgY: C, maxX: D, maxY: Y };
    }
    const k = P(() => e.value.style.layout.marginalBars.tranches), A = P(() => Ct(ge.value, k.value)), yl = P(() => {
      const l = n.value.top - e.value.style.layout.marginalBars.offset, a = n.value.right + e.value.style.layout.marginalBars.offset;
      return ge.value.map((t) => {
        const o = Ct(t, k.value);
        return {
          coords: o,
          dX: Xl(o.avgX.map((s, u) => ({
            x: s,
            y: l - o.x[u] / o.maxX * e.value.style.layout.marginalBars.size
          }))),
          dY: Nl(o.avgY.map((s, u) => ({
            y: s,
            x: a + e.value.style.layout.marginalBars.size * o.y[u] / o.maxY
          }))),
          color: t.color,
          id: t.id
        };
      });
    }), R = $(void 0), g = $(null), Ze = $(null);
    function Ke(l, a) {
      R.value = l.id, g.value = l;
      let t = "";
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: l, seriesIndex: a }), Ze.value = {
        datapoint: l,
        seriesIndex: a,
        series: V.value,
        config: e.value
      };
      const o = e.value.style.tooltip.customFormat;
      jl(o) && Gl(() => o({
        datapoint: l,
        seriesIndex: a,
        series: V.value,
        config: e.value
      })) ? We.value = o({
        datapoint: l,
        seriesIndex: a,
        series: V.value,
        config: e.value
      }) : (l.clusterName && (t += `<div style="display:flex;gap:3px;align-items:center">${l.clusterName}</div>`), l.v.name && (t += `<div>${l.v.name}</div>`), t += `<div style="text-align:left;margin-top:6px;padding-top:6px;border-top:1px solid ${e.value.style.tooltip.borderColor}">`, t += `<div>${e.value.style.layout.dataLabels.xAxis.name}: <b>${isNaN(l.v.x) ? "-" : ve(
        e.value.style.layout.plots.selectors.labels.x.formatter,
        l.v.x,
        oe({
          p: e.value.style.tooltip.prefix,
          v: l.v.x,
          s: e.value.style.tooltip.suffix,
          r: e.value.style.tooltip.roundingValue
        }),
        { datapoint: l, seriesIndex: a }
      )}</b></div>`, t += `<div>${e.value.style.layout.dataLabels.yAxis.name}: <b>${isNaN(l.v.y) ? "-" : ve(
        e.value.style.layout.plots.selectors.labels.y.formatter,
        l.v.y,
        oe({
          p: e.value.style.tooltip.prefix,
          v: l.v.y,
          s: e.value.style.tooltip.suffix,
          r: e.value.style.tooltip.roundingValue
        }),
        { datapoint: l, seriesIndex: a }
      )}</b></div>`, t += `${e.value.style.layout.plots.deviation.translation}: <b>${oe({
        v: l.deviation,
        r: e.value.style.layout.plots.deviation.roundingValue
      })}</b>`, t += "</div>", We.value = `<div>${t}</div>`), Fe.value = !0;
    }
    function Be(l, a) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: a }), Fe.value = !1, R.value = void 0, g.value = null;
    }
    function Je(l, a) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: a });
    }
    function dl() {
      O.value.length ? O.value = [] : Z.value.forEach((l) => {
        O.value.push(l.id);
      });
    }
    function Pe(l) {
      O.value.includes(l) ? O.value = O.value.filter((a) => a !== l) : O.value.length < Ue.value.length - 1 && O.value.push(l);
    }
    function At(l) {
      if (!Z.value.length)
        return e.value.debug && console.warn("VueUiScatter - There are no series to show."), null;
      const a = Z.value.find((t) => t.name === l);
      return a || (e.value.debug && console.warn(`VueUiScatter - Series name not found "${l}"`), null);
    }
    function hl(l) {
      const a = At(l);
      a !== null && O.value.includes(a.id) && Pe(a.id);
    }
    function fl(l) {
      const a = At(l);
      a !== null && (O.value.includes(a.id) || Pe(a.id));
    }
    function Qe(l = null) {
      Nt(() => {
        const a = ["", e.value.table.translations.correlationCoefficient, e.value.table.translations.nbrPlots, `${e.value.style.layout.dataLabels.xAxis.name} ${e.value.table.translations.average}`, `${e.value.style.layout.dataLabels.yAxis.name} ${e.value.table.translations.average}`], t = V.value.map((u) => [
          u.name,
          u.correlation.coefficient,
          u.plots.length,
          u.plots.map((r) => r.v.x).reduce((r, f) => r + f, 0) / u.plots.length,
          u.plots.map((r) => r.v.y).reduce((r, f) => r + f, 0) / u.plots.length
        ]), o = [[e.value.style.title.text], [e.value.style.title.subtitle.text], [[""], [""], [""]]].concat([a]).concat(t), s = Ul(o);
        l ? l(s) : Vl({ csvContent: s, title: e.value.style.title.text || "vue-ui-heatmap" });
      });
    }
    const ze = P(() => {
      const l = [
        e.value.table.translations.series,
        e.value.table.translations.correlationCoefficient,
        e.value.table.translations.nbrPlots,
        `${e.value.style.layout.dataLabels.xAxis.name} ${e.value.table.translations.average}`,
        `${e.value.style.layout.dataLabels.yAxis.name} ${e.value.table.translations.average}`
      ], a = V.value.map((o) => [
        {
          shape: o.shape,
          content: o.name,
          color: o.color
        },
        Number((o.correlation.coefficient ?? 0).toFixed(e.value.table.td.roundingValue)).toLocaleString(),
        o.plots.length.toLocaleString(),
        Number((o.plots.map((s) => s.v.x ?? 0).reduce((s, u) => s + u, 0) / o.plots.length).toFixed(e.value.table.td.roundingAverage)).toLocaleString(),
        Number((o.plots.map((s) => s.v.y ?? 0).reduce((s, u) => s + u, 0) / o.plots.length).toFixed(e.value.table.td.roundingAverage)).toLocaleString()
      ]), t = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return { head: l, body: a, config: t, colNames: l };
    }), ye = $(!1);
    function St(l) {
      ye.value = l, it.value += 1;
    }
    function Tt() {
      H.value.showTable = !H.value.showTable;
    }
    function Lt() {
      H.value.showTooltip = !H.value.showTooltip;
    }
    const Oe = $(!1);
    function et() {
      Oe.value = !Oe.value;
    }
    async function gl({ scale: l = 2 } = {}) {
      if (!se.value) return;
      const { width: a, height: t } = se.value.getBoundingClientRect(), o = a / t, { imageUri: s, base64: u } = await ua({ domElement: se.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: s,
        base64: u,
        title: e.value.style.title.text,
        width: a,
        height: t,
        aspectRatio: o
      };
    }
    function pl(l) {
      be.value = l, e.value.style.layout.marginalBars.highlighter.highlightBothAxes && (ke.value = A.value.y.length - 2 - l);
    }
    function ml(l) {
      ke.value = l, e.value.style.layout.marginalBars.highlighter.highlightBothAxes && (be.value = l);
    }
    function tt() {
      be.value = null, ke.value = null;
    }
    const Bt = {
      circle: 1,
      square: 1,
      diamond: 1,
      triangle: 1.2,
      star: 1.3,
      pentagon: 1.3,
      hexagon: 1.3
    }, x = (l) => l.toFixed(3);
    function bl({ shape: l = "circle", cx: a, cy: t, r: o }) {
      if (!e.value.usePerformanceMode) return "";
      const s = Bt[l] * o;
      switch (l) {
        case "circle": {
          const u = x(a - s), r = x(t), f = x(a + s), v = x(s);
          return `M ${u} ${r} A ${v} ${v} 0 1 0 ${f} ${r} A ${v} ${v} 0 1 0 ${u} ${r} Z`;
        }
        case "square": {
          const u = x(a - s), r = x(t - s), f = x(a + s), v = x(t + s);
          return `M ${u} ${r} L ${f} ${r} L ${f} ${v} L ${u} ${v} Z`;
        }
        case "diamond": {
          const u = x(a), r = x(t);
          return `M ${u} ${x(t - s)} L ${x(a + s)} ${r} L ${u} ${x(t + s)} L ${x(a - s)} ${r} Z`;
        }
        case "triangle": {
          const u = s * Math.sqrt(3), r = a, f = t - 2 / 3 * u, v = a - s, p = t + 1 / 3 * u, L = a + s, d = p;
          return `M ${x(r)} ${x(f)} L ${x(v)} ${x(p)} L ${x(L)} ${x(d)} Z`;
        }
        case "star": {
          const u = s, r = s * 0.5, f = [];
          for (let p = 0; p < 10; p += 1) {
            const L = (-90 + p * 36) * Math.PI / 180, d = p % 2 === 0 ? u : r;
            f.push([a + d * Math.cos(L), t + d * Math.sin(L)]);
          }
          let v = `M ${x(f[0][0])} ${x(f[0][1])}`;
          for (let p = 1; p < f.length; p += 1)
            v += ` L ${x(f[p][0])} ${x(f[p][1])}`;
          return v + " Z";
        }
        case "pentagon": {
          const r = [];
          for (let v = 0; v < 5; v += 1) {
            const p = (-90 + v * 72) * Math.PI / 180;
            r.push([a + s * Math.cos(p), t + s * Math.sin(p)]);
          }
          let f = `M ${x(r[0][0])} ${x(r[0][1])}`;
          for (let v = 1; v < 5; v += 1) f += ` L ${x(r[v][0])} ${x(r[v][1])}`;
          return f + " Z";
        }
        case "hexagon": {
          const r = [];
          for (let v = 0; v < 6; v += 1) {
            const p = (-60 + v * 60) * Math.PI / 180;
            r.push([a + s * Math.cos(p), t + s * Math.sin(p)]);
          }
          let f = `M ${x(r[0][0])} ${x(r[0][1])}`;
          for (let v = 1; v < 6; v += 1) f += ` L ${x(r[v][0])} ${x(r[v][1])}`;
          return f + " Z";
        }
        default: {
          const u = x(a - s), r = x(t), f = x(a + s), v = x(s);
          return `M ${u} ${r} A ${v} ${v} 0 1 0 ${f} ${r} A ${v} ${v} 0 1 0 ${u} ${r} Z`;
        }
      }
    }
    const kl = P(() => {
      if (!e.value.usePerformanceMode) return [""];
      const { left: l, right: a, top: t, bottom: o } = n.value, s = Math.max(1, (a - l) * (o - t)), u = (d) => d / s * 1e4, r = 2.5, f = 1e3, v = e.value.style.layout.plots.stroke, p = e.value.style.layout.plots.strokeWidth, L = e.value.style.layout.plots.opacity;
      return V.value.map((d) => {
        const M = [];
        for (const C of d.plots) {
          const D = C.x, Y = C.y;
          if (D < l || D > a || Y < t || Y > o) continue;
          const B = Math.max(e.value.style.layout.plots.radius, C.weight);
          M.push(bl({
            shape: d.shape || "circle",
            cx: D,
            cy: Y,
            r: B
          }));
        }
        if (!M.length) return null;
        const F = u(d.plots.length) > r || d.plots.length > f;
        return {
          id: d.id,
          d: M.join(""),
          fill: Ie(d.color, L * 100),
          stroke: F ? "none" : v,
          strokeWidth: F ? 0 : p,
          strokeOpacity: 1
        };
      }).filter(Boolean);
    });
    function xl() {
      return e.value.usePerformanceMode ? (l) => {
        const a = Te.value;
        if (!a) return;
        const t = a.createSVGPoint();
        t.x = l.clientX, t.y = l.clientY;
        const o = a.getScreenCTM();
        if (!o) return;
        const s = o.inverse(), u = t.matrixTransform(s), r = 8, f = r * r;
        let v = null, p = 1 / 0, L = -1;
        if (V.value.forEach((d, M) => {
          d.plots.forEach((F) => {
            const C = F.x - u.x, D = F.y - u.y, Y = C * C + D * D;
            Y <= f && Y < p && (p = Y, v = F, L = M);
          });
        }), v)
          R.value !== v.id && (R.value = v.id, Ke(v, L));
        else if (R.value) {
          const d = g.value;
          R.value = void 0, Be(d, L);
        }
      } : () => null;
    }
    const Pt = xl();
    function wl() {
      if (R.value) {
        const l = g.value;
        R.value = void 0, Be(l, null);
      }
    }
    function $l(l) {
      const a = g.value;
      if (a) {
        const t = V.value.findIndex((o) => o.id === a.clusterId);
        Je(a, t >= 0 ? t : 0);
      }
    }
    const we = P(() => {
      const l = e.value.table.useDialog && !e.value.table.show, a = H.value.showTable;
      return {
        component: l ? el : Ht,
        title: `${e.value.style.title.text}${e.value.style.title.subtitle.text ? `: ${e.value.style.title.subtitle.text}` : ""}`,
        props: l ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: ye.value,
          fullscreenParent: se.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: ce.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.backgroundColor,
              color: e.value.style.color
            },
            head: {
              backgroundColor: e.value.style.backgroundColor,
              color: e.value.style.color
            }
          }
        }
      };
    });
    Ee(() => H.value.showTable, (l) => {
      e.value.table.show || (l && e.value.table.useDialog && xe.value ? xe.value.open() : "close" in xe.value && xe.value.close());
    });
    function zt() {
      H.value.showTable = !1, Ne.value && Ne.value.setTableIconState(!1);
    }
    const _l = P(() => e.value.style.backgroundColor), Ml = P(() => e.value.style.legend), Cl = P(() => e.value.style.title), { exportSvg: Al, getSvg: Sl } = ea({
      svg: Te,
      title: Cl,
      legend: Ml,
      legendItems: Z,
      backgroundColor: _l
    });
    async function Ot({ isCb: l }) {
      Se.value = !0, await Nt();
      try {
        if (l) {
          const { blob: a, url: t, text: o, dataUrl: s } = await Sl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: t, text: o, dataUrl: s }));
        } else
          await Promise.resolve(Al());
      } finally {
        Se.value = !1;
      }
    }
    function Tl(l) {
      if (l?.stage === "start") {
        Ae.value = !0;
        return;
      }
      if (l?.stage === "end") {
        Ae.value = !1;
        return;
      }
      $t();
    }
    async function Dt() {
      if (ol("copyAlt", {
        config: e.value,
        dataset: ge.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: ge.value
      }));
    }
    return Vt({
      getData: cl,
      getImage: gl,
      generatePdf: wt,
      generateCsv: Qe,
      generateImage: $t,
      generateSvg: Ot,
      hideSeries: fl,
      showSeries: hl,
      toggleTable: Tt,
      toggleTooltip: Lt,
      toggleAnnotator: et,
      toggleFullscreen: St,
      copyAlt: Dt
    }), (l, a) => (i(), c("div", {
      class: It(`vue-data-ui-component vue-ui-scatter ${ye.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      ref_key: "scatterChart",
      ref: se,
      id: `vue-ui-scatter_${X.value}`,
      style: ne(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.backgroundColor};${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: a[5] || (a[5] = () => y(mt)(!0)),
      onMouseleave: a[6] || (a[6] = () => y(mt)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (i(), le(y(Kt), {
        key: 0,
        svgRef: y(Te),
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        active: Oe.value,
        isCursorPointer: ce.value,
        onClose: et
      }, {
        "annotator-action-close": T(() => [
          z(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": T(({ color: t }) => [
          z(l.$slots, "annotator-action-color", K(J({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": T(({ mode: t }) => [
          z(l.$slots, "annotator-action-draw", K(J({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": T(({ disabled: t }) => [
          z(l.$slots, "annotator-action-undo", K(J({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": T(({ disabled: t }) => [
          z(l.$slots, "annotator-action-redo", K(J({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": T(({ disabled: t }) => [
          z(l.$slots, "annotator-action-delete", K(J({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : b("", !0),
      il.value ? (i(), c("div", {
        key: 1,
        ref_key: "noTitle",
        ref: dt,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : b("", !0),
      e.value.style.title.text ? (i(), c("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: vt,
        style: "width:100%;background:transparent"
      }, [
        (i(), le(na, {
          key: `title_${ht.value}`,
          config: {
            title: {
              cy: "scatter-div-title",
              ...e.value.style.title
            },
            subtitle: {
              cy: "scatter-div-subtitle",
              ...e.value.style.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : b("", !0),
      w("div", {
        id: `legend-top-${X.value}`
      }, null, 8, ca),
      e.value.userOptions.show && rt.value && (y(bt) || y(je)) ? (i(), le(y(Jt), {
        ref_key: "userOptionsRef",
        ref: Ne,
        key: `user_options_${it.value}`,
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        isImaging: y(He),
        isPrinting: y(qe),
        uid: X.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isTooltip: H.value.showTooltip,
        isFullscreen: ye.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: se.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Oe.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: ce.value,
        onToggleFullscreen: St,
        onGeneratePdf: y(wt),
        onGenerateCsv: Qe,
        onGenerateImage: Tl,
        onGenerateSvg: Ot,
        onToggleTable: Tt,
        onToggleTooltip: Lt,
        onToggleAnnotator: et,
        onCopyAlt: Dt,
        style: ne({
          visibility: y(bt) ? y(je) ? "visible" : "hidden" : "visible"
        })
      }, Ft({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: T(({ isOpen: t, color: o }) => [
            z(l.$slots, "menuIcon", K(J({ isOpen: t, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: T(() => [
            z(l.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: T(() => [
            z(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionCsv ? {
          name: "optionCsv",
          fn: T(() => [
            z(l.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: T(() => [
            z(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: T(() => [
            z(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionTable ? {
          name: "optionTable",
          fn: T(() => [
            z(l.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: T(({ toggleFullscreen: t, isFullscreen: o }) => [
            z(l.$slots, "optionFullscreen", K(J({ toggleFullscreen: t, isFullscreen: o })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: T(({ toggleAnnotator: t, isAnnotator: o }) => [
            z(l.$slots, "optionAnnotator", K(J({ toggleAnnotator: t, isAnnotator: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: T(({ altCopy: t }) => [
            z(l.$slots, "optionAltCopy", K(J({ altCopy: t })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isTooltip", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : b("", !0),
      (i(), c("svg", {
        ref_key: "svgRef",
        ref: Te,
        xmlns: y(Yl),
        class: It({ "vue-data-ui-fullscreen--on": ye.value, "vue-data-ui-fulscreen--off": !ye.value, animated: e.value.useCssAnimation }),
        viewBox: `0 0 ${N.value.width <= 0 ? 10 : N.value.width} ${N.value.height <= 0 ? 10 : N.value.height}`,
        style: ne(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.color}`),
        onMouseleave: tt
      }, [
        Ce(y(Qt)),
        l.$slots["chart-background"] ? (i(), c("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: N.value.width <= 0 ? 10 : N.value.width,
          height: N.value.height <= 0 ? 10 : N.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          z(l.$slots, "chart-background", {}, void 0, !0)
        ], 8, da)) : b("", !0),
        e.value.style.layout.axis.show ? (i(), c("g", ha, [
          w("line", {
            x1: I.value.x,
            x2: I.value.x,
            y1: n.value.top,
            y2: n.value.bottom,
            stroke: e.value.style.layout.axis.stroke,
            "stroke-width": e.value.style.layout.axis.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, fa),
          w("line", {
            x1: n.value.left,
            x2: n.value.right,
            y1: I.value.y,
            y2: I.value.y,
            stroke: e.value.style.layout.axis.stroke,
            "stroke-width": e.value.style.layout.axis.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, ga)
        ])) : b("", !0),
        e.value.style.layout.marginalBars.show ? (i(), c("g", pa, [
          w("defs", null, [
            w("linearGradient", {
              id: `marginal_x_${X.value}`,
              x1: "0%",
              y1: "0%",
              x2: "0%",
              y2: "100%"
            }, [
              w("stop", {
                offset: "0%",
                "stop-color": e.value.style.layout.marginalBars.fill
              }, null, 8, ba),
              w("stop", {
                offset: "100%",
                "stop-color": e.value.style.backgroundColor
              }, null, 8, ka)
            ], 8, ma),
            w("linearGradient", {
              id: `marginal_y_${X.value}`,
              x1: "0%",
              x2: "100%",
              y1: "0%",
              y2: "0%"
            }, [
              w("stop", {
                offset: "0%",
                "stop-color": e.value.style.backgroundColor
              }, null, 8, wa),
              w("stop", {
                offset: "100%",
                "stop-color": e.value.style.layout.marginalBars.fill
              }, null, 8, $a)
            ], 8, xa)
          ]),
          (i(!0), c(Q, null, ae(A.value.x, (t, o) => (i(), c("g", null, [
            t && A.value.avgX[o] ? (i(), c("rect", {
              key: 0,
              x: A.value.avgX[o] - n.value.width / k.value / 2,
              y: n.value.top - e.value.style.layout.marginalBars.offset - t / A.value.maxX * e.value.style.layout.marginalBars.size,
              width: n.value.width / k.value <= 0 ? 1e-4 : n.value.width / k.value,
              height: t / A.value.maxX * e.value.style.layout.marginalBars.size <= 0 ? 1e-4 : t / A.value.maxX * e.value.style.layout.marginalBars.size,
              fill: e.value.style.layout.marginalBars.useGradient ? `url(#marginal_x_${X.value})` : e.value.style.layout.marginalBars.fill,
              style: ne([`opacity:${e.value.style.layout.marginalBars.opacity}`, { "pointer-events": "none" }]),
              stroke: e.value.style.backgroundColor,
              "stroke-width": e.value.style.layout.marginalBars.strokeWidth,
              rx: e.value.style.layout.marginalBars.borderRadius
            }, null, 12, _a)) : b("", !0),
            A.value.avgX[o] ? (i(), c("rect", {
              key: 1,
              x: A.value.avgX[o] - n.value.width / k.value / 2,
              y: n.value.top - e.value.style.layout.marginalBars.offset - e.value.style.layout.marginalBars.size,
              width: n.value.width / k.value <= 0 ? 1e-4 : n.value.width / k.value,
              height: Math.max(0.1, e.value.style.layout.marginalBars.size),
              fill: "transparent",
              onMouseenter: (s) => pl(o),
              onMouseleave: a[0] || (a[0] = (s) => tt())
            }, null, 40, Ma)) : b("", !0),
            A.value.avgX[o] && be.value != null && be.value === o ? (i(), c("g", Ca, [
              w("rect", {
                x: A.value.avgX[o] - n.value.width / k.value / 2,
                y: n.value.top,
                width: n.value.width / k.value <= 0 ? 1e-4 : n.value.width / k.value,
                height: n.value.height,
                fill: e.value.style.layout.marginalBars.highlighter.color,
                "fill-opacity": e.value.style.layout.marginalBars.highlighter.opacity
              }, null, 8, Aa),
              w("line", {
                x1: A.value.avgX[o] - n.value.width / k.value / 2,
                x2: A.value.avgX[o] - n.value.width / k.value / 2,
                y1: 0,
                y2: n.value.top + n.value.height,
                stroke: e.value.style.layout.marginalBars.highlighter.stroke,
                "stroke-dasharray": e.value.style.layout.marginalBars.highlighter.strokeDasharray,
                "stroke-width": e.value.style.layout.marginalBars.highlighter.strokeWidth,
                style: { transition: "none !important", animation: "none !important" }
              }, null, 8, Sa),
              w("line", {
                x1: A.value.avgX[o] - n.value.width / k.value / 2 + (n.value.width / k.value <= 0 ? 1e-4 : n.value.width / k.value),
                x2: A.value.avgX[o] - n.value.width / k.value / 2 + (n.value.width / k.value <= 0 ? 1e-4 : n.value.width / k.value),
                y1: 0,
                y2: n.value.top + n.value.height,
                stroke: e.value.style.layout.marginalBars.highlighter.stroke,
                "stroke-dasharray": e.value.style.layout.marginalBars.highlighter.strokeDasharray,
                "stroke-width": e.value.style.layout.marginalBars.highlighter.strokeWidth,
                style: { transition: "none !important", animation: "none !important" }
              }, null, 8, Ta)
            ])) : b("", !0)
          ]))), 256)),
          (i(!0), c(Q, null, ae(A.value.y, (t, o) => (i(), c("g", null, [
            t && A.value.avgY[o] ? (i(), c("rect", {
              key: 0,
              x: n.value.right + e.value.style.layout.marginalBars.offset,
              y: A.value.avgY[o] - n.value.height / k.value / 2,
              height: n.value.height / k.value <= 0 ? 1e-4 : n.value.height / k.value,
              width: t / A.value.maxY * e.value.style.layout.marginalBars.size <= 0 ? 1e-4 : t / A.value.maxY * e.value.style.layout.marginalBars.size,
              fill: e.value.style.layout.marginalBars.useGradient ? `url(#marginal_y_${X.value})` : e.value.style.layout.marginalBars.fill,
              style: ne([`opacity:${e.value.style.layout.marginalBars.opacity}`, { "pointer-events": "none" }]),
              stroke: e.value.style.backgroundColor,
              "stroke-width": e.value.style.layout.marginalBars.strokeWidth,
              rx: e.value.style.layout.marginalBars.borderRadius
            }, null, 12, La)) : b("", !0),
            A.value.avgY[o] ? (i(), c("rect", {
              key: 1,
              x: n.value.right + e.value.style.layout.marginalBars.offset,
              y: A.value.avgY[o] - n.value.height / k.value / 2,
              width: Math.max(0.1, e.value.style.layout.marginalBars.size),
              height: n.value.height / k.value <= 0 ? 1e-4 : n.value.height / k.value,
              fill: "transparent",
              onMouseenter: (s) => ml(o),
              onMouseleave: a[1] || (a[1] = (s) => tt())
            }, null, 40, Ba)) : b("", !0),
            A.value.avgY[o] && ke.value != null && ke.value === o ? (i(), c("g", Pa, [
              w("rect", {
                x: n.value.left,
                y: A.value.avgY[o] - n.value.height / k.value / 2,
                width: n.value.width,
                height: n.value.height / k.value <= 0 ? 1e-4 : n.value.height / k.value,
                fill: e.value.style.layout.marginalBars.highlighter.color,
                "fill-opacity": e.value.style.layout.marginalBars.highlighter.opacity
              }, null, 8, za),
              w("line", {
                x1: n.value.left,
                x2: N.value.width,
                y1: A.value.avgY[o] - n.value.height / k.value / 2,
                y2: A.value.avgY[o] - n.value.height / k.value / 2,
                stroke: e.value.style.layout.marginalBars.highlighter.stroke,
                "stroke-dasharray": e.value.style.layout.marginalBars.highlighter.strokeDasharray,
                "stroke-width": e.value.style.layout.marginalBars.highlighter.strokeWidth,
                style: { transition: "none !important", animation: "none !important" }
              }, null, 8, Oa),
              w("line", {
                x1: n.value.left,
                x2: N.value.width,
                y1: A.value.avgY[o] - n.value.height / k.value / 2 + (n.value.height / k.value <= 0 ? 1e-4 : n.value.height / k.value),
                y2: A.value.avgY[o] - n.value.height / k.value / 2 + (n.value.height / k.value <= 0 ? 1e-4 : n.value.height / k.value),
                stroke: e.value.style.layout.marginalBars.highlighter.stroke,
                "stroke-dasharray": e.value.style.layout.marginalBars.highlighter.strokeDasharray,
                "stroke-width": e.value.style.layout.marginalBars.highlighter.strokeWidth,
                style: { transition: "none !important", animation: "none !important" }
              }, null, 8, Da)
            ])) : b("", !0)
          ]))), 256)),
          e.value.style.layout.marginalBars.showLines ? (i(), c("g", Ea, [
            (i(!0), c(Q, null, ae(yl.value, (t) => (i(), c(Q, null, [
              O.value.includes(t.id) ? b("", !0) : (i(), c("path", {
                key: 0,
                d: `M ${t.dX}`,
                stroke: e.value.style.backgroundColor,
                "stroke-width": e.value.style.layout.marginalBars.linesStrokeWidth + 1,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                fill: "none"
              }, null, 8, Ia)),
              O.value.includes(t.id) ? b("", !0) : (i(), c("path", {
                key: 1,
                d: `M ${t.dX}`,
                stroke: t.color,
                "stroke-width": e.value.style.layout.marginalBars.linesStrokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                fill: "none"
              }, null, 8, Fa)),
              O.value.includes(t.id) ? b("", !0) : (i(), c("path", {
                key: 2,
                d: `M ${t.dY}`,
                stroke: e.value.style.backgroundColor,
                "stroke-width": e.value.style.layout.marginalBars.linesStrokeWidth + 1,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                fill: "none"
              }, null, 8, Wa)),
              O.value.includes(t.id) ? b("", !0) : (i(), c("path", {
                key: 3,
                d: `M ${t.dY}`,
                stroke: t.color,
                "stroke-width": e.value.style.layout.marginalBars.linesStrokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                fill: "none"
              }, null, 8, Na))
            ], 64))), 256))
          ])) : b("", !0)
        ])) : b("", !0),
        e.value.style.layout.plots.giftWrap.show ? (i(), c("g", Xa, [
          (i(!0), c(Q, null, ae(V.value, (t, o) => (i(), c("g", null, [
            t.plots.length > 2 ? (i(), c("polygon", {
              key: 0,
              points: y(Rl)({ series: t.plots }),
              fill: y(Ie)(t.color, e.value.style.layout.plots.giftWrap.fillOpacity * 100),
              "stroke-width": e.value.style.layout.plots.giftWrap.strokeWidth,
              "stroke-dasharray": e.value.style.layout.plots.giftWrap.strokeDasharray,
              stroke: t.color,
              "stroke-linejoin": "round",
              "stroke-linecap": "round"
            }, null, 8, Ya)) : b("", !0)
          ]))), 256))
        ])) : b("", !0),
        e.value.usePerformanceMode ? b("", !0) : (i(!0), c(Q, { key: 4 }, ae(V.value, (t, o) => (i(), c("g", null, [
          !t.shape || t.shape === "circle" ? (i(), c("g", Ra, [
            (i(!0), c(Q, null, ae(t.plots, (s, u) => (i(), c("circle", {
              cx: s.x,
              cy: s.y,
              r: R.value && R.value === s.id ? s.weight * 2 : s.weight,
              fill: y(Ie)(t.color, e.value.style.layout.plots.opacity * 100),
              stroke: e.value.style.layout.plots.stroke,
              "stroke-width": e.value.style.layout.plots.strokeWidth,
              style: ne(`opacity:${R.value && R.value === s.id ? 1 : e.value.style.layout.plots.significance.useDistanceOpacity ? 1 - Math.abs(s.deviation) / Mt.value : e.value.style.layout.plots.significance.show && Math.abs(s.deviation) > e.value.style.layout.plots.significance.deviationThreshold ? e.value.style.layout.plots.significance.opacity : 1}`),
              onMouseover: (r) => Ke(s, o),
              onMouseleave: (r) => Be(s, o),
              onClick: (r) => Je(s, o)
            }, null, 44, Ua))), 256))
          ])) : (i(), c("g", Va, [
            (i(!0), c(Q, null, ae(t.plots, (s, u) => (i(), le(ut, {
              plot: { x: s.x, y: s.y },
              radius: R.value && R.value === s.id ? s.weight * 2 : s.weight,
              shape: t.shape,
              color: y(Ie)(t.color, e.value.style.layout.plots.opacity * 100),
              stroke: e.value.style.layout.plots.stroke,
              strokeWidth: e.value.style.layout.plots.strokeWidth,
              style: ne(`opacity:${R.value && R.value === s.id ? 1 : e.value.style.layout.plots.significance.useDistanceOpacity ? 1 - Math.abs(s.deviation) / Mt.value : e.value.style.layout.plots.significance.show && Math.abs(s.deviation) > e.value.style.layout.plots.significance.deviationThreshold ? e.value.style.layout.plots.significance.opacity : 1}`),
              onMouseover: (r) => Ke(s, o),
              onMouseleave: (r) => Be(s, o),
              onClick: (r) => Je(s, o)
            }, null, 8, ["plot", "radius", "shape", "color", "stroke", "strokeWidth", "style", "onMouseover", "onMouseleave", "onClick"]))), 256))
          ]))
        ]))), 256)),
        e.value.usePerformanceMode ? (i(), c(Q, { key: 5 }, [
          w("g", {
            "clip-path": `url(#clip_path_${X.value})`
          }, [
            (i(!0), c(Q, null, ae(kl.value, (t) => (i(), c("path", {
              key: t.id,
              d: t.d,
              fill: t.fill,
              stroke: t.stroke,
              "stroke-width": t.strokeWidth,
              "stroke-opacity": t.strokeOpacity,
              "vector-effect": "non-scaling-stroke",
              "paint-order": "fill"
            }, null, 8, Ga))), 128))
          ], 8, ja),
          g.value && e.value.style.layout.plots.selectors.show ? (i(), c("g", qa, [
            Ce(ut, {
              shape: g.value.shape || "circle",
              color: g.value.color,
              plot: { x: g.value.x, y: g.value.y },
              radius: Math.max(4 * Bt[g.value.shape || "circle"], g.value.weight * 2),
              stroke: e.value.style.layout.plots.stroke,
              strokeWidth: e.value.style.layout.plots.strokeWidth
            }, null, 8, ["shape", "color", "plot", "radius", "stroke", "strokeWidth"])
          ])) : b("", !0),
          w("rect", {
            x: n.value.left,
            y: n.value.top,
            width: Math.max(1e-4, n.value.width),
            height: Math.max(1e-4, n.value.height),
            fill: "transparent",
            onMousemove: a[2] || (a[2] = (...t) => y(Pt) && y(Pt)(...t)),
            onMouseleave: wl,
            onClick: $l
          }, null, 40, Ha)
        ], 64)) : b("", !0),
        g.value && e.value.style.layout.plots.selectors.show ? (i(), c("g", Za, [
          w("line", {
            x1: I.value.x,
            x2: g.value.x,
            y1: g.value.y,
            y2: g.value.y,
            stroke: e.value.style.layout.plots.selectors.stroke,
            "stroke-width": e.value.style.layout.plots.selectors.strokeWidth,
            "stroke-dasharray": e.value.style.layout.plots.selectors.strokeDasharray,
            "stroke-linecap": "round",
            class: "line-pointer"
          }, null, 8, Ka),
          w("line", {
            x1: g.value.x,
            x2: g.value.x,
            y1: I.value.y,
            y2: g.value.y,
            stroke: e.value.style.layout.plots.selectors.stroke,
            "stroke-width": e.value.style.layout.plots.selectors.strokeWidth,
            "stroke-dasharray": e.value.style.layout.plots.selectors.strokeDasharray,
            "stroke-linecap": "round",
            class: "line-pointer"
          }, null, 8, Ja),
          w("text", {
            x: I.value.x + (g.value.x > I.value.x ? -6 : 6),
            y: g.value.y + e.value.style.layout.plots.selectors.labels.fontSize / 3,
            "font-size": e.value.style.layout.plots.selectors.labels.fontSize,
            fill: e.value.style.layout.plots.selectors.labels.color,
            "font-weight": e.value.style.layout.plots.selectors.labels.bold ? "bold" : "normal",
            "text-anchor": g.value.x > I.value.x ? "end" : "start"
          }, G(y(ve)(
            e.value.style.layout.plots.selectors.labels.y.formatter,
            y(q)(g.value.v.y),
            y(oe)({
              p: e.value.style.layout.plots.selectors.labels.prefix,
              v: y(q)(g.value.v.y),
              s: e.value.style.layout.plots.selectors.labels.suffix,
              r: e.value.style.layout.plots.selectors.labels.rounding
            }),
            { datapoint: g.value }
          )), 9, Qa),
          w("text", {
            x: g.value.x,
            y: I.value.y + (g.value.y > I.value.y ? -6 : e.value.style.layout.plots.selectors.labels.fontSize + 6),
            "font-size": e.value.style.layout.plots.selectors.labels.fontSize,
            fill: e.value.style.layout.plots.selectors.labels.color,
            "font-weight": e.value.style.layout.plots.selectors.labels.bold ? "bold" : "normal",
            "text-anchor": "middle"
          }, G(y(ve)(
            e.value.style.layout.plots.selectors.labels.y.formatter,
            y(q)(g.value.v.x),
            y(oe)({
              p: e.value.style.layout.plots.selectors.labels.prefix,
              v: y(q)(g.value.v.x),
              s: e.value.style.layout.plots.selectors.labels.suffix,
              r: e.value.style.layout.plots.selectors.labels.rounding
            }),
            { datapoint: g.value }
          )), 9, eo),
          w("circle", {
            cx: I.value.x,
            cy: g.value.y,
            r: e.value.style.layout.plots.selectors.markers.radius,
            fill: e.value.style.layout.plots.selectors.markers.fill,
            stroke: e.value.style.layout.plots.selectors.markers.stroke,
            "stroke-width": e.value.style.layout.plots.selectors.markers.strokeWidth,
            class: "line-pointer"
          }, null, 8, to),
          w("circle", {
            cx: g.value.x,
            cy: I.value.y,
            r: e.value.style.layout.plots.selectors.markers.radius,
            fill: e.value.style.layout.plots.selectors.markers.fill,
            stroke: e.value.style.layout.plots.selectors.markers.stroke,
            "stroke-width": e.value.style.layout.plots.selectors.markers.strokeWidth,
            class: "line-pointer"
          }, null, 8, lo),
          e.value.style.layout.plots.selectors.labels.showName ? (i(), c("text", {
            key: 0,
            x: g.value.x,
            y: g.value.y + (g.value.y < I.value.y ? -e.value.style.layout.plots.selectors.labels.fontSize / 2 : e.value.style.layout.plots.selectors.labels.fontSize),
            "font-size": e.value.style.layout.plots.selectors.labels.fontSize,
            fill: e.value.style.layout.plots.selectors.labels.color,
            "font-weight": e.value.style.layout.plots.selectors.labels.bold ? "bold" : "normal",
            "text-anchor": g.value.x < n.value.left + 100 ? "start" : g.value.x > n.value.right - 100 ? "end" : g.value.x > I.value.x ? "start" : "end"
          }, G(g.value.v.name), 9, ao)) : b("", !0)
        ])) : b("", !0),
        e.value.style.layout.dataLabels.xAxis.show ? (i(), c("g", {
          key: 7,
          ref_key: "xAxisLabelLeft",
          ref: Xe
        }, [
          w("text", {
            id: `vue-ui-scatter-xAxis-label-${X.value}`,
            transform: `translate(${e.value.style.layout.dataLabels.xAxis.fontSize}, ${n.value.top + n.value.height / 2}), rotate(-90)`,
            "text-anchor": "middle",
            "font-size": e.value.style.layout.dataLabels.xAxis.fontSize,
            "font-weight": e.value.style.layout.dataLabels.xAxis.bold ? "bold" : "normal",
            fill: e.value.style.layout.dataLabels.xAxis.color
          }, G(e.value.style.layout.dataLabels.xAxis.name), 9, oo),
          w("text", {
            "text-anchor": "middle",
            "font-size": e.value.style.layout.dataLabels.xAxis.fontSize,
            fill: e.value.style.layout.dataLabels.xAxis.color,
            transform: `translate(${e.value.style.layout.dataLabels.xAxis.name ? e.value.style.layout.dataLabels.xAxis.fontSize * 3 : 0}, ${I.value.y + e.value.style.layout.dataLabels.xAxis.fontSize / 3}), rotate(-90)`
          }, G(y(ve)(
            e.value.style.layout.plots.selectors.labels.x.formatter,
            y(q)(E.value.xMin),
            y(oe)({
              p: e.value.style.layout.plots.selectors.labels.prefix,
              v: y(q)(E.value.xMin),
              s: e.value.style.layout.plots.selectors.labels.suffix,
              r: e.value.style.layout.dataLabels.xAxis.rounding
            })
          )), 9, so)
        ], 512)) : b("", !0),
        e.value.style.layout.dataLabels.xAxis.show ? (i(), c("text", {
          key: 8,
          ref_key: "xAxisLabelRight",
          ref: sl,
          "text-anchor": "middle",
          transform: `translate(${n.value.right + e.value.style.layout.padding.right + 6}, ${I.value.y + e.value.style.layout.dataLabels.xAxis.fontSize / 3}), rotate(-90)`,
          "font-size": e.value.style.layout.dataLabels.xAxis.fontSize,
          fill: e.value.style.layout.dataLabels.xAxis.color
        }, G(y(ve)(
          e.value.style.layout.plots.selectors.labels.x.formatter,
          y(q)(E.value.xMax),
          y(oe)({
            p: e.value.style.layout.plots.selectors.labels.prefix,
            v: y(q)(E.value.xMax),
            s: e.value.style.layout.plots.selectors.labels.suffix,
            r: e.value.style.layout.dataLabels.xAxis.rounding
          })
        )), 9, no)) : b("", !0),
        e.value.style.layout.dataLabels.yAxis.show ? (i(), c("text", {
          key: 9,
          ref_key: "yAxisLabelTop",
          ref: nl,
          x: I.value.x,
          y: n.value.top - e.value.style.layout.dataLabels.yAxis.fontSize,
          "text-anchor": "middle",
          "font-size": e.value.style.layout.dataLabels.yAxis.fontSize,
          fill: e.value.style.layout.dataLabels.yAxis.color
        }, G(y(ve)(
          e.value.style.layout.plots.selectors.labels.y.formatter,
          y(q)(E.value.yMax),
          y(oe)({
            p: e.value.style.layout.plots.selectors.labels.prefix,
            v: y(q)(E.value.yMax),
            s: e.value.style.layout.plots.selectors.labels.suffix,
            r: e.value.style.layout.dataLabels.yAxis.rounding
          })
        )), 9, uo)) : b("", !0),
        e.value.style.layout.dataLabels.yAxis.show ? (i(), c("g", {
          key: 10,
          ref_key: "yAxisLabelBottom",
          ref: Ye
        }, [
          w("text", {
            x: I.value.x,
            y: N.value.height - e.value.style.layout.dataLabels.yAxis.fontSize * 2,
            "text-anchor": "middle",
            "font-size": e.value.style.layout.dataLabels.yAxis.fontSize,
            fill: e.value.style.layout.dataLabels.yAxis.color
          }, G(y(ve)(
            e.value.style.layout.plots.selectors.labels.y.formatter,
            y(q)(E.value.yMin),
            y(oe)({
              p: e.value.style.layout.plots.selectors.labels.prefix,
              v: y(q)(E.value.yMin),
              s: e.value.style.layout.plots.selectors.labels.suffix,
              r: e.value.style.layout.dataLabels.yAxis.rounding
            })
          )), 9, ro),
          w("text", {
            "text-anchor": "middle",
            "font-size": e.value.style.layout.dataLabels.yAxis.fontSize,
            "font-weight": e.value.style.layout.dataLabels.yAxis.bold ? "bold" : "normal",
            fill: e.value.style.layout.dataLabels.yAxis.color,
            x: n.value.left + n.value.width / 2,
            y: N.value.height
          }, G(e.value.style.layout.dataLabels.yAxis.name), 9, io)
        ], 512)) : b("", !0),
        w("clipPath", {
          id: `clip_path_${X.value}`
        }, [
          w("rect", {
            x: n.value.left,
            y: n.value.top,
            width: n.value.width <= 0 ? 1e-4 : n.value.width,
            height: n.value.height <= 0 ? 1e-4 : n.value.height
          }, null, 8, co)
        ], 8, vo),
        e.value.style.layout.correlation.show ? (i(), c("g", yo, [
          (i(!0), c(Q, null, ae(V.value, (t, o) => (i(), c("line", {
            x1: t.correlation.x1,
            x2: t.correlation.x2,
            y1: t.correlation.y1,
            y2: t.correlation.y2,
            "stroke-dasharray": e.value.style.layout.correlation.strokeDasharray,
            stroke: t.color,
            "stroke-width": e.value.style.layout.correlation.strokeWidth,
            "clip-path": `url(#clip_path_${X.value})`
          }, null, 8, ho))), 256)),
          (i(!0), c(Q, null, ae(V.value, (t, o) => (i(), c("g", null, [
            e.value.style.layout.correlation.label.show ? (i(), c("text", {
              key: 0,
              x: t.correlation.x2,
              y: t.correlation.y2,
              fill: e.value.style.layout.correlation.label.useSerieColor ? t.color : e.value.style.layout.correlation.label.color,
              "text-anchor": "end",
              "font-size": e.value.style.layout.correlation.label.fontSize,
              "font-weight": e.value.style.layout.correlation.label.bold ? "bold" : "normal"
            }, G(y(oe)({
              v: y(q)(t.correlation.coefficient),
              r: e.value.style.layout.correlation.label.roundingValue
            })), 9, fo)) : b("", !0)
          ]))), 256))
        ])) : b("", !0),
        z(l.$slots, "svg", {
          svg: {
            ...N.value,
            drawingArea: {
              ...n.value,
              zero: I.value
            },
            data: ge.value,
            isPrintingImg: y(qe) | y(He) | Ae.value,
            isPrintingSvg: Se.value
          }
        }, void 0, !0)
      ], 46, ya)),
      l.$slots.watermark ? (i(), c("div", go, [
        z(l.$slots, "watermark", K(J({ isPrinting: y(qe) || y(He) || Ae.value || Se.value })), void 0, !0)
      ])) : b("", !0),
      w("div", {
        id: `legend-bottom-${X.value}`
      }, null, 8, po),
      pt.value ? (i(), le(zl, {
        key: 5,
        to: e.value.style.legend.position === "top" ? `#legend-top-${X.value}` : `#legend-bottom-${X.value}`
      }, [
        w("div", {
          ref_key: "chartLegend",
          ref: ct
        }, [
          e.value.style.legend.show ? (i(), le(sa, {
            key: `legend_${gt.value}`,
            legendSet: Z.value,
            config: vl.value,
            isCursorPointer: ce.value,
            onClickMarker: a[3] || (a[3] = ({ legend: t }) => Pe(t.id))
          }, {
            item: T(({ legend: t }) => [
              w("div", {
                onClick: (o) => t.segregate(),
                style: ne(`opacity:${O.value.includes(t.id) ? 0.5 : 1}`)
              }, G(t.name), 13, mo)
            ]),
            legendToggle: T(() => [
              Z.value.length > 2 && e.value.style.legend.selectAllToggle.show && !y(Re) ? (i(), le(ra, {
                key: 0,
                backgroundColor: e.value.style.legend.selectAllToggle.backgroundColor,
                color: e.value.style.legend.selectAllToggle.color,
                fontSize: e.value.style.legend.fontSize,
                checked: O.value.length > 0,
                isCursorPointer: ce.value,
                onToggle: dl
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : b("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : z(l.$slots, "legend", {
            key: 1,
            legend: Z.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : b("", !0),
      l.$slots.source ? (i(), c("div", {
        key: 6,
        ref_key: "source",
        ref: yt,
        dir: "auto"
      }, [
        z(l.$slots, "source", {}, void 0, !0)
      ], 512)) : b("", !0),
      Ce(y(Gt), {
        teleportTo: e.value.style.tooltip.teleportTo,
        show: H.value.showTooltip && Fe.value,
        backgroundColor: e.value.style.tooltip.backgroundColor,
        color: e.value.style.tooltip.color,
        borderRadius: e.value.style.tooltip.borderRadius,
        borderColor: e.value.style.tooltip.borderColor,
        borderWidth: e.value.style.tooltip.borderWidth,
        fontSize: e.value.style.tooltip.fontSize,
        backgroundOpacity: e.value.style.tooltip.backgroundOpacity,
        position: e.value.style.tooltip.position,
        offsetY: e.value.style.tooltip.offsetY,
        parent: se.value,
        content: We.value,
        isFullscreen: ye.value,
        isCustom: e.value.style.tooltip.customFormat && typeof e.value.style.tooltip.customFormat == "function",
        smooth: e.value.style.tooltip.smooth,
        backdropFilter: e.value.style.tooltip.backdropFilter,
        smoothForce: e.value.style.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": T(() => [
          z(l.$slots, "tooltip-before", K(J({ ...Ze.value })), void 0, !0)
        ]),
        "tooltip-after": T(() => [
          z(l.$slots, "tooltip-after", K(J({ ...Ze.value })), void 0, !0)
        ]),
        default: T(() => [
          e.value.style.tooltip.showShape ? (i(), c("div", bo, [
            (i(), c("svg", ko, [
              Ce(ut, {
                shape: g.value.shape,
                color: g.value.color,
                plot: { x: 10, y: 10 },
                radius: 7
              }, null, 8, ["shape", "color"])
            ]))
          ])) : b("", !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      rt.value && e.value.userOptions.buttons.table ? (i(), le(Ol(we.value.component), Dl({ key: 7 }, we.value.props, {
        ref_key: "tableUnit",
        ref: xe,
        onClose: zt
      }), Ft({
        content: T(() => [
          (i(), le(y(Zt), {
            key: `table_${ft.value}`,
            colNames: ze.value.colNames,
            head: ze.value.head,
            body: ze.value.body,
            config: ze.value.config,
            title: e.value.table.useDialog ? "" : we.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: ce.value,
            onClose: zt
          }, {
            th: T(({ th: t }) => [
              Wt(G(t), 1)
            ]),
            td: T(({ td: t }) => [
              t.shape ? (i(), c("div", xo, [
                w("span", null, G(t.content), 1)
              ])) : (i(), c("div", {
                key: 1,
                innerHTML: t
              }, null, 8, wo))
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: T(() => [
            Wt(G(we.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: T(() => [
            w("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[4] || (a[4] = (t) => Qe(e.value.userOptions.callbacks.csv)),
              style: ne({ cursor: ce.value ? "pointer" : "default" })
            }, [
              Ce(y(qt), {
                name: "fileCsv",
                stroke: we.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : b("", !0),
      z(l.$slots, "skeleton", {}, () => [
        y(Re) ? (i(), le(Jl, { key: 0 })) : b("", !0)
      ], !0)
    ], 46, va));
  }
}, Yo = /* @__PURE__ */ ia($o, [["__scopeId", "data-v-b904e445"]]);
export {
  Yo as default
};
