// vue-data-ui/dist/vue-ui-table-BHuXI4Eq.js
import { useCssVars as Gt, defineAsyncComponent as qt, ref as f, computed as x, onMounted as Qt, watch as Fe, openBlock as r, createElementBlock as i, normalizeStyle as h, createElementVNode as s, toDisplayString as d, withDirectives as le, withKeys as ve, withModifiers as de, vModelText as he, createVNode as je, createCommentVNode as b, Fragment as L, renderList as j, normalizeClass as D, createTextVNode as z, unref as V, vModelSelect as bt, withCtx as mt, createBlock as kt, nextTick as ke } from "vue";
import { j as el, t as tl, y as ll, x as Ve, u as wt, W as al, Y as ol, l as ae, a0 as nl, a1 as sl, p as We } from "./lib-GwPnJSrY.js";
import { u as ul } from "./useConfig-CIgyKZ6R.js";
import Ct from "./vue-ui-xy-f67V1xjd.js";
import rl from "./vue-ui-donut-CzCgCVsx.js";
import il from "./BaseIcon-CNx8Tzl5.js";
import { _ as cl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const vl = ["innerHTML"], dl = ["data-is-open"], hl = { class: "vue-ui-table-export-hub-options" }, gl = { class: "vue-ui-table-export-hub-option-wrapper" }, pl = { class: "label" }, fl = ["innerHTML"], yl = { class: "vue-ui-table-export-hub-option-wrapper" }, bl = { class: "label" }, ml = ["innerHTML"], kl = { class: "vue-ui-table-dialog-field" }, wl = {
  class: "label vue-ui-table-dialog-input-label",
  style: { width: "100%" }
}, Cl = { style: { width: "fit-content" } }, $l = { class: "vue-ui-table" }, Ml = { key: 0 }, xl = { key: 0 }, Tl = { key: 1 }, _l = {
  key: 0,
  style: { display: "flex", "align-items": "center", "justify-content": "flex-end" }
}, Sl = ["innerHTML"], Nl = {
  key: 0,
  style: { "margin-left": "3px" }
}, Ol = { key: 0 }, Ll = { class: "th-filter" }, Dl = {
  key: 0,
  class: "th-date"
}, El = { class: "date-wrapper--inputs" }, Il = { class: "date-fieldset" }, Al = ["for"], Pl = ["id", "onUpdate:modelValue", "onInput"], Hl = { class: "date-fieldset" }, Bl = ["for"], Rl = ["id", "onUpdate:modelValue", "onInput"], zl = { class: "date-wrapper--button" }, Ul = ["onClick"], Fl = ["innerHTML"], jl = ["innerHTML"], Vl = ["innerHTML"], Wl = ["onClick", "disabled"], Jl = ["placeholder", "onUpdate:modelValue", "name"], Kl = ["onClick"], Xl = ["innerHTML"], Yl = ["innerHTML"], Zl = ["innerHTML"], Gl = ["onClick", "innerHTML"], ql = ["innerHTML"], Ql = {
  key: 5,
  class: "th-range-filter"
}, ea = ["for"], ta = ["id", "max", "min", "onUpdate:modelValue"], la = ["id", "max", "min", "onUpdate:modelValue"], aa = ["for"], oa = ["onClick", "disabled"], na = ["id"], sa = ["onClick"], ua = ["onClick", "onKeyup"], ra = ["innerHTML", "onClick", "onKeyup"], ia = ["data-row"], ca = ["data-row"], va = ["data-row", "onClick", "onKeyup", "id"], da = ["innerHTML"], ha = { key: 1 }, ga = { key: 2 }, pa = { key: 5 }, fa = ["innerHTML"], ya = { key: 0 }, ba = { style: { "margin-left": "12px" } }, ma = { class: "format-num" }, ka = { style: { "margin-left": "12px" } }, wa = {
  key: 0,
  class: "format-num"
}, Ca = {
  key: 1,
  class: "format-num"
}, $a = { key: 2 }, Ma = { style: { "margin-left": "12px" } }, xa = {
  key: 0,
  class: "format-num"
}, Ta = {
  key: 1,
  class: "format-num"
}, _a = { key: 2 }, Sa = {
  key: 1,
  class: "vue-ui-table-paginator format-num"
}, Na = {
  key: 2,
  class: "vue-ui-table-size-warning"
}, Oa = ["innerHTML"], La = {
  key: 4,
  class: "vue-ui-table-pagination format-num"
}, Da = ["innerHTML", "disabled"], Ea = ["disabled"], Ia = {
  for: "pageScroller",
  style: { "font-size": "14px" }
}, Aa = ["max", "value"], Pa = { key: 1 }, Ha = ["disabled"], Ba = ["innerHTML", "disabled"], Ra = { class: "vue-ui-table-chart-modal-options" }, za = ["innerHTML"], Ua = ["innerHTML"], Fa = ["innerHTML"], ja = { style: { width: "100%", height: "fit-content" } }, Va = { class: "vue-ui-table-fieldset" }, Wa = { class: "vue-ui-table-fieldset-wrapper" }, Ja = ["name", "id", "checked", "onInput"], Ka = ["for"], Xa = ["disabled"], Ya = ["innerHTML"], Za = { style: { width: "100%", "margin-bottom": "12px" } }, Ga = {
  key: 2,
  style: { width: "100%", "margin-bottom": "32px" }
}, qa = {
  __name: "vue-ui-table",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["page-change"],
  setup(m, { expose: $t, emit: Mt }) {
    Gt((e) => ({
      v86d313de: Dt.value,
      v5f87fb24: Et.value,
      v57e0136c: At.value,
      v1633ca2c: De.value,
      v785d6714: Pt.value,
      v79493d14: Ht.value,
      v43f025bc: Ee.value,
      v56d8582e: It.value,
      v0057ee80: Bt.value
    }));
    const xt = qt(() => import("./BaseDraggableDialog-DSpKTn3R.js")), g = m, Tt = Mt, { vue_ui_table: Je } = ul(), G = el();
    f(null);
    const Oe = f(null), se = f(null), S = f({
      height: 316,
      type: "bar",
      width: 512
    }), c = f({
      ASC: 1,
      BAR: "bar",
      DATE: "date",
      DESC: -1,
      DONUT: "donut",
      LINE: "line",
      NUMERIC: "numeric",
      PERCENTAGE: "percentage",
      TEXT: "text"
    }), _t = f(100), St = f(100);
    f(0), f(0), f(400), f(200);
    const Nt = f({
      CELL: "smart-td-selected",
      FIRST_TD: "smart-td-selected-first",
      LAST_TD: "smart-td-selected-last",
      ROW: "smart-td-selected-neighbor"
    }), we = f(void 0), Ce = f(void 0), v = f({
      col: void 0,
      rows: []
    }), k = f(0), y = f(20), ge = f(!1), $e = f(!1), U = f({}), E = f({}), pe = f({}), W = f({}), C = f({}), J = f({}), H = f({}), P = f({}), w = f({}), oe = f(g.config.rowsPerPage ? g.config.rowsPerPage : 25), Ot = f([.../* @__PURE__ */ new Set([10, 25, 50, 100, 250, 500, g.config.rowsPerPage ? g.config.rowsPerPage : 25, g.dataset.body.length])].sort((e, t) => e - t)), ue = f(void 0), F = f(void 0), Lt = f(void 0), X = f(!1), Ke = f(0), Me = f(""), re = f(""), I = f(!1), N = f(JSON.parse(JSON.stringify(g.dataset.body)).map((e, t) => ({
      ...e,
      absoluteIndex: t
    }))), Le = f(JSON.parse(JSON.stringify(g.dataset.body)).map((e, t) => ({
      ...e,
      absoluteIndex: t
    }))), K = f(JSON.parse(JSON.stringify(g.dataset.header)).map((e, t) => ({
      average: Object.hasOwn(e, "average") ? e.average : !1,
      decimals: Object.hasOwn(e, "decimals") ? e.decimals : 0,
      isMultiselect: Object.hasOwn(e, "isMultiselect") ? e.isMultiselect : !1,
      isPercentage: Object.hasOwn(e, "isPercentage") ? e.isPercentage : !1,
      isSearch: Object.hasOwn(e, "isSearch") ? e.isSearch : !1,
      isSort: Object.hasOwn(e, "isSort") ? e.isSort : !1,
      name: e.name,
      // this attribute is mandatory
      percentageTo: Object.hasOwn(e, "percentageTo") ? e.percentageTo : void 0,
      prefix: Object.hasOwn(e, "prefix") ? e.prefix : "",
      rangeFilter: Object.hasOwn(e, "rangeFilter") ? e.rangeFilter : !1,
      suffix: Object.hasOwn(e, "suffix") ? e.suffix : "",
      sum: Object.hasOwn(e, "sum") ? e.sum : !1,
      type: e.type,
      // this attribute is mandatory
      index: t
    }))), Xe = x(() => Oe.value ? Oe.value.getBoundingClientRect().height + 3 : 3), l = x(() => {
      if (!Object.keys(g.config || {}).length)
        return Je;
      const e = tl({
        defaultConfig: Je,
        userConfig: g.config
      });
      return ll(e);
    }), T = x(() => l.value.useCursorPointer), Dt = x(() => l.value.style.th.buttons.cancel.inactive.backgroundColor), Et = x(() => l.value.style.th.buttons.cancel.inactive.color), De = x(() => l.value.style.th.buttons.cancel.active.backgroundColor), Ee = x(() => l.value.style.th.buttons.filter.active.backgroundColor), It = x(() => l.value.style.th.buttons.filter.active.color), At = x(() => Ve(De.value, 0.33)), Pt = x(() => wt(De.value, 33)), Ht = x(() => Ve(Ee.value, 0.33)), Bt = x(() => wt(Ee, 33)), Ie = x(() => [...K.value].filter((e) => e.type === c.value.DATE)), Ye = x(() => ["", ...Ie.value.map((e) => e.name)]), Ze = x(() => {
      const e = Ie.value.find((t) => t.name === Me.value);
      return e ? e.index : null;
    }), $ = x(() => {
      const e = [];
      if (N.value.length)
        for (let t = 0; t < N.value.length; t += oe.value)
          e.push(N.value.slice(t, t + oe.value));
      return e;
    }), ne = x(() => $.value[k.value]), Ae = x(() => Ze.value == null ? [] : ne.value.map((e) => e.td[Ze.value])), xe = x(() => Object.keys(C.value).map((e) => ({
      index: e,
      name: g.dataset.header[e].name,
      options: C.value[e]
    }))), Ge = x(() => l.value.useChart && v.value.rows.length > 1), qe = x(() => g.dataset.header.map((e) => e.type).includes(c.value.NUMERIC)), M = x(() => ({
      arrowSort: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M3 9l4 -4l4 4m-4 -4v14" /><path d="M21 15l-4 4l-4 -4m4 4v-14" /></svg>`,
      bar: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M3 12m0 1a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v6a1 1 0 0 1 -1 1h-4a1 1 0 0 1 -1 -1z" /><path d="M9 8m0 1a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v10a1 1 0 0 1 -1 1h-4a1 1 0 0 1 -1 -1z" /><path d="M15 4m0 1a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v14a1 1 0 0 1 -1 1h-4a1 1 0 0 1 -1 -1z" /><path d="M4 20l14 0" /></svg>`,
      chart: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M4 19l16 0" /><path d="M4 15l4 -6l4 2l4 -5l4 4" /></svg>`,
      chevronDown: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M6 9l6 6l6 -6" /></svg>`,
      chevronLeft: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value * 1.6}" height="${y.value * 1.6}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M15 6l-6 6l6 6" /></svg>`,
      chevronRight: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value * 1.6}" height="${y.value * 1.6}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M9 6l6 6l-6 6" /></svg>`,
      donut: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value * 0.8}" height="${y.value * 0.8}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M12 3v5m4 4h5" /><path d="M8.929 14.582l-3.429 2.918" /><path d="M12 12m-4 0a4 4 0 1 0 8 0a4 4 0 1 0 -8 0" /><path d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0" /></svg>`,
      export: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M12.5 21h-7.5a2 2 0 0 1 -2 -2v-14a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v7.5" /><path d="M3 10h18" /><path d="M10 3v18" /><path d="M16 19h6" /><path d="M19 16l3 3l-3 3" /></svg>`,
      fileDownload: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M14 3v4a1 1 0 0 0 1 1h4" /><path d="M17 21h-10a2 2 0 0 1 -2 -2v-14a2 2 0 0 1 2 -2h7l5 5v11a2 2 0 0 1 -2 2z" /><path d="M12 17v-6" /><path d="M9.5 14.5l2.5 2.5l2.5 -2.5" /></svg>`,
      filter: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M4 4h16v2.172a2 2 0 0 1 -.586 1.414l-4.414 4.414v7l-6 2v-8.5l-4.48 -4.928a2 2 0 0 1 -.52 -1.345v-2.227z" /></svg>`,
      move: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 9l3 3l-3 3" /><path d="M15 12h6" /><path d="M6 9l-3 3l3 3" /><path d="M3 12h6" /><path d="M9 18l3 3l3 -3" /><path d="M12 15v6" /><path d="M15 6l-3 -3l-3 3" /><path d="M12 3v6" /></svg>`,
      sort09: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M4 15l3 3l3 -3" /><path d="M7 6v12" /><path d="M17 3a2 2 0 0 1 2 2v3a2 2 0 1 1 -4 0v-3a2 2 0 0 1 2 -2z" /><path d="M17 16m-2 0a2 2 0 1 0 4 0a2 2 0 1 0 -4 0" /><path d="M19 16v3a2 2 0 0 1 -2 2h-1.5" /></svg>`,
      sort90: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M4 15l3 3l3 -3" /><path d="M7 6v12" /><path d="M17 14a2 2 0 0 1 2 2v3a2 2 0 1 1 -4 0v-3a2 2 0 0 1 2 -2z" /><path d="M17 5m-2 0a2 2 0 1 0 4 0a2 2 0 1 0 -4 0" /><path d="M19 5v3a2 2 0 0 1 -2 2h-1.5" /></svg>`,
      sortAZ: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M15 10v-5c0 -1.38 .62 -2 2 -2s2 .62 2 2v5m0 -3h-4" /><path d="M19 21h-4l4 -7h-4" /><path d="M4 15l3 3l3 -3" /><path d="M7 6v12" /></svg>`,
      sortZA: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M15 21v-5c0 -1.38 .62 -2 2 -2s2 .62 2 2v5m0 -3h-4" /><path d="M19 10h-4l4 -7h-4" /><path d="M4 15l3 3l3 -3" /><path d="M7 6v12" /></svg>`,
      sum: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>`,
      table: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" fill="white" d="M 10 2, 21 2, 21 21, 10 21Z"/><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M3 5a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2v-14z" /><path d="M3 10h18" /><path d="M10 3v18" /></svg>`,
      warning: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value * 0.8}" height="${y.value * 0.8}" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M10.24 3.957l-8.422 14.06a1.989 1.989 0 0 0 1.7 2.983h16.845a1.989 1.989 0 0 0 1.7 -2.983l-8.423 -14.06a1.989 1.989 0 0 0 -3.4 0z" /><path d="M12 9v4" /><path d="M12 17h.01" /></svg>`,
      grip: `<svg xmlns="http://www.w3.org/2000/svg" width="${y.value}" height="${y.value}" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1" stroke-linecap="round" stroke-linejoin="round"><path d="M5 9m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0" /><path d="M5 15m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0" /><path d="M12 9m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0" /><path d="M12 15m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0" /><path d="M19 9m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0" /><path d="M19 15m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0" /></svg>`
    })), Qe = x(() => N.value.map((e) => e.td)), Te = x(() => ({
      sumPercentage: Number((v.value.rows.map((e) => e.value).reduce((e, t) => e + t, 0) * 100).toFixed(g.dataset.header[v.value.col].decimals)).toLocaleString(),
      sumRegular: Number(v.value.rows.map((e) => e.value).reduce((e, t) => e + t, 0).toFixed(g.dataset.header[v.value.col].decimals)).toLocaleString(),
      averagePercentage: Number((v.value.rows.map((e) => e.value).reduce((e, t) => e + t, 0) / v.value.rows.length * 100).toFixed(g.dataset.header[v.value.col].decimals)).toLocaleString(),
      averageRegular: Number((v.value.rows.map((e) => e.value).reduce((e, t) => e + t, 0) / v.value.rows.length).toFixed(g.dataset.header[v.value.col].decimals)).toLocaleString()
    })), Pe = f({
      width: 1e3,
      height: 500
    });
    function Rt({ width: e, height: t }) {
      Pe.value = { width: e, height: t };
    }
    const ie = x(() => {
      if (!Ge.value) return [];
      const e = Pe.value.height, t = Pe.value.width, o = v.value.rows.length, a = t / o, n = Math.max(...v.value.rows.map((R) => R.value)), u = Math.min(...v.value.rows.map((R) => R.value)), p = g.dataset.header[v.value.col].isPercentage && g.dataset.header[v.value.col].percentageTo, O = S.value.type === c.value.DONUT && F.value && F.value.name, Y = g.dataset.header[v.value.col].name + (p ? ` / ${g.dataset.header[J.value[v.value.col].referenceIndex].name}` : "") + (O ? ` ${l.value.translations.by} ${F.value.name}` : ""), Z = g.dataset.header[v.value.col].prefix, ee = g.dataset.header[v.value.col].suffix, be = [
        {
          name: Y,
          series: v.value.rows.map((R) => R.value),
          type: "line",
          useProgression: !0,
          smooth: l.value.style.chart.layout.line.smooth,
          color: l.value.style.chart.layout.line.stroke,
          useArea: l.value.style.chart.layout.line.useArea
        }
      ], me = [
        {
          name: Y,
          series: v.value.rows.map((R) => R.value),
          type: "bar",
          useProgression: !0,
          color: l.value.style.chart.layout.bar.fill
        }
      ], te = l.value.style.chart.modal.backgroundColor, _ = al(te), A = g.dataset.header[v.value.col].decimals, Xt = {
        useCssAnimation: !1,
        chart: {
          width: t,
          height: e - 250,
          backgroundColor: te,
          color: _,
          labels: {
            fontSize: 18,
            prefix: Z,
            suffix: ee
          },
          grid: {
            stroke: Ve(_, 0.5),
            labels: {
              color: _,
              xAxisLabels: {
                color: _,
                show: Ae.value.length,
                values: Ae.value,
                datetimeFormatter: l.value.style.chart.layout.datetimeFormatter,
                showOnlyAtModulo: l.value.style.chart.layout.timeLabels.showOnlyAtModulo,
                modulo: l.value.style.chart.layout.timeLabels.modulo
              }
            }
          },
          highlighter: {
            color: _
          },
          legend: {
            color: _,
            show: !1
            // not necessary since we only show a single series and the title is displayed above
          },
          title: {
            text: Y,
            color: _,
            textAlign: "center",
            fontSize: 18
          },
          tooltip: {
            showTimeLabel: Ae.value.length,
            backgroundOpacity: 30,
            color: _,
            backgroundColor: te,
            showPercentage: !1,
            roundingValue: A
          },
          userOptions: {
            buttons: {
              pdf: !1,
              csv: !1,
              table: !1,
              annotator: !1
            }
          },
          zoom: {
            show: l.value.style.chart.layout.zoom.show,
            focusOnDrag: !0,
            minimap: {
              show: !0
            }
          }
        },
        line: {
          labels: {
            show: !0,
            color: _,
            rounding: A
          },
          dot: {
            useSerieColor: !1,
            fill: l.value.style.chart.layout.line.plot.fill,
            strokeWidth: l.value.style.chart.layout.line.plot.strokeWidth
          }
        },
        bar: {
          useGradient: !1,
          border: {
            useSerieColor: !1,
            stroke: l.value.style.chart.layout.bar.stroke,
            strokeWidth: l.value.style.chart.layout.bar.strokeWidth
          },
          labels: {
            show: !0,
            color: _,
            rounding: A
          }
        }
      }, ft = u >= 0 ? 0 : Math.abs(u), Yt = n + ft, Ne = g.dataset.header[v.value.col].isPercentage, yt = v.value.rows.map((R, Zt) => ({
        x: a * Zt + a / 2,
        y: (1 - (R.value + ft) / Yt) * e,
        value: Ne ? R.value * 100 : R.value,
        suffix: Ne ? "%" : g.dataset.header[v.value.col].suffix ? g.dataset.header[v.value.col].suffix : "",
        prefix: g.dataset.header[v.value.col].prefix ? g.dataset.header[v.value.col].prefix : "",
        index: R.index,
        absoluteValue: Ne ? Math.abs(R.value) * 100 : Math.abs(R.value)
      }));
      return {
        donutConfig: {
          useCssAnimation: !1,
          userOptions: {
            position: "left",
            buttons: {
              pdf: !1,
              csv: !1,
              table: !1,
              annotator: !1
            }
          },
          style: {
            chart: {
              backgroundColor: te,
              color: _,
              layout: {
                curvedMarkers: !0,
                donut: {
                  strokeWidth: 64
                },
                labels: {
                  dataLabels: {
                    suffix: Ne ? "%" : "",
                    prefix: Z
                  },
                  value: {
                    rounding: A
                  },
                  percentage: {
                    color: _,
                    rounding: A
                  },
                  name: {
                    color: _
                  },
                  hollow: {
                    average: {
                      color: _,
                      text: l.value.translations.average,
                      value: {
                        color: _
                      }
                    },
                    total: {
                      color: _,
                      offsetY: -12,
                      text: l.value.translations.total,
                      value: {
                        color: _,
                        offsetY: -12
                      }
                    }
                  }
                }
              },
              legend: {
                backgroundColor: te,
                color: _,
                roundingValue: A,
                roundingPercentage: A
              },
              padding: {
                left: 12,
                right: 12
              },
              title: {
                text: Y,
                color: _,
                fontSize: 18
              },
              tooltip: {
                backgroundOpacity: 30,
                color: _,
                backgroundColor: te,
                roundingValue: A,
                roundingValue: A
              }
            }
          }
        },
        xyConfig: Xt,
        xyDatasetLine: be,
        xyDatasetBar: me,
        progression: yt.length >= 2 ? ol(yt) : !1
      };
    });
    function et() {
      const e = F.value.options.map((t, o) => ({
        name: t,
        color: We[o] || We[o % We.length],
        values: [ne.value.filter((a, n) => a.td[F.value.index] === t && v.value.rows.map((u) => u.index).includes(n)).map((a) => a.td[v.value.col]).reduce((a, n) => Math.abs(a) + Math.abs(n), 0)],
        absoluteValue: ne.value.filter((a, n) => a.td[F.value.index] === t && v.value.rows.map((u) => u.index).includes(n)).map((a) => a.td[v.value.col]).reduce((a, n) => a + n, 0)
      })).sort((t, o) => o.value - t.value);
      we.value = e, ke(() => {
        S.value.type = c.value.DONUT, I.value = !1;
      });
    }
    function zt(e, t) {
      return !E.value[e] && (t.isSort || t.isSearch || t.isMultiselect || t.rangeFilter) && ![c.value.DATE].includes(t.type);
    }
    function tt(e = "all") {
      const t = g.dataset.header.map((u) => u.name), o = e === "all" ? N.value.map((u) => u.td) : ne.value.map((u) => u.td), a = [t].concat(o), n = nl(a);
      sl({ csvContent: n, title: re.value });
    }
    function Ut() {
      const e = document.getElementsByClassName("th-dropdown");
      e.length && Array.from(e).forEach((t) => {
        t.dataset.isOpen = !1;
      });
    }
    function He(e, t) {
      let o;
      clearTimeout(o), o = setTimeout(e, t);
    }
    function q() {
      const e = document.getElementsByClassName(`tr_${G}`);
      Array.from(e).forEach((t) => {
        Array.from(t.getElementsByTagName("td")).forEach((o) => {
          o.dataset.row === "even" ? (o.style.background = l.value.style.rows.even.backgroundColor, o.style.color = l.value.style.rows.even.color) : (o.style.background = l.value.style.rows.odd.backgroundColor, o.style.color = l.value.style.rows.odd.color);
        });
      }), Array.from(e).forEach((t) => t.dataset.selected = "false"), k.value > $.value.length - 1 && (k.value = $.value.length - 1), X.value = !1, we.value = void 0, ue.value = void 0, S.value.type = c.value.BAR, v.value = {
        col: void 0,
        rows: []
      }, _t.value = 100, St.value = 100;
    }
    function Ft(e, t) {
      N.value = e.filter((o) => o.td[t] >= H.value[t].min && o[t] <= H.value[t].max);
    }
    function lt(e, t) {
      if (w.value[t] === c.value.ASC && (e = e.sort((o, a) => o[t] - a[t])), w.value[t] === c.value.DESC)
        e = e.sort((o, a) => a[t] - o[t]);
      else
        return 0;
    }
    function at() {
      q(), Object.keys(H.value).forEach((e) => {
        Ft(N.value, e);
      }), Object.keys(w.value).forEach((e) => {
        lt(N.value, e);
      }), Ce.value !== void 0 && lt(N.value, Ce.value), g.dataset.header.forEach((e, t) => {
        if (e.isPercentage) {
          const o = J.value[t].referenceIndex, a = N.value.map((n) => n.td[o]).reduce((n, u) => n + u, 0);
          N.value.forEach((n) => {
            n.td[t] = n.td[o] / a;
          });
        }
      }), k.value > $.value.length - 1 && (k.value = $.value.length - 1), [-1].includes(k.value) && (k.value = 0);
    }
    function Q() {
      N.value = Le.value.filter((e) => {
        for (const t in P.value)
          if (!e.td[t].toUpperCase().includes(P.value[t].toUpperCase()))
            return !1;
        for (const t in C.value)
          if (!C.value[t].some((a) => a === e.td[t]))
            return !1;
        for (const t in U.value) {
          const o = new Date(e.td[t]), a = new Date(U.value[t].from), n = new Date(U.value[t].to);
          if (o < a || o > n)
            return !1;
        }
        return !0;
      }), at();
    }
    function ot(e) {
      return Qe.value.map((t) => t[e]).map((t) => isNaN(Number(t)) ? 0 : t).reduce((t, o) => t + o, 0) / N.value.length;
    }
    function fe(e) {
      const t = g.dataset.body.map((me) => new Date(me.td[e])), o = new Date(Math.min(...t)), a = new Date(Math.max(...t)), n = o.getFullYear(), u = a.getFullYear(), p = String(o.getMonth() + 1).padStart(2, "0"), O = String(a.getMonth() + 1).padStart(2, "0"), Y = String(o.getDate()).padStart(2, "0"), Z = String(a.getDate()).padStart(2, "0"), ee = `${n}-${p}-${Y}`, be = `${u}-${O}-${Z}`;
      return {
        from: ee,
        to: be
      };
    }
    function B(e) {
      return [...new Set(g.dataset.body.map((t) => t.td[e]))];
    }
    function ce(e) {
      return Qe.value.map((t) => t[e]).map((t) => isNaN(Number(t)) ? 0 : t).reduce((t, o) => t + o, 0);
    }
    function jt(e) {
      return e.includes(NaN);
    }
    function nt(e, t) {
      return C.value[t] ? C.value[t].includes(e) : !0;
    }
    function Vt(e) {
      return !isNaN(Number(String(e).replaceAll("%", "")));
    }
    function st(e, t) {
      const o = t.isSort, a = t.isSearch, n = t.isMultiselect && C.value[e], u = t.rangeFilter, p = (O) => {
        if (u && H.value[O])
          return Math.round(H.value[O].min) === W.value[O].min && Math.round(H.value[O].max) === W.value[O].max;
      };
      if (o && a && n && u)
        return ["", void 0].includes(P.value[e]) && [void 0].includes(w.value[e]) && C.value[e].length === B(e).length && p(e);
      if (o && a && n)
        return ["", void 0].includes(P.value[e]) && [void 0].includes(w.value[e]) && C.value[e].length === B(e).length;
      if (o && a && u)
        return ["", void 0].includes(P.value[e]) && [void 0].includes(w.value[e]) && p(e);
      if (o && a)
        return ["", void 0].includes(P.value[e]) && [void 0].includes(w.value[e]);
      if (o && n && u)
        return [void 0].includes(w.value[e]) && C.value[e].length === B(e).length && p(e);
      if (o && n)
        return [void 0].includes(w.value[e]) && C.value[e].length === B(e).length;
      if (a && n && u)
        return ["", void 0].includes(P.value[e]) && C.value[e].length === B(e).length && p(e);
      if (a && n)
        return ["", void 0].includes(P.value[e]) && C.value[e].length === B(e).length;
      if (a && u)
        return ["", void 0].includes(P.value[e]) && p(e);
      if (a)
        return ["", void 0].includes(P.value[e]);
      if (o && u)
        return [void 0].includes(w.value[e]) && p(e);
      if (o)
        return [void 0].includes(w.value[e]);
      if (n && u)
        return C.value[e].length === B(e).length && p(e);
      if (n)
        return C.value[e].length === B(e).length;
    }
    function ut() {
      return {
        totalPages: $.value.length,
        itemsPerPage: oe.value,
        currentPage: k.value,
        currentPageData: ne.value.map((e) => e.td)
      };
    }
    $t({
      getCurrentPageData: ut
    });
    function Be() {
      Tt("page-change", ut());
    }
    const Re = f(null);
    function _e(e) {
      if (q(), e === "next" && k.value < $.value.length) {
        if (k.value + 1 > $.value.length - 1)
          return;
        k.value += 1;
      } else if (e === "previous" && k.value >= 1)
        k.value -= 1;
      else {
        if (e - 1 < 0 || e > $.value.length || e === "previous")
          return;
        k.value = e - 1;
      }
      Be(), Re.value && Re.value.scrollTo({
        top: 0,
        left: 0,
        behavior: "smooth"
      });
    }
    function Wt(e) {
      e.preventDefault();
      const t = e.keyCode, o = 38, a = 40, n = 37, u = 39;
      if (![o, a, n, u].includes(t)) return;
      const p = e.target.id, O = /cell_(\d+)_(\d+)_([0-9a-fA-F-]{36})/, Y = p.match(O), Z = parseInt(Y[1]), ee = parseInt(Y[2]), be = document.getElementById(`cell_${Z}_${ee + 1}_${G}`), me = document.getElementById(`cell_${Z}_${ee - 1}_${G}`), te = document.getElementById(`cell_${Z + 1}_${ee}_${G}`), _ = document.getElementById(`cell_${Z - 1}_${ee}_${G}`);
      let A;
      switch (!0) {
        case t === u:
          A = be;
          break;
        case t === n:
          A = me;
          break;
        case t === o:
          A = _;
          break;
        case t === a:
          A = te;
          break;
        default:
          return;
      }
      A && (A.focus(), A.scrollIntoView({ behavior: "smooth", block: "center" }));
    }
    function Jt(e) {
      let t = 0;
      for (let o = 0; o < e.length; o += 1)
        t += e.charCodeAt(o);
      return t;
    }
    async function rt() {
      return new Promise((e) => {
        const t = [];
        K.value.forEach((o, a) => {
          if (o.isSearch && Object.assign(P.value, { [a]: "" }), o.isMultiselect && Object.assign(C.value, { [a]: B(a) }), o.type === c.value.DATE && (Object.assign(U.value, { [a]: fe(a) }), Object.assign(pe.value, { [a]: !1 })), (o.isPercentage || o.percentageTo) && Object.assign(J.value, {
            [a]: {
              reference: o.percentageTo,
              referenceIndex: g.dataset.header.map((n) => n.name).indexOf(o.percentageTo)
            }
          }), o.rangeFilter && (Object.assign(H.value, {
            [a]: {
              min: Math.round(Math.min(...g.dataset.body.map((n) => n.td).map((n) => n[a]))),
              max: Math.round(Math.max(...g.dataset.body.map((n) => n.td).map((n) => n[a])))
            }
          }), Object.assign(W.value, {
            [a]: {
              min: Math.round(Math.min(...g.dataset.body.map((n) => n.td).map((n) => n[a]))),
              max: Math.round(Math.max(...g.dataset.body.map((n) => n.td).map((n) => n[a])))
            }
          })), o.isPercentage) {
            const n = g.dataset.header.map((p) => p.name).indexOf(o.percentageTo), u = g.dataset.body.map((p) => p.td[n]).reduce((p, O) => p + O, 0);
            t.push([a, n, u]);
          }
          o.type === c.value.NUMERIC && !o.isPercentage && Object.assign(E.value, { [a]: jt(g.dataset.body.map((n) => Number(n.td[a]))) });
        }), N.value.forEach((o, a) => {
          t.map((n) => {
            const [u, p, O] = n;
            o.td[u] = o.td[p] / O;
          }), o.td.forEach((n, u) => {
            g.dataset.header[u].type === c.value.TEXT && g.dataset.header[u].isSearch && (o[u] = Jt(n)), g.dataset.header[u].type === c.value.DATE && (o[u] = new Date(n).getTime()), g.dataset.header[u].type === c.value.NUMERIC && (o[u] = isNaN(Number(n)) ? u : n), Le.value[a][u] = o[u];
          });
        }), e(!0);
      });
    }
    function it(e, t) {
      return new Promise((o, a) => {
        e().then((n) => {
          try {
            const u = t(n);
            o(u);
          } catch (u) {
            a(u);
          }
        }).catch((n) => {
          a(n);
        });
      });
    }
    async function Kt(e) {
      U.value[e] = {
        from: fe(e).from,
        to: fe(e).to
      }, pe.value[e] = !1, await ke(), Q();
    }
    function ct(e, t, o) {
      const a = o.currentTarget;
      clearTimeout(se.value), a.classList.add("clicked"), se.value = setTimeout(() => {
        a.classList.remove("clicked");
      }, 200), Ce.value = void 0, t.rangeFilter && (H.value[e].min = W.value[e].min, H.value[e].max = W.value[e].max), t.isMultiselect ? (C.value[e] = B(e), t.type === c.value.TEXT && (w.value[e] = void 0), t.isSearch && (P.value[e] = "")) : t.type === c.value.NUMERIC ? w.value[e] = void 0 : t.type === c.value.TEXT ? (w.value[e] = void 0, P.value[e] = "") : t.type === c.value.DATE && (w.value[e] = void 0), Q();
    }
    function Se({ td: e, rowIndex: t, colIndex: o, headerType: a, event: n }) {
      if (a !== c.value.NUMERIC || isNaN(Number(e))) {
        q();
        return;
      }
      v.value.col !== o && q();
      const u = n.currentTarget.parentNode;
      v.value.col = o, v.value.rows.map((p) => p.index).includes(t) ? (u.dataset.selected = "false", v.value.rows = v.value.rows.filter((p) => p.index !== t), n.currentTarget.classList.remove(Nt.value.CELL), Array.from(u.children).forEach((p, O) => {
        p.dataset.row === "even" ? (p.style.background = l.value.style.rows.even.backgroundColor, p.style.color = l.value.style.rows.even.olor) : (p.style.background = l.value.style.rows.odd.backgroundColor, p.style.color = l.value.style.rows.odd.color);
      }), n.currentTarget.dataset.row === "even" ? (n.currentTarget.style.background = l.value.style.rows.even.backgroundColor, n.currentTarget.style.color = l.value.style.rows.even.color) : (n.currentTarget.style.background = l.value.style.rows.odd.backgroundColor, n.currentTarget.style.color = l.value.style.rows.odd.color)) : (u.dataset.selected = "true", v.value.rows.push({
        index: t,
        value: e
      }), Array.from(u.children).forEach((p, O) => {
        p.dataset.row === "even" ? (p.style.background = l.value.style.rows.even.selectedNeighbors.backgroundColor, p.style.color = l.value.style.rows.even.selectedNeighbors.color) : (p.style.background = l.value.style.rows.odd.selectedNeighbors.backgroundColor, p.style.color = l.value.style.rows.odd.selectedNeighbors.color);
      }), n.currentTarget.dataset.row === "odd" ? (n.currentTarget.style.background = l.value.style.rows.odd.selectedCell.backgroundColor, n.currentTarget.style.color = l.value.style.rows.odd.selectedCell.color) : (n.currentTarget.style.background = l.value.style.rows.even.selectedCell.backgroundColor, n.currentTarget.style.color = l.value.style.rows.even.selectedCell.color)), v.value.rows = v.value.rows.sort((p, O) => p.index - O.index), S.value.type === c.value.DONUT && v.value.rows.length > 0 && et();
    }
    function vt(e) {
      v.value.col !== e ? (ne.value.forEach((t, o) => {
        Se({
          td: t.td[e],
          rowIndex: o,
          colIndex: e,
          headerType: c.value.NUMERIC,
          event: {
            currentTarget: document.getElementById(`cell_${o}_${e}_${G}`)
          }
        });
      }), ue.value = e) : (ue.value = void 0, q());
    }
    async function ze(e, t) {
      C.value[t].includes(e) ? C.value[t] = C.value[t].filter((o) => o !== e) : C.value[t].push(e), await ke(), Q();
    }
    function dt(e) {
      pe.value[e] = !(fe(e).from === U.value[e].from && fe(e).to === U.value[e].to);
    }
    function ht(e, t) {
      Ce.value = e;
      const o = t.currentTarget;
      clearTimeout(se.value), o.classList.add("clicked"), se.value = setTimeout(() => {
        o.classList.remove("clicked");
      }, 200), w.value[e] === 1 ? w.value[e] = c.value.DESC : w.value[e] = c.value.ASC, at();
    }
    function gt(e, t, o) {
      const a = o.currentTarget;
      clearTimeout(se.value), a.classList.add("clicked"), se.value = setTimeout(() => {
        a.classList.remove("clicked");
      }, 200);
      const n = document.getElementById(`th_dropdown_${e}`);
      n.dataset.isOpen === "false" ? n.dataset.isOpen = "true" : n.dataset.isOpen = "false";
    }
    function pt(e) {
      q(), !e || !e.target.value ? k.value = 0 : k.value = Number(e.target.value), Be();
    }
    const ye = f(null);
    Qt(() => {
      if (g.dataset.header.length === 0)
        throw new Error(`vue-ui-table error: missing header data.
Provide an array of objects of type:
{
 name: string;
 type: string; ('text' | 'numeric' | 'date')
 average: boolean;
 decimals: number | undefined;
 sum: boolean;
 isSort:boolean;
 isSearch: boolean;
 isMultiselect: boolean;
 isPercentage: boolean;
 percentageTo: string; (or '')
}`);
      if (g.dataset.body.length === 0)
        throw new Error("vue-ui-table error: missing body data");
      $e.value = !0, it(rt, async () => {
        await ke(), $e.value = !1;
      }), document.addEventListener("keydown", (e) => {
        const t = document.activeElement;
        (t && Array.from(t.classList).includes("td-focusable") && e.key.includes("Arrow") || e.code === "Space") && e.preventDefault();
      }), re.value = l.value.style.exportMenu.filename, Me.value = Ie.value[0]?.name ?? "";
    }), Fe(() => g.dataset, (e) => {
      $e.value = !0, N.value = JSON.parse(JSON.stringify(e.body)).map((t, o) => ({
        ...t,
        absoluteIndex: o
      })), Le.value = JSON.parse(JSON.stringify(e.body)).map((t, o) => ({
        ...t,
        absoluteIndex: o
      })), K.value = JSON.parse(JSON.stringify(e.header)).map((t, o) => ({
        average: Object.hasOwn(t, "average") ? t.average : !1,
        decimals: Object.hasOwn(t, "decimals") ? t.decimals : 0,
        isMultiselect: Object.hasOwn(t, "isMultiselect") ? t.isMultiselect : !1,
        isPercentage: Object.hasOwn(t, "isPercentage") ? t.isPercentage : !1,
        isSearch: Object.hasOwn(t, "isSearch") ? t.isSearch : !1,
        isSort: Object.hasOwn(t, "isSort") ? t.isSort : !1,
        name: t.name,
        percentageTo: Object.hasOwn(t, "percentageTo") ? t.percentageTo : void 0,
        prefix: Object.hasOwn(t, "prefix") ? t.prefix : "",
        rangeFilter: Object.hasOwn(t, "rangeFilter") ? t.rangeFilter : !1,
        suffix: Object.hasOwn(t, "suffix") ? t.suffix : "",
        sum: Object.hasOwn(t, "sum") ? t.sum : !1,
        type: t.type,
        index: o
      })), v.value = { col: void 0, rows: [] }, ue.value = void 0, oe.value = g.config.rowsPerPage ? g.config.rowsPerPage : 25, J.value = {}, U.value = {}, pe.value = {}, E.value = {}, W.value = {}, H.value = {}, C.value = {}, P.value = {}, w.value = {}, X.value = !1, we.value = void 0, F.value = void 0, Lt.value = void 0, I.value = !1, v.value.col = void 0, v.value.rows = [], it(rt, async () => {
        pt(), await ke(), $e.value = !1;
      });
    }, { immediate: !0, deep: !0 });
    const Ue = f(null);
    return Fe(ge, (e) => {
      e && Ue.value && Ue.value.focus();
    }), Fe(X, (e) => {
      e ? ye.value && ye.value.open() : ye.value && ye.value.close();
    }), (e, t) => (r(), i("div", {
      class: "vue-data-ui-component vue-ui-table-main",
      style: h(`font-family: ${l.value.fontFamily}`)
    }, [
      l.value.style.exportMenu.show ? (r(), i("div", {
        key: 0,
        class: "vue-ui-table-export-hub",
        style: h({ top: Xe.value + "px" })
      }, [
        s("button", {
          onClick: t[0] || (t[0] = (o) => ge.value = !ge.value),
          innerHTML: M.value.export,
          style: h(`background:${l.value.style.exportMenu.backgroundColor};color:${l.value.style.exportMenu.color};cursor:${T.value ? "pointer" : "default"}`)
        }, null, 12, vl),
        s("div", {
          class: "vue-ui-table-export-hub-dropdown",
          "data-is-open": ge.value || "false",
          style: h(`background:${l.value.style.exportMenu.backgroundColor};color:${l.value.style.exportMenu.color}`)
        }, [
          t[24] || (t[24] = s("b", { class: "vue-ui-table-export-hub-title" }, " Export ", -1)),
          s("button", {
            class: "close-dropdown",
            onClick: t[1] || (t[1] = (o) => ge.value = !1),
            style: h(`background:${l.value.style.closeButtons.backgroundColor};color:${l.value.style.closeButtons.color};border-radius:${l.value.style.closeButtons.borderRadius}`)
          }, " ✖ ", 4),
          s("div", hl, [
            s("div", gl, [
              s("div", pl, d(l.value.translations.exportAllLabel) + " (" + d(N.value.length) + ") ", 1),
              s("button", {
                id: "exportAll",
                onClick: t[2] || (t[2] = (o) => tt("all")),
                style: h(`background:${l.value.style.exportMenu.buttons.backgroundColor};color:${l.value.style.exportMenu.buttons.color};cursor:${T.value ? "pointer" : "default"}`)
              }, [
                s("div", {
                  innerHTML: M.value.fileDownload
                }, null, 8, fl),
                s("span", null, d(l.value.translations.exportAllButton), 1)
              ], 4)
            ]),
            s("div", yl, [
              s("div", bl, d(l.value.translations.exportPageLabel), 1),
              s("button", {
                id: "exportPage",
                onClick: t[3] || (t[3] = (o) => tt("page")),
                style: h(`background:${l.value.style.exportMenu.buttons.backgroundColor};color:${l.value.style.exportMenu.buttons.color};cursor:${T.value ? "pointer" : "default"}`)
              }, [
                s("div", {
                  innerHTML: M.value.fileDownload
                }, null, 8, ml),
                s("span", null, d(l.value.translations.exportPageButton), 1)
              ], 4)
            ]),
            s("div", kl, [
              s("label", wl, [
                s("span", Cl, d(l.value.translations.filename), 1),
                le(s("input", {
                  name: "filename",
                  ref_key: "filenameInputRef",
                  ref: Ue,
                  onKeydown: t[4] || (t[4] = ve(de(() => {
                  }, ["stop"]), ["space"])),
                  pattern: ".*",
                  class: "vue-ui-table-dialog-input",
                  type: "text",
                  "onUpdate:modelValue": t[5] || (t[5] = (o) => re.value = o)
                }, null, 544), [
                  [he, re.value]
                ])
              ]),
              re.value ? (r(), i("button", {
                key: 0,
                class: "vue-ui-table-dialog-field-button",
                onClick: t[6] || (t[6] = (o) => re.value = ""),
                style: h({ cursor: T.value ? "pointer" : "default" })
              }, [
                je(il, {
                  name: "close",
                  stroke: l.value.style.exportMenu.color,
                  size: 18
                }, null, 8, ["stroke"])
              ], 4)) : b("", !0)
            ])
          ])
        ], 12, dl)
      ], 4)) : b("", !0),
      s("div", {
        class: "vue-ui-table__wrapper",
        style: h(`max-height:${l.value.maxHeight}px`),
        ref_key: "tableWrapper",
        ref: Re
      }, [
        s("table", $l, [
          l.value.style.title.text ? (r(), i("caption", {
            key: 0,
            class: "vue-ui-table__caption",
            ref_key: "tableCaption",
            ref: Oe,
            style: h({
              textAlign: l.value.style.title.textAlign,
              paddingLeft: l.value.style.title.paddingLeft + "px",
              paddingRight: l.value.style.title.paddingRight + "px",
              backgroundColor: l.value.style.title.backgroundColor,
              boxShadow: `${l.value.style.title.backgroundColor} -1px 0px 0px 0px`
            })
          }, [
            s("span", {
              style: h({
                fontSize: l.value.style.title.fontSize + "px",
                fontWeight: l.value.style.title.bold ? "bold" : "normal",
                color: l.value.style.title.color
              })
            }, d(l.value.style.title.text), 5),
            l.value.style.title.subtitle.text ? (r(), i(L, { key: 0 }, [
              t[25] || (t[25] = s("br", null, null, -1)),
              s("span", {
                style: h({
                  fontSize: l.value.style.title.subtitle.fontSize,
                  fontWeight: l.value.style.title.subtitle.bold ? "bold" : "normal",
                  color: l.value.style.title.subtitle.color
                })
              }, d(l.value.style.title.subtitle.text), 5)
            ], 64)) : b("", !0)
          ], 4)) : b("", !0),
          s("thead", {
            id: "tableHead",
            class: "vue-ui-table__head",
            style: h({
              background: l.value.style.th.backgroundColor,
              boxShadow: `-1px 0 0 ${l.value.style.th.backgroundColor}`,
              top: Xe.value - 3 + "px"
            })
          }, [
            s("tr", null, [
              t[26] || (t[26] = s("th", { class: "invisible-cell" }, null, -1)),
              (r(!0), i(L, null, j(K.value, (o, a) => (r(), i("th", {
                key: `thead_${a}`,
                style: h(`overflow: visible;background:${l.value.style.th.backgroundColor};color:${l.value.style.th.color};outline:${l.value.style.th.outline}`),
                class: D({ "th-has-nan": E.value[a] })
              }, [
                qe.value && ([c.value.TEXT, c.value.DATE].includes(o.type) || o.isPercentage) ? (r(), i("span", Ml, [
                  z(d(o.name) + " ", 1),
                  o.isPercentage ? (r(), i("span", xl, " / " + d(o.percentageTo), 1)) : b("", !0)
                ])) : (r(), i("span", Tl, d(o.name), 1))
              ], 6))), 128))
            ]),
            qe.value ? (r(), i(L, { key: 0 }, [
              s("tr", null, [
                t[27] || (t[27] = s("th", { class: "invisible-cell" }, null, -1)),
                (r(!0), i(L, null, j(K.value, (o, a) => (r(), i("th", {
                  key: `thead_${a}`,
                  class: D({ "th-numeric": !0, "th-has-nan": E.value[a] }),
                  style: h(`background:${l.value.style.th.backgroundColor};color:${l.value.style.th.color};outline:${l.value.style.th.outline}`)
                }, [
                  o.sum && !E.value[a] ? (r(), i("span", _l, [
                    s("span", {
                      innerHTML: M.value.sum,
                      style: { "margin-bottom": "-4px", "margin-right": "3px" }
                    }, null, 8, Sl),
                    z(" " + d(V(ae)({
                      p: o.prefix,
                      v: Number(ce(a)),
                      s: o.suffix,
                      r: o.decimals
                    })) + " ", 1),
                    J.value[a] && o.percentageTo && !o.isPercentage ? (r(), i("span", Nl, " (" + d(isNaN(ce(a) / ce(J.value[a].referenceIndex)) ? "-" : V(ae)({
                      v: ce(a) / ce(J.value[a].referenceIndex) * 100,
                      s: "%",
                      r: o.decimals
                    })) + ") ", 1)) : b("", !0)
                  ])) : b("", !0)
                ], 6))), 128))
              ]),
              s("tr", null, [
                t[28] || (t[28] = s("th", { class: "invisible-cell" }, null, -1)),
                (r(!0), i(L, null, j(K.value, (o, a) => (r(), i("th", {
                  key: `thead_${a}`,
                  class: D({ "th-numeric": !0, "th-has-nan": E.value[a] }),
                  style: h(`background:${l.value.style.th.backgroundColor};color:${l.value.style.th.color};outline:${l.value.style.th.outline}`)
                }, [
                  o.average && !E.value[a] ? (r(), i("span", Ol, " ~ " + d(isNaN(ot(a)) ? "" : V(ae)({
                    p: o.prefix,
                    v: Number(ot(a)),
                    s: o.suffix,
                    r: o.decimals
                  })), 1)) : b("", !0)
                ], 6))), 128))
              ]),
              s("tr", null, [
                t[31] || (t[31] = s("th", { class: "invisible-cell" }, null, -1)),
                (r(!0), i(L, null, j(K.value, (o, a) => (r(), i("th", {
                  key: `thead_${a}`,
                  class: D({ "th-has-nan": E.value[a] }),
                  style: h(`background:${l.value.style.th.backgroundColor};color:${l.value.style.th.color};outline:${l.value.style.th.outline}`)
                }, [
                  s("div", Ll, [
                    o.type === c.value.DATE && U.value[a] ? (r(), i("div", Dl, [
                      s("div", El, [
                        s("div", Il, [
                          s("label", {
                            for: `from_${a}`
                          }, d(l.value.translations.from), 9, Al),
                          le(s("input", {
                            id: `from_${a}`,
                            type: "date",
                            "onUpdate:modelValue": (n) => U.value[a].from = n,
                            onInput: (n) => {
                              Q(), dt(a);
                            },
                            style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border}`)
                          }, null, 44, Pl), [
                            [he, U.value[a].from]
                          ])
                        ]),
                        s("div", Hl, [
                          s("label", {
                            for: `to_${a}`
                          }, d(l.value.translations.to), 9, Bl),
                          le(s("input", {
                            id: `to_${a}`,
                            type: "date",
                            "onUpdate:modelValue": (n) => U.value[a].to = n,
                            onInput: (n) => {
                              Q(), dt(a);
                            },
                            style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border}`)
                          }, null, 44, Rl), [
                            [he, U.value[a].to]
                          ])
                        ])
                      ]),
                      s("div", zl, [
                        o.isSort ? (r(), i("button", {
                          key: 0,
                          onClick: (n) => ht(a, n),
                          class: D({ "th-button-active": [c.value.DESC, c.value.ASC].includes(w.value[a]) }),
                          style: h(`cursor:${T.value ? "pointer" : "default"}; background:${[c.value.DESC, c.value.ASC].includes(w.value[a]) ? "" : l.value.style.th.buttons.filter.inactive.backgroundColor};color:${[c.value.DESC, c.value.ASC].includes(w.value[a]) ? "" : l.value.style.th.buttons.filter.inactive.color}`)
                        }, [
                          w.value[a] === c.value.ASC ? (r(), i("span", {
                            key: 0,
                            innerHTML: [c.value.DATE].includes(o.type) ? M.value.sort09 : M.value.sortAZ
                          }, null, 8, Fl)) : w.value[a] === c.value.DESC ? (r(), i("span", {
                            key: 1,
                            innerHTML: [c.value.DATE].includes(o.type) ? M.value.sort90 : M.value.sortZA
                          }, null, 8, jl)) : (r(), i("span", {
                            key: 2,
                            innerHTML: M.value.arrowSort
                          }, null, 8, Vl))
                        ], 14, Ul)) : b("", !0),
                        s("button", {
                          onClick: (n) => {
                            Kt(a), ct(a, o, n);
                          },
                          disabled: !pe.value[a] && st(a, o),
                          class: "th-reset",
                          style: h({ cursor: T.value ? "pointer" : "default" })
                        }, " ✖ ", 12, Wl)
                      ])
                    ])) : b("", !0),
                    o.isSearch ? le((r(), i("input", {
                      key: 1,
                      placeholder: l.value.translations.inputPlaceholder,
                      "onUpdate:modelValue": (n) => P.value[a] = n,
                      onInput: t[7] || (t[7] = (n) => He(Q, 400)),
                      name: `search_${a}`,
                      style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border}`)
                    }, null, 44, Jl)), [
                      [he, P.value[a]]
                    ]) : b("", !0),
                    !E.value[a] && o.isSort && o.type !== c.value.DATE ? (r(), i("button", {
                      key: 2,
                      onClick: (n) => ht(a, n),
                      class: D({ "th-button-active": [c.value.DESC, c.value.ASC].includes(w.value[a]) }),
                      style: h(`cursor:${T.value ? "pointer" : "default"};background:${[c.value.DESC, c.value.ASC].includes(w.value[a]) ? "" : l.value.style.th.buttons.filter.inactive.backgroundColor};color:${[c.value.DESC, c.value.ASC].includes(w.value[a]) ? "" : l.value.style.th.buttons.filter.inactive.color}`)
                    }, [
                      w.value[a] === c.value.ASC ? (r(), i("span", {
                        key: 0,
                        innerHTML: [c.value.NUMERIC].includes(o.type) ? M.value.sort09 : M.value.sortZA
                      }, null, 8, Xl)) : w.value[a] === c.value.DESC ? (r(), i("span", {
                        key: 1,
                        innerHTML: [c.value.NUMERIC].includes(o.type) ? M.value.sort90 : M.value.sortAZ
                      }, null, 8, Yl)) : (r(), i("span", {
                        key: 2,
                        innerHTML: M.value.arrowSort
                      }, null, 8, Zl))
                    ], 14, Kl)) : b("", !0),
                    o.isMultiselect ? (r(), i("button", {
                      key: 3,
                      onClick: (n) => gt(a, o, n),
                      innerHTML: M.value.filter,
                      class: D({ "th-button-active": C.value[a] && C.value[a].length !== B(a).length }),
                      style: h(`cursor:${T.value ? "pointer" : "default"};background:${C.value[a] && C.value[a].length !== B(a).length ? "" : l.value.style.th.buttons.filter.inactive.backgroundColor};color:${C.value[a] && C.value[a].length !== B(a).length ? "" : l.value.style.th.buttons.filter.inactive.color}`)
                    }, null, 14, Gl)) : b("", !0),
                    v.value.col === a && Ge.value ? (r(), i("button", {
                      key: 4,
                      onClick: t[8] || (t[8] = (n) => X.value = !X.value),
                      innerHTML: M.value.chart,
                      class: D({ "th-button-active": X.value }),
                      style: h(`cursor:${T.value ? "pointer" : "default"};background:${X.value ? "" : l.value.style.th.buttons.filter.inactive.backgroundColor};color:${X.value ? "" : l.value.style.th.buttons.filter.inactive.color}`)
                    }, null, 14, ql)) : b("", !0),
                    o.rangeFilter && H.value[a] && !E.value[a] ? (r(), i("div", Ql, [
                      s("label", {
                        for: `rangeMin${a}`
                      }, [...t[29] || (t[29] = [
                        s("span", { style: { color: "grey" } }, "ᒥ", -1),
                        z(" min ", -1),
                        s("span", { style: { color: "grey" } }, "ᒣ", -1)
                      ])], 8, ea),
                      le(s("input", {
                        type: "number",
                        id: `rangeMin${a}`,
                        max: W.value[a].max,
                        min: W.value[a].min,
                        "onUpdate:modelValue": (n) => H.value[a].min = n,
                        onInput: t[9] || (t[9] = (n) => He(Q, 400)),
                        style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border}`)
                      }, null, 44, ta), [
                        [
                          he,
                          H.value[a].min,
                          void 0,
                          { number: !0 }
                        ]
                      ]),
                      le(s("input", {
                        type: "number",
                        id: `rangeMax${a}`,
                        max: W.value[a].max,
                        min: W.value[a].min,
                        "onUpdate:modelValue": (n) => H.value[a].max = n,
                        onInput: t[10] || (t[10] = (n) => He(Q, 400)),
                        style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border}`)
                      }, null, 44, la), [
                        [
                          he,
                          H.value[a].max,
                          void 0,
                          { number: !0 }
                        ]
                      ]),
                      s("label", {
                        for: `rangeMax${a}`
                      }, [...t[30] || (t[30] = [
                        s("span", { style: { color: "grey" } }, "ᒪ", -1),
                        z(" max ", -1),
                        s("span", { style: { color: "grey" } }, "ᒧ", -1)
                      ])], 8, aa)
                    ])) : b("", !0),
                    zt(a, o) ? (r(), i("button", {
                      key: 6,
                      onClick: (n) => ct(a, o, n),
                      disabled: st(a, o),
                      class: "th-reset",
                      style: h({ cursor: T.value ? "pointer" : "default" })
                    }, " ✖ ", 12, oa)) : b("", !0),
                    o.isMultiselect ? (r(), i("div", {
                      key: 7,
                      class: "th-dropdown",
                      "data-is-open": "false",
                      id: `th_dropdown_${a}`,
                      style: h(`background:${l.value.style.dropdowns.backgroundColor};color:${l.value.style.dropdowns.color}`)
                    }, [
                      s("button", {
                        class: "close-dropdown",
                        onClick: (n) => gt(a, o, n),
                        style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.closeButtons.backgroundColor};color:${l.value.style.closeButtons.color}`)
                      }, " ✖ ", 12, sa),
                      (r(!0), i(L, null, j(B(a), (n, u) => (r(), i("span", {
                        class: "th-option",
                        key: `th_option_${a}_${u}`,
                        onClick: (p) => ze(n, a),
                        onKeyup: [
                          ve((p) => ze(n, a), ["enter"]),
                          ve((p) => ze(n, a), ["space"])
                        ],
                        style: h(`opacity:${nt(n, a) ? 1 : 0.5}`),
                        tabindex: "0"
                      }, [
                        nt(n, a) ? (r(), i("span", {
                          key: 0,
                          style: h(`color:${l.value.style.dropdowns.icons.selected.color};margin-right:5px`),
                          class: "th-icon-green"
                        }, d(l.value.style.dropdowns.icons.selected.unicode), 5)) : (r(), i("span", {
                          key: 1,
                          style: h(`color:${l.value.style.dropdowns.icons.unselected.color};margin-right:5px`),
                          class: "th-icon-red"
                        }, d(l.value.style.dropdowns.icons.unselected.unicode), 5)),
                        s("span", null, d(n), 1)
                      ], 44, ua))), 128))
                    ], 12, na)) : b("", !0)
                  ])
                ], 6))), 128))
              ])
            ], 64)) : b("", !0),
            s("tr", null, [
              t[32] || (t[32] = s("th", { class: "invisible-cell" }, null, -1)),
              (r(!0), i(L, null, j(K.value, (o, a) => (r(), i("th", {
                key: `col_selector_${a}`,
                class: D({ "vue-ui-table-col-selector": !E.value[a], "th-has-nan": E.value[a] }),
                style: h(`background:${a === ue.value && !E.value[a] ? l.value.style.th.selected.backgroundColor : l.value.style.th.backgroundColor};color:${a === ue.value && !E.value[a] ? l.value.style.th.selected.color : l.value.style.th.color};outline:${l.value.style.th.outline}`)
              }, [
                E.value[a] ? b("", !0) : (r(), i("div", {
                  key: 0,
                  innerHTML: K.value[a].type === c.value.NUMERIC ? M.value.chevronDown : "",
                  class: D({ "col-selector": K.value[a].type === c.value.NUMERIC }),
                  tabindex: "0",
                  style: h({ cursor: T.value ? "pointer" : "default" }),
                  onClick: de((n) => vt(a), ["stop"]),
                  onKeyup: ve((n) => vt(a), ["enter"])
                }, null, 46, ra))
              ], 6))), 128))
            ])
          ], 4),
          s("tbody", {
            onClick: Ut,
            onKeydown: t[11] || (t[11] = (o) => Wt(o))
          }, [
            (r(!0), i(L, null, j(ne.value, (o, a) => (r(), i("tr", {
              key: `tbody_${a}`,
              "data-row": a % 2 === 0 ? "odd" : "even",
              class: D(`tr_${V(G)}`),
              style: h(`${a % 2 === 0 ? `background:${l.value.style.rows.odd.backgroundColor};color:${l.value.style.rows.odd.color}` : `background:${l.value.style.rows.even.backgroundColor};color:${l.value.style.rows.even.color}`}`)
            }, [
              s("td", {
                class: "vue-ui-table-td-iteration",
                "data-row": a % 2 === 0 ? "odd" : "even",
                style: h({
                  outline: l.value.style.rows.outline
                })
              }, d(o.absoluteIndex + 1), 13, ca),
              (r(!0), i(L, null, j(o.td, (n, u) => (r(), i("td", {
                "data-row": a % 2 === 0 ? "odd" : "even",
                key: `td_${a}_${u}`,
                style: h(Vt(n) || m.dataset.header[u].type === c.value.DATE ? `text-align:right;font-variant-numeric: tabular-nums;outline:${l.value.style.rows.outline}` : `outline:${l.value.style.rows.outline}`),
                onClick: (p) => Se({
                  td: n,
                  rowIndex: a,
                  colIndex: u,
                  headerType: m.dataset.header[u].type,
                  event: p
                }),
                onKeyup: [
                  ve((p) => Se({
                    td: n,
                    rowIndex: a,
                    colIndex: u,
                    headerType: m.dataset.header[u].type,
                    event: p
                  }), ["enter"]),
                  ve((p) => Se({
                    td: n,
                    rowIndex: a,
                    colIndex: u,
                    headerType: m.dataset.header[u].type,
                    event: p
                  }), ["space"])
                ],
                class: D({ "td-numeric": m.dataset.header[u].type === c.value.NUMERIC, "td-focusable": !0, "td-has-nan": E.value[u] }),
                id: `cell_${a}_${u}_${V(G)}`,
                tabindex: "0"
              }, [
                o.meta && o.meta.markerIndices.includes(u) && o.meta.unicodeIcon ? (r(), i("span", {
                  key: 0,
                  style: h(`color:${o.meta.color};margin-right:3px`),
                  innerHTML: o.meta.unicodeIcon
                }, null, 12, da)) : b("", !0),
                m.dataset.header[u].type === c.value.DATE ? (r(), i("span", ha, d(m.dataset.header[u].prefix) + " " + d(new Date(n).toLocaleString().slice(0, 10)) + " " + d(m.dataset.header[u].suffix), 1)) : m.dataset.header[u].isPercentage ? (r(), i("span", ga, d(V(ae)({
                  v: Number(n * 100),
                  s: "%",
                  r: m.dataset.header[u].decimals
                })), 1)) : J.value[u] && m.dataset.header[u].percentageTo && !m.dataset.header[u].isPercentage ? (r(), i("span", {
                  key: 3,
                  class: D({ "td-nan": isNaN(Number(n)) })
                }, d(isNaN(Number(n)) ? `${n} is not ${c.value.NUMERIC}` : V(ae)({
                  p: m.dataset.header[u].prefix,
                  v: Number(n),
                  s: m.dataset.header[u].suffix,
                  r: m.dataset.header[u].decimals
                })) + " (" + d(isNaN(Number(n)) ? "" : V(ae)({
                  v: Number(n / ce(J.value[u].referenceIndex) * 100),
                  s: "%",
                  r: m.dataset.header[u].decimals
                })) + ") ", 3)) : m.dataset.header[u].type === c.value.NUMERIC ? (r(), i("span", {
                  key: 4,
                  class: D({ "td-nan": isNaN(Number(n)) })
                }, d(isNaN(Number(n)) ? `${n} is not ${c.value.NUMERIC}` : V(ae)({
                  p: m.dataset.header[u].prefix,
                  v: Number(n.toFixed(m.dataset.header[u].decimals)),
                  s: m.dataset.header[u].suffix,
                  r: m.dataset.header[u].decimals
                })), 3)) : (r(), i("span", pa, d(m.dataset.header[u].prefix) + " " + d(n) + " " + d(m.dataset.header[u].suffix), 1))
              ], 46, va))), 128))
            ], 14, ia))), 128))
          ], 32)
        ])
      ], 4),
      s("div", {
        class: D({ "td-selector-info": !0, "td-selector-info--active": v.value.col !== void 0 && v.value.rows.length }),
        style: h(`background:${l.value.style.infoBar.backgroundColor};color:${l.value.style.infoBar.color}`)
      }, [
        v.value.col !== void 0 && v.value.rows.length ? (r(), i(L, { key: 0 }, [
          s("div", {
            innerHTML: M.value.table,
            class: "td-selector-icon"
          }, null, 8, fa),
          s("span", null, [
            s("b", null, [
              z(d(m.dataset.header[v.value.col].name) + " ", 1),
              m.dataset.header[v.value.col].isPercentage ? (r(), i("span", ya, " / " + d(m.dataset.header[J.value[v.value.col].referenceIndex].name), 1)) : b("", !0)
            ]),
            s("span", ba, [
              z(d(l.value.translations.nb) + " : ", 1),
              s("b", ma, d(v.value.rows.length), 1)
            ]),
            s("span", ka, [
              z(d(l.value.translations.sum) + " : ", 1),
              m.dataset.header[v.value.col].isPercentage ? (r(), i("b", wa, d(Te.value.sumPercentage), 1)) : (r(), i("b", Ca, d(m.dataset.header[v.value.col].prefix) + " " + d(Te.value.sumRegular) + " " + d(m.dataset.header[v.value.col].suffix), 1)),
              m.dataset.header[v.value.col].isPercentage ? (r(), i("b", $a, "%")) : b("", !0)
            ]),
            s("span", Ma, [
              z(d(l.value.translations.average) + " : ", 1),
              m.dataset.header[v.value.col].isPercentage ? (r(), i("b", xa, d(Te.value.averagePercentage), 1)) : (r(), i("b", Ta, d(m.dataset.header[v.value.col].prefix) + " " + d(Te.value.averageRegular) + " " + d(m.dataset.header[v.value.col].suffix), 1)),
              m.dataset.header[v.value.col].isPercentage ? (r(), i("b", _a, "%")) : b("", !0)
            ])
          ]),
          s("button", {
            onClick: q,
            class: "td-selector-info-reset",
            style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.closeButtons.backgroundColor};color:${l.value.style.closeButtons.color};border-radius:${l.value.style.closeButtons.borderRadius}`)
          }, " ✖ ", 4)
        ], 64)) : b("", !0)
      ], 6),
      N.value.length > 10 ? (r(), i("div", Sa, [
        z(d(l.value.translations.totalRows) + " : " + d(m.dataset.body.length) + " | " + d(l.value.translations.paginatorLabel) + " : ", 1),
        N.value.length > 10 ? le((r(), i("select", {
          key: 0,
          id: "paginatorSelector",
          "onUpdate:modelValue": t[12] || (t[12] = (o) => oe.value = o),
          onChange: t[13] || (t[13] = (o) => {
            q(), Be();
          }),
          style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border}`)
        }, [
          (r(!0), i(L, null, j(Ot.value, (o, a) => (r(), i(L, null, [
            N.value.length > o || m.dataset.body.length === o ? (r(), i("option", {
              key: `paginator_option_${a}`
            }, d(o), 1)) : b("", !0)
          ], 64))), 256))
        ], 36)), [
          [
            bt,
            oe.value,
            void 0,
            { number: !0 }
          ]
        ]) : b("", !0)
      ])) : b("", !0),
      oe.value >= 250 ? (r(), i("div", Na, [
        s("span", {
          innerHTML: M.value.warning
        }, null, 8, Oa),
        z(d(l.value.translations.sizeWarning), 1)
      ])) : b("", !0),
      $.value.length > 1 && $.value.length <= 10 ? (r(), i("div", {
        key: 3,
        class: "vue-ui-table-navigation-indicator",
        style: h(`background:${l.value.style.pagination.navigationIndicator.backgroundColor};width:calc(${k.value / ($.value.length - 1) * 100}%)`)
      }, null, 4)) : b("", !0),
      $.value.length > 1 ? (r(), i("div", La, [
        s("button", {
          class: "vue-ui-table-navigation",
          onClick: t[14] || (t[14] = de((o) => _e("previous"), ["stop"])),
          innerHTML: M.value.chevronLeft,
          disabled: k.value === 0,
          style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.pagination.buttons.backgroundColor};color:${l.value.style.pagination.buttons.color};opacity:${k.value === 0 ? l.value.style.pagination.buttons.opacityDisabled : 1}`)
        }, null, 12, Da),
        $.value.length > 3 ? (r(), i(L, { key: 0 }, [
          s("button", {
            class: "vue-ui-table-navigation",
            onClick: t[15] || (t[15] = de((o) => _e(1), ["stop"])),
            disabled: k.value === 0,
            style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.pagination.buttons.backgroundColor};color:${l.value.style.pagination.buttons.color};opacity:${k.value === 0 ? l.value.style.pagination.buttons.opacityDisabled : 1}`)
          }, " 1 ", 12, Ea),
          $.value.length > 10 ? (r(), i("div", {
            key: 0,
            class: "vue-ui-table-page-scroller-wrapper",
            style: h({ cursor: T.value ? "pointer" : "default" })
          }, [
            s("label", Ia, d(l.value.translations.page) + " " + d(k.value + 1) + " / " + d($.value.length), 1),
            s("input", {
              class: "vue-ui-table-page-scroller",
              id: "pageScroller",
              type: "range",
              step: "1",
              min: 0,
              max: $.value.length - 1,
              onInput: t[16] || (t[16] = (o) => pt(o)),
              value: k.value,
              style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border};accent-color:${l.value.style.inputs.accentColor}`)
            }, null, 44, Aa)
          ], 4)) : (r(), i("span", Pa, d(l.value.translations.page) + " " + d(k.value + 1) + " / " + d($.value.length), 1)),
          s("button", {
            class: "vue-ui-table-navigation",
            onClick: t[17] || (t[17] = de((o) => _e($.value.length), ["stop"])),
            disabled: k.value === $.value.length - 1,
            style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.pagination.buttons.backgroundColor};color:${l.value.style.pagination.buttons.color};opacity:${k.value === $.value.length - 1 ? l.value.style.pagination.buttons.opacityDisabled : 1}`)
          }, d($.value.length), 13, Ha)
        ], 64)) : (r(), i(L, { key: 1 }, [
          z(d(l.value.translations.page) + " " + d(k.value + 1) + " / " + d($.value.length), 1)
        ], 64)),
        s("button", {
          class: "vue-ui-table-navigation",
          onClick: t[18] || (t[18] = de((o) => _e("next"), ["stop"])),
          innerHTML: M.value.chevronRight,
          disabled: k.value === $.value.length - 1,
          style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.pagination.buttons.backgroundColor};color:${l.value.style.pagination.buttons.color};opacity:${k.value === $.value.length - 1 ? l.value.style.pagination.buttons.opacityDisabled : 1}`)
        }, null, 12, Ba)
      ])) : b("", !0),
      je(V(xt), {
        ref_key: "chartModal",
        ref: ye,
        backgroundColor: l.value.style.chart.modal.backgroundColor,
        headerColor: l.value.style.chart.modal.color,
        color: l.value.style.chart.modal.color,
        forcedHeight: 500,
        isCursorPointer: T.value,
        withPadding: "",
        withFullWidth: "",
        noLayerUpdate: "",
        onClose: t[23] || (t[23] = (o) => X.value = !1),
        onSize: Rt
      }, {
        before: mt(() => [
          s("div", Ra, [
            xe.value.length ? (r(), i("button", {
              key: 0,
              onClick: t[19] || (t[19] = (o) => I.value = !0),
              innerHTML: M.value.donut,
              class: D({ "is-active-chart": S.value.type === c.value.DONUT || I.value }),
              style: h(`cursor:${T.value ? "pointer" : "default"};background:${S.value.type === c.value.DONUT || I.value ? l.value.style.chart.modal.buttons.selected.backgroundColor : l.value.style.chart.modal.buttons.unselected.backgroundColor};color:${S.value.type === c.value.DONUT || I.value ? l.value.style.chart.modal.buttons.selected.color : l.value.style.chart.modal.buttons.unselected.color}`)
            }, null, 14, za)) : b("", !0),
            s("button", {
              onClick: t[20] || (t[20] = (o) => {
                S.value.type = c.value.LINE, I.value = !1;
              }),
              innerHTML: M.value.chart,
              class: D({ "is-active-chart": S.value.type === c.value.LINE && !I.value }),
              style: h(`cursor:${T.value ? "pointer" : "default"};background:${S.value.type === c.value.LINE && !I.value ? l.value.style.chart.modal.buttons.selected.backgroundColor : l.value.style.chart.modal.buttons.unselected.backgroundColor};color:${S.value.type === c.value.LINE && !I.value ? l.value.style.chart.modal.buttons.selected.color : l.value.style.chart.modal.buttons.unselected.color}`)
            }, null, 14, Ua),
            s("button", {
              onClick: t[21] || (t[21] = (o) => {
                S.value.type = c.value.BAR, I.value = !1;
              }),
              innerHTML: M.value.bar,
              class: D({ "is-active-chart": S.value.type === c.value.BAR && !I.value }),
              style: h(`cursor:${T.value ? "pointer" : "default"};background:${S.value.type === c.value.BAR && !I.value ? l.value.style.chart.modal.buttons.selected.backgroundColor : l.value.style.chart.modal.buttons.unselected.backgroundColor};color:${S.value.type === c.value.BAR && !I.value ? l.value.style.chart.modal.buttons.selected.color : l.value.style.chart.modal.buttons.unselected.color}`)
            }, null, 14, Fa)
          ])
        ]),
        content: mt(() => [
          s("div", ja, [
            I.value && xe.value.length ? (r(), i("div", {
              key: 0,
              style: h(`background:${l.value.style.chart.modal.backgroundColor};color:${l.value.style.chart.modal.color}`)
            }, [
              s("fieldset", Va, [
                s("legend", null, d(l.value.translations.chooseCategoryColumn), 1),
                s("div", Wa, [
                  (r(!0), i(L, null, j(xe.value, (o, a) => (r(), i("div", {
                    key: `donut_radio_${a}`,
                    class: "vue-ui-table-fieldset-option"
                  }, [
                    s("input", {
                      type: "radio",
                      name: o.name,
                      id: o.name,
                      checked: F.value && o.name === F.value.name,
                      onInput: (n) => F.value = xe.value[a],
                      style: h(`background:${l.value.style.inputs.backgroundColor};color:${l.value.style.inputs.color};border:${l.value.style.inputs.border};accent-color:${l.value.style.inputs.accentColor}`)
                    }, null, 44, Ja),
                    s("label", {
                      for: o.name
                    }, d(o.name), 9, Ka)
                  ]))), 128))
                ]),
                s("button", {
                  class: "vue-ui-table-generate-donut",
                  disabled: !F.value,
                  onClick: et,
                  style: h(`cursor:${T.value ? "pointer" : "default"};background:${l.value.style.chart.modal.buttons.selected.backgroundColor};color:${l.value.style.chart.modal.buttons.selected.color}`)
                }, [
                  s("div", {
                    style: { "margin-bottom": "-3px" },
                    innerHTML: M.value.donut
                  }, null, 8, Ya),
                  z(" " + d(l.value.translations.makeDonut), 1)
                ], 12, Xa)
              ])
            ], 4)) : b("", !0),
            [c.value.BAR, c.value.LINE].includes(S.value.type) && !I.value ? (r(), i(L, { key: 1 }, [
              Ye.value.length > 1 ? (r(), i("label", {
                key: 0,
                style: h({
                  color: l.value.style.chart.modal.color
                })
              }, [
                z(d(l.value.translations.xAxisLabels) + " ", 1),
                le(s("select", {
                  "onUpdate:modelValue": t[22] || (t[22] = (o) => Me.value = o)
                }, [
                  (r(!0), i(L, null, j(Ye.value, (o) => (r(), i("option", null, d(o), 1))), 256))
                ], 512), [
                  [bt, Me.value]
                ])
              ], 4)) : b("", !0),
              s("div", Za, [
                S.value.type === c.value.LINE ? (r(), kt(Ct, {
                  key: `chart_line_${Ke.value}`,
                  dataset: ie.value.xyDatasetLine,
                  config: ie.value.xyConfig
                }, null, 8, ["dataset", "config"])) : b("", !0),
                S.value.type === c.value.BAR ? (r(), kt(Ct, {
                  key: `chart_bar_${Ke.value}`,
                  dataset: ie.value.xyDatasetBar,
                  config: ie.value.xyConfig
                }, null, 8, ["dataset", "config"])) : b("", !0)
              ]),
              v.value.rows.length >= 2 ? (r(), i("div", {
                key: 1,
                class: "chart-trend",
                style: h(`color:${l.value.style.chart.modal.color}`)
              }, [
                t[33] || (t[33] = s("span", null, "---", -1)),
                z(" Trend: " + d(V(ae)({
                  v: ie.value.progression.trend * 100,
                  s: "%",
                  r: 1
                })), 1)
              ], 4)) : b("", !0)
            ], 64)) : b("", !0),
            [c.value.DONUT].includes(S.value.type) && !I.value ? (r(), i("div", Ga, [
              je(rl, {
                dataset: we.value,
                config: ie.value.donutConfig
              }, null, 8, ["dataset", "config"])
            ])) : b("", !0)
          ])
        ]),
        _: 1
      }, 8, ["backgroundColor", "headerColor", "color", "isCursorPointer"])
    ], 4));
  }
}, uo = /* @__PURE__ */ cl(qa, [["__scopeId", "data-v-760e1817"]]);
export {
  uo as default
};
