// vue-data-ui/dist/vue-ui-geo-DykhEWVq.js
import { defineAsyncComponent as me, ref as m, computed as d, toRefs as Co, watch as ie, nextTick as J, onBeforeUnmount as It, onMounted as wo, openBlock as b, createElementBlock as A, normalizeStyle as ve, normalizeClass as xt, createCommentVNode as j, createBlock as K, unref as y, withCtx as w, renderSlot as g, normalizeProps as P, guardReactiveProps as $, createSlots as No, createVNode as St, createElementVNode as ae, Fragment as he, renderList as pe, mergeProps as Ao, toDisplayString as Fo } from "vue";
import { j as Po, t as Ot, s as Be, X as $o, T as Rt } from "./lib-GwPnJSrY.js";
import { u as To } from "./usePanZoom-CDF7rb3o.js";
import { t as Io, u as xo } from "./useResponsive-DfdjqQps.js";
import { u as So } from "./useConfig-CIgyKZ6R.js";
import { u as Oo, B as Ro } from "./BaseScanner-BhhkKVoj.js";
import { u as Mo } from "./usePrinter-DiK5R6k-.js";
import { u as jo } from "./useSvgExport-ByUukOZt.js";
import { u as Ee } from "./useNestedProp-BBT0mbHh.js";
import { u as Bo } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Eo } from "./useUserOptionState-BIvW1Kz7.js";
import { u as _o } from "./useChartAccessibility-9icAAmYg.js";
import Go from "./img-BYYzY_QU.js";
import { g as zo } from "./geoProjections-D0Vxsz_O.js";
import Zo from "./Title-BfyrP-jp.js";
import Lo from "./vue_ui_geo--kg4VoVX.js";
import { B as Mt } from "./BaseZoomControls-CoFievy9.js";
import { _ as Wo } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Yo = ["id"], Do = ["xmlns", "viewBox", "id"], qo = ["transform"], Ho = ["d", "fill", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Xo = ["d", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Uo = ["cx", "cy", "r", "fill", "stroke", "stroke-width", "onMouseenter"], Vo = ["cx", "cy", "r", "fill", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Jo = ["x", "y", "fill", "font-size"], Ko = {
  key: 5,
  class: "vue-data-ui-watermark"
}, jt = 1.5, Qo = {
  __name: "vue-ui-geo",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: [Array, Object],
      default() {
        return [];
      }
    }
  },
  emits: ["copyAlt"],
  setup(Bt, { expose: Et, emit: _t }) {
    const Gt = me(() => import("./Tooltip-DL3QgjZf.js")), zt = me(() => import("./UserOptions-CrLqEimB.js")), Zt = me(() => import("./PenAndPaper-DWV1FMIi.js")), Lt = me(() => import("./PackageVersion-DN9d0ME5.js")), W = Bt, Wt = _t, { vue_ui_geo: Yt } = So(), { isThemeValid: Dt, warnInvalidTheme: qt } = Bo(), B = m(Po()), ye = m(0), _e = m(0), le = m(!1), se = m(!1), S = m(null), E = m(!1), ue = m(null), Ge = m(null), ze = m(null), Ze = m(null), Y = m(null), D = m(null), Le = m(null), We = m(null), _ = m(!1);
    function Ye(e) {
      _.value = e, ye.value += 1;
    }
    function ge() {
      const e = Ee({
        userConfig: W.config,
        defaultConfig: Yt
      }), o = e.theme;
      if (!o) return e;
      if (!Dt.value(e))
        return qt(e), e;
      const t = Ee({
        userConfig: Lo[o] || W.config,
        defaultConfig: e
      });
      return Ee({
        userConfig: W.config,
        defaultConfig: t
      });
    }
    const n = m(ge()), Q = d(() => n.value.userOptions.useCursorPointer), { userOptionsVisible: De, keepUserOptionState: qe } = Eo({
      config: n.value
    }), { svgRef: ce } = _o({
      config: n.value.style.chart.title
    }), Ht = d(() => Ot({
      defaultConfig: {
        map: {
          geoJson: {
            type: "FeatureCollection",
            features: [
              {
                type: "Feature",
                properties: { name: "Island A" },
                geometry: {
                  type: "Polygon",
                  coordinates: [
                    [
                      [-6, 2],
                      [-4, 2],
                      [-4, 4],
                      [-6, 4],
                      [-6, 2]
                    ]
                  ]
                }
              },
              {
                type: "Feature",
                properties: { name: "Island B" },
                geometry: {
                  type: "Polygon",
                  coordinates: [
                    [
                      [-2, -1],
                      [1, -1],
                      [1, 2],
                      [-2, 2],
                      [-2, -1]
                    ]
                  ]
                }
              },
              {
                type: "Feature",
                properties: { name: "Island C" },
                geometry: {
                  type: "Polygon",
                  coordinates: [
                    [
                      [3, -3],
                      [5, -3],
                      [5, -1],
                      [3, -1],
                      [3, -3]
                    ]
                  ]
                }
              },
              {
                type: "Feature",
                properties: { name: "Island D" },
                geometry: {
                  type: "Polygon",
                  coordinates: [
                    [
                      [4, 3],
                      [7, 3],
                      [7, 5],
                      [4, 5],
                      [4, 3]
                    ]
                  ]
                }
              }
            ]
          }
        },
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            territory: {
              fill: "#99999950",
              stroke: "#8A8A8A",
              strokeWidth: 0.5
            }
          }
        }
      },
      userConfig: n.value.skeletonConfig ?? {}
    })), { loading: be, FINAL_DATASET: ke } = Oo({
      ...Co(W),
      FINAL_CONFIG: n,
      prepareConfig: ge,
      allowEmptyDataset: !0,
      skeletonDataset: W.config?.skeletonDataset ?? [],
      skeletonConfig: Ot({
        defaultConfig: n.value,
        userConfig: Ht.value
      })
    }), { isPrinting: Ce, isImaging: we, generatePdf: He, generateImage: Xe } = Mo({
      elementId: `vue-ui-geo_${B.value}`,
      fileName: n.value.style.chart.title.text || "vue-ui-geo",
      options: n.value.userOptions.print
    }), Xt = d(() => n.value.style.chart.controls.show ? n.value.style.chart.controls.position === "top" ? Le.value?.$el ?? null : n.value.style.chart.controls.position === "bottom" ? We.value?.$el ?? null : null : null);
    function Ne() {
      Y.value && (D.value && Y.value.unobserve(D.value), Y.value.disconnect(), Y.value = null, D.value = null);
    }
    const q = d(() => !!n.value.responsive && !_.value);
    ie(
      () => _.value,
      async (e) => {
        if (e) {
          Ne(), H.value = Number(n.value.style.chart.dimensions.width) || H.value, G.value = Number(n.value.style.chart.dimensions.height) || G.value, await J(), E.value = !1, requestAnimationFrame(() => {
            E.value || de();
          });
          return;
        }
        n.value.responsive && (await J(), Ae());
      }
    );
    function Ae() {
      if (Ne(), !q.value) return;
      const e = Io(() => {
        if (!S.value) return;
        const { width: o, height: t } = xo({
          chart: S.value,
          noTitle: Ue.value ? ze.value : null,
          title: n.value.style.chart.title.text ? Ge.value : null,
          legend: Xt.value,
          source: Ze.value
        });
        requestAnimationFrame(() => {
          H.value = Math.max(0.1, Number(o) || 0.1), G.value = Math.max(0.1, Number(t) || 0.1) - 24;
        });
      });
      Y.value = new ResizeObserver(e), D.value = S.value ? S.value.parentNode : null, D.value && Y.value.observe(D.value), e();
    }
    const Ue = d(() => n.value.userOptions.show && !n.value.style.chart.title.text), ee = m({
      showTooltip: n.value.style.chart.tooltip.show
    }), O = m(n.value.style.chart.zoom.active);
    ie(
      () => W.config,
      () => {
        n.value = ge(), ee.value.showTooltip = n.value.style.chart.tooltip.show, O.value = n.value.style.chart.zoom.active, H.value = n.value.style.chart.dimensions.width, G.value = n.value.style.chart.dimensions.height, Ae(), _e.value += 1, ye.value += 1;
      },
      { deep: !0 }
    );
    const { projections: Ve } = zo, Je = d(() => n.value?.projection || "equirectangular"), Ut = d(() => {
      const e = Ve?.[Je.value];
      return typeof e == "function" ? e : Ve.equirectangular;
    }), Ke = d(() => n.value.map.geoJson ?? null), Qe = {
      aitoff: { width: 1e3, height: 500 },
      azimuthalEquidistant: { width: 1e3, height: 1e3 },
      bonne: { width: 1e3, height: 1e3 },
      equirectangular: { width: 1e3, height: 700 },
      gallPeters: { width: 1e3, height: 800 },
      globe: { width: 1e3, height: 1e3 },
      hammer: { width: 1e3, height: 500 },
      mercator: { width: 1e3, height: 750 },
      mollweide: { width: 1e3, height: 600 },
      robinson: { width: 1e3, height: 600 },
      sinusoidal: { width: 1e3, height: 500 },
      vanDerGrinten: { width: 1e3, height: 1e3 },
      winkelTripel: { width: 1e3, height: 1e3 }
    }, H = m(n.value.style.chart.dimensions.width), G = m(n.value.style.chart.dimensions.height), Fe = d(() => {
      const e = Qe[Je.value] || Qe.equirectangular;
      return { width: e.width, height: e.height };
    }), T = d(() => {
      const e = Fe.value, o = H.value, t = G.value, r = Number.isFinite(o) && o > 0, i = Number.isFinite(t) && t > 0;
      if (r && i)
        return { width: o, height: t };
      if (r && !i) {
        const a = e.height / e.width;
        return { width: o, height: o * a };
      }
      if (!r && i) {
        const a = e.width / e.height;
        return { width: t * a, height: t };
      }
      return { width: e.width, height: e.height };
    }), Vt = d(() => {
      const e = n.value?.map?.center, o = [0, 0];
      if (!Array.isArray(e) || e.length !== 2) return o;
      const t = Number(e[0]), r = Number(e[1]);
      return !Number.isFinite(t) || !Number.isFinite(r) ? o : [t, r];
    });
    function X([e, o]) {
      return Ut.value(
        [e, o],
        Fe.value.width,
        Fe.value.height,
        Vt.value
      );
    }
    function Jt(e, o = {}) {
      const {
        defaultName: t = "",
        namePropertyCandidates: r = [
          "name",
          "nom",
          "admin",
          "NAME",
          "label",
          "title",
          "description",
          "DESCRIPTION",
          "NAME_1",
          "NAME_2",
          "NAME_3",
          "NAME_EN",
          "name:en",
          "name_en",
          "localname",
          "local_name"
        ],
        includeNullGeometries: i = !1
      } = o, a = { type: "FeatureCollection", features: [] };
      function l(u) {
        return u !== null && typeof u == "object" && !Array.isArray(u);
      }
      function c(u) {
        return !l(u) || typeof u.type != "string" ? !1 : u.type === "GeometryCollection" ? Array.isArray(u.geometries) : Object.hasOwn(u, "coordinates");
      }
      function f(u) {
        if (!l(u)) return "";
        for (const h of r) {
          const C = u[h];
          if (typeof C == "string" && C.trim())
            return C.trim();
        }
        return "";
      }
      function s({ geometry: u, properties: h, fallbackName: C, featureIndex: v }) {
        const p = l(h) ? { ...h } : {}, M = f(p), N = t ? `${t} ${v + 1}` : "", V = M || (typeof C == "string" ? C.trim() : "") || N;
        return !p.name && V && (p.name = V), {
          type: "Feature",
          geometry: u,
          properties: p
        };
      }
      function F(u, h, C, v, p) {
        const M = Array.isArray(u?.geometries) ? u.geometries : [];
        let N = v;
        for (let V = 0; V < M.length; V += 1) {
          const Z = M[V];
          if (Z) {
            if (l(Z) && Z.type === "GeometryCollection") {
              N = F(
                Z,
                h,
                C,
                N,
                p
              );
              continue;
            }
            if (c(Z)) {
              if (Z.coordinates == null && Z.type !== "GeometryCollection") {
                i && (p.push(
                  s({
                    geometry: null,
                    properties: h,
                    fallbackName: C,
                    featureIndex: N
                  })
                ), N += 1);
                continue;
              }
              p.push(
                s({
                  geometry: Z,
                  properties: h,
                  fallbackName: C,
                  featureIndex: N
                })
              ), N += 1;
            }
          }
        }
        return N;
      }
      function k(u, h, C = 0) {
        let v = C;
        if (u == null) return v;
        if (Array.isArray(u)) {
          for (const p of u)
            v = k(p, h, v);
          return v;
        }
        if (!l(u) || typeof u.type != "string")
          return v;
        if (u.type === "FeatureCollection") {
          const p = Array.isArray(u.features) ? u.features : [];
          for (const M of p)
            v = k(M, h, v);
          return v;
        }
        if (u.type === "Feature") {
          const p = l(u.properties) ? u.properties : {}, M = f(p), N = u.geometry ?? null;
          return N == null ? (i && (h.push(
            s({
              geometry: null,
              properties: p,
              fallbackName: M,
              featureIndex: v
            })
          ), v += 1), v) : l(N) && N.type === "GeometryCollection" ? F(
            N,
            p,
            M,
            v,
            h
          ) : (c(N) && (h.push(
            s({
              geometry: N,
              properties: p,
              fallbackName: M,
              featureIndex: v
            })
          ), v += 1), v);
        }
        return u.type === "GeometryCollection" ? F(u, {}, "", v, h) : (c(u) && (h.push(
          s({
            geometry: u,
            properties: {},
            fallbackName: "",
            featureIndex: v
          })
        ), v += 1), v);
      }
      const x = [];
      return k(e, x, 0), { ...a, features: x };
    }
    function Kt(e) {
      const o = Array.isArray(e?.features) ? e.features : [];
      function t(r) {
        if (!r || typeof r != "object") return "";
        const i = [
          r.name,
          r.nom,
          r.admin,
          r.NAME,
          r.label,
          r.title,
          r.description,
          r.DESCRIPTION,
          r.NAME_1,
          r.NAME_2,
          r.NAME_3,
          r.NAME_EN,
          r["name:en"],
          r.name_en,
          r.localname,
          r.local_name
        ];
        for (const a of i)
          if (typeof a == "string" && a.trim()) return a.trim();
        for (const a of Object.values(r))
          if (typeof a == "string" && a.trim()) return a.trim();
        return "";
      }
      return o.filter((r) => r && r.type === "Feature" && r.geometry).map((r, i) => {
        const a = r.properties || {}, l = t(a);
        return {
          feature: r,
          geometry: r.geometry,
          properties: a,
          name: l,
          uid: `map-feature-${B.value}-${i}`,
          index: i
        };
      });
    }
    const Qt = d(
      () => Jt(Ke.value, {
        defaultName: "",
        includeNullGeometries: !1
      })
    ), fe = d(
      () => Kt(Qt.value)
    );
    function Pe(e) {
      if (!e || typeof e.type != "string") return "";
      if (e.type === "GeometryCollection")
        return (Array.isArray(e.geometries) ? e.geometries : []).map((a) => Pe(a)).filter(Boolean).join(" ");
      const o = (i) => {
        if (!Array.isArray(i) || i.length < 2) return null;
        const a = Number(i[0]), l = Number(i[1]);
        if (!Number.isFinite(a) || !Number.isFinite(l)) return null;
        const c = X([a, l]);
        if (!Array.isArray(c) || c.length < 2) return null;
        const f = Number(c[0]), s = Number(c[1]);
        return !Number.isFinite(f) || !Number.isFinite(s) ? null : [f, s];
      }, t = (i) => {
        const a = (i || []).map(o).filter(Boolean);
        return a.length < 2 ? "" : "M" + a.map(([l, c]) => `${l},${c}`).join("L");
      }, r = (i) => (i || []).map((a) => {
        const l = t(a);
        return l ? l + "Z" : "";
      }).filter(Boolean).join(" ");
      return e.type === "Polygon" ? r(e.coordinates) : e.type === "MultiPolygon" ? (e.coordinates || []).map(r).filter(Boolean).join(" ") : e.type === "LineString" ? t(e.coordinates) : e.type === "MultiLineString" ? (e.coordinates || []).map(t).filter(Boolean).join(" ") : "";
    }
    function et(e) {
      return e ? e.type === "GeometryCollection" ? (Array.isArray(e.geometries) ? e.geometries : []).flatMap((t) => et(t)) : e.type === "Point" ? Array.isArray(e.coordinates) ? [e.coordinates] : [] : e.type === "MultiPoint" ? Array.isArray(e.coordinates) ? e.coordinates : [] : [] : [];
    }
    function tt(e) {
      return !e || typeof e.type != "string" ? !1 : e.type === "GeometryCollection" ? (Array.isArray(e.geometries) ? e.geometries : []).some((t) => tt(t)) : e.type === "Point" || e.type === "MultiPoint";
    }
    const eo = d(
      () => fe.value.filter((e) => {
        const o = e.geometry;
        return o ? o.type === "GeometryCollection" ? (Array.isArray(o.geometries) ? o.geometries : []).some(
          (r) => r?.type === "Polygon" || r?.type === "MultiPolygon"
        ) : o.type === "Polygon" || o.type === "MultiPolygon" : !1;
      })
    ), to = d(
      () => fe.value.filter((e) => {
        const o = e.geometry;
        return o ? o.type === "GeometryCollection" ? (Array.isArray(o.geometries) ? o.geometries : []).some(
          (r) => r?.type === "LineString" || r?.type === "MultiLineString"
        ) : o.type === "LineString" || o.type === "MultiLineString" : !1;
      })
    ), oo = d(
      () => fe.value.filter((e) => {
        const o = e.geometry;
        return o ? tt(o) : !1;
      })
    ), ot = d(() => eo.value.map((e) => {
      const o = Pe(e.geometry);
      return o ? { ...e, path: o } : null;
    }).filter(Boolean)), nt = d(() => to.value.map((e) => {
      const o = Pe(e.geometry);
      return o ? { ...e, path: o } : null;
    }).filter(Boolean)), rt = d(() => {
      const e = [];
      function o(t = {}) {
        const r = t?.style && typeof t.style == "object" ? t.style : {}, i = t.radius ?? t.r ?? r.radius ?? r.r, a = t.color ?? t.fill ?? r.color ?? r.fill, l = t.stroke ?? r.stroke, c = t.strokeWidth ?? t.stroke_width ?? r.strokeWidth ?? r.stroke_width, f = Number.isFinite(Number(i)) ? Number(i) : Number(n.value.style.chart.points.radius), s = a != null && String(a).trim() ? Be(String(a).trim()) : n.value.style.chart.points.fill, F = l != null && String(l).trim() ? Be(String(l).trim()) : n.value.style.chart.points.stroke, k = Number.isFinite(Number(c)) ? Number(c) : Number(n.value.style.chart.points.strokeWidth);
        return { radius: f, fill: s, stroke: F, strokeWidth: k };
      }
      for (const t of oo.value) {
        const r = et(t.geometry);
        for (let i = 0; i < r.length; i += 1) {
          const a = r[i];
          if (!Array.isArray(a) || a.length < 2) continue;
          const l = Number(a[0]), c = Number(a[1]);
          if (!Number.isFinite(l) || !Number.isFinite(c)) continue;
          const f = X([l, c]), s = f?.[0], F = f?.[1];
          if (!Number.isFinite(s) || !Number.isFinite(F)) continue;
          const k = t?.feature?.properties && typeof t.feature.properties == "object" ? t.feature.properties : {}, x = o(k);
          e.push({
            uid: `${t.uid}-geojson-point-${i}`,
            name: t.name || "",
            x: s,
            y: F,
            ...x,
            originalFeature: t.feature,
            featureIndex: t.index,
            pointIndex: i
          });
        }
      }
      return e;
    });
    function it(e) {
      if (!e || typeof e.type != "string") return [];
      const o = [], t = (i) => {
        if (!Array.isArray(i) || i.length < 2) return;
        const a = Number(i[0]), l = Number(i[1]);
        if (!Number.isFinite(a) || !Number.isFinite(l)) return;
        const c = X([a, l]), f = c?.[0], s = c?.[1];
        !Number.isFinite(f) || !Number.isFinite(s) || o.push([f, s]);
      }, r = (i) => {
        if (Array.isArray(i)) {
          if (i.length >= 2 && typeof i[0] == "number" && typeof i[1] == "number") {
            t(i);
            return;
          }
          for (const a of i)
            r(a);
        }
      };
      if (e.type === "GeometryCollection" && Array.isArray(e.geometries)) {
        for (const i of e.geometries)
          o.push(...it(i));
        return o;
      }
      return e.type === "Point" ? (t(e.coordinates), o) : e.type === "MultiPoint" ? ((e.coordinates || []).forEach(t), o) : (r(e.coordinates), o);
    }
    function no(e) {
      const o = [];
      for (const f of e)
        o.push(...it(f.geometry));
      if (!o.length) return null;
      let t = 1 / 0, r = 1 / 0, i = -1 / 0, a = -1 / 0;
      for (const [f, s] of o)
        f < t && (t = f), s < r && (r = s), f > i && (i = f), s > a && (a = s);
      const l = i - t, c = a - r;
      return !(l > 0) || !(c > 0) ? null : { minX: t, minY: r, maxX: i, maxY: a, width: l, height: c };
    }
    const ro = d(() => no(fe.value));
    function io({ bounds: e, targetWidth: o, targetHeight: t, padding: r }) {
      if (!e || !(e.width > 0) || !(e.height > 0))
        return { scale: 1, translateX: 0, translateY: 0, transform: "" };
      const i = Number.isFinite(Number(r)) ? Number(r) : 0, a = Math.max(1, o - i * 2), l = Math.max(1, t - i * 2), c = a / e.width, f = l / e.height, s = Math.min(c, f), F = e.width * s, k = e.height * s, x = i + (a - F) / 2, u = i + (l - k) / 2, h = x - e.minX * s, C = u - e.minY * s;
      return {
        scale: s,
        translateX: h,
        translateY: C,
        transform: `translate(${h} ${C}) scale(${s})`
      };
    }
    const at = d(() => {
      const e = ro.value;
      if (!e) return { scale: 1, translateX: 0, translateY: 0, transform: "" };
      const o = n.value?.map?.fitPadding, t = Number.isFinite(Number(o)) ? Number(o) : 0;
      return io({
        bounds: e,
        targetWidth: T.value.width,
        targetHeight: T.value.height,
        padding: t
      });
    }), ao = d(() => {
      const e = T.value.width, o = T.value.height;
      return `0 0 ${e} ${o}`;
    }), {
      viewBox: te,
      resetZoom: lt,
      setInitialViewBox: $e,
      scale: st,
      zoomByFactor: ut
    } = To(
      ce,
      { x: 0, y: 0, width: T.value.width, height: T.value.height },
      1,
      O
    );
    function lo() {
      O.value = !O.value;
    }
    function ct(e) {
      const o = at.value, t = Number(o?.scale) || 1, r = Number(o?.translateX) || 0, i = Number(o?.translateY) || 0;
      return {
        x: e.x * t + r,
        y: e.y * t + i
      };
    }
    function ft() {
      const e = n.value?.map?.center, o = Number(e?.[0]), t = Number(e?.[1]), r = Array.isArray(e) && e.length === 2 && Number.isFinite(o) && Number.isFinite(t) && (o !== 0 || t !== 0), i = T.value.width, a = T.value.height;
      if (!r)
        return { x: 0, y: 0, width: i, height: a };
      const l = X([o, t]), c = l?.[0], f = l?.[1];
      if (!Number.isFinite(c) || !Number.isFinite(f))
        return { x: 0, y: 0, width: i, height: a };
      const s = ct({ x: c, y: f });
      return {
        x: s.x - i / 2,
        y: s.y - a / 2,
        width: i,
        height: a
      };
    }
    function de() {
      const e = ft();
      $e(e, { overwriteCurrentIfNotZoomed: !0 }), E.value = !0;
    }
    const dt = d(() => {
      const e = te.value;
      return e ? `${e.x} ${e.y} ${e.width} ${e.height}` : ao.value;
    });
    function Te() {
      ut(jt, !0);
    }
    function Ie() {
      ut(1 / jt, !0);
    }
    async function mt() {
      const e = ft();
      $e(e, { overwriteCurrentIfNotZoomed: !0 }), lt(!0);
    }
    ie(
      () => [
        n.value.projection,
        n.value.map.center?.[0],
        n.value.map.center?.[1],
        n.value.map.fitPadding,
        Ke.value
      ],
      async () => {
        E.value = !1, await J(), requestAnimationFrame(() => {
          E.value || de();
        });
      },
      { deep: !1 }
    ), ie(
      () => [H.value, G.value, q.value],
      async () => {
        q.value && (await J(), $e(
          { x: 0, y: 0, width: T.value.width, height: T.value.height },
          { overwriteCurrentIfNotZoomed: !0 }
        ), E.value = !1, requestAnimationFrame(() => {
          E.value || de();
        }));
      },
      { flush: "post" }
    );
    function so(e) {
      return e ? e.type === "FeatureCollection" && Array.isArray(e.features) ? e.features.filter((o) => o?.type === "Feature" && o.geometry?.type === "Point").map((o, t) => {
        const r = o.properties || {}, i = r.name || r.label || r.title || `Point ${t + 1}`, a = o.geometry.coordinates;
        return {
          uid: `map-point-${B.value}-${t}`,
          name: i,
          coordinates: a,
          color: r.color ?? null,
          radius: r.radius ?? null,
          hoverRadiusRatio: n.value.style.chart.points.hoverRadiusRatio,
          description: r.description ?? null,
          original: o,
          index: t
        };
      }) : Array.isArray(e) ? e.map((o, t) => {
        let r = null, i = `Point ${t + 1}`, a = null, l = null, c = null;
        Array.isArray(o) && o.length >= 2 ? r = [Number(o[0]), Number(o[1])] : o && typeof o == "object" && (Array.isArray(o.coordinates) && o.coordinates.length >= 2 ? r = [Number(o.coordinates[0]), Number(o.coordinates[1])] : Number.isFinite(Number(o.lon)) && Number.isFinite(Number(o.lat)) && (r = [Number(o.lon), Number(o.lat)]), typeof o.name == "string" && o.name.trim() && (i = o.name), o.description != null && (a = o.description), o.color != null && (l = o.color), o.radius != null && (c = o.radius));
        const f = r ? r[0] : NaN, s = r ? r[1] : NaN;
        return !Number.isFinite(f) || !Number.isFinite(s) ? null : {
          uid: `map-point-${B.value}-${t}`,
          name: i,
          coordinates: [f, s],
          color: l,
          radius: c,
          hoverRadiusRatio: n.value.style.chart.points.hoverRadiusRatio,
          description: a,
          original: o,
          index: t
        };
      }).filter(Boolean) : typeof e == "object" ? Object.entries(e).map(([o, t], r) => {
        if (!t || typeof t != "object") return null;
        const i = Array.isArray(t.coordinates) ? t.coordinates : null;
        if (!i || i.length < 2) return null;
        const a = Number(i[0]), l = Number(i[1]);
        return !Number.isFinite(a) || !Number.isFinite(l) ? null : {
          uid: `map-point-${B.value}-${r}-${o}`,
          name: t.name || o,
          coordinates: [a, l],
          color: t.color ?? null,
          radius: t.radius ?? null,
          hoverRadiusRatio: n.value.style.chart.points.hoverRadiusRatio,
          description: t.description ?? null,
          original: t,
          index: r
        };
      }).filter(Boolean) : [] : [];
    }
    const vt = d(() => so(ke.value)), ht = d(() => vt.value.map((e) => {
      const o = X(e.coordinates), t = o?.[0], r = o?.[1];
      return !Number.isFinite(t) || !Number.isFinite(r) ? null : {
        ...e,
        x: t,
        y: r,
        fill: e.color ? Be(e.color) : n.value.style.chart.points.fill,
        radius: Number.isFinite(Number(e.radius)) ? Number(e.radius) : n.value.style.chart.points.radius
      };
    }).filter(Boolean)), xe = m(!1), Se = m(""), z = m(!1), oe = m(null), I = m(null);
    function Oe(e) {
      Se.value = e, xe.value = !0;
    }
    function ne() {
      xe.value = !1, Se.value = "";
    }
    function Re(e) {
      return String(e ?? "").replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;").replaceAll("'", "&#039;");
    }
    function uo(e) {
      ue.value = {
        datapoint: e,
        seriesIndex: e.index,
        series: vt.value,
        config: n.value
      }, z.value = !1;
      const o = n.value.style.chart.tooltip.customFormat;
      if (Rt(o))
        try {
          const i = o(ue.value);
          if (typeof i == "string")
            return z.value = !0, i;
        } catch {
          console.warn("Custom format cannot be applied."), z.value = !1;
        }
      const t = Re(e.name), r = e.description != null ? `<div style="margin-top:6px">${Re(e.description)}</div>` : "";
      return `<div><div style="font-weight:600">${t}</div>${r}</div>`;
    }
    function pt(e) {
      z.value = !1;
      const o = n.value.style.chart.tooltip.customFormat;
      if (Rt(o))
        try {
          const r = o({
            datapoint: e,
            config: n.value
          });
          if (typeof r == "string")
            return z.value = !0, r;
        } catch {
          console.warn("Custom format cannot be applied."), z.value = !1;
        }
      const t = typeof e?.name == "string" ? e.name.trim() : "";
      return t ? `<div><div style="font-weight:600">${Re(t)}</div></div>` : (z.value = !1, "");
    }
    function yt(e) {
      if (n.value.events.territoryEnter && n.value.events.territoryEnter({
        datapoint: e,
        seriesIndex: e.index
      }), !(typeof e?.name == "string" ? e.name.trim() : "") && !n.value.style.chart.territory.hover.enabledWhenEmpty) {
        I.value = null, ne();
        return;
      }
      I.value = e.uid, Oe(pt(e));
    }
    function gt(e) {
      I.value = null, ne(), n.value.events.territoryLeave && n.value.events.territoryLeave({
        datapoint: e,
        seriesIndex: e.index
      });
    }
    function bt(e) {
      n.value.events.territoryClick && n.value.events.territoryClick({
        datapoint: e,
        seriesIndex: e.index
      });
    }
    function kt(e) {
      oe.value = e.uid, Oe(uo(e)), n.value.events.datapointEnter && n.value.events.datapointEnter({
        datapoint: e,
        seriesIndex: e.index
      });
    }
    function Ct(e) {
      oe.value = null, ne(), n.value.events.datapointLeave && n.value.events.datapointLeave({
        datapoint: e,
        seriesIndex: e.index
      });
    }
    const wt = d(() => !!n.value.events.datapointClick && typeof n.value.events.datapointClick == "function");
    function Nt(e) {
      wt.value && n.value.events.datapointClick({
        datapoint: e,
        seriesIndex: e.index
      });
    }
    const R = d(() => n.value.style.chart.territory), At = d(() => n.value.style.chart.points);
    function co(e) {
      const o = typeof e?.name == "string" ? e.name.trim() : "";
      if (!o) {
        I.value = null, ne();
        return;
      }
      const t = {
        name: o,
        uid: e.uid,
        index: e.featureIndex,
        properties: {}
      };
      I.value = e.uid, Oe(pt(t));
    }
    function fo() {
      I.value = null, ne();
    }
    const re = m(!1);
    function Me() {
      re.value = !re.value;
    }
    ie(
      () => re.value,
      (e) => {
        O.value = !e;
      }
    );
    function Ft() {
      ee.value.showTooltip = !ee.value.showTooltip;
    }
    const mo = d(() => n.value.style.chart.backgroundColor), vo = d(() => n.value.style.chart.title), { exportSvg: ho, getSvg: po } = jo({
      svg: ce,
      title: vo,
      legend: void 0,
      legendItems: void 0,
      backgroundColor: mo
    });
    async function Pt({ isCb: e }) {
      se.value = !0, await J();
      try {
        if (e) {
          const { blob: o, url: t, text: r, dataUrl: i } = await po();
          await Promise.resolve(n.value.userOptions.callbacks.svg({ blob: o, url: t, text: r, dataUrl: i }));
        } else
          await Promise.resolve(ho());
      } finally {
        se.value = !1;
      }
    }
    function yo(e) {
      if (e?.stage === "start") {
        le.value = !0;
        return;
      }
      if (e?.stage === "end") {
        le.value = !1;
        return;
      }
      Xe();
    }
    async function go({ scale: e = 2 } = {}) {
      if (!S.value) return;
      const { width: o, height: t } = S.value.getBoundingClientRect(), r = o / t, { imageUri: i, base64: a } = await Go({
        domElement: S.value,
        base64: !0,
        img: !0,
        scale: e
      });
      return {
        imageUri: i,
        base64: a,
        title: n.value.style.chart.title.text,
        width: o,
        height: t,
        aspectRatio: r
      };
    }
    const L = m(!1);
    let U = null;
    function bo() {
      O.value && (L.value = !0);
    }
    function je() {
      L.value = !1;
    }
    function ko() {
      O.value && (L.value = !0, U && clearTimeout(U), U = setTimeout(() => {
        L.value = !1, U = null;
      }, 140));
    }
    It(() => {
      U && clearTimeout(U);
    });
    async function $t([e, o], { animated: t = !0 } = {}) {
      const r = Number(e), i = Number(o);
      if (!Number.isFinite(r) || !Number.isFinite(i)) return;
      const a = X([r, i]), l = a?.[0], c = a?.[1];
      if (!Number.isFinite(l) || !Number.isFinite(c)) return;
      const f = ct({ x: l, y: c }), s = te.value ? { ...te.value } : { x: 0, y: 0, width: T.value.width, height: T.value.height }, F = {
        x: f.x - s.width / 2,
        y: f.y - s.height / 2,
        width: s.width,
        height: s.height
      };
      if (!t) {
        te.value = F;
        return;
      }
      const k = { ...s }, x = { ...F }, u = 250;
      let h = null;
      const C = (v) => {
        h == null && (h = v);
        const p = Math.min((v - h) / u, 1);
        te.value = {
          x: k.x + (x.x - k.x) * p,
          y: k.y + (x.y - k.y) * p,
          width: k.width + (x.width - k.width) * p,
          height: k.height + (x.height - k.height) * p
        }, p < 1 && requestAnimationFrame(C);
      };
      requestAnimationFrame(C);
    }
    wo(async () => {
      Ae(), await J(), requestAnimationFrame(() => {
        E.value || (de(), (n.value.map.center?.[0] || n.value.map.center?.[1]) && $t(n.value.map.center, { animated: !1 }));
      });
    }), It(() => {
      Ne();
    });
    async function Tt() {
      if (Wt("copyAlt", {
        config: n.value,
        dataset: ke.value
      }), !n.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(n.value.userOptions.callbacks.altCopy({
        config: n.value,
        dataset: ke.value
      }));
    }
    return Et({
      getImage: go,
      generatePdf: He,
      generateImage: Xe,
      generateSvg: Pt,
      toggleTooltip: Ft,
      toggleAnnotator: Me,
      toggleFullscreen: Ye,
      zoomIn: Te,
      zoomOut: Ie,
      resetZoom: lt,
      focusLocation: $t,
      copyAlt: Tt
    }), (e, o) => (b(), A("div", {
      class: xt(`vue-data-ui-component vue-ui-geo ${_.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${q.value ? "vue-ui-geo-responsive" : ""} ${L.value ? "vue-ui-geo-interacting" : ""}`),
      ref_key: "geoChart",
      ref: S,
      id: `map_${B.value}`,
      dir: "auto",
      style: ve({
        fontFamily: n.value.style.fontFamily,
        width: "100%",
        backgroundColor: n.value.style.chart.backgroundColor,
        height: q.value ? `${G.value}px` : void 0
      })
    }, [
      Ue.value ? (b(), A("div", {
        key: 0,
        ref_key: "noTitle",
        ref: ze,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : j("", !0),
      n.value.style.chart.title.text ? (b(), A("div", {
        key: 1,
        ref_key: "chartTitle",
        ref: Ge,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (b(), K(Zo, {
          key: `title_${_e.value}`,
          config: {
            title: {
              cy: "geo-title",
              ...n.value.style.chart.title
            },
            subtitle: {
              cy: "geo-subtitle",
              ...n.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : j("", !0),
      n.value.userOptions.buttons.annotator ? (b(), K(y(Zt), {
        key: 2,
        svgRef: y(ce),
        backgroundColor: n.value.style.chart.backgroundColor,
        color: n.value.style.chart.color,
        active: re.value,
        isCursorPointer: Q.value,
        onClose: Me
      }, {
        "annotator-action-close": w(() => [
          g(e.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": w(({ color: t }) => [
          g(e.$slots, "annotator-action-color", P($({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": w(({ mode: t }) => [
          g(e.$slots, "annotator-action-draw", P($({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": w(({ disabled: t }) => [
          g(e.$slots, "annotator-action-undo", P($({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": w(({ disabled: t }) => [
          g(e.$slots, "annotator-action-redo", P($({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": w(({ disabled: t }) => [
          g(e.$slots, "annotator-action-delete", P($({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : j("", !0),
      n.value.userOptions.show && (y(qe) || y(De)) ? (b(), K(y(zt), {
        ref: "userOptionsRef",
        key: `user_options_${ye.value}`,
        backgroundColor: n.value.style.chart.backgroundColor,
        color: n.value.style.chart.color,
        isPrinting: y(Ce),
        isImaging: y(we),
        uid: B.value,
        hasXls: !1,
        hasTable: !1,
        hasTooltip: n.value.userOptions.buttons.tooltip && n.value.style.chart.tooltip.show,
        hasPdf: n.value.userOptions.buttons.pdf,
        hasImg: n.value.userOptions.buttons.img,
        hasSvg: n.value.userOptions.buttons.svg,
        hasFullscreen: n.value.userOptions.buttons.fullscreen,
        hasAltCopy: n.value.userOptions.buttons.altCopy,
        hasAnnotator: n.value.userOptions.buttons.annotator,
        hasZoom: n.value.userOptions.buttons.zoom,
        isZoom: O.value,
        isFullscreen: _.value,
        isTooltip: ee.value.showTooltip,
        titles: { ...n.value.userOptions.buttonTitles },
        chartElement: S.value,
        position: n.value.userOptions.position,
        isAnnotation: re.value,
        callbacks: n.value.userOptions.callbacks,
        printScale: n.value.userOptions.print.scale,
        isCursorPointer: Q.value,
        onToggleFullscreen: Ye,
        onGeneratePdf: y(He),
        onGenerateImage: yo,
        onGenerateSvg: Pt,
        onToggleTooltip: Ft,
        onToggleAnnotator: Me,
        onToggleZoom: lo,
        onCopyAlt: Tt,
        style: ve({
          visibility: y(qe) ? y(De) ? "visible" : "hidden" : "visible"
        })
      }, No({ _: 2 }, [
        e.$slots.menuIcon ? {
          name: "menuIcon",
          fn: w(({ isOpen: t, color: r }) => [
            g(e.$slots, "menuIcon", P($({ isOpen: t, color: r })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        e.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: w(() => [
            g(e.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        e.$slots.optionPdf ? {
          name: "optionPdf",
          fn: w(() => [
            g(e.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        e.$slots.optionImg ? {
          name: "optionImg",
          fn: w(() => [
            g(e.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        e.$slots.optionSvg ? {
          name: "optionSvg",
          fn: w(() => [
            g(e.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        e.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: w(({ toggleFullscreen: t, isFullscreen: r }) => [
            g(e.$slots, "optionFullscreen", P($({ toggleFullscreen: t, isFullscreen: r })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        e.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: w(({ toggleAnnotator: t, isAnnotator: r }) => [
            g(e.$slots, "optionAnnotator", P($({ toggleAnnotator: t, isAnnotator: r })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        e.$slots.optionZoom ? {
          name: "optionZoom",
          fn: w(({ toggleZoom: t, isZoomLocked: r }) => [
            g(e.$slots, "optionZoom", P($({ toggleZoom: t, isZoomLocked: r })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        e.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: w(({ altCopy: t }) => [
            g(e.$slots, "optionAltCopy", P($({ altCopy: t })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "hasAnnotator", "hasZoom", "isZoom", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "isAnnotation", "callbacks", "printScale", "isCursorPointer", "onGeneratePdf", "style"])) : j("", !0),
      n.value.style.chart.controls.position === "top" && n.value.style.chart.controls.show && !y(be) ? (b(), K(Mt, {
        key: 4,
        ref_key: "zoomControlsTop",
        ref: Le,
        config: n.value,
        scale: y(st),
        isFullscreen: _.value,
        isCursorPointer: Q.value,
        onZoomIn: Te,
        onZoomOut: Ie,
        onResetZoom: mt
      }, null, 8, ["config", "scale", "isFullscreen", "isCursorPointer"])) : j("", !0),
      (b(), A("svg", {
        ref_key: "svgRef",
        ref: ce,
        xmlns: y($o),
        viewBox: dt.value,
        preserveAspectRatio: "xMidYMid meet",
        style: ve({
          display: "block",
          width: "100%",
          height: q.value ? `${G.value}px` : "auto",
          background: n.value.style.chart.backgroundColor,
          touchAction: O.value ? "none" : "auto",
          cursor: O.value ? L.value ? "grabbing" : "grab" : "default"
        }),
        id: `vue-ui-geo_${B.value}`,
        onPointerdown: bo,
        onPointerup: je,
        onPointercancel: je,
        onPointerleave: je,
        onWheel: ko
      }, [
        St(y(Lt)),
        ae("g", {
          transform: at.value.transform,
          style: ve({ pointerEvents: L.value ? "none" : "auto" })
        }, [
          ae("g", null, [
            (b(!0), A(he, null, pe(ot.value, (t) => (b(), A("path", {
              class: "vue-ui-geo-territory",
              key: t.uid,
              d: t.path,
              fill: I.value === t.uid ? R.value.hover.fill : R.value.fill,
              stroke: I.value === t.uid ? R.value.hover.stroke : R.value.stroke,
              "stroke-width": I.value === t.uid ? R.value.hover.strokeWidth : R.value.strokeWidth,
              "vector-effect": "non-scaling-stroke",
              onMouseenter: (r) => yt(t),
              onMouseleave: (r) => gt(t),
              onClick: (r) => bt(t)
            }, null, 40, Ho))), 128))
          ]),
          ae("g", null, [
            (b(!0), A(he, null, pe(nt.value, (t) => (b(), A("path", {
              class: "vue-ui-geo-territory",
              key: t.uid,
              d: t.path,
              fill: "none",
              stroke: I.value === t.uid ? R.value.hover.stroke : R.value.stroke,
              "stroke-width": I.value === t.uid ? R.value.hover.strokeWidth : R.value.strokeWidth,
              "vector-effect": "non-scaling-stroke",
              onMouseenter: (r) => yt(t),
              onMouseleave: (r) => gt(t),
              onClick: (r) => bt(t)
            }, null, 40, Xo))), 128))
          ]),
          ae("g", null, [
            (b(!0), A(he, null, pe(rt.value, (t) => (b(), A("circle", {
              key: t.uid,
              cx: t.x,
              cy: t.y,
              r: t.radius,
              fill: t.fill,
              stroke: t.stroke,
              "stroke-width": t.strokeWidth,
              "vector-effect": "non-scaling-stroke",
              onMouseenter: (r) => co(t),
              onMouseleave: fo
            }, null, 40, Uo))), 128))
          ]),
          (b(!0), A(he, null, pe(ht.value, (t) => (b(), A("g", {
            key: t.uid
          }, [
            g(e.$slots, "datapoint", Ao({ ref_for: !0 }, {
              point: t,
              onPointEnter: kt,
              onPointLeave: Ct,
              onPointClick: Nt,
              highlighted: oe.value === t.uid
            }), () => [
              ae("circle", {
                class: xt({
                  "vue-ui-geo-point": !0,
                  "vue-ui-geo-point-with-event": wt.value && Q.value
                }),
                cx: t.x,
                cy: t.y,
                r: oe.value === t.uid ? t.radius * t.hoverRadiusRatio : t.radius,
                fill: t.fill,
                stroke: At.value.stroke,
                "stroke-width": At.value.strokeWidth,
                "vector-effect": "non-scaling-stroke",
                onMouseenter: (r) => kt(t),
                onMouseleave: (r) => Ct(t),
                onClick: (r) => Nt(t)
              }, null, 42, Vo)
            ], !0),
            n.value.style.chart.points.labels.show ? (b(), A("text", {
              key: 0,
              class: "vue-ui-geo-point-label",
              x: t.x,
              y: t.y + (oe.value === t.uid ? t.radius * t.hoverRadiusRatio : t.radius) + n.value.style.chart.points.labels.offsetY + 1 * n.value.style.chart.points.labels.fontSizeRatio,
              "text-anchor": "middle",
              fill: n.value.style.chart.points.labels.color,
              "font-size": 1 * n.value.style.chart.points.labels.fontSizeRatio
            }, Fo(t.name), 9, Jo)) : j("", !0)
          ]))), 128)),
          g(e.$slots, "svg", {
            svg: {
              drawingArea: dt.value,
              isPrintingImg: y(Ce) | y(we) | le.value,
              isPrintingSvg: se.value,
              data: {
                areaPaths: ot.value,
                linePaths: nt.value,
                geoJsonPoints: rt.value,
                projectedPoints: ht.value
              }
            }
          }, void 0, !0)
        ], 12, qo)
      ], 44, Do)),
      e.$slots.watermark ? (b(), A("div", Ko, [
        g(e.$slots, "watermark", P($({ isPrinting: y(Ce) || y(we) || le.value || se.value })), void 0, !0)
      ])) : j("", !0),
      n.value.style.chart.controls.position === "bottom" && n.value.style.chart.controls.show && !y(be) ? (b(), K(Mt, {
        key: 6,
        ref_key: "zoomControlsBottom",
        ref: We,
        config: n.value,
        scale: y(st),
        isFullscreen: _.value,
        isCursorPointer: Q.value,
        onZoomIn: Te,
        onZoomOut: Ie,
        onResetZoom: mt
      }, null, 8, ["config", "scale", "isFullscreen", "isCursorPointer"])) : j("", !0),
      St(y(Gt), {
        teleportTo: n.value.style.chart.tooltip.teleportTo,
        show: ee.value.showTooltip && xe.value,
        backgroundColor: n.value.style.chart.tooltip.backgroundColor,
        color: n.value.style.chart.tooltip.color,
        fontSize: n.value.style.chart.tooltip.fontSize,
        borderRadius: n.value.style.chart.tooltip.borderRadius,
        borderColor: n.value.style.chart.tooltip.borderColor,
        borderWidth: n.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: n.value.style.chart.tooltip.backgroundOpacity,
        position: n.value.style.chart.tooltip.position,
        offsetY: n.value.style.chart.tooltip.offsetY,
        parent: S.value,
        content: Se.value,
        isCustom: z.value,
        isFullscreen: _.value,
        smooth: n.value.style.chart.tooltip.smooth,
        backdropFilter: n.value.style.chart.tooltip.backdropFilter,
        smoothForce: n.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: n.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": w(() => [
          g(e.$slots, "tooltip-before", P($({ ...ue.value })), void 0, !0)
        ]),
        "tooltip-after": w(() => [
          g(e.$slots, "tooltip-after", P($({ ...ue.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isCustom", "isFullscreen", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      e.$slots.source ? (b(), A("div", {
        key: 7,
        ref_key: "source",
        ref: Ze,
        dir: "auto"
      }, [
        g(e.$slots, "source", {}, void 0, !0)
      ], 512)) : j("", !0),
      g(e.$slots, "skeleton", {}, () => [
        y(be) ? (b(), K(Ro, { key: 0 })) : j("", !0)
      ], !0)
    ], 14, Yo));
  }
}, bn = /* @__PURE__ */ Wo(Qo, [["__scopeId", "data-v-753cf3dd"]]);
export {
  bn as default
};
