// vue-data-ui/dist/vue-ui-mood-radar-A2LBLgUi.js
import { defineAsyncComponent as _, ref as u, computed as v, toRefs as gt, onMounted as pt, watch as Te, openBlock as r, createElementBlock as h, unref as a, normalizeStyle as D, normalizeClass as Oe, createBlock as k, withCtx as i, renderSlot as d, normalizeProps as b, guardReactiveProps as C, createCommentVNode as y, createVNode as K, createElementVNode as n, createSlots as Me, Fragment as _e, renderList as Ne, toDisplayString as j, Teleport as mt, resolveDynamicComponent as kt, mergeProps as bt, createTextVNode as Se, nextTick as re } from "vue";
import { j as Ct, t as Ie, o as wt, q as Le, A as $t, k as De, l as q, X as xt, u as P, v as At, ag as Pt, a0 as Tt, a1 as Ot } from "./lib-GwPnJSrY.js";
import { t as Mt, u as _t } from "./useResponsive-DfdjqQps.js";
import { u as Nt } from "./useConfig-CIgyKZ6R.js";
import { u as St, B as It } from "./BaseScanner-BhhkKVoj.js";
import { u as Lt } from "./usePrinter-DiK5R6k-.js";
import { u as Dt } from "./useSvgExport-ByUukOZt.js";
import { u as ue } from "./useNestedProp-BBT0mbHh.js";
import { u as Ft } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Bt } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Et } from "./useChartAccessibility-9icAAmYg.js";
import Rt from "./img-BYYzY_QU.js";
import Vt from "./Title-BfyrP-jp.js";
import Ut from "./vue_ui_mood_radar-BUs6Fdn8.js";
import jt from "./Legend-1LMMsPxL.js";
import { _ as Wt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const zt = ["id"], Gt = ["id"], Ht = ["xmlns", "viewBox"], Xt = ["width", "height"], Kt = ["id"], qt = ["stop-color"], Yt = ["stop-color"], Jt = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], Qt = ["d", "stroke", "stroke-width"], Zt = ["transform"], eo = ["stroke"], to = ["fill"], oo = ["transform"], lo = ["stroke"], ao = ["fill"], no = ["transform"], so = ["stroke"], ro = ["fill"], uo = ["transform"], io = ["stroke"], co = ["fill"], vo = ["transform"], yo = ["stroke"], fo = ["fill"], ho = ["d", "stroke", "stroke-width", "fill"], go = ["x1", "y1", "x2", "y2", "stroke"], po = ["cx", "cy", "fill", "stroke"], mo = ["cx", "cy", "fill", "stroke"], ko = ["x", "y", "fill", "font-weight"], bo = ["x", "y", "fill"], Co = {
  key: 4,
  class: "vue-data-ui-watermark"
}, wo = ["id"], $o = ["onClick"], xo = ["innerHTML"], Ao = {
  __name: "vue-ui-mood-radar",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["copyAlt"],
  setup(Fe, { expose: Be, emit: Ee }) {
    const N = _(() => import("./BaseIcon-CNx8Tzl5.js")), Re = _(() => import("./vue-ui-accordion-BGHTqIBt.js")), Ve = _(() => import("./DataTable-Bo1pmAqG.js")), Ue = _(() => import("./PenAndPaper-DWV1FMIi.js")), je = _(() => import("./UserOptions-CrLqEimB.js")), We = _(() => import("./PackageVersion-DN9d0ME5.js")), ze = _(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_mood_radar: Ge } = Nt(), { isThemeValid: He, warnInvalidTheme: Xe } = Ft(), w = Fe, Ke = Ee, $ = u(Ct()), x = u(null), g = u(null), ie = u(null), qe = u(0), Ye = u(0), de = u(0), ce = u(null), ve = u(null), F = u(null), W = u(null), ye = u(null), fe = u(!1), B = u(null), Y = u(null), z = u(!1), G = u(!1), he = v(() => !!w.dataset && Object.keys(w.dataset).length), e = u(ee()), S = v(() => e.value.userOptions.useCursorPointer), Je = v(() => Ie({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              grid: {
                stroke: "#6A6A6A"
              },
              outerPolygon: {
                stroke: "#6A6A6A"
              },
              dataPolygon: {
                color: "#6A6A6A",
                opacity: 30,
                stroke: "#6A6A6A"
              },
              smileys: {
                colors: {
                  1: "#DBDBDB",
                  2: "#C4C4C4",
                  3: "#ADADAD",
                  4: "#969696",
                  5: "#808080"
                }
              },
              dataLabel: {
                color: "transparent"
              }
            },
            legend: {
              backgroundColor: "transparent"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: J, FINAL_DATASET: T } = St({
      ...gt(w),
      FINAL_CONFIG: e,
      prepareConfig: ee,
      callback: () => {
        Promise.resolve().then(async () => {
          await re(), O.value.showTable = e.value.table.show;
        });
      },
      skeletonDataset: w.config?.skeletonDataset ?? {
        1: 1,
        2: 1,
        3: 1,
        4: 1,
        5: 1
      },
      skeletonConfig: Ie({
        defaultConfig: e.value,
        userConfig: Je.value
      })
    });
    pt(() => {
      fe.value = !0, ge();
    });
    const Qe = v(() => e.value.debug);
    function ge() {
      if (wt(w.dataset) && Le({
        componentName: "VueUiMoodRadar",
        type: "dataset",
        debug: Qe.value
      }), e.value.responsive) {
        const t = Mt(() => {
          const { width: l, height: o } = _t({
            chart: x.value,
            title: e.value.style.chart.title.text ? ce.value : null,
            legend: e.value.style.chart.legend.show ? ve.value : null,
            noTitle: ie.value,
            source: ye.value
          });
          requestAnimationFrame(() => {
            s.value.width = l, s.value.height = o - 12;
          });
        });
        F.value && (W.value && F.value.unobserve(W.value), F.value.disconnect()), F.value = new ResizeObserver(t), W.value = x.value.parentNode, F.value.observe(W.value);
      }
    }
    const { userOptionsVisible: Q, setUserOptionsVisibility: pe, keepUserOptionState: me } = Bt({ config: e.value }), { svgRef: Z } = Et({ config: e.value.style.chart.title });
    function ee() {
      const t = ue({
        userConfig: w.config,
        defaultConfig: Ge
      }), l = t.theme;
      if (!l) return t;
      if (!He.value(t))
        return Xe(t), t;
      const o = ue({
        userConfig: Ut[l] || w.config,
        defaultConfig: t
      });
      return ue({
        userConfig: w.config,
        defaultConfig: o
      });
    }
    Te(() => w.config, (t) => {
      e.value = ee(), Q.value = !e.value.userOptions.showOnChartHover, ge(), qe.value += 1, Ye.value += 1, de.value += 1, O.value.showTable = e.value.table.show;
    }, { deep: !0 });
    const { isPrinting: te, isImaging: oe, generatePdf: ke, generateImage: be } = Lt({
      elementId: $.value,
      fileName: e.value.style.chart.title.text || "vue-ui-mood-radar",
      options: e.value.userOptions.print
    }), Ze = v(() => e.value.userOptions.show && !e.value.style.chart.title.text), O = u({
      showTable: e.value.table.show
    }), s = u({
      height: 256,
      width: 256
    }), et = {
      5: { x: 128, y: 35 },
      4: { x: 218, y: 98.5 },
      3: { x: 185, y: 204 },
      2: { x: 70, y: 204 },
      1: { x: 38.5, y: 98.5 }
    }, tt = v(() => {
      const t = {};
      return ae.value.forEach((l) => {
        t[l.key] = { x: l.x, y: l.y };
      }), t;
    });
    function E(t) {
      const l = et[t], o = tt.value[t] || l;
      return `translate(${o.x - l.x}, ${o.y - l.y})`;
    }
    const le = v(() => $t({
      plot: { x: s.value.width / 2, y: s.value.height / 2 },
      radius: Math.min(s.value.height, s.value.width) * 0.35,
      sides: 5,
      rotation: 11
    }));
    function ot({ centerX: t, centerY: l, apexX: o, apexY: c, proportion: p, key: m, value: ht }) {
      return {
        x: t + (o - t) * p,
        y: l + (c - l) * p,
        key: m,
        value: ht
      };
    }
    const lt = v(() => Math.max(...Object.values(T.value).map((t) => isNaN(t) ? 0 : t))), M = v(() => Object.values(T.value).reduce((t, l) => (isNaN(t) ? 0 : t) + (isNaN(l) ? 0 : l), 0)), f = v(() => Object.keys(T.value).map((t, l) => {
      const o = typeof T.value[t] != "number" || isNaN(T.value[t]) ? 0 : T.value[t];
      return {
        index: l,
        key: t,
        value: o,
        proportion: o / M.value,
        color: e.value.style.chart.layout.smileys.colors[t]
      };
    }).sort((t, l) => l.key - t.key)), Ce = v(() => f.value.map((t, l) => ({
      ...t,
      display: `${De(
        e.value.style.chart.layout.dataLabel.formatter,
        t.value,
        q({
          p: e.value.style.chart.layout.dataLabel.prefix,
          v: t.value,
          s: e.value.style.chart.layout.dataLabel.suffix,
          r: e.value.style.chart.layout.dataLabel.roundingValue
        })
      )}${J.value ? "" : ` (${q({
        v: t.proportion * 100,
        s: "%",
        r: e.value.style.chart.legend.roundingPercentage
      })})`}`
    }))), ae = v(() => (["1", "2", "3", "4", "5"].forEach((t) => {
      [null, void 0].includes(T.value[t]) && Le({
        componentName: "VueUiMoodRadar",
        type: "datasetAttribute",
        property: t
      });
    }), le.value.coordinates.map((t, l) => {
      const o = ot({
        centerX: s.value.width / 2,
        centerY: s.value.height / 2,
        apexX: t.x,
        apexY: t.y,
        proportion: f.value[l].value / lt.value,
        key: f.value[l].key,
        value: f.value[l].value
      });
      return {
        ...t,
        plots: o,
        key: f.value[l].key
      };
    }))), at = v(() => ({
      cy: "mood-radar-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    }));
    function nt(t) {
      t === g.value ? g.value = null : (g.value = t, I(t));
    }
    function R(t) {
      g.value = t;
      const l = f.value.find((o) => o.key === t);
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: l, seriesIndex: l.index });
    }
    function V(t) {
      g.value = null;
      const l = f.value.find((o) => o.key === t);
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: l.index });
    }
    function I(t) {
      const l = f.value.find((o) => o.key === t);
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: l.index });
    }
    const A = v(() => {
      const t = f.value.map((o) => ({
        name: o.key,
        color: o.color
      })), l = f.value.map((o) => isNaN(o.value) ? 0 : o.value);
      return { head: t, body: l };
    });
    function ne(t = null) {
      re(() => {
        const l = A.value.head.map((p, m) => [[
          p.name
        ], [A.value.body[m]], [isNaN(A.value.body[m] / M.value) ? "-" : A.value.body[m] / M.value * 100]]), o = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[""], ["val"], ["%"]]].concat(l), c = Tt(o);
        t ? t(c) : Ot({ csvContent: c, title: e.value.style.chart.title.text || "vue-ui-mood-radar" });
      });
    }
    const H = v(() => {
      const t = [
        ' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>',
        Number(M.value.toFixed(e.value.table.td.roundingValue)).toLocaleString(),
        "100%"
      ], l = A.value.head.map((p, m) => [
        {
          color: p.color,
          name: p.name
        },
        A.value.body[m].toFixed(e.value.table.td.roundingValue),
        isNaN(A.value.body[m] / M.value) ? "-" : (A.value.body[m] / M.value * 100).toFixed(e.value.table.td.roundingPercentage) + "%"
      ]), o = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, c = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.value,
        e.value.table.columnNames.percentage
      ];
      return {
        head: t,
        body: l,
        config: o,
        colNames: c
      };
    });
    function st() {
      return f.value;
    }
    function we() {
      O.value.showTable = !O.value.showTable;
    }
    const L = u(!1);
    function $e(t) {
      L.value = t;
    }
    const X = u(!1);
    function se() {
      X.value = !X.value;
    }
    async function rt({ scale: t = 2 } = {}) {
      if (!x.value) return;
      const { width: l, height: o } = x.value.getBoundingClientRect(), c = l / o, { imageUri: p, base64: m } = await Rt({ domElement: x.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: p,
        base64: m,
        title: e.value.style.chart.title.text,
        width: l,
        height: o,
        aspectRatio: c
      };
    }
    const U = v(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = O.value.showTable;
      return {
        component: t ? ze : Re,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: L.value,
          fullscreenParent: x.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: S.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    Te(() => O.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && B.value ? B.value.open() : "close" in B.value && B.value.close());
    });
    function xe() {
      O.value.showTable = !1, Y.value && Y.value.setTableIconState(!1);
    }
    const ut = v(() => Ce.value.map((t) => ({
      ...t,
      name: t.display,
      shape: "circle"
    }))), it = v(() => e.value.style.chart.backgroundColor), dt = v(() => e.value.style.chart.legend), ct = v(() => e.value.style.chart.title), { exportSvg: vt, getSvg: yt } = Dt({
      svg: Z,
      title: ct,
      legend: dt,
      legendItems: ut,
      backgroundColor: it
    });
    async function Ae({ isCb: t }) {
      G.value = !0, await re();
      try {
        if (t) {
          const { blob: l, url: o, text: c, dataUrl: p } = await yt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: o, text: c, dataUrl: p }));
        } else
          await Promise.resolve(vt());
      } finally {
        G.value = !1;
      }
    }
    function ft(t) {
      if (t?.stage === "start") {
        z.value = !0;
        return;
      }
      if (t?.stage === "end") {
        z.value = !1;
        return;
      }
      be();
    }
    async function Pe() {
      if (Ke("copyAlt", {
        config: e.value,
        dataset: f.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: f.value
      }));
    }
    return Be({
      getData: st,
      getImage: rt,
      generatePdf: ke,
      generateCsv: ne,
      generateImage: be,
      generateSvg: Ae,
      toggleTable: we,
      toggleAnnotator: se,
      toggleFullscreen: $e,
      copyAlt: Pe
    }), (t, l) => (r(), h("div", {
      class: Oe(`vue-data-ui-component vue-ui-mood-radar ${L.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      ref_key: "moodRadarChart",
      ref: x,
      id: `${$.value}`,
      style: D(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor}`),
      onMouseenter: l[16] || (l[16] = () => a(pe)(!0)),
      onMouseleave: l[17] || (l[17] = () => a(pe)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (r(), k(a(Ue), {
        key: 0,
        svgRef: a(Z),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: X.value,
        isCursorPointer: S.value,
        onClose: se
      }, {
        "annotator-action-close": i(() => [
          d(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": i(({ color: o }) => [
          d(t.$slots, "annotator-action-color", b(C({ color: o })), void 0, !0)
        ]),
        "annotator-action-draw": i(({ mode: o }) => [
          d(t.$slots, "annotator-action-draw", b(C({ mode: o })), void 0, !0)
        ]),
        "annotator-action-undo": i(({ disabled: o }) => [
          d(t.$slots, "annotator-action-undo", b(C({ disabled: o })), void 0, !0)
        ]),
        "annotator-action-redo": i(({ disabled: o }) => [
          d(t.$slots, "annotator-action-redo", b(C({ disabled: o })), void 0, !0)
        ]),
        "annotator-action-delete": i(({ disabled: o }) => [
          d(t.$slots, "annotator-action-delete", b(C({ disabled: o })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : y("", !0),
      Ze.value ? (r(), h("div", {
        key: 1,
        ref_key: "noTitle",
        ref: ie,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : y("", !0),
      e.value.style.chart.title.text ? (r(), h("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: ce,
        style: "width:100%;background:transparent"
      }, [
        K(Vt, {
          config: {
            title: {
              cy: "mood-radar-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "mood-radar-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"])
      ], 512)) : y("", !0),
      n("div", {
        id: `legend-top-${$.value}`
      }, null, 8, Gt),
      e.value.userOptions.show && he.value && (a(me) || a(Q)) ? (r(), k(a(je), {
        key: 3,
        ref_key: "userOptionsRef",
        ref: Y,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: a(te),
        isImaging: a(oe),
        uid: $.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: L.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: x.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: X.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: S.value,
        onToggleFullscreen: $e,
        onGeneratePdf: a(ke),
        onGenerateCsv: ne,
        onGenerateImage: ft,
        onGenerateSvg: Ae,
        onToggleTable: we,
        onToggleAnnotator: se,
        onCopyAlt: Pe,
        style: D({
          visibility: a(me) ? a(Q) ? "visible" : "hidden" : "visible"
        })
      }, Me({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: i(({ isOpen: o, color: c }) => [
            d(t.$slots, "menuIcon", b(C({ isOpen: o, color: c })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: i(() => [
            d(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: i(() => [
            d(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: i(() => [
            d(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: i(() => [
            d(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: i(() => [
            d(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: i(({ toggleFullscreen: o, isFullscreen: c }) => [
            d(t.$slots, "optionFullscreen", b(C({ toggleFullscreen: o, isFullscreen: c })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: i(({ toggleAnnotator: o, isAnnotator: c }) => [
            d(t.$slots, "optionAnnotator", b(C({ toggleAnnotator: o, isAnnotator: c })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: i(({ altCopy: o }) => [
            d(t.$slots, "optionAltCopy", b(C({ altCopy: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : y("", !0),
      (r(), h("svg", {
        ref_key: "svgRef",
        ref: Z,
        xmlns: a(xt),
        viewBox: `0 0 ${s.value.width} ${s.value.height}`,
        class: Oe({ "vue-data-ui-fullscreen--on": L.value, "vue-data-ui-fulscreen--off": !L.value }),
        style: D(`overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        K(a(We)),
        t.$slots["chart-background"] ? (r(), h("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: s.value.width,
          height: s.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          d(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Xt)) : y("", !0),
        n("defs", null, [
          n("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `mood_radar_gradient_${$.value}`
          }, [
            n("stop", {
              offset: "0%",
              "stop-color": a(P)(e.value.style.chart.layout.dataPolygon.color, e.value.style.chart.layout.dataPolygon.opacity)
            }, null, 8, qt),
            n("stop", {
              offset: "100%",
              "stop-color": a(P)(a(At)(e.value.style.chart.layout.dataPolygon.color, e.value.style.chart.layout.dataPolygon.gradient.intensity / 100), e.value.style.chart.layout.dataPolygon.opacity)
            }, null, 8, Yt)
          ], 8, Kt)
        ]),
        (r(!0), h(_e, null, Ne(le.value.coordinates, (o) => (r(), h("line", {
          x1: s.value.width / 2,
          y1: s.value.height / 2,
          x2: o.x,
          y2: o.y,
          stroke: e.value.style.chart.layout.grid.stroke,
          "stroke-width": e.value.style.chart.layout.grid.strokeWidth
        }, null, 8, Jt))), 256)),
        n("path", {
          d: le.value.path,
          fill: "none",
          stroke: e.value.style.chart.layout.outerPolygon.stroke,
          "stroke-width": e.value.style.chart.layout.outerPolygon.strokeWidth,
          "stroke-linejoin": "round",
          "stroke-linecap": "round"
        }, null, 8, Qt),
        n("g", {
          transform: E("5")
        }, [
          n("path", {
            fill: "none",
            stroke: e.value.style.chart.layout.smileys.colors[5],
            "stroke-width": "1",
            "stroke-linecap": "round",
            d: "M119 25A1 1 0 00137 25 1 1 0 00119 25M123 26C124 33 132 33 133 26L123 26M123 22A1 1 0 00126 22 1 1 0 00123 22M130 22A1 1 0 00133 22 1 1 0 00130 22"
          }, null, 8, eo),
          n("circle", {
            class: "vue-ui-mood-radar-trap",
            cx: "128",
            cy: "25",
            r: "20",
            fill: g.value === "5" ? a(P)(e.value.style.chart.layout.smileys.colors[5], 20) : "transparent",
            onMouseenter: l[0] || (l[0] = (o) => R("5")),
            onMouseleave: l[1] || (l[1] = (o) => V("5")),
            onClick: l[2] || (l[2] = (o) => I("5"))
          }, null, 40, to)
        ], 8, Zt),
        n("g", {
          transform: E("4")
        }, [
          n("path", {
            fill: "none",
            stroke: e.value.style.chart.layout.smileys.colors[4],
            "stroke-width": "1",
            "stroke-linecap": "round",
            d: "M218 95A1 1 0 00236 95 1 1 0 00218 95M222 97C225 99 229 99 232 97M222 92A1 1 0 00225 92 1 1 0 00222 92M229 92A1 1 0 00232 92 1 1 0 00229 92"
          }, null, 8, lo),
          n("circle", {
            class: "vue-ui-mood-radar-trap",
            cx: "227",
            cy: "95.5",
            r: "20",
            fill: g.value === "4" ? a(P)(e.value.style.chart.layout.smileys.colors[4], 20) : "transparent",
            onMouseenter: l[3] || (l[3] = (o) => R("4")),
            onMouseleave: l[4] || (l[4] = (o) => V("4")),
            onClick: l[5] || (l[5] = (o) => I("4"))
          }, null, 40, ao)
        ], 8, oo),
        n("g", {
          transform: E("3")
        }, [
          n("path", {
            fill: "none",
            stroke: e.value.style.chart.layout.smileys.colors[3],
            "stroke-width": "1",
            "stroke-linecap": "round",
            d: "M181 213A1 1 0 00199 213 1 1 0 00181 213M185 210A1 1 0 00188 210 1 1 0 00185 210M192 210A1 1 0 00195 210 1 1 0 00192 210M185 215 195 215"
          }, null, 8, so),
          n("circle", {
            class: "vue-ui-mood-radar-trap",
            cx: "190",
            cy: "213.5",
            r: "20",
            fill: g.value === "3" ? a(P)(e.value.style.chart.layout.smileys.colors[3], 20) : "transparent",
            onMouseenter: l[6] || (l[6] = (o) => R("3")),
            onMouseleave: l[7] || (l[7] = (o) => V("3")),
            onClick: l[8] || (l[8] = (o) => I("3"))
          }, null, 40, ro)
        ], 8, no),
        n("g", {
          transform: E("2")
        }, [
          n("path", {
            fill: "none",
            stroke: e.value.style.chart.layout.smileys.colors[2],
            "stroke-width": "1",
            "stroke-linecap": "round",
            d: "M56 213A1 1 0 0074 213 1 1 0 0056 213M60 216C63 214 67 214 70 216M60 210A1 1 0 0063 210 1 1 0 0060 210M67 210A1 1 0 0070 210 1 1 0 0067 210"
          }, null, 8, io),
          n("circle", {
            class: "vue-ui-mood-radar-trap",
            cx: "65",
            cy: "213.5",
            r: "20",
            fill: g.value === "2" ? a(P)(e.value.style.chart.layout.smileys.colors[2], 20) : "transparent",
            onMouseenter: l[9] || (l[9] = (o) => R("2")),
            onMouseleave: l[10] || (l[10] = (o) => V("2")),
            onClick: l[11] || (l[11] = (o) => I("2"))
          }, null, 40, co)
        ], 8, uo),
        n("g", {
          transform: E("1")
        }, [
          n("path", {
            fill: "none",
            stroke: e.value.style.chart.layout.smileys.colors[1],
            "stroke-width": "1",
            "stroke-linecap": "round",
            d: "M20 96A1 1 0 0038 96 1 1 0 0020 96M24 100C25 95 33 95 34 100L24 100M24 93A1 1 0 0027 93 1 1 0 0024 93M31 93A1 1 0 0034 93 1 1 0 0031 93"
          }, null, 8, yo),
          n("circle", {
            class: "vue-ui-mood-radar-trap",
            cx: "29",
            cy: "95.5",
            r: "20",
            fill: g.value === "1" ? a(P)(e.value.style.chart.layout.smileys.colors[1], 20) : "transparent",
            onMouseenter: l[12] || (l[12] = (o) => R("1")),
            onMouseleave: l[13] || (l[13] = (o) => V("1")),
            onClick: l[14] || (l[14] = (o) => I("1"))
          }, null, 40, fo)
        ], 8, vo),
        n("path", {
          d: a(Pt)(ae.value.map((o) => o.plots)),
          stroke: e.value.style.chart.layout.dataPolygon.stroke,
          "stroke-width": e.value.style.chart.layout.dataPolygon.strokeWidth,
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          fill: e.value.style.chart.layout.dataPolygon.gradient.show ? `url(#mood_radar_gradient_${$.value})` : a(P)(e.value.style.chart.layout.dataPolygon.color, e.value.style.chart.layout.dataPolygon.opacity)
        }, null, 8, ho),
        (r(!0), h(_e, null, Ne(ae.value.map((o) => o.plots), (o, c) => (r(), h("g", {
          class: "vue-ui-mood-radar-trap",
          style: D(`opacity:${g.value == o.key ? "1" : "0"}`)
        }, [
          n("line", {
            x1: o.x,
            y1: o.y,
            x2: s.value.width / 2,
            y2: s.value.height / 2,
            stroke: e.value.style.chart.layout.smileys.colors[o.key]
          }, null, 8, go),
          n("circle", {
            cx: o.x,
            cy: o.y,
            fill: e.value.style.chart.layout.smileys.colors[o.key],
            r: "3",
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": 0.5
          }, null, 8, po),
          n("circle", {
            cx: s.value.width / 2,
            cy: s.value.height / 2,
            fill: e.value.style.chart.layout.smileys.colors[o.key],
            r: "3",
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": 0.5
          }, null, 8, mo),
          n("text", {
            x: s.value.width / 2,
            y: ["5", 5].includes(o.key) ? s.value.height / 2 * 1.13 : s.value.height / 2 * 0.9375,
            fill: e.value.style.chart.layout.dataLabel.color,
            "font-size": "12",
            "text-anchor": "middle",
            "font-weight": e.value.style.chart.layout.dataLabel.bold ? "bold" : "normal"
          }, j(a(De)(
            e.value.style.chart.layout.dataLabel.formatter,
            o.value,
            a(q)({
              p: e.value.style.chart.layout.dataLabel.prefix,
              v: o.value,
              s: e.value.style.chart.layout.dataLabel.suffix,
              r: e.value.style.chart.layout.dataLabel.roundingValue
            }),
            { datapoint: o, seriesIndex: c }
          )), 9, ko),
          n("text", {
            x: s.value.width / 2,
            y: ["5", 5].includes(o.key) ? s.value.height / 2 * 1.273 : s.value.height / 2 * 0.7968,
            fill: e.value.style.chart.layout.dataLabel.color,
            "font-size": "12",
            "text-anchor": "middle"
          }, " (" + j(a(q)({
            v: o.value / M.value * 100,
            s: "%",
            r: e.value.style.chart.layout.dataLabel.roundingPercentage
          })) + ") ", 9, bo)
        ], 4))), 256)),
        d(t.$slots, "svg", {
          svg: {
            ...s.value,
            isPrintingImg: a(te) | a(oe) | z.value,
            isPrintingSvg: G.value
          }
        }, void 0, !0)
      ], 14, Ht)),
      t.$slots.watermark ? (r(), h("div", Co, [
        d(t.$slots, "watermark", b(C({ isPrinting: a(te) || a(oe) || z.value || G.value })), void 0, !0)
      ])) : y("", !0),
      n("div", {
        id: `legend-bottom-${$.value}`
      }, null, 8, wo),
      fe.value ? (r(), k(mt, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${$.value}` : `#legend-bottom-${$.value}`
      }, [
        n("div", {
          ref_key: "chartLegend",
          ref: ve
        }, [
          e.value.style.chart.legend.show ? (r(), k(jt, {
            legendSet: Ce.value,
            config: at.value,
            key: `legend_${de.value}`,
            isCursorPointer: S.value,
            style: { display: "flex", "row-gap": "6px" }
          }, {
            item: i(({ legend: o, index: c }) => [
              n("div", {
                onClick: () => nt(o.key),
                style: { display: "flex", "flex-direction": "row", gap: "3px", "align-items": "center", margin: "3px 0" }
              }, [
                o.key == 1 ? (r(), k(a(N), {
                  key: 0,
                  strokeWidth: 1,
                  name: "moodSad",
                  stroke: e.value.style.chart.layout.smileys.colors[o.key]
                }, null, 8, ["stroke"])) : y("", !0),
                o.key == 2 ? (r(), k(a(N), {
                  key: 1,
                  strokeWidth: 1,
                  name: "moodFlat",
                  stroke: e.value.style.chart.layout.smileys.colors[o.key]
                }, null, 8, ["stroke"])) : y("", !0),
                o.key == 3 ? (r(), k(a(N), {
                  key: 2,
                  strokeWidth: 1,
                  name: "moodNeutral",
                  stroke: e.value.style.chart.layout.smileys.colors[o.key]
                }, null, 8, ["stroke"])) : y("", !0),
                o.key == 4 ? (r(), k(a(N), {
                  key: 3,
                  strokeWidth: 1,
                  name: "smiley",
                  stroke: e.value.style.chart.layout.smileys.colors[o.key]
                }, null, 8, ["stroke"])) : y("", !0),
                o.key == 5 ? (r(), k(a(N), {
                  key: 4,
                  strokeWidth: 1,
                  name: "moodHappy",
                  stroke: e.value.style.chart.layout.smileys.colors[o.key]
                }, null, 8, ["stroke"])) : y("", !0),
                a(J) ? y("", !0) : (r(), h("span", {
                  key: 5,
                  style: D({ fontWeight: e.value.style.chart.legend.bold ? "bold" : "normal" })
                }, j(o.display), 5))
              ], 8, $o)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : y("", !0),
          d(t.$slots, "legend", { legend: f.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : y("", !0),
      t.$slots.source ? (r(), h("div", {
        key: 6,
        ref_key: "source",
        ref: ye,
        dir: "auto"
      }, [
        d(t.$slots, "source", {}, void 0, !0)
      ], 512)) : y("", !0),
      he.value && e.value.userOptions.buttons.table ? (r(), k(kt(U.value.component), bt({ key: 7 }, U.value.props, {
        ref_key: "tableUnit",
        ref: B,
        onClose: xe
      }), Me({
        content: i(() => [
          K(a(Ve), {
            colNames: H.value.colNames,
            head: H.value.head,
            body: H.value.body,
            config: H.value.config,
            title: e.value.table.useDialog ? "" : U.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: S.value,
            onClose: xe
          }, {
            th: i(({ th: o }) => [
              n("div", {
                innerHTML: o,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, xo)
            ]),
            td: i(({ td: o }) => [
              Se(j(o.name || o), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"])
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: i(() => [
            Se(j(U.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: i(() => [
            n("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[15] || (l[15] = (o) => ne(e.value.userOptions.callbacks.csv)),
              style: D({ cursor: S.value ? "pointer" : "default" })
            }, [
              K(a(N), {
                name: "fileCsv",
                stroke: U.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : y("", !0),
      d(t.$slots, "skeleton", {}, () => [
        a(J) ? (r(), k(It, { key: 0 })) : y("", !0)
      ], !0)
    ], 46, zt));
  }
}, jo = /* @__PURE__ */ Wt(Ao, [["__scopeId", "data-v-22f6b271"]]);
export {
  jo as default
};
