// vue-data-ui/dist/vue-ui-sparkline-B2RDrdiR.js
import { ref as C, onMounted as Ze, onBeforeUnmount as De, computed as y, watch as z, nextTick as U, openBlock as p, createElementBlock as k, Fragment as be, renderList as Ce, createElementVNode as L, createCommentVNode as A, defineAsyncComponent as ot, toRefs as At, shallowRef as st, watchEffect as Pt, normalizeStyle as ae, renderSlot as ge, normalizeProps as it, guardReactiveProps as ct, toDisplayString as Te, unref as g, createVNode as _t, createBlock as Ie, withCtx as Nt, createTextVNode as Tt } from "vue";
import { j as It, t as dt, aC as We, o as vt, q as ht, r as Rt, v as Re, Y as Bt, ao as Et, c as Oe, b as ft, s as Dt, X as Ft, u as Be, V as qt, G as pt, k as ze, l as je, aD as Gt } from "./lib-GwPnJSrY.js";
import { t as Vt, u as Wt } from "./useResponsive-DfdjqQps.js";
import { u as Ot } from "./useConfig-CIgyKZ6R.js";
import { u as zt, B as jt } from "./BaseScanner-BhhkKVoj.js";
import { u as Ye } from "./useNestedProp-BBT0mbHh.js";
import { u as Yt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Xt } from "./useTimeLabels-C-A2uZKi.js";
import { u as Kt } from "./useChartAccessibility-9icAAmYg.js";
import Ut from "./vue_ui_sparkline-CQ8DJVJx.js";
import { _ as Ht } from "./_plugin-vue_export-helper-CHgC5LLL.js";
function Zt() {
  const e = C(!1);
  return Ze(() => {
    const a = window.matchMedia("(prefers-reduced-motion: reduce)"), l = () => {
      e.value = a.matches;
    };
    l(), a.addEventListener("change", l), De(() => {
      a.removeEventListener("change", l);
    });
  }), e;
}
const Jt = {
  key: 0,
  style: { "pointer-events": "none" }
}, Qt = ["r", "fill", "filter"], ea = ["begin", "dur", "repeatCount", "fill", "calcMode", "keySplines", "keyTimes", "keyPoints"], ta = ["href"], aa = ["dur", "repeatCount", "fill", "values"], la = ["r", "fill"], na = ["begin", "dur", "repeatCount", "calcMode", "keySplines", "keyTimes", "keyPoints"], ra = ["href"], ua = ["dur", "repeatCount"], oa = {
  __name: "SparklinePulse",
  props: {
    uid: { type: String, required: !0 },
    svgRef: { type: Object, default: null },
    pulsePathId: { type: String, required: !0 },
    pulsePathLength: { type: Number, required: !0 },
    pulseDur: { type: String, required: !0 },
    pulseBegin: { type: String, required: !0 },
    pulseRepeatCount: { type: String, required: !0 },
    pulseFillMode: { type: String, default: void 0 },
    pulseKeyPoints: { type: String, required: !0 },
    pulseMotion: { type: Object, required: !0 },
    pulse: { type: Object, required: !0 },
    pulseTrail: { type: Object, required: !0 },
    pulseTrailLength: { type: Number, required: !0 },
    prefersReducedMotion: { type: Boolean, required: !0 },
    loading: { type: Boolean, required: !0 },
    isBar: { type: Boolean, required: !0 }
  },
  setup(e) {
    const a = e, l = y(() => Math.min(a.pulseTrailLength, 45)), o = y(() => !!a.pulse?.show && !a.isBar && !a.prefersReducedMotion && !a.loading && Number(a.pulsePathLength) > 0);
    function r() {
      const i = a.svgRef?.value;
      i && (typeof i.pauseAnimations == "function" && i.pauseAnimations(), typeof i.setCurrentTime == "function" && i.setCurrentTime(0), typeof i.unpauseAnimations == "function" && i.unpauseAnimations());
    }
    function c(i, x) {
      if (typeof i != "string") return i;
      const v = i.trim().match(/^([\d.]+)\s*(ms|s)$/);
      if (!v) return i;
      const m = Number(v[1]), P = v[2];
      return Number.isNaN(m) ? i : `${m + x}${P}`;
    }
    function d(i) {
      return (l.value - i) / l.value * a.pulse.radius;
    }
    function f(i) {
      const x = a.pulse.trail.opacity, v = (l.value - i) / l.value, m = i === 0 ? 1 : v * x;
      return `0;${m};${m};0`;
    }
    return z(
      () => a.loading,
      async (i) => {
        if (i) {
          r();
          return;
        }
        await U(), r();
      }
    ), z(
      () => a.pulsePathId,
      async () => {
        await U(), r();
      }
    ), De(() => {
      r();
    }), (i, x) => o.value ? (p(), k("g", Jt, [
      (p(!0), k(be, null, Ce(l.value, (v, m) => (p(), k(be, null, [
        m % 3 === 0 ? (p(), k("circle", {
          key: `sparkline_dot_${m}_${e.pulsePathId}`,
          r: d(m),
          fill: e.pulse.color,
          filter: `url(#sparkline_pulse_glow_${e.uid})`,
          opacity: "0"
        }, [
          L("animateMotion", {
            begin: c(e.pulseBegin, m * 10),
            dur: e.pulseDur,
            repeatCount: e.pulseRepeatCount,
            fill: e.pulseFillMode,
            calcMode: e.pulseMotion.calcMode,
            keySplines: e.pulseMotion.keySplines || void 0,
            keyTimes: e.pulseMotion.keyTimes || void 0,
            keyPoints: e.pulseKeyPoints,
            rotate: "auto"
          }, [
            L("mpath", {
              href: `#${e.pulsePathId}`
            }, null, 8, ta)
          ], 8, ea),
          L("animate", {
            attributeName: "opacity",
            dur: e.pulseDur,
            repeatCount: e.pulseRepeatCount,
            fill: e.pulseFillMode,
            values: f(m),
            keyTimes: "0;0.1;0.9;1"
          }, null, 8, aa)
        ], 8, Qt)) : A("", !0)
      ], 64))), 256)),
      (p(), k("circle", {
        key: `sparkline_halo_${e.pulsePathId}`,
        r: Math.max(e.pulse.radius * 1.3),
        fill: e.pulse.color,
        opacity: "0"
      }, [
        L("animateMotion", {
          begin: e.pulseBegin,
          dur: e.pulseDur,
          repeatCount: e.pulseRepeatCount,
          calcMode: e.pulseMotion.calcMode,
          keySplines: e.pulseMotion.keySplines || void 0,
          keyTimes: e.pulseMotion.keyTimes || void 0,
          keyPoints: e.pulseKeyPoints,
          rotate: "auto"
        }, [
          L("mpath", {
            href: `#${e.pulsePathId}`
          }, null, 8, ra)
        ], 8, na),
        L("animate", {
          attributeName: "opacity",
          values: "0;0.35;0.35;0",
          keyTimes: "0;0.15;0.85;1",
          dur: e.pulseDur,
          repeatCount: e.pulseRepeatCount
        }, null, 8, ua)
      ], 8, la))
    ])) : A("", !0);
  }
}, Ee = {
  canvas: null,
  ctx: null
};
function sa() {
  if (typeof document > "u")
    throw new Error("color-utils: document is not available (browser-only).");
  if (!Ee.canvas) {
    const e = document.createElement("canvas");
    e.width = 1, e.height = 1;
    const a = e.getContext("2d", { willReadFrequently: !0 });
    if (!a)
      throw new Error("color-utils: unable to get 2D canvas context.");
    Ee.canvas = e, Ee.ctx = a;
  }
  return Ee.ctx;
}
function ia(e) {
  return e < 0 ? 0 : e > 255 ? 255 : Math.round(e);
}
function Qe(e) {
  return e < 0 ? 0 : e > 1 ? 1 : e;
}
function ca(e, a = null) {
  if (typeof e != "string" || e.trim().length === 0)
    throw new Error("colorToRgba: inputColor must be a non-empty string.");
  const l = e.trim();
  if (l.toLowerCase() === "transparent")
    return { red: 0, green: 0, blue: 0, alpha: 0 };
  const o = da(l, a), r = sa();
  r.clearRect(0, 0, 1, 1), r.fillStyle = "#000";
  const c = r.fillStyle;
  if (r.fillStyle = o, r.fillStyle === c && !va(o))
    throw new Error(`colorToRgba: unsupported or invalid color "${e}".`);
  r.fillRect(0, 0, 1, 1);
  const f = r.getImageData(0, 0, 1, 1).data;
  return {
    red: f[0],
    green: f[1],
    blue: f[2],
    alpha: f[3] / 255
  };
}
function gt(e, a = null) {
  if (!Array.isArray(e) || e.length === 0)
    throw new Error("colorsToRgba: colors must be a non-empty array.");
  const l = new Array(e.length);
  for (let o = 0; o < e.length; o += 1)
    l[o] = ca(e[o], a);
  return l;
}
function da(e, a = null) {
  return e !== "currentColor" || typeof window > "u" || !a ? e : window.getComputedStyle(a).color || e;
}
function va(e) {
  const a = e.trim().toLowerCase();
  return a === "black" || a === "#000" || a === "#000000" || a === "rgb(0, 0, 0)" || a === "rgba(0, 0, 0, 1)" || a === "rgba(0, 0, 0, 0)";
}
function ye(e) {
  const a = Qe(e / 255);
  return a <= 0.04045 ? a / 12.92 : Math.pow((a + 0.055) / 1.055, 2.4);
}
function Xe(e) {
  const a = Qe(e);
  let l;
  return a <= 31308e-7 ? l = a * 12.92 : l = 1.055 * Math.pow(a, 1 / 2.4) - 0.055, ia(l * 255);
}
function yt(e, a, l, o) {
  const r = Qe(l);
  if (o === "linearRGB") {
    const c = ye(e.red), d = ye(e.green), f = ye(e.blue), i = ye(a.red), x = ye(a.green), v = ye(a.blue), m = c + (i - c) * r, P = d + (x - d) * r, N = f + (v - f) * r;
    return {
      red: Xe(m),
      green: Xe(P),
      blue: Xe(N),
      alpha: e.alpha + (a.alpha - e.alpha) * r
    };
  }
  return {
    red: e.red + (a.red - e.red) * r,
    green: e.green + (a.green - e.green) * r,
    blue: e.blue + (a.blue - e.blue) * r,
    alpha: e.alpha + (a.alpha - e.alpha) * r
  };
}
function V(e, a, l) {
  let o = e;
  return o < a && (o = a), o > l && (o = l), o;
}
function H(e, a) {
  const l = Math.pow(10, a);
  return Math.round(e * l) / l;
}
function le(e) {
  return Number.isFinite(e) ? e.toString() : "0";
}
function Ke(e) {
  const l = V(e, 0, 255).toString(16);
  return l.length === 1 ? "0" + l : l;
}
function ha(e) {
  if (typeof e != "string") return;
  const a = Number.parseFloat(e);
  return Number.isFinite(a) ? a : void 0;
}
function fa(e, a, l, o = !0) {
  const r = l - a;
  if (r === 0) return 0;
  const c = (e - a) / r;
  return o ? V(c, 0, 1) : c;
}
const mt = "http://www.w3.org/2000/svg";
function pa(e) {
  if (typeof document > "u")
    throw new Error("SvgPathGradientAsync: document is not available (browser-only implementation).");
  const a = document.createElementNS(mt, "svg");
  a.setAttribute("width", "0"), a.setAttribute("height", "0"), a.setAttribute("viewBox", "0 0 0 0"), a.style.position = "absolute", a.style.left = "-10000px", a.style.top = "-10000px", a.style.visibility = "hidden", a.style.pointerEvents = "none";
  const l = document.createElementNS(mt, "path");
  return e === "M " && (e = "M 0,0"), l.setAttribute("d", e), a.appendChild(l), document.body.appendChild(a), { svgElement: a, pathElement: l };
}
function Ue(e) {
  if (e)
    try {
      e.svgElement?.remove?.();
    } catch {
    }
}
function ga(e, a, l) {
  if (typeof l == "number" && l > 0)
    return l;
  let o = a * 0.75;
  o < 0.5 && (o = 0.5), o > 8 && (o = 8);
  const r = e / 2;
  return o > r && r > 0 && (o = r), o <= 0 && (o = 1), o;
}
function ya(e, a, l) {
  if (typeof a == "number" && a >= 1)
    return Math.floor(a);
  const o = typeof l == "number" && l > 0 ? l : 2, r = Math.ceil(e / o);
  return r < 1 ? 1 : r;
}
function ma(e, a, l, o, r, c) {
  const d = l - a;
  let f = Math.ceil(d / o) + 1;
  f < 2 && (f = 2), f > c && (f = c);
  const i = f > 1 ? d / (f - 1) : 0, x = e.getPointAtLength(a);
  let v = `M ${le(H(x.x, r))} ${le(
    H(x.y, r)
  )}`;
  for (let m = 1; m < f; m += 1) {
    const P = a + m * i, N = e.getPointAtLength(P);
    v += ` L ${le(H(N.x, r))} ${le(
      H(N.y, r)
    )}`;
  }
  return v;
}
function ba(e, a, l, o, r, c) {
  const d = l - a;
  let f = Math.ceil(d / o) + 1;
  f < 2 && (f = 2), f > c && (f = c);
  const i = f > 1 ? d / (f - 1) : 0, x = e.getPointAtLength(a);
  let v = x.x, m = x.x, P = x.y, N = x.y, t = `M ${le(H(x.x, r))} ${le(
    H(x.y, r)
  )}`;
  for (let q = 1; q < f; q += 1) {
    const ne = a + q * i, w = e.getPointAtLength(ne);
    w.x < v && (v = w.x), w.x > m && (m = w.x), w.y < P && (P = w.y), w.y > N && (N = w.y), t += ` L ${le(H(w.x, r))} ${le(
      H(w.y, r)
    )}`;
  }
  return {
    pathData: t,
    bounds: { minX: v, maxX: m, minY: P, maxY: N }
  };
}
function ka(e) {
  if (e.length === 1) return [{ position: 0, rgba: e[0] }];
  const a = new Array(e.length), l = e.length - 1;
  for (let o = 0; o < e.length; o += 1)
    a[o] = { position: o / l, rgba: e[o] };
  return a;
}
function wa(e, a, l, o) {
  const r = V(a, 0, 1);
  let c = o;
  for (c < 0 && (c = 0), c > e.length - 2 && (c = e.length - 2); c > 0 && r < e[c].position; ) c -= 1;
  for (; c < e.length - 2 && r > e[c + 1].position; ) c += 1;
  const d = e[c], f = e[c + 1], i = f.position - d.position, x = i > 0 ? (r - d.position) / i : 0, v = xa(d.rgba, f.rgba, x, l);
  return Je(v);
}
function xa(e, a, l, o) {
  const r = V(l, 0, 1);
  if (o === "linearRGB") {
    const c = me(e.red), d = me(e.green), f = me(e.blue), i = me(a.red), x = me(a.green), v = me(a.blue), m = c + (i - c) * r, P = d + (x - d) * r, N = f + (v - f) * r;
    return {
      red: He(m),
      green: He(P),
      blue: He(N),
      alpha: e.alpha + (a.alpha - e.alpha) * r
    };
  }
  return {
    red: e.red + (a.red - e.red) * r,
    green: e.green + (a.green - e.green) * r,
    blue: e.blue + (a.blue - e.blue) * r,
    alpha: e.alpha + (a.alpha - e.alpha) * r
  };
}
function me(e) {
  const a = V(e / 255, 0, 1);
  return a <= 0.04045 ? a / 12.92 : Math.pow((a + 0.055) / 1.055, 2.4);
}
function He(e) {
  const a = V(e, 0, 1);
  let l;
  a <= 31308e-7 ? l = a * 12.92 : l = 1.055 * Math.pow(a, 1 / 2.4) - 0.055;
  const o = Math.round(l * 255);
  return V(o, 0, 255);
}
function Je(e) {
  const a = V(e.alpha, 0, 1), l = V(Math.round(e.red), 0, 255), o = V(Math.round(e.green), 0, 255), r = V(Math.round(e.blue), 0, 255);
  if (a >= 1)
    return `#${Ke(l)}${Ke(o)}${Ke(r)}`;
  const c = H(a, 4).toString();
  return `rgba(${l}, ${o}, ${r}, ${c})`;
}
function Ma(e) {
  const a = e.fill ?? "none";
  let l = `fill="${Z(a)}"`;
  const o = Object.keys(e);
  for (let r = 0; r < o.length; r += 1) {
    const c = o[r];
    if (c === "stroke" || c === "d") continue;
    const d = e[c];
    l += ` ${c}="${Z(d)}"`;
  }
  return l;
}
function bt(e, a, l) {
  return `<path d="${Z(e)}" stroke="${Z(a)}" ${l} />`;
}
function $a(e) {
  const a = [], l = e.segmentAttributeMap.fill ?? "none";
  a.push(`fill="${Z(l)}"`), a.push(`stroke="${Z(e.stroke)}"`);
  for (const o of Object.keys(e.segmentAttributeMap)) {
    if (o === "stroke" || o === "d") continue;
    const r = e.segmentAttributeMap[o];
    a.push(`${o}="${Z(r)}"`);
  }
  return `<path d="${Z(e.pathData)}" ${a.join(" ")} />`;
}
function kt(e, a) {
  const l = [];
  for (const r of Object.keys(a))
    l.push(`${r}="${Z(a[r])}"`);
  return `<g data-svg-path-gradient="true"${l.length ? " " + l.join(" ") : ""}>${e.join("")}</g>`;
}
function Z(e) {
  return String(e).replaceAll("&", "&amp;").replaceAll('"', "&quot;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
}
function Sa(e, a = null, l = {}) {
  return new Promise((o, r) => {
    let c = null;
    try {
      if (!e || typeof e != "string")
        throw new Error("SvgPathGradientAsync: pathData must be a non-empty string.");
      const d = l.temperatureMode ?? null;
      if (d === null && (!Array.isArray(a) || a.length === 0))
        throw new Error("SvgPathGradientAsync: colors must be a non-empty array when temperatureMode is not enabled.");
      if ((l.returnMode ?? "string") !== "string")
        throw new Error('SvgPathGradientAsync: only returnMode="string" is supported for the async version.');
      const i = typeof l.decimalPlaces == "number" ? l.decimalPlaces : 3, x = typeof l.flattenTolerance == "number" && l.flattenTolerance > 0 ? l.flattenTolerance : 0.25, v = l.attrs ? { ...l.attrs } : {}, m = l.groupAttrs ? { ...l.groupAttrs } : {}, P = ha(v["stroke-width"]), N = typeof l.strokeWidth == "number" && l.strokeWidth > 0 ? l.strokeWidth : typeof P == "number" && P > 0 ? P : 1;
      c = pa(e);
      const t = c.pathElement, q = t.getTotalLength(), ne = d === null ? gt(a, l.colorReferenceElement) : [];
      if (d === null && (!(q > 0) || ne.length === 1)) {
        const _ = Je(ne[0]), I = kt(
          [
            $a({
              pathData: e,
              stroke: _,
              segmentAttributeMap: v
            })
          ],
          m
        );
        Ue(c), o(I);
        return;
      }
      const w = l.colorSpace ?? "linearRGB";
      if (d !== null) {
        const _ = l.temperatureColors;
        if (!Array.isArray(_) || _.length !== 2)
          throw new Error("SvgPathGradientAsync: temperatureColors must be a tuple of exactly 2 colors when temperatureMode is enabled.");
      }
      const J = d !== null ? gt(l.temperatureColors, l.colorReferenceElement) : null, re = d === null ? (() => {
        const _ = ka(ne), I = new Array(_.length);
        for (let X = 0; X < _.length; X += 1)
          I[X] = { position: _[X].position, rgba: _[X].rgba };
        return I;
      })() : [], ce = ga(q, N, l.maxSegmentLength), Q = ya(q, l.segments, ce), T = q / Q;
      let de = typeof l.overlap == "number" && l.overlap >= 0 ? l.overlap : N * 0.5;
      const ue = T * 0.45;
      de > ue && (de = ue);
      const Ae = typeof l.samplePointLimitPerSegment == "number" && l.samplePointLimitPerSegment > 10 ? l.samplePointLimitPerSegment : 250, Pe = Ma(v), ve = [], _e = d !== null ? [] : null, oe = d !== null ? [] : null;
      let ee = 0;
      const ke = d === null ? re.length - 2 : 0;
      let Y = 0;
      const we = () => {
        const _ = performance.now(), I = typeof l.frameBudgetMs == "number" ? l.frameBudgetMs : 8;
        for (; Y < Q && performance.now() - _ < I; ) {
          const F = Y * T, G = (Y + 1) * T;
          let R = F, B = G;
          if (Y !== 0 && (R = F - de), Y !== Q - 1 && (B = G + de), R = V(R, 0, q), B = V(B, 0, q), B > R)
            if (d !== null) {
              const E = ba(
                t,
                R,
                B,
                x,
                i,
                Ae
              );
              _e.push(E.pathData), oe.push(E.bounds);
            } else {
              const D = (R + B) * 0.5 / q;
              for (; ee < ke && D > re[ee + 1].position; )
                ee += 1;
              const j = wa(
                re,
                D,
                w,
                ee
              ), W = ma(
                t,
                R,
                B,
                x,
                i,
                Ae
              );
              ve.push(
                bt(W, j, Pe)
              );
            }
          Y += 1;
        }
        if (Y < Q) {
          requestAnimationFrame(we);
          return;
        }
        if (d !== null) {
          let F = Number.POSITIVE_INFINITY, G = Number.NEGATIVE_INFINITY;
          for (let E = 0; E < oe.length; E += 1) {
            const D = oe[E], j = d === "vertical" ? D.minY : D.minX, W = d === "vertical" ? D.maxY : D.maxX;
            j < F && (F = j), W > G && (G = W);
          }
          const R = J[0], B = J[1];
          ve.length = 0;
          for (let E = 0; E < oe.length; E += 1) {
            const D = oe[E], j = d === "vertical" ? D.minY : D.minX, W = d === "vertical" ? D.maxY : D.maxX, xe = (j + W) * 0.5, Me = fa(xe, F, G, !0), se = d === "vertical" ? yt(R, B, Me, w) : yt(B, R, Me, w), $e = Je(se), Ne = _e[E];
            ve.push(
              bt(Ne, $e, Pe)
            );
          }
        }
        const X = kt(ve, m);
        Ue(c), o(X);
      };
      requestAnimationFrame(we);
    } catch (d) {
      Ue(c), r(d);
    }
  });
}
const La = ["innerHTML"], Ca = {
  __name: "SparklineGradientPath",
  props: {
    svgPathData: { type: String, required: !0 },
    enabled: { type: Boolean, required: !0 },
    strokeWidth: { type: Number, required: !0 },
    highColor: { type: String, required: !0 },
    lowColor: { type: String, required: !0 },
    segments: { type: Number, required: !0 }
  },
  setup(e) {
    const a = e, l = C("");
    let o = 0;
    return z(
      () => [
        a.enabled,
        a.svgPathData,
        a.strokeWidth,
        a.highColor,
        a.lowColor,
        a.segments
      ],
      async ([r]) => {
        const c = ++o;
        if (!r) {
          l.value = "";
          return;
        }
        const d = await Sa(a.svgPathData, null, {
          segments: a.segments,
          temperatureMode: "vertical",
          temperatureColors: [a.highColor, a.lowColor],
          attrs: {
            "stroke-width": a.strokeWidth,
            "stroke-linecap": "round",
            "stroke-linejoin": "round"
          },
          groupAttrs: { class: "vue-ui-sparkline-gradient" }
        });
        c === o && (l.value = d);
      },
      { immediate: !0 }
    ), (r, c) => (p(), k("g", { innerHTML: l.value }, null, 8, La));
  }
}, Aa = ["id"], Pa = ["xmlns", "viewBox"], _a = ["width", "height"], Na = ["id"], Ta = ["stop-color"], Ia = ["stop-color"], Ra = ["id"], Ba = ["stop-color"], Ea = ["stop-color"], Da = ["id"], Fa = ["stop-color"], qa = ["stop-color"], Ga = ["id"], Va = ["id"], Wa = ["stop-color", "offset"], Oa = { key: 1 }, za = ["d", "fill"], ja = ["d", "fill"], Ya = ["id", "d", "stroke", "stroke-width"], Xa = ["id", "d", "stroke", "stroke-width"], Ka = ["x", "y", "width", "height", "fill", "rx"], Ua = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Ha = ["x1", "x2", "y1", "y2", "stroke", "stroke-dasharray", "stroke-width"], Za = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], Ja = ["x", "y", "font-size", "font-weight", "fill"], Qa = ["x", "y", "height", "width", "onMouseenter", "onMouseleave", "onClick"], el = {
  __name: "vue-ui-sparkline",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    showInfo: {
      type: Boolean,
      default: !0
    },
    selectedIndex: {
      type: Number,
      default: void 0
    },
    heightRatio: {
      type: Number,
      default: 1
    },
    forcedPadding: {
      type: Number,
      default: 30
    }
  },
  emits: ["hoverIndex", "selectDatapoint"],
  setup(e, { emit: a }) {
    const l = ot(() => import("./PackageVersion-DN9d0ME5.js")), o = ot(() => import("./SparkTooltip-BiseJed3.js")), { vue_ui_sparkline: r } = Ot(), { isThemeValid: c, warnInvalidTheme: d } = Yt(), f = Zt(), i = e, x = y(() => Array.isArray(i.dataset) && i.dataset.length > 0), v = C(It()), m = C(null), P = C(null), N = C(null), t = C(Q());
    function q(n) {
      return i.config?.skeletonDataset && Array.isArray(i.config.skeletonDataset) ? i.config.skeletonDataset.map((u) => ({ period: "-", value: u })) : Gt(n).map((u) => ({ period: "-", value: u }));
    }
    const ne = y(() => dt({
      defaultConfig: {
        gradientPath: { show: !1 },
        temperatureColors: { show: !1 },
        style: {
          backgroundColor: "#99999930",
          scaleMin: 0,
          scaleMax: null,
          animation: { show: !1 },
          line: { color: "#AAAAAA", pulse: { show: !1 } },
          bar: { color: "#AAAAAA" },
          area: { color: "#CACACA" },
          zeroLine: { color: "#6A6A6A" },
          dataLabel: { show: !1 },
          tooltip: { show: !1 }
        }
      },
      userConfig: t.value.skeletonConfig ?? {}
    })), { loading: w, FINAL_DATASET: J, manualLoading: re } = zt({
      ...At(i),
      FINAL_CONFIG: t,
      prepareConfig: Q,
      callback: () => {
        Promise.resolve().then(async () => {
          await U(), xe();
        });
      },
      skeletonDataset: q(12),
      skeletonConfig: dt({
        defaultConfig: t.value,
        userConfig: ne.value
      })
    }), { svgRef: ce } = Kt({ config: t.value.style.title });
    function Q() {
      const n = Ye({
        userConfig: i.config,
        defaultConfig: r
      });
      let u = {};
      const s = n.theme;
      if (s)
        if (!c.value(n))
          d(n), u = n;
        else {
          const h = Ye({
            userConfig: Ut[s] || i.config,
            defaultConfig: n
          });
          u = {
            ...Ye({
              userConfig: i.config,
              defaultConfig: h
            })
          };
        }
      else
        u = n;
      return u;
    }
    const T = y(() => t.value?.style?.line.pulse || {}), de = y(() => `${Math.max(200, Number(T.value.durationMs) || 4e3) / 1e3}s`), ue = C(0), Ae = y(() => T.value?.begin || "0ms"), Pe = C("0;1"), ve = y(() => T.value?.loop === !1 ? "1" : "indefinite"), _e = y(() => T.value?.loop === !1 ? "freeze" : void 0), oe = y(() => T.value.trail.show && we.value.lengthPx || 1), ee = y(() => !!T.value?.show && !K.value && !f.value && !w.value && (M.value?.length || 0) > 1);
    function ke() {
      if (!ee.value) {
        ue.value = 0;
        return;
      }
      const n = ce.value;
      if (!n) return;
      const u = `#${se.value}`, s = n.querySelector?.(u);
      if (s && typeof s.getTotalLength == "function") {
        const h = s.getTotalLength();
        Number.isFinite(h) && h > 0 && (ue.value = h);
        return;
      }
      requestAnimationFrame(() => {
        const h = n.querySelector?.(u);
        if (h && typeof h.getTotalLength == "function") {
          const O = h.getTotalLength();
          Number.isFinite(O) && O > 0 && (ue.value = O);
        }
      });
    }
    const Y = y(() => {
      const n = T.value?.easing || "ease-in-out", u = {
        ease: [0.25, 0.1, 0.25, 1],
        "ease-in": [0.42, 0, 1, 1],
        "ease-out": [0, 0, 0.58, 1],
        "ease-in-out": [0.42, 0, 0.58, 1]
      };
      return n === "linear" ? {
        calcMode: "linear",
        keySplines: null,
        keyTimes: "0;1"
      } : n === "steps" ? {
        calcMode: "discrete",
        keySplines: null,
        keyTimes: "0;1"
      } : {
        calcMode: "spline",
        keySplines: (n === "cubic-bezier" ? Array.isArray(T.value?.cubicBezier) && T.value.cubicBezier.length === 4 ? T.value.cubicBezier : [0.4, 0, 0.2, 1] : u[n] || u["ease-in-out"]).join(" "),
        keyTimes: "0;1"
      };
    }), we = y(() => {
      const n = T.value?.trail || {}, u = t.value.style.line.strokeWidth || 1;
      return {
        show: n.show !== !1,
        lengthPx: n.length,
        width: Math.max(1, Number(n.strokeWidth) || u * 2.2),
        opacity: Math.min(1, Math.max(0, Number(n.opacity) ?? 0.6)),
        fadeIn: 0.5,
        fadeOut: 0.2
      };
    }), _ = y(() => We({
      data: J.value,
      threshold: t.value.downsample.threshold
    }));
    z(() => i.config, (n) => {
      w.value || (t.value = Q()), tt(), b.value.chartWidth = t.value.style.chartWidth;
    }, { deep: !0 }), z(() => i.dataset, (n) => {
      Array.isArray(n) && n.length > 0 && (re.value = !1), I.value = We({
        data: J.value.map((u) => ({
          ...u,
          value: [void 0].includes(u.value) ? null : u.value
        })),
        threshold: t.value.downsample.threshold
      });
    }, { deep: !0 });
    const I = C(X());
    function X() {
      return We({
        data: J.value.map((n) => t.value.style.animation.show && J.value.length > 1 ? {
          ...n,
          value: null
        } : {
          ...n,
          value: [void 0].includes(n.value) ? null : n.value
        }),
        threshold: t.value.downsample.threshold
      });
    }
    const F = st(null), G = st(null), R = C(!1), B = C(0), E = C([]), D = C(""), j = y(() => {
      const u = (_.value || []).map((O) => `${O.period}::${Number.isFinite(O.value) ? O.value : 0}`).join("|"), s = t.value?.style?.animation || {}, h = !!t.value?.gradientPath?.show && !t.value.temperatureColors.show;
      return `${u}#${!!s.show}#${s.animationFrames || 0}#${h}`;
    });
    function W() {
      B.value && (cancelAnimationFrame(B.value), B.value = 0), E.value.forEach((n) => clearTimeout(n)), E.value = [], R.value = !1;
    }
    function xe() {
      const n = t.value?.style?.animation || {}, u = _.value || [], s = j.value, h = !!t.value.gradientPath.show && !t.value.temperatureColors.show;
      if (s && s === D.value && (R.value || I.value.length === u.length))
        return;
      if (W(), h || !n.show || w.value || u.length <= 1) {
        I.value = u, D.value = s;
        return;
      }
      R.value = !0, D.value = s, I.value = [];
      const O = Math.max(1, Number(n.animationFrames) || 1), Lt = Math.max(1, Math.floor(O / u.length));
      let Ve = 0;
      const ut = () => {
        if (s !== j.value) {
          W();
          return;
        }
        if (Ve < u.length) {
          I.value.push(u[Ve]);
          const Ct = setTimeout(() => {
            B.value = requestAnimationFrame(ut);
          }, Lt);
          E.value.push(Ct), Ve += 1;
        } else
          I.value = u, W();
      };
      B.value = requestAnimationFrame(ut);
    }
    z(j, () => {
      xe();
    }), Ze(() => {
      tt(), xe();
    }), De(() => {
      W();
    });
    const Me = C(0);
    z(
      () => w.value,
      async (n) => {
        n || (await U(), Me.value += 1);
      }
    ), z(
      () => R.value,
      async (n) => {
        n || w.value || (await U(), Me.value += 1);
      }
    );
    const se = y(() => `sparkline_line_path_${v.value}`), $e = C(!0);
    async function Ne() {
      $e.value = !1, await U(), $e.value = !0, await U(), ke();
    }
    const et = y(() => !!t.value.debug);
    function tt() {
      if (vt(i.dataset) ? (ht({
        componentName: "VueUiSparkline",
        type: "dataset",
        debug: et.value
      }), re.value = !0) : i.dataset.forEach((n, u) => {
        Rt({
          datasetObject: n,
          requiredAttributes: ["period", "value"]
        }).forEach((s) => {
          ht({
            componentName: "VueUiSparkline",
            type: "datasetSerieAttribute",
            property: s,
            index: u,
            debug: et.value
          });
        });
      }), vt(i.dataset) || (re.value = t.value.loading), t.value.responsive) {
        const n = Vt(() => {
          const { width: u, height: s } = Wt({
            chart: m.value,
            title: t.value.style.title.show && i.showInfo ? P.value : null,
            source: N.value
          });
          requestAnimationFrame(() => {
            b.value.width = u, b.value.height = s, b.value.chartWidth = t.value.style.chartWidth / 500 * u, b.value.padding = i.forcedPadding / 500 * u;
          });
        });
        F.value && (G.value && F.value.unobserve(G.value), F.value.disconnect()), F.value = new ResizeObserver(n), G.value = m.value.parentNode, F.value.observe(G.value);
      }
    }
    De(() => {
      F.value && (G.value && F.value.unobserve(G.value), F.value.disconnect());
    });
    const b = C({
      height: 80 * i.heightRatio,
      width: 500,
      chartWidth: t.value.style.chartWidth,
      padding: i.forcedPadding
    }), Fe = a, $ = y(() => {
      const { top: n, right: u, bottom: s, left: h } = t.value.style.padding;
      return {
        top: n,
        left: h,
        right: b.value.width - u,
        bottom: b.value.height - s,
        start: i.showInfo && t.value.style.dataLabel.show && t.value.style.dataLabel.position === "left" ? b.value.width - b.value.chartWidth + h : b.value.padding + h,
        width: i.showInfo && t.value.style.dataLabel.show ? b.value.chartWidth - h - u : b.value.width - b.value.padding - h - u,
        height: b.value.height - n - s
      };
    }), qe = y(() => [null, void 0].includes(t.value.style.scaleMin) ? Math.min(...I.value.map((n) => isNaN(n.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(n.value) ? 0 : n.value || 0)) : t.value.style.scaleMin), wt = y(() => [null, void 0].includes(t.value.style.scaleMax) ? Math.max(...I.value.map((n) => isNaN(n.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(n.value) ? 0 : n.value || 0)) : t.value.style.scaleMax), Se = y(() => {
      const n = qe.value >= 0 ? 0 : qe.value;
      return Math.abs(n);
    }), at = y(() => wt.value + Se.value), he = y(() => $.value.bottom - $.value.height * Ge(Se.value));
    function Ge(n) {
      return isNaN(n / at.value) ? 0 : n / at.value;
    }
    const fe = y(() => _.value.length - 1 || 1), Le = C([]);
    let lt = 0;
    Pt(() => {
      const n = ++lt;
      (async () => {
        const u = await Xt({
          values: _.value.map((s) => s.period),
          maxDatapoints: _.value.length,
          formatter: t.value.style.dataLabel.datetimeFormatter,
          start: 0,
          end: _.value.length
        });
        n === lt && (Le.value = u);
      })();
    });
    const M = y(() => I.value.map((n, u) => {
      const s = isNaN(n.value) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(n.value) ? 0 : n.value || 0, h = $.value.width / fe.value;
      return {
        absoluteValue: s,
        period: Le.value && Le.value[u] && Le.value[u].text ? Le.value[u].text : n.period,
        plotValue: s + Se.value,
        toMax: Ge(s + Se.value),
        x: $.value.start + u * h,
        y: $.value.bottom - $.value.height * Ge(s + Se.value),
        id: `plot_${v.value}_${u}`,
        color: K.value ? t.value.style.bar.color : t.value.style.area.useGradient ? Re(t.value.style.line.color, 0.05 * (1 - u / fe.value)) : t.value.style.line.color,
        width: h
      };
    })), xt = y(() => {
      const n = { x: M.value[0].x || 0, y: (b.value.height || 0) - 6 }, u = { x: M.value[M.value.length - 1].x || 0, y: (b.value.height || 0) - 6 }, s = [];
      return M.value.forEach((h) => {
        s.push(`${h.x || 0},${h.y || 0} `);
      }), [n.x, n.y, ...s, u.x, u.y].toString();
    }), S = C(void 0), pe = C(void 0);
    function Mt(n, u) {
      t.value.events.datapointEnter && t.value.events.datapointEnter({ datapoint: n, seriesIndex: u }), S.value = n, pe.value || (pe.value = n), Fe("hoverIndex", { index: u });
    }
    function $t(n, u) {
      t.value.events.datapointLeave && t.value.events.datapointLeave({ datapoint: n, seriesIndex: u }), pe.value = S.value, S.value = void 0, Fe("hoverIndex", { index: void 0 });
    }
    const te = y(() => {
      if (x.value) {
        const n = M.value.map((s) => s.absoluteValue), u = n.reduce((s, h) => s + h, 0);
        return {
          latest: M.value[M.value.length - 1] ? M.value[M.value.length - 1].absoluteValue : 0,
          sum: u,
          average: u / M.value.length,
          median: Et(n),
          trend: Bt(M.value.map(({ x: s, y: h, absoluteValue: O }) => ({
            x: s,
            y: h,
            value: O
          }))).trend
        };
      } else
        return {
          latest: null,
          sum: null,
          average: null,
          median: null,
          trend: null
        };
    }), nt = y(() => x.value ? t.value.style.dataLabel.valueType === "latest" ? te.value.latest : t.value.style.dataLabel.valueType === "sum" ? te.value.sum : t.value.style.dataLabel.valueType === "average" ? te.value.average : 0 : 0), K = y(() => t.value.type && t.value.type === "bar");
    function St(n, u) {
      t.value.events.datapointClick && t.value.events.datapointClick({ datapoint: n, seriesIndex: u }), Fe("selectDatapoint", { datapoint: n, index: u });
    }
    const rt = y(() => K.value || !t.value.gradientPath.show || t.value.temperatureColors.show ? "" : `M ${(t.value.style.line.smooth ? Oe(M.value) : ft(M.value)) || "0,0"}`), ie = y(() => !t.value.temperatureColors.show || t.value.temperatureColors.colors.length === 0 ? null : t.value.temperatureColors.colors.map((n) => Dt(n)));
    return z(
      () => [
        ee.value,
        se.value,
        M.value.length,
        b.value.width,
        b.value.height,
        t.value?.style?.line?.smooth
      ],
      async () => {
        await U(), ke();
      },
      { immediate: !0 }
    ), Ze(async () => {
      await U(), ke();
    }), z(
      () => w.value,
      async (n) => {
        n || await Ne();
      }
    ), z(
      () => R.value,
      async (n) => {
        n || w.value || await Ne();
      }
    ), (n, u) => (p(), k("div", {
      ref_key: "sparklineChart",
      ref: m,
      class: "vue-data-ui-component vue-ui-sparkline",
      id: v.value,
      style: ae(`width:100%;font-family:${t.value.style.fontFamily};`)
    }, [
      ge(n.$slots, "before", it(ct({
        selected: S.value,
        latest: te.value.latest,
        sum: te.value.sum,
        average: te.value.average,
        median: te.value.median,
        trend: te.value.trend
      })), void 0, !0),
      t.value.style.title.show && e.showInfo ? (p(), k("div", {
        key: 0,
        ref_key: "chartTitle",
        ref: P,
        class: "vue-ui-sparkline-title",
        style: ae(`display:flex;align-items:center;width:100%;color:${t.value.style.title.color};background:${t.value.style.backgroundColor};justify-content:${t.value.style.title.textAlign === "left" ? "flex-start" : t.value.style.title.textAlign === "right" ? "flex-end" : "center"};height:${t.value.style.title.fontSize * 2}px;font-size:${t.value.style.title.fontSize}px;font-weight:${t.value.style.title.bold ? "bold" : "normal"};`)
      }, [
        L("span", {
          style: ae(`padding:${t.value.style.title.textAlign === "left" ? "0 0 0 12px" : t.value.style.title.textAlign === "right" ? "0 12px 0 0" : "0"}`)
        }, Te(S.value ? S.value.period : t.value.style.title.text), 5)
      ], 4)) : A("", !0),
      (p(), k("svg", {
        ref_key: "svgRef",
        ref: ce,
        xmlns: g(Ft),
        viewBox: `0 0 ${b.value.width} ${b.value.height}`,
        style: ae(`background:${t.value.style.backgroundColor};overflow:visible;direction:ltr`),
        onMouseleave: u[0] || (u[0] = (s) => pe.value = void 0)
      }, [
        _t(g(l)),
        n.$slots["chart-background"] ? (p(), k("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: b.value.width <= 0 ? 10 : b.value.width,
          height: b.value.height <= 0 ? 10 : b.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          ge(n.$slots, "chart-background", {}, void 0, !0)
        ], 8, _a)) : A("", !0),
        L("defs", null, [
          L("linearGradient", {
            x1: "0%",
            y1: "0%",
            x2: "100%",
            y2: "0%",
            id: `sparkline_gradient_${v.value}`
          }, [
            L("stop", {
              offset: "0%",
              "stop-color": g(Be)(g(Re)(t.value.style.area.color, 0.05), t.value.style.area.opacity)
            }, null, 8, Ta),
            L("stop", {
              offset: "100%",
              "stop-color": g(Be)(t.value.style.area.color, t.value.style.area.opacity)
            }, null, 8, Ia)
          ], 8, Na),
          L("linearGradient", {
            x2: "0%",
            y2: "100%",
            id: `sparkline_bar_gradient_pos_${v.value}`
          }, [
            L("stop", {
              offset: "0%",
              "stop-color": t.value.style.bar.color
            }, null, 8, Ba),
            L("stop", {
              offset: "100%",
              "stop-color": g(Re)(t.value.style.bar.color, 0.05)
            }, null, 8, Ea)
          ], 8, Ra),
          L("linearGradient", {
            x2: "0%",
            y2: "100%",
            id: `sparkline_bar_gradient_neg_${v.value}`
          }, [
            L("stop", {
              offset: "0%",
              "stop-color": g(Re)(t.value.style.bar.color, 0.05)
            }, null, 8, Fa),
            L("stop", {
              offset: "100%",
              "stop-color": t.value.style.bar.color
            }, null, 8, qa)
          ], 8, Da),
          L("filter", {
            id: `sparkline_pulse_glow_${v.value}`,
            filterUnits: "userSpaceOnUse",
            x: "-50",
            y: "-50",
            width: "100",
            height: "100"
          }, [...u[1] || (u[1] = [
            L("feGaussianBlur", {
              in: "SourceGraphic",
              stdDeviation: "3",
              result: "blur"
            }, null, -1),
            L("feMerge", null, [
              L("feMergeNode", { in: "blur" }),
              L("feMergeNode", { in: "SourceGraphic" })
            ], -1)
          ])], 8, Ga),
          t.value.temperatureColors.show && ie.value ? (p(), k("linearGradient", {
            key: 0,
            id: `temperature_grad_sparkline_${v.value}`,
            gradientTransform: "rotate(90)"
          }, [
            (p(!0), k(be, null, Ce(ie.value, (s, h) => (p(), k("stop", {
              key: `temperature_grad_stop_${h}_${v.value}`,
              "stop-color": s,
              offset: g(qt)(h, ie.value.length)
            }, null, 8, Wa))), 128))
          ], 8, Va)) : A("", !0)
        ]),
        t.value.style.area.show && !K.value && M.value[0] && M.value.length > 1 ? (p(), k("g", Oa, [
          t.value.style.line.smooth ? (p(), k("path", {
            key: 0,
            class: "vue-ui-sparkline-area",
            d: `M ${M.value[0].x},${$.value.bottom} ${g(Oe)(M.value)} L ${M.value.at(-1).x},${$.value.bottom} Z`,
            fill: t.value.style.area.useGradient ? `url(#sparkline_gradient_${v.value})` : g(Be)(t.value.style.area.color, t.value.style.area.opacity),
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: ae({
              transition: g(w) ? void 0 : "all 0.2s"
            })
          }, null, 12, za)) : (p(), k("path", {
            key: 1,
            class: "vue-ui-sparkline-area",
            d: `M${xt.value}Z`,
            fill: t.value.style.area.useGradient ? `url(#sparkline_gradient_${v.value})` : g(Be)(t.value.style.area.color, t.value.style.area.opacity),
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: ae({
              transition: g(w) ? void 0 : "all 0.2s"
            })
          }, null, 12, ja))
        ])) : A("", !0),
        t.value.style.line.smooth && !K.value ? (p(), k("path", {
          key: 2,
          class: "vue-ui-sparkline-path",
          id: se.value,
          d: `M ${g(Oe)(M.value) || "0,0"}`,
          stroke: ie.value ? `url(#temperature_grad_sparkline_${v.value})` : t.value.style.line.color,
          fill: "none",
          "stroke-width": t.value.style.line.strokeWidth,
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          style: ae({
            transition: g(w) ? void 0 : "all 0.2s"
          })
        }, null, 12, Ya)) : A("", !0),
        !t.value.style.line.smooth && !K.value ? (p(), k("path", {
          key: 3,
          class: "vue-ui-sparkline-path",
          id: se.value,
          d: `M ${g(ft)(M.value) || "0,0"}`,
          stroke: ie.value ? `url(#temperature_grad_sparkline_${v.value})` : t.value.style.line.color,
          fill: "none",
          "stroke-width": t.value.style.line.strokeWidth,
          "stroke-linecap": "round",
          "stroke-linejoin": "round",
          style: ae({
            transition: g(w) ? void 0 : "all 0.2s"
          })
        }, null, 12, Xa)) : A("", !0),
        rt.value && !ie.value ? (p(), Ie(Ca, {
          key: 4,
          svgPathData: rt.value,
          enabled: t.value.gradientPath.show && !K.value && !ie.value,
          strokeWidth: t.value.style.line.strokeWidth,
          highColor: t.value.gradientPath.colors.high,
          lowColor: t.value.gradientPath.colors.low,
          segments: t.value.gradientPath.segments
        }, null, 8, ["svgPathData", "enabled", "strokeWidth", "highColor", "lowColor", "segments"])) : A("", !0),
        $e.value && ee.value ? (p(), Ie(oa, {
          key: 5,
          uid: v.value,
          svgRef: g(ce),
          pulsePathId: se.value,
          pulsePathLength: ue.value,
          pulseDur: de.value,
          pulseBegin: Ae.value,
          pulseRepeatCount: ve.value,
          pulseFillMode: _e.value,
          pulseKeyPoints: Pe.value,
          pulseMotion: Y.value,
          pulse: T.value,
          pulseTrail: we.value,
          pulseTrailLength: oe.value,
          prefersReducedMotion: g(f),
          loading: g(w),
          isBar: K.value
        }, null, 8, ["uid", "svgRef", "pulsePathId", "pulsePathLength", "pulseDur", "pulseBegin", "pulseRepeatCount", "pulseFillMode", "pulseKeyPoints", "pulseMotion", "pulse", "pulseTrail", "pulseTrailLength", "prefersReducedMotion", "loading", "isBar"])) : A("", !0),
        (p(!0), k(be, null, Ce(M.value, (s, h) => (p(), k("g", null, [
          K.value ? (p(), k("rect", {
            key: 0,
            x: s.x - s.width / 2,
            y: isNaN(s.absoluteValue > 0 ? s.y : he.value) ? 0 : s.absoluteValue > 0 ? s.y : he.value,
            width: s.width,
            height: isNaN(Math.abs(s.y - he.value)) ? 0 : Math.abs(s.y - he.value),
            fill: s.absoluteValue > 0 ? `url(#sparkline_bar_gradient_pos_${v.value})` : `url(#sparkline_bar_gradient_neg_${v.value})`,
            rx: t.value.style.bar.borderRadius
          }, null, 8, Ka)) : A("", !0),
          t.value.style.verticalIndicator.show && (S.value && s.id === S.value.id || e.selectedIndex === h) ? (p(), k("line", {
            key: 1,
            x1: s.x,
            x2: s.x,
            y1: $.value.top - 6,
            y2: $.value.bottom,
            stroke: t.value.style.verticalIndicator.color || s.color,
            "stroke-width": t.value.style.verticalIndicator.strokeWidth,
            "stroke-linecap": "round",
            "stroke-dasharray": t.value.style.verticalIndicator.strokeDasharray || 0
          }, null, 8, Ua)) : A("", !0)
        ]))), 256)),
        qe.value < 0 ? (p(), k("line", {
          key: 6,
          x1: $.value.start,
          x2: $.value.start + $.value.width,
          y1: g(pt)(he.value, $.value.bottom),
          y2: g(pt)(he.value, $.value.bottom),
          stroke: t.value.style.zeroLine.color,
          "stroke-dasharray": t.value.style.zeroLine.strokeWidth * 2,
          "stroke-width": t.value.style.zeroLine.strokeWidth,
          "stroke-linecap": "round"
        }, null, 8, Ha)) : A("", !0),
        t.value.style.plot.show ? (p(!0), k(be, { key: 7 }, Ce(M.value, (s, h) => (p(), k("g", null, [
          S.value && s.id === S.value.id || e.selectedIndex === h || g(J).length === 1 ? (p(), k("circle", {
            key: 0,
            cx: s.x,
            cy: s.y,
            r: t.value.style.plot.radius,
            fill: s.color,
            stroke: t.value.style.plot.stroke,
            "stroke-width": t.value.style.plot.strokeWidth
          }, null, 8, Za)) : A("", !0)
        ]))), 256)) : A("", !0),
        e.showInfo && t.value.style.dataLabel.show ? (p(), k("text", {
          key: 8,
          x: t.value.style.dataLabel.position === "left" ? 12 + t.value.style.dataLabel.offsetX : $.value.width + 12 + t.value.style.dataLabel.offsetX,
          y: b.value.height / 2 + t.value.style.dataLabel.fontSize / 2.5 + t.value.style.dataLabel.offsetY,
          "font-size": t.value.style.dataLabel.fontSize,
          "font-weight": t.value.style.dataLabel.bold ? "bold" : "normal",
          fill: t.value.style.dataLabel.color
        }, Te(S.value ? g(ze)(
          t.value.style.dataLabel.formatter,
          S.value.absoluteValue,
          g(je)({ p: t.value.style.dataLabel.prefix, v: S.value.absoluteValue, s: t.value.style.dataLabel.suffix, r: t.value.style.dataLabel.roundingValue }),
          { datapoint: S.value }
        ) : g(ze)(
          t.value.style.dataLabel.formatter,
          nt.value,
          g(je)({ p: t.value.style.dataLabel.prefix, v: nt.value, s: t.value.style.dataLabel.suffix, r: t.value.style.dataLabel.roundingValue })
        )), 9, Ja)) : A("", !0),
        (p(!0), k(be, null, Ce(M.value, (s, h) => (p(), k("rect", {
          x: s.x - ($.value.width / (fe.value + 1) > b.value.padding ? b.value.padding : $.value.width / (fe.value + 1)) / 2,
          y: $.value.top - 6,
          height: $.value.height + 6,
          width: $.value.width / (fe.value + 1) > b.value.padding ? b.value.padding : $.value.width / (fe.value + 1),
          fill: "transparent",
          onMouseenter: () => Mt(s, h),
          onMouseleave: () => $t(s, h),
          onClick: () => St(s, h)
        }, null, 40, Qa))), 256)),
        ge(n.$slots, "svg", { svg: b.value }, void 0, !0)
      ], 44, Pa)),
      S.value && t.value.style.tooltip.show ? (p(), Ie(g(o), {
        key: 1,
        x: S.value.x,
        y: S.value.y,
        prevX: pe.value.x,
        prevY: pe.value.y,
        offsetY: t.value.style.plot.radius * 3 + t.value.style.tooltip.offsetY,
        svgRef: g(ce),
        background: t.value.style.tooltip.backgroundColor,
        color: t.value.style.tooltip.color,
        fontSize: t.value.style.tooltip.fontSize,
        borderWidth: t.value.style.tooltip.borderWidth,
        borderColor: t.value.style.tooltip.borderColor,
        borderRadius: t.value.style.tooltip.borderRadius,
        backgroundOpacity: t.value.style.tooltip.backgroundOpacity
      }, {
        default: Nt(() => [
          ge(n.$slots, "tooltip", it(ct({ ...S.value })), () => [
            Tt(Te(S.value.period) + ": " + Te(g(ze)(
              t.value.style.dataLabel.formatter,
              S.value.absoluteValue,
              g(je)({
                p: t.value.style.dataLabel.prefix,
                v: S.value.absoluteValue,
                s: t.value.style.dataLabel.suffix,
                r: t.value.style.dataLabel.roundingValue
              }),
              { datapoint: S.value }
            )), 1)
          ], !0)
        ]),
        _: 3
      }, 8, ["x", "y", "prevX", "prevY", "offsetY", "svgRef", "background", "color", "fontSize", "borderWidth", "borderColor", "borderRadius", "backgroundOpacity"])) : A("", !0),
      n.$slots.source ? (p(), k("div", {
        key: 2,
        ref_key: "source",
        ref: N,
        dir: "auto"
      }, [
        ge(n.$slots, "source", {}, void 0, !0)
      ], 512)) : A("", !0),
      ge(n.$slots, "skeleton", {}, () => [
        g(w) ? (p(), Ie(jt, { key: 0 })) : A("", !0)
      ], !0)
    ], 12, Aa));
  }
}, vl = /* @__PURE__ */ Ht(el, [["__scopeId", "data-v-051de641"]]);
export {
  vl as default
};
