// vue-data-ui/dist/vue-ui-chord-uQYQgyGM.js
import { useCssVars as ia, defineAsyncComponent as Q, ref as v, shallowRef as ft, computed as m, toRefs as ca, watch as Te, onMounted as gt, onBeforeUnmount as pt, openBlock as r, createElementBlock as i, unref as d, normalizeStyle as E, normalizeClass as U, createBlock as G, withCtx as p, renderSlot as h, normalizeProps as N, guardReactiveProps as R, createCommentVNode as f, createElementVNode as I, createSlots as Oe, withModifiers as bt, createVNode as be, Fragment as V, renderList as Z, mergeProps as mt, toDisplayString as J, Teleport as va, resolveDynamicComponent as da, createTextVNode as yt, nextTick as Be } from "vue";
import { j as Le, t as kt, m as ha, p as me, o as fa, q as Ct, r as ga, n as pa, s as ba, X as ma, W as ya, l as De, ar as ka, at as Ca, a0 as $a, a1 as wa, k as Aa } from "./lib-GwPnJSrY.js";
import { t as Ma, u as Pa } from "./useResponsive-DfdjqQps.js";
import { u as Ia } from "./useConfig-CIgyKZ6R.js";
import { u as xa, B as _a } from "./BaseScanner-BhhkKVoj.js";
import { u as Sa } from "./usePrinter-DiK5R6k-.js";
import { u as Ta } from "./useSvgExport-ByUukOZt.js";
import { u as Ee } from "./useNestedProp-BBT0mbHh.js";
import { u as Oa } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Ba } from "./useUserOptionState-BIvW1Kz7.js";
import { u as La } from "./useChartAccessibility-9icAAmYg.js";
import Da from "./img-BYYzY_QU.js";
import { _ as Ea } from "./Shape-CSyWYgXc.js";
import Na from "./Title-BfyrP-jp.js";
import Ra from "./vue_ui_chord-Bc1_ocwK.js";
import Fa from "./Legend-1LMMsPxL.js";
import { _ as za } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Ua = ["id"], Va = ["id"], Xa = ["xmlns", "viewBox"], ja = ["width", "height"], Ga = { key: 1 }, Ya = ["id", "d"], Wa = { key: 2 }, qa = ["transform"], Ha = ["d", "fill", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Qa = ["d", "fill", "stroke", "stroke-width"], Za = ["d", "fill"], Ja = ["d", "fill", "stroke", "stroke-width", "onMouseenter", "onClick", "onMouseleave"], Ka = ["d", "fill", "stroke", "stroke-width"], eo = ["d", "fill", "stroke", "stroke-width"], to = ["d", "fill", "stroke", "stroke-width"], ao = { key: 0 }, oo = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], lo = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], no = ["cx", "cy", "r", "stroke", "stroke-width", "fill"], so = ["transform", "fill", "text-anchor", "font-size", "font-weight"], ro = { key: 1 }, uo = ["font-size", "font-weight", "fill"], io = ["href"], co = ["transform", "text-anchor", "font-size", "font-weight", "fill", "innerHTML"], vo = {
  key: 4,
  class: "vue-data-ui-watermark"
}, ho = ["id"], fo = ["onClick"], go = {
  key: 7,
  "data-dom-to-png-ignore": "",
  class: "reset-wrapper"
}, po = { style: { "text-align": "right", width: "100%" } }, bo = {
  __name: "vue-ui-chord",
  props: {
    dataset: {
      type: Object,
      default() {
        return {};
      }
    },
    config: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["selectLegend", "selectGroup", "selectRibbon", "copyAlt"],
  setup($t, { expose: wt, emit: At }) {
    ia((e) => ({
      v78a6832a: e.slicerColor
    }));
    const Mt = Q(() => import("./vue-ui-accordion-BGHTqIBt.js")), Ne = Q(() => import("./BaseIcon-CNx8Tzl5.js")), Pt = Q(() => import("./DataTable-Bo1pmAqG.js")), It = Q(() => import("./PackageVersion-DN9d0ME5.js")), xt = Q(() => import("./PenAndPaper-DWV1FMIi.js")), _t = Q(() => import("./UserOptions-CrLqEimB.js")), St = Q(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_chord: Tt } = Ia(), { isThemeValid: Ot, warnInvalidTheme: Bt } = Oa(), x = $t, ae = At, F = v(!!x.dataset && Object.hasOwn(x.dataset, "matrix")), w = v(Le()), A = v(null), C = v(null), b = v(null), Re = v(0), B = v(null), Fe = v(null), ze = v(null), Ue = v(null), Ve = v(null), ye = v(0), ke = v(0), Ce = v(0), K = v(!1), X = ft(null), ee = ft(null), Xe = v(!1), oe = v(null), $e = v(null), ie = v(!1), ce = v(!1), t = v(Pe()), Y = m(() => t.value.userOptions.useCursorPointer), Lt = m(() => kt({
      defaultConfig: {
        useCssAnimation: !1,
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            legend: {
              backgroundColor: "transparent"
            },
            arcs: {
              stroke: "#6A6A6A",
              labels: {
                show: !1
              }
            },
            ribbons: {
              stroke: "#6A6A6A",
              underlayerOpacity: 0,
              labels: {
                show: !1
              }
            }
          }
        }
      },
      userConfig: t.value.skeletonConfig ?? {}
    })), { loading: je, FINAL_DATASET: M, manualLoading: le } = xa({
      ...ca(x),
      FINAL_CONFIG: t,
      prepareConfig: Pe,
      callback: () => {
        Promise.resolve().then(async () => {
          await Be(), W.value.showTable = t.value.table.show;
        });
      },
      skeletonDataset: x.config?.skeletonDataset ?? {
        matrix: [
          [12e3, 6e3, 9e3, 3e3],
          [2e3, 1e4, 2e3, 6001],
          [8e3, 1600, 8e3, 8001],
          [1e3, 1e3, 1e3, 7001]
        ],
        labels: [],
        colors: ["#DBDBDB", "#C4C4C4", "#ADADAD", "#969696"]
      },
      skeletonConfig: kt({
        defaultConfig: t.value,
        userConfig: Lt.value
      })
    }), { userOptionsVisible: we, setUserOptionsVisibility: Ge, keepUserOptionState: Ye } = Ba({ config: t.value }), { svgRef: te } = La({ config: t.value.style.chart.title }), { isPrinting: Ae, isImaging: Me, generatePdf: We, generateImage: qe } = Sa({
      elementId: `chord_${w.value}`,
      fileName: t.value.style.chart.title.text || "vue-ui-chord",
      options: t.value.userOptions.print
    }), W = v({
      showTable: t.value.table.show
    });
    function Pe() {
      const e = Ee({
        userConfig: x.config,
        defaultConfig: Tt
      }), o = e.theme;
      if (!o) return e;
      if (!Ot.value(e))
        return Bt(e), e;
      const a = Ee({
        userConfig: Ra[o] || x.config,
        defaultConfig: e
      }), l = Ee({
        userConfig: x.config,
        defaultConfig: a
      });
      return {
        ...l,
        customPalette: l.customPalette.length ? l.customPalette : ha[o] || me
      };
    }
    Te(() => x.config, (e) => {
      je.value || (t.value = Pe()), we.value = !t.value.userOptions.showOnChartHover, Ie(), ye.value += 1, ke.value += 1, Ce.value += 1, _.value = t.value.initialRotation, W.value.showTable = t.value.table.show;
    }, { deep: !0 });
    const He = m(() => t.value.debug);
    Te(() => x.dataset, () => {
      Qe(), Ie(), ye.value += 1, ke.value += 1, Ce.value += 1;
    });
    function Qe() {
      if (fa(x.dataset)) {
        Ct({
          componentName: "VueUiChord",
          type: "dataset",
          debug: He.value
        }), F.value = !1, le.value = !0;
        return;
      }
      const e = ga({
        datasetObject: x.dataset,
        requiredAttributes: ["matrix"]
      });
      if (e.length) {
        e.forEach((s) => {
          Ct({
            componentName: "VueUiChord",
            type: "datasetAttribute",
            property: s,
            debug: He.value
          });
        }), F.value = !1, le.value = !0;
        return;
      }
      const o = x.dataset.matrix;
      if (!Array.isArray(o) || o.length < 2) {
        console.warn(`VueUiChord: dataset.matrix requires a minimum of 2 datapoints, for example:
        
matrix:[
    [1, 1],
    [1, 1]
]`), F.value = !1, le.value = !0;
        return;
      }
      const a = o.length, l = o.findIndex((s) => !Array.isArray(s) || s.length !== a);
      if (l !== -1) {
        console.warn(
          `VueUiChord - Invalid matrix: dataset.matrix at index ${l} has ${Array.isArray(o[l]) ? o[l].length : "NaN"} elements instead of the required ${a}

dataset.matrix[${l}] = [${Array.isArray(o[l]) ? o[l].toString() : "invalid"}]`
        ), F.value = !1, le.value = !0;
        return;
      }
      F.value = !0, le.value = !1;
    }
    function Ie() {
      if (Qe(), t.value.responsive) {
        const e = Ma(() => {
          const { width: o, height: a, heightNoTitle: l, heightSource: s, heightTitle: n, heightLegend: u } = Pa({
            chart: B.value,
            title: t.value.style.chart.title.text ? Fe.value : null,
            legend: t.value.style.chart.legend.show ? ze.value : null,
            source: Ue.value,
            noTitle: Ve.value
          }), g = n + u + s + l;
          o < a ? B.value.style.width = "100%" : (B.value.style.height = "100%", te.value.style.height = `calc(100% - ${g}px)`);
        }, 100);
        X.value && (ee.value && X.value.unobserve(ee.value), X.value.disconnect()), X.value = new ResizeObserver(e), ee.value = B.value.parentNode, X.value.observe(ee.value), e();
      }
      Ze.value = setTimeout(() => {
        K.value = !0;
      }, 500);
    }
    function Dt() {
      if (!M.value || !Object.hasOwn(M.value, "matrix") || M.value.matrix.length < 2) {
        console.warn(`VueUiChord: dataset.matrix requires a minimum of 2 datapoints, for example:

matrix:[
  [1, 1],
  [1, 1]
]`), F.value = !1;
        return;
      }
      M.value.matrix.forEach((e, o) => {
        e.length !== M.value.matrix.length && (console.warn(`VueUiChord - Invalid matrix: dataset.matrix at index ${o} has ${e.length} elements instead of the required ${M.value.matrix.length}

dataset.matrix[${o}] = [${e.toString()}] has a length of ${e.length} but should have the same length as the matrix itself (${M.value.matrix.length})`), F.value = !1);
      });
    }
    const Ze = v(null);
    gt(() => {
      Xe.value = !0, Ie();
    });
    const L = v({
      height: 600,
      width: 600
    }), Et = m(() => t.value.userOptions.show && !t.value.style.chart.title.text), Nt = m(() => pa(t.value.customPalette)), y = m(() => ({
      inner: L.value.width * 0.3 * t.value.style.chart.arcs.innerRadiusRatio,
      outer: L.value.width * 0.34 * t.value.style.chart.arcs.outerRadiusRatio
    })), Rt = m(() => t.value.style.chart.arcs.padAngle / 100), k = m(() => ({
      matrix: M.value.matrix ?? [[0]],
      labels: M.value.labels ?? [""],
      colors: M.value.colors && Array.isArray(M.value.colors) && M.value.colors.length ? M.value.colors.map((e) => ba(e)) : M.value.matrix.map((e, o) => Nt.value[o] || me[o] || me[o % me.length])
    }));
    function Ft(e, o) {
      const a = e.length, l = y.value.inner, s = new Array(a).fill(0);
      let n = 0;
      for (let c = 0; c < a; c += 1)
        for (let $ = 0; $ < a; $ += 1)
          s[c] += e[c][$], n += e[c][$];
      const u = (2 * Math.PI - o * a) / n, g = [];
      let S = 0;
      for (let c = 0; c < a; c += 1) {
        const $ = S, O = $ + s[c] * u;
        g.push({
          index: c,
          pattern: `pattern_${w.value}_${c}`,
          startAngle: $,
          endAngle: O,
          name: k.value.labels[c],
          id: Le(),
          color: k.value.colors[c],
          proportion: s[c] / n
        }), S = O + o;
      }
      const T = [];
      for (let c = 0; c < a; c += 1) {
        const $ = e[c].map((D, j) => ({ j, v: D }));
        $.sort((D, j) => j.v - D.v);
        let O = g[c].startAngle;
        for (const { j: D, v: j } of $) {
          const H = O, ue = H + j * u;
          T.push({
            index: c,
            subIndex: D,
            pattern: `pattern_${w.value}_${c}`,
            startAngle: H,
            endAngle: ue,
            value: j,
            groupName: k.value.labels[c],
            groupId: g[c].id,
            groupColor: k.value.colors[c],
            midAngle: (H + ue) / 2,
            midBaseX: Math.cos((H + ue) / 2 - Math.PI / 2) * l,
            midBaseY: Math.sin((H + ue) / 2 - Math.PI / 2) * l
          }), O = ue;
        }
      }
      const z = [];
      for (const c of T) {
        const $ = T.find(
          (O) => O.index === c.subIndex && O.subIndex === c.index
        );
        z.push({ source: c, target: $, id: Le() });
      }
      return { groups: g, chords: z };
    }
    const P = m(() => {
      const e = Ft(k.value.matrix, Rt.value);
      return Dt(), e.chords.sort(
        (o, a) => Math.max(a.source.value, a.target.value) - Math.max(o.source.value, o.target.value)
      ), e;
    }), zt = m(() => {
      const e = y.value.outer + t.value.style.chart.ribbons.labels.offset + 12, a = t.value.style.chart.ribbons.labels.fontSize * 0.6, l = [];
      if (C.value) {
        const n = C.value;
        if (n.source.value) {
          const u = String(n.source.value);
          l.push({
            id: n.id + "-src",
            theta: n.source.midAngle,
            w: u.length * a / e,
            midBaseX: n.source.midBaseX,
            midBaseY: n.source.midBaseY,
            groupColor: n.source.groupColor,
            value: n.source.value
          });
        }
        if (n.target && n.target.value && n.target.value !== n.source.value) {
          const u = String(n.target.value);
          l.push({
            id: n.id + "-tgt",
            theta: n.target.midAngle,
            w: u.length * a / e,
            midBaseX: n.target.midBaseX,
            midBaseY: n.target.midBaseY,
            groupColor: n.target.groupColor,
            value: n.target.value
          });
        }
        return l;
      }
      function s(n) {
        P.value.chords.filter((u) => u.source.groupId === n && u.source.value).forEach((u) => {
          {
            const g = String(u.source.value);
            l.push({
              id: u.id + "-src",
              theta: u.source.midAngle,
              w: g.length * a / e,
              midBaseX: u.source.midBaseX,
              midBaseY: u.source.midBaseY,
              groupColor: u.source.groupColor,
              value: u.source.value
            });
          }
          if (u.target && u.target.value && u.target.value !== u.source.value) {
            const g = String(u.target.value);
            l.push({
              id: u.id + "-tgt",
              theta: u.target.midAngle,
              w: g.length * a / e,
              midBaseX: u.target.midBaseX,
              midBaseY: u.target.midBaseY,
              groupColor: u.target.groupColor,
              value: u.target.value
            });
          }
        });
      }
      return b.value && s(b.value), A.value && s(A.value.id), l;
    }), Ut = m(() => {
      const e = zt.value.map((s) => ({ ...s })).sort((s, n) => s.theta - n.theta), o = t.value.style.chart.ribbons.labels.minSeparationDeg * Math.PI / 180;
      let a = !0, l = 0;
      for (; a && l++ < 10; ) {
        a = !1;
        for (let g = 1; g < e.length; g += 1) {
          const S = e[g - 1], T = e[g], z = S.theta + S.w + o;
          T.theta < z && (T.theta = z, a = !0);
        }
        const s = e[0], n = e[e.length - 1], u = n.theta + n.w + o - 2 * Math.PI;
        s.theta < u && (s.theta = u, a = !0);
      }
      return e;
    });
    function Je(e, o, a, l) {
      const s = e - Math.PI / 2, n = o - Math.PI / 2, u = Math.cos(s) * a, g = Math.sin(s) * a, S = Math.cos(n) * a, T = Math.sin(n) * a, z = Math.cos(n) * l, c = Math.sin(n) * l, $ = Math.cos(s) * l, O = Math.sin(s) * l, D = o - e > Math.PI ? 1 : 0;
      return `M${u},${g} A${a},${a} 0 ${D} 1 ${S},${T} L${z},${c} A${l},${l} 0 ${D} 0 ${$},${O} Z`;
    }
    function ve(e, o) {
      const a = y.value.inner, l = e.startAngle - Math.PI / 2, s = e.endAngle - Math.PI / 2, n = o.startAngle - Math.PI / 2, u = o.endAngle - Math.PI / 2, g = Math.cos(l) * a, S = Math.sin(l) * a, T = Math.cos(s) * a, z = Math.sin(s) * a, c = Math.cos(n) * a, $ = Math.sin(n) * a, O = Math.cos(u) * a, D = Math.sin(u) * a, j = e.endAngle - e.startAngle > Math.PI ? 1 : 0, H = o.endAngle - o.startAngle > Math.PI ? 1 : 0;
      return `M${g},${S}A${a},${a} 0 ${j} 1 ${T},${z}Q0,0 ${c},${$}A${a},${a} 0 ${H} 1 ${O},${D}Q0,0 ${g},${S}Z`;
    }
    const Ke = m(() => _.value * Math.PI / 180);
    function Vt(e) {
      return (e.startAngle + e.endAngle) / 2;
    }
    function Xt(e) {
      return ((Vt(e) + Ke.value) % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI);
    }
    function jt(e) {
      let o = e + Ke.value;
      return o = (o % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI), o > Math.PI ? "end" : "start";
    }
    function Gt(e) {
      const o = e - Math.PI / 2, a = Math.cos(o) * (y.value.outer + t.value.style.chart.arcs.labels.offset + 24), l = Math.sin(o) * (y.value.outer + t.value.style.chart.arcs.labels.offset + 24);
      return `translate(${a},${l})`;
    }
    function Yt(e, o, a) {
      const l = e - Math.PI / 2, s = o - Math.PI / 2, n = Math.cos(l) * a, u = Math.sin(l) * a, g = Math.cos(s) * a, S = Math.sin(s) * a, T = o - e > Math.PI ? 1 : 0;
      return `M${n},${u} A${a},${a} 0 ${T} 1 ${g},${S}`;
    }
    const _ = v(t.value.initialRotation), ne = v(!1);
    let et = 0, tt = 0;
    function at(e) {
      const o = te.value.getBoundingClientRect(), a = o.left + o.width / 2, l = o.top + o.height / 2, s = e.clientX ?? e.touches[0].clientX, n = e.clientY ?? e.touches[0].clientY;
      return Math.atan2(n - l, s - a);
    }
    function ot(e) {
      !t.value.enableRotation || se.value || (e.preventDefault(), ne.value = !0, et = _.value, tt = at(e));
    }
    function de(e) {
      if (!ne.value) return;
      const o = at(e);
      _.value = et + (o - tt) * 180 / Math.PI;
    }
    function he() {
      ne.value = !1;
    }
    function lt(e) {
      return Aa(
        t.value.style.chart.ribbons.labels.formatter,
        e,
        De({
          p: t.value.style.chart.ribbons.labels.prefix,
          v: e,
          s: t.value.style.chart.ribbons.labels.suffix,
          r: t.value.style.chart.ribbons.labels.rounding
        })
      );
    }
    gt(() => {
      window.addEventListener("mousemove", de), window.addEventListener("mouseup", he), window.addEventListener("touchmove", de, { passive: !1 }), window.addEventListener("touchend", he);
    }), pt(() => {
      window.removeEventListener("mousemove", de), window.removeEventListener("mouseup", he), window.removeEventListener("touchmove", de), window.removeEventListener("touchend", he), clearTimeout(Ze.value), X.value && (ee.value && X.value.unobserve(ee.value), X.value.disconnect());
    });
    const fe = v(!1);
    function nt(e) {
      fe.value = e, Re.value += 1;
    }
    function Wt(e, o) {
      t.value.events.datapointEnter && t.value.events.datapointEnter({ datapoint: e, seriesIndex: o }), !b.value && (A.value = e);
    }
    function qt(e, o) {
      A.value = null, t.value.events.datapointLeave && t.value.events.datapointLeave({ datapoint: e, seriesIndex: o });
    }
    function Ht(e, o) {
      ae("selectGroup", e), t.value.events.datapointClick && t.value.events.datapointClick({ datapoint: e, seriesIndex: o });
    }
    function Qt(e, o) {
      t.value.events.datapointEnter && t.value.events.datapointEnter({ datapoint: e, seriesIndex: o }), !b.value && (C.value = e);
    }
    function Zt(e, o) {
      C.value = null, t.value.events.datapointLeave && t.value.events.datapointLeave({ datapoint: e, seriesIndex: o });
    }
    function Jt(e, o) {
      ae("selectRibbon", e), t.value.events.datapointClick && t.value.events.datapointClick({ datapoint: e, seriesIndex: o });
    }
    function st(e) {
      return !b.value && !A.value && !C.value ? 0.8 : (b.value ? b.value === e.source.groupId : A.value ? A.value.id === e.source.groupId : C.value?.id === e.id) ? 1 : 0.1;
    }
    function rt(e) {
      return !b.value && !A.value && !C.value || (b.value ? b.value === e.id : A.value ? A.value.id === e.id : [C.value?.source.groupId, C.value?.target.groupId].includes(e.id)) ? 1 : 0.3;
    }
    const se = v(!1);
    function xe() {
      se.value = !se.value;
    }
    function ut() {
      W.value.showTable = !W.value.showTable;
    }
    function Kt() {
      return P.value;
    }
    const q = v(null);
    function it() {
      const e = t.value.initialRotation;
      q.value !== null && cancelAnimationFrame(q.value);
      const o = () => {
        _.value += (e - _.value) * 0.05, Math.abs(_.value - e) < 0.1 ? (_.value = e, q.value = null) : q.value = requestAnimationFrame(o);
      };
      q.value = requestAnimationFrame(o);
    }
    pt(() => {
      q.value !== null && cancelAnimationFrame(q.value);
    });
    function ct(e) {
      e === b.value ? (b.value = null, ae("selectLegend", null)) : (b.value = e, ae("selectLegend", P.value.groups.find((o) => o.id === e)));
    }
    const _e = m(() => P.value.groups.map((e, o) => ({
      name: e.name,
      color: e.color,
      shape: "circle",
      patternIndex: o,
      pattern: `pattern_${w.value}_${o}`,
      id: e.id,
      select: () => ct(e.id),
      opacity: b.value ? b.value === e.id ? 1 : 0.3 : 1
    }))), ea = m(() => ({
      cy: "chord-div-legend",
      backgroundColor: t.value.style.chart.legend.backgroundColor,
      color: t.value.style.chart.legend.color,
      fontSize: t.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: t.value.style.chart.legend.bold ? "bold" : ""
    })), ge = m(() => {
      const e = P.value.groups.map((a, l) => ({
        name: a.name || l,
        color: a.color
      })), o = k.value.matrix;
      return { head: e, body: o };
    }), pe = m(() => {
      const e = [{ name: "", color: null }, ...ge.value.head], o = ge.value.body.map((s, n) => [
        ge.value.head[n],
        ...ge.value.body[n]
      ]), a = {
        th: {
          backgroundColor: t.value.table.th.backgroundColor,
          color: t.value.table.th.color,
          outline: t.value.table.th.outline
        },
        td: {
          backgroundColor: t.value.table.td.backgroundColor,
          color: t.value.table.td.color,
          outline: t.value.table.td.outline
        },
        breakpoint: t.value.table.responsiveBreakpoint
      };
      return {
        colNames: e,
        head: e,
        body: o,
        config: a
      };
    });
    function Se(e = null) {
      Be(() => {
        const o = k.value.matrix.map((s, n) => [
          [k.value.labels[n] || n],
          s
        ]), a = [
          [t.value.style.chart.title.text],
          [t.value.style.chart.title.subtitle.text],
          [
            [""],
            ...k.value.labels.map((s, n) => [s || n])
          ]
        ].concat(o), l = $a(a);
        e ? e(l) : wa({ csvContent: l, title: t.value.style.chart.title.text || "vue-ui-chord" });
      });
    }
    async function ta({ scale: e = 2 } = {}) {
      if (!B.value) return;
      const { width: o, height: a } = B.value.getBoundingClientRect(), l = o / a, { imageUri: s, base64: n } = await Da({ domElement: B.value, base64: !0, img: !0, scale: e });
      return {
        imageUri: s,
        base64: n,
        title: t.value.style.chart.title.text,
        width: o,
        height: a,
        aspectRatio: l
      };
    }
    function aa(e, o) {
      return `${k.value.labels[o]}${t.value.style.chart.arcs.labels.showPercentage ? De({
        p: " (",
        v: isNaN(e.proportion) ? 0 : e.proportion * 100,
        s: "%)",
        r: t.value.style.chart.arcs.labels.roundingPercentage
      }) : ""}`;
    }
    const re = m(() => {
      const e = t.value.table.useDialog && !t.value.table.show, o = W.value.showTable;
      return {
        component: e ? St : Mt,
        title: `${t.value.style.chart.title.text}${t.value.style.chart.title.subtitle.text ? `: ${t.value.style.chart.title.subtitle.text}` : ""}`,
        props: e ? {
          backgroundColor: t.value.table.th.backgroundColor,
          color: t.value.table.th.color,
          headerColor: t.value.table.th.color,
          headerBg: t.value.table.th.backgroundColor,
          isFullscreen: fe.value,
          fullscreenParent: B.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: Y.value
        } : {
          hideDetails: !0,
          config: {
            open: o,
            maxHeight: 1e4,
            body: {
              backgroundColor: t.value.style.chart.backgroundColor,
              color: t.value.style.chart.color
            },
            head: {
              backgroundColor: t.value.style.chart.backgroundColor,
              color: t.value.style.chart.color
            }
          }
        }
      };
    });
    Te(() => W.value.showTable, (e) => {
      t.value.table.show || (e && t.value.table.useDialog && oe.value ? oe.value.open() : "close" in oe.value && oe.value.close());
    });
    function vt() {
      W.value.showTable = !1, $e.value && $e.value.setTableIconState(!1);
    }
    const oa = m(() => t.value.style.chart.backgroundColor), la = m(() => t.value.style.chart.legend), na = m(() => t.value.style.chart.title), { exportSvg: sa, getSvg: ra } = Ta({
      svg: te,
      title: na,
      legend: la,
      legendItems: _e,
      backgroundColor: oa
    });
    async function dt({ isCb: e }) {
      ce.value = !0, await Be();
      try {
        if (e) {
          const { blob: o, url: a, text: l, dataUrl: s } = await ra();
          await Promise.resolve(t.value.userOptions.callbacks.svg({ blob: o, url: a, text: l, dataUrl: s }));
        } else
          await Promise.resolve(sa());
      } finally {
        ce.value = !1;
      }
    }
    function ua(e) {
      if (e?.stage === "start") {
        ie.value = !0;
        return;
      }
      if (e?.stage === "end") {
        ie.value = !1;
        return;
      }
      qe();
    }
    async function ht() {
      if (ae("copyAlt", {
        config: t.value,
        dataset: k.value
      }), !t.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(t.value.userOptions.callbacks.altCopy({
        config: t.value,
        dataset: k.value
      }));
    }
    return wt({
      getData: Kt,
      getImage: ta,
      generateCsv: Se,
      generateImage: qe,
      generateSvg: dt,
      generatePdf: We,
      toggleAnnotator: xe,
      toggleTable: ut,
      toggleFullscreen: nt,
      copyAlt: ht
    }), (e, o) => (r(), i("div", {
      ref_key: "chordChart",
      ref: B,
      class: U({
        "vue-data-ui-component": !0,
        "vue-ui-chord": !0,
        "vue-data-ui-wrapper-fullscreen": fe.value,
        "vue-data-ui-responsive": t.value.responsive
      }),
      style: E(`font-family:${t.value.style.fontFamily};width:100%; text-align:center;background:${t.value.style.chart.backgroundColor}`),
      id: `chord_${w.value}`,
      onMouseenter: o[2] || (o[2] = () => d(Ge)(!0)),
      onMouseleave: o[3] || (o[3] = () => d(Ge)(!1))
    }, [
      t.value.userOptions.buttons.annotator && d(te) ? (r(), G(d(xt), {
        key: 0,
        color: t.value.style.chart.color,
        backgroundColor: t.value.style.chart.backgroundColor,
        active: se.value,
        svgRef: d(te),
        isCursorPointer: Y.value,
        onClose: xe
      }, {
        "annotator-action-close": p(() => [
          h(e.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": p(({ color: a }) => [
          h(e.$slots, "annotator-action-color", N(R({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": p(({ mode: a }) => [
          h(e.$slots, "annotator-action-draw", N(R({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": p(({ disabled: a }) => [
          h(e.$slots, "annotator-action-undo", N(R({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": p(({ disabled: a }) => [
          h(e.$slots, "annotator-action-redo", N(R({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": p(({ disabled: a }) => [
          h(e.$slots, "annotator-action-delete", N(R({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["color", "backgroundColor", "active", "svgRef", "isCursorPointer"])) : f("", !0),
      h(e.$slots, "userConfig", {}, void 0, !0),
      Et.value ? (r(), i("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Ve,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : f("", !0),
      t.value.style.chart.title.text ? (r(), i("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Fe,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (r(), G(Na, {
          key: `title_${ye.value}`,
          config: {
            title: {
              cy: "chord-div-title",
              ...t.value.style.chart.title
            },
            subtitle: {
              cy: "chord-div-subtitle",
              ...t.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : f("", !0),
      I("div", {
        id: `legend-top-${w.value}`
      }, null, 8, Va),
      t.value.userOptions.show && F.value && (d(Ye) || d(we)) ? (r(), G(d(_t), {
        ref_key: "userOptionsRef",
        ref: $e,
        key: `user_option_${Re.value}`,
        backgroundColor: t.value.style.chart.backgroundColor,
        color: t.value.style.chart.color,
        isPrinting: d(Ae),
        isImaging: d(Me),
        uid: w.value,
        hasTooltip: !1,
        hasPdf: t.value.userOptions.buttons.pdf,
        hasImg: t.value.userOptions.buttons.img,
        hasSvg: t.value.userOptions.buttons.svg,
        hasXls: t.value.userOptions.buttons.csv,
        hasTable: t.value.userOptions.buttons.table,
        hasLabel: !1,
        hasFullscreen: t.value.userOptions.buttons.fullscreen,
        hasAltCopy: t.value.userOptions.buttons.altCopy,
        isFullscreen: fe.value,
        chartElement: B.value,
        position: t.value.userOptions.position,
        titles: { ...t.value.userOptions.buttonTitles },
        hasAnnotator: t.value.userOptions.buttons.annotator,
        isAnnotation: se.value,
        callbacks: t.value.userOptions.callbacks,
        printScale: t.value.userOptions.print.scale,
        tableDialog: t.value.table.useDialog,
        isCursorPointer: Y.value,
        onToggleFullscreen: nt,
        onGeneratePdf: d(We),
        onGenerateCsv: Se,
        onGenerateImage: ua,
        onGenerateSvg: dt,
        onToggleTable: ut,
        onToggleAnnotator: xe,
        onCopyAlt: ht,
        style: E({
          visibility: d(Ye) ? d(we) ? "visible" : "hidden" : "visible"
        })
      }, Oe({ _: 2 }, [
        e.$slots.menuIcon ? {
          name: "menuIcon",
          fn: p(({ isOpen: a, color: l }) => [
            h(e.$slots, "menuIcon", N(R({ isOpen: a, color: l })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        e.$slots.optionPdf ? {
          name: "optionPdf",
          fn: p(() => [
            h(e.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        e.$slots.optionCsv ? {
          name: "optionCsv",
          fn: p(() => [
            h(e.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        e.$slots.optionImg ? {
          name: "optionImg",
          fn: p(() => [
            h(e.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        e.$slots.optionSvg ? {
          name: "optionSvg",
          fn: p(() => [
            h(e.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        e.$slots.optionTable ? {
          name: "optionTable",
          fn: p(() => [
            h(e.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        e.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: p(({ toggleFullscreen: a, isFullscreen: l }) => [
            h(e.$slots, "optionFullscreen", N(R({ toggleFullscreen: a, isFullscreen: l })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        e.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: p(({ toggleAnnotator: a, isAnnotator: l }) => [
            h(e.$slots, "optionAnnotator", N(R({ toggleAnnotator: a, isAnnotator: l })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        e.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: p(({ altCopy: a }) => [
            h(e.$slots, "optionAltCopy", N(R({ altCopy: a })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : f("", !0),
      (r(), i("svg", {
        xmlns: d(ma),
        ref_key: "svgRef",
        ref: te,
        viewBox: `0 0 ${L.value.width} ${L.value.height}`,
        preserveAspectRatio: "xMidYMid meet",
        style: {
          overflow: "visible"
        },
        class: U({ "vue-ui-chord-rotating": ne.value, "vue-ui-chord-idle": !ne.value }),
        onMousedown: bt(ot, ["prevent"]),
        onTouchstart: bt(ot, ["prevent"])
      }, [
        be(d(It)),
        e.$slots["chart-background"] ? (r(), i("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: L.value.width <= 0 ? 10 : L.value.width,
          height: L.value.height <= 0 ? 10 : L.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          h(e.$slots, "chart-background", {}, void 0, !0)
        ], 8, ja)) : f("", !0),
        t.value.style.chart.arcs.labels.curved ? (r(), i("defs", Ga, [
          (r(!0), i(V, null, Z(P.value.groups, (a, l) => (r(), i("path", {
            key: `labelPath-${l}`,
            id: `labelPath-${l}_${w.value}`,
            d: Yt(
              a.startAngle,
              a.endAngle,
              (y.value.inner + y.value.outer) / 2 + t.value.style.chart.arcs.labels.offset
            ),
            fill: "none"
          }, null, 8, Ya))), 128))
        ])) : f("", !0),
        e.$slots.pattern ? (r(), i("g", Wa, [
          (r(!0), i(V, null, Z(P.value.groups, (a, l) => (r(), i("defs", null, [
            h(e.$slots, "pattern", mt({ ref_for: !0 }, { seriesIndex: a.index, patternId: `pattern_${w.value}_${l}` }), void 0, !0)
          ]))), 256))
        ])) : f("", !0),
        I("g", {
          transform: `translate(${L.value.width / 2}, ${L.value.height / 2}) rotate(${_.value})`
        }, [
          (r(!0), i(V, null, Z(P.value.groups, (a, l) => (r(), i("g", null, [
            (r(), i("path", {
              class: U({ "vue-ui-chord-arc": !0, "vue-ui-chord-arc-animated": t.value.useCssAnimation && !K.value }),
              key: `arc-${l}`,
              d: Je(
                a.startAngle,
                a.endAngle,
                y.value.outer,
                y.value.inner
              ),
              fill: k.value.colors[l],
              stroke: t.value.style.chart.arcs.stroke,
              "stroke-width": t.value.style.chart.arcs.strokeWidth,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: E({
                opacity: rt(a)
              }),
              onMouseenter: (s) => Wt(a, l),
              onMouseleave: (s) => qt(a, l),
              onClick: (s) => Ht(a, l)
            }, null, 46, Ha)),
            e.$slots.pattern ? (r(), i("path", {
              class: U({ "vue-ui-chord-arc": !0, "vue-ui-chord-arc-animated": t.value.useCssAnimation && !K.value }),
              key: `arc-${l}`,
              d: Je(
                a.startAngle,
                a.endAngle,
                y.value.outer,
                y.value.inner
              ),
              fill: `url(#${a.pattern})`,
              stroke: t.value.style.chart.arcs.stroke,
              "stroke-width": t.value.style.chart.arcs.strokeWidth,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: E({
                opacity: rt(a),
                pointerEvents: "none"
              })
            }, null, 14, Qa)) : f("", !0)
          ]))), 256)),
          I("g", null, [
            (r(!0), i(V, null, Z(A.value ? P.value.chords.filter((a) => a.source.groupId === A.value.id) : b.value ? P.value.chords.filter((a) => a.source.groupId === b.value) : P.value.chords, (a, l) => (r(), i(V, {
              key: `ribbon-${a.id}`
            }, [
              a.source.value ? (r(), i("path", {
                key: 0,
                class: U({ "vue-ui-chord-ribbon": !0, "vue-ui-chord-ribbon-animated": t.value.useCssAnimation && !K.value }),
                d: ve(a.source, a.target),
                fill: t.value.style.chart.backgroundColor,
                style: E({ opacity: t.value.style.chart.ribbons.underlayerOpacity })
              }, null, 14, Za)) : f("", !0),
              a.source.value ? (r(), i("path", {
                key: 1,
                class: U({ "vue-ui-chord-ribbon": !0, "vue-ui-chord-ribbon-animated": t.value.useCssAnimation && !K.value }),
                d: ve(a.source, a.target),
                fill: k.value.colors[a.source.index],
                stroke: t.value.style.chart.ribbons.stroke,
                "stroke-width": t.value.style.chart.ribbons.strokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                style: E({ opacity: st(a) }),
                onMouseenter: (s) => Qt({
                  ...a,
                  path: ve(a.source, a.target),
                  color: k.value.colors[a.source.index]
                }, l),
                onClick: (s) => Jt({
                  ...a,
                  color: k.value.colors[a.source.index]
                }, l),
                onMouseleave: (s) => Zt({
                  ...a,
                  color: k.value.colors[a.source.index]
                }, l)
              }, null, 46, Ja)) : f("", !0),
              a.source.value && e.$slots.pattern ? (r(), i("path", {
                key: 2,
                class: U({ "vue-ui-chord-ribbon": !0, "vue-ui-chord-ribbon-animated": t.value.useCssAnimation && !K.value }),
                d: ve(a.source, a.target),
                fill: `url(#pattern_${w.value}_${a.source.index})`,
                stroke: t.value.style.chart.ribbons.stroke,
                "stroke-width": t.value.style.chart.ribbons.strokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                style: E({
                  opacity: st(a),
                  pointerEvents: "none"
                })
              }, null, 14, Ka)) : f("", !0)
            ], 64))), 128)),
            C.value ? (r(), i("path", {
              key: 0,
              d: C.value.path,
              fill: C.value.color,
              stroke: t.value.style.chart.ribbons.stroke,
              "stroke-width": t.value.style.chart.ribbons.strokeWidth,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: U({ "vue-ui-chord-ribbon": !0 }),
              style: {
                pointerEvents: "none"
              }
            }, null, 8, eo)) : f("", !0),
            C.value && e.$slots.pattern ? (r(), i("path", {
              key: 1,
              d: C.value.path,
              fill: `url(#${C.value.source.pattern})`,
              stroke: t.value.style.chart.ribbons.stroke,
              "stroke-width": t.value.style.chart.ribbons.strokeWidth,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: U({ "vue-ui-chord-ribbon": !0 }),
              style: {
                pointerEvents: "none"
              }
            }, null, 8, to)) : f("", !0)
          ]),
          (A.value || C.value || b.value) && t.value.style.chart.ribbons.labels.show ? (r(), i("g", ao, [
            (r(!0), i(V, null, Z(Ut.value, (a) => (r(), i(V, {
              key: a.id
            }, [
              I("line", {
                x1: a.midBaseX,
                y1: a.midBaseY,
                x2: Math.cos(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 12),
                y2: Math.sin(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 12),
                stroke: t.value.style.chart.backgroundColor,
                "stroke-width": t.value.style.chart.ribbons.labels.connector.strokeWidth * 3,
                "pointer-events": "none"
              }, null, 8, oo),
              I("line", {
                x1: a.midBaseX,
                y1: a.midBaseY,
                x2: Math.cos(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 12),
                y2: Math.sin(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 12),
                stroke: t.value.style.chart.ribbons.labels.connector.stroke,
                "stroke-width": t.value.style.chart.ribbons.labels.connector.strokeWidth,
                "pointer-events": "none"
              }, null, 8, lo),
              t.value.style.chart.ribbons.labels.marker.show ? (r(), i("circle", {
                key: 0,
                cx: Math.cos(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 12),
                cy: Math.sin(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 12),
                r: t.value.style.chart.ribbons.labels.marker.radius,
                stroke: t.value.style.chart.ribbons.labels.marker.stroke,
                "stroke-width": t.value.style.chart.ribbons.labels.marker.strokeWidth,
                fill: a.groupColor,
                "pointer-events": "none"
              }, null, 8, no)) : f("", !0),
              I("text", {
                transform: `
                            translate(
                                ${Math.cos(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 24)},
                                ${Math.sin(a.theta - Math.PI / 2) * (y.value.outer + t.value.style.chart.ribbons.labels.offset + 24)}
                            ) rotate(${-_.value})
                            `,
                fill: t.value.style.chart.ribbons.labels.useSerieColor ? a.groupColor : t.value.style.chart.ribbons.labels.color,
                "text-anchor": jt(a.theta),
                "font-size": t.value.style.chart.ribbons.labels.fontSize,
                "font-weight": t.value.style.chart.ribbons.labels.bold ? "bold" : "normal",
                dy: ".35em",
                "pointer-events": "none"
              }, J(lt(a.value)), 9, so)
            ], 64))), 128))
          ])) : f("", !0),
          t.value.style.chart.arcs.labels.show ? (r(), i("g", ro, [
            t.value.style.chart.arcs.labels.curved ? (r(!0), i(V, { key: 0 }, Z(P.value.groups, (a, l) => (r(), i("text", {
              class: "vue-ui-chord-label-curved",
              key: `curved-label-${l}`,
              "font-size": t.value.style.chart.arcs.labels.fontSize,
              "font-weight": t.value.style.chart.arcs.labels.bold ? "bold" : "normal",
              fill: t.value.style.chart.arcs.labels.adaptColorToBackground ? d(ya)(k.value.colors[l]) : t.value.style.chart.arcs.labels.color
            }, [
              I("textPath", {
                href: `#labelPath-${l}_${w.value}`,
                startOffset: "50%",
                "text-anchor": "middle"
              }, J(k.value.labels[l]) + J(t.value.style.chart.arcs.labels.showPercentage ? d(De)({
                p: " (",
                v: isNaN(a.proportion) ? 0 : a.proportion * 100,
                s: "%)",
                r: t.value.style.chart.arcs.labels.roundingPercentage
              }) : ""), 9, io)
            ], 8, uo))), 128)) : !A.value && !C.value && !b.value ? (r(!0), i(V, { key: 1 }, Z(P.value.groups, (a, l) => (r(), i("text", {
              class: "vue-ui-chord-label-straight",
              key: `label-${l}`,
              transform: `
                                ${Gt((a.startAngle + a.endAngle) / 2)}
                                rotate(${-_.value})
                            `,
              dy: ".35em",
              "text-anchor": Xt(a) > Math.PI ? "end" : "start",
              "font-size": t.value.style.chart.arcs.labels.fontSize,
              "font-weight": t.value.style.chart.arcs.labels.bold ? "bold" : "normal",
              fill: t.value.style.chart.arcs.labels.color,
              innerHTML: d(ka)({
                content: d(Ca)(aa(a, l)),
                fontSize: t.value.style.chart.arcs.labels.fontSize,
                fill: t.value.style.chart.arcs.labels.color,
                x: 0,
                y: 0
              })
            }, null, 8, co))), 128)) : f("", !0)
          ])) : f("", !0)
        ], 8, qa),
        h(e.$slots, "svg", {
          svg: {
            height: 600,
            width: 600,
            isPrintingImg: d(Ae) | d(Me) | ie.value,
            isPrintingSvg: ce.value
          }
        }, void 0, !0)
      ], 42, Xa)),
      e.$slots.watermark ? (r(), i("div", vo, [
        h(e.$slots, "watermark", N(R({ isPrinting: d(Ae) || d(Me) || ie.value || ce.value })), void 0, !0)
      ])) : f("", !0),
      I("div", {
        id: `legend-bottom-${w.value}`
      }, null, 8, ho),
      Xe.value ? (r(), G(va, {
        key: 5,
        to: t.value.style.chart.legend.position === "top" ? `#legend-top-${w.value}` : `#legend-bottom-${w.value}`
      }, [
        I("div", {
          ref_key: "chartLegend",
          ref: ze
        }, [
          t.value.style.chart.legend.show ? (r(), G(Fa, {
            key: `legend_${Ce.value}`,
            legendSet: _e.value,
            config: ea.value,
            isCursorPointer: Y.value,
            onClickMarker: o[0] || (o[0] = ({ legend: a }) => ct(a.id))
          }, Oe({
            item: p(({ legend: a }) => [
              I("div", {
                style: E({
                  opacity: b.value ? b.value === a.id ? 1 : 0.3 : 1
                }),
                onClick: (l) => a.select()
              }, J(a.name), 13, fo)
            ]),
            _: 2
          }, [
            e.$slots.pattern ? {
              name: "legend-pattern",
              fn: p(({ legend: a, index: l }) => [
                be(Ea, {
                  shape: a.shape,
                  radius: 30,
                  stroke: "none",
                  plot: { x: 30, y: 30 },
                  fill: `url(#pattern_${w.value}_${l})`
                }, null, 8, ["shape", "fill"])
              ]),
              key: "0"
            } : void 0
          ]), 1032, ["legendSet", "config", "isCursorPointer"])) : f("", !0),
          h(e.$slots, "legend", { legend: _e.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : f("", !0),
      e.$slots.source ? (r(), i("div", {
        key: 6,
        ref_key: "source",
        ref: Ue,
        dir: "auto"
      }, [
        h(e.$slots, "source", {}, void 0, !0)
      ], 512)) : f("", !0),
      _.value !== t.value.initialRotation ? (r(), i("div", go, [
        h(e.$slots, "reset-action", { reset: it }, () => [
          I("button", {
            "data-cy-reset": "",
            tabindex: "0",
            role: "button",
            class: "vue-data-ui-refresh-button",
            style: E({
              background: t.value.style.chart.backgroundColor,
              cursor: Y.value ? "pointer" : "default"
            }),
            onClick: it
          }, [
            be(d(Ne), {
              name: "refresh",
              stroke: t.value.style.chart.color
            }, null, 8, ["stroke"])
          ], 4)
        ], !0)
      ])) : f("", !0),
      F.value && t.value.userOptions.buttons.table ? (r(), G(da(re.value.component), mt({ key: 8 }, re.value.props, {
        ref_key: "tableUnit",
        ref: oe,
        onClose: vt
      }), Oe({
        content: p(() => [
          (r(), G(d(Pt), {
            key: `table_${ke.value}`,
            colNames: pe.value.colNames,
            head: pe.value.head,
            body: pe.value.body,
            config: pe.value.config,
            title: t.value.table.useDialog ? "" : re.value.title,
            withCloseButton: !t.value.table.useDialog,
            isCursorPointer: Y.value,
            onClose: vt
          }, {
            th: p(({ th: a }) => [
              yt(J(a.name), 1)
            ]),
            td: p(({ td: a }) => [
              I("div", po, J(a.name ? a.name : lt(a)), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        t.value.table.useDialog ? {
          name: "title",
          fn: p(() => [
            yt(J(re.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        t.value.table.useDialog ? {
          name: "actions",
          fn: p(() => [
            I("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: o[1] || (o[1] = (a) => Se(t.value.userOptions.callbacks.csv)),
              style: E({ cursor: Y.value ? "pointer" : "default" })
            }, [
              be(d(Ne), {
                name: "fileCsv",
                stroke: re.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : f("", !0),
      h(e.$slots, "skeleton", {}, () => [
        d(je) ? (r(), G(_a, { key: 0 })) : f("", !0)
      ], !0)
    ], 46, Ua));
  }
}, Do = /* @__PURE__ */ za(bo, [["__scopeId", "data-v-9e6da50e"]]);
export {
  Do as default
};
