// vue-data-ui/dist/vue-ui-gauge-J2VYHsQe.js
import { defineAsyncComponent as ae, computed as d, ref as f, toRefs as ft, watch as le, shallowRef as Pe, onMounted as yt, onBeforeUnmount as gt, openBlock as s, createElementBlock as n, unref as u, normalizeStyle as oe, normalizeClass as Me, createBlock as V, withCtx as k, renderSlot as h, normalizeProps as $, guardReactiveProps as A, createCommentVNode as c, toDisplayString as T, createSlots as mt, createVNode as pt, createElementVNode as b, Fragment as S, renderList as z, mergeProps as L, nextTick as Ne } from "vue";
import { j as kt, t as Ie, m as bt, p as Xe, n as St, o as Re, s as wt, q as H, r as Ye, a3 as se, ad as M, ae as _t, K as Ct, X as xt, u as re, Z as $t, a6 as At, k as ne, l as ue, a2 as ie } from "./lib-GwPnJSrY.js";
import { t as zt, u as Ft } from "./useResponsive-DfdjqQps.js";
import { u as Ot, B as Pt } from "./BaseScanner-BhhkKVoj.js";
import { u as Mt } from "./usePrinter-DiK5R6k-.js";
import { u as Nt } from "./useConfig-CIgyKZ6R.js";
import { u as It } from "./useSvgExport-ByUukOZt.js";
import { u as ce } from "./useNestedProp-BBT0mbHh.js";
import { u as Xt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Rt } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Yt } from "./useChartAccessibility-9icAAmYg.js";
import { u as Tt } from "./useAutoSizeLabelsInsideViewbox-80fzdfxt.js";
import Et from "./img-BYYzY_QU.js";
import Lt from "./vue_ui_gauge-BMEuRC4J.js";
import Gt from "./Title-BfyrP-jp.js";
import { _ as Ut } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const jt = ["id"], Bt = { key: 0 }, Dt = ["xmlns", "viewBox"], Wt = ["width", "height"], Vt = ["id"], Ht = ["stop-color"], qt = ["stop-color"], Zt = ["stop-color"], Kt = ["id"], Jt = ["stdDeviation"], Qt = { key: 1 }, ea = ["d", "fill", "stroke"], ta = ["d", "fill", "stroke"], aa = ["d", "fill"], la = ["id", "d"], oa = ["fill", "font-size", "font-weight"], sa = ["href", "startOffset"], ra = ["text-anchor", "fill", "font-size", "font-weight", "innerHTML"], na = ["d", "filter"], ua = ["stroke", "stroke-width"], ia = ["stroke", "stroke-width"], ca = ["stroke", "stroke-width"], va = ["stroke", "stroke-width"], da = { key: 8 }, ha = ["x", "y", "text-anchor", "font-size", "font-weight", "fill"], fa = ["x", "y", "font-size", "font-weight", "fill"], ya = { key: 0 }, ga = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], ma = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], pa = ["x1", "y1", "x2", "y2", "stroke-width", "filter"], ka = { key: 1 }, ba = ["d", "fill", "stroke", "stroke-width"], Sa = ["cx", "cy", "fill", "r", "stroke-width", "stroke"], wa = ["x", "y", "font-size", "fill"], _a = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Ca = {
  __name: "vue-ui-gauge",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["copyAlt"],
  setup(Te, { expose: Ee, emit: Le }) {
    const Ge = ae(() => import("./PenAndPaper-DWV1FMIi.js")), Ue = ae(() => import("./UserOptions-CrLqEimB.js")), je = ae(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_gauge: Be } = Nt(), { isThemeValid: De, warnInvalidTheme: We } = Xt(), g = Te, Ve = Le, He = d(() => !!g.dataset && Object.keys(g.dataset).length > 0 && g.dataset.series && g.dataset.series.length), C = f(kt()), qe = f(null), ve = f(0), N = f(null), de = f(null), he = f(null), fe = f(null), ye = f(null), ge = f(0), G = f(!1), U = f(!1), e = f(Z()), me = d(() => e.value.userOptions.useCursorPointer), Ze = d(() => Ie({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            animation: { use: !1 },
            layout: {
              markers: { show: !1 },
              segmentSeparators: { stroke: "#6A6A6A" },
              segmentNames: { show: !1 },
              indicatorArc: { fill: "#6A6A6A50" },
              pointer: {
                stroke: "#6A6A6A",
                useRatingColor: !0,
                circle: {
                  stroke: "#6A6A6A",
                  color: "#6A6A6A"
                }
              }
            },
            legend: {
              show: !1
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: pe, FINAL_DATASET: x, manualLoading: X } = Ot({
      ...ft(g),
      FINAL_CONFIG: e,
      prepareConfig: Z,
      callback: () => {
        Promise.resolve().then(async () => {
          await Ne();
        });
      },
      skeletonDataset: g.config?.skeletonDataset ?? {
        value: 0,
        series: [
          { from: -1, to: 0, name: "_", color: "#A1A1A1" },
          { from: 0, to: 1, name: "__", color: "#CACACA" }
        ]
      },
      skeletonConfig: Ie({
        defaultConfig: e.value,
        userConfig: Ze.value
      })
    }), { userOptionsVisible: q, setUserOptionsVisibility: ke, keepUserOptionState: be } = Rt({ config: e.value }), { svgRef: j } = Yt({ config: e.value.style.chart.title });
    le(() => g.config, (t) => {
      pe.value || (e.value = Z()), q.value = !e.value.userOptions.showOnChartHover, ee(), ge.value += 1;
    }, { deep: !0 }), le(() => g.dataset, (t) => {
      ee(), t && Object.keys(t).length > 0 && (X.value = e.value?.loading ?? !1);
    }, { deep: !0, immediate: !1 });
    function Z() {
      const t = ce({
        userConfig: g.config,
        defaultConfig: Be
      }), r = t.theme;
      if (!r) return t;
      if (!De.value(t))
        return We(t), t;
      const a = ce({
        userConfig: Lt[r] || g.config,
        defaultConfig: t
      }), l = ce({
        userConfig: g.config,
        defaultConfig: a
      });
      return {
        ...l,
        customPalette: l.customPalette.length ? l.customPalette : bt[r] || Xe
      };
    }
    const { isPrinting: K, isImaging: J, generatePdf: Se, generateImage: we } = Mt({
      elementId: `vue-ui-gauge_${C.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-gauge",
      options: e.value.userOptions.print
    }), Ke = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), Je = d(() => St(e.value.customPalette)), F = d(() => {
      if (Re(x.value.series || {}))
        return {
          value: 0,
          series: [
            {
              from: 0,
              to: 0
            }
          ]
        };
      const t = [];
      (x.value.series || []).forEach((a) => {
        t.push(a.from || 1e-7), t.push(a.to || 1e-7);
      });
      const r = Math.max(...t);
      return {
        ...x.value,
        series: (x.value.series || []).map((a, l) => ({
          ...a,
          color: wt(a.color) || Je.value[l] || Xe[l],
          value: ((a.to || 0) - (a.from || 0)) / r * 100
        }))
      };
    }), w = f(512), o = f({
      height: 358.4,
      width: w.value,
      top: 0,
      bottom: 358.4,
      centerX: 179.2,
      centerY: w.value / 2,
      labelFontSize: 18,
      legendFontSize: e.value.style.chart.legend.fontSize,
      pointerRadius: e.value.style.chart.layout.pointer.circle.radius,
      trackSize: e.value.style.chart.layout.track.size,
      pointerSize: e.value.style.chart.layout.pointer.size,
      pointerStrokeWidth: e.value.style.chart.layout.pointer.strokeWidth,
      markerOffset: e.value.style.chart.layout.markers.offsetY + 3,
      segmentFontSize: e.value.style.chart.layout.segmentNames.fontSize
    }), R = f(0), O = f(0), m = f(
      e.value.style.chart.animation.use ? Math.min(...x.value.series.map((t) => t.from)) : x.value.value
    );
    le(() => x.value.value, () => {
      xe(x.value.value);
    });
    const i = d(() => {
      const t = o.value.width / 2, r = v.value.base, a = Math.PI * ((m.value + 0 - O.value) / (R.value - O.value)) + Math.PI;
      return {
        x1: t,
        y1: r,
        x2: t + v.value.pointerSize * o.value.pointerSize * 0.9 * Math.cos(a),
        y2: r + v.value.pointerSize * o.value.pointerSize * 0.9 * Math.sin(a)
      };
    }), _e = d(() => {
      const t = o.value.width / 2, r = v.value.base, a = Math.PI * ((m.value + 0 - O.value) / (R.value - O.value)) + Math.PI, l = t + v.value.pointerSize * o.value.pointerSize * 0.9 * Math.cos(a), p = r + v.value.pointerSize * o.value.pointerSize * 0.9 * Math.sin(a), _ = o.value.pointerRadius, ct = t + _ * Math.cos(a + Math.PI / 2), vt = r + _ * Math.sin(a + Math.PI / 2), dt = t + _ * Math.cos(a - Math.PI / 2), ht = r + _ * Math.sin(a - Math.PI / 2);
      return isNaN(l) ? null : `M ${l},${p} ${ct},${vt} ${dt},${ht} Z`;
    }), Q = d(() => {
      for (let t = 0; t < F.value.series.length; t += 1) {
        const { color: r, from: a, to: l } = F.value.series[t];
        if (m.value >= a && m.value <= l)
          return r;
      }
      return "#2D353C";
    }), P = Pe(null), Y = Pe(null), B = d({
      get: () => o.value.segmentFontSize,
      set: (t) => t
    }), { autoSizeLabels: Ce } = Tt({
      svgRef: j,
      fontSize: o.value.segmentFontSize,
      minFontSize: e.value.style.chart.layout.segmentNames.minFontSize,
      sizeRef: B,
      labelClass: ".vue-ui-gauge-label-flat"
    }), D = d(() => !!e.value.debug);
    function ee() {
      let t = !1;
      if (Re(g.dataset) ? (H({
        componentName: "VueUiGauge",
        type: "dataset",
        debug: D.value
      }), X.value = !0, t = !0) : (Ye({
        datasetObject: g.dataset,
        requiredAttributes: ["value", "series"]
      }).forEach((r) => {
        H({
          componentName: "VueUiGauge",
          type: "datasetAttribute",
          property: r,
          debug: D.value
        }), X.value = !0, t = !0;
      }), Object.hasOwn(g.dataset, "series") && (g.dataset.series.length ? g.dataset.series.forEach((r, a) => {
        Ye({
          datasetObject: r,
          requiredAttributes: ["from", "to"]
        }).forEach((l) => {
          H({
            componentName: "VueUiGauge",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: D.value
          }), X.value = !0, t = !0;
        });
      }) : (H({
        componentName: "VueUiGauge",
        type: "datasetAttributeEmpty",
        property: "series",
        debug: D.value
      }), X.value = !0, t = !0))), X.value = t, xe(x.value.value || 0), e.value.responsive) {
        const r = zt(() => {
          let { width: a, height: l } = Ft({
            chart: N.value,
            title: e.value.style.chart.title.text ? de.value : null,
            legend: he.value,
            source: fe.value,
            noTitle: ye.value
          });
          l -= 12, requestAnimationFrame(() => {
            o.value.width = a, o.value.height = l, o.value.centerX = a / 2, o.value.centerY = w.value / 2 / 358.4 * l, o.value.bottom = l, o.value.labelFontSize = 18 / w.value * Math.min(l, a) < 10 ? 10 : 18 / w.value * Math.min(l, a), o.value.legendFontSize = e.value.style.chart.legend.fontSize / w.value * Math.min(l, a) < 14 ? 14 : e.value.style.chart.legend.fontSize / w.value * Math.min(l, a), o.value.pointerRadius = e.value.style.chart.layout.pointer.circle.radius / w.value * Math.min(l, a), o.value.trackSize = e.value.style.chart.layout.track.size / w.value * Math.min(l, a), o.value.pointerStrokeWidth = ie({
              relator: Math.min(a, l),
              adjuster: w.value,
              source: e.value.style.chart.layout.pointer.strokeWidth,
              threshold: 2,
              fallback: 2
            }), o.value.markerOffset = ie({
              relator: Math.max(a, l),
              adjuster: w.value,
              source: e.value.style.chart.layout.markers.offsetY + 3,
              threshold: 2,
              fallback: 2
            }), o.value.segmentFontSize = ie({
              relator: Math.min(a, l),
              adjuster: w.value,
              source: e.value.style.chart.layout.segmentNames.fontSize,
              threshold: 8,
              fallback: 8
            });
          }), Ce();
        });
        P.value && (Y.value && P.value.unobserve(Y.value), P.value.disconnect()), P.value = new ResizeObserver(r), Y.value = N.value.parentNode, P.value.observe(Y.value);
      }
      Ce();
    }
    yt(() => {
      ee();
    }), gt(() => {
      P.value && (Y.value && P.value.unobserve(Y.value), P.value.disconnect());
    });
    function xe(t) {
      const r = [];
      (F.value.series || []).forEach((_) => {
        r.push(_.from || 0), r.push(_.to || 0);
      }), R.value = Math.max(...r), O.value = Math.min(...r);
      let a = e.value.style.chart.animation.speed;
      const l = Math.abs(t - m.value) / (a * 60);
      function p() {
        m.value < t ? m.value = Math.min(m.value + l, t) : m.value > t && (m.value = Math.max(m.value - l, t)), m.value !== t && requestAnimationFrame(p);
      }
      p();
    }
    const v = d(() => {
      const t = e.value.responsive ? Math.min(o.value.width, o.value.height) : o.value.width, r = 2.5 / e.value.style.chart.layout.radiusRatio;
      return {
        arcs: t / r,
        gradients: t / (r * 1.1),
        base: e.value.responsive ? o.value.height / 1.618 : o.value.height * 0.7,
        ratingBase: e.value.responsive ? o.value.height / 2 + o.value.height / 4 : o.value.height * 0.9,
        pointerSize: e.value.responsive ? Math.min(o.value.width, o.value.height) / 3 : o.value.width / 3.2
      };
    }), y = d(() => se(
      { series: F.value.series },
      o.value.width / 2,
      v.value.base,
      v.value.arcs,
      v.value.arcs,
      1,
      1,
      1,
      180,
      109.9495,
      40 * o.value.trackSize
    )), Qe = d(() => se(
      { series: F.value.series },
      o.value.width / 2,
      v.value.base,
      v.value.arcs * e.value.style.chart.layout.segmentNames.offsetRatio,
      v.value.arcs * e.value.style.chart.layout.segmentNames.offsetRatio,
      1,
      1,
      1,
      180,
      109.9495,
      40 * o.value.trackSize
    )), I = d(() => y.value.map((t) => v.value.arcs * e.value.style.chart.layout.segmentNames.offsetRatio * (t.nameOffsetRatio || 1))), $e = d(() => {
      const { x: t, y: r } = M({
        initX: y.value[0].firstSeparator.x,
        initY: y.value[0].firstSeparator.y,
        centerX: i.value.x1,
        centerY: i.value.y1,
        offset: -e.value.style.chart.layout.segmentSeparators.offsetIn
      }), { x: a, y: l } = M({
        initX: y.value[0].startX,
        initY: y.value[0].startY,
        centerX: i.value.x1,
        centerY: i.value.y1,
        offset: e.value.style.chart.layout.segmentSeparators.offsetOut
      });
      return { x1: t, y1: r, x2: a, y2: l };
    }), Ae = d(() => y.value.map((t) => {
      const { x: r, y: a } = M({
        initX: t.separator.x,
        initY: t.separator.y,
        centerX: i.value.x1,
        centerY: i.value.y1,
        offset: -e.value.style.chart.layout.segmentSeparators.offsetIn
      }), { x: l, y: p } = M({
        initX: t.endX,
        initY: t.endY,
        centerX: i.value.x1,
        centerY: i.value.y1,
        offset: e.value.style.chart.layout.segmentSeparators.offsetOut
      });
      return { x1: r, y1: a, x2: l, y2: p };
    }));
    function et(t) {
      if (t.reduce((p, _) => p + _, 0) > 100)
        throw new Error("Total % must not exceed 100");
      const a = 50;
      let l = 0;
      return t.map((p) => (l += p, `${l / 100 * a - p / 4}%`));
    }
    const tt = d(() => et(y.value.map((t) => t.proportion * 100))), at = d(() => se(
      { series: F.value.series },
      o.value.width / 2,
      v.value.base,
      v.value.gradients,
      v.value.gradients,
      0.95,
      1,
      1,
      180,
      110.02,
      2 * o.value.trackSize
    )), lt = d(() => {
      const t = O.value >= 0 ? -O.value : Math.abs(O.value);
      return _t({
        radius: e.value.style.chart.layout.indicatorArc.radius * o.value.trackSize,
        centerX: o.value.width / 2,
        centerY: v.value.base,
        percentage: Ct((m.value + t) / (R.value + t))
      });
    }), E = f(!1);
    function ze(t) {
      E.value = t, ve.value += 1;
    }
    const W = f(!1);
    function te() {
      W.value = !W.value;
    }
    async function ot({ scale: t = 2 } = {}) {
      if (!N.value) return;
      const { width: r, height: a } = N.value.getBoundingClientRect(), l = r / a, { imageUri: p, base64: _ } = await Et({ domElement: N.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: p,
        base64: _,
        title: e.value.style.chart.title.text,
        width: r,
        height: a,
        aspectRatio: l
      };
    }
    const st = d(() => e.value.style.chart.backgroundColor), rt = d(() => e.value.style.chart.title), { exportSvg: nt, getSvg: ut } = It({
      svg: j,
      title: rt,
      backgroundColor: st
    });
    async function Fe({ isCb: t }) {
      U.value = !0, await Ne();
      try {
        if (t) {
          const { blob: r, url: a, text: l, dataUrl: p } = await ut();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: r, url: a, text: l, dataUrl: p }));
        } else
          await Promise.resolve(nt());
      } finally {
        U.value = !1;
      }
    }
    function it(t) {
      if (t?.stage === "start") {
        G.value = !0;
        return;
      }
      if (t?.stage === "end") {
        G.value = !1;
        return;
      }
      we();
    }
    async function Oe() {
      if (Ve("copyAlt", {
        config: e.value,
        dataset: F.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: F.value
      }));
    }
    return Ee({
      getImage: ot,
      generatePdf: Se,
      generateImage: we,
      generateSvg: Fe,
      toggleAnnotator: te,
      toggleFullscreen: ze,
      copyAlt: Oe
    }), (t, r) => (s(), n("div", {
      class: Me(`vue-data-ui-component vue-ui-gauge ${E.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      ref_key: "gaugeChart",
      ref: N,
      id: `vue-ui-gauge_${C.value}`,
      style: oe(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: r[0] || (r[0] = () => u(ke)(!0)),
      onMouseleave: r[1] || (r[1] = () => u(ke)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (s(), V(u(Ge), {
        key: 0,
        svgRef: u(j),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: W.value,
        isCursorPointer: me.value,
        onClose: te
      }, {
        "annotator-action-close": k(() => [
          h(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": k(({ color: a }) => [
          h(t.$slots, "annotator-action-color", $(A({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": k(({ mode: a }) => [
          h(t.$slots, "annotator-action-draw", $(A({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": k(({ disabled: a }) => [
          h(t.$slots, "annotator-action-undo", $(A({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": k(({ disabled: a }) => [
          h(t.$slots, "annotator-action-redo", $(A({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": k(({ disabled: a }) => [
          h(t.$slots, "annotator-action-delete", $(A({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : c("", !0),
      Ke.value ? (s(), n("div", {
        key: 1,
        ref_key: "noTitle",
        ref: ye,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : c("", !0),
      e.value.style.chart.title.text ? (s(), n("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: de,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (s(), V(Gt, {
          key: `title_${ge.value}`,
          config: {
            title: {
              cy: "gauge-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "gauge-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, {
          default: k(() => [
            e.value.translations.base && u(x).base ? (s(), n("span", Bt, T(e.value.translations.base) + ": " + T(u(x).base), 1)) : c("", !0)
          ]),
          _: 1
        }, 8, ["config"]))
      ], 512)) : c("", !0),
      e.value.userOptions.show && He.value && (u(be) || u(q)) ? (s(), V(u(Ue), {
        ref_key: "details",
        ref: qe,
        key: `user_options_${ve.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isImaging: u(J),
        isPrinting: u(K),
        uid: C.value,
        hasXls: !1,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: E.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: N.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: W.value,
        isCursorPointer: me.value,
        onToggleFullscreen: ze,
        onGeneratePdf: u(Se),
        onGenerateImage: it,
        onGenerateSvg: Fe,
        onToggleAnnotator: te,
        onCopyAlt: Oe,
        style: oe({
          visibility: u(be) ? u(q) ? "visible" : "hidden" : "visible"
        })
      }, mt({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: k(({ isOpen: a, color: l }) => [
            h(t.$slots, "menuIcon", $(A({ isOpen: a, color: l })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: k(() => [
            h(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: k(() => [
            h(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: k(() => [
            h(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: k(({ toggleFullscreen: a, isFullscreen: l }) => [
            h(t.$slots, "optionFullscreen", $(A({ toggleFullscreen: a, isFullscreen: l })), void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: k(({ toggleAnnotator: a, isAnnotator: l }) => [
            h(t.$slots, "optionAnnotator", $(A({ toggleAnnotator: a, isAnnotator: l })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: k(({ altCopy: a }) => [
            h(t.$slots, "optionAltCopy", $(A({ altCopy: a })), void 0, !0)
          ]),
          key: "6"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "callbacks", "printScale", "position", "hasAnnotator", "isAnnotation", "isCursorPointer", "onGeneratePdf", "style"])) : c("", !0),
      (s(), n("svg", {
        ref_key: "svgRef",
        ref: j,
        xmlns: u(xt),
        class: Me({ "vue-data-ui-fullscreen--on": E.value, "vue-data-ui-fulscreen--off": !E.value }),
        viewBox: `0 0 ${o.value.width <= 0 ? 10 : o.value.width} ${o.value.height <= 0 ? 10 : o.value.height}`,
        style: oe(`max-width:100%;overflow:hidden !important;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        pt(u(je)),
        t.$slots["chart-background"] ? (s(), n("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: o.value.width <= 0 ? 10 : o.value.width,
          height: o.value.height <= 0 ? 10 : o.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          h(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Wt)) : c("", !0),
        b("defs", null, [
          b("radialGradient", {
            id: `gradient_${C.value}`,
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%"
          }, [
            b("stop", {
              offset: "0%",
              "stop-color": u(re)("#FFFFFF", 1)
            }, null, 8, Ht),
            b("stop", {
              offset: "80%",
              "stop-color": u(re)("#FFFFFF", e.value.style.chart.layout.track.gradientIntensity)
            }, null, 8, qt),
            b("stop", {
              offset: "100%",
              "stop-color": u(re)("#FFFFFF", 1)
            }, null, 8, Zt)
          ], 8, Vt)
        ]),
        b("defs", null, [
          b("filter", {
            id: `blur_${C.value}`,
            x: "-50%",
            y: "-50%",
            width: "200%",
            height: "200%"
          }, [
            b("feGaussianBlur", {
              in: "SourceGraphic",
              stdDeviation: 100 / e.value.style.chart.layout.track.gradientIntensity
            }, null, 8, Jt)
          ], 8, Kt)
        ]),
        t.$slots.pattern ? (s(), n("g", Qt, [
          (s(!0), n(S, null, z(y.value, (a, l) => (s(), n("defs", null, [
            h(t.$slots, "pattern", L({ ref_for: !0 }, { seriesIndex: l, patternId: `pattern_${C.value}_${l}` }), void 0, !0)
          ]))), 256))
        ])) : c("", !0),
        (s(!0), n(S, null, z(y.value, (a, l) => (s(), n("path", {
          key: `arc_${l}`,
          d: a.arcSlice,
          fill: a.color,
          stroke: e.value.style.chart.backgroundColor,
          "stroke-linecap": "round"
        }, null, 8, ea))), 128)),
        t.$slots.pattern ? (s(!0), n(S, { key: 2 }, z(y.value, (a, l) => (s(), n("path", {
          key: `arc_${l}`,
          d: a.arcSlice,
          fill: `url(#pattern_${C.value}_${l})`,
          stroke: e.value.style.chart.backgroundColor,
          "stroke-linecap": "round"
        }, null, 8, ta))), 128)) : c("", !0),
        e.value.style.chart.layout.indicatorArc.show ? (s(), n("path", {
          key: 3,
          d: lt.value,
          fill: e.value.style.chart.layout.indicatorArc.fill
        }, null, 8, aa)) : c("", !0),
        e.value.style.chart.layout.segmentNames.show && e.value.style.chart.layout.segmentNames.curved ? (s(), n(S, { key: 4 }, [
          (s(!0), n(S, null, z(y.value, (a, l) => (s(), n("path", {
            id: `curve_${C.value}_${l}`,
            d: `M ${i.value.x1},${i.value.y1} m -${I.value[l]},0 a ${I.value[l]},${I.value[l]} 0 1,1 ${2 * I.value[l]},0 a ${I.value[l]},${I.value[l]} 0 1,1 -${2 * I.value[l]},0`,
            fill: "transparent"
          }, null, 8, la))), 256)),
          (s(!0), n(S, null, z(y.value, (a, l) => (s(), n("text", {
            fill: e.value.style.chart.layout.segmentNames.useSerieColor ? a.color : e.value.style.chart.layout.segmentNames.color,
            "font-size": B.value,
            "font-weight": e.value.style.chart.layout.segmentNames.bold ? "bold" : "normal",
            "text-anchor": "middle"
          }, [
            b("textPath", {
              href: `#curve_${C.value}_${l}`,
              startOffset: tt.value[l]
            }, T(a.name || ""), 9, sa)
          ], 8, oa))), 256))
        ], 64)) : c("", !0),
        e.value.style.chart.layout.segmentNames.show && !e.value.style.chart.layout.segmentNames.curved ? (s(!0), n(S, { key: 5 }, z(Qe.value, (a, l) => (s(), n("text", {
          class: "vue-ui-gauge-label-flat",
          "text-anchor": u(At)(a, !1, 12).anchor,
          fill: e.value.style.chart.layout.segmentNames.useSerieColor ? a.color : e.value.style.chart.layout.segmentNames.color,
          "font-size": B.value,
          "font-weight": e.value.style.chart.layout.segmentNames.bold ? "bold" : "normal",
          innerHTML: u($t)({
            content: String(a.name ?? ""),
            fontSize: B.value,
            fill: e.value.style.chart.layout.segmentNames.useSerieColor ? a.color : e.value.style.chart.layout.segmentNames.color,
            x: a.center.endX,
            y: a.center.endY
          })
        }, null, 8, ra))), 256)) : c("", !0),
        e.value.style.chart.layout.track.useGradient ? (s(!0), n(S, { key: 6 }, z(at.value, (a, l) => (s(), n("path", {
          key: `arc_${l}`,
          d: a.arcSlice,
          fill: "#FFFFFF",
          stroke: "none",
          "stroke-linecap": "round",
          filter: `url(#blur_${C.value})`
        }, null, 8, na))), 128)) : c("", !0),
        e.value.style.chart.layout.segmentSeparators.show ? (s(), n(S, { key: 7 }, [
          b("line", L($e.value, {
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": e.value.style.chart.layout.segmentSeparators.strokeWidth + 2,
            "stroke-linecap": "round"
          }), null, 16, ua),
          b("line", L($e.value, {
            stroke: e.value.style.chart.layout.segmentSeparators.stroke,
            "stroke-width": e.value.style.chart.layout.segmentSeparators.strokeWidth,
            "stroke-linecap": "round"
          }), null, 16, ia),
          (s(!0), n(S, null, z(Ae.value, (a) => (s(), n("line", L({ ref_for: !0 }, a, {
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": e.value.style.chart.layout.segmentSeparators.strokeWidth + 2,
            "stroke-linecap": "round"
          }), null, 16, ca))), 256)),
          (s(!0), n(S, null, z(Ae.value, (a) => (s(), n("line", L({ ref_for: !0 }, a, {
            stroke: e.value.style.chart.layout.segmentSeparators.stroke,
            "stroke-width": e.value.style.chart.layout.segmentSeparators.strokeWidth,
            "stroke-linecap": "round"
          }), null, 16, va))), 256))
        ], 64)) : c("", !0),
        e.value.style.chart.layout.markers.show ? (s(), n("g", da, [
          (s(!0), n(S, null, z(y.value, (a, l) => (s(), n("text", {
            x: u(M)({
              centerX: i.value.x1,
              centerY: v.value.base,
              initX: a.center.startX,
              initY: a.center.startY,
              offset: o.value.markerOffset
            }).x,
            y: u(M)({
              centerX: i.value.x1,
              centerY: v.value.base,
              initX: a.center.startX,
              initY: a.center.startY,
              offset: o.value.markerOffset
            }).y,
            "text-anchor": a.center.startX < i.value.x1 - 5 ? "end" : a.center.startX > i.value.x1 + 5 ? "start" : "middle",
            "font-size": o.value.labelFontSize * e.value.style.chart.layout.markers.fontSizeRatio,
            "font-weight": `${e.value.style.chart.layout.markers.bold ? "bold" : "normal"}`,
            fill: e.value.style.chart.layout.markers.color
          }, T(u(ne)(
            e.value.style.chart.layout.markers.formatter,
            a.from,
            u(ue)({
              p: e.value.style.chart.layout.markers.prefix,
              v: a.from,
              s: e.value.style.chart.layout.markers.suffix,
              r: e.value.style.chart.layout.markers.roundingValue
            })
          )), 9, ha))), 256))
        ])) : c("", !0),
        e.value.style.chart.layout.markers.show ? (s(), n("text", {
          key: 9,
          x: u(M)({
            centerX: o.value.width / 2,
            centerY: v.value.base,
            initX: y.value.at(-1).endX,
            initY: y.value.at(-1).endY,
            offset: o.value.markerOffset
          }).x,
          y: u(M)({
            centerX: o.value.width / 2,
            centerY: v.value.base,
            initX: y.value.at(-1).endX,
            initY: y.value.at(-1).endY,
            offset: o.value.markerOffset
          }).y,
          "text-anchor": "start",
          "font-size": o.value.labelFontSize * e.value.style.chart.layout.markers.fontSizeRatio,
          "font-weight": `${e.value.style.chart.layout.markers.bold ? "bold" : "normal"}`,
          fill: e.value.style.chart.layout.markers.color
        }, T(u(ne)(
          e.value.style.chart.layout.markers.formatter,
          R.value,
          u(ue)({
            p: e.value.style.chart.layout.markers.prefix,
            v: R.value,
            s: e.value.style.chart.layout.markers.suffix,
            r: e.value.style.chart.layout.markers.roundingValue
          })
        )), 9, fa)) : c("", !0),
        e.value.style.chart.layout.pointer.show ? (s(), n(S, { key: 10 }, [
          e.value.style.chart.layout.pointer.type === "rounded" ? (s(), n("g", ya, [
            isNaN(i.value.x2) ? c("", !0) : (s(), n("line", {
              key: 0,
              x1: i.value.x1,
              y1: i.value.y1,
              x2: i.value.x2,
              y2: i.value.y2,
              stroke: e.value.style.chart.layout.pointer.stroke,
              "stroke-width": o.value.pointerStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, ga)),
            isNaN(i.value.x2) ? c("", !0) : (s(), n("line", {
              key: 1,
              x1: i.value.x1,
              y1: i.value.y1,
              x2: i.value.x2,
              y2: i.value.y2,
              stroke: e.value.style.chart.layout.pointer.useRatingColor ? Q.value : e.value.style.chart.layout.pointer.color,
              "stroke-linecap": "round",
              "stroke-width": o.value.pointerStrokeWidth * 0.7
            }, null, 8, ma)),
            !isNaN(i.value.x2) && e.value.style.chart.layout.track.useGradient ? (s(), n("line", {
              key: 2,
              x1: i.value.x1,
              y1: i.value.y1,
              x2: i.value.x2,
              y2: i.value.y2,
              stroke: "white",
              "stroke-linecap": "round",
              "stroke-width": o.value.pointerStrokeWidth * 0.3,
              filter: `url(#blur_${C.value})`
            }, null, 8, pa)) : c("", !0)
          ])) : (s(), n("g", ka, [
            _e.value ? (s(), n("path", {
              key: 0,
              d: _e.value,
              fill: e.value.style.chart.layout.pointer.useRatingColor ? Q.value : e.value.style.chart.layout.pointer.color,
              stroke: e.value.style.chart.layout.pointer.stroke,
              "stroke-width": e.value.style.chart.layout.pointer.circle.strokeWidth,
              "stroke-linejoin": "round"
            }, null, 8, ba)) : c("", !0)
          ])),
          b("circle", {
            cx: o.value.width / 2,
            cy: v.value.base,
            fill: e.value.style.chart.layout.pointer.circle.color,
            r: o.value.pointerRadius <= 0 ? 1e-4 : o.value.pointerRadius,
            "stroke-width": e.value.style.chart.layout.pointer.circle.strokeWidth,
            stroke: e.value.style.chart.layout.pointer.circle.stroke
          }, null, 8, Sa)
        ], 64)) : c("", !0),
        e.value.style.chart.legend.show ? (s(), n("text", {
          key: 11,
          x: o.value.width / 2,
          y: v.value.ratingBase,
          "text-anchor": "middle",
          "font-size": o.value.legendFontSize,
          "font-weight": "bold",
          fill: e.value.style.chart.legend.useRatingColor ? Q.value : e.value.style.chart.legend.color
        }, T(u(ne)(
          e.value.style.chart.legend.formatter,
          m.value,
          u(ue)({
            p: e.value.style.chart.legend.prefix + (e.value.style.chart.legend.showPlusSymbol && m.value > 0 ? "+" : ""),
            v: m.value,
            s: e.value.style.chart.legend.suffix,
            r: e.value.style.chart.legend.roundingValue
          })
        )), 9, wa)) : c("", !0),
        h(t.$slots, "svg", {
          svg: {
            ...o.value,
            isPrintingImg: u(K) | u(J) | G.value,
            isPrintingSvg: U.value
          }
        }, void 0, !0)
      ], 14, Dt)),
      t.$slots.watermark ? (s(), n("div", _a, [
        h(t.$slots, "watermark", $(A({ isPrinting: u(K) || u(J) || G.value || U.value })), void 0, !0)
      ])) : c("", !0),
      b("div", {
        ref_key: "chartLegend",
        ref: he
      }, [
        h(t.$slots, "legend", { legend: F.value }, void 0, !0)
      ], 512),
      t.$slots.source ? (s(), n("div", {
        key: 5,
        ref_key: "source",
        ref: fe,
        dir: "auto"
      }, [
        h(t.$slots, "source", {}, void 0, !0)
      ], 512)) : c("", !0),
      h(t.$slots, "skeleton", {}, () => [
        u(pe) ? (s(), V(Pt, { key: 0 })) : c("", !0)
      ], !0)
    ], 46, jt));
  }
}, Ga = /* @__PURE__ */ Ut(Ca, [["__scopeId", "data-v-82d3235a"]]);
export {
  Ga as default
};
