// vue-data-ui/dist/vue-ui-treemap-BS5EilZf.js
import { defineAsyncComponent as Ie, useSlots as $l, onMounted as Oo, computed as b, ref as $, shallowRef as ye, toRefs as Tl, watch as Le, onBeforeUnmount as Sl, openBlock as h, createElementBlock as M, unref as y, normalizeStyle as me, normalizeClass as jt, createBlock as be, withCtx as S, renderSlot as x, normalizeProps as X, guardReactiveProps as j, createCommentVNode as z, createElementVNode as q, createSlots as Po, Fragment as qe, renderList as ut, withKeys as _o, withModifiers as qt, mergeProps as it, createVNode as ct, createTextVNode as dt, toDisplayString as vt, Teleport as Al, resolveDynamicComponent as Ml, nextTick as Io } from "vue";
import { j as ht, t as Lo, m as Ol, p as G, n as Pl, o as Fo, q as _l, s as Ge, k as ft, l as Ke, X as Gt, x as pt, W as gt, T as Bo, a0 as Il, a1 as Ll, ay as yt, az as Kt, U as Fl } from "./lib-GwPnJSrY.js";
import { t as Bl, u as Rl } from "./useResponsive-DfdjqQps.js";
import { u as zl } from "./useConfig-CIgyKZ6R.js";
import { u as Nl, B as Dl } from "./BaseScanner-BhhkKVoj.js";
import { u as Hl } from "./usePrinter-DiK5R6k-.js";
import { u as Zt } from "./useNestedProp-BBT0mbHh.js";
import { u as Vl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Wl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as El } from "./useChartAccessibility-9icAAmYg.js";
import Ul from "./img-BYYzY_QU.js";
import Yl from "./Title-BfyrP-jp.js";
import Xl from "./vue_ui_treemap-D9-AIkT-.js";
import jl from "./Legend-1LMMsPxL.js";
import Ro from "./BaseIcon-CNx8Tzl5.js";
import { u as ql } from "./useSvgExport-ByUukOZt.js";
import { B as Gl } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Kl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
function zo(r, p) {
  const d = r.length;
  if (d === 0)
    throw new Error(`Max aspect ratio cannot be computed: ${r} is an empty array`);
  {
    let g = 1 / 0, f = -1 / 0, w = 0;
    for (let O = 0; O < d; O += 1) {
      const T = r[O].normalizedValue;
      T < g && (g = T), T > f && (f = T), w += T;
    }
    return Math.max(
      p ** 2 * f / w ** 2,
      w ** 2 / (p ** 2 * g)
    );
  }
}
function No(r) {
  const { xOffset: p, yOffset: d, width: g, height: f } = r;
  return {
    x0: p,
    y0: d,
    x1: p + g,
    y1: d + f
  };
}
function Zl(r, p, d) {
  if (r.length === 0)
    return !0;
  {
    const g = r.concat(p), f = zo(
      r,
      d
    ), w = zo(g, d);
    return f >= w;
  }
}
function Jl(r) {
  const p = [], d = r.length;
  for (let g = 0; g < d; g += 1) {
    const f = r[g], w = f.length;
    for (let m = 0; m < w; m += 1)
      p.push(f[m]);
  }
  return p;
}
function Do(r, p) {
  const d = {
    ...p,
    children: r
  };
  return Vo(d);
}
function Ql(r) {
  return (r.x1 - r.x0) * (r.y1 - r.y0);
}
function Ho(r, p) {
  const { width: d, height: g, xOffset: f, yOffset: w } = Jt(p), m = r.length, O = r.map((k) => k.normalizedValue || 0).reduce((k, N) => k + N, 0), T = O / g, we = O / d;
  let ne = f, ce = w;
  const K = [];
  if (d >= g) {
    for (let k = 0; k < m; k += 1) {
      const N = r[k], de = ce + N.normalizedValue / T, ke = {
        x0: ne,
        y0: ce,
        x1: ne + T,
        y1: de
      }, ve = Object.assign({}, N, ke);
      ce = de, K.push(ve);
    }
    return K;
  } else {
    for (let k = 0; k < m; k += 1) {
      const N = r[k], de = ne + N.normalizedValue / we, ke = {
        x0: ne,
        y0: ce,
        x1: de,
        y1: ce + we
      }, ve = Object.assign({}, N, ke);
      ne = de, K.push(ve);
    }
    return K;
  }
}
function ea(r) {
  const p = Jt(r), d = p.width, g = p.height;
  return Math.min(d, g);
}
function ta(r, p) {
  const d = r.length, g = r.map((T) => T.value ?? 0).reduce((T, we) => T + we, 0), f = p / g, w = [];
  let m, O;
  for (let T = 0; T < d; T += 1)
    O = r[T], m = Object.assign({}, O, {
      normalizedValue: O.value * (f || 0)
    }), w.push(m);
  return w;
}
function Jt(r) {
  const { x0: p, y0: d, x1: g, y1: f } = r;
  return {
    xOffset: p,
    yOffset: d,
    width: g - p,
    height: f - d
  };
}
function oa(r, p, d, g) {
  let f = r, w = p, m = d, O = g;
  for (; ; ) {
    const T = f.length;
    if (T === 0) {
      const K = Ho(w, m);
      return O.concat(K);
    }
    const we = ea(m), ne = f[0], ce = f.slice(1, T);
    if (Zl(w, ne, we)) {
      const K = w.concat(ne);
      f = ce, w = K, m = m, O = O;
    } else {
      const K = w.length;
      let k = 0;
      for (let ve = 0; ve < K; ve += 1)
        k += w[ve].normalizedValue;
      const N = la(m, k), de = Ho(w, m), ke = O.concat(de);
      f = f, w = [], m = N, O = ke;
    }
  }
}
function Vo(r) {
  if (typeof r.children > "u" || !r.children.length)
    return [r];
  {
    const p = ta(
      r.children,
      Ql(r)
    ), d = oa(p, [], r, []), g = d.length, f = [];
    for (let m = 0; m < g; m += 1)
      f.push(Vo(d[m]));
    return Jl(f);
  }
}
function la(r, p) {
  const { width: d, height: g, xOffset: f, yOffset: w } = Jt(r);
  if (d >= g) {
    const m = p / g, O = d - m, T = {
      xOffset: f + m,
      yOffset: w,
      width: O,
      height: g
    };
    return No(T);
  } else {
    const m = p / d, O = g - m, T = {
      xOffset: f,
      yOffset: w + m,
      width: d,
      height: O
    };
    return No(T);
  }
}
const aa = ["id"], na = ["id"], sa = ["tabindex", "onClick", "onKeydown", "data-last-crumb", "onMouseenter", "onFocus"], ra = { class: "vue-ui-treemap-crumb-unit" }, ua = { class: "vue-ui-treemap-crumb-unit-label" }, ia = {
  key: 0,
  style: { width: "24px", display: "flex", "align-items": "center" }
}, ca = {
  key: 0,
  class: "vue-ui-treemap-crumb-unit-arrow"
}, da = ["xmlns", "viewBox"], va = { key: 0 }, ha = ["id"], fa = ["stop-color"], pa = ["stop-color"], ga = ["x", "y", "height", "width", "fill", "stroke", "stroke-width"], ya = ["width", "height", "x", "y"], ma = ["innerHTML"], ba = ["x", "y", "height", "width", "fill", "rx", "stroke", "stroke-width", "onClick", "onMouseenter", "onMouseleave"], wa = ["innerHTML"], ka = ["x", "y", "height", "width"], Ca = {
  style: {
    width: "100%",
    height: "100%"
  },
  class: "vue-ui-treemap-cell"
}, xa = {
  key: 5,
  class: "vue-data-ui-watermark"
}, $a = ["id"], Ta = ["onClick"], Sa = ["innerHTML"], Aa = {
  __name: "vue-ui-treemap",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "selectDatapoint", "copyAlt"],
  setup(r, { expose: p, emit: d }) {
    const g = Ie(() => import("./DataTable-Bo1pmAqG.js")), f = Ie(() => import("./PenAndPaper-DWV1FMIi.js")), w = Ie(() => import("./vue-ui-accordion-BGHTqIBt.js")), m = Ie(() => import("./Tooltip-DL3QgjZf.js")), O = Ie(() => import("./UserOptions-CrLqEimB.js")), T = Ie(() => import("./PackageVersion-DN9d0ME5.js")), we = Ie(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_treemap: ne } = zl(), { isThemeValid: ce, warnInvalidTheme: K } = Vl(), k = r, N = d, de = $l();
    Oo(() => {
      de["chart-background"] && console.warn("VueUiTreemap does not support the #chart-background slot.");
    });
    const ke = b(() => !!k.dataset && k.dataset.length), ve = b(() => /^((?!chrome|android).)*safari/i.test(navigator.userAgent)), Ce = $(ht()), mt = $(!1), bt = $(""), Oe = $(!1), Qt = $(0), F = $([]), se = ye(null), eo = ye(null), to = ye(null), oo = ye(null), lo = ye(null), ao = $(0), wt = $(0), kt = $(0), I = $([]), Ct = $(null), no = $(!1), He = $(null), xt = $(null), Ve = $(null), U = $(/* @__PURE__ */ new Map()), Ze = $(!1), Je = $(!1), t = $(At()), xe = b(() => t.value.userOptions.useCursorPointer), Wo = b(() => Lo({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#999999",
            layout: {
              labels: {
                showDefaultLabels: !1
              },
              rects: {
                stroke: "#6A6A6A"
              }
            },
            legend: {
              backgroundColor: "transparent"
            }
          }
        }
      },
      userConfig: t.value.skeletonConfig ?? {}
    })), { loading: $e, FINAL_DATASET: $t, manualLoading: Eo } = Nl({
      ...Tl(k),
      FINAL_CONFIG: t,
      prepareConfig: At,
      skeletonDataset: k.config?.skeletonDataset ?? [
        {
          name: "_",
          value: 53,
          color: "#CACACA90",
          children: [
            { name: "_", value: 21 },
            { name: "_", value: 13 },
            { name: "_", value: 8 },
            { name: "_", value: 5 },
            { name: "_", value: 3 },
            { name: "_", value: 2 },
            { name: "_", value: 1 }
          ]
        }
      ],
      skeletonConfig: Lo({
        defaultConfig: t.value,
        userConfig: Wo.value
      })
    }), { userOptionsVisible: Tt, setUserOptionsVisibility: so, keepUserOptionState: ro } = Wl({ config: t.value }), { svgRef: St } = El({ config: t.value.style.chart.title });
    function At() {
      const e = Zt({
        userConfig: k.config,
        defaultConfig: ne
      }), l = e.theme;
      if (!l) return e;
      if (!ce.value(e))
        return K(e), e;
      const o = Zt({
        userConfig: Xl[l] || k.config,
        defaultConfig: e
      }), a = Zt({
        userConfig: k.config,
        defaultConfig: o
      });
      return {
        ...a,
        customPalette: a.customPalette.length ? a.customPalette : Ol[l] || G
      };
    }
    Le(() => k.config, (e) => {
      $e.value || (t.value = At()), Tt.value = !t.value.userOptions.showOnChartHover, ao.value += 1, wt.value += 1, kt.value += 1, Z.value.showTable = t.value.table.show, Z.value.showTooltip = t.value.style.chart.tooltip.show;
    }, { deep: !0 });
    const { isPrinting: Mt, isImaging: Ot, generatePdf: uo, generateImage: io } = Hl({
      elementId: `treemap_${Ce.value}`,
      fileName: t.value.style.chart.title.text || "vue-ui-treemap",
      options: t.value.userOptions.print
    }), Uo = b(() => t.value.userOptions.show && !t.value.style.chart.title.text), Qe = b(() => Pl(t.value.customPalette)), Z = $({
      showTable: t.value.table.show,
      showTooltip: t.value.style.chart.tooltip.show
    });
    Le(t, () => {
      Z.value = {
        showTable: t.value.table.show,
        showTooltip: t.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const Pe = $({
      height: t.value.style.chart.height,
      width: t.value.style.chart.width
    });
    function Yo(e) {
      const l = se.value, o = Pe.value.height;
      if (!l || !o) return e;
      const a = l.getBoundingClientRect();
      return e * (o / a.height);
    }
    function Xo() {
      if (!Ct.value || he.value.length <= 1) return 0;
      const e = Ct.value, l = getComputedStyle(e), o = e.offsetHeight + parseFloat(l.marginTop || "0") + parseFloat(l.marginBottom || "0") + parseFloat(l.paddingTop || "0") + parseFloat(l.paddingBottom || "0");
      return Yo(o);
    }
    const B = b(() => {
      const e = t.value.style.chart.padding, l = Pe.value.width, o = Pe.value.height, a = Xo(), n = e.left, u = l - e.right, s = e.top, i = o - e.bottom - a;
      return {
        left: n,
        top: s,
        right: u,
        bottom: i,
        width: u - n,
        height: Math.max(0, i - s),
        vbWidth: l,
        vbHeight: o - a,
        offsetY: a
      };
    }), W = $($t.value), Te = $(W.value), co = ye(/* @__PURE__ */ new Map());
    Le(
      [W, () => t.value],
      () => {
        Ve.value = null, U.value = /* @__PURE__ */ new Map();
      },
      { deep: !0 }
    ), Le(
      () => F.value,
      () => {
        Ve.value = null, U.value = /* @__PURE__ */ new Map();
      },
      { deep: !0 }
    ), Le(
      [() => Pe.value.width, () => Pe.value.height],
      () => {
        U.value = /* @__PURE__ */ new Map();
      }
    );
    function vo(e) {
      Array.isArray(e) && e.forEach((l, o) => {
        l.id || (l.id = ht());
        let a = Ge(l.color) || co.value.get(l.id) || Qe.value[o] || G[o] || G[o % G.length];
        a = Ge(a), co.value.set(l.id, a), l.color = a, ho(l, a);
      });
    }
    function ho(e, l) {
      Array.isArray(e.children) && e.children.forEach((o) => {
        o.id || (o.id = ht()), o.parentId = e.id, o.color = l, ho(o, l);
      });
    }
    function jo() {
      if (!I.value.length)
        Te.value = W.value.slice();
      else {
        const e = I.value[I.value.length - 1], l = J(e);
        Te.value = l?.children?.slice() || [];
      }
    }
    Le(
      () => $t.value,
      () => {
        W.value = $t.value, vo(W.value), jo(), kt.value += 1, wt.value += 1;
      },
      { deep: !0, immediate: !0, flush: "post" }
    );
    const Se = ye(null), Fe = ye(null);
    Oo(() => {
      no.value = !0, Go();
    });
    const qo = b(() => t.value.debug);
    function Go() {
      if (Fo(k.dataset) && _l({
        componentName: "VueUiTreemap",
        type: "dataset",
        debug: qo.value
      }), vo(W.value), Fo(k.dataset) || (Eo.value = t.value.loading), t.value.responsive) {
        const e = Bl(() => {
          const { width: l, height: o } = Rl({
            chart: se.value,
            title: t.value.style.chart.title.text ? eo.value : null,
            legend: t.value.style.chart.legend.show ? to.value : null,
            source: oo.value,
            noTitle: lo.value
          });
          requestAnimationFrame(() => {
            Pe.value.width = l, Pe.value.height = o - 12;
          });
        });
        Se.value && (Fe.value && Se.value.unobserve(Fe.value), Se.value.disconnect()), Se.value = new ResizeObserver(e), Fe.value = se.value.parentNode, Se.value.observe(Fe.value);
      }
    }
    Sl(() => {
      Se.value && (Fe.value && Se.value.unobserve(Fe.value), Se.value.disconnect());
    });
    const We = b(() => Te.value.map((e, l) => ({
      ...e,
      color: Ge(e.color) || Qe.value[l] || G[l] || G[l % G.length]
    })).filter((e) => !F.value.includes(e.id))), Be = b(() => W.value.filter((e) => !F.value.includes(e.id)).map((e) => e.value || 0).reduce((e, l) => e + l, 0)), Pt = b({
      get() {
        let e = [...We.value];
        return t.value.style.chart.layout.sorted && (e = [...We.value].sort((l, o) => o.value - l.value)), e.map((l) => ({
          ...l
        }));
      },
      set(e) {
        return e;
      }
    });
    function fo(e, l) {
      return e.value / l;
    }
    function Ko(e, l, o) {
      const a = t.value.style.chart.layout.rects.colorRatio - fo(l, o);
      return pt(e, a < 0 ? 0 : a);
    }
    function po(e) {
      return Array.isArray(e.children) && e.children.length > 0;
    }
    function Zo(e) {
      return [...e].sort((l, o) => {
        const a = po(l), n = po(o);
        if (a !== n)
          return a - n;
        const u = Number(l.value) || 0;
        return (Number(o.value) || 0) - u;
      });
    }
    function go(e, l, o, a, n) {
      return Zo(e).map((s, i) => {
        const c = Ko(
          Ge(l) || Qe.value[i] || G[i] || G[i % G.length],
          s,
          a
        ), v = fo(s, a), A = n ?? s.parentId ?? s.id;
        return {
          ...s,
          color: c,
          proportion: v,
          parentName: o,
          rootId: A,
          children: Array.isArray(s.children) && s.children.length ? go(
            s.children,
            c,
            s.name,
            a,
            A
          ) : void 0
        };
      });
    }
    const yo = b(() => {
      const e = Pt.value.map((n) => Number(n.value) || 0).reduce((n, u) => n + u, 0) || 1, l = Pt.value.map((n) => {
        const u = n.children ? n.children.reduce((Y, Ae) => Y + (Number(Ae.value) || 0), 0) : n.value, { children: s, parentId: i, id: c, value: v, name: A, color: R, ...C } = n, D = n.id || ht();
        return {
          ...C,
          id: D,
          name: n.name,
          value: n.value,
          color: n.color,
          proportion: (Number(n.value) || 0) / e,
          children: Array.isArray(n.children) && n.children.length ? go(
            n.children,
            n.color,
            n.name,
            u || 1
          ) : void 0
        };
      });
      if (I.value.length === 0) {
        if (!Ve.value) {
          const i = { x0: 0, y0: 0, x1: 1, y1: 1 };
          Ve.value = Do(l, i);
        }
        const n = {
          x0: B.value.left,
          y0: B.value.top,
          x1: B.value.left + B.value.width,
          y1: B.value.top + B.value.height
        }, u = n.x1 - n.x0, s = n.y1 - n.y0;
        return Ve.value.map((i) => ({
          ...i,
          x0: n.x0 + i.x0 * u,
          x1: n.x0 + i.x1 * u,
          y0: n.y0 + i.y0 * s,
          y1: n.y0 + i.y1 * s
        }));
      }
      const a = {
        x0: B.value.left,
        y0: B.value.top,
        x1: B.value.left + B.value.width,
        y1: B.value.top + B.value.height
      };
      return Do(l, a);
    });
    function Jo(e) {
      const l = I.value.length ? I.value[I.value.length - 1] : null, o = /* @__PURE__ */ new Map(), a = /* @__PURE__ */ new Map(), n = /* @__PURE__ */ new Map(), u = /* @__PURE__ */ new Map();
      for (const i of e) {
        let c = ol(i.id);
        if (l) {
          const v = c.indexOf(l);
          if (v >= 0)
            c = c.slice(v);
          else
            continue;
        }
        c = c.filter((v) => {
          const A = J(v);
          return A && Array.isArray(A.children) && A.children.length;
        }), c.length && (c.forEach((v, A) => {
          const R = J(v);
          if (!R) return;
          let C = o.get(v);
          C ? (C.x0 = Math.min(C.x0, i.x0), C.y0 = Math.min(C.y0, i.y0), C.x1 = Math.max(C.x1, i.x1), C.y1 = Math.max(C.y1, i.y1), C.depth = Math.min(C.depth, A)) : (C = {
            id: v,
            name: R.name,
            value: R.value,
            color: R.color,
            x0: i.x0,
            y0: i.y0,
            x1: i.x1,
            y1: i.y1,
            depth: A
            // depth from visible root
          }, o.set(v, C));
          let D = n.get(v);
          D || (D = /* @__PURE__ */ new Set(), n.set(v, D)), D.add(i.id);
          const Y = u.get(v);
          (Y == null || A < Y) && u.set(v, A);
        }), a.set(i.id, c));
      }
      return {
        wrappers: Array.from(o.values()).map((i) => ({
          ...i,
          depth: u.get(i.id) ?? i.depth ?? 0,
          childrenIds: Array.from(n.get(i.id) || [])
        })),
        ancestorsByLeaf: a
      };
    }
    const _t = b(() => Jo(yo.value)), Qo = b(() => {
      const e = _t.value.wrappers;
      return e.length ? e.filter((l) => el(l)) : [];
    });
    function It(e) {
      return Re(e) * 1.2;
    }
    function el(e) {
      const l = It(e);
      return ee(e) < l * 1.1 ? !1 : (e.labelY0 = e.y0, !0);
    }
    const _e = b(() => tl({
      rects: yo.value,
      wrappers: _t.value.wrappers,
      ancestorsByLeaf: _t.value.ancestorsByLeaf
    }));
    function tl({ rects: e, wrappers: l, ancestorsByLeaf: o }) {
      if (!l.length || !o) return e;
      const a = new Map(l.map((n) => [n.id, n]));
      return e.map((n) => {
        const u = o.get(n.id);
        if (!u || !u.length) return n;
        let s = { ...n };
        for (const i of u) {
          const c = a.get(i);
          if (!c) continue;
          const v = ee(c), A = It(c);
          if (v < A * 1.1) continue;
          const R = Re(c), C = R * 1.2, D = R * 0.5, Y = R * 0.5, Ae = C * 0.5, Ye = te(c) / 3, Xe = Math.min(D, Ye), Me = v * 0.25, zt = Math.min(Y, Me), Dt = (typeof c.labelY0 == "number" ? c.labelY0 : c.y0) + A, re = Math.max(s.x0, c.x0 + Xe), pe = Math.min(s.x1, c.x1 - Xe), ge = Math.max(s.y0, Dt + Ae), Ht = Math.min(s.y1, c.y1 - zt);
          if (pe <= re || Ht <= ge)
            break;
          s = {
            ...s,
            x0: re,
            y0: ge,
            x1: pe,
            y1: Ht
          };
        }
        return s;
      });
    }
    function ee({ y0: e, y1: l }) {
      return l - e <= 0 ? 1e-4 : l - e;
    }
    function te({ x0: e, x1: l }) {
      return l - e <= 0 ? 1e-4 : l - e;
    }
    function Re(e) {
      const l = t.value.style.chart.layout.labels, o = l.fontSize, a = l.minFontSize, n = l.fontSize * 2, u = 0.5, s = 0.6, i = 1, c = 0.9;
      let v = e.proportion;
      if (!(typeof v == "number" && isFinite(v))) {
        const Y = Math.max(1e-6, te(e) * ee(e)), Ae = Math.max(1e-6, B.value.width * B.value.height);
        v = Y / Ae;
      }
      const A = Math.pow(Math.min(1, Math.max(0, v)), u), R = s + (i - s) * A;
      let C = o * R;
      const D = Math.max(1e-4, Math.min(te(e), ee(e)));
      return C = Math.min(C, D * c), C = Math.max(a, Math.min(n, C)), C;
    }
    function mo(e) {
      Oe.value = e, Qt.value += 1;
    }
    const ze = b(() => ({
      startX: 0,
      startY: 0,
      width: B.value.vbWidth,
      height: B.value.vbHeight
    }));
    function J(e, l = W.value) {
      for (const o of l) {
        if (o.id === e)
          return o;
        if (o.children) {
          const a = J(e, o.children);
          if (a)
            return a;
        }
      }
      return null;
    }
    function ol(e) {
      const l = [];
      let o = J(e);
      for (; o && o.parentId; ) {
        const a = J(o.parentId);
        if (!a) break;
        l.unshift(a.id), o = a;
      }
      return l;
    }
    const Ne = b(() => I.value.length > 0);
    function et(e, l) {
      if (!e) {
        Te.value = W.value.slice(), N("selectDatapoint", void 0), I.value = [];
        return;
      }
      t.value.events.datapointClick && t.value.events.datapointClick({ datapoint: e, seriesIndex: l });
      const o = J(e.id);
      if (o && o.children?.length)
        I.value.push(o.id), Te.value = o.children.slice(), N("selectDatapoint", e);
      else if (e.parentId) {
        I.value.push(e.parentId);
        const a = J(e.parentId);
        Te.value = a.children.slice(), N("selectDatapoint", e);
      } else if (I.value.length > 0) {
        I.value.pop();
        const a = I.value[I.value.length - 1];
        if (a) {
          const n = J(a);
          Te.value = n.children.slice();
        } else
          Te.value = W.value.slice(), I.value = [], N("selectDatapoint", void 0);
      }
    }
    function ll(e) {
      return !!(J(e.id)?.children?.length || e.parentId);
    }
    const he = b(() => {
      const e = [
        { id: null, label: "All" }
      ];
      if (I.value.length > 0) {
        let l = J(I.value[I.value.length - 1]);
        const o = [];
        for (; l; )
          o.unshift(l), l = l.parentId ? J(l.parentId) : null;
        for (const a of o)
          e.push({
            id: a.id,
            label: a.name,
            node: a
          });
      }
      return e;
    }), E = ye(null), De = b(() => W.value.map((e, l) => ({
      ...e,
      color: Ge(e.color) || Qe.value[l] || G[l] || G[l % G.length],
      shape: "square"
    })).sort((e, l) => l.value - e.value).map((e, l) => {
      const o = e.value / W.value.map((a) => a.value).reduce((a, n) => a + n, 0);
      return {
        ...e,
        proportion: o,
        opacity: F.value.includes(e.id) ? 0.5 : 1,
        display: `${e.name}${t.value.style.chart.legend.showPercentage || t.value.style.chart.legend.showValue ? ": " : ""}${t.value.style.chart.legend.showValue ? ft(
          t.value.style.chart.layout.labels.formatter,
          e.value,
          Ke({
            p: t.value.style.chart.layout.labels.prefix,
            v: e.value,
            s: t.value.style.chart.layout.labels.suffix,
            r: t.value.style.chart.legend.roundingValue
          }),
          { datapoint: e }
        ) : ""}${t.value.style.chart.legend.showPercentage ? F.value.includes(e.id) ? `${t.value.style.chart.legend.showValue ? " (" : ""}- %${t.value.style.chart.legend.showValue ? ")" : ""}` : `${t.value.style.chart.legend.showValue ? " (" : ""}${isNaN(e.value / Be.value) ? "-" : (e.value / Be.value * 100).toFixed(t.value.style.chart.legend.roundingPercentage)}%${t.value.style.chart.legend.showValue ? ")" : ""}` : ""}`
      };
    })), al = b(() => ({
      cy: "treemap-div-legend",
      backgroundColor: t.value.style.chart.legend.backgroundColor,
      color: t.value.style.chart.legend.color,
      fontSize: t.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: t.value.style.chart.legend.bold ? "bold" : ""
    })), Ee = b(() => F.value.length === De.value.length);
    function nl() {
      F.value.length ? F.value = [] : De.value.forEach((e) => {
        F.value.push(e.id);
      });
    }
    function tt(e) {
      E.value = null, F.value.includes(e.id) ? F.value = F.value.filter((l) => l !== e.id) : F.value.length < k.dataset.length - 1 && F.value.push(e.id), N("selectLegend", Pt.value);
    }
    function bo(e) {
      if (!W.value.length)
        return t.value.debug && console.warn("VueUiTreemap - There are no series to show."), null;
      const l = W.value.find((o) => o.name === e);
      return l || (t.value.debug && console.warn(`VueUiTreemap - Series name not found "${e}"`), null);
    }
    function sl(e) {
      const l = bo(e);
      l !== null && F.value.includes(l.id) && tt({ id: l.id });
    }
    function rl(e) {
      const l = bo(e);
      l !== null && (F.value.includes(l.id) || tt({ id: l.id }));
    }
    function ul({ datapoint: e, seriesIndex: l }) {
      E.value = null, mt.value = !1, t.value.events.datapointLeave && t.value.events.datapointLeave({ datapoint: e, seriesIndex: l });
    }
    const Lt = $(null);
    function il({ datapoint: e, seriesIndex: l }) {
      if (Ee.value) return;
      t.value.events.datapointEnter && t.value.events.datapointEnter({ datapoint: e, seriesIndex: l }), E.value = e, Lt.value = { datapoint: e, seriesIndex: l, config: t.value, series: We.value };
      const o = t.value.style.chart.tooltip.customFormat;
      if (Bo(o) && Fl(() => o({
        seriesIndex: l,
        datapoint: e,
        series: We.value,
        config: t.value
      })))
        bt.value = o({
          seriesIndex: l,
          datapoint: e,
          series: We.value,
          config: t.value
        });
      else {
        let a = "";
        a += `<div style="width:100%;text-align:center;border-bottom:1px solid ${t.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${e.name}</div>`, a += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 12 12" height="14" width="14"><rect x="0" y="0" height="12" width="12" stroke="none" fill="${e.color}"/></svg>`, a += `<b>${ft(
          t.value.style.chart.layout.labels.formatter,
          e.value,
          Ke({
            p: t.value.style.chart.layout.labels.prefix,
            v: e.value,
            s: t.value.style.chart.layout.labels.suffix,
            r: t.value.style.chart.tooltip.roundingValue
          }),
          { datapoint: e, seriesIndex: l }
        )}</b>`, bt.value = `<div>${a}</div>`;
      }
      mt.value = !0;
    }
    const fe = b(() => {
      const e = _e.value.map((o) => ({
        name: o.name,
        color: o.color
      })), l = _e.value.map((o) => o.value);
      return { head: e, body: l };
    });
    function Ft(e = null) {
      Io(() => {
        const l = fe.value.head.map((n, u) => [[
          n.name
        ], [fe.value.body[u]], [isNaN(fe.value.body[u] / Be.value) ? "-" : fe.value.body[u] / Be.value * 100]]), o = [[t.value.style.chart.title.text], [t.value.style.chart.title.subtitle.text], [[""], ["val"], ["%"]]].concat(l), a = Il(o);
        e ? e(a) : Ll({ csvContent: a, title: t.value.style.chart.title.text || "vue-ui-treemap" });
      });
    }
    const ot = b(() => {
      const e = [
        t.value.table.columnNames.series,
        t.value.table.columnNames.value,
        t.value.table.columnNames.percentage
      ], l = fe.value.head.map((n, u) => {
        const s = ft(
          t.value.style.chart.layout.labels.formatter,
          fe.value.body[u],
          Ke({
            p: t.value.style.chart.layout.labels.prefix,
            v: fe.value.body[u],
            s: t.value.style.chart.layout.labels.suffix,
            r: t.value.table.td.roundingValue
          })
        );
        return [
          {
            color: n.color,
            name: n.name,
            shape: "square"
          },
          s,
          isNaN(fe.value.body[u] / Be.value) ? "-" : Ke({
            v: fe.value.body[u] / Be.value * 100,
            s: "%",
            r: t.value.table.td.roundingPercentage
          })
        ];
      }), o = {
        th: {
          backgroundColor: t.value.table.th.backgroundColor,
          color: t.value.table.th.color,
          outline: t.value.table.th.outline
        },
        td: {
          backgroundColor: t.value.table.td.backgroundColor,
          color: t.value.table.td.color,
          outline: t.value.table.td.outline
        },
        breakpoint: t.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          t.value.table.columnNames.series,
          t.value.table.columnNames.value
        ],
        head: e,
        body: l,
        config: o
      };
    });
    function cl() {
      return _e.value;
    }
    function wo() {
      Z.value.showTable = !Z.value.showTable;
    }
    function ko() {
      Z.value.showTooltip = !Z.value.showTooltip;
    }
    const lt = $(!1);
    function Bt() {
      lt.value = !lt.value;
    }
    const at = $(null);
    function Co(e) {
      at.value = e;
    }
    function xo() {
      at.value = null;
    }
    async function dl({ scale: e = 2 } = {}) {
      if (!se.value) return;
      const { width: l, height: o } = se.value.getBoundingClientRect(), a = l / o, { imageUri: n, base64: u } = await Ul({ domElement: se.value, base64: !0, img: !0, scale: e });
      return {
        imageUri: n,
        base64: u,
        title: t.value.style.chart.title.text,
        width: l,
        height: o,
        aspectRatio: a
      };
    }
    const Ue = b(() => {
      const e = t.value.table.useDialog && !t.value.table.show, l = Z.value.showTable;
      return {
        component: e ? we : w,
        title: `${t.value.style.chart.title.text}${t.value.style.chart.title.subtitle.text ? `: ${t.value.style.chart.title.subtitle.text}` : ""}`,
        props: e ? {
          backgroundColor: t.value.table.th.backgroundColor,
          color: t.value.table.th.color,
          headerColor: t.value.table.th.color,
          headerBg: t.value.table.th.backgroundColor,
          isFullscreen: Oe.value,
          fullscreenParent: se.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: xe.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: t.value.style.chart.backgroundColor,
              color: t.value.style.chart.color
            },
            head: {
              backgroundColor: t.value.style.chart.backgroundColor,
              color: t.value.style.chart.color
            }
          }
        }
      };
    });
    Le(() => Z.value.showTable, (e) => {
      t.value.table.show || (e && t.value.table.useDialog && He.value ? He.value.open() : "close" in He.value && He.value.close());
    });
    function $o() {
      Z.value.showTable = !1, xt.value && xt.value.setTableIconState(!1);
    }
    const vl = b(() => De.value.map((e) => ({
      ...e,
      name: e.display
    }))), hl = b(() => t.value.style.chart.backgroundColor), fl = b(() => t.value.style.chart.legend), pl = b(() => t.value.style.chart.title), { exportSvg: gl, getSvg: yl } = ql({
      svg: St,
      title: pl,
      legend: fl,
      legendItems: vl,
      backgroundColor: hl
    });
    async function To({ isCb: e }) {
      Je.value = !0, await Io();
      try {
        if (e) {
          const { blob: l, url: o, text: a, dataUrl: n } = await yl();
          await Promise.resolve(t.value.userOptions.callbacks.svg({ blob: l, url: o, text: a, dataUrl: n }));
        } else
          await Promise.resolve(gl());
      } finally {
        Je.value = !1;
      }
    }
    function ml(e) {
      if (e?.stage === "start") {
        Ze.value = !0;
        return;
      }
      if (e?.stage === "end") {
        Ze.value = !1;
        return;
      }
      io();
    }
    function bl(e) {
      const o = Re(e) * 0.6;
      return Math.max(0, te(e) - 2 * o);
    }
    function So({ rect: e, seriesIndex: l, isTitle: o = !1 }) {
      const a = t.value.style.chart.layout.labels.name.show, n = t.value.style.chart.layout.labels.value.show;
      if (!a && !n) return "";
      const u = I.value.length === 0, s = u ? `${e.id}:${o ? "title" : "rect"}` : null;
      if (u && U.value.has(s))
        return U.value.get(s);
      const i = Math.max(2, Re(e) / 3), c = Math.max(8, Re(e)), v = c * 1.2, A = Math.max(0, te(e) - i * 2), R = Math.max(0, ee(e) - i * 2);
      if (A <= 2 || R <= c * 0.8)
        return u && s && U.value.set(s, ""), "";
      const C = t.value.style.fontFamily, D = t.value.style.chart.layout.labels.name.bold ? "600" : "400", Y = t.value.style.chart.layout.labels.value.bold ? "600" : "400", Ae = o ? t.value.style.chart.layout.rects.group.label.adaptColorToBackground ? t.value.style.chart.layout.rects.group.useSeriesBackgroundColor ? gt(pt(e.color, 0.4)) : gt(t.value.style.chart.backgroundColor) : t.value.style.chart.layout.rects.group.label.color : gt(e.color), Rt = a ? String(e.name ?? "") : "", Ye = n ? ft(
        t.value.style.chart.layout.labels.formatter,
        e.value ?? 0,
        Ke({
          p: t.value.style.chart.layout.labels.prefix,
          v: e.value ?? 0,
          s: t.value.style.chart.layout.labels.suffix,
          r: t.value.style.chart.layout.labels.rounding
        }),
        { datapoint: e, seriesIndex: l }
      ) : "", Xe = o ? bl(e) : A;
      function Me(L, oe = c, je = "400") {
        const Q = document.createElementNS(Gt, "svg");
        Q.setAttribute("width", "0"), Q.setAttribute("height", "0"), Q.style.position = "absolute", Q.style.visibility = "hidden", Q.style.pointerEvents = "none";
        const le = document.createElementNS(Gt, "text");
        le.setAttribute("font-size", String(oe)), le.setAttribute("font-family", C), le.setAttribute("font-weight", String(je)), le.textContent = L || "", Q.appendChild(le), document.body.appendChild(Q);
        const nt = le.getComputedTextLength();
        return document.body.removeChild(Q), nt;
      }
      if (o) {
        let L = "";
        if (a && (L = Rt), n && (L = L ? `${L} (${Ye})` : Ye), !L)
          return u && s && U.value.set(s, ""), "";
        let oe = L;
        if (Me(oe, c, D) > Xe) {
          let V = oe;
          for (; V.length && Me(V + "…", c, D) > Xe; )
            V = V.slice(0, -1);
          oe = V ? V + "…" : "";
        }
        if (!oe)
          return u && s && U.value.set(s, ""), "";
        const je = It(e), Q = typeof e.depth == "number" ? e.depth : 0, le = je * 1.4, nt = typeof e.labelY0 == "number" ? e.labelY0 : e.y0 + Q * le, Et = e.x0 + i + c / 2, H = nt + i + c, ae = `<text
            x="${Et}"
            y="${H}"
            font-size="${c}"
            font-family="${yt(C)}"
            font-weight="${D}"
            fill="${yt(Ae)}"
            paint-order="stroke"
            stroke="transparent"
            stroke-width="0"
            style="transition: all 0.2s ease-in-out;"
        >
            ${Kt(oe)}
        </text>`;
        return u && s && U.value.set(s, ae), ae;
      }
      const zt = n ? 1 : 0, Nt = Math.max(0, Math.floor(R / v) - zt);
      if (Nt <= 0 && !n)
        return u && s && U.value.set(s, ""), "";
      function Dt(L, oe, je, Q, le, nt) {
        const Et = String(L).split(/\s+/).filter(Boolean), H = [];
        let ae = "";
        function V(_) {
          return Me(_, je, Q) <= oe;
        }
        function Cl(_) {
          if (V(_)) return [_];
          const ue = [];
          let P = 0;
          for (; P < _.length; ) {
            let ie = 1, Yt = _.length - P, Xt = 1;
            for (; ie <= Yt; ) {
              const rt = ie + Yt >> 1, xl = _.slice(P, P + rt);
              V(xl) ? (Xt = rt, ie = rt + 1) : Yt = rt - 1;
            }
            ue.push(_.slice(P, P + Xt)), P += Xt;
          }
          return ue;
        }
        const st = Et.flatMap((_) => Cl(_));
        let Ut = !1;
        for (let _ = 0; _ < st.length; _ += 1) {
          const ue = st[_], P = ae ? ae + " " + ue : ue;
          if (V(P))
            ae = P;
          else {
            if (ae)
              H.push(ae);
            else {
              let ie = ue;
              for (; ie.length && !V(ie); ) ie = ie.slice(0, -1);
              ie && H.push(ie);
            }
            if (ae = "", H.length === le) {
              Ut = !0;
              break;
            }
          }
          _ === st.length - 1 && ae && H.length < le && H.push(ae);
        }
        if (H.length && H.join(" ") !== st.join(" ") && (Ut = !0), Ut && H.length > 0) {
          let _ = H[H.length - 1];
          if (!V(_)) {
            let P = _;
            for (; P.length && !V(P); ) P = P.slice(0, -1);
            _ = P;
          }
          let ue = _ + "…";
          if (!V(ue)) {
            let P = _;
            for (; P.length && !V(P + "…"); )
              P = P.slice(0, -1);
            ue = P + "…";
          }
          H[H.length - 1] = ue;
        }
        return H;
      }
      const re = a ? Dt(
        Rt,
        A,
        c,
        D,
        Nt
      ) : [];
      let pe = c, ge = String(Ye);
      if (n) {
        for (; Me(ge, pe, Y) > A && pe > Math.max(8, c * 0.75); )
          pe -= 1;
        if (Me(ge, pe, Y) > A) {
          let L = ge;
          for (; L.length && Me(L + "…", pe, Y) > A; )
            L = L.slice(0, -1);
          if (ge = L ? L + "…" : "", !ge.length)
            return u && s && U.value.set(s, ""), "";
        }
      }
      if (re.length * v + (n ? v : 0) > R) {
        for (; re.length && re.length * v + (n ? v : 0) > R; )
          re.pop();
        if (!re.length && (!n || v > R))
          return u && s && U.value.set(s, ""), "";
      }
      const Vt = e.x0 + i, kl = e.y0 + i + c, Wt = [];
      a && re.forEach((L, oe) => {
        Wt.push(`
            <tspan 
                x="${Vt}" 
                dy="${oe === 0 ? 0 : v}" 
                font-weight="${D}"
            >
                ${Kt(L)}
            </tspan>`);
      }), n && Wt.push(`
        <tspan 
            x="${Vt}" 
            dy="${re.length ? v : 0}"
            font-weight="${Y}" 
            font-size="${pe}"
        >
            ${Kt(ge)}
        </tspan>`);
      const Mo = `<text 
        x="${Vt}" 
        y="${kl}" 
        font-size="${c}" 
        font-family="${yt(C)}" 
        fill="${yt(Ae)}" 
        paint-order="stroke" 
        stroke="transparent" 
        stroke-width="0"
        style="transition: all 0.2s ease-in-out;"
    >
        ${Wt.join("")}
    </text>`;
      return u && s && U.value.set(s, Mo), Mo;
    }
    function wl(e) {
      const l = t.value.style.chart.layout.rects.borderRadius, o = te(e), a = ee(e);
      return Math.min(l, Math.min(o, a) / 6);
    }
    async function Ao() {
      if (N("copyAlt", {
        config: t.value,
        dataset: _e.value
      }), !t.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(t.value.userOptions.callbacks.altCopy({
        config: t.value,
        dataset: _e.value
      }));
    }
    return p({
      getData: cl,
      getImage: dl,
      generateCsv: Ft,
      generateImage: io,
      generateSvg: To,
      generatePdf: uo,
      hideSeries: rl,
      showSeries: sl,
      toggleTable: wo,
      toggleTooltip: ko,
      toggleAnnotator: Bt,
      toggleFullscreen: mo,
      copyAlt: Ao
    }), (e, l) => (h(), M("div", {
      ref_key: "treemapChart",
      ref: se,
      class: jt(`vue-data-ui-component vue-ui-treemap ${Oe.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${t.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: me(`font-family:${t.value.style.fontFamily};width:100%; ${t.value.responsive ? "height: 100%;" : ""} text-align:center;background:${t.value.style.chart.backgroundColor}`),
      id: `treemap_${Ce.value}`,
      onMouseenter: l[2] || (l[2] = () => y(so)(!0)),
      onMouseleave: l[3] || (l[3] = () => y(so)(!1))
    }, [
      t.value.userOptions.buttons.annotator ? (h(), be(y(f), {
        key: 0,
        svgRef: y(St),
        backgroundColor: t.value.style.chart.backgroundColor,
        color: t.value.style.chart.color,
        active: lt.value,
        isCursorPointer: xe.value,
        onClose: Bt
      }, {
        "annotator-action-close": S(() => [
          x(e.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": S(({ color: o }) => [
          x(e.$slots, "annotator-action-color", X(j({ color: o })), void 0, !0)
        ]),
        "annotator-action-draw": S(({ mode: o }) => [
          x(e.$slots, "annotator-action-draw", X(j({ mode: o })), void 0, !0)
        ]),
        "annotator-action-undo": S(({ disabled: o }) => [
          x(e.$slots, "annotator-action-undo", X(j({ disabled: o })), void 0, !0)
        ]),
        "annotator-action-redo": S(({ disabled: o }) => [
          x(e.$slots, "annotator-action-redo", X(j({ disabled: o })), void 0, !0)
        ]),
        "annotator-action-delete": S(({ disabled: o }) => [
          x(e.$slots, "annotator-action-delete", X(j({ disabled: o })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : z("", !0),
      Uo.value ? (h(), M("div", {
        key: 1,
        ref_key: "noTitle",
        ref: lo,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : z("", !0),
      t.value.style.chart.title.text ? (h(), M("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: eo,
        style: me(`width:100%;background:${t.value.style.chart.backgroundColor};padding-bottom:6px`)
      }, [
        (h(), be(Yl, {
          key: `title_${ao.value}`,
          config: {
            title: {
              cy: "treemap-div-title",
              ...t.value.style.chart.title
            },
            subtitle: {
              cy: "treemap-div-subtitle",
              ...t.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 4)) : z("", !0),
      q("div", {
        id: `legend-top-${Ce.value}`
      }, null, 8, na),
      t.value.userOptions.show && ke.value && (y(ro) || y(Tt)) ? (h(), be(y(O), {
        ref_key: "userOptionsRef",
        ref: xt,
        key: `user_option_${Qt.value}`,
        backgroundColor: t.value.style.chart.backgroundColor,
        color: t.value.style.chart.color,
        isPrinting: y(Mt),
        isImaging: y(Ot),
        uid: Ce.value,
        hasTooltip: t.value.userOptions.buttons.tooltip && t.value.style.chart.tooltip.show,
        hasPdf: t.value.userOptions.buttons.pdf,
        hasXls: t.value.userOptions.buttons.csv,
        hasImg: t.value.userOptions.buttons.img,
        hasSvg: t.value.userOptions.buttons.svg,
        hasTable: t.value.userOptions.buttons.table,
        hasFullscreen: t.value.userOptions.buttons.fullscreen,
        hasAltCopy: t.value.userOptions.buttons.altCopy,
        isFullscreen: Oe.value,
        isTooltip: Z.value.showTooltip,
        titles: { ...t.value.userOptions.buttonTitles },
        chartElement: se.value,
        position: t.value.userOptions.position,
        hasAnnotator: t.value.userOptions.buttons.annotator,
        isAnnotation: lt.value,
        callbacks: t.value.userOptions.callbacks,
        printScale: t.value.userOptions.print.scale,
        tableDialog: t.value.table.useDialog,
        isCursorPointer: xe.value,
        onToggleFullscreen: mo,
        onGeneratePdf: y(uo),
        onGenerateCsv: Ft,
        onGenerateImage: ml,
        onGenerateSvg: To,
        onToggleTable: wo,
        onToggleTooltip: ko,
        onToggleAnnotator: Bt,
        onCopyAlt: Ao,
        style: me({
          visibility: y(ro) ? y(Tt) ? "visible" : "hidden" : "visible"
        })
      }, Po({ _: 2 }, [
        e.$slots.menuIcon ? {
          name: "menuIcon",
          fn: S(({ isOpen: o, color: a }) => [
            x(e.$slots, "menuIcon", X(j({ isOpen: o, color: a })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        e.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: S(() => [
            x(e.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        e.$slots.optionPdf ? {
          name: "optionPdf",
          fn: S(() => [
            x(e.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        e.$slots.optionCsv ? {
          name: "optionCsv",
          fn: S(() => [
            x(e.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        e.$slots.optionImg ? {
          name: "optionImg",
          fn: S(() => [
            x(e.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        e.$slots.optionSvg ? {
          name: "optionSvg",
          fn: S(() => [
            x(e.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        e.$slots.optionTable ? {
          name: "optionTable",
          fn: S(() => [
            x(e.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        e.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: S(({ toggleFullscreen: o, isFullscreen: a }) => [
            x(e.$slots, "optionFullscreen", X(j({ toggleFullscreen: o, isFullscreen: a })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        e.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: S(({ toggleAnnotator: o, isAnnotator: a }) => [
            x(e.$slots, "optionAnnotator", X(j({ toggleAnnotator: o, isAnnotator: a })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        e.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: S(({ altCopy: o }) => [
            x(e.$slots, "optionAltCopy", X(j({ altCopy: o })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : z("", !0),
      he.value.length > 1 ? (h(), M("nav", {
        key: 4,
        class: "vue-ui-treemap-breadcrumbs",
        "data-dom-to-png-ignore": "",
        ref_key: "breadcrumbsNav",
        ref: Ct
      }, [
        (h(!0), M(qe, null, ut(he.value, (o, a) => (h(), M("span", {
          role: "button",
          tabindex: a < he.value.length - 1 ? 0 : void 0,
          key: o.id || "root",
          onClick: (n) => a === he.value.length - 1 ? () => {
          } : et(o.node),
          onKeydown: [
            _o(qt((n) => a === he.value.length - 1 ? void 0 : et(o.node), ["prevent"]), ["enter"]),
            _o(qt((n) => a === he.value.length - 1 ? void 0 : et(o.node), ["prevent"]), ["space"])
          ],
          class: "vue-ui-treemap-crumb",
          "data-last-crumb": a === he.value.length - 1,
          style: me({
            color: t.value.style.chart.color,
            cursor: xe.value ? "pointer" : "default"
          }),
          onMouseenter: (n) => Co(a),
          onMouseleave: xo,
          onFocus: (n) => Co(a),
          onBlur: xo
        }, [
          q("span", ra, [
            q("span", ua, [
              x(e.$slots, "breadcrumb-label", it({ ref_for: !0 }, { crumb: o, isRoot: a === 0, isFocus: at.value === a }), () => [
                a === 0 ? (h(), M("div", ia, [
                  ct(Ro, {
                    name: at.value === 0 ? "homeFilled" : "home",
                    stroke: t.value.style.chart.color
                  }, null, 8, ["name", "stroke"])
                ])) : (h(), M(qe, { key: 1 }, [
                  dt(vt(o.label), 1)
                ], 64))
              ], !0)
            ]),
            a < he.value.length - 1 ? (h(), M("span", ca, [
              x(e.$slots, "breadcrumb-arrow", {}, () => [
                l[4] || (l[4] = dt(" › ", -1))
              ], !0)
            ])) : z("", !0)
          ])
        ], 44, sa))), 128))
      ], 512)) : z("", !0),
      (h(), M("svg", {
        ref_key: "svgRef",
        ref: St,
        xmlns: y(Gt),
        class: jt({ "vue-data-ui-fullscreen--on": Oe.value, "vue-data-ui-fulscreen--off": !Oe.value, "vue-data-ui-zoom-plus": !Ne.value, "vue-data-ui-zoom-minus": Ne.value, loading: y($e) }),
        viewBox: `${ze.value.startX} ${ze.value.startY} ${ze.value.width <= 0 ? 10 : ze.value.width} ${ze.value.height <= 0 ? 10 : ze.value.height}`,
        style: me(`max-width:100%; overflow: hidden; background:transparent;color:${t.value.style.chart.color}`)
      }, [
        ct(y(T)),
        (h(!0), M(qe, null, ut(_e.value, (o, a) => (h(), M("g", {
          key: `tgrad_${o.id}`
        }, [
          t.value.style.chart.layout.rects.gradient.show ? (h(), M("defs", va, [
            q("radialGradient", {
              id: `tgrad_${o.id}`,
              gradientTransform: "translate(-1, -1.000001) scale(2, 2)"
            }, [
              q("stop", {
                offset: "18%",
                "stop-color": o.color
              }, null, 8, fa),
              q("stop", {
                offset: "100%",
                "stop-color": y(pt)(o.color, t.value.style.chart.layout.rects.gradient.intensity / 100)
              }, null, 8, pa)
            ], 8, ha)
          ])) : z("", !0)
        ]))), 128)),
        (h(!0), M(qe, null, ut(Qo.value, (o, a) => (h(), M("g", {
          key: `parent_${o.id}`,
          class: "vue-ui-treemap-parent-wrapper"
        }, [
          q("rect", {
            x: o.x0,
            y: o.y0,
            height: ee(o),
            width: te(o),
            fill: t.value.style.chart.layout.rects.group.useSeriesBackgroundColor ? y(pt)(o.color ?? t.value.style.chart.backgroundColor, t.value.style.chart.layout.rects.group.backgroundLighterRatio) : t.value.style.chart.backgroundColor,
            rx: 0,
            stroke: t.value.style.chart.layout.rects.group.stroke,
            "stroke-width": t.value.style.chart.layout.rects.group.strokeWidth,
            class: "vue-data-ui-cursor-default vue-ui-treemap-rect"
          }, null, 8, ga),
          e.$slots["group-label"] && !Ee.value ? (h(), M("foreignObject", {
            key: 0,
            width: te(o),
            height: ee(o),
            x: o.x0,
            y: o.y0
          }, [
            x(e.$slots, "group-label", it({ ref_for: !0 }, { group: o }), void 0, !0)
          ], 8, ya)) : !y($e) && t.value.style.chart.layout.labels.showDefaultLabels && !Ee.value ? (h(), M("g", {
            key: 1,
            style: { "pointer-events": "none" },
            innerHTML: So({ rect: o, seriesIndex: 0, isTitle: !0 }),
            class: "vue-data-ui-cursor-default"
          }, null, 8, ma)) : z("", !0)
        ]))), 128)),
        (h(!0), M(qe, null, ut(_e.value, (o, a) => (h(), M("g", {
          key: `k_${o.id}`
        }, [
          q("rect", {
            x: o.x0,
            y: o.y0,
            height: ee(o),
            width: te(o),
            fill: ve.value ? o.color ?? t.value.style.chart.backgroundColor : t.value.style.chart.layout.rects.gradient.show ? Ee.value ? t.value.style.chart.backgroundColor : `url(#tgrad_${o.id})` : o.color ?? t.value.style.chart.backgroundColor,
            rx: wl(o),
            stroke: E.value && E.value.id === o.id ? t.value.style.chart.layout.rects.selected.stroke : t.value.style.chart.layout.rects.stroke,
            "stroke-width": E.value && E.value.id === o.id ? t.value.style.chart.layout.rects.selected.strokeWidth : t.value.style.chart.layout.rects.strokeWidth,
            onClick: qt((n) => et(o, a), ["stop"]),
            onMouseenter: () => il({
              datapoint: o,
              seriesIndex: a
            }),
            onMouseleave: (n) => ul({ datapoint: o, seriesIndex: a }),
            style: me(`opacity:${E.value ? E.value.id === o.id ? 1 : t.value.style.chart.layout.rects.selected.unselectedOpacity : 1}`),
            class: jt([
              "vue-ui-treemap-rect",
              ll(o) ? "vue-data-ui-zoom-plus" : Ne.value ? "vue-data-ui-zoom-minus" : ""
            ])
          }, null, 46, ba),
          !e.$slots.rect && !y($e) && t.value.style.chart.layout.labels.showDefaultLabels && !Ee.value ? (h(), M("g", {
            key: 0,
            style: me(`pointer-events:none; opacity:${E.value ? E.value.id === o.id ? 1 : t.value.style.chart.layout.rects.selected.unselectedOpacity : 1}`),
            innerHTML: So({ rect: o, seriesIndex: a })
          }, null, 12, wa)) : z("", !0),
          (h(), M("foreignObject", {
            x: o.x0,
            y: o.y0,
            height: ee(o),
            width: te(o),
            class: "vue-ui-treemap-cell-foreignObject"
          }, [
            q("div", Ca, [
              y($e) ? z("", !0) : x(e.$slots, "rect", it({
                key: 0,
                ref_for: !0
              }, {
                rect: {
                  ...o,
                  height: ee(o),
                  width: te(o),
                  isSelected: E.value ? E.value.id === o.id : !0
                },
                shouldShow: o.proportion > t.value.style.chart.layout.labels.hideUnderProportion || Ne.value,
                fontSize: Re(o),
                isZoom: Ne.value,
                textColor: y(gt)(o.color)
              }), void 0, !0)
            ])
          ], 8, ka))
        ]))), 128)),
        x(e.$slots, "svg", X(j({ svg: B.value, isZoom: Ne.value, rect: E.value, config: t.value, isPrintingImg: y(Mt) | y(Ot) | Ze.value, isPrintingSvg: Je.value })), void 0, !0)
      ], 14, da)),
      e.$slots.watermark ? (h(), M("div", xa, [
        x(e.$slots, "watermark", X(j({ isPrinting: y(Mt) || y(Ot) || Ze.value || Je.value })), void 0, !0)
      ])) : z("", !0),
      q("div", {
        id: `legend-bottom-${Ce.value}`
      }, null, 8, $a),
      no.value ? (h(), be(Al, {
        key: 6,
        to: t.value.style.chart.legend.position === "top" ? `#legend-top-${Ce.value}` : `#legend-bottom-${Ce.value}`
      }, [
        q("div", {
          ref_key: "chartLegend",
          ref: to
        }, [
          t.value.style.chart.legend.show ? (h(), be(jl, {
            key: `legend_${kt.value}`,
            legendSet: De.value,
            config: al.value,
            id: `treemap_legend_${Ce.value}`,
            isCursorPointer: xe.value,
            onClickMarker: l[0] || (l[0] = ({ legend: o }) => tt(o))
          }, {
            item: S(({ legend: o, index: a }) => [
              y($e) ? z("", !0) : (h(), M("div", {
                key: 0,
                onClick: (n) => tt(o),
                style: me(`opacity:${F.value.includes(o.id) ? 0.5 : 1}`)
              }, vt(o.display), 13, Ta))
            ]),
            legendToggle: S(() => [
              De.value.length > 2 && t.value.style.chart.legend.selectAllToggle.show && !y($e) ? (h(), be(Gl, {
                key: 0,
                backgroundColor: t.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: t.value.style.chart.legend.selectAllToggle.color,
                fontSize: t.value.style.chart.legend.fontSize,
                checked: F.value.length > 0,
                isCursorPointer: xe.value,
                onToggle: nl
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : z("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "id", "isCursorPointer"])) : x(e.$slots, "legend", {
            key: 1,
            legend: De.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : z("", !0),
      e.$slots.source ? (h(), M("div", {
        key: 7,
        ref_key: "source",
        ref: oo,
        dir: "auto"
      }, [
        x(e.$slots, "source", {}, void 0, !0)
      ], 512)) : z("", !0),
      ct(y(m), {
        teleportTo: t.value.style.chart.tooltip.teleportTo,
        show: Z.value.showTooltip && mt.value,
        backgroundColor: t.value.style.chart.tooltip.backgroundColor,
        color: t.value.style.chart.tooltip.color,
        fontSize: t.value.style.chart.tooltip.fontSize,
        borderRadius: t.value.style.chart.tooltip.borderRadius,
        borderColor: t.value.style.chart.tooltip.borderColor,
        borderWidth: t.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: t.value.style.chart.tooltip.backgroundOpacity,
        position: t.value.style.chart.tooltip.position,
        offsetY: t.value.style.chart.tooltip.offsetY,
        parent: se.value,
        content: bt.value,
        isFullscreen: Oe.value,
        isCustom: y(Bo)(t.value.style.chart.tooltip.customFormat),
        smooth: t.value.style.chart.tooltip.smooth,
        backdropFilter: t.value.style.chart.tooltip.backdropFilter,
        smoothForce: t.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: t.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": S(() => [
          x(e.$slots, "tooltip-before", X(j({ ...Lt.value })), void 0, !0)
        ]),
        "tooltip-after": S(() => [
          x(e.$slots, "tooltip-after", X(j({ ...Lt.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      ke.value && t.value.userOptions.buttons.table ? (h(), be(Ml(Ue.value.component), it({ key: 8 }, Ue.value.props, {
        ref_key: "tableUnit",
        ref: He,
        onClose: $o
      }), Po({
        content: S(() => [
          (h(), be(y(g), {
            key: `table_${wt.value}`,
            colNames: ot.value.colNames,
            head: ot.value.head,
            body: ot.value.body,
            config: ot.value.config,
            title: t.value.table.useDialog ? "" : Ue.value.title,
            withCloseButton: !t.value.table.useDialog,
            isCursorPointer: xe.value,
            onClose: $o
          }, {
            th: S(({ th: o }) => [
              q("div", {
                innerHTML: o,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Sa)
            ]),
            td: S(({ td: o }) => [
              dt(vt(o.name || o), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        t.value.table.useDialog ? {
          name: "title",
          fn: S(() => [
            dt(vt(Ue.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        t.value.table.useDialog ? {
          name: "actions",
          fn: S(() => [
            q("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (o) => Ft(t.value.userOptions.callbacks.csv)),
              style: me({ cursor: xe.value ? "pointer" : "default" })
            }, [
              ct(Ro, {
                name: "fileCsv",
                stroke: Ue.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : z("", !0),
      x(e.$slots, "skeleton", {}, () => [
        y($e) ? (h(), be(Dl, { key: 0 })) : z("", !0)
      ], !0)
    ], 46, aa));
  }
}, Xa = /* @__PURE__ */ Kl(Aa, [["__scopeId", "data-v-c5ea578e"]]);
export {
  Xa as default
};
