// vue-data-ui/dist/vue-ui-galaxy-DWKHDciu.js
import { defineAsyncComponent as O, ref as s, computed as v, toRefs as Bt, onMounted as Vt, watch as Re, openBlock as u, createElementBlock as p, unref as n, normalizeStyle as z, normalizeClass as q, createBlock as $, withCtx as r, renderSlot as i, normalizeProps as y, guardReactiveProps as b, createCommentVNode as g, createElementVNode as S, createSlots as Ue, createVNode as ce, Fragment as We, renderList as Ge, Teleport as Et, toDisplayString as ve, resolveDynamicComponent as Mt, mergeProps as zt, createTextVNode as He, nextTick as de } from "vue";
import { j as Xe, t as je, o as Rt, q as Ye, r as Ut, m as Wt, p as J, n as Gt, a7 as qe, s as Ht, ac as Xt, k as Je, l as L, X as jt, T as Ke, a0 as Yt, a1 as qt, U as Jt } from "./lib-GwPnJSrY.js";
import { b as Kt } from "./labelUtils-CTXUc3V9.js";
import { t as Qt, u as Zt } from "./useResponsive-DfdjqQps.js";
import { u as el } from "./useConfig-CIgyKZ6R.js";
import { u as tl, B as ll } from "./BaseScanner-BhhkKVoj.js";
import { u as al } from "./usePrinter-DiK5R6k-.js";
import { u as ol } from "./useSvgExport-ByUukOZt.js";
import { u as he } from "./useNestedProp-BBT0mbHh.js";
import { u as sl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as nl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as ul } from "./useChartAccessibility-9icAAmYg.js";
import rl from "./img-BYYzY_QU.js";
import il from "./Title-BfyrP-jp.js";
import cl from "./Legend-1LMMsPxL.js";
import vl from "./vue_ui_galaxy-D95AN4FT.js";
import { B as dl } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as hl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const gl = ["id"], pl = ["id"], fl = ["xmlns", "viewBox"], yl = ["width", "height"], bl = ["id"], ml = ["stdDeviation"], kl = ["d", "stroke", "stroke-width"], Cl = ["d", "stroke", "stroke-width"], wl = ["filter"], Tl = ["d", "stroke", "stroke-width"], $l = ["d", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Sl = {
  key: 4,
  class: "vue-data-ui-watermark"
}, xl = ["id"], Pl = ["onClick"], Ol = ["innerHTML"], Al = {
  __name: "vue-ui-galaxy",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "selectDatapoint", "copyAlt"],
  setup(Qe, { expose: Ze, emit: et }) {
    const tt = O(() => import("./Tooltip-DL3QgjZf.js")), lt = O(() => import("./BaseIcon-CNx8Tzl5.js")), at = O(() => import("./vue-ui-accordion-BGHTqIBt.js")), ot = O(() => import("./DataTable-Bo1pmAqG.js")), st = O(() => import("./PenAndPaper-DWV1FMIi.js")), nt = O(() => import("./UserOptions-CrLqEimB.js")), ut = O(() => import("./PackageVersion-DN9d0ME5.js")), rt = O(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_galaxy: it } = el(), { isThemeValid: ct, warnInvalidTheme: vt } = sl(), k = Qe, C = s(Xe()), w = s(null), K = s(!1), Q = s(""), A = s(null), ge = s(0), pe = s(0), fe = s(0), ye = s(0), be = s(null), me = s(null), ke = s(null), Ce = s(null), D = s(null), R = s(null), we = s(!1), F = s(null), Z = s(null), U = s(!1), W = s(!1), Te = v(() => !!k.dataset && k.dataset.length), e = s(le()), _ = v(() => e.value.userOptions.useCursorPointer), dt = v(() => je({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        useCssAnimation: !1,
        style: {
          chart: {
            backgroundColor: "#99999930",
            legend: {
              backgroundColor: "transparent"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: G, FINAL_DATASET: $e } = tl({
      ...Bt(k),
      FINAL_CONFIG: e,
      prepareConfig: le,
      callback: () => {
        Promise.resolve().then(async () => {
          await de(), f.value.showTable = e.value.table.show;
        });
      },
      skeletonDataset: k.config?.skeletonDataset ?? [
        { name: "_", values: [21], color: "#DBDBDB" },
        { name: "_", values: [13], color: "#C4C4C4" },
        { name: "_", values: [8], color: "#ADADAD" }
      ],
      skeletonConfig: je({
        defaultConfig: e.value,
        userConfig: dt.value
      })
    });
    Vt(() => {
      we.value = !0, xe();
    });
    const Se = v(() => !!e.value.debug);
    function xe() {
      if (Rt(k.dataset) ? Ye({
        componentName: "VueUiGalaxy",
        type: "dataset",
        debug: Se.value
      }) : k.dataset.forEach((t, a) => {
        Ut({
          datasetObject: t,
          requiredAttributes: ["name", "values"]
        }).forEach((l) => {
          Ye({
            componentName: "VueUiGalaxy",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: Se.value
          });
        });
      }), e.value.responsive) {
        const t = Qt(() => {
          const { width: a, height: l } = Zt({
            chart: w.value,
            title: e.value.style.chart.title.text ? be.value : null,
            legend: e.value.style.chart.legend.show ? me.value : null,
            noTitle: Ce.value,
            source: ke.value
          });
          requestAnimationFrame(() => {
            H.value = Math.max(0.1, a), X.value = Math.max(0.1, l - 12);
          });
        });
        D.value && (R.value && D.value.unobserve(R.value), D.value.disconnect()), D.value = new ResizeObserver(t), R.value = w.value.parentNode, D.value.observe(R.value);
      }
    }
    const { userOptionsVisible: ee, setUserOptionsVisibility: Pe, keepUserOptionState: Oe } = nl({ config: e.value }), { svgRef: te } = ul({ config: e.value.style.chart.title });
    function le() {
      const t = he({
        userConfig: k.config,
        defaultConfig: it
      }), a = t.theme;
      if (!a) return t;
      if (!ct.value(t))
        return vt(t), t;
      const l = he({
        userConfig: vl[a] || k.config,
        defaultConfig: t
      }), o = he({
        userConfig: k.config,
        defaultConfig: l
      });
      return {
        ...o,
        customPalette: o.customPalette.length ? o.customPalette : Wt[a] || J
      };
    }
    Re(() => k.config, (t) => {
      e.value = le(), ee.value = !e.value.userOptions.showOnChartHover, xe(), pe.value += 1, fe.value += 1, ye.value += 1, f.value.dataLabels.show = e.value.style.chart.layout.labels.dataLabels.show, f.value.showTable = e.value.table.show, f.value.showTooltip = e.value.style.chart.tooltip.show;
    }, { deep: !0 });
    const { isPrinting: ae, isImaging: oe, generatePdf: Ae, generateImage: Ie } = al({
      elementId: `galaxy_${C.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-galaxy",
      options: e.value.userOptions.print
    }), ht = v(() => e.value.userOptions.show && !e.value.style.chart.title.text), gt = v(() => Gt(e.value.customPalette)), f = s({
      dataLabels: {
        show: e.value.style.chart.layout.labels.dataLabels.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    }), H = s(250), X = s(180), pt = s(0), ft = s(0), x = v(() => ({
      width: H.value,
      height: X.value,
      viewBox: `${pt.value} ${ft.value} ${H.value} ${X.value}`
    })), se = et, h = s([]);
    function yt() {
      h.value.length ? h.value = [] : V.value.forEach((t) => {
        h.value.push(t.id);
      });
    }
    function N(t) {
      h.value.includes(t.id) ? h.value = h.value.filter((a) => a !== t.id) : h.value.push(t.id), se("selectLegend", B.value.map((a) => ({
        name: a.name,
        color: a.color,
        value: a.value
      })));
    }
    function Le(t) {
      if (!m.value.length)
        return e.value.debug && console.warn("VueUiGalaxy - There are no series to show."), null;
      const a = m.value.find((l) => l.name === t);
      return a || (e.value.debug && console.warn(`VueUiGalaxy - Series name not found "${t}"`), null);
    }
    function bt(t) {
      const a = Le(t);
      a !== null && h.value.includes(a.id) && N({ id: a.id });
    }
    function mt(t) {
      const a = Le(t);
      a !== null && (h.value.includes(a.id) || N({ id: a.id }));
    }
    const m = v(() => $e.value.map((t, a) => ({
      name: t.name,
      color: Ht(t.color) || gt.value[a] || J[a] || J[a % J.length],
      value: t.values ? qe(t.values).reduce((l, o) => l + o, 0) : 0,
      absoluteValues: qe(t.values),
      id: Xe()
    })).sort((t, a) => a.value - t.value).map((t, a) => ({
      ...t,
      absoluteIndex: a
    })));
    function kt() {
      return m.value.map((t) => ({
        name: t.name,
        color: t.color,
        value: t.value
      }));
    }
    const T = v(() => m.value.filter((t) => !h.value.includes(t.id)).map((t) => t.value).reduce((t, a) => t + a, 0)), De = s(190), Fe = v(() => m.value.filter((t) => !h.value.includes(t.id))), Ct = v(
      () => (e.value.style.chart.layout.arcs.strokeWidth + e.value.style.chart.layout.arcs.borderWidth) / 2 + (e.value.style.chart.layout.padding ?? 12)
    ), wt = v(() => Xt({
      maxPoints: De.value,
      a: e.value.style.chart.layout.arcs.a ?? 6,
      b: e.value.style.chart.layout.arcs.b ?? 6,
      angleStep: e.value.style.chart.layout.arcs.angleStep ?? 0.07,
      startX: x.value.width / 2 + e.value.style.chart.layout.arcs.offsetX,
      startY: x.value.height / 2 + e.value.style.chart.layout.arcs.offsetY,
      boxWidth: x.value.width,
      boxHeight: x.value.height,
      padding: Ct.value
    })), B = v(() => {
      const t = [];
      for (let a = 0; a < Fe.value.length; a += 1) {
        const l = Fe.value[a], o = l.value / T.value * De.value + (a > 0 && t.length ? t[a - 1].points : 0);
        t.push({
          points: o,
          ...l,
          seriesIndex: a,
          proportion: l.value / T.value,
          path: wt.value(o)
        });
      }
      return t.filter((a) => !h.value.includes(a.id)).toSorted((a, l) => l.points - a.points);
    });
    function ne(t) {
      const a = Math.min(H.value, X.value), l = A.value === t.id && e.value.style.chart.layout.arcs.hoverEffect.show ? e.value.style.chart.layout.arcs.hoverEffect.multiplicator : 1, o = (e.value.style.chart.layout.arcs.strokeWidth + e.value.style.chart.layout.arcs.borderWidth) * l, c = e.value.style.chart.layout.arcs.strokeWidth * l, d = e.value.style.chart.layout.arcs.strokeWidth / 2 * l;
      return {
        border: o / 180 * a,
        path: c / 180 * a,
        blur: d / 180 * a
      };
    }
    const I = s(!1);
    function _e(t) {
      I.value = t, ge.value += 1;
    }
    const ue = s(null);
    function Tt(t) {
      K.value = !1, A.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: t.absoluteIndex });
    }
    function $t(t) {
      se("selectDatapoint", t), e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: t.absoluteIndex });
    }
    function Ne({
      val: t,
      percentage: a,
      showVal: l,
      showPercentage: o
    }) {
      const c = e.value.style.chart.layout.labels.dataLabels;
      return Kt({
        config: c,
        val: t,
        percentage: a,
        showVal: l,
        showPercentage: o
      });
    }
    function St({ datapoint: t, _relativeIndex: a, seriesIndex: l, show: o = !1 }) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: t.absoluteIndex }), ue.value = {
        datapoint: t,
        seriesIndex: l,
        series: m.value,
        config: e.value
      }, K.value = o, A.value = t.id;
      let c = "";
      const d = e.value.style.chart.tooltip.customFormat;
      Ke(d) && Jt(() => d({
        seriesIndex: l,
        datapoint: t,
        series: m.value,
        config: e.value
      })) ? Q.value = d({
        seriesIndex: l,
        datapoint: t,
        series: m.value,
        config: e.value
      }) : (c += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${t.name}</div>`, c += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="none" fill="${t.color}"/></svg>`, c += `<b>${Ne({
        showVal: e.value.style.chart.tooltip.showValue,
        showPercentage: e.value.style.chart.tooltip.showPercentage,
        val: Je(
          e.value.style.chart.layout.labels.dataLabels.formatter,
          t.value,
          L({
            p: e.value.style.chart.layout.labels.dataLabels.prefix,
            v: t.value,
            s: e.value.style.chart.layout.labels.dataLabels.suffix,
            r: e.value.style.chart.tooltip.roundingValue
          }),
          { datapoint: t, seriesIndex: l }
        ),
        percentage: L({
          v: t.proportion * 100,
          s: "%",
          r: e.value.style.chart.tooltip.roundingPercentage
        })
      })}</b></div>`, Q.value = `<div>${c}</div>`);
    }
    const V = v(() => m.value.map((t, a) => {
      const l = Je(e.value.style.chart.layout.labels.dataLabels.formatter, t.value, L({
        p: e.value.style.chart.layout.labels.dataLabels.prefix,
        v: t.value,
        s: e.value.style.chart.layout.labels.dataLabels.suffix,
        r: e.value.style.chart.legend.roundingValue
      }), { datapoint: t, index: a }), o = isNaN(t.value / T.value) || h.value.includes(t.id) ? "-" : L({ v: t.value / T.value * 100, s: "%", r: e.value.style.chart.legend.roundingPercentage }), c = Ne({
        showVal: e.value.style.chart.legend.showValue,
        showPercentage: e.value.style.chart.legend.showPercentage,
        val: l,
        percentage: o
      });
      return {
        ...t,
        proportion: (t.value || 0) / $e.value.map((d) => (d.values || []).reduce((M, Nt) => M + Nt, 0)).reduce((d, M) => d + M, 0),
        opacity: h.value.includes(t.id) ? 0.5 : 1,
        shape: t.shape || "circle",
        segregate: () => N(t),
        isSegregated: h.value.includes(t.id),
        display: `${t.name}${e.value.style.chart.legend.showPercentage || e.value.style.chart.legend.showValue ? ": " : ""}${c}`
      };
    })), xt = v(() => ({
      cy: "galaxy-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), P = v(() => {
      const t = B.value.map((l) => ({
        name: l.name,
        color: l.color
      })), a = B.value.map((l) => l.value);
      return { head: t, body: a };
    });
    function re(t = null) {
      de(() => {
        const a = P.value.head.map((c, d) => [[
          c.name
        ], [P.value.body[d]], [isNaN(P.value.body[d] / T.value) ? "-" : P.value.body[d] / T.value * 100]]), l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[""], ["val"], ["%"]]].concat(a), o = Yt(l);
        t ? t(o) : qt({ csvContent: o, title: e.value.style.chart.title.text || "vue-ui-galaxy" });
      });
    }
    const j = v(() => {
      const t = [
        ' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>',
        L({ p: e.value.style.chart.layout.labels.dataLabels.prefix, v: T.value, s: e.value.style.chart.layout.labels.dataLabels.suffix, r: e.value.table.td.roundingValue }),
        "100%"
      ], a = P.value.head.map((c, d) => {
        const M = L({ p: e.value.style.chart.layout.labels.dataLabels.prefix, v: P.value.body[d], s: e.value.style.chart.layout.labels.dataLabels.suffix, r: e.value.table.td.roundingValue });
        return [
          {
            color: c.color,
            name: c.name
          },
          M,
          isNaN(P.value.body[d] / T.value) ? "-" : L({
            v: P.value.body[d] / T.value * 100,
            s: "%",
            r: e.value.table.td.roundingPercentage
          })
        ];
      }), l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.series,
          e.value.table.columnNames.value,
          e.value.table.columnNames.percentage
        ],
        head: t,
        body: a,
        config: l
      };
    });
    function Be() {
      f.value.showTable = !f.value.showTable;
    }
    function Ve() {
      f.value.showTooltip = !f.value.showTooltip;
    }
    const Y = s(!1);
    function ie() {
      Y.value = !Y.value;
    }
    async function Pt({ scale: t = 2 } = {}) {
      if (!w.value) return;
      const { width: a, height: l } = w.value.getBoundingClientRect(), o = a / l, { imageUri: c, base64: d } = await rl({ domElement: w.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: c,
        base64: d,
        title: e.value.style.chart.title.text,
        width: a,
        height: l,
        aspectRatio: o
      };
    }
    const E = v(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = f.value.showTable;
      return {
        component: t ? rt : at,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: I.value,
          fullscreenParent: w.value,
          forcedWidth: Math.min(500, window.innerWidth * 0.8),
          isCursorPointer: _.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    Re(() => f.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && F.value ? F.value.open() : "close" in F.value && F.value.close());
    });
    function Ee() {
      f.value.showTable = !1, Z.value && Z.value.setTableIconState(!1);
    }
    const Ot = v(() => V.value.map((t) => ({
      ...t,
      name: t.display
    }))), At = v(() => e.value.style.chart.backgroundColor), It = v(() => e.value.style.chart.legend), Lt = v(() => e.value.style.chart.title), { exportSvg: Dt, getSvg: Ft } = ol({
      svg: te,
      title: Lt,
      legend: It,
      legendItems: Ot,
      backgroundColor: At
    });
    async function Me({ isCb: t }) {
      W.value = !0, await de();
      try {
        if (t) {
          const { blob: a, url: l, text: o, dataUrl: c } = await Ft();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: o, dataUrl: c }));
        } else
          await Promise.resolve(Dt());
      } finally {
        W.value = !1;
      }
    }
    function _t(t) {
      if (t?.stage === "start") {
        U.value = !0;
        return;
      }
      if (t?.stage === "end") {
        U.value = !1;
        return;
      }
      Ie();
    }
    async function ze() {
      if (se("copyAlt", {
        config: e.value,
        dataset: m.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: m.value
      }));
    }
    return Ze({
      getData: kt,
      getImage: Pt,
      generatePdf: Ae,
      generateCsv: re,
      generateImage: Ie,
      generateSvg: Me,
      hideSeries: mt,
      showSeries: bt,
      toggleTable: Be,
      toggleTooltip: Ve,
      toggleAnnotator: ie,
      toggleFullscreen: _e,
      copyAlt: ze
    }), (t, a) => (u(), p("div", {
      ref_key: "galaxyChart",
      ref: w,
      class: q(`vue-data-ui-component vue-ui-galaxy ${I.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"} ${n(G) ? "loading" : ""}`),
      style: z(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;${e.value.style.chart.title.text ? "" : "padding-top:36px"};background:${e.value.style.chart.backgroundColor}`),
      id: `galaxy_${C.value}`,
      onMouseenter: a[2] || (a[2] = () => n(Pe)(!0)),
      onMouseleave: a[3] || (a[3] = () => n(Pe)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (u(), $(n(st), {
        key: 0,
        svgRef: n(te),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: Y.value,
        isCursorPointer: _.value,
        onClose: ie
      }, {
        "annotator-action-close": r(() => [
          i(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": r(({ color: l }) => [
          i(t.$slots, "annotator-action-color", y(b({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": r(({ mode: l }) => [
          i(t.$slots, "annotator-action-draw", y(b({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": r(({ disabled: l }) => [
          i(t.$slots, "annotator-action-undo", y(b({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": r(({ disabled: l }) => [
          i(t.$slots, "annotator-action-redo", y(b({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": r(({ disabled: l }) => [
          i(t.$slots, "annotator-action-delete", y(b({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : g("", !0),
      ht.value ? (u(), p("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Ce,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : g("", !0),
      e.value.style.chart.title.text ? (u(), p("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: be,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (u(), $(il, {
          key: `title_${pe.value}`,
          config: {
            title: {
              cy: "galaxy-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "galaxy-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : g("", !0),
      S("div", {
        id: `legend-top-${C.value}`
      }, null, 8, pl),
      e.value.userOptions.show && Te.value && (n(Oe) || n(ee)) ? (u(), $(n(nt), {
        ref_key: "userOptionsRef",
        ref: Z,
        key: `user_option_${ge.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: n(ae),
        isImaging: n(oe),
        uid: C.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isTooltip: f.value.showTooltip,
        isFullscreen: I.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: w.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Y.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: _.value,
        onToggleFullscreen: _e,
        onGeneratePdf: n(Ae),
        onGenerateCsv: re,
        onGenerateImage: _t,
        onGenerateSvg: Me,
        onToggleTable: Be,
        onToggleTooltip: Ve,
        onToggleAnnotator: ie,
        onCopyAlt: ze,
        style: z({
          visibility: n(Oe) ? n(ee) ? "visible" : "hidden" : "visible"
        })
      }, Ue({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: r(({ isOpen: l, color: o }) => [
            i(t.$slots, "menuIcon", y(b({ isOpen: l, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: r(() => [
            i(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: r(() => [
            i(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: r(() => [
            i(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: r(() => [
            i(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: r(() => [
            i(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: r(() => [
            i(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: r(({ toggleFullscreen: l, isFullscreen: o }) => [
            i(t.$slots, "optionFullscreen", y(b({ toggleFullscreen: l, isFullscreen: o })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: r(({ toggleAnnotator: l, isAnnotator: o }) => [
            i(t.$slots, "optionAnnotator", y(b({ toggleAnnotator: l, isAnnotator: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: r(({ altCopy: l }) => [
            i(t.$slots, "optionAltCopy", y(b({ altCopy: l })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasFullscreen", "hasAltCopy", "isTooltip", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : g("", !0),
      (u(), p("svg", {
        ref_key: "svgRef",
        ref: te,
        xmlns: n(jt),
        class: q({ "vue-data-ui-fullscreen--on": I.value, "vue-data-ui-fulscreen--off": !I.value }),
        viewBox: x.value.viewBox,
        style: z(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color}`)
      }, [
        ce(n(ut)),
        t.$slots["chart-background"] ? (u(), p("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: x.value.width,
          height: x.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          i(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, yl)) : g("", !0),
        S("defs", null, [
          S("filter", {
            id: `blur_${C.value}`,
            x: "-50%",
            y: "-50%",
            width: "200%",
            height: "200%"
          }, [
            S("feGaussianBlur", {
              in: "SourceGraphic",
              stdDeviation: 100 / e.value.style.chart.layout.arcs.gradient.intensity
            }, null, 8, ml)
          ], 8, bl)
        ]),
        (u(!0), p(We, null, Ge(B.value, (l) => (u(), p("g", null, [
          l.value ? (u(), p("path", {
            key: 0,
            d: l.path,
            fill: "none",
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": ne(l).border,
            "stroke-linecap": "round"
          }, null, 8, kl)) : g("", !0),
          l.value ? (u(), p("path", {
            key: 1,
            d: l.path,
            fill: "none",
            stroke: l.color,
            "stroke-width": ne(l).path,
            "stroke-linecap": "round",
            class: q(`${A.value && A.value !== l.id && e.value.useBlurOnHover ? "vue-ui-galaxy-blur" : ""}`)
          }, null, 10, Cl)) : g("", !0),
          l.value && e.value.style.chart.layout.arcs.gradient.show ? (u(), p("g", {
            key: 2,
            filter: `url(#blur_${C.value})`
          }, [
            S("path", {
              d: l.path,
              fill: "none",
              stroke: e.value.style.chart.layout.arcs.gradient.color,
              "stroke-width": ne(l).blur,
              "stroke-linecap": "round",
              class: q(`vue-ui-galaxy-gradient ${A.value && A.value !== l.id && e.value.useBlurOnHover ? "vue-ui-galaxy-blur" : ""}`)
            }, null, 10, Tl)
          ], 8, wl)) : g("", !0)
        ]))), 256)),
        (u(!0), p(We, null, Ge(B.value, (l, o) => (u(), p("g", null, [
          l.value ? (u(), p("path", {
            key: 0,
            d: l.path,
            fill: "none",
            stroke: "transparent",
            "stroke-width": e.value.style.chart.layout.arcs.strokeWidth + e.value.style.chart.layout.arcs.borderWidth,
            "stroke-linecap": "round",
            onMouseenter: (c) => St({
              datapoint: l,
              seriesIndex: l.seriesIndex,
              show: !0
            }),
            onMouseleave: (c) => Tt(l),
            onClick: (c) => $t(l)
          }, null, 40, $l)) : g("", !0)
        ]))), 256)),
        i(t.$slots, "svg", {
          svg: {
            ...x.value,
            isPrintingImg: n(ae) | n(oe) | U.value,
            isPrintingSvg: W.value
          }
        }, void 0, !0)
      ], 14, fl)),
      t.$slots.watermark ? (u(), p("div", Sl, [
        i(t.$slots, "watermark", y(b({ isPrinting: n(ae) || n(oe) || U.value || W.value })), void 0, !0)
      ])) : g("", !0),
      S("div", {
        id: `legend-bottom-${C.value}`
      }, null, 8, xl),
      we.value ? (u(), $(Et, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${C.value}` : `#legend-bottom-${C.value}`
      }, [
        S("div", {
          ref_key: "chartLegend",
          ref: me
        }, [
          e.value.style.chart.legend.show ? (u(), $(cl, {
            key: `legend_${ye.value}`,
            legendSet: V.value,
            config: xt.value,
            onClickMarker: a[0] || (a[0] = ({ legend: l }) => N(l))
          }, {
            item: r(({ legend: l, index: o }) => [
              n(G) ? g("", !0) : (u(), p("div", {
                key: 0,
                onClick: (c) => N(l),
                style: z(`opacity:${h.value.includes(l.id) ? 0.5 : 1}`)
              }, ve(l.display), 13, Pl))
            ]),
            legendToggle: r(() => [
              V.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !n(G) ? (u(), $(dl, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: h.value.length > 0,
                onToggle: yt
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked"])) : g("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config"])) : g("", !0),
          i(t.$slots, "legend", { legend: V.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : g("", !0),
      t.$slots.source ? (u(), p("div", {
        key: 6,
        ref_key: "source",
        ref: ke,
        dir: "auto"
      }, [
        i(t.$slots, "source", {}, void 0, !0)
      ], 512)) : g("", !0),
      ce(n(tt), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: f.value.showTooltip && K.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        fontSize: e.value.style.chart.tooltip.fontSize,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: w.value,
        content: Q.value,
        isFullscreen: I.value,
        isCustom: n(Ke)(e.value.style.chart.tooltip.customFormat),
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": r(() => [
          i(t.$slots, "tooltip-before", y(b({ ...ue.value })), void 0, !0)
        ]),
        "tooltip-after": r(() => [
          i(t.$slots, "tooltip-after", y(b({ ...ue.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      Te.value && e.value.userOptions.buttons.table ? (u(), $(Mt(E.value.component), zt({ key: 7 }, E.value.props, {
        ref_key: "tableUnit",
        ref: F,
        onClose: Ee
      }), Ue({
        content: r(() => [
          (u(), $(n(ot), {
            key: `table_${fe.value}`,
            colNames: j.value.colNames,
            head: j.value.head,
            body: j.value.body,
            config: j.value.config,
            title: e.value.table.useDialog ? "" : E.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: _.value,
            onClose: Ee
          }, {
            th: r(({ th: l }) => [
              S("div", {
                innerHTML: l,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Ol)
            ]),
            td: r(({ td: l }) => [
              He(ve(l.name || l), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: r(() => [
            He(ve(E.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: r(() => [
            S("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[1] || (a[1] = (l) => re(e.value.userOptions.callbacks.csv)),
              style: z({ cursor: _.value ? "pointer" : "default" })
            }, [
              ce(n(lt), {
                name: "fileCsv",
                stroke: E.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : g("", !0),
      i(t.$slots, "skeleton", {}, () => [
        n(G) ? (u(), $(ll, { key: 0 })) : g("", !0)
      ], !0)
    ], 46, gl));
  }
}, Yl = /* @__PURE__ */ hl(Al, [["__scopeId", "data-v-39ee1d9f"]]);
export {
  Yl as default
};
