// vue-data-ui/dist/vue-ui-waffle-CbX5K9Q2.js
import { defineAsyncComponent as V, useSlots as al, computed as c, ref as r, toRefs as ol, watch as le, shallowRef as vt, onMounted as nl, onBeforeUnmount as ul, openBlock as s, createElementBlock as v, unref as d, normalizeStyle as ae, normalizeClass as dt, createBlock as B, withCtx as h, renderSlot as p, normalizeProps as $, guardReactiveProps as x, createCommentVNode as b, createElementVNode as k, createSlots as Fe, createVNode as de, Fragment as S, renderList as N, mergeProps as De, toDisplayString as he, Teleport as sl, resolveDynamicComponent as rl, createTextVNode as ht, nextTick as pt } from "vue";
import { j as il, t as ft, m as cl, p as E, o as gt, q as ze, r as vl, n as dl, k as oe, l as F, X as hl, u as pl, v as fl, W as gl, s as yl, a0 as ml, a1 as bl, a as wl, T as kl, U as Cl } from "./lib-GwPnJSrY.js";
import { b as $l } from "./labelUtils-CTXUc3V9.js";
import { t as xl, u as Tl } from "./useResponsive-DfdjqQps.js";
import { u as Sl } from "./useConfig-CIgyKZ6R.js";
import { u as Al, B as Pl } from "./BaseScanner-BhhkKVoj.js";
import { u as Il } from "./usePrinter-DiK5R6k-.js";
import { u as Ol } from "./useSvgExport-ByUukOZt.js";
import { u as Ve } from "./useNestedProp-BBT0mbHh.js";
import { u as _l } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Ll } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Bl } from "./useChartAccessibility-9icAAmYg.js";
import Fl from "./img-BYYzY_QU.js";
import { _ as Dl } from "./Shape-CSyWYgXc.js";
import zl from "./Title-BfyrP-jp.js";
import Vl from "./vue_ui_waffle-M2h-uJQ_.js";
import Nl from "./Legend-1LMMsPxL.js";
import { B as Ml } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Rl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Ul = ["id"], El = ["id"], Wl = ["xmlns", "viewBox"], Gl = ["id"], jl = ["stop-color"], Hl = ["stop-color"], ql = ["id"], Xl = ["x", "y", "height", "width"], Yl = ["height", "width"], Jl = { key: 0 }, Kl = ["rx", "x", "y", "height", "width", "stroke", "stroke-width", "filter"], Ql = ["rx", "x", "y", "height", "width", "fill", "stroke", "stroke-width", "filter"], Zl = { key: 1 }, ea = ["rx", "x", "y", "height", "width", "fill", "filter"], ta = ["textContent", "x", "y", "font-size", "fill", "filter"], la = ["x", "y", "height", "width", "onMouseover", "onMouseleave", "onClick"], aa = {
  key: 4,
  class: "vue-data-ui-watermark"
}, oa = ["id"], na = ["onClick"], ua = ["innerHTML"], sa = {
  __name: "vue-ui-waffle",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "copyAlt"],
  setup(yt, { expose: mt, emit: bt }) {
    const wt = V(() => import("./Tooltip-DL3QgjZf.js")), kt = V(() => import("./BaseIcon-CNx8Tzl5.js")), Ct = V(() => import("./DataTable-Bo1pmAqG.js")), $t = V(() => import("./vue-ui-accordion-BGHTqIBt.js")), xt = V(() => import("./UserOptions-CrLqEimB.js")), Tt = V(() => import("./PenAndPaper-DWV1FMIi.js")), St = V(() => import("./PackageVersion-DN9d0ME5.js")), At = V(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_waffle: Pt } = Sl(), { isThemeValid: It, warnInvalidTheme: Ot } = _l(), y = yt, _t = al(), pe = c({
      get() {
        return !!y.dataset && y.dataset.length;
      },
      set(t) {
        return t;
      }
    }), m = r(il()), fe = r(!1), ge = r(""), W = r(null), Ne = r(0), A = r(null), Me = r(null), Re = r(null), Ue = r(null), Ee = r(null), We = r(0), Ge = r(0), je = r(0), He = r(!1), Y = r(null), ye = r(null), ne = r(!1), ue = r(!1), e = r(Ce()), M = c(() => e.value.userOptions.useCursorPointer), Lt = c(() => ft({
      defaultConfig: {
        useCustomCells: !1,
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              labels: {
                captions: { show: !1 }
              },
              rect: {
                stroke: "#999999"
              }
            },
            legend: {
              backgroundColor: "transparent",
              showValue: !1,
              showPercentage: !1
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: me, FINAL_DATASET: G, manualLoading: be } = Al({
      ...ol(y),
      FINAL_CONFIG: e,
      prepareConfig: Ce,
      callback: () => {
        Promise.resolve().then(async () => {
          g.value = Te();
        });
      },
      skeletonDataset: y.config?.skeletonDataset ?? [
        {
          name: "",
          values: [1],
          color: "#AAAAAA"
        },
        {
          name: "",
          values: [1],
          color: "#BABABA"
        },
        {
          name: "",
          values: [1],
          color: "#CACACA"
        }
      ],
      skeletonConfig: ft({
        defaultConfig: e.value,
        userConfig: Lt.value
      })
    }), { userOptionsVisible: we, setUserOptionsVisibility: qe, keepUserOptionState: Xe } = Ll({ config: e.value }), { svgRef: ke } = Bl({ config: e.value.style.chart.title });
    function Ce() {
      const t = Ve({
        userConfig: y.config,
        defaultConfig: Pt
      }), a = t.theme;
      if (!a) return t;
      if (!It.value(t))
        return Ot(t), t;
      const l = Ve({
        userConfig: Vl[a] || y.config,
        defaultConfig: t
      }), o = Ve({
        userConfig: y.config,
        defaultConfig: l
      });
      return {
        ...o,
        customPalette: o.customPalette.length ? o.customPalette : cl[a] || E
      };
    }
    le(() => y.config, (t) => {
      me.value || (e.value = Ce()), we.value = !e.value.userOptions.showOnChartHover, Je(), We.value += 1, Ge.value += 1, je.value += 1, C.value.showTable = e.value.table.show, C.value.showTooltip = e.value.style.chart.tooltip.show;
    }, { deep: !0 });
    const D = vt(null), j = vt(null), Ye = c(() => !!e.value.debug);
    function Je() {
      if (gt(y.dataset) ? (ze({
        componentName: "VueUiWaffle",
        type: "dataset",
        debug: Ye.value
      }), pe.value = !1, be.value = !0) : y.dataset.forEach((t, a) => {
        vl({
          datasetObject: t,
          requiredAttributes: ["name", "values"]
        }).forEach((l) => {
          ze({
            componentName: "VueUiWaffle",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: Ye.value
          });
        });
      }), gt(y.dataset) || (be.value = e.value.loading), e.value.responsive) {
        const t = xl(() => {
          const { width: a, height: l } = Tl({
            chart: A.value,
            title: e.value.style.chart.title.text ? Me.value : null,
            legend: e.value.style.chart.legend.show ? Re.value : null,
            source: Ue.value,
            noTitle: Ee.value
          });
          requestAnimationFrame(() => {
            R.value.width = a, R.value.height = l, P.value.width = a, P.value.height = l;
          });
        });
        D.value && (j.value && D.value.unobserve(j.value), D.value.disconnect()), D.value = new ResizeObserver(t), j.value = A.value.parentNode, D.value.observe(j.value);
      }
    }
    nl(() => {
      He.value = !0, Je();
    }), ul(() => {
      D.value && (j.value && D.value.unobserve(j.value), D.value.disconnect());
    });
    const { isPrinting: $e, isImaging: xe, generatePdf: Ke, generateImage: Qe } = Il({
      elementId: `vue-ui-waffle_${m.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-waffle",
      options: e.value.userOptions.print
    }), Bt = c(() => e.value.userOptions.show && !e.value.style.chart.title.text), Ze = c(() => dl(e.value.customPalette)), C = r({
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    le(e, () => {
      C.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const R = r({
      height: 512,
      width: 512
    }), P = r({
      top: 0,
      left: 0,
      height: 512,
      width: 512
    }), I = c(() => (P.value.width - e.value.style.chart.layout.grid.size * e.value.style.chart.layout.grid.spaceBetween) / e.value.style.chart.layout.grid.size), O = c(() => (P.value.height - e.value.style.chart.layout.grid.size * e.value.style.chart.layout.grid.spaceBetween) / e.value.style.chart.layout.grid.size), Ft = c(() => Math.max(1e-4, P.value.width / e.value.style.chart.layout.grid.size)), et = c(() => Math.max(1e-4, P.value.height / e.value.style.chart.layout.grid.size));
    function tt(t) {
      const a = e.value.style.chart.layout.grid.size * e.value.style.chart.layout.grid.size, l = t.reduce((T, Be) => T + Be, 0), o = t.map((T) => T / l * a), n = o.map(Math.floor), u = o.map((T) => T % 1);
      let w = a - n.reduce((T, Be) => T + Be, 0);
      for (; w > 0; ) {
        let T = u.indexOf(Math.max(...u));
        n[T] += 1, u[T] = 0, w -= 1;
      }
      return n;
    }
    const J = r(!1);
    function Te() {
      return J.value = G.value.flatMap((t) => t.values.reduce((a, l) => a + l, 0)).reduce((t, a) => t + a, 0) === 0, G.value.map((t, a) => ({
        ...t,
        color: yl(t.color) || Ze.value[a] || E[a] || E[a % E.length],
        uid: `serie_${a}`,
        absoluteIndex: a
      }));
    }
    const se = c(() => Te()), g = r(se.value);
    le(() => y.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (be.value = !1);
    }, { immediate: !0 }), le(() => y.dataset, (t) => {
      g.value = Te();
    }, { deep: !0 });
    const Dt = c(() => {
      const t = g.value.filter((a, l) => !f.value.includes(a.uid)).map((a, l) => J.value ? 1 : (a.values || []).reduce((o, n) => o + n, 0));
      return tt(t);
    }), zt = c(() => {
      const t = g.value.map((a, l) => J.value ? 1 : (a.values || []).reduce((o, n) => o + n));
      return tt(t);
    }), H = c(() => (G.value.forEach((t, a) => {
      [null, void 0].includes(t.values) && ze({
        componentName: "VueUiWaffle",
        type: "datasetSerieAttribute",
        property: "values (number[])",
        index: a
      });
    }), g.value.filter((t, a) => !f.value.includes(t.uid)).map((t, a) => ({
      absoluteIndex: t.absoluteIndex,
      uid: t.uid,
      name: t.name,
      color: t.color,
      value: (t.values || []).reduce((l, o) => l + o, 0),
      absoluteValues: t.values || [],
      proportion: Dt.value[a]
    })))), Se = c(() => g.value.map((t, a) => ({
      absoluteIndex: t.absoluteIndex,
      uid: t.uid,
      name: t.name,
      color: t.color,
      value: (t.values || []).reduce((l, o) => l + o, 0),
      absoluteValues: t.values || [],
      proportion: zt.value[a]
    })));
    function Vt() {
      return Se.value.map((t) => ({
        name: t.name,
        color: t.color,
        value: t.value,
        proportion: t.proportion
      }));
    }
    const Nt = c(() => {
      let t = 0;
      return H.value.map((a, l) => {
        const o = t, n = o + a.proportion, u = [];
        for (let w = Math.floor(o); w < Math.floor(n); w += 1)
          u.push(w);
        return t = n, {
          ...a,
          start: o,
          rects: u
        };
      });
    }), i = c(() => Nt.value.flatMap((t, a) => t.rects.map((l, o) => ({
      isFirst: o === 0,
      isLongEnough: l.length > 2,
      name: t.name,
      color: t.color,
      value: t.value,
      serieIndex: a,
      absoluteStartIndex: o < 3,
      serieId: t.uid,
      ...t
    }))).map((t, a) => ({
      ...t,
      isAbsoluteFirst: a % e.value.style.chart.layout.grid.size === 0
    }))), q = c(() => {
      const t = [];
      for (let a = 0; a < e.value.style.chart.layout.grid.size; a += 1)
        for (let l = 0; l < e.value.style.chart.layout.grid.size; l += 1)
          t.push({
            isStartOfLine: l === 0,
            position: e.value.style.chart.layout.grid.vertical ? a : l,
            x: (e.value.style.chart.layout.grid.vertical ? a : l) * (I.value + e.value.style.chart.layout.grid.spaceBetween),
            y: (e.value.style.chart.layout.grid.vertical ? l : a) * (O.value + e.value.style.chart.layout.grid.spaceBetween) + P.value.top
          });
      return t;
    }), f = r([]), re = r(!1), K = r(/* @__PURE__ */ Object.create(null)), Q = r(/* @__PURE__ */ Object.create(null)), Z = r(0);
    function ee(t, a) {
      g.value = g.value.map((l) => l.uid === t ? {
        ...l,
        values: [a]
      } : l);
    }
    function lt(t, a) {
      const l = a.find((o) => o.uid === t);
      return l ? (l.values || []).reduce((o, n) => o + n, 0) : 0;
    }
    function Mt(t) {
      K.value[t] && (cancelAnimationFrame(K.value[t]), delete K.value[t]), Q.value[t] && (cancelAnimationFrame(Q.value[t]), delete Q.value[t]);
    }
    function at({
      seriesId: t,
      fromValue: a,
      toValue: l,
      mode: o
    }) {
      Mt(t), Z.value += 1, re.value = !0;
      let n = a;
      return new Promise((u) => {
        function w() {
          if (o === "increase") {
            if (n >= l) {
              ee(t, l), Z.value -= 1, Z.value <= 0 && (re.value = !1), delete K.value[t], u();
              return;
            }
            n += l * 0.025, n > l && (n = l), ee(t, n), K.value[t] = requestAnimationFrame(w);
          } else {
            if (n <= a / 100) {
              ee(t, 0), Z.value -= 1, Z.value <= 0 && (re.value = !1), delete Q.value[t], u();
              return;
            }
            n /= 1.15, ee(t, n), Q.value[t] = requestAnimationFrame(w);
          }
        }
        w();
      });
    }
    function Rt() {
      g.value = se.value;
      const t = f.value.length > 0;
      if (!e.value.useAnimation) {
        t ? f.value = [] : f.value = z.value.map((a) => a.uid);
        return;
      }
      t ? [...f.value].forEach((l) => {
        X(l, !0);
      }) : z.value.map((l) => l.uid).forEach((l) => {
        X(l, !0);
      });
    }
    function X(t, a = !1) {
      if (J.value && !a) return;
      const l = a || f.value.length < z.value.length - 1 && z.value.length > 1;
      if (!e.value.useAnimation) {
        f.value.includes(t) ? f.value = f.value.filter((u) => u !== t) : l && f.value.push(t);
        return;
      }
      const o = lt(t, se.value), n = lt(t, g.value);
      if (!(n === 0 && o === 0)) {
        if (f.value.includes(t)) {
          f.value = f.value.filter((u) => u !== t), at({
            seriesId: t,
            fromValue: n,
            toValue: o,
            mode: "increase"
          }).then(() => {
            Oe(
              "selectLegend",
              H.value.map((u) => ({
                name: u.name,
                color: u.color,
                value: u.value,
                proportion: u.proportion / Math.pow(e.value.style.chart.layout.grid.size, 2)
              }))
            );
          });
          return;
        }
        l && at({
          seriesId: t,
          fromValue: n,
          toValue: 0,
          mode: "decrease"
        }).then(() => {
          f.value.includes(t) || f.value.push(t), ee(t, 0), Oe(
            "selectLegend",
            H.value.map((u) => ({
              name: u.name,
              color: u.color,
              value: u.value,
              proportion: u.proportion / Math.pow(e.value.style.chart.layout.grid.size, 2)
            }))
          );
        });
      }
    }
    function ot(t) {
      if (!Se.value.length)
        return e.value.debug && console.warn("VueUiWaffle - There are no series to show."), null;
      const a = Se.value.find((l) => l.name === t);
      return a || (e.value.debug && console.warn(`VueUiWaffle - Series name not found "${t}"`), null);
    }
    function Ut(t) {
      const a = ot(t);
      a !== null && f.value.includes(a.uid) && X(a.uid);
    }
    function Et(t) {
      const a = ot(t);
      a !== null && (f.value.includes(a.uid) || X(a.uid));
    }
    function Ae({
      val: t,
      percentage: a,
      showVal: l,
      showPercentage: o,
      config: n
    }) {
      return $l({
        config: n,
        val: t,
        percentage: a,
        showVal: l,
        showPercentage: o
      });
    }
    const z = c(() => g.value.map((t, a) => {
      const l = (t.values || []).reduce((n, u) => n + u, 0), o = l / g.value.map((n) => (n.values || []).reduce((u, w) => u + w, 0)).reduce((n, u) => n + u, 0);
      return {
        name: t.name,
        color: t.color || Ze[a] || E[a] || E[a % E.length],
        value: l,
        proportion: o,
        uid: t.uid,
        shape: "square"
      };
    }).map((t, a) => {
      const l = oe(e.value.style.chart.layout.labels.dataLabels.formatter, t.value, F({
        p: e.value.style.chart.layout.labels.dataLabels.prefix,
        v: t.value,
        s: e.value.style.chart.layout.labels.dataLabels.suffix,
        r: e.value.style.chart.legend.roundingValue
      }), { datapoint: t, index: a }), o = isNaN(t.value / _.value) ? "-" : F({ v: t.value / _.value * 100, s: "%", r: e.value.style.chart.legend.roundingPercentage }), n = Ae({
        val: l,
        percentage: o,
        showVal: e.value.style.chart.legend.showValue,
        showPercentage: e.value.style.chart.legend.showPercentage,
        config: e.value.style.chart.legend
      });
      return {
        ...t,
        opacity: f.value.includes(t.uid) ? 0.5 : 1,
        segregate: () => X(t.uid),
        isSegregated: f.value.includes(t.uid),
        display: `${t.name}${e.value.style.chart.legend.showPercentage || e.value.style.chart.legend.showValue ? ": " : ""}${n}`
      };
    })), Wt = c(() => ({
      cy: "waffle-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), _ = c(() => H.value.map((t) => t.value).reduce((t, a) => t + a, 0)), Pe = r(null), Ie = r(null);
    function Gt(t) {
      const a = i.value[t];
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: a, seriesIndex: a.serieIndex });
    }
    function jt(t) {
      const a = i.value[t];
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: a, seriesIndex: a.serieIndex }), Ie.value = null, fe.value = !1, W.value = null;
    }
    function Ht(t) {
      if (f.value.length === y.dataset.length) return;
      const a = i.value[t];
      Pe.value = {
        datapoint: a,
        seriesIndex: a.absoluteIndex,
        series: g.value,
        config: e.value
      }, e.value.events.datapointEnter && Ie.value !== a.serieIndex && e.value.events.datapointEnter({ datapoint: a, seriesIndex: a.serieIndex }), Ie.value = a.serieIndex, fe.value = !0, W.value = i.value[t].serieIndex;
      const l = e.value.style.chart.tooltip.customFormat;
      if (kl(l) && Cl(() => l({
        seriesIndex: i.value[t].absoluteIndex,
        datapoint: a,
        series: g.value,
        config: e.value
      })))
        ge.value = l({
          seriesIndex: i.value[t].absoluteIndex,
          datapoint: a,
          series: g.value,
          config: e.value
        });
      else {
        let o = "";
        o += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${a.name}</div>`, o += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 60 60" height="14" width="14"><rect x="0" y="0" height="60" width="60" stroke="none" rx="1" fill="${a.color}" />${_t.pattern ? `<rect x="0" y="0" height="60" width="60" stroke="none" rx="1" stroke="none" fill="url(#pattern_${m.value}_${a.absoluteIndex})"/>` : ""}</svg>`, o += `<b>${Ae({
          config: e.value.style.chart.tooltip,
          showVal: e.value.style.chart.tooltip.showValue,
          showPercentage: e.value.style.chart.tooltip.showPercentage,
          val: `<span>${oe(
            e.value.style.chart.layout.labels.dataLabels.formatter,
            a.value,
            F({
              p: e.value.style.chart.layout.labels.dataLabels.prefix,
              v: a.value,
              s: e.value.style.chart.layout.labels.dataLabels.suffix,
              r: e.value.style.chart.tooltip.roundingValue
            }),
            {
              datapoint: a,
              seriesIndex: i.value[t].absoluteIndex,
              series: g.value
            }
          )}</span>`,
          percentage: `<span>${F({
            v: J.value ? 1 / G.value.length * 100 : a.value / _.value * 100,
            s: "%",
            r: e.value.style.chart.tooltip.roundingPercentage
          })}</span>`
        })}</b></div>`, ge.value = o;
      }
    }
    const Oe = bt, L = c(() => {
      const t = H.value.map((l) => ({
        name: l.name,
        color: l.color
      })), a = H.value.map((l) => l.value);
      return { head: t, body: a };
    });
    function ie(t) {
      return e.value.useBlurOnHover && ![null, void 0].includes(W.value) && W.value !== t ? `url(#blur_${m.value})` : "";
    }
    function qt(t, a) {
      return e.value.style.chart.layout.labels.captions.show ? i.value.length && !re.value && !e.value.style.chart.layout.grid.vertical && (i.value[t].isFirst && a.position < e.value.style.chart.layout.grid.size - 2 || i.value[t].isAbsoluteFirst && t % e.value.style.chart.layout.grid.size === 0 && i.value[t].absoluteStartIndex) : !1;
    }
    function Xt(t, a = null) {
      const l = oe(
        e.value.style.chart.layout.labels.dataLabels.formatter,
        i.value[t].value,
        F({
          p: e.value.style.chart.layout.labels.dataLabels.prefix,
          v: i.value[t].value,
          s: e.value.style.chart.layout.labels.dataLabels.suffix,
          r: e.value.style.chart.layout.labels.captions.roundingValue
        }),
        { datapoint: i.value[t], position: a }
      ), o = F({ v: i.value[t].proportion, s: "%", r: e.value.style.chart.layout.labels.captions.roundingPercentage }), n = (e.value.style.chart.layout.labels.captions.serieNameAbbreviation ? wl({ source: i.value[t].name, length: e.value.style.chart.layout.labels.captions.serieNameMaxAbbreviationSize }) : i.value[t].name) + (e.value.style.chart.layout.labels.captions.showPercentage || e.value.style.chart.layout.labels.captions.showValue ? ":" : ""), u = e.value.style.chart.layout.labels.captions.showSerieName ? n : "", w = Ae({
        val: l,
        percentage: o,
        showVal: e.value.style.chart.layout.labels.captions.showValue,
        showPercentage: e.value.style.chart.layout.labels.captions.showPercentage,
        config: e.value.style.chart.layout.labels.dataLabels
      });
      return `${u} ${w}`;
    }
    function _e(t = null) {
      pt(() => {
        const a = L.value.head.map((n, u) => [[
          n.name
        ], [L.value.body[u]], [isNaN(L.value.body[u] / _.value) ? "-" : L.value.body[u] / _.value * 100]]), l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[""], ["val"], ["%"]]].concat(a), o = ml(l);
        t ? t(o) : bl({ csvContent: o, title: e.value.style.chart.title.text || "vue-ui-waffle" });
      });
    }
    const ce = c(() => {
      const t = [
        ' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>',
        oe(
          e.value.style.chart.layout.labels.dataLabels.formatter,
          _.value,
          F({
            p: e.value.style.chart.layout.labels.dataLabels.prefix,
            v: _.value,
            s: e.value.style.chart.layout.labels.dataLabels.suffix,
            r: e.value.table.td.roundingValue
          })
        ),
        "100%"
      ], a = L.value.head.map((n, u) => [
        {
          color: n.color,
          name: n.name
        },
        oe(
          e.value.style.chart.layout.labels.dataLabels.formatter,
          L.value.body[u],
          F({
            p: e.value.style.chart.layout.labels.dataLabels.prefix,
            v: L.value.body[u],
            s: e.value.style.chart.layout.labels.dataLabels.suffix,
            r: e.value.table.td.roundingValue
          })
        ),
        isNaN(L.value.body[u] / _.value) ? "-" : F({
          v: L.value.body[u] / _.value * 100,
          s: "%",
          r: e.value.table.td.roundingPercentage
        })
      ]), l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        shape: "square",
        breakpoint: e.value.table.responsiveBreakpoint
      }, o = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.value,
        e.value.table.columnNames.percentage
      ];
      return {
        head: t,
        body: a,
        config: l,
        colNames: o
      };
    }), U = r(!1);
    function nt(t) {
      U.value = t, Ne.value += 1;
    }
    function ut() {
      C.value.showTable = !C.value.showTable;
    }
    function st() {
      C.value.showTooltip = !C.value.showTooltip;
    }
    const ve = r(!1);
    function Le() {
      ve.value = !ve.value;
    }
    async function Yt({ scale: t = 2 } = {}) {
      if (!A.value) return;
      const { width: a, height: l } = A.value.getBoundingClientRect(), o = a / l, { imageUri: n, base64: u } = await Fl({ domElement: A.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: n,
        base64: u,
        title: e.value.style.chart.title.text,
        width: a,
        height: l,
        aspectRatio: o
      };
    }
    const te = c(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = C.value.showTable;
      return {
        component: t ? At : $t,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: U.value,
          fullscreenParent: A.value,
          forcedWidth: Math.min(500, window.innerWidth * 0.8),
          isCursorPointer: M.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    le(() => C.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && Y.value ? Y.value.open() : "close" in Y.value && Y.value.close());
    });
    function rt() {
      C.value.showTable = !1, ye.value && ye.value.setTableIconState(!1);
    }
    const Jt = c(() => z.value.map((t) => ({
      ...t,
      name: t.display
    }))), Kt = c(() => e.value.style.chart.backgroundColor), Qt = c(() => e.value.style.chart.legend), Zt = c(() => e.value.style.chart.title), { exportSvg: el, getSvg: tl } = Ol({
      svg: ke,
      title: Zt,
      legend: Qt,
      legendItems: Jt,
      backgroundColor: Kt
    });
    async function it({ isCb: t }) {
      ue.value = !0, await pt();
      try {
        if (t) {
          const { blob: a, url: l, text: o, dataUrl: n } = await tl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: o, dataUrl: n }));
        } else
          await Promise.resolve(el());
      } finally {
        ue.value = !1;
      }
    }
    function ll(t) {
      if (t?.stage === "start") {
        ne.value = !0;
        return;
      }
      if (t?.stage === "end") {
        ne.value = !1;
        return;
      }
      Qe();
    }
    async function ct() {
      if (Oe("copyAlt", {
        config: e.value,
        dataset: G.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: G.value
      }));
    }
    return mt({
      getData: Vt,
      getImage: Yt,
      generatePdf: Ke,
      generateCsv: _e,
      generateImage: Qe,
      generateSvg: it,
      hideSeries: Et,
      showSeries: Ut,
      toggleTable: ut,
      toggleTooltip: st,
      toggleAnnotator: Le,
      toggleFullscreen: nt,
      copyAlt: ct
    }), (t, a) => (s(), v("div", {
      class: dt(`vue-data-ui-component vue-ui-waffle ${U.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      ref_key: "waffleChart",
      ref: A,
      id: `vue-ui-waffle_${m.value}`,
      style: ae(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: a[2] || (a[2] = () => d(qe)(!0)),
      onMouseleave: a[3] || (a[3] = () => d(qe)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (s(), B(d(Tt), {
        key: 0,
        svgRef: d(ke),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: ve.value,
        isCursorPointer: M.value,
        onClose: Le
      }, {
        "annotator-action-close": h(() => [
          p(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": h(({ color: l }) => [
          p(t.$slots, "annotator-action-color", $(x({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": h(({ mode: l }) => [
          p(t.$slots, "annotator-action-draw", $(x({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": h(({ disabled: l }) => [
          p(t.$slots, "annotator-action-undo", $(x({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": h(({ disabled: l }) => [
          p(t.$slots, "annotator-action-redo", $(x({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": h(({ disabled: l }) => [
          p(t.$slots, "annotator-action-delete", $(x({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : b("", !0),
      Bt.value ? (s(), v("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Ee,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : b("", !0),
      e.value.style.chart.title.text ? (s(), v("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Me,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (s(), B(zl, {
          key: `title_${We.value}`,
          config: {
            title: {
              cy: "waffle-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "waffle-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : b("", !0),
      k("div", {
        id: `legend-top-${m.value}`
      }, null, 8, El),
      e.value.userOptions.show && pe.value && (d(Xe) || d(we)) ? (s(), B(d(xt), {
        ref_key: "userOptionsRef",
        ref: ye,
        key: `user_options_${Ne.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: d($e),
        isImaging: d(xe),
        uid: m.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: U.value,
        isTooltip: C.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: A.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: ve.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: M.value,
        onToggleFullscreen: nt,
        onGeneratePdf: d(Ke),
        onGenerateCsv: _e,
        onGenerateImage: ll,
        onGenerateSvg: it,
        onToggleTable: ut,
        onToggleTooltip: st,
        onToggleAnnotator: Le,
        onCopyAlt: ct,
        style: ae({
          visibility: d(Xe) ? d(we) ? "visible" : "hidden" : "visible"
        })
      }, Fe({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: h(({ isOpen: l, color: o }) => [
            p(t.$slots, "menuIcon", $(x({ isOpen: l, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: h(() => [
            p(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: h(() => [
            p(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: h(() => [
            p(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: h(() => [
            p(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: h(() => [
            p(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: h(() => [
            p(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: h(({ toggleFullscreen: l, isFullscreen: o }) => [
            p(t.$slots, "optionFullscreen", $(x({ toggleFullscreen: l, isFullscreen: o })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: h(({ toggleAnnotator: l, isAnnotator: o }) => [
            p(t.$slots, "optionAnnotator", $(x({ toggleAnnotator: l, isAnnotator: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: h(({ altCopy: l }) => [
            p(t.$slots, "optionAltCopy", $(x({ altCopy: l })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : b("", !0),
      (s(), v("svg", {
        ref_key: "svgRef",
        ref: ke,
        xmlns: d(hl),
        class: dt({ "vue-data-ui-fullscreen--on": U.value, "vue-data-ui-fulscreen--off": !U.value }),
        viewBox: `0 0 ${R.value.width <= 0 ? 10 : R.value.width} ${R.value.height <= 0 ? 10 : R.value.height}`,
        style: ae(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        de(d(St)),
        k("defs", null, [
          (s(!0), v(S, null, N(i.value, (l, o) => (s(), v("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `gradient_${m.value}_${o}`
          }, [
            k("stop", {
              offset: "0%",
              "stop-color": d(pl)(d(fl)(l.color, 0.05), 100 - e.value.style.chart.layout.rect.gradientIntensity)
            }, null, 8, jl),
            k("stop", {
              offset: "100%",
              "stop-color": l.color
            }, null, 8, Hl)
          ], 8, Gl))), 256))
        ]),
        k("defs", null, [
          k("filter", {
            id: `blur_${m.value}`,
            x: "-50%",
            y: "-50%",
            width: "200%",
            height: "200%"
          }, [...a[4] || (a[4] = [
            k("feGaussianBlur", {
              in: "SourceGraphic",
              stdDeviation: 2
            }, null, -1),
            k("feColorMatrix", {
              type: "saturate",
              values: "0"
            }, null, -1)
          ])], 8, ql)
        ]),
        e.value.useCustomCells && i.value.length ? (s(!0), v(S, { key: 0 }, N(q.value, (l, o) => (s(), v("foreignObject", {
          x: l.x,
          y: l.y,
          height: O.value <= 0 ? 1e-4 : O.value,
          width: I.value <= 0 ? 1e-4 : I.value,
          class: "vue-ui-waffle-custom-cell-foreignObject"
        }, [
          p(t.$slots, "cell", De({ ref_for: !0 }, { cell: { ...l, color: i.value[o].color, ...i.value[o] }, isSelected: [null, void 0].includes(W.value) ? !0 : i.value[o].serieIndex === W.value }), void 0, !0)
        ], 8, Xl))), 256)) : b("", !0),
        !i.value.length && !e.value.useCustomCells ? (s(), v("rect", {
          key: 1,
          x: 12,
          y: 12,
          height: P.value.height - 24,
          width: P.value.width - 24,
          rx: 3,
          fill: "none",
          stroke: "black"
        }, null, 8, Yl)) : i.value.length && !e.value.useCustomCells ? (s(), v(S, { key: 2 }, [
          t.$slots.pattern ? (s(), v("g", Jl, [
            (s(!0), v(S, null, N(se.value, (l) => (s(), v("defs", null, [
              p(t.$slots, "pattern", De({ ref_for: !0 }, { seriesIndex: l.absoluteIndex, patternId: `pattern_${m.value}_${l.absoluteIndex}` }), void 0, !0)
            ]))), 256))
          ])) : b("", !0),
          (s(!0), v(S, null, N(q.value, (l, o) => (s(), v("rect", {
            rx: e.value.style.chart.layout.rect.rounded ? e.value.style.chart.layout.rect.rounding : 0,
            x: l.x + e.value.style.chart.layout.grid.spaceBetween / 2,
            y: l.y + e.value.style.chart.layout.grid.spaceBetween / 2,
            height: O.value <= 0 ? 1e-4 : O.value,
            width: I.value <= 0 ? 1e-4 : I.value,
            fill: "white",
            stroke: e.value.style.chart.layout.rect.stroke,
            "stroke-width": e.value.style.chart.layout.rect.strokeWidth,
            filter: ie(i.value[o].serieIndex)
          }, null, 8, Kl))), 256)),
          (s(!0), v(S, null, N(q.value, (l, o) => (s(), v("rect", {
            rx: e.value.style.chart.layout.rect.rounded ? e.value.style.chart.layout.rect.rounding : 0,
            x: l.x + e.value.style.chart.layout.grid.spaceBetween / 2,
            y: l.y + e.value.style.chart.layout.grid.spaceBetween / 2,
            height: O.value <= 0 ? 1e-4 : O.value,
            width: I.value <= 0 ? 1e-4 : I.value,
            fill: e.value.style.chart.layout.rect.useGradient && e.value.style.chart.layout.rect.gradientIntensity > 0 ? `url(#gradient_${m.value}_${o})` : i.value[o].color,
            stroke: e.value.style.chart.layout.rect.stroke,
            "stroke-width": e.value.style.chart.layout.rect.strokeWidth,
            filter: ie(i.value[o].serieIndex)
          }, null, 8, Ql))), 256)),
          t.$slots.pattern ? (s(), v("g", Zl, [
            (s(!0), v(S, null, N(q.value, (l, o) => (s(), v("rect", {
              rx: e.value.style.chart.layout.rect.rounded ? e.value.style.chart.layout.rect.rounding : 0,
              x: l.x + e.value.style.chart.layout.grid.spaceBetween / 2,
              y: l.y + e.value.style.chart.layout.grid.spaceBetween / 2,
              height: O.value <= 0 ? 1e-4 : O.value,
              width: I.value <= 0 ? 1e-4 : I.value,
              fill: `url(#pattern_${m.value}_${i.value[o].absoluteIndex})`,
              stroke: "none",
              filter: ie(i.value[o].serieIndex)
            }, null, 8, ea))), 256))
          ])) : b("", !0)
        ], 64)) : b("", !0),
        (s(!0), v(S, null, N(q.value, (l, o) => (s(), v(S, null, [
          qt(o, l) ? (s(), v("text", {
            key: `datalabel_${o}`,
            textContent: he(Xt(o, l)),
            x: l.x + e.value.style.chart.layout.labels.captions.offsetX + e.value.style.chart.layout.grid.spaceBetween / 2 + 6,
            y: l.y + e.value.style.chart.layout.labels.captions.offsetY + e.value.style.chart.layout.grid.spaceBetween / 2 + et.value / 2 + e.value.style.chart.layout.labels.captions.fontSize / 3,
            "font-size": e.value.style.chart.layout.labels.captions.fontSize,
            fill: d(gl)(i.value[o].color),
            filter: ie(i.value[o].serieIndex)
          }, null, 8, ta)) : b("", !0)
        ], 64))), 256)),
        (s(!0), v(S, null, N(q.value, (l, o) => (s(), v("rect", {
          x: l.x + e.value.style.chart.layout.grid.spaceBetween / 2,
          y: l.y + e.value.style.chart.layout.grid.spaceBetween / 2,
          height: et.value,
          width: Ft.value,
          fill: "transparent",
          stroke: "none",
          onMouseover: (n) => Ht(o),
          onMouseleave: (n) => jt(o),
          onClick: (n) => Gt(o)
        }, null, 40, la))), 256)),
        p(t.$slots, "svg", {
          svg: {
            ...R.value,
            isPrintingImg: d($e) | d(xe) | ne.value,
            isPrintingSvg: ue.value
          }
        }, void 0, !0)
      ], 14, Wl)),
      t.$slots.watermark ? (s(), v("div", aa, [
        p(t.$slots, "watermark", $(x({ isPrinting: d($e) || d(xe) || ne.value || ue.value })), void 0, !0)
      ])) : b("", !0),
      k("div", {
        id: `legend-bottom-${m.value}`
      }, null, 8, oa),
      He.value ? (s(), B(sl, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${m.value}` : `#legend-bottom-${m.value}`
      }, [
        k("div", {
          ref_key: "chartLegend",
          ref: Re
        }, [
          e.value.style.chart.legend.show ? (s(), B(Nl, {
            key: `legend_${je.value}`,
            legendSet: z.value,
            config: Wt.value,
            isCursorPointer: M.value,
            onClickMarker: a[0] || (a[0] = ({ legend: l }) => X(l.uid))
          }, Fe({
            item: h(({ legend: l }) => [
              k("div", {
                onClick: (o) => l.segregate(),
                style: ae(`opacity:${f.value.includes(l.uid) ? 0.5 : 1}`)
              }, he(l.display), 13, na)
            ]),
            legendToggle: h(() => [
              z.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !d(me) ? (s(), B(Ml, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: f.value.length > 0,
                isCursorPointer: M.value,
                onToggle: Rt
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : b("", !0)
            ]),
            _: 2
          }, [
            t.$slots.pattern ? {
              name: "legend-pattern",
              fn: h(({ legend: l, index: o }) => [
                de(Dl, {
                  shape: l.shape,
                  radius: 30,
                  stroke: "none",
                  plot: { x: 30, y: 30 },
                  fill: `url(#pattern_${m.value}_${o})`
                }, null, 8, ["shape", "fill"])
              ]),
              key: "0"
            } : void 0
          ]), 1032, ["legendSet", "config", "isCursorPointer"])) : p(t.$slots, "legend", {
            key: 1,
            legend: z.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : b("", !0),
      t.$slots.source ? (s(), v("div", {
        key: 6,
        ref_key: "source",
        ref: Ue,
        dir: "auto"
      }, [
        p(t.$slots, "source", {}, void 0, !0)
      ], 512)) : b("", !0),
      de(d(wt), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: C.value.showTooltip && fe.value && f.value.length < y.dataset.length,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: A.value,
        content: ge.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        fontSize: e.value.style.chart.tooltip.fontSize,
        isFullscreen: U.value,
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": h(() => [
          p(t.$slots, "tooltip-before", $(x({ ...Pe.value })), void 0, !0)
        ]),
        "tooltip-after": h(() => [
          p(t.$slots, "tooltip-after", $(x({ ...Pe.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isCustom", "fontSize", "isFullscreen", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      pe.value && e.value.userOptions.buttons.table ? (s(), B(rl(te.value.component), De({ key: 7 }, te.value.props, {
        ref_key: "tableUnit",
        ref: Y,
        onClose: rt
      }), Fe({
        content: h(() => [
          (s(), B(d(Ct), {
            key: `table_${Ge.value}`,
            colNames: ce.value.colNames,
            head: ce.value.head,
            body: ce.value.body,
            config: ce.value.config,
            title: e.value.table.useDialog ? "" : te.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: M.value,
            onClose: rt
          }, {
            th: h(({ th: l }) => [
              k("div", {
                innerHTML: l,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, ua)
            ]),
            td: h(({ td: l }) => [
              ht(he(l.name || l), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: h(() => [
            ht(he(te.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: h(() => [
            k("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[1] || (a[1] = (l) => _e(e.value.userOptions.callbacks.csv)),
              style: ae({ cursor: M.value ? "pointer" : "default" })
            }, [
              de(d(kt), {
                name: "fileCsv",
                stroke: te.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : b("", !0),
      p(t.$slots, "skeleton", {}, () => [
        d(me) ? (s(), B(Pl, { key: 0 })) : b("", !0)
      ], !0)
    ], 46, Ul));
  }
}, Aa = /* @__PURE__ */ Rl(sa, [["__scopeId", "data-v-8be3d29f"]]);
export {
  Aa as default
};
