// vue-data-ui/dist/vue-ui-tiremarks-UBDpAIVr.js
import { defineAsyncComponent as Y, computed as d, ref as i, toRefs as De, watch as ge, onMounted as Ge, openBlock as r, createElementBlock as c, unref as n, normalizeStyle as K, normalizeClass as W, createBlock as B, withCtx as f, renderSlot as u, normalizeProps as m, guardReactiveProps as k, createCommentVNode as w, createSlots as je, createVNode as He, Fragment as pe, renderList as ye, toDisplayString as Ve, nextTick as Xe } from "vue";
import { j as qe, t as me, K as Ye, o as Ke, q as We, v as ke, X as Je, k as Qe, l as Ze } from "./lib-GwPnJSrY.js";
import { t as et, u as tt } from "./useResponsive-DfdjqQps.js";
import { u as at } from "./useConfig-CIgyKZ6R.js";
import { u as ot, B as lt } from "./BaseScanner-BhhkKVoj.js";
import { u as nt } from "./usePrinter-DiK5R6k-.js";
import { u as st } from "./useSvgExport-ByUukOZt.js";
import { u as J } from "./useNestedProp-BBT0mbHh.js";
import { u as rt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ut } from "./useUserOptionState-BIvW1Kz7.js";
import { u as it } from "./useChartAccessibility-9icAAmYg.js";
import ct from "./vue_ui_tiremarks-DBotWfi8.js";
import vt from "./Title-BfyrP-jp.js";
import ht from "./img-BYYzY_QU.js";
import { _ as dt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ft = ["id"], gt = ["xmlns", "viewBox"], pt = ["width", "height"], yt = { key: 1 }, mt = ["d", "stroke-width", "stroke"], kt = { key: 2 }, bt = ["x1", "y1", "x2", "y2", "stroke-width", "stroke"], Ct = { key: 3 }, wt = ["x", "y", "height"], $t = ["x", "y", "font-size", "fill", "font-weight", "text-anchor"], At = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Pt = {
  __name: "vue-ui-tiremarks",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["copyAlt"],
  setup(be, { expose: Ce, emit: we }) {
    const $e = Y(() => import("./PenAndPaper-DWV1FMIi.js")), Ae = Y(() => import("./UserOptions-CrLqEimB.js")), Pe = Y(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_tiremarks: Oe } = at(), { isThemeValid: Se, warnInvalidTheme: _e } = rt(), b = be, xe = we, Ie = d(() => !!b.dataset && Object.keys(b.dataset).length), D = i(qe()), O = i(null), ze = i(null), Q = i(null), Z = i(null), ee = i(0), te = i(0), S = i(null), z = i(null), T = i(!1), F = i(!1), e = i(H()), ae = d(() => e.value.userOptions.useCursorPointer), Te = d(() => me({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            animation: { use: !1 },
            layout: {
              activeColor: "#6A6A6A80",
              inactiveColor: "#CACACA80",
              ticks: {
                gradient: { show: !1 }
              }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: oe, FINAL_DATASET: _ } = ot({
      ...De(b),
      FINAL_CONFIG: e,
      prepareConfig: H,
      skeletonDataset: b.config?.skeletonDataset ?? { percentage: 50 },
      skeletonConfig: me({
        defaultConfig: e.value,
        userConfig: Te.value
      })
    }), { userOptionsVisible: G, setUserOptionsVisibility: le, keepUserOptionState: ne } = ut({ config: e.value }), { svgRef: j } = it({ config: e.value.style.chart.title });
    function H() {
      const t = J({
        userConfig: b.config,
        defaultConfig: Oe
      }), o = t.theme;
      if (!o) return t;
      if (!Se.value(t))
        return _e(t), t;
      const a = J({
        userConfig: ct[o] || b.config,
        defaultConfig: t
      });
      return J({
        userConfig: b.config,
        defaultConfig: a
      });
    }
    ge(() => b.config, (t) => {
      e.value = H(), G.value = !e.value.userOptions.showOnChartHover, E.value = e.value.style.chart.width, M.value = e.value.style.chart.height, ie(), te.value += 1;
    }, { deep: !0 });
    const { isPrinting: V, isImaging: X, generatePdf: se, generateImage: re } = nt({
      elementId: D.value,
      fileName: e.value.style.chart.title.text || "vue-ui-tiremarks",
      options: e.value.userOptions.print
    }), Fe = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), v = i(e.value.style.chart.animation.use ? 0 : Ye(_.value.percentage));
    ge(() => _.value, (t) => {
      e.value.style.chart.animation.use ? ue(t.percentage) : v.value = t.percentage || 0;
    }, { deep: !0 }), Ge(() => {
      ie();
    });
    function ue(t) {
      let o = e.value.style.chart.animation.speed;
      const a = Math.abs(t - v.value) / (o * 120);
      function l() {
        v.value < t ? v.value = Math.min(v.value + a, t) : v.value > t && (v.value = Math.max(v.value - a, t)), v.value !== t && requestAnimationFrame(l);
      }
      l();
    }
    const Ne = d(() => !!e.value.debug);
    function ie() {
      if (Ke(b.dataset) && We({
        componentName: "VueUiTiremarks",
        type: "dataset",
        debug: Ne.value
      }), e.value.responsive) {
        const t = et(() => {
          const { width: o, height: a } = tt({
            chart: O.value,
            title: e.value.style.chart.title.text ? Q.value : null,
            source: Z.value
          });
          requestAnimationFrame(() => {
            E.value = Math.max(0.1, o), M.value = Math.max(0.1, a - 12);
          });
        });
        S.value && (z.value && S.value.unobserve(z.value), S.value.disconnect()), S.value = new ResizeObserver(t), z.value = O.value.parentNode, S.value.observe(z.value);
      }
      ue(_.value.percentage || 0);
    }
    const N = d(() => e.value.style.chart.layout.display === "vertical"), h = d(() => {
      const t = e.value.style.chart.percentage.show, o = {
        top: t ? 48 : 12,
        left: t ? 64 : 16,
        right: t ? 64 : 16,
        bottom: t ? 48 : 12
      };
      return N.value ? {
        top: e.value.style.chart.percentage.verticalPosition === "top" ? o.top : 3,
        left: 3,
        right: 3,
        bottom: e.value.style.chart.percentage.verticalPosition === "bottom" ? o.bottom : 3
      } : {
        top: 0,
        bottom: 0,
        left: e.value.style.chart.percentage.horizontalPosition === "left" ? o.left : 16,
        right: e.value.style.chart.percentage.horizontalPosition === "right" ? o.right : 10
      };
    }), R = d(() => Object.values(h.value).reduce((t, o) => t + o, 0)), E = i(e.value.style.chart.width), M = i(e.value.style.chart.height), s = d(() => ({
      height: M.value,
      width: E.value
    })), ce = d(() => ({
      horizontal: {
        x: y.value.x + (e.value.style.chart.percentage.horizontalPosition === "left" ? 6 : 3),
        y: s.value.height / 2 - y.value.fontSize / 2
      },
      vertical: {
        x: s.value.width / 2 - 20,
        y: y.value.y - y.value.fontSize / 2
      }
    })[e.value.style.chart.layout.display]), g = d(() => N.value ? {
      mark: (s.value.height - R.value) / 100 * 0.5,
      space: (s.value.height - R.value) / 100 * 0.5
    } : {
      mark: (s.value.width - R.value) / 100 * 0.5,
      space: (s.value.width - R.value) / 100 * 0.5
    }), ve = d(() => {
      const t = [];
      for (let a = 0; a < 100; a += 1) {
        const l = e.value.style.chart.layout.ticks.gradient.show ? ke(e.value.style.chart.layout.activeColor, a / 100 * (e.value.style.chart.layout.ticks.gradient.shiftHueIntensity / 100)) : e.value.style.chart.layout.activeColor;
        if (N.value) {
          const C = e.value.style.chart.layout.crescendo ? (100 - a) * (s.value.width - h.value.left - h.value.right) / 100 / 3 : 0, p = h.value.left + 4 + C, $ = s.value.width - h.value.right - 4 - C, A = s.value.height - h.value.bottom - a * g.value.mark - a * g.value.space - g.value.mark, P = s.value.height - h.value.bottom - a * g.value.mark - a * g.value.space - g.value.mark, x = ($ - p) / e.value.style.chart.layout.curveAngleX, I = e.value.style.chart.layout.curveAngleY * ((1 + a) / 100);
          t.push({
            x1: p,
            x2: $,
            y1: A,
            y2: P,
            curve: `M ${p} ${A} C ${p + x} ${A - I}, ${$ - x} ${P - I}, ${$} ${P}`,
            color: l
          });
        } else {
          const C = e.value.style.chart.layout.crescendo ? (100 - a) * (s.value.height - h.value.top - h.value.bottom) / 100 / 3 : 0, p = h.value.left + a * g.value.mark + a * g.value.space - g.value.mark, $ = p, A = h.value.top + 4 + C, P = s.value.height - h.value.bottom - 4 - C, x = e.value.style.chart.layout.curveAngleY * ((1 + a) / 100), I = (P - A) / e.value.style.chart.layout.curveAngleX;
          t.push({
            x1: p,
            x2: $,
            y1: A,
            y2: P,
            curve: `M ${p} ${A} C ${p + x} ${A + I}, ${$ + x} ${P - I}, ${$} ${P}`,
            color: l
          });
        }
      }
      return t;
    }), y = d(() => {
      let t, o, a;
      const l = e.value.style.chart.percentage.fontSize / 3;
      return N.value ? e.value.style.chart.percentage.verticalPosition === "top" ? (t = s.value.width / 2, o = h.value.top / 2, a = "middle") : e.value.style.chart.percentage.verticalPosition === "bottom" && (t = s.value.width / 2, o = s.value.height - h.value.bottom / 2 + l, a = "middle") : e.value.style.chart.percentage.horizontalPosition === "left" ? (t = 4, o = s.value.height / 2 + l, a = "start") : e.value.style.chart.percentage.horizontalPosition === "right" && (t = s.value.width - h.value.right + 8, o = s.value.height / 2 + l, a = "start"), {
        x: t,
        y: o,
        textAnchor: a,
        bold: e.value.style.chart.percentage.bold,
        fontSize: e.value.style.chart.percentage.fontSize,
        fill: e.value.style.chart.percentage.color
      };
    }), U = i(!1);
    function he(t) {
      U.value = t, ee.value += 1;
    }
    const L = i(!1);
    function q() {
      L.value = !L.value;
    }
    async function Re({ scale: t = 2 } = {}) {
      if (!O.value) return;
      const { width: o, height: a } = O.value.getBoundingClientRect(), l = o / a, { imageUri: C, base64: p } = await ht({ domElement: O.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: C,
        base64: p,
        title: e.value.style.chart.title.text,
        width: o,
        height: a,
        aspectRatio: l
      };
    }
    const Ee = d(() => e.value.style.chart.backgroundColor), Me = d(() => e.value.style.chart.title), { exportSvg: Ue, getSvg: Le } = st({
      svg: j,
      title: Me,
      backgroundColor: Ee
    });
    async function de({ isCb: t }) {
      F.value = !0, await Xe();
      try {
        if (t) {
          const { blob: o, url: a, text: l, dataUrl: C } = await Le();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: o, url: a, text: l, dataUrl: C }));
        } else
          await Promise.resolve(Ue());
      } finally {
        F.value = !1;
      }
    }
    function Be(t) {
      if (t?.stage === "start") {
        T.value = !0;
        return;
      }
      if (t?.stage === "end") {
        T.value = !1;
        return;
      }
      re();
    }
    async function fe() {
      if (xe("copyAlt", {
        config: e.value,
        dataset: _.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: _.value
      }));
    }
    return Ce({
      getImage: Re,
      generatePdf: se,
      generateImage: re,
      generateSvg: de,
      toggleAnnotator: q,
      toggleFullscreen: he,
      copyAlt: fe
    }), (t, o) => (r(), c("div", {
      ref_key: "tiremarksChart",
      ref: O,
      class: W(`vue-data-ui-component vue-ui-tiremarks ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: K(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor}`),
      id: D.value,
      onMouseenter: o[0] || (o[0] = () => n(le)(!0)),
      onMouseleave: o[1] || (o[1] = () => n(le)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (r(), B(n($e), {
        key: 0,
        svgRef: n(j),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: L.value,
        isCursorPointer: ae.value,
        onClose: q
      }, {
        "annotator-action-close": f(() => [
          u(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": f(({ color: a }) => [
          u(t.$slots, "annotator-action-color", m(k({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": f(({ mode: a }) => [
          u(t.$slots, "annotator-action-draw", m(k({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": f(({ disabled: a }) => [
          u(t.$slots, "annotator-action-undo", m(k({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": f(({ disabled: a }) => [
          u(t.$slots, "annotator-action-redo", m(k({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": f(({ disabled: a }) => [
          u(t.$slots, "annotator-action-delete", m(k({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : w("", !0),
      Fe.value ? (r(), c("div", {
        key: 1,
        ref_key: "noTitle",
        ref: ze,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : w("", !0),
      e.value.style.chart.title.text ? (r(), c("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Q,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (r(), B(vt, {
          key: `title_${te.value}`,
          config: {
            title: {
              cy: "wheel-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "wheel-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : w("", !0),
      e.value.userOptions.show && Ie.value && (n(ne) || n(G)) ? (r(), B(n(Ae), {
        ref: "details",
        key: `user_options_${ee.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: n(V),
        isImaging: n(X),
        uid: D.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        hasXls: !1,
        isFullscreen: U.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: O.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: L.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        isCursorPointer: ae.value,
        onToggleFullscreen: he,
        onGeneratePdf: n(se),
        onGenerateImage: Be,
        onGenerateSvg: de,
        onToggleAnnotator: q,
        onCopyAlt: fe,
        style: K({
          visibility: n(ne) ? n(G) ? "visible" : "hidden" : "visible"
        })
      }, je({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: f(({ isOpen: a, color: l }) => [
            u(t.$slots, "menuIcon", m(k({ isOpen: a, color: l })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: f(() => [
            u(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: f(() => [
            u(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: f(() => [
            u(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: f(({ toggleFullscreen: a, isFullscreen: l }) => [
            u(t.$slots, "optionFullscreen", m(k({ toggleFullscreen: a, isFullscreen: l })), void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: f(({ toggleAnnotator: a, isAnnotator: l }) => [
            u(t.$slots, "optionAnnotator", m(k({ toggleAnnotator: a, isAnnotator: l })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: f(({ altCopy: a }) => [
            u(t.$slots, "optionAltCopy", m(k({ altCopy: a })), void 0, !0)
          ]),
          key: "6"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "isCursorPointer", "onGeneratePdf", "style"])) : w("", !0),
      (r(), c("svg", {
        ref_key: "svgRef",
        ref: j,
        xmlns: n(Je),
        class: W({ "vue-data-ui-fullscreen--on": U.value, "vue-data-ui-fulscreen--off": !U.value }),
        viewBox: `0 0 ${E.value} ${M.value}`,
        style: K(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color}`)
      }, [
        He(n(Pe)),
        t.$slots["chart-background"] ? (r(), c("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: s.value.width,
          height: s.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          u(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, pt)) : w("", !0),
        e.value.style.chart.layout.curved ? (r(), c("g", yt, [
          (r(!0), c(pe, null, ye(ve.value, (a, l) => (r(), c("path", {
            d: a.curve,
            "stroke-width": g.value.mark,
            stroke: v.value >= l ? a.color : e.value.style.chart.layout.inactiveColor,
            "stroke-linecap": "round",
            fill: "none",
            class: W({ "vue-ui-tick-animated": e.value.style.chart.animation.use && l <= v.value })
          }, null, 10, mt))), 256))
        ])) : (r(), c("g", kt, [
          (r(!0), c(pe, null, ye(ve.value, (a, l) => (r(), c("line", {
            x1: a.x1,
            y1: a.y1,
            x2: a.x2,
            y2: a.y2,
            "stroke-width": g.value.mark,
            stroke: v.value >= l ? a.color : e.value.style.chart.layout.inactiveColor,
            "stroke-linecap": "round"
          }, null, 8, bt))), 256))
        ])),
        e.value.style.chart.percentage.show ? (r(), c("g", Ct, [
          n(oe) ? (r(), c("rect", {
            key: 0,
            x: ce.value.x,
            y: ce.value.y,
            width: 40,
            height: y.value.fontSize,
            fill: "#6A6A6A80",
            rx: 3
          }, null, 8, wt)) : (r(), c("text", {
            key: 1,
            x: y.value.x,
            y: y.value.y,
            "font-size": y.value.fontSize,
            fill: e.value.style.chart.layout.ticks.gradient.show && e.value.style.chart.percentage.useGradientColor ? n(ke)(e.value.style.chart.layout.activeColor, v.value / 100 * (e.value.style.chart.layout.ticks.gradient.shiftHueIntensity / 100)) : e.value.style.chart.percentage.color,
            "font-weight": y.value.bold ? "bold" : "normal",
            "text-anchor": y.value.textAnchor
          }, Ve(n(Qe)(
            e.value.style.chart.percentage.formatter,
            v.value,
            n(Ze)({
              v: v.value,
              s: "%",
              r: e.value.style.chart.percentage.rounding
            })
          )), 9, $t))
        ])) : w("", !0),
        u(t.$slots, "svg", {
          svg: {
            ...s.value,
            isPrintingImg: n(V) | n(X) | T.value,
            isPrintingSvg: F.value
          }
        }, void 0, !0)
      ], 14, gt)),
      t.$slots.watermark ? (r(), c("div", At, [
        u(t.$slots, "watermark", m(k({ isPrinting: n(V) || n(X) || T.value || F.value })), void 0, !0)
      ])) : w("", !0),
      t.$slots.source ? (r(), c("div", {
        key: 5,
        ref_key: "source",
        ref: Z,
        dir: "auto"
      }, [
        u(t.$slots, "source", {}, void 0, !0)
      ], 512)) : w("", !0),
      u(t.$slots, "skeleton", {}, () => [
        n(oe) ? (r(), B(lt, { key: 0 })) : w("", !0)
      ], !0)
    ], 46, ft));
  }
}, Dt = /* @__PURE__ */ dt(Pt, [["__scopeId", "data-v-f381cb10"]]);
export {
  Dt as default
};
