// vue-data-ui/dist/vue-data-ui-C2kPofmi.js
import { toRefs as ft, defineAsyncComponent as e, computed as c, ref as o, onMounted as St, watch as ot, shallowReactive as ht, openBlock as g, createElementBlock as kt, normalizeStyle as bt, createElementVNode as yt, createBlock as d, createCommentVNode as it, createTextVNode as Ct, toDisplayString as At, resolveDynamicComponent as Tt, mergeProps as Pt, toHandlers as zt, createSlots as Dt, renderList as Lt, withCtx as It, renderSlot as Et, normalizeProps as Mt, guardReactiveProps as Ot } from "vue";
import at from "./BaseIcon-CNx8Tzl5.js";
const Nt = { style: { width: "36px" } }, Ft = {
  __name: "vue-data-ui",
  props: {
    component: { type: String },
    dataset: { type: [Object, Array, Number, String] },
    config: { type: Object }
  },
  emits: [
    "selectLegend",
    "selectDatapoint",
    "toggleOpenState",
    "saveAnnotations",
    "selectRoot",
    "selectBranch",
    "selectNut",
    "change",
    "selectPlot",
    "selectSide",
    "rate",
    "postImage",
    "hoverIndex",
    "selectX",
    "toggleLock",
    "toggleTooltip",
    "start",
    "play",
    "pause",
    "reset",
    "restart",
    "lap",
    "toggleAnnotator",
    "selectGroup",
    "selectRibbon",
    "toggleTable",
    "resetZoom",
    "showSeries",
    "hideSeries",
    "toggleZoom",
    "onNodeClick",
    "onMidpointEnter",
    "onMidpointLeave",
    "focusLocation",
    "zoomStart",
    "zoomEnd",
    "zoomReset",
    "copyAlt"
  ],
  setup(nt, { expose: lt, emit: rt }) {
    const a = nt, { component: Bt, dataset: ut, config: st } = ft(a), p = {
      VueUi3dBar: e(() => import("./vue-ui-3d-bar-BhNgFaJA.js")),
      VueUiAgePyramid: e(() => import("./vue-ui-age-pyramid-CuYRf5v-.js")),
      VueUiAnnotator: e(() => import("./vue-ui-annotator-B1GWVtPp.js")),
      VueUiCandlestick: e(() => import("./vue-ui-candlestick-BY-_E7eZ.js")),
      VueUiChestnut: e(() => import("./vue-ui-chestnut-C-bQYaNI.js")),
      VueUiDashboard: e(() => import("./vue-ui-dashboard-Bo5FX73U.js")),
      VueUiDigits: e(() => import("./vue-ui-digits-mpo2IcM5.js")),
      VueUiDonut: e(() => import("./vue-ui-donut-CzCgCVsx.js")),
      VueUiDonutEvolution: e(() => import("./vue-ui-donut-evolution-C_ck_TOR.js")),
      VueUiGalaxy: e(() => import("./vue-ui-galaxy-DWKHDciu.js")),
      VueUiGauge: e(() => import("./vue-ui-gauge-J2VYHsQe.js")),
      VueUiHeatmap: e(() => import("./vue-ui-heatmap-BEDirv8T.js")),
      VueUiKpi: e(() => import("./vue-ui-kpi-DlUEHKuw.js")),
      VueUiMiniLoader: e(() => import("./vue-ui-mini-loader-CoOHvZUk.js")),
      VueUiMolecule: e(() => import("./vue-ui-molecule-Chq1qWDK.js")),
      VueUiMoodRadar: e(() => import("./vue-ui-mood-radar-A2LBLgUi.js")),
      VueUiNestedDonuts: e(() => import("./vue-ui-nested-donuts-DFX251aT.js")),
      VueUiOnion: e(() => import("./vue-ui-onion-CW0YO9Ro.js")),
      VueUiQuadrant: e(() => import("./vue-ui-quadrant-C1pLW2zA.js")),
      VueUiRadar: e(() => import("./vue-ui-radar-DE70LPBZ.js")),
      VueUiRating: e(() => import("./vue-ui-rating-zV58MDQh.js")),
      VueUiRelationCircle: e(() => import("./vue-ui-relation-circle-iuRXIyRU.js")),
      VueUiRings: e(() => import("./vue-ui-rings-80ZUki42.js")),
      VueUiScatter: e(() => import("./vue-ui-scatter-FdAAYJOK.js")),
      VueUiSkeleton: e(() => import("./vue-ui-skeleton-CbqNz9IE.js")),
      VueUiSmiley: e(() => import("./vue-ui-smiley-BK5EKp9u.js")),
      VueUiSparkbar: e(() => import("./vue-ui-sparkbar-DUDWvTo-.js")),
      VueUiSparkgauge: e(() => import("./vue-ui-sparkgauge-DdRtHQDe.js")),
      VueUiSparkHistogram: e(() => import("./vue-ui-sparkhistogram-C9DvzQae.js")),
      VueUiSparkline: e(() => import("./vue-ui-sparkline-B2RDrdiR.js")),
      VueUiSparkStackbar: e(() => import("./vue-ui-sparkstackbar-BsIOHUE5.js")),
      VueUiTable: e(() => import("./vue-ui-table-BHuXI4Eq.js")),
      VueUiTableSparkline: e(() => import("./vue-ui-table-sparkline-ClpLLKi2.js")),
      VueUiThermometer: e(() => import("./vue-ui-thermometer-sjruUGQx.js")),
      VueUiTiremarks: e(() => import("./vue-ui-tiremarks-UBDpAIVr.js")),
      VueUiTreemap: e(() => import("./vue-ui-treemap-BS5EilZf.js")),
      VueUiVerticalBar: e(() => import("./vue-ui-horizontal-bar-ClQcB9nL.js")),
      // legacy support
      VueUiHorizontalBar: e(() => import("./vue-ui-horizontal-bar-ClQcB9nL.js")),
      // v3 renaming
      VueUiWaffle: e(() => import("./vue-ui-waffle-CbX5K9Q2.js")),
      VueUiWheel: e(() => import("./vue-ui-wheel-BIJ09H5D.js")),
      VueUiXy: e(() => import("./vue-ui-xy-f67V1xjd.js")),
      VueUiTableHeatmap: e(() => import("./vue-ui-table-heatmap-y88sBfu3.js")),
      VueUiAccordion: e(() => import("./vue-ui-accordion-BGHTqIBt.js")),
      VueUiQuickChart: e(() => import("./vue-ui-quick-chart-FaxttnJY.js")),
      VueUiCursor: e(() => import("./vue-ui-cursor-B4mHR7QC.js")),
      VueUiSparkTrend: e(() => import("./vue-ui-spark-trend-C68p8-Xp.js")),
      VueUiStripPlot: e(() => import("./vue-ui-strip-plot-B9nBSRyY.js")),
      VueUiDumbbell: e(() => import("./vue-ui-dumbbell-CVSA90b2.js")),
      VueUiWordCloud: e(() => import("./vue-ui-word-cloud-DJmJ-oYT.js")),
      VueUiXyCanvas: e(() => import("./vue-ui-xy-canvas-4auyhlTr.js")),
      VueUiFlow: e(() => import("./vue-ui-flow-CsbSaFnu.js")),
      VueUiParallelCoordinatePlot: e(() => import("./vue-ui-parallel-coordinate-plot-UFg10bHP.js")),
      VueUiTimer: e(() => import("./vue-ui-timer-CgdgMKIc.js")),
      VueUiCarouselTable: e(() => import("./vue-ui-carousel-table-CmiHTJ8E.js")),
      VueUiGizmo: e(() => import("./vue-ui-gizmo-DEec3-v9.js")),
      VueUiStackbar: e(() => import("./vue-ui-stackbar-BZl0kEkb.js")),
      VueUiStackline: e(() => import("./vue-ui-stackline-2VKxwFei.js")),
      VueUiBullet: e(() => import("./vue-ui-bullet-DPoGGtaG.js")),
      VueUiFunnel: e(() => import("./vue-ui-funnel-BRkaEkBb.js")),
      VueUiHistoryPlot: e(() => import("./vue-ui-history-plot-Bs96PebH.js")),
      VueUiCirclePack: e(() => import("./vue-ui-circle-pack-CdbDcuDw.js")),
      VueUiWorld: e(() => import("./vue-ui-world-D-3wgoTn.js")),
      VueUiRidgeline: e(() => import("./vue-ui-ridgeline-CwvDh5pG.js")),
      VueUiChord: e(() => import("./vue-ui-chord-uQYQgyGM.js")),
      VueUiDag: e(() => import("./vue-ui-dag-DsHUvjvm.js")),
      VueUiGeo: e(() => import("./vue-ui-geo-DykhEWVq.js")),
      VueUiBump: e(() => import("./vue-ui-bump-C_fKharc.js"))
    }, ct = {
      VueUi3dBar: ["config", "dataset"],
      VueUiAgePyramid: ["config", "dataset"],
      VueUiAnnotator: ["config", "dataset"],
      VueUiCandlestick: ["config", "dataset"],
      VueUiChestnut: ["config", "dataset"],
      VueUiDashboard: ["config", "dataset"],
      VueUiDigits: ["config", "dataset"],
      VueUiDonut: ["config", "dataset"],
      VueUiDonutEvolution: ["config", "dataset"],
      VueUiGalaxy: ["config", "dataset"],
      VueUiGauge: ["config", "dataset"],
      VueUiHeatmap: ["config", "dataset"],
      VueUiKpi: ["config", "dataset"],
      VueUiMiniLoader: ["config"],
      VueUiMolecule: ["config", "dataset"],
      VueUiMoodRadar: ["config", "dataset"],
      VueUiNestedDonuts: ["config", "dataset"],
      VueUiOnion: ["config", "dataset"],
      VueUiQuadrant: ["config", "dataset"],
      VueUiRadar: ["config", "dataset"],
      VueUiRating: ["config", "dataset"],
      VueUiRelationCircle: ["config", "dataset"],
      VueUiRings: ["config", "dataset"],
      VueUiScatter: ["config", "dataset"],
      VueUiSkeleton: ["config"],
      VueUiSmiley: ["config", "dataset"],
      VueUiSparkbar: ["config", "dataset"],
      VueUiSparkgauge: ["config", "dataset"],
      VueUiSparkHistogram: ["config", "dataset"],
      VueUiSparkline: ["config", "dataset"],
      VueUiSparkStackbar: ["config", "dataset"],
      VueUiTable: ["config", "dataset"],
      VueUiTableSparkline: ["config", "dataset"],
      VueUiThermometer: ["config", "dataset"],
      VueUiTiremarks: ["config", "dataset"],
      VueUiTreemap: ["config", "dataset"],
      VueUiVerticalBar: ["config", "dataset"],
      // delete in v4
      VueUiHorizontalBar: ["config", "dataset"],
      // v3 renaming
      VueUiWaffle: ["config", "dataset"],
      VueUiWheel: ["config", "dataset"],
      VueUiXy: ["config", "dataset"],
      VueUiTableHeatmap: ["config", "dataset"],
      VueUiAccordion: ["config"],
      VueUiQuickChart: ["config", "dataset"],
      VueUiCursor: ["config"],
      VueUiSparkTrend: ["config", "dataset"],
      VueUiStripPlot: ["config", "dataset"],
      VueUiDumbbell: ["config", "dataset"],
      VueUiWordCloud: ["config", "dataset"],
      VueUiXyCanvas: ["config", "dataset"],
      VueUiFlow: ["config", "dataset"],
      VueUiParallelCoordinatePlot: ["config", "dataset"],
      VueUiTimer: ["config"],
      VueUiCarouselTable: ["config", "dataset"],
      VueUiGizmo: ["config", "dataset"],
      VueUiStackbar: ["config", "dataset"],
      VueUiStackline: ["config", "dataset"],
      VueUiBullet: ["config", "dataset"],
      VueUiFunnel: ["config", "dataset"],
      VueUiHistoryPlot: ["config", "dataset"],
      VueUiCirclePack: ["config", "dataset"],
      VueUiWorld: ["config", "dataset"],
      VueUiRidgeline: ["config", "dataset"],
      VueUiChord: ["config", "dataset"],
      VueUiDag: ["config", "dataset"],
      VueUiGeo: ["config", "dataset"],
      VueUiBump: ["config", "dataset"]
    }, gt = rt, V = c(() => !p[a.component]), pt = c(() => p[a.component] || null), l = o(null), mt = c(() => {
      const t = ct[a.component] || [], i = {};
      return t.includes("config") && (i.config = st.value), t.includes("dataset") && (i.dataset = ut.value), i;
    }), U = o(() => null), v = o(() => null), f = o(() => null), S = o(() => null), h = o(() => null), k = o(() => null), b = o(() => null), y = o(() => null), C = o(() => null), A = o(() => null), T = o(() => null), P = o(() => null), z = o(() => null), D = o(() => null), L = o(() => null), I = o(() => null), E = o(() => null), M = o(() => null), O = o(() => null), N = o(() => null), B = o(() => null), H = o(() => null), x = o(() => null), F = o(() => null), G = o(() => null), Z = o(() => null), _ = o(() => null), $ = o(() => null), W = o(() => null), X = o(() => null), j = o(() => null), Q = o(() => null), q = o(() => null), dt = o(() => null), Vt = o(() => null), Ut = o(() => null), K = o(() => null), Y = o(() => null), J = o(() => null), R = o(() => null), w = o(() => null);
    St(() => {
      V.value && console.error(`

Vue Data UI exception:
The provided component "${a.component}" does not exist. Check the spelling.

Available components:

${Object.keys(p).map((t) => `. ${t}
`).join("")}`);
    }), ot(l, async (t) => {
      t && (t.generatePdf && (U.value = t.generatePdf), t.generateImage && (f.value = t.generateImage), t.generateSvg && (S.value = t.generateSvg), t.generateCsv && (v.value = t.generateCsv), t.getItemsPositions && (h.value = t.getItemsPositions), t.toggleReadonly && (k.value = t.toggleReadonly), t.shoot && (b.value = t.shoot), t.close && (y.value = t.close), t.restoreOrder && (C.value = t.restoreOrder), t.recalculateHeight && (A.value = t.recalculateHeight), t.toggleLock && (T.value = t.toggleLock), t.toggleTable && (P.value = t.toggleTable), t.toggleLabels && (z.value = t.toggleLabels), t.toggleSort && (D.value = t.toggleSort), t.toggleStack && (L.value = t.toggleStack), t.toggleTooltip && (I.value = t.toggleTooltip), t.start && (E.value = t.start), t.pause && (M.value = t.pause), t.reset && (O.value = t.reset), t.restart && (N.value = t.restart), t.lap && (B.value = t.lap), t.toggleAnimation && (H.value = t.toggleAnimation), t.pauseAnimation && (x.value = t.pauseAnimation), t.resumeAnimation && (F.value = t.resumeAnimation), t.toggleAnnotator && (G.value = t.toggleAnnotator), t.selectNode && (Z.value = t.selectNode), t.selectGroup && (_.value = t.selectGroup), t.selectRibbon && ($.value = t.selectRibbon), t.autoSize && (W.value = t.autoSize), t.resetZoom && (X.value = t.resetZoom), t.showSeries && (j.value = t.showSeries), t.hideSeries && (Q.value = t.hideSeries), t.toggleZoom && (q.value = t.toggleZoom), t.onNodeClick && (dt.value = t.onNodeClick), t.onMidpointEnter && (Vt.value = t.onMidpointEnter), t.onMidpointLeave && (Ut.value = t.onMidpointLeave), t.zoomIn && (K.value = t.zoomIn), t.zoomOut && (Y.value = t.zoomOut), t.switchOrientation && (J.value = t.switchOrientation), t.focusLocation && (R.value = t.focusLocation), t.copyAlt && (w.value = t.copyAlt));
    });
    const vt = () => {
      const t = [
        "selectLegend",
        "selectDatapoint",
        "toggleOpenState",
        "saveAnnotations",
        "selectRoot",
        "selectBranch",
        "selectNut",
        "change",
        "selectPlot",
        "selectSide",
        "rate",
        "postImage",
        "hoverIndex",
        "selectX",
        "toggleLock",
        "toggleTooltip",
        "start",
        "pause",
        "reset",
        "restart",
        "lap",
        "toggleAnimation",
        "pauseAnimation",
        "resumeAnimation",
        "toggleAnnotator",
        "selectNode",
        "selectGroup",
        "selectRibbon",
        "autoSize",
        "toggleTable",
        "resetZoom",
        "showSeries",
        "hideSeries",
        "toggleZoom",
        "onNodeClick",
        "onMidpointEnter",
        "onMidpointLeave",
        "zoomIn",
        "zoomOut",
        "switchOrientation",
        "focusLocation",
        "zoomStart",
        "zoomEnd",
        "zoomReset",
        "copyAlt"
      ], i = {};
      return t.forEach((r) => {
        i[r] = (...n) => gt(r, ...n);
      }), i;
    }, m = ht([]);
    function tt(t, i) {
      return new Promise((r, n) => {
        m.push({ method: t, args: i, resolve: r, reject: n });
      });
    }
    ot(l, (t) => {
      if (t)
        for (; m.length; ) {
          const { method: i, args: r, resolve: n, reject: s } = m.shift(), et = t[i];
          typeof et == "function" ? Promise.resolve().then(() => et(...r)).then(n).catch(s) : s(new Error(`Method ${i} not found on ${a.component}`));
        }
    }), lt({
      getData(...t) {
        return l.value?.getData ? l.value.getData(...t) : tt("getData", t);
      },
      getImage(t = {}) {
        const { scale: i = 2 } = t;
        return l.value?.getImage ? l.value.getImage({ scale: i }) : tt("getImage", [{ scale: i }]);
      },
      autoSize: W,
      generatePdf: U,
      generateCsv: v,
      generateImage: f,
      generateSvg: S,
      getItemsPositions: h,
      toggleReadonly: k,
      shoot: b,
      close: y,
      restoreOrder: C,
      recalculateHeight: A,
      toggleLock: T,
      toggleTable: P,
      toggleLabels: z,
      toggleSort: D,
      toggleStack: L,
      toggleTooltip: I,
      start: E,
      pause: M,
      reset: O,
      restart: N,
      lap: B,
      pauseAnimation: x,
      resumeAnimation: F,
      toggleAnimation: H,
      toggleAnnotator: G,
      selectNode: Z,
      selectGroup: _,
      selectRibbon: $,
      resetZoom: X,
      showSeries: j,
      hideSeries: Q,
      toggleZoom: q,
      zoomIn: K,
      zoomOut: Y,
      switchOrientation: J,
      focusLocation: R,
      copyAlt: w
    });
    const u = c(() => {
      let t = `The provided component ${a.component} does not exist.`;
      return ["VueUiIcon", "VueUiPattern", "Arrow"].includes(a.component) ? (t = `${a.component} is not supported by the VueDataUi universal component. You must import it individually.`, console.warn(t), {
        status: "notSupported",
        message: t
      }) : (console.warn(t), {
        status: "unknown",
        message: t
      });
    });
    return (t, i) => V.value ? (g(), kt("div", {
      key: 0,
      style: bt({
        width: "100%",
        display: "flex",
        gap: "6px",
        alignItems: "center",
        color: u.value.status === "notSupported" ? "#FF9000" : "#FF0000"
      })
    }, [
      yt("div", Nt, [
        u.value.status === "unknown" ? (g(), d(at, {
          key: 0,
          name: "moodFlat",
          stroke: "#FF0000"
        })) : it("", !0),
        u.value.status === "notSupported" ? (g(), d(at, {
          key: 1,
          name: "circleExclamation",
          stroke: "#FF9000"
        })) : it("", !0)
      ]),
      Ct(" " + At(u.value.message), 1)
    ], 4)) : (g(), d(Tt(pt.value), Pt({
      key: 1,
      ref_key: "currentComponentRef",
      ref: l
    }, mt.value, zt(vt())), Dt({ _: 2 }, [
      Lt(t.$slots, (r, n) => ({
        name: n,
        fn: It((s) => [
          Et(t.$slots, n, Mt(Ot(s)))
        ])
      }))
    ]), 1040));
  }
};
export {
  Ft as default
};
