// vue-data-ui/dist/vue-ui-bump-C_fKharc.js
import { defineAsyncComponent as _, computed as p, ref as d, toRefs as jt, onMounted as qt, watch as ne, shallowRef as qe, onBeforeUnmount as Ge, watchEffect as Ze, openBlock as i, createElementBlock as c, normalizeStyle as L, normalizeClass as Je, createBlock as Y, unref as h, withCtx as g, renderSlot as y, normalizeProps as $, guardReactiveProps as S, createCommentVNode as b, createSlots as Ke, createVNode as we, Fragment as w, renderList as M, createElementVNode as se, toDisplayString as j, mergeProps as Le, resolveDynamicComponent as Gt, createTextVNode as Qe, nextTick as Zt } from "vue";
import { j as et, t as tt, m as Jt, p as ue, n as Kt, o as at, q as lt, r as Qt, s as ot, k as ea, l as ta, W as nt, a8 as q, c as aa, b as la, I as oa, X as na, Z as xe, a0 as sa, a1 as ua } from "./lib-GwPnJSrY.js";
import { t as st, u as ra } from "./useResponsive-DfdjqQps.js";
import { u as ia } from "./useConfig-CIgyKZ6R.js";
import { u as ca, B as va } from "./BaseScanner-BhhkKVoj.js";
import { u as ha } from "./usePrinter-DiK5R6k-.js";
import { u as da } from "./useSvgExport-ByUukOZt.js";
import { u as $e } from "./useNestedProp-BBT0mbHh.js";
import { u as ya } from "./useThemeCheck-D1ZYXUwG.js";
import { u as fa } from "./useTimeLabels-C-A2uZKi.js";
import { u as pa } from "./useUserOptionState-BIvW1Kz7.js";
import { u as ma } from "./useChartAccessibility-9icAAmYg.js";
import { u as ga } from "./useTimeLabelCollider-CIsgDrl9.js";
import ba from "./vue_ui_bump-CIht647H.js";
import ka from "./Title-BfyrP-jp.js";
import Ca from "./BaseDraggableDialog-DSpKTn3R.js";
import { _ as wa } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const La = ["id"], xa = ["xmlns", "viewBox"], $a = ["x", "y", "width", "height"], Sa = ["d", "stroke", "stroke-width", "onMouseenter"], Aa = ["d", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Ma = ["cx", "cy", "r", "fill", "stroke", "stroke-width", "onMouseenter", "onMouseleave", "onClick"], Ta = ["x", "y", "fill", "font-size"], Pa = ["onMouseenter", "onMouseleave", "onClick"], Ia = ["x", "y", "fill", "font-size"], za = ["x", "y", "fill", "font-size", "font--weight", "innerHTML", "onMouseenter", "onMouseleave", "onClick"], Oa = ["x", "y", "fill", "font-size", "font--weight", "innerHTML", "onMouseenter", "onMouseleave", "onClick"], Ra = { key: 0 }, Ha = { key: 1 }, Da = ["text-anchor", "font-size", "font-weight", "fill", "transform"], _a = ["text-anchor", "font-size", "fill", "transform", "innerHTML"], Wa = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Ea = ["innerHTML"], Fa = {
  __name: "vue-ui-bump",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["copyAlt"],
  setup(ut, { expose: rt, emit: it }) {
    const ct = _(() => import("./BaseIcon-CNx8Tzl5.js")), vt = _(() => import("./vue-ui-accordion-BGHTqIBt.js")), ht = _(() => import("./DataTable-Bo1pmAqG.js")), dt = _(() => import("./PenAndPaper-DWV1FMIi.js")), yt = _(() => import("./UserOptions-CrLqEimB.js")), ft = _(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_bump: pt } = ia(), { isThemeValid: mt, warnInvalidTheme: gt } = ya(), C = ut, bt = it, G = p({
      get() {
        return !!C.dataset && C.dataset.length;
      },
      set(a) {
        return a;
      }
    }), T = d(null), re = d(et()), Se = d(0), Ae = d(null), Me = d(null), Te = d(null), Z = d(!1), ie = d(!1), Pe = d(0), kt = d(0), W = d(null), ce = d(null), Ie = d(!1), J = d(!1), K = d(!1), ve = d(null), he = d(null), de = d(null), m = d(null), e = d(ye()), E = p(() => e.value.userOptions.useCursorPointer), Ct = p(() => tt({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              timeLabels: { show: !1 },
              lines: {
                coatingColor: "#4A4A4A"
              },
              plots: {
                stroke: "#4A4A4A",
                labels: {
                  show: !1,
                  displayedValue: "rank"
                }
              },
              nameLabels: {
                useSerieColor: !0
              }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: Q, FINAL_DATASET: wt, manualLoading: ee } = ca({
      ...jt(C),
      FINAL_CONFIG: e,
      prepareConfig: ye,
      skeletonDataset: C.config?.skeletonDataset ?? [
        { name: "————", values: [1, 1, 1, 2, 2, 2, 3, 3, 2, 2], color: "#4A4A4A" },
        { name: "————", values: [2, 2, 2, 1, 3, 3, 2, 2, 3, 3], color: "#6A6A6A" },
        { name: "————", values: [3, 3, 3, 3, 1, 1, 1, 1, 1, 1], color: "#8A8A8A" }
      ],
      skeletonConfig: tt({
        defaultConfig: e.value,
        userConfig: Ct.value
      })
    });
    qt(We);
    function ye() {
      const a = $e({
        userConfig: C.config,
        defaultConfig: pt
      });
      let l = {};
      const t = a.theme;
      if (t)
        if (!mt.value(a))
          gt(a), l = a;
        else {
          const o = $e({
            userConfig: ba[t] || C.config,
            defaultConfig: a
          });
          l = {
            ...$e({
              userConfig: C.config,
              defaultConfig: o
            }),
            customPalette: a.customPalette.length ? a.customPalette : Jt[t] || ue
          };
        }
      else
        l = a;
      return l;
    }
    const { userOptionsVisible: fe, setUserOptionsVisibility: ze, keepUserOptionState: Oe } = pa({ config: e.value }), { svgRef: pe } = ma({ config: e.value.style.chart.title });
    function Lt() {
      Ie.value = !0, ze(!0);
    }
    function xt() {
      ze(!1), Ie.value = !1;
    }
    const z = d({
      showTable: e.value.table.show
    });
    ne(e, () => {
      z.value = {
        showTable: e.value.table.show
      };
    }, { immediate: !0 }), ne(() => C.config, (a) => {
      Q.value || (e.value = ye()), fe.value = !e.value.userOptions.showOnChartHover, We(), Pe.value += 1, kt.value += 1, z.value.showTable = e.value.table.show, A.value.width = e.value.style.chart.width, A.value.height = e.value.style.chart.height, A.value.paddingRatio = {
        top: e.value.style.chart.padding.top / e.value.style.chart.height,
        right: e.value.style.chart.padding.right / e.value.style.chart.width,
        bottom: e.value.style.chart.padding.bottom / e.value.style.chart.height,
        left: e.value.style.chart.padding.left / e.value.style.chart.width
      };
    }), ne(() => C.dataset, (a) => {
      Array.isArray(a) && a.length > 0 && (ee.value = !1);
    }, { deep: !0 });
    const { isPrinting: me, isImaging: ge, generatePdf: Re, generateImage: He } = ha({
      elementId: `bump_${re.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-bump",
      options: e.value.userOptions.print
    }), $t = p(() => e.value.userOptions.show && !e.value.style.chart.title.text), A = d({
      width: e.value.style.chart.width,
      height: e.value.style.chart.height,
      paddingRatio: {
        top: e.value.style.chart.padding.top / e.value.style.chart.height,
        right: e.value.style.chart.padding.right / e.value.style.chart.width,
        bottom: e.value.style.chart.padding.bottom / e.value.style.chart.height,
        left: e.value.style.chart.padding.left / e.value.style.chart.width
      }
    }), St = p(() => Kt(e.value.customPalette)), P = qe(null), H = qe(null), De = d(null), _e = p(() => !!e.value.debug);
    function We() {
      if (at(C.dataset) ? (lt({
        componentName: "VueUiBump",
        type: "dataset",
        debug: _e.value
      }), ee.value = !0) : C.dataset.forEach((a, l) => {
        Qt({
          datasetObject: a,
          requiredAttributes: ["name", "values"]
        }).forEach((t) => {
          G.value = !1, lt({
            componentName: "VueUiBump",
            type: "datasetSerieAttribute",
            property: t,
            index: l,
            debug: _e.value
          }), ee.value = !0;
        });
      }), at(C.dataset) || (ee.value = e.value.loading), setTimeout(() => {
        ie.value = !0;
      }, 10), e.value.responsive) {
        const a = st(() => {
          ie.value = !1;
          const { width: l, height: t } = ra({
            chart: T.value,
            noTitle: Me.value,
            title: e.value.style.chart.title.text ? Ae.value : null,
            legend: null,
            slicer: null,
            source: Te.value
          });
          requestAnimationFrame(() => {
            A.value.width = l, A.value.height = t - 12, clearTimeout(De.value), De.value = setTimeout(() => {
              ie.value = !0;
            }, 10);
          });
        });
        P.value && (H.value && P.value.unobserve(H.value), P.value.disconnect()), P.value = new ResizeObserver(a), H.value = T.value.parentNode, P.value.observe(H.value);
      }
    }
    Ge(() => {
      P.value && (H.value && P.value.unobserve(H.value), P.value.disconnect());
    });
    const te = p(() => wt.value.map((a, l) => {
      const t = ot(a.color) || St.value[l] || ue[l] || ue[l % ue.length];
      return {
        ...a,
        absoluteIndex: l,
        id: et(),
        color: t
      };
    })), D = p(() => Math.max(...te.value.map((a) => a.values.length)));
    function At(a) {
      const l = a.map(() => Array(D.value).fill(null));
      for (let t = 0; t < D.value; t += 1) {
        const o = a.map((n, s) => ({
          seriesIndex: s,
          value: n.values?.[t],
          previousPosition: t > 0 ? l[s][t - 1] : s
        })).filter((n) => Number.isFinite(n.value));
        o.sort((n, s) => s.value !== n.value ? s.value - n.value : n.previousPosition - s.previousPosition);
        for (let n = 0; n < o.length; n += 1) {
          const s = o[n];
          l[s.seriesIndex][t] = n;
        }
      }
      return l;
    }
    function Mt() {
      let a = 0;
      return ve.value && (a = Array.from(ve.value.querySelectorAll("tspan")).reduce((t, o) => {
        const n = o.getComputedTextLength();
        return n > t ? n : t;
      }, 0)), a;
    }
    function Tt() {
      let a = 0;
      return he.value && (a = Array.from(he.value.querySelectorAll("tspan")).reduce((t, o) => {
        const n = o.getComputedTextLength();
        return n > t ? n : t;
      }, 0)), a;
    }
    const Ee = d(0), Pt = st((a) => {
      Ee.value = a;
    }, 100);
    Ze((a) => {
      const l = de.value;
      if (!l) return;
      const t = new ResizeObserver((o) => {
        Pt(o[0].contentRect.height);
      });
      t.observe(l), a(() => t.disconnect());
    }), Ge(() => {
      Ee.value = 0;
    });
    const It = p(() => {
      const a = e.value.style.chart.layout.timeLabels;
      if (!a.show) return 0;
      const l = a.fontSize, t = a.rotation, o = Math.PI / 180 * t, n = F.value || [];
      if (!n.length) return 0;
      const s = n.reduce((U, X) => {
        const oe = String(X?.text ?? "").split(`
`).length;
        return Math.max(U, oe);
      }, 1), u = n.reduce((U, X) => {
        const Ut = String(X?.text ?? "").split(`
`).reduce((Xt, Yt) => Math.max(Xt, Yt.length), 0);
        return Math.max(U, Ut);
      }, 0), v = l * 1.3, f = s * v, k = u * l * (u === 1 ? 1 : 0.6), I = Math.abs(Math.sin(o)) * k + Math.abs(Math.cos(o)) * f, B = l * 0.3;
      return Math.max(0, I + Math.abs(a.offsetY) + B);
    }), r = p(() => {
      const { height: a, width: l } = A.value, { right: t, left: o } = A.value.paddingRatio;
      let n = Mt(), s = Tt();
      const u = e.value.style.chart.padding.top, v = l - l * t - s, f = l * o + n, k = l - l * t - l * o - n - s, I = a - u - e.value.style.chart.padding.bottom - It.value, B = u + I, U = u + I, X = Math.max(0, a), oe = Math.max(0, l);
      return {
        chartHeight: U,
        chartWidth: oe,
        top: u,
        right: Math.max(0, v),
        bottom: Math.max(0, B),
        left: Math.max(0, f),
        width: Math.max(0, k),
        height: Math.max(0, I),
        unitH: Math.max(0, I) / te.value.length,
        unitW: Math.max(0, k) / D.value,
        svgHeight: X
      };
    }), x = p(() => {
      if (!G.value && !Q.value) return [];
      const a = At(te.value);
      return te.value.map((l, t) => ({
        ...l,
        positions: a[t]
      })).map((l) => {
        const t = l.positions.map((s, u) => {
          const v = ea(
            e.value.style.chart.layout.plots.labels.formatter,
            l.values[u],
            ta({
              p: e.value.style.chart.layout.plots.labels.prefix,
              v: l.values[u],
              s: e.value.style.chart.layout.plots.labels.suffix,
              r: e.value.style.chart.layout.plots.labels.rounding
            })
          );
          return {
            name: l.name,
            id: l.id,
            x: r.value.left + u * r.value.unitW + r.value.unitW / 2,
            y: r.value.top + s * r.value.unitH + r.value.unitH / 2,
            value: l.values[u],
            displayValue: v,
            rank: l.positions[u] + 1,
            color: l.color,
            labelColor: e.value.style.chart.layout.plots.labels.color === "auto" ? nt(l.color) : ot(e.value.style.chart.layout.plots.labels.color) ?? nt(l.color)
          };
        }), o = t.filter((s) => q(s.value)), n = e.value.style.chart.layout.lines.smooth ? aa(o) : la(o);
        return {
          ...l,
          coordinates: t,
          path: n
        };
      });
    });
    function Fe(a, l, t, o) {
      const n = a.toSorted((u, v) => u.y - v.y);
      for (let u = 1; u < n.length; u += 1) {
        const v = n[u - 1], f = n[u];
        f.y < v.y + l && (f.y = v.y + l);
      }
      const s = n.length - 1;
      if (s >= 0 && n[s].y > o) {
        n[s].y = o;
        for (let u = s - 1; u >= 0; u -= 1) {
          const v = n[u + 1], f = n[u];
          f.y > v.y - l && (f.y = v.y - l);
        }
      }
      if (n.length && n[0].y < t) {
        const u = t - n[0].y;
        for (let v = 0; v < n.length; v += 1)
          n[v].y += u;
      }
      return n;
    }
    const Ve = p(() => {
      const a = [], l = [];
      x.value.forEach((f) => {
        const k = f.coordinates.filter((B) => Number.isFinite(B.rank));
        if (!k.length) return;
        a.push({ ...k[0] });
        const I = k[k.length - 1];
        l.push({ ...I });
      });
      const t = e.value.style.chart.layout.nameLabels.fontSize, o = t * 1.4, n = r.value.top + t, s = r.value.bottom - t, u = Fe(
        l,
        o,
        n,
        s
      );
      return {
        left: Fe(
          a,
          o,
          n,
          s
        ),
        right: u
      };
    });
    function zt(a) {
      const l = e.value.style.chart.layout.plots.labels.fontSize * 0.4, t = a.displayValue, o = t.length * (e.value.style.chart.layout.plots.labels.fontSize * (t.length === 1 ? 1 : 0.6)), n = e.value.style.chart.layout.plots.labels.fontSize, s = o + l * 2, u = n + l * 2, v = a.x - s / 2, f = a.y - u / 2, k = e.value.style.chart.layout.plots.strokeWidth;
      return {
        x: v - k / 2,
        y: f - k / 2,
        width: s + k,
        height: u + k,
        fill: a.color,
        stroke: e.value.style.chart.layout.plots.stroke,
        "stroke-width": k,
        rx: u / 2
      };
    }
    const O = d([]);
    let Ne = 0;
    Ze(() => {
      const a = ++Ne, l = e.value.style.chart.layout.timeLabels, t = l.values, o = l.datetimeFormatter, n = D.value;
      if (!n || !Array.isArray(t) || t.length === 0) {
        O.value = [];
        return;
      }
      (async () => {
        const s = await fa({
          values: t,
          maxDatapoints: n,
          formatter: o,
          start: 0,
          end: n
        });
        a === Ne && (O.value = s);
      })();
    });
    const Ot = p(() => {
      const a = e.value.style.chart.layout.timeLabels.modulo;
      return O.value.length ? Math.min(a, [...new Set(O.value.map((l) => l.text))].length) : a;
    }), F = p(() => {
      const a = e.value.style.chart.layout.timeLabels, l = O.value || [], t = O.value || [], o = 0, n = null, s = D.value, u = l.map((f) => f?.text ?? ""), v = t.map((f) => f?.text ?? "");
      return oa(
        !!a.showOnlyFirstAndLast,
        !!a.showOnlyAtModulo,
        Math.max(1, Ot.value || 1),
        u,
        v,
        o,
        n,
        s
      );
    }), Rt = p(() => A.value.width), Ht = p(() => A.value.height), Dt = p(() => ({
      start: 0,
      end: D.value
    }));
    ga({
      timeLabelsEls: de,
      timeLabels: O,
      slicer: Dt,
      configRef: e,
      rotationPath: ["style", "chart", "layout", "timeLabels", "rotation"],
      autoRotatePath: ["style", "chart", "layout", "timeLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: Rt,
      height: Ht,
      rotation: e.value.style.chart.layout.timeLabels.autoRotate.angle
    });
    const ae = d(!1);
    function be() {
      ae.value = !ae.value;
    }
    function Be(a) {
      Z.value = a, Se.value += 1;
    }
    function Ue() {
      z.value.showTable = !z.value.showTable;
    }
    ne(() => z.value.showTable, (a) => {
      e.value.table.show || (a && e.value.table.useDialog && W.value ? W.value.open() : "close" in W.value && W.value.close());
    });
    function Xe() {
      z.value.showTable = !1, ce.value && ce.value.setTableIconState(!1);
    }
    function _t() {
      return x.value;
    }
    function ke(a = null) {
      const l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], t = ["", ...F.value.map((v) => v?.text ?? "")], o = x.value.map((v) => [
        v.name,
        ...v.coordinates.map((f) => `${f.displayValue}`)
      ]), n = x.value.map((v) => [
        v.name,
        ...v.coordinates.map((f) => `${f.rank}`)
      ]), s = l.concat([[e.value.table.columnNames.values]]).concat([t]).concat(o).concat([[""], [e.value.table.columnNames.ranking], [t]]).concat(n), u = sa(s);
      a ? a(u) : ua({ csvContent: u, title: e.value.style.chart.title.text || "vue-ui-bump" });
    }
    const V = p(() => {
      const a = e.value.table.useDialog && !e.value.table.show, l = z.value.showTable;
      return {
        component: a ? Ca : vt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: a ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: Z.value,
          fullscreenParent: T.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: E.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    }), le = p(() => {
      const a = [""].concat(F.value.map((n) => n.text)), l = x.value.map((n, s) => [
        n.name,
        ...n.coordinates.map((u) => `${u.displayValue} (${u.rank})`)
      ]), t = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, o = [e.value.table.columnNames.series];
      return { head: a, body: l, config: t, colNames: o };
    }), Wt = p(() => e.value.style.chart.backgroundColor), Et = p(() => e.value.style.chart.title), { exportSvg: Ft, getSvg: Vt } = da({
      svg: pe,
      title: Et,
      legend: null,
      legendItems: null,
      backgroundColor: Wt
    });
    async function Ye({ isCb: a }) {
      K.value = !0, await Zt();
      try {
        if (a) {
          const { blob: l, url: t, text: o, dataUrl: n } = await Vt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: t, text: o, dataUrl: n }));
        } else
          await Promise.resolve(Ft());
      } finally {
        K.value = !1;
      }
    }
    async function Nt({ scale: a = 2 } = {}) {
      if (!T.value) return;
      const { width: l, height: t } = T.value.getBoundingClientRect(), o = l / t, { imageUri: n, base64: s } = await img({ domElement: T.value, base64: !0, img: !0, scale: a });
      return {
        imageUri: n,
        base64: s,
        title: e.value.style.chart.title.text,
        width: l,
        height: t,
        aspectRatio: o
      };
    }
    function R(a, l) {
      m.value = l.id, e.value.events.datapointEnter && e.value.events.datapointEnter({
        datapoint: a,
        seriesIndex: a?.pointIndex ?? null
      });
    }
    function N(a, l) {
      m.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({
        datapoint: a,
        seriesIndex: a?.pointIndex ?? null
      });
    }
    function Ce(a, l) {
      e.value.events.datapointClick && e.value.events.datapointClick({
        datapoint: a,
        seriesIndex: a?.pointIndex ?? null
      });
    }
    function Bt(a) {
      if (a?.stage === "start") {
        J.value = !0;
        return;
      }
      if (a?.stage === "end") {
        J.value = !1;
        return;
      }
      He();
    }
    async function je() {
      if (bt("copyAlt", {
        config: e.value,
        dataset: x.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: x.value
      }));
    }
    return rt({
      getData: _t,
      getImage: Nt,
      generatePdf: Re,
      generateCsv: ke,
      generateImage: He,
      generateSvg: Ye,
      toggleTable: Ue,
      toggleAnnotator: be,
      toggleFullscreen: Be,
      copyAlt: je
    }), (a, l) => (i(), c("div", {
      id: `bump_${re.value}`,
      ref_key: "bumpChart",
      ref: T,
      class: Je({ "vue-data-ui-component": !0, "vue-ui-bump": !0, "vue-data-ui-wrapper-fullscreen": Z.value }),
      style: L(`background:${e.value.style.chart.backgroundColor};color:${e.value.style.chart.color};font-family:${e.value.style.fontFamily}; position: relative; ${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: Lt,
      onMouseleave: xt
    }, [
      e.value.userOptions.buttons.annotator ? (i(), Y(h(dt), {
        key: 0,
        svgRef: h(pe),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: ae.value,
        isCursorPointer: E.value,
        onClose: be
      }, {
        "annotator-action-close": g(() => [
          y(a.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": g(({ color: t }) => [
          y(a.$slots, "annotator-action-color", $(S({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": g(({ mode: t }) => [
          y(a.$slots, "annotator-action-draw", $(S({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": g(({ disabled: t }) => [
          y(a.$slots, "annotator-action-undo", $(S({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": g(({ disabled: t }) => [
          y(a.$slots, "annotator-action-redo", $(S({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": g(({ disabled: t }) => [
          y(a.$slots, "annotator-action-delete", $(S({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : b("", !0),
      y(a.$slots, "userConfig", {}, void 0, !0),
      $t.value ? (i(), c("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Me,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : b("", !0),
      e.value.style.chart.title.text ? (i(), c("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Ae,
        style: "width:100%;background:transparent;"
      }, [
        (i(), Y(ka, {
          key: `title_${Pe.value}`,
          config: {
            title: {
              cy: "bump-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "bump-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : b("", !0),
      e.value.userOptions.show && G.value && (h(Oe) || h(fe)) ? (i(), Y(h(yt), {
        ref_key: "userOptionsRef",
        ref: ce,
        key: `user_option_${Se.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: h(me),
        isImaging: h(ge),
        uid: re.value,
        hasTooltip: !1,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: !1,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        isFullcreen: Z.value,
        chartElement: T.value,
        position: e.value.userOptions.position,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: ae.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: E.value,
        onToggleFullscreen: Be,
        onGeneratePdf: h(Re),
        onGenerateCsv: ke,
        onGenerateImage: Bt,
        onGenerateSvg: Ye,
        onToggleTable: Ue,
        onToggleAnnotator: be,
        onCopyAlt: je,
        style: L({
          visibility: h(Oe) ? h(fe) ? "visible" : "hidden" : "visible"
        })
      }, Ke({ _: 2 }, [
        a.$slots.menuIcon ? {
          name: "menuIcon",
          fn: g(({ isOpen: t, color: o }) => [
            y(a.$slots, "menuIcon", $(S({ isOpen: t, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        a.$slots.optionPdf ? {
          name: "optionPdf",
          fn: g(() => [
            y(a.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        a.$slots.optionCsv ? {
          name: "optionCsv",
          fn: g(() => [
            y(a.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        a.$slots.optionImg ? {
          name: "optionImg",
          fn: g(() => [
            y(a.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        a.$slots.optionSvg ? {
          name: "optionSvg",
          fn: g(() => [
            y(a.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        a.$slots.optionTable ? {
          name: "optionTable",
          fn: g(() => [
            y(a.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        a.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: g(({ toggleFullscreen: t, isFullscreen: o }) => [
            y(a.$slots, "optionFullscreen", $(S({ toggleFullscreen: t, isFullscreen: o })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        a.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: g(({ toggleAnnotator: t, isAnnotator: o }) => [
            y(a.$slots, "optionAnnotator", $(S({ toggleAnnotator: t, isAnnotator: o })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        a.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: g(({ altCopy: t }) => [
            y(a.$slots, "optionAltCopy", $(S({ altCopy: t })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasAltCopy", "hasFullscreen", "isFullcreen", "chartElement", "position", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : b("", !0),
      (i(), c("svg", {
        ref_key: "svgRef",
        ref: pe,
        xmlns: h(na),
        viewBox: `0 0 ${r.value.chartWidth <= 0 ? 10 : r.value.chartWidth} ${r.value.svgHeight <= 0 ? 10 : r.value.svgHeight}`,
        class: Je({ "vue-data-ui-loading": h(Q), "no-transition": !e.value.useCssAnimation, "with-transition": e.value.useCssAnimation }),
        style: L(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`),
        role: "img",
        "aria-live": "polite",
        preserveAspectRatio: "xMidYMid"
      }, [
        we(h(ft)),
        a.$slots["chart-background"] ? (i(), c("foreignObject", {
          key: 0,
          x: r.value.left,
          y: r.value.top,
          width: r.value.width <= 0 ? 10 : r.value.width,
          height: r.value.height <= 0 ? 10 : r.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          y(a.$slots, "chart-background", {}, void 0, !0)
        ], 8, $a)) : b("", !0),
        (i(!0), c(w, null, M(x.value, (t) => (i(), c(w, null, [
          se("path", {
            class: "transition-opacity",
            d: `M${t.path}`,
            stroke: e.value.style.chart.layout.lines.coatingColor,
            "stroke-width": e.value.style.chart.layout.lines.strokeWidth + 2,
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            fill: "none",
            style: L({
              opacity: m.value == null || m.value === t.id ? 1 : 0.1
            }),
            onMouseenter: (o) => m.value = t.id,
            onMouseleave: l[0] || (l[0] = (o) => m.value = null)
          }, null, 44, Sa),
          se("path", {
            class: "transition-opacity",
            d: `M${t.path}`,
            stroke: t.color,
            "stroke-width": e.value.style.chart.layout.lines.strokeWidth,
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            fill: "none",
            style: L({
              opacity: m.value == null || m.value === t.id ? 1 : 0.1
            }),
            onMouseenter: (o) => R(t, t),
            onMouseleave: (o) => N(t),
            onClick: (o) => Ce(t)
          }, null, 44, Aa)
        ], 64))), 256)),
        (i(!0), c(w, null, M(x.value, (t) => (i(), c(w, null, [
          e.value.style.chart.layout.plots.labels.displayedValue === "rank" ? (i(!0), c(w, { key: 0 }, M(t.coordinates, (o, n) => (i(), c("circle", {
            class: "transition-opacity",
            cx: o.x,
            cy: o.y,
            r: e.value.style.chart.layout.plots.radius,
            fill: t.color,
            stroke: e.value.style.chart.layout.plots.stroke,
            "stroke-width": e.value.style.chart.layout.plots.strokeWidth,
            style: L({
              opacity: m.value == null || m.value === t.id ? 1 : 0.1
            }),
            onMouseenter: (s) => R({ ...o, pointIndex: n }, t),
            onMouseleave: (s) => N({ ...o, pointIndex: n }),
            onClick: (s) => R({ ...o, pointIndex: n }, t)
          }, null, 44, Ma))), 256)) : b("", !0),
          e.value.style.chart.layout.plots.labels.show && e.value.style.chart.layout.plots.labels.displayedValue === "rank" ? (i(!0), c(w, { key: 1 }, M(t.coordinates, (o) => (i(), c("text", {
            class: "transition-opacity",
            x: o.x,
            y: o.y + e.value.style.chart.layout.plots.labels.fontSize / 3,
            fill: o.labelColor,
            "font-size": e.value.style.chart.layout.plots.labels.fontSize,
            "text-anchor": "middle",
            style: L({
              userSelect: "none",
              pointerEvents: "none",
              opacity: m.value == null || m.value === t.id ? 1 : 0.1
            })
          }, j(o.rank), 13, Ta))), 256)) : b("", !0),
          e.value.style.chart.layout.plots.labels.displayedValue === "value" ? (i(!0), c(w, { key: 2 }, M(t.coordinates, (o) => (i(), c(w, null, [
            h(q)(o.value) ? (i(), c("rect", Le({
              key: 0,
              class: "transition-opacity"
            }, { ref_for: !0 }, zt(o), {
              style: {
                opacity: m.value == null || m.value === t.id ? 1 : 0.1
              },
              onMouseenter: (n) => R({ ...a.point, pointIndex: a.i }, t),
              onMouseleave: (n) => N({ ...a.point, pointIndex: a.i }),
              onClick: (n) => R({ ...a.point, pointIndex: a.i }, t)
            }), null, 16, Pa)) : b("", !0),
            h(q)(o.value) && e.value.style.chart.layout.plots.labels.show ? (i(), c("text", {
              key: 1,
              class: "transition-opacity",
              x: o.x,
              y: o.y + e.value.style.chart.layout.plots.labels.fontSize / 3,
              fill: o.labelColor,
              "font-size": e.value.style.chart.layout.plots.labels.fontSize,
              "text-anchor": "middle",
              style: L({
                userSelect: "none",
                pointerEvents: "none",
                opacity: m.value == null || m.value === t.id ? 1 : 0.1
              })
            }, j(o.displayValue), 13, Ia)) : b("", !0)
          ], 64))), 256)) : b("", !0)
        ], 64))), 256)),
        e.value.style.chart.layout.nameLabels.leftLabels.show ? (i(), c("g", {
          key: 1,
          ref_key: "labelsLeft",
          ref: ve
        }, [
          (i(!0), c(w, null, M(Ve.value.left.filter((t) => h(q)(t.value)), (t, o) => (i(), c("text", {
            class: "transition-opacity",
            x: r.value.left,
            y: t.y + e.value.style.chart.layout.nameLabels.fontSize / 3,
            fill: e.value.style.chart.layout.nameLabels.useSerieColor ? t.color : e.value.style.chart.layout.nameLabels.color,
            "font-size": e.value.style.chart.layout.nameLabels.fontSize,
            "font--weight": e.value.style.chart.layout.nameLabels.bold ? "bold" : "normal",
            "text-anchor": "end",
            innerHTML: h(xe)({
              content: t.name,
              fontSize: e.value.style.chart.layout.nameLabels.fontSize,
              fill: e.value.style.chart.layout.nameLabels.useSerieColor ? t.color : e.value.style.chart.layout.nameLabels.color,
              x: r.value.left - e.value.style.chart.layout.nameLabels.offsetX,
              y: t.y + e.value.style.chart.layout.nameLabels.fontSize / 3,
              translateY: !0
            }),
            style: L({
              opacity: m.value == null || m.value === t.id ? 1 : 0.1
            }),
            onMouseenter: (n) => R(t, t),
            onMouseleave: (n) => N(t),
            onClick: (n) => Ce(t)
          }, null, 44, za))), 256))
        ], 512)) : b("", !0),
        e.value.style.chart.layout.nameLabels.rightLabels.show ? (i(), c("g", {
          key: 2,
          ref_key: "labelsRight",
          ref: he
        }, [
          (i(!0), c(w, null, M(Ve.value.right.filter((t) => h(q)(t.value)), (t, o) => (i(), c("text", {
            class: "transition-opacity",
            x: r.value.right,
            y: t.y + e.value.style.chart.layout.nameLabels.fontSize / 3,
            fill: e.value.style.chart.layout.nameLabels.useSerieColor ? t.color : e.value.style.chart.layout.nameLabels.color,
            "font-size": e.value.style.chart.layout.nameLabels.fontSize,
            "font--weight": e.value.style.chart.layout.nameLabels.bold ? "bold" : "normal",
            "text-anchor": "start",
            innerHTML: h(xe)({
              content: t.name,
              fontSize: e.value.style.chart.layout.nameLabels.fontSize,
              fill: e.value.style.chart.layout.nameLabels.useSerieColor ? t.color : e.value.style.chart.layout.nameLabels.color,
              x: r.value.right + e.value.style.chart.layout.nameLabels.offsetX,
              y: t.y + e.value.style.chart.layout.nameLabels.fontSize / 3,
              translateY: !0
            }),
            style: L({
              opacity: m.value == null || m.value === t.id ? 1 : 0.1
            }),
            onMouseenter: (n) => R(t, t),
            onMouseleave: (n) => N(t),
            onClick: (n) => Ce(t)
          }, null, 44, Oa))), 256))
        ], 512)) : b("", !0),
        e.value.style.chart.layout.timeLabels.show ? (i(), c("g", {
          key: 3,
          ref_key: "timeLabelsEls",
          ref: de
        }, [
          a.$slots["time-label"] ? (i(), c("g", Ra, [
            (i(!0), c(w, null, M(F.value, (t, o) => (i(), c("g", null, [
              y(a.$slots, "time-label", Le({ ref_for: !0 }, {
                x: r.value.unitW * o + r.value.unitW / 2 + r.value.left,
                y: r.value.chartHeight + e.value.style.chart.layout.timeLabels.offsetY,
                fontSize: e.value.style.chart.layout.timeLabels.fontSize,
                fill: e.value.style.chart.layout.timeLabels.color,
                transform: `translate(${r.value.unitW * o + r.value.unitW / 2 + r.value.left}, ${r.value.chartHeight + e.value.style.chart.layout.timeLabels.offsetY}), rotate(${e.value.style.chart.layout.timeLabels.rotation})`,
                absoluteIndex: t.absoluteIndex,
                content: t.text,
                textAnchor: e.value.style.chart.layout.timeLabels.rotation > 0 ? "start" : e.value.style.chart.layout.timeLabels.rotation < 0 ? "end" : "middle",
                show: !0
              }), void 0, !0)
            ]))), 256))
          ])) : (i(), c("g", Ha, [
            (i(!0), c(w, null, M(F.value, (t, o) => (i(), c("g", null, [
              String(t.text).includes(`
`) ? (i(), c("text", {
                key: o + "-multi",
                "text-anchor": e.value.style.chart.layout.timeLabels.rotation > 0 ? "start" : e.value.style.chart.layout.timeLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.layout.timeLabels.fontSize,
                fill: e.value.style.chart.layout.timeLabels.color,
                transform: `
                                    translate(
                                    ${r.value.unitW * o + r.value.unitW / 2 + r.value.left},
                                    ${r.value.chartHeight + e.value.style.chart.layout.timeLabels.fontSize * 1.3 + e.value.style.chart.layout.timeLabels.offsetY}
                                    ),
                                    rotate(${e.value.style.chart.layout.timeLabels.rotation})
                                `,
                innerHTML: h(xe)({
                  content: String(t.text),
                  fontSize: e.value.style.chart.layout.timeLabels.fontSize,
                  fill: e.value.style.chart.layout.timeLabels.color,
                  x: 0,
                  y: 0
                })
              }, null, 8, _a)) : (i(), c("text", {
                class: "vue-data-ui-time-label",
                key: o,
                "text-anchor": e.value.style.chart.layout.timeLabels.rotation > 0 ? "start" : e.value.style.chart.layout.timeLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.layout.timeLabels.fontSize,
                "font-weight": e.value.style.chart.layout.timeLabels.bold ? "bold" : "normal",
                fill: e.value.style.chart.layout.timeLabels.color,
                transform: `translate(${r.value.unitW * o + r.value.unitW / 2 + r.value.left}, ${r.value.chartHeight + e.value.style.chart.layout.timeLabels.offsetY}), rotate(${e.value.style.chart.layout.timeLabels.rotation})`
              }, j(t.text), 9, Da))
            ]))), 256))
          ]))
        ], 512)) : b("", !0),
        y(a.$slots, "svg", {
          svg: {
            drawingArea: r.value,
            data: x.value,
            isPrintingImg: h(me) | h(ge) | J.value,
            isPrintingSvg: K.value
          }
        }, void 0, !0)
      ], 14, xa)),
      a.$slots.watermark ? (i(), c("div", Wa, [
        y(a.$slots, "watermark", $(S({ isPrinting: h(me) || h(ge) || J.value || K.value })), void 0, !0)
      ])) : b("", !0),
      G.value && e.value.userOptions.buttons.table ? (i(), Y(Gt(V.value.component), Le({ key: 5 }, V.value.props, {
        ref_key: "tableUnit",
        ref: W,
        onClose: Xe
      }), Ke({
        content: g(() => [
          we(h(ht), {
            colNames: le.value.colNames,
            head: le.value.head,
            body: le.value.body,
            config: le.value.config,
            title: e.value.table.useDialog ? "" : V.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: E.value,
            onClose: Xe
          }, {
            th: g(({ th: t }) => [
              se("div", { innerHTML: t }, null, 8, Ea)
            ]),
            td: g(({ td: t }) => [
              Qe(j(t), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"])
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: g(() => [
            Qe(j(V.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: g(() => [
            se("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (t) => ke(e.value.userOptions.callbacks.csv)),
              style: L({ cursor: E.value ? "pointer" : "default" })
            }, [
              we(h(ct), {
                name: "fileCsv",
                stroke: V.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : b("", !0),
      a.$slots.source ? (i(), c("div", {
        key: 6,
        ref_key: "source",
        ref: Te,
        dir: "auto"
      }, [
        y(a.$slots, "source", {}, void 0, !0)
      ], 512)) : b("", !0),
      y(a.$slots, "skeleton", {}, () => [
        h(Q) ? (i(), Y(va, { key: 0 })) : b("", !0)
      ], !0)
    ], 46, La));
  }
}, ol = /* @__PURE__ */ wa(Fa, [["__scopeId", "data-v-2133380e"]]);
export {
  ol as default
};
