// vue-data-ui/dist/vue-ui-quadrant-C1pLW2zA.js
import { defineAsyncComponent as V, computed as g, ref as h, toRefs as Ut, watch as ze, shallowRef as vt, onMounted as Et, onBeforeUnmount as Qt, openBlock as s, createElementBlock as r, unref as i, normalizeStyle as _, normalizeClass as O, createBlock as P, withCtx as f, renderSlot as p, normalizeProps as z, guardReactiveProps as A, createCommentVNode as y, createElementVNode as d, createSlots as ct, createVNode as he, Fragment as q, renderList as M, toDisplayString as k, mergeProps as dt, Teleport as Gt, resolveDynamicComponent as jt, createTextVNode as ht, nextTick as Ae } from "vue";
import { j as yt, t as Jt, m as Zt, p as We, o as Kt, q as J, n as el, s as tl, aj as ll, k as I, l as D, X as al, u as ye, v as ol, ak as ul, Z as sl, W as nl, a0 as rl, a1 as il, T as vl, U as cl } from "./lib-GwPnJSrY.js";
import { t as dl, u as hl } from "./useResponsive-DfdjqQps.js";
import { u as yl } from "./useConfig-CIgyKZ6R.js";
import { u as bl, B as fl } from "./BaseScanner-BhhkKVoj.js";
import { u as pl } from "./usePrinter-DiK5R6k-.js";
import { u as gl } from "./useSvgExport-ByUukOZt.js";
import { u as ml } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Xe } from "./useNestedProp-BBT0mbHh.js";
import { u as xl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as $l } from "./useChartAccessibility-9icAAmYg.js";
import kl from "./img-BYYzY_QU.js";
import Ll from "./Title-BfyrP-jp.js";
import { _ as bt } from "./Shape-CSyWYgXc.js";
import wl from "./vue_ui_quadrant-DjRs--Uz.js";
import Cl from "./Legend-1LMMsPxL.js";
import { B as Sl } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as _l } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ql = ["id"], Tl = ["id"], zl = ["xmlns", "viewBox", "id"], Al = ["x", "y", "width", "height"], Wl = ["id"], Xl = ["stop-color"], Yl = ["stop-color"], Ol = { key: 1 }, Pl = ["fill", "x", "y", "height", "width", "stroke-width", "stroke", "rx"], Ml = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], Nl = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], Bl = { key: 2 }, Fl = ["points", "fill"], Rl = ["points", "fill"], Vl = ["points", "fill"], Il = ["points", "fill"], Dl = { key: 3 }, Hl = ["y", "fill", "font-size"], Ul = ["x", "y", "fill", "font-size"], El = ["x", "y", "fill", "font-size"], Ql = ["y", "fill", "font-size"], Gl = { key: 4 }, jl = ["x", "y", "font-size", "fill"], Jl = ["x", "y", "font-size", "fill"], Zl = ["x", "y", "font-size", "fill"], Kl = ["id", "font-size", "transform", "fill"], ea = ["id", "font-size", "transform", "fill"], ta = ["id", "font-size", "transform", "fill"], la = { key: 5 }, aa = ["fill", "points"], oa = ["x", "y", "width", "height"], ua = ["x", "y", "width", "height"], sa = ["x", "y", "width", "height"], na = ["x", "y", "width", "height"], ra = ["x", "y", "onMouseenter", "onMouseleave", "onClick"], ia = {
  key: 0,
  style: { "pointer-events": "none" }
}, va = ["x", "y", "font-size", "fill"], ca = ["x", "y", "font-size", "fill", "innerHTML"], da = { key: 0 }, ha = ["x", "y", "onMouseover", "onMouseleave", "onClick"], ya = {
  key: 8,
  class: "vue-ui-dna"
}, ba = ["points", "fill"], fa = ["points", "fill"], pa = ["points", "fill"], ga = ["points", "fill"], ma = { key: 9 }, xa = ["x", "y", "font-size", "fill", "font-weight"], $a = { key: 10 }, ka = ["x", "y", "fill"], La = ["x", "y", "fill"], wa = ["x", "y", "fill"], Ca = ["x", "y", "fill"], Sa = ["stroke", "d"], _a = ["stroke", "d"], qa = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Ta = ["id"], za = ["onClick"], Aa = {
  key: 0,
  height: "14",
  width: "14",
  viewBox: "0 0 20 20"
}, Wa = ["innerHTML"], Xa = {
  __name: "vue-ui-quadrant",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectPlot", "selectSide", "selectLegend", "copyAlt"],
  setup(ft, { expose: pt, emit: gt }) {
    const mt = V(() => import("./Tooltip-DL3QgjZf.js")), xt = V(() => import("./BaseIcon-CNx8Tzl5.js")), $t = V(() => import("./vue-ui-accordion-BGHTqIBt.js")), kt = V(() => import("./DataTable-Bo1pmAqG.js")), Lt = V(() => import("./PenAndPaper-DWV1FMIi.js")), wt = V(() => import("./UserOptions-CrLqEimB.js")), Ct = V(() => import("./PackageVersion-DN9d0ME5.js")), St = V(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_quadrant: _t } = yl(), { isThemeValid: qt, warnInvalidTheme: Tt } = ml(), Y = ft, Ye = g(() => !!Y.dataset && Y.dataset.length), w = h(yt()), ae = gt, be = h(!1), fe = h(""), Oe = h(0), x = h(!1), N = h(null), Pe = h(null), Me = h(null), Ne = h(null), Be = h(null), Fe = h(0), Re = h(0), Ve = h(0), Ie = h(!1), Z = h(null), pe = h(null), oe = h(!1), ue = h(!1), e = h(xe()), W = g(() => e.value.userOptions.useCursorPointer), zt = g(() => ({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              grid: {
                stroke: "#6A6A6A",
                graduations: {
                  stroke: "#6A6A6A",
                  color: "#6A6A6A90"
                },
                xAxis: {
                  auto: !0
                },
                yAxis: {
                  auto: !0
                }
              },
              labels: {
                quadrantLabels: { show: !1 },
                plotLabels: { show: !1 },
                axisLabels: { show: !1 }
              },
              plots: {
                outlineColor: "#6A6A6A"
              }
            },
            legend: {
              backgroundColor: "transparent"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: se, FINAL_DATASET: De } = bl({
      ...Ut(Y),
      FINAL_CONFIG: e,
      prepareConfig: xe,
      callback: () => {
        Promise.resolve().then(async () => {
          await Ae(), $.value.showTable = e.value.table.show;
        });
      },
      skeletonDataset: Y.config?.skeletonDataset ?? [
        {
          name: "_",
          shape: "circle",
          color: "#CACACA",
          series: [
            { name: "_", x: -6, y: -4 },
            { name: "_", x: -5, y: -2 },
            { name: "_", x: -4, y: -1 },
            { name: "_", x: -3, y: -0.5 },
            { name: "_", x: -2, y: -0.25 },
            { name: "_", x: -1, y: -0.135 },
            { name: "_", x: 0, y: 0 },
            { name: "_", x: 1, y: 0.135 },
            { name: "_", x: 2, y: 0.25 },
            { name: "_", x: 3, y: 0.5 },
            { name: "_", x: 4, y: 1 },
            { name: "_", x: 5, y: 2 },
            { name: "_", x: 6, y: 4 }
          ]
        }
      ],
      skeletonConfig: Jt({
        defaultConfig: e.value,
        userConfig: zt.value
      })
    }), { userOptionsVisible: ge, setUserOptionsVisibility: He, keepUserOptionState: Ue } = xl({ config: e.value }), { svgRef: me } = $l({ config: e.value.style.chart.title });
    function xe() {
      const l = Xe({
        userConfig: Y.config,
        defaultConfig: _t
      }), o = l.theme;
      if (!o) return l;
      if (!qt.value(l))
        return Tt(l), l;
      const a = Xe({
        userConfig: wl[o] || Y.config,
        defaultConfig: l
      }), u = Xe({
        userConfig: Y.config,
        defaultConfig: a
      });
      return {
        ...u,
        customPalette: u.customPalette.length ? u.customPalette : Zt[o] || We
      };
    }
    ze(() => Y.config, (l) => {
      e.value = xe(), ge.value = !e.value.userOptions.showOnChartHover, Ee(), Fe.value += 1, Re.value += 1, Ve.value += 1, $.value.plotLabels.show = e.value.style.chart.layout.labels.plotLabels.show, $.value.showTable = e.value.table.show, $.value.showTooltip = e.value.style.chart.tooltip.show;
    }, { deep: !0 });
    const F = vt(null), Q = vt(null), G = g(() => e.value.debug);
    function Ee() {
      if (Kt(Y.dataset) ? J({
        componentName: "VueUiQuadrant",
        type: "dataset",
        debug: G.value
      }) : Y.dataset.forEach((l, o) => {
        [null, void 0].includes(l.name) && J({
          componentName: "VueUiQuadrant",
          type: "datasetSerieAttribute",
          property: "name",
          index: o,
          debug: G.value
        }), [null, void 0].includes(l.series) ? J({
          componentName: "VueUiQuadrant",
          type: "datasetSerieAttribute",
          property: "series",
          index: o,
          debug: G.value
        }) : l.series.forEach((a, u) => {
          [null, void 0].includes(a.name) && J({
            componentName: "VueUiQuadrant",
            type: "datasetSerieAttribute",
            property: "name",
            key: "series",
            index: u,
            debug: G.value
          });
        });
      }), e.value.responsive) {
        const l = dl(() => {
          const { width: o, height: a } = hl({
            chart: N.value,
            title: e.value.style.chart.title.text ? Pe.value : null,
            legend: e.value.style.chart.legend.show ? Me.value : null,
            source: Ne.value,
            noTitle: Be.value
          });
          L.value = 48, requestAnimationFrame(() => {
            t.value.height = a, t.value.usableHeight = a - L.value * 2, t.value.width = o, t.value.usableWidth = o - L.value * 2, t.value.top = L.value, t.value.left = L.value, t.value.right = o - L.value, t.value.bottom = a - L.value, t.value.centerX = o / 2, t.value.centerY = a - a / 2, t.value.padding = L.value;
          });
        });
        F.value && (Q.value && F.value.unobserve(Q.value), F.value.disconnect()), F.value = new ResizeObserver(l), Q.value = N.value.parentNode, F.value.observe(Q.value);
      }
    }
    Et(() => {
      Ie.value = !0, Ee();
    }), Qt(() => {
      F.value && (Q.value && F.value.unobserve(Q.value), F.value.disconnect());
    });
    const { isPrinting: $e, isImaging: ke, generatePdf: Qe, generateImage: Ge } = pl({
      elementId: `vue-ui-quadrant_${w.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-quadrant",
      options: e.value.userOptions.print
    }), At = g(() => e.value.userOptions.show && !e.value.style.chart.title.text), je = g(() => el(e.value.customPalette)), $ = h({
      plotLabels: {
        show: e.value.style.chart.layout.labels.plotLabels.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    }), L = h(48), t = h({
      height: e.value.style.chart.height,
      usableHeight: e.value.style.chart.height - L.value * 2,
      width: e.value.style.chart.width,
      usableWidth: e.value.style.chart.width - L.value * 2,
      top: L.value,
      left: L.value,
      right: e.value.style.chart.width - L.value,
      bottom: e.value.style.chart.height - L.value,
      centerX: e.value.style.chart.width / 2,
      centerY: e.value.style.chart.height - e.value.style.chart.height / 2,
      padding: L.value
    }), v = h({
      ...JSON.parse(JSON.stringify(t.value)),
      startX: 0,
      startY: 0
    });
    ze(() => t.value, (l) => {
      l && (v.value = {
        ...JSON.parse(JSON.stringify(t.value)),
        startX: 0,
        startY: 0
      });
    }, { deep: !0 });
    const c = h(null), H = g(() => {
      switch (c.value) {
        case "TL":
          return {
            x: v.value.startX + v.value.width / 2,
            y: v.value.height,
            text: e.value.style.chart.layout.labels.quadrantLabels.tl.text || "Top Left",
            fontSize: e.value.style.chart.layout.labels.quadrantLabels.tl.fontSize,
            fill: e.value.style.chart.layout.labels.quadrantLabels.tl.color,
            bold: e.value.style.chart.layout.labels.quadrantLabels.tl.bold
          };
        case "TR":
          return {
            x: v.value.startX + v.value.width / 2,
            y: v.value.height,
            text: e.value.style.chart.layout.labels.quadrantLabels.tr.text || "Top Right",
            fontSize: e.value.style.chart.layout.labels.quadrantLabels.tr.fontSize,
            fill: e.value.style.chart.layout.labels.quadrantLabels.tr.color,
            bold: e.value.style.chart.layout.labels.quadrantLabels.tr.bold
          };
        case "BR":
          return {
            x: v.value.startX + v.value.width / 2,
            y: v.value.height * 1.678,
            text: e.value.style.chart.layout.labels.quadrantLabels.br.text || "Bottom Right",
            fontSize: e.value.style.chart.layout.labels.quadrantLabels.br.fontSize,
            fill: e.value.style.chart.layout.labels.quadrantLabels.br.color,
            bold: e.value.style.chart.layout.labels.quadrantLabels.br.bold
          };
        case "BL":
          return {
            x: v.value.startX + v.value.width / 2,
            y: v.value.height * 1.678,
            text: e.value.style.chart.layout.labels.quadrantLabels.bl.text || "Bottom Left",
            fontSize: e.value.style.chart.layout.labels.quadrantLabels.bl.fontSize,
            fill: e.value.style.chart.layout.labels.quadrantLabels.bl.color,
            bold: e.value.style.chart.layout.labels.quadrantLabels.bl.bold
          };
        default:
          return { x: 0, y: 0, text: "", fontSize: 0, fill: "none", bold: !1 };
      }
    }), Wt = h(null), ne = h(!1);
    function K({ targetX: l, targetY: o, targetW: a, targetH: u, side: n }) {
      c.value && te(n);
      const b = {
        x: l - v.value.startX,
        y: o - v.value.startY,
        w: a - v.value.width,
        h: u - v.value.height
      }, m = e.value.zoomAnimationFrames;
      let B = 0;
      function it() {
        ne.value = !0, v.value.startX += b.x / m, v.value.startY += b.y / m, v.value.width += b.w / m, v.value.height += b.h / m, B += 1, B < m ? Wt.value = requestAnimationFrame(it) : ne.value = !1;
      }
      it();
    }
    function R(l) {
      if (!ne.value)
        if (x.value && c.value === l)
          K({
            targetX: 0,
            targetY: 0,
            targetW: t.value.width,
            targetH: t.value.height
          }), c.value = null, x.value = !1;
        else {
          switch (c.value = l, l) {
            case "TL":
              K({
                targetX: 0,
                targetY: 0,
                targetW: t.value.width / 2 + t.value.left,
                targetH: t.value.height / 2 + t.value.top,
                side: "tl"
              });
              break;
            case "TR":
              K({
                targetX: t.value.width / 2 - t.value.left,
                targetY: 0,
                targetW: t.value.width / 2 + t.value.left,
                targetH: t.value.height / 2 + t.value.top,
                side: "tr"
              });
              break;
            case "BR":
              K({
                targetX: t.value.width / 2 - t.value.left,
                targetY: t.value.height / 2 - t.value.top,
                targetW: t.value.width / 2 + t.value.left,
                targetH: t.value.height / 2 + t.value.top,
                side: "br"
              });
              break;
            case "BL":
              K({
                targetX: 0,
                targetY: t.value.height / 2 - t.value.top,
                targetW: t.value.width / 2 + t.value.left,
                targetH: t.value.height / 2 + t.value.top,
                side: "bl"
              });
              break;
            default:
              v.value.startX = 0, v.value.startY = 0, v.value.width = t.value.width, v.value.height = t.value.height;
              break;
          }
          x.value = !0;
        }
    }
    const Xt = g(() => {
      const l = e.value.style.chart.layout.grid.graduations.steps, o = t.value.usableWidth / (l * 2), a = t.value.top, u = [];
      for (let n = 0; n < l; n += 1)
        u.push({
          x: t.value.padding + o * n,
          y: a + t.value.usableHeight * (n / l / 2),
          height: t.value.usableHeight * (1 - n / l),
          width: t.value.usableWidth * (1 - n / l),
          opacity: Math.round((n + 1) / l * 20)
        });
      return u;
    }), T = g(() => {
      let l = e.value.style.chart.layout.grid.xAxis.max, o = e.value.style.chart.layout.grid.xAxis.min, a = e.value.style.chart.layout.grid.yAxis.max, u = e.value.style.chart.layout.grid.yAxis.min;
      return e.value.style.chart.layout.grid.xAxis.auto && (l = Math.max(...j.value.flatMap((n) => n.series.map((b) => b.x))), o = Math.min(...j.value.flatMap((n) => n.series.map((b) => b.x)))), e.value.style.chart.layout.grid.yAxis.auto && (a = Math.max(...j.value.flatMap((n) => n.series.map((b) => b.y))), u = Math.min(...j.value.flatMap((n) => n.series.map((b) => b.y)))), {
        x: {
          max: l,
          min: o
        },
        y: {
          max: a,
          min: u
        }
      };
    }), C = h([]), j = g(() => De.value.map((l, o) => ({
      ...l,
      series: ll({
        data: l.series,
        threshold: e.value.downsample.threshold
      }),
      id: `cat_${o}_${w.value}`,
      color: tl(l.color) || je.value[o] || We[o]
    }))), Le = g(() => j.value.filter((l) => !C.value.includes(l.id))), Yt = g(() => j.value.map((l, o) => ({
      ...l,
      shape: l.shape || "circle",
      series: l.series.map((a) => ({
        ...a,
        x: Je(a.x),
        y: Ze(a.y),
        xValue: a.x,
        yValue: a.y,
        quadrantSide: et({ x: a.x, y: a.y }),
        categoryName: l.name,
        shape: l.shape,
        color: l.color
      }))
    }))), X = g(() => (De.value.forEach((l, o) => {
      l.series.forEach((a, u) => {
        [null, void 0].includes(a.x) && J({
          componentName: "VueUiQuadrant",
          type: "datasetSerieAttribute",
          property: "x",
          key: "series",
          index: u,
          debug: G.value
        }), [null, void 0].includes(a.y) && J({
          componentName: "VueUiQuadrant",
          type: "datasetSerieAttribute",
          property: "y",
          key: "series",
          index: u,
          debug: G.value
        });
      });
    }), Le.value.map((l, o) => ({
      ...l,
      shape: l.shape || "circle",
      color: l.color || je.value[o] || We[o],
      series: l.series.map((a) => ({
        ...a,
        x: Je(a.x),
        y: Ze(a.y),
        xValue: a.x,
        yValue: a.y,
        quadrantSide: et({ x: a.x, y: a.y }),
        categoryName: l.name,
        shape: l.shape,
        color: l.color,
        uid: yt()
      }))
    }))));
    function Je(l) {
      if (l >= 0) {
        const o = l / T.value.x.max;
        return t.value.centerX + t.value.usableWidth / 2 * o;
      } else {
        const o = Math.abs(l) / Math.abs(T.value.x.min);
        return t.value.centerX - t.value.usableWidth / 2 * o;
      }
    }
    function Ze(l) {
      if (l >= 0) {
        const o = l / T.value.y.max;
        return t.value.centerY + (1 - t.value.usableHeight / 2 * o);
      } else {
        const o = Math.abs(l) / Math.abs(T.value.y.min);
        return t.value.centerY - (1 - t.value.usableHeight / 2 * o);
      }
    }
    const re = g(() => {
      const l = X.value.flatMap((m) => m.series.map((B) => ({
        x: B.xValue,
        y: B.yValue,
        name: B.name,
        category: B.categoryName,
        quadrantSide: B.quadrantSide,
        sideName: e.value.style.chart.layout.labels.quadrantLabels[B.quadrantSide].text,
        color: m.color,
        shape: m.shape
      }))), o = e.value.style.chart.layout.grid.xAxis.name || "x", a = e.value.style.chart.layout.grid.yAxis.name || "y", u = [e.value.translations.category, e.value.translations.item, o, a, e.value.translations.side], n = l.map((m) => [m.category, m.name, m.x, m.y, m.sideName || m.quadrantSide]), b = l.map((m) => ({
        shape: m.shape,
        color: m.color
      }));
      return { head: u, body: n, itsShapes: b, tableData: l };
    }), ie = g(() => {
      const l = re.value.head, o = re.value.tableData.map((u) => [
        {
          shape: u.shape,
          color: u.color,
          name: u.category
        },
        u.name,
        I(
          e.value.style.chart.layout.labels.plotLabels.x.formatter,
          u.x,
          D({
            v: u.x,
            r: e.value.style.chart.layout.labels.plotLabels.rounding
          })
        ),
        I(
          e.value.style.chart.layout.labels.plotLabels.y.formatter,
          u.y,
          D({
            v: u.y,
            r: e.value.style.chart.layout.labels.plotLabels.rounding
          })
        ),
        u.sideName || u.quadrantSide
      ]), a = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return { head: l, body: o, config: a, colNames: l };
    });
    function Ot() {
      C.value.length ? C.value = [] : U.value.forEach((l) => {
        C.value.push(l.id);
      });
    }
    function ee(l) {
      C.value.includes(l) ? C.value = C.value.filter((a) => a !== l) : C.value.push(l);
      const o = tt();
      ae("selectLegend", o);
    }
    const U = g(() => Yt.value.map((l) => ({
      name: l.name,
      shape: l.shape,
      color: l.color,
      id: l.id,
      opacity: C.value.includes(l.id) ? 0.5 : 1,
      segregate: () => ee(l.id),
      isSegregated: C.value.includes(l.id)
    })));
    function Ke(l) {
      if (!U.value.length)
        return e.value.debug && console.warn("VueUiQuadrant - There are no series to show."), null;
      const o = U.value.find((a) => a.name === l);
      return o || (e.value.debug && console.warn(`VueUiQuadrant - Series name not found "${l}"`), null);
    }
    function Pt(l) {
      const o = Ke(l);
      o !== null && C.value.includes(o.id) && ee(o.id);
    }
    function Mt(l) {
      const o = Ke(l);
      o !== null && (C.value.includes(o.id) || ee(o.id));
    }
    const Nt = g(() => ({
      cy: "quadrant-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    }));
    function et(l) {
      switch (!0) {
        case (l.x >= 0 && l.y >= 0):
          return "tr";
        case (l.x >= 0 && l.y < 0):
          return "br";
        case (l.x < 0 && l.y < 0):
          return "bl";
        case (l.x < 0 && l.y >= 0):
          return "tl";
        default:
          return "";
      }
    }
    const ve = h(null), ce = h(null), we = h(null);
    function Ce(l, o, a) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: o, seriesIndex: a }), ve.value = o.uid, ce.value = {
        color: l.color,
        shape: l.shape
      }, we.value = {
        datapoint: o,
        seriesIndex: a,
        series: X.value,
        config: e.value
      }, be.value = !0;
      const u = e.value.style.chart.tooltip.customFormat;
      if (vl(u) && cl(() => u({
        seriesIndex: a,
        datapoint: o,
        series: X.value,
        config: e.value
      })))
        fe.value = u({
          seriesIndex: a,
          datapoint: o,
          series: X.value,
          config: e.value
        });
      else {
        let n = "";
        o.quadrantSide && (n += `<div style="color:${e.value.style.chart.layout.labels.quadrantLabels[o.quadrantSide].color};font-weight:${e.value.style.chart.layout.labels.quadrantLabels[o.quadrantSide].bold ? "bold" : "400"}">${e.value.style.chart.layout.labels.quadrantLabels[o.quadrantSide].text}</div>`), n += `<div>${l.name}</div>`, n += `<div style="padding-bottom:6px;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};margin-bottom:3px">${o.name}</div>`, n += `<div>${e.value.style.chart.layout.grid.xAxis.name ? e.value.style.chart.layout.grid.xAxis.name : "x"}: <b>${I(
          e.value.style.chart.layout.labels.plotLabels.x.formatter,
          o.xValue,
          D({
            v: o.xValue,
            r: e.value.style.chart.tooltip.roundingValue
          }),
          { datapoint: o, category: l, categoryIndex: a }
        )}</b></div>`, n += `<div>${e.value.style.chart.layout.grid.yAxis.name ? e.value.style.chart.layout.grid.yAxis.name : "y"}: <b>${I(
          e.value.style.chart.layout.labels.plotLabels.y.formatter,
          o.yValue,
          D({
            v: o.yValue,
            r: e.value.style.chart.tooltip.roundingValue
          }),
          { datapoint: o, category: l, categoryIndex: a }
        )}</b></div>`, fe.value = `<div style="text-align:left;font-size:${e.value.style.chart.tooltip.fontSize}px">${n}</div>`;
      }
    }
    function Se(l, o) {
      be.value = !1, ve.value = null, ce.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: o });
    }
    function _e(l, o, a) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: o, seriesIndex: a });
      const u = {
        category: l.name,
        shape: l.shape,
        itemName: o.name,
        x: o.xValue,
        y: o.yValue,
        quadrantSide: o.quadrantSide,
        sideName: e.value.style.chart.layout.labels.quadrantLabels[o.quadrantSide].text
      };
      ae("selectPlot", u);
    }
    function te(l) {
      if (!l) return;
      const o = X.value.flatMap((u) => u.series.filter((n) => n.quadrantSide === l).map((n) => ({
        category: n.categoryName,
        itemName: n.name,
        x: n.xValue,
        y: n.yValue
      }))), a = {
        quadrantSide: l,
        sideName: e.value.style.chart.layout.labels.quadrantLabels[l].text,
        items: [...o]
      };
      ae("selectSide", a);
    }
    const S = g(() => ({
      TL: {
        tl: {
          x: t.value.left + t.value.usableWidth / 4 - 20,
          y: 0,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tl.color
        },
        tr: {
          x: t.value.left + t.value.usableWidth / 4,
          y: 0,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tr.color
        },
        br: {
          x: t.value.left + t.value.usableWidth / 4,
          y: 20,
          fill: e.value.style.chart.layout.labels.quadrantLabels.br.color
        },
        bl: {
          x: t.value.left + t.value.usableWidth / 4 - 20,
          y: 20,
          fill: e.value.style.chart.layout.labels.quadrantLabels.bl.color
        },
        crosshairs: {
          horizontal: `M ${t.value.left + t.value.usableWidth / 4 - 20},20 ${t.value.left + t.value.usableWidth / 4 + 20},20`,
          vertical: `M ${t.value.left + t.value.usableWidth / 4},0 ${t.value.left + t.value.usableWidth / 4},40`
        }
      },
      TR: {
        tl: {
          x: t.value.centerX + t.value.usableWidth / 4 - 20,
          y: 0,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tl.color
        },
        tr: {
          x: t.value.centerX + t.value.usableWidth / 4,
          y: 0,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tr.color
        },
        br: {
          x: t.value.centerX + t.value.usableWidth / 4,
          y: 20,
          fill: e.value.style.chart.layout.labels.quadrantLabels.br.color
        },
        bl: {
          x: t.value.centerX + t.value.usableWidth / 4 - 20,
          y: 20,
          fill: e.value.style.chart.layout.labels.quadrantLabels.bl.color
        },
        crosshairs: {
          horizontal: `M ${t.value.centerX + t.value.usableWidth / 4 - 20},20 ${t.value.centerX + t.value.usableWidth / 4 + 20},20`,
          vertical: `M ${t.value.centerX + t.value.usableWidth / 4},0 ${t.value.centerX + t.value.usableWidth / 4},40`
        }
      },
      BR: {
        tl: {
          x: t.value.centerX + t.value.usableWidth / 4 - 20,
          y: t.value.centerY - 48,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tl.color
        },
        tr: {
          x: t.value.centerX + t.value.usableWidth / 4,
          y: t.value.centerY - 48,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tr.color
        },
        br: {
          x: t.value.centerX + t.value.usableWidth / 4,
          y: t.value.centerY - 28,
          fill: e.value.style.chart.layout.labels.quadrantLabels.br.color
        },
        bl: {
          x: t.value.centerX + t.value.usableWidth / 4 - 20,
          y: t.value.centerY - 28,
          fill: e.value.style.chart.layout.labels.quadrantLabels.bl.color
        },
        crosshairs: {
          horizontal: `M ${t.value.centerX + t.value.usableWidth / 4 - 20},${t.value.centerY - 28} ${t.value.centerX + t.value.usableWidth / 4 + 20},${t.value.centerY - 28}`,
          vertical: `M ${t.value.centerX + t.value.usableWidth / 4},${t.value.centerY - 48} ${t.value.centerX + t.value.usableWidth / 4},${t.value.centerY - 8}`
        }
      },
      BL: {
        tl: {
          x: t.value.left + t.value.usableWidth / 4 - 20,
          y: t.value.centerY - 48,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tl.color
        },
        tr: {
          x: t.value.left + t.value.usableWidth / 4,
          y: t.value.centerY - 48,
          fill: e.value.style.chart.layout.labels.quadrantLabels.tr.color
        },
        br: {
          x: t.value.left + t.value.usableWidth / 4,
          y: t.value.centerY - 28,
          fill: e.value.style.chart.layout.labels.quadrantLabels.br.color
        },
        bl: {
          x: t.value.left + t.value.usableWidth / 4 - 20,
          y: t.value.centerY - 28,
          fill: e.value.style.chart.layout.labels.quadrantLabels.bl.color
        },
        crosshairs: {
          horizontal: `M ${t.value.left + t.value.usableWidth / 4 - 20},${t.value.centerY - 28} ${t.value.left + t.value.usableWidth / 4 + 20},${t.value.centerY - 28}`,
          vertical: `M ${t.value.left + t.value.usableWidth / 4},${t.value.centerY - 48} ${t.value.left + t.value.usableWidth / 4},${t.value.centerY - 8}`
        }
      }
    }));
    function tt() {
      return X.value.map((l) => ({
        color: l.color,
        name: l.name,
        shape: l.shape,
        series: l.series.map((o) => ({
          name: o.name,
          x: o.xValue,
          y: o.yValue,
          quadrantSide: o.quadrantSide,
          sideName: e.value.style.chart.layout.labels.quadrantLabels[o.quadrantSide].text
        }))
      }));
    }
    function qe(l = null) {
      Ae(() => {
        const o = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], a = re.value.head, u = re.value.body, n = o.concat([a]).concat(u), b = rl(n);
        l ? l(b) : il({ csvContent: b, title: e.value.style.chart.title.text || "vue-ui-quadrant" });
      });
    }
    const E = h(!1);
    function lt(l) {
      E.value = l, Oe.value += 1;
    }
    function at() {
      $.value.showTable = !$.value.showTable;
    }
    function ot() {
      $.value.plotLabels.show = !$.value.plotLabels.show;
    }
    function ut() {
      $.value.showTooltip = !$.value.showTooltip;
    }
    const de = h(!1);
    function Te() {
      de.value = !de.value;
    }
    async function Bt({ scale: l = 2 } = {}) {
      if (!N.value) return;
      const { width: o, height: a } = N.value.getBoundingClientRect(), u = o / a, { imageUri: n, base64: b } = await kl({ domElement: N.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: n,
        base64: b,
        title: e.value.style.chart.title.text,
        width: o,
        height: a,
        aspectRatio: u
      };
    }
    const le = g(() => {
      const l = e.value.table.useDialog && !e.value.table.show, o = $.value.showTable;
      return {
        component: l ? St : $t,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: l ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: E.value,
          fullscreenParent: N.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: W.value
        } : {
          hideDetails: !0,
          config: {
            open: o,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ze(() => $.value.showTable, (l) => {
      e.value.table.show || (l && e.value.table.useDialog && Z.value ? Z.value.open() : "close" in Z.value && Z.value.close());
    });
    function st() {
      $.value.showTable = !1, pe.value && pe.value.setTableIconState(!1);
    }
    const Ft = g(() => e.value.style.chart.backgroundColor), Rt = g(() => e.value.style.chart.legend), Vt = g(() => e.value.style.chart.title), { exportSvg: It, getSvg: Dt } = gl({
      svg: me,
      title: Vt,
      legend: Rt,
      legendItems: U,
      backgroundColor: Ft
    });
    async function nt({ isCb: l }) {
      ue.value = !0, await Ae();
      try {
        if (l) {
          const { blob: o, url: a, text: u, dataUrl: n } = await Dt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: o, url: a, text: u, dataUrl: n }));
        } else
          await Promise.resolve(It());
      } finally {
        ue.value = !1;
      }
    }
    function Ht(l) {
      if (l?.stage === "start") {
        oe.value = !0;
        return;
      }
      if (l?.stage === "end") {
        oe.value = !1;
        return;
      }
      Ge();
    }
    async function rt() {
      if (ae("copyAlt", {
        config: e.value,
        dataset: Le.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: Le.value
      }));
    }
    return pt({
      getData: tt,
      getImage: Bt,
      generatePdf: Qe,
      generateCsv: qe,
      generateImage: Ge,
      generateSvg: nt,
      hideSeries: Mt,
      showSeries: Pt,
      toggleTable: at,
      toggleLabels: ot,
      toggleTooltip: ut,
      toggleAnnotator: Te,
      toggleFullscreen: lt,
      copyAlt: rt
    }), (l, o) => (s(), r("div", {
      class: O(`vue-data-ui-component vue-ui-quadrant ${E.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      ref_key: "quadrantChart",
      ref: N,
      id: `vue-ui-quadrant_${w.value}`,
      style: _(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: o[14] || (o[14] = () => i(He)(!0)),
      onMouseleave: o[15] || (o[15] = () => i(He)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (s(), P(i(Lt), {
        key: 0,
        svgRef: i(me),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: de.value,
        isCursorPointer: W.value,
        onClose: Te
      }, {
        "annotator-action-close": f(() => [
          p(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": f(({ color: a }) => [
          p(l.$slots, "annotator-action-color", z(A({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": f(({ mode: a }) => [
          p(l.$slots, "annotator-action-draw", z(A({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": f(({ disabled: a }) => [
          p(l.$slots, "annotator-action-undo", z(A({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": f(({ disabled: a }) => [
          p(l.$slots, "annotator-action-redo", z(A({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": f(({ disabled: a }) => [
          p(l.$slots, "annotator-action-delete", z(A({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : y("", !0),
      At.value ? (s(), r("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Be,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : y("", !0),
      e.value.style.chart.title.text ? (s(), r("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Pe,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (s(), P(Ll, {
          key: `table_${Fe.value}`,
          config: {
            title: {
              cy: "quadrant-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "quadrant-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : y("", !0),
      d("div", {
        id: `legend-top-${w.value}`
      }, null, 8, Tl),
      e.value.userOptions.show && Ye.value && (i(Ue) || i(ge)) ? (s(), P(i(wt), {
        ref_key: "userOptionsRef",
        ref: pe,
        key: `user_options_${Oe.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isImaging: i(ke),
        isPrinting: i($e),
        uid: w.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: E.value,
        isTooltip: $.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: N.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: de.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: W.value,
        onToggleFullscreen: lt,
        onGeneratePdf: i(Qe),
        onGenerateCsv: qe,
        onGenerateImage: Ht,
        onGenerateSvg: nt,
        onToggleTable: at,
        onToggleLabels: ot,
        onToggleTooltip: ut,
        onToggleAnnotator: Te,
        onCopyAlt: rt,
        style: _({
          visibility: i(Ue) ? i(ge) ? "visible" : "hidden" : "visible"
        })
      }, ct({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: f(({ isOpen: a, color: u }) => [
            p(l.$slots, "menuIcon", z(A({ isOpen: a, color: u })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: f(() => [
            p(l.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: f(() => [
            p(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionCsv ? {
          name: "optionCsv",
          fn: f(() => [
            p(l.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: f(() => [
            p(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: f(() => [
            p(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionTable ? {
          name: "optionTable",
          fn: f(() => [
            p(l.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionLabels ? {
          name: "optionLabels",
          fn: f(() => [
            p(l.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: f(({ toggleFullscreen: a, isFullscreen: u }) => [
            p(l.$slots, "optionFullscreen", z(A({ toggleFullscreen: a, isFullscreen: u })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: f(({ toggleAnnotator: a, isAnnotator: u }) => [
            p(l.$slots, "optionAnnotator", z(A({ toggleAnnotator: a, isAnnotator: u })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: f(({ altCopy: a }) => [
            p(l.$slots, "optionAltCopy", z(A({ altCopy: a })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : y("", !0),
      (s(), r("svg", {
        ref_key: "svgRef",
        ref: me,
        xmlns: i(al),
        class: O({ "vue-data-ui-fullscreen--on": E.value, "vue-data-ui-fulscreen--off": !E.value }),
        viewBox: `${v.value.startX} ${v.value.startY} ${v.value.width} ${v.value.height}`,
        style: _(`max-width:100%;overflow:hidden;background:transparent;color:${e.value.style.chart.color}`),
        id: `svg_${w.value}`
      }, [
        he(i(Ct)),
        l.$slots["chart-background"] ? (s(), r("foreignObject", {
          key: 0,
          x: v.value.startX,
          y: v.value.startY,
          width: v.value.width,
          height: v.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          p(l.$slots, "chart-background", {}, void 0, !0)
        ], 8, Al)) : y("", !0),
        d("defs", null, [
          (s(!0), r(q, null, M(X.value, (a, u) => (s(), r("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `quadrant_gradient_${w.value}_${u}`
          }, [
            d("stop", {
              offset: "0%",
              "stop-color": i(ye)(i(ol)(a.color, 0.05), e.value.style.chart.layout.areas.opacity)
            }, null, 8, Xl),
            d("stop", {
              offset: "100%",
              "stop-color": i(ye)(a.color, e.value.style.chart.layout.areas.opacity)
            }, null, 8, Yl)
          ], 8, Wl))), 256))
        ]),
        e.value.style.chart.layout.grid.graduations.show ? (s(), r("g", Ol, [
          (s(!0), r(q, null, M(Xt.value, (a) => (s(), r("rect", {
            fill: e.value.style.chart.layout.grid.graduations.fill ? i(ye)(e.value.style.chart.layout.grid.graduations.color, a.opacity) : "none",
            x: a.x,
            y: a.y,
            height: a.height <= 0 ? 1e-3 : a.height,
            width: a.width <= 0 ? 1e-3 : a.width,
            "stroke-width": e.value.style.chart.layout.grid.graduations.strokeWidth,
            stroke: e.value.style.chart.layout.grid.graduations.stroke,
            rx: e.value.style.chart.layout.grid.graduations.roundingForce
          }, null, 8, Pl))), 256))
        ])) : y("", !0),
        d("line", {
          x1: t.value.left,
          y1: t.value.centerY,
          x2: t.value.right,
          y2: t.value.centerY,
          stroke: e.value.style.chart.layout.grid.stroke,
          "stroke-width": e.value.style.chart.layout.grid.strokeWidth
        }, null, 8, Ml),
        d("line", {
          x1: t.value.centerX,
          y1: t.value.top,
          x2: t.value.centerX,
          y2: t.value.bottom,
          stroke: e.value.style.chart.layout.grid.stroke,
          "stroke-width": e.value.style.chart.layout.grid.strokeWidth
        }, null, 8, Nl),
        e.value.style.chart.layout.grid.showArrows ? (s(), r("g", Bl, [
          d("polygon", {
            points: `${t.value.right - 8},${t.value.centerY - 6} ${t.value.right},${t.value.centerY} ${t.value.right - 8},${t.value.centerY + 6}`,
            fill: e.value.style.chart.layout.grid.stroke,
            stroke: "none"
          }, null, 8, Fl),
          d("polygon", {
            points: `${t.value.left + 8},${t.value.centerY - 6} ${t.value.left},${t.value.centerY} ${t.value.left + 8},${t.value.centerY + 6}`,
            fill: e.value.style.chart.layout.grid.stroke,
            stroke: "none"
          }, null, 8, Rl),
          d("polygon", {
            points: `${t.value.centerX - 6},${t.value.top + 8} ${t.value.centerX},${t.value.top} ${t.value.centerX + 6},${t.value.top + 8}`,
            fill: e.value.style.chart.layout.grid.stroke,
            stroke: "none"
          }, null, 8, Vl),
          d("polygon", {
            points: `${t.value.centerX - 6},${t.value.bottom - 8} ${t.value.centerX},${t.value.bottom} ${t.value.centerX + 6},${t.value.bottom - 8}`,
            fill: e.value.style.chart.layout.grid.stroke,
            stroke: "none"
          }, null, 8, Il)
        ])) : y("", !0),
        e.value.style.chart.layout.labels.quadrantLabels.show && !x.value ? (s(), r("g", Dl, [
          e.value.style.chart.layout.labels.quadrantLabels.tl.text ? (s(), r("text", {
            key: 0,
            x: 0,
            y: t.value.top - t.value.padding / 2,
            "text-anchor": "start",
            fill: e.value.style.chart.layout.labels.quadrantLabels.tl.color,
            "font-size": e.value.style.chart.layout.labels.quadrantLabels.tl.fontSize,
            style: _(`font-weight:${e.value.style.chart.layout.labels.quadrantLabels.tl.bold ? "bold" : ""}`),
            onClick: o[0] || (o[0] = (a) => te("tl"))
          }, k(e.value.style.chart.layout.labels.quadrantLabels.tl.text), 13, Hl)) : y("", !0),
          e.value.style.chart.layout.labels.quadrantLabels.tr.text ? (s(), r("text", {
            key: 1,
            x: t.value.width,
            y: t.value.top - t.value.padding / 2,
            "text-anchor": "end",
            fill: e.value.style.chart.layout.labels.quadrantLabels.tr.color,
            "font-size": e.value.style.chart.layout.labels.quadrantLabels.tr.fontSize,
            style: _(`font-weight:${e.value.style.chart.layout.labels.quadrantLabels.tr.bold ? "bold" : ""}`),
            onClick: o[1] || (o[1] = (a) => te("tr"))
          }, k(e.value.style.chart.layout.labels.quadrantLabels.tr.text), 13, Ul)) : y("", !0),
          e.value.style.chart.layout.labels.quadrantLabels.br.text ? (s(), r("text", {
            key: 2,
            x: t.value.width,
            y: t.value.bottom + t.value.padding * 0.7,
            "text-anchor": "end",
            fill: e.value.style.chart.layout.labels.quadrantLabels.br.color,
            "font-size": e.value.style.chart.layout.labels.quadrantLabels.br.fontSize,
            style: _(`font-weight:${e.value.style.chart.layout.labels.quadrantLabels.br.bold ? "bold" : ""}`),
            onClick: o[2] || (o[2] = (a) => te("br"))
          }, k(e.value.style.chart.layout.labels.quadrantLabels.br.text), 13, El)) : y("", !0),
          e.value.style.chart.layout.labels.quadrantLabels.bl.text ? (s(), r("text", {
            key: 3,
            x: 0,
            y: t.value.bottom + t.value.padding * 0.7,
            "text-anchor": "start",
            fill: e.value.style.chart.layout.labels.quadrantLabels.bl.color,
            "font-size": e.value.style.chart.layout.labels.quadrantLabels.bl.fontSize,
            style: _(`font-weight:${e.value.style.chart.layout.labels.quadrantLabels.bl.bold ? "bold" : ""}`),
            onClick: o[3] || (o[3] = (a) => te("bl"))
          }, k(e.value.style.chart.layout.labels.quadrantLabels.bl.text), 13, Ql)) : y("", !0)
        ])) : y("", !0),
        e.value.style.chart.layout.labels.axisLabels.show ? (s(), r("g", Gl, [
          d("text", {
            x: t.value.centerX,
            y: t.value.top - t.value.padding / 6,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.labels.axisLabels.fontSize,
            fill: e.value.style.chart.layout.labels.axisLabels.color.positive
          }, k(i(I)(
            e.value.style.chart.layout.labels.plotLabels.y.formatter,
            T.value.y.max,
            i(D)({
              v: T.value.y.max,
              r: e.value.style.chart.layout.labels.plotLabels.rounding
            })
          )), 9, jl),
          d("text", {
            x: t.value.centerX,
            y: t.value.top - t.value.padding / 2,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.labels.axisLabels.fontSize,
            fill: e.value.style.chart.layout.labels.axisLabels.color.positive
          }, k(e.value.style.chart.layout.grid.yAxis.name), 9, Jl),
          d("text", {
            x: t.value.centerX,
            y: t.value.bottom + t.value.padding * 0.35,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.labels.axisLabels.fontSize,
            fill: e.value.style.chart.layout.labels.axisLabels.color.negative
          }, k(i(I)(
            e.value.style.chart.layout.labels.plotLabels.y.formatter,
            T.value.y.min,
            i(D)({
              v: T.value.y.min,
              r: e.value.style.chart.layout.labels.plotLabels.rounding
            })
          )), 9, Zl),
          d("text", {
            id: `xLabelMin_${w.value}`,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.labels.axisLabels.fontSize,
            transform: `translate(${t.value.padding - e.value.style.chart.layout.labels.axisLabels.fontSize}, ${t.value.height / 2}), rotate(-90)`,
            fill: e.value.style.chart.layout.labels.axisLabels.color.negative
          }, k(i(I)(
            e.value.style.chart.layout.labels.plotLabels.x.formatter,
            T.value.x.min,
            i(D)({
              v: T.value.x.min,
              r: e.value.style.chart.layout.labels.plotLabels.rounding
            })
          )), 9, Kl),
          d("text", {
            id: `xLabelMax_${w.value}`,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.labels.axisLabels.fontSize,
            transform: `translate(${t.value.width - t.value.padding + e.value.style.chart.layout.labels.axisLabels.fontSize}, ${t.value.height / 2}), rotate(90)`,
            fill: e.value.style.chart.layout.labels.axisLabels.color.positive
          }, k(i(I)(
            e.value.style.chart.layout.labels.plotLabels.x.formatter,
            T.value.x.max,
            i(D)({
              v: T.value.x.max,
              r: e.value.style.chart.layout.labels.plotLabels.rounding
            })
          )), 9, ea),
          d("text", {
            id: `xLabelMaxName_${w.value}`,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.labels.axisLabels.fontSize,
            transform: `translate(${t.value.width - e.value.style.chart.layout.labels.axisLabels.fontSize}, ${t.value.height / 2}), rotate(90)`,
            fill: e.value.style.chart.layout.labels.axisLabels.color.positive
          }, k(e.value.style.chart.layout.grid.xAxis.name), 9, ta)
        ])) : y("", !0),
        e.value.style.chart.layout.areas.show ? (s(), r("g", la, [
          (s(!0), r(q, null, M(X.value, (a, u) => (s(), r("g", null, [
            a.series.length > 2 ? (s(), r("polygon", {
              key: 0,
              "data-cy-quadrant-area": "",
              fill: e.value.style.chart.layout.areas.useGradient ? `url(#quadrant_gradient_${w.value}_${u})` : i(ye)(a.color, e.value.style.chart.layout.areas.opacity),
              points: i(ul)(a)
            }, null, 8, aa)) : y("", !0)
          ]))), 256))
        ])) : y("", !0),
        d("g", null, [
          d("rect", {
            onClick: o[4] || (o[4] = (a) => R("TL")),
            x: t.value.left,
            y: t.value.top,
            width: t.value.usableWidth / 2 <= 0 ? 1e-3 : t.value.usableWidth / 2,
            height: t.value.usableHeight / 2 <= 0 ? 1e-3 : t.value.usableHeight / 2,
            fill: "transparent",
            class: O({ "vue-data-ui-zoom-plus": !x.value, "vue-data-ui-zoom-minus": x.value })
          }, null, 10, oa),
          d("rect", {
            onClick: o[5] || (o[5] = (a) => R("TR")),
            x: t.value.centerX,
            y: t.value.top,
            width: t.value.usableWidth / 2 <= 0 ? 1e-3 : t.value.usableWidth / 2,
            height: t.value.usableHeight / 2 <= 0 ? 1e-3 : t.value.usableHeight / 2,
            fill: "transparent",
            class: O({ "vue-data-ui-zoom-plus": !x.value, "vue-data-ui-zoom-minus": x.value })
          }, null, 10, ua),
          d("rect", {
            onClick: o[6] || (o[6] = (a) => R("BR")),
            x: t.value.centerX,
            y: t.value.centerY,
            width: t.value.usableWidth / 2 <= 0 ? 1e-3 : t.value.usableWidth / 2,
            height: t.value.usableHeight / 2 <= 0 ? 1e-3 : t.value.usableHeight / 2,
            fill: "transparent",
            class: O({ "vue-data-ui-zoom-plus": !x.value, "vue-data-ui-zoom-minus": x.value })
          }, null, 10, sa),
          d("rect", {
            onClick: o[7] || (o[7] = (a) => R("BL")),
            x: t.value.left,
            y: t.value.centerY,
            width: t.value.usableWidth / 2 <= 0 ? 1e-3 : t.value.usableWidth / 2,
            height: t.value.usableHeight / 2 <= 0 ? 1e-3 : t.value.usableHeight / 2,
            fill: "transparent",
            class: O({ "vue-data-ui-zoom-plus": !x.value, "vue-data-ui-zoom-minus": x.value })
          }, null, 10, na)
        ]),
        e.value.style.chart.layout.labels.plotLabels.showAsTag ? (s(), r(q, { key: 7 }, [
          $.value.plotLabels.show ? (s(), r("g", da, [
            (s(!0), r(q, null, M(X.value, (a, u) => (s(), r(q, null, [
              (s(!0), r(q, null, M(a.series, (n) => (s(), r("foreignObject", {
                style: { overflow: "visible" },
                height: "10",
                width: "100",
                x: n.x - 50,
                y: n.y - e.value.style.chart.layout.labels.plotLabels.fontSize,
                onMouseover: (b) => Ce(a, n, u),
                onMouseleave: (b) => Se(n, u),
                onClick: (b) => _e(a, n, u)
              }, [
                d("div", {
                  style: _(`color:${i(nl)(a.color)};margin: 0 auto; font-size:${e.value.style.chart.layout.labels.plotLabels.fontSize}px; text-align:center;background:${a.color}; padding: 2px 4px; border-radius: 12px; height: fit-content;`)
                }, k(n.name), 5)
              ], 40, ha))), 256))
            ], 64))), 256))
          ])) : y("", !0)
        ], 64)) : (s(), r(q, { key: 6 }, [
          (s(!0), r(q, null, M(X.value, (a, u) => (s(), r("g", null, [
            l.$slots.datapoint && !i(se) ? (s(!0), r(q, { key: 0 }, M(a.series, (n) => (s(), r("foreignObject", {
              x: n.x - 1,
              y: n.y - 1,
              width: "2",
              height: "2",
              style: { overflow: "visible" },
              onMouseenter: (b) => Ce(a, n, u),
              onMouseleave: (b) => Se(n, u),
              onClick: (b) => _e(a, n, u)
            }, [
              p(l.$slots, "datapoint", dt({ ref_for: !0 }, { datapoint: n }), void 0, !0)
            ], 40, ra))), 256)) : (s(!0), r(q, { key: 1 }, M(a.series, (n) => (s(), P(bt, {
              color: a.color,
              isSelected: ve.value && n.uid === ve.value,
              plot: n,
              radius: e.value.style.chart.layout.plots.radius / (x.value ? 1.5 : 1),
              shape: a.shape,
              stroke: e.value.style.chart.layout.plots.outline ? e.value.style.chart.layout.plots.outlineColor : "none",
              strokeWidth: e.value.style.chart.layout.plots.outlineWidth,
              onMouseenter: (b) => Ce(a, n, u),
              onMouseleave: (b) => Se(n, u),
              onClick: (b) => _e(a, n, u)
            }, null, 8, ["color", "isSelected", "plot", "radius", "shape", "stroke", "strokeWidth", "onMouseenter", "onMouseleave", "onClick"]))), 256))
          ]))), 256)),
          $.value.plotLabels.show ? (s(), r("g", ia, [
            (s(!0), r(q, null, M(X.value, (a) => (s(), r("g", null, [
              (s(!0), r(q, null, M(a.series, (u) => (s(), r("g", null, [
                String(u.name).includes(`
`) ? (s(), r("text", {
                  key: 1,
                  x: u.x,
                  y: u.y + e.value.style.chart.layout.labels.plotLabels.offsetY + e.value.style.chart.layout.plots.radius,
                  "text-anchor": "middle",
                  "font-size": e.value.style.chart.layout.labels.plotLabels.fontSize / (x.value ? 1.5 : 1),
                  fill: e.value.style.chart.layout.labels.plotLabels.color,
                  innerHTML: i(sl)({
                    content: String(u.name),
                    fontSize: e.value.style.chart.layout.labels.plotLabels.fontSize / (x.value ? 1.5 : 1),
                    fill: e.value.style.chart.layout.labels.plotLabels.color,
                    x: u.x,
                    y: u.y + e.value.style.chart.layout.labels.plotLabels.offsetY + e.value.style.chart.layout.plots.radius
                  })
                }, null, 8, ca)) : (s(), r("text", {
                  key: 0,
                  x: u.x,
                  y: u.y + e.value.style.chart.layout.labels.plotLabels.offsetY + e.value.style.chart.layout.plots.radius,
                  "text-anchor": "middle",
                  "font-size": e.value.style.chart.layout.labels.plotLabels.fontSize / (x.value ? 1.5 : 1),
                  fill: e.value.style.chart.layout.labels.plotLabels.color
                }, k(u.name), 9, va))
              ]))), 256))
            ]))), 256))
          ])) : y("", !0)
        ], 64)),
        x.value ? (s(), r("g", ya, [
          c.value === "TL" ? (s(), r("polygon", {
            key: 0,
            points: `${t.value.left - 1},${t.value.centerY} ${t.value.centerX},${t.value.centerY} ${t.value.centerX},${t.value.top - 1} ${t.value.right},${t.value.top - 1} ${t.value.right},${t.value.bottom} ${t.value.left - 1},${t.value.bottom} ${t.value.left - 1},${t.value.centerY}`,
            fill: e.value.style.chart.backgroundColor,
            style: { opacity: "1" }
          }, null, 8, ba)) : y("", !0),
          c.value === "TR" ? (s(), r("polygon", {
            key: 1,
            points: `${t.value.left},${t.value.top - 1} ${t.value.centerX},${t.value.top - 1} ${t.value.centerX},${t.value.centerY} ${t.value.right + 1},${t.value.centerY} ${t.value.right + 1},${t.value.bottom} ${t.value.left},${t.value.bottom} ${t.value.left},${t.value.top - 1}`,
            fill: e.value.style.chart.backgroundColor,
            style: { opacity: "1" }
          }, null, 8, fa)) : y("", !0),
          c.value === "BR" ? (s(), r("polygon", {
            key: 2,
            points: `${t.value.left},${t.value.top} ${t.value.right + 1},${t.value.top} ${t.value.right + 1},${t.value.centerY} ${t.value.centerX},${t.value.centerY} ${t.value.centerX},${t.value.bottom + 1} ${t.value.left},${t.value.bottom + 1} ${t.value.left},${t.value.top}`,
            fill: e.value.style.chart.backgroundColor,
            style: { opacity: "1" }
          }, null, 8, pa)) : y("", !0),
          c.value === "BL" ? (s(), r("polygon", {
            key: 3,
            points: `${t.value.left - 1},${t.value.top} ${t.value.right},${t.value.top} ${t.value.right},${t.value.bottom + 1} ${t.value.centerX},${t.value.bottom + 1} ${t.value.centerX},${t.value.centerY} ${t.value.left - 1},${t.value.centerY} ${t.value.left - 1},${t.value.top}`,
            fill: e.value.style.chart.backgroundColor,
            style: { opacity: "1" }
          }, null, 8, ga)) : y("", !0)
        ])) : y("", !0),
        c.value && !ne.value ? (s(), r("g", ma, [
          d("text", {
            x: H.value.x,
            y: H.value.y - H.value.fontSize / 1.5,
            "font-size": H.value.fontSize / 1.5,
            fill: H.value.fill,
            "text-anchor": "middle",
            "font-weight": H.value.bold ? "bold" : "normal"
          }, k(H.value.text), 9, xa)
        ])) : y("", !0),
        x.value && c.value ? (s(), r("g", $a, [
          d("rect", {
            x: S.value[c.value].tl.x,
            y: S.value[c.value].tl.y,
            height: "20",
            width: "20",
            fill: S.value[c.value].tl.fill,
            style: _(`cursor: ${W.value ? "pointer" : "default"}; opacity: ${c.value === "TL" ? 1 : 0.2}`),
            onClick: o[8] || (o[8] = (a) => R("TL")),
            class: O({ "vue-ui-quadrant-mini-map-cell": !0, "vue-ui-quadrant-mini-map-cell-selectable": c.value !== "TL" })
          }, null, 14, ka),
          d("rect", {
            x: S.value[c.value].tr.x,
            y: S.value[c.value].tr.y,
            height: "20",
            width: "20",
            fill: S.value[c.value].tr.fill,
            style: _(`cursor: ${W.value ? "pointer" : "default"}; opacity: ${c.value === "TR" ? 1 : 0.2}`),
            onClick: o[9] || (o[9] = (a) => R("TR")),
            class: O({ "vue-ui-quadrant-mini-map-cell": !0, "vue-ui-quadrant-mini-map-cell-selectable": c.value !== "TR" })
          }, null, 14, La),
          d("rect", {
            x: S.value[c.value].br.x,
            y: S.value[c.value].br.y,
            height: "20",
            width: "20",
            fill: S.value[c.value].br.fill,
            style: _(`cursor: ${W.value ? "pointer" : "default"}; opacity: ${c.value === "BR" ? 1 : 0.2}`),
            onClick: o[10] || (o[10] = (a) => R("BR")),
            class: O({ "vue-ui-quadrant-mini-map-cell": !0, "vue-ui-quadrant-mini-map-cell-selectable": c.value !== "BR" })
          }, null, 14, wa),
          d("rect", {
            x: S.value[c.value].bl.x,
            y: S.value[c.value].bl.y,
            height: "20",
            width: "20",
            fill: S.value[c.value].bl.fill,
            style: _(`cursor: ${W.value ? "pointer" : "default"}; opacity: ${c.value === "BL" ? 1 : 0.2}`),
            onClick: o[11] || (o[11] = (a) => R("BL")),
            class: O({ "vue-ui-quadrant-mini-map-cell": !0, "vue-ui-quadrant-mini-map-cell-selectable": c.value !== "BL" })
          }, null, 14, Ca),
          d("path", {
            class: "vue-ui-quadrant-minimap-crosshairs",
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": 1,
            d: S.value[c.value].crosshairs.horizontal
          }, null, 8, Sa),
          d("path", {
            class: "vue-ui-quadrant-minimap-crosshairs",
            stroke: e.value.style.chart.backgroundColor,
            "stroke-width": 1,
            d: S.value[c.value].crosshairs.vertical
          }, null, 8, _a)
        ])) : y("", !0),
        p(l.$slots, "svg", {
          svg: {
            ...t.value,
            isPrintingImg: i($e) | i(ke) | oe.value,
            isPrintingSvg: ue.value
          }
        }, void 0, !0)
      ], 14, zl)),
      l.$slots.watermark ? (s(), r("div", qa, [
        p(l.$slots, "watermark", z(A({ isPrinting: i($e) || i(ke) || oe.value || ue.value })), void 0, !0)
      ])) : y("", !0),
      d("div", {
        id: `legend-bottom-${w.value}`
      }, null, 8, Ta),
      Ie.value ? (s(), P(Gt, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${w.value}` : `#legend-bottom-${w.value}`
      }, [
        d("div", {
          ref_key: "chartLegend",
          ref: Me
        }, [
          e.value.style.chart.legend.show ? (s(), P(Cl, {
            key: `legend_${Ve.value}`,
            legendSet: U.value,
            config: Nt.value,
            isCursorPointer: W.value,
            onClickMarker: o[12] || (o[12] = ({ legend: a }) => ee(a.id))
          }, {
            item: f(({ legend: a }) => [
              i(se) ? y("", !0) : (s(), r("div", {
                key: 0,
                onClick: (u) => ee(a.id),
                style: _(`opacity:${C.value.includes(a.id) ? 0.5 : 1}`)
              }, k(a.name), 13, za))
            ]),
            legendToggle: f(() => [
              U.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !i(se) ? (s(), P(Sl, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: C.value.length > 0,
                isCursorPointer: W.value,
                onToggle: Ot
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : y("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : p(l.$slots, "legend", {
            key: 1,
            legend: U.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : y("", !0),
      l.$slots.source ? (s(), r("div", {
        key: 6,
        ref_key: "source",
        ref: Ne,
        dir: "auto"
      }, [
        p(l.$slots, "source", {}, void 0, !0)
      ], 512)) : y("", !0),
      he(i(mt), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: $.value.showTooltip && be.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        fontSize: e.value.style.chart.tooltip.fontSize,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: N.value,
        content: fe.value,
        isFullscreen: E.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": f(() => [
          p(l.$slots, "tooltip-before", z(A({ ...we.value })), void 0, !0)
        ]),
        "tooltip-after": f(() => [
          p(l.$slots, "tooltip-after", z(A({ ...we.value })), void 0, !0)
        ]),
        default: f(() => [
          e.value.style.chart.tooltip.showShape ? (s(), r("svg", Aa, [
            he(bt, {
              plot: { x: 10, y: 10 },
              shape: ce.value.shape,
              color: ce.value.color,
              radius: 8,
              stroke: e.value.style.chart.layout.plots.outline ? e.value.style.chart.layout.plots.outlineColor : "none",
              "stroke-width": e.value.style.chart.layout.plots.outlineWidth
            }, null, 8, ["shape", "color", "stroke", "stroke-width"])
          ])) : y("", !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      Ye.value && e.value.userOptions.buttons.table ? (s(), P(jt(le.value.component), dt({ key: 7 }, le.value.props, {
        ref_key: "tableUnit",
        ref: Z,
        onClose: st
      }), ct({
        content: f(() => [
          (s(), P(i(kt), {
            key: `table_${Re.value}`,
            colNames: ie.value.colNames,
            head: ie.value.head,
            body: ie.value.body,
            config: ie.value.config,
            title: e.value.table.useDialog ? "" : le.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: W.value,
            onClose: st
          }, {
            th: f(({ th: a }) => [
              ht(k(a), 1)
            ]),
            td: f(({ td: a }) => [
              d("div", {
                innerHTML: a.name || a
              }, null, 8, Wa)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: f(() => [
            ht(k(le.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: f(() => [
            d("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: o[13] || (o[13] = (a) => qe(e.value.userOptions.callbacks.csv)),
              style: _({ cursor: W.value ? "pointer" : "default" })
            }, [
              he(i(xt), {
                name: "fileCsv",
                stroke: le.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : y("", !0),
      p(l.$slots, "skeleton", {}, () => [
        i(se) ? (s(), P(fl, { key: 0 })) : y("", !0)
      ], !0)
    ], 46, ql));
  }
}, Za = /* @__PURE__ */ _l(Xa, [["__scopeId", "data-v-2b3162a9"]]);
export {
  Za as default
};
