// vue-data-ui/dist/vue-ui-rings-80ZUki42.js
import { defineAsyncComponent as D, useSlots as Ht, ref as u, computed as c, toRefs as Xt, watch as ae, shallowRef as lt, onMounted as Gt, onBeforeUnmount as jt, openBlock as i, createElementBlock as f, unref as s, normalizeStyle as N, normalizeClass as z, createBlock as _, withCtx as r, renderSlot as v, normalizeProps as T, guardReactiveProps as L, createCommentVNode as b, createElementVNode as C, createSlots as me, createVNode as oe, Fragment as se, renderList as ke, mergeProps as at, Teleport as Yt, toDisplayString as Ce, resolveDynamicComponent as qt, createTextVNode as ot, nextTick as st } from "vue";
import { j as nt, t as ut, m as Kt, p as E, o as rt, q as we, r as Zt, n as Jt, a7 as Qt, s as it, k as $e, l as R, X as el, u as tl, v as ll, Z as al, a0 as ol, a1 as sl, K as nl, T as ul, U as rl } from "./lib-GwPnJSrY.js";
import { b as il } from "./labelUtils-CTXUc3V9.js";
import { t as vl, u as cl } from "./useResponsive-DfdjqQps.js";
import { u as dl } from "./useConfig-CIgyKZ6R.js";
import { u as hl } from "./usePrinter-DiK5R6k-.js";
import { u as gl, B as pl } from "./BaseScanner-BhhkKVoj.js";
import { u as fl } from "./useSvgExport-ByUukOZt.js";
import { u as Te } from "./useNestedProp-BBT0mbHh.js";
import { u as yl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as bl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as ml } from "./useChartAccessibility-9icAAmYg.js";
import kl from "./img-BYYzY_QU.js";
import { _ as Cl } from "./Shape-CSyWYgXc.js";
import wl from "./Title-BfyrP-jp.js";
import $l from "./vue_ui_rings-CxOrekqA.js";
import Tl from "./Legend-1LMMsPxL.js";
import { B as Ll } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as xl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Sl = ["id"], Al = ["id"], Pl = ["xmlns", "viewBox"], Ol = ["width", "height"], _l = ["id"], Il = ["stop-color"], Fl = ["stop-color"], Dl = { key: 1 }, Nl = ["stroke", "cx", "cy", "r", "fill"], zl = ["stroke", "stroke-width", "cx", "cy", "r", "fill"], Rl = ["stroke", "stroke-width", "cx", "cy", "r", "fill"], Vl = ["cx", "cy", "r", "onMouseenter", "onMouseleave", "onClick"], Ml = ["x", "y", "width", "height", "fill", "rx"], Bl = ["cx", "cy", "r", "fill", "stroke"], Ul = ["x", "y", "text-anchor", "font-size", "fill", "font-weight", "onMouseenter", "onMouseleave", "onClick", "innerHTML"], El = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Wl = ["id"], Hl = ["onClick"], Xl = ["innerHTML"], Gl = {
  __name: "vue-ui-rings",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "copyAlt"],
  setup(vt, { expose: ct, emit: dt }) {
    const ht = D(() => import("./Tooltip-DL3QgjZf.js")), gt = D(() => import("./BaseIcon-CNx8Tzl5.js")), pt = D(() => import("./vue-ui-accordion-BGHTqIBt.js")), ft = D(() => import("./DataTable-Bo1pmAqG.js")), yt = D(() => import("./PenAndPaper-DWV1FMIi.js")), bt = D(() => import("./UserOptions-CrLqEimB.js")), mt = D(() => import("./PackageVersion-DN9d0ME5.js")), kt = D(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_rings: Ct } = dl(), { isThemeValid: wt, warnInvalidTheme: $t } = yl(), Tt = Ht(), m = vt, w = u(!1), Le = c(() => !!m.dataset && m.dataset.length), $ = u(nt()), Z = u(!1), ne = u(""), y = u(null), xe = u(0), S = u(null), Se = u(null), Ae = u(null), Pe = u(null), Oe = u(null), _e = u(0), Ie = u(0), Fe = u(0), ue = u(!1), De = u(!1), H = u(null), re = u(null), J = u(!1), Q = u(!1), e = u(ce()), V = c(() => e.value.userOptions.useCursorPointer), Lt = c(() => ut({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              rings: {
                stroke: "#6A6A6A",
                gradient: {
                  underlayerColor: "#FFFFFF"
                }
              },
              labels: {
                dataLabels: {
                  show: !1
                }
              }
            },
            legend: {
              backgroundColor: "transparent"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: x, FINAL_DATASET: Ne, manualLoading: ee } = gl({
      ...Xt(m),
      FINAL_CONFIG: e,
      prepareConfig: ce,
      skeletonDataset: m.config?.skeletonDataset ?? [
        { name: "_", values: [13], color: "#808080" },
        { name: "_", values: [8], color: "#969696" },
        { name: "_", values: [5], color: "#ADADAD" },
        { name: "_", values: [3], color: "#C4C4C4" },
        { name: "_", values: [2], color: "#DBDBDB" }
      ],
      skeletonConfig: ut({
        defaultConfig: e.value,
        userConfig: Lt.value
      })
    }), { userOptionsVisible: ie, setUserOptionsVisibility: ze, keepUserOptionState: Re } = bl({ config: e.value }), { svgRef: ve } = ml({ config: e.value.style.chart.title });
    function ce() {
      const t = Te({
        userConfig: m.config,
        defaultConfig: Ct
      }), a = t.theme;
      if (!a) return t;
      if (!wt.value(t))
        return $t(t), t;
      const l = Te({
        userConfig: $l[a] || m.config,
        defaultConfig: t
      });
      return {
        ...Te({
          userConfig: m.config,
          defaultConfig: l
        }),
        customPalette: t.customPalette.length ? t.customPalette : Kt[t.theme] || E
      };
    }
    ae(() => m.config, (t) => {
      x.value || (e.value = ce()), ie.value = !e.value.userOptions.showOnChartHover, Me(), _e.value += 1, Ie.value += 1, Fe.value += 1, p.value.showTable = e.value.table.show, p.value.showTooltip = e.value.style.chart.tooltip.show, p.value.showLabels = e.value.style.chart.layout.labels.dataLabels.show, d.value.width = e.value.style.chart.size, d.value.height = e.value.style.chart.size;
    }, { deep: !0 });
    const Ve = c(() => {
      const { markers: t } = e.value.style.chart.layout.labels.dataLabels, a = B.value / 2, l = p.value.showLabels ? t.position === "left" ? a : -a : 0, o = p.value.showLabels ? d.value.width / 2 - a : 0, n = t.position === "left" ? o : -o;
      return {
        x: e.value.responsive ? n : l / e.value.style.chart.size * d.value.width,
        y: 0 / e.value.style.chart.size * d.value.height
      };
    }), A = c(() => ({
      x: d.value.width / 2 + Ve.value.x,
      y: d.value.height / 2 + Ve.value.y
    }));
    ae(() => m.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (ee.value = !1);
    }, { deep: !0 });
    const I = lt(null), W = lt(null);
    Gt(() => {
      De.value = !0, Me();
    });
    const de = c(() => !!e.value.debug);
    function Me() {
      if (rt(m.dataset) ? (we({
        componentName: "VueUiRings",
        type: "dataset",
        debug: de.value
      }), ee.value = !0) : m.dataset.forEach((t, a) => {
        t.values.length || (we({
          componentName: "VueUiRings",
          type: "dataset",
          debug: de.value
        }), ee.value = !0), Zt({
          datasetObject: t,
          requiredAttributes: ["name", "values"]
        }).forEach((l) => {
          we({
            componentName: "VueUiRings",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: de.value
          });
        });
      }), rt(m.dataset) || (ee.value = e.value.loading), e.value.responsive) {
        const t = vl(() => {
          ue.value = !0;
          const { width: a, height: l } = cl({
            chart: S.value,
            title: e.value.style.chart.title.text ? Se.value : null,
            legend: e.value.style.chart.legend.show ? Ae.value : null,
            source: Pe.value,
            noTitle: Oe.value
          });
          requestAnimationFrame(() => {
            d.value.width = a, d.value.height = l - 12, ue.value = !1;
          });
        });
        I.value && (W.value && I.value.unobserve(W.value), I.value.disconnect()), I.value = new ResizeObserver(t), W.value = S.value.parentNode, I.value.observe(W.value);
      }
      setTimeout(() => {
        w.value = !0;
      }, 600);
    }
    jt(() => {
      I.value && (W.value && I.value.unobserve(W.value), I.value.disconnect());
    });
    const { isPrinting: he, isImaging: ge, generatePdf: Be, generateImage: Ue } = hl({
      elementId: `rings_${$.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-rings",
      options: e.value.userOptions.print
    }), xt = c(() => e.value.userOptions.show && !e.value.style.chart.title.text), Ee = c(() => Jt(e.value.customPalette)), p = u({
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show,
      showLabels: e.value.style.chart.layout.labels.dataLabels.show
    });
    ae(e, () => {
      p.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show,
        showLabels: e.value.style.chart.layout.labels.dataLabels.show
      };
    }, { immediate: !0 });
    const d = u({
      height: e.value.style.chart.size,
      width: e.value.style.chart.size
    }), St = c(() => Math.min(d.value.height, d.value.width)), We = dt, h = u([]);
    function At() {
      h.value.length ? h.value = [] : G.value.forEach((t) => {
        h.value.push(t.uid);
      });
    }
    function X(t) {
      if (h.value.includes(t))
        h.value = h.value.filter((a) => a !== t);
      else {
        if (h.value.length === M.value.length - 1) return;
        h.value.push(t);
      }
      We("selectLegend", k.value.map((a) => ({
        name: a.name,
        color: a.color,
        value: a.value
      })));
    }
    function He(t) {
      if (!M.value.length)
        return e.value.debug && console.warn("VueUiRings - There are no series to show."), null;
      const a = M.value.find((l) => l.name === t);
      return a || (e.value.debug && console.warn(`VueUiRings - Series name not found "${t}"`), null);
    }
    function Pt(t) {
      const a = He(t);
      a !== null && h.value.includes(a.uid) && X(a.uid);
    }
    function Ot(t) {
      const a = He(t);
      a !== null && (h.value.includes(a.uid) || X(a.uid));
    }
    const _t = c(() => Math.max(
      ...M.value.filter((t) => !h.value.includes(t.uid)).map(({ value: t }) => t)
    ));
    function Xe(t) {
      return t / _t.value;
    }
    const M = c(() => Ne.value.map(({ values: t, name: a, color: l = null }, o) => {
      const n = Qt(t).reduce((g, K) => g + K, 0);
      return {
        name: a,
        color: l || it(l) || Ee.value[o] || E[o] || E[o % E.length],
        value: n,
        proportion: n / Ne.value.map((g) => (g.values || []).reduce((K, Wt) => K + Wt, 0)).reduce((g, K) => g + K, 0),
        uid: nt(),
        absoluteIndex: o
      };
    })), G = c(() => M.value.map((t, a) => {
      const l = $e(e.value.style.chart.layout.labels.dataLabels.formatter, t.value, R({
        p: e.value.style.chart.layout.labels.dataLabels.prefix,
        v: t.value,
        s: e.value.style.chart.layout.labels.dataLabels.suffix,
        r: e.value.style.chart.legend.roundingValue
      }), { datapoint: t, index: a }), o = isNaN(t.value / P.value) ? "-" : R({ v: t.value / P.value * 100, s: "%", r: e.value.style.chart.legend.roundingPercentage }), n = pe({
        showVal: e.value.style.chart.legend.showValue,
        showPercentage: e.value.style.chart.legend.showPercentage,
        val: l,
        percentage: h.value.includes(t.uid) ? "-%" : o
      });
      return {
        ...t,
        shape: "circle",
        opacity: h.value.includes(t.uid) ? 0.5 : 1,
        segregate: () => X(t.uid),
        isSegregated: h.value.includes(t.uid),
        display: `${t.name}${e.value.style.chart.legend.showPercentage || e.value.style.chart.legend.showValue ? ": " : ""}${n}`
      };
    }).toSorted((t, a) => a.value - t.value)), It = c(() => ({
      cy: "rings-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), P = c(() => M.value.filter((t) => !h.value.includes(t.uid)).map(({ value: t }) => t).reduce((t, a) => t + a, 0)), k = c(() => M.value.filter((t) => !h.value.includes(t.uid)).map(({ name: t, value: a, color: l = null, uid: o, absoluteIndex: n }, g) => ({
      absoluteIndex: n,
      uid: o,
      name: t,
      color: l || it(l) || Ee.value[g] || E[g] || E[g % E.length],
      value: a,
      proportion: Xe(a),
      percentage: a / P.value * 100,
      strokeWidth: e.value.style.chart.layout.rings.strokeWidth * Xe(a)
    })).toSorted((t, a) => a.value - t.value));
    function Ft() {
      return k.value.map(
        ({ name: t, color: a, value: l, absoluteValues: o, percentage: n }) => ({
          name: t,
          color: a,
          value: l,
          absoluteValues: o,
          percentage: n
        })
      );
    }
    const B = c(() => St.value - e.value.style.chart.layout.rings.strokeWidth * 2);
    function Ge(t, a) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: a });
    }
    function je(t, a) {
      y.value = null, Z.value = !1, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: a });
    }
    function pe({
      val: t,
      percentage: a,
      showVal: l,
      showPercentage: o
    }) {
      const n = e.value.style.chart.layout.labels.dataLabels;
      return il({
        config: n,
        val: t,
        percentage: a,
        showVal: l,
        showPercentage: o
      });
    }
    function Dt(t) {
      const a = e.value.style.chart.layout.labels.dataLabels, l = $e(
        a.formatter,
        t.value,
        R({
          p: a.prefix,
          v: t.value,
          s: a.suffix,
          r: a.roundingValue
        })
      ), o = R({
        v: t.percentage,
        s: "%",
        r: a.roundingPercentage
      });
      return `${t.name}
${pe({
        val: l,
        percentage: o,
        showVal: a.showValue,
        showPercentage: a.showPercentage
      })}`;
    }
    function j(t) {
      return nl(B.value * t.proportion / 2 * 0.9 <= 0 ? 1e-4 : B.value * t.proportion / 2 * 0.9);
    }
    function Y(t, a) {
      return a === 0 ? A.value.y : A.value.y + B.value * k.value[0].proportion / 2 - B.value * t.proportion / 2 - 2 * (a + 1);
    }
    function O(t, a) {
      const o = e.value.style.chart.layout.labels.dataLabels.markers.position === "left" ? -(B.value / 2) : B.value / 2;
      return {
        x: A.value.x + o,
        y: Y(t, a) - j(t)
      };
    }
    const fe = u(null);
    function Ye(t, a) {
      if (e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: a }), h.value.length === m.dataset.length) return;
      fe.value = {
        datapoint: t,
        seriesIndex: a,
        series: k.value,
        config: e.value
      }, y.value = a;
      const l = k.value[a], o = e.value.style.chart.tooltip.customFormat;
      if (ul(o) && rl(() => o({
        seriesIndex: a,
        datapoint: t,
        series: k.value,
        config: e.value
      })))
        ne.value = o({
          seriesIndex: a,
          datapoint: t,
          series: k.value,
          config: e.value
        });
      else {
        let n = "";
        n += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${l.name}</div>`, n += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 60 60" height="14" width="14"><circle cx="30" cy="30" r="30" stroke="none" fill="${l.color}" />${Tt.pattern ? `<circle cx="30" cy="30" r="30" stroke="none" fill="url(#pattern_${$.value}_${t.absoluteIndex})" />` : ""}</svg>`, n += `<b>${pe({
          showVal: e.value.style.chart.tooltip.showValue,
          showPercentage: e.value.style.chart.tooltip.showPercentage,
          val: `<span>${$e(
            e.value.style.chart.layout.labels.dataLabels.formatter,
            l.value,
            R({
              p: e.value.style.chart.layout.labels.dataLabels.prefix,
              v: l.value,
              s: e.value.style.chart.layout.labels.dataLabels.suffix,
              r: e.value.style.chart.tooltip.roundingValue
            }),
            { datapoint: t, seriesIndex: a }
          )}</span>`,
          percentage: R({
            v: l.value / P.value * 100,
            s: "%",
            r: e.value.style.chart.tooltip.roundingPercentage
          })
        })}</b></div>`, ne.value = n;
      }
      Z.value = !0;
    }
    const F = c(() => {
      const t = k.value.map((l) => ({
        name: l.name,
        color: l.color
      })), a = k.value.map((l) => l.value);
      return { head: t, body: a };
    }), te = c(() => {
      const t = [
        ' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>',
        R({ p: e.value.style.chart.layout.labels.dataLabels.prefix, v: P.value, s: e.value.style.chart.layout.labels.dataLabels.suffix, r: e.value.table.td.roundingValue }),
        "100%"
      ], a = F.value.head.map((n, g) => [
        {
          color: n.color,
          name: n.name
        },
        R({ p: e.value.style.chart.layout.labels.dataLabels.prefix, v: F.value.body[g], s: e.value.style.chart.layout.labels.dataLabels.suffix, r: e.value.table.td.roundingValue }),
        isNaN(F.value.body[g] / P.value) ? "-" : (F.value.body[g] / P.value * 100).toFixed(e.value.table.td.roundingPercentage) + "%"
      ]), l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, o = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.value,
        e.value.table.columnNames.percentage
      ];
      return {
        head: t,
        body: a,
        config: l,
        colNames: o
      };
    });
    function ye(t = null) {
      st(() => {
        const a = F.value.head.map((n, g) => [[
          n.name
        ], [F.value.body[g]], [isNaN(F.value.body[g] / P.value) ? "-" : F.value.body[g] / P.value * 100]]), l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[""], ["val"], ["%"]]].concat(a), o = ol(l);
        t ? t(o) : sl({ csvContent: o, title: e.value.style.chart.title.text || "vue-ui-rings" });
      });
    }
    const U = u(!1);
    function qe(t) {
      U.value = t, xe.value += 1;
    }
    function Ke() {
      p.value.showTable = !p.value.showTable;
    }
    function Ze() {
      p.value.showTooltip = !p.value.showTooltip;
    }
    function Je() {
      p.value.showLabels = !p.value.showLabels;
    }
    const le = u(!1);
    function be() {
      le.value = !le.value;
    }
    async function Nt({ scale: t = 2 } = {}) {
      if (!S.value) return;
      const { width: a, height: l } = S.value.getBoundingClientRect(), o = a / l, { imageUri: n, base64: g } = await kl({ domElement: S.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: n,
        base64: g,
        title: e.value.style.chart.title.text,
        width: a,
        height: l,
        aspectRatio: o
      };
    }
    const q = c(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = p.value.showTable;
      return {
        component: t ? kt : pt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: U.value,
          fullscreenParent: S.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: V.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ae(() => p.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && H.value ? H.value.open() : "close" in H.value && H.value.close());
    });
    function Qe() {
      p.value.showTable = !1, re.value && re.value.setTableIconState(!1);
    }
    const zt = c(() => G.value.map((t) => ({
      ...t,
      name: t.display
    }))), Rt = c(() => e.value.style.chart.backgroundColor), Vt = c(() => e.value.style.chart.legend), Mt = c(() => e.value.style.chart.title), { exportSvg: Bt, getSvg: Ut } = fl({
      svg: ve,
      title: Mt,
      legend: Vt,
      legendItems: zt,
      backgroundColor: Rt
    });
    async function et({ isCb: t }) {
      Q.value = !0, await st();
      try {
        if (t) {
          const { blob: a, url: l, text: o, dataUrl: n } = await Ut();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: o, dataUrl: n }));
        } else
          await Promise.resolve(Bt());
      } finally {
        Q.value = !1;
      }
    }
    function Et(t) {
      if (t?.stage === "start") {
        J.value = !0;
        return;
      }
      if (t?.stage === "end") {
        J.value = !1;
        return;
      }
      Ue();
    }
    async function tt() {
      if (We("copyAlt", {
        config: e.value,
        dataset: k.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: k.value
      }));
    }
    return ct({
      getData: Ft,
      getImage: Nt,
      generatePdf: Be,
      generateCsv: ye,
      generateImage: Ue,
      generateSvg: et,
      hideSeries: Ot,
      showSeries: Pt,
      toggleTable: Ke,
      toggleTooltip: Ze,
      toggleAnnotator: be,
      toggleFullscreen: qe,
      toggleLabels: Je,
      copyAlt: tt
    }), (t, a) => (i(), f("div", {
      ref_key: "ringsChart",
      ref: S,
      class: z(`vue-data-ui-component vue-ui-rings ${U.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: N(`font-family:${e.value.style.fontFamily};text-align:center;width:100%;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height: 100%" : ""}`),
      id: `rings_${$.value}`,
      onMouseleave: a[2] || (a[2] = (l) => {
        y.value = null, Z.value = !1, s(ze)(!1);
      }),
      onMouseenter: a[3] || (a[3] = () => s(ze)(!0))
    }, [
      e.value.userOptions.buttons.annotator ? (i(), _(s(yt), {
        key: 0,
        svgRef: s(ve),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: le.value,
        isCursorPointer: V.value,
        onClose: be
      }, {
        "annotator-action-close": r(() => [
          v(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": r(({ color: l }) => [
          v(t.$slots, "annotator-action-color", T(L({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": r(({ mode: l }) => [
          v(t.$slots, "annotator-action-draw", T(L({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": r(({ disabled: l }) => [
          v(t.$slots, "annotator-action-undo", T(L({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": r(({ disabled: l }) => [
          v(t.$slots, "annotator-action-redo", T(L({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": r(({ disabled: l }) => [
          v(t.$slots, "annotator-action-delete", T(L({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : b("", !0),
      xt.value ? (i(), f("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Oe,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : b("", !0),
      e.value.style.chart.title.text ? (i(), f("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Se,
        style: "width:100%;background:transparent"
      }, [
        (i(), _(wl, {
          key: `title_${_e.value}`,
          config: {
            title: {
              cy: "rings-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "rings-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : b("", !0),
      C("div", {
        id: `legend-top-${$.value}`
      }, null, 8, Al),
      e.value.userOptions.show && Le.value && (s(Re) || s(ie)) ? (i(), _(s(bt), {
        ref_key: "userOptionsRef",
        ref: re,
        key: `user_options_${xe.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: s(he),
        isImaging: s(ge),
        uid: $.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        hasLabel: e.value.userOptions.buttons.labels,
        isTooltip: p.value.showTooltip,
        isFullscreen: U.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: S.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: le.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: V.value,
        onToggleFullscreen: qe,
        onGeneratePdf: s(Be),
        onGenerateCsv: ye,
        onGenerateImage: Et,
        onGenerateSvg: et,
        onToggleTable: Ke,
        onToggleTooltip: Ze,
        onToggleAnnotator: be,
        onToggleLabels: Je,
        onCopyAlt: tt,
        style: N({
          visibility: s(Re) ? s(ie) ? "visible" : "hidden" : "visible"
        })
      }, me({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: r(({ isOpen: l, color: o }) => [
            v(t.$slots, "menuIcon", T(L({ isOpen: l, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: r(() => [
            v(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: r(() => [
            v(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: r(() => [
            v(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: r(() => [
            v(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: r(() => [
            v(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: r(() => [
            v(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: r(() => [
            v(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: r(({ toggleFullscreen: l, isFullscreen: o }) => [
            v(t.$slots, "optionFullscreen", T(L({ toggleFullscreen: l, isFullscreen: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: r(({ toggleAnnotator: l, isAnnotator: o }) => [
            v(t.$slots, "optionAnnotator", T(L({ toggleAnnotator: l, isAnnotator: o })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: r(({ altCopy: l }) => [
            v(t.$slots, "optionAltCopy", T(L({ altCopy: l })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasFullscreen", "hasAltCopy", "hasLabel", "isTooltip", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : b("", !0),
      (i(), f("svg", {
        ref_key: "svgRef",
        ref: ve,
        xmlns: s(el),
        class: z({ "vue-data-ui-fullscreen--on": U.value, "vue-data-ui-fulscreen--off": !U.value, resizing: ue.value || s(x) }),
        viewBox: `0 0 ${d.value.width <= 0 ? 10 : d.value.width} ${d.value.height <= 0 ? 10 : d.value.height}`,
        style: N(`max-width:100%;overflow:hidden;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        oe(s(mt)),
        t.$slots["chart-background"] ? (i(), f("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: d.value.width <= 0 ? 10 : d.value.width,
          height: d.value.height <= 0 ? 10 : d.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          v(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Ol)) : b("", !0),
        C("defs", null, [
          (i(!0), f(se, null, ke(k.value, (l, o) => (i(), f("radialGradient", {
            cx: "50%",
            cy: "30%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `gradient_${$.value}_${o}`
          }, [
            C("stop", {
              offset: "0%",
              "stop-color": s(tl)(s(ll)(l.color, 0.05), 100 - e.value.style.chart.layout.rings.gradient.intensity)
            }, null, 8, Il),
            C("stop", {
              offset: "100%",
              "stop-color": l.color
            }, null, 8, Fl)
          ], 8, _l))), 256))
        ]),
        t.$slots.pattern ? (i(), f("g", Dl, [
          (i(!0), f(se, null, ke(k.value, (l) => (i(), f("defs", null, [
            v(t.$slots, "pattern", at({ ref_for: !0 }, { seriesIndex: l.absoluteIndex, patternId: `pattern_${$.value}_${l.absoluteIndex}` }), void 0, !0)
          ]))), 256))
        ])) : b("", !0),
        (i(!0), f(se, null, ke(k.value, (l, o) => (i(), f("g", null, [
          C("circle", {
            class: z({
              "vue-ui-rings-item": w.value && e.value.useCssAnimation,
              "vue-rings-item-onload": !w.value && e.value.useCssAnimation && !s(x),
              "vue-ui-rings-opacity": y.value !== null && y.value !== o
            }),
            style: N(`animation-delay:${o * 100}ms`),
            stroke: e.value.style.chart.layout.rings.stroke,
            cx: A.value.x,
            cy: Y(l, o),
            r: j(l),
            fill: e.value.style.chart.layout.rings.gradient.underlayerColor
          }, null, 14, Nl),
          C("circle", {
            class: z({
              "vue-ui-rings-item": w.value && e.value.useCssAnimation,
              "vue-rings-item-onload": !w.value && e.value.useCssAnimation && !s(x),
              "vue-ui-rings-shadow": e.value.style.chart.layout.rings.useShadow,
              "vue-ui-rings-blur": y.value !== null && y.value !== o
            }),
            style: N(`animation-delay:${o * 100}ms`),
            stroke: e.value.style.chart.layout.rings.stroke,
            "stroke-width": l.strokeWidth < 0.5 ? 0.5 : l.strokeWidth,
            cx: A.value.x,
            cy: Y(l, o),
            r: j(l),
            fill: e.value.style.chart.layout.rings.gradient.show ? `url(#gradient_${$.value}_${o})` : l.color
          }, null, 14, zl),
          t.$slots.pattern ? (i(), f("circle", {
            key: 0,
            class: z({
              "vue-ui-rings-item": w.value && e.value.useCssAnimation,
              "vue-rings-item-onload": !w.value && e.value.useCssAnimation && !s(x),
              "vue-ui-rings-shadow": e.value.style.chart.layout.rings.useShadow,
              "vue-ui-rings-blur": y.value !== null && y.value !== o
            }),
            style: N(`animation-delay:${o * 100}ms`),
            stroke: e.value.style.chart.layout.rings.stroke,
            "stroke-width": l.strokeWidth < 0.5 ? 0.5 : l.strokeWidth,
            cx: A.value.x,
            cy: Y(l, o),
            r: j(l),
            fill: `url(#pattern_${$.value}_${l.absoluteIndex})`
          }, null, 14, Rl)) : b("", !0),
          C("circle", {
            stroke: "none",
            cx: A.value.x,
            cy: Y(l, o),
            r: j(l),
            fill: "transparent",
            onMouseenter: (n) => Ye(l, o),
            onMouseleave: (n) => je(l, o),
            onClick: (n) => Ge(l, o)
          }, null, 40, Vl),
          p.value.showLabels ? (i(), f(se, { key: 1 }, [
            C("rect", {
              x: e.value.style.chart.layout.labels.dataLabels.markers.position === "left" ? O(l, o).x : A.value.x,
              y: O(l, o).y - e.value.style.chart.layout.labels.dataLabels.markers.strokeWidth / 2,
              width: Math.abs(A.value.x - O(l, o).x),
              height: e.value.style.chart.layout.labels.dataLabels.markers.strokeWidth,
              fill: e.value.style.chart.layout.labels.dataLabels.markers.stroke,
              rx: e.value.style.chart.layout.labels.dataLabels.markers.strokeWidth,
              class: z({
                "vue-ui-rings-item": w.value && e.value.useCssAnimation,
                "vue-rings-item-onload": !w.value && e.value.useCssAnimation && !s(x),
                "vue-ui-rings-shadow": e.value.style.chart.layout.rings.useShadow,
                "vue-ui-rings-blur": y.value !== null && y.value !== o
              })
            }, null, 10, Ml),
            C("circle", {
              cx: O(l, o).x,
              cy: O(l, o).y,
              r: e.value.style.chart.layout.labels.dataLabels.markers.radius,
              fill: l.color,
              stroke: e.value.style.chart.backgroundColor,
              class: z({
                "vue-ui-rings-item": w.value && e.value.useCssAnimation,
                "vue-rings-item-onload": !w.value && e.value.useCssAnimation && !s(x),
                "vue-ui-rings-shadow": e.value.style.chart.layout.rings.useShadow,
                "vue-ui-rings-blur": y.value !== null && y.value !== o
              })
            }, null, 10, Bl),
            C("text", {
              x: O(l, o).x + (e.value.style.chart.layout.labels.dataLabels.markers.position === "left" ? -e.value.style.chart.layout.labels.dataLabels.offsetX : e.value.style.chart.layout.labels.dataLabels.offsetX),
              y: O(l, o).y + e.value.style.chart.layout.labels.dataLabels.fontSize / 3,
              "text-anchor": e.value.style.chart.layout.labels.dataLabels.markers.position === "left" ? "end" : "start",
              "font-size": e.value.style.chart.layout.labels.dataLabels.fontSize,
              fill: e.value.style.chart.layout.labels.dataLabels.color,
              "font-weight": e.value.style.chart.layout.labels.dataLabels.bold ? "bold" : "normal",
              class: z({
                "vue-ui-rings-item": w.value && e.value.useCssAnimation,
                "vue-rings-item-onload": !w.value && e.value.useCssAnimation && !s(x),
                "vue-ui-rings-shadow": e.value.style.chart.layout.rings.useShadow,
                "vue-ui-rings-blur": y.value !== null && y.value !== o
              }),
              onMouseenter: (n) => Ye(l, o),
              onMouseleave: (n) => je(l, o),
              onClick: (n) => Ge(l, o),
              innerHTML: s(al)({
                content: Dt(l),
                fontSize: e.value.style.chart.layout.labels.dataLabels.fontSize,
                fill: e.value.style.chart.layout.labels.dataLabels.color,
                x: O(l, o).x + (e.value.style.chart.layout.labels.dataLabels.markers.position === "left" ? -e.value.style.chart.layout.labels.dataLabels.offsetX - 6 - e.value.style.chart.layout.labels.dataLabels.markers.radius : e.value.style.chart.layout.labels.dataLabels.offsetX + 6 + e.value.style.chart.layout.labels.dataLabels.markers.radius),
                y: O(l, o).y + e.value.style.chart.layout.labels.dataLabels.fontSize / 3,
                translateY: !0
              })
            }, null, 42, Ul)
          ], 64)) : b("", !0)
        ]))), 256)),
        v(t.$slots, "svg", {
          svg: {
            ...d.value,
            isPrintingImg: s(he) | s(ge) | J.value,
            isPrintingSvg: Q.value
          }
        }, void 0, !0)
      ], 14, Pl)),
      t.$slots.watermark ? (i(), f("div", El, [
        v(t.$slots, "watermark", T(L({ isPrinting: s(he) || s(ge) || J.value || Q.value })), void 0, !0)
      ])) : b("", !0),
      C("div", {
        id: `legend-bottom-${$.value}`
      }, null, 8, Wl),
      De.value ? (i(), _(Yt, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${$.value}` : `#legend-bottom-${$.value}`
      }, [
        C("div", {
          ref_key: "chartLegend",
          ref: Ae
        }, [
          e.value.style.chart.legend.show ? (i(), _(Tl, {
            key: `legend_${Fe.value}`,
            legendSet: G.value,
            config: It.value,
            isCursorPointer: V.value,
            onClickMarker: a[0] || (a[0] = ({ legend: l }) => X(l.uid))
          }, me({
            item: r(({ legend: l, index: o }) => [
              s(x) ? b("", !0) : (i(), f("div", {
                key: 0,
                onClick: (n) => X(l.uid),
                style: N(`opacity:${h.value.includes(l.uid) ? 0.5 : 1}`)
              }, Ce(l.display), 13, Hl))
            ]),
            legendToggle: r(() => [
              G.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !s(x) ? (i(), _(Ll, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: h.value.length > 0,
                isCursorPointer: V.value,
                onToggle: At
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : b("", !0)
            ]),
            _: 2
          }, [
            t.$slots.pattern ? {
              name: "legend-pattern",
              fn: r(({ legend: l, index: o }) => [
                oe(Cl, {
                  shape: l.shape,
                  radius: 30,
                  stroke: "none",
                  plot: { x: 30, y: 30 },
                  fill: `url(#pattern_${$.value}_${o})`
                }, null, 8, ["shape", "fill"])
              ]),
              key: "0"
            } : void 0
          ]), 1032, ["legendSet", "config", "isCursorPointer"])) : v(t.$slots, "legend", {
            key: 1,
            legend: G.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : b("", !0),
      t.$slots.source ? (i(), f("div", {
        key: 6,
        ref_key: "source",
        ref: Pe,
        dir: "auto"
      }, [
        v(t.$slots, "source", {}, void 0, !0)
      ], 512)) : b("", !0),
      oe(s(ht), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: p.value.showTooltip && Z.value && h.value.length < m.dataset.length,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        fontSize: e.value.style.chart.tooltip.fontSize,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: S.value,
        content: ne.value,
        isFullscreen: U.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": r(() => [
          v(t.$slots, "tooltip-before", T(L({ ...fe.value })), void 0, !0)
        ]),
        "tooltip-after": r(() => [
          v(t.$slots, "tooltip-after", T(L({ ...fe.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      Le.value && e.value.userOptions.buttons.table ? (i(), _(qt(q.value.component), at({ key: 7 }, q.value.props, {
        ref_key: "tableUnit",
        ref: H,
        onClose: Qe
      }), me({
        content: r(() => [
          (i(), _(s(ft), {
            key: `table_${Ie.value}`,
            colNames: te.value.colNames,
            head: te.value.head,
            body: te.value.body,
            config: te.value.config,
            title: e.value.table.useDialog ? "" : q.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: V.value,
            onClose: Qe
          }, {
            th: r(({ th: l }) => [
              C("div", {
                innerHTML: l,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Xl)
            ]),
            td: r(({ td: l }) => [
              ot(Ce(l.name || l), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: r(() => [
            ot(Ce(q.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: r(() => [
            C("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[1] || (a[1] = (l) => ye(e.value.userOptions.callbacks.csv)),
              style: N({ cursor: V.value ? "pointer" : "default" })
            }, [
              oe(s(gt), {
                name: "fileCsv",
                stroke: q.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : b("", !0),
      v(t.$slots, "skeleton", {}, () => [
        s(x) ? (i(), _(pl, { key: 0 })) : b("", !0)
      ], !0)
    ], 46, Sl));
  }
}, da = /* @__PURE__ */ xl(Gl, [["__scopeId", "data-v-3b6f84e5"]]);
export {
  da as default
};
