// vue-data-ui/dist/vue-ui-donut-CzCgCVsx.js
import { computed as g, defineAsyncComponent as Te, useSlots as ql, shallowRef as Pe, ref as k, onMounted as ul, onBeforeUnmount as rl, toRefs as Wl, watch as Ge, openBlock as u, createElementBlock as i, normalizeStyle as S, normalizeClass as Et, unref as n, createBlock as be, withCtx as x, renderSlot as m, normalizeProps as J, guardReactiveProps as Z, createCommentVNode as d, createElementVNode as w, createSlots as Ut, createVNode as _t, Fragment as T, renderList as W, mergeProps as Ht, toDisplayString as we, withDirectives as At, vShow as Tt, Teleport as jl, resolveDynamicComponent as Ql, createTextVNode as il, nextTick as cl } from "vue";
import { o as vl, q as Gt, r as Kl, j as Jl, t as dl, n as Zl, p as Ne, K as ea, s as hl, k as de, l as ue, a3 as fl, aw as ta, X as la, u as Pt, v as aa, a4 as oa, ad as Y, G as yl, a5 as ke, a6 as _, ar as gl, C as na, m as sa, a0 as ua, a1 as ra, T as ia, ah as pl, ax as ca } from "./lib-GwPnJSrY.js";
import { f as va, b as da } from "./labelUtils-CTXUc3V9.js";
import { t as ha, u as fa } from "./useResponsive-DfdjqQps.js";
import { u as ya } from "./useConfig-CIgyKZ6R.js";
import { u as ga } from "./usePrinter-DiK5R6k-.js";
import { u as pa, B as ma } from "./BaseScanner-BhhkKVoj.js";
import { u as ba } from "./useSvgExport-ByUukOZt.js";
import { u as qt } from "./useNestedProp-BBT0mbHh.js";
import { u as wa } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ka } from "./useUserOptionState-BIvW1Kz7.js";
import { u as xa } from "./useChartAccessibility-9icAAmYg.js";
import $a from "./img-BYYzY_QU.js";
import { _ as Ca } from "./Shape-CSyWYgXc.js";
import _a from "./Title-BfyrP-jp.js";
import Aa from "./vue_ui_donut-CA4dgV8J.js";
import Ta from "./Legend-1LMMsPxL.js";
import { B as Pa } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Sa } from "./_plugin-vue_export-helper-CHgC5LLL.js";
function La(ot) {
  const {
    FINAL_CONFIG: nt,
    noGhostDonut: St,
    svg: qe,
    padding: st,
    labels_inline_fontSize: Lt,
    minSize: ut,
    findArcMidpoint: Mt,
    calcMarkerOffsetX: It,
    calcMarkerOffsetY: Ft,
    animatingIndex: zt,
    segregated: Ot,
    isSmallArc: Yt
  } = ot;
  return { smallArcLayoutsClassic: g(() => {
    if (nt.value.type !== "classic") return {};
    const We = {}, V = St.value || [];
    if (!V.length) return We;
    const z = qe.value.width / 2, L = qe.value.height / 2, Se = st.value.top + 16, rt = qe.value.height - st.value.bottom - 16, je = Lt.value, it = je / 3, ct = je * 1.5, le = 8, he = 6, Qe = z - (ut.value + he), Ke = z + (ut.value + he), vt = !!nt.value.style.chart.layout.curvedMarkers;
    function dt({ midX: h, midY: f, bandX: p, bandY: $ }) {
      if (!vt)
        return `M ${h} ${f} L ${h} ${$} L ${p} ${$}`;
      const I = p < z ? -1 : 1, y = p - h, ee = $ - f, te = Math.sqrt(y * y + ee * ee) || 1, D = h - z, xe = f - L, re = Math.sqrt(D * D + xe * xe) || 1, e = D / re, j = xe / re, fe = re + 9;
      function ye({ x: Ce, y: Fe }) {
        const X = Ce - z, oe = Fe - L, ze = Math.sqrt(X * X + oe * oe) || 1;
        if (ze >= fe) return { x: Ce, y: Fe };
        const ge = fe / ze;
        return {
          x: z + X * ge,
          y: L + oe * ge
        };
      }
      if (te < 56) {
        const Ce = y / te;
        let X = -(ee / te), oe = Ce;
        const ze = (h + p) * 0.5, ge = (f + $) * 0.5, Ue = ze + X, He = ge + oe, mt = (Ue - z) * (Ue - z) + (He - L) * (He - L), et = ze - X, bt = ge - oe;
        (et - z) * (et - z) + (bt - L) * (bt - L) > mt && (X = -X, oe = -oe);
        const b = 0.78, s = h + y * b, pe = f + ee * b, Oe = Math.max(0, Math.min(1, (te - 18) / 44)), H = Oe * Oe * (3 - 2 * Oe);
        let O = 2.5 + H * 4, wt = 1 + H * 2.5;
        O *= 0.9;
        const F = {
          x: s + X * O + e * wt,
          y: pe + oe * O + j * wt
        }, ne = ye(F);
        return `M ${h} ${f} Q ${ne.x} ${ne.y} ${p} ${$}`;
      }
      let $e = te * 0.34;
      $e < 20 && ($e = 20), $e > 46 && ($e = 46);
      let A = te * 0.46;
      A < 22 && (A = 22), A > 70 && (A = 70);
      const Me = {
        x: h + e * $e,
        y: f + j * $e
      }, gt = z + I * Math.max(Math.abs(p - z), fe), Ie = {
        x: p - I * Math.min(A, Math.abs(gt - p) * 0.75),
        y: $
      }, pt = ye(Me), Ee = ye(Ie);
      return `M ${h} ${f} C ${pt.x} ${pt.y} ${Ee.x} ${Ee.y} ${p} ${$}`;
    }
    function ht(h) {
      const f = String(h ?? "").split(/\n/g), p = Math.max(0, f.length - 1), $ = je * 1.2, I = p * $;
      return ct + I;
    }
    function ft({ arc: h, index: f }) {
      const { x: p, y: $ } = Mt(h.path), I = It(h).x, y = Ft(h) - 3.5, ee = ht(h.name);
      return {
        arc: h,
        index: f,
        midX: p,
        midY: $,
        inlineMarkerX: I,
        inlineMarkerY: y,
        labelHeight: ee
      };
    }
    function Le(h) {
      const { arc: f } = h, p = f.seriesIndex ?? 0;
      return zt.value === p || Ot.value.includes(p) ? !1 : Yt(f, p);
    }
    function Je(h) {
      const f = h.inlineMarkerY < L, p = h.inlineMarkerX < z;
      return f && p ? "TL" : f && !p ? "TR" : !f && p ? "BL" : "BR";
    }
    function Re(h, f) {
      if (f.startsWith("T")) {
        h.sort(($, I) => $.inlineMarkerY - I.inlineMarkerY || $.index - I.index);
        return;
      }
      h.sort(($, I) => I.inlineMarkerY - $.inlineMarkerY || $.index - I.index);
    }
    function Ve({ side: h, markerX: f, markerY: p, labelY: $, connectorPath: I }) {
      return {
        side: h,
        labelX: h === "left" ? f - le : f + le,
        labelY: $ + it,
        textAnchor: h === "left" ? "end" : "start",
        markerX: f,
        markerY: p,
        connectorPath: I
      };
    }
    function ae({
      candidateList: h,
      side: f,
      bandMarkerX: p,
      startY: $,
      direction: I
    }) {
      let y = $;
      h.forEach((ee) => {
        const { index: te, midX: D, midY: xe, labelHeight: re } = ee;
        let e;
        I === "down" ? (e = y, y += re) : (y -= re, e = y);
        const j = e, yt = dt({
          midX: D,
          midY: xe,
          bandX: p,
          bandY: j
        });
        We[te] = Ve({
          side: f,
          markerX: p,
          markerY: j,
          labelY: e,
          connectorPath: yt
        });
      });
    }
    const E = V.map((h, f) => ft({ arc: h, index: f })).filter(Le), U = {
      TL: [],
      TR: [],
      BL: [],
      BR: []
    };
    return E.forEach((h) => {
      U[Je(h)].push(h);
    }), Object.keys(U).forEach((h) => {
      Re(U[h], h);
    }), ae({
      candidateList: U.TL,
      side: "left",
      bandMarkerX: Qe,
      startY: Se,
      direction: "down"
    }), ae({
      candidateList: U.TR,
      side: "right",
      bandMarkerX: Ke,
      startY: Se,
      direction: "down"
    }), U.BL.length > 1 && ae({
      candidateList: U.BL,
      side: "left",
      bandMarkerX: Qe,
      startY: rt,
      direction: "up"
    }), U.BR.length > 1 && ae({
      candidateList: U.BR,
      side: "right",
      bandMarkerX: Ke,
      startY: rt,
      direction: "up"
    }), We;
  }) };
}
const Ma = ["id"], Ia = ["id"], Fa = ["xmlns", "viewBox"], za = ["width", "height"], Oa = { key: 1 }, Ya = ["id"], Da = ["stop-color"], Xa = ["offset", "stop-color"], Ba = ["offset", "stop-color"], Na = ["stop-color"], Ra = { key: 2 }, Va = ["id", "cx", "cy"], Ea = ["stop-color", "stop-opacity"], Ua = ["stop-color"], Ha = ["id"], Ga = ["id"], qa = ["id"], Wa = ["flood-color"], ja = ["id"], Qa = ["flood-color"], Ka = ["d", "stroke", "filter"], Ja = ["d", "stroke", "filter"], Za = ["cx", "cy", "r", "fill", "filter"], eo = { key: 6 }, to = ["stroke", "d"], lo = ["d", "fill", "stroke", "stroke-width", "filter"], ao = { key: 0 }, oo = ["d", "fill", "stroke", "stroke-width", "filter"], no = { key: 0 }, so = ["stroke", "d"], uo = { key: 0 }, ro = ["d", "stroke", "stroke-width", "filter"], io = { key: 1 }, co = ["d", "fill", "stroke", "stroke-width", "filter"], vo = ["d", "fill", "stroke", "stroke-width", "filter"], ho = { key: 1 }, fo = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], yo = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], go = { key: 0 }, po = ["d", "fill", "stroke", "stroke-width"], mo = ["cx", "cy", "r", "stroke"], bo = ["cx", "cy", "r", "fill"], wo = ["cx", "cy", "r"], ko = { key: 0 }, xo = ["d", "stroke", "fill", "onMouseenter", "onMouseleave", "onClick"], $o = { key: 1 }, Co = ["cx", "cy", "r"], _o = ["x", "y", "fill", "font-size"], Ao = ["x", "y", "fill", "font-size"], To = ["x", "y", "fill", "font-size"], Po = ["x", "y", "fill", "font-size"], So = ["filter", "opacity"], Lo = { key: 0 }, Mo = ["x", "y"], Io = { key: 1 }, Fo = ["cx", "cy", "fill", "stroke", "filter", "onClick", "onMouseenter", "onMouseleave"], zo = ["text-anchor", "x", "y", "onClick", "onMouseenter", "onMouseleave", "innerHTML"], Oo = ["text-anchor", "x", "y", "fill", "font-size", "onClick", "onMouseenter", "onMouseleave"], Yo = ["text-anchor", "x", "y", "fill", "font-size", "onClick", "onMouseenter", "onMouseleave", "innerHTML"], Do = ["cx", "cy", "fill", "stroke", "filter", "onClick", "onMouseenter", "onMouseleave"], Xo = ["text-anchor", "x", "y", "onClick", "onMouseenter", "onMouseleave", "innerHTML"], Bo = ["text-anchor", "x", "y", "fill", "font-size", "onClick", "onMouseenter", "onMouseleave"], No = ["text-anchor", "x", "y", "fill", "font-size", "onClick", "onMouseenter", "onMouseleave", "innerHTML"], Ro = { key: 2 }, Vo = ["x", "y", "width"], Eo = ["x", "y", "width"], Uo = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Ho = ["id"], Go = ["onClick"], qo = {
  key: 7,
  class: "vue-ui-donut-hollow"
}, Wo = ["innerHTML"], jo = {
  __name: "vue-ui-donut",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "selectDatapoint", "copyAlt"],
  setup(ot, { expose: nt, emit: St }) {
    const qe = Te(() => import("./Tooltip-DL3QgjZf.js")), st = Te(() => import("./BaseIcon-CNx8Tzl5.js")), Lt = Te(() => import("./vue-ui-accordion-BGHTqIBt.js")), ut = Te(() => import("./DataTable-Bo1pmAqG.js")), Mt = Te(() => import("./PenAndPaper-DWV1FMIi.js")), It = Te(() => import("./UserOptions-CrLqEimB.js")), Ft = Te(() => import("./PackageVersion-DN9d0ME5.js")), zt = Te(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_donut: Ot } = ya(), { isThemeValid: Yt, warnInvalidTheme: Wt } = wa(), We = ql(), V = ot, z = g({
      get() {
        return !!V.dataset && V.dataset.length;
      },
      set(l) {
        return l;
      }
    }), L = Pe(null), Se = k(null), rt = k(null), je = k(null), it = Pe(null), ct = Pe(null), le = Pe(null), he = Pe(null), Qe = Pe(null), Ke = Pe(null), vt = k(0), dt = k(0), ht = k(0), ft = k(!1), Le = k(null), Je = k(null), Re = k(!1), Ve = k(!1), ae = g({
      get: () => e.value.style.chart.layout.labels.percentage.fontSize,
      set: (l) => l
    }), E = g({
      get: () => e.value.style.chart.layout.labels.name.fontSize,
      set: (l) => l
    }), U = g({
      get: () => e.value.style.chart.layout.labels.dataLabels.smallArcClusterFontSize,
      set: (l) => l
    });
    let h = !1;
    const f = () => {
      !e.value.autoSize || h || (h = !0, requestAnimationFrame(() => {
        h = !1;
        const l = e.value, o = Se.value, t = X.value;
        if (!l.autoSize || !o || !t) return;
        const [a, r, c, C] = t.getAttribute("viewBox").split(" ").map(Number), P = { x: a, y: r, width: c, height: C }, N = [
          {
            selector: ".vue-data-ui-datalabel-value",
            baseSize: l.style.chart.layout.labels.percentage.fontSize,
            minSize: l.style.chart.layout.labels.percentage.minFontSize,
            sizeRef: ae
          },
          {
            selector: ".vue-data-ui-datalabel-name",
            baseSize: l.style.chart.layout.labels.name.fontSize,
            minSize: l.style.chart.layout.labels.name.minFontSize,
            sizeRef: E
          },
          {
            selector: ".vue-data-ui-datalabel-inline",
            baseSize: l.style.chart.layout.labels.dataLabels.smallArcClusterFontSize,
            minSize: l.style.chart.layout.labels.name.minFontSize,
            sizeRef: U
          }
        ];
        N.map((q) => o.querySelectorAll(q.selector).length).reduce((q, Be) => q + Be, 0) !== 0 && N.forEach(({ selector: q, baseSize: Be, minSize: El, sizeRef: Ul }) => {
          o.querySelectorAll(q).forEach((Hl) => {
            const Gl = na({
              el: Hl,
              bounds: P,
              currentFontSize: Be,
              minFontSize: El,
              attempts: 200,
              padding: 1
            });
            Ul.value = Gl;
          });
        });
      }));
    };
    ul(async () => {
      ft.value = !0, I(), requestAnimationFrame(f);
    });
    let p;
    ul(() => {
      L.value && (p = new ResizeObserver((l) => {
        for (const o of l) {
          const { width: t, height: a } = o.contentRect;
          if (t > 0 && a > 0) {
            f();
            break;
          }
        }
      }), p.observe(L.value.parentElement));
    }), rl(() => {
      p?.disconnect();
    }), rl(() => {
      le.value && (he.value && le.value.unobserve(he.value), le.value.disconnect());
    });
    const $ = g(() => !!e.value.debug);
    function I() {
      if (vl(V.dataset) ? (Gt({
        componentName: "VueUiDonut",
        type: "dataset",
        debug: $.value
      }), z.value = !1, Ze.value = !0) : (V.dataset.forEach((l, o) => {
        Kl({
          datasetObject: l,
          requiredAttributes: ["name", "values"]
        }).forEach((t) => {
          Gt({
            componentName: "VueUiDonut",
            type: "datasetSerieAttribute",
            property: t,
            index: o,
            debug: $.value
          }), z.value = !1, Ze.value = !0;
        });
      }), V.dataset.forEach((l, o) => {
        (!l.name || l.name === "") && Gt({
          componentName: "VueUiDonut",
          type: "datasetAttributeEmpty",
          property: "name",
          index: o,
          debug: $.value
        });
      })), vl(V.dataset) || (Ze.value = e.value.loading), e.value.responsive) {
        const l = ha(() => {
          const { width: o, height: t } = fa({
            chart: L.value,
            title: e.value.style.chart.title.text ? it.value : null,
            legend: e.value.style.chart.legend.show ? ct.value : null,
            source: Qe.value,
            noTitle: Ke.value,
            padding: e.value.autoSize ? void 0 : ge.value
          });
          requestAnimationFrame(() => {
            s.value.width = o, s.value.height = t, f();
          });
        });
        le.value && (he.value && le.value.unobserve(he.value), le.value.disconnect()), le.value = new ResizeObserver(l), he.value = L.value.parentNode, le.value.observe(he.value);
      }
    }
    const y = k(Jl()), ee = k(!1), te = k(""), D = k(null), xe = k(0);
    function re() {
      const l = qt({
        userConfig: V.config,
        defaultConfig: Ot
      });
      let o = {};
      const t = l.theme;
      if (!t)
        o = l;
      else if (!Yt.value(l))
        Wt(l), o = l;
      else {
        const a = qt({
          userConfig: Aa[t] || V.config,
          defaultConfig: l
        });
        o = {
          ...qt({
            userConfig: V.config,
            defaultConfig: a
          }),
          customPalette: l.customPalette.length ? l.customPalette : sa[t] || Ne
        };
      }
      return o;
    }
    const e = k(re()), j = g(() => e.value.userOptions.useCursorPointer), yt = g(() => dl({
      defaultConfig: {
        useCssAnimation: !1,
        table: { show: !1 },
        startAnimation: {
          show: !1
        },
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              labels: {
                dataLabels: { show: !1 },
                hollow: {
                  average: { show: !1 },
                  total: { show: !1 }
                },
                value: { show: !1 }
              }
            },
            legend: {
              backgroundColor: "transparent",
              showValue: !1,
              showPercentage: !1
            },
            title: {
              color: "#1A1A1A",
              subtitle: {
                color: "#5A5A5A"
              }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: fe, FINAL_DATASET: ye, manualLoading: Ze, skeletonDataset: $e } = pa({
      ...Wl(V),
      FINAL_CONFIG: e,
      prepareConfig: re,
      skeletonDataset: V.config?.skeletonDataset ?? [
        {
          name: "",
          values: [3],
          color: "#BABABA"
        },
        {
          name: "",
          values: [2],
          color: "#AAAAAA"
        },
        {
          name: "",
          values: [1],
          color: "#CACACA"
        }
      ],
      skeletonConfig: dl({
        defaultConfig: e.value,
        userConfig: yt.value
      })
    }), A = k(!0), Me = k(0), gt = k(!1);
    let Ie = null;
    function pt(l = 1e3) {
      return new Promise((o) => {
        const t = performance.now();
        function a(r) {
          const c = r - t, C = Math.min(
            c / l,
            1
          ), P = pl(C);
          Me.value = P, C < 1 ? Ie = requestAnimationFrame(a) : (Me.value = 1, Ie = null, o());
        }
        Ie !== null && cancelAnimationFrame(Ie), Me.value = 0, Ie = requestAnimationFrame(a);
      });
    }
    Ge(
      () => fe.value,
      async (l) => {
        if (l || gt.value)
          return;
        const o = ye.value === $e, t = e.value.startAnimation?.show;
        if (!o && t) {
          gt.value = !0;
          const a = e.value.startAnimation.durationMs || 1e3;
          await pt(a);
        } else
          Me.value = 1;
        A.value = !1;
      },
      { immediate: !0 }
    );
    const { userOptionsVisible: Ee, setUserOptionsVisibility: Ce, keepUserOptionState: Fe } = ka({ config: e.value }), { svgRef: X } = xa({ config: e.value.style.chart.title });
    function oe() {
      Ce(!0);
    }
    function ze() {
      Ce(!1);
    }
    Ge(() => V.config, (l) => {
      fe.value || (e.value = re()), Ee.value = !e.value.userOptions.showOnChartHover, I(), vt.value += 1, dt.value += 1, ht.value += 1, b.value.dataLabels.show = e.value.style.chart.layout.labels.dataLabels.show, b.value.showTable = e.value.table.show, b.value.showTooltip = e.value.style.chart.tooltip.show, s.value.height = e.value.style.chart.height, s.value.width = e.value.style.chart.width;
    }, { deep: !0 });
    const ge = g(() => {
      const { top: l, right: o, bottom: t, left: a } = e.value.style.chart.padding;
      return {
        css: `padding:${l}px ${o}px ${t}px ${a}px`,
        top: l,
        right: o,
        bottom: t,
        left: a
      };
    }), { isPrinting: Ue, isImaging: He, generatePdf: mt, generateImage: et } = ga({
      elementId: `donut__${y.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-donut",
      options: e.value.userOptions.print
    }), bt = g(() => e.value.userOptions.show && !e.value.style.chart.title.text), Dt = g(() => Zl(e.value.customPalette)), b = k({
      dataLabels: {
        show: e.value.style.chart.layout.labels.dataLabels.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    Ge(e, () => {
      b.value = {
        dataLabels: {
          show: e.value.style.chart.layout.labels.dataLabels.show
        },
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const s = k({
      height: e.value.style.chart.height,
      width: e.value.style.chart.width
    }), pe = g(() => {
      if (e.value.pie)
        return M.value;
      const l = e.value.style.chart.layout.donut.strokeWidth / 512, o = Math.min(s.value.width, s.value.height) * l, t = o > M.value ? M.value : o;
      return Math.max(t, 12 * (1 + l));
    }), Oe = St, H = g(() => ye.value.sort((l, o) => {
      const t = Array.isArray(l.values) ? l.values.reduce((r, c) => r + c, 0) : l.value ?? 0, a = Array.isArray(o.values) ? o.values.reduce((r, c) => r + c, 0) : o.value ?? 0;
      return l.ghost && !o.ghost ? 1 : o.ghost && !l.ghost ? -1 : a - t;
    }).map((l, o) => ({
      name: l.name,
      color: hl(l.color) || Dt.value[o] || Ne[o] || Ne[o % Ne.length],
      value: ea(l.values.reduce((t, a) => t + a, 0)),
      absoluteValues: l.values,
      comment: l.comment || "",
      patternIndex: o,
      seriesIndex: o,
      ghost: !1,
      pattern: `pattern_${y.value}_${o}`
    })));
    Ge(() => V.dataset, (l) => {
      Array.isArray(l) && l.length > 0 && (Ze.value = !1);
    }, { immediate: !0 });
    const O = Pe(H.value);
    Ge(() => H.value, (l) => {
      O.value = l, requestAnimationFrame(f);
    });
    function wt() {
      return H.value.map((l) => ({
        name: l.name,
        color: l.color,
        value: l.value
      }));
    }
    const F = k([]), ne = k(!1), Ye = k(null);
    function jt({ from: l, to: o, duration: t, onUpdate: a, onDone: r, easing: c = pl }) {
      const C = performance.now();
      function P(N) {
        const R = Math.min((N - C) / t, 1), q = c(R), Be = l + (o - l) * q;
        a(Be, R), R < 1 ? requestAnimationFrame(P) : (a(o, 1), r && r());
      }
      requestAnimationFrame(P);
    }
    function ml() {
      F.value.length ? F.value = [] : tt.value.forEach((l, o) => {
        F.value.push(o);
      });
    }
    function kt(l) {
      const o = H.value.find((r, c) => c === l);
      let a = O.value.find((r, c) => c === l).value;
      if (F.value.includes(l)) {
        let c = function() {
          O.value = O.value.map(
            (P, N) => l === N ? { ...P, value: r } : P
          );
        }, C = function() {
          ne.value = !0, Ye.value = l, jt({
            from: a,
            to: r,
            duration: e.value.serieToggleAnimation.durationMs,
            onUpdate: (P, N) => {
              O.value = O.value.map(
                (R, q) => l === q ? { ...R, value: P } : R
              ), requestAnimationFrame(f);
            },
            onDone: () => {
              c(), ne.value = !1, Ye.value = null;
            }
          });
        };
        F.value = F.value.filter((P) => P !== l);
        const r = o.value;
        e.value.serieToggleAnimation.show && e.value.type === "classic" ? C() : (c(), requestAnimationFrame(f));
      } else if (F.value.length < H.value.length - 1) {
        let r = function() {
          F.value.push(l), O.value = O.value.map(
            (C, P) => l === P ? { ...C, value: 0 } : C
          );
        }, c = function() {
          ne.value = !0, Ye.value = l, jt({
            from: a,
            to: 0,
            duration: e.value.serieToggleAnimation.durationMs,
            onUpdate: (C, P) => {
              O.value = O.value.map(
                (N, R) => l === R ? { ...N, value: C } : N
              ), requestAnimationFrame(f);
            },
            onDone: () => {
              r(), requestAnimationFrame(f), ne.value = !1, Ye.value = null;
            }
          });
        };
        e.value.serieToggleAnimation.show && e.value.type === "classic" ? c() : r();
      }
      Oe("selectLegend", me.value.map((r) => ({
        name: r.name,
        color: r.color,
        value: r.value
      })));
    }
    function Qt(l) {
      if (!H.value.length)
        return e.value.debug && console.warn("VueUiDonut - There are no series to show."), null;
      const o = H.value.find((t) => t.name === l);
      return o || (e.value.debug && console.warn(`VueUiDonut - Series name not found "${l}"`), null);
    }
    function bl(l) {
      const o = Qt(l);
      o !== null && F.value.includes(o.seriesIndex) && kt(o.seriesIndex);
    }
    function wl(l) {
      const o = Qt(l);
      o !== null && (F.value.includes(o.seriesIndex) || kt(o.seriesIndex));
    }
    const kl = g(
      () => ye.value.reduce(
        (l, o) => l + o.values.reduce((t, a) => t + a, 0),
        0
      )
    ), me = g(() => {
      if (A.value && !fe.value) {
        const l = Me.value, o = H.value.map((a) => ({
          ...a,
          value: a.value * l,
          color: a.color,
          ghost: !1
        })), t = kl.value * (1 - l);
        return t > 0 && o.push({
          name: "__ghost__",
          value: t,
          color: "transparent",
          ghost: !0
        }), o;
      } else
        return O.value.forEach((l) => {
          if ([null, void 0].includes(l.values))
            return {
              ...l,
              values: []
            };
        }), O.value.map((l, o) => ({
          ...l,
          seriesIndex: o
        })).filter((l, o) => !F.value.includes(o));
    }), tt = g(() => ye.value.map((l, o) => {
      const t = (l.values || []).reduce((r, c) => r + c, 0), a = t / ye.value.map((r) => (r.values || []).reduce((c, C) => c + C, 0)).reduce((r, c) => r + c, 0);
      return {
        name: l.name,
        color: hl(l.color) || Dt.value[o] || Ne[o] || Ne[o % Ne.length],
        value: t,
        shape: "circle",
        patternIndex: o,
        proportion: a
      };
    }).map((l, o) => {
      const t = de(e.value.style.chart.layout.labels.value.formatter, l.value, ue({
        p: e.value.style.chart.layout.labels.dataLabels.prefix,
        v: l.value,
        s: e.value.style.chart.layout.labels.dataLabels.suffix,
        r: e.value.style.chart.legend.roundingValue
      }), { datapoint: l, index: o }), a = de(e.value.style.chart.layout.labels.percentage.formatter, Jt(l), ue({
        v: Jt(l),
        s: "%",
        r: e.value.style.chart.legend.roundingPercentage
      })), r = Rt({
        val: t,
        percentage: F.value.includes(o) ? `${Fl(l.proportion * 100)}%` : a,
        showVal: e.value.style.chart.legend.showValue,
        showPercentage: e.value.style.chart.legend.showPercentage,
        config: e.value.style.chart.legend
      });
      return {
        ...l,
        opacity: F.value.includes(o) ? 0.5 : 1,
        segregate: () => !ne.value && kt(o),
        isSegregated: F.value.includes(o),
        display: `${l.name}${e.value.style.chart.legend.showPercentage || e.value.style.chart.legend.showValue ? ": " : ""}${r}`
      };
    })), xl = g(() => ({
      cy: "donut-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), M = g(() => {
      const l = e.value.style.chart.layout.donut.radiusRatio, o = Math.max(0.1, Math.min(0.50001, l)), t = Math.min(s.value.width * o, s.value.height * o);
      return Math.max(12, t);
    }), G = g(() => fl({ series: me.value }, s.value.width / 2, s.value.height / 2, M.value, M.value, 1.99999, 2, 1, 360, 105.25, pe.value)), $l = g(() => Math.abs(me.value.map((l) => l.value).reduce((l, o) => l + o, 0)) > 0), Cl = g(() => fl({
      series: [
        {
          value: 1,
          color: e.value.style.chart.layout.donut.emptyFill,
          name: "_",
          seriesIndex: 0,
          patternIndex: -1,
          ghost: !1,
          absoluteValues: [1]
        }
      ]
    }, s.value.width / 2, s.value.height / 2, M.value, M.value, 1.99999, 2, 1, 360, 105.25, pe.value)), se = g(() => G.value.filter((l) => !l.ghost)), v = g(() => {
      const l = Math.max(...me.value.map((t) => t.value)), o = me.value.map((t) => t.value / l);
      return ta({
        series: o,
        center: {
          x: s.value.width / 2,
          y: s.value.height / 2
        },
        maxRadius: Math.min(s.value.width, s.value.height) / 3,
        hasGhost: A.value
      });
    });
    function _e(l) {
      return l.x > s.value.width / 2 + 6 ? "start" : l.x < s.value.width / 2 - 6 ? "end" : "middle";
    }
    function _l(l) {
      return l.middlePoint.y > s.value.height / 2 ? Y({ initX: l.middlePoint.x, initY: l.middlePoint.y, offset: 100, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y : Y({ initX: l.middlePoint.x, initY: l.middlePoint.y, offset: 0, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y - 100;
    }
    function Q(l) {
      return l.proportion * 100 > e.value.style.chart.layout.labels.dataLabels.hideUnderValue;
    }
    function Al(l, o) {
      const t = e.value.style.chart.layout.labels.dataLabels.hideUnderValue, a = e.value.style.chart.layout.labels.dataLabels.smallArcClusterThreshold, c = (Pl.value[o] ?? l.proportion ?? 0) * 100;
      return c > t && c <= a;
    }
    const { smallArcLayoutsClassic: B } = La({
      FINAL_CONFIG: e,
      noGhostDonut: se,
      svg: s,
      padding: ge,
      labels_inline_fontSize: U,
      minSize: M,
      findArcMidpoint: ca,
      calcMarkerOffsetX: _,
      calcMarkerOffsetY: ke,
      animatingIndex: Ye,
      segregated: F,
      isSmallArc: Al
    });
    function Kt(l, o) {
      const t = l.value / Tl(o);
      return isNaN(t) ? 0 : de(
        e.value.style.chart.layout.labels.percentage.formatter,
        t * 100,
        ue({
          v: t * 100,
          s: "%",
          r: e.value.style.chart.layout.labels.percentage.rounding
        }),
        { datapoint: l }
      );
    }
    function Tl(l) {
      return [...l].map((o) => o.value).reduce((o, t) => o + t, 0);
    }
    const K = g(() => me.value.map((l) => l.value).reduce((l, o) => l + o, 0)), Pl = g(() => {
      const l = H.value.reduce((o, t) => o + t.value, 0);
      return l <= 0 ? [] : H.value.map((o) => o.value / l);
    }), Xt = g(() => K.value / me.value.length);
    function Jt(l) {
      return ne.value ? l.proportion * 100 : l.value / K.value * 100;
    }
    const Bt = k(null), lt = k(!1);
    function ie({ datapoint: l, seriesIndex: o }) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: o }), ee.value = !1, D.value = null;
    }
    function ce({ datapoint: l, relativeIndex: o, seriesIndex: t, show: a = !1 }) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: l, seriesIndex: t }), Bt.value = { datapoint: l, seriesIndex: t, config: e.value, series: H.value }, ee.value = a, D.value = o;
      let r = "";
      const c = e.value.style.chart.tooltip.customFormat;
      if (lt.value = !1, ia(c))
        try {
          const C = c({
            seriesIndex: t,
            datapoint: l,
            series: H.value,
            config: e.value
          });
          typeof C == "string" && (te.value = C, lt.value = !0);
        } catch {
          console.warn("Custom format cannot be applied."), lt.value = !1;
        }
      lt.value || (r += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${l.name}</div>`, r += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 60 60" height="14" width="14"><circle cx="30" cy="30" r="30" stroke="none" fill="${l.color}"/>${We.pattern ? `<circle cx="30" cy="30" r="30" stroke="none" fill="url(#pattern_${y.value}_${t})"/>` : ""}</svg>`, r += `<b>${Rt({
        showVal: e.value.style.chart.tooltip.showValue,
        showPercentage: e.value.style.chart.tooltip.showPercentage,
        val: `<span>${de(
          e.value.style.chart.layout.labels.value.formatter,
          l.value,
          ue({
            p: e.value.style.chart.layout.labels.dataLabels.prefix,
            v: l.value,
            s: e.value.style.chart.layout.labels.dataLabels.suffix,
            r: e.value.style.chart.tooltip.roundingValue
          }),
          {
            datapoint: l,
            relativeIndex: o,
            seriesIndex: t
          }
        )}</span>`,
        percentage: de(
          e.value.style.chart.layout.labels.percentage.formatter,
          l.proportion * 100,
          ue({
            v: l.proportion * 100,
            s: "%",
            r: e.value.style.chart.tooltip.roundingPercentage
          }),
          {
            datapoint: l,
            relativeIndex: o,
            seriesIndex: t
          }
        ),
        config: e.value.style.chart.tooltip
      })}</b></div>`, e.value.style.chart.comments.showInTooltip && l.comment && (r += `<div class="vue-data-ui-tooltip-comment" style="background:${l.color}20; padding: 6px; margin-bottom: 6px; margin-top:6px; border-left: 1px solid ${l.color}">${l.comment}</div>`), te.value = `<div>${r}</div>`);
    }
    function Sl(l, o) {
      const t = B.value[o];
      if (t)
        return {
          textAnchor: t.textAnchor,
          x: t.labelX,
          y: t.labelY
        };
      const a = _(l, !0, 12);
      return {
        textAnchor: a.anchor,
        x: a.x,
        y: ke(l)
      };
    }
    function Ll(l, o, t) {
      const { textAnchor: a, x: r } = Sl(l, o), c = $t(l), C = `
        <tspan
            class="vue-data-ui-datalabel-inline"
            fill="${e.value.style.chart.layout.labels.percentage.color}"
            font-size="${t ? U.value : ae.value}px"
            style="font-weight:${e.value.style.chart.layout.labels.percentage.bold ? "bold" : ""}"
        >${c}</tspan>
    `, P = String(l.name ?? "").split(/\n/g);
      let N = "";
      return P.forEach((R, q) => {
        q === 0 ? N += `
                <tspan
                    class="${t ? "vue-data-ui-datalabel-inline" : "vue-data-ui-datalabel-name"}"
                    fill="${e.value.style.chart.layout.labels.name.color}"
                    font-size="${t ? U.value : E.value}px"
                    style="font-weight:${e.value.style.chart.layout.labels.name.bold ? "bold" : ""}"
                >${R}</tspan>
            ` : N += `
                <tspan
                    class="${t ? "vue-data-ui-datalabel-inline" : "vue-data-ui-datalabel-name"}"
                    x="${r}"
                    dy="${(t ? U.value : E.value) * 1.2}"
                    fill="${e.value.style.chart.layout.labels.name.color}"
                    font-size="${t ? U.value : E.value}px"
                    style="font-weight:${e.value.style.chart.layout.labels.name.bold ? "bold" : ""}"
                >${R}</tspan>
            `;
      }), a === "end" ? `${N}${C}` : `${C}${N}`;
    }
    function Ml(l, o) {
      const t = v.value[o].middlePoint, a = _e(t), r = Y({
        initX: t.x,
        initY: t.y,
        offset: 42,
        centerX: s.value.width / 2,
        centerY: s.value.height / 2
      }), c = r.x;
      r.y;
      const C = $t(l), P = `
        <tspan
            class="vue-data-ui-datalabel-value"
            fill="${e.value.style.chart.layout.labels.percentage.color}"
            font-size="${ae.value}px"
            style="font-weight:${e.value.style.chart.layout.labels.percentage.bold ? "bold" : "normal"}"
        >${C}</tspan>
    `, N = String(l.name ?? "").split(/\n/g);
      let R = "";
      return N.forEach((q, Be) => {
        Be === 0 ? R += `
                <tspan
                    class="vue-data-ui-datalabel-name"
                    fill="${e.value.style.chart.layout.labels.name.color}"
                    font-size="${E.value}px"
                    style="font-weight:${e.value.style.chart.layout.labels.name.bold ? "bold" : "normal"}"
                >${q}</tspan>
            ` : R += `
                <tspan
                    class="vue-data-ui-datalabel-name"
                    x="${c}"
                    dy="${E.value * 1.2}"
                    fill="${e.value.style.chart.layout.labels.name.color}"
                    font-size="${E.value}px"
                    style="font-weight:${e.value.style.chart.layout.labels.name.bold ? "bold" : "normal"}"
                >${q}</tspan>
            `;
      }), a === "end" ? `${R}${P}` : `${P}${R}`;
    }
    function De(l) {
      return e.value.useBlurOnHover && ![null, void 0].includes(D.value) && D.value !== l ? `url(#blur_${y.value})` : "";
    }
    function Il(l) {
      if (!ne.value || Ye.value === null || l.seriesIndex !== Ye.value)
        return 1;
      const o = (l.proportion ?? 0) * 100, t = e.value.style.chart.layout.labels.dataLabels.hideUnderValue, r = e.value.style.chart.layout.labels.dataLabels.smallArcClusterThreshold + 2, c = t;
      return o >= r ? 1 : o <= c ? 0 : (o - c) / (r - c);
    }
    const Ae = g(() => {
      const l = me.value.map((t) => ({
        name: t.name,
        color: t.color
      })), o = me.value.map((t) => t.value);
      return { head: l, body: o };
    });
    function Nt(l = null) {
      cl(() => {
        const o = Ae.value.head.map((r, c) => [[
          r.name
        ], [Ae.value.body[c]], [isNaN(Ae.value.body[c] / K.value) ? "-" : Ae.value.body[c] / K.value * 100]]), t = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[""], ["val"], ["%"]]].concat(o), a = ua(t);
        l ? l(a) : ra({ csvContent: a, title: e.value.style.chart.title.text || "vue-ui-donut" });
      });
    }
    const xt = g(() => {
      const l = [
        ' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>',
        ue({ p: e.value.style.chart.layout.labels.dataLabels.prefix, v: K.value, s: e.value.style.chart.layout.labels.dataLabels.suffix, r: e.value.table.td.roundingValue }),
        "100%"
      ], o = Ae.value.head.map((r, c) => [
        {
          color: r.color,
          name: r.name || "-"
        },
        Ae.value.body[c],
        isNaN(Ae.value.body[c] / K.value) ? "-" : (Ae.value.body[c] / K.value * 100).toFixed(e.value.table.td.roundingPercentage) + "%"
      ]), t = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.series,
          e.value.table.columnNames.value,
          e.value.table.columnNames.percentage
        ],
        head: l,
        body: o,
        config: t
      };
    }), Xe = k(!1);
    function Zt(l) {
      Xe.value = l, xe.value += 1;
    }
    const el = g(() => /^((?!chrome|android).)*safari/i.test(navigator.userAgent));
    function Fl(l) {
      return va({
        num: l,
        rounding: e.value.style.chart.legend.roundingPercentage
      });
    }
    function Rt({
      val: l,
      percentage: o,
      showVal: t,
      showPercentage: a,
      config: r
    }) {
      return da({
        config: r,
        val: l,
        percentage: o,
        showVal: t,
        showPercentage: a
      });
    }
    function $t(l) {
      return Rt({
        val: de(
          e.value.style.chart.layout.labels.value.formatter,
          l.value,
          ue({
            p: e.value.style.chart.layout.labels.dataLabels.prefix,
            v: l.value,
            s: e.value.style.chart.layout.labels.dataLabels.suffix,
            r: e.value.style.chart.layout.labels.value.rounding
          }),
          { datapoint: l }
        ),
        percentage: Kt(l, se.value),
        showVal: e.value.style.chart.layout.labels.value.show,
        showPercentage: !0,
        config: e.value.style.chart.layout.labels.dataLabels
      });
    }
    function ve(l, o) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: l.seriesIndex }), Oe("selectDatapoint", { datapoint: l, index: o });
    }
    function tl() {
      b.value.showTable = !b.value.showTable;
    }
    function ll() {
      b.value.dataLabels.show = !b.value.dataLabels.show;
    }
    function al() {
      b.value.showTooltip = !b.value.showTooltip;
    }
    const Ct = k(!1);
    function Vt() {
      Ct.value = !Ct.value;
    }
    async function zl({ scale: l = 2 } = {}) {
      if (!L.value) return;
      const { width: o, height: t } = L.value.getBoundingClientRect(), a = o / t, { imageUri: r, base64: c } = await $a({ domElement: L.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: r,
        base64: c,
        title: e.value.style.chart.title.text,
        width: o,
        height: t,
        aspectRatio: a
      };
    }
    function Ol() {
      if (!Se.value) return;
      const { x: l, y: o, width: t, height: a } = Se.value.getBBox();
      X.value && X.value.setAttribute("viewBox", `${l} ${o} ${t + Math.min(0, l)} ${a + Math.min(0, o)}`);
    }
    const at = g(() => {
      const l = e.value.table.useDialog && !e.value.table.show, o = b.value.showTable;
      return {
        component: l ? zt : Lt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: l ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: Xe.value,
          fullscreenParent: L.value,
          forcedWidth: Math.min(500, window.innerWidth * 0.8),
          isCursorPointer: j.value
        } : {
          hideDetails: !0,
          config: {
            open: o,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    Ge(() => b.value.showTable, (l) => {
      e.value.table.show || (l && e.value.table.useDialog && Le.value ? Le.value.open() : "close" in Le.value && Le.value.close());
    });
    function ol() {
      b.value.showTable = !1, Je.value && Je.value.setTableIconState(!1);
    }
    const Yl = g(() => tt.value.map((l) => ({
      ...l,
      name: l.display
    }))), Dl = g(() => e.value.style.chart.backgroundColor), Xl = g(() => e.value.style.chart.legend), Bl = g(() => e.value.style.chart.title), { exportSvg: Nl, getSvg: Rl } = ba({
      svg: X,
      title: Bl,
      legend: Xl,
      legendItems: Yl,
      backgroundColor: Dl
    });
    async function nl({ isCb: l }) {
      Ve.value = !0, await cl();
      try {
        if (l) {
          const { blob: o, url: t, text: a, dataUrl: r } = await Rl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: o, url: t, text: a, dataUrl: r }));
        } else
          await Promise.resolve(Nl());
      } finally {
        Ve.value = !1;
      }
    }
    function Vl(l) {
      if (l?.stage === "start") {
        Re.value = !0;
        return;
      }
      if (l?.stage === "end") {
        Re.value = !1;
        return;
      }
      et();
    }
    async function sl() {
      if (Oe("copyAlt", {
        config: e.value,
        dataset: O.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: O.value
      }));
    }
    return nt({
      autoSize: Ol,
      getData: wt,
      getImage: zl,
      generatePdf: mt,
      generateCsv: Nt,
      generateImage: et,
      generateSvg: nl,
      hideSeries: wl,
      showSeries: bl,
      toggleTable: tl,
      toggleLabels: ll,
      toggleTooltip: al,
      toggleAnnotator: Vt,
      toggleFullscreen: Zt,
      copyAlt: sl
    }), (l, o) => (u(), i("div", {
      ref_key: "donutChart",
      ref: L,
      class: Et(`vue-data-ui-component vue-ui-donut ${Xe.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: S(`font-family:${e.value.style.fontFamily};width:100%; ${e.value.responsive ? "height:100%;" : ""} text-align:center;background:${e.value.style.chart.backgroundColor}`),
      id: `donut__${y.value}`,
      onMouseenter: oe,
      onMouseleave: ze
    }, [
      e.value.userOptions.buttons.annotator && n(X) ? (u(), be(n(Mt), {
        key: 0,
        color: e.value.style.chart.color,
        backgroundColor: e.value.style.chart.backgroundColor,
        active: Ct.value,
        svgRef: n(X),
        isCursorPointer: j.value,
        onClose: Vt
      }, {
        "annotator-action-close": x(() => [
          m(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": x(({ color: t }) => [
          m(l.$slots, "annotator-action-color", J(Z({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": x(({ mode: t }) => [
          m(l.$slots, "annotator-action-draw", J(Z({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": x(({ disabled: t }) => [
          m(l.$slots, "annotator-action-undo", J(Z({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": x(({ disabled: t }) => [
          m(l.$slots, "annotator-action-redo", J(Z({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": x(({ disabled: t }) => [
          m(l.$slots, "annotator-action-delete", J(Z({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["color", "backgroundColor", "active", "svgRef", "isCursorPointer"])) : d("", !0),
      m(l.$slots, "userConfig", {}, void 0, !0),
      bt.value ? (u(), i("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Ke,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : d("", !0),
      e.value.style.chart.title.text ? (u(), i("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: it,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (u(), be(_a, {
          key: `title_${vt.value}`,
          config: {
            title: {
              cy: "donut-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "donut-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : d("", !0),
      w("div", {
        id: `legend-top-${y.value}`
      }, null, 8, Ia),
      e.value.userOptions.show && z.value && (n(Fe) || n(Ee)) ? (u(), be(n(It), {
        ref_key: "userOptionsRef",
        ref: Je,
        key: `user_option_${xe.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: n(Ue),
        isImaging: n(He),
        uid: y.value,
        hasTooltip: e.value.style.chart.tooltip.show && e.value.userOptions.buttons.tooltip,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: Xe.value,
        chartElement: L.value,
        position: e.value.userOptions.position,
        callbacks: e.value.userOptions.callbacks,
        isTooltip: b.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Ct.value,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: j.value,
        onToggleFullscreen: Zt,
        onGeneratePdf: n(mt),
        onGenerateCsv: Nt,
        onGenerateImage: Vl,
        onGenerateSvg: nl,
        onToggleTable: tl,
        onToggleLabels: ll,
        onToggleTooltip: al,
        onToggleAnnotator: Vt,
        onCopyAlt: sl,
        style: S({ visibility: n(Fe) ? n(Ee) ? "visible" : "hidden" : "visible" })
      }, Ut({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: x(({ isOpen: t, color: a }) => [
            m(l.$slots, "menuIcon", J(Z({ isOpen: t, color: a })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: x(() => [
            m(l.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: x(() => [
            m(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionCsv ? {
          name: "optionCsv",
          fn: x(() => [
            m(l.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: x(() => [
            m(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: x(() => [
            m(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionTable ? {
          name: "optionTable",
          fn: x(() => [
            m(l.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionLabels ? {
          name: "optionLabels",
          fn: x(() => [
            m(l.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: x(({ toggleFullscreen: t, isFullscreen: a }) => [
            m(l.$slots, "optionFullscreen", J(Z({ toggleFullscreen: t, isFullscreen: a })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: x(({ toggleAnnotator: t, isAnnotator: a }) => [
            m(l.$slots, "optionAnnotator", J(Z({ toggleAnnotator: t, isAnnotator: a })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: x(({ altCopy: t }) => [
            m(l.$slots, "optionAltCopy", J(Z({ altCopy: t })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "callbacks", "isTooltip", "titles", "hasAnnotator", "isAnnotation", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : d("", !0),
      (u(), i("svg", {
        ref_key: "svgRef",
        ref: X,
        xmlns: n(la),
        class: Et({ "vue-data-ui-fullscreen--on": Xe.value, "vue-data-ui-fulscreen--off": !Xe.value, "vue-data-ui-svg": !0 }),
        viewBox: `0 0 ${s.value.width <= 0 ? 10 : s.value.width} ${s.value.height <= 0 ? 10 : s.value.height}`,
        style: S(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color};${ge.value.css}`)
      }, [
        w("g", {
          ref_key: "G",
          ref: Se,
          class: "vue-data-ui-g"
        }, [
          _t(n(Ft)),
          l.$slots["chart-background"] ? (u(), i("foreignObject", {
            key: 0,
            x: 0,
            y: 0,
            width: s.value.width <= 0 ? 10 : s.value.width,
            height: s.value.height <= 0 ? 10 : s.value.height,
            style: {
              pointerEvents: "none"
            }
          }, [
            m(l.$slots, "chart-background", {}, void 0, !0)
          ], 8, za)) : d("", !0),
          e.value.type === "classic" && !isNaN(pe.value / M.value) ? (u(), i("defs", Oa, [
            e.value.style.chart.useGradient ? (u(), i("radialGradient", {
              key: 0,
              id: `gradient_${y.value}`
            }, [
              w("stop", {
                offset: "0%",
                "stop-color": n(Pt)(e.value.style.chart.backgroundColor, 0),
                "stop-opacity": "0"
              }, null, 8, Da),
              w("stop", {
                offset: `${(1 - pe.value / M.value) * 100}%`,
                "stop-color": n(Pt)("#FFFFFF", 0),
                "stop-opacity": "0"
              }, null, 8, Xa),
              w("stop", {
                offset: `${(1 - pe.value / M.value / 2) * 100}%`,
                "stop-color": n(Pt)("#FFFFFF", e.value.style.chart.gradientIntensity)
              }, null, 8, Ba),
              w("stop", {
                offset: "100%",
                "stop-color": n(Pt)(e.value.style.chart.backgroundColor, 0),
                "stop-opacity": "0"
              }, null, 8, Na)
            ], 8, Ya)) : d("", !0)
          ])) : d("", !0),
          e.value.type === "polar" ? (u(), i("defs", Ra, [
            (u(!0), i(T, null, W(v.value, (t, a) => (u(), i("radialGradient", {
              id: `polar_gradient_${a}_${y.value}`,
              cx: (isNaN(t.middlePoint.x / s.value.width * 100) ? 0 : t.middlePoint.x / s.value.width * 100) + "%",
              cy: (isNaN(t.middlePoint.y / s.value.height * 100) ? 0 : t.middlePoint.y / s.value.height * 100) + "%",
              r: "62%"
            }, [
              w("stop", {
                offset: "0%",
                "stop-color": n(aa)(G.value[a].color, 0.05),
                "stop-opacity": e.value.style.chart.gradientIntensity / 100
              }, null, 8, Ea),
              w("stop", {
                offset: "100%",
                "stop-color": G.value[a].color
              }, null, 8, Ua)
            ], 8, Va))), 256))
          ])) : d("", !0),
          w("defs", null, [
            w("filter", {
              id: `blur_${y.value}`,
              x: "-50%",
              y: "-50%",
              width: "200%",
              height: "200%"
            }, [
              w("feGaussianBlur", {
                in: "SourceGraphic",
                stdDeviation: 2,
                id: `blur_std_${y.value}`
              }, null, 8, Ga),
              o[5] || (o[5] = w("feColorMatrix", {
                type: "saturate",
                values: "0"
              }, null, -1))
            ], 8, Ha),
            w("filter", {
              id: `shadow_${y.value}`,
              "color-interpolation-filters": "sRGB"
            }, [
              w("feDropShadow", {
                dx: "0",
                dy: "0",
                stdDeviation: "10",
                "flood-opacity": "0.5",
                "flood-color": e.value.style.chart.layout.donut.shadowColor
              }, null, 8, Wa)
            ], 8, qa),
            w("filter", {
              id: `drop_shadow_${y.value}`,
              "color-interpolation-filters": "sRGB",
              x: "-50%",
              y: "-50%",
              width: "200%",
              height: "200%"
            }, [
              w("feDropShadow", {
                dx: "0",
                dy: "0",
                stdDeviation: "3",
                "flood-opacity": "1",
                "flood-color": e.value.style.chart.layout.donut.shadowColor
              }, null, 8, Qa)
            ], 8, ja)
          ]),
          e.value.type === "classic" ? (u(!0), i(T, { key: 3 }, W(G.value.filter((t) => !t.ghost), (t, a) => (u(), i("g", null, [
            Q(t) && b.value.dataLabels.show ? (u(), i("path", {
              key: 0,
              d: n(B)[a]?.connectorPath || n(oa)(
                t,
                { x: s.value.width / 2, y: s.value.height / 2 },
                16,
                16,
                !1,
                !1,
                pe.value,
                12,
                e.value.style.chart.layout.curvedMarkers
              ),
              stroke: t.color,
              "stroke-width": "1",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              fill: "none",
              filter: De(a)
            }, null, 8, Ka)) : d("", !0)
          ]))), 256)) : d("", !0),
          e.value.type === "polar" ? (u(!0), i(T, { key: 4 }, W(G.value.filter((t) => !t.ghost), (t, a) => (u(), i("g", null, [
            Q(t) && b.value.dataLabels.show ? (u(), i("path", {
              key: 0,
              d: `M ${n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 24, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x},${n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 24, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y} ${v.value[a].middlePoint.x},${v.value[a].middlePoint.y}`,
              stroke: t.color,
              "stroke-width": "1",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              fill: "none",
              filter: De(a),
              style: S({
                transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`
              })
            }, null, 12, Ja)) : d("", !0)
          ]))), 256)) : d("", !0),
          e.value.type === "classic" && e.value.style.chart.layout.donut.useShadow ? (u(), i("circle", {
            key: 5,
            cx: s.value.width / 2,
            cy: s.value.height / 2,
            r: M.value <= 0 ? 10 : M.value,
            fill: e.value.style.chart.backgroundColor,
            filter: `url(#shadow_${y.value})`
          }, null, 8, Za)) : d("", !0),
          l.$slots.pattern ? (u(), i("g", eo, [
            (u(!0), i(T, null, W(ot.dataset, (t, a) => (u(), i("defs", {
              key: `pattern-${t.patternIndex}`
            }, [
              m(l.$slots, "pattern", Ht({ ref_for: !0 }, { seriesIndex: a, patternId: `pattern_${y.value}_${a}` }), void 0, !0)
            ]))), 128))
          ])) : d("", !0),
          K.value && e.value.type === "classic" ? (u(), i(T, { key: 7 }, [
            (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
              stroke: e.value.style.chart.backgroundColor,
              d: t.arcSlice,
              fill: "#FFFFFF"
            }, null, 8, to))), 256)),
            (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
              class: "vue-ui-donut-arc-path",
              d: t.arcSlice,
              fill: t.color,
              stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
              "stroke-width": e.value.style.chart.layout.donut.borderWidth,
              filter: De(a)
            }, null, 8, lo))), 256)),
            l.$slots.pattern ? (u(), i("g", ao, [
              (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
                class: "vue-ui-donut-arc-path",
                d: t.arcSlice,
                fill: `url(#${t.pattern})`,
                stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
                "stroke-width": e.value.style.chart.layout.donut.borderWidth,
                filter: De(a)
              }, null, 8, oo))), 256))
            ])) : d("", !0)
          ], 64)) : d("", !0),
          K.value && e.value.type === "polar" ? (u(), i(T, { key: 8 }, [
            G.value.length > 1 ? (u(), i("g", no, [
              (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
                stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
                d: v.value[a].path,
                fill: "#FFFFFF",
                style: S({
                  transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`
                })
              }, null, 12, so))), 256)),
              e.value.style.chart.layout.donut.useShadow ? (u(), i("g", uo, [
                (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
                  class: "vue-ui-donut-arc-path",
                  d: v.value[a].path,
                  fill: "transparent",
                  stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
                  "stroke-width": e.value.style.chart.layout.donut.borderWidth,
                  filter: `url(#drop_shadow_${y.value})`,
                  style: S({
                    transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`
                  })
                }, null, 12, ro))), 256))
              ])) : d("", !0),
              l.$slots.pattern ? (u(), i("g", io, [
                (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
                  class: "vue-ui-donut-arc-path",
                  d: v.value[a].path,
                  fill: `url(#${t.pattern})`,
                  stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
                  "stroke-width": e.value.style.chart.layout.donut.borderWidth,
                  filter: De(a),
                  style: S({
                    transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`,
                    transformOrigin: "center"
                  })
                }, null, 12, co))), 256))
              ])) : d("", !0),
              (u(!0), i(T, null, W(se.value, (t, a) => (u(), i("path", {
                class: "vue-ui-donut-arc-path",
                d: v.value[a].path,
                fill: e.value.style.chart.useGradient ? `url(#polar_gradient_${a}_${y.value})` : t.color,
                stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
                "stroke-width": e.value.style.chart.layout.donut.borderWidth,
                filter: De(a),
                style: S({
                  transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`
                })
              }, null, 12, vo))), 256))
            ])) : (u(), i("g", ho, [
              l.$slots.pattern ? (u(), i("circle", {
                key: 0,
                cx: s.value.width / 2,
                cy: s.value.height / 2,
                r: M.value,
                fill: `url(#pattern_${y.value}_${G.value[0].patternIndex})`,
                stroke: e.value.style.chart.backgroundColor,
                "stroke-width": e.value.style.chart.layout.donut.borderWidth
              }, null, 8, fo)) : d("", !0),
              w("circle", {
                cx: s.value.width / 2,
                cy: s.value.height / 2,
                r: M.value,
                fill: e.value.style.chart.useGradient ? `url(#polar_gradient_0_${y.value})` : G.value[0].color,
                stroke: e.value.style.chart.backgroundColor,
                "stroke-width": e.value.style.chart.layout.donut.borderWidth
              }, null, 8, yo)
            ]))
          ], 64)) : (u(), i(T, { key: 9 }, [
            e.value.type === "classic" && !$l.value ? (u(), i("g", go, [
              (u(!0), i(T, null, W(Cl.value, (t, a) => (u(), i("path", {
                class: "vue-ui-donut-arc-path",
                d: t.arcSlice,
                fill: t.color,
                stroke: e.value.style.chart.backgroundColor,
                "stroke-width": e.value.style.chart.layout.donut.borderWidth
              }, null, 8, po))), 256))
            ])) : d("", !0),
            w("circle", {
              cx: s.value.width / 2,
              cy: s.value.height / 2,
              r: M.value <= 0 ? 10 : M.value,
              fill: "transparent",
              stroke: e.value.style.chart.backgroundColor
            }, null, 8, mo)
          ], 64)),
          e.value.style.chart.useGradient && e.value.type === "classic" ? (u(), i("circle", {
            key: 10,
            cx: s.value.width / 2,
            cy: s.value.height / 2,
            r: (
              /* This might require adjustments */
              M.value <= 0 ? 10 : M.value
            ),
            fill: `url(#gradient_${y.value})`
          }, null, 8, bo)) : d("", !0),
          w("circle", {
            ref_key: "circle_hollow",
            ref: je,
            style: { pointerEvents: "none" },
            fill: "none",
            cx: s.value.width / 2,
            cy: s.value.height / 2,
            r: Math.max(0.1, pe.value * 1.7)
          }, null, 8, wo),
          K.value ? (u(), i(T, { key: 11 }, [
            G.value.length > 1 || e.value.type === "classic" ? (u(), i("g", ko, [
              (u(!0), i(T, null, W(G.value.filter((t) => !t.ghost), (t, a) => (u(), i("path", {
                d: e.value.type === "classic" ? t.arcSlice : v.value[a].path,
                stroke: e.value.style.chart.layout.donut.borderColorAuto ? e.value.style.chart.backgroundColor : e.value.style.chart.layout.donut.borderColor,
                fill: D.value === a ? e.value.style.chart.layout.donut.selectedColor : "transparent",
                onMouseenter: (r) => ce({
                  datapoint: t,
                  relativeIndex: a,
                  seriesIndex: t.seriesIndex,
                  show: !0
                }),
                onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex }),
                onClick: (r) => ve(t, a)
              }, null, 40, xo))), 256))
            ])) : (u(), i("g", $o, [
              w("circle", {
                cx: s.value.width / 2,
                cy: s.value.height / 2,
                r: M.value,
                fill: "transparent",
                onMouseenter: o[0] || (o[0] = (t) => ce({
                  datapoint: G.value[0],
                  relativeIndex: 0,
                  seriesIndex: G.value[0].seriesIndex,
                  show: !0
                })),
                onMouseleave: o[1] || (o[1] = (t) => ie({ datapoint: G.value[0], seriesIndex: G.value[0].seriesIndex })),
                onClick: o[2] || (o[2] = (t) => ve(G.value[0], l.i))
              }, null, 40, Co)
            ]))
          ], 64)) : d("", !0),
          e.value.type === "classic" ? (u(), i("g", {
            key: 12,
            ref_key: "G_hollow",
            ref: rt,
            class: "vue-data-ui-donut-hollow-labels"
          }, [
            e.value.style.chart.layout.labels.hollow.total.show ? (u(), i("text", {
              key: 0,
              "text-anchor": "middle",
              x: s.value.width / 2,
              y: s.value.height / 2 - (e.value.style.chart.layout.labels.hollow.average.show ? e.value.style.chart.layout.labels.hollow.total.fontSize : 0) + e.value.style.chart.layout.labels.hollow.total.offsetY,
              fill: e.value.style.chart.layout.labels.hollow.total.color,
              "font-size": e.value.style.chart.layout.labels.hollow.total.fontSize,
              style: S(`font-weight:${e.value.style.chart.layout.labels.hollow.total.bold ? "bold" : ""}`)
            }, we(e.value.style.chart.layout.labels.hollow.total.text), 13, _o)) : d("", !0),
            e.value.style.chart.layout.labels.hollow.total.show ? (u(), i("text", {
              key: 1,
              "text-anchor": "middle",
              x: s.value.width / 2,
              y: s.value.height / 2 + e.value.style.chart.layout.labels.hollow.total.fontSize - (e.value.style.chart.layout.labels.hollow.average.show ? e.value.style.chart.layout.labels.hollow.total.fontSize : 0) + e.value.style.chart.layout.labels.hollow.total.value.offsetY,
              fill: e.value.style.chart.layout.labels.hollow.total.value.color,
              "font-size": e.value.style.chart.layout.labels.hollow.total.value.fontSize,
              style: S(`font-weight:${e.value.style.chart.layout.labels.hollow.total.value.bold ? "bold" : ""}`)
            }, we(n(de)(
              e.value.style.chart.layout.labels.hollow.total.value.formatter,
              K.value,
              n(ue)({
                p: e.value.style.chart.layout.labels.hollow.total.value.prefix,
                v: K.value,
                s: e.value.style.chart.layout.labels.hollow.total.value.suffix
              })
            )), 13, Ao)) : d("", !0),
            e.value.style.chart.layout.labels.hollow.average.show ? (u(), i("text", {
              key: 2,
              "text-anchor": "middle",
              x: s.value.width / 2,
              y: s.value.height / 2 + (e.value.style.chart.layout.labels.hollow.total.show ? e.value.style.chart.layout.labels.hollow.average.fontSize : 0) + e.value.style.chart.layout.labels.hollow.average.offsetY,
              fill: e.value.style.chart.layout.labels.hollow.average.color,
              "font-size": e.value.style.chart.layout.labels.hollow.average.fontSize,
              style: S(`font-weight:${e.value.style.chart.layout.labels.hollow.average.bold ? "bold" : ""}`)
            }, we(e.value.style.chart.layout.labels.hollow.average.text), 13, To)) : d("", !0),
            e.value.style.chart.layout.labels.hollow.average.show ? (u(), i("text", {
              key: 3,
              "text-anchor": "middle",
              x: s.value.width / 2,
              y: s.value.height / 2 + (e.value.style.chart.layout.labels.hollow.total.show ? e.value.style.chart.layout.labels.hollow.average.fontSize : 0) + e.value.style.chart.layout.labels.hollow.average.fontSize + e.value.style.chart.layout.labels.hollow.average.value.offsetY,
              fill: e.value.style.chart.layout.labels.hollow.average.value.color,
              "font-size": e.value.style.chart.layout.labels.hollow.average.value.fontSize,
              style: S(`font-weight:${e.value.style.chart.layout.labels.hollow.average.value.bold ? "bold" : ""}`)
            }, we(ne.value || A.value ? "--" : n(de)(
              e.value.style.chart.layout.labels.hollow.average.value.formatter,
              n(yl)(Xt.value),
              n(ue)({
                p: e.value.style.chart.layout.labels.hollow.average.value.prefix,
                v: n(yl)(Xt.value),
                s: e.value.style.chart.layout.labels.hollow.average.value.suffix,
                r: e.value.style.chart.layout.labels.hollow.average.value.rounding
              })
            )), 13, Po)) : d("", !0)
          ], 512)) : d("", !0),
          (u(!0), i(T, null, W(se.value.filter((t) => !t.ghost), (t, a) => (u(), i("g", {
            filter: De(a),
            class: Et({ animated: e.value.useCssAnimation }),
            key: t.seriesIndex,
            opacity: Il(t)
          }, [
            e.value.style.chart.layout.labels.dataLabels.useLabelSlots ? (u(), i("g", Lo, [
              (u(), i("foreignObject", {
                x: n(_)(t, !0).anchor === "end" ? n(_)(t).x - 120 : n(_)(t, !0).anchor === "middle" ? n(_)(t).x - 60 : n(_)(t).x,
                y: n(ke)(t) - (el.value ? 20 : 0),
                width: "120",
                height: "60",
                style: { overflow: "visible" }
              }, [
                w("div", null, [
                  m(l.$slots, "dataLabel", Ht({ ref_for: !0 }, {
                    datapoint: t,
                    isBlur: !e.value.useBlurOnHover || [null, void 0].includes(D.value) || D.value === a,
                    isSafari: el.value,
                    isVisible: Q(t) && b.value.dataLabels.show,
                    textAlign: n(_)(t, !0, 16, !0).anchor,
                    flexAlign: n(_)(t, !0, 16).anchor,
                    percentage: Kt(t, se.value)
                  }), void 0, !0)
                ])
              ], 8, Mo))
            ])) : (u(), i("g", Io, [
              e.value.type === "classic" ? (u(), i(T, { key: 0 }, [
                Q(t) && b.value.dataLabels.show ? (u(), i("circle", {
                  key: 0,
                  cx: n(B)[a]?.markerX ?? n(_)(t).x,
                  cy: n(B)[a]?.markerY ?? n(ke)(t) - 3.5,
                  fill: t.color,
                  stroke: e.value.style.chart.backgroundColor,
                  "stroke-width": 1,
                  r: 3,
                  filter: !e.value.useBlurOnHover || [null, void 0].includes(D.value) || D.value === a ? "" : `url(#blur_${y.value})`,
                  onClick: (r) => ve(t, a),
                  onMouseenter: (r) => ce({
                    datapoint: t,
                    relativeIndex: a,
                    seriesIndex: t.seriesIndex,
                    show: !0
                  }),
                  onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex })
                }, null, 40, Fo)) : d("", !0),
                e.value.style.chart.layout.labels.dataLabels.oneLine || n(B)[a] ? At((u(), i("text", {
                  key: 1,
                  class: "vue-data-ui-datalabel-inline",
                  "text-anchor": n(B)[a]?.textAnchor || n(_)(t, !0, 12).anchor,
                  x: n(B)[a]?.labelX ?? n(_)(t, !0, 12).x,
                  y: n(B)[a]?.labelY ?? n(ke)(t),
                  onClick: (r) => ve(t, a),
                  onMouseenter: (r) => ce({
                    datapoint: t,
                    relativeIndex: a,
                    seriesIndex: t.seriesIndex,
                    show: !0
                  }),
                  onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex }),
                  innerHTML: Ll(t, a, !!n(B)[a])
                }, null, 40, zo)), [
                  [Tt, Q(t) && b.value.dataLabels.show]
                ]) : (u(), i(T, { key: 2 }, [
                  At(w("text", {
                    class: "vue-data-ui-datalabel-value",
                    "text-anchor": n(B)[a]?.textAnchor || n(_)(t, !0, 12).anchor,
                    x: n(B)[a]?.labelX ?? n(_)(t, !0, 12).x,
                    y: n(B)[a]?.labelY ?? n(ke)(t),
                    fill: e.value.style.chart.layout.labels.percentage.color,
                    "font-size": ae.value + "px",
                    style: S(`font-weight:${e.value.style.chart.layout.labels.percentage.bold ? "bold" : ""}`),
                    onClick: (r) => ve(t, a),
                    onMouseenter: (r) => ce({
                      datapoint: t,
                      relativeIndex: a,
                      seriesIndex: t.seriesIndex,
                      show: !0
                    }),
                    onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex })
                  }, we($t(t)), 45, Oo), [
                    [Tt, Q(t) && b.value.dataLabels.show]
                  ]),
                  At(w("text", {
                    class: "vue-data-ui-datalabel-name",
                    "text-anchor": n(B)[a]?.textAnchor || n(_)(t).anchor,
                    x: n(B)[a]?.labelX ?? n(_)(t, !0, 12).x,
                    y: (n(B)[a]?.labelY ?? n(ke)(t)) + E.value * 1.2,
                    fill: e.value.style.chart.layout.labels.name.color,
                    "font-size": E.value + "px",
                    style: S(`font-weight:${e.value.style.chart.layout.labels.name.bold ? "bold" : ""}`),
                    onClick: (r) => ve(t, a),
                    onMouseenter: (r) => ce({
                      datapoint: t,
                      relativeIndex: a,
                      seriesIndex: t.seriesIndex,
                      show: !0
                    }),
                    onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex }),
                    innerHTML: n(gl)({
                      content: t.name,
                      fontSize: E.value,
                      fill: e.value.style.chart.layout.labels.name.color,
                      x: n(B)[a]?.labelX ?? n(_)(t, !0, 12).x,
                      y: (n(B)[a]?.labelY ?? n(ke)(t)) + E.value
                    })
                  }, null, 44, Yo), [
                    [Tt, Q(t) && b.value.dataLabels.show]
                  ])
                ], 64))
              ], 64)) : d("", !0),
              e.value.type === "polar" ? (u(), i(T, { key: 1 }, [
                Q(t) && b.value.dataLabels.show ? (u(), i("circle", {
                  key: 0,
                  cx: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 24, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x,
                  cy: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 24, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y,
                  fill: t.color,
                  stroke: e.value.style.chart.backgroundColor,
                  "stroke-width": 1,
                  r: 3,
                  filter: !e.value.useBlurOnHover || [null, void 0].includes(D.value) || D.value === a ? "" : `url(#blur_${y.value})`,
                  onClick: (r) => ve(t, a),
                  onMouseenter: (r) => ce({
                    datapoint: t,
                    relativeIndex: a,
                    seriesIndex: t.seriesIndex,
                    show: !0
                  }),
                  onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex }),
                  style: S({
                    transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`
                  })
                }, null, 44, Do)) : d("", !0),
                e.value.style.chart.layout.labels.dataLabels.oneLine ? At((u(), i("text", {
                  key: 1,
                  class: "vue-data-ui-datalabel-inline",
                  "text-anchor": _e(v.value[a].middlePoint),
                  x: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x,
                  y: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y,
                  style: S({
                    transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`
                  }),
                  onClick: (r) => ve(t, a),
                  onMouseenter: (r) => ce({
                    datapoint: t,
                    relativeIndex: a,
                    seriesIndex: t.seriesIndex,
                    show: !0
                  }),
                  onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex }),
                  innerHTML: Ml(t, a)
                }, null, 44, Xo)), [
                  [Tt, Q(t) && b.value.dataLabels.show]
                ]) : (u(), i(T, { key: 2 }, [
                  Q(t) && b.value.dataLabels.show ? (u(), i("text", {
                    key: 0,
                    class: "vue-data-ui-datalabel-value",
                    "text-anchor": _e(v.value[a].middlePoint),
                    x: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x,
                    y: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y,
                    fill: e.value.style.chart.layout.labels.percentage.color,
                    "font-size": ae.value,
                    style: S({
                      transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`,
                      fontWeight: e.value.style.chart.layout.labels.percentage.bold ? "bold" : "normal"
                    }),
                    onClick: (r) => ve(t, a),
                    onMouseenter: (r) => ce({
                      datapoint: t,
                      relativeIndex: a,
                      seriesIndex: t.seriesIndex,
                      show: !0
                    }),
                    onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex })
                  }, we($t(t)), 45, Bo)) : d("", !0),
                  Q(t) && b.value.dataLabels.show ? (u(), i("text", {
                    key: 1,
                    class: "vue-data-ui-datalabel-name",
                    "text-anchor": _e(v.value[a].middlePoint),
                    x: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x,
                    y: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y + E.value * 1.2,
                    fill: e.value.style.chart.layout.labels.name.color,
                    "font-size": E.value,
                    style: S({
                      transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`,
                      fontWeight: e.value.style.chart.layout.labels.name.bold ? "bold" : "normal"
                    }),
                    onClick: (r) => ve(t, a),
                    onMouseenter: (r) => ce({
                      datapoint: t,
                      relativeIndex: a,
                      seriesIndex: t.seriesIndex,
                      show: !0
                    }),
                    onMouseleave: (r) => ie({ datapoint: t, seriesIndex: t.seriesIndex }),
                    innerHTML: n(gl)({
                      content: t.name,
                      fontSize: ae.value,
                      fill: e.value.style.chart.layout.labels.name.color,
                      x: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x,
                      y: n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).y + E.value * 1.2
                    })
                  }, null, 44, No)) : d("", !0)
                ], 64))
              ], 64)) : d("", !0)
            ])),
            b.value.dataLabels.show && e.value.style.chart.comments.show && t.comment ? (u(), i("g", Ro, [
              Q(t) && e.value.type === "classic" ? (u(), i("foreignObject", {
                key: 0,
                x: e.value.style.chart.comments.offsetX + (n(_)(t, !0).anchor === "end" ? n(_)(t).x - e.value.style.chart.comments.width : n(_)(t, !0).anchor === "middle" ? n(_)(t).x - e.value.style.chart.comments.width / 2 : n(_)(t).x),
                y: n(ke)(t) + 24 + e.value.style.chart.comments.offsetY,
                width: e.value.style.chart.comments.width,
                height: "200",
                style: { overflow: "visible", "pointer-events": "none" }
              }, [
                w("div", null, [
                  m(l.$slots, "plot-comment", {
                    plot: { ...t, textAlign: n(_)(t, !0, 16, !0).anchor, flexAlign: n(_)(t, !0, 16).anchor, isFirstLoad: A.value }
                  }, void 0, !0)
                ])
              ], 8, Vo)) : d("", !0),
              Q(t) && e.value.type === "polar" ? (u(), i("foreignObject", {
                key: 1,
                x: e.value.style.chart.comments.offsetX + (_e(v.value[a].middlePoint) === "end" ? n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x - e.value.style.chart.comments.width : _e(v.value[a].middlePoint) === "middle" ? n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x - e.value.style.chart.comments.width / 2 : n(Y)({ initX: v.value[a].middlePoint.x, initY: v.value[a].middlePoint.y, offset: 42, centerX: s.value.width / 2, centerY: s.value.height / 2 }).x),
                y: _l(v.value[a]) + e.value.style.chart.comments.offsetY,
                width: e.value.style.chart.comments.width,
                height: "200",
                style: S({
                  transition: A.value || !e.value.serieToggleAnimation.show ? "none" : `all ${e.value.serieToggleAnimation.durationMs}ms ease-in-out`,
                  overflow: "visible",
                  pointerEvents: "none"
                })
              }, [
                w("div", null, [
                  m(l.$slots, "plot-comment", {
                    plot: { ...t, textAlign: _e(v.value[a].middlePoint), flexAlign: _e(v.value[a].middlePoint), isFirstLoad: A.value }
                  }, void 0, !0)
                ])
              ], 12, Eo)) : d("", !0)
            ])) : d("", !0)
          ], 10, So))), 128)),
          m(l.$slots, "svg", {
            svg: {
              ...s.value,
              isPrintingImg: n(Ue) | n(He) | Re.value,
              isPrintingSvg: Ve.value
            }
          }, void 0, !0)
        ], 512)
      ], 14, Fa)),
      l.$slots.watermark ? (u(), i("div", Uo, [
        m(l.$slots, "watermark", J(Z({ isPrinting: n(Ue) || n(He) || Re.value || Ve.value })), void 0, !0)
      ])) : d("", !0),
      w("div", {
        id: `legend-bottom-${y.value}`
      }, null, 8, Ho),
      ft.value ? (u(), be(jl, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${y.value}` : `#legend-bottom-${y.value}`
      }, [
        w("div", {
          ref_key: "chartLegend",
          ref: ct
        }, [
          e.value.style.chart.legend.show ? (u(), be(Ta, {
            key: `legend_${ht.value}`,
            legendSet: tt.value,
            config: xl.value,
            onClickMarker: o[3] || (o[3] = ({ i: t }) => kt(t)),
            isCursorPointer: j.value
          }, Ut({
            item: x(({ legend: t, index: a }) => [
              w("div", {
                style: S(`opacity:${F.value.includes(a) ? 0.5 : 1}`),
                onClick: (r) => t.segregate()
              }, we(t.display), 13, Go)
            ]),
            legendToggle: x(() => [
              tt.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !n(fe) ? (u(), be(Pa, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: F.value.length > 0,
                isCursorPointer: j.value,
                onToggle: ml
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : d("", !0)
            ]),
            _: 2
          }, [
            l.$slots.pattern ? {
              name: "legend-pattern",
              fn: x(({ legend: t, index: a }) => [
                _t(Ca, {
                  shape: t.shape,
                  radius: 30,
                  stroke: "none",
                  plot: { x: 30, y: 30 },
                  fill: `url(#pattern_${y.value}_${a})`
                }, null, 8, ["shape", "fill"])
              ]),
              key: "0"
            } : void 0
          ]), 1032, ["legendSet", "config", "isCursorPointer"])) : d("", !0),
          m(l.$slots, "legend", { legend: tt.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : d("", !0),
      l.$slots.source ? (u(), i("div", {
        key: 6,
        ref_key: "source",
        ref: Qe,
        dir: "auto"
      }, [
        m(l.$slots, "source", {}, void 0, !0)
      ], 512)) : d("", !0),
      l.$slots.hollow ? (u(), i("div", qo, [
        m(l.$slots, "hollow", J(Z({
          total: K.value,
          average: Xt.value,
          dataset: H.value
        })), void 0, !0)
      ])) : d("", !0),
      _t(n(qe), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: b.value.showTooltip && ee.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: L.value,
        content: te.value,
        isCustom: lt.value,
        isFullscreen: Xe.value,
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThrehsold
      }, {
        "tooltip-before": x(() => [
          m(l.$slots, "tooltip-before", J(Z({ ...Bt.value })), void 0, !0)
        ]),
        "tooltip-after": x(() => [
          m(l.$slots, "tooltip-after", J(Z({ ...Bt.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isCustom", "isFullscreen", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      z.value && e.value.userOptions.buttons.table ? (u(), be(Ql(at.value.component), Ht({ key: 8 }, at.value.props, {
        ref_key: "tableUnit",
        ref: Le,
        onClose: ol
      }), Ut({
        content: x(() => [
          (u(), be(n(ut), {
            key: `table_${dt.value}`,
            colNames: xt.value.colNames,
            head: xt.value.head,
            body: xt.value.body,
            config: xt.value.config,
            title: e.value.table.useDialog ? "" : at.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: j.value,
            onClose: ol
          }, {
            th: x(({ th: t }) => [
              w("div", {
                innerHTML: t,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Wo)
            ]),
            td: x(({ td: t }) => [
              il(we(t.name ? t.name : isNaN(Number(t)) ? t.includes("%") ? t : n(de)(
                e.value.style.chart.layout.labels.percentage.formatter,
                t,
                n(ue)({
                  v: t,
                  s: "%",
                  r: e.value.style.chart.layout.labels.percentage.rounding
                })
              ) : n(de)(
                e.value.style.chart.layout.labels.value.formatter,
                t,
                n(ue)({
                  p: e.value.style.chart.layout.labels.dataLabels.prefix,
                  v: t,
                  s: e.value.style.chart.layout.labels.dataLabels.suffix,
                  r: e.value.style.chart.layout.labels.value.rounding
                })
              )), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: x(() => [
            il(we(at.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: x(() => [
            w("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: o[4] || (o[4] = (t) => Nt(e.value.userOptions.callbacks.csv)),
              style: S({ cursor: j.value ? "pointer" : "default" })
            }, [
              _t(n(st), {
                name: "fileCsv",
                stroke: at.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : d("", !0),
      m(l.$slots, "skeleton", {}, () => [
        n(fe) ? (u(), be(ma, { key: 0 })) : d("", !0)
      ], !0)
    ], 46, Ma));
  }
}, gn = /* @__PURE__ */ Sa(jo, [["__scopeId", "data-v-da3fce96"]]);
export {
  gn as default
};
