// vue-data-ui/dist/vue-ui-sparkstackbar-BsIOHUE5.js
import { defineAsyncComponent as Z, useSlots as Ie, onMounted as ee, ref as g, computed as _, toRefs as Pe, watch as te, openBlock as s, createElementBlock as n, normalizeStyle as p, createElementVNode as v, toDisplayString as F, createCommentVNode as f, unref as c, createVNode as le, Fragment as $, renderList as I, normalizeClass as L, createBlock as oe, withCtx as ae, renderSlot as N, normalizeProps as se, guardReactiveProps as ne } from "vue";
import { j as re, t as ue, m as De, p as r, s as A, n as ze, o as Be, q as ie, r as Le, l as R, X as Ne, u as ce, v as de, k as ve, T as Ve } from "./lib-GwPnJSrY.js";
import { u as Ee } from "./useConfig-CIgyKZ6R.js";
import { u as Ue, B as Me } from "./BaseScanner-BhhkKVoj.js";
import { u as G } from "./useNestedProp-BBT0mbHh.js";
import { u as Oe } from "./useThemeCheck-D1ZYXUwG.js";
import { u as je } from "./useChartAccessibility-9icAAmYg.js";
import Re from "./vue_ui_sparkstackbar-BSexvOuM.js";
import { B as Ge } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as We } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Ye = ["xmlns", "viewBox"], qe = ["id"], He = ["stop-color"], Xe = ["stop-color"], Je = ["stop-color"], Ke = {
  id: "stackPill",
  clipPathUnits: "objectBoundingBox"
}, Qe = ["fill"], Ze = {
  key: 0,
  "clip-path": "url(#stackPill)"
}, et = ["x", "width", "height", "fill"], tt = ["x", "width", "height", "fill", "stroke"], lt = ["x", "width", "height", "onClick", "onMouseenter", "onMouseleave"], ot = ["width", "height", "rx"], at = ["onClick"], st = { style: { display: "flex", "flex-direction": "row", "align-items": "center", gap: "4px", "justify-content": "center" } }, nt = ["height", "width"], rt = ["id"], ut = ["stop-color"], it = ["stop-color"], ct = ["fill"], dt = {
  key: 2,
  ref: "source",
  dir: "auto"
}, vt = {
  __name: "vue-ui-sparkstackbar",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectDatapoint"],
  setup(ge, { expose: pe, emit: fe }) {
    const he = Z(() => import("./PackageVersion-DN9d0ME5.js")), ye = Z(() => import("./Tooltip-DL3QgjZf.js")), { vue_ui_sparkstackbar: me } = Ee(), { isThemeValid: ke, warnInvalidTheme: be } = Oe(), m = ge, _e = Ie();
    ee(() => {
      _e["chart-background"] && console.warn("VueUiSparkStackbar does not support the #chart-background slot.");
    });
    const W = g(null), P = g(re()), V = g(!1), E = g(""), e = g(U()), Y = _(() => e.value.useCursorPointer), Ce = _(() => ue({
      defaultConfig: {
        style: {
          backgroundColor: "#99999930",
          animation: { show: !1 },
          bar: { gradient: { inderlayerColor: "#6A6A6A" } },
          title: {
            backgroundColor: "transparent"
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: C, FINAL_DATASET: h } = Ue({
      ...Pe(m),
      FINAL_CONFIG: e,
      prepareConfig: U,
      skeletonDataset: m.config?.skeletonDataset ?? [
        { name: "_", value: 8, color: "#808080" },
        { name: "_", value: 5, color: "#ADADAD" },
        { name: "_", value: 3, color: "#DBDBDB" }
      ],
      skeletonConfig: ue({
        defaultConfig: e.value,
        userConfig: Ce.value
      })
    }), { svgRef: xe } = je({ config: e.value.style.title });
    function U() {
      const t = G({
        userConfig: m.config,
        defaultConfig: me
      }), l = t.theme;
      if (!l) return t;
      if (!ke.value(t))
        return be(t), t;
      const o = G({
        userConfig: Re[l] || m.config,
        defaultConfig: t
      }), a = G({
        userConfig: m.config,
        defaultConfig: o
      });
      return {
        ...a,
        customPalette: a.customPalette.length ? a.customPalette : De[l] || r
      };
    }
    te(() => m.config, (t) => {
      e.value = U(), X();
    }, { deep: !0 }), te(() => h.value, (t) => {
      x.value = h.value.map((l, o) => ({
        ...l,
        color: l.color ? A(l.color) : S.value[o] || r[o] || r[o % r.length]
      })), q();
    }, { deep: !0 });
    const S = _(() => ze(e.value.customPalette)), x = g(h.value.map((t, l) => ({
      ...t,
      value: e.value.style.animation.show ? 0 : t.value || 0,
      color: t.color ? A(t.color) : S.value[l] || r[l] || r[l % r.length]
    }))), w = g(!0);
    function q() {
      if (!e.value.style.animation.show) {
        w.value = !1;
        return;
      }
      const t = e.value.style.animation.animationFrames, l = h.value.map((u) => u.value || 0), o = l.map((u) => u / t), a = l.reduce((u, i) => u + i, 0);
      let b = 0;
      w.value = !0, x.value = h.value.map((u, i) => ({
        ...u,
        value: 0,
        color: u.color ? A(u.color) : S.value[i] || r[i] || r[i % r.length]
      }));
      function B() {
        b += a / t, b < a ? (x.value = x.value.map((u, i) => ({
          ...u,
          value: Math.min(u.value + o[i], l[i]),
          color: u.color ? A(u.color) : S.value[i] || r[i] || r[i % r.length]
        })), requestAnimationFrame(B)) : (w.value = !1, x.value = h.value.map((u, i) => ({
          ...u,
          value: l[i],
          color: u.color ? A(u.color) : S.value[i] || r[i] || r[i % r.length],
          id: re()
        })));
      }
      B();
    }
    ee(() => {
      X();
    });
    const H = _(() => e.value.debug);
    function X() {
      Be(m.dataset) ? ie({
        componentName: "VueUiSparkStackbar",
        type: "dataset",
        debug: H.value
      }) : m.dataset.forEach((t, l) => {
        Le({
          datasetObject: t,
          requiredAttributes: ["name", "value"]
        }).forEach((o) => {
          ie({
            componentName: "VueUiSparkStackbar",
            type: "datasetSerieAttribute",
            property: o,
            index: l,
            debug: H.value
          });
        });
      }), q();
    }
    const y = g({
      width: 500,
      height: 16
    }), d = g([]), J = _(() => h.value.map((t) => t.value || 0).filter((t, l) => !d.value.includes(l)).reduce((t, l) => t + l, 0)), k = _(() => x.value.map((t, l) => {
      const o = t.value || 0, a = o / J.value, b = isNaN(a) ? 0 : a, B = b * y.value.width;
      return {
        ...t,
        color: A(h.value[l]?.color ? h.value[l]?.color : S.value[l] || r[l] || r[l % r.length]),
        value: o,
        proportion: b,
        width: B,
        seriesIndex: l,
        proportionLabel: R({
          v: b * 100,
          s: "%",
          r: e.value.style.legend.percentage.rounding
        })
      };
    })), M = _(() => k.value.filter((t, l) => !d.value.includes(l)));
    function we() {
      d.value.length ? d.value = [] : k.value.forEach((t, l) => {
        d.value.push(l);
      });
    }
    function O(t) {
      d.value.includes(t) ? d.value = d.value.filter((l) => l !== t) : d.value.length < x.value.length - 1 && d.value.push(t);
    }
    function K(t) {
      if (!k.value.length)
        return e.value.debug && console.warn("VueUiSparkStackbar - There are no series to show."), null;
      const l = k.value.find((o) => o.name === t);
      return l || (e.value.debug && console.warn(`VueUiSparkStackbar - Series name not found "${t}"`), null);
    }
    function $e(t) {
      const l = K(t);
      l !== null && d.value.includes(l.id) && O(l.seriesIndex);
    }
    function Se(t) {
      const l = K(t);
      l !== null && (d.value.includes(l.id) || O(l.seriesIndex));
    }
    const D = _(() => {
      let t = 0;
      const l = [];
      for (let o = 0; o < M.value.length; o += 1)
        l.push({
          ...M.value[o],
          start: t
        }), t += M.value[o].width;
      return l;
    }), Te = fe;
    function Q(t, l, o = !1) {
      Te("selectDatapoint", { datapoint: t, index: l }), e.value.events.datapointClick && !o && e.value.events.datapointClick({ datapoint: t, seriesIndex: t.seriesIndex });
    }
    const j = g(null), z = g(!1), T = g(null);
    function Ae({ datapoint: t, seriesIndex: l }) {
      V.value = !1, T.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: t.seriesIndex });
    }
    function Fe({ datapoint: t, seriesIndex: l }) {
      if (e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: t.seriesIndex }), !e.value.style.tooltip.show)
        return;
      j.value = { datapoint: t, seriesIndex: l, config: e.value, series: k.value }, V.value = !0, T.value = l;
      const o = e.value.style.tooltip.customFormat;
      if (Ve(o))
        try {
          const a = o({
            seriesIndex: t.seriesIndex,
            datapoint: t,
            series: k.value,
            config: e.value
          });
          typeof a == "string" && (E.value = a, z.value = !0);
        } catch {
          console.warn("Custom format cannot be applied."), z.value = !1;
        }
      if (!z.value) {
        let a = "";
        a += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${t.name}</div>`, a += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="none" fill="${t.color}"/></svg>`, a += `<b>${t.proportionLabel}</b>`, a += `<span>(${ve(
          e.value.style.legend.value.formatter,
          t.value,
          R({
            p: e.value.style.legend.value.prefix,
            v: t.value,
            s: e.value.style.legend.value.suffix,
            r: e.value.style.legend.value.rounding
          }),
          {
            datapoint: t,
            seriesIndex: l
          }
        )})</span>`, E.value = `<div>${a}</div>`;
      }
    }
    return pe({
      hideSeries: Se,
      showSeries: $e
    }), (t, l) => (s(), n("div", {
      class: "vue-data-ui-component vue-ui-spark-stackbar",
      ref_key: "sparkstackbarChart",
      ref: W,
      style: p(`width:100%; background:${e.value.style.backgroundColor}`)
    }, [
      e.value.style.title.text ? (s(), n("div", {
        key: 0,
        style: p(`width:calc(100% - 12px);background:transparent;margin:0 auto;margin:${e.value.style.title.margin};padding: 0 6px;text-align:${e.value.style.title.textAlign}`)
      }, [
        v("div", {
          class: "atom-title",
          style: p(`font-size:${e.value.style.title.fontSize}px;color:${e.value.style.title.color};font-weight:${e.value.style.title.bold ? "bold" : "normal"}`)
        }, F(e.value.style.title.text), 5),
        e.value.style.title.subtitle.text ? (s(), n("div", {
          key: 0,
          class: "atom-subtitle",
          style: p(`font-size:${e.value.style.title.subtitle.fontSize}px;color:${e.value.style.title.subtitle.color};font-weight:${e.value.style.title.subtitle.bold ? "bold" : "normal"}`)
        }, F(e.value.style.title.subtitle.text), 5)) : f("", !0)
      ], 4)) : f("", !0),
      (s(), n("svg", {
        ref_key: "svgRef",
        ref: xe,
        xmlns: c(Ne),
        width: "100%",
        viewBox: `0 0 ${y.value.width} ${y.value.height}`
      }, [
        le(c(he)),
        v("defs", null, [
          (s(!0), n($, null, I(D.value, (o, a) => (s(), n("linearGradient", {
            key: `stack_gradient_${a}`,
            gradientTransform: "rotate(90)",
            id: `stack_gradient_${a}_${P.value}`
          }, [
            v("stop", {
              offset: "0%",
              "stop-color": o.color
            }, null, 8, He),
            v("stop", {
              offset: "50%",
              "stop-color": c(ce)(c(de)(o.color, 0.05), 100 - e.value.style.bar.gradient.intensity)
            }, null, 8, Xe),
            v("stop", {
              offset: "100%",
              "stop-color": o.color
            }, null, 8, Je)
          ], 8, qe))), 128)),
          v("clipPath", Ke, [
            v("rect", {
              x: "0.005",
              y: "-2",
              width: "0.99",
              height: "5",
              rx: "3",
              ry: "3",
              fill: e.value.style.backgroundColor
            }, null, 8, Qe)
          ])
        ]),
        J.value > 0 ? (s(), n("g", Ze, [
          (s(!0), n($, null, I(D.value, (o, a) => (s(), n("rect", {
            key: `stack_underlayer_${a}`,
            x: o.start,
            y: 0,
            width: o.width,
            height: y.value.height,
            fill: e.value.style.bar.gradient.underlayerColor,
            class: L({ animated: !w.value && !c(C) }),
            style: p({
              opacity: T.value !== null && e.value.style.tooltip.show ? T.value === a ? 1 : 0.5 : 1
            })
          }, null, 14, et))), 128)),
          (s(!0), n($, null, I(D.value, (o, a) => (s(), n("rect", {
            key: `stack_${a}`,
            x: o.start,
            y: 0,
            width: o.width,
            height: y.value.height,
            fill: e.value.style.bar.gradient.show ? `url(#stack_gradient_${a}_${P.value})` : o.color,
            stroke: e.value.style.backgroundColor,
            "stroke-linecap": "round",
            class: L({ animated: !w.value && !c(C) }),
            style: p({
              opacity: T.value !== null && e.value.style.tooltip.show ? T.value === a ? 1 : 0.5 : 1
            })
          }, null, 14, tt))), 128)),
          (s(!0), n($, null, I(D.value, (o, a) => (s(), n("rect", {
            key: `stack_trap_${a}`,
            x: o.start,
            y: 0,
            width: o.width,
            height: y.value.height,
            fill: "transparent",
            stroke: "none",
            class: L({ animated: !w.value && !c(C) }),
            onClick: () => Q(o, a),
            onMouseenter: () => Fe({ datapoint: o, seriesIndex: a }),
            onMouseleave: (b) => Ae({ datapoint: o, seriesIndex: a })
          }, null, 42, lt))), 128))
        ])) : (s(), n("rect", {
          key: 1,
          x: 2,
          y: 1,
          width: y.value.width - 4,
          height: y.value.height - 2,
          stroke: "#CCCCCC",
          "stroke-width": "2",
          fill: "transparent",
          rx: (y.value.height - 4) / 2
        }, null, 8, ot))
      ], 8, Ye)),
      e.value.style.legend.show ? (s(), n("div", {
        key: 1,
        style: p(`background:transparent;margin:0 auto;margin:${e.value.style.legend.margin};justify-content:${e.value.style.legend.textAlign === "left" ? "flex-start" : e.value.style.legend.textAlign === "right" ? "flex-end" : "center"}`),
        class: "vue-ui-sparkstackbar-legend"
      }, [
        e.value.style.legend.selectAllToggle.show && k.value.length > 2 && !c(C) ? (s(), oe(Ge, {
          key: 0,
          backgroundColor: e.value.style.legend.selectAllToggle.backgroundColor,
          color: e.value.style.legend.selectAllToggle.color,
          fontSize: e.value.style.legend.fontSize,
          checked: d.value.length > 0,
          isCursorPointer: Y.value,
          onToggle: we
        }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : f("", !0),
        (s(!0), n($, null, I(k.value, (o, a) => (s(), n("div", {
          style: p(`font-size:${e.value.style.legend.fontSize}px;cursor:${Y.value ? "pointer" : "default"}`),
          class: L({ "vue-ui-sparkstackbar-legend-item": !0, "vue-ui-sparkstackbar-legend-item-unselected": d.value.includes(a) }),
          onClick: (b) => {
            O(a), Q(o, a, !0);
          }
        }, [
          v("div", st, [
            (s(), n("svg", {
              height: `${e.value.style.legend.fontSize}px`,
              width: `${e.value.style.legend.fontSize}px`,
              viewBox: "0 0 10 10"
            }, [
              v("defs", null, [
                v("radialGradient", {
                  id: `legend_grad_${a}-${P.value}`
                }, [
                  v("stop", {
                    offset: "0%",
                    "stop-color": c(C) ? "#FFFFFF" : c(ce)(c(de)(o.color, 0.05), 100 - e.value.style.bar.gradient.intensity)
                  }, null, 8, ut),
                  v("stop", {
                    offset: "100%",
                    "stop-color": o.color
                  }, null, 8, it)
                ], 8, rt)
              ]),
              v("circle", {
                cx: 5,
                cy: 5,
                r: 5,
                fill: e.value.style.bar.gradient.show ? `url(#legend_grad_${a}-${P.value})` : o.color
              }, null, 8, ct)
            ], 8, nt)),
            c(C) ? f("", !0) : (s(), n($, { key: 0 }, [
              v("span", {
                style: p(`color:${e.value.style.legend.name.color}; font-weight:${e.value.style.legend.name.bold ? "bold" : "normal"}`)
              }, F(o.name), 5),
              w.value ? f("", !0) : (s(), n($, { key: 0 }, [
                e.value.style.legend.percentage.show ? (s(), n("span", {
                  key: 0,
                  style: p(`font-weight:${e.value.style.legend.percentage.bold ? "bold" : "normal"};color:${e.value.style.legend.percentage.color}`)
                }, F(d.value.includes(a) ? " - " : o.proportionLabel), 5)) : f("", !0),
                e.value.style.legend.value.show ? (s(), n("span", {
                  key: 1,
                  style: p(`font-weight:${e.value.style.legend.value.bold ? "bold" : "normal"};color:${e.value.style.legend.value.color}`)
                }, " (" + F(c(ve)(
                  e.value.style.legend.value.formatter,
                  o.value,
                  c(R)({
                    p: e.value.style.legend.value.prefix,
                    v: o.value,
                    s: e.value.style.legend.value.suffix,
                    r: e.value.style.legend.value.rounding
                  }),
                  { datapoint: o, seriesIndex: a }
                )) + ") ", 5)) : f("", !0)
              ], 64))
            ], 64))
          ])
        ], 14, at))), 256))
      ], 4)) : f("", !0),
      le(c(ye), {
        teleportTo: e.value.style.tooltip.teleportTo,
        show: V.value && e.value.style.tooltip.show,
        parent: W.value,
        backgroundColor: e.value.style.tooltip.backgroundColor,
        color: e.value.style.tooltip.color,
        fontSize: e.value.style.tooltip.fontSize,
        borderRadius: e.value.style.tooltip.borderRadius,
        borderColor: e.value.style.tooltip.borderColor,
        borderWidth: e.value.style.tooltip.borderWidth,
        backgroundOpacity: e.value.style.tooltip.backgroundOpacity,
        position: e.value.style.tooltip.position,
        content: E.value,
        isCustom: z.value,
        offsetY: -124 + e.value.style.tooltip.offsetY,
        blockShiftY: !0,
        smooth: e.value.style.tooltip.smooth,
        backdropFilter: e.value.style.tooltip.backdropFilter,
        smoothForce: e.value.style.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": ae(() => [
          N(t.$slots, "tooltip-before", se(ne({ ...j.value })), void 0, !0)
        ]),
        "tooltip-after": ae(() => [
          N(t.$slots, "tooltip-after", se(ne({ ...j.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "parent", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "content", "isCustom", "offsetY", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      t.$slots.source ? (s(), n("div", dt, [
        N(t.$slots, "source", {}, void 0, !0)
      ], 512)) : f("", !0),
      N(t.$slots, "skeleton", {}, () => [
        c(C) ? (s(), oe(Me, { key: 0 })) : f("", !0)
      ], !0)
    ], 4));
  }
}, xt = /* @__PURE__ */ We(vt, [["__scopeId", "data-v-2d316e9c"]]);
export {
  xt as default
};
