// vue-data-ui/dist/vue-ui-carousel-table-CmiHTJ8E.js
import { defineAsyncComponent as me, ref as n, useSlots as Ne, onMounted as w, computed as O, watch as N, onBeforeUnmount as G, shallowRef as Ue, nextTick as ge, openBlock as r, createElementBlock as p, unref as u, createElementVNode as U, normalizeClass as Ve, normalizeStyle as h, createTextVNode as D, toDisplayString as X, renderSlot as d, Fragment as j, renderList as J, mergeProps as be, createCommentVNode as R, createBlock as ye, createSlots as Me, withCtx as b, normalizeProps as K, guardReactiveProps as Q } from "vue";
import { u as qe } from "./useConfig-CIgyKZ6R.js";
import { u as Le } from "./useNestedProp-BBT0mbHh.js";
import { j as Ge, o as W, q as Y, u as De, a0 as Xe, a1 as je } from "./lib-GwPnJSrY.js";
import { u as Je } from "./usePrinter-DiK5R6k-.js";
import { u as Ke } from "./useUserOptionState-BIvW1Kz7.js";
import { _ as Qe } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const We = ["id"], Ye = ["height"], Ze = ["data-cell", "height"], _e = {
  key: 2,
  ref: "source",
  dir: "auto"
}, et = {
  __name: "vue-ui-carousel-table",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["copyAlt"],
  setup(k, { expose: Ce, emit: $e }) {
    const Ae = me(() => import("./vue-ui-skeleton-CbqNz9IE.js")), ke = me(() => import("./UserOptions-CrLqEimB.js")), { vue_ui_carousel_table: Te } = qe(), l = k, we = $e, V = n(Ge()), Z = n(!1), I = n(!!l.dataset), Oe = Ne();
    w(() => {
      _();
    }), w(() => {
      Oe["chart-background"] && console.warn("VueUiCarouselTable does not support the #chart-background slot.");
    });
    function _() {
      W(l.dataset) ? Y({
        componentName: "VueUiCarouselTable",
        type: "dataset"
      }) : ((!l.dataset.head || W(l.dataset.head)) && (Y({
        componentName: "VueUiCarouselTable",
        type: "datasetAttribute",
        property: "head"
      }), I.value = !1), (!l.dataset.body || W(l.dataset.body)) && (Y({
        componentName: "VueUiCarouselTable",
        type: "datasetAttribute",
        property: "body"
      }), I.value = !1));
    }
    const e = O({
      get: () => ae(),
      set: (t) => t
    }), Ie = O(() => e.value.userOptions.useCursorPointer), { userOptionsVisible: M, setUserOptionsVisibility: ee, keepUserOptionState: te } = Ke({ config: e.value });
    function ae() {
      return Le({
        userConfig: l.config,
        defaultConfig: Te
      });
    }
    N(() => l.config, (t) => {
      e.value = ae(), M.value = !e.value.userOptions.showOnChartHover, _();
    }, { deep: !0 }), N(() => l.dataset, (t) => {
      q();
    }, { deep: !0 });
    const { isPrinting: oe, isImaging: ne, generatePdf: xe, generateImage: le } = Je({
      elementId: `carousel-table_${V.value}`,
      fileName: e.value.caption.text || "vue-ui-carousel-table",
      options: e.value.userOptions.print
    }), m = n({
      showAnimation: e.value.animation.use
    }), i = n(null), ie = n(null), x = n(null), P = n(null), F = n(0), z = n(0), E = n(!1), S = n(null), y = n(null), T = n(0);
    function q() {
      S.value && (y.value = {
        elements: S.value.getElementsByTagName("tr"),
        heights: Array.from(S.value.getElementsByTagName("tr")).map((t) => t.getBoundingClientRect().height)
      });
    }
    w(q);
    const Pe = O(() => !y.value || !y.value.heights.length ? 0 : Math.max(...y.value.heights) + F.value + z.value), Fe = O(() => l.dataset.body ? e.value.tbody.tr.visible <= l.dataset.body.length ? e.value.tbody.tr.visible : l.dataset.body.length : 0), Se = O(() => (e.value.tbody.tr.height + e.value.tbody.tr.td.padding.top + e.value.tbody.tr.td.padding.bottom + e.value.tbody.tr.border.size * 2) * Fe.value + F.value + z.value), c = n(0), C = n(null), v = n(0), $ = n(!1), se = n(0);
    w(() => {
      x.value && (F.value = x.value.getBoundingClientRect().height), P.value && (z.value = P.value.getBoundingClientRect().height), m.value.showAnimation && y.value && ue();
    }), w(() => {
      if (i.value) {
        let a = function() {
          const s = t.getBoundingClientRect().bottom;
          o.forEach((f) => {
            f.getBoundingClientRect().top < s ? f.style.visibility = "hidden" : f.style.visibility = "visible";
          });
        };
        const t = i.value.querySelector("thead"), o = Array.from(S.value.querySelectorAll("tr"));
        i.value.addEventListener("scroll", a), a(), G(() => {
          i.value.removeEventListener("scroll", a);
        });
      }
    });
    function Be(t) {
      Z.value = t, se.value += 1;
    }
    function ue() {
      !C.value && !$.value && (e.value.animation.type === "scroll" ? C.value = requestAnimationFrame(re) : C.value = requestAnimationFrame(de));
    }
    function Re() {
      if (!i.value) return !1;
      const { scrollTop: t, scrollHeight: o, clientHeight: a } = i.value;
      return t + a >= o;
    }
    function re(t) {
      if ($.value) return;
      v.value || (v.value = t), t - v.value >= e.value.animation.speedMs && (c.value += y.value.heights[T.value], (Re() || T.value >= y.value.heights.length) && (c.value = 0, T.value = -1), T.value += 1, i.value && i.value.scrollTo({
        top: c.value,
        behavior: "smooth"
      }), v.value = t), C.value = requestAnimationFrame(re);
    }
    function de(t) {
      if ($.value) return;
      v.value || (v.value = t);
      const o = t - v.value, a = e.value.animation.speedMs / 4 / 1e3;
      o >= a && (c.value += a, c.value >= i.value.scrollHeight - i.value.clientHeight && (c.value = 0), i.value && i.value.scrollTo({
        top: c.value,
        behavior: "auto"
      }), v.value = t), C.value = requestAnimationFrame(de);
    }
    function g() {
      $.value = !0, cancelAnimationFrame(C.value), C.value = null;
    }
    G(g);
    function A() {
      !$.value || !m.value.showAnimation || ($.value = !1, v.value = 0, ue());
    }
    function ze() {
      e.value.animation.pauseOnHover && g();
    }
    const L = n(null);
    function Ee() {
      g(), clearTimeout(L.value);
    }
    function ve() {
      clearTimeout(L.value), L.value = setTimeout(A, 1e3);
    }
    N(
      () => e.value.animation.use,
      (t) => {
        t ? (m.value.showAnimation = !0, A()) : (m.value.showAnimation = !1, g());
      }
    ), N(
      () => e.value.animation.type,
      (t) => {
        g(), c.value = 0, T.value = 0, i.value.scrollTo({
          top: 0,
          behavior: "auto"
        }), A();
      }
    );
    const He = O(() => e.value.responsiveBreakpoint), H = Ue(null);
    w(() => {
      H.value = new ResizeObserver((t) => {
        t.forEach((o) => {
          E.value = o.contentRect.width < He.value;
        }), F.value = x.value ? x.value.getBoundingClientRect().height : 0, z.value = P.value ? P.value.getBoundingClientRect().height : 0, T.value = 0, ge(() => {
          g(), v.value = 0, c.value = 0, q(), A();
        });
      }), i.value && H.value.observe(i.value);
    }), G(() => {
      H.value && H.value.disconnect();
    });
    function pe() {
      xe();
    }
    function ce() {
      m.value.showAnimation = !m.value.showAnimation, m.value.showAnimation ? A() : g();
    }
    function fe(t = null) {
      ge(() => {
        const o = l.dataset.head.map((f, B) => [
          [l.dataset.body[B]]
        ]), a = [
          [e.value.caption.text],
          [l.dataset.head.map((f) => [f])]
        ].concat(o), s = Xe(a);
        t ? t(s) : je({
          csvContent: s,
          title: e.value.caption.text || "vue-ui-carousel-table"
        });
      });
    }
    async function he() {
      if (we("copyAlt", {
        config: e.value,
        dataset: l.dataset
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: l.dataset
      }));
    }
    return Ce({
      pauseAnimation: g,
      resumeAnimation: A,
      toggleAnimation: ce,
      generateCsv: fe,
      generatePdf: pe,
      generateImage: le,
      copyAlt: he
    }), (t, o) => (r(), p("div", {
      class: "vue-data-ui-component vue-ui-carousel-table",
      style: { position: "relative", overflow: "visible" },
      ref_key: "chartContainer",
      ref: ie,
      onMouseenter: o[5] || (o[5] = () => u(ee)(!0)),
      onMouseleave: o[6] || (o[6] = () => u(ee)(!1))
    }, [
      U("div", {
        ref_key: "tableContainer",
        ref: i,
        id: `carousel-table_${V.value}`,
        style: h({
          height: u(oe) || u(ne) ? "auto" : `${Math.max(Se.value, Pe.value)}px`,
          containerType: "inline-size",
          position: "relative",
          overflow: "auto",
          fontFamily: e.value.fontFamily
        }),
        class: Ve({ "vue-ui-responsive": E.value, "is-playing": e.value.scrollbar.hide || !$.value && e.value.scrollbar.showOnlyOnHover }),
        onMouseover: o[0] || (o[0] = (a) => ze()),
        onMouseleave: o[1] || (o[1] = (a) => A()),
        onTouchstart: o[2] || (o[2] = (a) => Ee()),
        onTouchend: o[3] || (o[3] = (a) => ve()),
        onTouchcancel: o[4] || (o[4] = (a) => ve())
      }, [
        I.value ? (r(), p("table", {
          key: 0,
          class: "vue-data-ui-carousel-table",
          style: h({
            ...e.value.style,
            border: `${e.value.border.size}px solid ${e.value.border.color}`,
            width: "100%",
            borderCollapse: "collapse",
            backgroundColor: e.value.tbody.backgroundColor
          })
        }, [
          U("caption", {
            ref_key: "caption",
            ref: x,
            class: "vue-data-ui-carousel-table-caption",
            style: h({
              ...e.value.caption.style,
              fontFamily: "inherit",
              position: "sticky",
              top: 0,
              zIndex: 2,
              paddingTop: e.value.caption.padding.top + "px",
              paddingRight: e.value.caption.padding.right + "px",
              paddingBottom: e.value.caption.padding.bottom + "px",
              paddingLeft: e.value.caption.padding.left + "px",
              boxShadow: E.value ? e.value.thead.tr.style.boxShadow : "none",
              minHeight: "36px",
              display: t.$slots.caption || e.value.caption.text || e.value.userOptions.show ? "" : "none"
            })
          }, [
            D(X(e.value.caption.text && !t.$slots.caption ? e.value.caption.text : "") + " ", 1),
            d(t.$slots, "caption", {}, void 0, !0)
          ], 4),
          U("thead", {
            style: h({ ...e.value.thead.style, position: "sticky", top: `${t.$slots.caption || e.value.caption.text || e.value.userOptions.show ? F.value : 0}px`, zIndex: 1 })
          }, [
            U("tr", {
              ref_key: "tableRow",
              ref: P,
              role: "row",
              style: h({
                ...e.value.thead.tr.style,
                border: e.value.thead.tr.border.size ? `${e.value.thead.tr.border.size}px solid ${e.value.thead.tr.border.color}` : "none",
                boxShadow: E.value ? "none" : e.value.thead.tr.style.boxShadow
              }),
              height: `${e.value.thead.tr.height}px`
            }, [
              (r(!0), p(j, null, J(k.dataset.head, (a, s) => (r(), p("th", {
                role: "cell",
                key: `th_${s}`,
                style: h({
                  ...e.value.thead.tr.th.style,
                  border: e.value.thead.tr.th.border.size ? `${e.value.thead.tr.th.border.size}px solid ${e.value.thead.tr.th.border.color}` : "none",
                  paddingTop: e.value.thead.tr.th.padding.top + "px",
                  paddingRight: e.value.thead.tr.th.padding.right + "px",
                  paddingBottom: e.value.thead.tr.th.padding.bottom + "px",
                  paddingLeft: e.value.thead.tr.th.padding.left + "px"
                })
              }, [
                D(X(t.$slots.th ? "" : a) + " ", 1),
                d(t.$slots, "th", be({ ref_for: !0 }, { th: a, colIndex: s }), void 0, !0)
              ], 4))), 128))
            ], 12, Ye)
          ], 4),
          k.dataset.body && k.dataset.head ? (r(), p("tbody", {
            key: 0,
            ref_key: "tbody",
            ref: S,
            style: {
              clipPath: "inset(0,0,0,0)"
            }
          }, [
            (r(!0), p(j, null, J(k.dataset.body, (a, s) => (r(), p("tr", {
              style: h({
                ...e.value.tbody.tr.style,
                border: `${e.value.tbody.tr.border.size}px solid ${e.value.tbody.tr.border.color}`,
                verticalAlign: "middle"
              })
            }, [
              (r(!0), p(j, null, J(a, (f, B) => (r(), p("td", {
                role: "cell",
                "data-cell": k.dataset.head[B] || "",
                style: h({
                  ...e.value.tbody.tr.td.style,
                  border: `${e.value.tbody.tr.td.border.size}px solid ${e.value.tbody.tr.td.border.color}`,
                  backgroundColor: u(De)(e.value.tbody.tr.td.style.backgroundColor, s % 2 === 0 && e.value.tbody.tr.td.alternateColor ? e.value.tbody.tr.td.alternateOpacity * 100 : 100),
                  paddingTop: e.value.tbody.tr.td.padding.top + "px",
                  paddingRight: e.value.tbody.tr.td.padding.right + "px",
                  paddingBottom: e.value.tbody.tr.td.padding.bottom + "px",
                  paddingLeft: e.value.tbody.tr.td.padding.left + "px",
                  verticalAlign: "middle"
                }),
                height: `${e.value.tbody.tr.height}px`
              }, [
                D(X(t.$slots.td ? "" : f) + " ", 1),
                d(t.$slots, "td", be({ ref_for: !0 }, { td: f, rowIndex: s, colIndex: B }), void 0, !0)
              ], 12, Ze))), 256))
            ], 4))), 256))
          ], 512)) : R("", !0)
        ], 4)) : R("", !0)
      ], 46, We),
      I.value ? R("", !0) : (r(), ye(u(Ae), {
        key: 0,
        config: {
          type: "table"
        }
      })),
      e.value.userOptions.show && I.value && (u(te) || u(M)) ? (r(), ye(u(ke), {
        ref: "details",
        key: `user_option_${se.value}`,
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        isPrinting: u(oe),
        isImaging: u(ne),
        uid: V.value,
        hasTooltip: !1,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: !1,
        hasLabel: !1,
        hasAnimation: e.value.userOptions.buttons.animation,
        isAnimation: !m.value.showAnimation,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: Z.value,
        chartElement: ie.value,
        position: e.value.userOptions.position,
        titles: { ...e.value.userOptions.buttonTitles },
        zIndex: 3,
        offsetX: 12,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        isCursorPointer: Ie.value,
        onGeneratePdf: pe,
        onGenerateCsv: fe,
        onGenerateImage: u(le),
        onToggleAnimation: ce,
        onToggleFullscreen: Be,
        onCopyAlt: he,
        style: h({
          visibility: u(te) ? u(M) ? "visible" : "hidden" : "visible"
        })
      }, Me({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: b(({ isOpen: a, color: s }) => [
            d(t.$slots, "menuIcon", K(Q({ isOpen: a, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: b(() => [
            d(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: b(() => [
            d(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: b(() => [
            d(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: b(() => [
            d(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionAnimation ? {
          name: "optionAnimation",
          fn: b(() => [
            d(t.$slots, "optionAnimation", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: b(({ toggleFullscreen: a, isFullscreen: s }) => [
            d(t.$slots, "optionFullscreen", K(Q({ toggleFullscreen: a, isFullscreen: s })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: b(({ altCopy: a }) => [
            d(t.$slots, "optionAltCopy", K(Q({ altCopy: a })), void 0, !0)
          ]),
          key: "7"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasXls", "hasAnimation", "isAnimation", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "titles", "callbacks", "printScale", "isCursorPointer", "onGenerateImage", "style"])) : R("", !0),
      t.$slots.source ? (r(), p("div", _e, [
        d(t.$slots, "source", {}, void 0, !0)
      ], 512)) : R("", !0)
    ], 544));
  }
}, ut = /* @__PURE__ */ Qe(et, [["__scopeId", "data-v-82bf3039"]]);
export {
  ut as default
};
