// vue-data-ui/dist/vue-ui-molecule-Chq1qWDK.js
import { useCssVars as St, defineAsyncComponent as m, computed as p, onMounted as It, ref as u, toRefs as Dt, watch as G, nextTick as Z, openBlock as d, createElementBlock as A, unref as a, normalizeStyle as E, normalizeClass as Le, createBlock as T, withCtx as i, renderSlot as r, normalizeProps as f, guardReactiveProps as h, createCommentVNode as k, createSlots as Be, createVNode as D, createElementVNode as _, Fragment as _t, renderList as Ft, resolveDynamicComponent as Nt, mergeProps as Lt, createTextVNode as Ue, toDisplayString as Ze } from "vue";
import { o as Bt, q as Ut, j as Ee, t as Re, m as Zt, p as ze, n as Et, af as Rt, X as zt, x as Vt, A as Mt, s as Ve, a0 as Ht, a1 as Gt, T as xt, U as Wt } from "./lib-GwPnJSrY.js";
import { u as jt } from "./useConfig-CIgyKZ6R.js";
import { u as Xt, B as Yt } from "./BaseScanner-BhhkKVoj.js";
import { u as qt } from "./usePrinter-DiK5R6k-.js";
import { u as Jt } from "./useSvgExport-ByUukOZt.js";
import { u as Kt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ne } from "./useNestedProp-BBT0mbHh.js";
import { u as Qt } from "./useUserOptionState-BIvW1Kz7.js";
import { u as eo } from "./useChartAccessibility-9icAAmYg.js";
import to from "./img-BYYzY_QU.js";
import oo from "./vue_ui_molecule-BtvlINUh.js";
import { u as lo } from "./usePanZoom-CDF7rb3o.js";
import { _ as ao } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const so = ["id"], no = {
  key: 1,
  ref: "noTitle",
  class: "vue-data-ui-no-title-space",
  style: "height:36px; width: 100%;background:transparent"
}, ro = {
  key: 2,
  style: "width:100%;background:transparent;"
}, io = ["xmlns", "viewBox"], uo = ["width", "height"], co = ["id"], vo = ["stop-color"], po = ["stop-color"], fo = {
  key: 4,
  class: "vue-data-ui-watermark"
}, ho = {
  key: 5,
  "data-dom-to-png-ignore": "",
  class: "reset-wrapper"
}, mo = {
  key: 6,
  ref: "source",
  dir: "auto"
}, go = ["innerHTML"], Co = {
  __name: "vue-ui-molecule",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectNode", "copyAlt"],
  setup(Me, { expose: He, emit: Ge }) {
    St((t) => ({
      v791dd11f: t.slicerColor
    }));
    const xe = m(() => import("./Title-BfyrP-jp.js")), We = m(() => import("./Tooltip-DL3QgjZf.js")), re = m(() => import("./BaseIcon-CNx8Tzl5.js")), je = m(() => import("./vue-ui-accordion-BGHTqIBt.js")), Xe = m(() => import("./DataTable-Bo1pmAqG.js")), Ye = m(() => import("./PenAndPaper-DWV1FMIi.js")), qe = m(() => import("./UserOptions-CrLqEimB.js")), Je = m(() => import("./PackageVersion-DN9d0ME5.js")), Ke = m(() => import("./RecursiveLinks-BYh0tHDW.js")), Qe = m(() => import("./RecursiveLabels-BhQMS9Sm.js")), et = m(() => import("./RecursiveCircles-BE6GoMmo.js")), tt = m(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_molecule: ot } = jt(), { isThemeValid: lt, warnInvalidTheme: at } = Kt(), w = Me, ie = Ge, ue = p(() => !!w.dataset && w.dataset.length);
    It(() => {
      ce();
    });
    const st = p(() => e.value.debug);
    function ce() {
      Bt(w.dataset) && Ut({
        componentName: "VueUiMolecule",
        type: "dataset",
        debug: st.value
      });
    }
    const x = u(Ee()), W = u(!1), R = u(""), $ = u(null), ve = u(0), de = u(0), pe = u(0), F = u(null), j = u(null), z = u(!1), V = u(!1), e = u(J()), P = p(() => e.value.userOptions.useCursorPointer), nt = p(() => Re({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            nodes: { stroke: "#6A6A6A" },
            links: { stroke: "#6A6A6A80" }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: X, FINAL_DATASET: Y } = Xt({
      ...Dt(w),
      FINAL_CONFIG: e,
      prepareConfig: J,
      skeletonDataset: w.config?.skeletonDataset ?? [
        {
          name: "_",
          color: "#CACACA",
          nodes: [
            { name: "_", color: "#CACACA", nodes: [
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" }
            ] },
            { name: "_", color: "#CACACA", nodes: [
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" }
            ] },
            { name: "_", color: "#CACACA", nodes: [
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" }
            ] },
            { name: "_", color: "#CACACA", nodes: [
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" }
            ] },
            { name: "_", color: "#CACACA", nodes: [
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" }
            ] },
            { name: "_", color: "#CACACA", nodes: [
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" },
              { name: "_", color: "#CACACA" }
            ] }
          ]
        }
      ],
      skeletonConfig: Re({
        defaultConfig: e.value,
        userConfig: nt.value
      })
    }), { userOptionsVisible: q, setUserOptionsVisibility: fe, keepUserOptionState: he } = Qt({ config: e.value }), { svgRef: N } = eo({ config: e.value.style.chart.title });
    function J() {
      const t = ne({
        userConfig: w.config,
        defaultConfig: ot
      }), l = t.theme;
      if (!l) return t;
      if (!lt.value(t))
        return at(t), t;
      const o = ne({
        userConfig: oo[l] || w.config,
        defaultConfig: t
      }), s = ne({
        userConfig: w.config,
        defaultConfig: o
      });
      return {
        ...s,
        customPalette: s.customPalette.length ? s.customPalette : Zt[l] || ze
      };
    }
    G(() => w.config, (t) => {
      e.value = J(), q.value = !e.value.userOptions.showOnChartHover, ce(), de.value += 1, pe.value += 1, c.value.showTable = e.value.table.show, c.value.showTooltip = e.value.style.chart.tooltip.show, c.value.showZoom = e.value.style.chart.zoom.show;
    }, { deep: !0 });
    const me = u(0), rt = () => {
      me.value += 1;
    };
    G([() => X.value, () => Y.value], async ([t]) => {
      t || (await Z(), g.value = ye(), await Z(), rt(), await Z(), bt({ x: 0, y: 0, width: 400, height: 400 }), le(!1));
    }, { flush: "post", deep: !1 });
    const { isPrinting: K, isImaging: Q, generatePdf: ge, generateImage: Ce } = qt({
      elementId: `cluster_${x.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-molecule",
      options: e.value.userOptions.print
    }), it = p(() => e.value.userOptions.show && !e.value.style.chart.title.text), ut = p(() => Et(e.value.customPalette)), c = u({
      showTable: e.value.table.show,
      showDataLabels: !0,
      showTooltip: e.value.style.chart.tooltip.show,
      showZoom: e.value.style.chart.zoom.show
    });
    G(e, () => {
      c.value = {
        showTable: e.value.table.show,
        showDataLabels: !0,
        showTooltip: e.value.style.chart.tooltip.show,
        showZoom: e.value.style.chart.zoom.show
      };
    }, { immediate: !0 });
    function be(t, l = 0) {
      return Array.isArray(t) && t.length > 0 && t[0].nodes ? be(t[0].nodes, l + 1) : l;
    }
    function ye() {
      const t = be(Y.value);
      let l = 100, o = l;
      for (let s = 0; s < t; s += 1)
        l /= 1, o += l;
      return {
        height: o,
        width: o
      };
    }
    const g = u(ye());
    function ke(t, l = { x: -g.value.width / 2.43, y: g.value.height / 2 }, o = g.value.width / 1.1, s = 24, n = 0, v = 0, b = "#BBBBBB", ae = 0) {
      if (!Array.isArray(t) || t.length === 0) return t;
      const Ot = Mt({
        plot: l,
        radius: o,
        sides: t.length,
        rotation: n
      });
      return t.forEach((y, _e) => {
        const Fe = Ot.coordinates[_e], Ne = y.color ? (() => {
          let se = Ve(y.color);
          return se.startsWith("#") ? se : `#${se}`;
        })() : null;
        let I;
        Ne ? I = Ne : ae === 0 ? I = b : ae === 1 ? (I = ut.value[v] || ze[v] || b, v += 1) : I = b, y.polygonPath = { coordinates: [Fe] }, y.circleRadius = s, y.color = I, y.uid = Ee(), Array.isArray(y.nodes) && y.nodes.length && (y.nodes = ke(
          y.nodes,
          Fe,
          o / 2.9,
          s / 2.2,
          n + Math.PI * _e / y.nodes.length,
          v,
          I,
          ae + 1
        ));
      }), t;
    }
    function ct(t) {
      const l = /* @__PURE__ */ new Set();
      function o(n) {
        n.forEach((v) => {
          if (!v.color) return;
          let b = v.color;
          /^#?[0-9A-F]{6}$/i.test(b) || (b = Ve(b)), b.startsWith("#") || (b = `#${b}`), l.add(b), Array.isArray(v.nodes) && v.nodes.length && o(v.nodes);
        });
      }
      o(t);
      const s = {};
      return Array.from(l).forEach((n) => {
        const v = n.slice(1);
        s[n] = `gradient_${v}`;
      }), s;
    }
    const vt = p(() => ct(C.value)), C = p(() => {
      const t = Rt(Y.value);
      return ke(t);
    }), ee = u(null);
    function dt(t) {
      ee.value = {
        datapoint: t,
        seriesIndex: -1,
        series: C.value,
        config: e.value
      };
      const l = e.value.style.chart.tooltip.customFormat;
      if (xt(l) && Wt(() => l({
        seriesIndex: -1,
        datapoint: t,
        series: C.value,
        config: e.value
      })))
        R.value = l({
          seriesIndex: -1,
          // well, ok
          datapoint: t,
          series: C.value,
          config: e.value
        });
      else {
        let o = "";
        o += `<div style="display:flex;align-items:center;gap:3px"><div style="color:${t.color}">⬤</div><div>${t.name}</div></div>`, t.details && (o += `<div style="width:100%;border-top:1px solid ${e.value.style.chart.tooltip.borderColor};margin-top: 2px">${t.details}</div>`), R.value = `<div style="font-family:inherit">${o}</div>`;
      }
    }
    const M = u(null), L = u(null), Ae = u(null);
    function pt(t) {
      Ae.value = M.value, M.value = t, t ? e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: -1 }) : e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: Ae.value || M.value, seriesIndex: -1 }), t ? (W.value = !0, dt(t), L.value = t.uid) : (W.value = !1, R.value = "", L.value = null);
    }
    function ft(t) {
      const l = [];
      function o(s) {
        s.forEach((n) => {
          const v = {
            name: n.name,
            details: n.details || "-",
            ancestor: n.ancestor && n.ancestor.name || "-",
            color: n.color || ""
          };
          l.push(v), n.nodes && n.nodes.length > 0 && o(n.nodes, n.name);
        });
      }
      return o(t), l;
    }
    const ht = p(() => ft(C.value)), O = p(() => {
      const t = [
        e.value.table.translations.nodeName,
        e.value.table.translations.details,
        e.value.table.translations.ancestor
      ], l = ht.value.map((n, v) => [
        {
          color: n.color,
          name: n.name
        },
        n.details,
        n.ancestor || ""
      ]), o = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, s = [
        e.value.table.translations.nodeName,
        e.value.table.translations.details,
        e.value.table.translations.ancestor
      ];
      return {
        head: t,
        body: l,
        config: o,
        colNames: s
      };
    });
    function te(t = null) {
      Z(() => {
        const l = O.value.body.map((n, v) => [[n[0].name], [n[1]], [n[2]]]), o = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[...O.value.head]]].concat(l), s = Ht(o);
        t ? t(s) : Gt({ csvContent: s, title: e.value.style.chart.title.text || "vue-ui-molecule" });
      });
    }
    function mt() {
      return C.value;
    }
    const S = u(!1);
    function we(t) {
      S.value = t, ve.value += 1;
    }
    function $e() {
      c.value.showTable = !c.value.showTable;
    }
    function Te() {
      c.value.showDataLabels = !c.value.showDataLabels;
    }
    function Pe() {
      c.value.showTooltip = !c.value.showTooltip;
    }
    function Oe() {
      c.value.showZoom = !c.value.showZoom;
    }
    const B = u(!1);
    function oe() {
      B.value = !B.value;
    }
    const gt = p(() => !B.value && c.value.showZoom), { viewBox: H, resetZoom: le, isZoom: Ct, setInitialViewBox: bt } = lo(N, {
      x: 0,
      y: 0,
      width: Math.max(10, g.value.width),
      height: Math.max(10, g.value.height)
    }, e.value.style.chart.zoom.speed, gt);
    function yt(t) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: -1 }), ie("selectNode", t);
    }
    async function kt({ scale: t = 2 } = {}) {
      if (!$.value) return;
      const { width: l, height: o } = $.value.getBoundingClientRect(), s = l / o, { imageUri: n, base64: v } = await to({ domElement: $.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: n,
        base64: v,
        title: e.value.style.chart.title.text,
        width: l,
        height: o,
        aspectRatio: s
      };
    }
    const U = p(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = c.value.showTable;
      return {
        component: t ? tt : je,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: S.value,
          fullscreenParent: $.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: P.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    G(() => c.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && F.value ? F.value.open() : "close" in F.value && F.value.close());
    });
    function Se() {
      c.value.showTable = !1, j.value && j.value.setTableIconState(!1);
    }
    const At = p(() => e.value.style.chart.backgroundColor), wt = p(() => e.value.style.chart.title), { exportSvg: $t, getSvg: Tt } = Jt({
      svg: N,
      title: wt,
      backgroundColor: At
    });
    async function Ie({ isCb: t }) {
      V.value = !0, await Z();
      try {
        if (t) {
          const { blob: l, url: o, text: s, dataUrl: n } = await Tt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: o, text: s, dataUrl: n }));
        } else
          await Promise.resolve($t());
      } finally {
        V.value = !1;
      }
    }
    function Pt(t) {
      if (t?.stage === "start") {
        z.value = !0;
        return;
      }
      if (t?.stage === "end") {
        z.value = !1;
        return;
      }
      Ce();
    }
    async function De() {
      if (ie("copyAlt", {
        config: e.value,
        dataset: C.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: C.value
      }));
    }
    return He({
      getData: mt,
      getImage: kt,
      generatePdf: ge,
      generateCsv: te,
      generateImage: Ce,
      generateSvg: Ie,
      toggleTable: $e,
      toggleLabels: Te,
      toggleTooltip: Pe,
      toggleAnnotator: oe,
      toggleFullscreen: we,
      toggleZoom: Oe,
      copyAlt: De
    }), (t, l) => (d(), A("div", {
      ref_key: "moleculeChart",
      ref: $,
      class: Le(`vue-data-ui-component vue-ui-molecule ${S.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      style: E(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor}`),
      id: `cluster_${x.value}`,
      onMouseleave: l[2] || (l[2] = (o) => {
        M.value = null, L.value = null, a(fe)(!1);
      }),
      onMouseenter: l[3] || (l[3] = () => a(fe)(!0))
    }, [
      e.value.userOptions.buttons.annotator && a(N) ? (d(), T(a(Ye), {
        key: 0,
        svgRef: a(N),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: B.value,
        isCursorPointer: P.value,
        onClose: oe
      }, {
        "annotator-action-close": i(() => [
          r(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": i(({ color: o }) => [
          r(t.$slots, "annotator-action-color", f(h({ color: o })), void 0, !0)
        ]),
        "annotator-action-draw": i(({ mode: o }) => [
          r(t.$slots, "annotator-action-draw", f(h({ mode: o })), void 0, !0)
        ]),
        "annotator-action-undo": i(({ disabled: o }) => [
          r(t.$slots, "annotator-action-undo", f(h({ disabled: o })), void 0, !0)
        ]),
        "annotator-action-redo": i(({ disabled: o }) => [
          r(t.$slots, "annotator-action-redo", f(h({ disabled: o })), void 0, !0)
        ]),
        "annotator-action-delete": i(({ disabled: o }) => [
          r(t.$slots, "annotator-action-delete", f(h({ disabled: o })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : k("", !0),
      it.value ? (d(), A("div", no, null, 512)) : k("", !0),
      e.value.style.chart.title.text ? (d(), A("div", ro, [
        (d(), T(a(xe), {
          key: `title_${de.value}`,
          config: {
            title: {
              cy: "molecule-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "molecule-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ])) : k("", !0),
      e.value.userOptions.show && ue.value && (a(he) || a(q)) ? (d(), T(a(qe), {
        ref_key: "userOptionsRef",
        ref: j,
        key: `user_options_${ve.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: a(K),
        isImaging: a(Q),
        uid: x.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isTooltip: c.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: $.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: B.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        hasZoom: e.value.userOptions.buttons.zoom,
        isZoom: c.value.showZoom,
        isCursorPointer: P.value,
        onToggleFullscreen: we,
        onGeneratePdf: a(ge),
        onGenerateCsv: te,
        onGenerateImage: Pt,
        onGenerateSvg: Ie,
        onToggleTable: $e,
        onToggleLabels: Te,
        onToggleTooltip: Pe,
        onToggleAnnotator: oe,
        onToggleZoom: Oe,
        onCopyAlt: De,
        style: E({
          visibility: a(he) ? a(q) ? "visible" : "hidden" : "visible"
        })
      }, Be({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: i(({ isOpen: o, color: s }) => [
            r(t.$slots, "menuIcon", f(h({ isOpen: o, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: i(() => [
            r(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: i(() => [
            r(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: i(() => [
            r(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: i(() => [
            r(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: i(() => [
            r(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: i(() => [
            r(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: i(() => [
            r(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: i(({ toggleFullscreen: o, isFullscreen: s }) => [
            r(t.$slots, "optionFullscreen", f(h({ toggleFullscreen: o, isFullscreen: s })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: i(({ toggleAnnotator: o, isAnnotator: s }) => [
            r(t.$slots, "optionAnnotator", f(h({ toggleAnnotator: o, isAnnotator: s })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        t.$slots.optionZoom ? {
          name: "optionZoom",
          fn: i(({ toggleZoom: o, isZoomLocked: s }) => [
            r(t.$slots, "optionZoom", f(h({ toggleZoom: o, isZoomLocked: s })), void 0, !0)
          ]),
          key: "10"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: i(({ altCopy: o }) => [
            r(t.$slots, "optionAltCopy", f(h({ altCopy: o })), void 0, !0)
          ]),
          key: "11"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "hasZoom", "isZoom", "isCursorPointer", "onGeneratePdf", "style"])) : k("", !0),
      (d(), A("svg", {
        ref_key: "svgRef",
        ref: N,
        key: `svg_${me.value}`,
        xmlns: a(zt),
        viewBox: `${a(H).x} ${a(H).y} ${a(H).width} ${a(H).height}`,
        class: Le({ "vue-data-ui-fullscreen--on": S.value, "vue-data-ui-fulscreen--off": !S.value }),
        style: E(`overflow: hidden; background:transparent;color:${e.value.style.chart.color}`)
      }, [
        D(a(Je)),
        t.$slots["chart-background"] ? (d(), A("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: g.value.width <= 0 ? 10 : g.value.width,
          height: g.value.height <= 0 ? 10 : g.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          r(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, uo)) : k("", !0),
        _("defs", null, [
          (d(!0), A(_t, null, Ft(Object.keys(vt.value), (o) => (d(), A("radialGradient", {
            id: `gradient_${o}`,
            cx: "50%",
            cy: "30%",
            r: "50%",
            fx: "50%",
            fy: "50%"
          }, [
            _("stop", {
              offset: "0%",
              "stop-color": a(Vt)(o, 0.5)
            }, null, 8, vo),
            _("stop", {
              offset: "100%",
              "stop-color": o
            }, null, 8, po)
          ], 8, co))), 256))
        ]),
        D(a(Ke), {
          dataset: C.value,
          color: e.value.style.chart.links.stroke,
          backgroundColor: e.value.style.chart.backgroundColor
        }, null, 8, ["dataset", "color", "backgroundColor"]),
        D(a(et), {
          dataset: C.value,
          hoveredUid: L.value,
          stroke: e.value.style.chart.nodes.stroke,
          strokeHovered: e.value.style.chart.nodes.strokeHovered,
          onClick: yt,
          onHover: pt
        }, {
          node: i(({ node: o }) => [
            r(t.$slots, "node", f(h({ node: o })), void 0, !0)
          ]),
          _: 3
        }, 8, ["dataset", "hoveredUid", "stroke", "strokeHovered"]),
        c.value.showDataLabels && !a(X) ? (d(), T(a(Qe), {
          key: 1,
          dataset: C.value,
          color: e.value.style.chart.color,
          hoveredUid: L.value
        }, null, 8, ["dataset", "color", "hoveredUid"])) : k("", !0),
        r(t.$slots, "svg", {
          svg: {
            ...g.value,
            isPrintingImg: a(K) | a(Q) | z.value,
            isPrintingSvg: V.value
          }
        }, void 0, !0)
      ], 14, io)),
      t.$slots.watermark ? (d(), A("div", fo, [
        r(t.$slots, "watermark", f(h({ isPrinting: a(K) || a(Q) || z.value || V.value })), void 0, !0)
      ])) : k("", !0),
      a(Ct) ? (d(), A("div", ho, [
        r(t.$slots, "reset-action", { reset: a(le) }, () => [
          _("button", {
            "data-cy-reset": "",
            tabindex: "0",
            role: "button",
            class: "vue-data-ui-refresh-button",
            style: E({
              background: e.value.style.chart.backgroundColor,
              cursor: P.value ? "pointer" : "default"
            }),
            onClick: l[0] || (l[0] = (o) => a(le)(!0))
          }, [
            D(a(re), {
              name: "refresh",
              stroke: e.value.style.chart.color
            }, null, 8, ["stroke"])
          ], 4)
        ], !0)
      ])) : k("", !0),
      t.$slots.source ? (d(), A("div", mo, [
        r(t.$slots, "source", {}, void 0, !0)
      ], 512)) : k("", !0),
      D(a(We), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: c.value.showTooltip && W.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        fontSize: e.value.style.chart.tooltip.fontSize,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: $.value,
        content: R.value,
        isFullscreen: S.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": i(() => [
          r(t.$slots, "tooltip-before", f(h({ ...ee.value })), void 0, !0)
        ]),
        "tooltip-after": i(() => [
          r(t.$slots, "tooltip-after", f(h({ ...ee.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      ue.value && e.value.userOptions.buttons.table ? (d(), T(Nt(U.value.component), Lt({ key: 7 }, U.value.props, {
        ref_key: "tableUnit",
        ref: F,
        onClose: Se
      }), Be({
        content: i(() => [
          (d(), T(a(Xe), {
            key: `table_${pe.value}`,
            colNames: O.value.colNames,
            head: O.value.head,
            body: O.value.body,
            config: O.value.config,
            title: e.value.table.useDialog ? "" : U.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: P.value,
            onClose: Se
          }, {
            th: i(({ th: o }) => [
              _("div", {
                innerHTML: o,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, go)
            ]),
            td: i(({ td: o }) => [
              Ue(Ze(o.name || o), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: i(() => [
            Ue(Ze(U.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: i(() => [
            _("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (o) => te(e.value.userOptions.callbacks.csv)),
              style: E({ cursor: P.value ? "pointer" : "default" })
            }, [
              D(a(re), {
                name: "fileCsv",
                stroke: U.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : k("", !0),
      r(t.$slots, "skeleton", {}, () => [
        a(X) ? (d(), T(Yt, { key: 0 })) : k("", !0)
      ], !0)
    ], 46, so));
  }
}, No = /* @__PURE__ */ ao(Co, [["__scopeId", "data-v-b0473b8f"]]);
export {
  No as default
};
