// vue-data-ui/dist/vue-ui-spark-trend-C68p8-Xp.js
import { defineAsyncComponent as ge, ref as f, computed as s, toRefs as me, watch as G, nextTick as Q, onMounted as ke, openBlock as o, createElementBlock as r, normalizeStyle as be, unref as i, createVNode as _e, renderSlot as R, createCommentVNode as y, createElementVNode as z, toDisplayString as Y, createBlock as we } from "vue";
import { j as $e, t as ee, H as te, o as Z, q as xe, K as $, am as Ce, X as Le, u as H, c as le, l as ae, k as Ae } from "./lib-GwPnJSrY.js";
import { t as Te, u as Ne } from "./useResponsive-DfdjqQps.js";
import { u as Se } from "./useConfig-CIgyKZ6R.js";
import { u as ze, B as Fe } from "./BaseScanner-BhhkKVoj.js";
import { u as Ie } from "./useFitSvgText-2crs6Fv5.js";
import { u as W } from "./useNestedProp-BBT0mbHh.js";
import { u as Me } from "./useThemeCheck-D1ZYXUwG.js";
import je from "./vue_ui_spark_trend-3BjrSprO.js";
import { _ as Ee } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Ve = ["id"], Be = ["xmlns", "viewBox"], De = ["width", "height"], Oe = ["id"], Ge = ["stop-color"], Re = ["stop-color"], Ze = { key: 1 }, He = ["d", "fill"], We = ["d", "fill"], qe = ["d", "stroke", "stroke-width", "stroke-linecap", "stroke-linejoin"], Ue = ["d", "stroke", "stroke-width", "stroke-linecap", "stroke-linejoin"], Pe = ["d"], Ke = ["fill", "d"], Xe = ["x", "y", "width", "height"], Je = ["x", "y", "fill", "font-size", "font-weight"], Qe = ["stroke", "cx", "cy", "fill"], Ye = ["x", "y", "font-size", "fill", "font-weight"], et = {
  __name: "vue-ui-spark-trend",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  setup(ne) {
    const ue = ge(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_spark_trend: se } = Se(), { isThemeValid: oe, warnInvalidTheme: re } = Me(), h = ne, F = f(null), q = f(null), U = f(null), x = f(null), L = f(null), A = f(!1), p = f(null), C = f($e()), e = f(j()), ie = s(() => ee({
      defaultConfig: {
        style: {
          animation: { show: !1 },
          backgroundColor: "#99999930",
          line: {
            stroke: "#6A6A6A",
            useColorTrend: !1
          },
          dataLabel: { show: !1, useColorTrend: !1, color: "#6A6A6A" }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: _, FINAL_DATASET: I, manualLoading: M } = ze({
      ...me(h),
      FINAL_CONFIG: e,
      prepareConfig: j,
      skeletonDataset: h.config?.skeletonDataset ?? [1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233],
      skeletonConfig: ee({
        defaultConfig: e.value,
        userConfig: ie.value
      })
    });
    function j() {
      const t = W({
        userConfig: h.config,
        defaultConfig: se
      }), l = t.theme;
      if (!l) return t;
      if (!oe.value(t))
        return re(t), t;
      const u = W({
        userConfig: je[l] || h.config,
        defaultConfig: t
      });
      return W({
        userConfig: h.config,
        defaultConfig: u
      });
    }
    const g = s(() => te({
      data: I.value,
      threshold: e.value.downsample.threshold
    }));
    G(() => h.config, (t) => {
      e.value = j(), B.value = e.value.style.width, w.value = e.value.style.height, P();
    }, { deep: !0 });
    function E(t) {
      return t.map((l) => $(l));
    }
    const b = f(te({
      data: I.value,
      threshold: e.value.downsample.threshold
    }).map((t) => e.value.style.animation.show || [void 0, 1 / 0, -1 / 0, null, NaN].includes(t) ? null : t));
    G(g, (t) => {
      p.value && (cancelAnimationFrame(p.value), p.value = null), e.value.style.animation.show ? b.value = Array(t.length).fill(null) : b.value = t.map(
        (l) => ![void 0, 1 / 0, -1 / 0, null].includes(l) && !Number.isNaN(l) ? l : null
      ), V(), Q(() => S(".vue-ui-sparktrend-progress-label", 6));
    }, { deep: !0, immediate: !0 }), G(
      () => JSON.stringify(h.dataset),
      () => {
        p.value && (cancelAnimationFrame(p.value), p.value = null), M.value = Z(h.dataset);
        const t = g.value;
        b.value = e.value.style.animation.show ? Array(t.length).fill(null) : t.map((l) => Number.isFinite(l) ? l : null), V(), Q(() => S(".vue-ui-sparktrend-progress-label", 6));
      },
      { deep: !1, immediate: !0 }
    );
    function V() {
      let l = 1e3 / e.value.style.animation.animationFrames, u = performance.now();
      if (!_.value && e.value.style.animation.show && e.value.style.animation.animationFrames && I.value.length > 1) {
        let k = function() {
          A.value = !0;
          let X = performance.now(), J = X - u;
          J > l ? (u = X - J % l, c < g.value.length ? (b.value.push(g.value[c]), c += 1, p.value = requestAnimationFrame(k)) : (cancelAnimationFrame(p.value), b.value = E(g.value), A.value = !1, S(".vue-ui-sparktrend-progress-label", 6))) : p.value = requestAnimationFrame(k);
        };
        b.value = [];
        let c = 0;
        k();
      }
    }
    ke(() => {
      P();
    });
    const ve = s(() => e.value.debug);
    function P() {
      if (Z(h.dataset) && (xe({
        componentName: "VueUiSparkTrend",
        type: "dataset",
        debug: ve.value
      }), M.value = !0), Z(h.dataset) || (M.value = !1), e.value.responsive) {
        const t = Te(() => {
          const { width: l, height: u } = Ne({
            chart: F.value,
            source: U.value
          });
          requestAnimationFrame(() => {
            B.value = l, w.value = u;
          });
        });
        x.value && (L.value && x.value.unobserve(L.value), x.value.disconnect()), x.value = new ResizeObserver(t), L.value = F.value.parentNode, x.value.observe(L.value);
      }
      V(), S(".vue-ui-sparktrend-progress-label", 6);
    }
    const B = f(e.value.style.width), w = f(e.value.style.height), v = s(() => ({
      height: w.value,
      width: B.value
    })), n = s(() => ({
      top: e.value.style.padding.top,
      left: e.value.style.padding.left,
      right: v.value.width - e.value.style.padding.right,
      bottom: v.value.height - e.value.style.padding.bottom,
      height: v.value.height - (e.value.style.padding.top + e.value.style.padding.bottom) - (e.value.style.dataLabel.show ? e.value.style.dataLabel.fontSize : 0),
      width: v.value.width - (e.value.style.padding.left + e.value.style.padding.right)
    })), D = s(() => {
      const t = E(g.value).filter(Number.isFinite);
      if (!t.length) return { max: 0, min: 0 };
      let l = t[0], u = t[0];
      for (let c = 1; c < t.length; c++) {
        const k = t[c];
        k > l && (l = k), k < u && (u = k);
      }
      return { max: l, min: u };
    }), de = s(() => {
      const t = g.value, l = t.length ? t[t.length - 1] : "x";
      return [
        C.value,
        t.length,
        Number.isFinite(l) ? l : "x",
        e.value.downsample.threshold,
        e.value.style.line.smooth ? "s" : "l"
      ].join("-");
    }), T = s(() => {
      const t = D.value.min >= 0 ? 0 : D.value.min;
      return Math.abs(t);
    }), fe = s(() => D.value.max + T.value);
    function K(t) {
      return t / fe.value;
    }
    const ce = s(() => g.value.length), a = s(() => b.value.map((t, l) => {
      const u = isNaN(t) || [void 0, null, "NaN", NaN, 1 / 0, -1 / 0].includes(t) ? 0 : t || 0;
      return {
        value: $(t),
        absoluteValue: $(u),
        plotValue: $(u + T.value),
        toMax: K(u + T.value),
        x: n.value.left + $(l * (n.value.width / (ce.value - 1))) - e.value.style.padding.right,
        y: n.value.bottom - $(n.value.height * K(u + T.value))
      };
    })), N = s(() => {
      const t = E(g.value);
      return e.value.style.trendLabel.trendType === "global" ? Ce(t) : e.value.style.trendLabel.trendType === "n-1" && t.length > 1 ? (t.at(-1) / t.at(-2) - 1) * 100 : e.value.style.trendLabel.trendType === "lastToFirst" ? (t.at(-1) / t[0] - 1) * 100 : 0;
    }), O = s(() => A.value || N.value === 0 ? "neutral" : N.value > 0 ? "positive" : "negative"), m = s(() => e.value.style.arrow.colors[O.value]), he = s(() => {
      const t = { x: a.value[0].x, y: v.value.height - 6 }, l = { x: a.value[a.value.length - 1].x, y: v.value.height - 6 }, u = [];
      return a.value.forEach((c) => {
        u.push(`${c.x},${c.y} `);
      }), [t.x, t.y, ...u, l.x, l.y].toString();
    }), ye = s(() => {
      let t = [];
      return a.value.forEach((l) => {
        t.push(`${l.x},${l.y} `);
      }), `M ${t.toString()}`;
    }), d = s(() => w.value / 2 - e.value.style.trendLabel.fontSize), pe = s(() => n.value.left * 0.8), { fitText: S } = Ie({
      svgRef: q,
      unitWidth: pe,
      fontSize: e.value.style.trendLabel.fontSize
    });
    return (t, l) => (o(), r("div", {
      ref_key: "sparkTrendChart",
      ref: F,
      class: "vue-data-ui-component vue-ui-spark-trend",
      id: C.value,
      style: be(`width:100%;font-family:${e.value.style.fontFamily};background:${e.value.style.backgroundColor}`)
    }, [
      (o(), r("svg", {
        key: de.value,
        ref_key: "svgRef",
        ref: q,
        xmlns: i(Le),
        viewBox: `0 0 ${v.value.width} ${v.value.height}`,
        style: "width:100%;background:transparent;overflow:visible"
      }, [
        _e(i(ue)),
        t.$slots["chart-background"] ? (o(), r("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: v.value.width <= 0 ? 10 : v.value.width,
          height: v.value.height <= 0 ? 10 : v.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          R(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, De)) : y("", !0),
        z("defs", null, [
          z("linearGradient", {
            x1: "0%",
            y1: "0%",
            x2: "0%",
            y2: "100%",
            id: `pill_gradient_${C.value}`
          }, [
            z("stop", {
              offset: "0%",
              "stop-color": i(H)(e.value.style.line.useColorTrend ? m.value : e.value.style.line.stroke, e.value.style.area.opacity)
            }, null, 8, Ge),
            z("stop", {
              offset: "100%",
              "stop-color": e.value.style.backgroundColor
            }, null, 8, Re)
          ], 8, Oe)
        ]),
        e.value.style.area.show && a.value[0] ? (o(), r("g", Ze, [
          e.value.style.line.smooth ? (o(), r("path", {
            key: 0,
            d: `M ${a.value[0].x},${n.value.bottom} ${i(le)(a.value)} L ${a.value.at(-1).x},${n.value.bottom} Z`,
            fill: e.value.style.area.useGradient ? `url(#pill_gradient_${C.value})` : i(H)(e.value.style.line.useColorTrend ? m.value : e.value.style.line.stroke, e.value.style.area.opacity),
            stroke: "none"
          }, null, 8, He)) : (o(), r("path", {
            key: 1,
            d: `M${he.value}Z`,
            fill: e.value.style.area.useGradient ? `url(#pill_gradient_${C.value})` : i(H)(e.value.style.line.useColorTrend ? m.value : e.value.style.line.stroke, e.value.style.area.opacity),
            stroke: "none"
          }, null, 8, We))
        ])) : y("", !0),
        e.value.style.line.smooth && a.value.length ? (o(), r("path", {
          key: 2,
          d: `M ${i(le)(a.value)}`,
          stroke: e.value.style.line.useColorTrend ? m.value : e.value.style.line.stroke,
          fill: "none",
          "stroke-width": e.value.style.line.strokeWidth,
          "stroke-linecap": e.value.style.line.strokeLinecap,
          "stroke-linejoin": e.value.style.line.strokeLinejoin
        }, null, 8, qe)) : y("", !0),
        !e.value.style.line.smooth && a.value.length ? (o(), r("path", {
          key: 3,
          d: ye.value,
          stroke: e.value.style.line.useColorTrend ? m.value : e.value.style.line.stroke,
          fill: "none",
          "stroke-width": e.value.style.line.strokeWidth,
          "stroke-linecap": e.value.style.line.strokeLinecap,
          "stroke-linejoin": e.value.style.line.strokeLinejoin
        }, null, 8, Ue)) : y("", !0),
        i(_) ? (o(), r("path", {
          key: 4,
          fill: "#6A6A6A",
          d: `M ${n.value.left / 2 + 6}, ${d.value + 7} ${n.value.left / 2 - 7}, ${d.value} ${n.value.left / 2 - 7}, ${d.value + 14} Z`
        }, null, 8, Pe)) : (o(), r("path", {
          key: 5,
          fill: m.value,
          d: O.value === "positive" ? `M ${n.value.left / 2}, ${d.value} ${n.value.left / 2 - 7}, ${d.value + 12} ${n.value.left / 2 + 7}, ${d.value + 12} Z` : O.value === "negative" ? `M ${n.value.left / 2}, ${d.value + 12} ${n.value.left / 2 - 7}, ${d.value} ${n.value.left / 2 + 7}, ${d.value} Z` : `M ${n.value.left / 2 + 6}, ${d.value + 7} ${n.value.left / 2 - 7}, ${d.value} ${n.value.left / 2 - 7}, ${d.value + 14} Z`
        }, null, 8, Ke)),
        i(_) ? (o(), r("rect", {
          key: 6,
          x: n.value.left / 2 - e.value.style.trendLabel.fontSize - 2,
          y: w.value / 2 + e.value.style.trendLabel.fontSize - 2,
          width: e.value.style.trendLabel.fontSize * 2,
          height: e.value.style.trendLabel.fontSize,
          fill: "#6A6A6A80",
          rx: "3"
        }, null, 8, Xe)) : y("", !0),
        !A.value && !i(_) ? (o(), r("text", {
          key: 7,
          class: "vue-ui-sparktrend-progress-label",
          x: n.value.left / 2,
          y: w.value / 2 + e.value.style.trendLabel.fontSize * 2,
          "text-anchor": "middle",
          fill: e.value.style.trendLabel.useColorTrend ? m.value : e.value.style.trendLabel.color,
          "font-size": e.value.style.trendLabel.fontSize,
          "font-weight": e.value.style.trendLabel.bold ? "bold" : "normal"
        }, Y(i(ae)({
          p: N.value > 0 ? "+" : "",
          v: N.value,
          s: "%",
          r: e.value.style.trendLabel.rounding
        })), 9, Je)) : y("", !0),
        a.value.length && a.value.at(-1).x !== void 0 ? (o(), r("circle", {
          key: 8,
          stroke: e.value.style.backgroundColor,
          "stroke-width": 2,
          cx: a.value.at(-1).x,
          cy: a.value.at(-1).y,
          r: 4,
          fill: i(_) ? "#6A6A6A" : m.value
        }, null, 8, Qe)) : y("", !0),
        a.value.length && a.value.at(-1).x !== void 0 && e.value.style.dataLabel.show ? (o(), r("text", {
          key: 9,
          "text-anchor": "middle",
          x: a.value.at(-1).x,
          y: a.value.at(-1).y - e.value.style.dataLabel.fontSize / 1.5,
          "font-size": e.value.style.dataLabel.fontSize,
          fill: e.value.style.dataLabel.useColorTrend ? m.value : e.value.style.dataLabel.color,
          "font-weight": e.value.style.dataLabel.bold ? "bold" : "normal"
        }, Y(i(Ae)(
          e.value.style.dataLabel.formatter,
          a.value.at(-1).value,
          i(ae)({
            p: e.value.style.dataLabel.prefix,
            v: a.value.at(-1).value,
            s: e.value.style.dataLabel.suffix,
            r: e.value.style.dataLabel.rounding
          }),
          { datapoint: a.value.at(-1) }
        )), 9, Ye)) : y("", !0)
      ], 8, Be)),
      t.$slots.source ? (o(), r("div", {
        key: 0,
        ref_key: "source",
        ref: U,
        dir: "auto"
      }, [
        R(t.$slots, "source", {}, void 0, !0)
      ], 512)) : y("", !0),
      R(t.$slots, "skeleton", {}, () => [
        i(_) ? (o(), we(Fe, { key: 0 })) : y("", !0)
      ], !0)
    ], 12, Ve));
  }
}, dt = /* @__PURE__ */ Ee(et, [["__scopeId", "data-v-4ee999e7"]]);
export {
  dt as default
};
