// vue-data-ui/dist/vue-ui-word-cloud-DJmJ-oYT.js
import { defineAsyncComponent as le, computed as G, ref as I, toRefs as Vt, watch as $e, shallowRef as it, onMounted as _t, onBeforeUnmount as Ut, openBlock as Y, createElementBlock as K, unref as C, normalizeStyle as Ae, createBlock as se, withCtx as A, renderSlot as P, normalizeProps as V, guardReactiveProps as _, createCommentVNode as L, createSlots as ut, normalizeClass as rt, createVNode as Be, createElementVNode as Xe, Fragment as Lt, renderList as qt, toDisplayString as Ve, resolveDynamicComponent as Ht, mergeProps as jt, createTextVNode as ct, nextTick as _e } from "vue";
import { d as dt, u as Gt } from "./useResponsive-DfdjqQps.js";
import { j as Kt, t as vt, e as Jt, K as Qt, m as eo, p as ze, o as mt, q as ft, r as to, l as oo, X as no, a0 as ao, a1 as lo, T as so } from "./lib-GwPnJSrY.js";
import { u as io } from "./useConfig-CIgyKZ6R.js";
import { u as uo, B as ro } from "./BaseScanner-BhhkKVoj.js";
import { u as co } from "./usePrinter-DiK5R6k-.js";
import { u as vo } from "./useSvgExport-ByUukOZt.js";
import { u as Ue } from "./useNestedProp-BBT0mbHh.js";
import { u as mo } from "./useThemeCheck-D1ZYXUwG.js";
import { u as fo } from "./useUserOptionState-BIvW1Kz7.js";
import { u as ho } from "./useChartAccessibility-9icAAmYg.js";
import po from "./img-BYYzY_QU.js";
import go from "./Title-BfyrP-jp.js";
import bo from "./vue_ui_word_cloud-mKAtBxHR.js";
import { u as yo } from "./usePanZoom-CDF7rb3o.js";
import { B as ht } from "./BaseZoomControls-CoFievy9.js";
import { _ as wo } from "./_plugin-vue_export-helper-CHgC5LLL.js";
function xo({
  word: o,
  fontSize: s,
  pad: m,
  canvas: c,
  ctx: r,
  svg: v
}) {
  const n = `${v.style && v.style.bold ? "bold " : ""}${s}px Arial`;
  r.font = n;
  const h = r.measureText(o.name), f = Math.ceil(h.width) + 2 + (m ? m * 2 : 0), p = Math.ceil(s) + 2 + (m ? m * 2 : 0);
  c.width = f, c.height = p, r.font = n, r.textAlign = "center", r.textBaseline = "middle", r.fillStyle = "black", r.fillText(o.name, f / 2, p / 2);
  const g = r.getImageData(0, 0, f, p).data, w = [], l = [];
  let u = f, $ = p, F = 0, k = 0, d = !1;
  for (let x = 0; x < p; x += 1) {
    const Z = x * f * 4;
    let W = -1, D = !1;
    for (let X = 0; X < f; X += 1) {
      const B = Z + X * 4 + 3;
      g[B] > 1 ? (w.push([X, x]), d = !0, X < u && (u = X), X > F && (F = X), x < $ && ($ = x), x > k && (k = x), D || (D = !0, W = X)) : D && (l.push([x, W, X - 1]), D = !1, W = -1);
    }
    D && l.push([x, W, f - 1]);
  }
  return d || (u = 0, $ = 0, F = 0, k = 0), {
    w: f,
    h: p,
    wordMask: w,
    runs: l,
    minX: u,
    minY: $,
    maxX: F,
    maxY: k
  };
}
function Co(o) {
  const s = [];
  if (!o.length) return s;
  let m = o[0][1], c = o[0][0], r = c;
  for (let v = 1; v < o.length; v += 1) {
    const i = o[v][0], n = o[v][1];
    n !== m ? (s.push([m, c, r]), m = n, c = i, r = i) : (i === r + 1 || (s.push([m, c, r]), c = i), r = i);
  }
  return s.push([m, c, r]), s;
}
function yt(o, s) {
  return 4294967295 >>> 32 - (s - o + 1) << o >>> 0;
}
function wt(o) {
  return 4294967295 << o >>> 0;
}
function xt(o) {
  return 4294967295 >>> 31 - o >>> 0;
}
function ko({
  maskBits: o,
  maskRowStride: s,
  maskW: m,
  maskH: c,
  wx: r,
  wy: v,
  runs: i
}) {
  for (let n = 0; n < i.length; n += 1) {
    const h = i[n][0], f = i[n][1], p = i[n][2], b = v + h;
    if (b < 0 || b >= c) return !1;
    const g = r + f, w = r + p;
    if (g < 0 || w >= m) return !1;
    const l = b * s, u = g >>> 5, $ = w >>> 5, F = g & 31, k = w & 31;
    if (u === $) {
      const d = o[l + u], x = yt(F, k);
      if (d & x) return !1;
    } else {
      {
        const d = o[l + u], x = wt(F);
        if (d & x) return !1;
      }
      for (let d = u + 1; d < $; d += 1)
        if (o[l + d]) return !1;
      {
        const d = o[l + $], x = xt(k);
        if (d & x) return !1;
      }
    }
  }
  return !0;
}
function Fo({
  maskBits: o,
  maskRowStride: s,
  maskW: m,
  maskH: c,
  wx: r,
  wy: v,
  runs: i
}) {
  for (let n = 0; n < i.length; n += 1) {
    const h = i[n][0], f = i[n][1], p = i[n][2], b = v + h;
    if (b < 0 || b >= c) continue;
    const g = r + f, w = r + p;
    if (w < 0 || g >= m) continue;
    const l = b * s, u = g >>> 5, $ = w >>> 5, F = g & 31, k = w & 31;
    if (u === $) {
      const d = yt(F, k);
      o[l + u] |= d;
    } else {
      {
        const d = wt(F);
        o[l + u] |= d;
      }
      for (let d = u + 1; d < $; d += 1)
        o[l + d] = 4294967295;
      {
        const d = xt(k);
        o[l + $] |= d;
      }
    }
  }
}
function So({ wordMask: o, w: s, h: m, dilation: c }) {
  const r = new Uint8Array(s * m), v = [];
  for (let n = 0; n < o.length; n += 1) {
    const h = o[n][0], p = o[n][1] * s + h;
    r[p] || (r[p] = 1, v.push(p));
  }
  for (let n = 0; n < v.length; n += 1) {
    const h = v[n], f = h / s | 0, p = h - f * s;
    for (let b = -c; b <= c; b += 1) {
      const g = f + b;
      if (g < 0 || g >= m) continue;
      const w = g * s;
      for (let l = -c; l <= c; l += 1) {
        if (l === 0 && b === 0) continue;
        const u = p + l;
        u < 0 || u >= s || (r[w + u] = 1);
      }
    }
  }
  const i = [];
  for (let n = 0; n < r.length; n += 1)
    if (r[n]) {
      const h = n / s | 0, f = n - h * s;
      i.push([f, h]);
    }
  return i;
}
function $o({
  runs: o,
  w: s,
  h: m,
  dilation: c
}) {
  const r = new Uint8Array(s * m), v = [];
  for (let n = 0; n < o.length; n += 1) {
    const h = o[n], f = h[0], p = h[1], b = h[2], g = f * s;
    for (let w = p; w <= b; w += 1) {
      const l = g + w;
      r[l] || (r[l] = 1, v.push(l));
    }
  }
  for (let n = 0; n < v.length; n += 1) {
    const h = v[n], f = h / s | 0, p = h - f * s;
    for (let b = -c; b <= c; b += 1) {
      const g = f + b;
      if (g < 0 || g >= m) continue;
      const w = g * s;
      for (let l = -c; l <= c; l += 1) {
        if (l === 0 && b === 0) continue;
        const u = p + l;
        u < 0 || u >= s || (r[w + u] = 1);
      }
    }
  }
  const i = [];
  for (let n = 0; n < m; n += 1) {
    const h = n * s;
    let f = -1, p = !1;
    for (let b = 0; b < s; b += 1)
      r[h + b] ? p || (p = !0, f = b) : p && (i.push([n, f, b - 1]), p = !1, f = -1);
    p && i.push([n, f, s - 1]);
  }
  return i;
}
function To(o, s) {
  const m = o.runs, c = o.w, r = o.h, v = Math.max(1, Math.round(c * s)), i = Math.max(1, Math.round(r * s)), n = new Array(i);
  let h = v, f = i, p = 0, b = 0, g = !1;
  for (let l = 0; l < m.length; l += 1) {
    const u = m[l], $ = u[0], F = u[1], k = u[2], d = Math.round($ * s);
    if (d < 0 || d >= i) continue;
    const x = Math.round(F * s), Z = Math.round((k + 1) * s) - 1;
    if (Z < x) continue;
    let W = n[d];
    W || (W = [], n[d] = W), W.push([x, Z]), g = !0, x < h && (h = x), Z > p && (p = Z), d < f && (f = d), d > b && (b = d);
  }
  if (!g)
    return {
      w: v,
      h: i,
      runs: [],
      minX: 0,
      minY: 0,
      maxX: 0,
      maxY: 0
    };
  const w = [];
  for (let l = 0; l < i; l += 1) {
    const u = n[l];
    if (!u || u.length === 0) continue;
    u.sort((k, d) => k[0] - d[0]);
    let $ = u[0][0], F = u[0][1];
    for (let k = 1; k < u.length; k += 1) {
      const d = u[k][0], x = u[k][1];
      d <= F + 1 ? x > F && (F = x) : (w.push([l, $, F]), $ = d, F = x);
    }
    w.push([l, $, F]);
  }
  return {
    w: v,
    h: i,
    runs: w,
    minX: h,
    minY: f,
    maxX: p,
    maxY: b
  };
}
const pt = /* @__PURE__ */ new Map(), gt = /* @__PURE__ */ new Map();
function Mo({ word: o, fontSize: s, pad: m, svg: c }) {
  const r = c.style && c.style.bold ? 1 : 0, v = m || 0;
  return `${o.name}::${s}::${v}::${r}`;
}
function Io({
  word: o,
  fontSize: s,
  pad: m,
  canvas: c,
  ctx: r,
  svg: v
}) {
  const i = Mo({ word: o, fontSize: s, pad: m, svg: v }), n = pt.get(i);
  if (n) return { key: i, bitmap: n };
  const h = xo({ word: o, fontSize: s, pad: m, canvas: c, ctx: r, svg: v });
  return pt.set(i, h), { key: i, bitmap: h };
}
function Po({
  bitmapKey: o,
  wordMask: s,
  w: m,
  h: c,
  dilation: r
}) {
  const v = `${o}::d${r}`, i = gt.get(v);
  if (i) return i;
  const n = So({ wordMask: s, w: m, h: c, dilation: r }), h = Co(n), f = { wordMask: n, runs: h };
  return gt.set(v, f), f;
}
const Oo = 6, Ao = 2, Ct = Math.PI / 180, kt = [], Ft = [];
for (let o = 0; o < 360; o += Oo) {
  const s = o * Ct;
  kt.push(Math.cos(s)), Ft.push(Math.sin(s));
}
const St = [], $t = [];
for (let o = 0; o < 360; o += Ao) {
  const s = o * Ct;
  St.push(Math.cos(s)), $t.push(Math.sin(s));
}
function Xo(o, s, m) {
  if (!o.length) return;
  let c = 1 / 0, r = -1 / 0, v = 1 / 0, i = -1 / 0;
  for (let l = 0; l < o.length; l += 1) {
    const u = o[l], $ = u.x + u.minX, F = u.x + u.maxX, k = u.y + u.minY, d = u.y + u.maxY;
    $ < c && (c = $), F > r && (r = F), k < v && (v = k), d > i && (i = d);
  }
  if (!isFinite(c) || !isFinite(r) || !isFinite(v) || !isFinite(i)) return;
  const n = Math.max(
    Math.abs(c),
    Math.abs(r)
  ), h = Math.max(
    Math.abs(v),
    Math.abs(i)
  );
  if (n === 0 || h === 0) return;
  const f = 0.9, p = s * 0.5 * f / n, b = m * 0.5 * f / h;
  let g = Math.min(p, b);
  if (g <= 1) return;
  const w = 4;
  g > w && (g = w);
  for (let l = 0; l < o.length; l += 1) {
    const u = o[l];
    u.x *= g, u.y *= g, u.width *= g, u.height *= g, u.fontSize *= g, u.minX *= g, u.maxX *= g, u.minY *= g, u.maxY *= g;
  }
}
function zo() {
  return typeof performance < "u" && typeof performance.now == "function" ? () => performance.now() : () => Date.now();
}
function Ro({
  value: o,
  minimumValue: s,
  maximumValue: m,
  configuredMinimumFontSize: c,
  maximumFontSize: r
}) {
  if (m === s)
    return r;
  const i = (o - s) / (m - s) * (r - c) + c;
  return Math.max(
    c,
    Math.min(r, i)
  );
}
function Yo({
  currentBitmap: o,
  strictPixelPadding: s,
  scaleFactor: m,
  baseBitmap: c,
  bitmapKey: r
}) {
  let v = o.runs;
  const i = o.w, n = o.h, h = o.minX, f = o.minY, p = o.maxX, b = o.maxY;
  return !s || !v.length ? {
    runs: v,
    bitmapWidth: i,
    bitmapHeight: n,
    bitmapMinimumX: h,
    bitmapMinimumY: f,
    bitmapMaximumX: p,
    bitmapMaximumY: b
  } : m === 1 ? {
    runs: Po({
      bitmapKey: r,
      wordMask: c.wordMask,
      w: c.w,
      h: c.h,
      dilation: 2
    }).runs,
    bitmapWidth: i,
    bitmapHeight: n,
    bitmapMinimumX: h,
    bitmapMinimumY: f,
    bitmapMaximumX: p,
    bitmapMaximumY: b
  } : {
    runs: $o({
      runs: v,
      w: i,
      h: n,
      dilation: 2
    }),
    bitmapWidth: i,
    bitmapHeight: n,
    bitmapMinimumX: h,
    bitmapMinimumY: f,
    bitmapMaximumX: p,
    bitmapMaximumY: b
  };
}
async function bt({
  baseBitmap: o,
  baseFontSize: s,
  minimumScaleFactor: m,
  maskBits: c,
  maskRowStride: r,
  maskWidth: v,
  maskHeight: i,
  centerX: n,
  centerY: h,
  maximumRadius: f,
  scaleStep: p,
  strictPixelPadding: b,
  bitmapKey: g,
  minimumFontSize: w,
  rawWord: l,
  cosineArray: u,
  sineArray: $,
  radiusStep: F,
  maximumAttempts: k,
  maybeYield: d
}) {
  let x = 1;
  for (; x >= m; ) {
    const Z = x === 1 ? o : To(o, x), W = Yo({
      currentBitmap: Z,
      strictPixelPadding: b,
      scaleFactor: x,
      baseBitmap: o,
      bitmapKey: g
    }), {
      runs: D,
      bitmapWidth: X,
      bitmapHeight: B,
      bitmapMinimumX: Q,
      bitmapMinimumY: ie,
      bitmapMaximumX: ue,
      bitmapMaximumY: re
    } = W;
    let H = 0, q = 0;
    for (; H < f && q < k; ) {
      for (let e = 0; e < u.length; e += 1) {
        q += 1;
        const N = Math.round(
          n + H * u[e] - X / 2
        ), ee = Math.round(
          h + H * $[e] - B / 2
        );
        if (N < 0 || ee < 0 || N + X > v || ee + B > i)
          continue;
        if (ko({
          maskBits: c,
          maskRowStride: r,
          maskW: v,
          maskH: i,
          wx: N,
          wy: ee,
          runs: D
        })) {
          const { __wcIndex: z, ...E } = l, te = Math.max(
            w,
            Math.round(s * x)
          ), pe = {
            ...E,
            x: N - v / 2,
            y: ee - i / 2,
            fontSize: te,
            width: X,
            height: B,
            angle: 0,
            minX: Q,
            minY: ie,
            maxX: ue,
            maxY: re
          };
          return Fo({
            maskBits: c,
            maskRowStride: r,
            maskW: v,
            maskH: i,
            wx: N,
            wy: ee,
            runs: D
          }), pe;
        }
      }
      H += F, (q & 1023) === 0 && await d();
    }
    x *= p, await d();
  }
  return null;
}
async function Wo({
  baseBitmap: o,
  baseFontSize: s,
  minimumScaleFactor: m,
  maskBits: c,
  maskRowStride: r,
  maskWidth: v,
  maskHeight: i,
  centerX: n,
  centerY: h,
  maximumRadius: f,
  scaleStep: p,
  strictPixelPadding: b,
  bitmapKey: g,
  minimumFontSize: w,
  rawWord: l,
  maybeYield: u,
  spiralRadiusStep: $,
  fallbackSpiralRadiusStep: F
}) {
  const k = await bt({
    baseBitmap: o,
    baseFontSize: s,
    minimumScaleFactor: m,
    maskBits: c,
    maskRowStride: r,
    maskWidth: v,
    maskHeight: i,
    centerX: n,
    centerY: h,
    maximumRadius: f,
    scaleStep: p,
    strictPixelPadding: b,
    bitmapKey: g,
    minimumFontSize: w,
    rawWord: l,
    cosineArray: kt,
    sineArray: Ft,
    radiusStep: $,
    maximumAttempts: 1e4,
    maybeYield: u
  });
  return k || await bt({
    baseBitmap: o,
    baseFontSize: s,
    minimumScaleFactor: m,
    maskBits: c,
    maskRowStride: r,
    maskWidth: v,
    maskHeight: i,
    centerX: n,
    centerY: h,
    maximumRadius: f,
    scaleStep: p,
    strictPixelPadding: b,
    bitmapKey: g,
    minimumFontSize: w,
    rawWord: l,
    cosineArray: St,
    sineArray: $t,
    radiusStep: F,
    maximumAttempts: 25e3,
    maybeYield: u
  });
}
async function Do({
  words: o,
  proximity: s = 0,
  svg: m,
  strictPixelPadding: c,
  onProgress: r,
  debugTiming: v = !1
}) {
  const i = zo(), n = i(), h = 12;
  let f = n;
  async function p() {
    i() - f >= h && (await new Promise((E) => setTimeout(E, 0)), f = i());
  }
  const b = m.width, g = m.height, w = Math.round(b), l = Math.round(g), u = 1, $ = m.minFontSize, F = Math.min(m.maxFontSize, 100), k = o.map((z) => z.value), d = Math.min(...k), x = Math.max(...k);
  if (w <= 0 || l <= 0) return [];
  const Z = w + 31 >>> 5, W = new Uint32Array(Z * l), D = document.createElement("canvas"), X = D.getContext("2d", { willReadFrequently: !0 });
  D.width = w, D.height = l;
  const B = 2, Q = 1, ie = Math.max(w, l), ue = Math.floor(w / 2), re = Math.floor(l / 2), q = [...o.map((z, E) => ({
    ...z,
    __wcIndex: E,
    id: z.id != null ? z.id : `${z.name}__${E}`
  }))].sort(
    (z, E) => E.value - z.value
  ), e = [], N = 0.9;
  for (let z = 0; z < q.length; z += 1) {
    const E = q[z], te = e.length, ce = Ro({
      value: E.value,
      minimumValue: d,
      maximumValue: x,
      configuredMinimumFontSize: $,
      maximumFontSize: F
    }), ge = Io({
      word: E,
      fontSize: ce,
      pad: s,
      canvas: D,
      ctx: X,
      svg: m
    }), Te = ge.key, oe = ge.bitmap, be = oe.w, ye = oe.h;
    if (be <= 0 || ye <= 0) {
      await p();
      continue;
    }
    const we = Math.max(
      u / ce,
      0.1
    ), Me = await Wo({
      baseBitmap: oe,
      baseFontSize: ce,
      minimumScaleFactor: we,
      maskBits: W,
      maskRowStride: Z,
      maskWidth: w,
      maskHeight: l,
      centerX: ue,
      centerY: re,
      maximumRadius: ie,
      scaleStep: N,
      strictPixelPadding: c,
      bitmapKey: Te,
      minimumFontSize: u,
      rawWord: E,
      maybeYield: p,
      spiralRadiusStep: B,
      fallbackSpiralRadiusStep: Q
    });
    if (Me && (e.push(Me), r && e.length > te)) {
      const me = e[e.length - 1];
      r({ word: me, all: e });
    }
    await p();
  }
  if (!e.length)
    return [];
  Xo(e, w, l);
  const ve = i() - n;
  return v && typeof console < "u" && console.log && console.log(
    "[vue-data-ui][word-cloud] positionWordsAsync:",
    `${ve.toFixed(2)} ms for ${o.length} words`
  ), e.sort((z, E) => E.fontSize - z.fontSize);
}
const Eo = ["id", "data-resizing", "data-relayout"], No = ["xmlns", "viewBox"], Zo = ["width", "height"], Bo = ["transform"], Vo = ["transform"], _o = ["x", "y", "width", "height", "onMouseover", "onMouseleave", "onClick"], Uo = ["fill", "font-weight", "font-size", "transform", "stroke", "stroke-width"], Lo = {
  key: 4,
  class: "vue-data-ui-watermark"
}, qo = {
  key: 5,
  "data-dom-to-png-ignore": "",
  class: "reset-wrapper"
}, Ho = ["innerHTML"], jo = {
  __name: "vue-ui-word-cloud",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: [Array, String],
      default() {
        return [];
      }
    }
  },
  emits: ["copyAlt"],
  setup(o, { expose: s, emit: m }) {
    const c = le(() => import("./Tooltip-DL3QgjZf.js")), r = le(() => import("./BaseIcon-CNx8Tzl5.js")), v = le(() => import("./vue-ui-accordion-BGHTqIBt.js")), i = le(() => import("./DataTable-Bo1pmAqG.js")), n = le(() => import("./PenAndPaper-DWV1FMIi.js")), h = le(() => import("./UserOptions-CrLqEimB.js")), f = le(() => import("./PackageVersion-DN9d0ME5.js")), p = le(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_word_cloud: b } = io(), { isThemeValid: g, warnInvalidTheme: w } = mo(), l = o, u = m, $ = G({
      get() {
        return !!l.dataset && l.dataset.length;
      },
      set(t) {
        return t;
      }
    }), F = I(Kt()), k = I(0), d = I(null), x = I(null), Z = I(null), W = I(0), D = I(0), X = I(!1), B = I(null), Q = I(null), ie = I(!1), ue = I(!1), re = I(null), H = I(!1), q = I(!1), e = I(be()), N = G(() => e.value.userOptions.useCursorPointer), ee = G(() => vt({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        useCssAnimation: !1,
        animationDelayMs: 0,
        nodeCategories: {},
        nodeCategoryColors: {},
        style: {
          chart: {
            backgroundColor: "#99999930",
            words: {
              color: "#6A6A6A",
              usePalette: !1,
              selectedStroke: "#CCCCCC"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: ve, FINAL_DATASET: z, manualLoading: E } = uo({
      ...Vt(l),
      FINAL_CONFIG: e,
      prepareConfig: be,
      callback: () => {
        Promise.resolve().then(() => {
          O.value.showTable = e.value.table.show, O.value.showTooltip = e.value.style.chart.tooltip.show, O.value.showZoom = e.value.style.chart.zoom.show;
        });
      },
      skeletonDataset: l.config?.skeletonDataset ?? [
        { name: "Lorem", value: 6 },
        { name: "ipsum", value: 3 },
        { name: "dolor", value: 1 },
        { name: "sit", value: 3 },
        { name: "amet", value: 3 },
        { name: "consectetur", value: 2 },
        { name: "adipiscing", value: 1 },
        { name: "elit", value: 2 },
        { name: "Vivamus", value: 2 },
        { name: "pulvinar", value: 1 },
        { name: "pretium", value: 1 },
        { name: "venenatis", value: 2 },
        { name: "Donec", value: 1 },
        { name: "imperdiet", value: 3 },
        { name: "id", value: 1 },
        { name: "porttitor", value: 2 },
        { name: "tristique", value: 1 },
        { name: "Aenean", value: 2 },
        { name: "ac", value: 5 },
        { name: "commodo", value: 2 },
        { name: "justo", value: 2 },
        { name: "Vestibulum", value: 2 },
        { name: "placerat", value: 1 },
        { name: "molestie", value: 1 },
        { name: "nisl", value: 1 },
        { name: "lacinia", value: 2 },
        { name: "nulla", value: 1 },
        { name: "posuere", value: 2 },
        { name: "quis", value: 3 },
        { name: "ullamcorper", value: 1 },
        { name: "eu", value: 1 },
        { name: "ex", value: 1 },
        { name: "vitae", value: 3 },
        { name: "facilisis", value: 1 },
        { name: "Aliquam", value: 1 },
        { name: "erat", value: 1 },
        { name: "volutpat", value: 1 },
        { name: "Proin", value: 1 },
        { name: "nunc", value: 1 },
        { name: "felis", value: 1 },
        { name: "gravida", value: 3 },
        { name: "sed", value: 1 },
        { name: "orci", value: 1 },
        { name: "Interdum", value: 1 },
        { name: "et", value: 1 },
        { name: "malesuada", value: 1 },
        { name: "fames", value: 1 },
        { name: "ante", value: 1 }
      ],
      skeletonConfig: vt({
        defaultConfig: e.value,
        userConfig: ee.value
      })
    }), te = I(pe());
    function pe() {
      return typeof z.value == "string" ? Jt(z.value) : z.value.map((t, y) => ({
        ...t,
        value: Qt(t.value)
      }));
    }
    const { userOptionsVisible: ce, setUserOptionsVisibility: ge, keepUserOptionState: Te } = fo({ config: e.value }), { svgRef: oe } = ho({ config: e.value.style.chart.title });
    function be() {
      const t = Ue({
        userConfig: l.config,
        defaultConfig: b
      }), y = t.theme;
      if (!y) return t;
      if (!g.value(t))
        return w(t), t;
      const a = Ue({
        userConfig: bo[y] || l.config,
        defaultConfig: t
      }), S = Ue({
        userConfig: l.config,
        defaultConfig: a
      });
      return {
        ...S,
        customPalette: S.customPalette.length ? S.customPalette : eo[y] || ze
      };
    }
    const ye = I({ x: 0, y: 0 });
    function we() {
      const t = {
        x: 0,
        y: 0,
        width: Math.max(10, R.value.width),
        height: Math.max(10, R.value.height)
      };
      Xt(t), ye.value = {
        x: t.x + t.width / 2,
        y: t.y + t.height / 2
      }, ke();
    }
    const Me = dt(() => {
      je();
    }, 100), me = I(!1);
    $e(() => me.value, (t) => {
      t === !1 && (ue.value = !0, Me(), we());
    }), $e(() => l.config, (t) => {
      e.value = be(), ce.value = !e.value.userOptions.showOnChartHover, Le(), W.value += 1, D.value += 1, O.value.showTable = e.value.table.show, O.value.showTooltip = e.value.style.chart.tooltip.show, O.value.showZoom = e.value.style.chart.zoom.show;
    }, { deep: !0 });
    const R = I({
      width: e.value.style.chart.width,
      height: e.value.style.chart.height,
      maxFontSize: e.value.style.chart.words.maxFontSize,
      minFontSize: e.value.style.chart.words.minFontSize,
      bold: e.value.style.chart.words.bold
    }), Tt = dt(() => {
      const { width: t, height: y } = Gt({
        chart: d.value,
        title: e.value.style.chart.title.text ? x.value : null,
        legend: e.value.style.chart.controls.show ? re.value?.$el : null,
        source: Z.value
      });
      requestAnimationFrame(async () => {
        R.value.width = Math.max(10, t), R.value.height = Math.max(10, y - 12), await _e(), me.value = !1;
      });
    }, 100), Mt = () => {
      me.value = !0, Tt();
    }, ne = it(null), fe = it(null);
    _t(Le);
    const Re = G(() => !!e.value.debug);
    function Le() {
      mt(l.dataset) ? ft({
        componentName: "VueUiWordCloud",
        type: "dataset",
        debug: Re.value
      }) : te.value.forEach((t, y) => {
        to({
          datasetObject: t,
          requiredAttributes: ["name", "value"]
        }).forEach((a) => {
          $.value = !1, ft({
            componentName: "VueUiWordCloud",
            type: "datasetSerieAttribute",
            property: a,
            index: y,
            debug: Re.value
          });
        });
      }), mt(l.dataset) || (E.value = e.value.loading), e.value.responsive && (ne.value && (fe.value && ne.value.unobserve(fe.value), ne.value.disconnect()), ne.value = new ResizeObserver(Mt), fe.value = d.value.parentNode, ne.value.observe(fe.value)), we();
    }
    Ut(() => {
      ne.value && (fe.value && ne.value.unobserve(fe.value), ne.value.disconnect());
    });
    const { isPrinting: Ye, isImaging: We, generatePdf: qe, generateImage: He } = co({
      elementId: `wordCloud_${F.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-word-cloud",
      options: e.value.userOptions.print
    }), O = I({
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show,
      showZoom: e.value.style.chart.zoom.show
    });
    $e(e, () => {
      O.value.showTable = e.value.table.show, O.value.showTooltip = e.value.style.chart.tooltip.show, O.value.showZoom = e.value.style.chart.zoom.show;
    }, { immediate: !0 });
    function It(t, y, a = "Arial") {
      const T = document.createElement("canvas").getContext("2d");
      return T.font = `${y}px ${e.value.style.chart.words.bold ? "bold" : "normal"} ${a}`, {
        width: T.measureText(t).width + e.value.style.chart.words.proximity,
        height: y
      };
    }
    const U = I([]), xe = /* @__PURE__ */ new Map();
    function je() {
      const t = [...te.value].map((T) => T.value), y = Math.max(...t), a = Math.min(...t), S = [...te.value].map((T, M) => {
        let j = (T.value - a) / (y - a) * (R.value.maxFontSize - R.value.minFontSize) + R.value.minFontSize;
        j = isNaN(j) ? R.value.minFontSize : j;
        const de = It(T.name, j);
        return {
          ...T,
          id: T.id ?? `${T.name}__${M}`,
          fontSize: j,
          width: de.width,
          height: de.height,
          color: e.value.style.chart.words.usePalette ? e.value.customPalette[M] || e.value.customPalette[M % e.value.customPalette.length] || ze[M] || ze[M % ze.length] : e.value.style.chart.words.color
        };
      });
      U.value.length = 0, xe.clear(), Do({
        debugTiming: Re.value,
        words: S,
        svg: R.value,
        proximity: e.value.style.chart.words.proximity,
        strictPixelPadding: e.value.strictPixelPadding,
        onProgress: ({ all: T }) => {
          for (const M of T) {
            const j = M.id;
            let de = xe.get(j);
            if (de === void 0)
              de = U.value.length, xe.set(j, de), U.value.push({ ...M });
            else {
              const J = U.value[de];
              J.x = M.x, J.y = M.y, J.width = M.width, J.height = M.height, J.fontSize = M.fontSize, J.minX = M.minX, J.minY = M.minY, J.maxX = M.maxX, J.maxY = M.maxY;
            }
          }
        }
      }), U.value.sort((T, M) => M.fontSize - T.fontSize), xe.clear(), U.value.forEach((T, M) => {
        xe.set(T.id, M);
      }), ie.value = !0, ue.value = !1;
    }
    const Ie = G(() => {
      const t = U.value.map((a) => ({
        name: a.name,
        color: a.color
      })), y = U.value.map((a) => a.value);
      return { head: t, body: y };
    });
    function De(t = null) {
      _e(() => {
        const y = Ie.value.head.map((T, M) => [[
          T.name
        ], [Ie.value.body[M]]]), a = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[""], [e.value.table.columnNames.value]]].concat(y), S = ao(a);
        t ? t(S) : lo({ csvContent: S, title: e.value.style.chart.title.text || "vue-ui-word-cloud" });
      });
    }
    const Pe = G(() => {
      const t = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.value
      ], y = Ie.value.head.map((T, M) => {
        const j = oo({ p: e.value.table.td.prefix, v: Ie.value.body[M], s: e.value.table.td.suffix, r: e.value.table.td.roundingValue });
        return [
          {
            color: T.color,
            name: T.name
          },
          j
        ];
      }), a = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.series,
          e.value.table.columnNames.value
        ],
        head: t,
        body: y,
        config: a
      };
    }), ae = I(!1);
    function Ge(t) {
      ae.value = t, k.value += 1;
    }
    function Pt() {
      return U.value;
    }
    function Ke() {
      O.value.showTable = !O.value.showTable;
    }
    function Je() {
      O.value.showTooltip = !O.value.showTooltip;
    }
    const Ce = I(!1);
    function Ee() {
      Ce.value = !Ce.value;
    }
    function Qe() {
      O.value.showZoom = !O.value.showZoom;
    }
    const Ot = G(() => !Ce.value && O.value.showZoom), { viewBox: Oe, resetZoom: ke, isZoom: At, setInitialViewBox: Xt, zoomByFactor: et, scale: tt } = yo(oe, {
      x: 0,
      y: 0,
      width: R.value.width <= 0 ? 10 : R.value.width,
      height: R.value.height <= 0 ? 10 : R.value.height
    }, 1, Ot);
    $e(() => l.dataset, () => {
      te.value = pe(), e.value.responsive || (je(), we());
    }, { immediate: !0 });
    async function zt({ scale: t = 2 } = {}) {
      if (!d.value) return;
      const { width: y, height: a } = d.value.getBoundingClientRect(), S = y / a, { imageUri: T, base64: M } = await po({ domElement: d.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: T,
        base64: M,
        title: e.value.style.chart.title.text,
        width: y,
        height: a,
        aspectRatio: S
      };
    }
    const Fe = G(() => {
      const t = e.value.table.useDialog && !e.value.table.show, y = O.value.showTable;
      return {
        component: t ? p : v,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: ae.value,
          fullscreenParent: d.value,
          forcedWidth: Math.min(500, window.innerWidth * 0.8),
          isCursorPointer: N.value
        } : {
          hideDetails: !0,
          config: {
            open: y,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    $e(() => O.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && B.value ? B.value.open() : "close" in B.value && B.value.close());
    });
    function ot() {
      O.value.showTable = !1, Q.value && Q.value.setTableIconState(!1);
    }
    const Rt = G(() => e.value.style.chart.backgroundColor), Yt = G(() => e.value.style.chart.title), { exportSvg: Wt, getSvg: Dt } = vo({
      svg: oe,
      title: Yt,
      backgroundColor: Rt
    });
    async function nt({ isCb: t }) {
      q.value = !0, await _e();
      try {
        if (t) {
          const { blob: y, url: a, text: S, dataUrl: T } = await Dt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: y, url: a, text: S, dataUrl: T }));
        } else
          await Promise.resolve(Wt());
      } finally {
        q.value = !1;
      }
    }
    function Et(t) {
      if (t?.stage === "start") {
        H.value = !0;
        return;
      }
      if (t?.stage === "end") {
        H.value = !1;
        return;
      }
      He();
    }
    function at() {
      et(1.5, !0);
    }
    function lt() {
      et(1 / 1.5, !0);
    }
    const he = I(null), Se = I(!1), Ne = I(""), Ze = I(null);
    function Nt(t, y) {
      he.value = null, X.value = !1, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: y });
    }
    function Zt(t, y) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: y });
    }
    function Bt(t, y) {
      if (e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: y }), !O.value.showTooltip) return;
      he.value = t.id, Ze.value = { datapoint: t, config: e.value };
      const a = e.value.style.chart.tooltip.customFormat;
      if (Se.value = !1, so(a))
        try {
          const S = a({
            datapoint: t,
            config: e.value
          });
          typeof S == "string" && (Ne.value = S, Se.value = !0);
        } catch {
          console.warn("Custom format cannot be applied."), Se.value = !1;
        }
      if (!Se.value) {
        let S = `<svg viewBox="0 0 10 10" height="${e.value.style.chart.tooltip.fontSize}"><circle cx="5" cy="5" r="5" fill="${t.color}"/></svg><span>${t.name}:</span><b>${(t.value || 0).toFixed(e.value.style.chart.tooltip.roundingValue)}</b>`;
        Ne.value = `<div dir="auto" style="display:flex; gap:4px; align-items:center; jsutify-content:center;">${S}</div>`;
      }
      X.value = !0;
    }
    async function st() {
      if (u("copyAlt", {
        config: e.value,
        dataset: U.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: U.value
      }));
    }
    return s({
      getData: Pt,
      getImage: zt,
      generateCsv: De,
      generatePdf: qe,
      generateImage: He,
      generateSvg: nt,
      resetZoom: ke,
      toggleTable: Ke,
      toggleTooltip: Je,
      toggleAnnotator: Ee,
      toggleFullscreen: Ge,
      toggleZoom: Qe,
      copyAlt: st
    }), (t, y) => (Y(), K("div", {
      class: "vue-data-ui-component vue-ui-word-cloud",
      ref_key: "wordCloudChart",
      ref: d,
      id: `wordCloud_${F.value}`,
      "data-resizing": me.value,
      "data-relayout": ue.value,
      style: Ae(`width: 100%; font-family:${e.value.style.fontFamily};background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height:100%" : ""}`),
      onMouseenter: y[3] || (y[3] = () => C(ge)(!0)),
      onMouseleave: y[4] || (y[4] = () => C(ge)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (Y(), se(C(n), {
        key: 0,
        svgRef: C(oe),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: Ce.value,
        isCursorPointer: N.value,
        onClose: Ee
      }, {
        "annotator-action-close": A(() => [
          P(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": A(({ color: a }) => [
          P(t.$slots, "annotator-action-color", V(_({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": A(({ mode: a }) => [
          P(t.$slots, "annotator-action-draw", V(_({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": A(({ disabled: a }) => [
          P(t.$slots, "annotator-action-undo", V(_({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": A(({ disabled: a }) => [
          P(t.$slots, "annotator-action-redo", V(_({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": A(({ disabled: a }) => [
          P(t.$slots, "annotator-action-delete", V(_({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : L("", !0),
      e.value.style.chart.title.text ? (Y(), K("div", {
        key: 1,
        ref_key: "chartTitle",
        ref: x,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (Y(), se(go, {
          key: `title_${W.value}`,
          config: {
            title: {
              ...e.value.style.chart.title
            },
            subtitle: {
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : L("", !0),
      e.value.userOptions.show && $.value && (C(Te) || C(ce)) ? (Y(), se(C(h), {
        ref_key: "userOptionsRef",
        ref: Q,
        key: `user_option_${k.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: C(Ye),
        isImaging: C(We),
        uid: F.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: ae.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: d.value,
        position: e.value.userOptions.position,
        hasTooltip: e.value.style.chart.tooltip.show && e.value.userOptions.buttons.tooltip,
        isTooltip: O.value.showTooltip,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Ce.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        hasZoom: e.value.userOptions.buttons.zoom,
        isZoom: O.value.showZoom,
        isCursorPointer: N.value,
        onToggleFullscreen: Ge,
        onGeneratePdf: C(qe),
        onGenerateCsv: De,
        onGenerateImage: Et,
        onGenerateSvg: nt,
        onToggleTable: Ke,
        onToggleTooltip: Je,
        onToggleAnnotator: Ee,
        onToggleZoom: Qe,
        onCopyAlt: st,
        style: Ae({
          visibility: C(Te) ? C(ce) ? "visible" : "hidden" : "visible"
        })
      }, ut({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: A(({ isOpen: a, color: S }) => [
            P(t.$slots, "menuIcon", V(_({ isOpen: a, color: S })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: A(() => [
            P(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: A(() => [
            P(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: A(() => [
            P(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: A(() => [
            P(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: A(() => [
            P(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: A(({ toggleFullscreen: a, isFullscreen: S }) => [
            P(t.$slots, "optionFullscreen", V(_({ toggleFullscreen: a, isFullscreen: S })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: A(({ toggleAnnotator: a, isAnnotator: S }) => [
            P(t.$slots, "optionAnnotator", V(_({ toggleAnnotator: a, isAnnotator: S })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionZoom ? {
          name: "optionZoom",
          fn: A(({ toggleZoom: a, isZoomLocked: S }) => [
            P(t.$slots, "optionZoom", V(_({ toggleZoom: a, isZoomLocked: S })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: A(({ altCopy: a }) => [
            P(t.$slots, "optionAltCopy", V(_({ altCopy: a })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasTooltip", "isTooltip", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "hasZoom", "isZoom", "isCursorPointer", "onGeneratePdf", "style"])) : L("", !0),
      e.value.style.chart.controls.position === "top" && e.value.style.chart.controls.show && !C(ve) ? (Y(), se(ht, {
        key: 3,
        ref_key: "zoomControls",
        ref: re,
        config: e.value,
        scale: C(tt),
        isFullscreen: ae.value,
        isCursorPointer: N.value,
        onZoomIn: at,
        onZoomOut: lt,
        onResetZoom: y[0] || (y[0] = (a) => C(ke)(!0))
      }, null, 8, ["config", "scale", "isFullscreen", "isCursorPointer"])) : L("", !0),
      (Y(), K("svg", {
        ref_key: "svgRef",
        ref: oe,
        class: rt({ "vue-data-ui-fullscreen--on": ae.value, "vue-data-ui-fulscreen--off": !ae.value }),
        xmlns: C(no),
        viewBox: `${C(Oe).x} ${C(Oe).y} ${C(Oe).width} ${C(Oe).height}`,
        style: "overflow:hidden;background:transparent;display:block"
      }, [
        Be(C(f)),
        t.$slots["chart-background"] ? (Y(), K("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: R.value.width <= 0 ? 10 : R.value.width,
          height: R.value.height <= 0 ? 10 : R.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          P(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Zo)) : L("", !0),
        Xe("g", {
          transform: `translate(${ye.value.x}, ${ye.value.y})`,
          class: rt({ "wc-finalized": ie.value })
        }, [
          (Y(!0), K(Lt, null, qt(U.value, (a, S) => (Y(), K("g", {
            key: a.id,
            class: "vue-ui-word-cloud-word",
            transform: `translate(${a.x}, ${a.y})`
          }, [
            a.minX !== void 0 ? (Y(), K("rect", {
              key: 0,
              x: a.minX,
              y: a.minY * 1.25,
              width: a.maxX - a.minX,
              height: a.maxY - a.minY,
              fill: "transparent",
              "pointer-events": "visiblePainted",
              onMouseover: (T) => Bt(a, S),
              onMouseleave: (T) => Nt(a, S),
              onClick: (T) => Zt(a, S)
            }, null, 40, _o)) : L("", !0),
            Xe("text", {
              fill: a.color,
              "font-weight": e.value.style.chart.words.bold ? "bold" : "normal",
              x: 0,
              y: 0,
              "font-size": a.fontSize,
              transform: `translate(${a.width / 2}, ${a.height / 2})`,
              "text-anchor": "middle",
              "dominant-baseline": "central",
              "paint-order": "stroke fill",
              stroke: !he.value || he.value === a.id ? e.value.style.chart.words.selectedStroke : void 0,
              "stroke-width": a.height * 0.05,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: Ae(`
                            pointer-events:none;
                            fill-opacity:${!he.value || he.value === a.id || !ie.value ? 1 : e.value.style.chart.words.hoverOpacity} !important;
                        `)
            }, Ve(a.name), 13, Uo)
          ], 8, Vo))), 128))
        ], 10, Bo),
        P(t.$slots, "svg", {
          svg: {
            ...R.value,
            isPrintingImg: C(Ye) | C(We) | H.value,
            isPrintingSvg: q.value
          }
        }, void 0, !0)
      ], 10, No)),
      t.$slots.watermark ? (Y(), K("div", Lo, [
        P(t.$slots, "watermark", V(_({ isPrinting: C(Ye) || C(We) || H.value || q.value })), void 0, !0)
      ])) : L("", !0),
      C(At) ? (Y(), K("div", qo, [
        P(t.$slots, "reset-action", { reset: C(ke) }, void 0, !0)
      ])) : L("", !0),
      e.value.style.chart.controls.position === "bottom" && e.value.style.chart.controls.show && !C(ve) ? (Y(), se(ht, {
        key: 6,
        ref_key: "zoomControls",
        ref: re,
        config: e.value,
        scale: C(tt),
        isFullscreen: ae.value,
        isCursorPointer: N.value,
        onZoomIn: at,
        onZoomOut: lt,
        onResetZoom: y[1] || (y[1] = (a) => C(ke)(!0))
      }, null, 8, ["config", "scale", "isFullscreen", "isCursorPointer"])) : L("", !0),
      Be(C(c), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: O.value.showTooltip && X.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: d.value,
        content: Ne.value,
        isCustom: Se.value,
        isFullscreen: ae.value,
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": A(() => [
          P(t.$slots, "tooltip-before", V(_({ ...Ze.value })), void 0, !0)
        ]),
        "tooltip-after": A(() => [
          P(t.$slots, "tooltip-after", V(_({ ...Ze.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isCustom", "isFullscreen", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      t.$slots.source ? (Y(), K("div", {
        key: 7,
        ref_key: "source",
        ref: Z,
        dir: "auto"
      }, [
        P(t.$slots, "source", {}, void 0, !0)
      ], 512)) : L("", !0),
      $.value && e.value.userOptions.buttons.table ? (Y(), se(Ht(Fe.value.component), jt({ key: 8 }, Fe.value.props, {
        ref_key: "tableUnit",
        ref: B,
        onClose: ot
      }), ut({
        content: A(() => [
          (Y(), se(C(i), {
            key: `table_${D.value}`,
            colNames: Pe.value.colNames,
            head: Pe.value.head,
            body: Pe.value.body,
            config: Pe.value.config,
            title: e.value.table.useDialog ? "" : Fe.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: N.value,
            onClose: ot
          }, {
            th: A(({ th: a }) => [
              Xe("div", {
                innerHTML: a,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Ho)
            ]),
            td: A(({ td: a }) => [
              ct(Ve(a.name || a), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: A(() => [
            ct(Ve(Fe.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: A(() => [
            Xe("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: y[2] || (y[2] = (a) => De(e.value.userOptions.callbacks.csv)),
              style: Ae({ cursor: N.value ? "pointer" : "default" })
            }, [
              Be(C(r), {
                name: "fileCsv",
                stroke: Fe.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : L("", !0),
      P(t.$slots, "skeleton", {}, () => [
        C(ve) ? (Y(), se(ro, { key: 0 })) : L("", !0)
      ], !0)
    ], 44, Eo));
  }
}, fn = /* @__PURE__ */ wo(jo, [["__scopeId", "data-v-ed7a2652"]]);
export {
  fn as default
};
