// vue-data-ui/dist/vue-ui-strip-plot-B9nBSRyY.js
import { defineAsyncComponent as T, computed as d, ref as i, toRefs as Ut, watch as Q, shallowRef as Qe, onMounted as Vt, onBeforeUnmount as et, watchEffect as Xt, openBlock as r, createElementBlock as u, unref as v, normalizeStyle as G, normalizeClass as H, createBlock as z, withCtx as c, renderSlot as h, normalizeProps as g, guardReactiveProps as k, createCommentVNode as f, createSlots as tt, createVNode as Ae, Fragment as x, renderList as S, createElementVNode as C, toDisplayString as _, mergeProps as lt, resolveDynamicComponent as jt, createTextVNode as at, nextTick as ot } from "vue";
import { j as Le, t as st, m as Yt, p as D, o as rt, q as $e, r as nt, n as qt, s as ut, K as Jt, D as Kt, l as ee, X as Zt, k as Te, Z as Qt, at as el, x as tl, w as ll, T as it, a2 as al, a0 as ol, a1 as sl, U as rl } from "./lib-GwPnJSrY.js";
import { t as vt, u as nl } from "./useResponsive-DfdjqQps.js";
import { u as ul } from "./useConfig-CIgyKZ6R.js";
import { u as il, B as vl } from "./BaseScanner-BhhkKVoj.js";
import { u as cl } from "./usePrinter-DiK5R6k-.js";
import { u as hl } from "./useSvgExport-ByUukOZt.js";
import { u as Se } from "./useNestedProp-BBT0mbHh.js";
import { u as dl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as pl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as fl } from "./useChartAccessibility-9icAAmYg.js";
import { u as bl } from "./useTimeLabelCollider-CIsgDrl9.js";
import yl from "./img-BYYzY_QU.js";
import { _ as ml } from "./Shape-CSyWYgXc.js";
import gl from "./Title-BfyrP-jp.js";
import kl from "./vue_ui_strip_plot-B9ypi_F3.js";
import { _ as xl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Cl = ["id"], wl = ["xmlns", "viewBox"], Al = ["x", "y", "width", "height"], Ll = { key: 1 }, $l = { key: 0 }, Tl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Sl = { key: 1 }, _l = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Pl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Ol = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], zl = ["x", "y", "fill", "font-size"], Dl = ["transform", "font-size", "fill", "text-anchor"], Nl = ["transform", "font-size", "fill", "text-anchor", "innerHTML"], Il = ["fill", "font-size", "transform"], Fl = ["fill", "font-size", "x", "y"], Ml = ["x1", "x2", "y1", "y2", "stroke"], Rl = ["cx", "cy", "fill"], Bl = ["cx", "cy", "fill"], Wl = ["id"], El = ["stop-color"], Gl = ["stop-color"], Hl = ["stop-color"], Ul = { key: 0 }, Vl = ["x", "y", "font-size", "fill"], Xl = {
  key: 4,
  class: "vue-data-ui-watermark"
}, jl = ["innerHTML"], Yl = {
  __name: "vue-ui-strip-plot",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectDatapoint", "copyAlt"],
  setup(ct, { expose: ht, emit: dt }) {
    const pt = T(() => import("./Tooltip-DL3QgjZf.js")), ft = T(() => import("./BaseIcon-CNx8Tzl5.js")), bt = T(() => import("./vue-ui-accordion-BGHTqIBt.js")), yt = T(() => import("./DataTable-Bo1pmAqG.js")), mt = T(() => import("./UserOptions-CrLqEimB.js")), gt = T(() => import("./PenAndPaper-DWV1FMIi.js")), kt = T(() => import("./PackageVersion-DN9d0ME5.js")), xt = T(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_strip_plot: Ct } = ul(), { isThemeValid: wt, warnInvalidTheme: At } = dl(), m = ct, _e = dt, U = d({
      get() {
        return !!m.dataset && m.dataset.length;
      },
      set(t) {
        return t;
      }
    }), te = i(Le()), Pe = i(0), le = i(!1), ae = i(""), A = i(null), Oe = i(null), ze = i(null), De = i(null), Ne = i(!1), Ie = i(0), Fe = i(0), M = i(null), oe = i(null), V = i(!1), X = i(!1), se = i(null), re = i(null), j = i(null), ne = i(null), e = i(de()), R = d(() => e.value.userOptions.useCursorPointer), Lt = d(() => st({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            padding: {
              top: 24,
              left: 24,
              right: 24,
              bottom: 24
            },
            grid: {
              stroke: "#6A6A6A",
              horizontalGrid: {
                stroke: "#6A6A6A"
              },
              verticalGrid: {
                stroke: "#6A6A6A"
              }
            },
            plots: {
              stroke: "#6A6A6A"
            },
            labels: {
              bestPlotLabel: { show: !1 },
              axis: {
                xLabel: "",
                yLabel: ""
              },
              xAxisLabels: { show: !1 },
              yAxisLabels: { show: !1 }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: ue, FINAL_DATASET: ie, manualLoading: ve } = il({
      ...Ut(m),
      FINAL_CONFIG: e,
      prepareConfig: de,
      skeletonDataset: m.config?.skeletonDataset ?? [
        {
          name: "_",
          color: "#DBDBDB",
          plots: [
            { name: "_", value: 1 },
            { name: "_", value: 2 },
            { name: "_", value: 3 }
          ]
        },
        {
          name: "_",
          color: "#C4C4C4",
          plots: [
            { name: "_", value: 3 },
            { name: "_", value: 5 },
            { name: "_", value: 8 }
          ]
        },
        {
          name: "_",
          color: "#ADADAD",
          plots: [
            { name: "_", value: 8 },
            { name: "_", value: 13 },
            { name: "_", value: 21 }
          ]
        },
        {
          name: "_",
          color: "#969696",
          plots: [
            { name: "_", value: 21 },
            { name: "_", value: 34 },
            { name: "_", value: 55 }
          ]
        },
        {
          name: "_",
          color: "#808080",
          plots: [
            { name: "_", value: 55 },
            { name: "_", value: 89 },
            { name: "_", value: 144 }
          ]
        }
      ],
      skeletonConfig: st({
        defaultConfig: e.value,
        userConfig: Lt.value
      })
    }), { userOptionsVisible: ce, setUserOptionsVisibility: Me, keepUserOptionState: Re } = pl({ config: e.value }), { svgRef: he } = fl({ config: e.value.style.chart.title });
    function de() {
      const t = Se({
        userConfig: m.config,
        defaultConfig: Ct
      }), a = t.theme;
      if (!a) return t;
      if (!wt.value(t))
        return At(t), t;
      const l = Se({
        userConfig: kl[a] || m.config,
        defaultConfig: t
      }), o = Se({
        userConfig: m.config,
        defaultConfig: l
      });
      return {
        ...o,
        customPalette: o.customPalette.length ? o.customPalette : Yt[a] || D
      };
    }
    Q(() => m.config, async (t) => {
      ue.value || (e.value = de()), ce.value = !e.value.userOptions.showOnChartHover, Be(), Ie.value += 1, Fe.value += 1, b.value.dataLabels.show = e.value.style.chart.labels.bestPlotLabel.show, b.value.showTable = e.value.table.show, b.value.showTooltip = e.value.style.chart.tooltip.show, I.value = e.value.style.chart.width, $.value = e.value.style.chart.height, W.value = e.value.style.chart.plots.radius;
    }, { deep: !0 }), Q(() => m.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (ve.value = !1);
    }, { deep: !0 });
    const L = Qe(null), N = Qe(null);
    Vt(() => {
      Be();
    });
    const pe = d(() => !!e.value.debug);
    function Be() {
      if (rt(m.dataset) ? ($e({
        componentName: "VueUiStripPlot",
        type: "dataset",
        debug: pe.value
      }), ve.value = !0) : m.dataset.forEach((t, a) => {
        nt({
          datasetObject: t,
          requiredAttributes: ["name", "plots"]
        }).forEach((l) => {
          U.value = !1, $e({
            componentName: "VueUiStripPlot",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: pe.value
          });
        }), t.plots && t.plots.forEach((l, o) => {
          nt({
            datasetObject: l,
            requiredAttributes: ["name", "value"]
          }).forEach((n) => {
            U.value = !1, $e({
              componentName: "VueUiStripPlot",
              type: "datasetSerieAttribute",
              property: n,
              index: `${a},${o}`,
              debug: pe.value
            });
          });
        });
      }), rt(m.dataset) || (ve.value = e.value.loading), e.value.responsive) {
        const t = vt(() => {
          const { width: a, height: l } = nl({
            chart: A.value,
            title: e.value.style.chart.title.text ? Oe.value : null,
            source: ze.value,
            noTitle: De.value
          });
          requestAnimationFrame(() => {
            Tt.value = l, I.value = Math.max(0.1, a), $.value = Math.max(0.1, l - 12), e.value.responsiveProportionalSizing ? W.value = al({
              relator: Math.min(l, a),
              adjuster: 600,
              source: e.value.style.chart.plots.radius,
              threshold: 6,
              fallback: 6
            }) : W.value = e.value.style.chart.plots.radius;
          });
        });
        L.value && (N.value && L.value.unobserve(N.value), L.value.disconnect()), L.value = new ResizeObserver(t), N.value = A.value.parentNode, L.value.observe(N.value);
      }
      Ne.value = !0, setTimeout(() => {
        B.value = !1;
      }, Ot.value * 50);
    }
    et(() => {
      L.value && (N.value && L.value.unobserve(N.value), L.value.disconnect());
    });
    const { isPrinting: fe, isImaging: be, generatePdf: We, generateImage: Ee } = cl({
      elementId: `strip-plot_${te.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-strip-plot",
      options: e.value.userOptions.print
    }), $t = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), Ge = d(() => qt(e.value.customPalette)), B = i(e.value.useCssAnimation), Y = i({
      top: e.value.style.chart.padding.top,
      bottom: e.value.style.chart.padding.bottom,
      left: e.value.style.chart.padding.left,
      right: e.value.style.chart.padding.right
    }), I = i(e.value.style.chart.width), $ = i(e.value.style.chart.height), Tt = i(e.value.style.chart.height), W = i(e.value.style.chart.plots.radius), b = i({
      showTable: e.value.table.show,
      dataLabels: {
        show: e.value.style.chart.labels.bestPlotLabel.show
      },
      showTooltip: e.value.style.chart.tooltip.show
    });
    Q(e, () => {
      b.value = {
        showTable: e.value.table.show,
        dataLabels: {
          show: e.value.style.chart.labels.bestPlotLabel.show
        },
        showTooltip: e.value.style.chart.tooltip.show
      }, I.value = e.value.style.chart.width, $.value = e.value.style.chart.height, W.value = e.value.style.chart.plots.radius;
    }, { immediate: !0 });
    const ye = d(() => Math.min(W.value, s.value.stripWidth / 2 * 0.9));
    function St() {
      let t = 0;
      ne.value && (t = Array.from(ne.value.querySelectorAll("text")).reduce((o, n) => {
        const p = n.getComputedTextLength();
        return p > o ? p : o;
      }, 0));
      const a = re.value ? re.value.getBoundingClientRect().width : 0;
      return t + a + (a ? 24 : 0);
    }
    const me = i(0), _t = vt((t) => {
      me.value = t;
    }, 100);
    Xt((t) => {
      const a = j.value;
      if (!a) return;
      const l = new ResizeObserver((o) => {
        _t(o[0].contentRect.height);
      });
      l.observe(a), t(() => l.disconnect());
    }), et(() => {
      me.value = 0;
    });
    const Pt = d(() => {
      let t = 0;
      se.value && (t = se.value.getBBox().height + e.value.style.chart.labels.axis.fontSize / 3 + 12 + e.value.style.chart.labels.axis.xLabelOffsetY);
      let a = 0;
      return j.value && (a = me.value + 12), t + a;
    }), s = d(() => {
      const t = St(), a = Y.value.left + t + e.value.style.chart.labels.axis.yLabelOffsetX + 5, l = I.value - Y.value.right, o = Math.max(0, l - a), n = Y.value.top + e.value.style.chart.plots.radius + e.value.style.chart.labels.bestPlotLabel.fontSize, p = $.value - Y.value.bottom - Pt.value, F = Math.max(0, p - n), Ze = Array.isArray(ie.value) ? ie.value.length : 0, Ht = Ze > 0 ? o / Ze : 0;
      return {
        left: a,
        right: l,
        top: n,
        bottom: p,
        width: o,
        height: F,
        stripWidth: Ht,
        absoluteHeight: $.value
      };
    }), q = d(() => ie.value.map((t, a) => {
      const l = Le();
      return {
        ...t,
        id: l,
        color: t.color ? ut(t.color) : Ge.value[a] || D[a] || D[a % D.length],
        plots: t.plots.map((o, n) => ({
          ...o,
          value: Jt(o.value),
          parentId: l,
          parentName: t.name,
          parentIndex: a,
          plotIndex: n,
          color: t.color ? ut(t.color) : Ge.value[a] || D[a] || D[a % D.length],
          id: Le()
        })).sort((o, n) => o.value - n.value)
      };
    })), w = d(() => (q.value || []).map((t, a) => ({
      ...t,
      plots: t.plots.map((l) => ({
        ...l,
        x: s.value.left + (a + 1) * s.value.stripWidth - s.value.stripWidth / 2
      }))
    }))), Ot = d(() => Math.max(...w.value.map((t) => t.plots.length))), ge = d(() => {
      const t = w.value.flatMap((a) => a.plots.map((l) => l.value));
      return {
        max: Math.max(...t),
        min: Math.min(...t)
      };
    }), P = d(() => Kt(ge.value.min < 0 ? ge.value.min : 0, ge.value.max, e.value.style.chart.grid.scaleSteps)), ke = d(() => (w.value || []).map((t, a) => ({
      ...t,
      plots: t.plots.map((l) => ({
        ...l,
        y: s.value.bottom - (l.value + Math.abs(P.value.min)) / (P.value.max + Math.abs(P.value.min)) * s.value.height
      }))
    }))), He = d(() => P.value.ticks.map((t) => ({
      y: s.value.bottom - s.value.height * ((t + Math.abs(P.value.min)) / (P.value.max + Math.abs(P.value.min))),
      x1: s.value.left,
      x2: s.value.right,
      value: t
    }))), xe = i(null), y = i(null);
    function zt({ datapoint: t, seriesIndex: a }) {
      le.value = !1, y.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: a });
    }
    function Dt({ datapoint: t, seriesIndex: a }) {
      _e("selectDatapoint", t), e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: a });
    }
    function Nt({ datapoint: t, seriesIndex: a }) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: a }), xe.value = { datapoint: t, seriesIndex: a, config: e.value, series: q.value }, le.value = !0, y.value = t;
      const l = e.value.style.chart.tooltip.customFormat;
      if (it(l) && rl(() => l({
        seriesIndex: a,
        datapoint: t,
        series: q.value,
        config: e.value
      })))
        ae.value = l({
          seriesIndex: a,
          datapoint: t,
          series: q.value,
          config: e.value
        });
      else {
        let o = "";
        o += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="none" fill="${e.value.style.chart.plots.gradient.show ? `url(#${t.parentId})` : t.color}"/></svg>${t.name}</div>`, o += `<div>${Te(
          e.value.style.chart.labels.formatter,
          t.value,
          ee({
            p: e.value.style.chart.labels.prefix,
            v: t.value,
            s: e.value.style.chart.labels.suffix,
            r: e.value.style.chart.tooltip.roundingValue
          }),
          { datapoint: t, seriesIndex: a }
        )}</div>`, ae.value = `<div>${o}</div>`;
      }
    }
    const J = d(() => {
      const t = w.value.flatMap((l) => JSON.parse(JSON.stringify(l.plots)).sort((o, n) => n.value - o.value).map((o) => ({
        name: `${l.name} - ${o.name}`,
        color: o.color
      }))), a = w.value.flatMap((l) => JSON.parse(JSON.stringify(l.plots)).sort((o, n) => n.value - o.value).map((o) => o.value));
      return { head: t, body: a };
    });
    function Ce(t = null) {
      ot(() => {
        const a = J.value.head.map((n, p) => [[
          n.name
        ], [J.value.body[p]]]), l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[e.value.table.columnNames.series], [e.value.table.columnNames.value]]].concat(a), o = ol(l);
        t ? t(o) : sl({ csvContent: o, title: e.value.style.chart.title.text || "vue-ui-strip-plot" });
      });
    }
    const K = d(() => {
      const t = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.value
      ], a = J.value.head.map((n, p) => {
        const F = ee({
          p: e.value.style.chart.labels.prefix,
          v: J.value.body[p],
          s: e.value.style.chart.labels.suffix,
          r: e.value.table.td.roundingValue
        });
        return [
          {
            color: n.color,
            name: n.name
          },
          F
        ];
      }), l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.series,
          e.value.table.columnNames.value
        ],
        head: t,
        body: a,
        config: l
      };
    }), O = i(!1);
    function Ue(t) {
      O.value = t, Pe.value += 1;
    }
    function It() {
      return w.value;
    }
    function Ve() {
      b.value.showTable = !b.value.showTable;
    }
    function Xe() {
      b.value.dataLabels.show = !b.value.dataLabels.show;
    }
    function je() {
      b.value.showTooltip = !b.value.showTooltip;
    }
    const Z = i(!1);
    function we() {
      Z.value = !Z.value;
    }
    async function Ft({ scale: t = 2 } = {}) {
      if (!A.value) return;
      const { width: a, height: l } = A.value.getBoundingClientRect(), o = a / l, { imageUri: n, base64: p } = await yl({ domElement: A.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: n,
        base64: p,
        title: e.value.style.chart.title.text,
        width: a,
        height: l,
        aspectRatio: o
      };
    }
    const Ye = d(() => w.value.map((t) => t.name)), Mt = i({ start: 0, end: w.value.length });
    bl({
      timeLabelsEls: j,
      timeLabels: Ye,
      slicer: Mt,
      configRef: e,
      rotationPath: ["style", "chart", "labels", "xAxisLabels", "rotation"],
      autoRotatePath: ["style", "chart", "labels", "xAxisLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: I,
      height: $,
      targetClass: ".vue-ui-strip-plot-category-name",
      rotation: e.value.style.chart.labels.xAxisLabels.autoRotate.angle
    });
    const E = d(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = b.value.showTable;
      return {
        component: t ? xt : bt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: O.value,
          fullscreenParent: A.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: R.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    Q(() => b.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && M.value ? M.value.open() : "close" in M.value && M.value.close());
    });
    function qe() {
      b.value.showTable = !1, oe.value && oe.value.setTableIconState(!1);
    }
    const Rt = d(() => e.value.style.chart.backgroundColor), Bt = d(() => e.value.style.chart.title), { exportSvg: Wt, getSvg: Et } = hl({
      svg: he,
      title: Bt,
      backgroundColor: Rt
    });
    async function Je({ isCb: t }) {
      X.value = !0, await ot();
      try {
        if (t) {
          const { blob: a, url: l, text: o, dataUrl: n } = await Et();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: o, dataUrl: n }));
        } else
          await Promise.resolve(Wt());
      } finally {
        X.value = !1;
      }
    }
    function Gt(t) {
      if (t?.stage === "start") {
        V.value = !0;
        return;
      }
      if (t?.stage === "end") {
        V.value = !1;
        return;
      }
      Ee();
    }
    async function Ke() {
      if (_e("copyAlt", {
        config: e.value,
        dataset: ke.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: ke.value
      }));
    }
    return ht({
      getData: It,
      getImage: Ft,
      generatePdf: We,
      generateCsv: Ce,
      generateImage: Ee,
      generateSvg: Je,
      toggleTable: Ve,
      toggleLabels: Xe,
      toggleTooltip: je,
      toggleAnnotator: we,
      toggleFullscreen: Ue,
      copyAlt: Ke
    }), (t, a) => (r(), u("div", {
      ref_key: "stripPlotChart",
      ref: A,
      class: H(`vue-data-ui-component vue-ui-strip-plot ${O.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: G(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height:100%" : ""}`),
      id: `strip-plot_${te.value}`,
      onMouseenter: a[1] || (a[1] = () => v(Me)(!0)),
      onMouseleave: a[2] || (a[2] = () => v(Me)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (r(), z(v(gt), {
        key: 0,
        svgRef: v(he),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: Z.value,
        isCursorPointer: R.value,
        onClose: we
      }, {
        "annotator-action-close": c(() => [
          h(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": c(({ color: l }) => [
          h(t.$slots, "annotator-action-color", g(k({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": c(({ mode: l }) => [
          h(t.$slots, "annotator-action-draw", g(k({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": c(({ disabled: l }) => [
          h(t.$slots, "annotator-action-undo", g(k({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": c(({ disabled: l }) => [
          h(t.$slots, "annotator-action-redo", g(k({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": c(({ disabled: l }) => [
          h(t.$slots, "annotator-action-delete", g(k({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : f("", !0),
      $t.value ? (r(), u("div", {
        key: 1,
        ref_key: "noTitle",
        ref: De,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : f("", !0),
      e.value.style.chart.title.text ? (r(), u("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Oe,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (r(), z(gl, {
          key: `title_${Ie.value}`,
          config: {
            title: {
              cy: "donut-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "donut-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : f("", !0),
      e.value.userOptions.show && U.value && (v(Re) || v(ce)) ? (r(), z(v(mt), {
        ref_key: "userOptionsRef",
        ref: oe,
        key: `user_option_${Pe.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: v(fe),
        isImaging: v(be),
        uid: te.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isTooltip: b.value.showTooltip,
        isFullscreen: O.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: A.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Z.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: R.value,
        onToggleFullscreen: Ue,
        onGeneratePdf: v(We),
        onGenerateCsv: Ce,
        onGenerateImage: Gt,
        onGenerateSvg: Je,
        onToggleTable: Ve,
        onToggleLabels: Xe,
        onToggleTooltip: je,
        onToggleAnnotator: we,
        onCopyAlt: Ke,
        style: G({
          visibility: v(Re) ? v(ce) ? "visible" : "hidden" : "visible"
        })
      }, tt({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: c(({ isOpen: l, color: o }) => [
            h(t.$slots, "menuIcon", g(k({ isOpen: l, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: c(() => [
            h(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: c(() => [
            h(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: c(() => [
            h(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: c(() => [
            h(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: c(() => [
            h(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: c(() => [
            h(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: c(() => [
            h(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: c(({ toggleFullscreen: l, isFullscreen: o }) => [
            h(t.$slots, "optionFullscreen", g(k({ toggleFullscreen: l, isFullscreen: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: c(({ toggleAnnotator: l, isAnnotator: o }) => [
            h(t.$slots, "optionAnnotator", g(k({ toggleAnnotator: l, isAnnotator: o })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: c(({ altCopy: l }) => [
            h(t.$slots, "optionAltCopy", g(k({ altCopy: l })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isTooltip", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : f("", !0),
      (r(), u("svg", {
        ref_key: "svgRef",
        ref: he,
        xmlns: v(Zt),
        class: H({ "vue-data-ui-fullscreen--on": O.value, "vue-data-ui-fulscreen--off": !O.value }),
        viewBox: `0 0 ${I.value} ${$.value}`,
        style: G(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color};`)
      }, [
        Ae(v(kt)),
        t.$slots["chart-background"] ? (r(), u("foreignObject", {
          key: 0,
          x: s.value.left,
          y: s.value.top,
          width: Math.max(0.1, s.value.width),
          height: Math.max(0.1, s.value.height),
          style: {
            pointerEvents: "none"
          }
        }, [
          h(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Al)) : f("", !0),
        e.value.style.chart.grid.show ? (r(), u("g", Ll, [
          e.value.style.chart.grid.horizontalGrid.show ? (r(), u("g", $l, [
            (r(!0), u(x, null, S(He.value, (l) => (r(), u("line", {
              x1: l.x1,
              x2: l.x2,
              y1: l.y,
              y2: l.y,
              stroke: e.value.style.chart.grid.horizontalGrid.stroke,
              "stroke-width": e.value.style.chart.grid.horizontalGrid.strokeWidth,
              "stroke-dasharray": e.value.style.chart.grid.horizontalGrid.strokeDasharray,
              "stroke-linecap": "round"
            }, null, 8, Tl))), 256))
          ])) : f("", !0),
          e.value.style.chart.grid.verticalGrid.show ? (r(), u("g", Sl, [
            (r(!0), u(x, null, S(w.value, (l, o) => (r(), u("line", {
              x1: s.value.left + (o + 1) * s.value.stripWidth,
              x2: s.value.left + (o + 1) * s.value.stripWidth,
              y1: s.value.top,
              y2: s.value.bottom,
              stroke: e.value.style.chart.grid.verticalGrid.stroke,
              "stroke-width": e.value.style.chart.grid.verticalGrid.strokeWidth,
              "stroke-dasharray": e.value.style.chart.grid.verticalGrid.strokeDasharray,
              "stroke-linecap": "round"
            }, null, 8, _l))), 256))
          ])) : f("", !0),
          C("line", {
            x1: s.value.left,
            x2: s.value.left,
            y1: s.value.top,
            y2: s.value.bottom,
            stroke: e.value.style.chart.grid.stroke,
            "stroke-width": e.value.style.chart.grid.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, Pl),
          C("line", {
            x1: s.value.left,
            x2: s.value.right,
            y1: s.value.bottom,
            y2: s.value.bottom,
            stroke: e.value.style.chart.grid.stroke,
            "stroke-width": e.value.style.chart.grid.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, Ol)
        ])) : f("", !0),
        e.value.style.chart.labels.yAxisLabels.show ? (r(), u("g", {
          key: 2,
          ref_key: "scaleLabels",
          ref: ne
        }, [
          (r(!0), u(x, null, S(He.value, (l, o) => (r(), u("text", {
            x: l.x1 + e.value.style.chart.labels.yAxisLabels.offsetX - 5,
            y: l.y + e.value.style.chart.labels.yAxisLabels.fontSize / 3,
            fill: e.value.style.chart.labels.yAxisLabels.color,
            "font-size": e.value.style.chart.labels.yAxisLabels.fontSize,
            "text-anchor": "end"
          }, _(v(Te)(
            e.value.style.chart.labels.formatter,
            l.value,
            v(ee)({
              p: e.value.style.chart.labels.prefix,
              v: l.value,
              s: e.value.style.chart.labels.suffix,
              r: e.value.style.chart.labels.yAxisLabels.rounding
            }),
            { datapoint: l, seriesIndex: o }
          )), 9, zl))), 256))
        ], 512)) : f("", !0),
        e.value.style.chart.labels.xAxisLabels.show ? (r(), u("g", {
          key: 3,
          ref_key: "timeLabelsEls",
          ref: j
        }, [
          (r(!0), u(x, null, S(Ye.value, (l, o) => (r(), u("g", null, [
            String(l).includes(`
`) ? (r(), u("text", {
              key: 1,
              class: "vue-ui-strip-plot-category-name",
              transform: `translate(${s.value.left + (o + 1) * s.value.stripWidth - s.value.stripWidth / 2}, ${s.value.bottom + e.value.style.chart.labels.xAxisLabels.fontSize * 2 + e.value.style.chart.labels.xAxisLabels.offsetY}), rotate(${e.value.style.chart.labels.xAxisLabels.rotation})`,
              "font-size": e.value.style.chart.labels.xAxisLabels.fontSize,
              fill: e.value.style.chart.labels.xAxisLabels.color,
              "text-anchor": e.value.style.chart.labels.xAxisLabels.rotation > 0 ? "start" : e.value.style.chart.labels.xAxisLabels.rotation < 0 ? "end" : "middle",
              innerHTML: v(Qt)({
                content: v(el)(String(l)),
                fontSize: e.value.style.chart.labels.xAxisLabels.fontSize,
                fill: e.value.style.chart.labels.xAxisLabels.color,
                x: 0,
                y: 0
              })
            }, null, 8, Nl)) : (r(), u("text", {
              key: 0,
              class: "vue-ui-strip-plot-category-name",
              transform: `translate(${s.value.left + (o + 1) * s.value.stripWidth - s.value.stripWidth / 2}, ${s.value.bottom + e.value.style.chart.labels.xAxisLabels.fontSize * 2 + e.value.style.chart.labels.xAxisLabels.offsetY}), rotate(${e.value.style.chart.labels.xAxisLabels.rotation})`,
              "font-size": e.value.style.chart.labels.xAxisLabels.fontSize,
              fill: e.value.style.chart.labels.xAxisLabels.color,
              "text-anchor": e.value.style.chart.labels.xAxisLabels.rotation > 0 ? "start" : e.value.style.chart.labels.xAxisLabels.rotation < 0 ? "end" : "middle"
            }, _(String(l)), 9, Dl))
          ]))), 256))
        ], 512)) : f("", !0),
        e.value.style.chart.labels.axis.yLabel ? (r(), u("text", {
          key: 4,
          ref_key: "yAxisLabel",
          ref: re,
          fill: e.value.style.chart.labels.axis.color,
          "font-size": e.value.style.chart.labels.axis.fontSize,
          transform: `translate(${e.value.style.chart.labels.axis.fontSize}, ${s.value.top + s.value.height / 2}) rotate(-90)`,
          "text-anchor": "middle"
        }, _(e.value.style.chart.labels.axis.yLabel), 9, Il)) : f("", !0),
        e.value.style.chart.labels.axis.xLabel ? (r(), u("text", {
          key: 5,
          ref_key: "xAxisLabel",
          ref: se,
          fill: e.value.style.chart.labels.axis.color,
          "font-size": e.value.style.chart.labels.axis.fontSize,
          x: s.value.left + s.value.width / 2,
          y: $.value - e.value.style.chart.labels.axis.fontSize / 3,
          "text-anchor": "middle"
        }, _(e.value.style.chart.labels.axis.xLabel), 9, Fl)) : f("", !0),
        y.value ? (r(), u(x, { key: 6 }, [
          C("line", {
            x1: s.value.left,
            x2: s.value.right,
            y1: y.value.y,
            y2: y.value.y,
            stroke: y.value.color,
            "stroke-width": 1,
            class: H({ "select-circle": e.value.useCssAnimation })
          }, null, 10, Ml),
          C("circle", {
            cx: s.value.left,
            cy: y.value.y,
            r: 3,
            fill: y.value.color,
            class: H({ "select-circle": e.value.useCssAnimation })
          }, null, 10, Rl),
          C("circle", {
            cx: s.value.right,
            cy: y.value.y,
            r: 3,
            fill: y.value.color,
            class: H({ "select-circle": e.value.useCssAnimation })
          }, null, 10, Bl)
        ], 64)) : f("", !0),
        C("defs", null, [
          (r(!0), u(x, null, S(w.value, (l) => (r(), u("radialGradient", {
            id: l.id,
            fy: "30%"
          }, [
            C("stop", {
              offset: "10%",
              "stop-color": v(tl)(l.color, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, El),
            C("stop", {
              offset: "90%",
              "stop-color": v(ll)(l.color, 0.1)
            }, null, 8, Gl),
            C("stop", {
              offset: "100%",
              "stop-color": l.color
            }, null, 8, Hl)
          ], 8, Wl))), 256))
        ]),
        (r(!0), u(x, null, S(ke.value, (l, o) => (r(), u(x, null, [
          (r(!0), u(x, null, S(l.plots, (n, p) => (r(), z(ml, lt({ ref_for: !0 }, t.$attrs, {
            plot: { x: n.x, y: Ne.value ? n.y : s.value.top },
            radius: y.value && y.value.id === n.id ? ye.value * 1.5 : ye.value,
            shape: e.value.style.chart.plots.shape,
            stroke: e.value.style.chart.plots.stroke,
            strokeWidth: e.value.style.chart.plots.strokeWidth,
            color: e.value.style.chart.plots.gradient.show ? `url(#${l.id})` : l.color,
            style: `transition: all 0.2s ease-in-out; opacity:${y.value ? y.value.id === n.id ? 1 : 0.2 : e.value.style.chart.plots.opacity};${B.value ? `transition-delay:${p * 50}ms` : ""}`,
            class: { "vue-ui-strip-plot-animated": e.value.useCssAnimation && B.value && !v(ue), "vue-ui-strip-plot-select-circle": e.value.useCssAnimation && !B.value },
            onMouseenter: (F) => Nt({ datapoint: n, seriesIndex: p }),
            onMouseleave: (F) => zt({ datapoint: n, seriesIndex: p }),
            onClick: (F) => Dt({ datapoint: n, seriesIndex: p })
          }), null, 16, ["plot", "radius", "shape", "stroke", "strokeWidth", "color", "style", "class", "onMouseenter", "onMouseleave", "onClick"]))), 256)),
          b.value.dataLabels.show ? (r(), u("g", Ul, [
            (r(!0), u(x, null, S(l.plots, (n, p) => (r(), u(x, null, [
              p === l.plots.length - 1 || y.value && y.value.id === n.id && !b.value.showTooltip ? (r(), u("text", {
                key: 0,
                x: n.x,
                y: n.y + e.value.style.chart.labels.bestPlotLabel.offsetY - ye.value * (y.value && y.value.id === n.id && !b.value.showTooltip ? 2 : 1.5),
                "font-size": e.value.style.chart.labels.bestPlotLabel.fontSize,
                fill: e.value.style.chart.labels.bestPlotLabel.color,
                "text-anchor": "middle",
                style: G(`opacity:${e.value.useCssAnimation && B.value ? 0 : 1};transition:opacity 0.2s ease-in;`)
              }, _(n.name) + " " + _(e.value.style.chart.labels.bestPlotLabel.showValue ? v(Te)(
                e.value.style.chart.labels.formatter,
                n.value,
                v(ee)({
                  p: `(${e.value.style.chart.labels.prefix}`,
                  v: n.value,
                  s: `${e.value.style.chart.labels.suffix})`,
                  r: e.value.style.chart.labels.bestPlotLabel.rounding
                }),
                { datapoint: n, seriesIndex: p }
              ) : ""), 13, Vl)) : f("", !0)
            ], 64))), 256))
          ])) : f("", !0)
        ], 64))), 256)),
        h(t.$slots, "svg", {
          svg: {
            ...s.value,
            isPrintingImg: v(fe) | v(be) | V.value,
            isPrintingSvg: X.value
          }
        }, void 0, !0)
      ], 14, wl)),
      t.$slots.watermark ? (r(), u("div", Xl, [
        h(t.$slots, "watermark", g(k({ isPrinting: v(fe) || v(be) || V.value || X.value })), void 0, !0)
      ])) : f("", !0),
      t.$slots.source ? (r(), u("div", {
        key: 5,
        ref_key: "source",
        ref: ze,
        dir: "auto"
      }, [
        h(t.$slots, "source", {}, void 0, !0)
      ], 512)) : f("", !0),
      Ae(v(pt), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: b.value.showTooltip && le.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        fontSize: e.value.style.chart.tooltip.fontSize,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: A.value,
        content: ae.value,
        isFullscreen: O.value,
        isCustom: v(it)(e.value.style.chart.tooltip.customFormat),
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": c(() => [
          h(t.$slots, "tooltip-before", g(k({ ...xe.value })), void 0, !0)
        ]),
        "tooltip-after": c(() => [
          h(t.$slots, "tooltip-after", g(k({ ...xe.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      U.value && e.value.userOptions.buttons.table ? (r(), z(jt(E.value.component), lt({ key: 6 }, E.value.props, {
        ref_key: "tableUnit",
        ref: M,
        onClose: qe
      }), tt({
        content: c(() => [
          (r(), z(v(yt), {
            key: `table_${Fe.value}`,
            colNames: K.value.colNames,
            head: K.value.head,
            body: K.value.body,
            config: K.value.config,
            title: e.value.table.useDialog ? "" : E.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: R.value,
            onClose: qe
          }, {
            th: c(({ th: l }) => [
              C("div", {
                innerHTML: l,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, jl)
            ]),
            td: c(({ td: l }) => [
              at(_(l.name || l), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: c(() => [
            at(_(E.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: c(() => [
            C("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[0] || (a[0] = (l) => Ce(e.value.userOptions.callbacks.csv)),
              style: G({ cursor: R.value ? "pointer" : "default" })
            }, [
              Ae(v(ft), {
                name: "fileCsv",
                stroke: E.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : f("", !0),
      h(t.$slots, "skeleton", {}, () => [
        v(ue) ? (r(), z(vl, { key: 0 })) : f("", !0)
      ], !0)
    ], 46, Cl));
  }
}, ha = /* @__PURE__ */ xl(Yl, [["__scopeId", "data-v-1ba81519"]]);
export {
  ha as default
};
