// vue-data-ui/dist/vue-ui-table-heatmap-y88sBfu3.js
import { useCssVars as ne, defineAsyncComponent as se, ref as h, useSlots as le, onMounted as E, computed as v, watch as re, openBlock as s, createElementBlock as r, unref as i, normalizeClass as G, normalizeStyle as C, createBlock as ue, createSlots as ie, withCtx as g, renderSlot as u, normalizeProps as B, guardReactiveProps as A, createCommentVNode as f, createElementVNode as b, Fragment as k, renderList as w, mergeProps as m, createVNode as de, nextTick as pe } from "vue";
import { j as H, o as ce, q as ve, ao as fe, W as me, an as he, a0 as ge, a1 as be } from "./lib-GwPnJSrY.js";
import { u as ye } from "./useConfig-CIgyKZ6R.js";
import { u as Ce } from "./usePrinter-DiK5R6k-.js";
import { u as O } from "./useNestedProp-BBT0mbHh.js";
import { u as ke } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Ie } from "./useUserOptionState-BIvW1Kz7.js";
import _e from "./vue_ui_table_heatmap-CkvipPVy.js";
import { _ as $e } from "./Shape-CSyWYgXc.js";
import { _ as we } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Oe = ["id"], xe = { role: "cell" }, Fe = { role: "row" }, Pe = ["data-cell"], Ne = {
  key: 0,
  style: { display: "flex", "flex-direction": "row", gap: "2px", "align-items": "center" }
}, Te = ["height", "width"], Ve = {
  key: 0,
  role: "cell",
  "data-cell": "sum"
}, Se = {
  key: 1,
  role: "cell",
  "data-cell": "average"
}, Re = {
  key: 2,
  role: "cell",
  "data-cell": "median"
}, Me = {
  key: 1,
  ref: "source",
  dir: "auto"
}, Ue = {
  __name: "vue-ui-table-heatmap",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  setup(D, { expose: L }) {
    ne((e) => ({
      v7bb1d4dc: oe.value,
      v4b487b6d: te.value
    }));
    const W = se(() => import("./UserOptions-CrLqEimB.js")), { vue_ui_table_heatmap: X } = ye(), { isThemeValid: q, warnInvalidTheme: J } = ke(), c = D, I = h(H()), d = h(!1), y = h(null), _ = h(!1), x = h(0), K = le();
    E(() => {
      K["chart-background"] && console.warn("VueUiTableHeatmap does not support the #chart-background slot.");
    });
    const t = v({
      get: () => N(),
      set: (e) => e
    }), { userOptionsVisible: $, setUserOptionsVisibility: F, keepUserOptionState: P } = Ie({ config: t.value });
    function N() {
      const e = O({
        userConfig: c.config,
        defaultConfig: X
      }), a = e.theme;
      if (!a) return e;
      if (!q.value(e))
        return J(e), e;
      const o = O({
        userConfig: _e[a] || c.config,
        defaultConfig: e
      });
      return O({
        userConfig: c.config,
        defaultConfig: o
      });
    }
    re(() => c.config, (e) => {
      t.value = N(), $.value = !t.value.userOptions.showOnChartHover, S();
    }, { deep: !0 });
    const { isPrinting: Q, isImaging: Y, generatePdf: T, generateImage: V } = Ce({
      elementId: `table_heatmap_${I.value}`,
      fileName: "vue-ui-table-heatmap",
      options: t.value.userOptions.print
    }), Z = v(() => t.value.table.responsiveBreakpoint), j = v(() => !!c.dataset && c.dataset.length);
    E(() => {
      S();
    });
    function S() {
      ce(c.dataset) && ve({
        componentName: "VueUiTableHeatmap",
        type: "dataset"
      });
      const e = new ResizeObserver((a) => {
        a.forEach((o) => {
          d.value = o.contentRect.width < Z.value;
        });
      });
      y.value && e.observe(y.value);
    }
    const R = v(() => c.dataset.map((e) => {
      const a = e.values.map((n) => isNaN(n) ? 0 : n), o = a.reduce((n, l) => n + l, 0);
      return {
        ...e,
        values: a,
        serieExtremes: {
          max: Math.max(...a),
          min: Math.min(...a)
        },
        sum: o,
        average: o / a.length,
        median: fe(a),
        displayValues: [e.name, ...e.values],
        id: H()
      };
    })), M = v(() => {
      const e = R.value.flatMap((a) => a.values);
      return {
        min: Math.min(...e),
        max: Math.max(...e)
      };
    });
    function ee(e, a) {
      const o = t.value.style.heatmapColors.useIndividualScale;
      return he(t.value.style.heatmapColors.min, t.value.style.heatmapColors.max, o ? a.min : M.value.min, o ? a.max : M.value.max, e);
    }
    const U = v(() => R.value.map((e) => ({
      ...e,
      colors: e.displayValues.map((a) => isNaN(a) ? t.value.style.backgroundColor : ee(a, e.serieExtremes))
    }))), te = v(() => t.value.style.backgroundColor), oe = v(() => `${t.value.table.borderWidth}px`);
    function z(e = null) {
      pe(() => {
        const a = U.value.map((l) => [
          [l.name],
          l.displayValues,
          [l.sum],
          [l.average],
          [l.median]
        ]), o = [[[""], t.value.table.head.values, ["sum"], ["average"], ["median"]]].concat(a), n = ge(o);
        e ? e(n) : be({ csvContent: n, title: "vue-ui-table-heatmap" });
      });
    }
    function ae(e) {
      _.value = e, x.value += 1;
    }
    return L({
      generatePdf: T,
      generateCsv: z,
      generateImage: V
    }), (e, a) => (s(), r("div", {
      ref_key: "tableContainer",
      ref: y,
      style: C(`width:100%; overflow-x:auto; container-type: inline-size;padding-top:${t.value.userOptions.show ? "36px" : ""}; ${_.value ? "vue-data-ui-wrapper-fullscreen" : ""}; position:relative;`),
      class: G({ "vue-data-ui-component": !0, "vue-ui-responsive": d.value }),
      id: `table_heatmap_${I.value}`,
      onMouseenter: a[0] || (a[0] = () => i(F)(!0)),
      onMouseleave: a[1] || (a[1] = () => i(F)(!1))
    }, [
      t.value.userOptions.show && j.value && (i(P) || i($)) ? (s(), ue(i(W), {
        ref: "details",
        key: `user_option_${x.value}`,
        backgroundColor: t.value.style.backgroundColor,
        color: t.value.style.color,
        isPrinting: i(Q),
        isImaging: i(Y),
        uid: I.value,
        hasPdf: t.value.userOptions.buttons.pdf,
        hasXls: t.value.userOptions.buttons.csv,
        hasImg: t.value.userOptions.buttons.img,
        hasFullscreen: t.value.userOptions.buttons.fullscreen,
        isFullscreen: _.value,
        titles: { ...t.value.userOptions.buttonTitles },
        chartElement: y.value,
        position: t.value.userOptions.position,
        callbacks: t.value.userOptions.callbacks,
        printScale: t.value.userOptions.print.scale,
        onToggleFullscreen: ae,
        onGeneratePdf: i(T),
        onGenerateCsv: z,
        onGenerateImage: i(V),
        style: C({
          visibility: i(P) ? i($) ? "visible" : "hidden" : "visible"
        })
      }, ie({ _: 2 }, [
        e.$slots.menuIcon ? {
          name: "menuIcon",
          fn: g(({ isOpen: o, color: n }) => [
            u(e.$slots, "menuIcon", B(A({ isOpen: o, color: n })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        e.$slots.optionPdf ? {
          name: "optionPdf",
          fn: g(() => [
            u(e.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        e.$slots.optionCsv ? {
          name: "optionCsv",
          fn: g(() => [
            u(e.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        e.$slots.optionImg ? {
          name: "optionImg",
          fn: g(() => [
            u(e.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        e.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: g(({ toggleFullscreen: o, isFullscreen: n }) => [
            u(e.$slots, "optionFullscreen", B(A({ toggleFullscreen: o, isFullscreen: n })), void 0, !0)
          ]),
          key: "4"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasXls", "hasImg", "hasFullscreen", "isFullscreen", "titles", "chartElement", "position", "callbacks", "printScale", "onGeneratePdf", "onGenerateImage", "style"])) : f("", !0),
      b("table", {
        class: G({ "vue-ui-table-heatmap": !0 }),
        style: C(`width:100%;font-family:${t.value.style.fontFamily};background:${t.value.style.backgroundColor};`)
      }, [
        b("caption", null, [
          u(e.$slots, "caption", {}, void 0, !0)
        ]),
        b("thead", null, [
          b("tr", {
            role: "row",
            style: C(`background:${t.value.table.head.backgroundColor};color:${t.value.table.head.color}`)
          }, [
            (s(!0), r(k, null, w(t.value.table.head.values, (o, n) => (s(), r("th", xe, [
              u(e.$slots, "head", m({ ref_for: !0 }, { value: o, rowIndex: n, type: typeof o, isResponsive: d.value }), void 0, !0)
            ]))), 256))
          ], 4)
        ]),
        b("tbody", null, [
          (s(!0), r(k, null, w(U.value, (o, n) => (s(), r("tr", Fe, [
            (s(!0), r(k, null, w(o.displayValues, (l, p) => (s(), r("td", {
              role: "cell",
              "data-cell": t.value.table.head.values[p]
            }, [
              o.color && p === 0 ? (s(), r("div", Ne, [
                o.color ? (s(), r("svg", {
                  key: 0,
                  height: t.value.style.shapeSize,
                  width: t.value.style.shapeSize,
                  viewBox: "0 0 20 20",
                  style: { background: "none", overflow: "visible" }
                }, [
                  de($e, {
                    plot: { x: 10, y: 10 },
                    color: o.color,
                    radius: 9,
                    shape: o.shape || "circle"
                  }, null, 8, ["color", "shape"])
                ], 8, Te)) : f("", !0),
                u(e.$slots, "rowTitle", m({ ref_for: !0 }, { value: l, rowIndex: n, colIndex: p, type: typeof l, isResponsive: d.value }), void 0, !0)
              ])) : (s(), r(k, { key: 1 }, [
                p === 0 ? u(e.$slots, "rowTitle", m({
                  key: 0,
                  ref_for: !0
                }, { value: l, rowIndex: n, colIndex: p, type: typeof l, isResponsive: d.value }), void 0, !0) : f("", !0),
                p > 0 ? u(e.$slots, "cell", m({
                  key: 1,
                  ref_for: !0
                }, { value: l, rowIndex: n, colIndex: p, type: typeof l, isResponsive: d.value, color: o.colors[p], textColor: i(me)(o.colors[p]) }), void 0, !0) : f("", !0)
              ], 64))
            ], 8, Pe))), 256)),
            t.value.table.showSum ? (s(), r("td", Ve, [
              u(e.$slots, "sum", m({ ref_for: !0 }, { value: o.sum, rowIndex: n, isResponsive: d.value }), void 0, !0)
            ])) : f("", !0),
            t.value.table.showAverage ? (s(), r("td", Se, [
              u(e.$slots, "average", m({ ref_for: !0 }, { value: o.average, rowIndex: n, isResponsive: d.value }), void 0, !0)
            ])) : f("", !0),
            t.value.table.showMedian ? (s(), r("td", Re, [
              u(e.$slots, "median", m({ ref_for: !0 }, { value: o.median, rowIndex: n, isResponsive: d.value }), void 0, !0)
            ])) : f("", !0)
          ]))), 256))
        ])
      ], 4),
      e.$slots.source ? (s(), r("div", Me, [
        u(e.$slots, "source", {}, void 0, !0)
      ], 512)) : f("", !0)
    ], 46, Oe));
  }
}, qe = /* @__PURE__ */ we(Ue, [["__scopeId", "data-v-9de1ab93"]]);
export {
  qe as default
};
