// vue-data-ui/dist/vue-ui-dag-DsHUvjvm.js
import { ref as E, watchEffect as vo, unref as p, defineAsyncComponent as lt, computed as W, toRefs as go, onMounted as Qt, nextTick as pe, watch as Re, onBeforeUnmount as en, openBlock as C, createElementBlock as x, normalizeStyle as ye, normalizeClass as tn, toDisplayString as Pe, createCommentVNode as T, createBlock as te, withCtx as V, renderSlot as M, normalizeProps as H, guardReactiveProps as G, createSlots as po, createVNode as ut, createElementVNode as J, Fragment as ue, renderList as De, mergeProps as we, withModifiers as nn, createTextVNode as yo, Transition as on, Teleport as sn } from "vue";
import { j as pt, aE as wo, t as rn, o as bo, q as an, s as ln, X as ko, ar as Eo } from "./lib-GwPnJSrY.js";
import { t as _o, u as Co } from "./useResponsive-DfdjqQps.js";
import { u as xo } from "./usePanZoom-CDF7rb3o.js";
import { u as No } from "./useConfig-CIgyKZ6R.js";
import { u as So, B as Lo } from "./BaseScanner-BhhkKVoj.js";
import { u as Mo } from "./usePrinter-DiK5R6k-.js";
import { u as Oo } from "./useSvgExport-ByUukOZt.js";
import { u as ct } from "./useNestedProp-BBT0mbHh.js";
import { u as To } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Ro } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Po } from "./useChartAccessibility-9icAAmYg.js";
import Do from "./Title-BfyrP-jp.js";
import $o from "./vue_ui_dag-ZOwuDBZ8.js";
import { B as un } from "./BaseZoomControls-CoFievy9.js";
import Ao from "./img-BYYzY_QU.js";
import { _ as Fo } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Bo = "\0", ce = "\0", cn = "";
class ne {
  _isDirected = !0;
  _isMultigraph = !1;
  _isCompound = !1;
  // Label for the graph itself
  _label;
  // Defaults to be set when creating a new node
  _defaultNodeLabelFn = () => {
  };
  // Defaults to be set when creating a new edge
  _defaultEdgeLabelFn = () => {
  };
  // v -> label
  _nodes = {};
  // v -> edgeObj
  _in = {};
  // u -> v -> Number
  _preds = {};
  // v -> edgeObj
  _out = {};
  // v -> w -> Number
  _sucs = {};
  // e -> edgeObj
  _edgeObjs = {};
  // e -> label
  _edgeLabels = {};
  /* Number of nodes in the graph. Should only be changed by the implementation. */
  _nodeCount = 0;
  /* Number of edges in the graph. Should only be changed by the implementation. */
  _edgeCount = 0;
  _parent;
  _children;
  constructor(t) {
    t && (this._isDirected = Object.hasOwn(t, "directed") ? t.directed : !0, this._isMultigraph = Object.hasOwn(t, "multigraph") ? t.multigraph : !1, this._isCompound = Object.hasOwn(t, "compound") ? t.compound : !1), this._isCompound && (this._parent = {}, this._children = {}, this._children[ce] = {});
  }
  /* === Graph functions ========= */
  /**
   * Whether graph was created with 'directed' flag set to true or not.
   */
  isDirected() {
    return this._isDirected;
  }
  /**
   * Whether graph was created with 'multigraph' flag set to true or not.
   */
  isMultigraph() {
    return this._isMultigraph;
  }
  /**
   * Whether graph was created with 'compound' flag set to true or not.
   */
  isCompound() {
    return this._isCompound;
  }
  /**
   * Sets the label of the graph.
   */
  setGraph(t) {
    return this._label = t, this;
  }
  /**
   * Gets the graph label.
   */
  graph() {
    return this._label;
  }
  /* === Node functions ========== */
  /**
   * Sets the default node label. If newDefault is a function, it will be
   * invoked each time when setting a label for a node. Otherwise, this label
   * will be assigned as default label in case if no label was specified while
   * setting a node.
   * Complexity: O(1).
   */
  setDefaultNodeLabel(t) {
    return this._defaultNodeLabelFn = t, typeof t != "function" && (this._defaultNodeLabelFn = () => t), this;
  }
  /**
   * Gets the number of nodes in the graph.
   * Complexity: O(1).
   */
  nodeCount() {
    return this._nodeCount;
  }
  /**
   * Gets all nodes of the graph. Note, the in case of compound graph subnodes are
   * not included in list.
   * Complexity: O(1).
   */
  nodes() {
    return Object.keys(this._nodes);
  }
  /**
   * Gets list of nodes without in-edges.
   * Complexity: O(|V|).
   */
  sources() {
    const t = this;
    return this.nodes().filter((n) => Object.keys(t._in[n]).length === 0);
  }
  /**
   * Gets list of nodes without out-edges.
   * Complexity: O(|V|).
   */
  sinks() {
    const t = this;
    return this.nodes().filter((n) => Object.keys(t._out[n]).length === 0);
  }
  /**
   * Invokes setNode method for each node in names list.
   * Complexity: O(|names|).
   */
  setNodes(t, n) {
    const o = arguments, s = this;
    return t.forEach((r) => {
      o.length > 1 ? s.setNode(r, n) : s.setNode(r);
    }), this;
  }
  /**
   * Creates or updates the value for the node v in the graph. If value is supplied
   * it is set as the value for the node. If value is not supplied and the node was
   * created by this call then the default node label will be assigned.
   * Complexity: O(1).
   */
  setNode(t, n) {
    return Object.hasOwn(this._nodes, t) ? (arguments.length > 1 && (this._nodes[t] = n), this) : (this._nodes[t] = arguments.length > 1 ? n : this._defaultNodeLabelFn(t), this._isCompound && (this._parent[t] = ce, this._children[t] = {}, this._children[ce][t] = !0), this._in[t] = {}, this._preds[t] = {}, this._out[t] = {}, this._sucs[t] = {}, ++this._nodeCount, this);
  }
  /**
   * Gets the label of node with specified name.
   * Complexity: O(1).
   */
  node(t) {
    return this._nodes[t];
  }
  /**
   * Detects whether graph has a node with specified name or not.
   */
  hasNode(t) {
    return Object.hasOwn(this._nodes, t);
  }
  /**
   * Remove the node with the name from the graph or do nothing if the node is not in
   * the graph. If the node was removed this function also removes any incident
   * edges.
   * Complexity: O(1).
   */
  removeNode(t) {
    const n = this;
    if (Object.hasOwn(this._nodes, t)) {
      const o = (s) => n.removeEdge(n._edgeObjs[s]);
      delete this._nodes[t], this._isCompound && (this._removeFromParentsChildList(t), delete this._parent[t], this.children(t).forEach((s) => {
        n.setParent(s);
      }), delete this._children[t]), Object.keys(this._in[t]).forEach(o), delete this._in[t], delete this._preds[t], Object.keys(this._out[t]).forEach(o), delete this._out[t], delete this._sucs[t], --this._nodeCount;
    }
    return this;
  }
  /**
   * Sets node parentId as a parent for node nodeId if it is defined, or removes the
   * parent for nodeId if parentId is undefined. Method throws an exception in case of
   * invoking it in context of noncompound graph.
   * Average-case complexity: O(1).
   */
  setParent(t, n) {
    if (!this._isCompound)
      throw new Error("Cannot set parent in a non-compound graph");
    if (n === void 0)
      n = ce;
    else {
      n += "";
      for (let o = n; o !== void 0; o = this.parent(o))
        if (o === t)
          throw new Error(
            "Setting " + n + " as parent of " + t + " would create a cycle"
          );
      this.setNode(n);
    }
    return this.setNode(t), this._removeFromParentsChildList(t), this._parent[t] = n, this._children[n][t] = !0, this;
  }
  _removeFromParentsChildList(t) {
    delete this._children[this._parent[t]][t];
  }
  /**
   * Gets parent node for node nodeId.
   * Complexity: O(1).
   */
  parent(t) {
    if (this._isCompound) {
      const n = this._parent[t];
      if (n !== ce)
        return n;
    }
  }
  /**
   * Gets list of direct children of node nodeId.
   * Complexity: O(1).
   */
  children(t = ce) {
    if (this._isCompound) {
      const n = this._children[t];
      if (n)
        return Object.keys(n);
    } else {
      if (t === ce)
        return this.nodes();
      if (this.hasNode(t))
        return [];
    }
  }
  /**
   * Return all nodes that are predecessors of the specified node or undefined if nodeId is not in
   * the graph. Behavior is undefined for undirected graphs - use neighbors instead.
   * Complexity: O(|V|).
   */
  predecessors(t) {
    const n = this._preds[t];
    if (n)
      return Object.keys(n);
  }
  /**
   * Return all nodes that are successors of the specified node or undefined if nodeId is not in
   * the graph. Behavior is undefined for undirected graphs - use neighbors instead.
   * Complexity: O(|V|).
   */
  successors(t) {
    const n = this._sucs[t];
    if (n)
      return Object.keys(n);
  }
  /**
   * Return all nodes that are predecessors or successors of the specified node or undefined if
   * nodeId is not in the graph.
   * Complexity: O(|V|).
   */
  neighbors(t) {
    const n = this.predecessors(t);
    if (n) {
      const o = new Set(n);
      for (const s of this.successors(t))
        o.add(s);
      return Array.from(o.values());
    }
  }
  isLeaf(t) {
    let n;
    return this.isDirected() ? n = this.successors(t) : n = this.neighbors(t), n.length === 0;
  }
  /**
   * Creates new graph with nodes filtered via filter. Edges incident to rejected node
   * are also removed. In case of compound graph, if parent is rejected by filter,
   * then all its children are rejected too.
   * Average-case complexity: O(|E|+|V|).
   */
  filterNodes(t) {
    const n = new this.constructor({
      directed: this._isDirected,
      multigraph: this._isMultigraph,
      compound: this._isCompound
    });
    n.setGraph(this.graph());
    const o = this;
    Object.entries(this._nodes).forEach(([i, a]) => {
      t(i) && n.setNode(i, a);
    }), Object.values(this._edgeObjs).forEach((i) => {
      n.hasNode(i.v) && n.hasNode(i.w) && n.setEdge(i, o.edge(i));
    });
    const s = {};
    function r(i) {
      const a = o.parent(i);
      return a === void 0 || n.hasNode(a) ? (s[i] = a, a) : a in s ? s[a] : r(a);
    }
    return this._isCompound && n.nodes().forEach((i) => n.setParent(i, r(i))), n;
  }
  /* === Edge functions ========== */
  /**
   * Sets the default edge label or factory function. This label will be
   * assigned as default label in case if no label was specified while setting
   * an edge or this function will be invoked each time when setting an edge
   * with no label specified and returned value will be used as a label for edge.
   * Complexity: O(1).
   */
  setDefaultEdgeLabel(t) {
    return this._defaultEdgeLabelFn = t, typeof t != "function" && (this._defaultEdgeLabelFn = () => t), this;
  }
  /**
   * Gets the number of edges in the graph.
   * Complexity: O(1).
   */
  edgeCount() {
    return this._edgeCount;
  }
  /**
   * Gets edges of the graph. In case of compound graph subgraphs are not considered.
   * Complexity: O(|E|).
   */
  edges() {
    return Object.values(this._edgeObjs);
  }
  /**
   * Establish an edges path over the nodes in nodeIds list. If some edge already
   * exists, it will update its label, otherwise it will create an edge between pair
   * of nodes with label provided or default label if no label provided.
   * Complexity: O(|nodeIds|).
   */
  setPath(t, n) {
    const o = this, s = arguments;
    return t.reduce((r, i) => (s.length > 1 ? o.setEdge(r, i, n) : o.setEdge(r, i), i)), this;
  }
  /**
   * Creates or updates the label for the edge (v, w) with the optionally supplied
   * name. If value is supplied it is set as the value for the edge. If value is not
   * supplied and the edge was created by this call then the default edge label will
   * be assigned. The name parameter is only useful with multigraphs.
   */
  setEdge() {
    let t, n, o, s, r = !1;
    const i = arguments[0];
    typeof i == "object" && i !== null && "v" in i ? (t = i.v, n = i.w, o = i.name, arguments.length === 2 && (s = arguments[1], r = !0)) : (t = i, n = arguments[1], o = arguments[3], arguments.length > 2 && (s = arguments[2], r = !0)), t = "" + t, n = "" + n, o !== void 0 && (o = "" + o);
    let a = $e(this._isDirected, t, n, o);
    if (Object.hasOwn(this._edgeLabels, a))
      return r && (this._edgeLabels[a] = s), this;
    if (o !== void 0 && !this._isMultigraph)
      throw new Error("Cannot set a named edge when isMultigraph = false");
    this.setNode(t), this.setNode(n), this._edgeLabels[a] = r ? s : this._defaultEdgeLabelFn(t, n, o);
    const u = zo(this._isDirected, t, n, o);
    return t = u.v, n = u.w, Object.freeze(u), this._edgeObjs[a] = u, dn(this._preds[n], t), dn(this._sucs[t], n), this._in[n][a] = u, this._out[t][a] = u, this._edgeCount++, this;
  }
  /**
   * Gets the label for the specified edge.
   * Complexity: O(1).
   */
  edge(t, n, o) {
    const s = arguments.length === 1 ? dt(this._isDirected, arguments[0]) : $e(this._isDirected, t, n, o);
    return this._edgeLabels[s];
  }
  /**
   * Gets the label for the specified edge and converts it to an object.
   * Complexity: O(1)
   */
  edgeAsObj() {
    const t = this.edge(...arguments);
    return typeof t != "object" ? { label: t } : t;
  }
  /**
   * Detects whether the graph contains specified edge or not. No subgraphs are considered.
   * Complexity: O(1).
   */
  hasEdge(t, n, o) {
    const s = arguments.length === 1 ? dt(this._isDirected, arguments[0]) : $e(this._isDirected, t, n, o);
    return Object.hasOwn(this._edgeLabels, s);
  }
  /**
   * Removes the specified edge from the graph. No subgraphs are considered.
   * Complexity: O(1).
   */
  removeEdge(t, n, o) {
    const s = arguments.length === 1 ? dt(this._isDirected, arguments[0]) : $e(this._isDirected, t, n, o), r = this._edgeObjs[s];
    return r && (t = r.v, n = r.w, delete this._edgeLabels[s], delete this._edgeObjs[s], fn(this._preds[n], t), fn(this._sucs[t], n), delete this._in[n][s], delete this._out[t][s], this._edgeCount--), this;
  }
  /**
   * Return all edges that point to the node v. Optionally filters those edges down to just those
   * coming from node u. Behavior is undefined for undirected graphs - use nodeEdges instead.
   * Complexity: O(|E|).
   */
  inEdges(t, n) {
    const o = this._in[t];
    if (o) {
      const s = Object.values(o);
      return n ? s.filter((r) => r.v === n) : s;
    }
  }
  /**
   * Return all edges that are pointed at by node v. Optionally filters those edges down to just
   * those pointing to w. Behavior is undefined for undirected graphs - use nodeEdges instead.
   * Complexity: O(|E|).
   */
  outEdges(t, n) {
    const o = this._out[t];
    if (o) {
      const s = Object.values(o);
      return n ? s.filter((r) => r.w === n) : s;
    }
  }
  /**
   * Returns all edges to or from node v regardless of direction. Optionally filters those edges
   * down to just those between nodes v and w regardless of direction.
   * Complexity: O(|E|).
   */
  nodeEdges(t, n) {
    const o = this.inEdges(t, n);
    if (o)
      return o.concat(this.outEdges(t, n));
  }
}
function dn(e, t) {
  e[t] ? e[t]++ : e[t] = 1;
}
function fn(e, t) {
  --e[t] || delete e[t];
}
function $e(e, t, n, o) {
  let s = "" + t, r = "" + n;
  if (!e && s > r) {
    const i = s;
    s = r, r = i;
  }
  return s + cn + r + cn + (o === void 0 ? Bo : o);
}
function zo(e, t, n, o) {
  let s = "" + t, r = "" + n;
  if (!e && s > r) {
    const a = s;
    s = r, r = a;
  }
  const i = { v: s, w: r };
  return o && (i.name = o), i;
}
function dt(e, t) {
  return $e(e, t.v, t.w, t.name);
}
class Vo {
  constructor() {
    const t = {};
    t._next = t._prev = t, this._sentinel = t;
  }
  dequeue() {
    const t = this._sentinel, n = t._prev;
    if (n !== t)
      return hn(n), n;
  }
  enqueue(t) {
    const n = this._sentinel;
    t._prev && t._next && hn(t), t._next = n._next, n._next._prev = t, n._next = t, t._prev = n;
  }
  toString() {
    const t = [], n = this._sentinel;
    let o = n._prev;
    for (; o !== n; )
      t.push(JSON.stringify(o, Io)), o = o._prev;
    return "[" + t.join(", ") + "]";
  }
}
function hn(e) {
  e._prev._next = e._next, e._next._prev = e._prev, delete e._next, delete e._prev;
}
function Io(e, t) {
  if (e !== "_next" && e !== "_prev")
    return t;
}
const Ho = () => 1;
function Go(e, t) {
  if (e.nodeCount() <= 1)
    return [];
  const n = Yo(e, t || Ho);
  return Wo(n.graph, n.buckets, n.zeroIndex).flatMap((s) => e.outEdges(s.v, s.w));
}
function Wo(e, t, n) {
  let o = [];
  const s = t[t.length - 1], r = t[0];
  let i;
  for (; e.nodeCount(); ) {
    for (; i = r.dequeue(); )
      ft(e, t, n, i);
    for (; i = s.dequeue(); )
      ft(e, t, n, i);
    if (e.nodeCount()) {
      for (let a = t.length - 2; a > 0; --a)
        if (i = t[a].dequeue(), i) {
          o = o.concat(ft(e, t, n, i, !0));
          break;
        }
    }
  }
  return o;
}
function ft(e, t, n, o, s) {
  const r = s ? [] : void 0;
  return e.inEdges(o.v).forEach((i) => {
    const a = e.edge(i), u = e.node(i.v);
    s && r.push({ v: i.v, w: i.w }), u.out -= a, yt(t, n, u);
  }), e.outEdges(o.v).forEach((i) => {
    const a = e.edge(i), u = e.node(i.w);
    u.in -= a, yt(t, n, u);
  }), e.removeNode(o.v), r;
}
function Yo(e, t) {
  const n = new ne();
  let o = 0, s = 0;
  e.nodes().forEach((a) => {
    n.setNode(a, { v: a, in: 0, out: 0 });
  }), e.edges().forEach((a) => {
    const u = n.edge(a.v, a.w) || 0, d = t(a), h = u + d;
    n.setEdge(a.v, a.w, h), s = Math.max(s, n.node(a.v).out += d), o = Math.max(o, n.node(a.w).in += d);
  });
  const r = jo(s + o + 3).map(() => new Vo()), i = o + 1;
  return n.nodes().forEach((a) => {
    yt(r, i, n.node(a));
  }), {
    graph: n,
    buckets: r,
    zeroIndex: i
  };
}
function yt(e, t, n) {
  n.out ? n.in ? e[n.out - n.in + t].enqueue(n) : e[e.length - 1].enqueue(n) : e[0].enqueue(n);
}
function jo(e) {
  const t = [];
  for (let n = 0; n < e; n++)
    t.push(n);
  return t;
}
function ke(e, t, n, o) {
  let s = o;
  for (; e.hasNode(s); )
    s = je(o);
  return n.dummy = t, e.setNode(s, n), s;
}
function Cn(e) {
  const t = new ne().setGraph(e.graph());
  return e.nodes().forEach((n) => {
    t.setNode(n, e.node(n));
  }), e.edges().forEach((n) => {
    const o = t.edge(n.v, n.w) || { weight: 0, minlen: 1 }, s = e.edge(n);
    t.setEdge(n.v, n.w, {
      weight: o.weight + s.weight,
      minlen: Math.max(o.minlen, s.minlen)
    });
  }), t;
}
function bt(e) {
  const t = new ne({ multigraph: e.isMultigraph() }).setGraph(e.graph());
  return e.nodes().forEach((n) => {
    e.children(n).length || t.setNode(n, e.node(n));
  }), e.edges().forEach((n) => {
    t.setEdge(n, e.edge(n));
  }), t;
}
function Uo(e) {
  const t = e.nodes().map((n) => {
    const o = {};
    return e.outEdges(n).forEach((s) => {
      o[s.w] = (o[s.w] || 0) + e.edge(s).weight;
    }), o;
  });
  return Ue(e.nodes(), t);
}
function Zo(e) {
  const t = e.nodes().map((n) => {
    const o = {};
    return e.inEdges(n).forEach((s) => {
      o[s.v] = (o[s.v] || 0) + e.edge(s).weight;
    }), o;
  });
  return Ue(e.nodes(), t);
}
function wt(e, t) {
  let n = e.x, o = e.y;
  const s = t.x - n, r = t.y - o;
  let i = e.width / 2, a = e.height / 2;
  if (!s && !r)
    throw new Error("Not possible to find intersection inside of the rectangle");
  let u, d;
  return Math.abs(r) * i > Math.abs(s) * a ? (r < 0 && (a = -a), u = a * s / r, d = a) : (s < 0 && (i = -i), u = i, d = i * r / s), { x: n + u, y: o + d };
}
function Fe(e) {
  const t = be(kt(e) + 1).map(() => []);
  return e.nodes().forEach((n) => {
    const o = e.node(n), s = o.rank;
    s !== void 0 && (t[s][o.order] = n);
  }), t;
}
function xn(e) {
  const t = e.nodes().map((o) => {
    const s = e.node(o).rank;
    return s === void 0 ? Number.MAX_VALUE : s;
  }), n = Z(Math.min, t);
  e.nodes().forEach((o) => {
    const s = e.node(o);
    Object.hasOwn(s, "rank") && (s.rank -= n);
  });
}
function Nn(e) {
  const t = e.nodes().map((i) => e.node(i).rank).filter((i) => i !== void 0), n = Z(Math.min, t), o = [];
  e.nodes().forEach((i) => {
    const a = e.node(i).rank - n;
    o[a] || (o[a] = []), o[a].push(i);
  });
  let s = 0;
  const r = e.graph().nodeRankFactor;
  Array.from(o).forEach((i, a) => {
    i === void 0 && a % r !== 0 ? --s : i !== void 0 && s && i.forEach((u) => {
      e.node(u).rank += s;
    });
  });
}
function Xo(e, t, n, o) {
  const s = {
    width: 0,
    height: 0
  };
  return arguments.length >= 4 && (s.rank = n, s.order = o), ke(e, "border", s, t);
}
const Sn = 65535;
function qo(e, t = Sn) {
  const n = [];
  for (let o = 0; o < e.length; o += t) {
    const s = e.slice(o, o + t);
    n.push(s);
  }
  return n;
}
function Z(e, t) {
  if (t.length > Sn) {
    const n = qo(t);
    return e.apply(
      null,
      n.map((o) => e.apply(null, o))
    );
  } else
    return e.apply(null, t);
}
function kt(e) {
  const n = e.nodes().map((o) => {
    const s = e.node(o).rank;
    return s === void 0 ? Number.MIN_VALUE : s;
  });
  return Z(Math.max, n);
}
function Ln(e, t) {
  const n = { lhs: [], rhs: [] };
  return e.forEach((o) => {
    t(o) ? n.lhs.push(o) : n.rhs.push(o);
  }), n;
}
function Mn(e, t) {
  const n = Date.now();
  try {
    return t();
  } finally {
    console.log(e + " time: " + (Date.now() - n) + "ms");
  }
}
function On(e, t) {
  return t();
}
let Ko = 0;
function je(e) {
  const t = ++Ko;
  return e + String(t);
}
function be(e, t, n = 1) {
  t == null && (t = e, e = 0);
  let o = (r) => r < t;
  n < 0 && (o = (r) => t < r);
  const s = [];
  for (let r = e; o(r); r += n)
    s.push(r);
  return s;
}
function Ae(e, t) {
  const n = {};
  for (const o of t)
    e[o] !== void 0 && (n[o] = e[o]);
  return n;
}
function Be(e, t) {
  let n = t;
  if (typeof t == "string") {
    const o = t;
    n = (s) => s[o];
  }
  return Object.entries(e).reduce((o, [s, r]) => (o[s] = n(r, s), o), {});
}
function Ue(e, t) {
  return e.reduce((n, o, s) => (n[o] = t[s], n), {});
}
const Ye = {
  addBorderNode: Xo,
  addDummyNode: ke,
  applyWithChunking: Z,
  asNonCompoundGraph: bt,
  buildLayerMatrix: Fe,
  intersectRect: wt,
  mapValues: Be,
  maxRank: kt,
  normalizeRanks: xn,
  notime: On,
  partition: Ln,
  pick: Ae,
  predecessorWeights: Zo,
  range: be,
  removeEmptyRanks: Nn,
  simplify: Cn,
  successorWeights: Uo,
  time: Mn,
  uniqueId: je,
  zipObject: Ue
};
function Jo(e) {
  (e.graph().acyclicer === "greedy" ? Go(e, n(e)) : es(e)).forEach((o) => {
    const s = e.edge(o);
    e.removeEdge(o), s.forwardName = o.name, s.reversed = !0, e.setEdge(o.w, o.v, s, je("rev"));
  });
  function n(o) {
    return (s) => o.edge(s).weight;
  }
}
function Qo(e) {
  e.edges().forEach((t) => {
    const n = e.edge(t);
    if (n.reversed) {
      e.removeEdge(t);
      const o = n.forwardName;
      delete n.reversed, delete n.forwardName, e.setEdge(t.w, t.v, n, o);
    }
  });
}
function es(e) {
  const t = [], n = {}, o = {};
  function s(r) {
    Object.hasOwn(o, r) || (o[r] = !0, n[r] = !0, e.outEdges(r).forEach((i) => {
      Object.hasOwn(n, i.w) ? t.push(i) : s(i.w);
    }), delete n[r]);
  }
  return e.nodes().forEach(s), t;
}
function ts(e) {
  e.graph().dummyChains = [], e.edges().forEach((t) => ns(e, t));
}
function ns(e, t) {
  let n = t.v, o = e.node(n).rank;
  const s = t.w, r = e.node(s).rank, i = t.name, a = e.edge(t), u = a.labelRank;
  if (r === o + 1)
    return;
  e.removeEdge(t);
  let d, h, m;
  for (m = 0, ++o; o < r; ++m, ++o)
    a.points = [], h = {
      width: 0,
      height: 0,
      edgeLabel: a,
      edgeObj: t,
      rank: o
    }, d = ke(e, "edge", h, "_d"), o === u && (h.width = a.width, h.height = a.height, h.dummy = "edge-label", h.labelpos = a.labelpos), e.setEdge(n, d, { weight: a.weight }, i), m === 0 && e.graph().dummyChains.push(d), n = d;
  e.setEdge(n, s, { weight: a.weight }, i);
}
function os(e) {
  e.graph().dummyChains.forEach((t) => {
    let n = e.node(t);
    const o = n.edgeLabel;
    e.setEdge(n.edgeObj, o);
    let s;
    for (; n.dummy; )
      s = e.successors(t)[0], e.removeNode(t), o.points.push({ x: n.x, y: n.y }), n.dummy === "edge-label" && (o.x = n.x, o.y = n.y, o.width = n.width, o.height = n.height), t = s, n = e.node(t);
  });
}
function Tn(e) {
  function t(n) {
    const o = e.node(n);
    if (o && Object.prototype.hasOwnProperty.call(o, "rank"))
      return o.rank;
    const s = e.outEdges(n) || [];
    if (!s.length)
      return o && (o.rank = 0), 0;
    const r = s.map((a) => {
      if (!e.node(a.w))
        return Number.POSITIVE_INFINITY;
      const d = t(a.w), h = e.edge(a).minlen;
      return d - h;
    });
    let i = Z(Math.min, r);
    return i === Number.POSITIVE_INFINITY && (i = 0), o && (o.rank = i), i;
  }
  (e.sources() || []).forEach(t);
}
function ss(e, t) {
  return e.node(t.w).rank - e.node(t.v).rank - e.edge(t).minlen;
}
function is(e) {
  rs(e);
  const t = new ne();
  e.nodes().forEach((r) => {
    t.setNode(r, {});
  });
  const n = e.nodes();
  if (!n.length)
    return t;
  const o = n[0], s = /* @__PURE__ */ new Set([o]);
  for (; s.size < n.length; ) {
    const r = as(e, s);
    if (!r) {
      const h = n.find((m) => !s.has(m));
      s.add(h), t.setNode(h, {});
      continue;
    }
    const { edgeObject: i, delta: a, attachFrom: u, attachTo: d } = r;
    ls(e, s, u, a), t.setEdge(i.v, i.w, {}), s.add(d);
  }
  return t;
}
function rs(e) {
  e.nodes().forEach((t) => {
    const n = e.node(t) || {};
    Object.prototype.hasOwnProperty.call(n, "rank") || (n.rank = 0, e.setNode(t, n));
  });
}
function as(e, t) {
  let n = null;
  return e.edges().forEach((o) => {
    const s = t.has(o.v), r = t.has(o.w);
    if (s === r)
      return;
    const i = ss(e, o), a = Math.abs(i);
    if (!n || a < n.absoluteSlack) {
      const u = s ? o.v : o.w, d = s ? o.w : o.v, h = s ? i : -i;
      n = {
        edgeObject: o,
        delta: h,
        attachFrom: u,
        attachTo: d,
        absoluteSlack: a
      };
    }
  }), n;
}
function ls(e, t, n, o) {
  o && e.nodes().forEach((s) => {
    if (!t.has(s)) {
      const r = e.node(s);
      r.rank += o;
    }
  });
}
function Ee(e) {
  const t = Cn(e);
  return Tn(t), t.nodes().forEach((n) => {
    const o = t.node(n), s = e.node(n) || {};
    s.rank = o.rank, e.setNode(n, s);
  }), e;
}
function us(e, t) {
}
function cs(e, t) {
}
function ds(e, t, n) {
  return 0;
}
function fs(e) {
  return null;
}
function hs(e, t) {
  return null;
}
Ee.initLowLimValues = us;
Ee.initCutValues = cs;
Ee.calcCutValue = ds;
Ee.leaveEdge = fs;
Ee.enterEdge = hs;
const Rn = Tn;
function ms(e) {
  const t = e.graph().ranker;
  if (t instanceof Function)
    return t(e);
  switch (e.graph().ranker) {
    case "network-simplex":
      mn(e);
      break;
    case "tight-tree":
      gs(e);
      break;
    case "longest-path":
      vs(e);
      break;
    case "none":
      break;
    default:
      mn(e);
  }
}
const vs = Rn;
function gs(e) {
  Rn(e), is(e);
}
function mn(e) {
  Ee(e);
}
function ps(e) {
  const t = ws(e);
  e.graph().dummyChains.forEach((n) => {
    let o = e.node(n);
    const s = o.edgeObj, r = ys(e, t, s.v, s.w), i = r.path, a = r.lca;
    let u = 0, d = i[u], h = !0, m = n;
    for (; m !== s.w; ) {
      if (o = e.node(m), h) {
        for (; (d = i[u]) !== a && e.node(d).maxRank < o.rank; )
          u++;
        d === a && (h = !1);
      }
      if (!h) {
        for (; u < i.length - 1 && e.node(d = i[u + 1]).minRank <= o.rank; )
          u++;
        d = i[u];
      }
      e.setParent(m, d), m = e.successors(m)[0];
    }
  });
}
function ys(e, t, n, o) {
  const s = [], r = [], i = Math.min(t[n].low, t[o].low), a = Math.max(t[n].lim, t[o].lim);
  let u, d;
  u = n;
  do
    u = e.parent(u), s.push(u);
  while (u && (t[u].low > i || a > t[u].lim));
  for (d = u, u = o; (u = e.parent(u)) !== d; )
    r.push(u);
  return {
    path: s.concat(r.reverse()),
    lca: d
  };
}
function ws(e) {
  const t = {};
  let n = 0;
  function o(s) {
    const r = n;
    e.children(s).forEach(o), t[s] = { low: r, lim: n++ };
  }
  return e.children().forEach(o), t;
}
function bs(e) {
  const t = Ye.addDummyNode(e, "root", {}, "_root"), n = ks(e), o = Object.values(n), s = Ye.applyWithChunking(Math.max, o) - 1, r = 2 * s + 1;
  e.graph().nestingRoot = t, e.edges().forEach((a) => {
    e.edge(a).minlen *= r;
  });
  const i = Es(e) + 1;
  e.children().forEach((a) => {
    Pn(e, t, r, i, s, n, a);
  }), e.graph().nodeRankFactor = r;
}
function Pn(e, t, n, o, s, r, i) {
  const a = e.children(i);
  if (!a.length) {
    i !== t && e.setEdge(t, i, { weight: 0, minlen: n });
    return;
  }
  const u = Ye.addBorderNode(e, "_bt"), d = Ye.addBorderNode(e, "_bb"), h = e.node(i);
  e.setParent(u, i), h.borderTop = u, e.setParent(d, i), h.borderBottom = d, a.forEach((m) => {
    Pn(e, t, n, o, s, r, m);
    const v = e.node(m), b = v.borderTop ? v.borderTop : m, L = v.borderBottom ? v.borderBottom : m, N = v.borderTop ? o : 2 * o, z = b !== L ? 1 : s - r[i] + 1;
    e.setEdge(u, b, {
      weight: N,
      minlen: z,
      nestingEdge: !0
    }), e.setEdge(L, d, {
      weight: N,
      minlen: z,
      nestingEdge: !0
    });
  }), e.parent(i) || e.setEdge(t, u, {
    weight: 0,
    minlen: s + r[i]
  });
}
function ks(e) {
  const t = {};
  function n(o, s) {
    const r = e.children(o);
    r && r.length && r.forEach((i) => n(i, s + 1)), t[o] = s;
  }
  return e.children().forEach((o) => n(o, 1)), t;
}
function Es(e) {
  return e.edges().reduce((t, n) => t + e.edge(n).weight, 0);
}
function _s(e) {
  const t = e.graph();
  e.removeNode(t.nestingRoot), delete t.nestingRoot, e.edges().forEach((n) => {
    e.edge(n).nestingEdge && e.removeEdge(n);
  });
}
function Cs(e) {
  function t(n) {
    const o = e.children(n), s = e.node(n);
    if (o.length && o.forEach(t), Object.hasOwn(s, "minRank")) {
      s.borderLeft = [], s.borderRight = [];
      for (let r = s.minRank, i = s.maxRank + 1; r < i; ++r)
        vn(e, "borderLeft", "_bl", n, s, r), vn(e, "borderRight", "_br", n, s, r);
    }
  }
  e.children().forEach(t);
}
function vn(e, t, n, o, s, r) {
  const i = {
    width: 0,
    height: 0,
    rank: r,
    borderType: t
  }, a = s[t][r - 1], u = ke(e, "border", i, n);
  s[t][r] = u, e.setParent(u, o), a && e.setEdge(a, u, { weight: 1 });
}
function xs(e) {
  const t = e.graph().rankdir.toLowerCase();
  (t === "lr" || t === "rl") && Dn(e);
}
function Ns(e) {
  const t = e.graph().rankdir.toLowerCase();
  (t === "bt" || t === "rl") && Ss(e), (t === "lr" || t === "rl") && (Ls(e), Dn(e));
}
function Dn(e) {
  e.nodes().forEach((t) => {
    gn(e.node(t));
  }), e.edges().forEach((t) => {
    gn(e.edge(t));
  });
}
function gn(e) {
  const t = e.width;
  e.width = e.height, e.height = t;
}
function Ss(e) {
  e.nodes().forEach((t) => {
    ht(e.node(t));
  }), e.edges().forEach((t) => {
    const n = e.edge(t);
    n.points.forEach(ht), Object.hasOwn(n, "y") && ht(n);
  });
}
function ht(e) {
  e.y = -e.y;
}
function Ls(e) {
  e.nodes().forEach((t) => {
    mt(e.node(t));
  }), e.edges().forEach((t) => {
    const n = e.edge(t);
    n.points.forEach(mt), Object.hasOwn(n, "x") && mt(n);
  });
}
function mt(e) {
  const t = e.x;
  e.x = e.y, e.y = t;
}
const pn = {
  adjust: xs,
  undo: Ns
};
function Ms(e) {
  const t = {}, n = e.nodes().filter((u) => e.children(u).length === 0), o = n.map((u) => e.node(u).rank), s = Z(Math.max, o), r = be(s + 1).map(() => []);
  function i(u) {
    if (t[u]) return;
    t[u] = !0;
    const d = e.node(u);
    r[d.rank].push(u), e.successors(u).forEach(i);
  }
  return n.sort(
    (u, d) => e.node(u).rank - e.node(d).rank
  ).forEach(i), r;
}
function Os(e, t) {
  let n = 0;
  for (let o = 1; o < t.length; ++o)
    n += Ts(e, t[o - 1], t[o]);
  return n;
}
function Ts(e, t, n) {
  const o = Ue(
    n,
    n.map((d, h) => h)
  ), s = t.flatMap((d) => e.outEdges(d).map((h) => ({
    pos: o[h.w],
    weight: e.edge(h).weight
  })).sort((h, m) => h.pos - m.pos));
  let r = 1;
  for (; r < n.length; ) r <<= 1;
  const i = 2 * r - 1;
  r -= 1;
  const a = new Array(i).fill(0);
  let u = 0;
  return s.forEach((d) => {
    let h = d.pos + r;
    a[h] += d.weight;
    let m = 0;
    for (; h > 0; )
      h % 2 && (m += a[h + 1]), h = h - 1 >> 1, a[h] += d.weight;
    u += d.weight * m;
  }), u;
}
function Rs(e, t = []) {
  return t.map((n) => {
    const o = e.inEdges(n);
    if (!o.length)
      return { v: n };
    const s = o.reduce(
      (r, i) => {
        const a = e.edge(i), u = e.node(i.v);
        return {
          sum: r.sum + a.weight * u.order,
          weight: r.weight + a.weight
        };
      },
      { sum: 0, weight: 0 }
    );
    return {
      v: n,
      barycenter: s.sum / s.weight,
      weight: s.weight
    };
  });
}
function Ps(e, t) {
  const n = {};
  e.forEach((s, r) => {
    const i = n[s.v] = {
      indegree: 0,
      in: [],
      out: [],
      vs: [s.v],
      i: r
    };
    s.barycenter !== void 0 && (i.barycenter = s.barycenter, i.weight = s.weight);
  }), t.edges().forEach((s) => {
    const r = n[s.v], i = n[s.w];
    r !== void 0 && i !== void 0 && (i.indegree++, r.out.push(i));
  });
  const o = Object.values(n).filter((s) => !s.indegree);
  return Ds(o);
}
function Ds(e) {
  const t = [];
  function n(s) {
    return (r) => {
      r.merged || (r.barycenter === void 0 || s.barycenter === void 0 || r.barycenter >= s.barycenter) && $s(s, r);
    };
  }
  function o(s) {
    return (r) => {
      r.in.push(s), --r.indegree === 0 && e.push(r);
    };
  }
  for (; e.length; ) {
    const s = e.pop();
    t.push(s), s.in.reverse().forEach(n(s)), s.out.forEach(o(s));
  }
  return t.filter((s) => !s.merged).map((s) => Ae(s, ["vs", "i", "barycenter", "weight"]));
}
function $s(e, t) {
  let n = 0, o = 0;
  e.weight && (n += e.barycenter * e.weight, o += e.weight), t.weight && (n += t.barycenter * t.weight, o += t.weight), e.vs = t.vs.concat(e.vs), e.barycenter = n / o, e.weight = o, e.i = Math.min(t.i, e.i), t.merged = !0;
}
function As(e, t) {
  const n = Ln(e, (h) => Object.hasOwn(h, "barycenter")), o = n.lhs, s = n.rhs.sort((h, m) => m.i - h.i), r = [];
  let i = 0, a = 0, u = 0;
  o.sort(Fs(!!t)), u = yn(r, s, u), o.forEach((h) => {
    u += h.vs.length, r.push(h.vs), i += h.barycenter * h.weight, a += h.weight, u = yn(r, s, u);
  });
  const d = { vs: r.flat(!0) };
  return a && (d.barycenter = i / a, d.weight = a), d;
}
function yn(e, t, n) {
  let o;
  for (; t.length && (o = t[t.length - 1]).i <= n; )
    t.pop(), e.push(o.vs), n++;
  return n;
}
function Fs(e) {
  return (t, n) => t.barycenter < n.barycenter ? -1 : t.barycenter > n.barycenter ? 1 : e ? n.i - t.i : t.i - n.i;
}
function $n(e, t, n, o) {
  let s = e.children(t);
  const r = e.node(t), i = r ? r.borderLeft : void 0, a = r ? r.borderRight : void 0, u = {};
  i && (s = s.filter((v) => v !== i && v !== a));
  const d = Rs(e, s);
  d.forEach((v) => {
    if (e.children(v.v).length) {
      const b = $n(e, v.v, n, o);
      u[v.v] = b, Object.hasOwn(b, "barycenter") && zs(v, b);
    }
  });
  let h = Ps(d, n);
  Bs(h, u);
  const m = As(h, o);
  if (i && (m.vs = [i, m.vs, a].flat(!0), e.predecessors(i).length)) {
    const v = e.node(e.predecessors(i)[0]), b = e.node(e.predecessors(a)[0]);
    Object.hasOwn(m, "barycenter") || (m.barycenter = 0, m.weight = 0), m.barycenter = (m.barycenter * m.weight + v.order + b.order) / (m.weight + 2), m.weight += 2;
  }
  return m;
}
function Bs(e, t) {
  e.forEach((n) => {
    n.vs = n.vs.flatMap((o) => t[o] ? t[o].vs : o);
  });
}
function zs(e, t) {
  e.barycenter !== void 0 ? (e.barycenter = (e.barycenter * e.weight + t.barycenter * t.weight) / (e.weight + t.weight), e.weight += t.weight) : (e.barycenter = t.barycenter, e.weight = t.weight);
}
function Vs(e, t, n, o) {
  o || (o = e.nodes());
  const s = Is(e), r = new ne({ compound: !0 }).setGraph({ root: s }).setDefaultNodeLabel((i) => e.node(i));
  return o.forEach((i) => {
    const a = e.node(i), u = e.parent(i);
    (a.rank === t || a.minRank <= t && t <= a.maxRank) && (r.setNode(i), r.setParent(i, u || s), e[n](i).forEach((h) => {
      const m = h.v === i ? h.w : h.v, v = r.edge(m, i), b = v ? v.weight : 0;
      r.setEdge(m, i, {
        weight: e.edge(h).weight + b
      });
    }), Object.hasOwn(a, "minRank") && r.setNode(i, {
      borderLeft: a.borderLeft[t],
      borderRight: a.borderRight[t]
    }));
  }), r;
}
function Is(e) {
  let t;
  for (; e.hasNode(t = je("_root")); ) ;
  return t;
}
function Hs(e, t, n) {
  let o = {}, s;
  n.forEach((r) => {
    let i = e.parent(r), a, u;
    for (; i; ) {
      if (a = e.parent(i), a ? (u = o[a], o[a] = i) : (u = s, s = i), u && u !== i) {
        t.setEdge(u, i);
        return;
      }
      i = a;
    }
  });
}
function An(e, t = {}) {
  if (typeof t.customOrder == "function") {
    t.customOrder(e, An);
    return;
  }
  const n = kt(e), o = wn(
    e,
    be(1, n + 1),
    "inEdges"
  ), s = wn(
    e,
    be(n - 1, -1, -1),
    "outEdges"
  );
  let r = Ms(e);
  if (bn(e, r), t.disableOptimalOrderHeuristic)
    return;
  let i = Number.POSITIVE_INFINITY, a;
  const u = t.constraints || [];
  for (let d = 0, h = 0; h < 4; ++d, ++h) {
    const m = d % 2 === 0 ? o : s, v = d % 4 >= 2;
    Gs(m, v, u), r = Fe(e);
    const b = Os(e, r);
    b < i ? (i = b, h = 0, a = structuredClone(r)) : b === i && (a = structuredClone(r));
  }
  bn(e, a);
}
function wn(e, t, n) {
  const o = /* @__PURE__ */ new Map();
  function s(r, i) {
    o.has(r) || o.set(r, []), o.get(r).push(i);
  }
  for (const r of e.nodes()) {
    const i = e.node(r);
    if (typeof i.rank == "number" && s(i.rank, r), typeof i.minRank == "number" && typeof i.maxRank == "number")
      for (let a = i.minRank; a <= i.maxRank; a++)
        a !== i.rank && s(a, r);
  }
  return t.map(
    (r) => Vs(
      e,
      r,
      n,
      o.get(r) || []
    )
  );
}
function Gs(e, t, n) {
  const o = new ne();
  e.forEach((s) => {
    n.forEach((a) => o.setEdge(a.left, a.right));
    const r = s.graph().root, i = $n(s, r, o, t);
    i.vs.forEach((a, u) => {
      s.node(a).order = u;
    }), Hs(s, o, i.vs);
  });
}
function bn(e, t) {
  Object.values(t).forEach((n) => {
    n.forEach((o, s) => {
      e.node(o).order = s;
    });
  });
}
function Ws(e, t) {
  const n = {};
  function o(s, r) {
    let i = 0, a = 0;
    const u = s.length, d = r[r.length - 1];
    return r.forEach((h, m) => {
      const v = js(e, h), b = v ? e.node(v).order : u;
      (v || h === d) && (r.slice(a, m + 1).forEach((L) => {
        e.predecessors(L).forEach((N) => {
          const z = e.node(N), Q = z.order;
          (Q < i || b < Q) && !(z.dummy && e.node(L).dummy) && Fn(n, N, L);
        });
      }), a = m + 1, i = b);
    }), r;
  }
  return t.length && t.reduce(o), n;
}
function Ys(e, t) {
  const n = {};
  function o(r, i, a, u, d) {
    let h;
    be(i, a).forEach((m) => {
      h = r[m], e.node(h).dummy && e.predecessors(h).forEach((v) => {
        const b = e.node(v);
        b.dummy && (b.order < u || b.order > d) && Fn(n, v, h);
      });
    });
  }
  function s(r, i) {
    let a = -1, u, d = 0;
    return i.forEach((h, m) => {
      if (e.node(h).dummy === "border") {
        const v = e.predecessors(h);
        v.length && (u = e.node(v[0]).order, o(i, d, m, a, u), d = m, a = u);
      }
      o(i, d, i.length, u, r.length);
    }), i;
  }
  return t.length && t.reduce(s), n;
}
function js(e, t) {
  if (e.node(t).dummy)
    return e.predecessors(t).find((n) => e.node(n).dummy);
}
function Fn(e, t, n) {
  if (t > n) {
    const s = t;
    t = n, n = s;
  }
  let o = e[t];
  o || (e[t] = o = {}), o[n] = !0;
}
function Us(e, t, n) {
  if (t > n) {
    const o = t;
    t = n, n = o;
  }
  return !!e[t] && Object.hasOwn(e[t], n);
}
function Zs(e, t, n, o) {
  const s = {}, r = {}, i = {};
  return t.forEach((a) => {
    a.forEach((u, d) => {
      s[u] = u, r[u] = u, i[u] = d;
    });
  }), t.forEach((a) => {
    let u = -1;
    a.forEach((d) => {
      let h = o(d);
      if (h.length) {
        h = h.sort((v, b) => i[v] - i[b]);
        const m = (h.length - 1) / 2;
        for (let v = Math.floor(m), b = Math.ceil(m); v <= b; ++v) {
          const L = h[v];
          r[d] === d && u < i[L] && !Us(n, d, L) && (r[L] = d, r[d] = s[d] = s[L], u = i[L]);
        }
      }
    });
  }), { root: s, align: r };
}
function Xs(e, t, n, o, s) {
  const r = {}, i = qs(e, t, n, s), a = s ? "borderLeft" : "borderRight";
  function u(m, v) {
    const b = i.nodes().slice(), L = {};
    let N = b.pop();
    for (; N; ) {
      if (L[N])
        m(N);
      else {
        L[N] = !0, b.push(N);
        for (const z of v(N))
          b.push(z);
      }
      N = b.pop();
    }
  }
  function d(m) {
    r[m] = i.inEdges(m).reduce((v, b) => Math.max(v, r[b.v] + i.edge(b)), 0);
  }
  function h(m) {
    const v = i.outEdges(m).reduce((L, N) => Math.min(L, r[N.w] - i.edge(N)), Number.POSITIVE_INFINITY), b = e.node(m);
    v !== Number.POSITIVE_INFINITY && b.borderType !== a && (r[m] = Math.max(r[m], v));
  }
  return u(d, i.predecessors.bind(i)), u(h, i.successors.bind(i)), Object.keys(o).forEach((m) => {
    r[m] = r[n[m]];
  }), r;
}
function qs(e, t, n, o) {
  const s = new e.constructor(), r = e.graph(), i = ti(r.nodesep, r.edgesep, o);
  return t.forEach((a) => {
    let u;
    a.forEach((d) => {
      const h = n[d];
      if (s.setNode(h), u) {
        const m = n[u], v = s.edge(m, h);
        s.setEdge(
          m,
          h,
          Math.max(i(e, d, u), v || 0)
        );
      }
      u = d;
    });
  }), s;
}
function Ks(e, t) {
  return Object.values(t).reduce(
    (n, o) => {
      let s = Number.NEGATIVE_INFINITY, r = Number.POSITIVE_INFINITY;
      Object.entries(o).forEach(([a, u]) => {
        const d = ni(e, a) / 2;
        s = Math.max(u + d, s), r = Math.min(u - d, r);
      });
      const i = s - r;
      return i < n[0] && (n = [i, o]), n;
    },
    [Number.POSITIVE_INFINITY, null]
  )[1];
}
function Js(e, t) {
  const n = Object.values(t), o = Z(Math.min, n), s = Z(Math.max, n);
  ["u", "d"].forEach((r) => {
    ["l", "r"].forEach((i) => {
      const a = r + i;
      let u = e[a];
      if (u === t) return;
      const d = Object.values(u);
      let h = o - Z(Math.min, d);
      i !== "l" && (h = s - Z(Math.max, d)), h && (u = Be(u, (m) => m + h), e[a] = u);
    });
  });
}
function Qs(e, t) {
  return Be(e.ul, (n, o) => {
    if (t)
      return e[t.toLowerCase()][o];
    const s = Object.values(e).map((r) => r[o]).sort((r, i) => r - i);
    return (s[1] + s[2]) / 2;
  });
}
function ei(e) {
  const t = Fe(e), n = Object.assign(
    Ws(e, t),
    Ys(e, t)
  ), o = {};
  let s;
  ["u", "d"].forEach((i) => {
    s = i === "u" ? t : Object.values(t).reverse(), ["l", "r"].forEach((a) => {
      let u = s;
      a === "r" && (u = u.map(
        (v) => Object.values(v).reverse()
      ));
      const d = (i === "u" ? e.predecessors : e.successors).bind(e), h = Zs(
        e,
        u,
        n,
        d
      );
      let m = Xs(
        e,
        u,
        h.root,
        h.align,
        a === "r"
      );
      a === "r" && (m = Be(m, (v) => -v)), o[i + a] = m;
    });
  });
  const r = Ks(e, o);
  return Js(o, r), Qs(o, e.graph().align);
}
function ti(e, t, n) {
  return (o, s, r) => {
    const i = o.node(s), a = o.node(r);
    let u = 0, d;
    if (u += i.width / 2, Object.hasOwn(i, "labelpos"))
      switch (i.labelpos.toLowerCase()) {
        case "l":
          d = -i.width / 2;
          break;
        case "r":
          d = i.width / 2;
          break;
      }
    if (d && (u += n ? d : -d), d = 0, u += (i.dummy ? t : e) / 2, u += (a.dummy ? t : e) / 2, u += a.width / 2, Object.hasOwn(a, "labelpos"))
      switch (a.labelpos.toLowerCase()) {
        case "l":
          d = a.width / 2;
          break;
        case "r":
          d = -a.width / 2;
          break;
      }
    return d && (u += n ? d : -d), d = 0, u;
  };
}
function ni(e, t) {
  return e.node(t).width;
}
function oi(e) {
  const t = bt(e);
  si(t);
  const n = ei(t);
  Object.entries(n).forEach(([o, s]) => {
    e.node(o).x = s;
  });
}
function si(e) {
  const t = Fe(e), n = e.graph().ranksep;
  let o = 0;
  t.forEach((s) => {
    const r = s.reduce((i, a) => {
      const u = e.node(a).height;
      return i > u ? i : u;
    }, 0);
    s.forEach((i) => {
      e.node(i).y = o + r / 2;
    }), o += r + n;
  });
}
function ii(e, t = {}) {
  const n = t.debugTiming ? Mn : On;
  return n("layout", () => {
    const o = n("  buildLayoutGraph", () => vi(e));
    return n("  runLayout", () => ri(o, n, t)), n("  updateInputGraph", () => ai(e, o)), o;
  });
}
function ri(e, t, n) {
  t("    makeSpaceForEdgeLabels", () => gi(e)), t("    removeSelfEdges", () => xi(e)), t("    acyclic", () => Jo(e)), t("    nestingGraph.run", () => bs(e)), t("    rank", () => ms(bt(e))), t("    injectEdgeLabelProxies", () => pi(e)), t("    removeEmptyRanks", () => Nn(e)), t("    nestingGraph.cleanup", () => _s(e)), t("    normalizeRanks", () => xn(e)), t("    assignRankMinMax", () => yi(e)), t("    removeEdgeLabelProxies", () => wi(e)), t("    normalize.run", () => ts(e)), t("    parentDummyChains", () => ps(e)), t("    addBorderSegments", () => Cs(e)), t("    order", () => An(e, n)), t("    insertSelfEdges", () => Ni(e)), t("    adjustCoordinateSystem", () => pn.adjust(e)), t("    position", () => oi(e)), t("    positionSelfEdges", () => Si(e)), t("    removeBorderNodes", () => Ci(e)), t("    normalize.undo", () => os(e)), t("    fixupEdgeLabelCoords", () => Ei(e)), t("    undoCoordinateSystem", () => pn.undo(e)), t("    translateGraph", () => bi(e)), t("    assignNodeIntersects", () => ki(e)), t("    reversePoints", () => _i(e)), t("    acyclic.undo", () => Qo(e));
}
function ai(e, t) {
  e.nodes().forEach((s) => {
    const r = e.node(s), i = t.node(s);
    r && (r.x = i.x, r.y = i.y, r.order = i.order, r.rank = i.rank, t.children(s).length && (r.width = i.width, r.height = i.height));
  }), e.edges().forEach((s) => {
    const r = e.edge(s), i = t.edge(s);
    r.points = i.points, Object.hasOwn(i, "x") && (r.x = i.x, r.y = i.y);
  });
  const n = t.graph(), o = e.graph();
  o.width = n.width, o.height = n.height;
}
const li = ["nodesep", "edgesep", "ranksep", "marginx", "marginy"], ui = { ranksep: 50, edgesep: 20, nodesep: 50, rankdir: "tb" }, ci = ["acyclicer", "ranker", "rankdir", "align"], di = ["width", "height", "rank"], kn = { width: 0, height: 0 }, fi = ["minlen", "weight", "width", "height", "labeloffset"], hi = {
  minlen: 1,
  weight: 1,
  width: 0,
  height: 0,
  labeloffset: 10,
  labelpos: "r"
}, mi = ["labelpos", "arrowshape"];
function vi(e) {
  const t = gt(e.graph()), n = {
    ...ui,
    ...vt(t, li),
    ...Ae(t, ci)
  }, o = new ne({ multigraph: !0, compound: !0 });
  return o.setGraph(n), e.nodes().forEach((s) => {
    const r = gt(e.node(s)), i = vt(r, di);
    Object.keys(kn).forEach((a) => {
      i[a] === void 0 && (i[a] = kn[a]);
    }), o.setNode(s, i), o.setParent(s, e.parent(s));
  }), e.edges().forEach((s) => {
    const r = gt(e.edge(s));
    o.setEdge(
      s,
      {
        ...hi,
        ...vt(r, fi),
        ...Ae(r, mi)
      }
    );
  }), o;
}
function gi(e) {
  const t = e.graph();
  t.ranksep /= 2, e.edges().forEach((n) => {
    const o = e.edge(n);
    o.minlen *= 2, o.labelpos.toLowerCase() !== "c" && (t.rankdir === "TB" || t.rankdir === "BT" ? o.width += o.labeloffset : o.height += o.labeloffset);
  });
}
function pi(e) {
  e.edges().forEach((t) => {
    const n = e.edge(t);
    if (n.width && n.height) {
      const o = e.node(t.v), r = { rank: (e.node(t.w).rank - o.rank) / 2 + o.rank, e: t };
      ke(e, "edge-proxy", r, "_ep");
    }
  });
}
function yi(e) {
  let t = 0;
  e.nodes().forEach((n) => {
    const o = e.node(n);
    o.borderTop && (o.minRank = e.node(o.borderTop).rank, o.maxRank = e.node(o.borderBottom).rank, t = Math.max(t, o.maxRank));
  }), e.graph().maxRank = t;
}
function wi(e) {
  e.nodes().forEach((t) => {
    const n = e.node(t);
    n.dummy === "edge-proxy" && (e.edge(n.e).labelRank = n.rank, e.removeNode(t));
  });
}
function bi(e) {
  let t = Number.POSITIVE_INFINITY, n = 0, o = Number.POSITIVE_INFINITY, s = 0;
  const r = e.graph(), i = r.marginx || 0, a = r.marginy || 0;
  function u(d) {
    const h = d.x, m = d.y, v = d.width, b = d.height;
    t = Math.min(t, h - v / 2), n = Math.max(n, h + v / 2), o = Math.min(o, m - b / 2), s = Math.max(s, m + b / 2);
  }
  e.nodes().forEach((d) => u(e.node(d))), e.edges().forEach((d) => {
    const h = e.edge(d);
    Object.hasOwn(h, "x") && u(h);
  }), t -= i, o -= a, e.nodes().forEach((d) => {
    const h = e.node(d);
    h.x -= t, h.y -= o;
  }), e.edges().forEach((d) => {
    const h = e.edge(d);
    h.points.forEach((m) => {
      m.x -= t, m.y -= o;
    }), Object.hasOwn(h, "x") && (h.x -= t), Object.hasOwn(h, "y") && (h.y -= o);
  }), r.width = n - t + i, r.height = s - o + a;
}
function En(e, t, n) {
  if (!t)
    return e;
  const o = t.x - e.x, s = t.y - e.y, r = Math.sqrt(o * o + s * s);
  if (!r || r <= n)
    return e;
  const i = n / r;
  return {
    x: e.x + o * i,
    y: e.y + s * i
  };
}
function ki(e) {
  e.edges().forEach((n) => {
    const o = e.edge(n), s = e.node(n.v), r = e.node(n.w);
    let i, a;
    !o.points || !o.points.length ? (o.points = [], i = r, a = s) : (i = o.points[0], a = o.points[o.points.length - 1]);
    const u = wt(s, i), d = wt(r, a), h = o.points.length ? o.points[0] : i, m = o.points.length ? o.points[o.points.length - 1] : a, v = o.arrowshape, b = v === "normal" || v === "vee", L = !!o.reversed;
    let N = u, z = d;
    b && (L ? N = En(u, h, 4) : z = En(d, m, 4)), o.points.unshift(N), o.points.push(z);
  });
}
function Ei(e) {
  e.edges().forEach((t) => {
    const n = e.edge(t);
    if (Object.hasOwn(n, "x"))
      switch ((n.labelpos === "l" || n.labelpos === "r") && (n.width -= n.labeloffset), n.labelpos) {
        case "l":
          n.x -= n.width / 2 + n.labeloffset;
          break;
        case "r":
          n.x += n.width / 2 + n.labeloffset;
          break;
      }
  });
}
function _i(e) {
  e.edges().forEach((t) => {
    const n = e.edge(t);
    n.reversed && n.points.reverse();
  });
}
function Ci(e) {
  e.nodes().forEach((t) => {
    if (e.children(t).length) {
      const n = e.node(t), o = e.node(n.borderTop), s = e.node(n.borderBottom), r = e.node(n.borderLeft[n.borderLeft.length - 1]), i = e.node(n.borderRight[n.borderRight.length - 1]);
      n.width = Math.abs(i.x - r.x), n.height = Math.abs(s.y - o.y), n.x = r.x + n.width / 2, n.y = o.y + n.height / 2;
    }
  }), e.nodes().forEach((t) => {
    e.node(t).dummy === "border" && e.removeNode(t);
  });
}
function xi(e) {
  e.edges().forEach((t) => {
    if (t.v === t.w) {
      const n = e.node(t.v);
      n.selfEdges || (n.selfEdges = []), n.selfEdges.push({ e: t, label: e.edge(t) }), e.removeEdge(t);
    }
  });
}
function Ni(e) {
  Fe(e).forEach((n) => {
    let o = 0;
    n.forEach((s, r) => {
      const i = e.node(s);
      i.order = r + o, (i.selfEdges || []).forEach((a) => {
        ke(e, "selfedge", {
          width: a.label.width,
          height: a.label.height,
          rank: i.rank,
          order: r + ++o,
          e: a.e,
          label: a.label
        }, "_se");
      }), delete i.selfEdges;
    });
  });
}
function Si(e) {
  e.nodes().forEach((t) => {
    const n = e.node(t);
    if (n.dummy === "selfedge") {
      const o = e.node(n.e.v), s = o.x + o.width / 2, r = o.y, i = n.x - s, a = o.height / 2;
      e.setEdge(n.e, n.label), e.removeNode(t), n.label.points = [
        { x: s + 2 * i / 3, y: r - a },
        { x: s + 5 * i / 6, y: r - a },
        { x: s + i, y: r },
        { x: s + 5 * i / 6, y: r + a },
        { x: s + 2 * i / 3, y: r + a }
      ], n.label.x = n.x, n.label.y = n.y;
    }
  });
}
function vt(e, t) {
  return Be(Ae(e, t), Number);
}
function gt(e) {
  const t = {};
  return e && Object.entries(e).forEach(([n, o]) => {
    let s = n;
    typeof s == "string" && (s = s.toLowerCase()), t[s] = o;
  }), t;
}
const Li = {
  rankDirection: "TB",
  // "TB", "BT", "LR", "RL"
  nodeSeparation: 50,
  rankSeparation: 50,
  edgeSeparation: 10,
  // "UL", "UR", "DL", "DR" or undefined force one of four extreme alignment strategies (Up/Down × Left/Right).
  align: void 0,
  nodeWidth: 100,
  nodeHeight: 40,
  curvedEdges: !1,
  padding: 20,
  // "undirected" | "normal" | "vee"
  arrowShape: "normal",
  arrowSize: 10
};
function Mi(e) {
  return e.length ? e.map(
    (t, n) => `${n === 0 ? "M" : "L"} ${t.x} ${t.y}`
  ).join(" ") : "";
}
function Oi(e) {
  if (!e.length) return "";
  if (e.length === 1) {
    const s = e[0];
    return `M ${s.x} ${s.y}`;
  }
  if (e.length === 2)
    return `M ${e[0].x} ${e[0].y} L ${e[1].x} ${e[1].y}`;
  const [t] = e;
  let n = `M ${t.x} ${t.y}`;
  for (let s = 1; s < e.length - 1; s += 1) {
    e[s - 1];
    const r = e[s], i = e[s + 1], a = r.x, u = r.y, d = (r.x + i.x) / 2, h = (r.y + i.y) / 2;
    n += ` Q ${a} ${u} ${d} ${h}`;
  }
  const o = e[e.length - 1];
  return n += ` L ${o.x} ${o.y}`, n;
}
function Ti(e) {
  const { nodes: t, edges: n, configuration: o } = e, s = E(null), r = E(null), i = `dag-arrow-${pt()}`;
  function a(u, d, h) {
    r.value = null, s.value = null;
    const m = {
      ...Li,
      ...h
    }, v = new ne({ multigraph: !0, compound: !0 });
    v.setGraph({
      rankdir: m.rankDirection,
      nodesep: m.nodeSeparation,
      ranksep: m.rankSeparation,
      edgesep: m.edgeSeparation,
      align: m.align
    }), u.forEach((w) => {
      v.setNode(w.id, {
        label: w.label,
        width: w.width ?? m.nodeWidth,
        height: w.height ?? m.nodeHeight
      });
    }), d.forEach((w) => {
      v.setEdge(w.from, w.to, {
        weight: w.weight ?? 1,
        minlen: w.minLength ?? 1,
        arrowShape: m.arrowShape ?? "normal"
      });
    }), ii(v);
    const b = u.map((w) => {
      const $ = v.node(w.id);
      return {
        id: w.id,
        label: w.label,
        x: $.x,
        y: $.y,
        width: $.width,
        height: $.height,
        original: w
      };
    }), L = v.edges().map((w) => {
      const $ = v.edge(w), c = $.points || [];
      if (!c.length)
        return null;
      const Ce = m.curvedEdges ? Oi(c) : Mi(c), xe = m.arrowShape !== "undirected" ? `url(#${i})` : null, Ve = d.find((A) => A?.from === w.v && A?.to === w.w);
      return {
        id: `${w.v}->${w.w}->${pt()}`,
        from: w.v,
        to: w.w,
        points: c,
        pathData: Ce,
        markerEnd: xe,
        original: {
          ...Ve,
          ...$
        }
      };
    }).filter(Boolean), N = m.padding;
    if (!b.length) {
      s.value = {
        nodes: [],
        edges: L,
        viewBox: "0 0 0 0",
        arrowShape: m.arrowShape,
        arrowSize: m.arrowSize
      };
      return;
    }
    const z = b.flatMap((w) => [
      w.x - w.width / 2,
      w.x + w.width / 2
    ]), Q = b.flatMap((w) => [
      w.y - w.height / 2,
      w.y + w.height / 2
    ]), _e = Math.min(...z) - N, de = Math.max(...z) + N, oe = Math.min(...Q) - N, fe = Math.max(...Q) + N;
    s.value = {
      nodes: b,
      edges: L.map((w) => ({
        ...w,
        midpoint: w.pathData ? wo(w.pathData) : { x: 0, y: 0 }
      })),
      viewBox: `${_e} ${oe} ${de - _e} ${fe - oe}`,
      arrowShape: m.arrowShape,
      arrowSize: m.arrowSize
    };
  }
  return vo(() => {
    try {
      const u = p(t) || [], d = p(n) || [], h = p(o) || {};
      a(u, d, h);
    } catch (u) {
      console.error("[useDag] layout error:", u), r.value = u, s.value = null;
    }
  }), {
    layoutData: s,
    lastError: r,
    arrowMarkerIdentifier: i,
    recomputeLayout: () => {
      const u = p(t) || [], d = p(n) || [], h = p(o) || {};
      a(u, d, h);
    }
  };
}
const Ri = ["id"], Pi = {
  key: 0,
  class: "dag-chart-error"
}, Di = ["viewBox", "xmlns"], $i = { key: 0 }, Ai = ["id", "width", "height"], Fi = ["cx", "cy", "r", "fill"], Bi = ["x", "y", "width", "height", "fill"], zi = { key: 2 }, Vi = ["id", "markerWidth", "markerHeight", "refX", "refY"], Ii = ["d", "fill", "stroke"], Hi = ["d", "fill", "stroke"], Gi = { class: "vue-ui-dag-edges" }, Wi = ["onMouseenter"], Yi = { class: "vue-ui-dag-nodes" }, ji = ["onClick", "onMouseenter", "onMouseleave"], Ui = ["x", "y", "width", "height"], Zi = { class: "vue-ui-dag-edges" }, Xi = ["d", "stroke-width", "marker-end"], qi = { class: "vue-ui-dag-node-labels" }, Ki = ["onClick", "onMouseenter", "onMouseleave"], Ji = ["x", "y", "font-size", "fill", "font-weight"], Qi = ["x", "y", "font-size", "fill", "font-weight", "innerHTML"], er = { key: 1 }, tr = {
  key: 6,
  class: "vue-data-ui-watermark"
}, nr = ["data-position"], or = { key: 0 }, sr = ["data-position"], ir = { key: 0 }, _n = 1.5, rr = {
  __name: "vue-ui-dag",
  props: {
    dataset: {
      type: Object,
      default() {
        return {
          nodes: [],
          edges: []
        };
      }
    },
    config: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["onNodeClick", "onMidpointEnter", "onMidpointLeave", "copyAlt", "rotate"],
  setup(e, { expose: t, emit: n }) {
    const o = lt(() => import("./PenAndPaper-DWV1FMIi.js")), s = lt(() => import("./UserOptions-CrLqEimB.js")), r = lt(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_dag: i } = No(), { isThemeValid: a, warnInvalidTheme: u } = To(), d = e, h = n, m = E(null), v = E(pt()), b = E(null), L = E(null), N = E(null), z = E(0), Q = E(0), _e = E(!1), de = E(!1), oe = E(!1), fe = E(!1), w = E(null), $ = E(null), c = E(Ke()), Ce = W(() => c.value.userOptions.useCursorPointer), ze = E(c.value.style.chart.width), xe = E(c.value.style.chart.height), Ve = E({ x: 0, y: 0 }), A = E(null), Et = E(null), _t = E({ left: "0px", top: "0px" }), Ct = E("top"), se = E(!1), xt = E({ x: 0, y: 0 }), Nt = E({ x: 0, y: 0 }), Ne = E(null), Ze = E(null), Xe = E({ left: "0px", top: "0px" }), St = E("top"), Ie = E(!1), Se = E(!1), { svgRef: ie } = Po({ config: c.value.style.chart.title }), { userOptionsVisible: qe, setUserOptionsVisibility: Lt, keepUserOptionState: Mt } = Ro({ config: c.value }), X = E(c.value.style.chart.layout.rankDirection), Bn = W(() => rn({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            nodes: {
              stroke: "#CCCCCC",
              backgroundColor: "#DDDDDD50"
            },
            edges: {
              stroke: "#CCCCCC"
            },
            midpoints: {
              stroke: "#CCCCCC",
              fill: "#CCCCCC"
            }
          }
        }
      },
      userConfig: c.value.skeletonConfig ?? {}
    })), { loading: Le, FINAL_DATASET: He, manualLoading: Ot } = So({
      ...go(d),
      FINAL_CONFIG: c,
      prepareConfig: Ke,
      skeletonDataset: d.config?.skeletonDataset ?? {
        nodes: [
          { id: "A", label: "" },
          { id: "B", label: "" },
          { id: "C", label: "" }
        ],
        edges: [
          { from: "A", to: "B" },
          { from: "A", to: "C" }
        ]
      },
      skeletonConfig: rn({
        defaultConfig: c.value,
        userConfig: Bn.value
      })
    });
    function Ke() {
      const l = ct({
        userConfig: d.config,
        defaultConfig: i
      }), y = l.theme;
      if (!y) return l;
      if (!a.value(l))
        return u(l), l;
      const f = ct({
        userConfig: $o[y] || d.config,
        defaultConfig: l
      });
      return ct({
        userConfig: d.config,
        defaultConfig: f
      });
    }
    const Je = W(() => !!c.value.debug);
    Qt(async () => {
      bo(d.dataset) && (an({
        componentName: "VueUiDag",
        type: "dataset",
        debug: Je.value
      }), de.value = !1, Ot.value = !0), d.dataset.nodes || (an({
        componentName: "VueUiDag",
        type: "datasetAttributeEmpty",
        property: "nodes",
        index: 0,
        debug: Je.value
      }), de.value = !1, Ot.value = !0), de.value = !0, await pe(), jt();
    }), Re(() => d.config, async (l) => {
      Le.value || (c.value = Ke()), qe.value = !c.value.userOptions.showOnChartHover, z.value += 1, X.value = c.value.style.chart.layout.rankDirection, ze.value = c.value.style.chart.width, xe.value = c.value.style.chart.height, he.value = c.value.style.chart.zoom.active, await pe(), jt();
    }, { deep: !0 });
    const { isPrinting: Qe, isImaging: et, generatePdf: Tt, generateImage: Rt } = Mo({
      elementId: `dag_${v.value}`,
      fileName: c.value.style.chart.title.text || "vue-ui-dag",
      options: c.value.userOptions.print
    }), zn = W(() => c.value.style.chart.backgroundColor), Vn = W(() => c.value.style.chart.title), { exportSvg: In, getSvg: Hn } = Oo({
      svg: ie,
      title: Vn,
      legend: void 0,
      legendItems: void 0,
      backgroundColor: zn
    });
    async function Pt({ isCb: l }) {
      fe.value = !0, await pe();
      try {
        if (l) {
          const { blob: y, url: f, text: g, dataUrl: S } = await Hn();
          await Promise.resolve(c.value.userOptions.callbacks.svg({ blob: y, url: f, text: g, dataUrl: S }));
        } else
          await Promise.resolve(In());
      } finally {
        fe.value = !1;
      }
    }
    function Gn() {
      _e.value = !0, Lt(!0);
    }
    function Wn() {
      Lt(!1), _e.value = !1;
    }
    function Dt({
      tooltipRef: l,
      isVisibleRef: y,
      anchorRef: f,
      styleRef: g,
      placementRef: S,
      offsetRef: R,
      margin: k = 24
    }) {
      return function() {
        const U = l.value;
        if (!U || !y.value) return;
        const O = U.getBoundingClientRect(), B = window.innerWidth, I = window.innerHeight, q = f.value.x, Y = f.value.y, K = R?.value?.x ?? 0, le = R?.value?.y ?? 0;
        let P = Y - le - O.height - k, j = q - O.width / 2, Oe = "top";
        if (P < k) {
          const ge = Y + le + k;
          ge + O.height <= I - k ? (P = ge, Oe = "bottom") : (P = Y - O.height / 2, P < k && (P = k), P + O.height > I - k && (P = I - O.height - k), Oe = "center");
        }
        j < k && (j = k), j + O.width > B - k && (j = B - O.width - k);
        const fo = j <= k, ho = j + O.width >= B - k;
        if ((fo || ho) && Oe !== "center") {
          let ge, Te;
          const mo = q - K - k;
          B - (q + K) - k >= mo ? (ge = "right", Te = q + K + k) : (ge = "left", Te = q - K - k - O.width), Te >= k && Te + O.width <= B - k && (j = Te, P = Y - O.height / 2, P < k && (P = k), P + O.height > I - k && (P = I - O.height - k), Oe = ge);
        }
        S.value = Oe, g.value = {
          left: `${j}px`,
          top: `${P}px`
        };
      };
    }
    const Yn = W(() => He.value.nodes.map((l) => ({
      ...l,
      backgroundColor: l.backgroundColor ? ln(l.backgroundColor) : c.value.style.chart.nodes.backgroundColor,
      color: l.color ? ln(l.color) : c.value.style.chart.nodes.labels.color
    }))), jn = W(() => He.value.edges), Un = W(() => ({
      ...c.value.style.chart.layout,
      rankDirection: X.value
    })), { layoutData: _, lastError: $t, arrowMarkerIdentifier: Zn } = Ti({
      nodes: Yn,
      edges: jn,
      configuration: Un
    });
    function At(l) {
      return _.value ? _.value.nodes.find((y) => y.id === l) : null;
    }
    W(() => {
      if (!_.value) return [];
      const l = c.value.style.chart.edges.stroke, y = /* @__PURE__ */ new Set();
      return _.value.edges.forEach((f) => {
        y.add({
          id: f.id,
          from: f.from,
          to: f.to,
          color: f.original?.color || l
        });
      }), Array.from(y);
    });
    const Ft = W(() => {
      const l = ze.value, y = xe.value, f = Number(l), g = Number(y), S = Number.isFinite(f) && f > 0, R = Number.isFinite(g) && g > 0;
      return !S && !R ? null : {
        width: S ? f : null,
        height: R ? g : null
      };
    }), he = E(c.value.style.chart.zoom.active), {
      viewBox: me,
      resetZoom: Ge,
      setInitialViewBox: Xn,
      scale: Bt,
      zoomByFactor: zt
    } = xo(
      ie,
      { x: 0, y: 0, width: 100, height: 100 },
      1,
      he,
      () => {
        se.value = !1;
      }
    );
    function qn() {
      he.value = !he.value;
    }
    function Vt() {
      const l = _.value && _.value.viewBox;
      if (!l) return;
      const y = String(l).split(" ").map(Number);
      if (y.length !== 4) return;
      const [f, g, S, R] = y;
      if (!Number.isFinite(f) || !Number.isFinite(g) || !Number.isFinite(S) || !Number.isFinite(R))
        return;
      let k = S, D = R, U = f, O = g;
      const B = Ft.value;
      B && (B.width !== null && (k = B.width), B.height !== null && (D = B.height), U = f - (k - S) / 2, O = g - (D - R) / 2), Xn(
        { x: U, y: O, width: k, height: D },
        { overwriteCurrentIfNotZoomed: !0 }
      );
    }
    Re(
      () => _.value && _.value.viewBox,
      () => {
        Vt();
      },
      { immediate: !0 }
    ), Re(
      () => Ft.value,
      () => {
        Vt();
      }
    ), Re(() => Se.value, (l) => {
      he.value = !l;
    });
    const Kn = W(() => {
      const l = me.value;
      return l ? `${l.x} ${l.y} ${l.width} ${l.height}` : "0 0 0 0";
    }), re = E(!1);
    function It(l) {
      re.value = l, Q.value += 1;
    }
    function tt() {
      Se.value = !Se.value;
    }
    function nt() {
      zt(_n, !0);
    }
    function ot() {
      zt(1 / _n, !0);
    }
    const st = ["TB", "RL", "BT", "LR"];
    function it() {
      X.value = st[(st.indexOf(X.value) + 1) % st.length], Ge(), h("rotate", X.value);
    }
    const Jn = Dt({
      tooltipRef: Et,
      isVisibleRef: Ie,
      anchorRef: Ve,
      styleRef: _t,
      placementRef: Ct
    }), Qn = Dt({
      tooltipRef: Ze,
      isVisibleRef: se,
      anchorRef: xt,
      styleRef: Xe,
      placementRef: St,
      offsetRef: Nt
    }), ee = E(null);
    async function eo(l) {
      h("onMidpointEnter", l);
      const y = ie.value;
      if (!y || !l?.midpoint) return;
      const f = y.createSVGPoint();
      f.x = l.midpoint.x, f.y = l.midpoint.y;
      const g = y.getScreenCTM();
      if (!g) return;
      const S = f.matrixTransform(g);
      Ve.value = {
        x: S.x,
        y: S.y
      }, A.value = l, Ie.value = !0, c.value.style.chart.midpoints.selectedEdge.animated === !0 && (ee.value = l.id, Me()), await pe(), Jn();
    }
    function to() {
      Ie.value = !1, A.value = null, h("onMidpointLeave"), c.value.style.chart.midpoints.selectedEdge.animated === !0 && (ee.value = null, Me());
    }
    async function Ht(l) {
      if (h("onNodeClick", l), !c.value.style.chart.nodes.tooltip.showOnClick) return;
      const y = ie.value;
      if (!y) return;
      const f = y.createSVGPoint();
      f.x = l.x, f.y = l.y;
      const g = y.getScreenCTM();
      if (!g) return;
      const S = f.matrixTransform(g), R = c.value.style.chart.layout.nodeWidth, k = c.value.style.chart.layout.nodeHeight, D = g.a, U = g.d, O = R * D, B = k * U;
      Nt.value = {
        x: O / 2,
        y: B / 2
      }, xt.value = {
        x: S.x,
        y: S.y
      }, Ne.value = l, se.value = !0, await pe(), Qn();
    }
    function Gt() {
      se.value = !1, Ne.value = null;
    }
    function Wt(l) {
      if (!se.value) return;
      const y = Ze.value;
      if (y && y.contains(l.target))
        return;
      const f = ie.value;
      f && f.contains(l.target) && l.target.closest(".vue-ui-dag-node") || Gt();
    }
    function Yt(l) {
      l.key === "Escape" && se.value && Gt();
    }
    Qt(() => {
      document.addEventListener("mousedown", Wt), document.addEventListener("keydown", Yt);
    }), en(() => {
      document.removeEventListener("mousedown", Wt), document.removeEventListener("keydown", Yt), w.value && ($.value && w.value.unobserve($.value), w.value.disconnect());
    });
    function jt() {
      if (!c.value.responsive) {
        w.value && ($.value && w.value.unobserve($.value), w.value.disconnect(), w.value = null, $.value = null);
        return;
      }
      const l = _o(() => {
        if (!m.value) return;
        const { width: y, height: f } = Co({
          chart: m.value,
          title: c.value.style.chart.title.text ? b.value : null,
          legend: c.value.style.chart.controls.show ? N.value?.$el : null,
          source: L.value
        });
        requestAnimationFrame(() => {
          ze.value = Math.max(0.1, y), xe.value = Math.max(0.1, f - 12);
        });
      });
      w.value && ($.value && w.value.unobserve($.value), w.value.disconnect()), w.value = new ResizeObserver(l), $.value = m.value ? m.value.parentNode : null, $.value && w.value.observe($.value), l();
    }
    function no(l, y = {}) {
      const {
        direction: f = -1,
        mode: g = "oneLapNearest",
        dasharray: S = null
      } = y;
      if (!l || typeof l.getTotalLength != "function")
        return Je.value && console.warn(
          "VueUiDag @getIdealDashoffsetDelta: invalid path element",
          l
        ), 0;
      const R = l.getTotalLength(), k = S ?? l.getAttribute("stroke-dasharray") ?? (typeof getComputedStyle == "function" ? getComputedStyle(l).strokeDasharray : ""), D = q(k);
      if (!Number.isFinite(D) || D <= 0)
        return f * R;
      const U = Math.max(1, Math.round(R / D)), O = Math.max(1, Math.ceil(R / D)), B = Math.max(1, Math.floor(R / D));
      let I;
      return g === "pattern" ? I = D : g === "oneLapCeil" ? I = O * D : g === "oneLapFloor" ? I = B * D : I = U * D, f * I;
      function q(Y) {
        if (!Y || Y === "none") return NaN;
        const K = String(Y).replace(/,/g, " ").trim().split(/\s+/).map((P) => Number.parseFloat(P)).filter((P) => Number.isFinite(P));
        if (!K.length) return NaN;
        const le = K.reduce((P, j) => P + j, 0);
        return K.length % 2 === 1 ? le * 2 : le;
      }
    }
    const rt = E(/* @__PURE__ */ new Map()), at = E(/* @__PURE__ */ new Map());
    function oo(l) {
      return function(f) {
        f ? rt.value.set(l, f) : rt.value.delete(l);
      };
    }
    function Ut() {
      at.value.forEach((l) => {
        try {
          l.cancel();
        } catch {
        }
      }), at.value.clear();
    }
    function Me() {
      Ut();
      const l = _.value?.edges ?? [];
      if (!l.length) return;
      const y = c.value.style.chart.edges.animations, f = Number(y.referenceDistance) > 0 ? Number(y.referenceDistance) : 24;
      l.forEach((g) => {
        const S = c.value.style.chart.midpoints.selectedEdge.animated === !0 && ee.value != null && g.id === ee.value, R = !!g?.original?.animated || !!g?.animated || S, k = rt.value.get(g.id);
        if (!k) return;
        if (!R) {
          k.style.strokeDasharray = "0", k.style.strokeDashoffset = "0";
          return;
        }
        const D = g?.original?.dasharray ?? y.dasharray;
        k.style.strokeDasharray = String(D), k.style.strokeDashoffset = "0";
        const U = ![void 0, null].includes(g?.original?.animationDirection), O = ![void 0, null].includes(y.animationDirection), B = U ? Number(g.original.animationDirection) : O ? Number(y.animationDirection) : -1, I = no(k, {
          direction: B,
          mode: "oneLapNearest",
          dasharray: String(D)
        }), q = g?.original?.animationDurationMs ?? y.animationDurationMs ?? 1e3, Y = Number(q), le = Number.isFinite(Y) && Y > 0 ? f / Y : f / 1e3, P = Math.max(1, Math.round(Math.abs(I) / Math.max(1e-9, le))), j = k.animate([{ strokeDashoffset: 0 }, { strokeDashoffset: I }], { duration: P, iterations: 1 / 0, easing: "linear" });
        at.value.set(g.id, j);
      });
    }
    Re(
      () => _.value && _.value.edges,
      async () => {
        await pe(), Me();
      },
      { deep: !0, immediate: !0 }
    ), en(() => {
      Ut();
    });
    async function so({ scale: l = 2 } = {}) {
      if (!m.value) return;
      const { width: y, height: f } = m.value.getBoundingClientRect(), g = y / f, { imageUri: S, base64: R } = await Ao({ domElement: m.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: S,
        base64: R,
        title: c.value.style.chart.title.text ?? "vue-ui-dag",
        width: y,
        height: f,
        aspectRatio: g
      };
    }
    const ae = W(() => {
      const l = Number(c.value.style.chart.layout.nodeHeight);
      return Number.isFinite(l) && l > 0 ? l / c.value.style.chart.backgroundPattern.spacingRatio : 12;
    }), io = W(() => ae.value * (c.value.style.chart.backgroundPattern.dotRadiusRatio / 100)), F = E(null);
    let ve = null, Zt = 0;
    function Xt(l) {
      ve && (clearTimeout(ve), ve = null), F.value !== l && (F.value = l, (c.value.style.chart.nodes.selected.downstreamEdges.animated === !0 || c.value.style.chart.nodes.selected.upstreamEdges.animated === !0) && Me());
    }
    function qt(l) {
      const y = ++Zt;
      ve && clearTimeout(ve), ve = setTimeout(() => {
        y === Zt && F.value === l && (F.value = null, (c.value.style.chart.nodes.selected.downstreamEdges.animated === !0 || c.value.style.chart.nodes.selected.upstreamEdges.animated === !0) && Me());
      }, 20);
    }
    function ro(l) {
      const y = l.from === F.value, f = l.to === F.value;
      let g = l.original.color ?? c.value.style.chart.edges.stroke;
      if ((ee.value === l.id || A.value?.id === l.id) && c.value.style.chart.midpoints.selectedEdge.stroke != null ? g = c.value.style.chart.midpoints.selectedEdge.stroke : y && c.value.style.chart.nodes.selected.downstreamEdges.stroke != null ? g = c.value.style.chart.nodes.selected.downstreamEdges.stroke : f && c.value.style.chart.nodes.selected.upstreamEdges.stroke != null && (g = c.value.style.chart.nodes.selected.upstreamEdges.stroke), y && c.value.style.chart.nodes.selected.downstreamEdges.animated === !0)
        l.animated = !0;
      else if (f && c.value.style.chart.nodes.selected.upstreamEdges.animated === !0)
        l.animated = !0;
      else {
        const S = _.value?.edges.find((R) => R.id === l.id);
        l.animated = S?.original?.animated ?? !1;
      }
      return {
        d: l.pathData,
        fill: "none",
        stroke: g,
        "stroke-width": c.value.style.chart.edges.strokeWidth * (ee.value === l.id || l.from === F.value || l.id === A.value?.id ? 2 : 1),
        "stroke-linecap": "round",
        "stroke-linejoin": "round"
      };
    }
    function ao(l) {
      const y = F.value === l.id, f = y && c.value.style.chart.nodes.selected.backgroundColor != null ? c.value.style.chart.nodes.selected.backgroundColor : l.original.backgroundColor, g = y && c.value.style.chart.nodes.selected.stroke != null ? c.value.style.chart.nodes.selected.stroke : c.value.style.chart.nodes.stroke, S = y && c.value.style.chart.nodes.selected.strokeWidth != null ? c.value.style.chart.nodes.selected.strokeWidth : c.value.style.chart.nodes.strokeWidth;
      return {
        x: l.x - l.width / 2,
        y: l.y - l.height / 2,
        width: l.width,
        height: l.height,
        rx: c.value.style.chart.nodes.borderRadius,
        fill: f,
        stroke: g,
        "stroke-width": S
      };
    }
    function lo(l) {
      const y = l.from === F.value, f = l.to === F.value;
      let g = l.original.color ?? c.value.style.chart.edges.stroke;
      return (ee.value === l.id || A.value?.id === l.id) && c.value.style.chart.midpoints.selectedEdge.stroke != null ? g = c.value.style.chart.midpoints.selectedEdge.stroke : y && c.value.style.chart.nodes.selected.downstreamEdges.stroke != null ? g = c.value.style.chart.nodes.selected.downstreamEdges.stroke : f && c.value.style.chart.nodes.selected.upstreamEdges.stroke != null && (g = c.value.style.chart.nodes.selected.upstreamEdges.stroke), {
        cx: l.midpoint.x,
        cy: l.midpoint.y,
        r: c.value.style.chart.midpoints.radius,
        fill: c.value.style.chart.midpoints.fill,
        stroke: g,
        "stroke-width": c.value.style.chart.edges.strokeWidth * (l.from === F.value || l.id === A.value?.id ? 2 : 1)
      };
    }
    function We(l) {
      const y = l.from === F.value, f = l.to === F.value;
      let g = l.color ?? l.original?.color ?? c.value.style.chart.edges.stroke;
      return (ee.value === l.id || A.value?.id === l.id) && c.value.style.chart.midpoints.selectedEdge.stroke != null ? g = c.value.style.chart.midpoints.selectedEdge.stroke : y && c.value.style.chart.nodes.selected.downstreamEdges.stroke != null ? g = c.value.style.chart.nodes.selected.downstreamEdges.stroke : f && c.value.style.chart.nodes.selected.upstreamEdges.stroke != null && (g = c.value.style.chart.nodes.selected.upstreamEdges.stroke), g;
    }
    function Kt(l) {
      return `${Zn}-${String(l).replace(/[^a-zA-Z0-9_-]/g, "_")}`;
    }
    function uo() {
      return _.value;
    }
    function co(l) {
      if (l?.stage === "start") {
        oe.value = !0;
        return;
      }
      if (l?.stage === "end") {
        oe.value = !1;
        return;
      }
      Rt();
    }
    async function Jt() {
      if (h("copyAlt", {
        config: c.value,
        dataset: He.value
      }), !c.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(c.value.userOptions.callbacks.altCopy({
        config: c.value,
        dataset: He.value
      }));
    }
    return t({
      getData: uo,
      getImage: so,
      generatePdf: Tt,
      generateSvg: Pt,
      generateImage: Rt,
      toggleAnnotator: tt,
      toggleFullscreen: It,
      zoomIn: nt,
      zoomOut: ot,
      resetZoom: Ge,
      switchDirection: it,
      copyAlt: Jt
    }), (l, y) => (C(), x("div", {
      class: tn(`vue-data-ui-component vue-ui-dag ${re.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${c.value.responsive ? "vue-ui-dag-responsive" : ""}`),
      id: `dag_${v.value}`,
      ref_key: "dagChart",
      ref: m,
      style: ye({
        fontFamily: c.value.style.fontFamily,
        backgroundColor: c.value.style.chart.backgroundColor,
        padding: "0.5rem"
      }),
      onMouseenter: Gn,
      onMouseleave: Wn
    }, [
      p($t) ? (C(), x("div", Pi, Pe(String(p($t))), 1)) : T("", !0),
      c.value.userOptions.buttons.annotator ? (C(), te(p(o), {
        key: 1,
        svgRef: p(ie),
        backgroundColor: c.value.style.chart.backgroundColor,
        color: c.value.style.chart.color,
        active: Se.value,
        isCursorPointer: Ce.value,
        onClose: tt
      }, {
        "annotator-action-close": V(() => [
          M(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": V(({ color: f }) => [
          M(l.$slots, "annotator-action-color", H(G({ color: f })), void 0, !0)
        ]),
        "annotator-action-draw": V(({ mode: f }) => [
          M(l.$slots, "annotator-action-draw", H(G({ mode: f })), void 0, !0)
        ]),
        "annotator-action-undo": V(({ disabled: f }) => [
          M(l.$slots, "annotator-action-undo", H(G({ disabled: f })), void 0, !0)
        ]),
        "annotator-action-redo": V(({ disabled: f }) => [
          M(l.$slots, "annotator-action-redo", H(G({ disabled: f })), void 0, !0)
        ]),
        "annotator-action-delete": V(({ disabled: f }) => [
          M(l.$slots, "annotator-action-delete", H(G({ disabled: f })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : T("", !0),
      c.value.userOptions.show && de.value && (p(Mt) || p(qe)) ? (C(), te(p(s), {
        ref: "userOptionsRef",
        key: `user_option_${Q.value}`,
        backgroundColor: c.value.style.chart.backgroundColor,
        color: c.value.style.chart.color,
        isPrinting: p(Qe),
        isImaging: p(et),
        uid: v.value,
        hasTooltip: !1,
        hasTable: !1,
        hasXls: !1,
        hasLabel: !1,
        hasPdf: c.value.userOptions.buttons.pdf,
        hasImg: c.value.userOptions.buttons.img,
        hasSvg: c.value.userOptions.buttons.svg,
        hasFullscreen: c.value.userOptions.buttons.fullscreen,
        hasAltCopy: c.value.userOptions.buttons.altCopy,
        isFullscreen: re.value,
        chartElement: m.value,
        position: c.value.userOptions.position,
        titles: { ...c.value.userOptions.buttonTitles },
        hasAnnotator: c.value.userOptions.buttons.annotator,
        isAnnotation: Se.value,
        callbacks: c.value.userOptions.callbacks,
        printScale: c.value.userOptions.print.scale,
        hasZoom: c.value.userOptions.buttons.zoom,
        isZoom: he.value,
        isCursorPointer: Ce.value,
        onToggleFullscreen: It,
        onGeneratePdf: p(Tt),
        onGenerateImage: co,
        onGenerateSvg: Pt,
        onToggleAnnotator: tt,
        onToggleZoom: qn,
        onCopyAlt: Jt,
        style: ye({
          visibility: p(Mt) ? p(qe) ? "visible" : "hidden" : "visible"
        })
      }, po({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: V(({ isOpen: f, color: g }) => [
            M(l.$slots, "menuIcon", H(G({ isOpen: f, color: g })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: V(() => [
            M(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: V(() => [
            M(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: V(() => [
            M(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: V(({ toggleFullscreen: f, isFullscreen: g }) => [
            M(l.$slots, "optionFullscreen", H(G({ toggleFullscreen: f, isFullscreen: g })), void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: V(({ toggleAnnotator: f, isAnnotator: g }) => [
            M(l.$slots, "optionAnnotator", H(G({ toggleAnnotator: f, isAnnotator: g })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionZoom ? {
          name: "optionZoom",
          fn: V(({ toggleZoom: f, isZoomLocked: g }) => [
            M(l.$slots, "optionZoom", H(G({ toggleZoom: f, isZoomLocked: g })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: V(({ altCopy: f }) => [
            M(l.$slots, "optionAltCopy", H(G({ altCopy: f })), void 0, !0)
          ]),
          key: "7"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "hasZoom", "isZoom", "isCursorPointer", "onGeneratePdf", "style"])) : T("", !0),
      c.value.style.chart.title.text ? (C(), x("div", {
        key: 3,
        ref_key: "chartTitle",
        ref: b,
        style: "width:100%;background:transparent;"
      }, [
        (C(), te(Do, {
          key: `title_${z.value}`,
          config: {
            title: {
              cy: "dag-title",
              ...c.value.style.chart.title
            },
            subtitle: {
              cy: "dag-subtitle",
              ...c.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : T("", !0),
      c.value.style.chart.controls.position === "top" && !p(Le) && c.value.style.chart.controls.show ? (C(), te(un, {
        key: 4,
        ref_key: "zoomControls",
        ref: N,
        config: c.value,
        scale: p(Bt),
        isFullscreen: re.value,
        withDirection: "",
        onZoomIn: nt,
        onZoomOut: ot,
        onResetZoom: y[0] || (y[0] = () => p(Ge)(!0)),
        onSwitchDirection: it
      }, null, 8, ["config", "scale", "isFullscreen"])) : T("", !0),
      p(_) ? (C(), x("svg", {
        key: 5,
        ref_key: "svgRef",
        ref: ie,
        class: tn({ "vue-ui-dag-svg": !0, "vue-data-ui-loading": p(Le) }),
        viewBox: Kn.value,
        xmlns: p(ko),
        style: ye({
          backgroundColor: c.value.style.chart.backgroundColor,
          height: "100%",
          width: "100%"
        })
      }, [
        ut(p(r)),
        c.value.style.chart.backgroundPattern.show ? (C(), x("defs", $i, [
          J("pattern", {
            id: `dag_bg_pattern_${v.value}`,
            patternUnits: "userSpaceOnUse",
            width: ae.value,
            height: ae.value
          }, [
            M(l.$slots, "background-pattern", H(G({
              x: ae.value / 2,
              y: ae.value / 2,
              color: c.value.style.chart.backgroundPattern.dotColor
            })), () => [
              J("circle", {
                cx: ae.value / 2,
                cy: ae.value / 2,
                r: io.value,
                fill: c.value.style.chart.backgroundPattern.dotColor
              }, null, 8, Fi)
            ], !0)
          ], 8, Ai)
        ])) : T("", !0),
        c.value.style.chart.backgroundPattern.show ? (C(), x("rect", {
          key: 1,
          x: p(me)?.x ?? 0,
          y: p(me)?.y ?? 0,
          width: p(me)?.width ?? 0,
          height: p(me)?.height ?? 0,
          fill: `url(#dag_bg_pattern_${v.value})`,
          style: ye({
            pointerEvents: "none",
            opacity: c.value.style.chart.backgroundPattern.opacity
          })
        }, null, 12, Bi)) : T("", !0),
        p(_).arrowShape !== "undirected" ? (C(), x("defs", zi, [
          (C(!0), x(ue, null, De(p(_).edges, (f) => (C(), x("marker", {
            key: `marker_${f.id}`,
            id: Kt(f.id),
            markerWidth: p(_).arrowSize,
            markerHeight: p(_).arrowSize,
            refX: p(_).arrowSize - 3,
            refY: p(_).arrowSize / 2,
            orient: "auto",
            markerUnits: "strokeWidth"
          }, [
            p(_).arrowShape === "normal" ? (C(), x("path", {
              key: 0,
              d: `M 0 0 L ${p(_).arrowSize} ${p(_).arrowSize / 2} L 0 ${p(_).arrowSize} Z`,
              fill: We(f),
              stroke: We(f),
              "stroke-width": "0",
              style: { transition: "stroke 0.2s ease-in-out, fill 0.2s ease-in-out, stroke-width 0.2s ease-in-out" }
            }, null, 8, Ii)) : (C(), x("path", {
              key: 1,
              d: `M 0 0 L ${p(_).arrowSize} ${p(_).arrowSize / 2} L 0 ${p(_).arrowSize} L ${p(_).arrowSize / 3} ${p(_).arrowSize / 2} Z`,
              fill: We(f),
              stroke: We(f),
              "stroke-width": "0",
              style: { transition: "stroke 0.2s ease-in-out, fill 0.2s ease-in-out, stroke-width 0.2s ease-in-out" }
            }, null, 8, Hi))
          ], 8, Vi))), 128))
        ])) : T("", !0),
        J("g", Gi, [
          (C(!0), x(ue, null, De(p(_).edges, (f) => (C(), x(ue, {
            key: f.id
          }, [
            J("path", we({
              "data-cy-edge": "",
              ref_for: !0,
              ref: oo(f.id)
            }, { ref_for: !0 }, ro(f), { style: { "pointer-events": "none", transition: "stroke-width 0.2s ease-in-out, stroke 0.2s ease-in-out" } }), null, 16),
            c.value.style.chart.midpoints.show ? (C(), x("circle", we({
              key: 0,
              "data-cy-midpoint": "",
              class: "vue-ui-dag-edge-midpoint"
            }, { ref_for: !0 }, lo(f), {
              style: { transition: "stroke-width 0.2s ease-in-out, stroke 0.2s ease-in-out" },
              onMouseenter: (g) => eo(f),
              onMouseleave: to
            }), null, 16, Wi)) : T("", !0)
          ], 64))), 128))
        ]),
        J("g", Yi, [
          (C(!0), x(ue, null, De(p(_).nodes, (f) => (C(), x("g", {
            key: f.id,
            class: "vue-ui-dag-node",
            onClick: nn((g) => c.value.style.chart.nodes.tooltip.showOnClick && Ht(f), ["stop"]),
            onMouseenter: (g) => Xt(f.id),
            onMouseleave: (g) => qt(f.id)
          }, [
            l.$slots.node ? T("", !0) : (C(), x("rect", we({
              key: 0,
              "data-cy-node": ""
            }, { ref_for: !0 }, ao(f), {
              style: {
                cursor: c.value.style.chart.nodes.tooltip.showOnClick && Ce.value ? "pointer" : "default",
                transition: "stroke 0.2s ease-in-out, stroke-width 0.2s ease-in-out, fill 0.2s ease-in-out"
              }
            }), null, 16)),
            l.$slots.node ? (C(), x("foreignObject", {
              key: 1,
              x: f.x - f.width / 2,
              y: f.y - f.height / 2,
              width: f.width,
              height: f.height
            }, [
              M(l.$slots, "node", we({ ref_for: !0 }, { node: f, orientation: X.value }), void 0, !0)
            ], 8, Ui)) : T("", !0)
          ], 40, ji))), 128))
        ]),
        J("g", Zi, [
          (C(!0), x(ue, null, De(p(_).edges, (f) => (C(), x("path", {
            key: f.id,
            d: f.pathData,
            fill: "none",
            stroke: "transparent",
            "stroke-width": c.value.style.chart.edges.strokeWidth * (f.from === F.value || f.id === A.value?.id ? 1.3 : 1),
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            "marker-end": p(_).arrowShape === "undirected" ? null : `url(#${Kt(f.id)})`,
            style: { "pointer-events": "none", transition: "stroke-width 0.2s ease-in-out, stroke 0.2s ease-in-out" }
          }, null, 8, Xi))), 128))
        ]),
        J("g", qi, [
          (C(!0), x(ue, null, De(p(_).nodes, (f) => (C(), x("g", {
            key: f.id,
            onClick: nn((g) => c.value.style.chart.nodes.tooltip.showOnClick && Ht(f), ["stop"]),
            onMouseenter: (g) => Xt(f.id),
            onMouseleave: (g) => qt(f.id)
          }, [
            l.$slots["free-node-label"] ? T("", !0) : (C(), x(ue, { key: 0 }, [
              l.$slots["node-label"] ? (C(), x("text", {
                key: 0,
                x: f.x,
                y: f.y + c.value.style.chart.nodes.labels.fontSize / 3,
                "text-anchor": "middle",
                "font-size": c.value.style.chart.nodes.labels.fontSize,
                fill: F.value === f.id && c.value.style.chart.nodes.selected.labelColor != null ? c.value.style.chart.nodes.selected.labelColor : f.original.color,
                "font-weight": c.value.style.chart.nodes.labels.bold ? "bold" : "normal",
                style: { transition: "fill 0.2s ease-in-out" }
              }, [
                M(l.$slots, "node-label", we({ ref_for: !0 }, { node: f, orientation: X.value }), () => [
                  yo(Pe(f.label), 1)
                ], !0)
              ], 8, Ji)) : !l.$slots["free-node-label"] && !l.$slots.node ? (C(), x("text", {
                key: 1,
                "data-cy-node-label": "",
                x: f.x,
                y: f.y + c.value.style.chart.nodes.labels.fontSize / 3,
                "text-anchor": "middle",
                "font-size": c.value.style.chart.nodes.labels.fontSize,
                fill: F.value === f.id && c.value.style.chart.nodes.selected.labelColor != null ? c.value.style.chart.nodes.selected.labelColor : f.original.color,
                "font-weight": c.value.style.chart.nodes.labels.bold ? "bold" : "normal",
                style: { transition: "fill 0.2s ease-in-out" },
                innerHTML: p(Eo)({
                  content: f.label,
                  fontSize: c.value.style.chart.nodes.labels.fontSize,
                  fontWeight: c.value.style.chart.nodes.labels.bold ? "bold" : "normal",
                  fill: F.value === f.id && c.value.style.chart.nodes.selected.labelColor != null ? c.value.style.chart.nodes.selected.labelColor : f.original.color,
                  x: f.x,
                  y: f.y,
                  autoOffset: !0
                })
              }, null, 8, Qi)) : T("", !0)
            ], 64)),
            l.$slots["free-node-label"] ? (C(), x("g", er, [
              M(l.$slots, "free-node-label", we({ ref_for: !0 }, { node: f, layoutData: p(_), orientation: X.value }), void 0, !0)
            ])) : T("", !0)
          ], 40, Ki))), 128))
        ]),
        M(l.$slots, "svg", {
          svg: {
            drawingArea: p(me),
            data: p(_),
            orientation: X.value,
            isPrintingImg: p(Qe) | p(et) | oe.value,
            isPrintingSvg: fe.value
          }
        }, void 0, !0)
      ], 14, Di)) : T("", !0),
      l.$slots.watermark ? (C(), x("div", tr, [
        M(l.$slots, "watermark", H(G({ isPrinting: p(Qe) || p(et) || oe.value || fe.value })), void 0, !0)
      ])) : T("", !0),
      ut(on, { name: "fade" }, {
        default: V(() => [
          Ie.value ? (C(), te(sn, {
            key: 0,
            to: re.value ? m.value : "body"
          }, [
            J("div", {
              "data-cy-tooltip-midpoint": "",
              ref_key: "tooltipRef",
              ref: Et,
              class: "vue-ui-dag-tooltip",
              style: ye({
                ..._t.value,
                maxWidth: c.value.style.chart.midpoints.tooltip.maxWidth,
                "--vue-data-ui-dag-tooltip-background": c.value.style.chart.midpoints.tooltip.backgroundColor,
                "--vue-data-ui-dag-tooltip-color": c.value.style.chart.midpoints.tooltip.color
              }),
              "data-position": Ct.value
            }, [
              M(l.$slots, "tooltip-midpoint", H(G({ edge: A.value, layoutData: p(_) })), () => [
                A.value ? (C(), x("div", or, Pe(At(A.value.from)?.label ?? A.value.from) + " → " + Pe(At(A.value.to)?.label ?? A.value.to), 1)) : T("", !0)
              ], !0)
            ], 12, nr)
          ], 8, ["to"])) : T("", !0)
        ]),
        _: 3
      }),
      ut(on, { name: "fade" }, {
        default: V(() => [
          se.value ? (C(), te(sn, {
            key: 0,
            to: re.value ? m.value : "body"
          }, [
            J("div", {
              "data-cy-tooltip-node": "",
              ref_key: "nodeTooltipRef",
              ref: Ze,
              class: "vue-ui-dag-node-tooltip",
              style: ye({
                maxWidth: c.value.style.chart.nodes.tooltip.maxWidth,
                left: Xe.value.left,
                top: Xe.value.top,
                "--vue-data-ui-dag-node-tooltip-background": c.value.style.chart.nodes.tooltip.backgroundColor,
                "--vue-data-ui-dag-node-tooltip-color": c.value.style.chart.nodes.tooltip.color
              }),
              "data-position": St.value
            }, [
              M(l.$slots, "tooltip-node", H(G({ node: Ne.value, layoutData: p(_) })), () => [
                Ne.value ? (C(), x("div", ir, Pe(Ne.value.label), 1)) : T("", !0)
              ], !0)
            ], 12, sr)
          ], 8, ["to"])) : T("", !0)
        ]),
        _: 3
      }),
      c.value.style.chart.controls.position === "bottom" && !p(Le) && c.value.style.chart.controls.show ? (C(), te(un, {
        key: 7,
        ref_key: "zoomControls",
        ref: N,
        config: c.value,
        scale: p(Bt),
        isFullscreen: re.value,
        withDirection: "",
        onZoomIn: nt,
        onZoomOut: ot,
        onResetZoom: y[1] || (y[1] = () => p(Ge)(!0)),
        onSwitchDirection: it
      }, null, 8, ["config", "scale", "isFullscreen"])) : T("", !0),
      l.$slots.source ? (C(), x("div", {
        key: 8,
        ref_key: "source",
        ref: L,
        dir: "auto"
      }, [
        M(l.$slots, "source", {}, void 0, !0)
      ], 512)) : T("", !0),
      M(l.$slots, "skeleton", {}, () => [
        p(Le) ? (C(), te(Lo, { key: 0 })) : T("", !0)
      ], !0)
    ], 46, Ri));
  }
}, xr = /* @__PURE__ */ Fo(rr, [["__scopeId", "data-v-f3c6f09c"]]);
export {
  xr as default
};
