// vue-data-ui/dist/vue-ui-radar-DE70LPBZ.js
import { defineAsyncComponent as O, computed as d, ref as i, toRefs as Yt, watch as me, shallowRef as et, onMounted as jt, onBeforeUnmount as qt, openBlock as n, createElementBlock as u, unref as r, normalizeStyle as q, normalizeClass as H, createBlock as I, withCtx as v, renderSlot as p, normalizeProps as b, guardReactiveProps as k, createCommentVNode as h, createElementVNode as $, createSlots as tt, createVNode as oe, Fragment as V, renderList as z, toDisplayString as D, Teleport as Ht, resolveDynamicComponent as Kt, mergeProps as Jt, createTextVNode as be, nextTick as ke } from "vue";
import { j as Qt, t as at, m as Zt, p as U, o as ea, q as B, n as ta, s as ot, A as lt, K as aa, l as G, k as Ce, X as oa, u as we, v as la, ag as nt, r as st, a0 as na, a1 as sa, T as ra, U as ua } from "./lib-GwPnJSrY.js";
import { t as ia, u as ca } from "./useResponsive-DfdjqQps.js";
import { u as va } from "./useConfig-CIgyKZ6R.js";
import { u as da, B as pa } from "./BaseScanner-BhhkKVoj.js";
import { u as ga } from "./usePrinter-DiK5R6k-.js";
import { u as ha } from "./useSvgExport-ByUukOZt.js";
import { u as $e } from "./useNestedProp-BBT0mbHh.js";
import { u as fa } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ya } from "./useUserOptionState-BIvW1Kz7.js";
import { u as ma } from "./useChartAccessibility-9icAAmYg.js";
import { u as ba } from "./useAutoSizeLabelsInsideViewbox-80fzdfxt.js";
import ka from "./img-BYYzY_QU.js";
import Ca from "./Title-BfyrP-jp.js";
import wa from "./Legend-1LMMsPxL.js";
import $a from "./vue_ui_radar-DryGBOrz.js";
import { B as Ta } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Pa } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Aa = ["id"], _a = ["id"], Sa = ["xmlns", "viewBox"], Oa = ["width", "height"], Ia = ["id"], Fa = ["stop-color"], xa = ["stop-color"], La = { key: 1 }, Na = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], Ra = { key: 0 }, Va = ["d", "stroke", "stroke-width"], za = ["d", "stroke", "stroke-width"], Da = { key: 2 }, Ua = ["x", "y", "text-anchor", "font-size", "fill", "onMouseenter", "onMouseleave", "onClick"], Ba = ["points", "stroke", "stroke-width"], Ea = ["points", "stroke", "stroke-width", "fill"], Ma = { key: 3 }, Wa = ["cx", "cy", "fill", "r", "stroke"], Ga = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Xa = ["id"], Ya = ["onClick"], ja = {
  key: 0,
  style: { "max-width": "200px", margin: "0 auto" }
}, qa = {
  class: "vue-ui-radar-tooltip-datalabel",
  style: { width: "100%" }
}, Ha = { class: "vue-ui-radar-tooltip-datalabel-name" }, Ka = { key: 0 }, Ja = {
  __name: "vue-ui-radar",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["selectLegend", "copyAlt"],
  setup(rt, { expose: ut, emit: it }) {
    const ct = O(() => import("./Tooltip-DL3QgjZf.js")), vt = O(() => import("./BaseIcon-CNx8Tzl5.js")), dt = O(() => import("./vue-ui-sparkbar-DUDWvTo-.js")), pt = O(() => import("./vue-ui-accordion-BGHTqIBt.js")), gt = O(() => import("./DataTable-Bo1pmAqG.js")), ht = O(() => import("./PenAndPaper-DWV1FMIi.js")), ft = O(() => import("./UserOptions-CrLqEimB.js")), yt = O(() => import("./PackageVersion-DN9d0ME5.js")), mt = O(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_radar: bt } = va(), { isThemeValid: kt, warnInvalidTheme: Ct } = fa(), _ = rt, Te = d(() => !!_.dataset && Object.keys(_.dataset).length), C = i(Qt()), le = i(!1), ne = i(""), Pe = i(0), S = i(null), Ae = i(null), _e = i(null), Se = i(null), Oe = i(null), Ie = i(0), Fe = i(0), xe = i(0), Le = i(!1), X = i(null), se = i(null), K = i(!1), J = i(!1), e = i(ie()), x = d(() => e.value.userOptions.useCursorPointer), wt = d(() => at({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        useCssAnimation: !1,
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              grid: {
                stroke: "#6A6A6A90"
              },
              labels: {
                dataLabels: { show: !1 }
              },
              outerPolygon: {
                stroke: "#6A6A6A"
              }
            },
            legend: {
              backgroundColor: "transparent"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: re, FINAL_DATASET: L } = da({
      ...Yt(_),
      FINAL_CONFIG: e,
      prepareConfig: ie,
      callback: () => {
        Promise.resolve().then(async () => {
          await ke(), y.value.showTable = e.value.table.show;
        });
      },
      skeletonDataset: _.config?.skeletonDataset ?? {
        categories: [{ name: "_", color: "#6A6A6A" }],
        series: [
          { name: "_", values: [0.6], target: 1 },
          { name: "_", values: [0.6], target: 1 },
          { name: "_", values: [0.6], target: 1 },
          { name: "_", values: [0.6], target: 1 },
          { name: "_", values: [0.6], target: 1 },
          { name: "_", values: [0.6], target: 1 }
        ]
      },
      skeletonConfig: at({
        defaultConfig: e.value,
        userConfig: wt.value
      })
    }), { userOptionsVisible: ue, setUserOptionsVisibility: Ne, keepUserOptionState: Re } = ya({ config: e.value }), { svgRef: Q } = ma({ config: e.value.style.chart.title });
    function ie() {
      const t = $e({
        userConfig: _.config,
        defaultConfig: bt
      }), o = t.theme;
      if (!o) return t;
      if (!kt.value(t))
        return Ct(t), t;
      const a = $e({
        userConfig: $a[o] || _.config,
        defaultConfig: t
      }), l = $e({
        userConfig: _.config,
        defaultConfig: a
      });
      return {
        ...l,
        customPalette: l.customPalette.length ? l.customPalette : Zt[o] || U
      };
    }
    me(() => _.config, (t) => {
      e.value = ie(), ue.value = !e.value.userOptions.showOnChartHover, Ve(), Ie.value += 1, Fe.value += 1, xe.value += 1, y.value.dataLabels.show = e.value.style.chart.layout.labels.dataLabels.show, y.value.showTable = e.value.table.show, y.value.showTooltip = e.value.style.chart.tooltip ? e.value.style.chart.tooltip.show : !1;
    }, { deep: !0 });
    const F = et(null), E = et(null), N = d(() => e.value.debug);
    function Ve() {
      if (ea(_.dataset) && B({
        componentName: "VueUiRadar",
        type: "dataset",
        debug: N.value
      }), e.value.responsive) {
        const t = ia(() => {
          const { width: o, height: a } = ca({
            chart: S.value,
            title: e.value.style.chart.title.text ? Ae.value : null,
            legend: e.value.style.chart.legend.show ? _e.value : null,
            source: Se.value,
            noTitle: Oe.value
          });
          requestAnimationFrame(() => {
            c.value.width = o, c.value.height = a, Xe();
          });
        });
        F.value && (E.value && F.value.unobserve(E.value), F.value.disconnect()), F.value = new ResizeObserver(t), E.value = S.value.parentNode, F.value.observe(E.value);
      }
      Xe();
    }
    jt(() => {
      Le.value = !0, Ve();
    }), qt(() => {
      F.value && (E.value && F.value.unobserve(E.value), F.value.disconnect());
    });
    const { isPrinting: ce, isImaging: ve, generatePdf: ze, generateImage: De } = ga({
      elementId: `vue-ui-radar_${C.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-radar",
      options: e.value.userOptions.print
    }), $t = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), Ue = d(() => ta(e.value.customPalette)), y = i({
      dataLabels: {
        show: e.value.style.chart.layout.labels.dataLabels.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    }), Tt = d(() => ({
      style: {
        backgroundColor: "#FFFFFF00",
        animation: {
          show: e.value.style.chart.tooltip.animation.show,
          animationFrames: e.value.style.chart.tooltip.animation.animationFrames
        },
        labels: {
          fontSize: e.value.style.chart.tooltip.fontSize,
          name: {
            color: e.value.style.chart.tooltip.color
          }
        },
        gutter: {
          backgroundColor: "#CCCCCC",
          opacity: 30
        }
      }
    })), c = i({
      height: 312,
      width: 512
    }), Be = it, g = i([]), Y = i(null), M = i(!1);
    function Pt() {
      g.value.length ? g.value = [] : A.value.forEach((t, o) => {
        g.value.push(o);
      });
    }
    function Z(t) {
      M.value = !0, g.value.includes(t) ? (Y.value = t, g.value = g.value.filter((o) => o !== t), setTimeout(() => {
        M.value = !1, Y.value = null;
      }, 500)) : (g.value.push(t), setTimeout(() => {
        M.value = !1;
      }, 500)), Be("selectLegend", A.value.filter((o, a) => !g.value.includes(a)).map((o) => ({
        name: o.name,
        color: o.color,
        proportion: o.totalProportion
      })));
    }
    function At() {
      return A.value.map((t) => ({
        name: t.name,
        color: t.color,
        proportion: t.totalProportion
      }));
    }
    function _t() {
      const t = L.value;
      if ([null, void 0].includes(t?.categories)) {
        B({
          componentName: "VueUiRadar",
          type: "dataset",
          debug: N.value
        }), B({
          componentName: "VueUiRadar",
          type: "datasetAttribute",
          property: "categories ({ name: string; prefix?: string; suffix?: string}[])",
          debug: N.value
        });
        return;
      }
      t.categories.length === 0 ? B({
        componentName: "VueUiRadar",
        type: "datasetAttributeEmpty",
        property: "categories",
        debug: N.value
      }) : t.categories.forEach((o, a) => {
        st({
          datasetObject: o,
          requiredAttributes: ["name"]
        }).forEach((l) => {
          B({
            componentName: "VueUiRadar",
            type: "datasetAttribute",
            property: `category.${l} at index ${a}`,
            index: a,
            debug: N.value
          });
        });
      }), [null, void 0].includes(t?.series) ? B({
        componentName: "VueUiRadar",
        type: "datasetAttribute",
        property: "series ({ name: string; values: number[]; color?: string; target: number}[])",
        debug: N.value
      }) : t.series.forEach((o, a) => {
        st({
          datasetObject: o,
          requiredAttributes: ["name", "values", "target"]
        }).forEach((l) => {
          B({
            componentName: "VueUiRadar",
            type: "datasetSerieAttribute",
            key: "series",
            property: l,
            index: a,
            debug: N.value
          });
        });
      });
    }
    me(
      () => L.value,
      () => _t(),
      { deep: !0, immediate: !0 }
    );
    const T = d(() => {
      const t = Array.isArray(L.value?.categories) ? L.value.categories : [], o = Ue.value ?? U;
      return t.map((a, l) => ({
        name: a?.name ?? "",
        categoryId: `radar_category_${C.value}_${l}`,
        color: ot(a?.color) || o[l] || U[l % U.length],
        prefix: a?.prefix ?? "",
        suffix: a?.suffix ?? ""
      }));
    }), w = d(() => L.value.series.map((t, o) => ({
      ...t,
      color: ot(t.color) || Ue.value[o] || U[o] || U[o % U.length],
      serieId: `radar_serie_${C.value}_${o}`,
      formatter: t.formatter || null,
      absoluteIndex: o
    })));
    function Ee(t) {
      if (!w.value.length)
        return e.value.debug && console.warn("VueUiRadar - There are no series to show."), null;
      const o = A.value.find((a) => a.name === t);
      return o || (e.value.debug && console.warn(`VueUiRadar - Series name not found "${t}"`), null);
    }
    function St(t) {
      const o = Ee(t);
      o !== null && g.value.includes(o.absoluteIndex) && Z(o.absoluteIndex);
    }
    function Ot(t) {
      const o = Ee(t);
      o !== null && (g.value.includes(o.absoluteIndex) || Z(o.absoluteIndex));
    }
    const Me = d(() => Math.max(...w.value.flatMap((t) => t.values))), We = d(() => w.value.length), de = d(() => Math.min(c.value.width, c.value.height) / 3), Ge = d(() => lt({
      plot: { x: c.value.width / 2, y: c.value.height / 2 },
      radius: de.value,
      sides: We.value,
      rotation: 0
    })), It = d(() => {
      const t = [];
      for (let o = 0; o < de.value; o += de.value / e.value.style.chart.layout.grid.graduations)
        t.push(o);
      return t;
    }), P = d(() => Ge.value.coordinates.map((t, o) => {
      const a = w.value[o].values.map((l) => xt({
        centerX: c.value.width / 2,
        centerY: c.value.height / 2,
        apexX: t.x,
        apexY: t.y,
        proportion: l / (w.value[o].target || Me.value)
      }));
      return {
        ...t,
        ...w.value[o],
        plots: a
      };
    }).map((t) => ({
      ...t,
      labelX: pe(t).x,
      labelY: pe(t).y,
      labelAnchor: pe(t).anchor
    })));
    function pe({ x: t, y: o }) {
      let a = "middle";
      return t = Math.round(t), o = Math.round(o), t > c.value.width / 2 && (t += 12, a = "start"), t < c.value.width / 2 && (t -= 12, a = "end"), o > c.value.height / 2 + 1 && (o += 20), o < c.value.height / 2 - 1 && (o -= 12), o === c.value.height / 2 && (o += 4), { x: t, y: o, anchor: a };
    }
    const Ft = d({
      get: () => e.value.style.chart.layout.labels.dataLabels.fontSize,
      set: (t) => t
    }), { autoSizeLabels: Xe } = ba({
      svgRef: Q,
      fontSize: e.value.style.chart.layout.labels.dataLabels.fontSize,
      minFontSize: 6,
      sizeRef: Ft,
      labelClass: ".vue-ui-radar-apex-label"
    });
    function xt({ centerX: t, centerY: o, apexX: a, apexY: l, proportion: s }) {
      return {
        x: t + (a - t) * s,
        y: o + (l - o) * s
      };
    }
    const A = d(() => {
      const t = w.value.map((o, a) => o.values.map((l) => l / (o.target || Me.value)));
      return T.value.map((o, a) => {
        const l = aa(t.map((s) => s[a]).reduce((s, f) => s + f, 0) / w.value.length);
        return {
          ...o,
          absoluteIndex: a,
          totalProportion: l,
          shape: "circle",
          opacity: g.value.includes(a) ? 0.5 : 1,
          segregate: () => Z(a),
          isSegregated: g.value.includes(a),
          display: `${o.name}: ${G({
            v: (l ?? 0) * 100,
            s: "%",
            r: e.value.style.chart.legend.roundingPercentage
          })}`
        };
      });
    }), Lt = d(() => ({
      cy: "radar-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), ee = d(() => {
      const t = [
        { name: e.value.translations.datapoint, color: "" },
        { name: e.value.translations.target, color: "" },
        ...A.value
      ], o = L.value.series.map((l) => [
        l.name,
        Ce(
          l.formatter,
          l.target,
          G({
            p: l.prefix,
            v: l.target,
            s: l.suffix,
            r: e.value.table.td.roundingValue
          })
        ),
        ...l.values.map((s, f) => `${Ce(
          l.formatter,
          s,
          G({ p: T.value[f]?.prefix ?? "", v: s, s: T.value[f]?.suffix ?? "", r: e.value.table.td.roundingValue })
        )} (${isNaN(s / l.target) ? "" : G({
          v: s / l.target * 100,
          s: "%",
          r: e.value.table.td.roundingPercentage
        })})`)
      ]), a = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return { head: t, body: o, config: a, colNames: t };
    }), te = i(null), ge = i([]), he = i(null);
    function Nt(t, o) {
      le.value = !1, te.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: o });
    }
    function Rt(t, o) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: o });
    }
    function Vt(t, o) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: o });
      const a = T.value.slice();
      ge.value = [], te.value = o, le.value = !0, he.value = {
        datapoint: t,
        seriesIndex: o,
        series: {
          categories: a,
          datapoints: w.value,
          radar: P.value
        },
        config: e.value
      };
      const l = e.value.style.chart.tooltip.customFormat;
      if (ra(l) && ua(() => l({
        seriesIndex: o,
        datapoint: t,
        series: { categories: a, datapoints: w.value, radar: P.value },
        config: e.value
      }))) {
        ne.value = l({
          seriesIndex: o,
          datapoint: t,
          series: { categories: a, datapoints: w.value, radar: P.value },
          config: e.value
        });
        return;
      }
      ne.value = `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;">${t.name}</div>`;
      for (let s = 0; s < t.values.length; s += 1)
        if (!g.value.includes(s)) {
          const f = a[s], m = t.values[s], Qe = isNaN(m / t.target) ? 0 : m / t.target * 100, W = Ce(
            t.formatter,
            m,
            G({
              p: T.value[s].prefix,
              v: m,
              s: T.value[s].suffix,
              r: e.value.style.chart.tooltip.roundingValue
            }),
            { datapoint: t }
          ), Ze = G({
            v: Qe,
            s: "%",
            r: e.value.style.chart.tooltip.roundingPercentage
          }), Xt = e.value.style.chart.tooltip.showValue && e.value.style.chart.tooltip.showPercentage ? `${W} (${Ze})` : e.value.style.chart.tooltip.showValue && !e.value.style.chart.tooltip.showPercentage ? W : !e.value.style.chart.tooltip.showValue && e.value.style.chart.tooltip.showPercentage ? `${Ze}` : "";
          ge.value.push({
            name: f?.name ?? `#${s + 1}`,
            value: t.values[s] / t.target * 100,
            color: f?.color,
            suffix: Xt,
            prefix: "",
            rounding: e.value.style.chart.tooltip.roundingPercentage,
            formatter: t.formatter
          });
        }
    }
    function fe(t = null) {
      ke(() => {
        const o = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], a = [[""], [e.value.translations.target], ...A.value.flatMap((m) => [[m.name], ["%"]])], l = L.value.series.map((m, Qe) => [m.name, m.target, ...m.values.flatMap((W) => [
          W,
          isNaN(W / m.target) ? "" : W / m.target * 100
        ])]), s = o.concat([a]).concat(l), f = na(s);
        t ? t(f) : sa({ csvContent: f, title: e.value.style.chart.title.text || "vue-ui-radar" });
      });
    }
    const R = i(!1);
    function Ye(t) {
      R.value = t, Pe.value += 1;
    }
    function je() {
      y.value.showTable = !y.value.showTable;
    }
    function qe() {
      y.value.showTooltip = !y.value.showTooltip;
    }
    const ae = i(!1);
    function ye() {
      ae.value = !ae.value;
    }
    async function zt({ scale: t = 2 } = {}) {
      if (!S.value) return;
      const { width: o, height: a } = S.value.getBoundingClientRect(), l = o / a, { imageUri: s, base64: f } = await ka({ domElement: S.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: s,
        base64: f,
        title: e.value.style.chart.title.text,
        width: o,
        height: a,
        aspectRatio: l
      };
    }
    const j = d(() => {
      const t = e.value.table.useDialog && !e.value.table.show, o = y.value.showTable;
      return {
        component: t ? mt : pt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: R.value,
          fullscreenParent: S.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: x.value
        } : {
          hideDetails: !0,
          config: {
            open: o,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    me(() => y.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && X.value ? X.value.open() : "close" in X.value && X.value.close());
    });
    function He() {
      y.value.showTable = !1, se.value && se.value.setTableIconState(!1);
    }
    const Dt = d(() => A.value.map((t) => ({
      ...t,
      name: t.display
    }))), Ut = d(() => e.value.style.chart.backgroundColor), Bt = d(() => e.value.style.chart.legend), Et = d(() => e.value.style.chart.title), { exportSvg: Mt, getSvg: Wt } = ha({
      svg: Q,
      title: Et,
      legend: Bt,
      legendItems: Dt,
      backgroundColor: Ut
    });
    async function Ke({ isCb: t }) {
      J.value = !0, await ke();
      try {
        if (t) {
          const { blob: o, url: a, text: l, dataUrl: s } = await Wt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: o, url: a, text: l, dataUrl: s }));
        } else
          await Promise.resolve(Mt());
      } finally {
        J.value = !1;
      }
    }
    function Gt(t) {
      if (t?.stage === "start") {
        K.value = !0;
        return;
      }
      if (t?.stage === "end") {
        K.value = !1;
        return;
      }
      De();
    }
    async function Je() {
      if (Be("copyAlt", {
        config: e.value,
        dataset: P.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: P.value
      }));
    }
    return ut({
      getData: At,
      getImage: zt,
      generatePdf: ze,
      generateCsv: fe,
      generateImage: De,
      generateSvg: Ke,
      hideSeries: Ot,
      showSeries: St,
      toggleTable: je,
      toggleTooltip: qe,
      toggleAnnotator: ye,
      toggleFullscreen: Ye,
      copyAlt: Je
    }), (t, o) => (n(), u("div", {
      class: H(`vue-data-ui-component vue-ui-radar ${R.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      ref_key: "radarChart",
      ref: S,
      id: `vue-ui-radar_${C.value}`,
      style: q(`font-family:${e.value.style.fontFamily};width:100%; ${e.value.responsive ? "height: 100%;" : ""} text-align:center;background:${e.value.style.chart.backgroundColor}`),
      onMouseenter: o[2] || (o[2] = () => r(Ne)(!0)),
      onMouseleave: o[3] || (o[3] = () => r(Ne)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (n(), I(r(ht), {
        key: 0,
        svgRef: r(Q),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: ae.value,
        isCursorPointer: x.value,
        onClose: ye
      }, {
        "annotator-action-close": v(() => [
          p(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": v(({ color: a }) => [
          p(t.$slots, "annotator-action-color", b(k({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": v(({ mode: a }) => [
          p(t.$slots, "annotator-action-draw", b(k({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": v(({ disabled: a }) => [
          p(t.$slots, "annotator-action-undo", b(k({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": v(({ disabled: a }) => [
          p(t.$slots, "annotator-action-redo", b(k({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": v(({ disabled: a }) => [
          p(t.$slots, "annotator-action-delete", b(k({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : h("", !0),
      $t.value ? (n(), u("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Oe,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : h("", !0),
      e.value.style.chart.title.text ? (n(), u("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Ae,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (n(), I(Ca, {
          key: `title_${Ie.value}`,
          config: {
            title: {
              cy: "radar-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "radar-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : h("", !0),
      $("div", {
        id: `legend-top-${C.value}`
      }, null, 8, _a),
      e.value.userOptions.show && Te.value && (r(Re) || r(ue)) ? (n(), I(r(ft), {
        ref_key: "userOptionsRef",
        ref: se,
        key: `user_options_${Pe.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isImaging: r(ve),
        isPrinting: r(ce),
        uid: C.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: R.value,
        isTooltip: y.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: S.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: ae.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: x.value,
        onToggleFullscreen: Ye,
        onGeneratePdf: r(ze),
        onGenerateCsv: fe,
        onGenerateImage: Gt,
        onGenerateSvg: Ke,
        onToggleTable: je,
        onToggleTooltip: qe,
        onToggleAnnotator: ye,
        onCopyAlt: Je,
        style: q({
          visibility: r(Re) ? r(ue) ? "visible" : "hidden" : "visible"
        })
      }, tt({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: v(({ isOpen: a, color: l }) => [
            p(t.$slots, "menuIcon", b(k({ isOpen: a, color: l })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: v(() => [
            p(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: v(() => [
            p(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: v(() => [
            p(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: v(() => [
            p(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: v(() => [
            p(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: v(() => [
            p(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: v(({ toggleFullscreen: a, isFullscreen: l }) => [
            p(t.$slots, "optionFullscreen", b(k({ toggleFullscreen: a, isFullscreen: l })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: v(({ toggleAnnotator: a, isAnnotator: l }) => [
            p(t.$slots, "optionAnnotator", b(k({ toggleAnnotator: a, isAnnotator: l })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: v(({ altCopy: a }) => [
            p(t.$slots, "optionAltCopy", b(k({ altCopy: a })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : h("", !0),
      (n(), u("svg", {
        ref_key: "svgRef",
        ref: Q,
        xmlns: r(oa),
        class: H({ "vue-data-ui-fullscreen--on": R.value, "vue-data-ui-fulscreen--off": !R.value }),
        viewBox: `0 0 ${c.value.width <= 0 ? 10 : c.value.width} ${c.value.height <= 0 ? 10 : c.value.height}`,
        style: q(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        oe(r(yt)),
        t.$slots["chart-background"] ? (n(), u("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: c.value.width <= 0 ? 10 : c.value.width,
          height: c.value.height <= 0 ? 10 : c.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          p(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Oa)) : h("", !0),
        $("defs", null, [
          (n(!0), u(V, null, z(T.value, (a, l) => (n(), u("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `radar_gradient_${C.value}_${l}`
          }, [
            $("stop", {
              offset: "0%",
              "stop-color": r(we)(r(la)(a.color, 0.05), e.value.style.chart.layout.dataPolygon.opacity)
            }, null, 8, Fa),
            $("stop", {
              offset: "100%",
              "stop-color": r(we)(a.color, e.value.style.chart.layout.dataPolygon.opacity)
            }, null, 8, xa)
          ], 8, Ia))), 256))
        ]),
        e.value.style.chart.layout.grid.show ? (n(), u("g", La, [
          (n(!0), u(V, null, z(P.value, (a) => (n(), u("line", {
            x1: c.value.width / 2,
            y1: c.value.height / 2,
            x2: a.x,
            y2: a.y,
            stroke: e.value.style.chart.layout.grid.stroke,
            "stroke-width": e.value.style.chart.layout.grid.strokeWidth
          }, null, 8, Na))), 256)),
          e.value.style.chart.layout.grid.graduations > 0 ? (n(), u("g", Ra, [
            (n(!0), u(V, null, z(It.value, (a) => (n(), u("path", {
              d: r(lt)({
                plot: { x: c.value.width / 2, y: c.value.height / 2 },
                radius: a,
                sides: We.value,
                rotation: 0
              }).path,
              fill: "none",
              stroke: e.value.style.chart.layout.grid.stroke,
              "stroke-width": e.value.style.chart.layout.grid.strokeWidth
            }, null, 8, Va))), 256))
          ])) : h("", !0)
        ])) : h("", !0),
        $("path", {
          d: Ge.value.path,
          fill: "none",
          stroke: e.value.style.chart.layout.outerPolygon.stroke,
          "stroke-width": e.value.style.chart.layout.outerPolygon.strokeWidth,
          "stroke-linejoin": "round",
          "stroke-linecap": "round"
        }, null, 8, za),
        e.value.style.chart.layout.labels.dataLabels.show ? (n(), u("g", Da, [
          (n(!0), u(V, null, z(P.value, (a, l) => (n(), u("text", {
            class: "vue-ui-radar-apex-label",
            x: a.labelX,
            y: a.labelY,
            "text-anchor": a.labelAnchor,
            "font-size": e.value.style.chart.layout.labels.dataLabels.fontSize,
            fill: e.value.style.chart.layout.labels.dataLabels.color,
            onMouseenter: (s) => Vt(a, l),
            onMouseleave: (s) => Nt(a, l),
            onClick: (s) => Rt(a, l)
          }, D(a.name), 41, Ua))), 256))
        ])) : h("", !0),
        (n(!0), u(V, null, z(T.value, (a, l) => (n(), u("g", null, [
          $("g", null, [
            e.value.useCssAnimation || !e.value.useCssAnimation && !g.value.includes(l) ? (n(), u("polygon", {
              key: 0,
              points: r(nt)(P.value.map((s) => s.plots[l]), !1, !0),
              stroke: e.value.style.chart.backgroundColor,
              "stroke-width": e.value.style.chart.layout.dataPolygon.strokeWidth + 1,
              fill: "none",
              class: H({ "animated-out": g.value.includes(l) && e.value.useCssAnimation, "animated-in": M.value && Y.value === l && e.value.useCssAnimation })
            }, null, 10, Ba)) : h("", !0),
            e.value.useCssAnimation || !e.value.useCssAnimation && !g.value.includes(l) ? (n(), u("polygon", {
              key: 1,
              points: r(nt)(P.value.map((s) => s.plots[l]), !1, !0),
              stroke: a.color,
              "stroke-width": e.value.style.chart.layout.dataPolygon.strokeWidth,
              fill: e.value.style.chart.layout.dataPolygon.transparent ? "transparent" : e.value.style.chart.layout.dataPolygon.useGradient ? `url(#radar_gradient_${C.value}_${l})` : r(we)(a.color, e.value.style.chart.layout.dataPolygon.opacity),
              class: H({ "animated-out": g.value.includes(l) && e.value.useCssAnimation, "animated-in": M.value && Y.value === l && e.value.useCssAnimation })
            }, null, 10, Ea)) : h("", !0)
          ])
        ]))), 256)),
        e.value.style.chart.layout.plots.show ? (n(), u("g", Ma, [
          (n(!0), u(V, null, z(P.value, (a, l) => (n(), u("g", null, [
            (n(!0), u(V, null, z(a.plots, (s, f) => (n(), u("circle", {
              cx: s.x,
              cy: s.y,
              fill: g.value.includes(f) ? "transparent" : T.value[f] ? T.value[f].color : "transparent",
              r: te.value !== null && te.value === l ? e.value.style.chart.layout.plots.radius * 1.6 : e.value.style.chart.layout.plots.radius,
              stroke: g.value.includes(f) ? "transparent" : e.value.style.chart.backgroundColor,
              "stroke-width": 0.5,
              class: H({ "animated-out": g.value.includes(f) && e.value.useCssAnimation, "animated-in": M.value && Y.value === f && e.value.useCssAnimation })
            }, null, 10, Wa))), 256))
          ]))), 256))
        ])) : h("", !0),
        p(t.$slots, "svg", {
          svg: {
            ...c.value,
            isPrintingImg: r(ce) | r(ve) | K.value,
            isPrintingSvg: J.value
          }
        }, void 0, !0)
      ], 14, Sa)),
      t.$slots.watermark ? (n(), u("div", Ga, [
        p(t.$slots, "watermark", b(k({ isPrinting: r(ce) || r(ve) || K.value || J.value })), void 0, !0)
      ])) : h("", !0),
      $("div", {
        id: `legend-bottom-${C.value}`
      }, null, 8, Xa),
      Le.value ? (n(), I(Ht, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${C.value}` : `#legend-bottom-${C.value}`
      }, [
        $("div", {
          ref_key: "chartLegend",
          ref: _e
        }, [
          e.value.style.chart.legend.show ? (n(), I(wa, {
            key: `legend_${xe.value}`,
            legendSet: A.value,
            config: Lt.value,
            isCursorPointer: x.value,
            onClickMarker: o[0] || (o[0] = ({ i: a }) => Z(a))
          }, {
            item: v(({ legend: a, index: l }) => [
              r(re) ? h("", !0) : (n(), u("div", {
                key: 0,
                onClick: (s) => a.segregate(),
                style: q(`opacity:${g.value.includes(l) ? 0.5 : 1}`)
              }, D(a.display), 13, Ya))
            ]),
            legendToggle: v(() => [
              A.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !r(re) ? (n(), I(Ta, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: g.value.length > 0,
                isCursorPointer: x.value,
                onToggle: Pt
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : h("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : p(t.$slots, "legend", {
            key: 1,
            legend: A.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : h("", !0),
      t.$slots.source ? (n(), u("div", {
        key: 6,
        ref_key: "source",
        ref: Se,
        dir: "auto"
      }, [
        p(t.$slots, "source", {}, void 0, !0)
      ], 512)) : h("", !0),
      oe(r(ct), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: y.value.showTooltip && le.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        fontSize: e.value.style.chart.tooltip.fontSize,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: S.value,
        content: ne.value,
        isFullscreen: R.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": v(() => [
          p(t.$slots, "tooltip-before", b(k({ ...he.value })), void 0, !0)
        ]),
        "tooltip-after": v(() => [
          ["function"].includes(typeof e.value.style.chart.tooltip.customFormat) ? h("", !0) : (n(), u("div", ja, [
            oe(r(dt), {
              dataset: ge.value,
              config: Tt.value,
              backgroundOpacity: 0
            }, {
              "data-label": v(({ bar: a }) => [
                $("div", qa, [
                  $("span", Ha, D(a.name + (e.value.style.chart.tooltip.showValue || e.value.style.chart.tooltip.showPercentage ? ":" : "")), 1),
                  e.value.style.chart.tooltip.showValue || e.value.style.chart.tooltip.showPercentage ? (n(), u("span", Ka, D(a.suffix), 1)) : h("", !0)
                ])
              ]),
              _: 1
            }, 8, ["dataset", "config"])
          ])),
          p(t.$slots, "tooltip-after", b(k({ ...he.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "fontSize", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      Te.value && e.value.userOptions.buttons.table ? (n(), I(Kt(j.value.component), Jt({ key: 7 }, j.value.props, {
        ref_key: "tableUnit",
        ref: X,
        onClose: He
      }), tt({
        content: v(() => [
          (n(), I(r(gt), {
            key: `table_${Fe.value}`,
            colNames: ee.value.colNames,
            head: ee.value.head,
            body: ee.value.body,
            config: ee.value.config,
            title: e.value.table.useDialog ? "" : j.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: x.value,
            onClose: He
          }, {
            th: v(({ th: a }) => [
              be(D(a.name), 1)
            ]),
            td: v(({ td: a }) => [
              be(D(a), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: v(() => [
            be(D(j.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: v(() => [
            $("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: o[1] || (o[1] = (a) => fe(e.value.userOptions.callbacks.csv)),
              style: q({ cursor: x.value ? "pointer" : "default" })
            }, [
              oe(r(vt), {
                name: "fileCsv",
                stroke: j.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : h("", !0),
      p(t.$slots, "skeleton", {}, () => [
        r(re) ? (n(), I(pa, { key: 0 })) : h("", !0)
      ], !0)
    ], 46, Aa));
  }
}, yo = /* @__PURE__ */ Pa(Ja, [["__scopeId", "data-v-ed2dc0a6"]]);
export {
  yo as default
};
