// vue-data-ui/dist/vue-ui-stackbar-BZl0kEkb.js
import { defineAsyncComponent as le, useSlots as Ql, computed as g, ref as h, shallowRef as wt, onMounted as jl, toRefs as eo, watch as Te, onBeforeUnmount as Je, watchEffect as $e, nextTick as Lt, openBlock as s, createElementBlock as n, normalizeStyle as X, normalizeClass as ye, createBlock as j, unref as m, withCtx as w, renderSlot as k, normalizeProps as B, guardReactiveProps as U, createCommentVNode as y, createElementVNode as oe, createSlots as Ct, createVNode as Ae, Fragment as x, renderList as L, mergeProps as Pe, toDisplayString as H, Teleport as to, resolveDynamicComponent as ao, createTextVNode as Ma } from "vue";
import { u as lo } from "./useConfig-CIgyKZ6R.js";
import { j as _a, t as Ta, m as oo, p as Ze, n as ro, o as $a, q as Aa, r as so, s as no, aq as Ne, E as Pa, D as Na, I as io, X as uo, x as co, G as ce, W as Fa, k as fe, l as re, Z as vo, ar as ho, T as Ra, a0 as go, a1 as mo, U as yo } from "./lib-GwPnJSrY.js";
import { t as Ia, u as fo } from "./useResponsive-DfdjqQps.js";
import { u as Oa, a as Da, b as bo } from "./useTimeLabels-C-A2uZKi.js";
import { u as po } from "./usePrinter-DiK5R6k-.js";
import { u as xo, B as ko } from "./BaseScanner-BhhkKVoj.js";
import { u as wo } from "./useSvgExport-ByUukOZt.js";
import { u as St } from "./useNestedProp-BBT0mbHh.js";
import { u as Lo } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Co } from "./useUserOptionState-BIvW1Kz7.js";
import { u as So } from "./useStableElementSize-DCNs0Tx7.js";
import { u as zo } from "./useChartAccessibility-9icAAmYg.js";
import { u as Mo } from "./useTimeLabelCollider-CIsgDrl9.js";
import _o from "./img-BYYzY_QU.js";
import To from "./Title-BfyrP-jp.js";
import { _ as $o } from "./Shape-CSyWYgXc.js";
import Ao from "./vue_ui_stackbar-CuDC4O4b.js";
import Po from "./Legend-1LMMsPxL.js";
import { S as No } from "./SlicerPreview-I2J0UUF_.js";
import { B as Fo } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Ro } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Io = ["id"], Oo = ["id"], Do = ["xmlns", "viewBox"], Eo = ["x", "y", "width", "height"], Vo = { key: 1 }, Bo = ["id"], Uo = ["stop-color"], Ho = ["stop-color"], Wo = ["stop-color"], qo = ["x", "y", "width", "height", "stroke", "stroke-width", "stroke-linecap", "stroke-linejoin", "stroke-dasharray"], Yo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Xo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Go = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Jo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Zo = { key: 0 }, Ko = ["x", "y", "height", "rx", "width", "fill", "stroke", "stroke-width"], Qo = { key: 0 }, jo = ["x", "y", "height", "rx", "width", "fill", "stroke", "stroke-width"], er = ["x", "y", "width", "rx", "height", "fill", "stroke", "stroke-width"], tr = { key: 0 }, ar = ["x", "y", "width", "rx", "height", "fill", "stroke", "stroke-width"], lr = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], or = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], rr = ["x", "y", "font-size", "fill", "font-weight"], sr = ["transform", "font-size", "fill", "font-weight"], nr = ["x", "y", "font-size", "fill", "font-weight"], ir = ["x", "y", "font-size", "font-weight", "fill"], ur = ["x", "y", "font-size", "fill", "font-weight"], cr = ["x", "y", "font-size", "font-weight", "fill"], vr = ["x1", "x2", "y1", "y2", "stroke"], hr = ["x", "y", "font-size", "font-weight", "fill"], dr = ["x1", "x2", "y1", "y2", "stroke"], gr = ["font-size", "font-weight", "fill", "text-anchor", "transform"], mr = { key: 0 }, yr = { key: 1 }, fr = ["text-anchor", "font-size", "font-weight", "fill", "transform", "onClick"], br = ["text-anchor", "font-size", "fill", "transform", "innerHTML", "onClick"], pr = { key: 0 }, xr = { key: 1 }, kr = ["font-size", "font-weight", "fill", "x", "y", "onClick"], wr = ["font-size", "font-weight", "fill", "x", "y", "onClick", "innerHTML"], Lr = ["x", "y", "width", "height", "onClick", "onMouseenter", "onMouseleave", "fill"], Cr = ["x", "y", "width", "height", "onClick", "onMouseenter", "onMouseleave", "fill"], Sr = ["data-start", "data-end"], zr = {
  key: 3,
  class: "vue-data-ui-watermark"
}, Mr = ["x", "y", "height", "width", "fill", "stroke"], _r = ["id"], Tr = ["onClick"], $r = ["innerHTML"], Ar = {
  __name: "vue-ui-stackbar",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    selectedXIndex: {
      type: Number,
      default: void 0
    }
  },
  emits: ["selectDatapoint", "selectLegend", "selectTimeLabel", "selectX", "copyAlt"],
  setup(Ke, { expose: Ea, emit: Va }) {
    const Ba = le(() => import("./Tooltip-DL3QgjZf.js")), Ua = le(() => import("./BaseIcon-CNx8Tzl5.js")), Ha = le(() => import("./vue-ui-accordion-BGHTqIBt.js")), Wa = le(() => import("./DataTable-Bo1pmAqG.js")), qa = le(() => import("./PenAndPaper-DWV1FMIi.js")), Ya = le(() => import("./UserOptions-CrLqEimB.js")), Xa = le(() => import("./PackageVersion-DN9d0ME5.js")), Ga = le(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_stackbar: Ja } = lo(), { isThemeValid: Za, warnInvalidTheme: Ka } = Lo(), Qa = Ql(), P = Ke, ve = Va, be = g({
      get() {
        return !!P.dataset && P.dataset.length;
      },
      set(t) {
        return t;
      }
    }), Z = h(null), W = h(_a()), Qe = h(!1), je = h(null), et = h(""), _ = h([]), zt = h(0), Mt = h(null), _t = h(null), pe = h(null), Tt = h(null), xe = h(!1), K = h(null), se = h(!1), $t = h(0), ja = h(0), el = h(0), At = h(!1), ke = h(null), tt = h(null), Fe = h(!1), Re = h(!1), Ie = h(null), at = h(null), ne = h(null), ie = h(null), tl = h(null), lt = h(null), Pt = wt(null), Nt = h(!1), Ft = h(0), Rt = h(0), It = So({
      elementRef: Pt,
      minimumWidth: 2,
      minimumHeight: 2,
      stableFramesRequired: 2,
      once: !1,
      onSizeAccepted: () => {
        Dt();
      }
    });
    function ot() {
      Pt.value = Z.value?.parentNode ?? null;
    }
    function Ot() {
      return new Promise((t) => {
        requestAnimationFrame(() => {
          requestAnimationFrame(t);
        });
      });
    }
    async function Dt() {
      const t = ++Rt.value;
      Nt.value = !1, await Lt(), await Ot(), await Ot(), t === Rt.value && (Ft.value += 1, Nt.value = !0);
    }
    const rt = h(null);
    jl(() => {
      At.value = !0, ot(), It.start(), Yt();
    });
    const e = h(ut()), q = g(() => e.value.userOptions.useCursorPointer), al = g(() => Ta({
      defaultConfig: {
        userOptions: { show: !1 },
        useCssAnimation: !1,
        table: { show: !1 },
        tooltip: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            bars: {
              totalValues: { show: !1 },
              dataLabels: { show: !1 }
            },
            grid: {
              frame: {
                stroke: "#6A6A6A"
              },
              scale: {
                scaleMin: 0,
                scaleMax: 40
              },
              x: {
                axisColor: "#6A6A6A",
                linesColor: "#6A6A6A",
                axisName: { show: !1 },
                timeLabels: { show: !1 }
              },
              y: {
                axisColor: "#6A6A6A",
                linesColor: "#6A6A6A",
                axisName: { show: !1 },
                axisLabels: { show: !1 }
              }
            },
            legend: {
              backgroundColor: "transparent"
            },
            padding: {
              left: 24,
              right: 24,
              bottom: 12
            },
            zoom: {
              show: !1,
              startIndex: null,
              endIndex: null
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: he, FINAL_DATASET: st, manualLoading: Oe } = xo({
      ...eo(P),
      FINAL_CONFIG: e,
      prepareConfig: ut,
      callback: () => {
        Promise.resolve().then(async () => {
          await Ye();
        });
      },
      skeletonDataset: P.config?.skeletonDataset ?? [
        {
          name: "",
          series: [2, 3, 5, 8, 13, 21],
          color: "#BABABA"
        },
        {
          name: "",
          series: [1, 2, 3, 5, 8, 13],
          color: "#CACACA"
        }
      ],
      skeletonConfig: Ta({
        defaultConfig: e.value,
        userConfig: al.value
      })
    }), { userOptionsVisible: nt, setUserOptionsVisibility: Et, keepUserOptionState: Vt } = Co({ config: e.value }), { svgRef: it } = zo({ config: e.value.style.chart.title });
    function ll() {
      Et(!0);
    }
    function ol() {
      Et(!1), ve("selectX", { seriesIndex: null, datapoint: null }), K.value = null;
    }
    function ut() {
      const t = St({
        userConfig: P.config,
        defaultConfig: Ja
      });
      let l = {};
      const a = t.theme;
      if (a)
        if (!Za.value(t))
          Ka(t), l = t;
        else {
          const o = St({
            userConfig: Ao[a] || P.config,
            defaultConfig: t
          });
          l = {
            ...St({
              userConfig: P.config,
              defaultConfig: o
            }),
            customPalette: t.customPalette.length ? t.customPalette : oo[a] || Ze
          };
        }
      else
        l = t;
      return l;
    }
    const Bt = g(() => e.value.style.chart.bars.dataLabels.hideUnderValue != null), Ut = g(() => e.value.style.chart.bars.dataLabels.hideUnderPercentage != null);
    Te(() => P.config, (t) => {
      he.value || (e.value = ut()), nt.value = !e.value.userOptions.showOnChartHover, Yt(), $t.value += 1, ja.value += 1, el.value += 1, $.value.dataLabels.show = e.value.style.chart.bars.dataLabels.show, $.value.showTable = e.value.table.show, $.value.showTooltip = e.value.style.chart.tooltip.show, ot(), Ve();
    }, { deep: !0 }), Te(() => P.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (Oe.value = !1), ot(), Zt();
    }, { deep: !0 });
    const $ = h({
      dataLabels: {
        show: e.value.style.chart.bars.dataLabels.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    Te(e, () => {
      $.value = {
        dataLabels: {
          show: e.value.style.chart.bars.dataLabels.show
        },
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const { isPrinting: ct, isImaging: vt, generatePdf: Ht, generateImage: Wt } = po({
      elementId: `stackbar_${W.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-stackbar",
      options: e.value.userOptions.print
    }), ee = h({
      width: e.value.style.chart.width,
      height: e.value.style.chart.height,
      paddingRatio: {
        top: e.value.style.chart.padding.top / e.value.style.chart.height,
        right: e.value.style.chart.padding.right / e.value.style.chart.width,
        bottom: e.value.style.chart.padding.bottom / e.value.style.chart.height,
        left: e.value.style.chart.padding.left / e.value.style.chart.width
      }
    }), rl = g(() => ro(e.value.customPalette)), te = wt(null), de = wt(null), qt = h(null), ht = g(() => !!e.value.debug);
    function Yt() {
      if ($a(P.dataset) ? (Aa({
        componentName: "VueUiStackbar",
        type: "dataset",
        debug: ht.value
      }), Oe.value = !0) : P.dataset.forEach((t, l) => {
        so({
          datasetObject: t,
          requiredAttributes: ["name", "series"]
        }).forEach((a) => {
          be.value = !1, Aa({
            componentName: "VueUiStackbar",
            type: "datasetSerieAttribute",
            property: a,
            index: l,
            debug: ht.value
          }), Oe.value = !0;
        });
      }), $a(P.dataset) || (Oe.value = e.value.loading), setTimeout(() => {
        se.value = !0;
      }, 10), e.value.responsive) {
        const t = Ia(() => {
          se.value = !1;
          const { width: l, height: a } = fo({
            chart: Z.value,
            title: e.value.style.chart.title.text ? Mt.value : null,
            legend: e.value.style.chart.legend.show ? _t.value : null,
            slicer: e.value.style.chart.zoom.show && N.value > 6 ? pe.value.$el : null,
            source: Tt.value
          });
          requestAnimationFrame(() => {
            ee.value.width = l, ee.value.height = a - 12, clearTimeout(qt.value), qt.value = setTimeout(() => {
              se.value = !0;
            }, 10);
          });
        });
        te.value && (de.value && te.value.unobserve(de.value), te.value.disconnect()), te.value = new ResizeObserver(t), de.value = Z.value.parentNode, te.value.observe(de.value);
      }
      Dt(), Ye();
    }
    Je(() => {
      It.stop(), te.value && (de.value && te.value.unobserve(de.value), te.value.disconnect());
    });
    function sl() {
      let t = 0;
      e.value.orientation === "vertical" && ne.value && (t = Array.from(ne.value.querySelectorAll("text")).reduce((o, c) => {
        const r = c.getComputedTextLength();
        return r > o ? r : o;
      }, 0)), e.value.orientation === "horizontal" && ie.value && (t = Array.from(ie.value.querySelectorAll("text")).reduce((o, c) => {
        const r = c.getComputedTextLength();
        return r > o ? r : o;
      }, 0));
      const l = at.value ? at.value.getBoundingClientRect().width : 0;
      return t + l + (l ? 24 : 0);
    }
    const nl = h(0), De = h(0), il = Ia((t) => {
      De.value = t + e.value.style.chart.bars.totalValues.fontSize;
    }, 100);
    function ul() {
      if (e.value.orientation !== "horizontal") return 0;
      const t = lt.value;
      if (!t) return 0;
      const l = Array.from(t.querySelectorAll("text"));
      if (!l.length) return 0;
      let a = -1 / 0;
      for (const c of l)
        try {
          const r = c.getBBox(), b = r.x + r.width;
          b > a && (a = b);
        } catch {
        }
      return Math.max(0, a - (u.value?.right ?? 0));
    }
    $e((t) => {
      if (e.value.orientation !== "horizontal") return;
      const l = lt.value;
      if (!l) return;
      const a = () => {
        const r = ul();
        il(r);
      };
      a();
      const o = new ResizeObserver(a);
      o.observe(l);
      const c = new MutationObserver(a);
      c.observe(l, { childList: !0, subtree: !0, characterData: !0 }), t(() => {
        o.disconnect(), c.disconnect();
      });
    }), Je(() => {
      nl.value = 0, De.value = 0;
    });
    const Ee = h(0), we = h(0);
    function cl() {
      const t = e.value.orientation === "vertical" ? ie.value : ne.value;
      if (!t) {
        Ee.value = 0;
        return;
      }
      try {
        const l = t.getBBox();
        Ee.value = Number.isFinite(l?.height) ? l.height : 0;
      } catch {
        Ee.value = 0;
      }
    }
    function vl() {
      we.value && cancelAnimationFrame(we.value), we.value = requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          cl();
        });
      });
    }
    Je(() => {
      we.value && cancelAnimationFrame(we.value);
    });
    const Xt = g(() => {
      let t = 0;
      if (Ie.value)
        try {
          t = Ie.value.getBBox().height;
        } catch {
          t = 0;
        }
      const l = e.value.style.chart.grid.x.timeLabels.show ? Ee.value : 0;
      return t + l;
    }), u = g(() => {
      Ft.value;
      const { height: t, width: l } = ee.value, { right: a } = ee.value.paddingRatio;
      let o = e.value.style.chart.bars.totalValues.show && P.dataset && P.dataset.length > 1 ? e.value.style.chart.bars.totalValues.fontSize * 1.3 : 0, c = 0;
      e.value.style.chart.grid.y.axisLabels.show && (c = sl());
      const r = e.value.style.chart.padding.top + o, b = l - l * a - De.value, S = t - e.value.style.chart.padding.bottom - Xt.value - o, T = e.value.style.chart.padding.left + c, R = l - T - l * a - De.value, A = t - r - e.value.style.chart.padding.bottom - Xt.value - o;
      return {
        chartHeight: t,
        chartWidth: l,
        top: r,
        right: b,
        bottom: S,
        left: T,
        width: Math.max(0, R),
        height: Math.max(0, A)
      };
    }), hl = g(() => {
      const { left: t, top: l, width: a, height: o } = u.value, c = v.value.start, r = v.value.end, b = Math.max(1, r - c), S = Math.max(0, Math.min(b, (E.value.start ?? c) - c)), T = Math.max(0, Math.min(b, (E.value.end ?? r) - c)), R = Math.max(0, T - S), A = {
        fill: e.value.style.chart.zoom.preview.fill,
        stroke: e.value.style.chart.zoom.preview.stroke,
        "stroke-width": e.value.style.chart.zoom.preview.strokeWidth,
        "stroke-dasharray": e.value.style.chart.zoom.preview.strokeDasharray,
        "stroke-linecap": "round",
        "stroke-linejoin": "round",
        style: {
          pointerEvents: "none",
          transition: "none !important",
          animation: "none !important"
        }
      };
      if (e.value.orientation === "horizontal") {
        const O = o / b;
        return {
          x: t,
          y: l + S * O,
          width: a,
          height: R * O,
          ...A
        };
      } else {
        const O = a / b;
        return {
          x: t + S * O,
          y: l,
          width: R * O,
          height: o,
          ...A
        };
      }
    }), D = g(() => st.value.map((t, l) => {
      const a = no(t.color) || rl.value[l] || Ze[l] || Ze[l % Ze.length];
      return {
        ...t,
        // In distributed mode, all values are converted to positive
        series: JSON.parse(JSON.stringify(t.series)).map((o) => e.value.style.chart.bars.distributed ? Math.abs(o) : o),
        // Store signs to manage display of neg values in distributed mode
        signedSeries: t.series.map((o) => o >= 0 ? 1 : -1),
        absoluteIndex: l,
        id: _a(),
        color: a
      };
    })), N = g(() => {
      const t = Math.max(...D.value.filter((l) => !_.value.includes(l.id)).map((l) => l.series.length));
      return isFinite(t) ? t : Math.max(...D.value.map((l) => l.series.length));
    });
    function dl(t) {
      rt.value = t;
    }
    const v = h({
      start: 0,
      end: Math.max(...st.value.map((t) => t.series.length))
    }), E = h({ start: 0, end: Math.max(...st.value.map((t) => t.series.length)) }), Gt = g(() => e.value.style.chart.zoom.preview.enable && (E.value.start !== v.value.start || E.value.end !== v.value.end));
    function Jt(t, l) {
      E.value[t] = l;
    }
    function Ve() {
      const t = N.value;
      let l = Math.max(0, Math.min(v.value.start ?? 0, t - 1)), a = Math.max(l + 1, Math.min(v.value.end ?? t, t));
      (!Number.isFinite(l) || !Number.isFinite(a) || a <= l) && (l = 0, a = t), v.value.start = l, v.value.end = a, E.value.start = l, E.value.end = a, pe.value && (pe.value.setStartValue(l), pe.value.setEndValue(a));
    }
    const Le = h(null);
    function gl() {
      return new Promise((t) => requestAnimationFrame(
        () => requestAnimationFrame(() => t())
      ));
    }
    Je(() => {
      Le.value && cancelAnimationFrame(Le.value);
    });
    async function Zt() {
      Ye(), await Lt(), Le.value && cancelAnimationFrame(Le.value), Le.value = requestAnimationFrame(async () => {
        await gl(), Ye();
      });
    }
    const p = g(() => {
      let t;
      return e.value.orientation === "vertical" ? t = u.value.width / (v.value.end - v.value.start) : t = u.value.height / (v.value.end - v.value.start), t <= 0 ? 0 : t;
    }), Ce = g(() => Ne(D.value.filter((t) => !_.value.includes(t.id))).slice(v.value.start, v.value.end)), Kt = g(() => _.value.length === D.value.length), ue = g(() => e.value.style.chart.zoom.minimap.show ? Ne(D.value.map((t) => ({
      ...t,
      series: t.series.map((l) => l ?? 0)
    })).filter((t) => Kt.value ? !0 : !_.value.includes(t.id))) : []), ml = g(() => e.value.style.chart.zoom.minimap.show ? [{
      name: "",
      series: ue.value,
      color: "#000000",
      isVisible: !0
    }] : []), yl = g(() => Ne(D.value.filter((t) => !_.value.includes(t.id)).map((t) => ({
      ...t,
      series: t.series.map((l, a) => t.signedSeries[a] === -1 && l >= 0 ? -l : l)
    }))).slice(v.value.start, v.value.end)), Be = g(() => {
      const t = D.value.filter((l) => !_.value.includes(l.id));
      return {
        positive: Ne(t.map((l) => ({
          ...l,
          series: l.series.slice(v.value.start, v.value.end).map((a) => a >= 0 ? a : 0)
        }))),
        negative: Ne(t.map((l) => ({
          ...l,
          series: l.series.slice(v.value.start, v.value.end).map((a) => a < 0 ? a : 0)
        })))
      };
    }), Y = g(() => {
      const t = e.value.style.chart.grid.scale.scaleMax !== null && !e.value.style.chart.bars.distributed ? e.value.style.chart.grid.scale.scaleMax : Math.max(...Be.value.positive), l = Math.min(...Be.value.negative), a = e.value.style.chart.grid.scale.scaleMin !== null && !e.value.style.chart.bars.distributed ? e.value.style.chart.grid.scale.scaleMin : [-1 / 0, 1 / 0, NaN, void 0, null].includes(l) ? 0 : l, o = !e.value.style.chart.bars.distributed && (e.value.style.chart.grid.scale.scaleMax !== null || e.value.style.chart.grid.scale.scaleMin !== null) ? Pa(a > 0 ? 0 : a, t < 0 ? 0 : t, e.value.style.chart.grid.scale.ticks) : Na(a > 0 ? 0 : a, t < 0 ? 0 : t, e.value.style.chart.grid.scale.ticks);
      return o.ticks.map((c) => ({
        zero: u.value.bottom - u.value.height * (Math.abs(o.min) / (o.max + Math.abs(o.min))),
        y: u.value.bottom - u.value.height * ((c + Math.abs(o.min)) / (o.max + Math.abs(o.min))),
        x: u.value.left - 8,
        horizontal_zero: u.value.left + u.value.width * (Math.abs(o.min) / (o.max + Math.abs(o.min))),
        horizontal_x: u.value.left + u.value.width * ((c + Math.abs(o.min)) / (o.max + Math.abs(o.min))),
        horizontal_y: u.value.bottom - 8,
        value: c
      }));
    }), F = h([]), ge = h([]);
    let Qt = 0;
    $e(() => {
      const t = ++Qt;
      (async () => {
        const l = await Oa({
          values: e.value.style.chart.grid.x.timeLabels.values,
          maxDatapoints: N.value,
          formatter: e.value.style.chart.grid.x.timeLabels.datetimeFormatter,
          start: v.value.start,
          end: v.value.end
        });
        t === Qt && (F.value = l);
      })();
    });
    let jt = 0;
    $e(() => {
      const t = ++jt;
      (async () => {
        const l = await Oa({
          values: e.value.style.chart.grid.x.timeLabels.values,
          maxDatapoints: N.value,
          formatter: e.value.style.chart.grid.x.timeLabels.datetimeFormatter,
          start: 0,
          end: N.value
        });
        t === jt && (ge.value = l);
      })();
    });
    const fl = g(() => {
      const t = e.value.style.chart.grid.x.timeLabels.modulo;
      return F.value.length ? Math.min(t, [...new Set(F.value.map((l) => l.text))].length) : t;
    }), dt = g(() => {
      const t = e.value.style.chart.grid.x.timeLabels, l = F.value || [], a = ge.value || [], o = v.value.start ?? 0, c = K.value, r = N.value, b = l.map((T) => T?.text ?? ""), S = a.map((T) => T?.text ?? "");
      return io(
        !!t.showOnlyFirstAndLast,
        !!t.showOnlyAtModulo,
        Math.max(1, fl.value || 1),
        b,
        S,
        o,
        c,
        r
      );
    });
    $e(() => {
      e.value.orientation, e.value.style.chart.grid.x.timeLabels.show, e.value.style.chart.grid.x.timeLabels.fontSize, e.value.style.chart.grid.x.timeLabels.rotation, e.value.style.chart.grid.x.timeLabels.offsetY, dt.value?.map((t) => t?.text ?? "").join("|"), Y.value?.map((t) => t?.value ?? "").join("|"), ee.value.width, ee.value.height, ie.value, ne.value, Ie.value, vl();
    }, { flush: "post" });
    const ea = h({ months: [], shortMonths: [], days: [], shortDays: [] });
    let ta = 0;
    $e(() => {
      const t = ++ta, l = e.value.style.chart.grid.x.timeLabels.datetimeFormatter;
      (async () => {
        const a = await Da(l.locale).catch(() => Da("en"));
        t === ta && (ea.value = a.data);
      })();
    });
    const aa = g(() => {
      const t = e.value.style.chart.grid.x.timeLabels.datetimeFormatter, l = bo({
        useUTC: t.useUTC,
        locale: ea.value,
        januaryAsYear: t.januaryAsYear
      });
      return (a, o) => {
        const r = e.value.style.chart.grid.x.timeLabels.values?.[a];
        return r == null ? "" : l.formatDate(new Date(r), o);
      };
    }), la = g(() => (e.value.style.chart.grid.x.timeLabels.values || []).map((l, a) => ({
      text: aa.value(a, e.value.style.chart.zoom.timeFormat),
      absoluteIndex: a
    }))), gt = g(() => (e.value.style.chart.grid.x.timeLabels.values || []).map((l, a) => ({
      text: aa.value(a, e.value.style.chart.tooltip.timeFormat),
      absoluteIndex: a
    }))), C = g(() => {
      if (!be.value && !he.value) return [];
      let t = Array(N.value).fill(0), l = Array(N.value).fill(0), a = Array(N.value).fill(0), o = Array(N.value).fill(0);
      const c = Math.max(...Be.value.positive) || 0, r = Math.min(...Be.value.negative), b = [-1 / 0, 1 / 0, NaN, void 0, null].includes(r) ? 0 : r, S = !e.value.style.chart.bars.distributed && (e.value.style.chart.grid.scale.scaleMax !== null || e.value.style.chart.grid.scale.scaleMin !== null) ? Pa(e.value.style.chart.grid.scale.scaleMin !== null ? e.value.style.chart.grid.scale.scaleMin : b > 0 ? 0 : b, e.value.style.chart.grid.scale.scaleMax !== null ? e.value.style.chart.grid.scale.scaleMax : c < 0 ? 0 : c, e.value.style.chart.grid.scale.ticks) : Na(e.value.style.chart.grid.scale.scaleMin !== null ? e.value.style.chart.grid.scale.scaleMin : b > 0 ? 0 : b, e.value.style.chart.grid.scale.scaleMax !== null ? e.value.style.chart.grid.scale.scaleMax : c < 0 ? 0 : c, e.value.style.chart.grid.scale.ticks), { min: T, max: R } = S, A = R + (T >= 0 ? 0 : Math.abs(T)) || 1, O = u.value.height, V = u.value.width, _e = Y.value[0] ? Y.value[0].zero : u.value.bottom, G = Y.value[0] ? Y.value[0].horizontal_zero : u.value.left;
      let wa = Array(ue.value.length).fill(0), La = Array(ue.value.length).fill(0);
      return D.value.filter((me) => !_.value.includes(me.id)).map((me) => {
        const ae = me.series.slice(), Ca = ae.map((d, f) => wa[f]), El = ae.map((d, f) => La[f]);
        ae.forEach((d, f) => {
          (Number(d) || 0) >= 0 ? wa[f] += Number(d) || 0 : La[f] += Math.abs(Number(d) || 0);
        });
        const Q = me.series.slice(v.value.start, v.value.end), Vl = me.signedSeries.slice(v.value.start, v.value.end), Bl = Q.map((d, f) => u.value.left + p.value * f + p.value * e.value.style.chart.bars.gapRatio / 4), Ge = (ue.value || []).filter(Number.isFinite), pt = Ge.length ? Math.max(...Ge) : 0, xt = Ge.length ? Math.min(...Ge) : 0;
        function Sa({ minimapH: d }) {
          const z = pt > 0, I = xt < 0;
          if (z && I) {
            const J = Math.max(pt, Math.abs(xt)) || 1e-9;
            return {
              pxPerUnit: d / 2 / J,
              zero: d / 2
            };
          }
          if (z) {
            const J = Math.max(1e-9, pt);
            return {
              pxPerUnit: d / J,
              zero: d
            };
          }
          const M = Math.max(1e-9, Math.abs(xt));
          return {
            pxPerUnit: d / M,
            zero: 0
          };
        }
        const Ul = ({ left: d, unitW: f }) => {
          const z = f * (e.value.style.chart.bars.gapRatio / 4);
          return ae.map((I, M) => d + f * M + z);
        }, Hl = Q.map((d, f) => u.value.top + p.value * f + p.value * e.value.style.chart.bars.gapRatio / 4), Wl = Q.map((d, f) => {
          const z = e.value.style.chart.bars.distributed ? (d || 0) / Ce.value[f] : (d || 0) / A;
          let I, M;
          return d > 0 ? (M = O * z, I = _e - M - t[f], t[f] += M) : (M = O * z, I = _e + a[f], a[f] += Math.abs(M)), I;
        }), ql = ({ minimapH: d }) => {
          if (e.value.style.chart.bars.distributed)
            return ae.map((M, J) => {
              const kt = Math.abs(Number(M) || 0), za = Math.abs(ue.value?.[J] || 0) || 1e-9, Kl = (Ca[J] + kt) / za;
              return d - Kl * d;
            });
          const { pxPerUnit: f, zero: z } = Sa({ minimapH: d });
          return ae.map((I, M) => {
            const J = Number(I) || 0;
            return J >= 0 ? z - (Ca[M] + J) * f : z + El[M] * f;
          });
        }, Yl = Q.map((d, f) => {
          const z = e.value.style.chart.bars.distributed ? (d || 0) / Ce.value[f] : (d || 0) / A;
          let I, M;
          return d > 0 ? (M = V * z, I = G + l[f], l[f] += M) : (M = V * z, I = G - Math.abs(M) - o[f], o[f] += Math.abs(M)), I;
        }), Xl = Q.map((d, f) => {
          const z = e.value.style.chart.bars.distributed ? (d || 0) / Ce.value[f] : (d || 0) / A;
          return d > 0 ? O * z : O * Math.abs(z);
        }), Gl = ({ minimapH: d }) => {
          if (e.value.style.chart.bars.distributed)
            return ae.map((I, M) => {
              const J = Math.abs(Number(I) || 0), kt = Math.abs(ue.value?.[M] || 0) || 1e-9;
              return J / kt * d;
            });
          const { pxPerUnit: f } = Sa({ minimapH: d });
          return ae.map((z) => Math.abs(Number(z) || 0) * f);
        }, Jl = Q.map((d, f) => {
          const z = e.value.style.chart.bars.distributed ? (d || 0) / Ce.value[f] : (d || 0) / A;
          return d > 0 ? V * z : V * Math.abs(z);
        }), Zl = Q.map((d) => Math.abs(d)).reduce((d, f) => d + f, 0);
        return {
          ...me,
          proportions: Q.map((d, f) => e.value.style.chart.bars.distributed ? (d || 0) / Ce.value[f] : (d || 0) / Zl),
          series: Q,
          signedSeries: Vl,
          x: Bl,
          y: Wl,
          height: Xl,
          horizontal_width: Jl,
          horizontal_y: Hl,
          horizontal_x: Yl,
          xMinimap: Ul,
          yMinimap: ql,
          heightMinimap: Gl
        };
      });
    }), oa = g(() => yl.value.map((t, l) => ({
      value: t,
      sign: t >= 0 ? 1 : -1
    })));
    function Ue(t, l, a, o, c) {
      const r = c === -1 && t >= 0 ? -t : t;
      return fe(
        e.value.style.chart.bars.dataLabels.formatter,
        r,
        re({
          p: e.value.style.chart.bars.dataLabels.prefix,
          v: r,
          s: e.value.style.chart.bars.dataLabels.suffix,
          r: e.value.style.chart.bars.dataLabels.rounding
        }),
        { datapoint: l, seriesIndex: a, datapointIndex: o }
      );
    }
    function ra(t, l, a, o) {
      return fe(
        e.value.style.chart.bars.dataLabels.formatter,
        t,
        re({
          v: isNaN(t) ? 0 : t,
          s: "%",
          r: e.value.style.chart.bars.dataLabels.rounding
        }),
        { datapoint: l, seriesIndex: a, datapointIndex: o }
      );
    }
    function sa(t) {
      const l = JSON.parse(JSON.stringify(C.value)).map((a) => ({
        name: a.name,
        value: a.series[t] === 0 ? 0 : a.series[t] || null,
        proportion: a.proportions[t] || null,
        color: a.color,
        id: a.id
      }));
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: t + v.value.start }), ve("selectDatapoint", { datapoint: l, period: F.value[t] });
    }
    function mt(t) {
      return JSON.parse(JSON.stringify(C.value)).map((l) => ({
        name: l.name,
        absoluteIndex: l.absoluteIndex,
        value: l.series[t] === 0 ? 0 : (l.signedSeries[t] === -1 && l.series[t] >= 0 ? -l.series[t] : l.series[t]) || null,
        proportion: l.proportions[t] || null,
        color: l.color,
        id: l.id,
        timeLabel: ge.value[t]
      }));
    }
    function na(t) {
      if (e.value.events.datapointLeave) {
        const l = mt(t);
        e.value.events.datapointLeave({ datapoint: l, seriesIndex: t + v.value.start });
      }
      Qe.value = null, K.value = null;
    }
    const bl = g(() => Array(N.value).fill(0).map((t, l) => mt(l)));
    Te(() => P.selectedXIndex, (t) => {
      if ([null, void 0].includes(P.selectedXIndex)) {
        K.value = null;
        return;
      }
      const l = t - v.value.start;
      l < 0 || t >= v.value.end ? K.value = null : K.value = l ?? null;
    }, { immediate: !0 });
    function ia(t) {
      if (Kt.value) return;
      K.value = t, Qe.value = !0;
      const l = e.value.style.chart.tooltip.customFormat, a = mt(t);
      Il({ seriesIndex: t, datapoint: a }), e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: a, seriesIndex: t + v.value.start }), je.value = {
        datapoint: a,
        seriesIndex: t,
        config: e.value,
        series: C.value
      };
      const o = a.map((r) => Math.abs(r.value)).reduce((r, b) => r + b, 0), c = a.map((r) => ce(r.value)).reduce((r, b) => r + b, 0);
      if (Ra(l) && yo(() => l({
        seriesIndex: t,
        datapoint: a,
        series: C.value,
        config: e.value
      })))
        et.value = l({
          seriesIndex: t,
          datapoint: a,
          series: C.value,
          config: e.value
        });
      else {
        const {
          showValue: r,
          showTotal: b,
          totalTranslation: S,
          showPercentage: T,
          borderColor: R,
          roundingValue: A,
          roundingPercentage: O
        } = e.value.style.chart.tooltip;
        let V = "";
        (F.value[t] && F.value[t].text || gt.value[t] && gt.value[t].text && e.value.style.chart.tooltip.showTimeLabel) && (V += `<div style="width:100%;text-align:center;border-bottom:1px solid ${R};padding-bottom:6px;margin-bottom:3px;">${e.value.style.chart.tooltip.useDefaultTimeFormat ? F.value[t]?.text : gt.value[t]?.text || ge.value[t]?.text || ""}</div>`), b && (V += `<div class="vue-data-ui-tooltip-total" style="display:flex;flex-direction:row;align-items:center;gap:4px">
                <span>${S}:</span>
                <span>
                    ${fe(
          e.value.style.chart.bars.dataLabels.formatter,
          c,
          re({
            p: e.value.style.chart.bars.dataLabels.prefix,
            v: c,
            s: e.value.style.chart.bars.dataLabels.suffix,
            r: A
          }),
          { datapoint: { name: S, value: c } }
        )}
                </span>
            </div>`);
        const _e = [
          r && T ? "(" : "",
          r && T ? ")" : ""
        ];
        a.reverse().forEach((G) => {
          V += `
                <div style="display:flex;flex-direction:row;align-items:center;gap:4px">
                    <svg viewBox="0 0 60 60" height="14" width="14"><rect rx="5" x="0" y="0" height="60" width="60" stroke="none" fill="${e.value.style.chart.bars.gradient.show ? `url(#gradient_${G.id})` : G.color}"/>${Qa.pattern ? `<rect rx="5" x="0" y="0" height="60" width="60" stroke="none" fill="url(#pattern_${W.value}_${G.absoluteIndex})"/>` : ""}</svg>
                    ${G.name}${r || T ? ":" : ""} ${r ? fe(
            e.value.style.chart.bars.dataLabels.formatter,
            G.value,
            re({
              p: e.value.style.chart.bars.dataLabels.prefix,
              v: G.value,
              s: e.value.style.chart.bars.dataLabels.suffix,
              r: A
            })
          ) : ""} ${_e[0]}${T ? re({
            v: isNaN(G.value / o) ? 0 : Math.abs(G.value) / o * 100,
            // Negs are absed to show relative proportion to absolute total. It's opinionated.
            s: "%",
            r: O
          }) : ""}${_e[1]}
                </div>
            `;
        }), et.value = `<div>${V}</div>`;
      }
    }
    const pl = g(() => ee.value.width), xl = g(() => ee.value.height);
    Mo({
      timeLabelsEls: e.value.orientation === "vertical" ? ie : ne,
      timeLabels: F,
      slicer: v,
      configRef: e,
      rotationPath: ["style", "chart", "grid", "x", "timeLabels", "rotation"],
      autoRotatePath: ["style", "chart", "grid", "x", "timeLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: pl,
      height: xl,
      rotation: e.value.style.chart.grid.x.timeLabels.autoRotate.angle
    });
    function ua(t) {
      xe.value = t, zt.value += 1;
    }
    function ca() {
      $.value.showTable = !$.value.showTable;
    }
    function va() {
      $.value.dataLabels.show = !$.value.dataLabels.show;
    }
    function ha() {
      $.value.showTooltip = !$.value.showTooltip;
    }
    function kl() {
      return C.value;
    }
    const da = g(() => {
      if (C.value.length === 0) return { head: [], body: [], config: {}, columnNames: [] };
      const t = C.value.map(({ name: a, color: o }) => ({
        label: a,
        color: o
      })), l = [];
      return F.value.forEach((a) => {
        const o = [e.value.style.chart.grid.x.timeLabels.values[a.absoluteIndex] ? a.text : i + 1];
        D.value.forEach((c) => {
          o.push(Number((c.series[a.absoluteIndex] || 0).toFixed(e.value.table.td.roundingValue)));
        }), l.push(o);
      }), { head: t, body: l };
    });
    function yt(t = null) {
      const l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], a = ["", ...da.value.head.map((b) => b.label)], o = da.value.body, c = l.concat([a]).concat(o), r = go(c);
      t ? t(r) : mo({ csvContent: r, title: e.value.style.chart.title.text || "vue-ui-stackbar" });
    }
    const He = g(() => {
      const t = [""].concat(C.value.map((c) => c.name)).concat(' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>');
      let l = [];
      for (let c = 0; c < N.value; c += 1) {
        const r = C.value.map((b) => b.series[c] ?? 0).reduce((b, S) => b + S, 0);
        l.push([e.value.style.chart.grid.x.timeLabels.values.slice(v.value.start, v.value.end)[c] ? F.value[c].text : c + 1].concat(C.value.map((b) => (b.series[c] ?? 0).toFixed(e.value.table.td.roundingValue))).concat((r ?? 0).toFixed(e.value.table.td.roundingValue)));
      }
      const a = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, o = [e.value.table.columnNames.period].concat(C.value.map((c) => c.name)).concat(e.value.table.columnNames.total);
      return { head: t, body: l.slice(0, v.value.end - v.value.start), config: a, colNames: o };
    });
    function wl() {
      _.value.length ? _.value = [] : Se.value.forEach((t) => {
        _.value.push(t.id);
      });
    }
    function ft(t) {
      if (_.value.includes(t.id))
        _.value = _.value.filter((l) => l !== t.id);
      else {
        if (_.value.length === D.value.length - 1) return;
        _.value.push(t.id);
      }
      ve("selectLegend", C.value);
    }
    function ga(t) {
      if (!D.value.length)
        return e.value.debug && console.warn("VueUiStackbar - There are no series to show."), null;
      const l = D.value.find((a) => a.name === t);
      return l || (e.value.debug && console.warn(`VueUiStackbar - Series name not found "${t}"`), null);
    }
    function Ll(t) {
      const l = ga(t);
      l !== null && _.value.includes(l.id) && ft({ id: l.id });
    }
    function Cl(t) {
      const l = ga(t);
      l !== null && (_.value.includes(l.id) || ft({ id: l.id }));
    }
    const Se = g(() => D.value.map((t, l) => ({
      ...t,
      shape: "square"
    })).map((t) => ({
      ...t,
      opacity: _.value.includes(t.id) ? 0.5 : 1,
      segregate: () => ft(t),
      isSegregated: _.value.includes(t.id)
    }))), Sl = g(() => ({
      cy: "stackbar-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), We = h(!1);
    function bt() {
      We.value = !We.value;
    }
    function qe(t, l) {
      const a = JSON.parse(JSON.stringify(C.value)).map((o) => ({
        name: o.name,
        value: o.series[l] === 0 ? 0 : (o.signedSeries[l] === -1 && o.series[l] >= 0 ? -o.series[l] : o.series[l]) || null,
        proportion: o.proportions[l] || null,
        color: o.color,
        id: o.id
      }));
      ve("selectTimeLabel", {
        datapoint: a,
        absoluteIndex: t.absoluteIndex,
        label: t.text
      });
    }
    const zl = g(() => Math.max(...C.value.flatMap((t) => t.series)));
    function ma(t, l) {
      return e.value.style.chart.bars.showDistributedPercentage && e.value.style.chart.bars.distributed ? Ut.value ? l * 100 >= e.value.style.chart.bars.dataLabels.hideUnderPercentage : e.value.style.chart.bars.dataLabels.hideEmptyPercentages ? l > 0 : !0 : Ut.value ? (Bt.value && ht.value && console.warn("Vue Data UI - VueUiStackbar - You cannot set both dataLabels.hideUnderPercentage and dataLabels.hideUnderValue. Note that dataLabels.hideUnderPercentage takes precedence in this case."), t > zl.value * e.value.style.chart.bars.dataLabels.hideUnderPercentage / 100) : Bt.value ? Math.abs(t) >= e.value.style.chart.bars.dataLabels.hideUnderValue : e.value.style.chart.bars.dataLabels.hideEmptyValues ? t !== 0 : !0;
    }
    async function Ml({ scale: t = 2 } = {}) {
      if (!Z.value) return;
      const { width: l, height: a } = Z.value.getBoundingClientRect(), o = l / a, { imageUri: c, base64: r } = await _o({ domElement: Z.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: c,
        base64: r,
        title: e.value.style.chart.title.text,
        width: l,
        height: a,
        aspectRatio: o
      };
    }
    const ze = g(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = $.value.showTable;
      return {
        component: t ? Ga : Ha,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: xe.value,
          fullscreenParent: Z.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: q.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    Te(() => $.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && ke.value ? ke.value.open() : "close" in ke.value && ke.value.close());
    });
    function ya() {
      $.value.showTable = !1, tt.value && tt.value.setTableIconState(!1);
    }
    const _l = g(() => e.value.style.chart.backgroundColor), Tl = g(() => e.value.style.chart.legend), $l = g(() => e.value.style.chart.title), { exportSvg: Al, getSvg: Pl } = wo({
      svg: it,
      title: $l,
      legend: Tl,
      legendItems: Se,
      backgroundColor: _l
    });
    async function fa({ isCb: t }) {
      Re.value = !0, await Lt();
      try {
        if (t) {
          const { blob: l, url: a, text: o, dataUrl: c } = await Pl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: a, text: o, dataUrl: c }));
        } else
          await Promise.resolve(Al());
      } finally {
        Re.value = !1;
      }
    }
    function Nl(t) {
      if (t?.stage === "start") {
        Fe.value = !0;
        return;
      }
      if (t?.stage === "end") {
        Fe.value = !1;
        return;
      }
      Wt();
    }
    function ba(t) {
      const l = N.value;
      return t > l ? l : t < 0 || t < v.value.start ? e.value.style.chart.zoom.startIndex !== null ? v.value.start + 1 : 1 : t;
    }
    const Me = h(!1), pa = h(!1);
    function Ye() {
      if (!Me.value) {
        Me.value = !0;
        try {
          const { startIndex: t, endIndex: l } = e.value.style.chart.zoom, a = N.value, o = t ?? 0, c = l != null ? Math.min(ba(l + 1), a) : a;
          Xe.value = !0, v.value.start = o, v.value.end = c, E.value.start = o, E.value.end = c, Ve(), pa.value = !0;
        } finally {
          queueMicrotask(() => {
            Xe.value = !1;
          }), Me.value = !1;
        }
      }
    }
    const Xe = h(!1);
    function Fl(t) {
      Me.value || Xe.value || t !== v.value.start && (v.value.start = t, E.value.start = t, Ve());
    }
    function Rl(t) {
      if (Me.value || Xe.value) return;
      const l = ba(t);
      l !== v.value.end && (v.value.end = l, E.value.end = l, Ve());
    }
    function Il({ seriesIndex: t, datapoint: l }) {
      const a = v.value.start + t;
      ve("selectX", {
        dataset: l,
        index: a,
        indexLabel: e.value.style.chart.grid.x.timeLabels.values[a]
      });
    }
    function xa() {
      return { y0: Y.value?.[0]?.zero ?? u.value.bottom };
    }
    function Ol(t) {
      const { y0: l } = xa(), a = e.value.style.chart.bars.totalValues, o = Math.max(2, a.fontSize * 0.3 + a.offsetY);
      let c = 1 / 0, r = !1;
      for (const R of C.value || []) {
        const A = R?.series?.[t] ?? 0, O = R?.height?.[t] ?? 0, V = R?.y?.[t];
        A > 0 && O > 0 && Number.isFinite(V) && (r = !0, V < c && (c = V));
      }
      const S = (r && Number.isFinite(c) ? c : l) - o;
      return Math.min(
        Math.max(S, 0),
        u.value.bottom
      );
    }
    function Dl(t) {
      const { x0: l } = xa(), a = Math.max(2, e.value.style.chart.bars.totalValues.fontSize * 0.3 + e.value.style.chart.bars.totalValues.offsetX);
      let o = -1 / 0, c = !1;
      for (const b of C.value || []) {
        const S = b?.series?.[t] ?? 0, T = b?.horizontal_x?.[t], R = b?.horizontal_width?.[t], A = Number.isFinite(R) ? Math.max(0, R) : 0;
        Number.isFinite(T) && S > 0 && A > 0 && (c = !0, o = Math.max(o, T + A));
      }
      return (c && Number.isFinite(o) ? o : l) + a;
    }
    async function ka() {
      if (ve("copyAlt", {
        config: e.value,
        dataset: C.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: C.value
      }));
    }
    return Ea({
      getData: kl,
      getImage: Ml,
      generatePdf: Ht,
      generateCsv: yt,
      generateImage: Wt,
      generateSvg: fa,
      hideSeries: Cl,
      showSeries: Ll,
      toggleTable: ca,
      toggleLabels: va,
      toggleTooltip: ha,
      toggleAnnotator: bt,
      toggleFullscreen: ua,
      copyAlt: ka
    }), (t, l) => (s(), n("div", {
      id: `stackbar_${W.value}`,
      ref_key: "stackbarChart",
      ref: Z,
      class: ye({ "vue-data-ui-component": !0, "vue-ui-stackbar": !0, "vue-data-ui-wrapper-fullscreen": xe.value }),
      style: X(`background:${e.value.style.chart.backgroundColor};color:${e.value.style.chart.color};font-family:${e.value.style.fontFamily}; position: relative; ${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: ll,
      onMouseleave: ol
    }, [
      e.value.userOptions.buttons.annotator ? (s(), j(m(qa), {
        key: 0,
        svgRef: m(it),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: We.value,
        isCursorPointer: q.value,
        onClose: bt
      }, {
        "annotator-action-close": w(() => [
          k(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": w(({ color: a }) => [
          k(t.$slots, "annotator-action-color", B(U({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": w(({ mode: a }) => [
          k(t.$slots, "annotator-action-draw", B(U({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": w(({ disabled: a }) => [
          k(t.$slots, "annotator-action-undo", B(U({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": w(({ disabled: a }) => [
          k(t.$slots, "annotator-action-redo", B(U({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": w(({ disabled: a }) => [
          k(t.$slots, "annotator-action-delete", B(U({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : y("", !0),
      k(t.$slots, "userConfig", {}, void 0, !0),
      e.value.style.chart.title.text ? (s(), n("div", {
        key: 1,
        ref_key: "chartTitle",
        ref: Mt,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (s(), j(To, {
          key: `title_${$t.value}`,
          config: {
            title: {
              cy: "stackbar-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "stackbar-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : y("", !0),
      oe("div", {
        id: `legend-top-${W.value}`
      }, null, 8, Oo),
      e.value.userOptions.show && be.value && (m(Vt) || m(nt)) ? (s(), j(m(Ya), {
        ref_key: "userOptionsRef",
        ref: tt,
        key: `user_option_${zt.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: m(ct),
        isImaging: m(vt),
        uid: W.value,
        hasTooltip: e.value.style.chart.tooltip.show && e.value.userOptions.buttons.tooltip,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: xe.value,
        chartElement: Z.value,
        position: e.value.userOptions.position,
        isTooltip: $.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: We.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: q.value,
        onToggleFullscreen: ua,
        onGeneratePdf: m(Ht),
        onGenerateCsv: yt,
        onGenerateImage: Nl,
        onGenerateSvg: fa,
        onToggleTable: ca,
        onToggleLabels: va,
        onToggleTooltip: ha,
        onToggleAnnotator: bt,
        onCopyAlt: ka,
        style: X({
          visibility: m(Vt) ? m(nt) ? "visible" : "hidden" : "visible"
        })
      }, Ct({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: w(({ isOpen: a, color: o }) => [
            k(t.$slots, "menuIcon", B(U({ isOpen: a, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: w(() => [
            k(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: w(() => [
            k(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: w(() => [
            k(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: w(() => [
            k(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: w(() => [
            k(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: w(() => [
            k(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: w(() => [
            k(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: w(({ toggleFullscreen: a, isFullscreen: o }) => [
            k(t.$slots, "optionFullscreen", B(U({ toggleFullscreen: a, isFullscreen: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: w(({ toggleAnnotator: a, isAnnotator: o }) => [
            k(t.$slots, "optionAnnotator", B(U({ toggleAnnotator: a, isAnnotator: o })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: w(({ altCopy: a }) => [
            k(t.$slots, "optionAltCopy", B(U({ altCopy: a })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "isTooltip", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : y("", !0),
      (s(), n("svg", {
        ref_key: "svgRef",
        ref: it,
        xmlns: m(uo),
        viewBox: `0 0 ${u.value.chartWidth <= 0 ? 10 : u.value.chartWidth} ${u.value.chartHeight <= 0 ? 10 : u.value.chartHeight}`,
        class: ye({ "vue-data-ui-loading": m(he) }),
        style: X(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        Ae(m(Xa)),
        t.$slots["chart-background"] ? (s(), n("foreignObject", {
          key: 0,
          x: u.value.left,
          y: u.value.top,
          width: u.value.width <= 0 ? 10 : u.value.width,
          height: u.value.height <= 0 ? 10 : u.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          k(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Eo)) : y("", !0),
        e.value.style.chart.bars.gradient.show ? (s(), n("defs", Vo, [
          (s(!0), n(x, null, L(C.value, (a, o) => (s(), n("linearGradient", {
            id: `gradient_${a.id}`,
            x1: "0%",
            y1: "0%",
            x2: "0%",
            y2: "100%"
          }, [
            oe("stop", {
              offset: "0%",
              "stop-color": a.color
            }, null, 8, Uo),
            oe("stop", {
              offset: "61.8%",
              "stop-color": m(co)(a.color, e.value.style.chart.bars.gradient.intensity / 100)
            }, null, 8, Ho),
            oe("stop", {
              offset: "100%",
              "stop-color": a.color
            }, null, 8, Wo)
          ], 8, Bo))), 256))
        ])) : y("", !0),
        e.value.style.chart.grid.frame.show ? (s(), n("rect", {
          key: 2,
          style: { pointerEvents: "none", transition: "none", animation: "none !important" },
          x: Math.max(0, u.value.left),
          y: Math.max(0, u.value.top),
          width: Math.max(0, u.value.width),
          height: Math.max(0, u.value.height),
          fill: "transparent",
          stroke: e.value.style.chart.grid.frame.stroke,
          "stroke-width": e.value.style.chart.grid.frame.strokeWidth,
          "stroke-linecap": e.value.style.chart.grid.frame.strokeLinecap,
          "stroke-linejoin": e.value.style.chart.grid.frame.strokeLinejoin,
          "stroke-dasharray": e.value.style.chart.grid.frame.strokeDasharray
        }, null, 8, qo)) : y("", !0),
        e.value.style.chart.grid.x.showHorizontalLines && e.value.orientation === "vertical" ? (s(!0), n(x, { key: 3 }, L(Y.value, (a, o) => (s(), n("line", {
          x1: u.value.left,
          x2: u.value.right,
          y1: a.y,
          y2: a.y,
          stroke: e.value.style.chart.grid.x.linesColor,
          "stroke-width": e.value.style.chart.grid.x.linesThickness,
          "stroke-dasharray": e.value.style.chart.grid.x.linesStrokeDasharray,
          "stroke-linecap": "round"
        }, null, 8, Yo))), 256)) : y("", !0),
        e.value.style.chart.grid.x.showHorizontalLines && e.value.orientation === "horizontal" ? (s(!0), n(x, { key: 4 }, L(v.value.end - v.value.start + 1, (a, o) => (s(), n("line", {
          x1: u.value.left,
          x2: u.value.right,
          y1: u.value.top + p.value * o,
          y2: u.value.top + p.value * o,
          stroke: e.value.style.chart.grid.x.linesColor,
          "stroke-width": e.value.style.chart.grid.x.linesThickness,
          "stroke-dasharray": e.value.style.chart.grid.x.linesStrokeDasharray,
          "stroke-linecap": "round"
        }, null, 8, Xo))), 256)) : y("", !0),
        e.value.style.chart.grid.y.showVerticalLines && e.value.orientation === "vertical" ? (s(!0), n(x, { key: 5 }, L(v.value.end - v.value.start + 1, (a, o) => (s(), n("line", {
          x1: u.value.left + p.value * o,
          x2: u.value.left + p.value * o,
          y1: u.value.top,
          y2: u.value.bottom,
          stroke: e.value.style.chart.grid.y.linesColor,
          "stroke-width": e.value.style.chart.grid.y.linesThickness,
          "stroke-dasharray": e.value.style.chart.grid.y.linesStrokeDasharray,
          "stroke-linecap": "round"
        }, null, 8, Go))), 256)) : y("", !0),
        e.value.style.chart.grid.y.showVerticalLines && e.value.orientation === "horizontal" ? (s(!0), n(x, { key: 6 }, L(Y.value, (a, o) => (s(), n("line", {
          x1: a.horizontal_x,
          x2: a.horizontal_x,
          y1: u.value.top,
          y2: u.value.bottom,
          stroke: e.value.style.chart.grid.y.linesColor,
          "stroke-width": e.value.style.chart.grid.y.linesThickness,
          "stroke-dasharray": e.value.style.chart.grid.y.linesStrokeDasharray,
          "stroke-linecap": "round"
        }, null, 8, Jo))), 256)) : y("", !0),
        (s(!0), n(x, null, L(C.value, (a, o) => (s(), n("g", null, [
          t.$slots.pattern ? (s(), n("defs", Zo, [
            k(t.$slots, "pattern", Pe({ ref_for: !0 }, { seriesIndex: a.absoluteIndex, patternId: `pattern_${W.value}_${a.absoluteIndex}` }), void 0, !0)
          ])) : y("", !0),
          e.value.orientation === "vertical" ? (s(), n(x, { key: 1 }, [
            (s(!0), n(x, null, L(a.x, (c, r) => (s(), n("rect", {
              x: c,
              y: m(ce)(a.y[r]),
              height: a.height[r] < 0 ? 1e-4 : a.height[r] || 0,
              rx: e.value.style.chart.bars.borderRadius > a.height[r] / 2 ? (a.height[r] < 0 ? 0 : a.height[r]) / 2 : e.value.style.chart.bars.borderRadius,
              width: p.value * (1 - e.value.style.chart.bars.gapRatio / 2),
              fill: e.value.style.chart.bars.gradient.show ? `url(#gradient_${a.id})` : a.color,
              stroke: e.value.style.chart.backgroundColor,
              "stroke-width": e.value.style.chart.bars.strokeWidth,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: ye({ "vue-data-ui-bar-animated": e.value.useCssAnimation, "vue-data-ui-bar-transition": se.value })
            }, null, 10, Ko))), 256)),
            t.$slots.pattern ? (s(), n("g", Qo, [
              (s(!0), n(x, null, L(a.x, (c, r) => (s(), n("rect", {
                x: c,
                y: m(ce)(a.y[r]),
                height: a.height[r] < 0 ? 1e-4 : a.height[r] || 0,
                rx: e.value.style.chart.bars.borderRadius > a.height[r] / 2 ? (a.height[r] < 0 ? 0 : a.height[r]) / 2 : e.value.style.chart.bars.borderRadius,
                width: p.value * (1 - e.value.style.chart.bars.gapRatio / 2),
                fill: `url(#pattern_${W.value}_${a.absoluteIndex})`,
                stroke: e.value.style.chart.backgroundColor,
                "stroke-width": e.value.style.chart.bars.strokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                class: ye({ "vue-data-ui-bar-animated": e.value.useCssAnimation, "vue-data-ui-bar-transition": se.value })
              }, null, 10, jo))), 256))
            ])) : y("", !0)
          ], 64)) : (s(), n(x, { key: 2 }, [
            (s(!0), n(x, null, L(a.horizontal_x, (c, r) => (s(), n("rect", {
              x: m(ce)(c, u.value.left),
              y: a.horizontal_y[r] < 0 ? 0 : a.horizontal_y[r],
              width: m(ce)(a.horizontal_width[r] < 0 ? 1e-4 : a.horizontal_width[r]),
              rx: e.value.style.chart.bars.borderRadius > a.height[r] / 2 ? (a.height[r] < 0 ? 0 : a.height[r]) / 2 : e.value.style.chart.bars.borderRadius,
              height: p.value * (1 - e.value.style.chart.bars.gapRatio / 2),
              fill: e.value.style.chart.bars.gradient.show ? `url(#gradient_${a.id})` : a.color,
              stroke: e.value.style.chart.backgroundColor,
              "stroke-width": e.value.style.chart.bars.strokeWidth,
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: ye({ "vue-data-ui-bar-animated": e.value.useCssAnimation, "vue-data-ui-bar-transition": se.value })
            }, null, 10, er))), 256)),
            t.$slots.pattern ? (s(), n("g", tr, [
              (s(!0), n(x, null, L(a.horizontal_x, (c, r) => (s(), n("rect", {
                x: m(ce)(c, u.value.left),
                y: a.horizontal_y[r] < 0 ? 0 : a.horizontal_y[r],
                width: m(ce)(a.horizontal_width[r] < 0 ? 1e-4 : a.horizontal_width[r]),
                rx: e.value.style.chart.bars.borderRadius > a.height[r] / 2 ? (a.height[r] < 0 ? 0 : a.height[r]) / 2 : e.value.style.chart.bars.borderRadius,
                height: p.value * (1 - e.value.style.chart.bars.gapRatio / 2),
                fill: `url(#pattern_${W.value}_${a.absoluteIndex})`,
                stroke: e.value.style.chart.backgroundColor,
                "stroke-width": e.value.style.chart.bars.strokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                class: ye({ "vue-data-ui-bar-animated": e.value.useCssAnimation, "vue-data-ui-bar-transition": se.value })
              }, null, 10, ar))), 256))
            ])) : y("", !0)
          ], 64))
        ]))), 256)),
        e.value.style.chart.grid.x.showAxis ? (s(), n("line", {
          key: 7,
          x1: u.value.left,
          x2: u.value.right,
          y1: u.value.bottom,
          y2: u.value.bottom,
          stroke: e.value.style.chart.grid.x.axisColor,
          "stroke-width": e.value.style.chart.grid.x.axisThickness,
          "stroke-linecap": "round",
          "stroke-linejoin": "round"
        }, null, 8, lr)) : y("", !0),
        e.value.style.chart.grid.y.showAxis && !e.value.style.chart.bars.distributed ? (s(), n("line", {
          key: 8,
          x1: u.value.left,
          x2: u.value.left,
          y1: u.value.top,
          y2: u.value.bottom,
          stroke: e.value.style.chart.grid.y.axisColor,
          "stroke-width": e.value.style.chart.grid.y.axisThickness,
          "stroke-linecap": "round",
          "stroke-linejoin": "round"
        }, null, 8, or)) : y("", !0),
        e.value.style.chart.grid.x.axisName.show && e.value.style.chart.grid.x.axisName.text ? (s(), n("text", {
          key: 9,
          ref_key: "xAxisLabel",
          ref: Ie,
          x: u.value.left + u.value.width / 2,
          y: u.value.chartHeight - 3,
          "font-size": e.value.style.chart.grid.x.axisName.fontSize,
          fill: e.value.style.chart.grid.x.axisName.color,
          "font-weight": e.value.style.chart.grid.x.axisName.bold ? "bold" : "normal",
          "text-anchor": "middle"
        }, H(e.value.style.chart.grid.x.axisName.text), 9, rr)) : y("", !0),
        e.value.style.chart.grid.y.axisName.show && e.value.style.chart.grid.y.axisName.text ? (s(), n("text", {
          key: 10,
          ref_key: "yAxisLabel",
          ref: at,
          transform: `translate(${e.value.style.chart.grid.y.axisName.fontSize}, ${u.value.top + u.value.height / 2}) rotate(-90)`,
          "font-size": e.value.style.chart.grid.y.axisName.fontSize,
          fill: e.value.style.chart.grid.y.axisName.color,
          "font-weight": e.value.style.chart.grid.y.axisName.bold ? "bold" : "normal",
          "text-anchor": "middle"
        }, H(e.value.style.chart.grid.y.axisName.text), 9, sr)) : y("", !0),
        $.value.dataLabels.show && e.value.orientation === "vertical" ? (s(), n(x, { key: 11 }, [
          (s(!0), n(x, null, L(C.value, (a, o) => (s(), n("g", null, [
            (s(!0), n(x, null, L(a.x, (c, r) => (s(), n(x, null, [
              ma(a.series[r], a.proportions[r]) ? (s(), n("text", {
                key: 0,
                x: c + p.value * (1 - e.value.style.chart.bars.gapRatio / 2) / 2,
                y: a.y[r] + a.height[r] / 2 + e.value.style.chart.bars.dataLabels.fontSize / 3,
                "font-size": e.value.style.chart.bars.dataLabels.fontSize,
                fill: e.value.style.chart.bars.dataLabels.adaptColorToBackground ? m(Fa)(a.color) : e.value.style.chart.bars.dataLabels.color,
                "font-weight": e.value.style.chart.bars.dataLabels.bold ? "bold" : "normal",
                "text-anchor": "middle"
              }, H(e.value.style.chart.bars.showDistributedPercentage && e.value.style.chart.bars.distributed ? ra(a.proportions[r] * 100, a, o, r) : Ue(a.series[r], a, o, r, a.signedSeries[r])), 9, nr)) : y("", !0)
            ], 64))), 256))
          ]))), 256)),
          e.value.style.chart.bars.totalValues.show && C.value.length > 1 ? (s(), n("g", {
            key: 0,
            ref_key: "sumTop",
            ref: tl
          }, [
            (s(!0), n(x, null, L(oa.value, (a, o) => (s(), n(x, null, [
              !e.value.style.chart.bars.dataLabels.hideEmptyValues || a.value !== 0 ? (s(), n("text", {
                key: 0,
                x: u.value.left + p.value * o + p.value / 2,
                y: Ol(o),
                "text-anchor": "middle",
                "font-size": e.value.style.chart.bars.totalValues.fontSize,
                "font-weight": e.value.style.chart.bars.totalValues.bold ? "bold" : "normal",
                fill: e.value.style.chart.bars.totalValues.color
              }, H(Ue(a.value, a, o, a.sign)), 9, ir)) : y("", !0)
            ], 64))), 256))
          ], 512)) : y("", !0)
        ], 64)) : y("", !0),
        $.value.dataLabels.show && e.value.orientation === "horizontal" ? (s(), n(x, { key: 12 }, [
          (s(!0), n(x, null, L(C.value, (a, o) => (s(), n("g", null, [
            (s(!0), n(x, null, L(a.horizontal_x, (c, r) => (s(), n(x, null, [
              ma(a.series[r], a.proportions[r]) ? (s(), n("text", {
                key: 0,
                x: c + (a.horizontal_width[r] < 0 ? 1e-4 : a.horizontal_width[r]) / 2,
                y: a.horizontal_y[r] + p.value * (1 - e.value.style.chart.bars.gapRatio / 2) / 2 + e.value.style.chart.bars.dataLabels.fontSize / 3,
                "font-size": e.value.style.chart.bars.dataLabels.fontSize,
                fill: e.value.style.chart.bars.dataLabels.adaptColorToBackground ? m(Fa)(a.color) : e.value.style.chart.bars.dataLabels.color,
                "font-weight": e.value.style.chart.bars.dataLabels.bold ? "bold" : "normal",
                "text-anchor": "middle"
              }, H(e.value.style.chart.bars.showDistributedPercentage && e.value.style.chart.bars.distributed ? ra(a.proportions[r] * 100, a, o, r) : Ue(a.series[r], a, o, r, a.signedSeries[r])), 9, ur)) : y("", !0)
            ], 64))), 256))
          ]))), 256)),
          e.value.style.chart.bars.totalValues.show && C.value.length > 1 ? (s(), n("g", {
            key: 0,
            ref_key: "sumRight",
            ref: lt
          }, [
            (s(!0), n(x, null, L(oa.value, (a, o) => (s(), n(x, null, [
              !e.value.style.chart.bars.dataLabels.hideEmptyValues || a.value !== 0 ? (s(), n("text", {
                key: 0,
                x: Dl(o),
                y: u.value.top + p.value * o + p.value / 2 + e.value.style.chart.bars.totalValues.fontSize / 3,
                "text-anchor": "start",
                "font-size": e.value.style.chart.bars.totalValues.fontSize,
                "font-weight": e.value.style.chart.bars.totalValues.bold ? "bold" : "normal",
                fill: e.value.style.chart.bars.totalValues.color
              }, H(Ue(a.value, a, o, a.sign)), 9, cr)) : y("", !0)
            ], 64))), 256))
          ], 512)) : y("", !0)
        ], 64)) : y("", !0),
        e.value.style.chart.grid.y.axisLabels.show && !e.value.style.chart.bars.distributed && e.value.orientation === "vertical" ? (s(), n("g", {
          key: 13,
          ref_key: "scaleLabels",
          ref: ne
        }, [
          (s(!0), n(x, null, L(Y.value, (a, o) => (s(), n("line", {
            x1: u.value.left,
            x2: u.value.left - 6,
            y1: a.y,
            y2: a.y,
            stroke: e.value.style.chart.grid.x.axisColor,
            "stroke-width": 1
          }, null, 8, vr))), 256)),
          (s(!0), n(x, null, L(Y.value, (a, o) => (s(), n("text", {
            x: a.x,
            y: a.y + e.value.style.chart.grid.y.axisLabels.fontSize / 3,
            "font-size": e.value.style.chart.grid.y.axisLabels.fontSize,
            "font-weight": e.value.style.chart.grid.y.axisLabels.bold ? "bold" : "normal",
            fill: e.value.style.chart.grid.y.axisLabels.color,
            "text-anchor": "end"
          }, H(m(fe)(
            e.value.style.chart.grid.y.axisLabels.formatter,
            a.value,
            m(re)({
              p: e.value.style.chart.bars.dataLabels.prefix,
              v: a.value,
              s: e.value.style.chart.bars.dataLabels.suffix,
              r: e.value.style.chart.grid.y.axisLabels.rounding
            }),
            { datapoint: a }
          )), 9, hr))), 256))
        ], 512)) : y("", !0),
        e.value.style.chart.grid.y.axisLabels.show && !e.value.style.chart.bars.distributed && e.value.orientation === "horizontal" ? (s(), n("g", {
          key: 14,
          ref_key: "scaleLabels",
          ref: ne
        }, [
          (s(!0), n(x, null, L(Y.value, (a, o) => (s(), n("line", {
            x1: a.horizontal_x,
            x2: a.horizontal_x,
            y1: u.value.bottom,
            y2: u.value.bottom + 6,
            stroke: e.value.style.chart.grid.x.axisColor,
            "stroke-width": 1,
            "stroke-linecap": "round"
          }, null, 8, dr))), 256)),
          (s(!0), n(x, null, L(Y.value, (a, o) => (s(), n("text", {
            class: "vue-data-ui-time-label",
            "font-size": e.value.style.chart.grid.x.timeLabels.fontSize,
            "font-weight": e.value.style.chart.grid.y.axisLabels.bold ? "bold" : "normal",
            fill: e.value.style.chart.grid.y.axisLabels.color,
            "text-anchor": e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
            transform: `translate(${a.horizontal_x}, ${u.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}), rotate(${e.value.style.chart.grid.x.timeLabels.rotation})`
          }, H(m(fe)(
            e.value.style.chart.grid.y.axisLabels.formatter,
            a.value,
            m(re)({
              p: e.value.style.chart.bars.dataLabels.prefix,
              v: a.value,
              s: e.value.style.chart.bars.dataLabels.suffix,
              r: e.value.style.chart.grid.y.axisLabels.rounding
            }),
            { datapoint: a }
          )), 9, gr))), 256))
        ], 512)) : y("", !0),
        e.value.style.chart.grid.x.timeLabels.show && e.value.orientation === "vertical" ? (s(), n("g", {
          key: 15,
          ref_key: "timeLabelsEls",
          ref: ie
        }, [
          t.$slots["time-label"] ? (s(), n("g", mr, [
            (s(!0), n(x, null, L(dt.value, (a, o) => (s(), n("g", null, [
              k(t.$slots, "time-label", Pe({ ref_for: !0 }, {
                x: u.value.left + p.value * o + p.value / 2,
                y: u.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY,
                fontSize: e.value.style.chart.grid.x.timeLabels.fontSize,
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: `translate(${u.value.left + p.value * o + p.value / 2}, ${u.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}), rotate(${e.value.style.chart.grid.x.timeLabels.rotation})`,
                absoluteIndex: a.absoluteIndex,
                content: a.text,
                textAnchor: e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
                show: !0
              }), void 0, !0)
            ]))), 256))
          ])) : (s(), n("g", yr, [
            (s(!0), n(x, null, L(dt.value, (a, o) => (s(), n("g", null, [
              String(a.text).includes(`
`) ? (s(), n("text", {
                key: o + "-multi",
                "text-anchor": e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.grid.x.timeLabels.fontSize,
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: `
                                    translate(
                                    ${u.value.left + p.value * o + p.value / 2},
                                    ${u.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}
                                    ),
                                    rotate(${e.value.style.chart.grid.x.timeLabels.rotation})
                                `,
                style: X({ cursor: q.value ? "pointer" : "default" }),
                innerHTML: m(vo)({
                  content: String(a.text),
                  fontSize: e.value.style.chart.grid.x.timeLabels.fontSize,
                  fill: e.value.style.chart.grid.x.timeLabels.color,
                  x: 0,
                  y: 0
                }),
                onClick: () => qe(a, o)
              }, null, 12, br)) : (s(), n("text", {
                class: "vue-data-ui-time-label",
                key: o,
                "text-anchor": e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.grid.x.timeLabels.fontSize,
                "font-weight": e.value.style.chart.grid.x.timeLabels.bold ? "bold" : "normal",
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: `translate(${u.value.left + p.value * o + p.value / 2}, ${u.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}), rotate(${e.value.style.chart.grid.x.timeLabels.rotation})`,
                style: X({ cursor: q.value ? "pointer" : "default" }),
                onClick: () => qe(a, o)
              }, H(a.text), 13, fr))
            ]))), 256))
          ]))
        ], 512)) : y("", !0),
        e.value.style.chart.grid.x.timeLabels.show && e.value.orientation === "horizontal" ? (s(), n("g", {
          key: 16,
          ref_key: "timeLabelsEls",
          ref: ie
        }, [
          t.$slots["time-label"] ? (s(), n("g", pr, [
            (s(!0), n(x, null, L(F.value, (a, o) => (s(), n("g", null, [
              k(t.$slots, "time-label", Pe({ ref_for: !0 }, {
                x: u.value.left - 8,
                y: u.value.top + p.value * o + p.value / 2 + e.value.style.chart.grid.y.axisLabels.fontSize / 3,
                fontSize: e.value.style.chart.grid.x.timeLabels.fontSize,
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: null,
                absoluteIndex: a.absoluteIndex,
                content: a.text,
                textAnchor: "end",
                show: !0
              }), void 0, !0)
            ]))), 256))
          ])) : (s(), n("g", xr, [
            (s(!0), n(x, null, L(F.value, (a, o) => (s(), n("g", null, [
              String(a.text).includes(`
`) ? (s(), n("text", {
                key: 1,
                "text-anchor": "end",
                "font-size": e.value.style.chart.grid.y.axisLabels.fontSize,
                "font-weight": e.value.style.chart.grid.y.axisLabels.bold ? "bold" : "normal",
                fill: e.value.style.chart.grid.y.axisLabels.color,
                x: u.value.left - 8,
                y: u.value.top + p.value * o + p.value / 2 + e.value.style.chart.grid.y.axisLabels.fontSize / 3,
                style: X({ cursor: q.value ? "pointer" : "default" }),
                onClick: () => qe(a, o),
                innerHTML: m(ho)({
                  content: String(a.text),
                  fontSize: e.value.style.chart.grid.y.axisLabels.fontSize,
                  fill: e.value.style.chart.grid.y.axisLabels.color,
                  x: u.value.left - 8,
                  y: 0
                })
              }, null, 12, wr)) : (s(), n("text", {
                key: 0,
                "text-anchor": "end",
                "font-size": e.value.style.chart.grid.y.axisLabels.fontSize,
                "font-weight": e.value.style.chart.grid.y.axisLabels.bold ? "bold" : "normal",
                fill: e.value.style.chart.grid.y.axisLabels.color,
                x: u.value.left - 8,
                y: u.value.top + p.value * o + p.value / 2 + e.value.style.chart.grid.y.axisLabels.fontSize / 3,
                style: X({
                  cursor: q.value ? "pointer" : "default"
                }),
                onClick: () => qe(a, o)
              }, H(a.text), 13, kr))
            ]))), 256))
          ]))
        ], 512)) : y("", !0),
        $.value.showTooltip && e.value.orientation === "vertical" ? (s(!0), n(x, { key: 17 }, L(v.value.end - v.value.start, (a, o) => (s(), n("rect", {
          x: u.value.left + o * p.value,
          y: u.value.top,
          width: p.value,
          height: u.value.height < 0 ? 0 : u.value.height,
          onClick: () => sa(o),
          onMouseenter: () => ia(o),
          onMouseleave: () => na(o),
          fill: o === K.value || o === rt.value ? e.value.style.chart.highlighter.color : "transparent",
          style: X({ opacity: e.value.style.chart.highlighter.opacity / 100 })
        }, null, 44, Lr))), 256)) : y("", !0),
        $.value.showTooltip && e.value.orientation === "horizontal" ? (s(!0), n(x, { key: 18 }, L(v.value.end - v.value.start, (a, o) => (s(), n("rect", {
          x: u.value.left,
          y: u.value.top + o * p.value,
          width: u.value.width < 0 ? 0 : u.value.width,
          height: p.value,
          onClick: () => sa(o),
          onMouseenter: () => ia(o),
          onMouseleave: () => na(o),
          fill: o === K.value || o === rt.value ? e.value.style.chart.highlighter.color : "transparent",
          style: X({ opacity: e.value.style.chart.highlighter.opacity / 100 })
        }, null, 44, Cr))), 256)) : y("", !0),
        Gt.value ? (s(), n("rect", Pe({ key: 19 }, hl.value, {
          "data-start": v.value.start,
          "data-end": v.value.end
        }), null, 16, Sr)) : y("", !0),
        k(t.$slots, "svg", {
          svg: {
            drawingArea: u.value,
            data: C.value,
            isPrintingImg: m(ct) | m(vt) | Fe.value,
            isPrintingSvg: Re.value
          }
        }, void 0, !0)
      ], 14, Do)),
      t.$slots.watermark ? (s(), n("div", zr, [
        k(t.$slots, "watermark", B(U({ isPrinting: m(ct) || m(vt) || Fe.value || Re.value })), void 0, !0)
      ])) : y("", !0),
      e.value.style.chart.zoom.show && be.value && pa.value && N.value > 6 ? (s(), j(No, {
        key: 4,
        ref_key: "chartSlicer",
        ref: pe,
        allMinimaps: ml.value,
        background: e.value.style.chart.zoom.color,
        borderColor: e.value.style.chart.backgroundColor,
        customFormat: e.value.style.chart.zoom.customFormat,
        cutNullValues: !1,
        forceZeroCenter: !0,
        enableRangeHandles: e.value.style.chart.zoom.enableRangeHandles,
        enableSelectionDrag: e.value.style.chart.zoom.enableSelectionDrag,
        end: v.value.end,
        focusOnDrag: e.value.style.chart.zoom.focusOnDrag,
        focusRangeRatio: e.value.style.chart.zoom.focusRangeRatio,
        fontSize: e.value.style.chart.zoom.fontSize,
        immediate: !e.value.style.chart.zoom.preview.enable,
        inputColor: e.value.style.chart.zoom.color,
        isPreview: Gt.value,
        labelLeft: e.value.style.chart.grid.x.timeLabels.values[v.value.start] ? F.value?.[0]?.text ?? "" : "",
        labelRight: e.value.style.chart.grid.x.timeLabels.values[v.value.end - 1] ? F.value?.at(-1)?.text ?? "" : "",
        max: Math.max(...Ke.dataset.map((a) => a.series.length)),
        min: 0,
        minimap: ue.value,
        minimapCompact: e.value.style.chart.zoom.minimap.compact,
        minimapFrameColor: e.value.style.chart.zoom.minimap.frameColor,
        minimapIndicatorColor: e.value.style.chart.zoom.minimap.indicatorColor,
        minimapMerged: !1,
        minimapSelectedColor: e.value.style.chart.zoom.minimap.selectedColor,
        minimapSelectedColorOpacity: e.value.style.chart.zoom.minimap.selectedColorOpacity,
        minimapSelectedIndex: K.value,
        minimapSelectionRadius: 1,
        preciseLabels: la.value.length ? la.value : ge.value,
        refreshEndPoint: e.value.style.chart.zoom.endIndex !== null ? e.value.style.chart.zoom.endIndex + 1 : Math.max(...Ke.dataset.map((a) => a.series.length)),
        refreshStartPoint: e.value.style.chart.zoom.startIndex !== null ? e.value.style.chart.zoom.startIndex : 0,
        selectColor: e.value.style.chart.zoom.highlightColor,
        selectedSeries: bl.value,
        smoothMinimap: !1,
        start: v.value.start,
        textColor: e.value.style.chart.color,
        timeLabels: ge.value,
        usePreciseLabels: e.value.style.chart.grid.x.timeLabels.datetimeFormatter.enable && !e.value.style.chart.zoom.useDefaultFormat,
        valueEnd: v.value.end,
        valueStart: v.value.start,
        verticalHandles: e.value.style.chart.zoom.minimap.verticalHandles,
        maxWidth: e.value.style.chart.zoom.maxWidth,
        isCursorPointer: q.value,
        additionalMinimapHeight: e.value.style.chart.zoom.minimap.additionalHeight,
        handleType: e.value.style.chart.zoom.minimap.handleType,
        handleIconColor: e.value.style.chart.zoom.minimap.handleIconColor,
        handleBorderWidth: e.value.style.chart.zoom.minimap.handleBorderWidth,
        handleBorderColor: e.value.style.chart.zoom.minimap.handleBorderColor,
        handleFill: e.value.style.chart.zoom.minimap.handleFill,
        handleWidth: e.value.style.chart.zoom.minimap.handleWidth,
        "onUpdate:end": Rl,
        "onUpdate:start": Fl,
        onTrapMouse: dl,
        onReset: Zt,
        onFutureEnd: l[0] || (l[0] = (a) => Jt("end", a)),
        onFutureStart: l[1] || (l[1] = (a) => Jt("start", a))
      }, {
        "reset-action": w(({ reset: a }) => [
          k(t.$slots, "reset-action", B(U({ reset: a })), void 0, !0)
        ]),
        slotMap: w(({ width: a, height: o, unitW: c }) => [
          (s(!0), n(x, null, L(C.value, (r) => (s(), n("g", {
            key: r.id
          }, [
            (s(!0), n(x, null, L(r.xMinimap({ left: 0, unitW: c }), (b, S) => (s(), n("rect", {
              key: S,
              x: S === 0 ? b - c * (e.value.style.chart.bars.gapRatio / 4) : b - c / 2,
              y: Math.max(0, Math.min(o, r.yMinimap({ minimapH: o })[S])),
              height: r.heightMinimap({ minimapH: o })[S],
              width: [0, N.value - 1].includes(S) ? c * (1 - e.value.style.chart.bars.gapRatio / 2) / 2 : c * (1 - e.value.style.chart.bars.gapRatio / 2),
              fill: e.value.style.chart.bars.gradient.show ? `url(#gradient_${r.id})` : r.color,
              stroke: e.value.style.chart.backgroundColor,
              "stroke-width": 0.5,
              rx: "0",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              style: X({
                opacity: S >= E.value.start && S <= E.value.end ? 1 : 0.62
              })
            }, null, 12, Mr))), 128))
          ]))), 128))
        ]),
        _: 3
      }, 8, ["allMinimaps", "background", "borderColor", "customFormat", "enableRangeHandles", "enableSelectionDrag", "end", "focusOnDrag", "focusRangeRatio", "fontSize", "immediate", "inputColor", "isPreview", "labelLeft", "labelRight", "max", "minimap", "minimapCompact", "minimapFrameColor", "minimapIndicatorColor", "minimapSelectedColor", "minimapSelectedColorOpacity", "minimapSelectedIndex", "preciseLabels", "refreshEndPoint", "refreshStartPoint", "selectColor", "selectedSeries", "start", "textColor", "timeLabels", "usePreciseLabels", "valueEnd", "valueStart", "verticalHandles", "maxWidth", "isCursorPointer", "additionalMinimapHeight", "handleType", "handleIconColor", "handleBorderWidth", "handleBorderColor", "handleFill", "handleWidth"])) : y("", !0),
      Ae(m(Ba), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: $.value.showTooltip && Qe.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: Z.value,
        content: et.value,
        isFullscreen: xe.value,
        isCustom: m(Ra)(e.value.style.chart.tooltip.customFormat),
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": w(() => [
          k(t.$slots, "tooltip-before", B(U({ ...je.value })), void 0, !0)
        ]),
        "tooltip-after": w(() => [
          k(t.$slots, "tooltip-after", B(U({ ...je.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      oe("div", {
        id: `legend-bottom-${W.value}`
      }, null, 8, _r),
      At.value ? (s(), j(to, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${W.value}` : `#legend-bottom-${W.value}`
      }, [
        oe("div", {
          ref_key: "chartLegend",
          ref: _t
        }, [
          e.value.style.chart.legend.show ? (s(), j(Po, {
            key: 0,
            legendSet: Se.value,
            config: Sl.value,
            isCursorPointer: q.value,
            onClickMarker: l[2] || (l[2] = ({ legend: a }) => a.segregate())
          }, Ct({
            item: w(({ legend: a }) => [
              m(he) ? y("", !0) : (s(), n("div", {
                key: 0,
                onClick: (o) => a.segregate(),
                style: X(`opacity:${_.value.includes(a.id) ? 0.5 : 1}`)
              }, H(a.name), 13, Tr))
            ]),
            legendToggle: w(() => [
              Se.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !m(he) ? (s(), j(Fo, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: _.value.length > 0,
                isCursorPointer: q.value,
                onToggle: wl
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : y("", !0)
            ]),
            _: 2
          }, [
            t.$slots.pattern ? {
              name: "legend-pattern",
              fn: w(({ legend: a, index: o }) => [
                Ae($o, {
                  shape: a.shape,
                  radius: 30,
                  stroke: "none",
                  plot: { x: 30, y: 30 },
                  fill: `url(#pattern_${W.value}_${o})`
                }, null, 8, ["shape", "fill"])
              ]),
              key: "0"
            } : void 0
          ]), 1032, ["legendSet", "config", "isCursorPointer"])) : k(t.$slots, "legend", {
            key: 1,
            legend: Se.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : y("", !0),
      t.$slots.source ? (s(), n("div", {
        key: 6,
        ref_key: "source",
        ref: Tt,
        dir: "auto"
      }, [
        k(t.$slots, "source", {}, void 0, !0)
      ], 512)) : y("", !0),
      be.value && e.value.userOptions.buttons.table ? (s(), j(ao(ze.value.component), Pe({ key: 7 }, ze.value.props, {
        ref_key: "tableUnit",
        ref: ke,
        onClose: ya
      }), Ct({
        content: w(() => [
          Ae(m(Wa), {
            colNames: He.value.colNames,
            head: He.value.head,
            body: He.value.body,
            config: He.value.config,
            title: e.value.table.useDialog ? "" : ze.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: q.value,
            onClose: ya
          }, {
            th: w(({ th: a }) => [
              oe("div", { innerHTML: a }, null, 8, $r)
            ]),
            td: w(({ td: a }) => [
              Ma(H(isNaN(Number(a)) ? a : m(re)({
                p: e.value.style.chart.bars.dataLabels.prefix,
                v: a,
                s: e.value.style.chart.bars.dataLabels.suffix,
                r: e.value.table.td.roundingValue
              })), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"])
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: w(() => [
            Ma(H(ze.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: w(() => [
            oe("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[3] || (l[3] = (a) => yt(e.value.userOptions.callbacks.csv)),
              style: X({ cursor: q.value ? "pointer" : "default" })
            }, [
              Ae(m(Ua), {
                name: "fileCsv",
                stroke: ze.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : y("", !0),
      k(t.$slots, "skeleton", {}, () => [
        m(he) ? (s(), j(ko, { key: 0 })) : y("", !0)
      ], !0)
    ], 46, Io));
  }
}, ts = /* @__PURE__ */ Ro(Ar, [["__scopeId", "data-v-30e2e23e"]]);
export {
  ts as default
};
