// vue-data-ui/dist/vue-ui-smiley-BK5EKp9u.js
import { openBlock as i, createElementBlock as s, withKeys as K, normalizeStyle as r, normalizeClass as T, createElementVNode as a, toDisplayString as v, renderSlot as $, createCommentVNode as f, unref as M, useSlots as q, ref as x, onMounted as N, computed as w, watch as V, Fragment as H, renderList as J, createVNode as Q, withCtx as z, createTextVNode as W } from "vue";
import { X as j, v as U, k as Z, l as _, o as ee, q as te } from "./lib-GwPnJSrY.js";
import { u as le } from "./useNestedProp-BBT0mbHh.js";
import { u as ae } from "./useConfig-CIgyKZ6R.js";
import { _ as oe } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const ne = ["xmlns", "stroke"], ie = ["id"], se = ["stop-color"], re = ["stop-color"], ue = ["stroke"], ce = ["xmlns", "viewBox", "stroke"], de = ["id"], ye = ["stop-color"], fe = ["stop-color"], ve = ["viewBox", "stroke"], ge = {
  __name: "BaseSmilingUnit",
  props: {
    config: { type: Object },
    unit: { type: Number },
    currentRating: { type: Number },
    getActiveColor: { type: Function },
    calcShapeFill: { type: Function },
    isReadonly: { type: Boolean },
    hasBreakdown: { type: Boolean },
    hoveredValue: { type: Number },
    isCursorPointer: { type: Boolean }
  },
  emits: ["rate", "mouseenter", "mouseleave"],
  setup(t, { emit: R }) {
    const S = t, g = R;
    return (o, c) => (i(), s("div", {
      tabindex: "0",
      class: T({ "vue-ui-smiley-rated": !t.config.readonly && t.currentRating === t.unit }),
      style: r({
        cursor: t.config.readonly ? "default" : S.isCursorPointer ? "pointer" : "default",
        height: t.config.style.itemSize + "px",
        aspectRatio: "1/1",
        position: "relative"
      }),
      onMouseenter: c[0] || (c[0] = (h) => g("mouseenter")),
      onMouseleave: c[1] || (c[1] = (h) => g("mouseleave")),
      onClick: c[2] || (c[2] = (h) => g("rate", t.unit)),
      onKeyup: c[3] || (c[3] = K((h) => g("rate", t.unit), ["enter"]))
    }, [
      t.config.style.tooltip.show && t.hasBreakdown && t.isReadonly ? (i(), s("div", {
        key: 0,
        class: "vue-ui-rating-tooltip",
        style: r({
          border: `1px solid ${t.config.style.tooltip.borderColor}`,
          position: "absolute",
          top: `${t.config.style.tooltip.offsetY - 48}px`,
          left: "50%",
          transform: "translateX(-50%)",
          width: "fit-content",
          textAlign: "center",
          background: t.config.style.tooltip.backgroundColor,
          display: t.hoveredValue === t.unit - 1 ? "block" : "none",
          padding: "2px 12px",
          borderRadius: t.config.style.tooltip.borderRadius + "px",
          boxShadow: t.config.style.tooltip.boxShadow
        })
      }, [
        a("div", {
          style: r({
            width: "100%",
            display: "flex",
            flexDirection: "row",
            gap: "6px",
            position: "relative",
            textAlign: "center",
            color: t.config.style.tooltip.color
          })
        }, [
          a("span", {
            style: r(`font-size:${t.config.style.tooltip.fontSize}px`)
          }, v(t.unit) + ":", 5),
          a("span", {
            style: r(`font-weight:${t.config.style.tooltip.bold ? "bold" : "normal"};font-size:${t.config.style.tooltip.fontSize}px`)
          }, [
            $(o.$slots, "rating", {}, void 0, !0)
          ], 4),
          a("div", {
            style: r(`font-family:Arial !important;position:absolute;top:calc(100% - 4px);left:50%;transform:translateX(-50%);color:${t.config.style.tooltip.borderColor}`)
          }, " ▼ ", 4)
        ], 4)
      ], 4)) : f("", !0),
      t.config.style.icons.filled ? (i(), s("svg", {
        key: 1,
        xmlns: M(j),
        style: { transition: "all 0.1s ease-in-out", position: "absolute", top: "0", left: "0" },
        height: "100%",
        viewBox: "0 0 24 24",
        "stroke-width": "1.5",
        stroke: t.getActiveColor(t.unit - 1),
        fill: "none",
        "stroke-linecap": "round",
        "stroke-linejoin": "round"
      }, [
        a("defs", null, [
          a("radialGradient", {
            id: `vueUiSmiley${t.unit - 1}`
          }, [
            a("stop", {
              offset: "0%",
              "stop-color": M(U)(t.config.style.colors.active[t.unit - 1], 0.05)
            }, null, 8, se),
            a("stop", {
              offset: "100%",
              "stop-color": t.config.style.colors.active[t.unit - 1]
            }, null, 8, re)
          ], 8, ie)
        ]),
        $(o.$slots, "path-icon-filled", {}, void 0, !0)
      ], 8, ne)) : (i(), s("svg", {
        key: 2,
        style: { position: "absolute", top: "0", left: "0", transition: "all 0.1s ease-in-out" },
        height: "100%",
        viewBox: "0 0 24 24",
        "stroke-width": "1.5",
        stroke: t.getActiveColor(t.unit - 1),
        fill: "none",
        "stroke-linecap": "round",
        "stroke-linejoin": "round"
      }, [
        $(o.$slots, "path-icon", {}, void 0, !0)
      ], 8, ue)),
      t.config.style.icons.filled && t.isReadonly ? (i(), s("svg", {
        key: 3,
        xmlns: M(j),
        style: { transition: "all 0.1s ease-in-out", position: "absolute", top: "0", left: "0" },
        height: "100%",
        viewBox: `0 0 ${t.calcShapeFill(t.unit - 1)} 24`,
        "stroke-width": "1.5",
        stroke: t.config.style.colors.activeReadonly[t.unit - 1],
        fill: "none",
        "stroke-linecap": "round",
        "stroke-linejoin": "round"
      }, [
        a("defs", null, [
          a("radialGradient", {
            id: `vueUiSmiley${t.unit - 1}`
          }, [
            a("stop", {
              offset: "0%",
              "stop-color": M(U)(t.config.style.colors.activeReadonly[t.unit - 1], 0.05)
            }, null, 8, ye),
            a("stop", {
              offset: "100%",
              "stop-color": t.config.style.colors.activeReadonly[t.unit - 1]
            }, null, 8, fe)
          ], 8, de)
        ]),
        $(o.$slots, "path-icon-filled-readonly", {}, void 0, !0)
      ], 8, ce)) : f("", !0),
      !t.config.style.icons.filled && t.isReadonly ? (i(), s("svg", {
        key: 4,
        style: { position: "absolute", top: "0", left: "0", transition: "all 0.1s ease-in-out" },
        height: "100%",
        viewBox: `0 0 ${t.calcShapeFill(t.unit - 1)} 24`,
        "stroke-width": "1.5",
        stroke: t.config.style.colors.activeReadonly[t.unit - 1],
        fill: "none",
        "stroke-linecap": "round",
        "stroke-linejoin": "round"
      }, [
        $(o.$slots, "path-icon-readonly", {}, void 0, !0)
      ], 8, ve)) : f("", !0)
    ], 38));
  }
}, me = /* @__PURE__ */ oe(ge, [["__scopeId", "data-v-843c3372"]]), he = {
  key: 0,
  class: "vue-ui-rating-title",
  style: { width: "100%" }
}, pe = ["d", "fill"], be = ["d"], $e = ["d", "fill"], ze = ["d"], Se = {
  __name: "vue-ui-smiley",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["rate"],
  setup(t, { expose: R, emit: S }) {
    const { vue_ui_smiley: g } = ae(), o = t, c = q(), h = S, m = x(null);
    N(() => {
      c["chart-background"] && console.warn("VueUiSmiley does not support the #chart-background slot.");
    });
    function C() {
      (!Object.hasOwn(o.dataset, "rating") || ee(o.dataset)) && te({
        componentName: "VueUiSmiley",
        type: "datasetAttribute",
        property: "rating"
      });
    }
    N(C);
    const e = x(I()), X = w(() => e.value.useCursorPointer);
    function I() {
      return le({
        userConfig: o.config,
        defaultConfig: g
      });
    }
    V(() => o.config, (l) => {
      e.value = I(), k.value = e.value.readonly, C();
    }, { deep: !0 });
    const F = w(() => {
      const l = o.dataset.rating;
      return l && typeof l == "object" && !Array.isArray(l) ? P(l) : l ?? null;
    }), G = w(() => typeof o.dataset.rating == "object" && !Array.isArray(o.dataset.rating)), d = x(F.value);
    V(
      F,
      (l) => {
        d.value = l;
      }
    );
    const k = x(e.value.readonly), O = [
      {
        key: "smiley_0",
        pathIconFilled: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-5 9.86a4.5 4.5 0 0 0 -3.214 1.35a1 1 0 1 0 1.428 1.4a2.5 2.5 0 0 1 3.572 0a1 1 0 0 0 1.428 -1.4a4.5 4.5 0 0 0 -3.214 -1.35zm-2.99 -4.2l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm6 0l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007z",
        pathIcon: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0 M9.5 15.25a3.5 3.5 0 0 1 5 0",
        pathIconFilledReadonly: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-5 9.86a4.5 4.5 0 0 0 -3.214 1.35a1 1 0 1 0 1.428 1.4a2.5 2.5 0 0 1 3.572 0a1 1 0 0 0 1.428 -1.4a4.5 4.5 0 0 0 -3.214 -1.35zm-2.99 -4.2l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm6 0l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007z",
        pathIconReadonly: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0 M9.5 15.25a3.5 3.5 0 0 1 5 0"
      },
      {
        key: "smiley_1",
        pathIconFilled: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-2 10.66h-6l-.117 .007a1 1 0 0 0 0 1.986l.117 .007h6l.117 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-5.99 -5l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm6 0l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007z",
        pathIcon: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0 M9 15l6 0",
        pathIconFilledReadonly: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-2 10.66h-6l-.117 .007a1 1 0 0 0 0 1.986l.117 .007h6l.117 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-5.99 -5l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm6 0l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007z",
        pathIconReadonly: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0 M9 15l6 0"
      },
      {
        key: "smiley_2",
        pathIconFilled: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-7.99 5.66l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm6 0l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007z",
        pathIcon: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0",
        pathIconFilledReadonly: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-7.99 5.66l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm6 0l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007z",
        pathIconReadonly: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0"
      },
      {
        key: "smiley_3",
        pathIconFilled: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-1.8 10.946a1 1 0 0 0 -1.414 .014a2.5 2.5 0 0 1 -3.572 0a1 1 0 0 0 -1.428 1.4a4.5 4.5 0 0 0 6.428 0a1 1 0 0 0 -.014 -1.414zm-6.19 -5.286l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993zm6 0l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993z",
        pathIcon: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0 M9.5 15a3.5 3.5 0 0 0 5 0",
        pathIconFilledReadonly: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-1.8 10.946a1 1 0 0 0 -1.414 .014a2.5 2.5 0 0 1 -3.572 0a1 1 0 0 0 -1.428 1.4a4.5 4.5 0 0 0 6.428 0a1 1 0 0 0 -.014 -1.414zm-6.19 -5.286l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993zm6 0l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993z",
        pathIconReadonly: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 10l.01 0 M15 10l.01 0 M9.5 15a3.5 3.5 0 0 0 5 0"
      },
      {
        key: "smiley_4",
        pathIconFilled: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-2 9.66h-6a1 1 0 0 0 -1 1v.05a3.975 3.975 0 0 0 3.777 3.97l.227 .005a4.026 4.026 0 0 0 3.99 -3.79l.006 -.206a1 1 0 0 0 -1 -1.029zm-5.99 -5l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993zm6 0l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993z",
        pathIcon: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 9l.01 0 M15 9l.01 0 M8 13a4 4 0 1 0 8 0h-8",
        pathIconFilledReadonly: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-2 9.66h-6a1 1 0 0 0 -1 1v.05a3.975 3.975 0 0 0 3.777 3.97l.227 .005a4.026 4.026 0 0 0 3.99 -3.79l.006 -.206a1 1 0 0 0 -1 -1.029zm-5.99 -5l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993zm6 0l-.127 .007a1 1 0 0 0 .117 1.993l.127 -.007a1 1 0 0 0 -.117 -1.993z",
        pathIconReadonly: "M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0 M9 9l.01 0 M15 9l.01 0 M8 13a4 4 0 1 0 8 0h-8"
      }
    ];
    function P(l) {
      if (l == null) return null;
      let n = 0, u = 0;
      for (const y in l) {
        const b = Number(y), B = l[y];
        Number.isFinite(b) && (n += b * B, u += B);
      }
      return u === 0 ? 0 : n / u;
    }
    function Y(l) {
      const n = d.value - l, u = 24;
      switch (!0) {
        case n < 0:
          return 0;
        case n > 1:
          return 1 * u;
        default:
          return n * u;
      }
    }
    function A(l) {
      return e.value.readonly ? e.value.style.colors.inactive[l] : d.value === l + 1 ? e.value.style.icons.useGradient && e.value.style.icons.filled ? `url(#vueUiSmiley${l})` : e.value.style.colors.active[l] : m.value !== null && m.value === l ? e.value.style.icons.useGradient && e.value.style.icons.filled ? `url(#vueUiSmiley${l})` : e.value.style.colors.active[l] : e.value.style.colors.inactive[l];
    }
    const p = w(() => ({ value: l, tooltip: n = !1 }) => Z(
      n ? e.value.style.tooltip.formatter : e.value.style.rating.formatter,
      l,
      _({
        v: l,
        r: n ? e.value.style.tooltip.roundingValue : e.value.style.rating.roundingValue
      }),
      e.value
    ));
    function L(l) {
      k.value || (d.value = l, h("rate", l));
    }
    function D() {
      return d.value;
    }
    function E(l = !0) {
      k.value = l;
    }
    return R({
      getData: D,
      toggleReadonly: E
    }), (l, n) => (i(), s("div", {
      class: "vue-data-ui-component vue-ui-smiley",
      style: r(`background:${e.value.style.backgroundColor};font-family:${e.value.style.fontFamily};width:100%;`),
      onMouseleave: n[1] || (n[1] = (u) => m.value = void 0)
    }, [
      e.value.style.title.text ? (i(), s("div", he, [
        a("div", {
          style: r(`color:${e.value.style.title.color};font-weight:${e.value.style.title.bold ? "bold" : "normal"};text-align:${e.value.style.title.textAlign};margin-bottom:${e.value.style.title.offsetY}px;font-size:${e.value.style.title.fontSize}px`)
        }, v(e.value.style.title.text), 5),
        e.value.style.title.subtitle.text ? (i(), s("div", {
          key: 0,
          style: r(`color:${e.value.style.title.subtitle.color};font-size:${e.value.style.title.subtitle.fontSize}px;text-align:${e.value.style.title.textAlign};margin-bottom:${e.value.style.title.subtitle.offsetY}px;font-weight:${e.value.style.title.subtitle.bold ? "bold" : "normal"}`)
        }, v(e.value.style.title.subtitle.text), 5)) : f("", !0)
      ])) : f("", !0),
      e.value.style.rating.show && e.value.style.rating.position === "top" ? (i(), s("div", {
        key: 1,
        style: r(`width:100%;text-align:center;margin-bottom:${e.value.style.rating.offsetY}px;font-size:${e.value.style.rating.fontSize}px;font-weight:${e.value.style.rating.bold ? "bold" : "normal"};margin-left:${e.value.style.rating.offsetX}px`)
      }, v(p.value({ value: d.value })), 5)) : f("", !0),
      a("div", {
        class: "vue-ui-smiley-wrapper",
        style: r(`overflow:visible;height:${e.value.style.itemSize}px;width:fit-content;margin:0 auto;display:flex;align-items:center;justify-content:center;`)
      }, [
        e.value.style.rating.show && e.value.style.rating.position === "left" ? (i(), s("div", {
          key: 0,
          style: r(`width:fit-content;text-align:center;margin-bottom:${e.value.style.rating.offsetY}px;font-size:${e.value.style.rating.fontSize}px;font-weight:${e.value.style.rating.bold ? "bold" : "normal"};padding-right:${e.value.style.rating.offsetX}px`)
        }, v(p.value({ value: d.value })), 5)) : f("", !0),
        (i(), s(H, null, J(O, (u, y) => Q(me, {
          key: u.key,
          config: e.value,
          unit: y + 1,
          currentRating: d.value,
          isReadonly: k.value,
          hasBreakdown: G.value,
          hoveredValue: m.value,
          getActiveColor: A,
          calcShapeFill: Y,
          isCursorPointer: X.value,
          onMouseenter: (b) => m.value = y,
          onMouseleave: n[0] || (n[0] = (b) => m.value = null),
          onRate: (b) => L(y + 1)
        }, {
          rating: z(() => [
            W(v(p.value({ value: o.dataset.rating[String(y + 1)], tooltip: !0 })), 1)
          ]),
          "path-icon-filled": z(() => [
            a("path", {
              d: u.pathIconFilled,
              "stroke-width": "0",
              fill: A(y)
            }, null, 8, pe)
          ]),
          "path-icon": z(() => [
            a("path", {
              d: u.pathIcon
            }, null, 8, be)
          ]),
          "path-icon-filled-readonly": z(() => [
            a("path", {
              d: u.pathIconFilledReadonly,
              "stroke-width": "0",
              fill: e.value.style.icons.useGradient ? `url(#vueUiSmiley${y})` : e.value.style.colors.activeReadonly[y]
            }, null, 8, $e)
          ]),
          "path-icon-readonly": z(() => [
            a("path", {
              d: u.pathIconReadonly
            }, null, 8, ze)
          ]),
          _: 2
        }, 1032, ["config", "unit", "currentRating", "isReadonly", "hasBreakdown", "hoveredValue", "isCursorPointer", "onMouseenter", "onRate"])), 64)),
        e.value.style.rating.show && e.value.style.rating.position === "right" ? (i(), s("div", {
          key: 1,
          style: r(`width:fit-content;text-align:center;margin-bottom:${e.value.style.rating.offsetY}px;font-size:${e.value.style.rating.fontSize}px;font-weight:${e.value.style.rating.bold ? "bold" : "normal"};padding-left:${e.value.style.rating.offsetX}px`)
        }, v(p.value({ value: d.value })), 5)) : f("", !0)
      ], 4),
      e.value.style.rating.show && e.value.style.rating.position === "bottom" ? (i(), s("div", {
        key: 2,
        style: r(`width:100%;text-align:center;margin-top:${e.value.style.rating.offsetY}px;font-size:${e.value.style.rating.fontSize}px;font-weight:${e.value.style.rating.bold ? "bold" : "normal"};margin-left:${e.value.style.rating.offsetX}px`)
      }, v(p.value({ value: d.value })), 5)) : f("", !0)
    ], 36));
  }
};
export {
  Se as default
};
