// vue-data-ui/dist/vue-ui-dumbbell-CVSA90b2.js
import { defineAsyncComponent as G, computed as y, ref as h, toRefs as Qt, watch as ue, shallowRef as nt, onMounted as rt, onBeforeUnmount as ut, watchEffect as el, openBlock as n, createElementBlock as v, unref as i, normalizeStyle as Z, normalizeClass as it, createBlock as I, withCtx as p, renderSlot as m, normalizeProps as w, guardReactiveProps as $, createCommentVNode as g, createElementVNode as c, createSlots as ct, createVNode as vt, Fragment as L, renderList as M, toDisplayString as _, withDirectives as J, vShow as Q, Teleport as tl, createTextVNode as Ne, resolveDynamicComponent as ll, mergeProps as al, nextTick as dt } from "vue";
import { j as ht, t as ft, o as sl, q as yt, r as ol, K as bt, D as nl, l as A, k as E, X as gt, u as pt, x as ee, w as te, a0 as rl, a1 as ul } from "./lib-GwPnJSrY.js";
import { t as Me, u as il } from "./useResponsive-DfdjqQps.js";
import { u as cl } from "./useConfig-CIgyKZ6R.js";
import { u as vl } from "./usePrinter-DiK5R6k-.js";
import { u as dl, B as hl } from "./BaseScanner-BhhkKVoj.js";
import { u as fl } from "./useSvgExport-ByUukOZt.js";
import { u as Xe } from "./useNestedProp-BBT0mbHh.js";
import { u as yl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as bl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as gl } from "./useChartAccessibility-9icAAmYg.js";
import { u as pl } from "./useTimeLabelCollider-CIsgDrl9.js";
import ml from "./img-BYYzY_QU.js";
import kl from "./Title-BfyrP-jp.js";
import xl from "./vue_ui_dumbbell-BCOHDiPk.js";
import _l from "./Legend-1LMMsPxL.js";
import { _ as Cl } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const wl = ["id"], $l = ["id"], Ll = ["xmlns", "viewBox"], Al = ["x", "y", "width", "height"], Sl = { key: 1 }, Nl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Ml = { key: 2 }, Xl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], zl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], Tl = ["transform", "font-size", "fill"], Ol = ["x", "y", "font-size", "fill", "font-weight", "onMouseenter", "onMouseleave", "onClick"], Pl = ["x", "y", "font-size", "fill", "onMouseenter", "onMouseleave", "onClick"], Il = ["x", "y", "font-size", "fill"], Dl = ["transform", "font-size", "fill", "font-weight", "text-anchor"], Fl = ["d", "stroke", "stroke-width", "stroke-dasharray"], Hl = ["d", "stroke", "stroke-width", "stroke-dasharray"], Gl = ["x", "y", "height", "width", "fill"], El = ["transform", "fill", "font-size"], Rl = ["id"], Bl = ["stop-color"], Vl = ["stop-color"], Ul = ["stop-color"], Wl = ["id"], ql = ["stop-color"], jl = ["stop-color"], Yl = ["stop-color"], Kl = ["id"], Zl = ["stop-color"], Jl = ["stop-color"], Ql = ["stop-color"], ea = ["id"], ta = ["stop-color"], la = ["stop-color"], aa = ["stop-color"], sa = ["id"], oa = ["stop-color"], na = ["stop-color"], ra = ["stop-color"], ua = ["id"], ia = ["stop-color"], ca = ["stop-color"], va = ["id"], da = ["stop-color"], ha = ["stop-color"], fa = { key: 0 }, ya = { key: 0 }, ba = ["d", "fill"], ga = { key: 1 }, pa = ["x", "y", "height", "width", "fill"], ma = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], ka = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], xa = { key: 7 }, _a = ["x", "y", "fill", "font-size"], Ca = { key: 8 }, wa = ["x", "y", "fill", "font-size"], $a = ["x", "y", "width", "height", "fill", "onMouseenter", "onMouseleave", "onClick"], La = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Aa = ["id"], Sa = ["xmlns", "height", "width"], Na = ["fill"], Ma = ["innerHTML"], Xa = {
  __name: "vue-ui-dumbbell",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectDatapoint", "copyAlt"],
  setup(mt, { expose: kt, emit: xt }) {
    const _t = G(() => import("./BaseIcon-CNx8Tzl5.js")), Ct = G(() => import("./vue-ui-accordion-BGHTqIBt.js")), wt = G(() => import("./DataTable-Bo1pmAqG.js")), $t = G(() => import("./PenAndPaper-DWV1FMIi.js")), Lt = G(() => import("./UserOptions-CrLqEimB.js")), At = G(() => import("./PackageVersion-DN9d0ME5.js")), St = G(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_dumbbell: Nt } = cl(), { isThemeValid: Mt, warnInvalidTheme: Xt } = yl(), C = mt, zt = xt, le = y({
      get() {
        return !!C.dataset && C.dataset.length;
      },
      set(l) {
        return l;
      }
    }), f = h(ht()), ze = h(0), X = h(null), Te = h(null), Oe = h(null), Pe = h(null), Ie = h(null), De = h(0), Fe = h(0), He = h(0), Ge = h(!1), q = h(null), ie = h(null), ae = h(!1), se = h(!1), ce = h(null), ve = h(null), oe = h(null), de = h(null), D = h(null), r = h(null), F = h(!1), e = h(ge()), j = y(() => e.value.userOptions.useCursorPointer), Tt = y(() => ft({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        useAnimation: !1,
        style: {
          chart: {
            backgroundColor: "#99999930",
            padding: {
              top: 12,
              right: 12,
              bottom: 12,
              left: 12
            },
            grid: {
              horizontalGrid: { stroke: "#6A6A6A" },
              verticalGrid: { stroke: "#6A6A6A" }
            },
            labels: {
              axis: {
                yLabel: "",
                xLabel: ""
              },
              xAxisLabels: { show: !1 },
              yAxisLabels: { show: !1 },
              endLabels: { show: !1 },
              startLabels: { show: !1 }
            },
            legend: {
              backgroundColor: "transparent"
            },
            plots: {
              endColor: "#969696",
              startColor: "#DBDBDB",
              stroke: "#6A6A6A",
              evaluationColors: { enable: !1 }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: he, FINAL_DATASET: fe, manualLoading: Ot } = dl({
      ...Qt(C),
      FINAL_CONFIG: e,
      prepareConfig: ge,
      skeletonDataset: C.config?.skeletonDataset ?? [
        { name: "_", start: 21, end: 34 },
        { name: "_", start: 13, end: 21 },
        { name: "_", start: 8, end: 13 },
        { name: "_", start: 5, end: 8 },
        { name: "_", start: 3, end: 5 }
      ],
      skeletonConfig: ft({
        defaultConfig: e.value,
        userConfig: Tt.value
      })
    }), { userOptionsVisible: ye, setUserOptionsVisibility: Ee, keepUserOptionState: Re } = bl({ config: e.value }), { svgRef: be } = gl({ config: e.value.style.chart.title });
    function ge() {
      const l = Xe({
        userConfig: C.config,
        defaultConfig: Nt
      }), s = l.theme;
      if (!s) return l;
      if (!Mt.value(l))
        return Xt(l), l;
      const t = Xe({
        userConfig: xl[s] || C.config,
        defaultConfig: l
      });
      return Xe({
        userConfig: C.config,
        defaultConfig: t
      });
    }
    ue(() => C.config, (l) => {
      he.value || (e.value = ge()), ye.value = !e.value.userOptions.showOnChartHover, Ve(), De.value += 1, Fe.value += 1, He.value += 1, V.value = e.value.style.chart.rowHeight, U.value = e.value.style.chart.width, H.value.showTable = e.value.table.show;
    }, { deep: !0 }), ue(() => fe.value, (l) => {
      Array.isArray(l) && l.length > 0 && (Ot.value = !1), Ke(), Ze();
    }, { deep: !0 });
    const z = nt(null), R = nt(null);
    rt(() => {
      Ge.value = !0, Ve();
    });
    const Be = y(() => !!e.value.debug);
    function Ve() {
      if (sl(C.dataset) ? yt({
        componentName: "VueUiDumbbell",
        type: "dataset",
        debug: Be.value
      }) : C.dataset.forEach((l, s) => {
        ol({
          datasetObject: l,
          requiredAttributes: ["name", "start", "end"]
        }).forEach((t) => {
          le.value = !1, yt({
            componentName: "VueUiDumbbell",
            type: "datasetSerieAttribute",
            property: t,
            index: s,
            debug: Be.value
          });
        });
      }), e.value.responsive) {
        const l = Me(() => {
          const { width: s, height: t } = il({
            chart: X.value,
            title: e.value.style.chart.title.text ? Te.value : null,
            legend: e.value.style.chart.legend.show ? Oe.value : null,
            source: Pe.value,
            noTitle: Ie.value
          }), a = e.value.style.chart.title.text ? 24 : 0, u = e.value.style.chart.legend.show ? 24 : 0;
          requestAnimationFrame(async () => {
            U.value = Math.max(0.1, s), V.value = Math.max(0.1, (Math.max(0.1, t) - (a + u)) / fe.value.length), tt();
          });
        });
        z.value && (R.value && z.value.unobserve(R.value), z.value.disconnect()), z.value = new ResizeObserver(l), R.value = X.value.parentNode, z.value.observe(R.value);
      }
      tt();
    }
    ut(() => {
      N.value != null && (cancelAnimationFrame(N.value), N.value = null), z.value && (R.value && z.value.unobserve(R.value), z.value.disconnect());
    });
    const { isPrinting: pe, isImaging: me, generatePdf: Ue, generateImage: We } = vl({
      elementId: `dumbbell_${f.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-dumbbell",
      options: e.value.userOptions.print
    }), Pt = y(() => e.value.userOptions.show && !e.value.style.chart.title.text), H = h({
      showTable: e.value.table.show
    });
    ue(e, () => {
      H.value = {
        showTable: e.value.table.show
      };
    }, { immediate: !0 });
    const B = y(() => fe.value.map((l, s) => ({
      ...l,
      start: bt(l.start),
      end: bt(l.end),
      id: l.id ?? `${String(l.name)}__${String(l.start)}__${String(l.end)}__${ht()}`
    }))), qe = y(() => {
      const l = e.value.style.chart.grid, s = B.value.flatMap((b) => [b.start, b.end]).map((b) => Number(b)).filter((b) => Number.isFinite(b)), t = s.length ? Math.min(...s) : 0, a = s.length ? Math.max(...s) : 0, u = l.scaleMin ?? Math.min(t, 0), d = l.scaleMax ?? a;
      return { min: u, max: d };
    }), S = y(() => nl(qe.value.min, qe.value.max, e.value.style.chart.grid.scaleSteps)), V = h(e.value.style.chart.rowHeight), U = h(e.value.style.chart.width);
    function It() {
      let l = 0;
      ve.value && (l = Array.from(ve.value.querySelectorAll("text")).reduce((a, u) => {
        const d = u.getComputedTextLength();
        return d > a ? d : a;
      }, 0));
      const s = ce.value ? ce.value.getBoundingClientRect().width : 0;
      return l + s + (s ? 24 + e.value.style.chart.labels.axis.yLabelOffsetX : 0);
    }
    const ke = h(0), Dt = Me((l) => {
      ke.value = l;
    }, 100);
    el((l) => {
      const s = oe.value;
      if (!s) return;
      const t = new ResizeObserver((a) => {
        Dt(a[0].contentRect.height);
      });
      t.observe(s), l(() => t.disconnect());
    }), ut(() => {
      ke.value = 0;
    });
    const je = y(() => {
      F.value;
      let l = 0;
      de.value && (l = de.value.getBBox().height);
      let s = 0;
      return oe.value && (s = ke.value), l + s;
    }), xe = y(() => B.value.length);
    function Ye(l, s, t) {
      const a = Number(l), u = Number(s.min), d = Number(s.max), b = Number(t.width);
      if (!Number.isFinite(a) || !Number.isFinite(u) || !Number.isFinite(d) || !Number.isFinite(b))
        return t.left;
      const k = d - u;
      return k <= 0 ? t.left : t.left + (a - u) / k * b;
    }
    const o = y(() => {
      F.value;
      const l = It(), s = e.value.style.chart.padding, t = e.value.style.chart.labels.axis.xLabel ? e.value.style.chart.labels.axis.xLabelOffsetY : 0, a = V.value * xe.value - je.value - s.top - s.bottom - t, u = a / xe.value, d = V.value * xe.value, b = U.value - l - s.left - s.right, k = S.value.ticks.length * (b / S.value.ticks.length);
      return {
        left: e.value.style.chart.padding.left + l,
        right: U.value - e.value.style.chart.padding.right,
        top: e.value.style.chart.padding.top,
        bottom: d - e.value.style.chart.padding.bottom - je.value - t,
        width: b,
        height: a,
        rowHeight: u,
        absoluteHeight: d,
        widthPlotReference: k
      };
    }), T = y(() => Math.min(V.value / 2 * 0.7, e.value.style.chart.plots.radius)), Y = h([]), x = y({
      get() {
        const l = o.value, s = S.value;
        return Y.value.map((t, a) => {
          const u = Ye(t.start, s, l), d = Ye(t.endVal, s, l), b = u + (d - u) / 2, k = ![null, void 0].includes(t.start) && ![null, void 0].includes(t.end) && t.end > t.start, P = ![null, void 0].includes(t.start) && ![null, void 0].includes(t.end) && t.end < t.start, ot = ![null, void 0].includes(t.start) && ![null, void 0].includes(t.end) && t.end === t.start || [null, void 0].includes(t.start) || [null, void 0].includes(t.end), Jt = k ? e.value.style.chart.plots.evaluationColors.positive : P ? e.value.style.chart.plots.evaluationColors.negative : e.value.style.chart.plots.evaluationColors.neutral;
          return {
            ...t,
            isPositive: k,
            isNegative: P,
            isNeutral: ot,
            evaluationColor: Jt,
            evaluationGrad: `url(#${k ? "positive" : P ? "negative" : "neutral"}_grad_${f.value})`,
            startX: u,
            endX: d,
            centerX: b,
            y: l.top + a * l.rowHeight + l.rowHeight / 2
          };
        });
      },
      set(l) {
        Y.value = l;
      }
    });
    function Ke() {
      Y.value = B.value.map((l) => {
        const s = Number(l.start);
        return {
          ...l,
          endVal: Number.isFinite(s) ? s : 0
        };
      });
    }
    const N = h(null);
    rt(() => {
      Ze();
    });
    function Ze() {
      if (N.value != null && (cancelAnimationFrame(N.value), N.value = null), Ke(), !e.value.useAnimation) {
        x.value = Y.value.map((a) => {
          const u = Number(a.end);
          return { ...a, endVal: Number.isFinite(u) ? u : a.endVal };
        });
        return;
      }
      const l = Math.max(1, Math.min(100, e.value.animationSpeed || 100)) / 100, s = B.value.map((a) => {
        const u = Number(a.start), d = Number(a.end), b = Number.isFinite(u) ? u : 0;
        return (Number.isFinite(d) ? d : b) - b;
      }), t = () => {
        let a = !0;
        x.value = Y.value.map((u, d) => {
          const b = Number.isFinite(Number(u.end)) ? Number(u.end) : u.endVal, k = u.endVal + s[d] * l, P = s[d] >= 0 ? Math.min(k, b) : Math.max(k, b);
          return P !== b && (a = !1), { ...u, endVal: P };
        }), a ? N.value = null : N.value = requestAnimationFrame(t);
      };
      N.value = requestAnimationFrame(t);
    }
    const _e = y(() => e.value.style.chart.plots.evaluationColors.enable ? [
      { name: e.value.style.chart.legend.labelNegative, color: e.value.style.chart.plots.gradient.show ? `url(#negative_grad_${f.value})` : e.value.style.chart.plots.evaluationColors.negative },
      { name: e.value.style.chart.legend.labelNeutral, color: e.value.style.chart.plots.gradient.show ? `url(#neutral_grad_${f.value})` : e.value.style.chart.plots.evaluationColors.neutral },
      { name: e.value.style.chart.legend.labelPositive, color: e.value.style.chart.plots.gradient.show ? `url(#positive_grad_${f.value})` : e.value.style.chart.plots.evaluationColors.positive }
    ] : [
      { name: e.value.style.chart.legend.labelStart, color: e.value.style.chart.plots.gradient.show ? `url(#start_grad_${f.value})` : e.value.style.chart.plots.startColor },
      { name: e.value.style.chart.legend.labelEnd, color: e.value.style.chart.plots.gradient.show ? `url(#end_grad_${f.value})` : e.value.style.chart.plots.endColor }
    ]), Ft = y(() => ({
      cy: "donut-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      paddingTop: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), O = y(() => {
      const l = x.value.map((t) => ({
        name: t.name
      })), s = x.value.map((t) => ({
        start: t.start,
        end: t.end
      }));
      return { head: l, body: s };
    }), ne = y(() => {
      const l = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.start,
        e.value.table.columnNames.end,
        e.value.table.columnNames.progression
      ], s = O.value.head.map((u, d) => {
        const b = A({
          p: e.value.style.chart.labels.prefix,
          v: O.value.body[d].start,
          s: e.value.style.chart.labels.suffix,
          r: e.value.table.td.roundingValue
        }), k = A({
          p: e.value.style.chart.labels.prefix,
          v: O.value.body[d].end,
          s: e.value.style.chart.labels.suffix,
          r: e.value.table.td.roundingValue
        }), P = A({
          v: 100 * (O.value.body[d].end / O.value.body[d].start - 1),
          s: "%",
          r: e.value.table.td.roundingPercentage
        });
        return [
          { name: u.name },
          b,
          k,
          P
        ];
      }), t = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.series,
          e.value.table.columnNames.start,
          e.value.table.columnNames.end,
          e.value.table.columnNames.progression
        ],
        head: l,
        body: s,
        config: t
      };
    });
    function Ce(l = null) {
      dt(() => {
        const s = O.value.head.map((u, d) => [[
          u.name
        ], [O.value.body[d].start], [O.value.body[d].end]]), t = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [[e.value.table.columnNames.series], [e.value.table.columnNames.start], [e.value.table.columnNames.end]]].concat(s), a = rl(t);
        l ? l(a) : ul({ csvContent: a, title: e.value.style.chart.title.text || "vue-ui-dumbbell" });
      });
    }
    const W = h(!1);
    function Je(l) {
      W.value = l, ze.value += 1;
    }
    function Ht() {
      return B.value;
    }
    function Qe() {
      H.value.showTable = !H.value.showTable;
    }
    const re = h(!1);
    function we() {
      re.value = !re.value;
    }
    async function Gt({ scale: l = 2 } = {}) {
      if (!X.value) return;
      const { width: s, height: t } = X.value.getBoundingClientRect(), a = s / t, { imageUri: u, base64: d } = await ml({ domElement: X.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: u,
        base64: d,
        title: e.value.style.chart.title.text,
        width: s,
        height: t,
        aspectRatio: a
      };
    }
    const et = y(() => S.value.ticks), Et = y(() => ({ start: 0, end: et.value.length }));
    pl({
      timeLabelsEls: oe,
      timeLabels: et,
      slicer: Et,
      configRef: e,
      rotationPath: ["style", "chart", "labels", "xAxisLabels", "rotation"],
      autoRotatePath: ["style", "chart", "labels", "xAxisLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: U,
      height: V,
      targetClass: ".vue-ui-dumbbell-scale-label",
      rotation: e.value.style.chart.labels.xAxisLabels.autoRotate.angle
    });
    function Rt({
      rowHeight: l,
      fontSize: s,
      showProgression: t
    }) {
      if (!t) return !1;
      const a = l / 3, u = l / 1.3, d = Math.abs(u - a), b = s * 1.2;
      return d < b;
    }
    const Bt = /* @__PURE__ */ (() => {
      let l = null, s = 0;
      const t = 1;
      return (a) => {
        if (a === F.value) {
          l = null, s = 0;
          return;
        }
        l === null || l !== a ? (l = a, s = 1) : (s += 1, s >= t && (F.value = a, l = null, s = 0));
      };
    })(), tt = Me(() => {
      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          const l = Rt({
            rowHeight: o.value.rowHeight,
            fontSize: e.value.style.chart.labels.yAxisLabels.fontSize,
            showProgression: e.value.style.chart.labels.yAxisLabels.showProgression
          });
          Bt(l);
        });
      });
    }, 100);
    function $e({ datapoint: l, seriesIndex: s }) {
      D.value = s, r.value = l, e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: l, seriesIndex: s });
    }
    function Le({ datapoint: l, seriesIndex: s }) {
      D.value = null, r.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: s });
    }
    function Ae({ datapoint: l, seriesIndex: s }) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: s });
    }
    const Se = y(() => {
      if (r.value === null) return 0;
      const l = ![null, void 0].includes(r.value.start), s = ![null, void 0].includes(r.value.end);
      return l && s ? Math.min(r.value.startX, r.value.endX) + Math.abs(r.value.startX - r.value.endX) / 2 : l && !s ? r.value.startX : s && !l ? r.value.endX : null;
    }), Vt = y(() => {
      if (r.value === null) return "";
      const l = ![null, void 0].includes(r.value.start), s = ![null, void 0].includes(r.value.end);
      let t = "", a = "";
      return l && (t = E(
        e.value.style.chart.labels.formatter,
        r.value.start,
        A({
          p: e.value.style.chart.labels.prefix,
          v: r.value.start,
          s: e.value.style.chart.labels.suffix,
          r: e.value.style.chart.labels.startLabels.rounding
        }),
        { datapoint: r.value, seriesIndex: D.value }
      )), s && (a = E(
        e.value.style.chart.labels.formatter,
        r.value.end,
        A({
          p: e.value.style.chart.labels.prefix,
          v: r.value.end,
          s: e.value.style.chart.labels.suffix,
          r: e.value.style.chart.labels.startLabels.rounding
        }),
        { datapoint: r.value, seriesIndex: D.value }
      )), l && s ? `${t} → ${a}` : l && !s ? t : s && !l ? a : "";
    }), K = y(() => {
      const l = e.value.table.useDialog && !e.value.table.show, s = H.value.showTable;
      return {
        component: l ? St : Ct,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: l ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: W.value,
          fullscreenParent: X.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: j.value
        } : {
          hideDetails: !0,
          config: {
            open: s,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ue(() => H.value.showTable, (l) => {
      e.value.table.show || (l && e.value.table.useDialog && q.value ? q.value.open() : "close" in q.value && q.value.close());
    });
    function lt() {
      H.value.showTable = !1, ie.value && ie.value.setTableIconState(!1);
    }
    const Ut = y(() => _e.value.map((l) => ({
      ...l,
      shape: "circle"
    }))), Wt = y(() => e.value.style.chart.backgroundColor), qt = y(() => e.value.style.chart.legend), jt = y(() => e.value.style.chart.title), { exportSvg: Yt, getSvg: Kt } = fl({
      svg: be,
      title: jt,
      legend: qt,
      legendItems: Ut,
      backgroundColor: Wt
    });
    async function at({ isCb: l }) {
      se.value = !0, await dt();
      try {
        if (l) {
          const { blob: s, url: t, text: a, dataUrl: u } = await Kt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: s, url: t, text: a, dataUrl: u }));
        } else
          await Promise.resolve(Yt());
      } finally {
        se.value = !1;
      }
    }
    function Zt(l) {
      if (l?.stage === "start") {
        ae.value = !0;
        return;
      }
      if (l?.stage === "end") {
        ae.value = !1;
        return;
      }
      We();
    }
    async function st() {
      if (zt("copyAlt", {
        config: e.value,
        dataset: x.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: x.value
      }));
    }
    return kt({
      getData: Ht,
      getImage: Gt,
      generatePdf: Ue,
      generateCsv: Ce,
      generateImage: We,
      generateSvg: at,
      toggleTable: Qe,
      toggleAnnotator: we,
      toggleFullscreen: Je,
      copyAlt: st
    }), (l, s) => (n(), v("div", {
      ref_key: "dumbbellChart",
      ref: X,
      class: it(`vue-data-ui-component vue-ui-dumbbell ${W.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      style: Z(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height:100%" : ""}`),
      id: `dumbbell_${f.value}`,
      onMouseenter: s[1] || (s[1] = () => i(Ee)(!0)),
      onMouseleave: s[2] || (s[2] = () => i(Ee)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (n(), I(i($t), {
        key: 0,
        svgRef: i(be),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: re.value,
        isCursorPointer: j.value,
        onClose: we
      }, {
        "annotator-action-close": p(() => [
          m(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": p(({ color: t }) => [
          m(l.$slots, "annotator-action-color", w($({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": p(({ mode: t }) => [
          m(l.$slots, "annotator-action-draw", w($({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": p(({ disabled: t }) => [
          m(l.$slots, "annotator-action-undo", w($({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": p(({ disabled: t }) => [
          m(l.$slots, "annotator-action-redo", w($({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": p(({ disabled: t }) => [
          m(l.$slots, "annotator-action-delete", w($({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : g("", !0),
      Pt.value ? (n(), v("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Ie,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : g("", !0),
      e.value.style.chart.title.text ? (n(), v("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Te,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (n(), I(kl, {
          key: `title_${De.value}`,
          config: {
            title: {
              cy: "donut-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "donut-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : g("", !0),
      c("div", {
        id: `legend-top-${f.value}`
      }, null, 8, $l),
      e.value.userOptions.show && le.value && (i(Re) || i(ye)) ? (n(), I(i(Lt), {
        ref_key: "userOptionsRef",
        ref: ie,
        key: `user_option_${ze.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: i(pe),
        isImaging: i(me),
        uid: f.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: W.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: X.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: re.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: j.value,
        onToggleFullscreen: Je,
        onGeneratePdf: i(Ue),
        onGenerateCsv: Ce,
        onGenerateImage: Zt,
        onGenerateSvg: at,
        onToggleTable: Qe,
        onToggleAnnotator: we,
        onCopyAlt: st,
        style: Z({
          visibility: i(Re) ? i(ye) ? "visible" : "hidden" : "visible"
        })
      }, ct({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: p(({ isOpen: t, color: a }) => [
            m(l.$slots, "menuIcon", w($({ isOpen: t, color: a })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: p(() => [
            m(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionCsv ? {
          name: "optionCsv",
          fn: p(() => [
            m(l.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: p(() => [
            m(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: p(() => [
            m(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionTable ? {
          name: "optionTable",
          fn: p(() => [
            m(l.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: p(({ toggleFullscreen: t, isFullscreen: a }) => [
            m(l.$slots, "optionFullscreen", w($({ toggleFullscreen: t, isFullscreen: a })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: p(({ toggleAnnotator: t, isAnnotator: a }) => [
            m(l.$slots, "optionAnnotator", w($({ toggleAnnotator: t, isAnnotator: a })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: p(({ altCopy: t }) => [
            m(l.$slots, "optionAltCopy", w($({ altCopy: t })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : g("", !0),
      (n(), v("svg", {
        ref_key: "svgRef",
        ref: be,
        xmlns: i(gt),
        class: it({ "vue-data-ui-fullscreen--on": W.value, "vue-data-ui-fulscreen--off": !W.value }),
        viewBox: `0 0 ${U.value} ${o.value.absoluteHeight <= 0 ? 10 : o.value.absoluteHeight}`,
        style: Z(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color}`)
      }, [
        vt(i(At)),
        l.$slots["chart-background"] ? (n(), v("foreignObject", {
          key: 0,
          x: o.value.left,
          y: o.value.top,
          width: Math.max(0.1, o.value.width),
          height: Math.max(0.1, o.value.height),
          style: {
            pointerEvents: "none"
          }
        }, [
          m(l.$slots, "chart-background", {}, void 0, !0)
        ], 8, Al)) : g("", !0),
        e.value.style.chart.grid.verticalGrid.show ? (n(), v("g", Sl, [
          (n(!0), v(L, null, M(S.value.ticks, (t, a) => (n(), v("line", {
            x1: o.value.left + a * o.value.width / (S.value.ticks.length - 1),
            x2: o.value.left + a * o.value.width / (S.value.ticks.length - 1),
            y1: o.value.top,
            y2: o.value.bottom,
            stroke: e.value.style.chart.grid.verticalGrid.stroke,
            "stroke-width": e.value.style.chart.grid.verticalGrid.strokeWidth,
            "stroke-dasharray": e.value.style.chart.grid.verticalGrid.strokeDasharray
          }, null, 8, Nl))), 256))
        ])) : g("", !0),
        e.value.style.chart.grid.horizontalGrid.show ? (n(), v("g", Ml, [
          (n(!0), v(L, null, M(B.value, (t, a) => (n(), v("line", {
            x1: o.value.left,
            x2: o.value.right,
            y1: o.value.top + a * o.value.rowHeight,
            y2: o.value.top + a * o.value.rowHeight,
            stroke: e.value.style.chart.grid.horizontalGrid.stroke,
            "stroke-width": e.value.style.chart.grid.horizontalGrid.strokeWidth,
            "stroke-dasharray": e.value.style.chart.grid.horizontalGrid.strokeDasharray
          }, null, 8, Xl))), 256)),
          c("line", {
            x1: o.value.left,
            x2: o.value.right,
            y1: o.value.bottom,
            y2: o.value.bottom,
            stroke: e.value.style.chart.grid.horizontalGrid.stroke,
            "stroke-width": e.value.style.chart.grid.horizontalGrid.strokeWidth,
            "stroke-dasharray": e.value.style.chart.grid.horizontalGrid.strokeDasharray
          }, null, 8, zl)
        ])) : g("", !0),
        e.value.style.chart.labels.axis.yLabel ? (n(), v("text", {
          key: 3,
          ref_key: "yAxisLabel",
          ref: ce,
          transform: `translate(${e.value.style.chart.labels.axis.fontSize}, ${o.value.absoluteHeight / 2}), rotate(-90)`,
          "font-size": e.value.style.chart.labels.axis.fontSize,
          fill: e.value.style.chart.labels.axis.color,
          "text-anchor": "middle"
        }, _(e.value.style.chart.labels.axis.yLabel), 9, Tl)) : g("", !0),
        e.value.style.chart.labels.yAxisLabels.show ? (n(), v("g", {
          key: 4,
          ref_key: "serieLabels",
          ref: ve
        }, [
          (n(!0), v(L, null, M(x.value, (t, a) => (n(), v("text", {
            class: "vue-ui-dumbbell-serie-name",
            key: `serieLabel_${t.id}_${a}`,
            x: o.value.left - 6 + e.value.style.chart.labels.yAxisLabels.offsetX,
            y: o.value.top + a * o.value.rowHeight + (!e.value.style.chart.labels.yAxisLabels.showProgression || F.value ? o.value.rowHeight / 2 : o.value.rowHeight / 3) + e.value.style.chart.labels.yAxisLabels.fontSize / 3,
            "font-size": e.value.style.chart.labels.yAxisLabels.fontSize,
            fill: e.value.style.chart.labels.yAxisLabels.color,
            "font-weight": e.value.style.chart.labels.yAxisLabels.bold ? "bold" : "normal",
            "text-anchor": "end",
            onMouseenter: (u) => $e({ datapoint: t, seriesIndex: a }),
            onMouseleave: (u) => Le({ datapoint: t, seriesIndex: a }),
            onClick: (u) => Ae({ datapoint: t, seriesIndex: a })
          }, _(t.name) + " " + _(F.value && e.value.style.chart.labels.yAxisLabels.showProgression ? [null, void 0].includes(t.start) || [null, void 0].includes(t.end) ? "" : `(${i(E)(
            e.value.style.chart.labels.yAxisLabels.formatter,
            100 * (t.end / t.start - 1),
            i(A)({
              v: 100 * (t.end / t.start - 1),
              s: "%",
              r: e.value.style.chart.labels.yAxisLabels.rounding
            }),
            { datapoint: t }
          )})` : ""), 41, Ol))), 128)),
          e.value.style.chart.labels.yAxisLabels.showProgression && !F.value ? (n(!0), v(L, { key: 0 }, M(x.value, (t, a) => (n(), v("text", {
            class: "vue-ui-dumbbell-serie-value",
            x: o.value.left - 6 + e.value.style.chart.labels.yAxisLabels.offsetX,
            y: o.value.top + a * o.value.rowHeight + o.value.rowHeight / 1.3 + e.value.style.chart.labels.yAxisLabels.fontSize / 3,
            "font-size": e.value.style.chart.labels.yAxisLabels.fontSize,
            fill: e.value.style.chart.labels.yAxisLabels.color,
            "text-anchor": "end",
            onMouseenter: (u) => $e({ datapoint: t, seriesIndex: a }),
            onMouseleave: (u) => Le({ datapoint: t, seriesIndex: a }),
            onClick: (u) => Ae({ datapoint: t, seriesIndex: a })
          }, _([null, void 0].includes(t.start) || [null, void 0].includes(t.end) ? "" : i(E)(
            e.value.style.chart.labels.yAxisLabels.formatter,
            100 * (t.end / t.start - 1),
            i(A)({
              v: 100 * (t.end / t.start - 1),
              s: "%",
              r: e.value.style.chart.labels.yAxisLabels.rounding
            }),
            { datapoint: t }
          )), 41, Pl))), 256)) : g("", !0)
        ], 512)) : g("", !0),
        e.value.style.chart.labels.axis.xLabel ? (n(), v("text", {
          key: 5,
          ref_key: "xAxisLabel",
          ref: de,
          x: o.value.left + o.value.width / 2,
          y: o.value.absoluteHeight - e.value.style.chart.labels.axis.fontSize / 3,
          "font-size": e.value.style.chart.labels.axis.fontSize,
          fill: e.value.style.chart.labels.axis.color,
          "text-anchor": "middle"
        }, _(e.value.style.chart.labels.axis.xLabel), 9, Il)) : g("", !0),
        e.value.style.chart.labels.xAxisLabels.show ? (n(), v("g", {
          key: 6,
          ref_key: "scaleLabels",
          ref: oe
        }, [
          (n(!0), v(L, null, M(S.value.ticks, (t, a) => (n(), v("text", {
            class: "vue-ui-dumbbell-scale-label",
            key: `tick_${a}`,
            transform: `translate(${o.value.left + a * (o.value.width / (S.value.ticks.length - 1))}, ${o.value.bottom + e.value.style.chart.labels.xAxisLabels.fontSize + e.value.style.chart.labels.xAxisLabels.offsetY}), rotate(${e.value.style.chart.labels.xAxisLabels.rotation})`,
            "font-size": e.value.style.chart.labels.xAxisLabels.fontSize,
            fill: e.value.style.chart.labels.xAxisLabels.color,
            "font-weight": e.value.style.chart.labels.xAxisLabels.bold ? "bold" : "normal",
            "text-anchor": e.value.style.chart.labels.xAxisLabels.rotation > 0 ? "start" : e.value.style.chart.labels.xAxisLabels.rotation < 0 ? "end" : "middle"
          }, _(i(E)(
            e.value.style.chart.labels.formatter,
            t,
            i(A)({
              p: e.value.style.chart.labels.prefix,
              v: t,
              s: e.value.style.chart.labels.suffix,
              r: e.value.style.chart.labels.xAxisLabels.rounding
            }),
            { datapoint: t, seriesIndex: a }
          )), 9, Dl))), 128))
        ], 512)) : g("", !0),
        J(c("g", null, [
          J(c("path", {
            d: `M ${r.value ? r.value.startX : o.value.left},${o.value.top} ${r.value ? r.value.startX : o.value.left},${o.value.bottom}`,
            stroke: r.value ? e.value.style.chart.plots.evaluationColors.enable ? r.value.evaluationColor : e.value.style.chart.plots.startColor : "transparent",
            "stroke-width": e.value.style.chart.comparisonLines.strokeWidth,
            "stroke-dasharray": e.value.style.chart.comparisonLines.strokeDasharray,
            style: { transition: "all 0.3s ease-in-out" }
          }, null, 8, Fl), [
            [Q, r.value !== null && ![null, void 0].includes(r.value.start)]
          ]),
          J(c("path", {
            d: `M ${r.value ? r.value.endX : o.value.left},${o.value.top} ${r.value ? r.value.endX : o.value.left},${o.value.bottom}`,
            stroke: r.value ? e.value.style.chart.plots.evaluationColors.enable ? r.value.evaluationColor : e.value.style.chart.plots.endColor : "transparent",
            "stroke-width": e.value.style.chart.comparisonLines.strokeWidth,
            "stroke-dasharray": e.value.style.chart.comparisonLines.strokeDasharray,
            style: { transition: "all 0.3s ease-in-out" }
          }, null, 8, Hl), [
            [Q, r.value !== null && ![null, void 0].includes(r.value.end)]
          ]),
          J(c("rect", {
            x: r.value ? Math.min(r.value.startX, r.value.endX) : o.value.left,
            y: o.value.top,
            height: Math.max(0.1, o.value.height),
            width: r.value ? Math.max(0.1, Math.abs(r.value.endX - r.value.startX)) : 0,
            fill: r.value ? i(pt)(e.value.style.chart.comparisonLines.rectColor, e.value.style.chart.comparisonLines.rectOpacity) : "transparent",
            style: { transition: "all 0.3s ease-in-out" }
          }, null, 8, Gl), [
            [Q, e.value.style.chart.comparisonLines.showRect && r.value !== null && ![null, void 0].includes(r.value.start) && ![null, void 0].includes(r.value.end)]
          ]),
          J(c("text", {
            transform: `translate(${Se.value == null ? 0 : Se.value}, ${o.value.top - 6})`,
            fill: e.value.style.chart.comparisonLines.labelColor,
            "font-size": e.value.style.chart.comparisonLines.labelFontSize,
            "text-anchor": "middle",
            style: { transition: "all 0.3s ease-in-out" }
          }, _(Vt.value), 9, El), [
            [Q, r.value !== null && Se.value !== null && e.value.style.chart.comparisonLines.showLabel]
          ])
        ], 512), [
          [Q, e.value.style.chart.comparisonLines.show && D.value !== null]
        ]),
        c("defs", null, [
          c("radialGradient", {
            id: `start_grad_${f.value}`,
            fy: "30%"
          }, [
            c("stop", {
              offset: "10%",
              "stop-color": i(ee)(e.value.style.chart.plots.startColor, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, Bl),
            c("stop", {
              offset: "90%",
              "stop-color": i(te)(e.value.style.chart.plots.startColor, 0.1)
            }, null, 8, Vl),
            c("stop", {
              offset: "100%",
              "stop-color": e.value.style.chart.plots.startColor
            }, null, 8, Ul)
          ], 8, Rl),
          c("radialGradient", {
            id: `end_grad_${f.value}`,
            fy: "30%"
          }, [
            c("stop", {
              offset: "10%",
              "stop-color": i(ee)(e.value.style.chart.plots.endColor, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, ql),
            c("stop", {
              offset: "90%",
              "stop-color": i(te)(e.value.style.chart.plots.endColor, 0.1)
            }, null, 8, jl),
            c("stop", {
              offset: "100%",
              "stop-color": e.value.style.chart.plots.endColor
            }, null, 8, Yl)
          ], 8, Wl),
          c("radialGradient", {
            id: `positive_grad_${f.value}`,
            fy: "30%"
          }, [
            c("stop", {
              offset: "10%",
              "stop-color": i(ee)(e.value.style.chart.plots.evaluationColors.positive, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, Zl),
            c("stop", {
              offset: "90%",
              "stop-color": i(te)(e.value.style.chart.plots.evaluationColors.positive, 0.1)
            }, null, 8, Jl),
            c("stop", {
              offset: "100%",
              "stop-color": e.value.style.chart.plots.evaluationColors.positive
            }, null, 8, Ql)
          ], 8, Kl),
          c("radialGradient", {
            id: `negative_grad_${f.value}`,
            fy: "30%"
          }, [
            c("stop", {
              offset: "10%",
              "stop-color": i(ee)(e.value.style.chart.plots.evaluationColors.negative, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, ta),
            c("stop", {
              offset: "90%",
              "stop-color": i(te)(e.value.style.chart.plots.evaluationColors.negative, 0.1)
            }, null, 8, la),
            c("stop", {
              offset: "100%",
              "stop-color": e.value.style.chart.plots.evaluationColors.negative
            }, null, 8, aa)
          ], 8, ea),
          c("radialGradient", {
            id: `neutral_grad_${f.value}`,
            fy: "30%"
          }, [
            c("stop", {
              offset: "10%",
              "stop-color": i(ee)(e.value.style.chart.plots.evaluationColors.neutral, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, oa),
            c("stop", {
              offset: "90%",
              "stop-color": i(te)(e.value.style.chart.plots.evaluationColors.neutral, 0.1)
            }, null, 8, na),
            c("stop", {
              offset: "100%",
              "stop-color": e.value.style.chart.plots.evaluationColors.neutral
            }, null, 8, ra)
          ], 8, sa)
        ]),
        (n(!0), v(L, null, M(x.value, (t, a) => (n(), v("g", {
          key: `plot_${a}_${t.id}`
        }, [
          c("defs", null, [
            c("linearGradient", {
              id: `grad_pos_${f.value}`,
              x1: "0%",
              x2: "100%",
              y1: "0%",
              y2: "0%"
            }, [
              c("stop", {
                offset: "0%",
                "stop-color": e.value.style.chart.plots.startColor
              }, null, 8, ia),
              c("stop", {
                offset: "100%",
                "stop-color": e.value.style.chart.plots.endColor
              }, null, 8, ca)
            ], 8, ua),
            c("linearGradient", {
              id: `grad_neg_${f.value}`,
              x1: "0%",
              x2: "100%",
              y1: "0%",
              y2: "0%"
            }, [
              c("stop", {
                offset: "0%",
                "stop-color": e.value.style.chart.plots.endColor
              }, null, 8, da),
              c("stop", {
                offset: "100%",
                "stop-color": e.value.style.chart.plots.startColor
              }, null, 8, ha)
            ], 8, va)
          ]),
          ![void 0, null].includes(t.end) && ![void 0, null].includes(t.start) ? (n(), v("g", fa, [
            e.value.style.chart.plots.link.type === "curved" ? (n(), v("g", ya, [
              c("path", {
                d: `M 
                                ${t.startX},${t.y + T.value / 2} 
                                C ${t.centerX},${t.y} ${t.centerX},${t.y} 
                                ${t.endX},${t.y + T.value / 2}
                                L ${t.endX},${t.y - T.value / 2}
                                C ${t.centerX},${t.y} ${t.centerX},${t.y}
                                ${t.startX},${t.y - T.value / 2}
                                Z
                            `,
                fill: e.value.style.chart.plots.evaluationColors.enable ? t.evaluationColor : t.endX > t.startX ? `url(#grad_pos_${f.value})` : `url(#grad_neg_${f.value})`
              }, null, 8, ba)
            ])) : (n(), v("g", ga, [
              c("rect", {
                x: t.endX > t.startX ? t.startX : t.endX,
                y: t.y - e.value.style.chart.plots.link.strokeWidth / 2,
                height: Math.max(0.01, e.value.style.chart.plots.link.strokeWidth),
                width: Math.max(0.01, Math.abs(t.endX - t.startX)),
                fill: e.value.style.chart.plots.evaluationColors.enable ? t.evaluationColor : t.endX > t.startX ? `url(#grad_pos_${f.value})` : `url(#grad_neg_${f.value})`
              }, null, 8, pa)
            ]))
          ])) : g("", !0),
          [null, void 0].includes(t.start) ? g("", !0) : (n(), v("circle", {
            key: 1,
            cx: t.startX,
            cy: t.y,
            r: T.value,
            fill: e.value.style.chart.plots.gradient.show ? e.value.style.chart.plots.evaluationColors.enable ? t.evaluationGrad : `url(#start_grad_${f.value})` : e.value.style.chart.plots.evaluationColors.enable ? t.evaluationColor : e.value.style.chart.plots.startColor,
            stroke: e.value.style.chart.plots.stroke,
            "stroke-width": e.value.style.chart.plots.strokeWidth
          }, null, 8, ma)),
          [null, void 0].includes(t.end) ? g("", !0) : (n(), v("circle", {
            key: 2,
            cx: t.endX,
            cy: t.y,
            r: T.value,
            fill: e.value.style.chart.plots.gradient.show ? e.value.style.chart.plots.evaluationColors.enable ? t.evaluationGrad : `url(#end_grad_${f.value})` : e.value.style.chart.plots.evaluationColors.enable ? t.evaluationColor : e.value.style.chart.plots.endColor,
            stroke: e.value.style.chart.plots.stroke,
            "stroke-width": e.value.style.chart.plots.strokeWidth
          }, null, 8, ka))
        ]))), 128)),
        e.value.style.chart.labels.startLabels.show ? (n(), v("g", xa, [
          (n(!0), v(L, null, M(x.value, (t, a) => (n(), v("g", {
            key: `start_label_${a}_${t.id}`
          }, [
            [null, void 0].includes(t.start) ? g("", !0) : (n(), v("text", {
              key: 0,
              x: t.startX,
              y: t.y + T.value * 2 + e.value.style.chart.labels.startLabels.fontSize / 2,
              fill: e.value.style.chart.plots.evaluationColors.enable && e.value.style.chart.labels.startLabels.useEvaluationColor ? t.evaluationColor : e.value.style.chart.labels.startLabels.useStartColor ? e.value.style.chart.plots.startColor : e.value.style.chart.labels.startLabels.color,
              "font-size": e.value.style.chart.labels.startLabels.fontSize,
              "text-anchor": "middle"
            }, _(i(E)(
              e.value.style.chart.labels.formatter,
              t.start,
              i(A)({
                p: e.value.style.chart.labels.prefix,
                v: t.start,
                s: e.value.style.chart.labels.suffix,
                r: e.value.style.chart.labels.startLabels.rounding
              }),
              { datapoint: t, seriesIndex: a }
            )), 9, _a))
          ]))), 128))
        ])) : g("", !0),
        e.value.style.chart.labels.endLabels.show ? (n(), v("g", Ca, [
          (n(!0), v(L, null, M(x.value, (t, a) => (n(), v("g", {
            key: `end_label_${a}_${t.id}`
          }, [
            [null, void 0].includes(t.end) ? g("", !0) : (n(), v("text", {
              key: 0,
              x: t.endX,
              y: t.y - (T.value * 2 - e.value.style.chart.labels.startLabels.fontSize / 3),
              fill: e.value.style.chart.plots.evaluationColors.enable && e.value.style.chart.labels.endLabels.useEvaluationColor ? t.evaluationColor : e.value.style.chart.labels.endLabels.useEndColor ? e.value.style.chart.plots.endColor : e.value.style.chart.labels.endLabels.color,
              "font-size": e.value.style.chart.labels.endLabels.fontSize,
              "text-anchor": "middle"
            }, _(i(E)(
              e.value.style.chart.labels.formatter,
              t.end,
              i(A)({
                p: e.value.style.chart.labels.prefix,
                v: t.end,
                s: e.value.style.chart.labels.suffix,
                r: e.value.style.chart.labels.endLabels.rounding
              }),
              { datapoint: t, seriesIndex: a }
            )), 9, wa))
          ]))), 128))
        ])) : g("", !0),
        c("g", null, [
          (n(!0), v(L, null, M(x.value, (t, a) => (n(), v("rect", {
            x: o.value.left,
            y: o.value.top + a * Math.max(0.1, o.value.rowHeight),
            width: Math.max(0.1, o.value.width),
            height: Math.max(0.1, o.value.rowHeight),
            fill: D.value !== null && D.value === a ? i(pt)(e.value.style.chart.highlighter.color, e.value.style.chart.highlighter.opacity) : "transparent",
            onMouseenter: (u) => $e({ datapoint: t, seriesIndex: a }),
            onMouseleave: (u) => Le({ datapoint: t, seriesIndex: a }),
            onClick: (u) => Ae({ datapoint: t, seriesIndex: a })
          }, null, 40, $a))), 256))
        ]),
        m(l.$slots, "svg", {
          svg: {
            ...o.value,
            isPrintingImg: i(pe) | i(me) | ae.value,
            isPrintingSvg: se.value
          }
        }, void 0, !0)
      ], 14, Ll)),
      l.$slots.watermark ? (n(), v("div", La, [
        m(l.$slots, "watermark", w($({ isPrinting: i(pe) || i(me) || ae.value || se.value })), void 0, !0)
      ])) : g("", !0),
      c("div", {
        id: `legend-bottom-${f.value}`
      }, null, 8, Aa),
      Ge.value ? (n(), I(tl, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${f.value}` : `#legend-bottom-${f.value}`
      }, [
        c("div", {
          ref_key: "chartLegend",
          ref: Oe
        }, [
          e.value.style.chart.legend.show && le.value ? (n(), I(_l, {
            key: `legend_${He.value}`,
            legendSet: _e.value,
            config: Ft.value,
            clickable: !1
          }, {
            item: p(({ legend: t }) => [
              c("div", {
                style: Z(`display:flex;align-items:center;gap:4px;font-size:${e.value.style.chart.legend.fontSize}px`)
              }, [
                (n(), v("svg", {
                  xmlns: i(gt),
                  viewBox: "0 0 20 20",
                  height: e.value.style.chart.legend.fontSize,
                  width: e.value.style.chart.legend.fontSize
                }, [
                  c("circle", {
                    cx: 10,
                    cy: 10,
                    r: 9,
                    fill: t.color
                  }, null, 8, Na)
                ], 8, Sa)),
                i(he) ? g("", !0) : (n(), v(L, { key: 0 }, [
                  Ne(_(t.name), 1)
                ], 64))
              ], 4)
            ]),
            _: 1
          }, 8, ["legendSet", "config"])) : m(l.$slots, "legend", {
            key: 1,
            legend: _e.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : g("", !0),
      l.$slots.source ? (n(), v("div", {
        key: 6,
        ref_key: "source",
        ref: Pe,
        dir: "auto"
      }, [
        m(l.$slots, "source", {}, void 0, !0)
      ], 512)) : g("", !0),
      le.value && e.value.userOptions.buttons.table ? (n(), I(ll(K.value.component), al({ key: 7 }, K.value.props, {
        ref_key: "tableUnit",
        ref: q,
        onClose: lt
      }), ct({
        content: p(() => [
          (n(), I(i(wt), {
            key: `table_${Fe.value}`,
            colNames: ne.value.colNames,
            head: ne.value.head,
            body: ne.value.body,
            config: ne.value.config,
            title: e.value.table.useDialog ? "" : K.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: j.value,
            onClose: lt
          }, {
            th: p(({ th: t }) => [
              c("div", {
                innerHTML: t,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Ma)
            ]),
            td: p(({ td: t }) => [
              Ne(_(t.name || t), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: p(() => [
            Ne(_(K.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: p(() => [
            c("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: s[0] || (s[0] = (t) => Ce(e.value.userOptions.callbacks.csv)),
              style: Z({ cursor: j.value ? "pointer" : "default" })
            }, [
              vt(i(_t), {
                name: "fileCsv",
                stroke: K.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : g("", !0),
      m(l.$slots, "skeleton", {}, () => [
        i(he) ? (n(), I(hl, { key: 0 })) : g("", !0)
      ], !0)
    ], 46, wl));
  }
}, Ya = /* @__PURE__ */ Cl(Xa, [["__scopeId", "data-v-aeadd9b2"]]);
export {
  Ya as default
};
