// vue-data-ui/dist/vue-ui-stackline-2VKxwFei.js
import { defineAsyncComponent as pe, useSlots as is, computed as m, ref as f, shallowRef as Vt, onMounted as us, toRefs as cs, watch as Pe, onBeforeUnmount as Bt, nextTick as Yt, watchEffect as ht, openBlock as i, createElementBlock as v, normalizeStyle as de, normalizeClass as ja, createBlock as re, unref as C, withCtx as N, renderSlot as z, normalizeProps as Q, guardReactiveProps as ee, createCommentVNode as k, createElementVNode as he, createSlots as Wt, createVNode as Ge, Fragment as F, renderList as q, mergeProps as dt, toDisplayString as me, Teleport as vs, resolveDynamicComponent as hs, createTextVNode as Ja } from "vue";
import { j as Za, t as Ka, m as ds, p as mt, n as ms, o as Qa, q as el, r as fs, s as gs, aq as je, E as ft, D as gt, I as ys, c as be, b as xe, F as ps, T as bs, U as xs, k as Je, l as Se, G as Ze, X as ks, x as ws, u as Ss, K as tl, a0 as Ms, a1 as $s, z as Cs, A as Ke } from "./lib-GwPnJSrY.js";
import { t as Ls, u as As } from "./useResponsive-DfdjqQps.js";
import { u as al, a as ll, b as Ts } from "./useTimeLabels-C-A2uZKi.js";
import { u as Ns } from "./useConfig-CIgyKZ6R.js";
import { u as zs, B as Ps } from "./BaseScanner-BhhkKVoj.js";
import { u as Is } from "./usePrinter-DiK5R6k-.js";
import { u as _s } from "./useSvgExport-ByUukOZt.js";
import { u as Ht } from "./useNestedProp-BBT0mbHh.js";
import { u as Fs } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Es } from "./useStableElementSize-DCNs0Tx7.js";
import { u as Os } from "./useChartAccessibility-9icAAmYg.js";
import { u as Rs } from "./useTimeLabelCollider-CIsgDrl9.js";
import { u as Ds } from "./useUserOptionState-BIvW1Kz7.js";
import Vs from "./img-BYYzY_QU.js";
import { _ as Ut } from "./Shape-CSyWYgXc.js";
import Bs from "./Title-BfyrP-jp.js";
import Ys from "./vue_ui_stackline-QTh-nuxE.js";
import Ws from "./Legend-1LMMsPxL.js";
import { S as Hs } from "./SlicerPreview-I2J0UUF_.js";
import { B as Us } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as qs } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Xs = ["id"], Gs = ["id"], js = ["xmlns", "viewBox"], Js = { key: 0 }, Zs = ["x", "y", "width", "height"], Ks = { key: 1 }, Qs = ["id"], eo = ["stop-color"], to = ["stop-color"], ao = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], lo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], so = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], oo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], no = ["x", "y", "font-size", "fill", "font-weight"], ro = ["transform", "font-size", "fill", "font-weight"], io = ["x", "y", "width", "height", "stroke", "stroke-width", "stroke-linecap", "stroke-linejoin", "stroke-dasharray"], uo = ["d", "fill", "opacity"], co = ["d", "stroke", "stroke-width"], vo = ["x1", "x2", "y1", "y2", "stroke"], ho = ["x", "y", "font-size", "font-weight", "fill"], mo = { key: 0 }, fo = { key: 1 }, go = ["text-anchor", "font-size", "font-weight", "fill", "transform", "onClick"], yo = ["text-anchor", "font-size", "fill", "transform", "innerHTML", "onClick"], po = ["x", "y", "height", "width", "fill"], bo = { key: 0 }, xo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], ko = { key: 0 }, wo = ["x", "y", "font-size", "fill", "font-weight"], So = ["x", "y", "font-size", "font-weight", "fill"], Mo = ["data-start", "data-end"], $o = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Co = ["onClick"], Lo = ["innerHTML"], Ao = ["innerHTML"], To = ["id"], No = {
  __name: "vue-ui-stackline",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    selectedXIndex: {
      type: Number,
      default: void 0
    }
  },
  emits: ["selectDatapoint", "selectLegend", "selectTimeLabel", "selectX", "copyAlt"],
  setup(yt, { expose: sl, emit: ol }) {
    const nl = pe(() => import("./Tooltip-DL3QgjZf.js")), rl = pe(() => import("./BaseIcon-CNx8Tzl5.js")), il = pe(() => import("./vue-ui-accordion-BGHTqIBt.js")), ul = pe(() => import("./DataTable-Bo1pmAqG.js")), cl = pe(() => import("./PenAndPaper-DWV1FMIi.js")), vl = pe(() => import("./UserOptions-CrLqEimB.js")), hl = pe(() => import("./PackageVersion-DN9d0ME5.js")), dl = pe(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_stackline: ml } = Ns(), { isThemeValid: fl, warnInvalidTheme: gl } = Fs(), qt = is(), X = yt, Me = ol, Ie = m({
      get() {
        return !!X.dataset && X.dataset.length;
      },
      set(t) {
        return t;
      }
    }), ie = f(null), te = f(Za()), Xt = f(!1), Gt = f(null), D = f([]), jt = f(0), Jt = f(null), Zt = f(null), _e = f(null), Kt = f(null), Qt = f(null), Fe = f(!1), pt = f(!1), ea = f(0), yl = f(0), pl = f(0), ta = f(!1), Ee = f(null), bt = f(null), $e = f(!1), Qe = f(!1), et = f(!1), tt = f(null), xt = f(null), kt = f(null), at = f(null), bl = f(null), aa = Vt(null), la = f(!1), sa = f(0), oa = f(0), na = Es({
      elementRef: aa,
      minimumWidth: 2,
      minimumHeight: 2,
      stableFramesRequired: 2,
      once: !1,
      onSizeAccepted: () => {
        lt();
      }
    });
    function wt() {
      aa.value = ie.value?.parentNode ?? null;
    }
    function ra() {
      return new Promise((t) => {
        requestAnimationFrame(() => {
          requestAnimationFrame(t);
        });
      });
    }
    async function lt() {
      const t = ++oa.value;
      la.value = !1, await Yt(), await ra(), await ra(), t === oa.value && (sa.value += 1, la.value = !0);
    }
    const Oe = f(null), Re = f(!1);
    function St() {
      Re.value = !Re.value;
    }
    us(() => {
      wt(), na.start(), ta.value = !0, fa(), lt();
    });
    const e = f(Ct()), ue = m(() => e.value.userOptions.useCursorPointer), xl = m(() => Ka({
      defaultConfig: {
        userOptions: { show: !1 },
        useCssAnimation: !1,
        table: { show: !1 },
        tooltip: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            lines: {
              totalValues: { show: !1 },
              dataLabels: { show: !1 }
            },
            grid: {
              frame: {
                stroke: "#6A6A6A"
              },
              scale: {
                scaleMin: 0,
                scaleMax: 144
              },
              x: {
                axisColor: "#6A6A6A",
                linesColor: "#6A6A6A",
                axisName: { show: !1 },
                timeLabels: { show: !1 }
              },
              y: {
                axisColor: "#6A6A6A",
                linesColor: "#6A6A6A",
                axisName: { show: !1 },
                axisLabels: { show: !1 }
              }
            },
            legend: {
              backgroundColor: "transparent"
            },
            padding: {
              left: 24,
              right: 24,
              bottom: 12
            },
            zoom: {
              show: !1,
              startIndex: null,
              endIndex: null
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: ce, FINAL_DATASET: Mt, manualLoading: st } = zs({
      ...cs(X),
      FINAL_CONFIG: e,
      prepareConfig: Ct,
      callback: () => {
        Promise.resolve().then(async () => {
          await ut();
        });
      },
      skeletonDataset: X.config?.skeletonDataset ?? [
        {
          name: "",
          series: [3, 2, 1, 5, 13, 21, 8, 89, 34, 55],
          color: "#8A8A8A"
        },
        {
          name: "",
          series: [1, 2, 3, 5, 8, 13, 21, 34, 55, 89],
          color: "#CACACA"
        }
      ],
      skeletonConfig: Ka({
        defaultConfig: e.value,
        userConfig: xl.value
      })
    }), { userOptionsVisible: $t, setUserOptionsVisibility: ia, keepUserOptionState: ua } = Ds({ config: e.value }), { svgRef: Ce } = Os({ config: e.value.style.chart.title });
    function kl() {
      $e.value = !0, ia(!0);
    }
    function wl() {
      ia(!1), $e.value = !1;
    }
    function Ct() {
      const t = Ht({
        userConfig: X.config,
        defaultConfig: ml
      });
      let l = {};
      const a = t.theme;
      if (a)
        if (!fl.value(t))
          gl(t), l = t;
        else {
          const s = Ht({
            userConfig: Ys[a] || X.config,
            defaultConfig: t
          });
          l = {
            ...Ht({
              userConfig: X.config,
              defaultConfig: s
            }),
            customPalette: t.customPalette.length ? t.customPalette : ds[a] || mt
          };
        }
      else
        l = t;
      return l;
    }
    const ca = m(() => e.value.style.chart.lines.dataLabels.hideUnderValue !== null), va = m(() => e.value.style.chart.lines.dataLabels.hideUnderPercentage !== null);
    Pe(() => X.config, (t) => {
      ce.value || (e.value = Ct()), $t.value = !e.value.userOptions.showOnChartHover, fa(), ea.value += 1, yl.value += 1, pl.value += 1, G.value.dataLabels.show = e.value.style.chart.lines.dataLabels.show, G.value.showTable = e.value.table.show, G.value.showTooltip = e.value.style.chart.tooltip.show, se.value.width = e.value.style.chart.width, se.value.height = e.value.style.chart.height, se.value.paddingRatio = {
        top: e.value.style.chart.padding.top / e.value.style.chart.height,
        right: e.value.style.chart.padding.right / e.value.style.chart.width,
        bottom: e.value.style.chart.padding.bottom / e.value.style.chart.height,
        left: e.value.style.chart.padding.left / e.value.style.chart.width
      }, wt(), lt(), nt();
    }, { deep: !0 }), Pe(() => X.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (st.value = !1), ba(), wt(), lt();
    }, { deep: !0 });
    const G = f({
      dataLabels: {
        show: e.value.style.chart.lines.dataLabels.show
      },
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    Pe(e, () => {
      G.value = {
        dataLabels: {
          show: e.value.style.chart.lines.dataLabels.show
        },
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const { isPrinting: Lt, isImaging: At, generatePdf: ha, generateImage: da } = Is({
      elementId: `stackline_${te.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-stackline",
      options: e.value.userOptions.print
    }), Sl = m(() => e.value.userOptions.show && !e.value.style.chart.title.text), se = f({
      width: e.value.style.chart.width,
      height: e.value.style.chart.height,
      paddingRatio: {
        top: e.value.style.chart.padding.top / e.value.style.chart.height,
        right: e.value.style.chart.padding.right / e.value.style.chart.width,
        bottom: e.value.style.chart.padding.bottom / e.value.style.chart.height,
        left: e.value.style.chart.padding.left / e.value.style.chart.width
      }
    }), Ml = m(() => ms(e.value.customPalette)), fe = Vt(null), Le = Vt(null), ma = f(null), Tt = m(() => !!e.value.debug);
    function fa() {
      if (Qa(X.dataset) ? (el({
        componentName: "VueUiStackline",
        type: "dataset",
        debug: Tt.value
      }), st.value = !0) : X.dataset.forEach((t, l) => {
        fs({
          datasetObject: t,
          requiredAttributes: ["name", "series"]
        }).forEach((a) => {
          Ie.value = !1, el({
            componentName: "VueUiStackline",
            type: "datasetSerieAttribute",
            property: a,
            index: l,
            debug: Tt.value
          }), st.value = !0;
        });
      }), Qa(X.dataset) || (st.value = e.value.loading), setTimeout(() => {
        pt.value = !0;
      }, 10), e.value.responsive) {
        const t = Ls(() => {
          pt.value = !1;
          const { width: l, height: a } = As({
            chart: ie.value,
            noTitle: Kt.value,
            title: e.value.style.chart.title.text ? Jt.value : null,
            legend: e.value.style.chart.legend.show ? Zt.value : null,
            slicer: e.value.style.chart.zoom.show && ve.value > 6 ? _e.value.$el : null,
            source: Qt.value
          });
          requestAnimationFrame(() => {
            se.value.width = l, se.value.height = a - 12, clearTimeout(ma.value), ma.value = setTimeout(() => {
              pt.value = !0;
            }, 10);
          });
        });
        fe.value && (Le.value && fe.value.unobserve(Le.value), fe.value.disconnect()), fe.value = new ResizeObserver(t), Le.value = ie.value.parentNode, fe.value.observe(Le.value);
      }
      ut();
    }
    Bt(() => {
      na.stop(), fe.value && (Le.value && fe.value.unobserve(Le.value), fe.value.disconnect());
    });
    function $l() {
      let t = 0;
      kt.value && (t = Array.from(kt.value.querySelectorAll("text")).reduce((s, n) => {
        const o = n.getComputedTextLength();
        return o > s ? o : s;
      }, 0));
      const l = xt.value ? xt.value.getBoundingClientRect().width : 0;
      return t + l + (l ? 24 : 0);
    }
    const Nt = f(0), De = f(0), Ve = f(0);
    function Cl() {
      const t = at.value;
      if (!t) {
        De.value = 0;
        return;
      }
      try {
        const l = t.getBBox();
        De.value = Number.isFinite(l?.height) ? l.height : 0;
      } catch {
        De.value = 0;
      }
    }
    function Ll() {
      Ve.value && cancelAnimationFrame(Ve.value), Ve.value = requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          Cl();
        });
      });
    }
    Bt(() => {
      Ve.value && cancelAnimationFrame(Ve.value), De.value = 0, Nt.value = 0;
    });
    const ga = m(() => {
      let t = 0;
      if (tt.value)
        try {
          t = tt.value.getBBox().height;
        } catch {
          t = 0;
        }
      const l = e.value.style.chart.grid.x.timeLabels.show ? De.value : 0;
      return t + l;
    }), h = m(() => {
      sa.value;
      const { height: t, width: l } = se.value, { right: a } = se.value.paddingRatio;
      let s = e.value.style.chart.lines.totalValues.show && X.dataset && X.dataset.length > 1 ? e.value.style.chart.lines.totalValues.fontSize * 1.3 : 0, n = 0;
      e.value.style.chart.grid.y.axisLabels.show && (n = $l());
      const o = e.value.style.chart.padding.top + s, r = l - l * a - Nt.value, p = t - e.value.style.chart.padding.bottom - ga.value - s, w = e.value.style.chart.padding.left + n, A = l - w - l * a - Nt.value, S = t - o - e.value.style.chart.padding.bottom - ga.value - s;
      return {
        chartHeight: Math.max(0, t),
        chartWidth: Math.max(0, l),
        top: o,
        right: Math.max(0, r),
        bottom: Math.max(0, p),
        left: Math.max(0, w),
        width: Math.max(0, A),
        height: Math.max(0, S)
      };
    }), Al = m(() => {
      const { left: t, top: l, width: a, height: s } = h.value, n = c.value.start, o = c.value.end, r = Math.max(1, o - n), p = Math.max(0, Math.min(r, (oe.value.start ?? n) - n)), w = Math.max(0, Math.min(r, (oe.value.end ?? o) - n)), A = Math.max(0, w - p), S = {
        fill: e.value.style.chart.zoom.preview.fill,
        stroke: e.value.style.chart.zoom.preview.stroke,
        "stroke-width": e.value.style.chart.zoom.preview.strokeWidth,
        "stroke-dasharray": e.value.style.chart.zoom.preview.strokeDasharray,
        "stroke-linecap": "round",
        "stroke-linejoin": "round",
        style: {
          pointerEvents: "none",
          transition: "none !important",
          animation: "none !important"
        }
      }, P = a / r;
      return {
        x: t + p * P,
        y: l,
        width: A * P,
        height: s,
        ...S
      };
    }), J = m(() => Mt.value.map((t, l) => {
      const a = gs(t.color) || Ml.value[l] || mt[l] || mt[l % mt.length];
      return {
        ...t,
        shape: t.shape || "circle",
        standalone: !!t.standalone,
        // In distributed mode, all values are converted to positive
        series: JSON.parse(JSON.stringify(t.series)).map((s) => e.value.style.chart.lines.distributed ? Math.abs(s) : s),
        seriesSource: t.series,
        // Store signs to manage display of neg values in distributed mode
        signedSeries: t.series.map((s) => s >= 0 ? 1 : -1),
        absoluteIndex: l,
        id: Za(),
        color: a
      };
    })), ke = m(
      () => J.value.filter((t) => !D.value.includes(t.id) && !t.standalone)
    ), ot = m(
      () => J.value.filter((t) => !D.value.includes(t.id) && t.standalone)
    ), ve = m(() => {
      const t = Math.max(...J.value.filter((l) => !D.value.includes(l.id)).map((l) => l.series.length));
      return isFinite(t) ? t : Math.max(...J.value.map((l) => l.series.length));
    });
    function Tl(t) {
      Oe.value = t;
    }
    const c = f({
      start: 0,
      end: Math.max(...Mt.value.map((t) => t.series.length))
    }), oe = f({ start: 0, end: Math.max(...Mt.value.map((t) => t.series.length)) });
    function zt(t) {
      return (c.value.start ?? 0) + (t ?? 0);
    }
    const ya = m(() => e.value.style.chart.zoom.preview.enable && (oe.value.start !== c.value.start || oe.value.end !== c.value.end));
    function pa(t, l) {
      oe.value[t] = l;
    }
    function nt() {
      const t = ve.value;
      let l = Math.max(0, Math.min(c.value.start ?? 0, t - 1)), a = Math.max(l + 1, Math.min(c.value.end ?? t, t));
      (!Number.isFinite(l) || !Number.isFinite(a) || a <= l) && (l = 0, a = t), c.value.start = l, c.value.end = a, oe.value.start = l, oe.value.end = a, _e.value && (_e.value.setStartValue(l), _e.value.setEndValue(a));
    }
    const Be = f(null);
    function Nl() {
      return new Promise((t) => requestAnimationFrame(
        () => requestAnimationFrame(() => t())
      ));
    }
    Bt(() => {
      Be.value && cancelAnimationFrame(Be.value);
    });
    async function ba() {
      ut(), await Yt(), Be.value && cancelAnimationFrame(Be.value), Be.value = requestAnimationFrame(async () => {
        await Nl(), ut();
      });
    }
    const rt = m(() => Math.max(0, h.value.width / (c.value.end - c.value.start))), xa = m(() => je(
      ke.value.map((t) => ({ ...t, series: t.series.map((l) => l ?? 0) }))
    ).slice(c.value.start, c.value.end)), Pt = m(() => D.value.length === J.value.length), ka = m(() => {
      if (!e.value.style.chart.zoom.minimap.show) return [];
      const t = J.value.filter((n) => Pt.value ? !0 : !D.value.includes(n.id) && !n.standalone);
      if (t.length)
        return je(
          t.map((n) => ({
            ...n,
            series: (n.series || []).map((o) => o ?? 0)
          }))
        );
      const l = ot.value;
      if (!l.length) return [];
      const a = Math.max(...l.map((n) => n.series.length || 0));
      return Array.from(
        { length: a },
        (n, o) => l.reduce((r, p) => r + Math.abs(p.series[o] ?? 0), 0)
      );
    }), zl = m(() => {
      if (!e.value.style.chart.zoom.minimap.show) return [];
      const l = J.value.filter((o) => Pt.value ? !0 : !D.value.includes(o.id) && !o.standalone), a = ot.value, s = l.length ? [{
        name: "",
        series: ka.value,
        color: "#000000",
        isVisible: !0
      }] : [], n = a.map((o) => ({
        name: o.name || "",
        series: (o.series || []).map((r) => r ?? 0),
        color: o.color,
        isVisible: !0
      }));
      return s.concat(n);
    }), Pl = m(() => je(
      ke.value.filter((t) => !D.value.includes(t.id)).map((t) => ({
        ...t,
        series: t.series.map((l, a) => {
          const s = l ?? 0;
          return t.signedSeries[a] === -1 && s >= 0 ? -s : s;
        })
      }))
    ).slice(c.value.start, c.value.end)), Ae = m(() => {
      const t = ke.value.filter((l) => !D.value.includes(l.id));
      return {
        positive: je(t.map((l) => ({
          ...l,
          series: l.series.slice(c.value.start, c.value.end).map((a) => (a ?? 0) >= 0 ? a ?? 0 : 0)
        }))),
        negative: je(t.map((l) => ({
          ...l,
          series: l.series.slice(c.value.start, c.value.end).map((a) => (a ?? 0) < 0 ? a ?? 0 : 0)
        })))
      };
    }), Te = m(() => {
      const t = Math.max(0, c.value.end - c.value.start), l = Array(t).fill(0), a = Array(t).fill(0);
      return ot.value.forEach((s) => {
        for (let n = 0; n < t; n += 1) {
          const o = s.series[c.value.start + n] ?? 0;
          o > 0 && (l[n] = Math.max(l[n], o)), o < 0 && (a[n] = Math.min(a[n], o));
        }
      }), {
        positive: l,
        negative: a
      };
    }), wa = m(() => Pl.value.map((t, l) => ({
      value: t,
      sign: t >= 0 ? 1 : -1
    })));
    function Il() {
      return { y0: Ne.value?.[0]?.zero ?? h.value.bottom };
    }
    function _l(t) {
      const { y0: l } = Il(), a = e.value.style.chart.lines.totalValues, s = Math.max(2, a.fontSize * 1.3 + a.offsetY), n = (r) => Math.min(Math.max(r, h.value.top - e.value.style.chart.lines.totalValues.fontSize * 1.3), h.value.bottom + e.value.style.chart.lines.totalValues.fontSize * 2);
      if ((wa.value?.[t]?.value ?? 0) >= 0 || e.value.style.chart.lines.distributed) {
        let r = 1 / 0;
        for (const w of T.value || []) {
          const A = w?.series?.[t], S = w?.topY?.[t];
          (A ?? 0) > 0 && Number.isFinite(S) && S < r && (r = S);
        }
        const p = Number.isFinite(r) ? r : l;
        return n(p - s - e.value.style.chart.lines.totalValues.offsetY);
      } else {
        let r = -1 / 0;
        for (const w of T.value || []) {
          const A = w?.series?.[t], S = w?.topY?.[t];
          (A ?? 0) < 0 && Number.isFinite(S) && S > r && (r = S);
        }
        const p = Number.isFinite(r) ? r : l;
        return n(p + s + a.fontSize * 0.7 + e.value.style.chart.lines.totalValues.offsetY);
      }
    }
    const Ne = m(() => {
      const t = Math.max(...Ae.value.positive, 0), l = Math.min(...Ae.value.negative, 0), a = Math.max(...Te.value.positive, 0), s = Math.min(...Te.value.negative, 0), n = Math.max(t, a), o = Math.min(l, s), r = [-1 / 0, 1 / 0, NaN, void 0, null].includes(o) ? 0 : o, p = e.value.style.chart.grid.scale.scaleMin, w = e.value.style.chart.grid.scale.scaleMax, A = !e.value.style.chart.lines.distributed && (p !== null || w !== null), S = p !== null && !e.value.style.chart.lines.distributed ? p : r > 0 ? 0 : r, P = w !== null && !e.value.style.chart.lines.distributed ? w : n < 0 ? 0 : n, E = A ? ft(S, P, e.value.style.chart.grid.scale.ticks) : gt(S, P, e.value.style.chart.grid.scale.ticks), M = Math.abs(Number(E.min) || 0), u = (Number(E.max) || 0) + M, b = u === 0 || !Number.isFinite(u) ? 1 : u, V = Array.isArray(E.ticks) && E.ticks.length ? E.ticks : [0], g = h.value.bottom - h.value.height * (M / b);
      return V.map(($) => {
        const O = Number($) || 0, _ = h.value.bottom - h.value.height * ((O + M) / b);
        return {
          zero: g,
          y: _,
          x: K.value.left - 8,
          value: O
        };
      });
    }), ae = f([]), ze = f([]);
    let Sa = 0;
    ht(() => {
      const t = ++Sa;
      (async () => {
        const l = await al({
          values: e.value.style.chart.grid.x.timeLabels.values,
          maxDatapoints: ve.value,
          formatter: e.value.style.chart.grid.x.timeLabels.datetimeFormatter,
          start: c.value.start,
          end: c.value.end
        });
        t === Sa && (ae.value = l);
      })();
    });
    let Ma = 0;
    ht(() => {
      const t = ++Ma;
      (async () => {
        const l = await al({
          values: e.value.style.chart.grid.x.timeLabels.values,
          maxDatapoints: ve.value,
          formatter: e.value.style.chart.grid.x.timeLabels.datetimeFormatter,
          start: 0,
          end: ve.value
        });
        t === Ma && (ze.value = l);
      })();
    });
    const Fl = m(() => {
      const t = e.value.style.chart.grid.x.timeLabels.modulo;
      return ae.value.length ? Math.min(t, [...new Set(ae.value.map((l) => l.text))].length) : t;
    }), it = m(() => {
      const t = e.value.style.chart.grid.x.timeLabels, l = ae.value || [], a = ze.value || [], s = c.value.start ?? 0, n = x.value, o = ve.value, r = l.map((w) => w?.text ?? ""), p = a.map((w) => w?.text ?? "");
      return ys(
        !!t.showOnlyFirstAndLast,
        !!t.showOnlyAtModulo,
        Math.max(1, Fl.value || 1),
        r,
        p,
        s,
        n,
        o
      );
    });
    ht(() => {
      e.value.style.chart.grid.x.timeLabels.show, e.value.style.chart.grid.x.timeLabels.fontSize, e.value.style.chart.grid.x.timeLabels.rotation, e.value.style.chart.grid.x.timeLabels.offsetY, (it.value || []).map((t) => t?.text ?? "").join("|"), se.value.width, se.value.height, at.value, tt.value, Ll();
    }, { flush: "post" });
    const $a = f({ months: [], shortMonths: [], days: [], shortDays: [] });
    let Ca = 0;
    ht(() => {
      const t = ++Ca, l = e.value.style.chart.grid.x.timeLabels.datetimeFormatter;
      (async () => {
        const a = await ll(l.locale).catch(() => ll("en"));
        t === Ca && ($a.value = a.data);
      })();
    });
    const La = m(() => {
      const t = e.value.style.chart.grid.x.timeLabels.datetimeFormatter, l = Ts({
        useUTC: t.useUTC,
        locale: $a.value,
        januaryAsYear: t.januaryAsYear
      });
      return (a, s) => {
        const o = e.value.style.chart.grid.x.timeLabels.values?.[a];
        return o == null ? "" : l.formatDate(new Date(o), s);
      };
    }), Aa = m(() => (e.value.style.chart.grid.x.timeLabels.values || []).map((l, a) => ({
      text: La.value(a, e.value.style.chart.zoom.timeFormat),
      absoluteIndex: a
    }))), It = m(() => (e.value.style.chart.grid.x.timeLabels.values || []).map((l, a) => ({
      text: La.value(a, e.value.style.chart.tooltip.timeFormat),
      absoluteIndex: a
    }))), El = m(() => {
      if (!Ie.value && !ce.value) return [];
      const t = h.value.height, l = Ne.value[0] ? Ne.value[0].zero : h.value.bottom, a = c.value.start ?? 0, s = c.value.end ?? 0, n = Math.max(1, s - a), o = Math.max(...Ae.value.positive, 0), r = Math.min(...Ae.value.negative, 0), p = Math.max(...Te.value.positive, 0), w = Math.min(...Te.value.negative, 0), A = Math.max(o, p), S = Math.min(r, w), P = !e.value.style.chart.lines.distributed && (e.value.style.chart.grid.scale.scaleMax !== null || e.value.style.chart.grid.scale.scaleMin !== null) ? ft(
        e.value.style.chart.grid.scale.scaleMin !== null ? e.value.style.chart.grid.scale.scaleMin : S > 0 ? 0 : S,
        e.value.style.chart.grid.scale.scaleMax !== null ? e.value.style.chart.grid.scale.scaleMax : A < 0 ? 0 : A,
        e.value.style.chart.grid.scale.ticks
      ) : gt(
        e.value.style.chart.grid.scale.scaleMin !== null ? e.value.style.chart.grid.scale.scaleMin : S > 0 ? 0 : S,
        e.value.style.chart.grid.scale.scaleMax !== null ? e.value.style.chart.grid.scale.scaleMax : A < 0 ? 0 : A,
        e.value.style.chart.grid.scale.ticks
      ), { min: E, max: M } = P, d = M + (E >= 0 ? 0 : Math.abs(E)) || 1, u = Array(n).fill(0), b = Array(n).fill(0), V = (g) => n <= 1 ? h.value.left + h.value.width / 2 : h.value.left + g / (n - 1) * h.value.width;
      return J.value.filter((g) => !D.value.includes(g.id)).map((g) => {
        const $ = [], O = [], _ = [], I = new Array(n).fill(null), B = new Array(n).fill(null);
        for (let y = 0; y < n; y += 1) {
          const L = a + y, U = g.series?.[L], H = g.signedSeries?.[L], ye = U == null || Number.isNaN(U) ? 0 : U, Ga = g.standalone ? ye / d : e.value.style.chart.lines.distributed ? ye / (xa.value[y] || 1) : ye / d;
          if (ye >= 0) {
            const Ue = t * Math.abs(Ga);
            if (g.standalone)
              I[y] = l, B[y] = l - Ue;
            else {
              const qe = u[y], Xe = qe + Ue;
              I[y] = l - qe, B[y] = l - Xe, u[y] = Xe;
            }
          } else {
            const Ue = t * Math.abs(Ga);
            if (g.standalone)
              I[y] = l, B[y] = l + Ue;
            else {
              const qe = b[y], Xe = qe + Ue;
              I[y] = l + qe, B[y] = l + Xe, b[y] = Xe;
            }
          }
          $.push(y), O.push(ye), _.push(H ?? (ye >= 0 ? 1 : -1));
        }
        const Y = $.map((y) => ({ x: V(y), y: B[y] })), Z = Y.map((y) => y.x), j = O.reduce((y, L) => y + Math.abs(L || 0), 0), W = j === 0 ? 1 : j, R = O.map((y, L) => {
          if (e.value.style.chart.lines.distributed && !g.standalone) {
            const U = $[L], H = xa.value[U] || 1;
            return (y || 0) / H;
          }
          return (y || 0) / W;
        });
        return {
          ...g,
          x: Z,
          points: Y,
          baseY: $.map((y) => I[y]),
          topY: $.map((y) => B[y]),
          series: O,
          signedSeries: _,
          proportions: R,
          rel: $,
          fullSeries: Array.isArray(g.fullSeries) ? g.fullSeries : g.series
        };
      });
    }), T = m(() => {
      const t = (M) => typeof M == "string" ? M.replace(
        /^M\s*[-+]?[\d.]+(?:e[-+]?\d+)?\s*,?\s*[-+]?[\d.]+(?:e[-+]?\d+)?\s*/i,
        ""
      ).trim() : "", l = Math.max(...Ae.value.positive, 0), a = Math.min(...Ae.value.negative, 0), s = Math.max(...Te.value.positive, 0), n = Math.min(...Te.value.negative, 0), o = Math.max(l, s), r = Math.min(a, n), p = !e.value.style.chart.lines.distributed && (e.value.style.chart.grid.scale.scaleMax !== null || e.value.style.chart.grid.scale.scaleMin !== null) ? ft(
        e.value.style.chart.grid.scale.scaleMin !== null ? e.value.style.chart.grid.scale.scaleMin : r > 0 ? 0 : r,
        e.value.style.chart.grid.scale.scaleMax !== null ? e.value.style.chart.grid.scale.scaleMax : o < 0 ? 0 : o,
        e.value.style.chart.grid.scale.ticks
      ) : gt(
        e.value.style.chart.grid.scale.scaleMin !== null ? e.value.style.chart.grid.scale.scaleMin : r > 0 ? 0 : r,
        e.value.style.chart.grid.scale.scaleMax !== null ? e.value.style.chart.grid.scale.scaleMax : o < 0 ? 0 : o,
        e.value.style.chart.grid.scale.ticks
      ), w = Number(p.min) || 0, A = Number(p.max) || 0, S = Math.abs(w), P = A + S || 1, E = (M) => {
        if (e.value.style.chart.lines.distributed) {
          const d = Math.max(0, Math.min(1, M));
          return minimapH - minimapH * d;
        }
        return clampY(minimapH - minimapH * (((M ?? 0) + S) / P));
      };
      return El.value.map((M) => {
        const d = M.x.length, u = Array.isArray(M.fullSeries) ? M.fullSeries : M.series, b = ({ left: W = 0, unitW: R }) => {
          const y = u.length;
          return !Number.isFinite(R) || R <= 0 || y <= 0 ? [] : y === 1 ? [W + R * 0.5] : u.map((L, U) => W + U * R);
        }, V = ({ minimapH: W }) => !Number.isFinite(W) || W <= 0 ? [] : u.map((R) => E(R || 0)), g = ({ minimapH: W }) => {
          if (!Number.isFinite(W) || W <= 0) return [];
          const R = E(0);
          return u.map(() => R);
        };
        if (d === 0)
          return {
            ...M,
            points: [],
            smoothPath: "",
            straightPath: "",
            smoothArea: "",
            straightArea: "",
            xMinimap: b,
            yMinimap: V,
            yMinimapBase: g
          };
        const $ = M.x.map((W, R) => ({ x: W, y: M.topY[R] })), O = M.x.map((W, R) => ({ x: W, y: M.baseY[R] })), _ = d >= 2 ? be($) : `M${$[0].x},${$[0].y}`, I = d >= 2 ? xe($) : `M${$[0].x},${$[0].y}`, B = d >= 2 ? `M${t(_)}` : _, Y = d >= 2 ? `M${t(I)}` : I;
        let Z = "", j = "";
        if (e.value.style.chart.lines.useArea && d >= 2) {
          const W = be([...O].reverse()), R = xe([...O].reverse()), y = O[O.length - 1];
          Z = `M${$[0].x},${$[0].y} ${t(_)} L${y.x},${y.y} ${t(W)} Z`, j = `M${$[0].x},${$[0].y} ${t(I)} L${y.x},${y.y} ${t(R)} Z`;
        }
        return {
          ...M,
          points: $,
          smoothPath: B,
          straightPath: Y,
          smoothArea: Z,
          straightArea: j,
          xMinimap: b,
          yMinimap: V,
          yMinimapBase: g
        };
      });
    }), Ye = f(!1), Ta = f(!1);
    function Na(t) {
      const l = ve.value;
      return t > l ? l : t < 0 || t < c.value.start ? e.value.style.chart.zoom.startIndex !== null ? c.value.start + 1 : 1 : t;
    }
    function ut() {
      if (!Ye.value) {
        Ye.value = !0;
        try {
          const { startIndex: t, endIndex: l } = e.value.style.chart.zoom, a = ve.value, s = t ?? 0, n = l != null ? Math.min(Na(l + 1), a) : a;
          ct.value = !0, c.value.start = s, c.value.end = n, oe.value.start = s, oe.value.end = n, nt(), Ta.value = !0;
        } finally {
          queueMicrotask(() => {
            ct.value = !1;
          }), Ye.value = !1;
        }
      }
    }
    function Ol(t) {
      Ye.value || ct.value || t !== c.value.start && (c.value.start = t, oe.value.start = t, nt());
    }
    function Rl(t) {
      if (Ye.value || ct.value) return;
      const l = Na(t);
      l !== c.value.end && (c.value.end = l, oe.value.end = l, nt());
    }
    const K = m(() => ({
      left: h.value.left,
      right: h.value.right,
      width: h.value.width
    })), ct = f(!1), Dl = m(() => se.value.width), Vl = m(() => se.value.height);
    Rs({
      timeLabelsEls: at,
      timeLabels: ae,
      slicer: c,
      configRef: e,
      rotationPath: ["style", "chart", "grid", "x", "timeLabels", "rotation"],
      autoRotatePath: ["style", "chart", "grid", "x", "timeLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: Dl,
      height: Vl,
      rotation: e.value.style.chart.grid.x.timeLabels.autoRotate.angle
    });
    const _t = f(null);
    function za(t) {
      const l = Ce.value;
      if (!l) return null;
      if (l.createSVGPoint && l.getScreenCTM) {
        const P = l.createSVGPoint();
        P.x = t.clientX, P.y = t.clientY;
        const E = l.getScreenCTM();
        if (E) {
          const M = P.matrixTransform(E.inverse());
          return { x: M.x, y: M.y, ok: !0 };
        }
      }
      const a = l.getBoundingClientRect(), s = l.viewBox?.baseVal || { x: 0, y: 0, width: a.width, height: a.height }, n = Math.min(a.width / s.width, a.height / s.height), o = s.width * n, r = s.height * n, p = (a.width - o) / 2, w = (a.height - r) / 2, A = (t.clientX - a.left - p) / n + s.x, S = (t.clientY - a.top - w) / n + s.y;
      return { x: A, y: S, ok: !0 };
    }
    let we = 0;
    const le = m(() => Math.max(1, c.value.end - c.value.start)), Ft = m(
      () => le.value > 1 ? h.value.width / (le.value - 1) : 0
    );
    function ne(t) {
      return le.value <= 1 ? h.value.left + h.value.width / 2 : h.value.left + t * Ft.value;
    }
    function Pa(t) {
      if (le.value <= 1)
        return { x: h.value.left, width: h.value.width };
      const l = t === 0 ? h.value.left : (ne(t - 1) + ne(t)) / 2, a = t === le.value - 1 ? h.value.left + h.value.width : (ne(t) + ne(t + 1)) / 2;
      return { x: l, width: Math.max(0, a - l) };
    }
    function Bl(t) {
      Re.value || (we && cancelAnimationFrame(we), we = requestAnimationFrame(() => {
        we = 0;
        const l = za(t);
        if (!l || !Ce.value) {
          Et();
          return;
        }
        const { left: a, right: s, top: n, bottom: o } = h.value;
        if (l.x < a || l.x > s || l.y < n || l.y > o) {
          Et();
          return;
        }
        let r = 0;
        if (le.value > 1) {
          const p = (l.x - a) / Ft.value;
          r = Math.round(p);
        } else
          r = 0;
        r < 0 && (r = 0), r > le.value - 1 && (r = le.value - 1), _t.value !== r && (_t.value = r, Ia(!0, r));
      }));
    }
    function Yl(t) {
      const l = za(t);
      if (!l || !Ce.value) return;
      const { left: a, right: s, top: n, bottom: o } = h.value;
      if (l.x < a || l.x > s || l.y < n || l.y > o) return;
      let r = 0;
      if (le.value > 1) {
        const p = (l.x - a) / Ft.value;
        r = Math.round(p);
      } else
        r = 0;
      r < 0 && (r = 0), r > le.value - 1 && (r = le.value - 1), qa({
        seriesIndex: r,
        datapoint: Wa(r)
      }), Wl(r);
    }
    function Wl(t) {
      const l = JSON.parse(JSON.stringify(T.value)).map((a) => ({
        name: a.name,
        value: a.series[t] === 0 ? 0 : a.series[t] || null,
        proportion: a.proportions[t] || null,
        color: a.color,
        id: a.id
      }));
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: t + c.value.start }), Me("selectDatapoint", { datapoint: l, period: ae.value[t] });
    }
    Pe(
      () => [c.value.start, c.value.end, T.value.length],
      () => {
        const t = le.value;
        if (x.value != null) {
          if (t <= 0) {
            x.value = null;
            return;
          }
          x.value < 0 && (x.value = 0), x.value > t - 1 && (x.value = t - 1);
        }
      }
    );
    function Et() {
      we && (cancelAnimationFrame(we), we = 0), _t.value = null, Ia(!1, null);
    }
    const x = f(null), Ot = m(() => {
      const t = x.value, l = t == null ? null : zt(t);
      return T.value.map((a) => ({
        slotAbsoluteIndex: a.absoluteIndex,
        shape: a.shape || "circle",
        name: a.name,
        color: a.color,
        value: t == null ? null : a.series.find((s, n) => n === t),
        sourceValue: l == null ? null : a.seriesSource?.[l],
        comments: a.comments || [],
        id: a.id,
        standalone: !!a.standalone
      }));
    }), Hl = m(() => {
      const t = e.value.style.chart.tooltip.customFormat, l = [...Ot.value].reverse(), a = l.filter((u) => !u.standalone), s = l.filter((u) => u.standalone), n = a.map((u) => u.value).filter((u) => ps(u) && u !== null).reduce((u, b) => Math.abs(u) + Math.abs(b), 0);
      if (bs(t) && xs(
        () => t({
          absoluteIndex: x.value + c.value.start,
          seriesIndex: x.value,
          datapoint: Ot.value,
          series: J.value,
          config: e.value
        })
      ))
        return t({
          absoluteIndex: x.value + c.value.start,
          seriesIndex: x.value,
          datapoint: Ot.value,
          series: J.value,
          config: e.value
        });
      const {
        showValue: o,
        showTotal: r,
        totalTranslation: p,
        showPercentage: w,
        borderColor: A,
        roundingValue: S,
        roundingPercentage: P
      } = e.value.style.chart.tooltip, E = (u) => !u.shape || !["star", "triangle", "square", "diamond", "pentagon", "hexagon"].includes(u.shape) ? `<svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" fill="${u.color}" /></svg>` : u.shape === "star" ? `<svg viewBox="0 0 12 12" height="14" width="14"><polygon stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" fill="${u.color}" points="${Cs({ plot: { x: 6, y: 6 }, radius: 5 })}" /></svg>` : u.shape === "triangle" ? `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Ke({ plot: { x: 6, y: 6 }, radius: 6, sides: 3, rotation: 0.52 }).path}" fill="${u.color}" stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>` : u.shape === "square" ? `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Ke({ plot: { x: 6, y: 6 }, radius: 6, sides: 4, rotation: 0.8 }).path}" fill="${u.color}" stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>` : u.shape === "diamond" ? `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Ke({ plot: { x: 6, y: 6 }, radius: 5, sides: 4, rotation: 0 }).path}" fill="${u.color}" stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>` : u.shape === "pentagon" ? `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Ke({ plot: { x: 6, y: 6 }, radius: 5, sides: 5, rotation: 0.95 }).path}" fill="${u.color}" stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>` : `<svg viewBox="0 0 12 12" height="14" width="14"><path d="${Ke({ plot: { x: 6, y: 6 }, radius: 5, sides: 6, rotation: 0 }).path}" fill="${u.color}" stroke="${e.value.style.chart.tooltip.backgroundColor}" stroke-width="1" /></svg>`, M = (u, b) => {
        const V = b ? Se({
          v: isNaN((u.value ?? 0) / (n || 1)) ? 0 : Math.abs(u.value ?? 0) / (n || 1) * 100,
          s: "%",
          r: P
        }) : "", g = o && b ? "(" : "", $ = o && b ? ")" : "";
        return `
        <div style="display:flex;flex-direction:row;align-items:center;gap:4px">
            <div style="width:20px;height:20px;display:flex;align-items:center;justify-content:center;">${E(u)}</div>
            ${u.name}${o || b ? ":" : ""} 
            ${o ? Je(
          e.value.style.chart.lines.dataLabels.formatter,
          u.sourceValue,
          Se({
            p: e.value.style.chart.lines.dataLabels.prefix,
            v: u.sourceValue,
            s: e.value.style.chart.lines.dataLabels.suffix,
            r: S
          })
        ) : ""} ${g}${V}${$}
        </div>
        `;
      };
      let d = "";
      return (ae.value[x.value] && ae.value[x.value].text || It.value[x.value] && It.value[x.value].text && e.value.style.chart.tooltip.showTimeLabel) && (d += `<div style="width:100%;text-align:center;border-bottom:1px solid ${A};padding-bottom:6px;margin-bottom:3px;">${e.value.style.chart.tooltip.useDefaultTimeFormat ? ae.value[x.value]?.text : It.value[x.value]?.text || ze.value[x.value]?.text || ""}</div>`), r && a.length > 1 && (d += `<div class="vue-data-ui-tooltip-total" style="display:flex;flex-direction:row;align-items:center;gap:4px">
        <span>${p}:</span>
        <span>${Je(
        e.value.style.chart.lines.dataLabels.formatter,
        Ze(n),
        Se({
          p: e.value.style.chart.lines.dataLabels.prefix,
          v: Ze(n),
          s: e.value.style.chart.lines.dataLabels.suffix,
          r: S
        }),
        { datapoint: { name: p, value: Ze(n) } }
      )}</span>
        </div>`), a.forEach((u) => {
        d += M(u, w);
      }), s.length && (d += `<div style="border-top:1px solid ${ke.value.length ? A : "transparent"}; margin:${ke.value.length ? "6px 0" : "0"};"></div>`, s.forEach((u) => {
        d += M(u, !1);
      })), `<div>${d}</div>`;
    });
    function Ia(t, l = null) {
      if (Pt.value) return;
      Xt.value = t;
      const a = T.value.map((s) => ({
        name: s.name,
        value: [null, void 0, NaN].includes(s.series[l]) ? null : s.series[l],
        color: s.color
      }));
      t ? (x.value = l, qa({ seriesIndex: l, datapoint: a }), e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: a, seriesIndex: l + c.value.start })) : (x.value = null, Me("selectX", { seriesIndex: null, dataset: null, indexLabel: null }), e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: a, seriesIndex: l + c.value.start }));
    }
    function _a(t, l) {
      const a = JSON.parse(JSON.stringify(T.value)).map((s) => ({
        name: s.name,
        value: s.series[l] === 0 ? 0 : (s.signedSeries[l] === -1 && s.series[l] >= 0 ? -s.series[l] : s.series[l]) || null,
        proportion: s.proportions[l] || null,
        color: s.color,
        id: s.id
      }));
      Me("selectTimeLabel", {
        datapoint: a,
        absoluteIndex: t.absoluteIndex,
        label: t.text
      });
    }
    function Ul() {
      D.value.length ? D.value = [] : We.value.forEach((t) => {
        D.value.push(t.id);
      });
    }
    function Rt(t) {
      if (D.value.includes(t.id))
        D.value = D.value.filter((l) => l !== t.id);
      else {
        if (D.value.length === J.value.length - 1) return;
        D.value.push(t.id);
      }
      Me("selectLegend", T.value);
    }
    const We = m(() => J.value.map((t) => ({
      ...t,
      opacity: D.value.includes(t.id) ? 0.5 : 1,
      segregate: () => Rt(t),
      isSegregated: D.value.includes(t.id)
    }))), ql = m(() => ({
      cy: "stackline-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    }));
    function Fa(t) {
      if (!J.value.length)
        return e.value.debug && console.warn("VueUiStackline - There are no series to show."), null;
      const l = J.value.find((a) => a.name === t);
      return l || (e.value.debug && console.warn(`VueUiStackline - Series name not found "${t}"`), null);
    }
    function Xl(t) {
      const l = Fa(t);
      l !== null && D.value.includes(l.id) && Rt({ id: l.id });
    }
    function Gl(t) {
      const l = Fa(t);
      l !== null && (D.value.includes(l.id) || Rt({ id: l.id }));
    }
    const Ea = m(() => {
      if (T.value.length === 0)
        return { head: [], body: [], config: {}, columnNames: [] };
      const t = T.value.map(({ name: a, color: s }) => ({ label: a, color: s })), l = [];
      return it.value.forEach((a) => {
        const s = a.absoluteIndex, o = [e.value.style.chart.grid.x.timeLabels.values?.[s] ? a.text : s + 1];
        J.value.forEach((r) => {
          const p = r.series?.[s], w = Number((p ?? 0).toFixed(e.value.table.td.roundingValue));
          o.push(w);
        }), l.push(o);
      }), { head: t, body: l };
    });
    function Dt(t = null) {
      const l = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], a = ["", ...Ea.value.head.map((r) => r.label)], s = Ea.value.body, n = l.concat([a]).concat(s), o = Ms(n);
      t ? t(o) : $s({ csvContent: o, title: e.value.style.chart.title.text || "vue-ui-stackline" });
    }
    const vt = m(() => {
      const t = [""].concat(T.value.map((o) => o.name)).concat(' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>'), l = [], a = Math.max(0, c.value.end - c.value.start);
      for (let o = 0; o < a; o += 1) {
        const r = zt(o), p = e.value.style.chart.grid.x.timeLabels.values?.[r] ? ae.value?.[o]?.text ?? r + 1 : r + 1, w = T.value.map((P) => {
          const E = P.series?.[o] ?? 0;
          return Number(E.toFixed(e.value.table.td.roundingValue));
        }), S = T.value.filter((P) => !P.standalone).map((P) => P.series?.[o] ?? 0).reduce((P, E) => P + E, 0);
        l.push([p].concat(w).concat(Number(S.toFixed(e.value.table.td.roundingValue))));
      }
      const s = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, n = [e.value.table.columnNames.period].concat(T.value.map((o) => o.name)).concat(e.value.table.columnNames.total);
      return { head: t, body: l, config: s, colNames: n };
    }), jl = m(() => e.value.style.chart.backgroundColor), Jl = m(() => e.value.style.chart.legend), Zl = m(() => e.value.style.chart.title), { exportSvg: Kl, getSvg: Ql } = _s({
      svg: Ce,
      title: Zl,
      legend: Jl,
      legendItems: We,
      backgroundColor: jl
    });
    async function Oa({ isCb: t }) {
      et.value = !0, await Yt();
      try {
        if (t) {
          const { blob: l, url: a, text: s, dataUrl: n } = await Ql();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: a, text: s, dataUrl: n }));
        } else
          await Promise.resolve(Kl());
      } finally {
        et.value = !1;
      }
    }
    async function es({ scale: t = 2 } = {}) {
      if (!ie.value) return;
      const { width: l, height: a } = ie.value.getBoundingClientRect(), s = l / a, { imageUri: n, base64: o } = await Vs({ domElement: ie.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: n,
        base64: o,
        title: e.value.style.chart.title.text,
        width: l,
        height: a,
        aspectRatio: s
      };
    }
    const He = m(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = G.value.showTable;
      return {
        component: t ? dl : il,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: Fe.value,
          fullscreenParent: ie.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: ue.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    Pe(() => G.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && Ee.value ? Ee.value.open() : "close" in Ee.value && Ee.value.close());
    });
    function Ra() {
      G.value.showTable = !1, bt.value && bt.value.setTableIconState(!1);
    }
    function Da(t) {
      Fe.value = t, jt.value += 1;
    }
    function Va() {
      G.value.showTable = !G.value.showTable;
    }
    function Ba() {
      G.value.dataLabels.show = !G.value.dataLabels.show;
    }
    function Ya() {
      G.value.showTooltip = !G.value.showTooltip;
    }
    function ts() {
      return T.value;
    }
    function Wa(t) {
      const l = t ?? 0, a = zt(l);
      return JSON.parse(JSON.stringify(T.value)).map((s) => {
        const n = s.series[l], o = s.signedSeries?.[l], r = n === 0 ? 0 : (o === -1 && n >= 0 ? -n : n) || null;
        return {
          name: s.name,
          absoluteIndex: s.absoluteIndex,
          value: r,
          proportion: s.proportions?.[l] || null,
          color: s.color,
          id: s.id,
          timeLabel: ze.value?.[a] || null
        };
      });
    }
    const as = m(() => Array(ve.value).fill(0).map((t, l) => Wa(l)));
    Pe(() => X.selectedXIndex, (t) => {
      if ([null, void 0].includes(X.selectedXIndex)) {
        x.value = null;
        return;
      }
      const l = t - c.value.start;
      l < 0 || t >= c.value.end ? x.value = null : x.value = l ?? null;
    }, { immediate: !0 });
    const Ha = m(() => {
      if (e.value.style.chart.lines.distributed)
        return { min: -1, max: 1 };
      const t = ke.value, l = ot.value, a = Math.max(
        1,
        ...[...t, ...l].map((d) => d.series?.length || 0)
      ), s = Array(a).fill(0), n = Array(a).fill(0);
      t.forEach((d) => {
        for (let u = 0; u < a; u += 1) {
          const b = d.series[u] ?? 0;
          b >= 0 ? s[u] += b : n[u] += b;
        }
      });
      const o = Array(a).fill(0), r = Array(a).fill(0);
      l.forEach((d) => {
        for (let u = 0; u < a; u += 1) {
          const b = d.series[u] ?? 0;
          b > 0 && (o[u] = Math.max(o[u], b)), b < 0 && (r[u] = Math.min(r[u], b));
        }
      });
      const p = Math.max(0, ...s, ...o), w = Math.min(0, ...n, ...r), A = e.value.style.chart.grid.scale.scaleMin, S = e.value.style.chart.grid.scale.scaleMax, P = e.value.style.chart.grid.scale.ticks, M = !e.value.style.chart.lines.distributed && (A !== null || S !== null) ? ft(
        A !== null ? A : w > 0 ? 0 : w,
        S !== null ? S : p < 0 ? 0 : p,
        P
      ) : gt(
        A !== null ? A : w > 0 ? 0 : w,
        S !== null ? S : p < 0 ? 0 : p,
        P
      );
      return { min: Number(M.min) || 0, max: Number(M.max) || 0 };
    });
    function ls({ minimapH: t, unitW: l }) {
      const a = (d) => typeof d == "string" ? d.replace(/^M\s*[-+]?[\d.]+(?:e[-+]?\d+)?\s*,?\s*[-+]?[\d.]+(?:e[-+]?\d+)?\s*/i, "").trim() : "";
      if (!Number.isFinite(t) || t <= 0 || !Number.isFinite(l) || l <= 0)
        return "";
      const s = T.value.filter((d) => !d.standalone), n = T.value.filter((d) => d.standalone), o = Math.max(0, ...T.value.map((d) => Array.isArray(d.fullSeries) ? d.fullSeries.length : d.series.length));
      if (o <= 0) return "";
      const r = o === 1 ? [l * 0.5] : Array.from({ length: o }, (d, u) => u * l);
      if (e.value.style.chart.lines.distributed) {
        const d = Array.from({ length: o }, (g, $) => {
          let O = 0;
          return s.forEach((_) => {
            const B = (Array.isArray(_.fullSeries) ? _.fullSeries : _.series)?.[$];
            B != null && !Number.isNaN(B) && (O += Math.abs(B));
          }), O || 1;
        }), u = (g) => {
          const $ = Math.max(0, Math.min(1, g || 0));
          return t - t * $;
        }, b = [], V = Array(o).fill(0);
        return s.forEach((g) => {
          const $ = Array.isArray(g.fullSeries) ? g.fullSeries : g.series, O = new Array(o), _ = new Array(o);
          for (let H = 0; H < o; H += 1) {
            const ge = $?.[H], ye = ge == null || Number.isNaN(ge) ? 0 : Math.abs(ge) / d[H];
            _[H] = V[H], V[H] += ye, O[H] = V[H];
          }
          const I = O.map(u), B = _.map(u), Y = r.map((H, ge) => ({ x: H, y: I[ge] })), Z = r.map((H, ge) => ({ x: H, y: B[ge] })), j = be(Y), W = be([...Z].reverse()), R = xe(Y), y = xe([...Z].reverse()), L = Z[Z.length - 1];
          if (e.value.style.chart.lines.useArea) {
            const H = e.value.style.chart.lines.smooth ? `M${Y[0].x},${Y[0].y} ${a(j)} L${L.x},${L.y} ${a(W)} Z` : `M${Y[0].x},${Y[0].y} ${a(R)} L${L.x},${L.y} ${a(y)} Z`;
            b.push(
              `<path d="${H}"
                    fill="${qt.pattern ? `url(#pattern_${te.value}_${g.absoluteIndex})` : (e.value.style.chart.lines.gradient.show, g.color)}"
                    opacity="${e.value.style.chart.lines.areaOpacity}"
                    stroke="none" />`
            );
          }
          const U = e.value.style.chart.lines.smooth ? `M${Y[0].x},${Y[0].y} ${a(j)}` : `M${Y[0].x},${Y[0].y} ${a(R)}`;
          b.push(
            `<path d="${U}"
                fill="none"
                stroke="${g.color}"
                stroke-width="${e.value.style.chart.lines.strokeWidth}"
                stroke-linecap="round" />`
          );
        }), n.length && n.forEach((g) => {
          const $ = Array.isArray(g.fullSeries) ? g.fullSeries : g.series, O = Array.from({ length: o }, (Z, j) => u(Math.abs($?.[j] ?? 0) / d[j])), _ = r.map((Z, j) => ({ x: Z, y: O[j] })), I = be(_), B = xe(_), Y = e.value.style.chart.lines.smooth ? `M${_[0].x},${_[0].y} ${a(I)}` : `M${_[0].x},${_[0].y} ${a(B)}`;
          b.push(
            `<path d="${Y}"
                    fill="none"
                    stroke="${g.color}"
                    stroke-width="${e.value.style.chart.lines.strokeWidth}"
                    stroke-linecap="round" />`
          );
        }), b.join("");
      }
      const p = Ha.value.min, w = Ha.value.max, A = Math.abs(p), S = w + A || 1, P = (d) => Math.max(0, Math.min(t, d)), E = (d) => P(t - t * (((d ?? 0) + A) / S)), M = [];
      if (s.length) {
        const d = Array(o).fill(0), u = Array(o).fill(0);
        s.forEach((b) => {
          const V = Array.isArray(b.fullSeries) ? b.fullSeries : b.series, g = new Array(o), $ = new Array(o);
          for (let L = 0; L < o; L += 1) {
            const U = V?.[L] ?? 0;
            U >= 0 ? ($[L] = d[L], d[L] += U, g[L] = d[L]) : ($[L] = u[L], u[L] += U, g[L] = u[L]);
          }
          const O = g.map(E), _ = $.map(E), I = r.map((L, U) => ({ x: L, y: O[U] })), B = r.map((L, U) => ({ x: L, y: _[U] })), Y = be(I), Z = be([...B].reverse()), j = xe(I), W = xe([...B].reverse()), R = B[B.length - 1];
          if (e.value.style.chart.lines.useArea) {
            const L = e.value.style.chart.lines.smooth ? `M${I[0].x},${I[0].y} ${a(Y)} L${R.x},${R.y} ${a(Z)} Z` : `M${I[0].x},${I[0].y} ${a(j)} L${R.x},${R.y} ${a(W)} Z`;
            M.push(
              `<path d="${L}"
                    fill="${qt.pattern ? `url(#pattern_${te.value}_${b.absoluteIndex})` : (e.value.style.chart.lines.gradient.show, b.color)}"
                    opacity="${e.value.style.chart.lines.areaOpacity}"
                    stroke="none" />`
            );
          }
          const y = e.value.style.chart.lines.smooth ? `M${I[0].x},${I[0].y} ${a(Y)}` : `M${I[0].x},${I[0].y} ${a(j)}`;
          M.push(
            `<path d="${y}"
                fill="none"
                stroke="${b.color}"
                stroke-width="${e.value.style.chart.lines.strokeWidth}"
                stroke-linecap="round" />`
          );
        });
      }
      return n.length && n.forEach((d) => {
        const b = (Array.isArray(d.fullSeries) ? d.fullSeries : d.series).map(E), V = r.map((_, I) => ({ x: _, y: b[I] })), g = be(V), $ = xe(V), O = e.value.style.chart.lines.smooth ? `M${V[0].x},${V[0].y} ${a(g)}` : `M${V[0].x},${V[0].y} ${a($)}`;
        M.push(
          `<path d="${O}"
                fill="none"
                stroke="${d.color}"
                stroke-width="${e.value.style.chart.lines.strokeWidth}"
                stroke-linecap="round" />`
        );
      }), M.join("");
    }
    function ss(t, l) {
      const a = e.value.style.chart.lines, s = a.dataLabels;
      return a.showDistributedPercentage && a.distributed ? va.value ? l === 0 ? !s.hideEmptyPercentages : Math.abs(l) * 100 >= s.hideUnderPercentage : s.hideEmptyPercentages ? Math.abs(l) > 0 : !0 : va.value ? (ca.value && Tt.value && console.warn("Vue Data UI - VueUiStackline - You cannot set both dataLabels.hideUnderPercentage and dataLabels.hideUnderValue. Note that dataLabels.hideUnderPercentage takes precedence in this case."), Math.abs(t) > os.value * s.hideUnderPercentage / 100) : ca.value ? Math.abs(t) >= s.hideUnderValue : s.hideEmptyValues ? t !== 0 : !0;
    }
    const os = m(() => Math.max(...T.value.flatMap((t) => t.series)));
    function Ua(t, l, a, s, n) {
      const o = n === -1 && t >= 0 ? -t : t;
      return Je(
        e.value.style.chart.lines.dataLabels.formatter,
        o,
        Se({
          p: e.value.style.chart.lines.dataLabels.prefix,
          v: o,
          s: e.value.style.chart.lines.dataLabels.suffix,
          r: e.value.style.chart.lines.dataLabels.rounding
        }),
        { datapoint: l, seriesIndex: a, datapointIndex: s }
      );
    }
    function ns(t, l, a, s) {
      return Je(
        e.value.style.chart.lines.dataLabels.formatter,
        t,
        Se({
          v: isNaN(t) ? 0 : t,
          s: "%",
          r: e.value.style.chart.lines.dataLabels.rounding
        }),
        { datapoint: l, seriesIndex: a, datapointIndex: s }
      );
    }
    function qa({ seriesIndex: t, datapoint: l }) {
      const a = c.value.start + t;
      Me("selectX", {
        dataset: l,
        index: a,
        indexLabel: e.value.style.chart.grid.x.timeLabels.values[a]
      });
    }
    function rs(t) {
      if (t?.stage === "start") {
        Qe.value = !0;
        return;
      }
      if (t?.stage === "end") {
        Qe.value = !1;
        return;
      }
      da();
    }
    async function Xa() {
      if (Me("copyAlt", {
        config: e.value,
        dataset: T.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: T.value
      }));
    }
    return sl({
      getData: ts,
      getImage: es,
      generatePdf: ha,
      generateCsv: Dt,
      generateImage: da,
      generateSvg: Oa,
      hideSeries: Gl,
      showSeries: Xl,
      toggleTable: Va,
      toggleLabels: Ba,
      toggleTooltip: Ya,
      toggleAnnotator: St,
      toggleFullscreen: Da,
      copyAlt: Xa
    }), (t, l) => (i(), v("div", {
      id: `stackline_${te.value}`,
      ref_key: "stacklineChart",
      ref: ie,
      class: ja({ "vue-data-ui-component": !0, "vue-ui-stackline": !0, "vue-data-ui-wrapper-fullscreen": Fe.value }),
      style: de(`background:${e.value.style.chart.backgroundColor};color:${e.value.style.chart.color};font-family:${e.value.style.fontFamily}; position: relative; ${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: kl,
      onMouseleave: wl
    }, [
      e.value.userOptions.buttons.annotator ? (i(), re(C(cl), {
        key: 0,
        svgRef: C(Ce),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: Re.value,
        isCursorPointer: ue.value,
        onClose: St
      }, {
        "annotator-action-close": N(() => [
          z(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": N(({ color: a }) => [
          z(t.$slots, "annotator-action-color", Q(ee({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": N(({ mode: a }) => [
          z(t.$slots, "annotator-action-draw", Q(ee({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": N(({ disabled: a }) => [
          z(t.$slots, "annotator-action-undo", Q(ee({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": N(({ disabled: a }) => [
          z(t.$slots, "annotator-action-redo", Q(ee({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": N(({ disabled: a }) => [
          z(t.$slots, "annotator-action-delete", Q(ee({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : k("", !0),
      z(t.$slots, "userConfig", {}, void 0, !0),
      Sl.value ? (i(), v("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Kt,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : k("", !0),
      e.value.style.chart.title.text ? (i(), v("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Jt,
        style: "width:100%;background:transparent;"
      }, [
        (i(), re(Bs, {
          key: `title_${ea.value}`,
          config: {
            title: {
              cy: "stackline-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "stackline-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : k("", !0),
      he("div", {
        id: `legend-top-${te.value}`
      }, null, 8, Gs),
      e.value.userOptions.show && Ie.value && (C(ua) || C($t)) ? (i(), re(C(vl), {
        ref_key: "userOptionsRef",
        ref: bt,
        key: `user_option_${jt.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: C(Lt),
        isImaging: C(At),
        uid: te.value,
        hasTooltip: e.value.style.chart.tooltip.show && e.value.userOptions.buttons.tooltip,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: e.value.userOptions.buttons.labels,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: Fe.value,
        chartElement: ie.value,
        position: e.value.userOptions.position,
        isTooltip: G.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Re.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: ue.value,
        onToggleFullscreen: Da,
        onGeneratePdf: C(ha),
        onGenerateCsv: Dt,
        onGenerateImage: rs,
        onGenerateSvg: Oa,
        onToggleTable: Va,
        onToggleLabels: Ba,
        onToggleTooltip: Ya,
        onToggleAnnotator: St,
        onCopyAlt: Xa,
        style: de({
          visibility: C(ua) ? C($t) ? "visible" : "hidden" : "visible"
        })
      }, Wt({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: N(({ isOpen: a, color: s }) => [
            z(t.$slots, "menuIcon", Q(ee({ isOpen: a, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: N(() => [
            z(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: N(() => [
            z(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: N(() => [
            z(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: N(() => [
            z(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: N(() => [
            z(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: N(() => [
            z(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: N(() => [
            z(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: N(({ toggleFullscreen: a, isFullscreen: s }) => [
            z(t.$slots, "optionFullscreen", Q(ee({ toggleFullscreen: a, isFullscreen: s })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: N(({ toggleAnnotator: a, isAnnotator: s }) => [
            z(t.$slots, "optionAnnotator", Q(ee({ toggleAnnotator: a, isAnnotator: s })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: N(({ altCopy: a }) => [
            z(t.$slots, "optionAltCopy", Q(ee({ altCopy: a })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasLabel", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "isTooltip", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : k("", !0),
      (i(), v("svg", {
        ref_key: "svgRef",
        ref: Ce,
        xmlns: C(ks),
        viewBox: `0 0 ${h.value.chartWidth <= 0 ? 10 : h.value.chartWidth} ${h.value.chartHeight <= 0 ? 10 : h.value.chartHeight}`,
        class: ja({ "vue-data-ui-loading": C(ce), "no-transition": !e.value.useCssAnimation }),
        style: de(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`),
        role: "img",
        "aria-live": "polite",
        preserveAspectRatio: "xMidYMid",
        onMousemove: Bl,
        onMouseleave: Et,
        onClick: Yl
      }, [
        Ge(C(hl)),
        (i(!0), v(F, null, q(ke.value, (a) => (i(), v(F, null, [
          t.$slots.pattern ? (i(), v("defs", Js, [
            z(t.$slots, "pattern", dt({ ref_for: !0 }, { seriesIndex: a.absoluteIndex, patternId: `pattern_${te.value}_${a.absoluteIndex}` }), void 0, !0)
          ])) : k("", !0)
        ], 64))), 256)),
        t.$slots["chart-background"] ? (i(), v("foreignObject", {
          key: 0,
          x: K.value.left,
          y: h.value.top,
          width: K.value.width,
          height: h.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          z(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Zs)) : k("", !0),
        e.value.style.chart.lines.gradient.show ? (i(), v("defs", Ks, [
          (i(!0), v(F, null, q(T.value, (a, s) => (i(), v("linearGradient", {
            id: `gradient_${a.id}`,
            x1: "0%",
            y1: "0%",
            x2: "0%",
            y2: "100%"
          }, [
            he("stop", {
              offset: "0%",
              "stop-color": a.color
            }, null, 8, eo),
            he("stop", {
              offset: "100%",
              "stop-color": C(ws)(a.color, e.value.style.chart.lines.gradient.intensity / 100)
            }, null, 8, to)
          ], 8, Qs))), 256))
        ])) : k("", !0),
        e.value.style.chart.grid.x.showHorizontalLines ? (i(!0), v(F, { key: 2 }, q(Ne.value, (a, s) => (i(), v("line", {
          x1: K.value.left,
          x2: K.value.right,
          y1: a.y,
          y2: a.y,
          stroke: e.value.style.chart.grid.x.linesColor,
          "stroke-width": e.value.style.chart.grid.x.linesThickness,
          "stroke-dasharray": e.value.style.chart.grid.x.linesStrokeDasharray,
          "stroke-linecap": "round"
        }, null, 8, ao))), 256)) : k("", !0),
        e.value.style.chart.grid.y.showVerticalLines ? (i(!0), v(F, { key: 3 }, q(c.value.end - c.value.start, (a, s) => (i(), v("line", {
          x1: ne(s),
          x2: ne(s),
          y1: h.value.top,
          y2: h.value.bottom,
          stroke: e.value.style.chart.grid.y.linesColor,
          "stroke-width": e.value.style.chart.grid.y.linesThickness,
          "stroke-dasharray": e.value.style.chart.grid.y.linesStrokeDasharray,
          "stroke-linecap": "round"
        }, null, 8, lo))), 256)) : k("", !0),
        e.value.style.chart.grid.x.showAxis ? (i(), v("line", {
          key: 4,
          x1: K.value.left,
          x2: K.value.right,
          y1: h.value.bottom,
          y2: h.value.bottom,
          stroke: e.value.style.chart.grid.x.axisColor,
          "stroke-width": e.value.style.chart.grid.x.axisThickness,
          "stroke-linecap": "round",
          "stroke-linejoin": "round"
        }, null, 8, so)) : k("", !0),
        e.value.style.chart.grid.y.showAxis && !e.value.style.chart.lines.distributed ? (i(), v("line", {
          key: 5,
          x1: K.value.left,
          x2: K.value.left,
          y1: h.value.top,
          y2: h.value.bottom,
          stroke: e.value.style.chart.grid.y.axisColor,
          "stroke-width": e.value.style.chart.grid.y.axisThickness,
          "stroke-linecap": "round",
          "stroke-linejoin": "round"
        }, null, 8, oo)) : k("", !0),
        e.value.style.chart.grid.x.axisName.show && e.value.style.chart.grid.x.axisName.text ? (i(), v("text", {
          key: 6,
          ref_key: "xAxisLabel",
          ref: tt,
          x: h.value.left + h.value.width / 2,
          y: h.value.chartHeight - 3,
          "font-size": e.value.style.chart.grid.x.axisName.fontSize,
          fill: e.value.style.chart.grid.x.axisName.color,
          "font-weight": e.value.style.chart.grid.x.axisName.bold ? "bold" : "normal",
          "text-anchor": "middle"
        }, me(e.value.style.chart.grid.x.axisName.text), 9, no)) : k("", !0),
        e.value.style.chart.grid.y.axisName.show && e.value.style.chart.grid.y.axisName.text ? (i(), v("text", {
          key: 7,
          ref_key: "yAxisLabel",
          ref: xt,
          transform: `translate(${e.value.style.chart.grid.y.axisName.fontSize}, ${h.value.top + h.value.height / 2}) rotate(-90)`,
          "font-size": e.value.style.chart.grid.y.axisName.fontSize,
          fill: e.value.style.chart.grid.y.axisName.color,
          "font-weight": e.value.style.chart.grid.y.axisName.bold ? "bold" : "normal",
          "text-anchor": "middle"
        }, me(e.value.style.chart.grid.y.axisName.text), 9, ro)) : k("", !0),
        e.value.style.chart.grid.frame.show ? (i(), v("rect", {
          key: 8,
          style: { pointerEvents: "none", transition: "none", animation: "none !important" },
          x: K.value.left,
          y: h.value.top,
          width: K.value.width,
          height: h.value.height,
          fill: "transparent",
          stroke: e.value.style.chart.grid.frame.stroke,
          "stroke-width": e.value.style.chart.grid.frame.strokeWidth,
          "stroke-linecap": e.value.style.chart.grid.frame.strokeLinecap,
          "stroke-linejoin": e.value.style.chart.grid.frame.strokeLinejoin,
          "stroke-dasharray": e.value.style.chart.grid.frame.strokeDasharray
        }, null, 8, io)) : k("", !0),
        (i(!0), v(F, null, q(T.value, (a) => (i(), v(F, null, [
          e.value.style.chart.lines.useArea && !a.standalone ? (i(), v("path", {
            key: 0,
            d: e.value.style.chart.lines.smooth ? a.smoothArea : a.straightArea,
            fill: t.$slots.pattern ? `url(#pattern_${te.value}_${a.absoluteIndex})` : e.value.style.chart.lines.gradient.show ? `url(#gradient_${a.id})` : a.color,
            opacity: e.value.style.chart.lines.areaOpacity,
            style: de({
              transition: C(ce) || !e.value.useCssAnimation ? void 0 : "all 0.3s ease-in-out"
            })
          }, null, 12, uo)) : k("", !0)
        ], 64))), 256)),
        (i(!0), v(F, null, q(T.value, (a) => (i(), v("path", {
          d: e.value.style.chart.lines.smooth ? a.smoothPath : a.straightPath,
          stroke: a.color,
          "stroke-width": e.value.style.chart.lines.strokeWidth,
          fill: "none",
          "stroke-linecap": "round",
          style: de({
            transition: C(ce) || !e.value.useCssAnimation ? void 0 : "all 0.3s ease-in-out"
          })
        }, null, 12, co))), 256)),
        e.value.style.chart.grid.y.axisLabels.show && !e.value.style.chart.lines.distributed ? (i(), v("g", {
          key: 9,
          ref_key: "scaleLabels",
          ref: kt
        }, [
          (i(!0), v(F, null, q(Ne.value, (a, s) => (i(), v("line", {
            x1: K.value.left,
            x2: K.value.left - 6,
            y1: a.y,
            y2: a.y,
            stroke: e.value.style.chart.grid.x.axisColor,
            "stroke-width": 1
          }, null, 8, vo))), 256)),
          (i(!0), v(F, null, q(Ne.value, (a, s) => (i(), v("text", {
            x: a.x,
            y: a.y + e.value.style.chart.grid.y.axisLabels.fontSize / 3,
            "font-size": e.value.style.chart.grid.y.axisLabels.fontSize,
            "font-weight": e.value.style.chart.grid.y.axisLabels.bold ? "bold" : "normal",
            fill: e.value.style.chart.grid.y.axisLabels.color,
            "text-anchor": "end"
          }, me(C(Je)(
            e.value.style.chart.grid.y.axisLabels.formatter,
            a.value,
            C(Se)({
              p: e.value.style.chart.lines.dataLabels.prefix,
              v: a.value,
              s: e.value.style.chart.lines.dataLabels.suffix,
              r: e.value.style.chart.grid.y.axisLabels.rounding
            }),
            { datapoint: a }
          )), 9, ho))), 256))
        ], 512)) : k("", !0),
        e.value.style.chart.grid.x.timeLabels.show ? (i(), v("g", {
          key: 10,
          ref_key: "timeLabelsEls",
          ref: at
        }, [
          t.$slots["time-label"] ? (i(), v("g", mo, [
            (i(!0), v(F, null, q(it.value, (a, s) => (i(), v("g", null, [
              z(t.$slots, "time-label", dt({ ref_for: !0 }, {
                x: h.value.left + rt.value * s + rt.value / 2,
                y: h.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY,
                fontSize: e.value.style.chart.grid.x.timeLabels.fontSize,
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: `translate(${h.value.left + rt.value * s + rt.value / 2}, ${h.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}), rotate(${e.value.style.chart.grid.x.timeLabels.rotation})`,
                absoluteIndex: a.absoluteIndex,
                content: a.text,
                textAnchor: e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
                show: !0
              }), void 0, !0)
            ]))), 256))
          ])) : (i(), v("g", fo, [
            (i(!0), v(F, null, q(it.value, (a, s) => (i(), v("g", null, [
              String(a.text).includes(`
`) ? (i(), v("text", {
                key: s + "-multi",
                "text-anchor": e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.grid.x.timeLabels.fontSize,
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: `
                                    translate(
                                    ${ne(s)},
                                    ${h.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}
                                    ),
                                    rotate(${e.value.style.chart.grid.x.timeLabels.rotation})
                                `,
                style: de({ cursor: ue.value ? "pointer" : "default" }),
                innerHTML: t.createTSpansFromLineBreaksOnX({
                  content: String(a.text),
                  fontSize: e.value.style.chart.grid.x.timeLabels.fontSize,
                  fill: e.value.style.chart.grid.x.timeLabels.color,
                  x: 0,
                  y: 0
                }),
                onClick: () => _a(a, s)
              }, null, 12, yo)) : (i(), v("text", {
                class: "vue-data-ui-time-label",
                key: s,
                "text-anchor": e.value.style.chart.grid.x.timeLabels.rotation > 0 ? "start" : e.value.style.chart.grid.x.timeLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.grid.x.timeLabels.fontSize,
                "font-weight": e.value.style.chart.grid.x.timeLabels.bold ? "bold" : "normal",
                fill: e.value.style.chart.grid.x.timeLabels.color,
                transform: `translate(${ne(s)}, ${h.value.bottom + e.value.style.chart.grid.x.timeLabels.fontSize * 1.3 + e.value.style.chart.grid.x.timeLabels.offsetY}), rotate(${e.value.style.chart.grid.x.timeLabels.rotation})`,
                style: de({ cursor: ue.value ? "pointer" : "default" }),
                onClick: () => _a(a, s)
              }, me(a.text), 13, go))
            ]))), 256))
          ]))
        ], 512)) : k("", !0),
        ($e.value || ![null, void 0].includes(x.value)) && !e.value.style.chart.highlighter.useLine ? (i(!0), v(F, { key: 11 }, q(c.value.end - c.value.start, (a, s) => (i(), v("g", {
          key: `tooltip_trap_highlighter_${s}`
        }, [
          he("rect", {
            x: Pa(s).x,
            y: h.value.top,
            height: h.value.height,
            width: Pa(s).width,
            fill: [Oe.value, x.value].includes(s) ? C(Ss)(e.value.style.chart.highlighter.color, e.value.style.chart.highlighter.opacity) : "transparent",
            style: { transition: "none !important", animation: "none !important" }
          }, null, 8, po)
        ]))), 128)) : k("", !0),
        ($e.value || ![null, void 0].includes(x.value)) && e.value.style.chart.highlighter.useLine ? (i(), v(F, { key: 12 }, [
          ![null, void 0].includes(x.value) || ![null, void 0].includes(Oe.value) ? (i(), v("g", bo, [
            he("line", {
              x1: ne((x.value ?? Oe.value) || 0),
              x2: ne((x.value ?? Oe.value) || 0),
              y1: C(Ze)(h.value.top),
              y2: C(Ze)(h.value.bottom),
              stroke: e.value.style.chart.highlighter.color,
              "stroke-width": e.value.style.chart.highlighter.lineWidth,
              "stroke-dasharray": e.value.style.chart.highlighter.lineDasharray,
              "stroke-linecap": "round",
              style: { transition: "none !important", animation: "none !important", "pointer-events": "none" }
            }, null, 8, xo)
          ])) : k("", !0)
        ], 64)) : k("", !0),
        (i(!0), v(F, null, q(T.value, (a) => (i(), v(F, null, [
          $e.value && c.value.end - c.value.start > e.value.style.chart.lines.dot.hideAboveMaxSerieLength ? (i(), v("g", ko, [
            x.value != null ? (i(), v(F, { key: 0 }, [
              a.rel.includes(x.value) && a.fullSeries?.[c.value.start + x.value] != null && !Number.isNaN(a.fullSeries?.[c.value.start + x.value]) ? (i(), re(Ut, {
                key: 0,
                shape: ["triangle", "square", "diamond", "pentagon", "hexagon", "star"].includes(a.shape) ? a.shape : "circle",
                color: e.value.style.chart.lines.dot.useSerieColor ? a.color : e.value.style.chart.lines.dot.fill,
                plot: {
                  x: C(tl)(a.points[a.rel.indexOf(x.value)].x),
                  y: C(tl)(a.points[a.rel.indexOf(x.value)].y)
                },
                radius: e.value.style.chart.lines.dot.radius * 1.3,
                stroke: e.value.style.chart.lines.dot.useSerieColor ? e.value.style.chart.backgroundColor : a.color,
                strokeWidth: e.value.style.chart.lines.dot.strokeWidth,
                transition: C(ce) ? void 0 : `all ${e.value.style.chart.lines.transitionDurationMs}ms ease-in-out`
              }, null, 8, ["shape", "color", "plot", "radius", "stroke", "strokeWidth", "transition"])) : k("", !0)
            ], 64)) : k("", !0)
          ])) : k("", !0)
        ], 64))), 256)),
        (i(!0), v(F, null, q(T.value, (a) => (i(), v(F, null, [
          c.value.end - c.value.start < e.value.style.chart.lines.dot.hideAboveMaxSerieLength ? (i(!0), v(F, { key: 0 }, q(a.points, (s, n) => (i(), v("g", { key: n }, [
            a.fullSeries?.[c.value.start + a.rel[n]] != null && !Number.isNaN(a.fullSeries?.[c.value.start + a.rel[n]]) ? (i(), re(Ut, {
              key: 0,
              shape: ["triangle", "square", "diamond", "pentagon", "hexagon", "star"].includes(a.shape) ? a.shape : "circle",
              color: e.value.style.chart.lines.dot.useSerieColor ? a.color : e.value.style.chart.lines.dot.fill,
              plot: { x: s.x, y: s.y },
              radius: $e.value && x.value === a.rel[n] ? e.value.style.chart.lines.dot.radius * 1.3 : e.value.style.chart.lines.dot.radius,
              stroke: e.value.style.chart.lines.dot.useSerieColor ? e.value.style.chart.backgroundColor : a.color,
              strokeWidth: e.value.style.chart.lines.dot.strokeWidth,
              transition: C(ce) ? void 0 : `all ${e.value.style.chart.lines.transitionDurationMs}ms ease-in-out`
            }, null, 8, ["shape", "color", "plot", "radius", "stroke", "strokeWidth", "transition"])) : k("", !0)
          ]))), 128)) : k("", !0)
        ], 64))), 256)),
        G.value.dataLabels.show && e.value.style.chart.lines.dataLabels.hideAboveMaxSerieLength > c.value.end - c.value.start ? (i(), v(F, { key: 13 }, [
          (i(!0), v(F, null, q(T.value, (a, s) => (i(), v("g", { key: s }, [
            (i(!0), v(F, null, q(a.points, (n, o) => (i(), v(F, { key: o }, [
              ss(a.series[o], a.proportions[o]) ? (i(), v("text", {
                key: 0,
                x: n.x,
                y: n.y + (a.series[o] >= 0 ? -e.value.style.chart.lines.dataLabels.fontSize / 2 + e.value.style.chart.lines.dataLabels.offsetY : e.value.style.chart.lines.dataLabels.fontSize * 1.2 - e.value.style.chart.lines.dataLabels.offsetY),
                "font-size": e.value.style.chart.lines.dataLabels.fontSize,
                fill: e.value.style.chart.lines.dataLabels.color,
                "font-weight": e.value.style.chart.lines.dataLabels.bold ? "bold" : "normal",
                "text-anchor": "middle"
              }, me(e.value.style.chart.lines.showDistributedPercentage && e.value.style.chart.lines.distributed ? ns(a.proportions[o] * 100, a, s, a.rel[o]) : Ua(a.series[o], a, s, a.rel[o], a.signedSeries[o])), 9, wo)) : k("", !0)
            ], 64))), 128))
          ]))), 128)),
          e.value.style.chart.lines.totalValues.show && T.value.length > 1 ? (i(), v("g", {
            key: 0,
            ref_key: "sumTop",
            ref: bl
          }, [
            (i(!0), v(F, null, q(wa.value, (a, s) => (i(), v(F, null, [
              !e.value.style.chart.lines.dataLabels.hideEmptyValues || a.value !== 0 ? (i(), v("text", {
                key: 0,
                x: ne(s),
                y: _l(s),
                "text-anchor": "middle",
                "font-size": e.value.style.chart.lines.totalValues.fontSize,
                "font-weight": e.value.style.chart.lines.totalValues.bold ? "bold" : "normal",
                fill: e.value.style.chart.lines.totalValues.color
              }, me(Ua(a.value, a, s, s, a.sign)), 9, So)) : k("", !0)
            ], 64))), 256))
          ], 512)) : k("", !0)
        ], 64)) : k("", !0),
        ya.value ? (i(), v("rect", dt({ key: 14 }, Al.value, {
          "data-start": c.value.start,
          "data-end": c.value.end
        }), null, 16, Mo)) : k("", !0),
        z(t.$slots, "svg", {
          svg: {
            drawingArea: h.value,
            data: T.value,
            isPrintingImg: C(Lt) | C(At) | Qe.value,
            isPrintingSvg: et.value
          }
        }, void 0, !0)
      ], 46, js)),
      t.$slots.watermark ? (i(), v("div", $o, [
        z(t.$slots, "watermark", Q(ee({ isPrinting: C(Lt) || C(At) || Qe.value || et.value })), void 0, !0)
      ])) : k("", !0),
      ta.value ? (i(), re(vs, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${te.value}` : `#legend-bottom-${te.value}`
      }, [
        he("div", {
          ref_key: "chartLegend",
          ref: Zt
        }, [
          e.value.style.chart.legend.show ? (i(), re(Ws, {
            key: 0,
            legendSet: We.value,
            config: ql.value,
            isCursorPointer: ue.value,
            onClickMarker: l[0] || (l[0] = ({ legend: a }) => a.segregate())
          }, Wt({
            item: N(({ legend: a }) => [
              C(ce) ? k("", !0) : (i(), v("div", {
                key: 0,
                onClick: (s) => a.segregate(),
                style: de(`opacity:${D.value.includes(a.id) ? 0.5 : 1}`)
              }, me(a.name), 13, Co))
            ]),
            legendToggle: N(() => [
              We.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !C(ce) ? (i(), re(Us, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: D.value.length > 0,
                isCursorPointer: ue.value,
                onToggle: Ul
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : k("", !0)
            ]),
            _: 2
          }, [
            t.$slots.pattern ? {
              name: "legend-pattern",
              fn: N(({ legend: a, index: s }) => [
                Ge(Ut, {
                  shape: a.shape,
                  radius: 30,
                  stroke: "none",
                  plot: { x: 30, y: 30 },
                  fill: `url(#pattern_${te.value}_${s})`
                }, null, 8, ["shape", "fill"])
              ]),
              key: "0"
            } : void 0
          ]), 1032, ["legendSet", "config", "isCursorPointer"])) : z(t.$slots, "legend", {
            key: 1,
            legend: We.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : k("", !0),
      Ie.value && e.value.userOptions.buttons.table ? (i(), re(hs(He.value.component), dt({ key: 6 }, He.value.props, {
        ref_key: "tableUnit",
        ref: Ee,
        onClose: Ra
      }), Wt({
        content: N(() => [
          Ge(C(ul), {
            colNames: vt.value.colNames,
            head: vt.value.head,
            body: vt.value.body,
            config: vt.value.config,
            title: e.value.table.useDialog ? "" : He.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: ue.value,
            onClose: Ra
          }, {
            th: N(({ th: a }) => [
              he("div", { innerHTML: a }, null, 8, Lo)
            ]),
            td: N(({ td: a }) => [
              Ja(me(isNaN(Number(a)) ? a : C(Se)({
                p: e.value.style.chart.lines.dataLabels.prefix,
                v: a,
                s: e.value.style.chart.lines.dataLabels.suffix,
                r: e.value.table.td.roundingValue
              })), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"])
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: N(() => [
            Ja(me(He.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: N(() => [
            he("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (a) => Dt(e.value.userOptions.callbacks.csv)),
              style: de({ cursor: ue.value ? "pointer" : "default" })
            }, [
              Ge(C(rl), {
                name: "fileCsv",
                stroke: He.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : k("", !0),
      e.value.style.chart.zoom.show && Ie.value && Ta.value && ve.value > 6 ? (i(), re(Hs, {
        key: 7,
        ref_key: "chartSlicer",
        ref: _e,
        allMinimaps: zl.value,
        background: e.value.style.chart.zoom.color,
        borderColor: e.value.style.chart.backgroundColor,
        customFormat: e.value.style.chart.zoom.customFormat,
        cutNullValues: !1,
        enableRangeHandles: e.value.style.chart.zoom.enableRangeHandles,
        enableSelectionDrag: e.value.style.chart.zoom.enableSelectionDrag,
        end: c.value.end,
        focusOnDrag: e.value.style.chart.zoom.focusOnDrag,
        focusRangeRatio: e.value.style.chart.zoom.focusRangeRatio,
        fontSize: e.value.style.chart.zoom.fontSize,
        immediate: !e.value.style.chart.zoom.preview.enable,
        inputColor: e.value.style.chart.zoom.color,
        isPreview: ya.value,
        labelLeft: e.value.style.chart.grid.x.timeLabels.values[c.value.start] ? ae.value?.[0]?.text ?? "" : "",
        labelRight: e.value.style.chart.grid.x.timeLabels.values[c.value.end - 1] ? ae.value?.at(-1)?.text ?? "" : "",
        max: Math.max(...yt.dataset.map((a) => a.series.length)),
        min: 0,
        minimap: ka.value,
        minimapCompact: e.value.style.chart.zoom.minimap.compact,
        minimapFrameColor: e.value.style.chart.zoom.minimap.frameColor,
        minimapIndicatorColor: e.value.style.chart.zoom.minimap.indicatorColor,
        minimapMerged: !1,
        minimapSelectedColor: e.value.style.chart.zoom.minimap.selectedColor,
        minimapSelectedColorOpacity: e.value.style.chart.zoom.minimap.selectedColorOpacity,
        minimapSelectedIndex: x.value,
        minimapSelectionRadius: 1,
        preciseLabels: Aa.value.length ? Aa.value : ze.value,
        refreshEndPoint: e.value.style.chart.zoom.endIndex !== null ? e.value.style.chart.zoom.endIndex + 1 : Math.max(...yt.dataset.map((a) => a.series.length)),
        refreshStartPoint: e.value.style.chart.zoom.startIndex !== null ? e.value.style.chart.zoom.startIndex : 0,
        selectColor: e.value.style.chart.zoom.highlightColor,
        selectedSeries: as.value,
        smoothMinimap: !1,
        start: c.value.start,
        textColor: e.value.style.chart.color,
        timeLabels: ze.value,
        usePreciseLabels: e.value.style.chart.grid.x.timeLabels.datetimeFormatter.enable && !e.value.style.chart.zoom.useDefaultFormat,
        valueEnd: c.value.end,
        valueStart: c.value.start,
        verticalHandles: e.value.style.chart.zoom.minimap.verticalHandles,
        maxWidth: e.value.style.chart.zoom.maxWidth,
        isCursorPointer: ue.value,
        additionalMinimapHeight: e.value.style.chart.zoom.minimap.additionalHeight,
        handleType: e.value.style.chart.zoom.minimap.handleType,
        handleIconColor: e.value.style.chart.zoom.minimap.handleIconColor,
        handleBorderWidth: e.value.style.chart.zoom.minimap.handleBorderWidth,
        handleBorderColor: e.value.style.chart.zoom.minimap.handleBorderColor,
        handleFill: e.value.style.chart.zoom.minimap.handleFill,
        handleWidth: e.value.style.chart.zoom.minimap.handleWidth,
        "onUpdate:end": Rl,
        "onUpdate:start": Ol,
        onTrapMouse: Tl,
        onReset: ba,
        onFutureEnd: l[2] || (l[2] = (a) => pa("end", a)),
        onFutureStart: l[3] || (l[3] = (a) => pa("start", a))
      }, {
        "reset-action": N(({ reset: a }) => [
          z(t.$slots, "reset-action", Q(ee({ reset: a })), void 0, !0)
        ]),
        slotMap: N(({ width: a, height: s, unitW: n }) => [
          he("g", {
            innerHTML: ls({ minimapH: s, unitW: n })
          }, null, 8, Ao)
        ]),
        _: 3
      }, 8, ["allMinimaps", "background", "borderColor", "customFormat", "enableRangeHandles", "enableSelectionDrag", "end", "focusOnDrag", "focusRangeRatio", "fontSize", "immediate", "inputColor", "isPreview", "labelLeft", "labelRight", "max", "minimap", "minimapCompact", "minimapFrameColor", "minimapIndicatorColor", "minimapSelectedColor", "minimapSelectedColorOpacity", "minimapSelectedIndex", "preciseLabels", "refreshEndPoint", "refreshStartPoint", "selectColor", "selectedSeries", "start", "textColor", "timeLabels", "usePreciseLabels", "valueEnd", "valueStart", "verticalHandles", "maxWidth", "isCursorPointer", "additionalMinimapHeight", "handleType", "handleIconColor", "handleBorderWidth", "handleBorderColor", "handleFill", "handleWidth"])) : k("", !0),
      he("div", {
        id: `legend-bottom-${te.value}`
      }, null, 8, To),
      Ge(C(nl), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: G.value.showTooltip && Xt.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: ie.value,
        content: Hl.value,
        isFullscreen: Fe.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": N(() => [
          z(t.$slots, "tooltip-before", Q(ee({ ...Gt.value })), void 0, !0)
        ]),
        "tooltip-after": N(() => [
          z(t.$slots, "tooltip-after", Q(ee({ ...Gt.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      t.$slots.source ? (i(), v("div", {
        key: 8,
        ref_key: "source",
        ref: Qt,
        dir: "auto"
      }, [
        z(t.$slots, "source", {}, void 0, !0)
      ], 512)) : k("", !0),
      z(t.$slots, "skeleton", {}, () => [
        C(ce) ? (i(), re(Ps, { key: 0 })) : k("", !0)
      ], !0)
    ], 46, Xs));
  }
}, Qo = /* @__PURE__ */ qs(No, [["__scopeId", "data-v-cdfe6451"]]);
export {
  Qo as default
};
