// vue-data-ui/dist/vue-ui-timer-CgdgMKIc.js
import { defineAsyncComponent as le, ref as v, shallowRef as K, onMounted as oe, onBeforeUnmount as re, computed as R, watch as ue, openBlock as l, createElementBlock as r, normalizeStyle as k, createBlock as ie, createCommentVNode as d, unref as Q, createVNode as T, renderSlot as H, createElementVNode as P, mergeProps as Y, normalizeProps as G, guardReactiveProps as X, toDisplayString as ne } from "vue";
import { u as ce } from "./useNestedProp-BBT0mbHh.js";
import { j as ve, X as de, a2 as q } from "./lib-GwPnJSrY.js";
import { t as he, u as pe } from "./useResponsive-DfdjqQps.js";
import { u as fe } from "./useConfig-CIgyKZ6R.js";
import { u as we } from "./useChartAccessibility-9icAAmYg.js";
import ke from "./Title-BfyrP-jp.js";
import S from "./BaseIcon-CNx8Tzl5.js";
import { _ as ge } from "./_plugin-vue_export-helper-CHgC5LLL.js";
class me {
  constructor(L, j, A, N = !0, D = !0) {
    this.interval = j, this.elapsed = 0, this.isPaused = !1;
    const m = `
            let interval;
            let elapsed = 0;
            let paused = false;
            let startTime;
            let tickInterval;

            onmessage = function(e) {
                const { action, data } = e.data;

                switch(action) {
                    case 'start':
                        startTime = Date.now();
                        tickInterval = data.interval;
                        elapsed = 0;
                        paused = false;
                        interval = setInterval(() => {
                            elapsed += tickInterval;
                            postMessage({ elapsed, timestamp: Date.now() });
                        }, tickInterval);
                        break;
                    
                    case 'pause':
                        paused = true;
                        clearInterval(interval);
                        elapsed = Date.now() - startTime;
                        break;

                    case 'resume':
                        if (paused) {
                            startTime = Date.now() - elapsed;
                            interval = setInterval(() => {
                                elapsed += tickInterval;
                                postMessage({ elapsed, timestamp: Date.now() });
                            }, tickInterval);
                        }
                        paused = false;
                        break;

                    case 'stop':
                        clearInterval(interval);
                        elapsed = 0;
                        postMessage({ elapsed });
                        break;

                    case 'reset':
                        elapsed = 0;
                        clearInterval(interval);
                        postMessage({ elapsed });
                        break;

                    case 'lap':
                        postMessage({
                            elapsed,
                            timestamp: Date.now(),
                            action: 'lap'
                        });
                        break;

                    default:
                        break;
                }
            };
        `, $ = new Blob([m], { type: "application/javascript" }), z = URL.createObjectURL($), i = new Worker(z);
    function n(o) {
      let h = Math.floor(o / 1e3), g = Math.floor(o % 1e3 / 10), y = Math.floor(h / 3600), e = Math.floor(h % 3600 / 60), p = h % 60, f = "";
      return D && (f += String(y).padStart(2, "0") + ":"), f += String(e).padStart(2, "0") + ":", f += String(p).padStart(2, "0"), N && (f += "." + String(g).padStart(2, "0")), f;
    }
    this.start = () => {
      this.isPaused = !1, i.postMessage({ action: "start", data: { interval: this.interval } });
    }, this.pause = () => {
      this.isPaused ? this.resume() : (this.isPaused = !0, i.postMessage({ action: "pause" }));
    }, this.resume = () => {
      this.isPaused && (i.postMessage({ action: "resume" }), this.isPaused = !1);
    }, this.stop = () => {
      i.postMessage({ action: "stop" }), this.isPaused = !1;
    }, this.reset = () => {
      i.postMessage({ action: "reset" }), this.elapsed = 0, this.isPaused = !1;
    }, this.restart = () => {
      this.stop(), this.start();
    }, this.lap = () => new Promise((o) => {
      i.postMessage({ action: "lap" });
      const h = (g) => {
        const { elapsed: y, timestamp: e, action: p } = g.data;
        if (p === "lap") {
          const f = n(y);
          o({
            timestamp: e || 0,
            elapsed: y,
            formatted: f
          });
        }
      };
      i.addEventListener("message", h, { once: !0 });
    }), i.onmessage = (o) => {
      const { elapsed: h, timestamp: g } = o.data;
      this.elapsed = h, L({
        timestamp: g || 0,
        elapsed: this.elapsed,
        formatted: n(this.elapsed)
      });
    }, i.onerror = (o) => {
      A && A(o);
    };
  }
}
const ye = ["xmlns", "viewBox"], be = ["width", "height"], _e = { key: 1 }, Ce = ["id"], Me = ["stop-color"], Te = ["stop-color"], Pe = ["cx", "cy", "r", "fill", "stroke", "stroke-width"], $e = ["d", "stroke", "stroke-width"], xe = ["r", "fill", "stroke", "stroke-width"], Ie = ["r", "fill", "stroke", "stroke-width"], Re = ["x", "y"], Se = ["x", "y", "font-size", "fill", "font-weight"], Ae = {
  key: 0,
  class: "vue-ui-timer-controls"
}, De = ["title"], ze = ["title"], Fe = ["title"], He = ["title"], Le = ["title"], je = {
  __name: "vue-ui-timer",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["start", "pause", "reset", "restart", "lap"],
  setup(J, { expose: L, emit: j }) {
    const A = le(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_timer: N } = fe(), D = J, m = j, $ = v(null), z = v(null), i = v(null), n = K(null), o = K(null), h = v(ve()), g = v(0);
    oe(() => {
      y();
    });
    function y() {
      if (e.value.responsive) {
        const a = he(() => {
          const { width: u, height: c } = pe({
            chart: $.value,
            title: e.value.style.title.text ? z.value : null,
            legend: i.value
          });
          requestAnimationFrame(() => {
            t.value.width = u, t.value.height = c, e.value.responsiveProportionalSizing ? (t.value.tracker.core = q({
              relator: Math.min(u, c),
              adjuster: e.value.style.width,
              source: 6 * e.value.stopwatch.tracker.radiusRatio,
              threshold: 1,
              fallback: 1
            }), t.value.tracker.aura = q({
              relator: Math.min(u, c),
              adjuster: e.value.style.width,
              source: 12 * e.value.stopwatch.tracker.aura.radiusRatio,
              threshold: 1,
              fallback: 1
            }), t.value.label = q({
              relator: Math.min(u, c),
              adjuster: e.value.style.width,
              source: e.value.stopwatch.label.fontSize,
              threshold: 10,
              fallback: 10
            })) : t.value.label = e.value.stopwatch.label.fontSize;
          });
        });
        n.value && (o.value && n.value.unobserve(o.value), n.value.disconnect()), n.value = new ResizeObserver(a), o.value = $.value.parentNode, n.value.observe(o.value);
      }
    }
    re(() => {
      n.value && (o.value && n.value.unobserve(o.value), n.value.disconnect());
    });
    const e = R({
      get: () => B(),
      set: (a) => a
    }), p = R(() => e.value.useCursorPointer), { svgRef: f } = we({ config: e.value.style.title });
    function B() {
      return ce({
        userConfig: D.config,
        defaultConfig: N
      });
    }
    ue(() => D.config, (a) => {
      e.value = B(), y(), g.value += 1;
    }, { deep: !0 });
    const Z = R(() => {
      if (e.value.stopwatch.showHours && e.value.stopwatch.showHundredth)
        return "00:00:00.00";
      if (e.value.stopwatch.showHours && !e.value.stopwatch.showHundredth)
        return "00:00:00";
      if (!e.value.stopwatch.showHours && e.value.stopwatch.showHundredth)
        return "00:00.00";
      if (!e.value.stopwatch.showHours && !e.value.stopwatch.showHundredth)
        return "00:00";
    }), t = v({
      height: e.value.style.height,
      width: e.value.style.width,
      tracker: {
        core: 6 * e.value.stopwatch.tracker.radiusRatio,
        aura: 12 * e.value.stopwatch.tracker.aura.radiusRatio
      },
      label: e.value.stopwatch.label.fontSize
    }), b = v(0), x = new me(
      (a) => ee(a),
      10,
      "",
      e.value.stopwatch.showHundredth,
      e.value.stopwatch.showHours
    ), E = v(!0), s = v(!1), w = v(!1);
    function O() {
      m("start"), E.value && x.start(), E.value = !1, s.value = !0;
    }
    function U() {
      s.value && (m("reset"), x.stop(), C.value = [], E.value = !0, s.value = !1);
    }
    function W() {
      w.value = !w.value, m("pause", b.value), x.pause();
    }
    function V() {
      s.value && (w.value = !1, m("restart"), C.value = [], x.restart());
    }
    const C = v([]);
    async function F() {
      if (!s.value || w.value) return;
      const a = await x.lap();
      a && (C.value.push(a), m("lap", C.value));
    }
    function ee({ timestamp: a, elapsed: u, formatted: c }) {
      b.value = { timestamp: a, elapsed: u, formatted: c };
    }
    const _ = R(() => Math.min(t.value.width, t.value.height) / 2.5 * e.value.stopwatch.track.radiusRatio);
    function te(a, u) {
      const I = 360 / (u * 1e3);
      return a * I % 360;
    }
    function ae(a) {
      let u = a * (Math.PI / 180), c = t.value.width / 2 + _.value * Math.cos(u), I = t.value.height / 2 + _.value * Math.sin(u);
      return { cx: c, cy: I };
    }
    const M = R(() => {
      const a = te(b.value.elapsed, e.value.stopwatch.cycleSeconds), { cx: u, cy: c } = ae(a - 90), I = a > 180 ? 1 : 0;
      return {
        cx: u || t.value.width / 2,
        cy: c || t.value.height / 2 - _.value,
        largeArcFlag: I,
        sweepFlag: 1
      };
    });
    return L({
      start: O,
      pause: W,
      reset: U,
      restart: V,
      lap: F
    }), (a, u) => (l(), r("div", {
      ref_key: "timerChart",
      ref: $,
      class: "vue-data-ui-component vue-ui-timer",
      style: k({
        fontFamily: e.value.style.fontFamily,
        width: "100%",
        height: e.value.responsive ? "100%" : "auto",
        textAlign: "center"
      })
    }, [
      e.value.style.title.text ? (l(), r("div", {
        key: 0,
        ref_key: "chartTitle",
        ref: z,
        style: k({
          width: "100%",
          background: e.value.style.backgroundColor
        })
      }, [
        (l(), ie(ke, {
          key: `title_${g.value}`,
          config: {
            title: {
              cy: "title",
              ...e.value.style.title
            },
            subtitle: {
              cy: "subtitle",
              ...e.value.style.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 4)) : d("", !0),
      (l(), r("svg", {
        ref_key: "svgRef",
        ref: f,
        xmlns: Q(de),
        viewBox: `0 0 ${t.value.width <= 0 ? 10 : t.value.width} ${t.value.height <= 0 ? 10 : t.value.height}`,
        style: k({
          maxWidth: "100%",
          overflow: "visible",
          background: e.value.style.backgroundColor
        })
      }, [
        T(Q(A)),
        a.$slots["chart-background"] ? (l(), r("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: t.value.width <= 0 ? 10 : t.value.width,
          height: t.value.height <= 0 ? 10 : t.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          H(a.$slots, "chart-background", {}, void 0, !0)
        ], 8, be)) : d("", !0),
        e.value.stopwatch.tracker.gradient.show ? (l(), r("defs", _e, [
          P("radialGradient", {
            id: `tracker_gradient_${h.value}`,
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%"
          }, [
            P("stop", {
              offset: "0%",
              "stop-color": e.value.stopwatch.tracker.gradient.color
            }, null, 8, Me),
            P("stop", {
              offset: "100%",
              "stop-color": e.value.stopwatch.tracker.fill
            }, null, 8, Te)
          ], 8, Ce)
        ])) : d("", !0),
        P("circle", {
          cx: t.value.width / 2,
          cy: t.value.height / 2,
          r: _.value,
          fill: e.value.stopwatch.track.fill,
          stroke: e.value.stopwatch.track.stroke,
          "stroke-width": e.value.stopwatch.track.strokeWidth
        }, null, 8, Pe),
        e.value.stopwatch.cycleTrack.show ? (l(), r("path", {
          key: 2,
          d: `M ${t.value.width / 2},${t.value.height / 2 - _.value} A ${_.value},${_.value} 0 ${M.value.largeArcFlag},${M.value.sweepFlag} ${M.value.cx},${M.value.cy}`,
          stroke: e.value.stopwatch.cycleTrack.stroke,
          "stroke-width": e.value.stopwatch.cycleTrack.strokeWidth,
          "stroke-linecap": "round",
          fill: "none"
        }, null, 8, $e)) : d("", !0),
        P("circle", Y(M.value, {
          r: t.value.tracker.core,
          fill: e.value.stopwatch.tracker.gradient.show ? `url(#tracker_gradient_${h.value})` : e.value.stopwatch.tracker.fill,
          stroke: e.value.stopwatch.tracker.stroke,
          "stroke-width": e.value.stopwatch.tracker.strokeWidth
        }), null, 16, xe),
        e.value.stopwatch.tracker.aura.show ? (l(), r("circle", Y({ key: 3 }, M.value, {
          r: t.value.tracker.aura,
          fill: `${e.value.stopwatch.tracker.aura.fill}20`,
          stroke: e.value.stopwatch.tracker.aura.stroke,
          "stroke-width": e.value.stopwatch.tracker.aura.strokeWidth
        }), null, 16, Ie)) : d("", !0),
        a.$slots.time ? (l(), r("foreignObject", {
          key: 4,
          x: t.value.width / 2,
          y: t.value.height / 2,
          height: "0.1",
          width: "0.1",
          style: { overflow: "visible" }
        }, [
          H(a.$slots, "time", G(X({ ...b.value })), void 0, !0)
        ], 8, Re)) : (l(), r("text", {
          key: 5,
          x: t.value.width / 2,
          y: t.value.height / 2 + t.value.label / 4,
          "font-size": t.value.label,
          "text-anchor": "middle",
          fill: e.value.stopwatch.label.color,
          "font-weight": e.value.stopwatch.label.bold ? "bold" : "normal",
          style: { "font-variant-numeric": "tabular-nums !important" }
        }, ne(b.value.formatted || Z.value), 9, Se))
      ], 12, ye)),
      P("div", {
        ref_key: "chartLegend",
        ref: i,
        style: k({ width: "100%", backgroundColor: e.value.stopwatch.legend.backgroundColor })
      }, [
        a.$slots.controls ? d("", !0) : (l(), r("div", Ae, [
          e.value.stopwatch.legend.buttons.start ? (l(), r("button", {
            key: 0,
            title: e.value.stopwatch.legend.buttonTitles.start,
            onClick: O,
            class: "vue-ui-timer-button",
            style: k({
              opacity: s.value ? 0.2 : 1,
              cursor: s.value ? "default" : p.value ? "pointer" : "default"
            })
          }, [
            T(S, {
              name: "play",
              stroke: e.value.stopwatch.legend.buttons.iconColor
            }, null, 8, ["stroke"])
          ], 12, De)) : d("", !0),
          e.value.stopwatch.legend.buttons.pause ? (l(), r("button", {
            key: 1,
            title: w.value ? e.value.stopwatch.legend.buttonTitles.resume : e.value.stopwatch.legend.buttonTitles.pause,
            onClick: W,
            class: "vue-ui-timer-button",
            style: k({
              opacity: s.value ? 1 : 0.2,
              cursor: s.value && p.value ? "pointer" : "default"
            })
          }, [
            T(S, {
              name: "pause",
              stroke: e.value.stopwatch.legend.buttons.iconColor
            }, null, 8, ["stroke"])
          ], 12, ze)) : d("", !0),
          e.value.stopwatch.legend.buttons.reset ? (l(), r("button", {
            key: 2,
            title: e.value.stopwatch.legend.buttonTitles.reset,
            onClick: U,
            class: "vue-ui-timer-button",
            style: k({
              opacity: s.value ? 1 : 0.2,
              cursor: s.value && p.value ? "pointer" : "default"
            })
          }, [
            T(S, {
              name: "stop",
              stroke: e.value.stopwatch.legend.buttons.iconColor
            }, null, 8, ["stroke"])
          ], 12, Fe)) : d("", !0),
          e.value.stopwatch.legend.buttons.restart ? (l(), r("button", {
            key: 3,
            title: e.value.stopwatch.legend.buttonTitles.restart,
            onClick: V,
            class: "vue-ui-timer-button",
            style: k({
              opacity: s.value ? 1 : 0.2,
              cursor: s.value && p.value ? "pointer" : "default"
            })
          }, [
            T(S, {
              name: "restart",
              stroke: e.value.stopwatch.legend.buttons.iconColor
            }, null, 8, ["stroke"])
          ], 12, He)) : d("", !0),
          e.value.stopwatch.legend.buttons.lap ? (l(), r("button", {
            key: 4,
            title: e.value.stopwatch.legend.buttonTitles.lap,
            onClick: F,
            class: "vue-ui-timer-button",
            style: k({
              opacity: s.value && !w.value ? 1 : 0.2,
              cursor: s.value && !w.value && p.value ? "pointer" : "default"
            })
          }, [
            T(S, {
              name: "lap",
              stroke: e.value.stopwatch.legend.buttons.iconColor
            }, null, 8, ["stroke"])
          ], 12, Le)) : d("", !0)
        ])),
        H(a.$slots, "controls", G(X({
          start: O,
          pause: W,
          reset: U,
          restart: V,
          lap: F,
          laps: C.value,
          isRunning: s.value,
          isPaused: w.value,
          ...b.value
        })), void 0, !0),
        H(a.$slots, "laps", G(X({
          laps: C.value,
          lap: F,
          isRunning: s.value,
          isPaused: w.value,
          ...b.value
        })), void 0, !0)
      ], 4)
    ], 4));
  }
}, qe = /* @__PURE__ */ ge(je, [["__scopeId", "data-v-4876d6f7"]]);
export {
  qe as default
};
