// vue-data-ui/dist/vue-ui-thermometer-sjruUGQx.js
import { useCssVars as Ke, defineAsyncComponent as Q, ref as d, computed as h, onMounted as Je, toRefs as Qe, watch as Ze, openBlock as r, createElementBlock as n, unref as u, normalizeStyle as pe, normalizeClass as V, createBlock as Y, withCtx as f, renderSlot as v, normalizeProps as y, guardReactiveProps as g, createCommentVNode as c, createSlots as et, createVNode as tt, createElementVNode as p, Fragment as E, renderList as ke, toDisplayString as ot, nextTick as st } from "vue";
import { j as at, t as be, o as lt, q as we, r as rt, m as nt, p as S, n as ut, K as b, s as H, X as it, u as ct, k as ht, l as vt } from "./lib-GwPnJSrY.js";
import { t as dt, u as ft } from "./useResponsive-DfdjqQps.js";
import { u as mt } from "./useConfig-CIgyKZ6R.js";
import { u as yt, B as gt } from "./BaseScanner-BhhkKVoj.js";
import { u as pt } from "./usePrinter-DiK5R6k-.js";
import { u as kt } from "./useSvgExport-ByUukOZt.js";
import { u as Z } from "./useNestedProp-BBT0mbHh.js";
import { u as bt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as wt } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Ct } from "./useChartAccessibility-9icAAmYg.js";
import { u as xt } from "./useAutoSizeLabelsInsideViewbox-80fzdfxt.js";
import _t from "./vue_ui_thermometer-BTmjRBH8.js";
import $t from "./Title-BfyrP-jp.js";
import At from "./img-BYYzY_QU.js";
import { _ as Pt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Ot = ["id"], St = {
  key: 1,
  ref: "noTitle",
  class: "vue-data-ui-no-title-space",
  style: "height:36px; width: 100%;background:transparent"
}, It = ["xmlns", "viewBox"], Ft = ["width", "height"], Tt = ["id"], Mt = ["x", "y", "width", "height", "rx", "ry"], Ut = ["id"], zt = ["stop-color"], Wt = ["stop-color"], Lt = ["stop-color"], Nt = ["clip-path"], Vt = ["x", "y", "height", "width"], Yt = ["x", "y", "height", "width", "fill"], Et = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], Ht = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], qt = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], Rt = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], Gt = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], Dt = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], Bt = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], jt = ["x1", "x2", "y1", "y2", "stroke-width", "stroke"], Xt = ["x", "y", "height", "width"], Kt = { key: 1 }, Jt = ["x", "y", "height"], Qt = ["y", "x", "fill", "font-size", "font-weight"], Zt = {
  key: 4,
  class: "vue-data-ui-watermark"
}, eo = {
  __name: "vue-ui-thermometer",
  props: {
    dataset: {
      type: Object,
      default() {
        return {};
      }
    },
    config: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["copyAlt"],
  setup(I, { expose: Ce, emit: xe }) {
    Ke((o) => ({
      v1ecd0f6c: Ne.value,
      b11332bc: Le.value,
      v4b513c0f: We.value
    }));
    const _e = Q(() => import("./PackageVersion-DN9d0ME5.js")), $e = Q(() => import("./PenAndPaper-DWV1FMIi.js")), Ae = Q(() => import("./UserOptions-CrLqEimB.js")), { vue_ui_thermometer: Pe } = mt(), { isThemeValid: Oe, warnInvalidTheme: Se } = bt(), m = I, Ie = xe, w = d(at()), C = d(null), ee = d(0), te = d(0), oe = d(null), se = d(null), _ = d(null), F = d(null), T = d(!1), M = d(!1), Fe = h(() => !!m.dataset && Object.keys(m.dataset).length);
    Je(() => {
      ne();
    });
    const e = d(D()), Te = d(e.value.style.chart.thermometer.width), q = d(e.value.style.chart.height), R = d(e.value.style.chart.width), ae = h(() => e.value.userOptions.useCursorPointer), Me = h(() => be({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            animation: { use: !1 },
            backgroundColor: "#99999930",
            graduations: {
              stroke: "#6A6A6A"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: le, FINAL_DATASET: i } = yt({
      ...Qe(m),
      FINAL_CONFIG: e,
      prepareConfig: D,
      skeletonDataset: m.config?.skeletonDataset ?? { value: 0, from: -100, to: 100, steps: 20, colors: { from: "#A1A1A1", to: "#CACACA" } },
      skeletonConfig: be({
        defaultConfig: e.value,
        userConfig: Me.value
      })
    }), re = h(() => !!e.value.debug);
    function ne() {
      if (lt(m.dataset) ? we({
        componentName: "VueUiThermometer",
        type: "dataset",
        debug: re.value
      }) : rt({
        datasetObject: m.dataset,
        requiredAttributes: ["value", "from", "to"]
      }).forEach((o) => {
        we({
          componentName: "VueUiThermometer",
          type: "datasetAttribute",
          property: o,
          debug: re.value
        });
      }), e.value.responsive) {
        const o = dt(() => {
          const { width: a, height: t } = ft({
            chart: C.value,
            title: e.value.style.title.text ? oe.value : null,
            source: se.value
          });
          requestAnimationFrame(() => {
            q.value = Math.max(0.1, t - 12), R.value = a, me();
          });
        });
        _.value && (F.value && _.value.unobserve(F.value), _.value.disconnect()), _.value = new ResizeObserver(o), F.value = C.value, _.value.observe(F.value);
      }
      me();
    }
    const { userOptionsVisible: G, setUserOptionsVisibility: ue, keepUserOptionState: ie } = wt({ config: e.value }), { svgRef: U } = Ct({ config: e.value.style.title });
    function D() {
      const o = Z({
        userConfig: m.config,
        defaultConfig: Pe
      }), a = o.theme;
      if (!a) return o;
      if (!Oe.value(o))
        return Se(o), o;
      const t = Z({
        userConfig: _t[a] || m.config,
        defaultConfig: o
      }), l = Z({
        userConfig: m.config,
        defaultConfig: t
      });
      return {
        ...l,
        customPalette: l.customPalette.length ? l.customPalette : nt[a] || S
      };
    }
    Ze(() => m.config, (o) => {
      e.value = D(), G.value = !e.value.userOptions.showOnChartHover, Te.value = e.value.style.chart.thermometer.width, q.value = e.value.style.chart.height, R.value = e.value.style.chart.width, ne(), te.value += 1;
    }, { deep: !0 });
    const { isPrinting: B, isImaging: j, generatePdf: ce, generateImage: he } = pt({
      elementId: `thermometer__${w.value}`,
      fileName: e.value.style.title.text || "vue-ui-thermometer",
      options: e.value.userOptions.print
    }), Ue = h(() => e.value.userOptions.show && !e.value.style.title.text), z = h(() => ut(e.value.customPalette)), k = h(() => i.value.steps || 10);
    function W(o, a, t) {
      const l = [], x = ve(o), P = ve(a);
      for (let O = 0; O < t; O += 1) {
        const De = X(x.red, P.red, O, t), Be = X(x.green, P.green, O, t), je = X(x.blue, P.blue, O, t), Xe = `#${K(De)}${K(Be)}${K(je)}`;
        l.push(Xe);
      }
      return l;
    }
    function ve(o) {
      const a = o.slice(1);
      return {
        red: parseInt(a.slice(0, 2), 16),
        green: parseInt(a.slice(2, 4), 16),
        blue: parseInt(a.slice(4, 6), 16)
      };
    }
    function X(o, a, t, l) {
      return Math.round(o + (a - o) * t / l);
    }
    function K(o) {
      return o.toString(16).padStart(2, "0");
    }
    const s = h(() => {
      const o = Math.max(0.1, R.value), a = Math.max(0.1, q.value), t = e.value.style.chart.padding;
      return {
        width: o,
        left: o / 2 - e.value.style.chart.thermometer.width / 2,
        right: o / 2 + e.value.style.chart.thermometer.width / 2,
        top: t.top,
        bottom: a - t.bottom - t.top,
        height: a,
        thermoHeight: a - t.top - t.bottom,
        thermoWidth: e.value.style.chart.thermometer.width
      };
    }), ze = h(() => s.value), L = h(() => {
      const o = b(i.value.from) < 0 ? Math.abs(b(i.value.from)) : b(i.value.from), a = b(i.value.to) < 0 ? Math.abs(b(i.value.to)) : b(i.value.to);
      let t = 0;
      return b(i.value.to) > 0 ? t = o + a : o > a ? t = o - a : t = a - o, (1 - (Math.abs(b(i.value.from)) + b(i.value.value)) / t) * s.value.thermoHeight;
    }), We = h(() => `${L.value}px`), Le = h(() => `${s.value.thermoHeight}px`), Ne = h(() => `${e.value.style.chart.animation.speedMs}ms`), Ve = h(() => {
      if (i.value.colors) {
        if (!i.value.colors.from)
          return W(z.value[0] || S[0], H(i.value.colors.to), k.value || 10);
        if (!i.value.colors.to)
          return W(H(i.value.colors.from), z.value[1] || S[1], k.value || 10);
      } else
        return W(z.value[1] || S[1], z.value[0] || S[0], k.value || 10);
      return W(H(i.value.colors.from), H(i.value.colors.to), k.value || 10);
    }), de = h(() => {
      const o = [];
      let a = 0;
      const t = s.value.thermoHeight;
      for (let l = 0; l < t - 1; l += t / k.value)
        o.push({
          x: s.value.left,
          y: s.value.top + l,
          qYLess: s.value.top + l + t / k.value / 4,
          halfY: s.value.top + l + t / k.value / 2,
          qYMore: s.value.top + l + t / k.value / 4 * 3,
          color: Ve.value[a],
          height: Math.max(0.1, t / k.value)
        }), a += 1;
      return o;
    }), $ = d(!1);
    function fe(o) {
      $.value = o, ee.value += 1;
    }
    const N = d(!1);
    function J() {
      N.value = !N.value;
    }
    async function Ye({ scale: o = 2 } = {}) {
      if (!C.value) return;
      const { width: a, height: t } = C.value.getBoundingClientRect(), l = a / t, { imageUri: x, base64: P } = await At({ domElement: C.value, base64: !0, img: !0, scale: o });
      return {
        imageUri: x,
        base64: P,
        title: e.value.style.title.text,
        width: a,
        height: t,
        aspectRatio: l
      };
    }
    const A = h({
      get: () => e.value.style.chart.label.fontSize,
      set: (o) => o
    }), { autoSizeLabels: me } = xt({
      svgRef: U,
      fontSize: e.value.style.chart.label.fontSize,
      minFontSize: e.value.style.chart.label.minFontSize,
      sizeRef: A,
      labelClass: ".vue-ui-thermometer-label"
    }), Ee = h(() => e.value.style.chart.backgroundColor), He = h(() => e.value.style.title), { exportSvg: qe, getSvg: Re } = kt({
      svg: U,
      title: He,
      backgroundColor: Ee
    });
    async function ye({ isCb: o }) {
      M.value = !0, await st();
      try {
        if (o) {
          const { blob: a, url: t, text: l, dataUrl: x } = await Re();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: t, text: l, dataUrl: x }));
        } else
          await Promise.resolve(qe());
      } finally {
        M.value = !1;
      }
    }
    function Ge(o) {
      if (o?.stage === "start") {
        T.value = !0;
        return;
      }
      if (o?.stage === "end") {
        T.value = !1;
        return;
      }
      he();
    }
    async function ge() {
      if (Ie("copyAlt", {
        config: e.value,
        dataset: i.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: i.value
      }));
    }
    return Ce({
      getImage: Ye,
      generatePdf: ce,
      generateImage: he,
      generateSvg: ye,
      toggleAnnotator: J,
      toggleFullscreen: fe,
      copyAlt: ge
    }), (o, a) => (r(), n("div", {
      ref_key: "thermoChart",
      ref: C,
      class: V(`vue-data-ui-component vue-ui-thermometer ${$.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      style: pe(`width:100%;background:${e.value.style.chart.backgroundColor};color:${e.value.style.chart.color};font-family:${e.value.style.fontFamily}`),
      id: `thermometer__${w.value}`,
      onMouseenter: a[0] || (a[0] = () => u(ue)(!0)),
      onMouseleave: a[1] || (a[1] = () => u(ue)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (r(), Y(u($e), {
        key: 0,
        svgRef: u(U),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: N.value,
        isCursorPointer: ae.value,
        onClose: J
      }, {
        "annotator-action-close": f(() => [
          v(o.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": f(({ color: t }) => [
          v(o.$slots, "annotator-action-color", y(g({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": f(({ mode: t }) => [
          v(o.$slots, "annotator-action-draw", y(g({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": f(({ disabled: t }) => [
          v(o.$slots, "annotator-action-undo", y(g({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": f(({ disabled: t }) => [
          v(o.$slots, "annotator-action-redo", y(g({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": f(({ disabled: t }) => [
          v(o.$slots, "annotator-action-delete", y(g({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : c("", !0),
      Ue.value ? (r(), n("div", St, null, 512)) : c("", !0),
      e.value.style.title.text ? (r(), n("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: oe,
        style: "width:100%"
      }, [
        (r(), Y($t, {
          key: `title_${te.value}`,
          config: {
            title: {
              cy: "thermometer-div-title",
              ...e.value.style.title
            },
            subtitle: {
              cy: "thermometer-div-subtitle",
              ...e.value.style.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : c("", !0),
      e.value.userOptions.show && Fe.value && (u(ie) || u(G)) ? (r(), Y(u(Ae), {
        ref: "details",
        key: `user_options_${ee.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isImaging: u(j),
        isPrinting: u(B),
        uid: w.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        hasXls: !1,
        isFullscreen: $.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: C.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: N.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        isCursorPointer: ae.value,
        onToggleFullscreen: fe,
        onGeneratePdf: u(ce),
        onGenerateImage: Ge,
        onGenerateSvg: ye,
        onToggleAnnotator: J,
        onCopyAlt: ge,
        style: pe({
          visibility: u(ie) ? u(G) ? "visible" : "hidden" : "visible"
        })
      }, et({ _: 2 }, [
        o.$slots.menuIcon ? {
          name: "menuIcon",
          fn: f(({ isOpen: t, color: l }) => [
            v(o.$slots, "menuIcon", y(g({ isOpen: t, color: l })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        o.$slots.optionPdf ? {
          name: "optionPdf",
          fn: f(() => [
            v(o.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        o.$slots.optionImg ? {
          name: "optionImg",
          fn: f(() => [
            v(o.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        o.$slots.optionSvg ? {
          name: "optionSvg",
          fn: f(() => [
            v(o.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        o.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: f(({ toggleFullscreen: t, isFullscreen: l }) => [
            v(o.$slots, "optionFullscreen", y(g({ toggleFullscreen: t, isFullscreen: l })), void 0, !0)
          ]),
          key: "4"
        } : void 0,
        o.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: f(({ toggleAnnotator: t, isAnnotator: l }) => [
            v(o.$slots, "optionAnnotator", y(g({ toggleAnnotator: t, isAnnotator: l })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        o.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: f(({ altCopy: t }) => [
            v(o.$slots, "optionAltCopy", y(g({ altCopy: t })), void 0, !0)
          ]),
          key: "6"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "isCursorPointer", "onGeneratePdf", "style"])) : c("", !0),
      (r(), n("svg", {
        ref_key: "svgRef",
        ref: U,
        xmlns: u(it),
        class: V({ "vue-data-ui-fullscreen--on": $.value, "vue-data-ui-fulscreen--off": !$.value }),
        width: "100%",
        viewBox: `0 0 ${s.value.width} ${s.value.height}`,
        style: "background:transparent"
      }, [
        tt(u(_e)),
        o.$slots["chart-background"] ? (r(), n("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: s.value.width,
          height: s.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          v(o.$slots, "chart-background", {}, void 0, !0)
        ], 8, Ft)) : c("", !0),
        p("defs", null, [
          p("clipPath", {
            id: `vueUiPill-${w.value}`,
            clipPathUnits: "userSpaceOnUse"
          }, [
            p("rect", {
              x: s.value.left,
              y: s.value.top,
              width: s.value.thermoWidth,
              height: s.value.thermoHeight,
              rx: s.value.thermoWidth / 2,
              ry: s.value.thermoWidth / 2
            }, null, 8, Mt)
          ], 8, Tt),
          (r(!0), n(E, null, ke(de.value, (t, l) => (r(), n("linearGradient", {
            id: `vueUiThermometerGradient_${l}_${w.value}`,
            x1: "0%",
            y1: "0%",
            x2: "100%",
            y2: "0%"
          }, [
            p("stop", {
              offset: "0%",
              "stop-color": t.color
            }, null, 8, zt),
            p("stop", {
              offset: "50%",
              "stop-color": u(ct)(t.color, 100 - e.value.style.chart.graduations.gradient.intensity)
            }, null, 8, Wt),
            p("stop", {
              offset: "100%",
              "stop-color": t.color
            }, null, 8, Lt)
          ], 8, Ut))), 256))
        ]),
        p("g", {
          "clip-path": `url(#vueUiPill-${w.value})`
        }, [
          p("rect", {
            x: s.value.left,
            y: s.value.top,
            height: s.value.thermoHeight,
            width: s.value.thermoWidth,
            fill: "#FFFFFF"
          }, null, 8, Vt),
          (r(!0), n(E, null, ke(de.value, (t, l) => (r(), n("g", {
            key: `graduation_${l}`
          }, [
            p("rect", {
              x: t.x,
              y: t.y,
              height: t.height,
              width: s.value.thermoWidth,
              fill: e.value.style.chart.graduations.gradient.show ? `url(#vueUiThermometerGradient_${l}_${w.value})` : t.color,
              "shape-rendering": "crispEdges"
            }, null, 8, Yt),
            e.value.style.chart.graduations.show && ["both", "left"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
              key: 0,
              x1: t.x,
              x2: t.x + 10,
              y1: t.y,
              y2: t.y,
              "stroke-width": e.value.style.chart.graduations.strokeWidth,
              stroke: e.value.style.chart.graduations.stroke,
              "stroke-linecap": "round"
            }, null, 8, Et)) : c("", !0),
            e.value.style.chart.graduations.showIntermediate ? (r(), n(E, { key: 1 }, [
              e.value.style.chart.graduations.show && ["both", "left"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
                key: 0,
                x1: t.x,
                x2: t.x + 5,
                y1: t.halfY,
                y2: t.halfY,
                "stroke-width": e.value.style.chart.graduations.strokeWidth / 2,
                stroke: e.value.style.chart.graduations.stroke,
                "stroke-linecap": "round"
              }, null, 8, Ht)) : c("", !0),
              e.value.style.chart.graduations.show && ["both", "left"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
                key: 1,
                x1: t.x,
                x2: t.x + 2.5,
                y1: t.qYLess,
                y2: t.qYLess,
                "stroke-width": e.value.style.chart.graduations.strokeWidth / 2,
                stroke: e.value.style.chart.graduations.stroke,
                "stroke-linecap": "round"
              }, null, 8, qt)) : c("", !0),
              e.value.style.chart.graduations.show && ["both", "left"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
                key: 2,
                x1: t.x,
                x2: t.x + 2.5,
                y1: t.qYMore,
                y2: t.qYMore,
                "stroke-width": e.value.style.chart.graduations.strokeWidth / 2,
                stroke: e.value.style.chart.graduations.stroke,
                "stroke-linecap": "round"
              }, null, 8, Rt)) : c("", !0)
            ], 64)) : c("", !0),
            e.value.style.chart.graduations.show && ["both", "right"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
              key: 2,
              x1: s.value.right,
              x2: s.value.right - 10,
              y1: t.y,
              y2: t.y,
              "stroke-width": e.value.style.chart.graduations.strokeWidth,
              stroke: e.value.style.chart.graduations.stroke,
              "stroke-linecap": "round"
            }, null, 8, Gt)) : c("", !0),
            e.value.style.chart.graduations.showIntermediate ? (r(), n(E, { key: 3 }, [
              e.value.style.chart.graduations.show && ["both", "right"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
                key: 0,
                x1: s.value.right,
                x2: s.value.right - 5,
                y1: t.halfY,
                y2: t.halfY,
                "stroke-width": e.value.style.chart.graduations.strokeWidth / 2,
                stroke: e.value.style.chart.graduations.stroke,
                "stroke-linecap": "round"
              }, null, 8, Dt)) : c("", !0),
              e.value.style.chart.graduations.show && ["both", "right"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
                key: 1,
                x1: s.value.right,
                x2: s.value.right - 2.5,
                y1: t.qYLess,
                y2: t.qYLess,
                "stroke-width": e.value.style.chart.graduations.strokeWidth / 2,
                stroke: e.value.style.chart.graduations.stroke,
                "stroke-linecap": "round"
              }, null, 8, Bt)) : c("", !0),
              e.value.style.chart.graduations.show && ["both", "right"].includes(e.value.style.chart.graduations.sides) ? (r(), n("line", {
                key: 2,
                x1: s.value.right,
                x2: s.value.right - 2.5,
                y1: t.qYMore,
                y2: t.qYMore,
                "stroke-width": e.value.style.chart.graduations.strokeWidth / 2,
                stroke: e.value.style.chart.graduations.stroke,
                "stroke-linecap": "round"
              }, null, 8, jt)) : c("", !0)
            ], 64)) : c("", !0)
          ]))), 128)),
          p("rect", {
            class: V({ "vue-ui-thermometer-temperature": e.value.style.chart.animation.use }),
            x: s.value.left,
            y: s.value.top,
            height: L.value,
            width: s.value.thermoWidth,
            fill: "#FFFFFF66"
          }, null, 10, Xt)
        ], 8, Nt),
        e.value.style.chart.label.show ? (r(), n("g", Kt, [
          u(le) ? (r(), n("rect", {
            key: 0,
            x: s.value.left - 60,
            y: L.value + s.value.top - A.value / 2,
            width: 50,
            height: A.value,
            fill: "#6A6A6A40",
            rx: "3"
          }, null, 8, Jt)) : (r(), n("text", {
            key: 1,
            class: V({ "vue-ui-thermometer-temperature-value": e.value.style.chart.animation.use, "vue-ui-thermometer-label": !0 }),
            y: L.value + s.value.top + A.value / 3,
            x: s.value.left - 10,
            "text-anchor": "end",
            fill: e.value.style.chart.label.color,
            "font-size": A.value,
            "font-weight": e.value.style.chart.label.bold ? "bold" : "normal"
          }, ot(u(ht)(
            e.value.style.chart.label.formatter,
            I.dataset.value,
            u(vt)({
              p: e.value.style.chart.label.prefix,
              v: I.dataset.value,
              s: e.value.style.chart.label.suffix,
              r: e.value.style.chart.label.rounding
            }),
            { datapoint: I.dataset }
          )), 11, Qt))
        ])) : c("", !0),
        v(o.$slots, "svg", {
          svg: {
            ...ze.value,
            isPrintingImg: u(B) | u(j) | T.value,
            isPrintingSvg: M.value
          }
        }, void 0, !0)
      ], 10, It)),
      o.$slots.watermark ? (r(), n("div", Zt, [
        v(o.$slots, "watermark", y(g({ isPrinting: u(B) || u(j) || T.value || M.value })), void 0, !0)
      ])) : c("", !0),
      o.$slots.source ? (r(), n("div", {
        key: 5,
        ref_key: "source",
        ref: se,
        dir: "auto"
      }, [
        v(o.$slots, "source", {}, void 0, !0)
      ], 512)) : c("", !0),
      v(o.$slots, "skeleton", {}, () => [
        u(le) ? (r(), Y(gt, { key: 0 })) : c("", !0)
      ], !0)
    ], 46, Ot));
  }
}, po = /* @__PURE__ */ Pt(eo, [["__scopeId", "data-v-10ce3bce"]]);
export {
  po as default
};
