// vue-data-ui/dist/vue-ui-accordion-BGHTqIBt.js
import { useCssVars as $, computed as u, ref as r, onMounted as O, watch as z, openBlock as i, createElementBlock as m, createElementVNode as l, normalizeStyle as s, normalizeClass as D, renderSlot as c, normalizeProps as d, mergeProps as N, createBlock as P, createCommentVNode as x, guardReactiveProps as p } from "vue";
import B from "./BaseIcon-CNx8Tzl5.js";
import { u as I } from "./useNestedProp-BBT0mbHh.js";
import { j as w } from "./lib-GwPnJSrY.js";
import { u as A } from "./useConfig-CIgyKZ6R.js";
import { _ as F } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const S = { class: "vue-data-ui-component" }, E = ["id"], V = {
  key: 0,
  class: "vue-ui-accordion-arrow"
}, j = {
  __name: "vue-ui-accordion",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    hideDetails: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["toggle"],
  setup(t, { emit: f }) {
    $((o) => ({
      efd64c80: b.value
    }));
    const { vue_ui_accordion: g } = A(), h = t, C = f, e = u(() => I({
      userConfig: h.config,
      defaultConfig: g
    })), k = u(() => e.value.useCursorPointer), a = r(e.value.open), y = r(w()), n = r(null), v = r(0);
    O(() => {
      n.value.open = e.value.open;
    }), z(() => e.value.open, (o) => {
      n.value.open = o;
    });
    function _() {
      (v.value > 0 || !e.value.open) && (a.value = !a.value), v.value += 1, C("toggle");
    }
    const b = u(() => `${e.value.maxHeight}px`);
    return (o, G) => (i(), m("div", S, [
      l("details", {
        id: `details_${y.value}`,
        ref_key: "details",
        ref: n,
        onToggle: _
      }, [
        l("summary", {
          class: D({ "vue-ui-accordion-headless": t.hideDetails }),
          style: s({ cursor: k.value ? "pointer" : "default" })
        }, [
          l("div", {
            class: "vue-ui-accordion-head",
            style: s(`background:${e.value.head.backgroundColor};padding:${e.value.head.padding}; ${t.hideDetails ? "height: 0px !important; padding: 0 !important;" : ""}`)
          }, [
            t.hideDetails ? x("", !0) : (i(), m("div", V, [
              e.value.head.useArrowSlot ? c(o.$slots, "arrow", d(N({ key: 0 }, { backgroundColor: e.value.head.backgroundColor, color: e.value.head.color, iconColor: e.value.head.iconColor, isOpen: a.value })), void 0, !0) : (i(), P(B, {
                key: 1,
                name: e.value.head.icon,
                stroke: e.value.head.iconColor,
                size: e.value.head.iconSize
              }, null, 8, ["name", "stroke", "size"]))
            ])),
            c(o.$slots, "title", d(p({ color: e.value.head.color, isOpen: a.value })), void 0, !0)
          ], 4)
        ], 6)
      ], 40, E),
      l("div", {
        class: "vue-ui-accordion-content",
        style: s(`background:${e.value.body.backgroundColor};color:${e.value.body.color}`)
      }, [
        c(o.$slots, "content", d(p({ backgroundColor: e.value.body.backgroundColor, color: e.value.body.color, isOpen: a.value })), void 0, !0)
      ], 4)
    ]));
  }
}, q = /* @__PURE__ */ F(j, [["__scopeId", "data-v-6254f78a"]]);
export {
  q as default
};
