// vue-data-ui/dist/vue-ui-history-plot-Bs96PebH.js
import { defineAsyncComponent as F, ref as v, shallowRef as Ge, computed as h, onMounted as qt, onBeforeUnmount as Zt, toRefs as Jt, watch as ae, watchEffect as Kt, openBlock as s, createElementBlock as n, unref as i, normalizeStyle as $, normalizeClass as X, renderSlot as d, createBlock as N, withCtx as y, normalizeProps as A, guardReactiveProps as S, createCommentVNode as f, createElementVNode as L, createSlots as qe, createVNode as ke, Fragment as x, renderList as b, toDisplayString as M, Teleport as Qt, resolveDynamicComponent as el, mergeProps as tl, createTextVNode as Ze, nextTick as Je } from "vue";
import { j as Ke, o as Qe, q as et, r as ll, t as tt, n as al, p as oe, s as ol, D as lt, k as H, l as E, X as sl, x as nl, w as rl, Z as ul, W as il, a2 as B, m as cl, a0 as vl, a1 as hl, T as dl } from "./lib-GwPnJSrY.js";
import { t as at, u as yl } from "./useResponsive-DfdjqQps.js";
import { u as fl } from "./useConfig-CIgyKZ6R.js";
import { u as pl, B as gl } from "./BaseScanner-BhhkKVoj.js";
import { u as ml } from "./usePrinter-DiK5R6k-.js";
import { u as xl } from "./useSvgExport-ByUukOZt.js";
import { u as Ce } from "./useNestedProp-BBT0mbHh.js";
import { u as bl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as kl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Cl } from "./useChartAccessibility-9icAAmYg.js";
import { u as wl } from "./useTimeLabelCollider-CIsgDrl9.js";
import Al from "./img-BYYzY_QU.js";
import Sl from "./Title-BfyrP-jp.js";
import Tl from "./vue_ui_history_plot-BCKk1CrQ.js";
import $l from "./Legend-1LMMsPxL.js";
import { B as Ll } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as _l } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const zl = ["id"], Pl = ["id"], Ol = ["xmlns", "viewBox"], Il = ["x", "y", "width", "height"], Nl = { key: 1 }, Ml = ["id"], Dl = ["stop-color"], Fl = ["stop-color"], Xl = ["stop-color"], Bl = { key: 2 }, Rl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Yl = { key: 3 }, Hl = ["stroke", "stroke-width", "x1", "x2", "y1", "y2"], El = ["x", "y", "fill", "font-size"], Wl = ["transform", "font-size", "fill", "font-weight"], Ul = { key: 6 }, jl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Vl = { key: 7 }, Gl = ["stroke", "stroke-width", "x1", "x2", "y1", "y2"], ql = ["transform", "fill", "font-size", "text-anchor"], Zl = ["x", "y", "font-size", "fill", "font-weight"], Jl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Kl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Ql = ["d", "stroke", "stroke-width"], ea = ["d", "stroke", "stroke-width"], ta = ["cx", "cy", "fill", "r"], la = ["cx", "cy", "fill", "r", "stroke", "stroke-width"], aa = { key: 1 }, oa = ["x", "y", "font-size", "fill", "font-weight"], sa = ["x", "y", "font-size", "fill", "font-weight", "innerHTML"], na = { key: 2 }, ra = ["x", "y", "font-size", "font-weight", "fill"], ua = ["cx", "cy", "r", "onMouseenter", "onMouseleave", "onClick"], ia = {
  key: 4,
  class: "vue-data-ui-watermark"
}, ca = ["id"], va = ["onClick"], ha = ["innerHTML"], da = {
  __name: "vue-ui-history-plot",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "selectDatapoint", "copyAlt"],
  setup(ot, { expose: st, emit: nt }) {
    const rt = F(() => import("./Tooltip-DL3QgjZf.js")), ut = F(() => import("./BaseIcon-CNx8Tzl5.js")), it = F(() => import("./vue-ui-accordion-BGHTqIBt.js")), ct = F(() => import("./DataTable-Bo1pmAqG.js")), vt = F(() => import("./PenAndPaper-DWV1FMIi.js")), ht = F(() => import("./UserOptions-CrLqEimB.js")), dt = F(() => import("./PackageVersion-DN9d0ME5.js")), yt = F(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_history_plot: ft } = fl(), { isThemeValid: pt, warnInvalidTheme: gt } = bl(), C = ot, z = v(null), we = v(null), Ae = v(0), Se = v(null), Te = v(0), $e = v(0), Le = v(0), D = Ge(null), R = Ge(null), _e = v(null), P = v(Ke()), se = v(!1), ne = v(""), p = v([]), W = v(!1), T = v(null), U = v(!1), ze = v(null), Pe = v(!1), j = v(null), re = v(null), q = v(!1), Z = v(!1), ue = v(null), ie = v(null), J = v(null), ce = v(null), K = h({
      get: () => !!C.dataset && C.dataset.length,
      set: (t) => t
    }), ve = nt;
    qt(() => {
      Pe.value = !0, Ie();
    });
    const Oe = h(() => !!e.value.debug);
    function Ie() {
      if (Qe(C.dataset) ? (et({
        componentName: "VueUiHistoryPlot",
        type: "dataset",
        debug: Oe.value
      }), de.value = !0) : C.dataset.forEach((t, a) => {
        ll({
          datasetObject: t,
          requiredAttributes: ["name", "values"]
        }).forEach((l) => {
          K.value = !1, et({
            componentName: "VueUiHistoryPlot",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: Oe.value
          });
        });
      }), Qe(C.dataset) || (de.value = e.value.loading), e.value.responsive) {
        const t = at(() => {
          const { width: a, height: l } = yl({
            chart: z.value,
            title: e.value.style.chart.title.text ? we.value : null,
            noTitle: _e.value,
            legend: e.value.style.chart.legend.show ? Se.value : null,
            source: ze.value
          });
          requestAnimationFrame(() => {
            g.value.width = a, g.value.height = l, e.value.responsiveProportionalSizing ? (r.value.plots = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.plots.radius,
              threshold: 3,
              fallback: 3
            }), r.value.indexLabels = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.plots.indexLabels.fontSize,
              threshold: 6,
              fallback: 6
            }), r.value.labels = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.plots.labels.fontSize,
              threshold: 6,
              fallback: 6
            }), r.value.xAxisLabels = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.axes.x.labels.fontSize,
              threshold: 6,
              fallback: 6
            }), r.value.xAxisName = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.axes.x.name.fontSize,
              threshold: 6,
              fallback: 6
            }), r.value.yAxisLabels = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.axes.y.labels.fontSize,
              threshold: 6,
              fallback: 6
            }), r.value.yAxisName = B({
              relator: Math.min(a, l),
              adjuster: 600,
              source: e.value.style.chart.axes.y.name.fontSize,
              threshold: 6,
              fallback: 6
            })) : (r.value.plots = e.value.style.chart.plots.radius, r.value.indexLabels = e.value.style.chart.plots.indexLabels.fontSize, r.value.labels = e.value.style.chart.plots.labels.fontSize, r.value.xAxisLabels = e.value.style.chart.axes.x.labels.fontSize, r.value.xAxisName = e.value.style.chart.axes.x.name.fontSize, r.value.yAxisLabels = e.value.style.chart.axes.y.labels.fontSize, r.value.yAxisName = e.value.style.chart.axes.y.name.fontSize);
          });
        });
        D.value && (R.value && D.value.unobserve(R.value), D.value.disconnect()), D.value = new ResizeObserver(t), R.value = z.value.parentNode, D.value.observe(R.value);
      }
    }
    Zt(() => {
      D.value && (R.value && D.value.unobserve(R.value), D.value.disconnect());
    });
    function he() {
      const t = Ce({
        userConfig: C.config,
        defaultConfig: ft
      });
      let a = {};
      const l = t.theme;
      if (l)
        if (!pt.value(t))
          gt(t), a = t;
        else {
          const u = Ce({
            userConfig: Tl[l] || C.config,
            defaultConfig: t
          });
          a = {
            ...Ce({
              userConfig: C.config,
              defaultConfig: u
            }),
            customPalette: t.customPalette.length ? t.customPalette : cl[l] || oe
          };
        }
      else
        a = t;
      return a;
    }
    const e = v(he()), Y = h(() => e.value.userOptions.useCursorPointer), mt = h(() => tt({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            axes: {
              x: {
                scaleMin: 0,
                scaleMax: 10,
                labels: { show: !1 },
                name: { text: "" }
              },
              y: {
                scaleMin: 0,
                scaleMax: 10,
                labels: { show: !1 },
                name: { text: "" }
              }
            },
            grid: {
              xAxis: {
                stroke: "#6A6A6A"
              },
              horizontalLines: {
                stroke: "#6A6A6A50"
              },
              yAxis: {
                stroke: "#6A6A6A"
              },
              verticalLines: {
                stroke: "#6A6A6A50"
              }
            },
            legend: {
              backgroundColor: "transparent"
            },
            paths: {
              useSerieColor: !1,
              stroke: "#6A6A6A"
            },
            plots: {
              stroke: "#6A6A6A",
              indexLabels: { show: !1 },
              labels: { show: !1 }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: O, FINAL_DATASET: xt, manualLoading: de } = pl({
      ...Jt(C),
      FINAL_CONFIG: e,
      prepareConfig: he,
      skeletonDataset: C.config?.skeletonDataset ?? [
        {
          name: "",
          color: "#CACACA",
          values: [
            { label: "", x: 1, y: 9 },
            { label: "", x: 4, y: 1 },
            { label: "", x: 7, y: 9 },
            { label: "", x: 9, y: 4 }
          ]
        }
      ],
      skeletonConfig: tt({
        defaultConfig: e.value,
        userConfig: mt.value
      })
    }), { userOptionsVisible: ye, setUserOptionsVisibility: Ne, keepUserOptionState: Me } = kl({ config: e.value }), { svgRef: fe } = Cl({ config: e.value.style.chart.title });
    ae(() => C.config, (t) => {
      O.value || (e.value = he()), ye.value = !e.value.userOptions.showOnChartHover, Ie(), Ae.value += 1, $e.value += 1, Te.value += 1, g.value.height = e.value.style.chart.height, g.value.width = e.value.style.chart.width, r.value.plots = e.value.style.chart.plots.radius, r.value.indexLabels = e.value.style.chart.plots.indexLabels.fontSize, r.value.labels = e.value.style.chart.plots.labels.fontSize, r.value.xAxisLabels = e.value.style.chart.axes.x.labels.fontSize, r.value.xAxisName = e.value.style.chart.axes.x.name.fontSize, r.value.yAxisLabels = e.value.style.chart.axes.y.labels.fontSize, r.value.yAxisName = e.value.style.chart.axes.y.name.fontSize, w.value.showTable = e.value.table.show, w.value.showTooltip = e.value.style.chart.tooltip.show;
    }, { deep: !0 }), ae(() => C.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (de.value = !1);
    }, { deep: !0 });
    const { isPrinting: pe, isImaging: ge, generatePdf: De, generateImage: Fe } = ml({
      elementId: `history_plot_${P.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-history-plot",
      options: e.value.userOptions.print
    }), bt = h(() => e.value.userOptions.show && !e.value.style.chart.title.text), kt = h(() => al(e.value.customPalette)), w = v({
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    ae(e, () => {
      w.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const g = v({
      height: e.value.style.chart.height,
      width: e.value.style.chart.width
    }), Ct = h(() => g.value.width), wt = h(() => g.value.height);
    function At() {
      let t = 0;
      ce.value && (t = Array.from(ce.value.querySelectorAll("text")).reduce((u, o) => {
        const k = o.getComputedTextLength();
        return k > u ? k : u;
      }, 0));
      const a = ie.value ? ie.value.getBoundingClientRect().width : 0;
      return t + a + (a ? 24 : 0);
    }
    const Xe = v(0), St = at((t) => {
      Xe.value = t;
    });
    Kt((t) => {
      const a = J.value;
      if (!a) return;
      const l = new ResizeObserver((u) => {
        St(u[0].contentRect.height);
      });
      l.observe(a), t(() => l.disconnect());
    });
    const Be = h(() => {
      let t = 0;
      ue.value && (t = ue.value.getBBox().height + r.value.xAxisName / 2);
      let a = 0;
      return J.value && (a = Xe.value), t + a;
    }), c = h(() => {
      const t = e.value.style.chart.padding.left, a = e.value.style.chart.padding.top, l = At(), u = e.value.style.chart.plots.radius;
      return {
        left: t + l + e.value.style.chart.axes.y.name.offsetX,
        top: a + u,
        right: g.value.width - e.value.style.chart.padding.right - u - e.value.style.chart.axes.y.name.offsetX,
        bottom: g.value.height - e.value.style.chart.padding.bottom - Be.value - u - e.value.style.chart.axes.x.name.offsetY,
        width: g.value.width - t - e.value.style.chart.padding.right - l - u - e.value.style.chart.axes.y.name.offsetX,
        height: g.value.height - a - e.value.style.chart.padding.bottom - Be.value - u * 2 - e.value.style.chart.axes.x.name.offsetY
      };
    }), r = v({
      plots: e.value.style.chart.plots.radius,
      indexLabels: e.value.style.chart.plots.indexLabels.fontSize,
      labels: e.value.style.chart.plots.labels.fontSize,
      xAxisLabels: e.value.style.chart.axes.x.labels.fontSize,
      xAxisName: e.value.style.chart.axes.x.name.fontSize,
      yAxisLabels: e.value.style.chart.axes.y.labels.fontSize,
      yAxisName: e.value.style.chart.axes.y.name.fontSize
    }), _ = h(() => xt.value.map((t, a) => ({
      ...t,
      color: t.color ? ol(t.color) : kt.value[a] || oe[a] || oe[a % oe.length],
      seriesIndex: a
    }))), Tt = h(() => Math.max(..._.value.filter((t) => !p.value.includes(t.seriesIndex)).flatMap((t) => t.values.map((a) => a.x)))), $t = h(() => {
      const t = Math.min(..._.value.filter((a) => !p.value.includes(a.seriesIndex)).flatMap((a) => a.values.map((l) => l.x)));
      return t < 0 ? t : 0;
    }), Lt = h(() => Math.max(..._.value.filter((t) => !p.value.includes(t.seriesIndex)).flatMap((t) => t.values.map((a) => a.y)))), _t = h(() => {
      const t = Math.min(..._.value.filter((a) => !p.value.includes(a.seriesIndex)).flatMap((a) => a.values.map((l) => l.y)));
      return t < 0 ? t : 0;
    }), m = h(() => {
      const t = lt(
        e.value.style.chart.axes.x.scaleMin ?? $t.value,
        e.value.style.chart.axes.x.scaleMax ?? Tt.value,
        e.value.style.chart.axes.x.ticks
      ), a = lt(
        e.value.style.chart.axes.y.scaleMin ?? _t.value,
        e.value.style.chart.axes.y.scaleMax ?? Lt.value,
        e.value.style.chart.axes.y.ticks
      );
      return {
        x: t,
        y: a,
        tickX: t.ticks.map((l) => ({
          x: c.value.left + (l - t.min) / (t.max - t.min) * c.value.width,
          y1: c.value.top,
          y2: c.value.bottom,
          value: l
        })),
        tickY: a.ticks.map((l) => ({
          y: c.value.bottom - (l - a.min) / (a.max - a.min) * c.value.height,
          x1: c.value.left,
          x2: c.value.right,
          value: l
        }))
      };
    });
    function zt(t) {
      const a = m.value.x.min < 0 ? Math.abs(m.value.x.min) : m.value.x.min > 0 ? -m.value.x.min : 0;
      return c.value.left + ((t || 0) + a) / (m.value.x.max + a) * c.value.width;
    }
    function Pt(t) {
      const a = m.value.y.min < 0 ? Math.abs(m.value.y.min) : m.value.y.min > 0 ? -m.value.y.min : 0;
      return c.value.bottom - ((t || 0) + a) / (m.value.y.max + a) * c.value.height;
    }
    const I = h(() => _.value.filter((t) => !p.value.includes(t.seriesIndex)).map((t, a) => {
      const l = t.values.map((o, k) => ({
        valueX: o.x,
        valueY: o.y,
        label: o.label,
        x: zt(o.x),
        y: Pt(o.y),
        color: t.color,
        seriesName: t.name,
        id: Ke()
      })), u = l.map((o) => `${o.x},${o.y} `).join("").trim();
      return {
        ...t,
        plots: l,
        path: `M${u}`
      };
    }));
    function Ot() {
      p.value.length ? p.value = [] : V.value.forEach((t) => {
        p.value.push(t.seriesIndex);
      });
    }
    function Q(t) {
      p.value.includes(t) ? p.value = p.value.filter((a) => a !== t) : p.value.push(t);
    }
    function Re(t) {
      if (!_.value.length)
        return e.value.debug && console.warn("VueUiHistoryPlot - There are no series to show."), null;
      const a = _.value.find((l) => l.name === t);
      return a || (e.value.debug && console.warn(`VueUiHistoryPlot - Series name not found "${t}"`), null);
    }
    function It(t) {
      const a = Re(t);
      a !== null && p.value.includes(a.seriesIndex) && Q(a.seriesIndex);
    }
    function Nt(t) {
      const a = Re(t);
      a !== null && (p.value.includes(a.seriesIndex) || Q(a.seriesIndex));
    }
    const V = h(() => _.value.map((t) => ({
      ...t,
      opacity: p.value.includes(t.seriesIndex) ? 0.5 : 1,
      segregate: () => Q(t.seriesIndex),
      isSegregated: p.value.includes(t.seriesIndex),
      shape: "circle"
    }))), Mt = h(() => ({
      cy: "history-plot-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    }));
    function Dt({ datapoint: t, plotIndex: a, seriesIndex: l }) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: { ...t, plotIndex: a, seriesIndex: l }, seriesIndex: l }), ve("selectDatapoint", t);
    }
    function Ye(t) {
      ve("selectLegend", {
        ...t,
        isSegregated: !t.isSegregated,
        opacity: t.isSegregated ? 1 : 0.5
      });
    }
    function He(t) {
      W.value = t, Le.value += 1;
    }
    const me = v(null);
    function Ft({ datapoint: t, plotIndex: a, seriesIndex: l }) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: { ...t, plotIndex: a, seriesIndex: l }, seriesIndex: l }), se.value = !1, T.value = null;
    }
    function Xt({ datapoint: t, plotIndex: a, seriesIndex: l }) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: { ...t, plotIndex: a, seriesIndex: l }, seriesIndex: l }), me.value = { datapoint: t, seriesIndex: l, plotIndex: a, config: e.value, series: _.value }, T.value = t;
      const u = e.value.style.chart.tooltip.customFormat;
      if (U.value = !1, dl(u))
        try {
          const o = u({
            seriesIndex: l,
            datapoint: t,
            plotIndex: a,
            series: _.value,
            config: e.value
          });
          typeof o == "string" && (ne.value = o, U.value = !0);
        } catch {
          console.warn("Custom format cannot be applied."), U.value = !1;
        }
      if (!U.value) {
        let o = "";
        o += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};margin-bottom:3px;padding-bottom:6px;"><svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="none" fill="${t.color}"/></svg><span>${t.seriesName}</span></div>`, o += `<div>${t.label}</div>`, o += `<div>${e.value.style.chart.axes.x.name.text || "x"}: ${H(
          e.value.style.chart.axes.x.labels.formatter,
          t.valueX,
          E({
            p: e.value.style.chart.axes.x.labels.prefix,
            v: t.valueX,
            s: e.value.style.chart.axes.x.labels.suffix,
            r: e.value.style.chart.axes.x.labels.rounding
          })
        )}</div>`, o += `<div>${e.value.style.chart.axes.y.name.text || "y"}: ${H(
          e.value.style.chart.axes.y.labels.formatter,
          t.valueY,
          E({
            p: e.value.style.chart.axes.y.labels.prefix,
            v: t.valueY,
            s: e.value.style.chart.axes.y.labels.suffix,
            r: e.value.style.chart.axes.y.labels.rounding
          })
        )}</div>`, ne.value = `<div>${o}</div>`;
      }
      se.value = !0;
    }
    h(() => ({ head: I.value.map((a) => ({
      name: a.name,
      color: a.color
    })) }));
    const ee = h(() => {
      const t = [
        e.value.table.columnNames.series,
        e.value.table.columnNames.datapoint,
        e.value.table.columnNames.x,
        e.value.table.columnNames.y
      ], a = I.value.flatMap((u) => u.plots.map((o) => [
        {
          color: o.color,
          name: o.seriesName
        },
        o.label,
        H(
          e.value.style.chart.axes.x.labels.formatter,
          o.valueX,
          E({
            p: e.value.style.chart.axes.x.labels.prefix,
            v: o.valueX,
            s: e.value.style.chart.axes.x.labels.suffix,
            r: e.value.style.chart.axes.x.labels.rounding
          })
        ),
        H(
          e.value.style.chart.axes.y.labels.formatter,
          o.valueY,
          E({
            p: e.value.style.chart.axes.y.labels.prefix,
            v: o.valueY,
            s: e.value.style.chart.axes.y.labels.suffix,
            r: e.value.style.chart.axes.y.labels.rounding
          })
        )
      ])), l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        head: t,
        body: a,
        config: l,
        colNames: t
      };
    });
    function xe(t = null) {
      Je(() => {
        const a = [
          [e.value.table.columnNames.series],
          [e.value.table.columnNames.datapoint],
          [e.value.table.columnNames.x],
          [e.value.table.columnNames.y]
        ], l = I.value.flatMap((k) => k.plots.map((le) => [
          [le.seriesName],
          [le.label],
          [le.valueX],
          [le.valueY]
        ])), u = [
          [e.value.style.chart.title.text],
          [e.value.style.chart.title.subtitle.text],
          a
        ].concat(l), o = vl(u);
        t ? t(o) : hl({
          csvContent: o,
          title: e.value.style.chart.title.text || "vue-ui-history-plot"
        });
      });
    }
    const te = v(!1);
    function be() {
      te.value = !te.value;
    }
    function Ee() {
      w.value.showTable = !w.value.showTable;
    }
    function We() {
      w.value.showTooltip = !w.value.showTooltip;
    }
    function Bt() {
      return I.value;
    }
    async function Rt({ scale: t = 2 } = {}) {
      if (!z.value) return;
      const { width: a, height: l } = z.value.getBoundingClientRect(), u = a / l, { imageUri: o, base64: k } = await Al({ domElement: z.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: o,
        base64: k,
        title: e.value.style.chart.title.text,
        width: a,
        height: l,
        aspectRatio: u
      };
    }
    const Yt = h(() => m.value.tickX), Ht = h(() => ({ start: 0, end: m.value.tickX.length }));
    wl({
      timeLabelsEls: J,
      timeLabels: Yt,
      slicer: Ht,
      configRef: e,
      rotationPath: ["style", "chart", "axes", "x", "labels", "rotation"],
      autoRotatePath: ["style", "chart", "axes", "x", "labels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: Ct,
      height: wt,
      targetClass: ".vue-ui-history-plot-x-axis-scale",
      rotation: e.value.style.chart.axes.x.labels.autoRotate.angle
    });
    const G = h(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = w.value.showTable;
      return {
        component: t ? yt : it,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: W.value,
          fullscreenParent: z.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8)
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ae(() => w.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && j.value ? j.value.open() : "close" in j.value && j.value.close());
    });
    function Ue() {
      w.value.showTable = !1, re.value && re.value.setTableIconState(!1);
    }
    const Et = h(() => e.value.style.chart.backgroundColor), Wt = h(() => e.value.style.chart.legend), Ut = h(() => e.value.style.chart.title), { exportSvg: jt, getSvg: Vt } = xl({
      svg: fe,
      title: Ut,
      legend: Wt,
      legendItems: V,
      backgroundColor: Et
    });
    async function je({ isCb: t }) {
      Z.value = !0, await Je();
      try {
        if (t) {
          const { blob: a, url: l, text: u, dataUrl: o } = await Vt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: u, dataUrl: o }));
        } else
          await Promise.resolve(jt());
      } finally {
        Z.value = !1;
      }
    }
    function Gt(t) {
      if (t?.stage === "start") {
        q.value = !0;
        return;
      }
      if (t?.stage === "end") {
        q.value = !1;
        return;
      }
      Fe();
    }
    async function Ve() {
      if (ve("copyAlt", {
        config: e.value,
        dataset: I.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: I.value
      }));
    }
    return st({
      getData: Bt,
      getImage: Rt,
      generatePdf: De,
      generateCsv: xe,
      generateImage: Fe,
      generateSvg: je,
      hideSeries: Nt,
      showSeries: It,
      toggleTable: Ee,
      toggleTooltip: We,
      toggleAnnotator: be,
      toggleFullscreen: He,
      copyAlt: Ve
    }), (t, a) => (s(), n("div", {
      id: `history_plot_${P.value}`,
      ref_key: "historyPlotChart",
      ref: z,
      class: X({ "vue-data-ui-component": !0, "vue-ui-history-plot": !0, "vue-data-ui-wrapper-fullscreen": W.value }),
      style: $(`background:${e.value.style.chart.backgroundColor};color:${e.value.style.chart.color};font-family:${e.value.style.fontFamily}; position: relative; ${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: a[2] || (a[2] = () => i(Ne)(!0)),
      onMouseleave: a[3] || (a[3] = () => i(Ne)(!1))
    }, [
      d(t.$slots, "userConfig", {}, void 0, !0),
      e.value.userOptions.buttons.annotator ? (s(), N(i(vt), {
        key: 0,
        svgRef: i(fe),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: te.value,
        isCursorPointer: Y.value,
        onClose: be
      }, {
        "annotator-action-close": y(() => [
          d(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": y(({ color: l }) => [
          d(t.$slots, "annotator-action-color", A(S({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": y(({ mode: l }) => [
          d(t.$slots, "annotator-action-draw", A(S({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": y(({ disabled: l }) => [
          d(t.$slots, "annotator-action-undo", A(S({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": y(({ disabled: l }) => [
          d(t.$slots, "annotator-action-redo", A(S({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": y(({ disabled: l }) => [
          d(t.$slots, "annotator-action-delete", A(S({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : f("", !0),
      bt.value ? (s(), n("div", {
        key: 1,
        ref_key: "noTitle",
        ref: _e,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : f("", !0),
      e.value.style.chart.title.text ? (s(), n("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: we,
        class: "vue-ui-xy-title",
        style: $(`font-family:${e.value.style.fontFamily}`)
      }, [
        (s(), N(Sl, {
          key: `title_${Ae.value}`,
          config: {
            title: {
              cy: "history-plot-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "history-plot-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 4)) : f("", !0),
      L("div", {
        id: `legend-top-${P.value}`
      }, null, 8, Pl),
      e.value.userOptions.show && K.value && (i(Me) || i(ye)) ? (s(), N(i(ht), {
        ref_key: "userOptionsRef",
        ref: re,
        key: `user_option_${Le.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: i(pe),
        isImaging: i(ge),
        uid: P.value,
        hasTooltip: e.value.style.chart.tooltip.show && e.value.userOptions.buttons.tooltip,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasLabel: !1,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: W.value,
        chartElement: z.value,
        position: e.value.userOptions.position,
        isTooltip: w.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: te.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: Y.value,
        onToggleFullscreen: He,
        onGeneratePdf: i(De),
        onGenerateCsv: xe,
        onGenerateImage: Gt,
        onGenerateSvg: je,
        onToggleTable: Ee,
        onToggleTooltip: We,
        onToggleAnnotator: be,
        onCopyAlt: Ve,
        style: $({
          visibility: i(Me) ? i(ye) ? "visible" : "hidden" : "visible"
        })
      }, qe({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: y(({ isOpen: l, color: u }) => [
            d(t.$slots, "menuIcon", A(S({ isOpen: l, color: u })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: y(() => [
            d(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: y(() => [
            d(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: y(() => [
            d(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: y(() => [
            d(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: y(() => [
            d(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: y(() => [
            d(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: y(({ toggleFullscreen: l, isFullscreen: u }) => [
            d(t.$slots, "optionFullscreen", A(S({ toggleFullscreen: l, isFullscreen: u })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: y(({ toggleAnnotator: l, isAnnotator: u }) => [
            d(t.$slots, "optionAnnotator", A(S({ toggleAnnotator: l, isAnnotator: u })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: y(({ altCopy: l }) => [
            d(t.$slots, "optionAltCopy", A(S({ altCopy: l })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "isTooltip", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : f("", !0),
      (s(), n("svg", {
        ref_key: "svgRef",
        ref: fe,
        xmlns: i(sl),
        viewBox: `0 0 ${g.value.width < 0 ? 0.1 : g.value.width} ${g.value.height < 0 ? 0.1 : g.value.height}`,
        style: $(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        ke(i(dt)),
        t.$slots["chart-background"] ? (s(), n("foreignObject", {
          key: 0,
          x: c.value.left,
          y: c.value.top,
          width: c.value.width <= 0 ? 10 : c.value.width,
          height: c.value.height <= 0 ? 10 : c.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          d(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, Il)) : f("", !0),
        e.value.style.chart.plots.gradient.show ? (s(), n("defs", Nl, [
          (s(!0), n(x, null, b(I.value, (l, u) => (s(), n("radialGradient", {
            id: `gradient_${u}_${P.value}`,
            fy: "30%"
          }, [
            L("stop", {
              offset: "10%",
              "stop-color": i(nl)(l.color, e.value.style.chart.plots.gradient.intensity / 100)
            }, null, 8, Dl),
            L("stop", {
              offset: "90%",
              "stop-color": i(rl)(l.color, 0.1)
            }, null, 8, Fl),
            L("stop", {
              offset: "100%",
              "stop-color": l.color
            }, null, 8, Xl)
          ], 8, Ml))), 256))
        ])) : f("", !0),
        e.value.style.chart.grid.verticalLines.show ? (s(), n("g", Bl, [
          (s(!0), n(x, null, b(m.value.tickX, (l) => (s(), n("line", {
            x1: l.x,
            x2: l.x,
            y1: l.y1,
            y2: l.y2,
            stroke: e.value.style.chart.grid.verticalLines.stroke,
            "stroke-width": e.value.style.chart.grid.verticalLines.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, Rl))), 256))
        ])) : f("", !0),
        e.value.style.chart.axes.y.labels.show ? (s(), n("g", Yl, [
          (s(!0), n(x, null, b(m.value.tickY, (l) => (s(), n("line", {
            stroke: e.value.style.chart.grid.verticalLines.stroke,
            "stroke-width": e.value.style.chart.grid.verticalLines.strokeWidth,
            "stroke-linecap": "round",
            x1: c.value.left - 5,
            x2: c.value.left,
            y1: l.y,
            y2: l.y
          }, null, 8, Hl))), 256))
        ])) : f("", !0),
        e.value.style.chart.axes.y.labels.show ? (s(), n("g", {
          key: 4,
          ref_key: "yAxisScales",
          ref: ce
        }, [
          (s(!0), n(x, null, b(m.value.tickY, (l) => (s(), n("text", {
            x: c.value.left + e.value.style.chart.axes.y.labels.offsetX - 4 - e.value.style.chart.plots.radius,
            y: l.y + r.value.yAxisLabels / 3,
            fill: e.value.style.chart.axes.y.labels.color,
            "font-size": r.value.yAxisLabels,
            "text-anchor": "end"
          }, M(i(H)(
            e.value.style.chart.axes.y.labels.formatter,
            l.value,
            i(E)({
              p: e.value.style.chart.axes.y.labels.prefix,
              v: l.value,
              s: e.value.style.chart.axes.y.labels.suffix,
              r: e.value.style.chart.axes.y.labels.rounding
            })
          )), 9, El))), 256))
        ], 512)) : f("", !0),
        e.value.style.chart.axes.y.name.text ? (s(), n("text", {
          key: 5,
          ref_key: "yAxisLabel",
          ref: ie,
          transform: `translate(${r.value.yAxisName}, ${g.value.height / 2 + e.value.style.chart.axes.y.name.offsetY}), rotate(-90)`,
          "font-size": r.value.yAxisName,
          fill: e.value.style.chart.axes.y.name.color,
          "font-weight": e.value.style.chart.axes.y.name.bold ? "bold" : "normal",
          "text-anchor": "middle"
        }, M(e.value.style.chart.axes.y.name.text), 9, Wl)) : f("", !0),
        e.value.style.chart.grid.horizontalLines.show ? (s(), n("g", Ul, [
          (s(!0), n(x, null, b(m.value.tickY, (l) => (s(), n("line", {
            x1: l.x1,
            x2: l.x2,
            y1: l.y,
            y2: l.y,
            stroke: e.value.style.chart.grid.horizontalLines.stroke,
            "stroke-width": e.value.style.chart.grid.horizontalLines.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, jl))), 256))
        ])) : f("", !0),
        e.value.style.chart.axes.x.labels.show ? (s(), n("g", Vl, [
          (s(!0), n(x, null, b(m.value.tickX, (l) => (s(), n("line", {
            stroke: e.value.style.chart.grid.verticalLines.stroke,
            "stroke-width": e.value.style.chart.grid.verticalLines.strokeWidth,
            "stroke-linecap": "round",
            x1: l.x,
            x2: l.x,
            y1: c.value.bottom,
            y2: c.value.bottom + 5
          }, null, 8, Gl))), 256))
        ])) : f("", !0),
        e.value.style.chart.axes.x.labels.show ? (s(), n("g", {
          key: 8,
          ref_key: "xAxisScales",
          ref: J
        }, [
          (s(!0), n(x, null, b(m.value.tickX, (l) => (s(), n("text", {
            class: "vue-ui-history-plot-x-axis-scale",
            transform: `translate(${l.x}, ${c.value.bottom + e.value.style.chart.axes.x.labels.offsetY + r.value.xAxisLabels + e.value.style.chart.plots.radius}), rotate(${e.value.style.chart.axes.x.labels.rotation})`,
            fill: e.value.style.chart.axes.x.labels.color,
            "font-size": r.value.xAxisLabels,
            "text-anchor": e.value.style.chart.axes.x.labels.rotation > 0 ? "start" : e.value.style.chart.axes.x.labels.rotation < 0 ? "end" : "middle"
          }, M(i(H)(
            e.value.style.chart.axes.x.labels.formatter,
            l.value,
            i(E)({
              p: e.value.style.chart.axes.x.labels.prefix,
              v: l.value,
              s: e.value.style.chart.axes.x.labels.suffix,
              r: e.value.style.chart.axes.x.labels.rounding
            })
          )), 9, ql))), 256))
        ], 512)) : f("", !0),
        e.value.style.chart.axes.x.name.text ? (s(), n("text", {
          key: 9,
          ref_key: "xAxisLabel",
          ref: ue,
          x: g.value.width / 2 + e.value.style.chart.axes.x.name.offsetX,
          y: g.value.height,
          "font-size": r.value.xAxisName,
          fill: e.value.style.chart.axes.x.name.color,
          "font-weight": e.value.style.chart.axes.x.name.bold ? "bold" : "normal",
          "text-anchor": "middle"
        }, M(e.value.style.chart.axes.x.name.text), 9, Zl)) : f("", !0),
        e.value.style.chart.grid.xAxis.show ? (s(), n("line", {
          key: 10,
          x1: c.value.left,
          x2: c.value.left + c.value.width,
          y1: c.value.bottom,
          y2: c.value.bottom,
          stroke: e.value.style.chart.grid.xAxis.stroke,
          "stroke-width": e.value.style.chart.grid.xAxis.strokeWidth,
          "stroke-linecap": "round"
        }, null, 8, Jl)) : f("", !0),
        e.value.style.chart.grid.yAxis.show ? (s(), n("line", {
          key: 11,
          x1: c.value.left,
          x2: c.value.left,
          y1: c.value.top,
          y2: c.value.bottom,
          stroke: e.value.style.chart.grid.yAxis.stroke,
          "stroke-width": e.value.style.chart.grid.yAxis.strokeWidth,
          "stroke-linecap": "round"
        }, null, 8, Kl)) : f("", !0),
        (s(!0), n(x, null, b(I.value, (l, u) => (s(), n("g", null, [
          e.value.style.chart.paths.show ? (s(), n("g", {
            key: 0,
            style: $({
              opacity: T.value === null ? 1 : 0.3
            })
          }, [
            L("path", {
              d: l.path,
              stroke: e.value.style.chart.backgroundColor,
              "stroke-width": e.value.style.chart.paths.strokeWidth + 4,
              fill: "none",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: X({ animated: e.value.useCssAnimation && !i(O) })
            }, null, 10, Ql),
            L("path", {
              d: l.path,
              stroke: e.value.style.chart.paths.useSerieColor ? l.color : e.value.style.chart.paths.stroke,
              "stroke-width": e.value.style.chart.paths.strokeWidth,
              fill: "none",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              class: X({ animated: e.value.useCssAnimation && !i(O) })
            }, null, 10, ea)
          ], 4)) : f("", !0),
          (s(!0), n(x, null, b(l.plots, (o) => (s(), n("circle", {
            cx: o.x,
            cy: o.y,
            fill: e.value.style.chart.backgroundColor,
            r: r.value.plots,
            stroke: "none"
          }, null, 8, ta))), 256)),
          (s(!0), n(x, null, b(l.plots, (o) => (s(), n("circle", {
            cx: o.x,
            cy: o.y,
            fill: e.value.style.chart.plots.gradient.show ? `url(#gradient_${u}_${P.value})` : o.color,
            r: r.value.plots,
            stroke: e.value.style.chart.plots.stroke,
            "stroke-width": e.value.style.chart.plots.strokeWidth,
            class: X({ animated: e.value.useCssAnimation && !i(O) }),
            style: $({
              opacity: T.value === null || T.value.id === o.id ? 1 : 0.3,
              transition: "opacity 0.2s ease-in-out"
            })
          }, null, 14, la))), 256)),
          e.value.style.chart.plots.labels.show ? (s(), n("g", aa, [
            (s(!0), n(x, null, b(l.plots, (o) => (s(), n("g", null, [
              String(o.label).includes(`
`) ? (s(), n("text", {
                key: 1,
                x: o.x + e.value.style.chart.plots.labels.offsetX,
                y: o.y + e.value.style.chart.plots.labels.offsetY + r.value.plots + r.value.labels,
                "font-size": r.value.labels,
                fill: e.value.style.chart.plots.labels.color,
                "font-weight": e.value.style.chart.plots.labels.bold ? "bold" : "normal",
                "text-anchor": "middle",
                class: X({ animated: e.value.useCssAnimation && !i(O) }),
                style: $({
                  opacity: T.value === null || T.value.id === o.id ? 1 : 0.3,
                  transition: "opacity 0.2s ease-in-out"
                }),
                innerHTML: i(ul)({
                  content: String(o.label),
                  fontSize: r.value.labels,
                  fill: e.value.style.chart.plots.labels.color,
                  x: o.x + e.value.style.chart.plots.labels.offsetX,
                  y: o.y + e.value.style.chart.plots.labels.offsetY + r.value.plots + r.value.labels
                })
              }, null, 14, sa)) : (s(), n("text", {
                key: 0,
                x: o.x + e.value.style.chart.plots.labels.offsetX,
                y: o.y + e.value.style.chart.plots.labels.offsetY + r.value.plots + r.value.labels,
                "font-size": r.value.labels,
                fill: e.value.style.chart.plots.labels.color,
                "font-weight": e.value.style.chart.plots.labels.bold ? "bold" : "normal",
                "text-anchor": "middle",
                class: X({ animated: e.value.useCssAnimation && !i(O) }),
                style: $({
                  opacity: T.value === null || T.value.id === o.id ? 1 : 0.3,
                  transition: "opacity 0.2s ease-in-out"
                })
              }, M(o.label), 15, oa))
            ]))), 256))
          ])) : f("", !0),
          e.value.style.chart.plots.indexLabels.show ? (s(), n("g", na, [
            (s(!0), n(x, null, b(l.plots, (o, k) => (s(), n("text", {
              x: o.x + e.value.style.chart.plots.indexLabels.offsetX,
              y: o.y + e.value.style.chart.plots.indexLabels.offsetY + r.value.indexLabels / 3,
              "font-size": r.value.indexLabels,
              "font-weight": e.value.style.chart.plots.indexLabels.bold ? "bold" : "normal",
              fill: e.value.style.chart.plots.indexLabels.adaptColorToBackground ? i(il)(l.color) : e.value.style.chart.plots.indexLabels.color,
              "text-anchor": "middle",
              class: X({ animated: e.value.useCssAnimation && !i(O) }),
              style: $({
                opacity: T.value === null || T.value.id === o.id ? 1 : 0.3,
                transition: "opacity 0.2s ease-in-out"
              })
            }, M(e.value.style.chart.plots.indexLabels.startAtZero ? k : k + 1), 15, ra))), 256))
          ])) : f("", !0)
        ]))), 256)),
        (s(!0), n(x, null, b(I.value, (l) => (s(), n("g", null, [
          (s(!0), n(x, null, b(l.plots, (u, o) => (s(), n("circle", {
            cx: u.x,
            cy: u.y,
            fill: "transparent",
            r: r.value.plots,
            stroke: "none",
            onMouseenter: (k) => Xt({
              datapoint: u,
              plotIndex: o,
              seriesIndex: l.seriesIndex
            }),
            onMouseleave: (k) => Ft({
              datapoint: u,
              plotIndex: o,
              seriesIndex: l.seriesIndex
            }),
            onClick: (k) => Dt({
              datapoint: u,
              plotIndex: o,
              seriesIndex: l.seriesIndex
            })
          }, null, 40, ua))), 256))
        ]))), 256)),
        d(t.$slots, "svg", {
          svg: {
            ...g.value,
            isPrintingImg: i(pe) | i(ge) | q.value,
            isPrintingSvg: Z.value
          }
        }, void 0, !0)
      ], 12, Ol)),
      t.$slots.watermark ? (s(), n("div", ia, [
        d(t.$slots, "watermark", A(S({ isPrinting: i(pe) || i(ge) || q.value || Z.value })), void 0, !0)
      ])) : f("", !0),
      L("div", {
        id: `legend-bottom-${P.value}`
      }, null, 8, ca),
      Pe.value ? (s(), N(Qt, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${P.value}` : `#legend-bottom-${P.value}`
      }, [
        L("div", {
          ref_key: "chartLegend",
          ref: Se
        }, [
          e.value.style.chart.legend.show && K.value ? (s(), N($l, {
            key: `legend_${Te.value}`,
            legendSet: V.value,
            config: Mt.value,
            isCursorPointer: Y.value,
            onClickMarker: a[0] || (a[0] = ({ legend: l }) => {
              Q(l.seriesIndex), Ye(l);
            })
          }, {
            item: y(({ legend: l, index: u }) => [
              L("div", {
                onClick: (o) => {
                  l.segregate(), Ye(l);
                },
                style: $(`opacity:${p.value.includes(l.seriesIndex) ? 0.5 : 1}`)
              }, M(l.name), 13, va)
            ]),
            legendToggle: y(() => [
              V.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !i(O) ? (s(), N(Ll, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: p.value.length > 0,
                isCursorPointer: Y.value,
                onToggle: Ot
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : f("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : d(t.$slots, "legend", {
            key: 1,
            legend: V.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : f("", !0),
      t.$slots.source ? (s(), n("div", {
        key: 6,
        ref_key: "source",
        ref: ze,
        dir: "auto"
      }, [
        d(t.$slots, "source", {}, void 0, !0)
      ], 512)) : f("", !0),
      ke(i(rt), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: w.value.showTooltip && se.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: z.value,
        content: ne.value,
        isCustom: U.value,
        isFullscreen: W.value,
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": y(() => [
          d(t.$slots, "tooltip-before", A(S({ ...me.value })), void 0, !0)
        ]),
        "tooltip-after": y(() => [
          d(t.$slots, "tooltip-after", A(S({ ...me.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isCustom", "isFullscreen", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      K.value && e.value.userOptions.buttons.table ? (s(), N(el(G.value.component), tl({ key: 7 }, G.value.props, {
        ref_key: "tableUnit",
        ref: j,
        onClose: Ue
      }), qe({
        content: y(() => [
          (s(), N(i(ct), {
            key: `table_${$e.value}`,
            colNames: ee.value.colNames,
            head: ee.value.head,
            body: ee.value.body,
            config: ee.value.config,
            title: e.value.table.useDialog ? "" : G.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: Y.value,
            onClose: Ue
          }, {
            th: y(({ th: l }) => [
              L("div", {
                innerHTML: l,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, ha)
            ]),
            td: y(({ td: l }) => [
              Ze(M(l.name || l), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: y(() => [
            Ze(M(G.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: y(() => [
            L("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[1] || (a[1] = (l) => xe(e.value.userOptions.callbacks.csv)),
              style: $({ cursor: Y.value ? "pointer" : "default" })
            }, [
              ke(i(ut), {
                name: "fileCsv",
                stroke: G.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : f("", !0),
      d(t.$slots, "skeleton", {}, () => [
        i(O) ? (s(), N(gl, { key: 0 })) : f("", !0)
      ], !0)
    ], 46, zl));
  }
}, Oa = /* @__PURE__ */ _l(da, [["__scopeId", "data-v-e7c471f3"]]);
export {
  Oa as default
};
