// vue-data-ui/dist/vue-ui-dashboard-Bo5FX73U.js
import { useCssVars as Te, defineAsyncComponent as t, ref as r, computed as y, watch as W, openBlock as U, createElementBlock as A, createElementVNode as T, normalizeStyle as $, Fragment as x, renderList as $e, normalizeClass as Oe, createCommentVNode as N, renderSlot as d, createBlock as E, resolveDynamicComponent as Pe, mergeProps as Ie, withCtx as p, normalizeProps as V, guardReactiveProps as C, unref as O, createSlots as Me } from "vue";
import { u as Ae } from "./useNestedProp-BBT0mbHh.js";
import { u as Se } from "./useConfig-CIgyKZ6R.js";
import { j as Re } from "./lib-GwPnJSrY.js";
import { u as ze } from "./usePrinter-DiK5R6k-.js";
import De from "./UserOptions-CrLqEimB.js";
import { u as Xe } from "./useUserOptionState-BIvW1Kz7.js";
import He from "./PenAndPaper-DWV1FMIi.js";
import { _ as Ye } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const We = ["id"], Ne = ["onMousedown", "onTouchstart"], Ee = ["onMousedown", "onTouchstart"], Le = ["onMousedown", "onTouchstart"], _e = ["onMousedown", "onTouchstart"], Be = ["onMousedown", "onTouchstart"], X = 1, Ge = {
  __name: "vue-ui-dashboard",
  props: {
    dataset: Array,
    config: Object
  },
  emits: ["change", "copyAlt"],
  setup(ee, { expose: oe, emit: te }) {
    Te((e) => ({
      v41df80aa: me.value,
      v6d59dbf0: ge.value
    }));
    const ne = {
      VueDataUi: t(() => import("./vue-data-ui-C2kPofmi.js")),
      VueUi3dBar: t(() => import("./vue-ui-3d-bar-BhNgFaJA.js")),
      VueUiAccordion: t(() => import("./vue-ui-accordion-BGHTqIBt.js")),
      VueUiAgePyramid: t(() => import("./vue-ui-age-pyramid-CuYRf5v-.js")),
      VueUiAnnotator: t(() => import("./vue-ui-annotator-B1GWVtPp.js")),
      VueUiCandlestick: t(() => import("./vue-ui-candlestick-BY-_E7eZ.js")),
      VueUiChestnut: t(() => import("./vue-ui-chestnut-C-bQYaNI.js")),
      VueUiDigits: t(() => import("./vue-ui-digits-mpo2IcM5.js")),
      VueUiDonut: t(() => import("./vue-ui-donut-CzCgCVsx.js")),
      VueUiDonutEvolution: t(() => import("./vue-ui-donut-evolution-C_ck_TOR.js")),
      VueUiDumbbell: t(() => import("./vue-ui-dumbbell-CVSA90b2.js")),
      VueUiFlow: t(() => import("./vue-ui-flow-CsbSaFnu.js")),
      VueUiGalaxy: t(() => import("./vue-ui-galaxy-DWKHDciu.js")),
      VueUiGauge: t(() => import("./vue-ui-gauge-J2VYHsQe.js")),
      VueUiHeatmap: t(() => import("./vue-ui-heatmap-BEDirv8T.js")),
      VueUiKpi: t(() => import("./vue-ui-kpi-DlUEHKuw.js")),
      VueUiMiniLoader: t(() => import("./vue-ui-mini-loader-CoOHvZUk.js")),
      VueUiMolecule: t(() => import("./vue-ui-molecule-Chq1qWDK.js")),
      VueUiMoodRadar: t(() => import("./vue-ui-mood-radar-A2LBLgUi.js")),
      VueUiNestedDonuts: t(() => import("./vue-ui-nested-donuts-DFX251aT.js")),
      VueUiOnion: t(() => import("./vue-ui-onion-CW0YO9Ro.js")),
      VueUiParallelCoordinatePlot: t(() => import("./vue-ui-parallel-coordinate-plot-UFg10bHP.js")),
      VueUiQuadrant: t(() => import("./vue-ui-quadrant-C1pLW2zA.js")),
      VueUiQuickChart: t(() => import("./vue-ui-quick-chart-FaxttnJY.js")),
      VueUiRadar: t(() => import("./vue-ui-radar-DE70LPBZ.js")),
      VueUiRating: t(() => import("./vue-ui-rating-zV58MDQh.js")),
      VueUiRelationCircle: t(() => import("./vue-ui-relation-circle-iuRXIyRU.js")),
      VueUiRings: t(() => import("./vue-ui-rings-80ZUki42.js")),
      VueUiScatter: t(() => import("./vue-ui-scatter-FdAAYJOK.js")),
      VueUiSkeleton: t(() => import("./vue-ui-skeleton-CbqNz9IE.js")),
      VueUiSmiley: t(() => import("./vue-ui-smiley-BK5EKp9u.js")),
      VueUiSparkHistogram: t(() => import("./vue-ui-sparkhistogram-C9DvzQae.js")),
      VueUiSparkStackbar: t(() => import("./vue-ui-sparkstackbar-BsIOHUE5.js")),
      VueUiSparkTrend: t(() => import("./vue-ui-spark-trend-C68p8-Xp.js")),
      VueUiSparkbar: t(() => import("./vue-ui-sparkbar-DUDWvTo-.js")),
      VueUiSparkgauge: t(() => import("./vue-ui-sparkgauge-DdRtHQDe.js")),
      VueUiSparkline: t(() => import("./vue-ui-sparkline-B2RDrdiR.js")),
      VueUiStripPlot: t(() => import("./vue-ui-strip-plot-B9nBSRyY.js")),
      VueUiTable: t(() => import("./vue-ui-table-BHuXI4Eq.js")),
      VueUiTableHeatmap: t(() => import("./vue-ui-table-heatmap-y88sBfu3.js")),
      VueUiTableSparkline: t(() => import("./vue-ui-table-sparkline-ClpLLKi2.js")),
      VueUiThermometer: t(() => import("./vue-ui-thermometer-sjruUGQx.js")),
      VueUiTimer: t(() => import("./vue-ui-timer-CgdgMKIc.js")),
      VueUiTiremarks: t(() => import("./vue-ui-tiremarks-UBDpAIVr.js")),
      VueUiTreemap: t(() => import("./vue-ui-treemap-BS5EilZf.js")),
      VueUiVerticalBar: t(() => import("./vue-ui-horizontal-bar-ClQcB9nL.js")),
      VueUiHorizontalBar: t(() => import("./vue-ui-horizontal-bar-ClQcB9nL.js")),
      VueUiWaffle: t(() => import("./vue-ui-waffle-CbX5K9Q2.js")),
      VueUiWheel: t(() => import("./vue-ui-wheel-BIJ09H5D.js")),
      VueUiWordCloud: t(() => import("./vue-ui-word-cloud-DJmJ-oYT.js")),
      VueUiXy: t(() => import("./vue-ui-xy-f67V1xjd.js")),
      VueUiXyCanvas: t(() => import("./vue-ui-xy-canvas-4auyhlTr.js")),
      VueUiCarouselTable: t(() => import("./vue-ui-carousel-table-CmiHTJ8E.js")),
      VueUiGizmo: t(() => import("./vue-ui-gizmo-DEec3-v9.js")),
      VueUiStackbar: t(() => import("./vue-ui-stackbar-BZl0kEkb.js")),
      VueUiStackline: t(() => import("./vue-ui-stackline-2VKxwFei.js")),
      VueUiBullet: t(() => import("./vue-ui-bullet-DPoGGtaG.js")),
      VueUiFunnel: t(() => import("./vue-ui-funnel-BRkaEkBb.js")),
      VueUiHistoryPlot: t(() => import("./vue-ui-history-plot-Bs96PebH.js")),
      VueUiCirclePack: t(() => import("./vue-ui-circle-pack-CdbDcuDw.js")),
      VueUiWorld: t(() => import("./vue-ui-world-D-3wgoTn.js")),
      VueUiChord: t(() => import("./vue-ui-chord-uQYQgyGM.js")),
      VueUiRidgeline: t(() => import("./vue-ui-ridgeline-CwvDh5pG.js")),
      VueUiDag: t(() => import("./vue-ui-dag-DsHUvjvm.js"))
    }, { vue_ui_dashboard: ae } = Se(), ie = r(null), ue = r(null), H = r(null), S = ee, i = y(() => Ae({
      userConfig: S.config,
      defaultConfig: ae
    })), L = y(() => i.value.userOptions.useCursorPointer), Y = r(Re()), f = r(i.value.locked);
    function le() {
      f.value = !f.value;
    }
    W(() => S.config, () => {
      f.value = i.value.locked, K.value = !i.value.userOptions.showOnChartHover;
    });
    function _() {
      return S.dataset.map((e, o) => ({
        ...e,
        index: o
      }));
    }
    const s = r(_());
    W(() => S.dataset, () => {
      s.value = _();
    });
    const k = y(
      () => s.value.map((e) => ({
        ...e,
        resolvedComponent: typeof e.component == "string" ? ne[e.component] : e.component
      }))
    ), h = r(null), l = r(null), m = r({ x: 0, y: 0 }), g = r({ x: 0, y: 0 }), c = r(null), v = r(!1), b = r(null), P = r(!1);
    function B(e) {
      const o = e.target;
      (o.tagName === "INPUT" && o.type === "range" || o.classList.contains("range-handle")) && (P.value = !0);
    }
    function G(e) {
      const o = e.target;
      o.tagName === "INPUT" && o.type === "range" && (P.value = !1);
    }
    const { isPrinting: re, isImaging: se, generatePdf: F, generateImage: de } = ze({
      elementId: `vue-ui-dashboard_${Y.value}`,
      fileName: i.value.userOptions.print.filename || "dashboard",
      options: {
        ...i.value.userOptions.print,
        aspectRatio: i.value.style.board.aspectRatio
      }
    });
    function ce(e) {
      if (!f.value && (v.value = !0, b.value = e, l.value === null)) {
        h.value = e, m.value = { x: event.clientX, y: event.clientY };
        const o = s.value[e], n = 100 - o.width, a = 100 - o.height;
        o.left < 0 && (o.left = 0), o.top < 0 && (o.top = 0), o.left > n && (o.left = n), o.top > a && (o.top = a), o.left < 0 && (o.left = 0), o.top < 0 && (o.top = 0), o.left + o.width > 100 && (o.left = 100 - o.width), o.top + o.height > 100 && (o.top = 100 - o.height);
      }
    }
    function R(e, o) {
      v.value = !0, b.value = e, l.value = { index: e, direction: o };
      const n = s.value[e];
      g.value = { x: event.clientX, y: event.clientY, initialWidth: n.width, initialHeight: n.height };
    }
    function j(e, o, n) {
      if (l.value.direction.includes("top")) {
        const a = e.height - n / c.value.offsetHeight * 100;
        a >= X && (e.top += n / c.value.offsetHeight * 100, e.height = a);
      }
      if (l.value.direction.includes("bottom")) {
        const a = e.height + n / c.value.offsetHeight * 100;
        a >= X && (e.height = a);
      }
      if (l.value.direction.includes("left")) {
        const a = e.width - o / c.value.offsetWidth * 100;
        a >= X && (e.left += o / c.value.offsetWidth * 100, e.width = a);
      }
      if (l.value.direction.includes("right")) {
        const a = e.width + o / c.value.offsetWidth * 100;
        a >= X && (e.width = a);
      }
    }
    function pe(e) {
      if (!(f.value || P.value)) {
        if (v.value = !0, h.value !== null) {
          const o = s.value[h.value], n = e.clientX - m.value.x, a = e.clientY - m.value.y, u = o.left + n / c.value.offsetWidth * 100, w = o.top + a / c.value.offsetHeight * 100;
          u >= 0 && w >= 0 && u + o.width <= 100 && w + o.height <= 100 && (o.left = u, o.top = w), m.value = { x: e.clientX, y: e.clientY };
        }
        if (l.value !== null) {
          const o = s.value[l.value.index], n = e.clientX - g.value.x, a = e.clientY - g.value.y;
          j(o, n, a), g.value = { x: e.clientX, y: e.clientY };
        }
      }
    }
    const Q = te;
    W(
      s,
      (e) => {
        e && !v.value && Q("change", s.value);
      },
      { deep: !0 }
    );
    function ve() {
      h.value = null, l.value = null, v.value = !1, b.value = null, s.value.forEach((e) => {
        e.left = Math.round(e.left / 100 * 100), e.top = Math.round(e.top / 100 * 100), e.width = Math.round(e.width / 100 * 100), e.height = Math.round(e.height / 100 * 100);
      });
    }
    function fe(e) {
      f.value || P.value || (v.value = !0, b.value = e, l.value === null && (h.value = e, m.value = { x: event.touches[0].clientX, y: event.touches[0].clientY }));
    }
    function z(e, o, n) {
      if (v.value = !0, b.value = e, l.value === null) {
        l.value = { index: e, direction: o };
        const a = s.value[e];
        g.value = { x: n.touches[0].clientX, y: n.touches[0].clientY, initialWidth: a.width, initialHeight: a.height };
      }
    }
    function D(e) {
      if (v.value = !0, e.preventDefault(), l.value !== null) {
        const o = s.value[l.value.index], n = e.touches[0].clientX - g.value.x, a = e.touches[0].clientY - g.value.y;
        j(o, n, a), g.value = { x: e.touches[0].clientX, y: e.touches[0].clientY };
      }
    }
    function he(e) {
      if (!(f.value || P.value) && (v.value = !0, e.preventDefault(), h.value !== null)) {
        const o = s.value[h.value], n = e.touches[0].clientX - m.value.x, a = e.touches[0].clientY - m.value.y, u = o.left + n / c.value.offsetWidth * 100, w = o.top + a / c.value.offsetHeight * 100;
        u >= 0 && w >= 0 && u + o.width <= 100 && w + o.height <= 100 && (o.left = u, o.top = w), m.value = { x: e.touches[0].clientX, y: e.touches[0].clientY };
      }
    }
    function I() {
      v.value = !1, b.value = null, h.value = null, l.value = null, s.value.forEach((e) => {
        e.left = Math.round(e.left / 100 * 100), e.top = Math.round(e.top / 100 * 100), e.width = Math.round(e.width / 100 * 100), e.height = Math.round(e.height / 100 * 100);
      });
    }
    const me = y(() => i.value.style.item.borderColor), ge = y(() => i.value.style.resizeHandles.backgroundColor), be = y(() => i.value.style.board.aspectRatio), ye = y(() => i.value.style.board.backgroundColor), Ue = y(() => i.value.style.board.border);
    function Ve() {
      return s.value;
    }
    const { userOptionsVisible: K, setUserOptionsVisibility: q, keepUserOptionState: Ce } = Xe({ config: i.value }), M = r(!1);
    function J() {
      M.value = !M.value;
    }
    function ke() {
      q(!0);
    }
    function we() {
      q(!1);
    }
    async function Z() {
      if (Q("copyAlt", {
        config: i.value,
        dataset: k.value
      }), !i.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(i.value.userOptions.callbacks.altCopy({
        config: i.value,
        dataset: k.value
      }));
    }
    return oe({
      generatePdf: F,
      getItemsPositions: Ve,
      toggleLock: le,
      copyAlt: Z
    }), (e, o) => (U(), A("div", {
      id: `vue-ui-dashboard_${Y.value}`,
      onMousedown: B,
      onMouseup: G,
      onTouchstart: B,
      onTouchend: G,
      onMouseenter: ke,
      onMouseleave: we,
      ref_key: "dashboardRef",
      ref: ie,
      style: {
        position: "relative"
      },
      class: "vue-data-ui-component"
    }, [
      T("div", {
        class: "vue-ui-dashboard-container",
        ref_key: "dashboardContainer",
        ref: c,
        style: $(`outline:${Ue.value}; background:${ye.value}; aspect-ratio:${be.value};${M.value ? "pointer-events:none" : ""}`)
      }, [
        T("div", {
          class: "vue-ui-dashboard-grid-container",
          ref: "container",
          onMousemove: pe,
          onMouseup: ve,
          onTouchmove: he,
          onTouchend: I,
          style: $(`background:${i.value.style.board.backgroundColor}`)
        }, [
          o[4] || (o[4] = T("div", { class: "vue-ui-dashboard-grid" }, null, -1)),
          (U(!0), A(x, null, $e(k.value, (n, a) => (U(), A("div", {
            key: n.id,
            class: Oe({ "vue-ui-dashboard-grid-item": !0, "vue-ui-dashboard-grid-item--locked": f.value }),
            style: $({
              width: `${n.width}%`,
              height: `${n.height}%`,
              left: `${n.left}%`,
              top: `${n.top}%`,
              cursor: "move",
              boxShadow: b.value === a ? "0 6px 12px -3px rgba(0,0,0,0.3)" : "",
              zIndex: b.value === a ? k.value.length + 1 : n.index,
              backgroundColor: i.value.style.item.backgroundColor
            }),
            onMousedown: (u) => ce(a),
            onTouchstart: (u) => fe(a)
          }, [
            f.value ? N("", !0) : (U(), A(x, { key: 0 }, [
              T("div", {
                class: "vue-ui-dashboard-resize-handle vue-ui-dashboard-top-left",
                onMousedown: (u) => R(a, "top-left"),
                onTouchstart: (u) => z(a, "top-left", u),
                onTouchmove: o[0] || (o[0] = (u) => D(u)),
                onTouchend: I
              }, null, 40, Ee),
              T("div", {
                class: "vue-ui-dashboard-resize-handle vue-ui-dashboard-top-right",
                onMousedown: (u) => R(a, "top-right"),
                onTouchstart: (u) => z(a, "top-right", u),
                onTouchmove: o[1] || (o[1] = (u) => D(u)),
                onTouchend: I
              }, null, 40, Le),
              T("div", {
                class: "vue-ui-dashboard-resize-handle vue-ui-dashboard-bottom-left",
                onMousedown: (u) => R(a, "bottom-left"),
                onTouchstart: (u) => z(a, "bottom-left", u),
                onTouchmove: o[2] || (o[2] = (u) => D(u)),
                onTouchend: I
              }, null, 40, _e),
              T("div", {
                class: "vue-ui-dashboard-resize-handle vue-ui-dashboard-bottom-right",
                onMousedown: (u) => R(a, "bottom-right"),
                onTouchstart: (u) => z(a, "bottom-right", u),
                onTouchmove: o[3] || (o[3] = (u) => D(u)),
                onTouchend: I
              }, null, 40, Be)
            ], 64)),
            d(e.$slots, "top", {
              item: n,
              index: a
            }, void 0, !0),
            n.resolvedComponent ? (U(), E(Pe(n.resolvedComponent), Ie({
              key: 1,
              ref_for: !0
            }, n.props), null, 16)) : d(e.$slots, "content", {
              key: 2,
              item: n,
              index: a,
              left: n.left,
              top: n.top,
              height: n.height,
              width: n.width
            }, void 0, !0),
            d(e.$slots, "bottom", {
              item: n,
              index: a
            }, void 0, !0)
          ], 46, Ne))), 128))
        ], 36)
      ], 4),
      (U(), A("svg", {
        style: $({
          width: "100%",
          height: "100%",
          pointerEvents: "none",
          position: "absolute",
          top: 0,
          left: 0,
          zIndex: k.value.length + 1
        }),
        ref_key: "svgRef",
        ref: H
      }, null, 4)),
      i.value.userOptions.buttons.annotator && c.value && H.value ? (U(), E(He, {
        key: 0,
        color: i.value.style.board.color,
        backgroundColor: i.value.style.board.backgroundColor,
        active: M.value,
        svgRef: H.value,
        isCursorPointer: L.value,
        onClose: J,
        style: $({
          zIndex: k.value.length + 1
        })
      }, {
        "annotator-action-close": p(() => [
          d(e.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": p(({ color: n }) => [
          d(e.$slots, "annotator-action-color", V(C({ color: n })), void 0, !0)
        ]),
        "annotator-action-draw": p(({ mode: n }) => [
          d(e.$slots, "annotator-action-draw", V(C({ mode: n })), void 0, !0)
        ]),
        "annotator-action-undo": p(({ disabled: n }) => [
          d(e.$slots, "annotator-action-undo", V(C({ disabled: n })), void 0, !0)
        ]),
        "annotator-action-redo": p(({ disabled: n }) => [
          d(e.$slots, "annotator-action-redo", V(C({ disabled: n })), void 0, !0)
        ]),
        "annotator-action-delete": p(({ disabled: n }) => [
          d(e.$slots, "annotator-action-delete", V(C({ disabled: n })), void 0, !0)
        ]),
        _: 3
      }, 8, ["color", "backgroundColor", "active", "svgRef", "isCursorPointer", "style"])) : N("", !0),
      (i.value.allowPrint || i.value.userOptions.show) && (i.value.userOptions.buttons.pdf || i.value.userOptions.buttons.img) ? (U(), E(De, {
        key: 1,
        ref_key: "userOptionsRef",
        ref: ue,
        backgroundColor: i.value.style.board.backgroundColor,
        color: i.value.style.board.color,
        isPrinting: O(re),
        isImaging: O(se),
        uid: Y.value,
        position: i.value.userOptions.position,
        hasTooltip: !1,
        hasPdf: i.value.userOptions.buttons.pdf,
        hasImg: i.value.userOptions.buttons.img,
        hasXls: !1,
        hasTable: !1,
        hasLabel: !1,
        hasFullscreen: !1,
        hasAltCopy: i.value.userOptions.buttons.altCopy,
        chartElement: c.value,
        callbacks: i.value.userOptions.callbacks,
        hasAnnotator: i.value.userOptions.buttons.annotator,
        isAnnotation: M.value,
        printScale: i.value.userOptions.print.scale,
        titles: { ...i.value.userOptions.buttonTitles },
        isCursorPointer: L.value,
        onGeneratePdf: O(F),
        onGenerateImage: O(de),
        onToggleAnnotator: J,
        onCopyAlt: Z,
        style: $({ visibility: O(Ce) ? O(K) ? "visible" : "hidden" : "visible", zIndex: k.value.length + 1 })
      }, Me({ _: 2 }, [
        e.$slots.menuIcon ? {
          name: "menuIcon",
          fn: p(({ isOpen: n, color: a }) => [
            d(e.$slots, "menuIcon", V(C({ isOpen: n, color: a })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        e.$slots.optionPdf ? {
          name: "optionPdf",
          fn: p(() => [
            d(e.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        e.$slots.optionImg ? {
          name: "optionImg",
          fn: p(() => [
            d(e.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        e.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: p(({ toggleAnnotator: n, isAnnotator: a }) => [
            d(e.$slots, "optionAnnotator", V(C({ toggleAnnotator: n, isAnnotator: a })), void 0, !0)
          ]),
          key: "3"
        } : void 0,
        e.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: p(({ altCopy: n }) => [
            d(e.$slots, "optionAltCopy", V(C({ altCopy: n })), void 0, !0)
          ]),
          key: "4"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "position", "hasPdf", "hasImg", "hasAltCopy", "chartElement", "callbacks", "hasAnnotator", "isAnnotation", "printScale", "titles", "isCursorPointer", "onGeneratePdf", "onGenerateImage", "style"])) : N("", !0)
    ], 40, We));
  }
}, oo = /* @__PURE__ */ Ye(Ge, [["__scopeId", "data-v-dd6926d2"]]);
export {
  oo as default
};
