// vue-data-ui/dist/vue-ui-chestnut-C-bQYaNI.js
import { useCssVars as zt, defineAsyncComponent as J, computed as g, ref as $, toRefs as St, watch as ke, onMounted as Nt, nextTick as ue, openBlock as a, createElementBlock as o, unref as n, normalizeStyle as y, normalizeClass as T, createBlock as re, withCtx as C, renderSlot as x, normalizeProps as V, guardReactiveProps as B, createCommentVNode as h, createSlots as Ee, createVNode as we, toDisplayString as c, createElementVNode as u, Fragment as _, renderList as m, createTextVNode as ee, resolveDynamicComponent as Ft, mergeProps as It, withKeys as Pt } from "vue";
import { j as Lt, t as Ge, m as Ot, p as M, n as Rt, r as Ce, q as ie, s as $e, o as Vt, k as E, l as F, X as Bt, u as I, v as Ue, W as We, a3 as Ye, a4 as Mt, a5 as te, a6 as j, a0 as jt, a1 as Dt } from "./lib-GwPnJSrY.js";
import { u as Xt } from "./useConfig-CIgyKZ6R.js";
import { u as Et, B as Gt } from "./BaseScanner-BhhkKVoj.js";
import { u as Ut } from "./usePrinter-DiK5R6k-.js";
import { u as Wt } from "./useSvgExport-ByUukOZt.js";
import { u as Te } from "./useNestedProp-BBT0mbHh.js";
import { u as Yt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as qt } from "./useTableResponsive-Cbx-ACcP.js";
import { u as Ht } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Kt } from "./useChartAccessibility-9icAAmYg.js";
import Zt from "./img-BYYzY_QU.js";
import Jt from "./vue_ui_chestnut-BloFvDuV.js";
import { _ as Qt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const el = ["id"], tl = {
  key: 1,
  ref: "noTitle",
  class: "vue-data-ui-no-title-space",
  style: "height:36px; width: 100%;background:transparent"
}, ll = ["xmlns", "viewBox"], al = ["width", "height"], ol = { key: 1 }, sl = ["fill", "font-weight", "font-size", "x", "y"], nl = ["fill", "font-weight", "font-size", "x", "y"], ul = ["id"], rl = ["stop-color"], il = ["stop-color"], cl = ["id"], dl = ["stop-color"], vl = ["stop-color"], yl = ["id"], hl = ["stop-color"], fl = ["stop-color"], bl = ["stop-color"], pl = ["id"], gl = ["stop-color"], _l = ["stop-color"], ml = ["stop-color"], xl = ["id"], kl = ["stop-color"], wl = ["stop-color"], Cl = ["stop-color"], $l = { key: 2 }, Tl = ["x", "y", "font-size", "font-weight", "fill"], Al = ["x", "y", "font-size", "font-weight", "fill"], zl = ["id"], Sl = ["stop-color"], Nl = ["stop-color"], Fl = ["d", "stroke", "fill"], Il = ["cx", "cy", "r", "fill"], Pl = ["cx", "cy", "r", "fill", "stroke", "stroke-width", "onClick"], Ll = { key: 3 }, Ol = ["x", "y", "font-size", "fill", "onClick"], Rl = { key: 0 }, Vl = ["x", "y", "fill", "font-size", "font-weight"], Bl = ["x", "y", "height", "width", "fill", "rx", "onClick"], Ml = ["x", "y", "height", "width", "fill", "rx", "stroke", "stroke-width", "onClick"], jl = { key: 4 }, Dl = ["x", "y", "fill", "font-size", "onClick"], Xl = ["d", "stroke"], El = ["fill", "cx", "cy", "r", "onClick"], Gl = { key: 5 }, Ul = ["x", "y", "font-size", "font-weight", "fill"], Wl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Yl = ["y", "height", "width"], ql = { style: { width: "100%", height: "100%", display: "flex", "align-items": "center", "justify-content": "center", "flex-direction": "column" } }, Hl = { style: { display: "flex", "align-items": "center", "justify-content": "center", gap: "12px", "flex-wrap": "wrap", "flex-direction": "row" } }, Kl = {
  viewBox: "0 0 20 20",
  height: "16",
  width: "16"
}, Zl = ["fill"], Jl = { key: 7 }, Ql = ["y", "height", "width"], ea = { style: { width: "100%", height: "100%", display: "flex", "align-items": "center", "justify-content": "center", "flex-direction": "column" } }, ta = { style: { display: "flex", "align-items": "center", "justify-content": "center", gap: "12px", "flex-wrap": "wrap", "flex-direction": "row" } }, la = {
  viewBox: "0 0 20 20",
  height: "16",
  width: "16"
}, aa = ["fill"], oa = ["cx", "cy", "fill"], sa = ["d", "stroke"], na = ["cx", "cy", "fill"], ua = ["d", "stroke"], ra = ["cx", "cy", "fill"], ia = ["cx", "cy", "fill"], ca = ["x", "y", "fill", "font-size", "font-weight"], da = ["x", "y", "fill", "font-size", "font-weight"], va = ["x", "text-anchor", "y", "fill", "font-size"], ya = ["x", "text-anchor", "y", "fill", "font-size"], ha = ["x", "text-anchor", "y", "fill", "font-size"], fa = ["x", "text-anchor", "y", "fill", "font-size"], ba = ["x", "text-anchor", "y", "fill", "font-size"], pa = ["x", "y", "font-size", "fill"], ga = ["x", "y", "font-size", "fill"], _a = ["x", "y", "font-size", "fill"], ma = {
  key: 4,
  class: "vue-data-ui-watermark"
}, xa = {
  key: 5,
  ref: "source",
  dir: "auto"
}, ka = { class: "vue-ui-data-table" }, wa = { key: 0 }, Ca = ["data-cell"], $a = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Ta = { key: 0 }, Aa = { key: 1 }, za = ["data-cell"], Sa = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Na = { key: 0 }, Fa = { key: 1 }, Ia = ["data-cell"], Pa = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, La = { key: 0 }, Oa = { key: 1 }, Ra = ["data-cell"], Va = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Ba = { key: 0 }, Ma = { key: 1 }, ja = ["data-cell"], Da = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Xa = { key: 0 }, Ea = { key: 1 }, Ga = ["data-cell"], Ua = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Wa = { key: 0 }, Ya = { key: 1 }, qa = ["data-cell"], Ha = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Ka = { key: 0 }, Za = { key: 1 }, Ja = ["data-cell"], Qa = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, eo = ["data-cell"], to = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, lo = ["data-cell"], ao = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, oo = ["data-cell"], so = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, no = ["data-cell"], uo = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, ro = {
  __name: "vue-ui-chestnut",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectRoot", "selectBranch", "selectNut", "copyAlt"],
  setup(qe, { expose: He, emit: Ke }) {
    zt((l) => ({
      v02951477: l.tdo
    }));
    const Ae = J(() => import("./BaseIcon-CNx8Tzl5.js")), Ze = J(() => import("./vue-ui-accordion-BGHTqIBt.js")), Je = J(() => import("./PenAndPaper-DWV1FMIi.js")), Qe = J(() => import("./UserOptions-CrLqEimB.js")), et = J(() => import("./PackageVersion-DN9d0ME5.js")), tt = J(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_chestnut: lt } = Xt(), { isThemeValid: at, warnInvalidTheme: ot } = Yt(), D = qe, ze = g(() => !!D.dataset && D.dataset.length), A = $(Lt()), H = $(null), Se = $(0), Q = $(null), ce = $(null), le = $(!1), ae = $(!1), e = $(he()), P = g(() => e.value.userOptions.useCursorPointer), st = g(() => Ge({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              grandTotal: { show: !1 },
              roots: {
                stroke: "#6A6A6A",
                labels: { show: !1 }
              },
              verticalSeparator: { stroke: "transparent" },
              branches: {
                stroke: "#6A6A6A",
                underlayerColor: "#6A6A6A90",
                labels: {
                  show: !1,
                  dataLabels: { show: !1 }
                }
              }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: Ne, FINAL_DATASET: de } = Et({
      ...St(D),
      FINAL_CONFIG: e,
      prepareConfig: he,
      skeletonDataset: D.config?.skeletonDataset ?? [
        {
          name: "_",
          color: "#969696",
          branches: [
            {
              name: "_",
              value: 32,
              breakdown: [
                { name: "_", value: 16, color: "#CACACA" },
                { name: "_", value: 16, color: "#6A6A6A" }
              ]
            },
            {
              name: "_",
              value: 16,
              breakdown: [
                { name: "_", value: 8, color: "#CACACA" },
                { name: "_", value: 8, color: "#6A6A6A" }
              ]
            },
            {
              name: "_",
              value: 8,
              breakdown: [
                { name: "_", value: 4, color: "#CACACA" },
                { name: "_", value: 4, color: "#6A6A6A" }
              ]
            },
            {
              name: "_",
              value: 4,
              breakdown: [
                { name: "_", value: 2, color: "#CACACA" },
                { name: "_", value: 2, color: "#6A6A6A" }
              ]
            }
          ]
        },
        {
          name: "_",
          color: "#C4C4C4",
          branches: [
            {
              name: "_",
              value: 24,
              breakdown: [
                { name: "_", value: 12, color: "#CACACA" },
                { name: "_", value: 12, color: "#6A6A6A" }
              ]
            },
            {
              name: "_",
              value: 12,
              breakdown: [
                { name: "_", value: 6, color: "#CACACA" },
                { name: "_", value: 6, color: "#6A6A6A" }
              ]
            },
            {
              name: "_",
              value: 6,
              breakdown: [
                { name: "_", value: 3, color: "#CACACA" },
                { name: "_", value: 3, color: "#6A6A6A" }
              ]
            },
            {
              name: "_",
              value: 2,
              breakdown: [
                { name: "_", value: 1, color: "#CACACA" },
                { name: "_", value: 1, color: "#6A6A6A" }
              ]
            }
          ]
        }
      ],
      skeletonConfig: Ge({
        defaultConfig: e.value,
        userConfig: st.value
      })
    }), { userOptionsVisible: ve, setUserOptionsVisibility: Fe, keepUserOptionState: Ie } = Ht({ config: e.value }), { svgRef: ye } = Kt({ config: e.value.style.chart.layout.title });
    function he() {
      const l = Te({
        userConfig: D.config,
        defaultConfig: lt
      }), r = l.theme;
      if (!r) return l;
      if (!at.value(l))
        return ot(l), l;
      const t = Te({
        userConfig: Jt[r] || D.config,
        defaultConfig: l
      }), s = Te({
        userConfig: D.config,
        defaultConfig: t
      });
      return {
        ...s,
        customPalette: s.customPalette.length ? s.customPalette : Ot[r] || M
      };
    }
    ke(() => D.config, (l) => {
      e.value = he(), ve.value = !e.value.userOptions.showOnChartHover, G.value.showTable = e.value.table.show, Be();
    }, { deep: !0 });
    const { isPrinting: fe, isImaging: be, generatePdf: Pe, generateImage: Le } = Ut({
      elementId: `vue-ui-chestnut_${A.value}`,
      fileName: e.value.style.chart.layout.title.text || "vue-ui-chestnut",
      options: e.value.userOptions.print
    }), nt = g(() => e.value.userOptions.show), pe = g(() => Rt(e.value.customPalette)), G = $({
      showTable: e.value.table.show
    });
    ke(e, () => {
      G.value = {
        showTable: e.value.table.show
      };
    }, { immediate: !0 });
    const Oe = $(null), ut = g(() => e.value.table.responsiveBreakpoint), i = $({
      gap: 6,
      padding: {
        top: 102,
        left: 12,
        right: 12,
        bottom: 96
      },
      width: 1024,
      height: 0,
      branchSize: 32,
      branchStart: 128
    }), S = g(() => ({
      left: i.value.padding.left,
      top: i.value.padding.top,
      right: i.value.width - i.value.padding.right,
      bottom: i.value.height - i.value.padding.bottom,
      width: i.value.width - (i.value.padding.left + i.value.padding.right),
      height: i.value.height - (i.value.padding.top + i.value.padding.bottom),
      seedX: i.value.padding.left + 64
    })), L = g(() => de.value.flatMap((l) => (l.branches || []).map((r) => r.value || 0)).reduce((l, r) => l + r, 0)), N = g(() => (de.value.forEach((l, r) => {
      Ce({
        datasetObject: l,
        requiredAttributes: ["name", "branches"]
      }).forEach((t) => {
        ie({
          componentName: "VueUiChestnut",
          type: "datasetSerieAttribute",
          property: t,
          index: r,
          debug: oe.value
        });
      }), l.branches && l.branches.forEach((t, s) => {
        Ce({
          datasetObject: t,
          requiredAttributes: ["name", "value"]
        }).forEach((b) => {
          ie({
            componentName: "VueUiChestnut",
            type: "datasetSerieAttribute",
            property: b,
            index: `${r} - ${s}`,
            debug: oe.value
          });
        }), t.breakdown && t.breakdown.forEach((b, p) => {
          Ce({
            datasetObject: b,
            requiredAttributes: ["name", "value"]
          }).forEach((f) => {
            ie({
              componentName: "VueUiChestnut",
              type: "datasetSerieAttribute",
              property: f,
              index: `${r} - ${s} - ${p}`,
              debug: oe.value
            });
          });
        });
      });
    }), de.value.map((l, r) => {
      const t = (l.branches || []).map((s) => s.value || 0).reduce((s, b) => s + b, 0);
      return {
        ...l,
        color: $e(l.color) || pe.value[r] || M[r] || M[r % M.length],
        id: l.id || `root_${r}_${A.value}`,
        type: "root",
        total: t,
        rootIndex: r,
        branches: (l.branches || []).map((s, b) => ({
          ...s,
          rootName: l.name,
          rootIndex: r,
          color: $e(l.color) || pe.value[r] || M[r] || M[r % M.length],
          value: s.value >= 0 ? s.value : 0,
          id: s.id || `branch_${r}_${b}_${A.value}`,
          proportionToRoot: s.value / t,
          type: "branch",
          breakdown: (s.breakdown || []).map((p, f) => ({
            table: {
              rootName: l.name,
              rootValue: t,
              rootToTotal: t / L.value,
              branchName: s.name,
              branchValue: s.value,
              branchToTotal: s.value / L.value,
              branchToRoot: s.value / t,
              nutName: p.name,
              nutValue: p.value,
              nutToTotal: p.value / L.value,
              nutToRoot: p.value / t,
              nutToBranch: p.value / s.value
            },
            ...p,
            type: "nut",
            branchName: s.name,
            rootName: l.name,
            branchTotal: s.value >= 0 ? s.value : 0,
            proportionToBranch: p.value / s.value,
            proportionToRoot: p.value / t,
            proportionToTree: p.value / L.value,
            rootIndex: r,
            id: p.id || `nut_${r}_${b}_${f}_${A.value}`,
            color: $e(p.color) || pe.value[f] || M[f] || M[f % M.length],
            value: p.value >= 0 ? p.value : 0
          }))
        }))
      };
    })));
    function rt() {
      return N.value;
    }
    const U = Ke, it = g(() => N.value.flatMap((l) => l.branches).length), ct = g(() => Math.max(...N.value.map((l) => l.branches.map((r) => r.value).reduce((r, t) => r + t, 0)))), dt = g(() => Math.max(...N.value.flatMap((l) => l.branches.map((r) => r.value)))), Re = g(() => 256 + i.value.padding.left), W = g(() => N.value.sort((l, r) => r.total - l.total).map((l, r) => {
      const t = S.value.height / N.value.length / 2, s = l.total / ct.value * (t > 64 ? 64 : t);
      return {
        ...l,
        x: S.value.seedX,
        y: S.value.top + S.value.height / N.value.length * (r + 1) - (S.value.height / N.value.length / 2 + i.value.gap / 2),
        r: s < i.value.branchSize / 2 ? i.value.branchSize / 2 : s
      };
    })), vt = g(() => e.value.style.chart.layout.branches.widthRatio <= 0 ? 0.1 : e.value.style.chart.layout.branches.widthRatio > 1.8 ? 1.8 : e.value.style.chart.layout.branches.widthRatio), yt = g(() => W.value.flatMap((l) => l.branches)), Y = g(() => yt.value.sort((l, r) => r.value - l.value).map((l, r) => ({
      ...l,
      y1: r * i.value.branchSize + S.value.top + r * i.value.gap,
      y2: r * i.value.branchSize + i.value.branchSize,
      x1: Re.value,
      x2: 384 * l.value / dt.value * vt.value + Re.value
    })));
    function ht(l) {
      const r = W.value.find((t) => t.rootIndex === l.rootIndex);
      return { x: r.x, y: r.y, r: r.r };
    }
    const d = $(null), X = $(null), k = $(null), O = $(null);
    function z() {
      d.value = null, k.value = null, O.value = null;
    }
    function R(l) {
      return O.value ? O.value === null ? !0 : l.type === "root" ? l.id === O.value.id : l.rootIndex === O.value.rootIndex : d.value === null && k.value === null ? !0 : l.type === "nut" ? d.value.id === l.id : l.type === "branch" ? k.value.id === l.id : l.type === "root" ? k.value && k.value.rootIndex === l.rootIndex || d.value && d.value.rootIndex === l.rootIndex : !1;
    }
    function ft(l) {
      z(), ue(() => {
        d.value = l, k.value = l, X.value = Ye(
          { series: l.breakdown },
          l.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
          l.y1 + i.value.branchSize / 2,
          80,
          80
        ), U("selectNut", l.breakdown);
      });
    }
    function q() {
      d.value = null, X.value = null, U("selectNut", null);
    }
    function ge(l) {
      k.value && k.value.id === l.id ? (k.value = null, z(), U("selectBranch", null)) : (z(), k.value = l, U("selectBranch", l));
    }
    function Ve(l) {
      O.value && O.value.id === l.id ? (z(), U("selectRoot", null)) : (z(), O.value = l, U("selectRoot", l));
    }
    function bt() {
      return S.value.bottom - (d.value.y1 + 180) < 0 ? 0 : S.value.bottom;
    }
    function K(l) {
      return l.proportion * 100 > e.value.style.chart.layout.nuts.selected.labels.dataLabels.hideUnderValue;
    }
    Nt(() => {
      Be();
    });
    const oe = g(() => e.value.debug);
    function Be() {
      Vt(D.dataset) && ie({
        componentName: "VueUiChestnut",
        type: "dataset",
        debug: oe.value
      });
      const l = it.value * (i.value.branchSize + i.value.gap) + i.value.padding.top + i.value.padding.bottom;
      i.value.height = l;
    }
    const v = g(() => {
      const l = [
        e.value.table.th.translations.rootName,
        e.value.table.th.translations.rootValue,
        e.value.table.th.translations.rootToTotal,
        e.value.table.th.translations.branchName,
        e.value.table.th.translations.branchValue,
        e.value.table.th.translations.branchToRoot,
        e.value.table.th.translations.branchToTotal,
        e.value.table.th.translations.nutName,
        e.value.table.th.translations.nutValue,
        e.value.table.th.translations.nutToBranch,
        e.value.table.th.translations.nutToRoot,
        e.value.table.th.translations.nutToTotal
      ], r = N.value.flatMap((t, s) => t.branches.flatMap((b, p) => b.breakdown.flatMap((f, w) => f.table)));
      return { head: l, body: r };
    });
    function _e(l = null) {
      ue(() => {
        const r = [[e.value.style.chart.layout.title.text], [e.value.style.chart.layout.title.subtitle.text], [""], ["Grand total", L.value], [""]], t = v.value.head, s = v.value.body.map((f, w) => [
          v.value.body[w - 1] && v.value.body[w - 1].rootName === f.rootName ? "" : f.rootName,
          v.value.body[w - 1] && v.value.body[w - 1].rootName === f.rootName ? "" : f.rootValue,
          v.value.body[w - 1] && v.value.body[w - 1].rootName === f.rootName ? "" : f.rootToTotal,
          v.value.body[w - 1] && v.value.body[w - 1].branchName === f.branchName ? "" : f.branchName,
          v.value.body[w - 1] && v.value.body[w - 1].branchName === f.branchName ? "" : f.branchValue,
          v.value.body[w - 1] && v.value.body[w - 1].branchName === f.branchName ? "" : f.branchToRoot,
          v.value.body[w - 1] && v.value.body[w - 1].branchName === f.branchName ? "" : f.branchToTotal,
          f.nutName,
          f.nutValue,
          f.nutToBranch,
          f.nutToRoot,
          f.nutToTotal
        ]), b = r.concat([t]).concat(s), p = jt(b);
        l ? l(p) : Dt({ csvContent: p, title: e.value.style.chart.layout.title.text || "vue-ui-chestnut" });
      });
    }
    const Z = $(!1);
    function Me(l) {
      Z.value = l, Se.value += 1;
    }
    function je() {
      G.value.showTable = !G.value.showTable;
    }
    const se = $(!1);
    function me() {
      se.value = !se.value;
    }
    async function pt({ scale: l = 2 } = {}) {
      if (!H.value) return;
      const { width: r, height: t } = H.value.getBoundingClientRect(), s = r / t, { imageUri: b, base64: p } = await Zt({ domElement: H.value, base64: !0, img: !0, scale: l });
      return {
        imageUri: b,
        base64: p,
        title: e.value.style.chart.layout.title.text,
        width: r,
        height: t,
        aspectRatio: s
      };
    }
    function gt(l) {
      const r = ht(l), t = l.x1, s = l.y1, b = i.value.branchSize, p = r.x + r.r / 2, f = r.y, w = 20;
      return [
        `M ${t},${s}`,
        `C ${t - w},${s} ${t - w},${s} ${p},${f}`,
        `C ${p},${f} ${t - w},${s + b} ${t},${s + b}`,
        "Z"
      ].join(" ");
    }
    const ne = g(() => {
      const l = e.value.table.useDialog && !e.value.table.show, r = G.value.showTable;
      return {
        component: l ? tt : Ze,
        title: `${e.value.style.chart.layout.title.text}${e.value.style.chart.layout.title.subtitle.text ? `: ${e.value.style.chart.layout.title.subtitle.text}` : ""}`,
        props: l ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: Z.value,
          fullscreenParent: H.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: P.value
        } : {
          hideDetails: !0,
          config: {
            open: r,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ke(() => G.value.showTable, async (l) => {
      e.value.table.show || (l && e.value.table.useDialog && Q.value ? (await ue(), Q.value.open()) : "close" in Q.value && Q.value.close());
    });
    const { isResponsive: _t } = qt(Oe, ut);
    function xe() {
      G.value.showTable = !1, ce.value && ce.value.setTableIconState(!1);
    }
    const mt = g(() => N.value.map((l, r) => ({
      ...l,
      display: `${l.name}: ${E(
        e.value.style.chart.layout.roots.labels.formatter,
        l.total,
        F({
          p: e.value.style.chart.layout.legend.prefix,
          v: l.total,
          s: e.value.style.chart.layout.legend.suffix,
          r: e.value.style.chart.layout.legend.roundingValue
        }),
        { datapoint: l }
      )} (${F({
        v: l.total / L.value * 100,
        s: "%",
        r: e.value.style.chart.layout.legend.roundingPercentage
      })})`
    }))), xt = g(() => mt.value.map((l) => ({
      color: l.color,
      name: l.display,
      shape: "circle"
    }))), kt = g(() => e.value.style.chart.backgroundColor), wt = g(() => ({
      ...e.value.style.chart.layout.legend,
      textAlign: "center",
      show: !0,
      position: "bottom"
    })), Ct = g(() => e.value.style.chart.layout.title), { exportSvg: $t, getSvg: Tt } = Wt({
      svg: ye,
      title: Ct,
      legend: wt,
      legendItems: xt,
      backgroundColor: kt,
      titleEmbedded: !0
    });
    async function De({ isCb: l }) {
      ae.value = !0, await ue();
      try {
        if (l) {
          const { blob: r, url: t, text: s, dataUrl: b } = await Tt();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: r, url: t, text: s, dataUrl: b }));
        } else
          await Promise.resolve($t());
      } finally {
        ae.value = !1;
      }
    }
    function At(l) {
      if (l?.stage === "start") {
        le.value = !0;
        return;
      }
      if (l?.stage === "end") {
        le.value = !1;
        return;
      }
      Le();
    }
    async function Xe() {
      if (U("copyAlt", {
        config: e.value,
        dataset: N.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: N.value
      }));
    }
    return He({
      getData: rt,
      getImage: pt,
      generatePdf: Pe,
      generateCsv: _e,
      generateImage: Le,
      generateSvg: De,
      toggleTable: je,
      toggleAnnotator: me,
      toggleFullscreen: Me,
      copyAlt: Xe
    }), (l, r) => (a(), o("div", {
      class: T(`vue-data-ui-component vue-ui-chestnut ${Z.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      ref_key: "chestnutChart",
      ref: H,
      id: `vue-ui-chestnut_${A.value}`,
      style: y(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor}`),
      onMouseenter: r[1] || (r[1] = () => n(Fe)(!0)),
      onMouseleave: r[2] || (r[2] = () => n(Fe)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (a(), re(n(Je), {
        key: 0,
        svgRef: n(ye),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: se.value,
        isCursorPointer: P.value,
        onClose: me
      }, {
        "annotator-action-close": C(() => [
          x(l.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": C(({ color: t }) => [
          x(l.$slots, "annotator-action-color", V(B({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": C(({ mode: t }) => [
          x(l.$slots, "annotator-action-draw", V(B({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": C(({ disabled: t }) => [
          x(l.$slots, "annotator-action-undo", V(B({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": C(({ disabled: t }) => [
          x(l.$slots, "annotator-action-redo", V(B({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": C(({ disabled: t }) => [
          x(l.$slots, "annotator-action-delete", V(B({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : h("", !0),
      nt.value ? (a(), o("div", tl, null, 512)) : h("", !0),
      e.value.userOptions.show && ze.value && (n(Ie) || n(ve)) ? (a(), re(n(Qe), {
        ref_key: "userOptionsRef",
        ref: ce,
        key: `user_options_${Se.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isImaging: n(be),
        isPrinting: n(fe),
        uid: A.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: Z.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: H.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: se.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: P.value,
        onToggleFullscreen: Me,
        onGeneratePdf: n(Pe),
        onGenerateCsv: _e,
        onGenerateImage: At,
        onGenerateSvg: De,
        onToggleTable: je,
        onToggleAnnotator: me,
        onCopyAlt: Xe,
        style: y({
          visibility: n(Ie) ? n(ve) ? "visible" : "hidden" : "visible"
        })
      }, Ee({ _: 2 }, [
        l.$slots.menuIcon ? {
          name: "menuIcon",
          fn: C(({ isOpen: t, color: s }) => [
            x(l.$slots, "menuIcon", V(B({ isOpen: t, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        l.$slots.optionPdf ? {
          name: "optionPdf",
          fn: C(() => [
            x(l.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        l.$slots.optionCsv ? {
          name: "optionCsv",
          fn: C(() => [
            x(l.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        l.$slots.optionImg ? {
          name: "optionImg",
          fn: C(() => [
            x(l.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        l.$slots.optionSvg ? {
          name: "optionSvg",
          fn: C(() => [
            x(l.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        l.$slots.optionTable ? {
          name: "optionTable",
          fn: C(() => [
            x(l.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        l.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: C(({ toggleFullscreen: t, isFullscreen: s }) => [
            x(l.$slots, "optionFullscreen", V(B({ toggleFullscreen: t, isFullscreen: s })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        l.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: C(({ toggleAnnotator: t, isAnnotator: s }) => [
            x(l.$slots, "optionAnnotator", V(B({ toggleAnnotator: t, isAnnotator: s })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        l.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: C(({ altCopy: t }) => [
            x(l.$slots, "optionAltCopy", V(B({ altCopy: t })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : h("", !0),
      i.value.height > 0 ? (a(), o("svg", {
        key: 3,
        ref_key: "svgRef",
        ref: ye,
        xmlns: n(Bt),
        class: T({ "vue-data-ui-fullscreen--on": Z.value, "vue-data-ui-fulscreen--off": !Z.value }),
        viewBox: `0 0 ${i.value.width <= 0 ? 10 : i.value.width} ${i.value.height <= 0 ? 10 : i.value.height}`,
        style: y(`overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        we(n(et)),
        l.$slots["chart-background"] ? (a(), o("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: i.value.width <= 0 ? 10 : i.value.width,
          height: i.value.height <= 0 ? 10 : i.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          x(l.$slots, "chart-background", {}, void 0, !0)
        ], 8, al)) : h("", !0),
        d.value ? h("", !0) : (a(), o("g", ol, [
          e.value.style.chart.layout.title.text ? (a(), o("text", {
            key: 0,
            "text-anchor": "middle",
            fill: e.value.style.chart.layout.title.color,
            "font-weight": e.value.style.chart.layout.title.bold ? "bold" : "normal",
            "font-size": e.value.style.chart.layout.title.fontSize,
            x: i.value.width / 2,
            y: 12 + e.value.style.chart.layout.title.fontSize + e.value.style.chart.layout.title.offsetY,
            onClick: z
          }, c(e.value.style.chart.layout.title.text), 9, sl)) : h("", !0),
          e.value.style.chart.layout.title.subtitle.text ? (a(), o("text", {
            key: 1,
            "text-anchor": "middle",
            fill: e.value.style.chart.layout.title.subtitle.color,
            "font-weight": e.value.style.chart.layout.title.subtitle.bold ? "bold" : "normal",
            "font-size": e.value.style.chart.layout.title.subtitle.fontSize,
            x: i.value.width / 2,
            y: 48 + e.value.style.chart.layout.title.subtitle.fontSize + e.value.style.chart.layout.title.subtitle.offsetY,
            onClick: z
          }, c(e.value.style.chart.layout.title.subtitle.text), 9, nl)) : h("", !0)
        ])),
        u("defs", null, [
          (a(!0), o(_, null, m(N.value, (t, s) => (a(), o("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `root_gradient_${A.value}_${t.rootIndex}`
          }, [
            u("stop", {
              offset: "0%",
              "stop-color": n(I)(n(Ue)(t.color, 0.05), 100 - e.value.style.chart.layout.roots.gradientIntensity)
            }, null, 8, rl),
            u("stop", {
              offset: "100%",
              "stop-color": t.color
            }, null, 8, il)
          ], 8, ul))), 256)),
          (a(!0), o(_, null, m(N.value, (t) => (a(), o("linearGradient", {
            x1: "0%",
            y1: "0%",
            x2: "100%",
            y2: "0%",
            id: `branch_gradient_${A.value}_${t.rootIndex}`
          }, [
            u("stop", {
              offset: "0%",
              "stop-color": t.color
            }, null, 8, dl),
            u("stop", {
              offset: "100%",
              "stop-color": n(I)(n(Ue)(t.color, 0.02), 100 - e.value.style.chart.layout.branches.gradientIntensity)
            }, null, 8, vl)
          ], 8, cl))), 256)),
          u("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `nutpick_${A.value}`
          }, [
            u("stop", {
              offset: "0%",
              "stop-color": n(I)("#FFFFFF", 0)
            }, null, 8, hl),
            u("stop", {
              offset: "80%",
              "stop-color": n(I)("#FFFFFF", e.value.style.chart.layout.nuts.selected.gradientIntensity)
            }, null, 8, fl),
            u("stop", {
              offset: "100%",
              "stop-color": n(I)("#FFFFFF", 0)
            }, null, 8, bl)
          ], 8, yl),
          u("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `nut_${A.value}`
          }, [
            u("stop", {
              offset: "0%",
              "stop-color": n(I)("#FFFFFF", 0)
            }, null, 8, gl),
            u("stop", {
              offset: "80%",
              "stop-color": n(I)("#FFFFFF", e.value.style.chart.layout.nuts.gradientIntensity)
            }, null, 8, _l),
            u("stop", {
              offset: "100%",
              "stop-color": n(I)("#FFFFFF", 0)
            }, null, 8, ml)
          ], 8, pl),
          u("radialGradient", {
            cx: "50%",
            cy: "50%",
            r: "50%",
            fx: "50%",
            fy: "50%",
            id: `nut_underlayer_${A.value}`
          }, [
            u("stop", {
              offset: "0%",
              "stop-color": n(I)(e.value.style.chart.backgroundColor, 100)
            }, null, 8, kl),
            u("stop", {
              offset: "80%",
              "stop-color": n(I)(e.value.style.chart.backgroundColor, 60)
            }, null, 8, wl),
            u("stop", {
              offset: "100%",
              "stop-color": n(I)(e.value.style.chart.backgroundColor, 0)
            }, null, 8, Cl)
          ], 8, xl)
        ]),
        e.value.style.chart.layout.grandTotal.show ? (a(), o("g", $l, [
          u("text", {
            x: S.value.seedX,
            y: 32 + e.value.style.chart.layout.grandTotal.offsetY,
            "font-size": e.value.style.chart.layout.grandTotal.fontSize,
            "font-weight": e.value.style.chart.layout.grandTotal.bold ? "bold" : "normal",
            fill: e.value.style.chart.layout.grandTotal.color,
            "text-anchor": "middle",
            onClick: z
          }, c(e.value.style.chart.layout.grandTotal.text), 9, Tl),
          u("text", {
            x: S.value.seedX,
            y: 38 + e.value.style.chart.layout.grandTotal.fontSize + e.value.style.chart.layout.grandTotal.offsetY,
            "font-size": e.value.style.chart.layout.grandTotal.fontSize,
            "font-weight": e.value.style.chart.layout.grandTotal.bold ? "bold" : "normal",
            fill: e.value.style.chart.layout.grandTotal.color,
            "text-anchor": "middle",
            onClick: z
          }, c(n(E)(
            e.value.style.chart.layout.grandTotal.formatter,
            L.value,
            n(F)({
              p: e.value.style.chart.layout.grandTotal.prefix,
              v: L.value,
              s: e.value.style.chart.layout.grandTotal.suffix,
              r: e.value.style.chart.layout.grandTotal.roundingValue
            })
          )), 9, Al)
        ])) : h("", !0),
        (a(!0), o(_, null, m(Y.value, (t) => (a(), o("g", null, [
          u("defs", null, [
            u("linearGradient", {
              id: `link_grad_${t.id}`
            }, [
              u("stop", {
                offset: "0%",
                "stop-color": t.color
              }, null, 8, Sl),
              u("stop", {
                offset: "100%",
                "stop-color": n(I)(t.color, e.value.style.chart.layout.links.opacity)
              }, null, 8, Nl)
            ], 8, zl)
          ]),
          u("path", {
            d: gt(t),
            stroke: n(I)(t.color, e.value.style.chart.layout.links.opacity),
            fill: `url(#link_grad_${t.id})`,
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            style: y(`opacity:${R(t) ? 1 : 0}`),
            onClick: z
          }, null, 12, Fl)
        ]))), 256)),
        (a(!0), o(_, null, m(W.value, (t) => (a(), o("circle", {
          cx: t.x,
          cy: t.y,
          r: t.r,
          fill: e.value.style.chart.layout.roots.underlayerColor,
          stroke: "none",
          style: y(`cursor:${P.value ? "pointer" : "default"}; opacity:${R(t) ? 1 : 0.05}`)
        }, null, 12, Il))), 256)),
        (a(!0), o(_, null, m(W.value, (t, s) => (a(), o("circle", {
          cx: t.x,
          cy: t.y,
          r: t.r,
          fill: e.value.style.chart.layout.roots.useGradient ? `url(#root_gradient_${A.value}_${t.rootIndex})` : t.color,
          stroke: e.value.style.chart.layout.roots.stroke,
          "stroke-width": e.value.style.chart.layout.roots.strokeWidth,
          style: y(`cursor:${P.value ? "pointer" : "default"}; opacity:${R(t) ? 1 : 0.05}`),
          onClick: (b) => Ve(t)
        }, null, 12, Pl))), 256)),
        e.value.style.chart.layout.roots.labels.show ? (a(), o("g", Ll, [
          (a(!0), o(_, null, m(W.value, (t, s) => (a(), o("text", {
            x: t.x,
            y: t.y + e.value.style.chart.layout.roots.labels.fontSize / 2.6,
            "text-anchor": "middle",
            "font-size": e.value.style.chart.layout.roots.labels.fontSize,
            fill: e.value.style.chart.layout.roots.labels.adaptColorToBackground ? n(We)(t.color) : e.value.style.chart.layout.roots.labels.color,
            "font-weight": "bold",
            style: y(`cursor:${P.value ? "pointer" : "default"}; opacity:${R(t) ? 1 : 0.05}`),
            onClick: (b) => Ve(t)
          }, c(n(E)(
            e.value.style.chart.layout.roots.labels.formatter,
            t.total,
            n(F)({
              p: e.value.style.chart.layout.roots.labels.prefix,
              v: t.total,
              s: e.value.style.chart.layout.roots.labels.suffix,
              r: e.value.style.chart.layout.roots.labels.roundingValue
            }),
            { datapoint: t }
          )), 13, Ol))), 256)),
          (a(!0), o(_, null, m(W.value, (t) => (a(), o("g", null, [
            d.value && t.rootIndex === d.value.rootIndex || k.value && t.rootIndex === k.value.rootIndex || O.value && t.rootIndex === O.value.rootIndex ? (a(), o("g", Rl, [
              u("text", {
                x: t.x,
                y: t.y + t.r + 24,
                "text-anchor": "middle",
                fill: e.value.style.chart.layout.roots.labels.name.color,
                "font-size": e.value.style.chart.layout.roots.labels.name.fontSize,
                "font-weight": e.value.style.chart.layout.roots.labels.name.bold ? "bold" : "normal",
                onClick: z
              }, c(t.name), 9, Vl)
            ])) : h("", !0)
          ]))), 256))
        ])) : h("", !0),
        (a(!0), o(_, null, m(Y.value, (t) => (a(), o("rect", {
          x: t.x1,
          y: t.y1,
          height: i.value.branchSize,
          width: t.x2 - t.x1,
          fill: e.value.style.chart.layout.branches.underlayerColor,
          rx: e.value.style.chart.layout.branches.borderRadius,
          stroke: "none",
          style: y(`opacity:${R(t) ? 1 : 0.05}`),
          onClick: (s) => ge(t)
        }, null, 12, Bl))), 256)),
        (a(!0), o(_, null, m(Y.value, (t, s) => (a(), o("rect", {
          x: t.x1,
          y: t.y1,
          height: i.value.branchSize,
          width: t.x2 - t.x1,
          fill: e.value.style.chart.layout.branches.useGradient ? `url(#branch_gradient_${A.value}_${t.rootIndex})` : t.color,
          rx: e.value.style.chart.layout.branches.borderRadius,
          stroke: e.value.style.chart.layout.branches.stroke,
          "stroke-width": e.value.style.chart.layout.branches.strokeWidth,
          style: y(`cursor:${P.value ? "pointer" : "default"}; opacity:${R(t) ? 1 : 0.05}`),
          onClick: (b) => ge(t)
        }, null, 12, Ml))), 256)),
        e.value.style.chart.layout.branches.labels.dataLabels.show ? (a(), o("g", jl, [
          (a(!0), o(_, null, m(Y.value, (t) => (a(), o("g", null, [
            t.proportionToRoot * 100 > e.value.style.chart.layout.branches.labels.dataLabels.hideUnderValue ? (a(), o("text", {
              key: 0,
              x: t.x1 + 6,
              y: t.y1 + i.value.branchSize / 1.5,
              "text-anchor": "start",
              fill: n(We)(t.color),
              "font-size": e.value.style.chart.layout.branches.labels.dataLabels.fontSize,
              "font-weight": "bold",
              style: y(`cursor:${P.value ? "pointer" : "default"}; opacity:${R(t) ? 1 : 0.05}`),
              onClick: (s) => ge(t)
            }, c(n(E)(
              e.value.style.chart.layout.branches.labels.dataLabels.formatter,
              t.value,
              n(F)({
                p: e.value.style.chart.layout.branches.labels.dataLabels.prefix,
                v: t.value,
                s: e.value.style.chart.layout.branches.labels.dataLabels.suffix,
                r: e.value.style.chart.layout.branches.labels.dataLabels.roundingValue
              }),
              { datapoint: t }
            )), 13, Dl)) : h("", !0)
          ]))), 256))
        ])) : h("", !0),
        (a(!0), o(_, null, m(Y.value, (t, s) => (a(), o("g", null, [
          (a(!0), o(_, null, m(n(Ye)(
            { series: t.breakdown, base: 1 },
            t.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            t.y1 + i.value.branchSize / 2,
            i.value.branchSize / 3,
            i.value.branchSize / 3
          ), (b, p) => (a(), o("path", {
            d: b.path,
            stroke: b.color,
            "stroke-width": 10,
            fill: "none",
            style: y(`opacity:${R(t) ? 1 : 0.1}`)
          }, null, 12, Xl))), 256)),
          u("circle", {
            fill: e.value.style.chart.layout.nuts.useGradient ? `url(#nut_${A.value})` : "transparent",
            cx: t.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            cy: t.y1 + i.value.branchSize / 2,
            r: i.value.branchSize / 2 + 2,
            onClick: (b) => ft(t),
            style: y(`cursor:${P.value ? "pointer" : "default"};opacity:${R(t) ? 1 : 0.1}`)
          }, null, 12, El)
        ]))), 256)),
        e.value.style.chart.layout.branches.labels.show && !k.value ? (a(), o("g", Gl, [
          (a(!0), o(_, null, m(Y.value, (t) => (a(), o("text", {
            x: t.x2 + i.value.branchSize + 24 + e.value.style.chart.layout.nuts.offsetX,
            y: t.y1 + i.value.branchSize / 2 + 5,
            "font-size": e.value.style.chart.layout.branches.labels.fontSize,
            "font-weight": e.value.style.chart.layout.branches.labels.bold ? "bold" : "normal",
            fill: e.value.style.chart.layout.branches.labels.color,
            "text-anchor": "start",
            style: y(`opacity:${R(t) ? 1 : 0.1}`)
          }, c(t.name), 13, Ul))), 256))
        ])) : h("", !0),
        u("line", {
          x1: 256 + i.value.padding.left,
          x2: 256 + i.value.padding.left,
          y1: S.value.top,
          y2: S.value.bottom,
          stroke: e.value.style.chart.layout.verticalSeparator.stroke,
          "stroke-width": e.value.style.chart.layout.verticalSeparator.strokeWidth
        }, null, 8, Wl),
        !d.value && !k.value ? (a(), o("foreignObject", {
          key: 6,
          x: 0,
          y: S.value.bottom,
          height: i.value.height - S.value.bottom,
          width: i.value.width,
          style: { overflow: "visible" },
          "data-no-svg-export": "",
          onClick: z
        }, [
          u("div", ql, [
            u("div", Hl, [
              (a(!0), o(_, null, m(W.value, (t) => (a(), o("div", {
                style: y(`display:flex;align-items:center;gap:3px;flex-direction:row;font-size:${e.value.style.chart.layout.legend.fontSize}px;`)
              }, [
                (a(), o("svg", Kl, [
                  u("circle", {
                    cx: "10",
                    cy: "10",
                    r: "10",
                    fill: t.color,
                    stroke: "none"
                  }, null, 8, Zl)
                ])),
                n(Ne) ? h("", !0) : (a(), o(_, { key: 0 }, [
                  u("span", null, c(t.name) + ":", 1),
                  u("b", null, c(n(E)(
                    e.value.style.chart.layout.roots.labels.formatter,
                    t.total,
                    n(F)({
                      p: e.value.style.chart.layout.legend.prefix,
                      v: t.total,
                      s: e.value.style.chart.layout.legend.suffix,
                      r: e.value.style.chart.layout.legend.roundingValue
                    }),
                    { datapoint: t }
                  )), 1),
                  ee(" (" + c(n(F)({
                    v: t.total / L.value * 100,
                    s: "%",
                    r: e.value.style.chart.layout.legend.roundingPercentage
                  })) + ") ", 1)
                ], 64))
              ], 4))), 256))
            ])
          ])
        ], 8, Yl)) : h("", !0),
        d.value && X.value ? (a(), o("g", Jl, [
          (a(), o("foreignObject", {
            x: 0,
            y: bt(),
            height: i.value.height - S.value.bottom,
            width: i.value.width,
            style: { overflow: "visible" },
            onClick: z
          }, [
            u("div", ea, [
              u("b", null, c(d.value.name), 1),
              u("div", ta, [
                (a(!0), o(_, null, m(d.value.breakdown, (t, s) => (a(), o("div", {
                  style: y(`display:flex;align-items:center;gap:6px;flex-direction:row;font-size:${e.value.style.chart.layout.legend.fontSize}px;`)
                }, [
                  (a(), o("svg", la, [
                    u("circle", {
                      cx: "10",
                      cy: "10",
                      r: "10",
                      fill: t.color,
                      stroke: "none"
                    }, null, 8, aa)
                  ])),
                  u("span", null, [
                    ee(c(t.name) + ": ", 1),
                    u("b", null, c(e.value.style.chart.layout.legend.prefix) + " " + c(t.value.toFixed(e.value.style.chart.layout.nuts.selected.labels.roundingValue)) + " " + c(e.value.style.chart.layout.legend.suffix), 1),
                    ee(" (" + c((t.proportionToBranch * 100).toFixed(e.value.style.chart.layout.nuts.selected.labels.roundingPercentage)) + "%)", 1)
                  ])
                ], 4))), 256))
              ])
            ])
          ], 8, Ql)),
          u("circle", {
            cx: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            cy: d.value.y1 + i.value.branchSize / 2,
            r: 256,
            fill: `url(#nut_underlayer_${A.value})`,
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, null, 10, oa),
          (a(!0), o(_, null, m(X.value, (t) => (a(), o("g", null, [
            K(t) ? (a(), o("path", {
              key: 0,
              d: n(Mt)(
                t,
                {
                  x: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
                  y: d.value.y1 + i.value.branchSize / 2
                },
                16,
                16,
                !1,
                !1,
                64
              ),
              stroke: t.color,
              "stroke-width": "1",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              fill: "none",
              class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
            }, null, 10, sa)) : h("", !0)
          ]))), 256)),
          u("circle", {
            cx: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            cy: d.value.y1 + i.value.branchSize / 2,
            r: 118,
            fill: e.value.style.chart.backgroundColor,
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, null, 10, na),
          (a(!0), o(_, null, m(X.value, (t) => (a(), o("path", {
            d: t.path,
            stroke: t.color,
            "stroke-width": 64,
            fill: "none",
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, null, 10, ua))), 256)),
          u("circle", {
            cx: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            cy: d.value.y1 + i.value.branchSize / 2,
            r: 110,
            fill: `url(#nutpick_${A.value})`,
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, null, 10, ra),
          u("circle", {
            cx: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            cy: d.value.y1 + i.value.branchSize / 2,
            r: 64,
            fill: e.value.style.chart.backgroundColor,
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, null, 10, ia),
          u("text", {
            x: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            y: d.value.y1 + 8,
            fill: e.value.style.chart.layout.nuts.selected.labels.core.total.color,
            "font-size": e.value.style.chart.layout.nuts.selected.labels.core.total.fontSize,
            "font-weight": e.value.style.chart.layout.nuts.selected.labels.core.total.bold ? "bold" : "normal",
            "text-anchor": "middle",
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, c(e.value.translations.total), 11, ca),
          u("text", {
            x: d.value.x2 + 24 + e.value.style.chart.layout.nuts.offsetX,
            y: d.value.y1 + 36,
            fill: e.value.style.chart.layout.nuts.selected.labels.core.value.color,
            "font-size": e.value.style.chart.layout.nuts.selected.labels.core.value.fontSize,
            "font-weight": e.value.style.chart.layout.nuts.selected.labels.core.value.bold ? "bold" : "normal",
            "text-anchor": "middle",
            onClick: q,
            class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
          }, c(n(E)(
            e.value.style.chart.layout.nuts.selected.labels.dataLabels.formatter,
            d.value.value,
            n(F)({
              p: e.value.style.chart.layout.nuts.selected.labels.core.value.prefix,
              v: d.value.value,
              s: e.value.style.chart.layout.nuts.selected.labels.core.value.suffix,
              r: e.value.style.chart.layout.nuts.selected.roundingValue
            }),
            { datapoint: d.value }
          )), 11, da),
          (a(!0), o(_, null, m(X.value, (t, s) => (a(), o("g", null, [
            K(t) ? (a(), o("text", {
              key: 0,
              x: n(j)(t).x,
              "text-anchor": n(j)(t).anchor,
              y: n(te)(t) - e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize / 6,
              fill: t.color,
              "font-size": e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize / 2,
              style: y(`font-weight:${e.value.style.chart.layout.nuts.selected.labels.dataLabels.bold ? "bold" : ""}`),
              class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
            }, " ⬤ ", 14, va)) : h("", !0),
            K(t) ? (a(), o("text", {
              key: 1,
              x: n(j)(t, !0).x,
              "text-anchor": n(j)(t, !0).anchor,
              y: n(te)(t),
              fill: e.value.style.chart.layout.nuts.selected.labels.dataLabels.color,
              "font-size": e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize,
              style: y(`font-weight:${e.value.style.chart.layout.nuts.selected.labels.dataLabels.bold ? "bold" : ""}`),
              class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
            }, c(d.value.breakdown[s].name), 15, ya)) : h("", !0)
          ]))), 256)),
          (a(!0), o(_, null, m(X.value, (t, s) => (a(), o("g", null, [
            K(t) ? (a(), o("text", {
              key: 0,
              x: n(j)(t, !0).x,
              "text-anchor": n(j)(t).anchor,
              y: n(te)(t) + e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize,
              fill: e.value.style.chart.layout.nuts.selected.labels.dataLabels.color,
              "font-size": e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize,
              style: y(`font-weight:${e.value.style.chart.layout.nuts.selected.labels.dataLabels.bold ? "bold" : ""}`),
              class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
            }, c(n(F)({
              v: d.value.breakdown[s].value / d.value.value * 100,
              s: "%",
              r: e.value.style.chart.layout.nuts.selected.labels.roundingPercentage
            })) + " " + c(e.value.translations.of) + " " + c(d.value.breakdown[s].branchName) + " " + c(n(E)(
              e.value.style.chart.layout.nuts.selected.labels.dataLabels.formatter,
              d.value.breakdown[s].value,
              n(F)({
                p: e.value.style.chart.layout.nuts.selected.labels.dataLabels.prefix,
                v: d.value.breakdown[s].value,
                s: e.value.style.chart.layout.nuts.selected.labels.dataLabels.suffix,
                r: e.value.style.chart.layout.nuts.selected.roundingValue
              }),
              { datapoint: X.value, seriesIndex: s }
            )), 15, ha)) : h("", !0),
            K(t) ? (a(), o("text", {
              key: 1,
              x: n(j)(t, !0).x,
              "text-anchor": n(j)(t).anchor,
              y: n(te)(t) + e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize * 2,
              fill: e.value.style.chart.layout.nuts.selected.labels.dataLabels.color,
              "font-size": e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize,
              style: y(`font-weight:${e.value.style.chart.layout.nuts.selected.labels.dataLabels.bold ? "bold" : ""}`),
              class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
            }, c(n(F)({
              v: d.value.breakdown[s].proportionToRoot * 100,
              s: "%",
              r: e.value.style.chart.layout.nuts.selected.labels.roundingPercentage
            })) + " " + c(e.value.translations.of) + " " + c(d.value.breakdown[s].rootName), 15, fa)) : h("", !0),
            K(t) ? (a(), o("text", {
              key: 2,
              x: n(j)(t, !0).x,
              "text-anchor": n(j)(t).anchor,
              y: n(te)(t) + e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize * 3,
              fill: e.value.style.chart.layout.nuts.selected.labels.dataLabels.color,
              "font-size": e.value.style.chart.layout.nuts.selected.labels.dataLabels.fontSize,
              style: y(`font-weight:${e.value.style.chart.layout.nuts.selected.labels.dataLabels.bold ? "bold" : ""}`),
              class: T(e.value.style.chart.layout.nuts.selected.useMotion ? "vue-ui-chestnut-animated" : "")
            }, c(n(F)({
              v: d.value.breakdown[s].proportionToTree * 100,
              s: "%",
              r: e.value.style.chart.layout.nuts.selected.labels.roundingPercentage
            })) + " " + c(e.value.translations.proportionToTree), 15, ba)) : h("", !0)
          ]))), 256))
        ])) : h("", !0),
        (a(!0), o(_, null, m(Y.value, (t) => (a(), o("g", null, [
          k.value && k.value.id === t.id && !d.value ? (a(), o("text", {
            key: 0,
            x: t.x1 + 6,
            y: t.y1 + i.value.branchSize + 24,
            "font-weight": "bold",
            "text-anchor": "start",
            "font-size": e.value.style.chart.layout.branches.labels.dataLabels.fontSize,
            fill: e.value.style.chart.layout.branches.labels.color,
            onClick: z
          }, c(t.name) + ": " + c(n(E)(
            e.value.style.chart.layout.branches.labels.dataLabels.formatter,
            t.value,
            n(F)({
              p: e.value.style.chart.layout.branches.labels.dataLabels.prefix,
              v: t.value,
              s: e.value.style.chart.layout.branches.labels.dataLabels.suffix,
              r: e.value.style.chart.layout.branches.labels.dataLabels.roundingValue
            }),
            { datapoint: t }
          )), 9, pa)) : h("", !0),
          k.value && k.value.id === t.id && !d.value ? (a(), o("text", {
            key: 1,
            x: t.x1 + 6,
            y: t.y1 + i.value.branchSize + 48,
            "text-anchor": "start",
            "font-size": e.value.style.chart.layout.branches.labels.dataLabels.fontSize,
            fill: e.value.style.chart.layout.branches.labels.color,
            onClick: z
          }, c(n(F)({
            v: t.proportionToRoot * 100,
            s: "%",
            r: e.value.style.chart.layout.branches.labels.dataLabels.roundingPercentage
          })) + " " + c(e.value.translations.of) + " " + c(t.rootName), 9, ga)) : h("", !0),
          k.value && k.value.id === t.id && !d.value ? (a(), o("text", {
            key: 2,
            x: t.x1 + 6,
            y: t.y1 + i.value.branchSize + 72,
            "text-anchor": "start",
            "font-size": e.value.style.chart.layout.branches.labels.dataLabels.fontSize,
            fill: e.value.style.chart.layout.branches.labels.color,
            onClick: z
          }, c(n(F)({
            v: t.value / L.value * 100,
            s: "%",
            r: e.value.style.chart.layout.branches.labels.dataLabels.roundingPercentage
          })) + " " + c(e.value.translations.proportionToTree), 9, _a)) : h("", !0)
        ]))), 256)),
        x(l.$slots, "svg", {
          svg: {
            ...i.value,
            isPrintingImg: n(fe) | n(be) | le.value,
            isPrintingSvg: ae.value
          }
        }, void 0, !0)
      ], 14, ll)) : h("", !0),
      l.$slots.watermark ? (a(), o("div", ma, [
        x(l.$slots, "watermark", V(B({ isPrinting: n(fe) || n(be) || le.value || ae.value })), void 0, !0)
      ])) : h("", !0),
      x(l.$slots, "legend", { legend: N.value }, void 0, !0),
      l.$slots.source ? (a(), o("div", xa, [
        x(l.$slots, "source", {}, void 0, !0)
      ], 512)) : h("", !0),
      ze.value && e.value.userOptions.buttons.table ? (a(), re(Ft(ne.value.component), It({ key: 6 }, ne.value.props, {
        ref_key: "tableUnit",
        ref: Q,
        onClose: xe
      }), Ee({
        content: C(() => [
          u("div", {
            ref_key: "tableContainer",
            ref: Oe,
            class: "vue-ui-chestnut-table",
            style: y(`${e.value.table.useDialog ? "" : "max-height: 300px;margin-top:24px"}`)
          }, [
            u("div", {
              style: y(`${e.value.table.useDialog ? "" : "padding-top:36px;"}position: relative`)
            }, [
              e.value.table.useDialog ? h("", !0) : (a(), o("div", {
                key: 0,
                role: "button",
                tabindex: "0",
                style: y(`width:32px; position: absolute; top: 0; left:4px; padding: 0 0px; display: flex; align-items:center;justify-content:center;height: 36px; width: 32px; cursor:${P.value ? "pointer" : "default"}; background:${e.value.table.th.backgroundColor};`),
                onClick: xe,
                onKeypress: Pt(xe, ["enter"])
              }, [
                we(n(Ae), {
                  name: "close",
                  stroke: e.value.table.th.color,
                  "stroke-width": 2
                }, null, 8, ["stroke"])
              ], 36)),
              u("div", {
                style: { width: "100%" },
                class: T({ "vue-ui-responsive": n(_t) })
              }, [
                u("table", ka, [
                  e.value.table.useDialog ? h("", !0) : (a(), o("caption", {
                    key: 0,
                    style: y({ backgroundColor: e.value.table.th.backgroundColor, color: e.value.table.th.color, outline: e.value.table.th.outline }),
                    class: "vue-ui-data-table__caption"
                  }, [
                    ee(c(e.value.style.chart.layout.title.text) + " ", 1),
                    e.value.style.chart.layout.title.subtitle.text ? (a(), o("span", wa, c(e.value.style.chart.layout.title.subtitle.text), 1)) : h("", !0)
                  ], 4)),
                  u("thead", null, [
                    u("tr", {
                      role: "row",
                      style: y(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color}`)
                    }, [
                      (a(!0), o(_, null, m(v.value.head, (t) => (a(), o("th", {
                        style: y(`outline:${e.value.table.th.outline}`)
                      }, c(t), 5))), 256))
                    ], 4)
                  ]),
                  u("tbody", null, [
                    (a(!0), o(_, null, m(v.value.body, (t, s) => (a(), o("tr", {
                      class: T({ "vue-ui-data-table__tbody__row": !0, "vue-ui-data-table__tbody__row-even": s % 2 === 0, "vue-ui-data-table__tbody__row-odd": s % 2 !== 0 }),
                      style: y(`background:${e.value.table.td.backgroundColor};color:${e.value.table.td.color}`)
                    }, [
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[0]
                      }, [
                        u("div", $a, [
                          v.value.body[s - 1] && v.value.body[s - 1].rootName === t.rootName ? (a(), o("span", Ta)) : (a(), o("span", Aa, c(t.rootName), 1))
                        ])
                      ], 12, Ca),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[1]
                      }, [
                        u("div", Sa, [
                          v.value.body[s - 1] && v.value.body[s - 1].rootName === t.rootName ? (a(), o("span", Na)) : (a(), o("span", Fa, c(t.rootValue.toFixed(e.value.table.td.roundingValue)), 1))
                        ])
                      ], 12, za),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[2]
                      }, [
                        u("div", Pa, [
                          v.value.body[s - 1] && v.value.body[s - 1].rootName === t.rootName ? (a(), o("span", La)) : (a(), o("span", Oa, c((t.rootToTotal * 100).toFixed(e.value.table.td.roundingPercentage)) + "% ", 1))
                        ])
                      ], 12, Ia),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[3]
                      }, [
                        u("div", Va, [
                          v.value.body[s - 1] && v.value.body[s - 1].branchName === t.branchName ? (a(), o("span", Ba)) : (a(), o("span", Ma, c(t.branchName), 1))
                        ])
                      ], 12, Ra),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[4]
                      }, [
                        u("div", Da, [
                          v.value.body[s - 1] && v.value.body[s - 1].branchName === t.branchName ? (a(), o("span", Xa)) : (a(), o("span", Ea, c(t.branchValue.toFixed(e.value.table.td.roundingValue)), 1))
                        ])
                      ], 12, ja),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[5]
                      }, [
                        u("div", Ua, [
                          v.value.body[s - 1] && v.value.body[s - 1].branchName === t.branchName ? (a(), o("span", Wa)) : (a(), o("span", Ya, c((t.branchToRoot * 100).toFixed(e.value.table.td.roundingPercentage)) + "% ", 1))
                        ])
                      ], 12, Ga),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[6]
                      }, [
                        u("div", Ha, [
                          v.value.body[s - 1] && v.value.body[s - 1].branchName === t.branchName ? (a(), o("span", Ka)) : (a(), o("span", Za, c((t.branchToTotal * 100).toFixed(e.value.table.td.roundingPercentage)) + "% ", 1))
                        ])
                      ], 12, qa),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[7]
                      }, [
                        u("div", Qa, c(t.nutName), 1)
                      ], 12, Ja),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[8]
                      }, [
                        u("div", to, c(t.nutValue.toFixed(e.value.table.td.roundingValue)), 1)
                      ], 12, eo),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[9]
                      }, [
                        u("div", ao, c((t.nutToBranch * 100).toFixed(e.value.table.td.roundingPercentage)) + "% ", 1)
                      ], 12, lo),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[10]
                      }, [
                        u("div", so, c((t.nutToRoot * 100).toFixed(e.value.table.td.roundingPercentage)) + "% ", 1)
                      ], 12, oo),
                      u("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: y(`outline:${e.value.table.td.outline}`),
                        "data-cell": v.value.head[11]
                      }, [
                        u("div", uo, c((t.nutToTotal * 100).toFixed(e.value.table.td.roundingPercentage)) + "% ", 1)
                      ], 12, no)
                    ], 6))), 256))
                  ])
                ])
              ], 2)
            ], 4)
          ], 4)
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: C(() => [
            ee(c(ne.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: C(() => [
            u("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: r[0] || (r[0] = (t) => _e(e.value.userOptions.callbacks.csv))
            }, [
              we(n(Ae), {
                name: "fileCsv",
                stroke: ne.value.props.color
              }, null, 8, ["stroke"])
            ])
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : h("", !0),
      x(l.$slots, "skeleton", {}, () => [
        n(Ne) ? (a(), re(Gt, { key: 0 })) : h("", !0)
      ], !0)
    ], 46, el));
  }
}, Co = /* @__PURE__ */ Qt(ro, [["__scopeId", "data-v-c9973f6a"]]);
export {
  Co as default
};
