// vue-data-ui/dist/vue-ui-horizontal-bar-ClQcB9nL.js
import { useCssVars as wa, defineAsyncComponent as le, useSlots as xa, computed as d, ref as i, toRefs as _a, watch as oe, onBeforeUnmount as $a, onMounted as Re, openBlock as n, createElementBlock as u, unref as r, normalizeStyle as p, normalizeClass as Q, createBlock as H, withCtx as h, renderSlot as f, normalizeProps as O, guardReactiveProps as I, createCommentVNode as g, createSlots as be, createElementVNode as s, toDisplayString as m, createVNode as ne, Fragment as V, renderList as M, mergeProps as $t, resolveDynamicComponent as Ta, withKeys as La, createTextVNode as Tt, nextTick as me } from "vue";
import { j as Sa, t as Lt, m as Pa, p as ee, $ as Xe, n as za, q as ie, o as He, K as P, s as Ue, k as ve, l as j, X as Aa, u as We, v as Na, Z as Ge, ap as Oa, a0 as Ia, a1 as Va, T as Ma, U as Fa } from "./lib-GwPnJSrY.js";
import { t as Da, u as Ba } from "./useResponsive-DfdjqQps.js";
import { u as Ea } from "./useConfig-CIgyKZ6R.js";
import { u as Ra, B as Xa } from "./BaseScanner-BhhkKVoj.js";
import { u as Ha } from "./usePrinter-DiK5R6k-.js";
import { u as Ua } from "./useSvgExport-ByUukOZt.js";
import { u as je } from "./useNestedProp-BBT0mbHh.js";
import { u as Wa } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Ga } from "./useTableResponsive-Cbx-ACcP.js";
import { u as ja } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Ya } from "./useChartAccessibility-9icAAmYg.js";
import qa from "./img-BYYzY_QU.js";
import { _ as St } from "./Shape-CSyWYgXc.js";
import Ka from "./Title-BfyrP-jp.js";
import Za from "./vue_ui_horizontal_bar-BEagrV0T.js";
import Pt from "./Legend-1LMMsPxL.js";
import Ja from "./vue-ui-accordion-BGHTqIBt.js";
import { B as zt } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Qa } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const el = ["id"], tl = ["onClick"], al = ["xmlns", "viewBox"], ll = ["width", "height"], ol = ["id"], nl = ["stop-color"], ul = ["stop-color"], sl = { key: 1 }, rl = ["y", "width", "height", "fill", "rx"], il = ["x", "y", "width", "height", "fill", "rx"], vl = ["x", "y", "width", "height", "fill", "rx", "stroke", "stroke-width"], cl = ["x", "y", "width", "height", "fill", "rx", "stroke", "stroke-width"], dl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], hl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], pl = { class: "vue-ui-horizontal-bar-parent-label" }, fl = ["y", "height", "width", "fill"], gl = ["y", "font-size", "fill", "x", "font-weight", "innerHTML"], yl = ["y", "font-size", "fill", "x", "font-weight"], bl = { class: "vue-ui-horizontal-bar-child-label" }, ml = ["x", "y", "font-size", "font-weight", "innerHTML"], kl = { class: "vue-ui-horizontal-bar-child-label" }, Cl = ["x", "y", "font-size", "fill", "font-weight", "innerHTML"], wl = { class: "vue-ui-horizontal-bar-child-label" }, xl = ["x", "y", "text-anchor", "font-size", "fill", "font-weight"], _l = ["y", "width", "height", "fill", "onMouseenter", "onMouseleave", "onClick"], $l = {
  key: 5,
  class: "vue-data-ui-watermark"
}, Tl = ["onClick"], Ll = { style: "width:100%;padding-top: 36px;position:relative" }, Sl = { class: "vue-ui-data-table" }, Pl = { key: 0 }, zl = { style: { width: "100%" } }, Al = ["data-cell"], Nl = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Ol = ["data-cell"], Il = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Vl = ["data-cell"], Ml = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Fl = ["data-cell"], Dl = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Bl = ["data-cell"], El = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Rl = ["data-cell"], Xl = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Hl = ["data-cell"], Ul = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, Wl = {
  __name: "vue-ui-horizontal-bar",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "copyAlt"],
  setup(At, { expose: Nt, emit: Ot }) {
    wa((t) => ({
      e04bdc8e: t.tdo
    }));
    const It = le(() => import("./Tooltip-DL3QgjZf.js")), Ye = le(() => import("./BaseIcon-CNx8Tzl5.js")), Vt = le(() => import("./PenAndPaper-DWV1FMIi.js")), Mt = le(() => import("./UserOptions-CrLqEimB.js")), Ft = le(() => import("./PackageVersion-DN9d0ME5.js")), Dt = le(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_vertical_bar: Bt } = Ea(), { isThemeValid: Et, warnInvalidTheme: Rt } = Wa(), Xt = xa(), k = At, ke = d({
      get() {
        return !!k.dataset && k.dataset.length;
      },
      set(t) {
        return t;
      }
    }), A = i(Sa()), Ce = i(!1), we = i(""), Ht = i(null), qe = i(0), B = i(null), Ke = i(null), xe = i(null), Ze = i(null), Je = i(null), Qe = i(0), _e = i(0), Ut = i(null), ue = i(null), $e = i(null), Wt = i(null), Te = i(null), w = i([]), Gt = i({ none: 0, asc: 1, desc: 2 }), U = i(0), ce = i(!1), se = i(0), te = i(0), Le = i(null), de = i(!1), he = i(!1), et = Ot, e = i(Ae()), Y = d(() => e.value.userOptions.useCursorPointer), jt = d(() => {
      const t = [
        { name: "", value: 6, color: "#d9d9d9" },
        { name: "", value: 5, color: "#d9d9d9" },
        { name: "", value: 4, color: "#d9d9d9" },
        { name: "", value: 3, color: "#d9d9d9" },
        { name: "", value: 2, color: "#d9d9d9" },
        { name: "", value: 1, color: "#d9d9d9" }
      ];
      return k.config?.skeletonDataset ?? t;
    }), Yt = d(() => Lt({
      defaultConfig: {
        useCssAnimation: !1,
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              bars: {
                offsetX: 110,
                dataLabels: {
                  color: "transparent",
                  value: { prefix: "", suffix: "", formatter: null }
                },
                nameLabels: { color: "transparent" }
              },
              separators: { color: "#7A7A7A" }
            },
            legend: {
              show: !0,
              backgroundColor: "transparent",
              color: "transparent",
              prefix: "",
              suffix: "",
              formatter: null
            },
            title: { color: "#1A1A1A", subtitle: { color: "#5A5A5A" } }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: pe, FINAL_DATASET: tt, manualLoading: Se } = Ra({
      ..._a(k),
      FINAL_CONFIG: e,
      prepareConfig: Ae,
      callback: () => {
        Promise.resolve().then(async () => {
          _.value.showTable = e.value.table.show, Jt();
        });
      },
      skeletonDataset: jt.value,
      skeletonConfig: Lt({
        defaultConfig: e.value,
        userConfig: Yt.value
      })
    }), { userOptionsVisible: Pe, setUserOptionsVisibility: at, keepUserOptionState: lt } = ja({ config: e.value }), { svgRef: ze } = Ya({
      config: e.value.style.chart.title
    });
    function Ae() {
      const t = je({
        userConfig: k.config,
        defaultConfig: Bt
      });
      let l = {};
      const a = t.theme;
      if (a)
        if (!Et.value(t))
          Rt(t), l = t;
        else {
          const o = je({
            userConfig: Za[a] || k.config,
            defaultConfig: t
          });
          l = {
            ...je({ userConfig: k.config, defaultConfig: o }),
            customPalette: t.customPalette.length ? t.customPalette : Pa[a] || ee
          };
        }
      else
        l = t;
      return k.config && Xe(k.config, "events.datapointEnter") ? l.events.datapointEnter = k.config.events.datapointEnter : l.events.datapointEnter = null, k.config && Xe(k.config, "events.datapointLeave") ? l.events.datapointLeave = k.config.events.datapointLeave : l.events.datapointLeave = null, k.config && Xe(k.config, "events.datapointClick") ? l.events.datapointClick = k.config.events.datapointClick : l.events.datapointClick = null, l;
    }
    const F = i(e.value.style.chart.width), W = i(e.value.style.chart.height), C = i(e.value.style.chart.layout.bars.gap);
    oe(
      () => k.config,
      (t) => {
        pe.value || (e.value = Ae()), Pe.value = !e.value.userOptions.showOnChartHover, Qe.value += 1, _e.value += 1, _.value.showTable = e.value.table.show, _.value.sortDesc = e.value.style.chart.layout.bars.sort === "desc", _.value.showTooltip = e.value.style.chart.tooltip.show, F.value = e.value.style.chart.width, W.value = e.value.style.chart.height, C.value = e.value.style.chart.layout.bars.gap;
      },
      { deep: !0 }
    ), oe(
      () => k.dataset,
      (t) => {
        Array.isArray(t) && t.length > 0 && (Se.value = !1);
      },
      { deep: !0 }
    );
    const { isPrinting: Ne, isImaging: Oe, generatePdf: ot, generateImage: nt } = Ha({
      elementId: `vue-ui-vertical-bar_${A.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-vertical-bar",
      options: e.value.userOptions.print
    }), qt = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), ut = d(() => za(e.value.customPalette)), st = i(null), Kt = d(() => e.value.table.responsiveBreakpoint), fe = d(() => !!e.value.debug), Zt = i(0), G = i(null), q = i(null);
    function Jt() {
      if (B.value && (He(k.dataset) && (ie({
        componentName: "VueUiVerticalBar",
        type: "dataset",
        debug: fe.value
      }), ke.value = !1, Se.value = !0), He(k.dataset) || (Se.value = e.value.loading), e.value.responsive)) {
        const t = Da(() => {
          const { width: l, height: a } = Ba({
            chart: B.value,
            title: e.value.style.chart.title.text ? Ke.value : null,
            legend: e.value.style.chart.legend.show ? xe.value : null,
            source: Ze.value,
            noTitle: Je.value
          });
          Zt.value = a / 3, requestAnimationFrame(async () => {
            W.value = a, F.value = l;
          });
        });
        G.value && (q.value && G.value.unobserve(q.value), G.value.disconnect()), G.value = new ResizeObserver(t), me(() => {
          q.value = B.value?.parentNode, q.value && G.value.observe(q.value);
        });
      }
    }
    $a(() => {
      G.value && (q.value && G.value.unobserve(q.value), G.value.disconnect());
    });
    const _ = i({
      showTable: e.value.table.show,
      sortDesc: e.value.style.chart.layout.bars.sort === "desc",
      showTooltip: e.value.style.chart.tooltip.show
    }), rt = d(() => _.value.sortDesc), $ = d(() => (tt.value.forEach((t, l) => {
      !t.value && !t.children && ie({
        componentName: "VueUiVerticalBar",
        type: "datasetAttributeEmpty",
        property: `value (index ${l})`,
        debug: fe.value
      }), t.children && (He(t.children) ? ie({
        componentName: "VueUiVerticalBar",
        type: "datasetAttributeEmpty",
        property: `children (index ${l})`,
        debug: fe.value
      }) : t.children.forEach((a, o) => {
        [null, void 0].includes(a.name) && ie({
          componentName: "VueUiVerticalBar",
          type: "datasetSerieAttribute",
          property: "name",
          key: "children",
          index: o,
          debug: fe.value
        });
      }));
    }), tt.value.map((t, l) => {
      const a = `vertical_parent_${l}_${A.value}`, o = t.children && t.children.length > 0, y = P(
        t.value ? t.value : o ? t.children.map((b) => b.value || 0).reduce((b, N) => b + N, 0) : 0
      ), x = y >= 0 ? 1 : -1;
      return {
        ...t,
        id: a,
        absoluteIndex: l,
        shape: "square",
        opacity: w.value.includes(a) ? 0.5 : 1,
        value: Math.abs(y),
        sign: x,
        hasChildren: o,
        isChild: !1,
        segregate: () => K(a),
        isSegregated: w.value.includes(a),
        color: Ue(t.color) || ut.value[l] || ee[l] || ee[l % ee.length],
        children: !t.children || !t.children.length ? [] : t.children.toSorted(
          ce.value ? () => 0 : (b, N) => rt.value ? N.value - b.value : b.value - N.value
        ).map((b, N) => ({
          ...b,
          value: P(Math.abs(b.value)),
          absoluteIndex: l,
          sign: b.value >= 0 ? 1 : -1,
          isChild: !0,
          parentId: a,
          parentName: t.name,
          parentValue: y,
          parentSign: x,
          id: `vertical_child_${l}_${N}_${A.value}`,
          childIndex: N,
          color: Ue(b.color) || Ue(t.color) || ut.value[l] || ee[l] || ee[l % ee.length]
        })).map((b, N) => ({
          ...b,
          isFirstChild: N === 0,
          isLastChild: N === t.children.length - 1
        }))
      };
    }).toSorted(
      ce.value ? () => 0 : (t, l) => rt.value ? l.value - t.value : t.value - l.value
    ))), R = d(() => $.value.filter(
      (t) => !w.value.includes(t.id)
    )), T = d(() => R.value.flatMap((t) => t.hasChildren ? t.children : t)), Qt = d(() => $.value.flatMap((t) => t.hasChildren ? t.children : t)), it = d(() => R.value.flatMap((t) => t.children && t.children.length > 0 ? t.children.length : 1).reduce((t, l) => t + l, 0)), vt = d(() => ({
      cy: "vertical-bar-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), L = d(() => e.value.style.chart.layout.bars.parentLabels.show ? e.value.style.chart.layout.bars.parentLabels.fontSize * 3 + e.value.style.chart.layout.bars.parentLabels.paddingBottom : 0), S = d(() => {
      const t = [];
      let l = 0;
      return T.value.forEach((a, o) => {
        a.isChild && a.childIndex === 0 && e.value.style.chart.layout.bars.parentLabels.show && (l += 1), t[o] = l;
      }), t;
    }), c = d(() => {
      const l = Math.max(0, ...S.value) * L.value;
      return (W.value - 24 - (it.value - 1) * C.value - l) / it.value;
    }), X = d(() => ({
      width: F.value,
      height: W.value,
      padding: {
        top: 12,
        left: 128 + e.value.style.chart.layout.bars.offsetX,
        right: 64 + e.value.style.chart.layout.bars.paddingRight,
        bottom: 12
      }
    })), Ie = i(0);
    function ct() {
      me(() => {
        if (!Te.value) {
          Ie.value = 0;
          return;
        }
        const l = Array.from(Te.value.querySelectorAll("text")).reduce((a, o) => {
          const y = o.getBBox?.().width ?? o.getComputedTextLength();
          return y > a ? y : a;
        }, 0);
        Ie.value = l + 12;
      });
    }
    oe(
      () => [
        T.value,
        F.value,
        W.value,
        e.value.style.chart.layout.bars.nameLabels,
        e.value.style.chart.layout.bars.offsetX
      ],
      () => ct(),
      { deep: !0 }
    ), Re(() => {
      ct();
    });
    function re() {
      me(() => {
        se.value = 0, te.value = 0, requestAnimationFrame(() => {
          if (!Le.value) {
            se.value = 0, te.value = 0;
            return;
          }
          const t = Array.from(
            Le.value.querySelectorAll("text")
          );
          if (!t.length) {
            se.value = 0, te.value = 0;
            return;
          }
          let l = 1 / 0, a = 0;
          t.forEach((Ca) => {
            const Ee = Ca.getBBox(), xt = Ee.x, _t = Ee.x + Ee.width;
            xt < l && (l = xt), _t > a && (a = _t);
          });
          const o = 0, y = F.value, x = 24, b = Math.max(0, o - l + x), N = Math.max(0, a - y + x);
          te.value = b, se.value = N;
        });
      });
    }
    oe(w, () => re(), { deep: !0 }), oe(
      () => [T.value, F.value, W.value],
      () => re(),
      { deep: !0 }
    ), Re(() => re());
    const v = d(() => {
      const t = Ie.value + Math.abs(e.value.style.chart.layout.bars.nameLabels.offsetX) + te.value + e.value.style.chart.layout.bars.offsetX, l = X.value.width - X.value.padding.right, a = Math.max(0, l - t - se.value - te.value - e.value.style.chart.layout.bars.offsetX);
      return {
        top: X.value.padding.top,
        bottom: X.value.padding.top + X.value.height,
        left: t,
        right: l,
        width: a,
        fullHeight: X.value.padding.top + X.value.padding.bottom + X.value.height
      };
    });
    function dt() {
      w.value.length ? w.value = [] : $.value.forEach((t) => {
        w.value.push(t.id);
      }), re();
    }
    async function K(t) {
      if (w.value.includes(t))
        w.value = w.value.filter((l) => l !== t);
      else {
        if (w.value.length === $.value.length - 1) return;
        w.value.push(t);
      }
      et("selectLegend", R.value), re();
    }
    function ht(t) {
      if (!$.value.length)
        return e.value.debug && console.warn("VueUiHorizontalBar - There are no series to show."), null;
      const l = $.value.find((a) => a.name === t);
      return l || (e.value.debug && console.warn(`VueUiHorizontalBar - Series name not found "${t}"`), null);
    }
    function ea(t) {
      const l = ht(t);
      l !== null && w.value.includes(l.id) && K(l.id);
    }
    function ta(t) {
      const l = ht(t);
      l !== null && (w.value.includes(l.id) || K(l.id));
    }
    const E = d(() => R.value.map((t) => Math.abs(t.value)).reduce((t, l) => t + l, 0));
    function pt(t, l = !1, a = 0) {
      return l ? j({
        v: P(Math.abs(t) / E.value * 100),
        s: "%",
        r: a
      }) : Math.abs(t) / E.value;
    }
    const Z = d(() => T.value.map((t) => t.sign).includes(-1)), aa = d(() => Math.max(
      ...R.value.flatMap((t) => t.children && t.children.length ? Math.max(...t.children.map((l) => l.value)) : t.value)
    ));
    function z(t) {
      const l = t / aa.value;
      return P(v.value.width / (Z.value ? 2 : 1) * l);
    }
    function la(t) {
      return z(t) + v.value.left;
    }
    function ae(t, l) {
      const a = R.value.find((b) => b.id === t.parentId);
      return {
        y: v.value.top + (C.value + c.value) * l + S.value[l] * L.value - L.value + e.value.style.chart.layout.bars.parentLabels.fontSize,
        name: a.name,
        value: [void 0, NaN, null].includes(a.value) ? "" : a.sign === 1 ? a.value : -a.value,
        percentageToTotal: isNaN(a.value / E.value) ? "" : pt(
          a.value,
          !0,
          e.value.style.chart.layout.bars.dataLabels.percentage.roundingPercentage
        ),
        sign: a.sign
      };
    }
    function oa() {
      return R.value;
    }
    function ft(t, l) {
      return ae(t, l).name;
    }
    function na(t, l) {
      const a = ae(t, l);
      return gt(a.value, a, l, t.parentSign || t.sign);
    }
    const Ve = i(null), Me = i(null);
    function ua({ datapoint: t, seriesIndex: l }) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: l });
    }
    function sa({ datapoint: t, seriesIndex: l }) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: l }), Ht.value = null, Ce.value = !1, Ve.value = null;
    }
    function ra(t, l) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datpoint: t, seriesIndex: l }), Me.value = {
        datapoint: t,
        seriesIndex: l,
        series: $.value,
        config: e.value
      }, Ce.value = !0, Ve.value = t.id;
      let a = "";
      const o = t.isChild ? t.parentName : t.name, y = t.isChild ? t.name : "", x = e.value.style.chart.tooltip.customFormat;
      Ma(x) && Fa(
        () => x({
          datapoint: t,
          series: $.value,
          config: e.value,
          seriesIndex: l
        })
      ) ? we.value = x({
        datapoint: t,
        series: $.value,
        config: e.value,
        seriesIndex: l
      }) : (a += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;text-align:left;">
                <div style="display:flex;align-items:center;gap:4px;"><svg viewBox="0 0 60 60" height="14" width="14"><rect x="0" y="0" height="60" width="60" rx="5" stroke="none" fill="${t.color}"/>${Xt.pattern ? `<rect x="0" y="0" height="60" width="60" rx="5" stroke="none" fill="url(#pattern_${A.value}_${t.absoluteIndex})"/>` : ""}</svg> ${o}</div>
                ${y ? `<div>${y}</div>` : ""}
            </div>`, e.value.style.chart.tooltip.showValue && (a += `<div>${e.value.translations.value}: <b>${ve(
        e.value.style.chart.layout.bars.dataLabels.value.formatter,
        t.sign === 1 ? t.value : -t.value,
        j({
          p: e.value.style.chart.tooltip.prefix,
          v: t.sign === 1 ? t.value : -t.value,
          s: e.value.style.chart.tooltip.suffix,
          r: e.value.style.chart.tooltip.roundingValue
        }),
        { datapoint: t, seriesIndex: l }
      )}</b></div>`), e.value.style.chart.tooltip.showPercentage && (a += `<div>${e.value.translations.percentageToTotal} : <b>${j({
        v: Math.abs(t.value) / E.value * 100,
        s: "%",
        r: e.value.style.chart.tooltip.roundingPercentage
      })}</b></div>`, t.isChild && (a += `<div>${e.value.translations.percentageToSerie}: <b>${j({
        v: Math.abs(t.value) / Math.abs(t.parentValue) * 100,
        s: "%",
        r: e.value.style.chart.tooltip.roundingPercentage
      })}</b></div>`)), we.value = `<div style="text-align:left">${a}</div>`);
    }
    function gt(t, l, a, o) {
      if (!e.value.style.chart.layout.bars.dataLabels.value.show)
        return "";
      const y = ve(
        e.value.style.chart.layout.bars.dataLabels.value.formatter,
        P(o === -1 && t >= 0 ? -t : t),
        j({
          p: e.value.style.chart.layout.bars.dataLabels.value.prefix,
          v: P(o === -1 && t >= 0 ? -t : t),
          s: e.value.style.chart.layout.bars.dataLabels.value.suffix,
          r: e.value.style.chart.layout.bars.dataLabels.value.roundingValue
        }),
        { datapoint: l, seriesIndex: a }
      ), x = `(${pt(
        t,
        !0,
        e.value.style.chart.layout.bars.dataLabels.percentage.roundingPercentage
      )})`;
      return `${y}${e.value.style.chart.layout.bars.dataLabels.percentage.show ? ` ${x}` : ""}`;
    }
    const D = d(() => {
      const t = [
        e.value.translations.parentName,
        e.value.translations.value,
        e.value.translations.percentageToTotal,
        e.value.translations.childName,
        e.value.translations.value,
        e.value.translations.percentageToSerie,
        e.value.translations.percentageToTotal
      ], l = T.value.map((a) => a.isChild ? a.isFirstChild ? {
        color: a.color,
        parentName: a.parentName,
        parentValue: a.parentValue,
        percentageToTotal: a.parentValue / E.value,
        childName: a.name,
        childValue: a.sign === 1 ? a.value : -a.value,
        childPercentageToParent: Math.abs(a.value) / Math.abs(a.parentValue),
        childPercentageToTotal: Math.abs(a.value) / E.value
      } : {
        color: "",
        parentName: "",
        parentValue: "",
        percentageToTotal: "",
        childName: a.name,
        childValue: a.sign === 1 ? a.value : -a.value,
        childPercentageToParent: Math.abs(a.value) / Math.abs(a.parentValue),
        childPercentageToTotal: Math.abs(a.value) / E.value
      } : {
        color: a.color,
        parentName: a.name,
        parentValue: a.sign === 1 ? a.value : -a.value,
        percentageToTotal: Math.abs(a.value) / E.value,
        childName: "",
        childValue: "",
        childPercentageToParent: "",
        childPercentageToTotal: ""
      });
      return { head: t, body: l };
    });
    function Fe(t = null) {
      const l = [
        [e.value.style.chart.title.text],
        [e.value.style.chart.title.subtitle.text],
        [""]
      ], a = D.value.head, o = D.value.body.map((b) => [
        b.parentName,
        b.parentValue,
        b.percentageToTotal,
        b.childName,
        b.childValue,
        b.childPercentageToParent,
        b.childPercentageToTotal
      ]), y = l.concat([a]).concat(o), x = Ia(y);
      t ? t(x) : Va({
        csvContent: x,
        title: e.value.style.chart.title.text || "vue-ui-vertical-bar"
      });
    }
    const J = i(!1);
    function yt(t) {
      J.value = t, qe.value += 1;
    }
    function bt() {
      _.value.showTable = !_.value.showTable;
    }
    Re(() => {
      ["none", "asc", "desc"].includes(
        e.value.style.chart.layout.bars.sort
      ) || ie({
        componentName: "VueUiVerticalBar",
        type: "attributeWrongValue",
        property: "style.chart.layout.bars.sort",
        key: e.value.style.chart.layout.bars.sort
      }), U.value = Gt.value[e.value.style.chart.layout.bars.sort], _.value.sortDesc = U.value === 2, ce.value = U.value === 0;
    });
    function ia() {
      U.value += 1, U.value > 2 && (U.value = 0);
    }
    function mt() {
      ia(), _.value.sortDesc = U.value === 2, ce.value = U.value === 0;
    }
    function kt() {
      _.value.showTooltip = !_.value.showTooltip;
    }
    const ge = i(!1);
    function De() {
      ge.value = !ge.value;
    }
    async function va({ scale: t = 2 } = {}) {
      if (!B.value) return;
      const { width: l, height: a } = B.value.getBoundingClientRect(), o = l / a, { imageUri: y, base64: x } = await qa({
        domElement: B.value,
        base64: !0,
        img: !0,
        scale: t
      });
      return {
        imageUri: y,
        base64: x,
        title: e.value.style.chart.title.text,
        width: l,
        height: a,
        aspectRatio: o
      };
    }
    const ye = d(() => {
      const t = e.value.table.useDialog && !e.value.table.show, l = _.value.showTable;
      return {
        component: t ? Dt : Ja,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: J.value,
          fullscreenParent: B.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: Y.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    oe(
      () => _.value.showTable,
      (t) => {
        e.value.table.show || (t && e.value.table.useDialog && ue.value ? ue.value.open() : "close" in ue.value && ue.value.close());
      }
    );
    const { isResponsive: ca } = Ga(st, Kt);
    function Be() {
      _.value.showTable = !1, $e.value && $e.value.setTableIconState(!1);
    }
    const da = d(() => $.value.map((t, l) => ({
      ...t,
      shape: "square",
      display: `${t.name}: ${ve(
        e.value.style.chart.layout.bars.dataLabels.value.formatter,
        t.value,
        j({
          p: e.value.style.chart.legend.prefix,
          v: t.value,
          s: e.value.style.chart.legend.suffix,
          r: e.value.style.chart.legend.roundingValue
        }),
        { datapoint: t, seriesIndex: l }
      )}`
    }))), ha = d(() => e.value.style.chart.backgroundColor), pa = d(() => e.value.style.chart.legend), fa = d(() => e.value.style.chart.title), ga = d(() => da.value.map((t) => ({
      ...t,
      name: t.display
    }))), { exportSvg: ya, getSvg: ba } = Ua({
      svg: ze,
      title: fa,
      legend: pa,
      legendItems: ga,
      backgroundColor: ha
    });
    async function Ct({ isCb: t }) {
      he.value = !0, await me();
      try {
        if (t) {
          const { blob: l, url: a, text: o, dataUrl: y } = await ba();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: a, text: o, dataUrl: y }));
        } else
          await Promise.resolve(ya());
      } finally {
        he.value = !1;
      }
    }
    function ma(t) {
      if (t?.stage === "start") {
        de.value = !0;
        return;
      }
      if (t?.stage === "end") {
        de.value = !1;
        return;
      }
      nt();
    }
    function ka() {
      console.warn("[autoSize]: this legacy method can be safely removed as it has no impact.");
    }
    async function wt() {
      if (et("copyAlt", {
        config: e.value,
        dataset: R.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: R.value
      }));
    }
    return Nt({
      autoSize: ka,
      getData: oa,
      getImage: va,
      generatePdf: ot,
      generateCsv: Fe,
      generateImage: nt,
      generateSvg: Ct,
      hideSeries: ta,
      showSeries: ea,
      toggleTable: bt,
      toggleSort: mt,
      toggleTooltip: kt,
      toggleAnnotator: De,
      toggleFullscreen: yt,
      copyAlt: wt
    }), (t, l) => (n(), u("div", {
      class: Q(`vue-data-ui-component vue-ui-vertical-bar ${J.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"} ${e.value.responsive ? "vue-ui-vertical-bar-autosize" : ""}`),
      ref_key: "verticalBarChart",
      ref: B,
      id: `vue-ui-vertical-bar_${A.value}`,
      style: p(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor};height:100%`),
      onMouseenter: l[3] || (l[3] = () => r(at)(!0)),
      onMouseleave: l[4] || (l[4] = () => r(at)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (n(), H(r(Vt), {
        key: 0,
        svgRef: r(ze),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: ge.value,
        isCursorPointer: Y.value,
        onClose: De
      }, {
        "annotator-action-close": h(() => [
          f(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": h(({ color: a }) => [
          f(t.$slots, "annotator-action-color", O(I({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": h(({ mode: a }) => [
          f(t.$slots, "annotator-action-draw", O(I({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": h(({ disabled: a }) => [
          f(t.$slots, "annotator-action-undo", O(I({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": h(({ disabled: a }) => [
          f(t.$slots, "annotator-action-redo", O(I({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": h(({ disabled: a }) => [
          f(t.$slots, "annotator-action-delete", O(I({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : g("", !0),
      qt.value ? (n(), u("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Je,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : g("", !0),
      e.value.style.chart.title.text ? (n(), u("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Ke,
        style: "width:100%;background:transparent;padding-bottom:12px"
      }, [
        (n(), H(Ka, {
          key: `title_${Qe.value}`,
          config: {
            title: {
              cy: "vertical-bar-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "vertical-bar-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : g("", !0),
      e.value.userOptions.show && ke.value && (r(lt) || r(Pe)) ? (n(), H(r(Mt), {
        ref_key: "userOptionsRef",
        ref: $e,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        key: `user_options_${qe.value}`,
        isImaging: r(Oe),
        isPrinting: r(Ne),
        uid: A.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.chart.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasSort: e.value.userOptions.buttons.sort,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: J.value,
        isTooltip: _.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: B.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: ge.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: Y.value,
        onToggleFullscreen: yt,
        onGeneratePdf: r(ot),
        onGenerateCsv: Fe,
        onGenerateImage: ma,
        onGenerateSvg: Ct,
        onToggleTable: bt,
        onToggleSort: mt,
        onToggleTooltip: kt,
        onToggleAnnotator: De,
        onCopyAlt: wt,
        style: p({
          visibility: r(lt) ? r(Pe) ? "visible" : "hidden" : "visible"
        })
      }, be({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: h(({ isOpen: a, color: o }) => [
            f(t.$slots, "menuIcon", O(I({ isOpen: a, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: h(() => [
            f(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: h(() => [
            f(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: h(() => [
            f(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: h(() => [
            f(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: h(() => [
            f(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: h(() => [
            f(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionLabels ? {
          name: "optionLabels",
          fn: h(() => [
            f(t.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionSort ? {
          name: "optionSort",
          fn: h(() => [
            f(t.$slots, "optionSort", {}, void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: h(({ toggleFullscreen: a, isFullscreen: o }) => [
            f(t.$slots, "optionFullscreen", O(I({ toggleFullscreen: a, isFullscreen: o })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: h(({ toggleAnnotator: a, isAnnotator: o }) => [
            f(t.$slots, "optionAnnotator", O(I({ toggleAnnotator: a, isAnnotator: o })), void 0, !0)
          ]),
          key: "10"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: h(({ altCopy: a }) => [
            f(t.$slots, "optionAltCopy", O(I({ altCopy: a })), void 0, !0)
          ]),
          key: "11"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasSort", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : g("", !0),
      e.value.style.chart.legend.show && e.value.style.chart.legend.position === "top" ? (n(), u("div", {
        key: 4,
        ref_key: "chartLegend",
        ref: xe
      }, [
        (n(), H(Pt, {
          key: `legend_top_${_e.value}`,
          legendSet: $.value,
          config: vt.value,
          onClickMarker: l[0] || (l[0] = ({ legend: a }) => K(a.id))
        }, be({
          item: h(({ legend: a, index: o }) => [
            s("div", {
              onClick: (y) => K(a.id),
              style: p(`opacity:${w.value.includes(a.id) ? 0.5 : 1}`)
            }, m(a.name) + ": " + m(r(ve)(
              e.value.style.chart.layout.bars.dataLabels.value.formatter,
              a.value,
              r(j)({
                p: e.value.style.chart.legend.prefix,
                v: a.value,
                s: e.value.style.chart.legend.suffix,
                r: e.value.style.chart.legend.roundingValue
              }),
              { datapoint: a, seriesIndex: o }
            )), 13, tl)
          ]),
          legendToggle: h(() => [
            $.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !r(pe) ? (n(), H(zt, {
              key: 0,
              backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
              color: e.value.style.chart.legend.selectAllToggle.color,
              fontSize: e.value.style.chart.legend.fontSize,
              checked: w.value.length > 0,
              onToggle: dt
            }, null, 8, ["backgroundColor", "color", "fontSize", "checked"])) : g("", !0)
          ]),
          _: 2
        }, [
          t.$slots.pattern ? {
            name: "legend-pattern",
            fn: h(({ legend: a, index: o }) => [
              ne(St, {
                shape: a.shape,
                radius: 30,
                stroke: "none",
                plot: { x: 30, y: 30 },
                fill: `url(#pattern_${A.value}_${o})`
              }, null, 8, ["shape", "fill"])
            ]),
            key: "0"
          } : void 0
        ]), 1032, ["legendSet", "config"]))
      ], 512)) : g("", !0),
      (n(), u("svg", {
        ref_key: "svgRef",
        ref: ze,
        xmlns: r(Aa),
        class: Q({
          "vue-data-ui-fullscreen--on": J.value,
          "vue-data-ui-fulscreen--off": !J.value
        }),
        viewBox: `0 0 ${F.value} ${W.value}`,
        style: p(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.chart.color}`)
      }, [
        s("g", {
          ref_key: "G",
          ref: Ut,
          class: "vue-data-ui-g"
        }, [
          ne(r(Ft)),
          t.$slots["chart-background"] ? (n(), u("foreignObject", {
            key: 0,
            x: 0,
            y: 0,
            width: F.value,
            height: W.value,
            style: {
              pointerEvents: "none"
            }
          }, [
            f(t.$slots, "chart-background", {}, void 0, !0)
          ], 8, ll)) : g("", !0),
          (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("linearGradient", {
            x1: "0%",
            y1: "0%",
            x2: "100%",
            y2: "0%",
            id: `vertical_bar_gradient_${A.value}_${o}`
          }, [
            s("stop", {
              offset: "0%",
              "stop-color": a.color
            }, null, 8, nl),
            s("stop", {
              offset: "100%",
              "stop-color": r(We)(
                r(Na)(a.color, 0.03),
                100 - e.value.style.chart.layout.bars.gradientIntensity
              )
            }, null, 8, ul)
          ], 8, ol))), 256)),
          t.$slots.pattern ? (n(), u("g", sl, [
            (n(!0), u(V, null, M(T.value, (a) => (n(), u("defs", null, [
              f(t.$slots, "pattern", $t({ ref_for: !0 }, {
                seriesIndex: a.absoluteIndex,
                patternId: `pattern_${A.value}_${a.absoluteIndex}`
              }), void 0, !0)
            ]))), 256))
          ])) : g("", !0),
          e.value.style.chart.layout.bars.rowColor ? (n(!0), u(V, { key: 2 }, M(T.value, (a, o) => (n(), u("g", null, [
            s("rect", {
              x: 0,
              y: v.value.top + (C.value + c.value) * o + S.value[o] * L.value,
              width: F.value,
              height: c.value <= 0 ? 1e-4 : c.value,
              fill: e.value.style.chart.layout.bars.rowColor,
              rx: e.value.style.chart.layout.bars.rowRadius,
              style: { pointerEvents: "none" }
            }, null, 8, rl)
          ]))), 256)) : g("", !0),
          (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("g", null, [
            s("rect", {
              x: r(P)(
                Z.value ? v.value.left + v.value.width / 2 - (a.sign === 1 ? 0 : z(a.value) <= 0 ? 1e-4 : z(a.value)) : v.value.left
              ),
              y: v.value.top + (C.value + c.value) * o + S.value[o] * L.value,
              width: r(P)(
                z(a.value) <= 0 ? 1e-4 : z(a.value)
              ),
              height: c.value <= 0 ? 1e-4 : c.value,
              fill: e.value.style.chart.layout.bars.underlayerColor,
              rx: e.value.style.chart.layout.bars.borderRadius,
              class: Q({ animated: e.value.useCssAnimation })
            }, null, 10, il)
          ]))), 256)),
          (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("g", null, [
            s("rect", {
              x: r(P)(
                Z.value ? v.value.left + v.value.width / 2 - (a.sign === 1 ? 0 : z(a.value) <= 0 ? 1e-4 : z(a.value)) : v.value.left
              ),
              y: v.value.top + (C.value + c.value) * o + S.value[o] * L.value,
              width: r(P)(
                z(a.value) <= 0 ? 1e-4 : z(a.value)
              ),
              height: c.value <= 0 ? 1e-4 : c.value,
              fill: e.value.style.chart.layout.bars.useGradient ? `url(#vertical_bar_gradient_${A.value}_${o})` : r(We)(
                a.color,
                e.value.style.chart.layout.bars.fillOpacity
              ),
              rx: e.value.style.chart.layout.bars.borderRadius,
              stroke: e.value.style.chart.layout.bars.useStroke ? a.color : "none",
              "stroke-width": e.value.style.chart.layout.bars.useStroke ? e.value.style.chart.layout.bars.strokeWidth : 0,
              class: Q({ animated: e.value.useCssAnimation })
            }, null, 10, vl),
            t.$slots.pattern ? (n(), u("rect", {
              key: 0,
              x: r(P)(
                Z.value ? v.value.left + v.value.width / 2 - (a.sign === 1 ? 0 : z(a.value) <= 0 ? 1e-4 : z(a.value)) : v.value.left
              ),
              y: v.value.top + (C.value + c.value) * o + S.value[o] * L.value,
              width: r(P)(
                z(a.value) <= 0 ? 1e-4 : z(a.value)
              ),
              height: c.value <= 0 ? 1e-4 : c.value,
              fill: `url(#pattern_${A.value}_${a.absoluteIndex})`,
              rx: e.value.style.chart.layout.bars.borderRadius,
              stroke: e.value.style.chart.layout.bars.useStroke ? a.color : "none",
              "stroke-width": e.value.style.chart.layout.bars.useStroke ? e.value.style.chart.layout.bars.strokeWidth : 0,
              class: Q({ animated: e.value.useCssAnimation })
            }, null, 10, cl)) : g("", !0),
            (!a.isChild || a.isLastChild) && e.value.style.chart.layout.separators.show && o !== T.value.length - 1 ? (n(), u("line", {
              key: 1,
              x1: F.value,
              x2: e.value.style.chart.layout.separators.fullWidth ? 0 : v.value.left,
              y1: c.value + C.value / 2 + v.value.top + (C.value + c.value) * o + S.value[o] * L.value,
              y2: c.value + C.value / 2 + v.value.top + (C.value + c.value) * o + S.value[o] * L.value,
              stroke: e.value.style.chart.layout.separators.color,
              "stroke-width": e.value.style.chart.layout.separators.strokeWidth,
              "stroke-linecap": "round",
              style: {
                transition: "none !important",
                animation: "none !important"
              }
            }, null, 8, dl)) : g("", !0),
            Z.value && e.value.style.chart.layout.separators.show ? (n(), u("line", {
              key: 2,
              x1: v.value.left + v.value.width / 2,
              x2: v.value.left + v.value.width / 2,
              y1: v.value.top,
              y2: v.value.bottom,
              stroke: e.value.style.chart.layout.separators.color,
              "stroke-width": e.value.style.chart.layout.separators.strokeWidth,
              "stroke-linecap": "round"
            }, null, 8, hl)) : g("", !0)
          ]))), 256)),
          s("g", {
            ref_key: "parentLabels",
            ref: Wt
          }, [
            (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("g", pl, [
              a.isChild && a.childIndex === 0 && e.value.style.chart.layout.bars.parentLabels.show ? (n(), u("rect", {
                key: 0,
                class: "vue-ui-horizontal-bar-parent-marker",
                x: 2,
                y: ae(a, o).y - e.value.style.chart.layout.bars.parentLabels.fontSize / 1.5,
                height: e.value.style.chart.layout.bars.parentLabels.fontSize / 1.5,
                width: e.value.style.chart.layout.bars.parentLabels.fontSize / 1.5,
                rx: 1,
                fill: a.color,
                style: { animation: "none !important", transition: "none !important" }
              }, null, 8, fl)) : g("", !0),
              a.isChild && a.childIndex === 0 && e.value.style.chart.layout.bars.parentLabels.show ? (n(), u("text", {
                key: 1,
                y: ae(a, o).y,
                "font-size": e.value.style.chart.layout.bars.parentLabels.fontSize,
                fill: e.value.style.chart.layout.bars.parentLabels.color,
                x: e.value.style.chart.layout.bars.parentLabels.offsetX + e.value.style.chart.layout.bars.parentLabels.fontSize,
                "font-weight": e.value.style.chart.layout.bars.parentLabels.bold ? "bold" : "normal",
                "text-anchor": "start",
                innerHTML: r(Ge)({
                  content: `${ft(a, o)}`,
                  fontSize: e.value.style.chart.layout.bars.parentLabels.fontSize,
                  fill: e.value.style.chart.layout.bars.parentLabels.color,
                  x: e.value.style.chart.layout.bars.parentLabels.offsetX + e.value.style.chart.layout.bars.parentLabels.fontSize,
                  y: ae(a, o).y,
                  translateY: !1
                })
              }, null, 8, gl)) : g("", !0),
              a.isChild && a.childIndex === 0 && e.value.style.chart.layout.bars.parentLabels.show ? (n(), u("text", {
                key: 2,
                y: ae(a, o).y + r(Oa)(ft(a, o)) * e.value.style.chart.layout.bars.parentLabels.fontSize * 1.1,
                "font-size": e.value.style.chart.layout.bars.parentLabels.fontSize,
                fill: e.value.style.chart.layout.bars.parentLabels.color,
                x: e.value.style.chart.layout.bars.parentLabels.offsetX + e.value.style.chart.layout.bars.parentLabels.fontSize,
                "font-weight": e.value.style.chart.layout.bars.dataLabels.bold ? "bold" : "normal",
                "text-anchor": "start"
              }, m(na(a, o)), 9, yl)) : g("", !0)
            ]))), 256))
          ], 512),
          s("g", {
            ref_key: "childLabels",
            ref: Te
          }, [
            (n(!0), u(V, null, M(Qt.value, (a, o) => (n(), u("g", bl, [
              (a.isChild || !a.hasChildren) && e.value.style.chart.layout.bars.nameLabels.show ? (n(), u("text", {
                key: 0,
                "text-anchor": "start",
                x: Math.abs(e.value.style.chart.layout.bars.nameLabels.offsetX),
                y: r(P)(v.value.top + (C.value + c.value) * o + c.value / 2 + e.value.style.chart.layout.bars.nameLabels.fontSize / 2 + S.value[o] * L.value),
                "font-size": e.value.style.chart.layout.bars.nameLabels.fontSize,
                fill: "transparent",
                "font-weight": e.value.style.chart.layout.bars.nameLabels.bold ? "bold" : "normal",
                style: { "user-select": "none" },
                innerHTML: r(Ge)({
                  content: a.name,
                  fontSize: e.value.style.chart.layout.bars.nameLabels.fontSize,
                  fill: "transparent",
                  x: Math.abs(e.value.style.chart.layout.bars.nameLabels.offsetX),
                  y: r(P)(v.value.top + (C.value + c.value) * o + c.value / 2 + e.value.style.chart.layout.bars.nameLabels.fontSize / 2 + S.value[o] * L.value),
                  translateY: !0
                })
              }, null, 8, ml)) : g("", !0)
            ]))), 256))
          ], 512),
          s("g", null, [
            (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("g", kl, [
              (a.isChild || !a.hasChildren) && e.value.style.chart.layout.bars.nameLabels.show ? (n(), u("text", {
                key: 0,
                "text-anchor": "end",
                x: v.value.left + e.value.style.chart.layout.bars.nameLabels.offsetX - 6,
                y: v.value.top + (C.value + c.value) * o + c.value / 2 + e.value.style.chart.layout.bars.nameLabels.fontSize / 3 + S.value[o] * L.value,
                "font-size": e.value.style.chart.layout.bars.nameLabels.fontSize,
                fill: e.value.style.chart.layout.bars.nameLabels.color,
                "font-weight": e.value.style.chart.layout.bars.nameLabels.bold ? "bold" : "normal",
                innerHTML: r(Ge)({
                  content: a.name,
                  fontSize: e.value.style.chart.layout.bars.nameLabels.fontSize,
                  fill: e.value.style.chart.layout.bars.nameLabels.color,
                  x: v.value.left + e.value.style.chart.layout.bars.nameLabels.offsetX - 6,
                  y: v.value.top + (C.value + c.value) * o + c.value / 2 + e.value.style.chart.layout.bars.nameLabels.fontSize / 3 + S.value[o] * L.value,
                  translateY: !0
                })
              }, null, 8, Cl)) : g("", !0)
            ]))), 256))
          ]),
          s("g", {
            ref_key: "dataLabels",
            ref: Le
          }, [
            (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("g", wl, [
              s("text", {
                x: Z.value ? v.value.left + v.value.width / 2 + (a.sign === 1 ? -12 : 12) + (a.sign === 1 ? -e.value.style.chart.layout.bars.dataLabels.offsetX : e.value.style.chart.layout.bars.dataLabels.offsetX) : Math.min(
                  la(a.value) + 3 + e.value.style.chart.layout.bars.dataLabels.offsetX,
                  v.value.right - 2
                ),
                y: v.value.top + (C.value + c.value) * o + c.value / 2 + S.value[o] * L.value + e.value.style.chart.layout.bars.dataLabels.fontSize / 3,
                "text-anchor": !Z.value || a.sign === -1 ? "start" : "end",
                "font-size": e.value.style.chart.layout.bars.dataLabels.fontSize,
                fill: e.value.style.chart.layout.bars.dataLabels.color,
                "font-weight": e.value.style.chart.layout.bars.dataLabels.bold ? "bold" : "normal"
              }, m(gt(a.value, a, o, a.sign)), 9, xl)
            ]))), 256))
          ], 512),
          (n(!0), u(V, null, M(T.value, (a, o) => (n(), u("g", null, [
            s("rect", {
              x: 0,
              y: v.value.top + (C.value + c.value) * o - C.value / 2 + S.value[o] * L.value,
              width: F.value,
              height: c.value + C.value <= 0 ? 1e-4 : c.value + C.value,
              fill: Ve.value === a.id ? r(We)(
                e.value.style.chart.layout.highlighter.color,
                e.value.style.chart.layout.highlighter.opacity
              ) : "transparent",
              onMouseenter: (y) => ra(a, o),
              onMouseleave: (y) => sa({ datapoint: a, seriesIndex: o }),
              onClick: (y) => ua({ datapoint: a, seriesIndex: o })
            }, null, 40, _l)
          ]))), 256)),
          f(t.$slots, "svg", {
            svg: {
              ...X.value,
              isPrintingImg: r(Ne) | r(Oe) | de.value,
              isPrintingSvg: he.value
            }
          }, void 0, !0)
        ], 512)
      ], 14, al)),
      t.$slots.watermark ? (n(), u("div", $l, [
        f(t.$slots, "watermark", O(I({ isPrinting: r(Ne) || r(Oe) || de.value || he.value })), void 0, !0)
      ])) : g("", !0),
      e.value.style.chart.legend.show && e.value.style.chart.legend.position === "bottom" ? (n(), u("div", {
        key: 6,
        ref_key: "chartLegend",
        ref: xe
      }, [
        (n(), H(Pt, {
          key: `legend_bottom_${_e.value}`,
          legendSet: $.value,
          config: vt.value,
          isCursorPointer: Y.value,
          onClickMarker: l[1] || (l[1] = ({ legend: a }) => K(a.id))
        }, be({
          item: h(({ legend: a, index: o }) => [
            s("div", {
              onClick: (y) => K(a.id),
              style: p(`opacity:${w.value.includes(a.id) ? 0.5 : 1}`)
            }, m(a.name) + ": " + m(r(ve)(
              e.value.style.chart.layout.bars.dataLabels.value.formatter,
              a.value,
              r(j)({
                p: e.value.style.chart.legend.prefix,
                v: a.value,
                s: e.value.style.chart.legend.suffix,
                r: e.value.style.chart.legend.roundingValue
              }),
              { datapoint: a, seriesIndex: o }
            )), 13, Tl)
          ]),
          legendToggle: h(() => [
            $.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !r(pe) ? (n(), H(zt, {
              key: 0,
              backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
              color: e.value.style.chart.legend.selectAllToggle.color,
              fontSize: e.value.style.chart.legend.fontSize,
              checked: w.value.length > 0,
              isCursorPointer: Y.value,
              onToggle: dt
            }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : g("", !0)
          ]),
          _: 2
        }, [
          t.$slots.pattern ? {
            name: "legend-pattern",
            fn: h(({ legend: a, index: o }) => [
              ne(St, {
                shape: a.shape,
                radius: 30,
                stroke: "none",
                plot: { x: 30, y: 30 },
                fill: `url(#pattern_${A.value}_${o})`
              }, null, 8, ["shape", "fill"])
            ]),
            key: "0"
          } : void 0
        ]), 1032, ["legendSet", "config", "isCursorPointer"]))
      ], 512)) : g("", !0),
      f(t.$slots, "legend", { legend: $.value }, void 0, !0),
      t.$slots.source ? (n(), u("div", {
        key: 7,
        ref_key: "source",
        ref: Ze,
        dir: "auto"
      }, [
        f(t.$slots, "source", {}, void 0, !0)
      ], 512)) : g("", !0),
      ne(r(It), {
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: _.value.showTooltip && Ce.value && w.value.length < k.dataset.length,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        fontSize: e.value.style.chart.tooltip.fontSize,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: B.value,
        content: we.value,
        isFullscreen: J.value,
        isCustom: e.value.style.chart.tooltip.customFormat && typeof e.value.style.chart.tooltip.customFormat == "function",
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": h(() => [
          f(t.$slots, "tooltip-before", O(I({ ...Me.value })), void 0, !0)
        ]),
        "tooltip-after": h(() => [
          f(t.$slots, "tooltip-after", O(I({ ...Me.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      ke.value && e.value.userOptions.buttons.table ? (n(), H(Ta(ye.value.component), $t({ key: 8 }, ye.value.props, {
        ref_key: "tableUnit",
        ref: ue,
        onClose: Be
      }), be({
        content: h(() => [
          s("div", {
            ref_key: "tableContainer",
            ref: st,
            class: "vue-ui-vertical-bar-table atom-data-table",
            style: p(`${e.value.table.useDialog ? "" : "max-height: 300px; margin-top: 24px"}`)
          }, [
            s("div", Ll, [
              e.value.table.useDialog ? g("", !0) : (n(), u("div", {
                key: 0,
                role: "button",
                tabindex: "0",
                style: p(`width:32px; position: absolute; top: 0; right:4px; padding: 0 0px; display: flex; align-items:center;justify-content:center;height: 36px; width: 32px; cursor:${Y.value ? "pointer" : "default"}; background:${e.value.table.th.backgroundColor};`),
                onClick: Be,
                onKeypress: La(Be, ["enter"])
              }, [
                ne(r(Ye), {
                  name: "close",
                  stroke: e.value.table.th.color,
                  "stroke-width": 2
                }, null, 8, ["stroke"])
              ], 36)),
              s("div", {
                style: { width: "100%", "container-type": "inline-size" },
                class: Q({ "vue-ui-responsive": r(ca) })
              }, [
                s("table", Sl, [
                  e.value.table.useDialog ? g("", !0) : (n(), u("caption", {
                    key: 0,
                    style: p({
                      backgroundColor: e.value.table.th.backgroundColor,
                      color: e.value.table.th.color,
                      outline: e.value.table.th.outline
                    }),
                    class: "vue-ui-data-table__caption"
                  }, [
                    Tt(m(e.value.style.chart.title.text) + " ", 1),
                    e.value.style.chart.title.subtitle.text ? (n(), u("span", Pl, m(e.value.style.chart.title.subtitle.text), 1)) : g("", !0)
                  ], 4)),
                  s("thead", null, [
                    s("tr", {
                      role: "row",
                      style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color}`)
                    }, [
                      (n(!0), u(V, null, M(D.value.head, (a) => (n(), u("th", {
                        style: p(`outline:${e.value.table.th.outline}`)
                      }, [
                        s("div", zl, m(a), 1)
                      ], 4))), 256))
                    ], 4),
                    s("tr", null, [
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline}`)
                      }, null, 4),
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline};text-align:right;padding-right:5px;font-weight:bold`)
                      }, " ∑ " + m(e.value.table.td.prefix) + m(isNaN(E.value) ? "" : E.value.toFixed(e.value.table.td.roundingValue)) + m(e.value.table.td.suffix), 5),
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline};text-align:right;padding-right:5px;font-weight:bold`)
                      }, " 100% ", 4),
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline}`)
                      }, null, 4),
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline}`)
                      }, null, 4),
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline}`)
                      }, null, 4),
                      s("th", {
                        style: p(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline}`)
                      }, null, 4)
                    ])
                  ]),
                  s("tbody", null, [
                    (n(!0), u(V, null, M(D.value.body, (a, o) => (n(), u("tr", {
                      class: Q({
                        "vue-ui-data-table__tbody__row": !0,
                        "vue-ui-data-table__tbody__row-even": o % 2 === 0,
                        "vue-ui-data-table__tbody__row-odd": o % 2 !== 0
                      }),
                      style: p(`background:${e.value.table.td.backgroundColor};color:${e.value.table.td.color}`)
                    }, [
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline};font-variant-numeric: tabular-nums;`),
                        "data-cell": D.value.head[0] ?? ""
                      }, [
                        s("div", Nl, [
                          a.color ? (n(), u("span", {
                            key: 0,
                            style: p(`color:${a.color};margin-right:3px`)
                          }, "⬤", 4)) : g("", !0),
                          s("span", null, m(a.parentName), 1)
                        ])
                      ], 12, Al),
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline}`),
                        "data-cell": D.value.head[1] ?? ""
                      }, [
                        s("div", Il, m(e.value.table.td.prefix) + m(["", NaN, void 0].includes(a.parentValue) ? "" : a.parentValue.toFixed(
                          e.value.table.td.roundingValue
                        )) + m(e.value.table.td.suffix), 1)
                      ], 12, Ol),
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline}`),
                        "data-cell": D.value.head[2] ?? ""
                      }, [
                        s("div", Ml, m(["", NaN, void 0].includes(a.percentageToTotal) ? "" : `${(a.percentageToTotal * 100).toFixed(
                          e.value.table.td.roundingPercentage
                        )}%`), 1)
                      ], 12, Vl),
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline}`),
                        "data-cell": D.value.head[3] ?? ""
                      }, [
                        s("div", Dl, m(a.childName), 1)
                      ], 12, Fl),
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline}`),
                        "data-cell": D.value.head[4] ?? ""
                      }, [
                        s("div", El, m(e.value.table.td.prefix) + m(["", NaN, void 0].includes(a.childValue) ? "" : a.childValue.toFixed(
                          e.value.table.td.roundingValue
                        )) + m(e.value.table.td.suffix), 1)
                      ], 12, Bl),
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline}`),
                        "data-cell": D.value.head[5] ?? ""
                      }, [
                        s("div", Xl, m(["", NaN, void 0].includes(
                          a.childPercentageToParent
                        ) ? "" : `${(a.childPercentageToParent * 100).toFixed(
                          e.value.table.td.roundingPercentage
                        )}%`), 1)
                      ], 12, Rl),
                      s("td", {
                        class: "vue-ui-data-table__tbody__td",
                        style: p(`outline:${e.value.table.td.outline}`),
                        "data-cell": D.value.head[6] ?? ""
                      }, [
                        s("div", Ul, m(["", NaN, void 0].includes(
                          a.childPercentageToTotal
                        ) ? "" : `${(a.childPercentageToTotal * 100).toFixed(
                          e.value.table.td.roundingPercentage
                        )}%`), 1)
                      ], 12, Hl)
                    ], 6))), 256))
                  ])
                ])
              ], 2)
            ])
          ], 4)
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: h(() => [
            Tt(m(ye.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: h(() => [
            s("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[2] || (l[2] = (a) => Fe(e.value.userOptions.callbacks.csv)),
              style: p({ cursor: Y.value ? "pointer" : "default" })
            }, [
              ne(r(Ye), {
                name: "fileCsv",
                stroke: ye.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : g("", !0),
      f(t.$slots, "skeleton", {}, () => [
        r(pe) ? (n(), H(Xa, { key: 0 })) : g("", !0)
      ], !0)
    ], 46, el));
  }
}, ho = /* @__PURE__ */ Qa(Wl, [["__scopeId", "data-v-26efb173"]]);
export {
  ho as default
};
