// vue-data-ui/dist/vue-ui-sparkbar-DUDWvTo-.js
import { defineAsyncComponent as ae, useSlots as oe, onMounted as P, ref as k, computed as x, toRefs as ne, watch as B, nextTick as se, openBlock as s, createElementBlock as i, normalizeStyle as d, unref as o, renderSlot as $, normalizeProps as ue, mergeProps as E, createCommentVNode as f, createElementVNode as v, toDisplayString as g, Fragment as V, renderList as ie, createVNode as re, createBlock as ve } from "vue";
import { j as ce, t as M, m as de, p as w, n as pe, o as fe, q as O, r as ge, s as me, u as D, k as _, l as A, X as ye, v as he } from "./lib-GwPnJSrY.js";
import { u as ke } from "./useConfig-CIgyKZ6R.js";
import { u as xe, B as be } from "./BaseScanner-BhhkKVoj.js";
import { u as I } from "./useNestedProp-BBT0mbHh.js";
import { u as Ce } from "./useThemeCheck-D1ZYXUwG.js";
import we from "./vue_ui_sparkbar-BsMzz5Um.js";
const _e = ["onClick", "onMouseenter", "onMouseleave"], Ae = {
  key: 1,
  class: "vue-ui-sparkbar-datapoint-value"
}, Se = ["xmlns", "viewBox"], $e = ["id"], De = ["stop-color"], Ie = ["stop-color"], Te = ["height", "width", "fill", "rx"], Le = ["height", "width", "fill", "rx"], ze = ["height", "width", "fill", "rx"], Ne = {
  key: 2,
  ref: "source",
  dir: "auto"
}, je = {
  __name: "vue-ui-sparkbar",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    /**
     * Used in VueUiRadar's tooltip exclusively
     */
    backgroundOpacity: {
      type: Number,
      default: null
    }
  },
  emits: ["selectDatapoint"],
  setup(j, { emit: U }) {
    const R = ae(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_sparkbar: W } = ke(), { isThemeValid: q, warnInvalidTheme: G } = Ce(), p = j, H = oe();
    P(() => {
      H["chart-background"] && console.warn("VueUiSparkbar does not support the #chart-background slot.");
    });
    const T = k(ce()), e = k(S()), X = x(() => M({
      defaultConfig: {
        style: {
          backgroundColor: "#99999930",
          animation: { show: !1 },
          layout: { independant: !0 },
          gutter: {
            backgroundColor: "#6A6A6A",
            opacity: 50
          },
          bar: {
            gradient: {
              underlayerColor: "#6A6A6A"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: b, FINAL_DATASET: c } = xe({
      ...ne(p),
      FINAL_CONFIG: e,
      prepareConfig: S,
      skeletonDataset: p.config?.skeletonDataset ?? [
        { name: "_", value: 21, target: 25, color: "#808080" },
        { name: "_", value: 13, target: 25, color: "#ADADAD" },
        { name: "_", value: 8, target: 25, color: "#DBDBDB" }
      ],
      skeletonConfig: M({
        defaultConfig: e.value,
        userConfig: X.value
      })
    }), L = x(() => !!e.value.debug);
    function S() {
      const t = I({
        userConfig: p.config,
        defaultConfig: W
      }), a = t.theme;
      if (!a) return t;
      if (!q.value(t))
        return G(t), t;
      const l = I({
        userConfig: we[a] || p.config,
        defaultConfig: t
      }), n = I({
        userConfig: p.config,
        defaultConfig: l
      });
      return {
        ...n,
        customPalette: n.customPalette.length ? n.customPalette : de[a] || w
      };
    }
    B(() => p.config, (t) => {
      e.value = S();
    }, { deep: !0 });
    const J = x(() => pe(e.value.customPalette)), m = k(c.value.map((t) => ({
      ...t,
      value: e.value.style.animation.show ? 0 : t.value || 0,
      formatter: t.formatter || null
    }))), z = k(null);
    P(async () => {
      fe(p.dataset) && O({
        componentName: "VueUiSparkbar",
        type: "dataset",
        debug: L.value
      }), N();
    });
    function N() {
      if (e.value.style.animation.show) {
        let h = function() {
          n += l / t, n < l ? (m.value = m.value.map((u, C) => ({
            ...u,
            value: u.value += a[C]
          })), z.value = requestAnimationFrame(h)) : m.value = c.value.map((u) => ({
            ...u,
            value: u.value || 0,
            formatter: u.formatter || null
          }));
        };
        const t = e.value.style.animation.animationFrames, a = c.value.map((u, C) => u.value / t), l = c.value.map((u) => u.value || 0).reduce((u, C) => u + C, 0);
        let n = 0;
        h();
      }
    }
    B(() => c.value, async (t) => {
      cancelAnimationFrame(z.value), m.value = c.value.map((a) => ({
        ...a,
        value: e.value.style.animation.show ? 0 : a.value || 0,
        formatter: a.formatter || null
      })), se(N);
    }, { deep: !0 });
    const r = k({
      width: 500,
      height: 16
    }), K = x(() => Math.max(...c.value.map((t) => t.value))), Q = x(() => (c.value.forEach((t, a) => {
      ge({
        datasetObject: t,
        requiredAttributes: ["name", "value"]
      }).forEach((l) => {
        O({
          componentName: "VueUiSparkbar",
          type: "datasetSerieAttribute",
          property: l,
          index: a,
          debug: L.value
        });
      });
    }), m.value.map((t, a) => ({
      ...t,
      value: t.value || 0,
      color: me(t.color) || J.value[a] || w[a] || w[a % w.length]
    }))));
    function Y(t) {
      return t / K.value;
    }
    function F(t) {
      return e.value.style.layout.independant ? t.target ? t.value / t.target : e.value.style.layout.percentage ? t.value > 100 ? 1 : t.value / 100 : e.value.style.layout.target === 0 ? 1 : t.value / e.value.style.layout.target : Y(t.value);
    }
    function y(t) {
      return e.value.style.layout.independant && t.target || e.value.style.layout.target;
    }
    const Z = U;
    function ee(t, a) {
      Z("selectDatapoint", { datapoint: t, index: a }), e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: a });
    }
    function te(t, a) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: a });
    }
    function le(t, a) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: a });
    }
    return (t, a) => (s(), i("div", {
      class: "vue-data-ui-component vue-ui-sparkbar",
      style: d({
        width: "100%",
        position: "relative",
        fontFamily: e.value.style.fontFamily,
        background: p.backgroundOpacity !== null ? o(D)(e.value.style.backgroundColor, p.backgroundOpacity) : e.value.style.backgroundColor
      })
    }, [
      t.$slots.title ? $(t.$slots, "title", ue(E({ key: 0 }, { title: { ...t.title, title: e.value.style.title.text, subtitle: e.value.style.title.subtitle.text } }))) : f("", !0),
      !t.$slots.title && e.value.style.title.text ? (s(), i("div", {
        key: 1,
        class: "vue-ui-sparkbar-title-container",
        style: d({
          background: e.value.style.title.backgroundColor,
          margin: e.value.style.title.margin,
          textAlign: e.value.style.title.textAlign
        })
      }, [
        v("div", {
          class: "vue-ui-sparkbar-title",
          style: d({
            fontSize: e.value.style.title.fontSize + "px",
            color: e.value.style.title.color,
            fontWeight: e.value.style.title.bold ? "bold" : "normal"
          })
        }, g(e.value.style.title.text), 5),
        e.value.style.title.subtitle.text ? (s(), i("div", {
          key: 0,
          class: "vue-ui-sparkbar-subtitle",
          style: d({
            fontSize: e.value.style.title.subtitle.fontSize + "px",
            color: e.value.style.title.subtitle.color,
            fontWeight: e.value.style.title.subtitle.bold ? "bold" : "normal"
          })
        }, g(e.value.style.title.subtitle.text), 5)) : f("", !0)
      ], 4)) : f("", !0),
      (s(!0), i(V, null, ie(Q.value, (l, n) => (s(), i("div", {
        style: d(`display:flex !important;${["left", "right"].includes(e.value.style.labels.name.position) ? `flex-direction: ${e.value.style.labels.name.position === "right" ? "row-reverse" : "row"} !important` : "flex-direction:column !important"};gap:${e.value.style.gap}px !important;align-items:center;${o(c).length > 0 && n !== o(c).length - 1 ? "margin-bottom:6px" : ""}`),
        onClick: (h) => ee(l, n),
        onMouseenter: (h) => te(l, n),
        onMouseleave: (h) => le(l, n)
      }, [
        o(b) ? f("", !0) : $(t.$slots, "data-label", E({
          key: 0,
          ref_for: !0
        }, { bar: {
          ...l,
          target: y(l),
          valueLabel: o(_)(
            l.formatter,
            l.value,
            o(A)({
              p: l.prefix || "",
              v: l.value,
              s: l.suffix || "",
              r: l.rounding || 0
            }),
            { datapoint: l, seriesIndex: n }
          ),
          targetLabel: o(_)(
            l.formatter,
            y(l),
            o(A)({
              p: l.prefix || "",
              v: y(l),
              s: l.suffix || "",
              r: l.rounding || 0
            }),
            { datapoint: l, seriesIndex: n }
          )
        } })),
        !t.$slots["data-label"] || o(b) ? (s(), i("div", {
          key: 1,
          style: d({
            display: "flex",
            justifyContent: ["right", "top-right"].includes(e.value.style.labels.name.position) ? "flex-end" : ["top-center"].includes(e.value.style.labels.name.position) ? "center" : "flex-start",
            alignItems: "center",
            width: e.value.style.labels.name.width,
            color: e.value.style.labels.name.color,
            fontSize: e.value.style.labels.fontSize + "px",
            fontWeight: e.value.style.labels.name.bold ? "bold" : "normal",
            flexWrap: "wrap"
          })
        }, [
          o(b) ? (s(), i("div", {
            key: 0,
            class: "vue-ui-sparkbar-skeleton-name",
            style: d({
              width: "60px",
              height: e.value.style.labels.fontSize + "px",
              borderRadius: e.value.style.labels.fontSize / 4 + "px",
              display: "flex",
              alignItems: "center",
              justifyContent: "space-between",
              marginTop: "5px"
            })
          }, [
            v("div", {
              style: d({ height: "100%", width: "40px", borderRadius: e.value.style.labels.fontSize / 4 + "px", backgroundColor: "#6A6A6A80" })
            }, null, 4),
            v("div", {
              style: d({ height: "100%", width: "15px", borderRadius: e.value.style.labels.fontSize / 4 + "px", backgroundColor: "#6A6A6A80" })
            }, null, 4)
          ], 4)) : (s(), i(V, { key: 1 }, [
            v("span", null, g(l.name), 1),
            e.value.style.labels.value.show ? (s(), i("span", {
              key: 0,
              style: d(`font-weight:${e.value.style.labels.value.bold ? "bold" : "normal"}`),
              class: "vue-ui-sparkbar-datapoint-name"
            }, ": " + g(o(_)(
              l.formatter,
              l.value,
              o(A)({
                p: l.prefix || "",
                v: l.value,
                s: l.suffix || "",
                r: l.rounding || 0
              }),
              { datapoint: l, seriesIndex: n }
            )), 5)) : f("", !0),
            e.value.style.layout.showTargetValue ? (s(), i("span", Ae, g(" " + e.value.style.layout.targetValueText) + " " + g(o(_)(
              l.formatter,
              y(l),
              o(A)({
                p: l.prefix || "",
                v: y(l),
                s: l.suffix || "",
                r: l.rounding || 0
              }),
              { datapoint: l, seriesIndex: n }
            )), 1)) : f("", !0)
          ], 64))
        ], 4)) : f("", !0),
        (s(), i("svg", {
          xmlns: o(ye),
          viewBox: `0 0 ${r.value.width} ${r.value.height}`,
          width: "100%"
        }, [
          re(o(R)),
          v("defs", null, [
            v("linearGradient", {
              x1: "0%",
              y1: "0%",
              x2: "100%",
              y2: "0%",
              id: `sparkbar_gradient_${n}_${T.value}`
            }, [
              v("stop", {
                offset: "0%",
                "stop-color": o(D)(o(he)(l.color, 0.03), 100 - e.value.style.bar.gradient.intensity)
              }, null, 8, De),
              v("stop", {
                offset: "100%",
                "stop-color": l.color
              }, null, 8, Ie)
            ], 8, $e)
          ]),
          v("rect", {
            height: r.value.height,
            width: r.value.width,
            x: 0,
            y: 0,
            fill: o(D)(e.value.style.gutter.backgroundColor, e.value.style.gutter.opacity),
            rx: r.value.height / 2
          }, null, 8, Te),
          v("rect", {
            height: r.value.height,
            width: r.value.width * F(l),
            x: 0,
            y: 0,
            fill: e.value.style.bar.gradient.underlayerColor,
            rx: r.value.height / 2
          }, null, 8, Le),
          v("rect", {
            height: r.value.height,
            width: r.value.width * F(l),
            x: 0,
            y: 0,
            fill: e.value.style.bar.gradient.show ? `url(#sparkbar_gradient_${n}_${T.value})` : l.color,
            rx: r.value.height / 2
          }, null, 8, ze)
        ], 8, Se))
      ], 44, _e))), 256)),
      t.$slots.source ? (s(), i("div", Ne, [
        $(t.$slots, "source")
      ], 512)) : f("", !0),
      o(b) ? (s(), ve(be, { key: 3 })) : f("", !0)
    ], 4));
  }
};
export {
  je as default
};
