// vue-data-ui/dist/vue-ui-flow-CsbSaFnu.js
import { defineAsyncComponent as U, ref as u, computed as d, onMounted as Mo, toRefs as Do, watch as _e, openBlock as c, createElementBlock as y, unref as r, normalizeStyle as R, normalizeClass as jt, createBlock as Q, withCtx as f, renderSlot as h, normalizeProps as E, guardReactiveProps as I, createCommentVNode as T, createElementVNode as G, createSlots as Xt, createVNode as Ee, Fragment as pe, renderList as ge, withKeys as lt, withModifiers as at, toDisplayString as me, Teleport as Bo, resolveDynamicComponent as Lo, mergeProps as zo, createTextVNode as qt, nextTick as nt } from "vue";
import { j as Ie, t as Kt, o as Zt, q as Ro, m as Yo, p as Y, n as Ho, K as x, s as Wo, l as ae, X as Uo, k as Me, W as Jt, a as Go, a0 as Vo, a1 as jo, T as Xo } from "./lib-GwPnJSrY.js";
import { t as qo, u as Ko } from "./useResponsive-DfdjqQps.js";
import { u as Zo } from "./useConfig-CIgyKZ6R.js";
import { u as Jo } from "./usePrinter-DiK5R6k-.js";
import { u as Qo, B as el } from "./BaseScanner-BhhkKVoj.js";
import { u as tl } from "./useSvgExport-ByUukOZt.js";
import { u as st } from "./useNestedProp-BBT0mbHh.js";
import { u as ol } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ll } from "./useUserOptionState-BIvW1Kz7.js";
import { u as al } from "./useChartAccessibility-9icAAmYg.js";
import nl from "./vue_ui_flow-BErINgox.js";
import sl from "./Legend-1LMMsPxL.js";
import rl from "./Title-BfyrP-jp.js";
import ul from "./img-BYYzY_QU.js";
import { _ as il } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const cl = ["id"], vl = {
  key: 1,
  ref: "noTitle",
  class: "vue-data-ui-no-title-space",
  style: "height:36px; width: 100%;background:transparent"
}, dl = ["id"], fl = ["xmlns", "viewBox"], hl = ["width", "height"], pl = ["id"], gl = ["stop-color"], ml = ["stop-color"], yl = ["d", "fill", "stroke", "stroke-width"], bl = ["x", "y", "height", "width", "fill", "stroke", "stroke-width", "rx", "aria-label", "onMouseenter", "onMouseleave", "onClick", "onKeydown", "onFocus", "onBlur"], Cl = { key: 1 }, kl = ["x", "y", "font-size", "fill"], wl = ["x", "y", "font-size", "fill"], $l = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Al = ["id"], Tl = ["onClick"], Sl = ["innerHTML"], xl = {
  __name: "vue-ui-flow",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["copyAlt"],
  setup(Qt, { expose: eo, emit: to }) {
    const oo = U(() => import("./Tooltip-DL3QgjZf.js")), lo = U(() => import("./BaseIcon-CNx8Tzl5.js")), ao = U(() => import("./vue-ui-accordion-BGHTqIBt.js")), no = U(() => import("./DataTable-Bo1pmAqG.js")), so = U(() => import("./PenAndPaper-DWV1FMIi.js")), ro = U(() => import("./UserOptions-CrLqEimB.js")), uo = U(() => import("./PackageVersion-DN9d0ME5.js")), io = U(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_flow: co } = Zo(), { isThemeValid: vo, warnInvalidTheme: fo } = ol(), S = Qt, ho = to, V = u(Ie()), M = u(null), rt = u(0), ut = u(0), De = u(!1), Be = u(""), it = u(null), ct = u(null), vt = u(null), ne = u(null), ye = u(null), dt = u(!1), se = u(null), Le = u(null), ze = u(null), be = u(!1), Ce = u(!1), ft = d(() => !!S.dataset && S.dataset.length);
    Mo(() => {
      dt.value = !0, pt();
    });
    const j = u(!1);
    function ht(o) {
      j.value = o, rt.value += 1;
    }
    const e = u(Ue()), ee = d(() => e.value.userOptions.useCursorPointer), po = d(() => Kt({
      defaultConfig: {
        userOptions: { show: !1 },
        nodeCategories: {
          B: "A",
          C: "B"
        },
        nodeCategoryColors: {
          A: "#CACACA",
          B: "#AAAAAA"
        },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            legend: {
              backgroundColor: "transparent"
            },
            nodes: {
              labels: {
                show: !1
              },
              stroke: "#666666"
            },
            links: {
              stroke: "#666666"
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: Re, FINAL_DATASET: ke, manualLoading: Ye } = Qo({
      ...Do(S),
      FINAL_CONFIG: e,
      prepareConfig: Ue,
      skeletonDataset: S.config?.skeletonDataset ?? [
        ["A", "B", 2, "#CACACA"],
        ["B", "C", 1, "#CACACA"],
        ["C", "D", 0.5, "#CACACA"],
        ["E", "F", 1, "#AAAAAA"],
        ["F", "G", 0.5, "#AAAAAA"],
        ["G", "H", 0.25, "#AAAAAA"]
      ],
      skeletonConfig: Kt({
        defaultConfig: e.value,
        userConfig: po.value
      })
    }), go = d(() => !!e.value.debug);
    function pt() {
      if (Zt(S.dataset) && (Ro({
        componentName: "VueUiFlow",
        type: "dataset",
        debug: go.value
      }), Ye.value = !0), Zt(S.dataset) || (Ye.value = e.value.loading), e.value.responsive) {
        const o = qo(() => {
          const { width: l, height: t } = Ko({
            chart: M.value,
            title: e.value.style.chart.title.text ? ct.value : null,
            legend: e.value.style.chart.legend.show ? it.value : null,
            source: vt.value
          });
          requestAnimationFrame(() => {
            yt.value = l, bt.value = t;
          });
        });
        ne.value && (ye.value && ne.value.unobserve(ye.value), ne.value.disconnect()), ne.value = new ResizeObserver(o), ye.value = M.value.parentNode, ne.value.observe(ye.value);
      }
    }
    const { userOptionsVisible: He, setUserOptionsVisibility: gt, keepUserOptionState: mt } = ll({ config: e.value }), { svgRef: We } = al({
      config: e.value.style.chart.title
    });
    function Ue() {
      const o = st({
        userConfig: S.config,
        defaultConfig: co
      });
      let l = o;
      const t = o.theme;
      if (t)
        if (!vo.value(o))
          fo(o), l = o;
        else {
          const a = st({
            userConfig: nl[t] || S.config,
            defaultConfig: o
          });
          l = {
            ...st({
              userConfig: S.config,
              defaultConfig: a
            }),
            customPalette: o.customPalette.length ? o.customPalette : Yo[t] || Y
          };
        }
      else
        l = o;
      return l.nodeCategories = S.config.nodeCategories || {}, l.nodeCategoryColors = S.config.nodeCategoryColors || {}, l;
    }
    const yt = u(e.value.style.chart.width), bt = u(e.value.style.chart.height);
    _e(
      () => S.config,
      (o) => {
        Re.value || (e.value = Ue()), He.value = !e.value.userOptions.showOnChartHover, pt(), ut.value += 1, O.value.showTable = e.value.table.show;
      },
      { deep: !0 }
    ), _e(() => S.dataset, (o) => {
      Array.isArray(o) && o.length > 0 && (Ye.value = !1);
    }, { deep: !0 });
    const { isPrinting: Ge, isImaging: Ve, generatePdf: Ct, generateImage: kt } = Jo({
      elementId: `flow_${V.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-flow",
      options: e.value.userOptions.print
    }), mo = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), we = d(() => Ho(e.value.customPalette)), $e = d(() => e.value.style.chart.nodes.width), O = u({
      showTable: e.value.table.show,
      showTooltip: e.value.style.chart.tooltip.show
    });
    _e(e, () => {
      O.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.chart.tooltip.show
      };
    }, { immediate: !0 });
    const wt = d(() => !ke.value || !ke.value.length ? [] : ke.value.map((o, l) => [
      o[0],
      o[1],
      x(o[2]),
      o[3] ? Wo(o[3]) : we.value[l] || we.value[l % we.value.length] || Y[l] || Y[l % Y.length]
    ]));
    function yo(o) {
      const l = {}, t = {};
      function a(n, i) {
        l[n] || (l[n] = {
          level: null,
          inflow: 0,
          outflow: 0,
          children: [],
          color: null,
          uid: Ie()
        }), l[n].level === null && (l[n].level = i), t[i] || (t[i] = []), t[i].includes(n) || t[i].push(n);
      }
      o.forEach(([n, i, p]) => {
        const g = l[n]?.level ?? 0, D = g + 1;
        a(n, g), a(i, D), l[n].children.push({ target: i, value: p }), l[n].outflow += p, l[i].inflow += p;
      });
      const s = new Set(o.map(([n]) => n)), A = new Set(o.map(([, n]) => n)), oe = Array.from(s).filter((n) => !A.has(n)), ve = {};
      oe.forEach((n, i) => {
        ve[n] = we.value[i] || Y[i % Y.length];
      });
      const de = {};
      o.forEach(([n, i, p, g]) => {
        g && (de[n] = g, de[i] = g);
      }), Object.keys(l).forEach((n, i) => {
        const p = e.value.nodeCategories?.[n], g = p ? e.value.nodeCategoryColors?.[p] : null;
        l[n].color = de[n] || g || (oe.includes(n) ? ve[n] : null) || Y[i % Y.length];
      }), Object.keys(l).forEach((n) => {
        l[n].value = Math.max(l[n].inflow, l[n].outflow);
      });
      const Pe = re.value, Je = Tt.value.width, le = Tt.value.height, L = Object.keys(t).map(Number).sort((n, i) => n - i), Ne = L.length || 1, Qe = Ne > 1 ? Je / (Ne - 1) : 0, P = Number($e.value), B = Number(
        e.value.style.chart.nodes.gapPx ?? e.value.style.chart.nodes.gap ?? 8
      ), fe = Number(
        e.value.style.chart.nodes.minHeight || 0
      );
      function v(n) {
        const i = t[n], p = i.length;
        if (!p) return 1 / 0;
        const g = Math.max(0, (p - 1) * B), D = Math.max(0, le - g), H = Math.min(fe, p ? D / p : 0);
        let C = 0, N = i.map((m) => l[m].value || 0), k = N.reduce((m, F) => m + F, 0);
        for (let m = 0; m < 12; m += 1) {
          const F = k > 0 ? (D - C) / k : 0, W = [];
          for (let _ = 0; _ < N.length; _ += 1) {
            const z = N[_];
            z < 0 || z * F < H && W.push(_);
          }
          if (!W.length) return Math.max(0, F);
          for (const _ of W)
            C += H, k -= N[_], N[_] = -1;
          if (k <= 0) return 0;
        }
        return k > 0 ? Math.max(0, (D - C) / k) : 0;
      }
      const b = L.map(v), q = b.length ? Math.min(...b) : 0, K = {};
      L.forEach((n) => {
        const i = t[n], p = i.length, g = Math.max(0, (p - 1) * B), D = Math.max(0, le - g), H = Math.min(fe, p ? D / p : 0), C = i.map(
          (m) => Math.max(H, (l[m].value || 0) * q)
        ), N = C.reduce((m, F) => m + F, 0) + g;
        let k = Math.max(0, (le - N) / 2);
        i.forEach((m, F) => {
          const W = C[F], _ = Pe.left + n * Qe, z = k;
          K[m] = {
            x: _,
            y: z,
            absoluteY: z,
            height: W,
            i: F,
            color: l[m].color,
            value: l[m].value,
            id: Ie()
          }, k += W, F < p - 1 && (k += B);
        });
      });
      const It = [], he = Pe.top, et = {}, tt = {};
      Object.keys(l).forEach((n) => {
        et[n] = K[n]?.y ?? 0, tt[n] = 0;
      });
      const Fo = 1e-6, _o = 0.25, Eo = !!e.value.style.chart.links.smooth, Mt = 0.5;
      return L.forEach((n) => {
        t[n].forEach((p) => {
          const g = l[p], D = K[p];
          if (!g.children || !g.children.length) return;
          let H = D.y;
          g.children.forEach(({ target: C, value: N }) => {
            const k = K[C], m = l[C], F = g.outflow > 0 ? N / g.outflow : 0, W = m.inflow > 0 ? N / m.inflow : 0, _ = x(H + he), z = x(
              H + F * D.height + he
            ), Dt = et[C];
            let Fe = Dt + W * k.height;
            tt[C] += N;
            const Io = m.inflow > 0 && tt[C] >= m.inflow - Fo, Bt = k.y + k.height;
            (Io || Fe > Bt - _o) && (Fe = Bt);
            const Lt = x(Dt + he), zt = x(Fe + he), Z = x(D.x + P), J = x(k.x);
            let ot;
            if (!Eo || J <= Z)
              ot = `M ${Z} ${_} L ${Z} ${z} L ${J} ${zt} L ${J} ${Lt} Z`;
            else {
              const Rt = J - Z, Yt = x(Z + Rt * Mt), Ht = x(J - Rt * Mt), Wt = _, Ut = Lt, Gt = z, Vt = zt;
              ot = `M ${Z} ${Wt} C ${Yt} ${Wt}, ${Ht} ${Ut}, ${J} ${Ut} L ${J} ${Vt} C ${Ht} ${Vt}, ${Yt} ${Gt}, ${Z} ${Gt} Z`;
            }
            It.push({
              id: Ie(),
              source: p,
              target: C,
              path: ot,
              value: N,
              sourceColor: g.color,
              targetColor: l[C].color
            }), H = z - he, et[C] = Fe;
          });
        });
      }), { nodeCoordinates: K, links: It };
    }
    const w = d(() => {
      const o = yo(ke.value);
      return {
        nodes: Object.keys(o.nodeCoordinates).map((l, t) => ({
          ...o.nodeCoordinates[l],
          name: l
        })),
        links: o.links
      };
    }), $t = d(() => yt.value), At = d(() => bt.value), re = d(() => e.value.style.chart.padding), Tt = d(() => ({
      width: Math.max(0, $t.value - 40 - re.value.right - re.value.left),
      height: Math.max(0, At.value - re.value.top - re.value.bottom)
    })), ue = d(() => ({
      width: $t.value,
      height: At.value
    }));
    function bo(o) {
      const l = {}, t = {}, a = /* @__PURE__ */ new Set();
      return wt.value.forEach(([s, A, oe]) => {
        l[s] || (l[s] = []), t[A] || (t[A] = []), l[s].push(A), t[A].push(s);
      }), l[o] && l[o].forEach((s) => a.add(s)), t[o] && t[o].forEach((s) => a.add(s)), Array.from(a).concat(o);
    }
    const $ = u(null), te = u(null), ie = u(null), Ae = u(!1), Te = u(null);
    function St(o, l) {
      X.value = [], $.value = bo(o.name), te.value = o.name, Te.value = o.id;
      const t = o.name, a = wt.value;
      let s = 0, A = 0, oe = [], ve = [];
      const de = new Set(a.map(([v]) => v)), Pe = new Set(a.map(([, v]) => v)), Je = Array.from(de).filter((v) => !Pe.has(v)), le = a.filter(([v]) => Je.includes(v)).reduce((v, [b, q, K]) => v + K, 0), L = {};
      w.value.nodes.forEach((v) => {
        L[v.name] = v.color;
      }), a.forEach(([v, b, q]) => {
        b === t && (s += q, oe.push({ source: v, value: q, color: L[v] })), v === t && (A += q, ve.push({ target: b, value: q, color: L[b] }));
      });
      const Ne = Math.max(s, A), Qe = le > 0 ? Ne / le * 100 : 0, P = {
        name: t,
        inflow: s,
        outflow: A,
        from: oe,
        to: ve,
        percentOfTotal: Qe,
        color: L[t] || "#000000"
      };
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: P, seriesIndex: l }), ie.value = { datapoint: P }, De.value = !0;
      let B = "";
      const fe = e.value.style.chart.tooltip.customFormat;
      if (Ae.value = !1, Xo(fe))
        try {
          const v = fe({
            datapoint: P,
            series: w.value,
            config: e.value
          });
          typeof v == "string" && (Be.value = v, Ae.value = !0);
        } catch {
          console.warn("Custom format cannot be applied.");
        }
      if (!Ae.value) {
        const v = e.value.style.chart.tooltip.showPercentage ? `<div>${ae({
          p: e.value.style.chart.tooltip.translations.percentOfTotal,
          v: P.percentOfTotal,
          s: "%",
          r: e.value.style.chart.tooltip.roundingPercentage
        })}</div>` : "";
        B += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.style.chart.tooltip.borderColor};padding-bottom:6px;margin-bottom:3px;"><span style="margin-right:4px; color:${P.color}">⏹</span>${P.name}${v}</div>`, P.from.length && (B += `<div>${e.value.style.chart.tooltip.translations.from}</div>`, P.from.forEach((b) => {
          B += `<div><span style="color:${b.color}">⏹←</span> ${b.source}: ${Me(
            e.value.style.chart.nodes.labels.formatter,
            b.value,
            ae({
              p: e.value.style.chart.nodes.labels.prefix,
              v: b.value,
              s: e.value.style.chart.nodes.labels.suffix,
              r: e.value.style.chart.nodes.labels.rounding
            })
          )}</div>`;
        })), P.to.length && (B += `<div style="margin-top:6px;">${e.value.style.chart.tooltip.translations.to}</div>`, P.to.forEach((b) => {
          B += `<div><span style="color:${b.color}">⏹→</span> ${b.target}: ${Me(
            e.value.style.chart.nodes.labels.formatter,
            b.value,
            ae({
              p: e.value.style.chart.nodes.labels.prefix,
              v: b.value,
              s: e.value.style.chart.nodes.labels.suffix,
              r: e.value.style.chart.nodes.labels.rounding
            })
          )}</div>`;
        })), Be.value = B;
      }
    }
    function Se(o) {
      Te.value = null;
      const l = ie.value;
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: l, seriesIndex: o }), $.value = null, te.value = null, De.value = !1;
    }
    function je(o) {
      const l = ie.value;
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: l, seriesIndex: o });
    }
    const xt = d(() => w.value.links.map(
      ({ source: o, target: l, sourceColor: t, targetColor: a, value: s }) => ({
        source: o,
        target: l,
        sourceColor: t,
        targetColor: a,
        value: s
      })
    ));
    function Xe(o = null) {
      nt(() => {
        const l = xt.value.map((s, A) => [[s.source], [s.target], [s.value]]), t = [
          [e.value.style.chart.title.text],
          [e.value.style.chart.title.subtitle.text],
          [
            [e.value.table.columnNames.source],
            [e.value.table.columnNames.target],
            [e.value.table.columnNames.value]
          ]
        ].concat(l), a = Vo(t);
        o ? o(a) : jo({
          csvContent: a,
          title: e.value.style.chart.title.text || "vue-ui-flow"
        });
      });
    }
    const xe = d(() => {
      const o = [
        e.value.table.columnNames.source,
        e.value.table.columnNames.target,
        e.value.table.columnNames.value
      ], l = xt.value.map((s, A) => [
        {
          color: s.sourceColor,
          name: s.source,
          shape: "square"
        },
        {
          color: s.targetColor,
          name: s.target,
          shape: "square"
        },
        ae({
          p: e.value.style.chart.nodes.labels.prefix,
          v: s.value,
          s: e.value.style.chart.nodes.labels.suffix,
          r: e.value.style.chart.nodes.labels.rounding
        })
      ]), t = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      };
      return {
        colNames: [
          e.value.table.columnNames.source,
          e.value.table.columnNames.target,
          e.value.table.columnNames.value
        ],
        head: o,
        body: l,
        config: t
      };
    });
    function Co() {
      return w.value;
    }
    function Ot() {
      O.value.showTable = !O.value.showTable;
    }
    const Oe = u(!1);
    function qe() {
      Oe.value = !Oe.value;
    }
    function Pt() {
      O.value.showTooltip = !O.value.showTooltip;
    }
    const Nt = d(() => {
      const o = new Set(
        w.value.nodes.map(
          (l) => e.value.nodeCategories[l.name] || "__uncategorized__"
        )
      );
      return Array.from(o).map((l) => ({
        name: l,
        color: e.value.nodeCategoryColors[l] || Y[0],
        shape: "square",
        count: w.value.nodes.filter(
          (t) => (e.value.nodeCategories[t.name] || "__uncategorized__") === l
        ).length
      })).map((l, t) => ({
        ...l,
        segregate: () => Ze({ legend: l, i: t }),
        opacity: X.value.length ? X.value.includes(t) ? 1 : 0.5 : 1,
        display: `${l.name} (${l.count})`
      }));
    }), Ke = d(
      () => Nt.value.filter((o) => o.name !== "__uncategorized__")
    ), X = u([]);
    function Ze({ legend: o, i: l }) {
      const t = o.name;
      if ($.value?.every(
        (a) => e.value.nodeCategories[a] === t
      )) {
        $.value = null, te.value = null, X.value = [];
        return;
      }
      X.value = [l], $.value = w.value.nodes.filter((a) => e.value.nodeCategories[a.name] === t).map((a) => a.name), te.value = null;
    }
    const ko = d(() => ({
      cy: "flow-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: e.value.style.chart.legend.paddingBottom,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : "normal"
    }));
    async function wo({ scale: o = 2 } = {}) {
      if (!M.value) return;
      const { width: l, height: t } = M.value.getBoundingClientRect(), a = l / t, { imageUri: s, base64: A } = await ul({ domElement: M.value, base64: !0, img: !0, scale: o });
      return {
        imageUri: s,
        base64: A,
        title: e.value.style.chart.title.text,
        width: l,
        height: t,
        aspectRatio: a
      };
    }
    const ce = d(() => {
      const o = e.value.table.useDialog && !e.value.table.show, l = O.value.showTable;
      return {
        component: o ? io : ao,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: o ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: j.value,
          fullscreenParent: M.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: ee.value
        } : {
          hideDetails: !0,
          config: {
            open: l,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    _e(() => O.value.showTable, (o) => {
      e.value.table.show || (o && e.value.table.useDialog && se.value ? se.value.open() : "close" in se.value && se.value.close());
    });
    function Ft() {
      O.value.showTable = !1, Le.value && Le.value.setTableIconState(!1);
    }
    const $o = d(() => Ke.value.map((o) => ({
      ...o,
      name: o.display
    }))), Ao = d(() => e.value.style.chart.backgroundColor), To = d(() => e.value.style.chart.legend), So = d(() => e.value.style.chart.title), { exportSvg: xo, getSvg: Oo } = tl({
      svg: We,
      title: So,
      legend: To,
      legendItems: $o,
      backgroundColor: Ao
    });
    async function _t({ isCb: o }) {
      Ce.value = !0, await nt();
      try {
        if (o) {
          const { blob: l, url: t, text: a, dataUrl: s } = await Oo();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: t, text: a, dataUrl: s }));
        } else
          await Promise.resolve(xo());
      } finally {
        Ce.value = !1;
      }
    }
    async function Po(o, l) {
      if (St(o, l), O.value.showTooltip && (await nt(), M.value && ze.value)) {
        const { left: t, top: a } = M.value.getBoundingClientRect();
        ze.value.placeTooltip({ x: (t ?? 0) + 12, y: (a ?? 0) + 12 });
      }
    }
    function No(o) {
      if (o?.stage === "start") {
        be.value = !0;
        return;
      }
      if (o?.stage === "end") {
        be.value = !1;
        return;
      }
      kt();
    }
    async function Et() {
      if (ho("copyAlt", {
        config: e.value,
        dataset: w.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: w.value
      }));
    }
    return eo({
      getData: Co,
      getImage: wo,
      generateCsv: Xe,
      generateImage: kt,
      generateSvg: _t,
      generatePdf: Ct,
      toggleTable: Ot,
      toggleAnnotator: qe,
      toggleTooltip: Pt,
      drillCategory: Ze,
      unselectNode: Se,
      toggleFullscreen: ht,
      copyAlt: Et
    }), (o, l) => (c(), y("div", {
      ref_key: "flowChart",
      ref: M,
      class: jt(`vue-data-ui-component vue-ui-flow ${j.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      style: R(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor}`),
      id: `flow_${V.value}`,
      onMouseenter: l[2] || (l[2] = () => r(gt)(!0)),
      onMouseleave: l[3] || (l[3] = () => r(gt)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (c(), Q(r(so), {
        key: 0,
        svgRef: r(We),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: Oe.value,
        onClose: qe,
        isCursorPointer: ee.value
      }, {
        "annotator-action-close": f(() => [
          h(o.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": f(({ color: t }) => [
          h(o.$slots, "annotator-action-color", E(I({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": f(({ mode: t }) => [
          h(o.$slots, "annotator-action-draw", E(I({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": f(({ disabled: t }) => [
          h(o.$slots, "annotator-action-undo", E(I({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": f(({ disabled: t }) => [
          h(o.$slots, "annotator-action-redo", E(I({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": f(({ disabled: t }) => [
          h(o.$slots, "annotator-action-delete", E(I({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : T("", !0),
      mo.value ? (c(), y("div", vl, null, 512)) : T("", !0),
      e.value.style.chart.title.text ? (c(), y("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: ct,
        style: "width:100%;background:transparent;padding-bottom:24px"
      }, [
        (c(), Q(rl, {
          key: `title_${ut.value}`,
          config: {
            title: {
              cy: "flow-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "flow-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : T("", !0),
      G("div", {
        id: `legend-top-${V.value}`
      }, null, 8, dl),
      e.value.userOptions.show && ft.value && (r(mt) || r(He)) ? (c(), Q(r(ro), {
        ref_key: "userOptionsRef",
        ref: Le,
        key: `user_option_${rt.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: r(Ge),
        isImaging: r(Ve),
        uid: V.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasXls: e.value.userOptions.buttons.csv,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasTable: e.value.userOptions.buttons.table,
        callbacks: e.value.userOptions.callbacks,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: j.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: M.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        printScale: e.value.userOptions.print.scale,
        isAnnotation: Oe.value,
        hasTooltip: e.value.style.chart.tooltip.show && e.value.userOptions.buttons.tooltip,
        isTooltip: O.value.showTooltip,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: ee.value,
        onToggleTooltip: Pt,
        onToggleFullscreen: ht,
        onGeneratePdf: r(Ct),
        onGenerateCsv: Xe,
        onGenerateImage: No,
        onGenerateSvg: _t,
        onToggleTable: Ot,
        onToggleAnnotator: qe,
        onCopyAlt: Et,
        style: R({
          visibility: r(mt) ? r(He) ? "visible" : "hidden" : "visible"
        })
      }, Xt({ _: 2 }, [
        o.$slots.menuIcon ? {
          name: "menuIcon",
          fn: f(({ isOpen: t, color: a }) => [
            h(o.$slots, "menuIcon", E(I({ isOpen: t, color: a })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        o.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: f(() => [
            h(o.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        o.$slots.optionPdf ? {
          name: "optionPdf",
          fn: f(() => [
            h(o.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        o.$slots.optionCsv ? {
          name: "optionCsv",
          fn: f(() => [
            h(o.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        o.$slots.optionImg ? {
          name: "optionImg",
          fn: f(() => [
            h(o.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        o.$slots.optionSvg ? {
          name: "optionSvg",
          fn: f(() => [
            h(o.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        o.$slots.optionTable ? {
          name: "optionTable",
          fn: f(() => [
            h(o.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        o.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: f(({ toggleFullscreen: t, isFullscreen: a }) => [
            h(o.$slots, "optionFullscreen", E(I({ toggleFullscreen: t, isFullscreen: a })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        o.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: f(({ toggleAnnotator: t, isAnnotator: a }) => [
            h(o.$slots, "optionAnnotator", E(I({ toggleAnnotator: t, isAnnotator: a })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        o.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: f(({ altCopy: t }) => [
            h(o.$slots, "optionAltCopy", E(I({ altCopy: t })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasXls", "hasImg", "hasSvg", "hasTable", "callbacks", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "printScale", "isAnnotation", "hasTooltip", "isTooltip", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : T("", !0),
      (c(), y("svg", {
        ref_key: "svgRef",
        ref: We,
        xmlns: r(Uo),
        viewBox: `0 0 ${ue.value.width} ${ue.value.height}`,
        class: jt({
          "vue-data-ui-fullscreen--on": j.value,
          "vue-data-ui-fulscreen--off": !j.value
        }),
        style: R({
          maxWidth: "100%",
          overflow: "visible",
          background: "transparent",
          color: e.value.style.chart.color
        })
      }, [
        Ee(r(uo)),
        o.$slots["chart-background"] ? (c(), y("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: ue.value.width,
          height: ue.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          h(o.$slots, "chart-background", {}, void 0, !0)
        ], 8, hl)) : T("", !0),
        G("defs", null, [
          (c(!0), y(pe, null, ge(w.value.links, (t, a) => (c(), y("linearGradient", {
            id: t.id,
            x1: "0%",
            y1: "0%",
            x2: "100%",
            y2: "0%"
          }, [
            G("stop", {
              offset: "0%",
              "stop-color": t.sourceColor
            }, null, 8, gl),
            G("stop", {
              offset: "100%",
              "stop-color": t.targetColor
            }, null, 8, ml)
          ], 8, pl))), 256))
        ]),
        (c(!0), y(pe, null, ge(w.value.links, (t) => (c(), y("path", {
          class: "vue-ui-flow-link",
          d: t.path,
          "stroke-linejoin": "round",
          "stroke-miterlimit": "1",
          fill: `url(#${t.id})`,
          stroke: e.value.style.chart.links.stroke,
          "stroke-width": e.value.style.chart.links.strokeWidth,
          style: R(`
                    opacity:${$.value ? $.value.includes(t.source) && $.value.includes(t.target) ? 1 : 0.3 : te.value ? [t.target, t.source].includes(te.value) ? 1 : 0.3 : e.value.style.chart.links.opacity}
                `)
        }, null, 12, yl))), 256)),
        (c(!0), y(pe, null, ge(w.value.nodes, (t, a) => (c(), y("rect", {
          class: "vue-ui-flow-node",
          x: t.x,
          y: r(x)(t.absoluteY) + e.value.style.chart.padding.top,
          height: r(x)(t.height),
          width: $e.value,
          fill: t.color,
          stroke: e.value.style.chart.nodes.stroke,
          "stroke-width": e.value.style.chart.nodes.strokeWidth,
          rx: e.value.style.chart.nodes.borderRadius,
          style: R({
            opacity: $.value ? $.value.includes(t.name) ? 1 : 0.3 : 1,
            outline: Te.value !== null && Te.value === t.id ? "2px solid currentColor" : void 0
          }),
          role: "button",
          tabindex: "0",
          "aria-label": `${t.name}: ${r(Me)(
            e.value.style.chart.nodes.labels.formatter,
            t.value,
            r(ae)({
              p: e.value.style.chart.nodes.labels.prefix,
              v: t.value,
              s: e.value.style.chart.nodes.labels.suffix,
              r: e.value.style.chart.nodes.labels.rounding
            })
          )}`,
          onMouseenter: (s) => St(t, a),
          onMouseleave: (s) => Se(a),
          onClick: (s) => je(a),
          onKeydown: [
            lt(at((s) => je(a), ["prevent"]), ["enter"]),
            lt(at((s) => je(a), ["prevent"]), ["space"]),
            lt(at((s) => Se(a), ["prevent"]), ["esc"])
          ],
          onFocus: (s) => Po(t, a),
          onBlur: (s) => Se(a)
        }, null, 44, bl))), 256)),
        e.value.style.chart.nodes.labels.show ? (c(), y("g", Cl, [
          (c(!0), y(pe, null, ge(w.value.nodes, (t, a) => (c(), y("text", {
            x: t.x + $e.value / 2,
            y: (e.value.style.chart.nodes.labels.showValue ? r(x)(
              t.absoluteY + t.height / 2 - e.value.style.chart.nodes.labels.fontSize / 4
            ) : t.absoluteY + t.height / 2 + e.value.style.chart.nodes.labels.fontSize / 3) + e.value.style.chart.padding.top,
            "font-size": e.value.style.chart.nodes.labels.fontSize,
            fill: r(Jt)(t.color),
            "text-anchor": "middle",
            style: R(`pointer-events: none; opacity:${$.value ? $.value.includes(t.name) ? 1 : 0 : 1}`)
          }, me(e.value.style.chart.nodes.labels.abbreviation.use ? r(Go)({
            source: t.name,
            length: e.value.style.chart.nodes.labels.abbreviation.length
          }) : t.name), 13, kl))), 256)),
          e.value.style.chart.nodes.labels.showValue ? (c(!0), y(pe, { key: 0 }, ge(w.value.nodes, (t, a) => (c(), y("text", {
            x: t.x + $e.value / 2,
            y: r(x)(
              t.absoluteY + t.height / 2 + e.value.style.chart.nodes.labels.fontSize / 1.3
            ) + e.value.style.chart.padding.top,
            "font-size": e.value.style.chart.nodes.labels.fontSize,
            fill: r(Jt)(t.color),
            "text-anchor": "middle",
            style: R(`pointer-events: none; opacity:${$.value ? $.value.includes(t.name) ? 1 : 0 : 1}`)
          }, me(r(Me)(
            e.value.style.chart.nodes.labels.formatter,
            t.value,
            r(ae)({
              p: e.value.style.chart.nodes.labels.prefix,
              v: t.value,
              s: e.value.style.chart.nodes.labels.suffix,
              r: e.value.style.chart.nodes.labels.rounding
            }),
            { datapoint: t, seriesIndex: a }
          )), 13, wl))), 256)) : T("", !0)
        ])) : T("", !0),
        h(o.$slots, "svg", {
          svg: {
            ...ue.value,
            isPrintingImg: r(Ge) | r(Ve) | be.value,
            isPrintingSvg: Ce.value
          }
        }, void 0, !0)
      ], 14, fl)),
      o.$slots.watermark ? (c(), y("div", $l, [
        h(o.$slots, "watermark", E(I({ isPrinting: r(Ge) || r(Ve) || be.value || Ce.value })), void 0, !0)
      ])) : T("", !0),
      G("div", {
        id: `legend-bottom-${V.value}`
      }, null, 8, Al),
      dt.value ? (c(), Q(Bo, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${V.value}` : `#legend-bottom-${V.value}`
      }, [
        G("div", {
          ref_key: "chartLegend",
          ref: it
        }, [
          e.value.style.chart.legend.show && Ke.value.length ? (c(), Q(sl, {
            key: 0,
            legendSet: Ke.value,
            config: ko.value,
            isCursorPointer: ee.value,
            onClickMarker: l[0] || (l[0] = (t) => Ze(t))
          }, {
            item: f(({ legend: t, index: a }) => [
              r(Re) ? T("", !0) : (c(), y("div", {
                key: 0,
                onClick: (s) => t.segregate(),
                style: R(`opacity:${X.value.length ? X.value.includes(a) ? 1 : 0.5 : 1}`)
              }, me(t.display), 13, Tl))
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : T("", !0),
          h(o.$slots, "legend", { legend: Nt.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : T("", !0),
      o.$slots.source ? (c(), y("div", {
        key: 6,
        ref_key: "source",
        ref: vt,
        dir: "auto"
      }, [
        h(o.$slots, "source", {}, void 0, !0)
      ], 512)) : T("", !0),
      Ee(r(oo), {
        ref_key: "tooltip",
        ref: ze,
        teleportTo: e.value.style.chart.tooltip.teleportTo,
        show: O.value.showTooltip && De.value,
        backgroundColor: e.value.style.chart.tooltip.backgroundColor,
        color: e.value.style.chart.tooltip.color,
        fontSize: e.value.style.chart.tooltip.fontSize,
        borderRadius: e.value.style.chart.tooltip.borderRadius,
        borderColor: e.value.style.chart.tooltip.borderColor,
        borderWidth: e.value.style.chart.tooltip.borderWidth,
        backgroundOpacity: e.value.style.chart.tooltip.backgroundOpacity,
        position: e.value.style.chart.tooltip.position,
        offsetY: e.value.style.chart.tooltip.offsetY,
        parent: M.value,
        content: Be.value,
        isCustom: Ae.value,
        isFullscreen: j.value,
        smooth: e.value.style.chart.tooltip.smooth,
        backdropFilter: e.value.style.chart.tooltip.backdropFilter,
        smoothForce: e.value.style.chart.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.chart.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": f(() => [
          h(o.$slots, "tooltip-before", E(I({ ...ie.value })), void 0, !0)
        ]),
        "tooltip-after": f(() => [
          h(o.$slots, "tooltip-after", E(I({ ...ie.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "backgroundOpacity", "position", "offsetY", "parent", "content", "isCustom", "isFullscreen", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      ft.value && e.value.userOptions.buttons.table ? (c(), Q(Lo(ce.value.component), zo({ key: 7 }, ce.value.props, {
        ref_key: "tableUnit",
        ref: se,
        onClose: Ft
      }), Xt({
        content: f(() => [
          Ee(r(no), {
            colNames: xe.value.colNames,
            head: xe.value.head,
            body: xe.value.body,
            config: xe.value.config,
            title: e.value.table.useDialog ? "" : ce.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: ee.value,
            onClose: Ft
          }, {
            th: f(({ th: t }) => [
              G("div", {
                innerHTML: t,
                style: { display: "flex", "align-items": "center" }
              }, null, 8, Sl)
            ]),
            td: f(({ td: t }) => [
              qt(me(t.name || t), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"])
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: f(() => [
            qt(me(ce.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: f(() => [
            G("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: l[1] || (l[1] = (t) => Xe(e.value.userOptions.callbacks.csv)),
              style: R({ cursor: ee.value ? "pointer" : "default" })
            }, [
              Ee(r(lo), {
                name: "fileCsv",
                stroke: ce.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : T("", !0),
      h(o.$slots, "skeleton", {}, () => [
        r(Re) ? (c(), Q(el, { key: 0 })) : T("", !0)
      ], !0)
    ], 46, cl));
  }
}, Ul = /* @__PURE__ */ il(xl, [["__scopeId", "data-v-e4271008"]]);
export {
  Ul as default
};
