// vue-data-ui/dist/vue-ui-candlestick-BY-_E7eZ.js
import { defineAsyncComponent as X, computed as h, ref as v, toRefs as Dl, watch as ie, shallowRef as Tt, onMounted as Bl, onBeforeUnmount as Ue, watchEffect as be, nextTick as He, openBlock as u, createElementBlock as i, normalizeStyle as Z, normalizeClass as V, createBlock as K, unref as y, withCtx as f, renderSlot as g, normalizeProps as z, guardReactiveProps as L, createCommentVNode as m, createSlots as It, createVNode as Xe, createElementVNode as p, Fragment as T, renderList as M, toDisplayString as ne, mergeProps as Ft, resolveDynamicComponent as Nl, createTextVNode as Mt } from "vue";
import { j as El, t as Rt, o as Pt, q as J, D as Vl, K as q, I as Wl, l as W, X as Ul, v as xe, u as Hl, a0 as Xl, a1 as ql, T as jl, U as Gl, a2 as Ot } from "./lib-GwPnJSrY.js";
import { t as Dt, u as Yl } from "./useResponsive-DfdjqQps.js";
import { u as Kl } from "./useConfig-CIgyKZ6R.js";
import { u as Bt, a as Nt, b as Jl } from "./useTimeLabels-C-A2uZKi.js";
import { u as Ql } from "./usePrinter-DiK5R6k-.js";
import { u as Zl, B as ea } from "./BaseScanner-BhhkKVoj.js";
import { u as ta } from "./useSvgExport-ByUukOZt.js";
import { u as qe } from "./useNestedProp-BBT0mbHh.js";
import { u as la } from "./useThemeCheck-D1ZYXUwG.js";
import { u as aa } from "./useUserOptionState-BIvW1Kz7.js";
import { u as oa } from "./useChartAccessibility-9icAAmYg.js";
import { u as sa } from "./useTimeLabelCollider-CIsgDrl9.js";
import ua from "./img-BYYzY_QU.js";
import ia from "./Title-BfyrP-jp.js";
import na from "./vue_ui_candlestick-BAMn-VGw.js";
import { S as ra } from "./SlicerPreview-I2J0UUF_.js";
import { _ as va } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const da = ["id"], ca = ["xmlns", "aria-label", "viewBox"], ya = ["x", "y", "width", "height"], ha = { key: 1 }, ma = ["id"], pa = ["stop-color"], fa = ["stop-color"], ga = ["stop-color"], ba = ["id"], xa = ["stop-color"], ka = ["stop-color"], wa = ["stop-color"], Ca = { key: 0 }, Aa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], $a = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Sa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], _a = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], za = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], La = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Ta = ["x", "y", "font-size", "fill", "font-weight"], Ia = ["transform", "text-anchor", "font-size", "fill", "font-weight"], Fa = ["transform", "text-anchor", "font-size", "fill", "font-weight"], Ma = ["x", "y", "width", "height", "fill", "rx"], Ra = { key: 0 }, Pa = ["cx", "cy", "r", "fill"], Oa = ["cx", "cy", "r", "fill"], Da = { key: 1 }, Ba = ["x", "y", "width", "height", "rx", "fill"], Na = ["x", "y", "width", "height", "fill", "rx"], Ea = ["x", "y", "height", "width", "fill", "rx"], Va = ["x", "y", "height", "width", "fill", "rx", "stroke", "stroke-width"], Wa = ["d", "stroke"], Ua = ["x", "y", "height", "width", "fill", "onMouseover", "onMouseleave", "onClick"], Ha = ["data-start", "data-end"], Xa = {
  key: 4,
  class: "vue-data-ui-watermark"
}, qa = ["d", "stroke"], ja = ["d", "stroke", "stroke-width"], Ga = ["innerHTML"], Ya = {
  __name: "vue-ui-candlestick",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    selectedXIndex: {
      type: Number,
      default: void 0
    }
  },
  emits: ["selectX", "copyAlt"],
  setup(Et, { expose: Vt, emit: Wt }) {
    const Ut = X(() => import("./BaseIcon-CNx8Tzl5.js")), Ht = X(() => import("./Tooltip-DL3QgjZf.js")), Xt = X(() => import("./vue-ui-accordion-BGHTqIBt.js")), qt = X(() => import("./DataTable-Bo1pmAqG.js")), jt = X(() => import("./PenAndPaper-DWV1FMIi.js")), Gt = X(() => import("./UserOptions-CrLqEimB.js")), Yt = X(() => import("./PackageVersion-DN9d0ME5.js")), Kt = X(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_candlestick: Jt } = Kl(), { isThemeValid: Qt, warnInvalidTheme: Zt } = la(), $ = Et, ke = Wt, we = h(() => !!$.dataset && $.dataset.length), O = v(!1), Ce = v(null), R = v(El()), Ae = v(!1), $e = v(""), B = v(void 0), je = v(0), N = v(null), Ge = v(null), Ye = v(null), ee = v(null), Ke = v(null), Je = v(null), el = v(0), Qe = v(0), Ze = v(0), Se = v(null), te = v(null), le = v(null), _e = v(null), et = v(null), re = v(!1), ve = v(!1), e = v(Ie()), ae = h(() => e.value.userOptions.useCursorPointer), tl = h(() => Rt({
      defaultConfig: {
        useCssAnimation: !1,
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          backgroundColor: "#99999930",
          layout: {
            candle: {
              colors: {
                bearish: "#BABABA",
                bullish: "#CACACA"
              }
            },
            grid: {
              stroke: "#6A6A6A",
              verticalLines: {
                stroke: "#6A6A6A"
              },
              horizontalLines: {
                stroke: "#6A6A6A"
              },
              yAxis: {
                dataLabels: { show: !1 },
                scale: {
                  min: null,
                  max: null
                }
              }
            },
            wick: {
              stroke: "#6A6A6A",
              extremity: {
                color: "#6A6A6A"
              }
            }
          },
          tooltip: { show: !1 },
          zoom: {
            show: !1,
            startIndex: null,
            endIndex: null
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: E, FINAL_DATASET: x, manualLoading: ze } = Zl({
      ...Dl($),
      FINAL_CONFIG: e,
      prepareConfig: Ie,
      callback: () => {
        Promise.resolve().then(async () => {
          await me();
        });
      },
      skeletonDataset: $.config?.skeletonDataset ?? [
        [17040672e5, 10, 20, 2, 10, 30],
        [17067456e5, 10, 30, 5, 20, 50],
        [17092512e5, 20, 50, 10, 30, 80],
        [17119296e5, 30, 80, 20, 50, 130],
        [17145216e5, 50, 130, 30, 100, 210],
        [17172e8, 80, 210, 50, 150, 340],
        [1719792e6, 130, 340, 80, 280, 550],
        [17224704e5, 210, 550, 130, 450, 890],
        [17251488e5, 340, 890, 210, 750, 1440],
        [17277408e5, 550, 1440, 340, 1230, 2330],
        [17304192e5, 890, 2330, 550, 1950, 3770],
        [17330112e5, 1440, 3770, 890, 3200, 5100]
      ],
      skeletonConfig: Rt({
        defaultConfig: e.value,
        userConfig: tl.value
      })
    }), { userOptionsVisible: Le, setUserOptionsVisibility: tt, keepUserOptionState: lt } = aa({ config: e.value }), { svgRef: Te } = oa({ config: e.value.style.title });
    function ll() {
      tt(!0);
    }
    function al() {
      tt(!1), ke("selectX", { seriesIndex: null, datapoint: null }), B.value = null;
    }
    function Ie() {
      const t = qe({
        userConfig: $.config,
        defaultConfig: Jt
      });
      let a = {};
      const l = t.theme;
      if (l)
        if (!Qt.value(t))
          Zt(t), a = t;
        else {
          const o = qe({
            userConfig: na[l] || $.config,
            defaultConfig: t
          });
          a = {
            ...qe({
              userConfig: $.config,
              defaultConfig: o
            })
          };
        }
      else
        a = t;
      return a;
    }
    ie(() => $.config, (t) => {
      E.value || (e.value = Ie()), Le.value = !e.value.userOptions.showOnChartHover, at(), el.value += 1, Ze.value += 1, Qe.value += 1, I.value.showTable = e.value.table.show, I.value.showTooltip = e.value.style.tooltip.show, de();
    }, { deep: !0 }), ie(() => $.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (ze.value = !1), bt();
    }, { deep: !0 });
    const b = v({
      height: e.value.style.height,
      width: e.value.style.width,
      xAxisFontSize: e.value.style.layout.grid.xAxis.dataLabels.fontSize,
      yAxisFontSize: e.value.style.layout.grid.yAxis.dataLabels.fontSize
    }), U = Tt(null), Q = Tt(null);
    Bl(() => {
      at();
    });
    const j = h(() => !!e.value.debug);
    function at() {
      if (Pt($.dataset) && (J({
        componentName: "VueUiCandlestick",
        type: "dataset",
        debug: j.value
      }), ze.value = !0), Pt($.dataset) || (ze.value = e.value.loading), setTimeout(() => {
        O.value = !0;
      }, 10), e.value.responsive) {
        const a = Dt(() => {
          O.value = !1;
          const { width: l, height: o } = Yl({
            chart: N.value,
            title: e.value.style.title.text ? Ge.value : null,
            slicer: e.value.style.zoom.show && A.value > 6 ? ee.value.$el : null,
            legend: Ye.value,
            source: Ke.value,
            noTitle: Je.value
          });
          requestAnimationFrame(() => {
            b.value.width = l, b.value.height = o - 12, e.value.responsiveProportionalSizing ? (b.value.xAxisFontSize = Ot({
              relator: Math.min(l, o),
              adjuster: e.value.style.width,
              source: e.value.style.layout.grid.xAxis.dataLabels.fontSize,
              threshold: 6,
              fallback: 6
            }), b.value.yAxisFontSize = Ot({
              relator: Math.min(l, o),
              adjuster: e.value.style.width,
              source: e.value.style.layout.grid.yAxis.dataLabels.fontSize,
              threshold: 6,
              fallback: 6
            })) : (b.value.xAxisFontSize = e.value.style.layout.grid.xAxis.dataLabels.fontSize, b.value.yAxisFontSize = e.value.style.layout.grid.yAxis.dataLabels.fontSize), Ce.value && clearTimeout(Ce.value), Ce.value = setTimeout(() => {
              O.value = !0;
            }, 10);
          });
        });
        U.value && (Q.value && U.value.unobserve(Q.value), U.value.disconnect()), U.value = new ResizeObserver(a), Q.value = N.value.parentNode, U.value.observe(Q.value);
      }
      me();
    }
    Ue(() => {
      U.value && (Q.value && U.value.unobserve(Q.value), U.value.disconnect());
    });
    const { isPrinting: Fe, isImaging: Me, generatePdf: ot, generateImage: st } = Ql({
      elementId: `vue-ui-candlestick_${R.value}`,
      fileName: e.value.style.title.text || "vue-ui-candlestick",
      options: e.value.userOptions.print
    }), ol = h(() => e.value.userOptions.show && !e.value.style.title.text), I = v({
      showTable: e.value.table.show,
      showTooltip: e.value.style.tooltip.show
    }), Re = v(0), sl = Dt((t) => {
      Re.value = t;
    }, 100);
    be((t) => {
      const a = te.value;
      if (!a) return;
      const l = new ResizeObserver((o) => {
        sl(o[0].contentRect.height);
      });
      l.observe(a), t(() => l.disconnect());
    }), Ue(() => {
      Re.value = 0;
    });
    const ut = h(() => {
      let t = 0;
      return te.value && (t = Re.value + b.value.xAxisFontSize), t;
    });
    function ul() {
      let t = e.value.style.layout.grid.yAxis.dataLabels.offsetX;
      return Se.value && (t = Array.from(Se.value.querySelectorAll("text")).reduce((o, s) => {
        const d = s.getComputedTextLength();
        return d > o ? d : o;
      }, 0)), t + 13;
    }
    const n = h(() => {
      const { top: t, right: a, bottom: l, left: o } = e.value.style.layout.padding, s = ul(), d = 12;
      return {
        top: t + d,
        right: b.value.width - a,
        left: o + s,
        bottom: b.value.height - l - ut.value,
        width: b.value.width - o - a - s,
        height: b.value.height - t - l - ut.value - d
      };
    }), il = h(() => {
      const { left: t, top: a, width: l, height: o } = n.value, s = r.value.start, k = r.value.end - s, F = l / k, _ = C.value.start - s, H = C.value.end - s, Y = Math.max(0, Math.min(k, _)), ge = Math.max(0, Math.min(k, H));
      return {
        x: t + Y * F,
        y: a,
        width: (ge - Y) * F,
        height: o,
        fill: e.value.style.zoom.preview.fill,
        stroke: e.value.style.zoom.preview.stroke,
        "stroke-width": e.value.style.zoom.preview.strokeWidth,
        "stroke-dasharray": e.value.style.zoom.preview.strokeDasharray,
        "stroke-linecap": "round",
        "stroke-linejoin": "round",
        style: {
          pointerEvents: "none",
          transition: "none !important",
          animation: "none !important"
        }
      };
    }), A = h(() => x.value.length), r = v({
      start: 0,
      end: A.value
    }), C = v({ start: 0, end: A.value }), it = h(() => e.value.style.zoom.preview.enable && (C.value.start !== r.value.start || C.value.end !== r.value.end));
    function nt(t, a) {
      C.value[t] = a;
    }
    function de() {
      let t = Math.max(0, Math.min(r.value.start ?? 0, A.value - 1)), a = Math.max(t + 1, Math.min(r.value.end ?? A.value, A.value));
      (!Number.isFinite(t) || !Number.isFinite(a) || a <= t) && (t = 0, a = A.value), r.value.start = t, r.value.end = a, C.value.start = t, C.value.end = a, ee.value && (ee.value.setStartValue(t), ee.value.setEndValue(a));
    }
    const rt = h(() => x.value.map((t, a) => ({
      ...t,
      absoluteIndex: a
    }))), vt = h(() => rt.value.slice(r.value.start, r.value.end)), S = h(() => (x.value.forEach((t, a) => {
      [null, void 0].includes(t[0]) && J({
        componentName: "VueUiCandlestick",
        type: "datasetAttribute",
        property: "period (index 0)",
        index: a,
        debug: j.value
      }), [null, void 0].includes(t[1]) && J({
        componentName: "VueUiCandlestick",
        type: "datasetAttribute",
        property: "open (index 1)",
        index: a,
        debug: j.value
      }), [null, void 0].includes(t[2]) && J({
        componentName: "VueUiCandlestick",
        type: "datasetAttribute",
        property: "high (index 2)",
        index: a,
        debug: j.value
      }), [null, void 0].includes(t[3]) && J({
        componentName: "VueUiCandlestick",
        type: "datasetAttribute",
        property: "low (index 3)",
        index: a,
        debug: j.value
      }), [null, void 0].includes(t[4]) && J({
        componentName: "VueUiCandlestick",
        type: "datasetAttribute",
        property: "close (index 4)",
        index: a,
        debug: j.value
      }), [null, void 0].includes(t[5]) && J({
        componentName: "VueUiCandlestick",
        type: "datasetAttribute",
        property: "volume (index 5)",
        index: a,
        debug: j.value
      });
    }), vt.value.map((t) => ({
      absoluteIndex: t.absoluteIndex,
      period: t[0],
      open: t[1],
      high: t[2],
      low: t[3],
      close: t[4],
      volume: t[5]
    })))), nl = h(() => rt.value.map((t) => ({
      absoluteIndex: t.absoluteIndex,
      period: t[0],
      open: t[1],
      high: t[2],
      low: t[3],
      close: t[4],
      volume: t[5]
    }))), c = h(() => n.value.width / vt.value.length), dt = h(() => {
      const t = e.value.style.layout.grid.yAxis.scale.max === null ? Math.max(...S.value.map((l) => l.high)) : e.value.style.layout.grid.yAxis.scale.max, a = e.value.style.layout.grid.yAxis.scale.min === null ? 0 : e.value.style.layout.grid.yAxis.scale.min;
      return {
        max: t,
        min: a
      };
    }), P = h(() => Vl(dt.value.min, dt.value.max, e.value.style.layout.grid.yAxis.dataLabels.steps));
    function ce(t, a, l = null, o = null) {
      return {
        ...t,
        x: q(n.value.left + a * c.value + c.value / 2),
        y: q(n.value.top + (1 - (t - P.value.min) / (P.value.max - P.value.min)) * n.value.height),
        value: q(t),
        isMax: t === l,
        isMin: t === o
      };
    }
    const w = h(() => {
      const t = {
        o: Math.max(...S.value.map((s) => s.open)),
        h: Math.max(...S.value.map((s) => s.high)),
        l: Math.max(...S.value.map((s) => s.low)),
        c: Math.max(...S.value.map((s) => s.low))
      }, a = {
        o: Math.min(...S.value.map((s) => s.open)),
        h: Math.min(...S.value.map((s) => s.high)),
        l: Math.min(...S.value.map((s) => s.low)),
        c: Math.min(...S.value.map((s) => s.low))
      }, l = Math.max(...S.value.map((s) => s.volume)), o = Math.min(...S.value.map((s) => s.volume));
      return S.value.map((s, d) => {
        const k = ce(s.open, d, t.o, a.o), F = ce(s.high, d, t.h, a.h), _ = ce(s.low, d, t.l, a.l), H = ce(s.close, d, t.c, a.c), Y = s.close > s.open, ge = s.volume === l, Ee = s.volume === o;
        return {
          period: s.period,
          open: k,
          high: F,
          low: _,
          close: H,
          volume: s.volume,
          isBullish: Y,
          absoluteIndex: s.absoluteIndex,
          isMaxVolume: ge,
          isMinVolume: Ee
        };
      });
    });
    function ye({ item: t, index: a, minimapH: l, unitW: o }) {
      const s = e.value.style.layout.grid.yAxis.scale.min ?? 0, d = e.value.style.layout.grid.yAxis.scale.max ?? Math.max(...x.value.map((k) => k[2]));
      return {
        ...t,
        x: q(a * o),
        y: q((1 - (t - s) / (d - s)) * l),
        value: q(t)
      };
    }
    const rl = h(() => ({ minimapH: t, unitW: a }) => nl.value.map((l, o) => {
      const s = ye({ item: l.open, index: o, minimapH: t, unitW: a }), d = ye({ item: l.high, index: o, minimapH: t, unitW: a }), k = ye({ item: l.low, index: o, minimapH: t, unitW: a }), F = ye({ item: l.close, index: o, minimapH: t, unitW: a }), _ = l.close > l.open;
      return {
        period: l.period,
        open: s,
        high: d,
        low: k,
        close: F,
        volume: l.volume,
        isBullish: _,
        absoluteIndex: l.absoluteIndex
      };
    })), vl = h(() => e.value.style.zoom.minimap.show ? [{
      name: "",
      series: x.value.map((t) => t[2]),
      color: "#000000",
      isVisible: !0
    }] : []);
    function dl(t) {
      return q((t - P.value.min) / (P.value.max - P.value.min));
    }
    const ct = h(() => P.value.ticks.map((t) => ({
      y: n.value.bottom - n.value.height * dl(t),
      value: q(t)
    }))), cl = h(() => S.value.map((t) => t.period)), D = v([]), Pe = v([]);
    let yt = 0;
    be(() => {
      const t = ++yt;
      (async () => {
        const a = await Bt({
          values: x.value.map((l) => l[0]),
          maxDatapoints: x.value.length,
          formatter: e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter,
          start: r.value.start,
          end: r.value.end
        });
        t === yt && (D.value = a);
      })();
    });
    let ht = 0;
    be(() => {
      const t = ++ht;
      (async () => {
        const a = await Bt({
          values: x.value.map((l) => l[0]),
          maxDatapoints: x.value.length,
          formatter: e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter,
          start: 0,
          end: A.value
        });
        t === ht && (Pe.value = a);
      })();
    });
    const yl = h(() => {
      const t = e.value.style.layout.grid.xAxis.dataLabels.modulo;
      return D.value.length ? Math.min(t, [...new Set(D.value.map((a) => a.text))].length) : t;
    }), mt = v({ months: [], shortMonths: [], days: [], shortDays: [] });
    let pt = 0;
    be(() => {
      const t = ++pt, a = e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter;
      (async () => {
        const l = await Nt(a.locale).catch(() => Nt("en"));
        t === pt && (mt.value = l.data);
      })();
    });
    const ft = h(() => {
      const t = e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter, a = Jl({
        useUTC: t.useUTC,
        locale: mt.value,
        januaryAsYear: t.januaryAsYear
      });
      return (l, o) => {
        const d = x.value.map((k) => k[0])?.[l];
        return d == null ? "" : a.formatDate(new Date(d), o);
      };
    }), hl = h(() => (x.value.map((a) => a[0]) || []).map((a, l) => ({
      text: ft.value(l, e.value.style.tooltip.timeFormat),
      absoluteIndex: l
    }))), ml = h(() => (x.value.map((a) => a[0]) || []).map((a, l) => ({
      text: ft.value(l, e.value.style.zoom.timeFormat),
      absoluteIndex: l
    }))), Oe = h(() => {
      const t = e.value.style.layout.grid.xAxis.dataLabels, a = D.value || [], l = Pe.value || [], o = r.value.start ?? 0, s = B.value, d = A.value, k = a.map((_) => _?.text ?? ""), F = l.map((_) => _?.text ?? "");
      return Wl(
        !!t.showOnlyFirstAndLast,
        !!t.showOnlyAtModulo,
        Math.max(1, yl.value || 1),
        k,
        F,
        o,
        s,
        d
      );
    }), gt = h(() => e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable ? {
      start: D.value.find((t) => t.absoluteIndex === r.value.start)?.text ?? "",
      end: D.value.find((t) => t.absoluteIndex === r.value.end - 1)?.text ?? ""
    } : {
      start: x.value[r.value.start] ? x.value[r.value.start][0] : x.value[0][0],
      end: x.value[r.value.end - 1] ? x.value[r.value.end - 1][0] : x.value.at(-1)[0]
    }), De = v(null);
    function pl(t, a) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: a, seriesIndex: t + r.value.start });
    }
    function fl(t, a) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: a, seriesIndex: t + r.value.start }), B.value = void 0, Ae.value = !1;
    }
    ie(() => $.selectedXIndex, (t) => {
      if ([null, void 0].includes($.selectedXIndex)) {
        B.value = null;
        return;
      }
      const a = t - r.value.start;
      a < 0 || t >= r.value.end ? B.value = null : B.value = a ?? null;
    }, { immediate: !0 });
    function gl(t, a) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: a, seriesIndex: t + r.value.start }), B.value = t, De.value = {
        datapoint: a,
        seriesIndex: t,
        series: w.value,
        config: e.value
      }, Tl({ seriesIndex: t, datapoint: a });
      const l = e.value.style.tooltip.customFormat;
      if (jl(l) && Gl(() => l({
        seriesIndex: t,
        datapoint: a,
        series: w.value,
        config: e.value
      })))
        $e.value = l({
          seriesIndex: t,
          datapoint: a,
          series: w.value,
          config: e.value
        });
      else if (e.value.style.tooltip.show) {
        let o = "";
        const { period: s, open: d, high: k, low: F, close: _, volume: H, isBullish: Y } = w.value[t], { period: ge, open: Ee, high: Fl, low: Ml, close: Rl, volume: Pl } = e.value.translations, Ol = e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable ? e.value.style.tooltip.useDefaultTimeFormat ? D.value[t].text : hl.value[t].text : s;
        o += `<div><svg style="margin-right:6px" viewBox="0 0 12 12" height="12" width="12"><rect x="0" y="0" height="12" width="12" rx="${e.value.style.layout.candle.borderRadius * 3}" stroke="${e.value.style.layout.candle.stroke}" stroke-width="${e.value.style.layout.candle.strokeWidth}" 
                fill="${e.value.style.layout.candle.gradient.show ? Y ? `url(#bullish_gradient_${R.value})` : `url(#bearish_gradient_${R.value})` : Y ? e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.colors.bearish}"/></svg>${Ol}</div>`, o += `${Pl} : <b>${isNaN(H) ? "-" : Number(H.toFixed(e.value.style.tooltip.roundingValue)).toLocaleString()}</b>`, o += `<div style="margin-top:6px;padding-top:6px;border-top:1px solid ${e.value.style.tooltip.borderColor}">`;
        const Ve = W({
          p: e.value.style.tooltip.prefix,
          v: d.value,
          s: e.value.style.tooltip.suffix,
          r: e.value.style.tooltip.roundingValue
        }), zt = W({
          p: e.value.style.tooltip.prefix,
          v: k.value,
          s: e.value.style.tooltip.suffix,
          r: e.value.style.tooltip.roundingValue
        }), Lt = W({
          p: e.value.style.tooltip.prefix,
          v: F.value,
          s: e.value.style.tooltip.suffix,
          r: e.value.style.tooltip.roundingValue
        }), We = W({
          p: e.value.style.tooltip.prefix,
          v: _.value,
          s: e.value.style.tooltip.suffix,
          r: e.value.style.tooltip.roundingValue
        });
        e.value.style.tooltip.showChart ? o += `<div style="width:100%;display:flex;align-items:center;justify-content:center;">
                    <svg viewBox="0 0 100 100" width="100px" style="background: transparent; overflow: visible">
                        <g>
                            <line x1="50" x2="50" y1="20" y2="80" stroke="${a.isBullish ? e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.colors.bearish}" stroke-width="2" stroke-linecap="round" />
                            ${a.isBullish ? `
                                <line x1="45" x2="50" y1="65" y2="65" stroke="${e.value.style.layout.candle.colors.bullish}" stroke-width="1.5" stroke-linecap="round" />
                                <line x1="50" x2="55" y1="35" y2="35" stroke="${e.value.style.layout.candle.colors.bullish}" stroke-width="1.5" stroke-linecap="round" />
                                <text x="38" y="70" text-anchor="end" fill="${e.value.style.tooltip.color}">${Ve}</text>
                                <text x="62" y="40" text-anchor="start" fill="${e.value.style.tooltip.color}">${We}</text>
                            ` : `
                                <line x1="45" x2="50" y1="35" y2="35" stroke="${e.value.style.layout.candle.colors.bearish}" stroke-width="1.5" stroke-linecap="round" />
                                <line x1="50" x2="55" y1="65" y2="65" stroke="${e.value.style.layout.candle.colors.bearish}" stroke-width="1.5" stroke-linecap="round" />
                                <text x="40" y="40" text-anchor="end" fill="${e.value.style.tooltip.color}">${Ve}</text>
                                <text x="60" y="70" text-anchor="start" fill="${e.value.style.tooltip.color}">${We}</text>
                            `}
                            <text x="50" y="13" text-anchor="middle" fill="${e.value.style.tooltip.color}">${zt}</text>
                            <text x="50" y="97" text-anchor="middle" fill="${e.value.style.tooltip.color}">${Lt}</text>
                        <g>
                    </svg>
                    <div>
                ` : (o += `<div>${Ee}: <b>${Ve}</b></div>`, o += `<div>${Fl}: <b>${zt}</b></div>`, o += `<div>${Ml}: <b>${Lt}</b></div>`, o += `<div>${Rl}: <b>${We}</b></div>`), o += "</div>", $e.value = `<div style="text-align:right">${o}</div>`;
      }
      Ae.value = !0;
    }
    const oe = v(null);
    function bl() {
      return new Promise((t) => requestAnimationFrame(
        () => requestAnimationFrame(() => t())
      ));
    }
    Ue(() => {
      oe.value && cancelAnimationFrame(oe.value);
    });
    async function bt() {
      me(), await He(), oe.value && cancelAnimationFrame(oe.value), oe.value = requestAnimationFrame(async () => {
        await bl(), me();
      });
    }
    const se = v(!1), xt = v(!1), he = v(!1);
    function me() {
      if (!se.value) {
        se.value = !0;
        try {
          const { startIndex: t, endIndex: a } = e.value.style.zoom, l = A.value, o = t ?? 0, s = a != null ? Math.min(kt(a + 1), l) : l;
          he.value = !0, r.value.start = o, r.value.end = s, C.value.start = o, C.value.end = s, de(), xt.value = !0;
        } finally {
          queueMicrotask(() => {
            he.value = !1;
          }), se.value = !1;
        }
      }
    }
    function kt(t) {
      const a = A.value;
      return t > a ? a : t < 0 || t < r.value.start ? e.value.style.zoom.startIndex !== null ? e.value.style.zoom.startIndex + 1 : 1 : t;
    }
    function Be(t = null) {
      He(() => {
        const a = [e.value.translations.period, e.value.translations.open, e.value.translations.high, e.value.translations.low, e.value.translations.close, e.value.translations.volume], l = w.value.map((d, k) => [
          e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable ? D.value[k].text : d.period,
          d.open.value,
          d.high.value,
          d.low.value,
          d.close.value,
          d.volume
        ]), o = [[e.value.style.title.text], [e.value.style.title.subtitle.text], [[""], [""], [""]]].concat([a]).concat(l), s = Xl(o);
        t ? t(s) : ql({ csvContent: s, title: e.value.style.title.text || "vue-ui-candlestick" });
      });
    }
    const pe = h(() => {
      const t = w.value.map((o, s) => {
        const d = e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable ? D.value?.[s]?.text ?? "" : o.period, k = W({
          p: e.value.table.td.prefix,
          v: o.open.value,
          s: e.value.table.td.suffix,
          r: e.value.table.td.roundingValue
        }), F = W({
          p: e.value.table.td.prefix,
          v: o.high.value,
          s: e.value.table.td.suffix,
          r: e.value.table.td.roundingValue
        }), _ = W({
          p: e.value.table.td.prefix,
          v: o.low.value,
          s: e.value.table.td.suffix,
          r: e.value.table.td.roundingValue
        }), H = W({
          p: e.value.table.td.prefix,
          v: o.close.value,
          s: e.value.table.td.suffix,
          r: e.value.table.td.roundingValue
        });
        return [
          `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 12" height="12" width="12" style="margin-right: 6px"><rect x="0" y="0" height="12" width="12" rx="${e.value.style.layout.candle.borderRadius * 3}" fill="${e.value.style.layout.candle.gradient.show ? o.isBullish ? `url(#bullish_gradient_${R.value}` : `url(#bearish_gradient_${R.value})` : o.isBullish ? e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.colors.bearish}"/></svg> ${d}`,
          k,
          F,
          _,
          H,
          `${isNaN(o.volume) ? "-" : o.volume.toLocaleString()}`
        ];
      }), a = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, l = [
        e.value.translations.period,
        e.value.translations.open,
        e.value.translations.high,
        e.value.translations.low,
        e.value.translations.last,
        e.value.translations.volume
      ];
      return { head: l, body: t, config: a, colNames: l };
    }), G = v(!1);
    function wt(t) {
      G.value = t, je.value += 1;
    }
    function Ct() {
      I.value.showTable = !I.value.showTable;
    }
    function At() {
      I.value.showTooltip = !I.value.showTooltip;
    }
    const fe = v(!1);
    function Ne() {
      fe.value = !fe.value;
    }
    async function xl({ scale: t = 2 } = {}) {
      if (!N.value) return;
      const { width: a, height: l } = N.value.getBoundingClientRect(), o = a / l, { imageUri: s, base64: d } = await ua({ domElement: N.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: s,
        base64: d,
        title: e.value.style.title.text,
        width: a,
        height: l,
        aspectRatio: o
      };
    }
    const kl = h(() => b.value.width), wl = h(() => b.value.height);
    sa({
      timeLabelsEls: te,
      timeLabels: D,
      slicer: r,
      configRef: e,
      rotationPath: ["style", "layout", "grid", "xAxis", "dataLabels", "rotation"],
      autoRotatePath: ["style", "layout", "grid", "xAxis", "dataLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      rotation: e.value.style.layout.grid.xAxis.dataLabels.autoRotate.angle,
      width: kl,
      height: wl
    }), ie(e, () => {
      I.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.tooltip.show
      };
    }, { immediate: !0 });
    const ue = h(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = I.value.showTable;
      return {
        component: t ? Kt : Xt,
        title: `${e.value.style.title.text}${e.value.style.title.subtitle.text ? `: ${e.value.style.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: G.value,
          fullscreenParent: N.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: ae.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.backgroundColor,
              color: e.value.style.color
            },
            head: {
              backgroundColor: e.value.style.backgroundColor,
              color: e.value.style.color
            }
          }
        }
      };
    });
    ie(() => I.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && le.value ? le.value.open() : "close" in le.value && le.value.close());
    });
    function $t() {
      I.value.showTable = !1, _e.value && _e.value.setTableIconState(!1);
    }
    const Cl = h(() => e.value.style.backgroundColor), Al = h(() => e.value.style.title), { exportSvg: $l, getSvg: Sl } = ta({
      svg: Te,
      title: Al,
      backgroundColor: Cl
    });
    async function St({ isCb: t }) {
      ve.value = !0, await He();
      try {
        if (t) {
          const { blob: a, url: l, text: o, dataUrl: s } = await Sl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: o, dataUrl: s }));
        } else
          await Promise.resolve($l());
      } finally {
        ve.value = !1;
      }
    }
    function _l(t) {
      et.value = t;
    }
    function zl(t) {
      se.value || he.value || t !== r.value.start && (r.value.start = t, C.value.start = t, de());
    }
    function Ll(t) {
      if (se.value || he.value) return;
      const a = kt(t);
      a !== r.value.end && (r.value.end = a, C.value.end = a, de());
    }
    function Tl({ seriesIndex: t, datapoint: a }) {
      const l = r.value.start + t;
      ke("selectX", {
        dataset: a,
        index: l,
        indexLabel: ""
      });
    }
    function Il(t) {
      if (t?.stage === "start") {
        re.value = !0;
        return;
      }
      if (t?.stage === "end") {
        re.value = !1;
        return;
      }
      st();
    }
    async function _t() {
      if (ke("copyAlt", {
        config: e.value,
        dataset: w.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: w.value
      }));
    }
    return Vt({
      getImage: xl,
      generatePdf: ot,
      generateCsv: Be,
      generateImage: st,
      generateSvg: St,
      toggleTable: Ct,
      toggleTooltip: At,
      toggleAnnotator: Ne,
      toggleFullscreen: wt,
      copyAlt: _t
    }), (t, a) => (u(), i("div", {
      ref_key: "candlestickChart",
      ref: N,
      class: V(`vue-data-ui-component vue-ui-candlestick ${G.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: Z(`position:relative;font-family:${e.value.style.fontFamily}; text-align:center;background:${e.value.style.backgroundColor}; ${e.value.responsive ? "height: 100%" : ""}`),
      id: `vue-ui-candlestick_${R.value}`,
      onMouseenter: ll,
      onMouseleave: al
    }, [
      e.value.userOptions.buttons.annotator ? (u(), K(y(jt), {
        key: 0,
        svgRef: y(Te),
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        active: fe.value,
        isCursorPointer: ae.value,
        onClose: Ne
      }, {
        "annotator-action-close": f(() => [
          g(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": f(({ color: l }) => [
          g(t.$slots, "annotator-action-color", z(L({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": f(({ mode: l }) => [
          g(t.$slots, "annotator-action-draw", z(L({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": f(({ disabled: l }) => [
          g(t.$slots, "annotator-action-undo", z(L({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": f(({ disabled: l }) => [
          g(t.$slots, "annotator-action-redo", z(L({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": f(({ disabled: l }) => [
          g(t.$slots, "annotator-action-delete", z(L({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : m("", !0),
      ol.value ? (u(), i("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Je,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : m("", !0),
      e.value.style.title.text ? (u(), i("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Ge,
        style: "width:100%;background:transparent"
      }, [
        (u(), K(ia, {
          key: `title_${Ze.value}`,
          config: {
            title: {
              cy: "candlestick-div-title",
              ...e.value.style.title
            },
            subtitle: {
              cy: "candlestick-div-subtitle",
              ...e.value.style.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : m("", !0),
      e.value.userOptions.show && we.value && (y(lt) || y(Le)) ? (u(), K(y(Gt), {
        ref_key: "userOptionsRef",
        ref: _e,
        key: `user_options_${je.value}`,
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        isImaging: y(Me),
        isPrinting: y(Fe),
        uid: R.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: G.value,
        isTooltip: I.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: N.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: fe.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: ae.value,
        onToggleFullscreen: wt,
        onGeneratePdf: y(ot),
        onGenerateCsv: Be,
        onGenerateImage: Il,
        onGenerateSvg: St,
        onToggleTable: Ct,
        onToggleTooltip: At,
        onToggleAnnotator: Ne,
        onCopyAlt: _t,
        style: Z({
          visibility: y(lt) ? y(Le) ? "visible" : "hidden" : "visible"
        })
      }, It({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: f(({ isOpen: l, color: o }) => [
            g(t.$slots, "menuIcon", z(L({ isOpen: l, color: o })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: f(() => [
            g(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: f(() => [
            g(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: f(() => [
            g(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: f(() => [
            g(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: f(() => [
            g(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: f(() => [
            g(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: f(({ toggleFullscreen: l, isFullscreen: o }) => [
            g(t.$slots, "optionFullscreen", z(L({ toggleFullscreen: l, isFullscreen: o })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: f(({ toggleAnnotator: l, isAnnotator: o }) => [
            g(t.$slots, "optionAnnotator", z(L({ toggleAnnotator: l, isAnnotator: o })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: f(({ altCopy: l }) => [
            g(t.$slots, "optionAltCopy", z(L({ altCopy: l })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : m("", !0),
      (u(), i("svg", {
        ref_key: "svgRef",
        ref: Te,
        xmlns: y(Ul),
        "aria-label": e.value.style.title.text || "candlestick chart",
        class: V({ "vue-data-ui-fullscreen--on": G.value, "vue-data-ui-fulscreen--off": !G.value }),
        viewBox: `0 0 ${b.value.width <= 0 ? 10 : b.value.width} ${b.value.height <= 0 ? 10 : b.value.height}`,
        style: Z(`max-width:100%;overflow:visible;background:transparent;color:${e.value.style.color}`)
      }, [
        Xe(y(Yt)),
        t.$slots["chart-background"] ? (u(), i("foreignObject", {
          key: 0,
          x: n.value.left,
          y: n.value.top,
          width: Math.max(0.1, n.value.width),
          height: Math.max(0.1, n.value.height),
          style: {
            pointerEvents: "none"
          }
        }, [
          g(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, ya)) : m("", !0),
        w.value.length > 0 ? (u(), i("g", ha, [
          p("defs", null, [
            p("linearGradient", {
              id: `bearish_gradient_${R.value}`,
              x2: "0%",
              y2: "100%"
            }, [
              p("stop", {
                offset: "0%",
                "stop-color": e.value.style.layout.candle.colors.bearish
              }, null, 8, pa),
              p("stop", {
                offset: "50%",
                "stop-color": `${y(xe)(e.value.style.layout.candle.colors.bearish, 0.02)}DE`
              }, null, 8, fa),
              p("stop", {
                offset: "100%",
                "stop-color": `${y(xe)(e.value.style.layout.candle.colors.bearish, 0.05)}66`
              }, null, 8, ga)
            ], 8, ma),
            p("linearGradient", {
              id: `bullish_gradient_${R.value}`,
              x2: "0%",
              y2: "100%"
            }, [
              p("stop", {
                offset: "0%",
                "stop-color": e.value.style.layout.candle.colors.bullish
              }, null, 8, xa),
              p("stop", {
                offset: "50%",
                "stop-color": `${y(xe)(e.value.style.layout.candle.colors.bullish, 0.02)}DE`
              }, null, 8, ka),
              p("stop", {
                offset: "100%",
                "stop-color": `${y(xe)(e.value.style.layout.candle.colors.bullish, 0.05)}66`
              }, null, 8, wa)
            ], 8, ba)
          ]),
          e.value.style.layout.grid.show ? (u(), i("g", Ca, [
            p("line", {
              x1: n.value.left,
              x2: n.value.left,
              y1: n.value.top,
              y2: n.value.bottom,
              stroke: e.value.style.layout.grid.stroke,
              "stroke-width": e.value.style.layout.grid.strokeWidth,
              "stroke-linecap": "round"
            }, null, 8, Aa),
            p("line", {
              x1: n.value.left,
              x2: n.value.right,
              y1: n.value.bottom,
              y2: n.value.bottom,
              stroke: e.value.style.layout.grid.stroke,
              "stroke-width": e.value.style.layout.grid.strokeWidth,
              "stroke-linecap": "round"
            }, null, 8, $a),
            e.value.style.layout.grid.horizontalLines.show ? (u(!0), i(T, { key: 0 }, M(ct.value, (l) => (u(), i("line", {
              x1: n.value.left,
              x2: n.value.right,
              y1: l.y,
              y2: l.y,
              stroke: e.value.style.layout.grid.horizontalLines.stroke,
              "stroke-width": e.value.style.layout.grid.horizontalLines.strokeWidth,
              "stroke-dasharray": e.value.style.layout.grid.horizontalLines.strokeDasharray,
              "stroke-linecap": "round"
            }, null, 8, Sa))), 256)) : m("", !0),
            e.value.style.layout.grid.verticalLines.show ? (u(!0), i(T, { key: 1 }, M(Oe.value, (l, o) => (u(), i("g", null, [
              l.text ? (u(), i("line", {
                key: 0,
                x1: n.value.left + c.value * o + c.value / 2,
                x2: n.value.left + c.value * o + c.value / 2,
                y1: n.value.top,
                y2: n.value.bottom,
                stroke: e.value.style.layout.grid.verticalLines.stroke,
                "stroke-width": e.value.style.layout.grid.verticalLines.strokeWidth,
                "stroke-dasharray": e.value.style.layout.grid.verticalLines.strokeDasharray,
                "stroke-linecap": "round"
              }, null, 8, _a)) : m("", !0)
            ]))), 256)) : m("", !0),
            e.value.style.layout.grid.xAxis.ticks.show ? (u(!0), i(T, { key: 2 }, M(Oe.value, (l, o) => (u(), i("g", null, [
              l.text ? (u(), i("line", {
                key: 0,
                x1: n.value.left + c.value * o + c.value / 2,
                x2: n.value.left + c.value * o + c.value / 2,
                y1: n.value.bottom,
                y2: n.value.bottom + 3,
                stroke: e.value.style.layout.grid.stroke,
                "stroke-width": e.value.style.layout.grid.strokeWidth,
                "stroke-linecap": "round"
              }, null, 8, za)) : m("", !0)
            ]))), 256)) : m("", !0)
          ])) : m("", !0),
          e.value.style.layout.grid.yAxis.dataLabels.show ? (u(), i("g", {
            key: 1,
            ref_key: "scaleLabels",
            ref: Se
          }, [
            (u(!0), i(T, null, M(ct.value, (l, o) => (u(), i("g", null, [
              l.value >= P.value.min && l.value <= P.value.max ? (u(), i("line", {
                key: 0,
                x1: n.value.left,
                x2: n.value.left - 5,
                y1: l.y,
                y2: l.y,
                stroke: e.value.style.layout.grid.stroke,
                "stroke-width": e.value.style.layout.grid.strokeWidth,
                "stroke-linecap": "round"
              }, null, 8, La)) : m("", !0),
              l.value >= P.value.min && l.value <= P.value.max ? (u(), i("text", {
                key: 1,
                x: n.value.left - 8 + e.value.style.layout.grid.yAxis.dataLabels.offsetX,
                y: l.y + b.value.yAxisFontSize / 3,
                "font-size": b.value.yAxisFontSize,
                "text-anchor": "end",
                fill: e.value.style.layout.grid.yAxis.dataLabels.color,
                "font-weight": e.value.style.layout.grid.yAxis.dataLabels.bold ? "bold" : "normal"
              }, ne(y(W)({
                p: e.value.style.layout.grid.yAxis.dataLabels.prefix,
                v: l.value,
                s: e.value.style.layout.grid.yAxis.dataLabels.suffix,
                r: e.value.style.layout.grid.yAxis.dataLabels.roundingValue
              })), 9, Ta)) : m("", !0)
            ]))), 256))
          ], 512)) : m("", !0),
          e.value.style.layout.grid.xAxis.dataLabels.show && !e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable ? (u(), i("g", {
            key: 2,
            ref_key: "timeLabelsEls",
            ref: te
          }, [
            (u(!0), i(T, null, M(cl.value, (l, o) => (u(), i("g", null, [
              p("text", {
                class: "vue-data-ui-time-label",
                transform: `translate(${n.value.left + c.value * o + c.value / 2}, ${n.value.bottom + b.value.xAxisFontSize * 1.5}), rotate(${e.value.style.layout.grid.xAxis.dataLabels.rotation})`,
                "text-anchor": e.value.style.layout.grid.xAxis.dataLabels.rotation > 0 ? "start" : e.value.style.layout.grid.xAxis.dataLabels.rotation < 0 ? "end" : "middle",
                "font-size": b.value.xAxisFontSize,
                fill: e.value.style.layout.grid.xAxis.dataLabels.color,
                "font-weight": e.value.style.layout.grid.xAxis.dataLabels.bold ? "bold" : "normal"
              }, ne(l), 9, Ia)
            ]))), 256))
          ], 512)) : m("", !0),
          e.value.style.layout.grid.xAxis.dataLabels.show && e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable ? (u(), i("g", {
            key: 3,
            ref_key: "timeLabelsEls",
            ref: te
          }, [
            (u(!0), i(T, null, M(Oe.value, (l, o) => (u(), i("g", null, [
              p("text", {
                class: "vue-data-ui-time-label",
                transform: `translate(${n.value.left + c.value * o + c.value / 2}, ${n.value.bottom + b.value.xAxisFontSize * 1.5}), rotate(${e.value.style.layout.grid.xAxis.dataLabels.rotation})`,
                "text-anchor": e.value.style.layout.grid.xAxis.dataLabels.rotation > 0 ? "start" : e.value.style.layout.grid.xAxis.dataLabels.rotation < 0 ? "end" : "middle",
                "font-size": b.value.xAxisFontSize,
                fill: e.value.style.layout.grid.xAxis.dataLabels.color,
                "font-weight": e.value.style.layout.grid.xAxis.dataLabels.bold ? "bold" : "normal"
              }, ne(l?.text ?? ""), 9, Fa)
            ]))), 256))
          ], 512)) : m("", !0),
          e.value.type === "candlestick" ? (u(), i(T, { key: 4 }, [
            p("g", null, [
              (u(!0), i(T, null, M(w.value, (l, o) => (u(), i("g", null, [
                p("rect", {
                  x: l.open.x - e.value.style.layout.wick.strokeWidth / 2,
                  y: l.high.y,
                  width: e.value.style.layout.wick.strokeWidth,
                  height: Math.abs(l.high.y - l.low.y),
                  fill: e.value.style.layout.wick.stroke,
                  stroke: "none",
                  rx: e.value.style.layout.wick.strokeWidth / 2,
                  class: V({ "vue-data-ui-transition": O.value && !y(E) })
                }, null, 10, Ma),
                e.value.style.layout.wick.extremity.shape === "circle" ? (u(), i("g", Ra, [
                  p("circle", {
                    cx: l.high.x,
                    cy: l.high.y,
                    r: e.value.style.layout.wick.extremity.size === "auto" ? c.value / 20 : e.value.style.layout.wick.extremity.size,
                    fill: e.value.style.layout.wick.extremity.color,
                    class: V({ "vue-data-ui-transition": O.value && !y(E) })
                  }, null, 10, Pa),
                  p("circle", {
                    cx: l.low.x,
                    cy: l.low.y,
                    r: e.value.style.layout.wick.extremity.size === "auto" ? c.value / 20 : e.value.style.layout.wick.extremity.size,
                    fill: e.value.style.layout.wick.extremity.color,
                    class: V({ "vue-data-ui-transition": O.value && !y(E) })
                  }, null, 10, Oa)
                ])) : m("", !0),
                e.value.style.layout.wick.extremity.shape === "line" ? (u(), i("g", Da, [
                  p("rect", {
                    x: l.high.x - (e.value.style.layout.wick.extremity.size === "auto" ? c.value * e.value.style.layout.candle.widthRatio : e.value.style.layout.wick.extremity.size) / 2,
                    y: l.high.y - e.value.style.layout.wick.strokeWidth / 2,
                    width: Math.abs(l.high.x - (e.value.style.layout.wick.extremity.size === "auto" ? c.value * e.value.style.layout.candle.widthRatio : e.value.style.layout.wick.extremity.size) / 2 - (l.high.x + (e.value.style.layout.wick.extremity.size === "auto" ? c.value * e.value.style.layout.candle.widthRatio : e.value.style.layout.wick.extremity.size) / 2)),
                    height: e.value.style.layout.wick.strokeWidth,
                    rx: e.value.style.layout.wick.strokeWidth / 2,
                    fill: e.value.style.layout.wick.extremity.color,
                    stroke: "none",
                    class: V({ "vue-data-ui-transition": O.value && !y(E) })
                  }, null, 10, Ba),
                  p("rect", {
                    x: l.low.x - (e.value.style.layout.wick.extremity.size === "auto" ? c.value * e.value.style.layout.candle.widthRatio : e.value.style.layout.wick.extremity.size) / 2,
                    y: l.low.y - e.value.style.layout.wick.strokeWidth / 2,
                    width: Math.abs(l.low.x - (e.value.style.layout.wick.extremity.size === "auto" ? c.value * e.value.style.layout.candle.widthRatio : e.value.style.layout.wick.extremity.size) / 2 - (l.low.x + (e.value.style.layout.wick.extremity.size === "auto" ? c.value * e.value.style.layout.candle.widthRatio : e.value.style.layout.wick.extremity.size) / 2)),
                    height: e.value.style.layout.wick.strokeWidth,
                    fill: e.value.style.layout.wick.extremity.color,
                    stroke: "none",
                    rx: e.value.style.layout.wick.strokeWidth / 2,
                    class: V({ "vue-data-ui-transition": O.value && !y(E) })
                  }, null, 10, Na)
                ])) : m("", !0)
              ]))), 256))
            ]),
            p("g", null, [
              (u(!0), i(T, null, M(w.value, (l, o) => (u(), i("rect", {
                x: l.open.x - c.value / 2 + c.value * (1 - e.value.style.layout.candle.widthRatio) / 2,
                y: l.isBullish ? l.close.y : l.open.y,
                height: Math.abs(l.close.y - l.open.y) <= 0 ? 1e-4 : Math.abs(l.close.y - l.open.y),
                width: c.value * e.value.style.layout.candle.widthRatio <= 0 ? 1e-4 : c.value * e.value.style.layout.candle.widthRatio,
                fill: e.value.style.layout.candle.gradient.underlayer,
                rx: e.value.style.layout.candle.borderRadius,
                stroke: "none",
                class: V({ "vue-data-ui-transition": O.value && !y(E) })
              }, null, 10, Ea))), 256)),
              (u(!0), i(T, null, M(w.value, (l, o) => (u(), i("rect", {
                x: l.open.x - c.value / 2 + c.value * (1 - e.value.style.layout.candle.widthRatio) / 2,
                y: l.isBullish ? l.close.y : l.open.y,
                height: Math.abs(l.close.y - l.open.y) <= 0 ? 1e-4 : Math.abs(l.close.y - l.open.y),
                width: c.value * e.value.style.layout.candle.widthRatio <= 0 ? 1e-4 : c.value * e.value.style.layout.candle.widthRatio,
                fill: l.isBullish ? e.value.style.layout.candle.gradient.show ? `url(#bullish_gradient_${R.value})` : e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.gradient.show ? `url(#bearish_gradient_${R.value})` : e.value.style.layout.candle.colors.bearish,
                rx: e.value.style.layout.candle.borderRadius,
                stroke: e.value.style.layout.candle.stroke,
                "stroke-width": e.value.style.layout.candle.strokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                class: V({ "vue-data-ui-transition": O.value && !y(E) })
              }, null, 10, Va))), 256))
            ])
          ], 64)) : m("", !0),
          e.value.type === "ohlc" ? (u(!0), i(T, { key: 5 }, M(w.value, (l, o) => (u(), i("g", null, [
            p("path", {
              d: `M ${l.high.x},${l.high.y} ${l.low.x},${l.low.y} M${l.open.x - Math.min(6, c.value / 3)},${l.open.y} ${l.open.x},${l.open.y} M${l.close.x},${l.close.y} ${l.close.x + Math.min(6, c.value / 3)},${l.close.y}`,
              stroke: l.isBullish ? e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.colors.bearish,
              "stroke-width": 1
            }, null, 8, Wa)
          ]))), 256)) : m("", !0),
          p("g", null, [
            (u(!0), i(T, null, M(w.value, (l, o) => (u(), i("rect", {
              x: n.value.left + o * c.value,
              y: n.value.top,
              height: n.value.height <= 0 ? 1e-4 : n.value.height,
              width: c.value <= 0 ? 1e-4 : c.value,
              fill: B.value === o || et.value === o ? y(Hl)(e.value.style.layout.selector.color, e.value.style.layout.selector.opacity) : "transparent",
              onMouseover: () => gl(o, l),
              onMouseleave: () => fl(o, l),
              onClick: () => pl(o, l)
            }, null, 40, Ua))), 256))
          ])
        ])) : m("", !0),
        it.value ? (u(), i("rect", Ft({ key: 2 }, il.value, {
          "data-start": r.value.start,
          "data-end": r.value.end
        }), null, 16, Ha)) : m("", !0),
        g(t.$slots, "svg", {
          svg: {
            ...b.value,
            data: w.value,
            drawingArea: n.value,
            isPrintingImg: y(Fe) | y(Me) | re.value,
            isPrintingSvg: ve.value
          }
        }, void 0, !0)
      ], 14, ca)),
      t.$slots.watermark ? (u(), i("div", Xa, [
        g(t.$slots, "watermark", z(L({ isPrinting: y(Fe) || y(Me) || re.value || ve.value })), void 0, !0)
      ])) : m("", !0),
      e.value.style.zoom.show && A.value > 6 && we.value && xt.value ? (u(), K(ra, {
        key: 5,
        ref_key: "chartSlicer",
        ref: ee,
        allMinimaps: vl.value,
        background: e.value.style.zoom.color,
        borderColor: e.value.style.backgroundColor,
        customFormat: e.value.style.zoom.customFormat,
        cutNullValues: !1,
        enableRangeHandles: e.value.style.zoom.enableRangeHandles,
        enableSelectionDrag: e.value.style.zoom.enableSelectionDrag,
        end: r.value.end,
        focusOnDrag: e.value.style.zoom.focusOnDrag,
        focusRangeRatio: e.value.style.zoom.focusRangeRatio,
        fontSize: e.value.style.zoom.fontSize,
        immediate: !e.value.style.zoom.preview.enable,
        inputColor: e.value.style.zoom.color,
        isPreview: it.value,
        labelLeft: gt.value.start || "",
        labelRight: gt.value.end || "",
        max: A.value,
        min: 0,
        minimap: e.value.style.zoom.minimap.show ? y(x).map((l) => l[2]) : [],
        minimapCompact: e.value.style.zoom.minimap.compact,
        minimapFrameColor: e.value.style.zoom.minimap.frameColor,
        minimapIndicatorColor: e.value.style.zoom.minimap.indicatorColor,
        minimapMerged: !1,
        minimapSelectedColor: e.value.style.zoom.minimap.selectedColor,
        minimapSelectedColorOpacity: e.value.style.zoom.minimap.selectedColorOpacity,
        minimapSelectedIndex: B.value,
        minimapSelectionRadius: 1,
        preciseLabels: ml.value,
        refreshEndPoint: e.value.style.zoom.endIndex !== null ? e.value.style.zoom.endIndex + 1 : A.value,
        refreshStartPoint: e.value.style.zoom.startIndex !== null ? e.value.style.zoom.startIndex : 0,
        selectColor: e.value.style.zoom.highlightColor,
        selectedSeries: y(x),
        smoothMinimap: !1,
        start: r.value.start,
        textColor: e.value.style.color,
        timeLabels: Pe.value,
        usePreciseLabels: e.value.style.layout.grid.xAxis.dataLabels.datetimeFormatter.enable && !e.value.style.zoom.useDefaultFormat,
        useResetSlot: e.value.style.zoom.useResetSlot,
        valueEnd: r.value.end,
        valueStart: r.value.start,
        verticalHandles: e.value.style.zoom.minimap.verticalHandles,
        minScale: e.value.style.layout.grid.yAxis.scale.min,
        maxScale: e.value.style.layout.grid.yAxis.scale.max,
        maxWidth: e.value.style.zoom.maxWidth,
        additionalMinimapHeight: e.value.style.zoom.minimap.additionalHeight,
        handleType: e.value.style.zoom.minimap.handleType,
        handleIconColor: e.value.style.zoom.minimap.handleIconColor,
        handleBorderWidth: e.value.style.zoom.minimap.handleBorderWidth,
        handleBorderColor: e.value.style.zoom.minimap.handleBorderColor,
        handleFill: e.value.style.zoom.minimap.handleFill,
        handleWidth: e.value.style.zoom.minimap.handleWidth,
        "onUpdate:end": Ll,
        "onUpdate:start": zl,
        onTrapMouse: _l,
        onReset: bt,
        onFutureEnd: a[0] || (a[0] = (l) => nt("end", l)),
        onFutureStart: a[1] || (a[1] = (l) => nt("start", l))
      }, {
        "reset-action": f(({ reset: l }) => [
          g(t.$slots, "reset-action", z(L({ reset: l })), void 0, !0)
        ]),
        slotMap: f(({ height: l, unitW: o }) => [
          (u(!0), i(T, null, M(rl.value({ minimapH: l, unitW: o }), (s, d) => (u(), i("g", null, [
            p("path", {
              d: `M ${s.high.x},${s.high.y} ${s.low.x},${s.low.y}`,
              stroke: s.isBullish ? e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.colors.bearish,
              "stroke-width": 1,
              style: Z({
                opacity: d >= C.value.start && d <= C.value.end ? 1 : 0.6
              })
            }, null, 12, qa),
            p("path", {
              d: `M ${s.open.x},${s.open.y} ${s.close.x},${s.close.y}`,
              stroke: s.isBullish ? e.value.style.layout.candle.colors.bullish : e.value.style.layout.candle.colors.bearish,
              "stroke-width": Math.min(6, o / 1.5),
              style: Z({
                opacity: d >= C.value.start && d <= C.value.end ? 1 : 0.6
              })
            }, null, 12, ja)
          ]))), 256))
        ]),
        _: 3
      }, 8, ["allMinimaps", "background", "borderColor", "customFormat", "enableRangeHandles", "enableSelectionDrag", "end", "focusOnDrag", "focusRangeRatio", "fontSize", "immediate", "inputColor", "isPreview", "labelLeft", "labelRight", "max", "minimap", "minimapCompact", "minimapFrameColor", "minimapIndicatorColor", "minimapSelectedColor", "minimapSelectedColorOpacity", "minimapSelectedIndex", "preciseLabels", "refreshEndPoint", "refreshStartPoint", "selectColor", "selectedSeries", "start", "textColor", "timeLabels", "usePreciseLabels", "useResetSlot", "valueEnd", "valueStart", "verticalHandles", "minScale", "maxScale", "maxWidth", "additionalMinimapHeight", "handleType", "handleIconColor", "handleBorderWidth", "handleBorderColor", "handleFill", "handleWidth"])) : m("", !0),
      p("div", {
        ref_key: "chartLegend",
        ref: Ye
      }, [
        g(t.$slots, "legend", { legend: w.value }, void 0, !0)
      ], 512),
      t.$slots.source ? (u(), i("div", {
        key: 6,
        ref_key: "source",
        ref: Ke,
        dir: "auto"
      }, [
        g(t.$slots, "source", {}, void 0, !0)
      ], 512)) : m("", !0),
      Xe(y(Ht), {
        teleportTo: e.value.style.tooltip.teleportTo,
        show: I.value.showTooltip && Ae.value,
        backgroundColor: e.value.style.tooltip.backgroundColor,
        color: e.value.style.tooltip.color,
        borderRadius: e.value.style.tooltip.borderRadius,
        borderColor: e.value.style.tooltip.borderColor,
        borderWidth: e.value.style.tooltip.borderWidth,
        fontSize: e.value.style.tooltip.fontSize,
        backgroundOpacity: e.value.style.tooltip.backgroundOpacity,
        position: e.value.style.tooltip.position,
        offsetY: e.value.style.tooltip.offsetY,
        parent: N.value,
        content: $e.value,
        isFullscreen: G.value,
        isCustom: e.value.style.tooltip.customFormat && typeof e.value.style.tooltip.customFormat == "function",
        smooth: e.value.style.tooltip.smooth,
        backdropFilter: e.value.style.tooltip.backdropFilter,
        smoothForce: e.value.style.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": f(() => [
          g(t.$slots, "tooltip-before", z(L({ ...De.value })), void 0, !0)
        ]),
        "tooltip-after": f(() => [
          g(t.$slots, "tooltip-after", z(L({ ...De.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      we.value && e.value.userOptions.buttons.table ? (u(), K(Nl(ue.value.component), Ft({ key: 7 }, ue.value.props, {
        ref_key: "tableUnit",
        ref: le,
        onClose: $t
      }), It({
        content: f(() => [
          (u(), K(y(qt), {
            key: `table_${Qe.value}`,
            colNames: pe.value.colNames,
            head: pe.value.head,
            body: pe.value.body,
            config: pe.value.config,
            title: e.value.table.useDialog ? "" : ue.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: ae.value,
            onClose: $t
          }, {
            th: f(({ th: l }) => [
              Mt(ne(l), 1)
            ]),
            td: f(({ td: l }) => [
              p("div", { innerHTML: l }, null, 8, Ga)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: f(() => [
            Mt(ne(ue.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: f(() => [
            p("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[2] || (a[2] = (l) => Be(e.value.userOptions.callbacks.csv)),
              style: Z({ cursor: ae.value ? "pointer" : "default" })
            }, [
              Xe(y(Ut), {
                name: "fileCsv",
                stroke: ue.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : m("", !0),
      g(t.$slots, "skeleton", {}, () => [
        y(E) ? (u(), K(ea, { key: 0 })) : m("", !0)
      ], !0)
    ], 46, da));
  }
}, mo = /* @__PURE__ */ va(Ya, [["__scopeId", "data-v-b191e88c"]]);
export {
  mo as default
};
