// vue-data-ui/dist/vue-ui-heatmap-BEDirv8T.js
import { watchEffect as Ke, nextTick as me, useCssVars as rl, defineAsyncComponent as ve, computed as g, ref as v, shallowRef as vl, toRefs as cl, watch as Se, onBeforeUnmount as At, onMounted as dl, reactive as Lt, openBlock as u, createElementBlock as i, normalizeStyle as A, normalizeClass as pe, createBlock as ge, unref as n, withCtx as x, renderSlot as b, normalizeProps as D, guardReactiveProps as I, createCommentVNode as f, createSlots as Tt, createElementVNode as p, createVNode as $e, Fragment as z, renderList as V, withDirectives as St, createTextVNode as We, toDisplayString as O, vShow as hl, resolveDynamicComponent as yl, mergeProps as fl, withKeys as pl } from "vue";
import { j as Ye, t as $t, o as zt, q as _t, an as be, r as gl, X as bl, W as ze, k as J, l as Y, K as Q, aA as ml, a0 as xl, a1 as wl, T as kl, U as Cl } from "./lib-GwPnJSrY.js";
import { t as _e, u as Al } from "./useResponsive-DfdjqQps.js";
import { u as Ll } from "./useConfig-CIgyKZ6R.js";
import { u as Tl } from "./usePrinter-DiK5R6k-.js";
import { u as Sl, B as $l } from "./BaseScanner-BhhkKVoj.js";
import { u as zl } from "./useSvgExport-ByUukOZt.js";
import { u as qe } from "./useNestedProp-BBT0mbHh.js";
import { u as Ot } from "./useTimeLabels-C-A2uZKi.js";
import { u as _l } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Ol } from "./useTableResponsive-Cbx-ACcP.js";
import { u as Ml } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Mt } from "./useTimeLabelCollider-CIsgDrl9.js";
import { u as Rl } from "./useChartAccessibility-9icAAmYg.js";
import Fl from "./img-BYYzY_QU.js";
import Nl from "./Title-BfyrP-jp.js";
import Pl from "./vue_ui_heatmap-C9u993qE.js";
import Dl from "./vue-ui-accordion-BGHTqIBt.js";
import { _ as Il } from "./_plugin-vue_export-helper-CHgC5LLL.js";
function je({
  elementRef: h,
  callback: B,
  attr: Z,
  earlyReturn: q = !1,
  retryFrames: ee = 12,
  alsoAfterFontsReady: M = !0
}) {
  if (q) return;
  const E = (S) => {
    if (!S) return;
    let L;
    if (typeof S.getBBox == "function")
      try {
        const R = S.getBBox();
        L = Z === "width" ? R.width : R.height;
      } catch {
      }
    if (typeof L != "number" || Number.isNaN(L))
      try {
        const R = S.getBoundingClientRect();
        L = Z === "width" ? R.width : R.height;
      } catch {
        L = void 0;
      }
    typeof L == "number" && !Number.isNaN(L) && B(L);
  };
  Ke((S) => {
    const L = h.value;
    if (!L) return;
    let R = !1;
    (async () => {
      await me();
      for (let T = 0; T < ee; T += 1) {
        if (R) return;
        await new Promise((le) => requestAnimationFrame(le));
        const H = h.value;
        if (!H) return;
        E(H);
      }
      if (M && typeof document < "u" && document.fonts && document.fonts.ready) {
        try {
          await document.fonts.ready;
        } catch {
        }
        !R && h.value && E(h.value);
      }
    })();
    const te = new MutationObserver(() => {
      h.value && requestAnimationFrame(() => {
        h.value && E(h.value);
      });
    });
    te.observe(L, {
      childList: !0,
      subtree: !0,
      characterData: !0,
      attributes: !0,
      attributeFilter: ["transform", "style", "class"]
    });
    let w;
    if (typeof ResizeObserver < "u") {
      const T = L.ownerSVGElement ? L.ownerSVGElement : L;
      w = new ResizeObserver(() => {
        const H = h.value;
        H && E(H);
      }), w.observe(T);
    }
    S(() => {
      R = !0, te.disconnect(), w && w.disconnect();
    });
  }, { flush: "post" });
}
const Hl = {
  mounted(h, B) {
    Rt(h, B.value);
  },
  updated(h, B) {
    Rt(h, B.value);
  }
};
function Rt(h, {
  cellWidth: B,
  cellHeight: Z,
  maxFontSize: q,
  minFontSize: ee,
  index: M,
  reportRotation: E,
  reportHide: S,
  rotateAll: L,
  hideAll: R
}) {
  h.removeAttribute("transform"), h.removeAttribute("visibility");
  let ue = !1, te = !1;
  h.setAttribute("font-size", q);
  const w = h.getComputedTextLength();
  if (w <= B)
    E(M, !1), S(M, !1);
  else {
    const T = Math.floor(q * B / w);
    T >= ee ? (h.setAttribute("font-size", T), E(M, !1), S(M, !1)) : (E(M, !0), S(M, !1), ue = !0);
  }
  if (L && (ue = !0), ue) {
    h.setAttribute("font-size", q);
    const T = h.getBBox(), H = T.x + T.width / 2, le = T.y + T.height / 2;
    h.setAttribute("transform", `rotate(-90 ${H} ${le})`);
    const _ = h.getBBox().width;
    if (_ <= Z)
      S(M, !1);
    else {
      const ne = Math.floor(q * Z / _);
      ne >= ee ? (h.setAttribute("font-size", ne), S(M, !1)) : (te = !0, S(M, !0));
    }
  }
  (R || te) && h.setAttribute("visibility", "hidden");
}
const Xl = ["id"], Vl = ["xmlns", "viewBox"], Bl = ["x", "y", "width", "height"], El = ["x", "y", "height", "width", "fill", "stroke", "stroke-width"], Ul = ["x", "y", "height", "width", "fill", "stroke", "stroke-width"], Gl = ["x", "y", "width", "height", "fill", "stroke", "stroke-width"], Wl = ["x", "y", "width", "height", "fill", "stroke", "stroke-width", "onMouseover", "onMouseout", "onClick"], Yl = ["font-size", "font-weight", "fill", "x", "y"], ql = ["x", "y", "width", "height", "fill", "stroke", "stroke-width"], jl = ["x", "y", "width", "height", "fill", "stroke", "stroke-width"], Kl = ["font-size", "fill", "x", "y", "font-weight"], Jl = ["font-size", "fill", "x", "y", "font-weight"], Ql = ["text-anchor", "font-size", "fill", "font-weight", "transform"], Zl = ["text-anchor", "font-size", "fill", "font-weight", "transform"], ea = { key: 6 }, ta = ["x", "y", "width", "height", "stroke", "stroke-width"], la = { key: 7 }, aa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], oa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width", "stroke-dasharray"], sa = { class: "vue-ui-heatmap-legend-gauge-right" }, ua = ["data-value"], na = { class: "vue-ui-heatmap-gauge-indicator-value" }, ia = {
  key: 4,
  class: "vue-data-ui-watermark"
}, ra = { class: "vue-ui-data-table" }, va = { key: 0 }, ca = ["data-cell"], da = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, ha = ["data-cell"], ya = { style: { display: "flex", "align-items": "center", gap: "5px", "justify-content": "flex-end", width: "100%", "padding-right": "3px" } }, fa = {
  __name: "vue-ui-heatmap",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectDatapoint", "copyAlt"],
  setup(h, { expose: B, emit: Z }) {
    rl((t) => ({
      v9b79bf1a: t.tdo
    }));
    const q = ve(() => import("./Tooltip-DL3QgjZf.js")), ee = ve(() => import("./BaseIcon-CNx8Tzl5.js")), M = ve(() => import("./PenAndPaper-DWV1FMIi.js")), E = ve(() => import("./UserOptions-CrLqEimB.js")), S = ve(() => import("./PackageVersion-DN9d0ME5.js")), L = ve(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_heatmap: R } = Ll(), { isThemeValid: ue, warnInvalidTheme: te } = _l(), w = h, T = Z, H = g({
      get() {
        return !!w.dataset && w.dataset.length;
      },
      set(t) {
        return t;
      }
    }), le = v(Ye()), _ = v(null), ne = v(!1), Oe = v(""), Je = v(void 0), N = v(null), Qe = v(0), Ze = v(null), et = v(0), Ft = v(null), ce = v(null), Me = v(null), xe = v(!1), we = v(!1), tt = v(null), lt = v(null), Nt = v(null), at = v(null), Pt = v(null), Re = v(null), Fe = v(null), Dt = v(null), ot = v(null), ie = vl(null), e = v(Ie()), ke = g(() => e.value.userOptions.useCursorPointer);
    function It() {
      const t = Array(7).fill("_"), o = [], a = 14, s = t.length;
      for (let c = 0; c < s; c += 1) {
        const d = [];
        for (let C = 0; C < a; C += 1)
          d.push(c + C * 2);
        o.push({
          name: `${t[c]}`,
          values: d
        });
      }
      return o;
    }
    const Ht = g(() => $t({
      defaultConfig: {
        table: { show: !1 },
        userOptions: { show: !1 },
        style: {
          backgroundColor: "#99999930",
          layout: {
            cells: {
              colors: {
                hot: "#999999",
                cold: "#CACACA"
              },
              columnTotal: { value: { show: !1 } },
              rowTotal: { value: { show: !1 } },
              value: { show: !1 }
            },
            dataLabels: {
              xAxis: { show: !1 },
              yAxis: { show: !1 }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: Ce, FINAL_DATASET: k, manualLoading: Ne } = Sl({
      ...cl(w),
      FINAL_CONFIG: e,
      prepareConfig: Ie,
      callback: () => {
        Promise.resolve().then(async () => {
          await me(), _.value && ml(_.value, { delta: 0.1, delay: 250 });
        });
      },
      skeletonDataset: w.config?.skeletonDataset ?? It(),
      skeletonConfig: $t({
        defaultConfig: e.value,
        userConfig: Ht.value
      })
    }), { userOptionsVisible: Pe, setUserOptionsVisibility: Xt, keepUserOptionState: st } = Ml({ config: e.value }), { svgRef: De } = Rl({ config: e.value.style.title });
    function ut(t) {
      Xt(t);
    }
    function Ie() {
      const t = qe({
        userConfig: w.config,
        defaultConfig: R
      });
      let o = {};
      const a = t.theme;
      if (a)
        if (!ue.value(t))
          te(t), o = t;
        else {
          const s = qe({
            userConfig: Pl[a] || w.config,
            defaultConfig: t
          });
          o = {
            ...qe({
              userConfig: w.config,
              defaultConfig: s
            })
          };
        }
      else
        o = t;
      return o;
    }
    Se(() => w.config, (t) => {
      Ce.value || (e.value = Ie()), Pe.value = !e.value.userOptions.showOnChartHover, Ve(), et.value += 1, P.value.showTable = e.value.table.show, P.value.showTooltip = e.value.style.tooltip.show, de.value = e.value.style.layout.width, he.value = e.value.style.layout.height;
    }, { deep: !0 }), Se(() => w.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (Ne.value = !1), Ve();
    }, { deep: !0 });
    const { isPrinting: He, isImaging: Xe, generatePdf: nt, generateImage: it } = Tl({
      elementId: `heatmap__${le.value}`,
      fileName: e.value.style.title.text || "vue-ui-heatmap",
      options: e.value.userOptions.print
    }), Vt = g(() => e.value.userOptions.show && !e.value.style.title.text), P = v({
      showTable: e.value.table.show,
      showTooltip: e.value.style.tooltip.show
    });
    Se(e, () => {
      P.value = {
        showTable: e.value.table.show,
        showTooltip: e.value.style.tooltip.show
      };
    }, { immediate: !0 });
    const Bt = g(() => e.value.table.responsiveBreakpoint), j = v(null), rt = g(() => !!e.value.debug);
    function Ve() {
      if (zt(w.dataset) && (_t({
        componentName: "VueUiHeatmap",
        type: "dataset",
        debug: rt.value
      }), H.value = !1, Ne.value = !0), zt(w.dataset) || (Ne.value = e.value.loading), e.value.responsive) {
        const t = _e(() => {
          const { width: o, height: a } = Al({
            chart: _.value,
            title: e.value.style.title.text ? tt.value : null,
            // add legend hirizontal in legend attr
            // add legend vertical in some other attr and modify composable to subtract width
            source: ot.value,
            noTitle: lt.value
          });
          requestAnimationFrame(() => {
            de.value = o, he.value = a;
          });
        });
        j.value && (ie.value && j.value.unobserve(ie.value), j.value.disconnect()), j.value = new ResizeObserver(t), ie.value = _.value.parentNode, j.value.observe(ie.value);
      }
    }
    At(() => {
      j.value && (ie.value && j.value.unobserve(ie.value), j.value.disconnect());
    }), dl(() => {
      Ve();
    });
    const U = g(() => Math.max(...k.value.flatMap((t) => (t.values || []).length))), de = v(e.value.style.layout.width), he = v(e.value.style.layout.height), F = g(() => ({
      width: Math.max(10, de.value),
      height: Math.max(10, he.value)
    })), re = v(0), Et = _e((t) => {
      t !== re.value && (re.value = t);
    }, 100);
    je({ elementRef: Re, callback: Et, attr: "height" });
    const ae = v(0), Ut = _e((t) => {
      t !== ae.value && (ae.value = t);
    }, 100);
    je({ elementRef: at, callback: Ut, attr: "width" });
    const ye = v(0), Gt = _e((t) => {
      t !== ye.value && (ye.value = t);
    }, 100);
    je({ elementRef: Fe, callback: Gt, attr: "height" }), At(() => {
      re.value = 0, ae.value = 0, ye.value = 0;
    });
    const r = g(() => Math.min(F.value.height, F.value.width) / 1e3 * e.value.style.layout.cells.spacing), l = g(() => {
      let t = 0;
      e.value.style.legend.show && (t = e.value.style.legend.width);
      const o = e.value.style.layout.padding, a = e.value.style.layout.dataLabels.xAxis.fontSize / 3, s = e.value.style.layout.dataLabels.xAxis.fontSize / 2, c = F.value.height - o.top - o.bottom - re.value - a, d = F.value.width / 60, C = {
        x: r.value * U.value,
        y: r.value * ((k.value || []).length + 1)
      }, G = F.value.width - o.left - o.right - C.x - d * 2 - 2 - t - ae.value, y = c - d - C.y - ye.value - s - e.value.style.layout.cells.columnTotal.value.offsetY, $ = {
        width: Math.max(3, G / U.value),
        height: Math.max(3, y / (k.value.length ?? 1))
      };
      return {
        top: o.top + re.value + d + a,
        topLabelsHeight: re.value,
        sumCellXHeight: d,
        height: y,
        left: o.left + ae.value + d / 2,
        right: o.right - t,
        bottom: F.value.height - o.bottom - ye.value,
        width: G,
        cellSize: $
      };
    }), K = g(() => Math.max(...k.value.flatMap((t) => t.values))), fe = g(() => Math.min(...k.value.flatMap((t) => t.values))), oe = g(() => {
      const t = k.value.flatMap((a) => a.values);
      return t.reduce((a, s) => a + s, 0) / t.length;
    }), vt = v([]), ct = v([]);
    let dt = 0;
    Ke(() => {
      const t = ++dt;
      (async () => {
        const o = e.value.style.layout.dataLabels.yAxis, a = await Ot({
          values: o.values.length ? o.values : k.value.map((s) => s.name),
          maxDatapoints: k.value.length,
          formatter: o.datetimeFormatter,
          start: 0,
          end: k.value.length
        });
        t === dt && (vt.value = a);
      })();
    });
    let ht = 0;
    Ke(() => {
      const t = ++ht;
      (async () => {
        const o = e.value.style.layout.dataLabels.xAxis, a = await Ot({
          values: o.values,
          maxDatapoints: U.value,
          formatter: o.datetimeFormatter,
          start: 0,
          end: U.value
        });
        t === ht && (ct.value = a);
      })();
    });
    const m = g(() => {
      const t = vt.value.map((y) => y.text), o = ct.value.map((y) => y.text), a = k.value.map((y) => y.values.reduce(($, Ge) => $ + Ge, 0)), s = Math.max(...a), c = Math.min(...a), d = [];
      for (let y = 0; y < U.value; y += 1)
        d.push(k.value.map(($) => $.values[y] || 0).reduce(($, Ge) => $ + Ge, 0));
      const C = Math.max(...d), G = Math.min(...d);
      return {
        yTotals: a.map((y) => {
          const $ = isNaN(y / s) ? 0 : y / s;
          return {
            total: y,
            proportion: $,
            color: be(e.value.style.layout.cells.colors.cold, e.value.style.layout.cells.colors.hot, c, s, y)
          };
        }),
        xTotals: d.map((y) => {
          const $ = isNaN(y / C) ? 0 : y / C;
          return {
            total: y,
            proportion: $,
            color: be(e.value.style.layout.cells.colors.cold, e.value.style.layout.cells.colors.hot, G, C, y)
          };
        }),
        yLabels: t,
        xLabels: o.slice(0, U.value)
      };
    }), W = g(() => (k.value.forEach((t, o) => {
      gl({
        datasetObject: t,
        requiredAttributes: ["values"]
      }).forEach((a) => {
        _t({
          componentName: "VueUiHeatmap",
          type: "datasetSerieAttribute",
          property: "values",
          index: o,
          debug: rt.value
        });
      });
    }), k.value.map((t, o) => ({
      ...t,
      temperatures: (t.values || []).map((a, s) => a >= oe.value ? {
        side: "up",
        color: be(e.value.style.layout.cells.colors.cold, e.value.style.layout.cells.colors.hot, fe.value, K.value, a),
        ratio: Math.abs(Math.abs(a - oe.value) / Math.abs(K.value - oe.value)) > 1 ? 1 : Math.abs(Math.abs(a - oe.value) / Math.abs(K.value - oe.value)),
        value: a,
        yAxisName: m.value.yLabels[o],
        xAxisName: m.value.xLabels[s],
        id: `vue-data-ui-heatmap-cell-${Ye()}`
      } : {
        side: "down",
        ratio: Math.abs(1 - Math.abs(a) / Math.abs(oe.value)) > 1 ? 1 : Math.abs(1 - Math.abs(a) / Math.abs(oe.value)),
        color: be(e.value.style.layout.cells.colors.cold, e.value.style.layout.cells.colors.hot, fe.value, K.value, a),
        value: a,
        yAxisName: m.value.yLabels[o],
        xAxisName: m.value.xLabels[s],
        id: `vue-data-ui-heatmap-cell-${Ye()}`
      })
    })))), yt = g(() => k.value.length), ft = Lt(Array(yt.value * U.value || 1).fill(!1)), Wt = g(() => ft.some((t) => t));
    function Yt(t, o) {
      ft[t] = o;
    }
    const pt = Lt(Array(yt.value * U.value || 1).fill(!1)), qt = g(() => pt.some((t) => t));
    function jt(t, o) {
      pt[t] = o;
    }
    const X = v(null), Ae = v(null);
    function Kt(t, o, a, s) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: o }), N.value = { x: a, y: s };
      const { value: c, yAxisName: d, xAxisName: C, id: G } = t;
      Je.value = G, X.value = c, Ae.value = {
        datapoint: t,
        seriesIndex: o,
        series: W.value,
        config: e.value
      }, ne.value = !0;
      let y = "";
      const $ = e.value.style.tooltip.customFormat;
      kl($) && Cl(() => $({
        datapoint: t,
        seriesIndex: o,
        series: W.value,
        config: e.value
      })) ? Oe.value = $({
        datapoint: t,
        seriesIndex: o,
        series: W.value,
        config: e.value
      }) : (y += `<div>${d} ${C ? d ? ` - ${C}` : `${C}` : ""}</div>`, y += `<div style="margin-top:6px;padding-top:6px;border-top:1px solid ${e.value.style.tooltip.borderColor};font-weight:bold;display:flex;flex-direction:row;gap:12px;align-items:center;justify-content:center"><span style="color:${be(e.value.style.layout.cells.colors.cold, e.value.style.layout.cells.colors.hot, fe.value, K.value, c)}">⬤</span><span>${isNaN(c) ? "-" : J(
        e.value.style.layout.cells.value.formatter,
        c,
        Y({
          p: e.value.style.layout.dataLabels.prefix,
          v: c,
          s: e.value.style.layout.dataLabels.suffix,
          r: e.value.style.tooltip.roundingValue
        }),
        { datapoint: t, seriesIndex: o }
      )}</span></div>`, Oe.value = `<div style="font-size:${e.value.style.tooltip.fontSize}px">${y}</div>`);
    }
    function Jt({ datapoint: t, seriesIndex: o }) {
      e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: o }), ne.value = !1, Je.value = void 0, X.value = null, N.value = null;
    }
    function Qt(t) {
      return J(
        e.value.style.layout.cells.value.formatter,
        m.value.yTotals[t].total,
        Y({
          p: e.value.style.layout.dataLabels.prefix,
          v: m.value.yTotals[t].total,
          s: e.value.style.layout.dataLabels.suffix,
          r: e.value.style.layout.cells.value.roundingValue
        }),
        { datapoint: m.value.yTotals[t], rowIndex: t }
      );
    }
    function Zt(t) {
      return J(
        e.value.style.layout.cells.value.formatter,
        m.value.xTotals[t].total,
        Y({
          p: e.value.style.layout.dataLabels.prefix,
          v: m.value.xTotals[t].total,
          s: e.value.style.layout.dataLabels.suffix,
          r: e.value.style.layout.cells.value.roundingValue
        }),
        { datapoint: m.value.xTotals[t], colIndex: t }
      );
    }
    g(() => {
      const t = k.value.map((a) => ({
        name: a.name
      })), o = k.value.map((a) => a.values);
      return { head: t, body: o };
    });
    function Be(t = null) {
      me(() => {
        const o = ["", ...k.value.map((d, C) => d.name)], a = [];
        for (let d = 0; d < m.value.xLabels.length; d += 1) {
          const C = [m.value.xLabels[d]];
          for (let G = 0; G < k.value.length; G += 1)
            C.push([k.value[G].values[d]]);
          a.push(C);
        }
        const s = [[e.value.style.title.text], [e.value.style.title.subtitle.text], [[""], [""], [""]]].concat([o]).concat(a), c = xl(s);
        t ? t(c) : wl({ csvContent: c, title: e.value.style.title.text || "vue-ui-heatmap" });
      });
    }
    const se = v(!1);
    function gt(t) {
      se.value = t, Qe.value += 1;
    }
    function bt() {
      P.value.showTable = !P.value.showTable;
    }
    function mt() {
      P.value.showTooltip = !P.value.showTooltip;
    }
    const Le = v(!1);
    function Ee() {
      Le.value = !Le.value;
    }
    function el(t, o) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: o }), T("selectDatapoint", t);
    }
    function tl() {
      return W.value;
    }
    async function ll({ scale: t = 2 } = {}) {
      if (!_.value) return;
      const { width: o, height: a } = _.value.getBoundingClientRect(), s = o / a, { imageUri: c, base64: d } = await Fl({ domElement: _.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: c,
        base64: d,
        title: e.value.style.title.text,
        width: o,
        height: a,
        aspectRatio: s
      };
    }
    const xt = g(() => ({
      start: 0,
      end: U.value
    })), wt = g(() => m.value.xLabels);
    Mt({
      timeLabelsEls: Re,
      timeLabels: wt,
      slicer: xt,
      configRef: e,
      rotationPath: ["style", "layout", "dataLabels", "xAxis", "rotation"],
      autoRotatePath: ["style", "layout", "dataLabels", "xAxis", "autoRotate", "enable"],
      isAutoSize: !1,
      targetClass: ".vue-ui-heatmap-col-name",
      rotation: e.value.style.layout.dataLabels.xAxis.autoRotate.angle,
      width: de,
      height: he
    }), Mt({
      timeLabelsEls: Fe,
      timeLabels: wt,
      slicer: xt,
      configRef: e,
      rotationPath: ["style", "layout", "cells", "columnTotal", "value", "rotation"],
      autoRotatePath: ["style", "layout", "cells", "columnTotal", "value", "autoRotate", "enable"],
      isAutoSize: !1,
      targetClass: ".vue-ui-heatmap-col-total",
      rotation: e.value.style.layout.cells.columnTotal.value.autoRotate.angle,
      width: de,
      height: he
    });
    const Te = g(() => {
      const t = e.value.table.useDialog && !e.value.table.show, o = P.value.showTable;
      return {
        component: t ? L : Dl,
        title: `${e.value.style.title.text}${e.value.style.title.subtitle.text ? `: ${e.value.style.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: se.value,
          fullscreenParent: _.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: ke.value
        } : {
          hideDetails: !0,
          config: {
            open: o,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.backgroundColor,
              color: e.value.style.color
            },
            head: {
              backgroundColor: e.value.style.backgroundColor,
              color: e.value.style.color
            }
          }
        }
      };
    });
    Se(() => P.value.showTable, async (t) => {
      e.value.table.show || (t && e.value.table.useDialog && ce.value ? (await me(), ce.value.open()) : "close" in ce.value && ce.value.close());
    });
    const { isResponsive: al } = Ol(Ze, Bt);
    function Ue() {
      P.value.showTable = !1, Me.value && Me.value.setTableIconState(!1);
    }
    const ol = g(() => e.value.style.backgroundColor), sl = g(() => e.value.style.title), { exportSvg: ul, getSvg: nl } = zl({
      svg: De,
      title: sl,
      backgroundColor: ol
    });
    async function kt({ isCb: t }) {
      we.value = !0, await me();
      try {
        if (t) {
          const { blob: o, url: a, text: s, dataUrl: c } = await nl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: o, url: a, text: s, dataUrl: c }));
        } else
          await Promise.resolve(ul());
      } finally {
        we.value = !1;
      }
    }
    function il(t) {
      if (t?.stage === "start") {
        xe.value = !0;
        return;
      }
      if (t?.stage === "end") {
        xe.value = !1;
        return;
      }
      it();
    }
    async function Ct() {
      if (T("copyAlt", {
        config: e.value,
        dataset: W.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: W.value
      }));
    }
    return B({
      getData: tl,
      getImage: ll,
      generatePdf: nt,
      generateCsv: Be,
      generateImage: it,
      generateSvg: kt,
      toggleTable: bt,
      toggleTooltip: mt,
      toggleAnnotator: Ee,
      toggleFullscreen: gt,
      copyAlt: Ct
    }), (t, o) => (u(), i("div", {
      ref_key: "heatmapChart",
      ref: _,
      class: pe(`vue-data-ui-component vue-ui-heatmap ${se.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      style: A(`font-family:${e.value.style.fontFamily};width:100%;${e.value.responsive ? "height: 100%;" : ""} text-align:center;background:${e.value.style.backgroundColor}`),
      id: `heatmap__${le.value}`,
      onMouseenter: o[1] || (o[1] = () => ut(!0)),
      onMouseleave: o[2] || (o[2] = () => ut(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (u(), ge(n(M), {
        key: 0,
        svgRef: n(De),
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        active: Le.value,
        isCursorPointer: ke.value,
        onClose: Ee
      }, {
        "annotator-action-close": x(() => [
          b(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": x(({ color: a }) => [
          b(t.$slots, "annotator-action-color", D(I({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": x(({ mode: a }) => [
          b(t.$slots, "annotator-action-draw", D(I({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": x(({ disabled: a }) => [
          b(t.$slots, "annotator-action-undo", D(I({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": x(({ disabled: a }) => [
          b(t.$slots, "annotator-action-redo", D(I({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": x(({ disabled: a }) => [
          b(t.$slots, "annotator-action-delete", D(I({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : f("", !0),
      Vt.value ? (u(), i("div", {
        key: 1,
        ref_key: "noTitle",
        ref: lt,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : f("", !0),
      e.value.style.title.text ? (u(), i("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: tt,
        style: "width:100%;background:transparent"
      }, [
        (u(), ge(Nl, {
          key: `title_${et.value}`,
          config: {
            title: {
              cy: "heatmap-div-title",
              ...e.value.style.title
            },
            subtitle: {
              cy: "heatmap-div-subtitle",
              ...e.value.style.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : f("", !0),
      e.value.userOptions.show && H.value && (n(st) || n(Pe)) ? (u(), ge(n(E), {
        ref_key: "userOptionsRef",
        ref: Me,
        key: `user_options_${Qe.value}`,
        backgroundColor: e.value.style.backgroundColor,
        color: e.value.style.color,
        isImaging: n(Xe),
        isPrinting: n(He),
        uid: le.value,
        hasTooltip: e.value.userOptions.buttons.tooltip && e.value.style.tooltip.show,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: se.value,
        isTooltip: P.value.showTooltip,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: _.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: Le.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: ke.value,
        onToggleFullscreen: gt,
        onGeneratePdf: n(nt),
        onGenerateCsv: Be,
        onGenerateImage: il,
        onGenerateSvg: kt,
        onToggleTable: bt,
        onToggleTooltip: mt,
        onToggleAnnotator: Ee,
        onCopyAlt: Ct,
        style: A({
          visibility: n(st) ? n(Pe) ? "visible" : "hidden" : "visible"
        })
      }, Tt({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: x(({ isOpen: a, color: s }) => [
            b(t.$slots, "menuIcon", D(I({ isOpen: a, color: s })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: x(() => [
            b(t.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: x(() => [
            b(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: x(() => [
            b(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: x(() => [
            b(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: x(() => [
            b(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: x(() => [
            b(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: x(({ toggleFullscreen: a, isFullscreen: s }) => [
            b(t.$slots, "optionFullscreen", D(I({ toggleFullscreen: a, isFullscreen: s })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: x(({ toggleAnnotator: a, isAnnotator: s }) => [
            b(t.$slots, "optionAnnotator", D(I({ toggleAnnotator: a, isAnnotator: s })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: x(({ altCopy: a }) => [
            b(t.$slots, "optionAltCopy", D(I({ altCopy: a })), void 0, !0)
          ]),
          key: "9"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isImaging", "isPrinting", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "isTooltip", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : f("", !0),
      p("div", {
        class: pe({
          "vue-ui-heatmap-chart-wrapper": !0,
          "vue-ui-heatmap-chart-wrapper-legend-right": e.value.style.legend.show
        })
      }, [
        (u(), i("svg", {
          ref_key: "svgRef",
          ref: De,
          xmlns: n(bl),
          class: pe({ "vue-data-ui-fullscreen--on": se.value, "vue-data-ui-fulscreen--off": !se.value }),
          viewBox: `0 0 ${F.value.width} ${F.value.height}`,
          width: "100%",
          style: A(`overflow: visible; background:transparent;color:${e.value.style.color}`),
          "aria-live": "polite",
          role: "img",
          preserveAspectRatio: "xMidYMid"
        }, [
          $e(n(S)),
          t.$slots["chart-background"] ? (u(), i("foreignObject", {
            key: 0,
            x: l.value.left,
            y: l.value.top,
            width: l.value.width,
            height: l.value.height,
            style: {
              pointerEvents: "none"
            }
          }, [
            b(t.$slots, "chart-background", {}, void 0, !0)
          ], 8, Bl)) : f("", !0),
          e.value.style.layout.cells.columnTotal.color.show ? (u(), i("g", {
            key: 1,
            ref_key: "xAxisSumRects",
            ref: Dt
          }, [
            (u(!0), i(z, null, V(m.value.xTotals, (a, s) => (u(), i("rect", {
              x: l.value.left + l.value.cellSize.width * s + r.value / 2 + l.value.sumCellXHeight,
              y: l.value.top - l.value.sumCellXHeight + r.value * (F.value.height / F.value.width),
              height: l.value.sumCellXHeight,
              width: l.value.cellSize.width - r.value,
              fill: e.value.style.layout.cells.colors.underlayer,
              stroke: e.value.style.backgroundColor,
              "stroke-width": r.value
            }, null, 8, El))), 256)),
            (u(!0), i(z, null, V(m.value.xTotals, (a, s) => (u(), i("rect", {
              x: l.value.left + l.value.cellSize.width * s + r.value / 2 + l.value.sumCellXHeight,
              y: l.value.top - l.value.sumCellXHeight + r.value * (F.value.height / F.value.width),
              height: l.value.sumCellXHeight,
              width: l.value.cellSize.width - r.value,
              fill: a.color,
              stroke: e.value.style.backgroundColor,
              "stroke-width": r.value
            }, null, 8, Ul))), 256))
          ], 512)) : f("", !0),
          p("g", {
            ref_key: "datapoints",
            ref: Ft
          }, [
            (u(!0), i(z, null, V(W.value, (a, s) => (u(), i(z, null, [
              (u(!0), i(z, null, V(a.temperatures, (c, d) => (u(), i("g", null, [
                p("rect", {
                  x: l.value.left + l.value.cellSize.width * d + r.value / 2 + l.value.sumCellXHeight,
                  y: l.value.top + l.value.cellSize.height * s + r.value / 2,
                  width: l.value.cellSize.width - r.value,
                  height: l.value.cellSize.height - r.value,
                  fill: e.value.style.layout.cells.colors.underlayer,
                  stroke: e.value.style.backgroundColor,
                  "stroke-width": r.value
                }, null, 8, Gl),
                p("rect", {
                  x: l.value.left + l.value.cellSize.width * d + r.value / 2 + l.value.sumCellXHeight,
                  y: l.value.top + l.value.cellSize.height * s + r.value / 2,
                  width: l.value.cellSize.width - r.value,
                  height: l.value.cellSize.height - r.value,
                  fill: c.color,
                  stroke: e.value.style.backgroundColor,
                  "stroke-width": r.value,
                  onMouseover: (C) => Kt(c, s, l.value.left + l.value.cellSize.width * d, l.value.top + l.value.cellSize.height * s),
                  onMouseout: () => Jt({ datapoint: c, seriesIndex: s }),
                  onClick: () => el(c, s)
                }, null, 40, Wl),
                e.value.style.layout.cells.value.show ? St((u(), i("text", {
                  key: 0,
                  "text-anchor": "middle",
                  "font-size": e.value.style.layout.cells.value.fontSize,
                  "font-weight": e.value.style.layout.cells.value.bold ? "bold" : "normal",
                  fill: n(ze)(c.color),
                  x: l.value.left + l.value.cellSize.width * d + l.value.cellSize.width / 2 + l.value.sumCellXHeight,
                  y: l.value.top + l.value.cellSize.height * s + l.value.cellSize.height / 2 + e.value.style.layout.cells.value.fontSize / 3,
                  style: { pointerEvents: "none", userSelect: "none" }
                }, [
                  We(O(n(J)(
                    e.value.style.layout.cells.value.formatter,
                    c.value,
                    n(Y)({
                      p: e.value.style.layout.dataLabels.prefix,
                      v: c.value,
                      s: e.value.style.layout.dataLabels.suffix,
                      r: e.value.style.layout.cells.value.roundingValue
                    }),
                    { datapoint: c }
                  )), 1)
                ], 8, Yl)), [
                  [n(Hl), {
                    cellWidth: l.value.cellSize.width - r.value,
                    cellHeight: l.value.cellSize.height - r.value,
                    maxFontSize: e.value.style.layout.cells.value.fontSize,
                    minFontSize: 10,
                    index: s * U.value + d,
                    reportHide: jt,
                    reportRotation: Yt,
                    hideAll: qt.value,
                    rotateAll: Wt.value
                  }]
                ]) : f("", !0)
              ]))), 256))
            ], 64))), 256))
          ], 512),
          e.value.style.layout.cells.rowTotal.color.show ? (u(), i("g", {
            key: 2,
            ref_key: "yAxisSumRects",
            ref: Pt
          }, [
            (u(!0), i(z, null, V(W.value, (a, s) => (u(), i(z, null, [
              p("rect", {
                x: l.value.left,
                y: l.value.top + l.value.cellSize.height * s,
                width: l.value.sumCellXHeight,
                height: l.value.cellSize.height - r.value,
                fill: e.value.style.layout.cells.colors.underlayer,
                stroke: e.value.style.backgroundColor,
                "stroke-width": r.value
              }, null, 8, ql),
              p("rect", {
                x: l.value.left,
                y: l.value.top + l.value.cellSize.height * s + r.value / 2,
                width: l.value.sumCellXHeight,
                height: l.value.cellSize.height - r.value,
                fill: m.value.yTotals[s].color,
                stroke: e.value.style.backgroundColor,
                "stroke-width": r.value
              }, null, 8, jl)
            ], 64))), 256))
          ], 512)) : f("", !0),
          e.value.style.layout.dataLabels.yAxis.show ? (u(), i("g", {
            key: 3,
            ref_key: "yAxisLabels",
            ref: at
          }, [
            (u(!0), i(z, null, V(W.value, (a, s) => (u(), i(z, null, [
              p("text", {
                class: "vue-ui-heatmap-row-name",
                "font-size": e.value.style.layout.dataLabels.yAxis.fontSize,
                fill: e.value.style.layout.dataLabels.yAxis.color,
                x: ae.value,
                y: l.value.top + l.value.cellSize.height * s + l.value.cellSize.height / 2 + e.value.style.layout.dataLabels.yAxis.fontSize / 3 + e.value.style.layout.dataLabels.yAxis.offsetY - (e.value.style.layout.cells.rowTotal.value.show ? e.value.style.layout.dataLabels.yAxis.fontSize / 1.5 : 0),
                "text-anchor": "end",
                "font-weight": e.value.style.layout.dataLabels.yAxis.bold ? "bold" : "normal"
              }, O(m.value.yLabels[s]), 9, Kl),
              e.value.style.layout.cells.rowTotal.value.show ? (u(), i("text", {
                key: 0,
                class: "vue-ui-heatmap-row-total",
                "font-size": e.value.style.layout.dataLabels.yAxis.fontSize,
                fill: e.value.style.layout.dataLabels.yAxis.color,
                x: ae.value,
                y: l.value.top + l.value.cellSize.height * s + l.value.cellSize.height / 2 + e.value.style.layout.dataLabels.yAxis.fontSize + e.value.style.layout.dataLabels.yAxis.offsetY,
                "text-anchor": "end",
                "font-weight": e.value.style.layout.dataLabels.yAxis.bold ? "bold" : "normal"
              }, O(Qt(s)), 9, Jl)) : f("", !0)
            ], 64))), 256))
          ], 512)) : f("", !0),
          e.value.style.layout.dataLabels.xAxis.show ? (u(), i("g", {
            key: 4,
            ref_key: "xAxisLabels",
            ref: Re
          }, [
            (u(!0), i(z, null, V(m.value.xLabels, (a, s) => (u(), i(z, null, [
              !e.value.style.layout.dataLabels.xAxis.showOnlyAtModulo || e.value.style.layout.dataLabels.xAxis.showOnlyAtModulo && s % e.value.style.layout.dataLabels.xAxis.showOnlyAtModulo === 0 ? (u(), i("text", {
                key: 0,
                class: "vue-ui-heatmap-col-name",
                "text-anchor": e.value.style.layout.dataLabels.xAxis.rotation === 0 ? "middle" : e.value.style.layout.dataLabels.xAxis.rotation < 0 ? "start" : "end",
                "font-size": e.value.style.layout.dataLabels.xAxis.fontSize,
                fill: e.value.style.layout.dataLabels.xAxis.color,
                "font-weight": e.value.style.layout.dataLabels.xAxis.bold ? "bold" : "normal",
                transform: `translate(${l.value.left + l.value.cellSize.width / 2 + l.value.width / m.value.xLabels.length * s + e.value.style.layout.dataLabels.xAxis.offsetX + l.value.sumCellXHeight}, ${l.value.topLabelsHeight}), rotate(${e.value.style.layout.dataLabels.xAxis.rotation})`
              }, O(a), 9, Ql)) : f("", !0)
            ], 64))), 256))
          ], 512)) : f("", !0),
          e.value.style.layout.cells.columnTotal.value.show ? (u(), i("g", {
            key: 5,
            ref_key: "xAxisSums",
            ref: Fe
          }, [
            (u(!0), i(z, null, V(m.value.xLabels, (a, s) => (u(), i("text", {
              class: "vue-ui-heatmap-col-total",
              "text-anchor": e.value.style.layout.cells.columnTotal.value.rotation === 0 ? "middle" : e.value.style.layout.cells.columnTotal.value.rotation < 0 ? "end" : "start",
              "font-size": e.value.style.layout.dataLabels.xAxis.fontSize,
              fill: e.value.style.layout.dataLabels.xAxis.color,
              "font-weight": e.value.style.layout.dataLabels.xAxis.bold ? "bold" : "normal",
              transform: `translate(${l.value.left + l.value.cellSize.width / 2 + l.value.width / m.value.xLabels.length * s + e.value.style.layout.dataLabels.xAxis.offsetX + e.value.style.layout.cells.columnTotal.value.offsetX + l.value.sumCellXHeight}, ${l.value.bottom + e.value.style.layout.dataLabels.xAxis.fontSize / 2}), rotate(${e.value.style.layout.cells.columnTotal.value.rotation})`
            }, O(Zt(s)), 9, Zl))), 256))
          ], 512)) : f("", !0),
          N.value ? (u(), i("g", ea, [
            p("rect", {
              style: { "pointer-events": "none" },
              x: N.value.x - e.value.style.layout.cells.selected.border / 2 + r.value + l.value.sumCellXHeight,
              y: N.value.y - e.value.style.layout.cells.selected.border / 2 + r.value,
              width: l.value.cellSize.width - r.value + e.value.style.layout.cells.selected.border - r.value,
              height: l.value.cellSize.height - r.value + e.value.style.layout.cells.selected.border - r.value,
              fill: "transparent",
              stroke: e.value.style.layout.cells.selected.color,
              "stroke-width": e.value.style.layout.cells.selected.border,
              rx: 1
            }, null, 8, ta)
          ])) : f("", !0),
          e.value.style.layout.crosshairs.show && N.value ? (u(), i("g", la, [
            p("line", {
              x1: l.value.left + l.value.sumCellXHeight,
              x2: N.value.x + l.value.sumCellXHeight,
              y1: N.value.y + (l.value.cellSize.height - r.value) / 2,
              y2: N.value.y + (l.value.cellSize.height - r.value) / 2,
              stroke: e.value.style.layout.crosshairs.stroke,
              "stroke-width": e.value.style.layout.crosshairs.strokeWidth,
              "stroke-dasharray": e.value.style.layout.crosshairs.strokeDasharray,
              "stroke-linecap": "round"
            }, null, 8, aa),
            p("line", {
              x1: N.value.x + l.value.sumCellXHeight + (l.value.cellSize.width - r.value) / 2,
              x2: N.value.x + l.value.sumCellXHeight + (l.value.cellSize.width - r.value) / 2,
              y1: N.value.y,
              y2: l.value.top,
              stroke: e.value.style.layout.crosshairs.stroke,
              "stroke-width": e.value.style.layout.crosshairs.strokeWidth,
              "stroke-dasharray": e.value.style.layout.crosshairs.strokeDasharray,
              "stroke-linecap": "round"
            }, null, 8, oa)
          ])) : f("", !0),
          b(t.$slots, "svg", {
            svg: {
              ...F.value,
              isPrintingImg: n(He) | n(Xe) | xe.value,
              isPrintingSvg: we.value
            }
          }, void 0, !0)
        ], 14, Vl)),
        e.value.style.legend.show ? (u(), i("div", {
          key: 0,
          ref_key: "legendRight",
          ref: Nt,
          class: "vue-ui-heatmap-legend-right",
          style: A({
            "--legend-width": e.value.style.legend.width + "px"
          })
        }, [
          n(Ce) ? f("", !0) : (u(), i("div", {
            key: 0,
            class: "vue-ui-heatmap-legend-label-max",
            style: A({
              fontSize: e.value.style.legend.fontSize + "px",
              color: e.value.style.legend.color
            })
          }, O(n(J)(
            e.value.style.layout.cells.value.formatter,
            n(Q)(K.value),
            n(Y)({
              p: e.value.style.layout.dataLabels.prefix,
              v: n(Q)(K.value),
              s: e.value.style.layout.dataLabels.suffix,
              r: e.value.style.legend.roundingValue
            })
          )), 5)),
          p("div", sa, [
            p("div", {
              class: "vue-ui-heatmap-gauge",
              style: A({
                background: `linear-gradient(to bottom, ${e.value.style.layout.cells.colors.hot}, ${e.value.style.layout.cells.colors.cold})`
              })
            }, [
              St(p("div", {
                class: "vue-ui-heatmap-gauge-indicator",
                "data-value": n(J)(
                  e.value.style.layout.cells.value.formatter,
                  n(Q)(X.value),
                  n(Y)({
                    p: e.value.style.layout.dataLabels.prefix,
                    v: n(Q)(X.value),
                    s: e.value.style.layout.dataLabels.suffix,
                    r: e.value.style.legend.roundingValue
                  })
                ),
                style: A({
                  position: "absolute",
                  width: "100%",
                  height: "2px",
                  background: [void 0, null].includes(X.value) ? "transparent" : n(ze)(Ae.value.datapoint.color),
                  top: `${[void 0, null].includes(X.value) ? 0 : (1 - X.value / K.value) * 100}%`,
                  transition: "all 0.2s ease-in-out",
                  "--background-color": e.value.style.backgroundColor,
                  "--gauge-arrow-color": n(ze)(e.value.style.backgroundColor),
                  "--gauge-arrow-text-color": n(ze)(e.value.style.backgroundColor),
                  "--gauge-arrow-value": X.value,
                  "--gauge-arrow-font-size": e.value.style.legend.fontSize + "px"
                })
              }, [
                p("div", na, O(n(J)(
                  e.value.style.layout.cells.value.formatter,
                  n(Q)(X.value),
                  n(Y)({
                    p: e.value.style.layout.dataLabels.prefix,
                    v: n(Q)(X.value),
                    s: e.value.style.layout.dataLabels.suffix,
                    r: e.value.style.legend.roundingValue
                  })
                )), 1)
              ], 12, ua), [
                [hl, ![void 0, null].includes(X.value)]
              ])
            ], 4)
          ]),
          n(Ce) ? f("", !0) : (u(), i("div", {
            key: 1,
            class: "vue-ui-heatmap-legend-label-min",
            style: A({
              fontSize: e.value.style.legend.fontSize + "px",
              color: e.value.style.legend.color
            })
          }, O(n(J)(
            e.value.style.layout.cells.value.formatter,
            n(Q)(fe.value),
            n(Y)({
              p: e.value.style.layout.dataLabels.prefix,
              v: n(Q)(fe.value),
              s: e.value.style.layout.dataLabels.suffix,
              r: e.value.style.legend.roundingValue
            })
          )), 5))
        ], 4)) : f("", !0)
      ], 2),
      t.$slots.watermark ? (u(), i("div", ia, [
        b(t.$slots, "watermark", D(I({ isPrinting: n(He) || n(Xe) || xe.value || we.value })), void 0, !0)
      ])) : f("", !0),
      t.$slots.source ? (u(), i("div", {
        key: 5,
        ref_key: "source",
        ref: ot,
        dir: "auto"
      }, [
        b(t.$slots, "source", {}, void 0, !0)
      ], 512)) : f("", !0),
      $e(n(q), {
        teleportTo: e.value.style.tooltip.teleportTo,
        show: P.value.showTooltip && ne.value,
        backgroundColor: e.value.style.tooltip.backgroundColor,
        color: e.value.style.tooltip.color,
        borderRadius: e.value.style.tooltip.borderRadius,
        borderColor: e.value.style.tooltip.borderColor,
        borderWidth: e.value.style.tooltip.borderWidth,
        fontSize: e.value.style.tooltip.fontSize,
        backgroundOpacity: e.value.style.tooltip.backgroundOpacity,
        position: e.value.style.tooltip.position,
        offsetY: e.value.style.tooltip.offsetY,
        parent: _.value,
        content: Oe.value,
        isFullscreen: se.value,
        isCustom: e.value.style.tooltip.customFormat && typeof e.value.style.tooltip.customFormat == "function",
        smooth: e.value.style.tooltip.smooth,
        backdropFilter: e.value.style.tooltip.backdropFilter,
        smoothForce: e.value.style.tooltip.smoothForce,
        smoothSnapThreshold: e.value.style.tooltip.smoothSnapThreshold
      }, {
        "tooltip-before": x(() => [
          b(t.$slots, "tooltip-before", D(I({ ...Ae.value })), void 0, !0)
        ]),
        "tooltip-after": x(() => [
          b(t.$slots, "tooltip-after", D(I({ ...Ae.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"]),
      H.value && e.value.userOptions.buttons.table ? (u(), ge(yl(Te.value.component), fl({ key: 6 }, Te.value.props, {
        ref_key: "tableUnit",
        ref: ce,
        onClose: Ue
      }), Tt({
        content: x(() => [
          p("div", {
            ref_key: "tableContainer",
            ref: Ze,
            class: "vue-ui-heatmap-table atom-data-table",
            style: A(`${e.value.table.useDialog ? "" : "max-height: 300px; margin-top: 24px;"}`)
          }, [
            p("div", {
              style: A(`width:100%;overflow-x:auto;position:relative;${e.value.table.useDialog ? "" : "padding-top:36px"};`),
              class: pe({ "vue-ui-responsive": n(al) })
            }, [
              e.value.table.useDialog ? f("", !0) : (u(), i("div", {
                key: 0,
                role: "button",
                tabindex: "0",
                style: A(`width:32px; position: absolute; top: 0; left:4px; padding: 0 0px; display: flex; align-items:center;justify-content:center;height: 36px; width: 32px; cursor:pointer; background:${e.value.table.th.backgroundColor};`),
                onClick: Ue,
                onKeypress: pl(Ue, ["enter"])
              }, [
                $e(n(ee), {
                  name: "close",
                  stroke: e.value.table.th.color,
                  "stroke-width": 2
                }, null, 8, ["stroke"])
              ], 36)),
              p("table", ra, [
                e.value.table.useDialog ? f("", !0) : (u(), i("caption", {
                  key: 0,
                  style: A(`backgroundColor:${e.value.table.th.backgroundColor};color:${e.value.table.th.color};outline:${e.value.table.th.outline}`)
                }, [
                  We(O(e.value.style.title.text) + " ", 1),
                  e.value.style.title.subtitle.text ? (u(), i("span", va, O(e.value.style.title.subtitle.text), 1)) : f("", !0)
                ], 4)),
                p("thead", null, [
                  p("tr", {
                    role: "row",
                    style: A(`background:${e.value.table.th.backgroundColor};color:${e.value.table.th.color}`)
                  }, [
                    p("th", {
                      style: A(`outline:${e.value.table.th.outline};padding-right:6px`)
                    }, null, 4),
                    (u(!0), i(z, null, V(h.dataset, (a, s) => (u(), i("th", {
                      align: "right",
                      style: A(`outline:${e.value.table.th.outline};padding-right:6px`)
                    }, O(a.name), 5))), 256))
                  ], 4)
                ]),
                p("tbody", null, [
                  (u(!0), i(z, null, V(m.value.xLabels, (a, s) => (u(), i("tr", {
                    role: "row",
                    class: pe({ "vue-ui-data-table__tbody__row": !0, "vue-ui-data-table__tbody__row-even": s % 2 === 0, "vue-ui-data-table__tbody__row-odd": s % 2 !== 0 }),
                    style: A(`background:${e.value.table.td.backgroundColor};color:${e.value.table.td.color}`)
                  }, [
                    p("td", {
                      "data-cell": e.value.table.colNames.xAxis,
                      class: "vue-ui-data-table__tbody__td",
                      style: A(`outline:${e.value.table.td.outline}`)
                    }, [
                      p("div", da, O(a), 1)
                    ], 12, ca),
                    (u(!0), i(z, null, V(h.dataset, (c, d) => (u(), i("td", {
                      class: "vue-ui-data-table__tbody__td",
                      "data-cell": h.dataset[d].name,
                      style: A(`outline:${e.value.table.td.outline}`)
                    }, [
                      p("div", ya, O(isNaN(c.values[s]) ? "-" : n(Y)({ p: e.value.style.layout.dataLabels.prefix, v: c.values[s], s: e.value.style.layout.dataLabels.suffix, r: e.value.table.td.roundingValue })), 1)
                    ], 12, ha))), 256))
                  ], 6))), 256))
                ])
              ])
            ], 6)
          ], 4)
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: x(() => [
            We(O(Te.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: x(() => [
            p("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: o[0] || (o[0] = (a) => Be(e.value.userOptions.callbacks.csv)),
              style: A({ cursor: ke.value ? "pointer" : "default" })
            }, [
              $e(n(ee), {
                name: "fileCsv",
                stroke: Te.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : f("", !0),
      b(t.$slots, "skeleton", {}, () => [
        n(Ce) ? (u(), ge($l, { key: 0 })) : f("", !0)
      ], !0)
    ], 46, Xl));
  }
}, Na = /* @__PURE__ */ Il(fa, [["__scopeId", "data-v-674fc54e"]]);
export {
  Na as default
};
