// vue-data-ui/dist/vue-ui-xy-canvas-4auyhlTr.js
import { defineAsyncComponent as Ce, ref as d, computed as g, useSlots as jt, onMounted as Ua, toRefs as Zt, watch as E, nextTick as ke, onBeforeUnmount as Je, watchEffect as Ke, shallowRef as Qe, openBlock as F, createElementBlock as Ee, normalizeClass as Jt, normalizeStyle as ie, createBlock as U, createCommentVNode as X, createElementVNode as K, unref as C, createSlots as Xa, withCtx as b, renderSlot as k, normalizeProps as Q, guardReactiveProps as ee, createVNode as _a, Teleport as Kt, toDisplayString as ea, resolveDynamicComponent as Qt, mergeProps as el, createTextVNode as Va } from "vue";
import { j as al, t as Ya, m as tl, p as oe, $ as ll, n as ol, D as aa, k as ve, l as ce, s as qa, H as ul, a7 as sl, J as nl, o as rl, q as il, T as Ga, u as we, x as vl, a0 as cl, a1 as dl, U as hl } from "./lib-GwPnJSrY.js";
import { c as fl, a as ta, d as ml, t as yl, u as gl, l as w, r as Ne, b as q, p as ja } from "./useResponsive-DfdjqQps.js";
import { u as pl, a as Za, b as bl } from "./useTimeLabels-C-A2uZKi.js";
import { u as xl } from "./useConfig-CIgyKZ6R.js";
import { u as Ll, B as Cl } from "./BaseScanner-BhhkKVoj.js";
import { u as kl } from "./usePrinter-DiK5R6k-.js";
import { u as Be } from "./useNestedProp-BBT0mbHh.js";
import { u as wl } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Sl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as zl } from "./useChartAccessibility-9icAAmYg.js";
import Tl from "./img-BYYzY_QU.js";
import Ml from "./vue_ui_xy_canvas-6BBDSowT.js";
import $l from "./Legend-1LMMsPxL.js";
import Al from "./Title-BfyrP-jp.js";
import Il from "./BaseIcon-CNx8Tzl5.js";
import Fl from "./vue-ui-accordion-BGHTqIBt.js";
import { S as Pl } from "./SlicerPreview-I2J0UUF_.js";
import { B as Rl } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Ol } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const Dl = ["id"], El = ["id"], Nl = ["id"], Bl = ["onClick"], Hl = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Wl = ["innerHTML"], Ul = {
  __name: "vue-ui-xy-canvas",
  props: {
    dataset: {
      type: Array,
      default() {
        return [];
      }
    },
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    selectedXIndex: {
      type: Number,
      default: void 0
    }
  },
  emits: ["selectLegend", "selectX", "copyAlt"],
  setup(la, { expose: Ja, emit: Ka }) {
    const Qa = Ce(() => import("./Tooltip-DL3QgjZf.js")), et = Ce(() => import("./DataTable-Bo1pmAqG.js")), at = Ce(() => import("./UserOptions-CrLqEimB.js")), tt = Ce(() => import("./NonSvgPenAndPaper-CGk_xThe.js")), lt = Ce(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_xy_canvas: ot } = xl(), { isThemeValid: ut, warnInvalidTheme: st } = wl(), P = la, ae = d(al()), _ = d(null), ue = d(null), v = d(null), y = d(1), G = d(1), de = d(!1), h = d(null), Se = d(""), oa = d(null), x = d([]), se = d(1), N = d(!0), V = d(!0), ze = d(null), ua = d(0), he = d(!1), sa = d(null), na = d(null), j = d(null), ra = d(null), He = d(0), We = d(0), Ue = d(0), H = d(null), ia = d(!1), fe = d(null), Xe = d(null), me = d(!1), va = d(!1), Te = d(!1), S = d(null), _e = d(!1), Y = g(() => Array.isArray(Ve.value) && Ve.value.length > 0), Me = Ka, nt = jt();
    Ua(() => {
      nt["chart-background"] && console.warn("VueUiXyCanvas does not support the #chart-background slot.");
    });
    const a = d(qe()), te = g(() => a.value.userOptions.useCursorPointer), rt = g(() => !!a.value.debug), it = g(() => Ya({
      defaultConfig: {
        userOptions: { show: !1 },
        table: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            grid: {
              x: {
                axisColor: "#6A6A6A",
                timeLabels: { show: !1 },
                axisName: "",
                horizontalLines: {
                  color: "#6A6A6A"
                }
              },
              y: {
                axisColor: "#6A6A6A",
                axisLabels: { show: !1 },
                axisName: "",
                verticalLines: {
                  color: "#6A6A6A"
                }
              },
              zeroLine: {
                color: "#6A6A6A"
              }
            },
            legend: {
              backgroundColor: "#99999930"
            },
            paddingProportions: {
              left: 0.05
            },
            scale: {
              max: null,
              min: null
            },
            zoom: {
              endIndex: null,
              startIndex: null
            }
          }
        }
      },
      userConfig: a.value.skeletonConfig ?? {}
    })), { loading: $e, FINAL_DATASET: Ve } = Ll({
      ...Zt(P),
      FINAL_CONFIG: a,
      prepareConfig: qe,
      skeletonDataset: P.config?.skeletonDataset ?? [
        {
          name: "",
          series: [0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 134],
          type: "line",
          smooth: !0,
          color: "#BABABA"
        },
        {
          name: "",
          series: [0, 0.5, 1, 1.5, 2.5, 4, 6.5, 10.5, 17, 27.5, 44.5, 67],
          type: "bar",
          color: "#AAAAAA"
        }
      ],
      skeletonConfig: Ya({
        defaultConfig: a.value,
        userConfig: it.value
      })
    }), { userOptionsVisible: Ye, setUserOptionsVisibility: ca, keepUserOptionState: da } = Sl({ config: a.value }), { svgRef: z } = zl({ config: a.value.style.chart.title });
    function vt() {
      ca(!0);
    }
    function ct() {
      ca(!1), Me("selectX", { seriesIndex: null, datapoint: null }), h.value = null;
    }
    function qe() {
      const e = Be({
        userConfig: P.config,
        defaultConfig: ot
      });
      let t = {};
      const l = e.theme;
      if (l)
        if (!ut.value(e))
          st(e), t = e;
        else {
          const n = Be({
            userConfig: Ml[l] || P.config,
            defaultConfig: e
          });
          t = {
            ...Be({
              userConfig: P.config,
              defaultConfig: n
            }),
            customPalette: e.customPalette.length ? e.customPalette : tl[l] || oe
          };
        }
      else
        t = e;
      return P.config && ll(P.config, "style.chart.grid.y.timeLabels") && (console.warn("VueUiXyCanvas: you are using the deprecated config.style.chart.grid.y.timeLabels. It is recommended to move this configuration to config.style.chart.grid.x.timeLabels."), t.style.chart.grid.x.timeLabels = Be({
        defaultConfig: t.style.chart.grid.x.timeLabels,
        userConfig: P.config.style.chart.grid.y.timeLabels
      })), t;
    }
    E(() => P.config, (e) => {
      $e.value || (a.value = qe()), Ye.value = !a.value.userOptions.showOnChartHover, Pa(), He.value += 1, We.value += 1, Ue.value += 1, c.value.showTable = a.value.table.show, c.value.showDataLabels = a.value.style.chart.dataLabels.show, c.value.stacked = a.value.style.chart.stacked, c.value.showTooltip = a.value.style.chart.tooltip.show;
    }, { deep: !0 }), E(() => P.dataset, async (e) => {
      !Array.isArray(e) || e.length === 0 || (await ke(), z.value && !v.value && (v.value = z.value.getContext("2d", { willReadFrequently: !0 })), N.value = !0, V.value = !0, await pe(), Ie(), He.value += 1, We.value += 1, Ue.value += 1);
    }, { deep: !0 });
    const ha = d(a.value.style.chart.aspectRatio), { isPrinting: fa, isImaging: ma, generatePdf: ya, generateImage: ga } = kl({
      elementId: `xy_canvas_${ae.value}`,
      fileName: a.value.style.chart.title.text || "vue-ui-xy-canvas",
      options: a.value.userOptions.print
    }), c = d({
      showTable: a.value.table.show,
      showDataLabels: a.value.style.chart.dataLabels.show,
      stacked: a.value.style.chart.stacked,
      showTooltip: a.value.style.chart.tooltip.show
    });
    function pa(e) {
      he.value = e, ua.value += 1;
    }
    const ba = g(() => ol(a.value.customPalette)), xa = g(() => x.value.length === $.value.length), O = g(() => $.value ? Math.max(...$.value.filter((e, t) => xa.value ? !0 : !x.value.includes(e.absoluteIndex)).map((e) => e.series.length)) : 0);
    function dt(e) {
      S.value = e, Z();
    }
    const La = d(0), ye = d(0);
    function ht() {
      if (!v.value || !a.value.style.chart.grid.x.timeLabels.show) return 0;
      const e = R.value || [], t = u.value.start ?? 0, l = u.value.end ?? 0;
      if (!Math.max(0, l - t)) return 0;
      const s = Math.round(
        y.value / 40 * a.value.style.chart.grid.x.timeLabels.fontSizeRatio
      ), i = `${a.value.style.chart.grid.x.timeLabels.bold ? "bold " : ""}${s}px ${a.value.style.fontFamily}`;
      v.value.save(), v.value.font = i;
      let r = 0;
      for (let W = t; W < l; W += 1) {
        const B = e[W]?.text ?? `${W + 1}`, De = v.value.measureText(String(B));
        De.width > r && (r = De.width);
      }
      v.value.restore();
      const p = (a.value.style.chart.grid.x.timeLabels.rotation || 0) * Math.PI / 180, M = s, L = Math.abs(Math.sin(p)) * r + Math.abs(Math.cos(p)) * M, D = a.value.style.chart.grid.x.timeLabels.offsetY || 1, le = y.value / D;
      return Math.max(0, le + L + 4);
    }
    function ft() {
      La.value = ht();
    }
    function mt() {
      ye.value && cancelAnimationFrame(ye.value), ye.value = requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          ft();
        });
      });
    }
    Je(() => {
      ye.value && cancelAnimationFrame(ye.value);
    });
    const o = g(() => {
      const e = y.value - y.value * (a.value.style.chart.paddingProportions.left + a.value.style.chart.paddingProportions.right), t = G.value * a.value.style.chart.paddingProportions.top, l = G.value * a.value.style.chart.paddingProportions.bottom, n = La.value, s = G.value - l - n, i = G.value - (t + l) - n;
      return {
        canvasWidth: y.value,
        canvasHeight: G.value,
        left: y.value * a.value.style.chart.paddingProportions.left,
        top: t,
        right: y.value - y.value * a.value.style.chart.paddingProportions.right,
        bottom: s,
        width: e,
        height: i,
        slot: e / (u.value.end - u.value.start)
      };
    });
    function Ca(e, t) {
      return e / t;
    }
    function yt({ hasAutoScale: e, series: t, min: l, max: n, scale: s, yOffset: i, individualHeight: r, stackIndex: f = null }) {
      return t.map((p, M) => {
        const L = s.min < 0 ? Math.abs(s.min) : 0, D = Ca(p + L, L + s.max);
        let le, W;
        e && (le = s.min, W = Ca(p - le, s.max - le));
        let B = 0;
        return f === null ? B = o.value.bottom - o.value.height * (e ? W : D) : B = o.value.bottom - i - r * (e ? W : D), {
          x: o.value.left + o.value.slot * M + o.value.slot / 2,
          y: B,
          value: p
        };
      });
    }
    const T = g(() => {
      const e = a.value.style.chart.scale.min !== null ? a.value.style.chart.scale.min : Math.min(...$.value.filter((r, f) => !x.value.includes(r.absoluteIndex)).flatMap((r) => r.series.slice(u.value.start, u.value.end))), t = a.value.style.chart.scale.max !== null ? a.value.style.chart.scale.max : Math.max(...$.value.filter((r, f) => !x.value.includes(r.absoluteIndex)).flatMap((r) => r.series.slice(u.value.start, u.value.end))), l = aa(e < 0 ? e : 0, t === e ? e + 1 < 0 ? 0 : e + 1 : t < 0 ? 0 : t, a.value.style.chart.scale.ticks), n = l.min < 0 ? Math.abs(l.min) : 0, s = o.value.bottom - o.value.height * (n / (l.max + n)), i = l.ticks.map((r) => ({
        y: o.value.bottom - o.value.height * ((r + n) / (l.max + n)),
        x: o.value.left - 8,
        value: r
      }));
      return {
        absoluteMin: n,
        max: t,
        min: e,
        scale: l,
        yLabels: i,
        zero: s
      };
    }), gt = g(() => m.value.map((e) => `
            <div style="display:flex;flex-direction:row;gap:6px;align-items:center;">
                <svg viewBox="0 0 10 10" height="12" width="12">
                    <circle cx="5" cy="5" r="5" fill="${e.color}"/>
                </svg>
                <span>${e.name ? e.name + ": " : ""}</span>
                <span>${ve(
      a.value.style.chart.dataLabels.formatter,
      e.series[h.value] ?? "-",
      ce({
        p: e.prefix || "",
        v: e.series[h.value] ?? "-",
        s: e.suffix || "",
        r: e.rounding || 0
      }),
      { datapoint: e, seriesIndex: h.value }
    )}</span>
            </div>
        `)), Ae = g(() => a.value.style.chart.line.cutNullValues), $ = g(() => Ve.value.map((e, t) => ({
      ...e,
      series: ul({
        data: sl(e.series, [], Ae.value),
        threshold: a.value.downsample.threshold
      }),
      absoluteIndex: t,
      color: qa(e.color || ba.value[t] || oe[t] || oe[t % oe.length])
    }))), pt = g(() => {
      if (!a.value.style.chart.zoom.minimap.show) return [];
      const e = $.value.filter((s) => !x.value.includes(s.absoluteIndex)), t = Math.max(...e.map((s) => s.series.length)), l = [];
      for (let s = 0; s < t; s += 1)
        l.push(e.map((i) => i.series[s] || 0).reduce((i, r) => (i || 0) + (r || 0), 0));
      const n = Math.min(...l);
      return l.map((s) => s + (n < 0 ? Math.abs(n) : 0));
    }), bt = g(() => a.value.style.chart.zoom.minimap.show ? $.value.map((t) => ({
      ...t,
      isVisible: !x.value.includes(t.absoluteIndex)
    })) : []);
    E(O, (e) => {
      e && pe();
    });
    const m = g(() => nl($.value.filter((e, t) => !x.value.includes(e.absoluteIndex))).map((e, t) => ({
      ...e,
      series: e.series.slice(u.value.start, u.value.end)
    })).map((e, t) => {
      let l = [null, void 0].includes(e.scaleMin) ? Math.min(...e.series) || 0 : e.scaleMin, n = [null, void 0].includes(e.scaleMax) ? Math.max(...e.series) || 1 : e.scaleMax;
      l === n && (l = l >= 0 ? n - 1 : l, n = n >= 0 ? n : l + 1), e.series.filter((B) => ![null, void 0].includes(B)).map((B) => (B - l) / (n - l));
      const s = {
        valueMin: l,
        valueMax: n
      }, i = e.scaleSteps || a.value.style.chart.scale.ticks;
      let r;
      e.autoScaling ? r = aa(s.valueMin, s.valueMax, i) : r = aa(s.valueMin < 0 ? s.valueMin : 0, s.valueMax <= 0 ? 0 : s.valueMax, i);
      const f = c.value.stacked ? o.value.height * (1 - e.cumulatedStackRatio) : 0, p = c.value.stacked ? o.value.height / a.value.style.chart.stackGap : 0, M = c.value.stacked ? o.value.height * e.stackRatio - p : o.value.height, L = r.min < 0 ? Math.abs(r.min) : 0;
      let D;
      e.autoScaling && c.value.stacked && n <= 0 ? D = o.value.bottom - f - M : D = o.value.bottom - f - M * (L / (r.max + L));
      const le = r.ticks.map((B, De) => ({
        y: o.value.bottom - f - M * (De / (r.ticks.length - 1)),
        x: o.value.left - 8,
        value: B
      })), W = yt({
        hasAutoScale: c.value.stacked && e.autoScaling,
        series: e.series,
        min: c.value.stacked ? l : T.value.min,
        max: c.value.stacked ? n : T.value.max,
        scale: c.value.stacked ? r : T.value.scale,
        yOffset: f,
        individualHeight: M,
        stackIndex: c.value.stacked ? t : null
      });
      return {
        ...e,
        coordinatesLine: W,
        min: l,
        max: n,
        localScale: r,
        localZero: D,
        localMin: L,
        localYLabels: le,
        yOffset: f,
        individualHeight: M
      };
    })), u = d({
      start: 0,
      end: O.value
    }), A = d({ start: 0, end: O.value }), ge = d(null);
    function xt() {
      return new Promise((e) => requestAnimationFrame(
        () => requestAnimationFrame(() => e())
      ));
    }
    Je(() => {
      ge.value && cancelAnimationFrame(ge.value);
    });
    async function pe() {
      wa(), await ke(), ge.value && cancelAnimationFrame(ge.value), ge.value = requestAnimationFrame(async () => {
        await xt(), wa();
      });
    }
    const Lt = g(() => a.value.style.chart.zoom.preview.enable && (A.value.start !== u.value.start || A.value.end !== u.value.end));
    function ka(e, t) {
      A.value[e] = t;
    }
    async function wa() {
      if (!me.value) {
        me.value = !0;
        try {
          const { startIndex: e, endIndex: t } = a.value.style.chart.zoom, l = Math.max(...$.value.map((i) => i.series.length)), n = e ?? 0, s = t != null ? Math.min(Sa(t + 1), l) : l;
          Te.value = !0, u.value.start = n, u.value.end = s, A.value.start = n, A.value.end = s, Ge(), va.value = !0, await ke(), j.value && (j.value.setStartValue(u.value.start), j.value.setEndValue(u.value.end));
        } finally {
          queueMicrotask(() => {
            Te.value = !1;
          }), me.value = !1;
        }
      }
    }
    function Ct(e) {
      me.value || Te.value || e !== u.value.start && (u.value.start = e, A.value.start = e, Ge());
    }
    function kt(e) {
      if (me.value || Te.value) return;
      const t = Sa(e);
      t !== u.value.end && (u.value.end = t, A.value.end = t, Ge());
    }
    function Sa(e) {
      const t = O.value;
      return e > t ? t : e < 0 || e < u.value.start ? a.value.style.chart.zoom.startIndex !== null ? a.value.style.chart.zoom.startIndex + 1 : 1 : e;
    }
    function Ge() {
      const e = Math.max(
        1,
        Math.max(...$.value.map((n) => n.series.length))
      );
      let t = Math.max(0, Math.min(u.value.start ?? 0, e - 1)), l = Math.max(t + 1, Math.min(u.value.end ?? e, e));
      (!Number.isFinite(t) || !Number.isFinite(l) || l <= t) && (t = 0, l = e), u.value = { start: t, end: l }, A.value.start = t, A.value.end = l, j.value && (j.value.setStartValue(t), j.value.setEndValue(l));
    }
    const wt = g(() => m.value.filter((e) => ["line", "plot", void 0].includes(e.type))), I = g(() => m.value.filter((e) => e.type === "bar"));
    function Ie() {
      if (!z.value || !ue.value) return;
      const e = ue.value.offsetWidth, t = ue.value.offsetHeight;
      z.value.width = e * se.value * 2, z.value.height = t * se.value * 2, y.value = e * se.value * 2, G.value = t * se.value * 2, v.value?.scale(se.value, se.value), Z();
    }
    E(Y, async (e) => {
      if (!e) {
        ze.value = null, de.value = !1, h.value = null, H.value = null;
        return;
      }
      await ke(), z.value && !v.value && (v.value = z.value.getContext("2d", { willReadFrequently: !0 })), N.value = !0, V.value = !0, await pe(), Ie(), Z();
    });
    function St() {
      if (v.value && (v.value.clearRect(0, 0, 1e4, 1e4), v.value.fillStyle = a.value.style.chart.backgroundColor, v.value.fillRect(0, 0, o.value.canvasWidth, o.value.canvasHeight), !!Y.value)) {
        if (c.value.stacked)
          a.value.style.chart.grid.y.verticalLines.show && u.value.end - u.value.start < a.value.style.chart.grid.y.verticalLines.hideUnderXLength ? m.value.forEach((e) => {
            for (let t = 0; t < u.value.end - u.value.start + 1; t += 1)
              w(
                v.value,
                [
                  {
                    x: o.value.left + o.value.slot * t,
                    y: o.value.bottom - e.yOffset - e.individualHeight
                  },
                  {
                    x: o.value.left + o.value.slot * t,
                    y: o.value.bottom - e.yOffset
                  }
                ],
                {
                  color: a.value.style.chart.grid.y.verticalLines.color
                }
              );
          }) : a.value.style.chart.grid.y.verticalLines.show && u.value.end - u.value.start >= a.value.style.chart.grid.y.verticalLines.hideUnderXLength && m.value.forEach((e) => {
            for (let t = u.value.start; t < u.value.end; t += 1)
              t % Math.floor((u.value.end - u.value.start) / a.value.style.chart.grid.x.timeLabels.modulo) === 0 && w(
                v.value,
                [
                  {
                    x: o.value.left + o.value.slot * (t - u.value.start) + o.value.slot / 2,
                    y: o.value.bottom - e.yOffset - e.individualHeight
                  },
                  {
                    x: o.value.left + o.value.slot * (t - u.value.start) + o.value.slot / 2,
                    y: o.value.bottom - e.yOffset
                  }
                ],
                {
                  color: a.value.style.chart.grid.y.verticalLines.color
                }
              );
          }), a.value.style.chart.grid.x.horizontalLines.show && (a.value.style.chart.grid.x.horizontalLines.alternate ? m.value.forEach((e) => {
            e.localYLabels.forEach((t, l) => {
              l < e.localYLabels.length - 1 && Ne(
                v.value,
                [
                  { x: o.value.left, y: t.y },
                  { x: o.value.right, y: t.y },
                  { x: o.value.right, y: e.localYLabels[l + 1].y },
                  { x: o.value.left, y: e.localYLabels[l + 1].y }
                ],
                {
                  fillColor: l % 2 === 0 ? "transparent" : we(a.value.style.chart.grid.x.horizontalLines.color, a.value.style.chart.grid.x.horizontalLines.opacity),
                  strokeColor: "transparent"
                }
              );
            });
          }) : m.value.forEach((e) => {
            e.localYLabels.slice(u.value.start, u.value.end).forEach((t) => {
              w(
                v.value,
                [
                  { x: o.value.left, y: t.y },
                  { x: o.value.right, y: t.y }
                ],
                {
                  color: a.value.style.chart.grid.x.horizontalLines.color
                }
              );
            });
          })), a.value.style.chart.grid.zeroLine.show && m.value.forEach((e) => {
            w(
              v.value,
              [
                { x: o.value.left, y: e.localZero },
                { x: o.value.right, y: e.localZero }
              ],
              {
                color: a.value.style.chart.grid.zeroLine.color,
                lineDash: a.value.style.chart.grid.zeroLine.dashed ? [10, 10] : [0, 0]
              }
            );
          }), a.value.style.chart.grid.y.axisLabels.show && m.value.forEach((e) => {
            w(
              v.value,
              [
                { x: o.value.left, y: o.value.bottom - e.yOffset },
                { x: o.value.left, y: o.value.bottom - e.yOffset - e.individualHeight }
              ],
              {
                color: e.color
              }
            ), w(
              v.value,
              [
                { x: o.value.right, y: o.value.bottom - e.yOffset },
                { x: o.value.right, y: o.value.bottom - e.yOffset - e.individualHeight }
              ],
              {
                color: e.color
              }
            );
          }), m.value.forEach((e) => {
            q(
              v.value,
              e.name,
              y.value / 35,
              o.value.bottom - e.yOffset - e.individualHeight / 2,
              {
                align: "center",
                rotation: -90,
                color: e.color,
                font: `${Math.round(y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`
              }
            );
          });
        else {
          if (a.value.style.chart.grid.y.verticalLines.show && u.value.end - u.value.start < a.value.style.chart.grid.y.verticalLines.hideUnderXLength)
            for (let e = 0; e < u.value.end - u.value.start + 1; e += 1)
              w(
                v.value,
                [
                  { x: o.value.left + o.value.slot * e, y: o.value.top },
                  { x: o.value.left + o.value.slot * e, y: o.value.bottom }
                ],
                {
                  color: a.value.style.chart.grid.y.verticalLines.color
                }
              );
          else if (a.value.style.chart.grid.y.verticalLines.show && u.value.end - u.value.start >= a.value.style.chart.grid.y.verticalLines.hideUnderXLength)
            for (let e = u.value.start; e < u.value.end; e += 1)
              e % Math.floor((u.value.end - u.value.start) / a.value.style.chart.grid.x.timeLabels.modulo) === 0 && w(
                v.value,
                [
                  { x: o.value.left + o.value.slot * (e - u.value.start) + o.value.slot / 2, y: o.value.top },
                  { x: o.value.left + o.value.slot * (e - u.value.start) + o.value.slot / 2, y: o.value.bottom }
                ],
                {
                  color: a.value.style.chart.grid.y.verticalLines.color
                }
              );
          a.value.style.chart.grid.x.horizontalLines.show && (a.value.style.chart.grid.x.horizontalLines.alternate ? T.value.yLabels.forEach((e, t) => {
            t < T.value.yLabels.length - 1 && Ne(
              v.value,
              [
                { x: o.value.left, y: e.y },
                { x: o.value.right, y: e.y },
                { x: o.value.right, y: T.value.yLabels[t + 1].y },
                { x: o.value.left, y: T.value.yLabels[t + 1].y }
              ],
              {
                fillColor: t % 2 === 0 ? "transparent" : we(a.value.style.chart.grid.x.horizontalLines.color, a.value.style.chart.grid.x.horizontalLines.opacity),
                strokeColor: "transparent"
              }
            );
          }) : T.value.yLabels.forEach((e) => {
            w(
              v.value,
              [
                { x: o.value.left, y: e.y },
                { x: o.value.right, y: e.y }
              ],
              {
                color: a.value.style.chart.grid.x.horizontalLines.color
              }
            );
          })), a.value.style.chart.grid.y.showAxis && w(
            v.value,
            [
              { x: o.value.left, y: o.value.top },
              { x: o.value.left, y: o.value.bottom }
            ],
            {
              color: a.value.style.chart.grid.y.axisColor,
              lineWidth: a.value.style.chart.grid.y.axisThickness
            }
          ), a.value.style.chart.grid.x.showAxis && w(
            v.value,
            [
              { x: o.value.left, y: o.value.bottom },
              { x: o.value.right, y: o.value.bottom }
            ],
            {
              color: a.value.style.chart.grid.x.axisColor,
              lineWidth: a.value.style.chart.grid.x.axisThickness
            }
          ), a.value.style.chart.grid.zeroLine.show && w(
            v.value,
            [
              { x: o.value.left, y: T.value.zero },
              { x: o.value.right, y: T.value.zero }
            ],
            {
              color: a.value.style.chart.grid.zeroLine.color,
              lineDash: a.value.style.chart.grid.zeroLine.dashed ? [10, 10] : [0, 0]
            }
          );
        }
        a.value.style.chart.grid.y.axisName && q(
          v.value,
          a.value.style.chart.grid.y.axisName,
          y.value - y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio * 1.2,
          o.value.bottom - o.value.height / 2,
          {
            font: `${a.value.style.chart.grid.y.axisLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
            color: a.value.style.chart.color,
            align: "center",
            rotation: 90
          }
        ), a.value.style.chart.grid.x.axisName && q(
          v.value,
          a.value.style.chart.grid.x.axisName,
          y.value / 2,
          G.value,
          {
            font: `${a.value.style.chart.grid.y.axisLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
            color: a.value.style.chart.color,
            align: "center"
          }
        );
      }
    }
    function zt(e) {
      for (let t = 0; t < e.coordinatesLine.length; t += 1) {
        const l = (h.value === t || S.value === t ? y.value / 150 : a.value.style.chart.line.plots.show || e.type === "plot" ? y.value / 200 : 0) * a.value.style.chart.line.plots.radiusRatio;
        ta(
          v.value,
          { x: e.coordinatesLine[t].x, y: e.coordinatesLine[t].y },
          l,
          {
            color: a.value.style.chart.backgroundColor,
            fillStyle: e.color
          }
        );
      }
    }
    function Tt() {
      m.value.forEach((e) => {
        e.showYMarker && xe(e) && q(
          v.value,
          ve(
            a.value.style.chart.dataLabels.formatter,
            xe(e).value,
            ce({
              p: e.prefix || a.value.style.chart.grid.y.axisLabels.prefix || "",
              v: xe(e).value,
              s: e.suffix || a.value.style.chart.grid.y.axisLabels.suffix || "",
              r: e.rounding || a.value.style.chart.grid.y.axisLabels.rounding || 0
            }),
            { datapoint: xe(e), seriesIndex: null }
          ),
          o.value.left - 8 + a.value.style.chart.grid.y.axisLabels.offsetX,
          xe(e).y,
          {
            align: "right",
            font: `${a.value.style.chart.grid.y.axisLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
            color: e.color
          }
        );
      });
    }
    function Mt() {
      a.value.style.chart.grid.y.axisLabels.show && (c.value.stacked ? m.value.forEach((e) => {
        e.localYLabels.forEach((t, l) => {
          q(
            v.value,
            ve(
              a.value.style.chart.dataLabels.formatter,
              t.value,
              ce({
                p: e.prefix || a.value.style.chart.grid.y.axisLabels.prefix || "",
                v: t.value,
                s: e.suffix || a.value.style.chart.grid.y.axisLabels.suffix || "",
                r: e.rounding || a.value.style.chart.grid.y.axisLabels.rounding || 0
              }),
              { datapoint: t, seriesIndex: l }
            ),
            t.x + a.value.style.chart.grid.y.axisLabels.offsetX,
            t.y,
            {
              align: "right",
              font: `${a.value.style.chart.grid.y.axisLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
              color: e.color,
              globalAlpha: e.showYMarker && ![null, void 0].includes(h.value ?? S.value) ? 0.2 : 1
            }
          );
        });
      }) : T.value.yLabels.forEach((e, t) => {
        q(
          v.value,
          ve(
            a.value.style.chart.dataLabels.formatter,
            e.value,
            ce({
              p: a.value.style.chart.grid.y.axisLabels.prefix || "",
              v: e.value,
              s: a.value.style.chart.grid.y.axisLabels.suffix || "",
              r: a.value.style.chart.grid.y.axisLabels.rounding || 0
            }),
            { datapoint: e, seriesIndex: t }
          ),
          e.x + a.value.style.chart.grid.y.axisLabels.offsetX,
          e.y,
          {
            align: "right",
            font: `${a.value.style.chart.grid.y.axisLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.grid.y.axisLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
            color: a.value.style.chart.grid.y.axisLabels.color,
            globalAlpha: m.value.some((l) => l.showYMarker) && ![null, void 0].includes(h.value ?? S.value) ? 0.2 : 1
          }
        );
      }));
    }
    function $t(e) {
      for (let t = 0; t < e.coordinatesLine.length; t += 1)
        q(
          v.value,
          ve(
            a.value.style.chart.dataLabels.formatter,
            e.coordinatesLine[t].value,
            ce({
              p: e.prefix || "",
              v: e.coordinatesLine[t].value,
              s: e.suffix || "",
              r: e.rounding || 0
            }),
            { datapoint: e.coordinatesLine[t], seriesIndex: t }
          ),
          e.coordinatesLine[t].x,
          e.coordinatesLine[t].y + a.value.style.chart.dataLabels.offsetY,
          {
            align: "center",
            font: `${a.value.style.chart.dataLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.dataLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
            color: a.value.style.chart.dataLabels.useSerieColor ? e.color : a.value.style.chart.dataLabels.color,
            strokeColor: a.value.style.chart.backgroundColor,
            lineWidth: 0.5
          }
        );
    }
    const R = d([]);
    let za = 0;
    Ke(() => {
      const e = ++za;
      (async () => {
        const t = await pl({
          values: a.value.style.chart.grid.x.timeLabels.values,
          maxDatapoints: O.value,
          formatter: a.value.style.chart.grid.x.timeLabels.datetimeFormatter,
          start: 0,
          end: O.value
        });
        e === za && (R.value = t);
      })();
    }), Ke(() => {
      a.value.style.chart.grid.x.timeLabels.show, a.value.style.chart.grid.x.timeLabels.rotation, a.value.style.chart.grid.x.timeLabels.offsetY, a.value.style.chart.grid.x.timeLabels.fontSizeRatio, a.value.style.chart.grid.x.timeLabels.bold, u.value.start, u.value.end, y.value, G.value, (R.value || []).map((e) => e?.text ?? "").join("|"), mt();
    }, { flush: "post" });
    const Ta = d({ months: [], shortMonths: [], days: [], shortDays: [] });
    let Ma = 0;
    Ke(() => {
      const e = ++Ma, t = a.value.style.chart.grid.x.timeLabels.datetimeFormatter;
      (async () => {
        const l = await Za(t.locale).catch(() => Za("en"));
        e === Ma && (Ta.value = l.data);
      })();
    });
    const $a = g(() => {
      const e = a.value.style.chart.grid.x.timeLabels.datetimeFormatter, t = bl({
        useUTC: e.useUTC,
        locale: Ta.value,
        januaryAsYear: e.januaryAsYear
      });
      return (l, n) => {
        const i = a.value.style.chart.grid.x.timeLabels.values?.[l];
        return i == null ? "" : t.formatDate(new Date(i), n);
      };
    }), Aa = g(() => (a.value.style.chart.grid.x.timeLabels.values || []).map((t, l) => ({
      text: $a.value(l, a.value.style.chart.zoom.timeFormat),
      absoluteIndex: l
    }))), At = g(() => (a.value.style.chart.grid.x.timeLabels.values || []).map((t, l) => ({
      text: $a.value(l, a.value.style.chart.tooltip.timeFormat),
      absoluteIndex: l
    })));
    function It() {
      for (let e = u.value.start; e < u.value.end; e += 1)
        (u.value.end - u.value.start < a.value.style.chart.grid.x.timeLabels.modulo || u.value.end - u.value.start >= a.value.style.chart.grid.x.timeLabels.modulo && (e % Math.floor((u.value.end - u.value.start) / a.value.style.chart.grid.x.timeLabels.modulo) === 0 || (e === h.value + u.value.start || e === S.value) && a.value.style.chart.grid.x.timeLabels.showMarker)) && q(
          v.value,
          R.value[e] ? R.value[e].text : e + 1,
          o.value.left + o.value.slot * (e - u.value.start) + o.value.slot / 2,
          o.value.bottom + y.value / a.value.style.chart.grid.x.timeLabels.offsetY,
          {
            align: a.value.style.chart.grid.x.timeLabels.rotation === 0 ? "center" : a.value.style.chart.grid.x.timeLabels.rotation > 0 ? "left" : "right",
            font: `${a.value.style.chart.grid.x.timeLabels.bold ? "bold " : ""}${Math.round(y.value / 40 * a.value.style.chart.grid.x.timeLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
            color: a.value.style.chart.grid.x.timeLabels.showMarker ? we(a.value.style.chart.grid.x.timeLabels.color, h.value !== null || S.value !== null ? h.value + u.value.start === e || S.value === e ? 100 : 20 : 100) : a.value.style.chart.grid.x.timeLabels.color,
            rotation: a.value.style.chart.grid.x.timeLabels.rotation
          }
        );
    }
    function Ia() {
      w(
        v.value,
        [
          { x: o.value.left + o.value.slot * (h.value ?? S.value) + o.value.slot / 2, y: o.value.top },
          { x: o.value.left + o.value.slot * (h.value ?? S.value) + o.value.slot / 2, y: o.value.bottom }
        ],
        {
          color: a.value.style.chart.selector.color,
          lineDash: a.value.style.chart.selector.dashed ? [8, 8] : [0, 0],
          lineWidth: 2
        }
      );
    }
    function Ft() {
      H.value && w(
        v.value,
        [
          { x: o.value.left, y: H.value },
          { x: o.value.right, y: H.value }
        ],
        {
          color: a.value.style.chart.selector.color,
          lineDash: a.value.style.chart.selector.dashed ? [8, 8] : [0, 0],
          lineWidth: 2
        }
      );
    }
    function Pt() {
      I.value.forEach((e, t) => {
        for (let l = 0; l < e.coordinatesLine.length; l += 1)
          Ne(
            v.value,
            [
              {
                x: o.value.left + o.value.slot * l + o.value.slot / 10 + (c.value.stacked ? 0 : o.value.slot / I.value.length * t - (t === 0 ? 0 : o.value.slot / (5 * I.value.length) * t)),
                y: c.value.stacked ? e.localZero : T.value.zero
              },
              {
                x: o.value.left + o.value.slot * l + o.value.slot / 10 + (c.value.stacked ? 0 : o.value.slot / I.value.length * t - (t === 0 ? 0 : o.value.slot / (5 * I.value.length) * t)) + o.value.slot * 0.8 / (c.value.stacked ? 1 : I.value.length),
                y: c.value.stacked ? e.localZero : T.value.zero
              },
              {
                x: o.value.left + o.value.slot * l + o.value.slot / 10 + (c.value.stacked ? 0 : o.value.slot / I.value.length * t - (t === 0 ? 0 : o.value.slot / (5 * I.value.length) * t)) + o.value.slot * 0.8 / (c.value.stacked ? 1 : I.value.length),
                y: e.coordinatesLine[l].y
              },
              {
                x: o.value.left + o.value.slot * l + o.value.slot / 10 + (c.value.stacked ? 0 : o.value.slot / I.value.length * t - (t === 0 ? 0 : o.value.slot / (5 * I.value.length) * t)),
                y: e.coordinatesLine[l].y
              }
            ],
            {
              strokeColor: a.value.style.chart.backgroundColor,
              gradient: {
                type: "linear",
                start: {
                  x: e.coordinatesLine[l].x,
                  y: e.coordinatesLine[l].y
                },
                end: {
                  x: e.coordinatesLine[l].x,
                  y: c.value.stacked ? e.localZero : T.value.zero
                },
                stops: [
                  { offset: 0, color: e.color },
                  { offset: 1, color: a.value.style.chart.bar.gradient.show ? vl(e.color, 0.5) : e.color }
                ]
              }
            }
          ), c.value.showDataLabels && [!0, void 0].includes(e.dataLabels) && q(
            v.value,
            ve(
              a.value.style.chart.dataLabels.formatter,
              e.coordinatesLine[l].value,
              ce({
                p: e.prefix || "",
                v: e.coordinatesLine[l].value,
                s: e.suffix || "",
                r: e.rounding || 0
              }),
              { datapoint: e.coordinatesLine[l], seriesIndex: l }
            ),
            o.value.left + o.value.slot * l + o.value.slot / 10 + (c.value.stacked ? 0 : o.value.slot / I.value.length * t - (t === 0 ? 0 : o.value.slot / (5 * I.value.length) * t)) + o.value.slot * 0.4 / (c.value.stacked ? 1 : I.value.length),
            (e.coordinatesLine[l].value < 0 ? c.value.stacked ? e.localZero : T.value.zero : e.coordinatesLine[l].y) + a.value.style.chart.dataLabels.offsetY,
            {
              align: "center",
              font: `${Math.round(y.value / 40 * a.value.style.chart.dataLabels.fontSizeRatio)}px ${a.value.style.fontFamily}`,
              color: a.value.style.chart.dataLabels.useSerieColor ? e.color : a.value.style.chart.dataLabels.color,
              strokeColor: a.value.style.chart.backgroundColor,
              lineWidth: 0.8
            }
          );
      });
    }
    function Rt(e, t) {
      const l = e.coordinatesLine.map((r, f) => {
        const p = e.series[f];
        return p != null && Number.isFinite(r?.y) ? r : null;
      }), n = [], s = [];
      let i = [];
      for (let r = 0; r < l.length; r += 1) {
        const f = l[r];
        if (f)
          i.push(f);
        else {
          if (i.length >= 2) {
            const p = i[0], M = i[i.length - 1];
            n.push([
              { x: p.x, y: t },
              ...i,
              { x: M.x, y: t }
            ]);
          } else i.length === 1 && s.push(i[0]);
          i = [];
        }
      }
      if (i.length >= 2) {
        const r = i[0], f = i[i.length - 1];
        n.push([
          { x: r.x, y: t },
          ...i,
          { x: f.x, y: t }
        ]);
      } else i.length === 1 && s.push(i[0]);
      return { polygons: n, singles: s };
    }
    function Ot(e) {
      const t = e.coordinatesLine.map((s, i) => {
        const r = e.series[i];
        return r != null && Number.isFinite(s?.y) ? s : null;
      }), l = [];
      let n = [];
      for (let s = 0; s < t.length; s += 1) {
        const i = t[s];
        i ? n.push(i) : (n.length >= 2 && l.push(n), n = []);
      }
      return n.length >= 2 && l.push(n), l;
    }
    function Fa(e) {
      if (Ae.value) {
        const t = Ot(e);
        for (const l of t)
          w(v.value, l, {
            color: e.color,
            lineWidth: 3
          });
      } else
        w(v.value, e.coordinatesLine, {
          color: e.color,
          lineWidth: 3
        });
    }
    function Dt(e) {
      const t = !!Ae.value;
      if (e.useArea) {
        const l = c.value.stacked ? e.localZero : T.value.zero;
        if (t) {
          const { polygons: n, singles: s } = Rt(e, l);
          for (const r of n)
            ja(v.value, r, {
              fillColor: we(e.color, a.value.style.chart.area.opacity),
              strokeColor: "transparent"
            });
          const i = y.value / 200 * a.value.style.chart.line.plots.radiusRatio;
          for (const r of s)
            ta(
              v.value,
              { x: r.x, y: r.y },
              i,
              {
                color: a.value.style.chart.backgroundColor,
                fillStyle: e.color
              }
            );
        } else {
          const n = { x: e.coordinatesLine[0].x, y: l }, s = { x: e.coordinatesLine.at(-1).x, y: l };
          ja(
            v.value,
            [n, ...e.coordinatesLine, s],
            {
              fillColor: we(e.color, a.value.style.chart.area.opacity),
              strokeColor: "transparent"
            }
          );
        }
        Fa(e);
        return;
      }
      Fa(e);
    }
    function Et() {
      m.value.forEach((e, t) => {
        w(
          v.value,
          [
            { x: o.value.left, y: o.value.bottom - e.yOffset },
            { x: o.value.right, y: o.value.bottom - e.yOffset }
          ],
          {
            color: a.value.style.chart.grid.x.horizontalLines.color,
            lineWidth: 1
          }
        );
      });
    }
    function Nt() {
      const { left: e, top: t, width: l, height: n } = o.value, s = u.value.start, r = u.value.end - s, f = l / r, p = A.value.start - s, M = A.value.end - s, L = Math.max(0, Math.min(r, p)), D = Math.max(0, Math.min(r, M));
      Ne(
        v.value,
        [
          { x: e + L * f, y: t },
          { x: e + L * f + (D - L) * f, y: t },
          { x: e + L * f + (D - L) * f, y: t + n },
          { x: e + L * f, y: t + n }
        ],
        {
          fillColor: a.value.style.chart.zoom.preview.fill,
          strokeColor: a.value.style.chart.zoom.preview.stroke,
          lineDash: Array(4).fill(a.value.style.chart.zoom.preview.strokeDasharray),
          lineWidth: a.value.style.chart.zoom.preview.strokeWidth
        }
      );
    }
    function Z() {
      !Y.value || !z.value || !v.value || (St(), N.value ? ((h.value !== null || S.value !== null) && a.value.style.chart.selector.show && Ia(), Pt(), c.value.stacked && a.value.style.chart.grid.x.showAxis && Et(), wt.value.forEach((e) => {
        (e.type === "line" || !e.type) && Dt(e), V.value && (zt(e), c.value.showDataLabels && [!0, void 0].includes(e.dataLabels) && $t(e));
      }), z.value && (ze.value = fl(z.value))) : (ze.value && (v.value.clearRect(0, 0, 1e4, 1e4), v.value.drawImage(ze.value, 0, 0)), (h.value !== null || S.value !== null) && a.value.style.chart.selector.show && Ia(), (h.value !== null || S.value !== null) && m.value.forEach((e) => {
        const t = h.value ?? S.value, l = e.coordinatesLine[t];
        (e.type === "line" || !e.type || e.type === "plot") && l && Number.isFinite(l.x) && Number.isFinite(l.y) && ta(
          v.value,
          { x: l.x, y: l.y },
          y.value / 150 * a.value.style.chart.line.plots.radiusRatio,
          {
            color: a.value.style.chart.backgroundColor,
            fillStyle: e.color
          }
        );
      })), a.value.style.chart.grid.x.timeLabels.show && It(), a.value.style.chart.selector.show && a.value.style.chart.selector.showHorizontalSelector && Ft(), Mt(), Tt(), a.value.style.chart.zoom.preview.enable && (u.value.start !== A.value.start || u.value.end !== A.value.end) && Nt(), N.value = !1);
    }
    const be = ml(() => {
      V.value = !0, Ie();
    }, O.value > 200 ? 10 : 1, !V.value);
    function xe(e) {
      if ([null, void 0].includes(h.value ?? S.value) || !e.coordinatesLine[h.value ?? S.value]) return !1;
      const { y: t, value: l } = e.coordinatesLine[h.value ?? S.value];
      return { y: t, value: l };
    }
    function Bt(e) {
      if (!Y.value || !z.value || xa.value) return;
      const { left: t, top: l } = z.value.getBoundingClientRect(), n = e.clientX - t;
      if (H.value = (e.clientY - l) * 2, (H.value < o.value.top || H.value > o.value.bottom) && (H.value = null), n * 2 < o.value.left || n * 2 > o.value.right) {
        de.value = !1, h.value = null;
        return;
      }
      const s = n * 2 - o.value.left;
      if (h.value = Math.floor(s / o.value.slot), de.value = !0, !V.value) return;
      let i = "";
      const r = a.value.style.chart.tooltip.customFormat, f = m.value.map((p) => ({
        shape: p.shape || null,
        name: p.name,
        color: p.color,
        type: p.type || "line",
        value: p.series.find((M, L) => L === h.value)
      }));
      Ht({ seriesIndex: h.value, datapoint: f }), Ga(r) && hl(() => r({
        seriesIndex: h.value,
        datapoint: f,
        series: m.value,
        config: a.value
      })) ? Se.value = r({
        seriesIndex: h.value,
        datapoint: f,
        series: m.value,
        config: a.value
      }) : (a.value.style.chart.grid.x.timeLabels.values.slice(u.value.start, u.value.end)[h.value] ? i += `<div style="padding-bottom: 6px; margin-bottom: 4px; border-bottom: 1px solid ${a.value.style.chart.tooltip.borderColor}; width:100%">${a.value.style.chart.tooltip.useDefaultTimeFormat ? R.value.slice(u.value.start, u.value.end)[h.value]?.text : At.value[h.value]?.text}</div>` : i += `<div style="padding-bottom: 6px; margin-bottom: 4px; border-bottom: 1px solid ${a.value.style.chart.tooltip.borderColor}; width:100%">${R.value[h.value + u.value.start]?.text ?? ""}</div>`, i += gt.value.join(""), Se.value = i), V.value = !1;
    }
    function Ht({ seriesIndex: e, datapoint: t }) {
      const l = u.value.start + e;
      Me("selectX", {
        dataset: t,
        index: l,
        indexLabel: ""
      });
    }
    E(() => P.selectedXIndex, (e) => {
      if ([null, void 0].includes(P.selectedXIndex)) {
        h.value = null;
        return;
      }
      const t = e - u.value.start;
      t < 0 || e >= u.value.end ? h.value = null : h.value = t ?? null;
    }, { immediate: !0 }), E(() => h.value, (e) => {
      be();
    }), E(() => u.value, (e) => {
      N.value = !0, Z();
    }, {
      deep: !0
    }), E(() => A.value, (e) => {
      Z();
    }, { deep: !0 }), E(() => c.value.showDataLabels, (e) => {
      N.value = !0, Z();
    }), E(() => H.value, (e) => {
      e && Z();
    }), E(() => c.value.stacked, (e) => {
      N.value = !0, V.value = !0, be();
    });
    function Wt() {
      de.value = !1, h.value = null, Se.value = "", H.value = null, Z();
    }
    const J = Qe(null), ne = Qe(null), re = Qe(null);
    Ua(() => {
      ia.value = !0, Pa();
    });
    function Pa() {
      if (rl(P.dataset) && rt.value && il({ componentName: "VueUiXyCanvas", type: "dataset" }), ke(() => {
        z.value && !v.value && (v.value = z.value.getContext("2d", { willReadFrequently: !0 })), v.value && Y.value && (N.value = !0, V.value = !0, Ie());
      }), a.value.responsive) {
        const e = yl(() => {
          const { width: t, height: l } = gl({
            chart: _.value,
            title: a.value.style.chart.title.text ? sa.value : null,
            legend: a.value.style.chart.legend.show ? na.value : null,
            slicer: a.value.style.chart.zoom.show && O.value > 6 ? j.value?.$el : null,
            source: ra.value
          });
          requestAnimationFrame(() => {
            ha.value = `${t} / ${l}`;
          });
        });
        J.value && (ne.value && J.value.unobserve(ne.value), J.value.disconnect()), J.value = new ResizeObserver(e), ne.value = _.value.parentNode, J.value.observe(ne.value);
      }
      re.value && re.value.disconnect(), re.value = new ResizeObserver(async (e) => {
        for (const t of e)
          t.contentBoxSize && ue.value && (N.value = !0, be());
      }), re.value.observe(ue.value), pe();
    }
    Je(() => {
      re.value && re.value.disconnect(), J.value && (ne.value && J.value.unobserve(ne.value), J.value.disconnect());
    });
    function Ut() {
      x.value.length ? x.value = [] : Pe.value.forEach((e, t) => {
        x.value.push(t);
      }), N.value = !0, be();
    }
    function Fe(e) {
      Me("selectLegend", m.value.find((t) => t.absoluteIndex === e)), x.value.includes(e) ? x.value = x.value.filter((t) => t !== e) : x.value.push(e), N.value = !0, be();
    }
    function Ra(e) {
      if (!$.value.length)
        return a.value.debug && console.warn("VueUiXyCanvas - There are no series to show."), null;
      const t = $.value.find((l) => l.name === e);
      return t || (a.value.debug && console.warn(`VueUiXyCanvas - Series name not found "${e}"`), null);
    }
    function Xt(e) {
      const t = Ra(e);
      t !== null && x.value.includes(t.absoluteIndex) && Fe(t.absoluteIndex);
    }
    function _t(e) {
      const t = Ra(e);
      t !== null && (x.value.includes(t.absoluteIndex) || Fe(t.absoluteIndex));
    }
    const Pe = g(() => $.value.map((e, t) => ({
      ...e,
      name: e.name,
      color: qa(e.color) || ba.value[t] || oe[t] || oe[t % oe.length],
      shape: e.shape || "circle",
      prefix: e.prefix || "",
      suffix: e.suffix || "",
      rounding: e.rounding || 0
    })).map((e) => ({
      ...e,
      opacity: x.value.includes(e.absoluteIndex) ? 0.5 : 1,
      segregate: () => Fe(e.absoluteIndex),
      isSegregated: x.value.includes(e.absoluteIndex)
    }))), Vt = g(() => ({
      cy: "donut-div-legend",
      backgroundColor: a.value.style.chart.legend.backgroundColor,
      color: a.value.style.chart.legend.color,
      fontSize: a.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: a.value.style.chart.legend.bold ? "bold" : ""
    })), Re = g(() => {
      const e = [""].concat(m.value.map((s) => s.name)).concat(' <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M18 16v2a1 1 0 0 1 -1 1h-11l6 -7l-6 -7h11a1 1 0 0 1 1 1v2" /></svg>');
      let t = [];
      for (let s = 0; s < O.value; s += 1) {
        const i = m.value.map((r) => r.series[s] ?? 0).reduce((r, f) => r + f, 0);
        t.push(
          [
            a.value.style.chart.grid.x.timeLabels.values.slice(u.value.start, u.value.end)[s] ? R?.value?.slice(u.value.start, u.value.end)?.[s]?.text ?? s + 1 : s + 1
          ].concat(
            m.value.map(
              (r) => (r.series[s] ?? 0).toFixed(a.value.table.rounding)
            )
          ).concat((i ?? 0).toFixed(a.value.table.rounding))
        );
      }
      const l = {
        th: {
          backgroundColor: a.value.table.th.backgroundColor,
          color: a.value.table.th.color,
          outline: a.value.table.th.outline
        },
        td: {
          backgroundColor: a.value.table.td.backgroundColor,
          color: a.value.table.td.color,
          outline: a.value.table.td.outline
        },
        breakpoint: a.value.table.responsiveBreakpoint
      }, n = [a.value.table.columnNames.period].concat(m.value.map((s) => s.name)).concat(a.value.table.columnNames.total);
      return { head: e, body: t.slice(0, u.value.end - u.value.start), config: l, colNames: n };
    }), Oa = g(() => {
      if (m.value.length === 0) return { head: [], body: [], config: {}, columnNames: [] };
      const e = m.value.map((l) => ({
        label: l.name,
        color: l.color,
        type: l.type
      })), t = [];
      for (let l = u.value.start; l < u.value.end; l += 1) {
        const n = [a.value.style.chart.grid.x.timeLabels.values[l] ? R.value[l].text : l + 1];
        m.value.forEach((s) => {
          n.push(Number((s.series[l] || 0).toFixed(a.value.table.rounding)));
        }), t.push(n);
      }
      return { head: e, body: t };
    });
    function je(e = null) {
      const t = [[a.value.style.chart.title.text], [a.value.style.chart.title.subtitle.text], [""]], l = ["", ...Oa.value.head.map((r) => r.label)], n = Oa.value.body, s = t.concat([l]).concat(n), i = cl(s);
      e ? e(i) : dl({ csvContent: i, title: a.value.style.chart.title.text || "vue-ui-xy-canvas" });
    }
    function Yt() {
      return m.value;
    }
    function Da() {
      c.value.showTable = !c.value.showTable;
    }
    function Ea() {
      c.value.showDataLabels = !c.value.showDataLabels;
    }
    function Na() {
      c.value.stacked = !c.value.stacked;
    }
    function Ba() {
      c.value.showTooltip = !c.value.showTooltip;
    }
    const Oe = d(!1);
    function Ze() {
      Oe.value = !Oe.value;
    }
    async function qt({ scale: e = 2 } = {}) {
      if (!_.value) return;
      const { width: t, height: l } = _.value.getBoundingClientRect(), n = t / l, { imageUri: s, base64: i } = await Tl({ domElement: _.value, base64: !0, img: !0, scale: e });
      return {
        imageUri: s,
        base64: i,
        title: a.value.style.chart.title.text,
        width: t,
        height: l,
        aspectRatio: n
      };
    }
    const Le = g(() => {
      const e = a.value.table.useDialog && !a.value.table.show, t = c.value.showTable;
      return {
        component: e ? lt : Fl,
        title: `${a.value.style.chart.title.text}${a.value.style.chart.title.subtitle.text ? `: ${a.value.style.chart.title.subtitle.text}` : ""}`,
        props: e ? {
          backgroundColor: a.value.table.th.backgroundColor,
          color: a.value.table.th.color,
          headerColor: a.value.table.th.color,
          headerBg: a.value.table.th.backgroundColor,
          isFullscreen: he.value,
          fullscreenParent: _.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: te.value
        } : {
          hideDetails: !0,
          config: {
            open: t,
            maxHeight: 1e4,
            body: {
              backgroundColor: a.value.style.chart.backgroundColor,
              color: a.value.style.chart.color
            },
            head: {
              backgroundColor: a.value.style.chart.backgroundColor,
              color: a.value.style.chart.color
            }
          }
        }
      };
    });
    E(() => c.value.showTable, (e) => {
      a.value.table.show || (e && a.value.table.useDialog && fe.value ? fe.value.open() : "close" in fe.value && fe.value.close());
    });
    function Ha() {
      c.value.showTable = !1, Xe.value && Xe.value.setTableIconState(!1);
    }
    function Gt(e) {
      if (e?.stage === "start") {
        _e.value = !0;
        return;
      }
      if (e?.stage === "end") {
        _e.value = !1;
        return;
      }
      ga();
    }
    async function Wa() {
      if (Me("copyAlt", {
        config: a.value,
        dataset: m.value
      }), !a.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(a.value.userOptions.callbacks.altCopy({
        config: a.value,
        dataset: m.value
      }));
    }
    return Ja({
      getData: Yt,
      getImage: qt,
      generateCsv: je,
      generatePdf: ya,
      generateImage: ga,
      hideSeries: _t,
      showSeries: Xt,
      toggleTable: Da,
      toggleLabels: Ea,
      toggleStack: Na,
      toggleTooltip: Ba,
      toggleAnnotator: Ze,
      toggleFullscreen: pa,
      copyAlt: Wa
    }), (e, t) => (F(), Ee("div", {
      style: ie(`width:100%; position:relative; ${a.value.responsive ? "height: 100%" : ""}; background:${a.value.style.chart.backgroundColor};`),
      ref_key: "xy",
      ref: _,
      id: `xy_canvas_${ae.value}`,
      class: Jt(`vue-data-ui-component vue-ui-xy-canvas ${he.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      onMouseenter: vt,
      onMouseleave: ct
    }, [
      a.value.style.chart.title.text ? (F(), Ee("div", {
        key: 0,
        ref_key: "chartTitle",
        ref: sa,
        style: ie(`width:100%;background:${a.value.style.chart.backgroundColor};`)
      }, [
        (F(), U(Al, {
          key: `title_${He.value}`,
          config: {
            title: {
              cy: "xy-canvas-title",
              ...a.value.style.chart.title
            },
            subtitle: {
              cy: "xy-canvas-subtitle",
              ...a.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 4)) : X("", !0),
      K("div", {
        id: `legend-top-${ae.value}`
      }, null, 8, El),
      a.value.userOptions.show && Y.value && (C(da) || C(Ye)) ? (F(), U(C(at), {
        ref_key: "userOptionsRef",
        ref: Xe,
        key: `user_option_${ua.value}`,
        backgroundColor: a.value.style.chart.backgroundColor,
        color: a.value.style.chart.color,
        isPrinting: C(fa),
        isImaging: C(ma),
        uid: ae.value,
        hasTooltip: a.value.userOptions.buttons.tooltip && a.value.style.chart.tooltip.show,
        hasPdf: a.value.userOptions.buttons.pdf,
        hasImg: a.value.userOptions.buttons.img,
        hasXls: a.value.userOptions.buttons.csv,
        hasLabel: a.value.userOptions.buttons.labels,
        hasStack: la.dataset.length > 1 && a.value.userOptions.buttons.stack,
        hasFullscreen: a.value.userOptions.buttons.fullscreen,
        hasAltCopy: a.value.userOptions.buttons.altCopy,
        hasTable: u.value.end - u.value.start <= 730 && a.value.userOptions.buttons.table,
        isFullscreen: he.value,
        isTooltip: c.value.showTooltip,
        isStacked: c.value.stacked,
        titles: { ...a.value.userOptions.buttonTitles },
        chartElement: _.value,
        position: a.value.userOptions.position,
        hasAnnotator: a.value.userOptions.buttons.annotator,
        isAnnotation: Oe.value,
        callbacks: a.value.userOptions.callbacks,
        printScale: a.value.userOptions.print.scale,
        tableDialog: a.value.table.useDialog,
        isCursorPointer: te.value,
        onToggleFullscreen: pa,
        onGeneratePdf: C(ya),
        onGenerateCsv: je,
        onGenerateImage: Gt,
        onToggleTable: Da,
        onToggleLabels: Ea,
        onToggleStack: Na,
        onToggleTooltip: Ba,
        onToggleAnnotator: Ze,
        onCopyAlt: Wa,
        style: ie({
          visibility: C(da) ? C(Ye) ? "visible" : "hidden" : "visible"
        })
      }, Xa({ _: 2 }, [
        e.$slots.menuIcon ? {
          name: "menuIcon",
          fn: b(({ isOpen: l, color: n }) => [
            k(e.$slots, "menuIcon", Q(ee({ isOpen: l, color: n })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        e.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: b(() => [
            k(e.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        e.$slots.optionPdf ? {
          name: "optionPdf",
          fn: b(() => [
            k(e.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        e.$slots.optionCsv ? {
          name: "optionCsv",
          fn: b(() => [
            k(e.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        e.$slots.optionImg ? {
          name: "optionImg",
          fn: b(() => [
            k(e.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        e.$slots.optionTable ? {
          name: "optionTable",
          fn: b(() => [
            k(e.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        e.$slots.optionLabels ? {
          name: "optionLabels",
          fn: b(() => [
            k(e.$slots, "optionLabels", {}, void 0, !0)
          ]),
          key: "6"
        } : void 0,
        e.$slots.optionStack ? {
          name: "optionStack",
          fn: b(() => [
            k(e.$slots, "optionStack", {}, void 0, !0)
          ]),
          key: "7"
        } : void 0,
        e.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: b(({ toggleFullscreen: l, isFullscreen: n }) => [
            k(e.$slots, "optionFullscreen", Q(ee({ toggleFullscreen: l, isFullscreen: n })), void 0, !0)
          ]),
          key: "8"
        } : void 0,
        e.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: b(({ toggleAnnotator: l, isAnnotator: n }) => [
            k(e.$slots, "optionAnnotator", Q(ee({ toggleAnnotator: l, isAnnotator: n })), void 0, !0)
          ]),
          key: "9"
        } : void 0,
        e.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: b(({ altCopy: l }) => [
            k(e.$slots, "optionAltCopy", Q(ee({ altCopy: l })), void 0, !0)
          ]),
          key: "10"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasXls", "hasLabel", "hasStack", "hasFullscreen", "hasAltCopy", "hasTable", "isFullscreen", "isTooltip", "isStacked", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : X("", !0),
      K("div", {
        class: "vue-ui-xy-canvas",
        style: ie(`position: relative; aspect-ratio: ${ha.value}`),
        ref_key: "container",
        ref: ue
      }, [
        K("canvas", {
          ref_key: "canvas",
          ref: z,
          style: { width: "100%", height: "100%" },
          onMousemove: t[0] || (t[0] = (l) => Bt(l)),
          onMouseleave: Wt
        }, null, 544),
        k(e.$slots, "skeleton", {}, () => [
          C($e) ? (F(), U(Cl, { key: 0 })) : X("", !0)
        ], !0),
        _a(C(Qa), {
          teleportTo: a.value.style.chart.tooltip.teleportTo,
          show: c.value.showTooltip && de.value,
          backgroundColor: a.value.style.chart.tooltip.backgroundColor,
          color: a.value.style.chart.tooltip.color,
          fontSize: a.value.style.chart.tooltip.fontSize,
          borderRadius: a.value.style.chart.tooltip.borderRadius,
          borderColor: a.value.style.chart.tooltip.borderColor,
          borderWidth: a.value.style.chart.tooltip.borderWidth,
          position: a.value.style.chart.tooltip.position,
          offsetY: a.value.style.chart.tooltip.offsetY,
          parent: C(z),
          content: Se.value,
          isFullscreen: he.value,
          backgroundOpacity: a.value.style.chart.tooltip.backgroundOpacity,
          isCustom: C(Ga)(a.value.style.chart.tooltip.customFormat),
          smooth: a.value.style.chart.tooltip.smooth,
          backdropFilter: a.value.style.chart.tooltip.backdropFilter,
          smoothForce: a.value.style.chart.tooltip.smoothForce,
          smoothSnapThreshold: a.value.style.chart.tooltip.smoothSnapThreshold
        }, {
          "tooltip-before": b(() => [
            k(e.$slots, "tooltip-before", Q(ee({ ...oa.value })), void 0, !0)
          ]),
          "tooltip-after": b(() => [
            k(e.$slots, "tooltip-after", Q(ee({ ...oa.value })), void 0, !0)
          ]),
          _: 3
        }, 8, ["teleportTo", "show", "backgroundColor", "color", "fontSize", "borderRadius", "borderColor", "borderWidth", "position", "offsetY", "parent", "content", "isFullscreen", "backgroundOpacity", "isCustom", "smooth", "backdropFilter", "smoothForce", "smoothSnapThreshold"])
      ], 4),
      a.value.style.chart.zoom.show && O.value > 6 && Y.value && va.value && !C($e) ? (F(), U(Pl, {
        key: 2,
        ref_key: "chartSlicer",
        ref: j,
        allMinimaps: bt.value,
        background: a.value.style.chart.zoom.color,
        borderColor: a.value.style.chart.backgroundColor,
        customFormat: a.value.style.chart.zoom.customFormat,
        cutNullValues: Ae.value,
        enableRangeHandles: a.value.style.chart.zoom.enableRangeHandles,
        enableSelectionDrag: a.value.style.chart.zoom.enableSelectionDrag,
        end: u.value.end,
        focusOnDrag: a.value.style.chart.zoom.focusOnDrag,
        focusRangeRatio: a.value.style.chart.zoom.focusRangeRatio,
        fontSize: a.value.style.chart.zoom.fontSize,
        immediate: !a.value.style.chart.zoom.preview.enable,
        inputColor: a.value.style.chart.zoom.color,
        isPreview: Lt.value,
        labelLeft: a.value.style.chart.grid.x.timeLabels.values[u.value.start] ? R.value[u.value.start]?.text : "",
        labelRight: a.value.style.chart.grid.x.timeLabels.values[u.value.end - 1] ? R.value[u.value.end - 1]?.text : "",
        max: O.value,
        min: 0,
        minimap: pt.value,
        minimapCompact: a.value.style.chart.zoom.minimap.compact,
        minimapFrameColor: a.value.style.chart.zoom.minimap.frameColor,
        minimapIndicatorColor: a.value.style.chart.zoom.minimap.indicatorColor,
        minimapLineColor: a.value.style.chart.zoom.minimap.lineColor,
        minimapMerged: a.value.style.chart.zoom.minimap.merged,
        minimapSelectedColor: a.value.style.chart.zoom.minimap.selectedColor,
        minimapSelectedColorOpacity: a.value.style.chart.zoom.minimap.selectedColorOpacity,
        minimapSelectedIndex: h.value,
        minimapSelectionRadius: a.value.style.chart.zoom.minimap.selectionRadius,
        preciseLabels: Aa.value?.length ? Aa.value : R.value,
        refreshEndPoint: a.value.style.chart.zoom.endIndex !== null ? a.value.style.chart.zoom.endIndex + 1 : O.value,
        refreshStartPoint: a.value.style.chart.zoom.startIndex !== null ? a.value.style.chart.zoom.startIndex : 0,
        selectColor: a.value.style.chart.zoom.highlightColor,
        selectedSeries: $.value,
        smoothMinimap: a.value.style.chart.zoom.minimap.smooth,
        start: u.value.start,
        timeLabels: R.value,
        usePreciseLabels: a.value.style.chart.grid.x.timeLabels.datetimeFormatter.enable && !a.value.style.chart.zoom.useDefaultFormat,
        textColor: a.value.style.chart.color,
        useResetSlot: a.value.style.chart.zoom.useResetSlot,
        valueEnd: u.value.end,
        valueStart: u.value.start,
        verticalHandles: a.value.style.chart.zoom.minimap.verticalHandles,
        maxWidth: a.value.style.chart.zoom.maxWidth,
        additionalMinimapHeight: a.value.style.chart.zoom.minimap.additionalHeight,
        handleType: a.value.style.chart.zoom.minimap.handleType,
        handleIconColor: a.value.style.chart.zoom.minimap.handleIconColor,
        handleBorderWidth: a.value.style.chart.zoom.minimap.handleBorderWidth,
        handleBorderColor: a.value.style.chart.zoom.minimap.handleBorderColor,
        handleFill: a.value.style.chart.zoom.minimap.handleFill,
        handleWidth: a.value.style.chart.zoom.minimap.handleWidth,
        onFutureEnd: t[1] || (t[1] = (l) => ka("end", l)),
        onFutureStart: t[2] || (t[2] = (l) => ka("start", l)),
        onReset: pe,
        onTrapMouse: dt,
        "onUpdate:end": kt,
        "onUpdate:start": Ct
      }, {
        "reset-action": b(({ reset: l }) => [
          k(e.$slots, "reset-action", Q(ee({ reset: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["allMinimaps", "background", "borderColor", "customFormat", "cutNullValues", "enableRangeHandles", "enableSelectionDrag", "end", "focusOnDrag", "focusRangeRatio", "fontSize", "immediate", "inputColor", "isPreview", "labelLeft", "labelRight", "max", "minimap", "minimapCompact", "minimapFrameColor", "minimapIndicatorColor", "minimapLineColor", "minimapMerged", "minimapSelectedColor", "minimapSelectedColorOpacity", "minimapSelectedIndex", "minimapSelectionRadius", "preciseLabels", "refreshEndPoint", "refreshStartPoint", "selectColor", "selectedSeries", "smoothMinimap", "start", "timeLabels", "usePreciseLabels", "textColor", "useResetSlot", "valueEnd", "valueStart", "verticalHandles", "maxWidth", "additionalMinimapHeight", "handleType", "handleIconColor", "handleBorderWidth", "handleBorderColor", "handleFill", "handleWidth"])) : X("", !0),
      K("div", {
        id: `legend-bottom-${ae.value}`
      }, null, 8, Nl),
      ia.value ? (F(), U(Kt, {
        key: 3,
        to: a.value.style.chart.legend.position === "top" ? `#legend-top-${ae.value}` : `#legend-bottom-${ae.value}`
      }, [
        K("div", {
          ref_key: "chartLegend",
          ref: na
        }, [
          a.value.style.chart.legend.show && Y.value ? (F(), U($l, {
            legendSet: Pe.value,
            config: Vt.value,
            key: `legend_${Ue.value}`,
            isCursorPointer: te.value,
            onClickMarker: t[3] || (t[3] = ({ i: l }) => Fe(l))
          }, {
            item: b(({ legend: l, index: n }) => [
              K("div", {
                onClick: (s) => l.segregate(),
                style: ie(`opacity:${x.value.includes(n) ? 0.5 : 1}`)
              }, ea(l.name), 13, Bl)
            ]),
            legendToggle: b(() => [
              Pe.value.length > 2 && a.value.style.chart.legend.selectAllToggle.show && !C($e) ? (F(), U(Rl, {
                key: 0,
                backgroundColor: a.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: a.value.style.chart.legend.selectAllToggle.color,
                fontSize: a.value.style.chart.legend.fontSize,
                checked: x.value.length > 0,
                isCursorPointer: te.value,
                onToggle: Ut
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : X("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config", "isCursorPointer"])) : k(e.$slots, "legend", {
            key: 1,
            legend: Pe.value
          }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : X("", !0),
      e.$slots.watermark ? (F(), Ee("div", Hl, [
        k(e.$slots, "watermark", Q(ee({ isPrinting: C(fa) || C(ma) || _e.value })), void 0, !0)
      ])) : X("", !0),
      e.$slots.source ? (F(), Ee("div", {
        key: 5,
        ref_key: "source",
        ref: ra,
        dir: "auto"
      }, [
        k(e.$slots, "source", {}, void 0, !0)
      ], 512)) : X("", !0),
      Y.value && a.value.userOptions.buttons.table ? (F(), U(Qt(Le.value.component), el({ key: 6 }, Le.value.props, {
        ref_key: "tableUnit",
        ref: fe,
        onClose: Ha
      }), Xa({
        content: b(() => [
          (F(), U(C(et), {
            key: `table_${We.value}`,
            colNames: Re.value.colNames,
            head: Re.value.head,
            body: Re.value.body,
            config: Re.value.config,
            title: a.value.table.useDialog ? "" : Le.value.title,
            withCloseButton: !a.value.table.useDialog,
            isCursorPointer: te.value,
            onClose: Ha
          }, {
            th: b(({ th: l }) => [
              K("div", { innerHTML: l }, null, 8, Wl)
            ]),
            td: b(({ td: l }) => [
              Va(ea(l), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        a.value.table.useDialog ? {
          name: "title",
          fn: b(() => [
            Va(ea(Le.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        a.value.table.useDialog ? {
          name: "actions",
          fn: b(() => [
            K("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: t[4] || (t[4] = (l) => je(a.value.userOptions.callbacks.csv)),
              style: ie({ cursor: te.value ? "pointer" : "default" })
            }, [
              _a(Il, {
                name: "fileCsv",
                stroke: Le.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : X("", !0),
      a.value.userOptions.buttons.annotator && m.value.length ? (F(), U(C(tt), {
        key: 7,
        parent: _.value,
        backgroundColor: a.value.style.chart.backgroundColor,
        color: a.value.style.chart.color,
        active: Oe.value,
        isCursorPointer: te.value,
        onClose: Ze
      }, null, 8, ["parent", "backgroundColor", "color", "active", "isCursorPointer"])) : X("", !0)
    ], 46, Dl));
  }
}, io = /* @__PURE__ */ Ol(Ul, [["__scopeId", "data-v-e0eee1ba"]]);
export {
  io as default
};
