// vue-data-ui/dist/vue-ui-quick-chart-FaxttnJY.js
import { defineAsyncComponent as Pe, ref as L, computed as O, toRefs as La, watch as Me, shallowRef as Ut, onMounted as Ca, onBeforeUnmount as Vt, nextTick as ot, watchEffect as qt, openBlock as o, createElementBlock as n, normalizeStyle as N, normalizeClass as fe, createBlock as we, unref as v, withCtx as H, renderSlot as M, normalizeProps as Z, guardReactiveProps as Y, createCommentVNode as m, createSlots as _a, toDisplayString as U, createElementVNode as $, createVNode as Ae, Fragment as x, renderList as I, Teleport as za, mergeProps as nt } from "vue";
import { j as $a, t as Ht, m as Ta, p as ee, n as Ea, a7 as Ia, q as Pa, o as Ma, s as it, a3 as Oa, D as Xe, X as Ra, a4 as Fa, a5 as ut, a6 as Oe, k as se, l as J, Z as Wt, c as Xt, K as ve, T as je, U as rt } from "./lib-GwPnJSrY.js";
import { t as jt, u as Ba } from "./useResponsive-DfdjqQps.js";
import { u as Na } from "./useConfig-CIgyKZ6R.js";
import { u as Da, B as Ua } from "./BaseScanner-BhhkKVoj.js";
import { u as Va } from "./usePrinter-DiK5R6k-.js";
import { u as qa } from "./useSvgExport-ByUukOZt.js";
import { u as st } from "./useNestedProp-BBT0mbHh.js";
import { u as Ha } from "./useTimeLabels-C-A2uZKi.js";
import { u as Wa } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Xa } from "./useChartAccessibility-9icAAmYg.js";
import { u as ja } from "./useTimeLabelCollider-CIsgDrl9.js";
import Ga from "./img-BYYzY_QU.js";
import { S as Za } from "./Slicer-DZJA9IG-.js";
import Ya from "./vue_ui_quick_chart-DMCjEjPe.js";
import { B as Ja } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as Ka } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const w = {
  LINE: "LINE",
  BAR: "BAR",
  DONUT: "DONUT"
}, Qa = ["SERIE", "SERIES", "DATA", "VALUE", "VALUES", "NUM"];
function el({ dataset: f, barLineSwitch: R = 6, debug: C = !0 }) {
  let V = null, G = null, ce = 0;
  if ((typeof f == "number" || typeof f == "string") && C && console.warn(`The provided dataset (${f}) is not sufficient to build a chart`), Ze(f) && (de(f) && (f.length < R ? V = w.BAR : V = w.LINE, G = f, ce = f.length), Ge(f))) {
    if (!al(f))
      return C && console.warn("The objects in the dataset array have a different data structure. Either keys or value types are different."), !1;
    const Ye = Object.keys(f[0]), Re = Object.values(f[0]);
    if (!Ye.some((X) => ll(X)))
      return C && console.warn("The data type of the dataset objects in the array must contain one of the following keys: DATA, SERIES, VALUE, VALUES, NUM. Casing is not important."), !1;
    Gt(Re, (X) => typeof X == "number") && (V = w.DONUT, G = f), Gt(Re, (X) => Array.isArray(X) && de(X)) && (Zt(f) > R ? V = w.LINE : V = w.BAR, ce = Zt(f), G = f.map((X) => ({
      ...X,
      data: ol(X, (te) => de(te))
    }))), f = f.map((X) => Yt(X)), G = G.map((X) => Yt(X));
  }
  return {
    dataset: f,
    type: V,
    usableDataset: G,
    maxSeriesLength: ce
  };
}
function Jt(f) {
  return !f || Ze(f) && !f.length;
}
function Ze(f) {
  return Array.isArray(f);
}
function de(f) {
  if (!Ze(f) || Jt(f)) return !1;
  const R = f.map((C) => Number(C));
  return ![...new Set(R.flatMap((C) => typeof C == "number" && !isNaN(C)))].includes(!1);
}
function Ge(f) {
  return !Ze(f) || Jt(f) || !![...new Set(f.flatMap((C) => typeof C == "object" && !Array.isArray(C)))].includes(!1) ? !1 : !f.map((C) => Object.keys(C).length > 0).includes(!1);
}
function tl(f, R) {
  const C = Object.keys(f).sort(), V = Object.keys(R).sort();
  if (C.length !== V.length)
    return !1;
  for (let G = 0; G < C.length; G += 1) {
    const ce = C[G], Le = V[G];
    if (ce !== Le || typeof f[ce] != typeof R[Le])
      return !1;
  }
  return !0;
}
function al(f) {
  if (f.length <= 1) return !0;
  for (let R = 0; R < f.length; R += 1)
    for (let C = R + 1; C < f.length; C += 1)
      if (!tl(f[R], f[C]))
        return !1;
  return !0;
}
function ll(f) {
  return Qa.includes(f.toUpperCase());
}
function Gt(f, R) {
  let C = [];
  for (let V = 0; V < f.length; V += 1)
    C.push(R(f[V]));
  return C.includes(!0);
}
function Zt(f) {
  return Math.max(...[...f].flatMap((R) => Object.values(R).filter((C) => de(C)).map((C) => C.length)));
}
function ol(f, R) {
  return Object.values(f).filter((C) => R(C))[0];
}
function Yt(f) {
  const R = {};
  for (let C in f)
    f.hasOwnProperty(C) && (R[C.toUpperCase()] = f[C]);
  return R;
}
const nl = ["id"], il = ["id"], ul = ["xmlns", "viewBox"], rl = ["x", "y", "width", "height"], sl = ["x", "y", "width", "height"], vl = ["width", "height"], dl = ["id"], cl = ["id"], hl = ["id"], fl = ["flood-color"], ml = {
  key: 0,
  class: "donut-label-connectors"
}, gl = ["d", "stroke", "stroke-width", "filter"], pl = ["cx", "cy", "r", "fill", "filter"], yl = { class: "donut" }, xl = ["d", "fill", "stroke", "stroke-width", "filter"], kl = ["d", "onMouseenter", "onMouseout", "onClick"], bl = {
  key: 1,
  class: "donut-labels quick-animation"
}, Sl = ["cx", "cy", "fill", "stroke", "filter"], wl = ["text-anchor", "x", "y", "fill", "font-size", "filter"], Al = ["text-anchor", "x", "y", "fill", "font-size", "filter"], Ll = {
  key: 2,
  class: "donut-hollow quick-animation"
}, Cl = ["x", "y", "font-size", "fill"], _l = ["x", "y", "font-size", "fill"], zl = {
  key: 0,
  class: "line-grid"
}, $l = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Tl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], El = {
  key: 1,
  class: "line-axis"
}, Il = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Pl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Ml = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Ol = ["x", "y", "font-size", "fill"], Rl = {
  key: 3,
  class: "periodLabels"
}, Fl = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Bl = { key: 0 }, Nl = ["font-size", "text-anchor", "fill", "transform"], Dl = ["font-size", "text-anchor", "fill", "transform", "innerHTML"], Ul = { class: "plots" }, Vl = { class: "line-plot-series" }, ql = ["d", "stroke", "stroke-width"], Hl = ["d", "stroke", "stroke-width"], Wl = ["d", "stroke", "stroke-width"], Xl = ["d", "stroke", "stroke-width"], jl = ["cx", "cy", "fill", "stroke"], Gl = {
  key: 4,
  class: "dataLabels"
}, Zl = ["font-size", "fill", "x", "y"], Yl = {
  key: 5,
  class: "tooltip-traps"
}, Jl = ["x", "y", "height", "width", "fill", "onMouseenter", "onMouseleave", "onClick"], Kl = {
  key: 0,
  class: "line-grid"
}, Ql = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], eo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], to = {
  key: 1,
  class: "line-axis"
}, ao = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], lo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], oo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], no = ["x", "y", "font-size", "fill"], io = {
  key: 3,
  class: "periodLabels"
}, uo = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], ro = { key: 0 }, so = ["font-size", "text-anchor", "fill", "transform"], vo = ["font-size", "text-anchor", "fill", "transform", "innerHTML"], co = { class: "plots" }, ho = ["x", "width", "height", "y", "fill", "stroke", "stroke-width"], fo = ["to"], mo = ["from", "to"], go = {
  key: 4,
  class: "dataLabels"
}, po = ["x", "y", "font-size", "fill"], yo = {
  key: 5,
  class: "tooltip-traps"
}, xo = ["x", "y", "height", "width", "fill", "onMouseenter", "onMouseleave", "onClick"], ko = {
  key: 6,
  class: "axis-labels"
}, bo = ["font-size", "fill", "x", "y"], So = ["font-size", "fill", "x", "y"], wo = ["font-size", "fill", "transform"], Ao = ["font-size", "fill", "transform"], Lo = {
  key: 5,
  class: "vue-data-ui-watermark"
}, Co = ["id"], _o = ["onClick"], zo = ["onClick"], $o = ["onClick"], To = {
  key: 1,
  class: "vue-ui-quick-chart-not-processable"
}, Eo = {
  __name: "vue-ui-quick-chart",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: [Array, Object, String, Number],
      default() {
        return null;
      }
    }
  },
  emits: ["selectDatapoint", "selectLegend", "copyAlt"],
  setup(f, { expose: R, emit: C }) {
    const V = Pe(() => import("./BaseIcon-CNx8Tzl5.js")), G = Pe(() => import("./PackageVersion-DN9d0ME5.js")), ce = Pe(() => import("./PenAndPaper-DWV1FMIi.js")), Le = Pe(() => import("./Tooltip-DL3QgjZf.js")), vt = Pe(() => import("./UserOptions-CrLqEimB.js")), { vue_ui_quick_chart: Ye } = Na(), { isThemeValid: Re, warnInvalidTheme: X } = Wa(), te = f, ne = L(null), dt = L(null), ct = L(null), ht = L(null), K = L($a()), me = L(!1), Ce = L(null), ge = L(""), oe = L(null), ft = L(null), mt = L(null), k = L([]), gt = L(0), Fe = L(0), pt = L(!1), _e = L(null), ze = L(null), ie = L(null), Be = L(null), yt = L(null), xt = L(null);
    L("#FFFFFF");
    const e = L(et()), Je = O(() => !!e.value.debug), xe = O(() => e.value.useCursorPointer), Kt = O(() => Ht({
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: D, FINAL_DATASET: Qt, manualLoading: kt } = Da({
      ...La(te),
      FINAL_CONFIG: e,
      prepareConfig: et,
      callback: () => {
        Promise.resolve().then(async () => {
          await ot(), _.value === w.LINE && e.value.lineAnimated && !D.value && fa({
            pathDuration: 1e3,
            pointDuration: 1200,
            labelDuration: 1200
          });
        });
      },
      skeletonDataset: te.config?.skeletonDataset ?? [1, 2, 3, 5, 8, 13, 21, 34, 55, 89],
      skeletonConfig: Ht({
        defaultConfig: e.value,
        userConfig: Kt.value
      })
    }), { svgRef: Ke } = Xa({ config: { text: e.value.title } }), ea = O(() => e.value.showUserOptionsOnChartHover), bt = O(() => e.value.keepUserOptionsStateOnChartLeave), Ne = L(!e.value.showUserOptionsOnChartHover), Qe = L(!1);
    function St(a = !1) {
      Qe.value = a, ea.value && (Ne.value = a);
    }
    function et() {
      const a = st({
        userConfig: te.config,
        defaultConfig: Ye
      });
      let r = {};
      const t = a.theme;
      if (t)
        if (!Re.value(a))
          X(a), r = a;
        else {
          const l = st({
            userConfig: Ya[t] || te.config,
            defaultConfig: a
          });
          r = {
            ...st({
              userConfig: te.config,
              defaultConfig: l
            }),
            customPalette: a.customPalette.length ? a.customPalette : Ta[t] || ee
          };
        }
      else
        r = a;
      return r;
    }
    Me(() => te.config, (a) => {
      D.value || (e.value = et()), A.value.width = e.value.width, A.value.height = e.value.height, Ne.value = !e.value.showUserOptionsOnChartHover, Lt(), ke.value.showTooltip = e.value.showTooltip;
    }, { deep: !0 }), Me(() => te.dataset, (a) => {
      b.value = Te.value, y.value.start = 0, y.value.end = b.value.maxSeriesLength, Fe.value += 1;
    }, { deep: !0 }), Me(() => te.dataset, (a) => {
      Array.isArray(a) && a.length > 0 && (kt.value = !1);
    }, { immediate: !0 });
    const $e = O(() => Ea(e.value.customPalette)), pe = C, Te = O(() => {
      const a = el({ debug: Je.value, dataset: Ia(Qt.value, [
        "serie",
        "series",
        "data",
        "value",
        "values",
        "num"
      ]), barLineSwitch: e.value.chartIsBarUnderDatasetLength });
      return !a && Je.value && console.error("VueUiQuickChart : Dataset is not processable"), a;
    }), b = L(Te.value), ta = O(() => !!b.value), _ = O(() => b.value ? b.value.type : null);
    Me(() => _.value, (a) => {
      a || Pa({
        componentName: "VueUiQuickChart",
        type: "dataset",
        debug: Je.value
      });
    }, { immediate: !0 });
    const { isPrinting: De, isImaging: Ue, generatePdf: wt, generateImage: At } = Va({
      elementId: `${_.value}_${K.value}`,
      fileName: e.value.title || _.value,
      options: e.value.userOptionsPrint
    }), aa = O(() => e.value.showUserOptions && !e.value.title), A = L({
      width: e.value.width,
      height: e.value.height
    }), ke = L({
      showTooltip: e.value.showTooltip
    });
    Me(e, () => {
      ke.value = {
        showTooltip: e.value.showTooltip
      };
    }, { immediate: !0 });
    const he = Ut(null), be = Ut(null);
    Ca(async () => {
      pt.value = !0, Lt();
    });
    function Lt() {
      if (Ma(te.dataset) || (kt.value = e.value.loading), e.value.responsive) {
        const a = jt(() => {
          const { width: r, height: t } = Ba({
            chart: ne.value,
            title: e.value.title ? dt.value : null,
            legend: e.value.showLegend ? ct.value : null,
            slicer: [w.BAR, w.LINE].includes(_.value) && e.value.zoomXy && b.value.maxSeriesLength > 1 ? ht.value : null,
            source: ft.value,
            noTitle: mt.value
          });
          requestAnimationFrame(() => {
            A.value.width = r, A.value.height = t;
          });
        });
        he.value && (be.value && he.value.unobserve(be.value), he.value.disconnect()), he.value = new ResizeObserver(a), be.value = ne.value.parentNode, he.value.observe(be.value);
      }
      It();
    }
    Vt(() => {
      he.value && (be.value && he.value.unobserve(be.value), he.value.disconnect());
    });
    const la = O(() => {
      switch (_.value) {
        case w.LINE:
          return `0 0 ${A.value.width <= 0 ? 10 : A.value.width} ${A.value.height <= 0 ? 10 : A.value.height}`;
        case w.BAR:
          return `0 0 ${A.value.width <= 0 ? 10 : A.value.width} ${A.value.height <= 0 ? 10 : A.value.height}`;
        case w.DONUT:
          return `0 0 ${A.value.width <= 0 ? 10 : A.value.width} ${A.value.height <= 0 ? 10 : A.value.height}`;
        default:
          return `0 0 ${A.value.width <= 0 ? 10 : A.value.width} ${A.value.height <= 0 ? 10 : A.value.height}`;
      }
    });
    function Ct(a) {
      return [...a].map((r) => r.value).reduce((r, t) => r + t, 0);
    }
    function Ee(a) {
      return e.value.blurOnHover && ![null, void 0].includes(oe.value) && oe.value !== a ? `url(#blur_${K.value})` : "";
    }
    function oa() {
      k.value.length ? k.value = [] : _.value === w.DONUT ? T.value.legend.forEach((a) => {
        k.value.push(a.id);
      }) : _.value === w.LINE ? d.value.legend.forEach((a) => {
        k.value.push(a.id);
      }) : _.value === w.BAR && s.value.legend.forEach((a) => {
        k.value.push(a.id);
      });
    }
    function _t(a, r) {
      k.value.includes(a) ? k.value = k.value.filter((t) => t !== a) : k.value.length < r && k.value.push(a);
    }
    const zt = L(null), $t = L(null), Ve = L(!1);
    function na(a, r) {
      Ve.value = !0;
      let t = a.value;
      const l = Te.value.dataset.find((u, p) => a.id === `donut_${p}`).VALUE;
      if (k.value.includes(a.id)) {
        let u = function() {
          t > l ? (Ve.value = !1, cancelAnimationFrame($t.value), b.value = {
            ...b.value,
            dataset: b.value.dataset.map((p, S) => a.id === `donut_${S}` ? {
              ...p,
              value: l,
              VALUE: l
            } : p)
          }) : (t += l * 0.025, b.value = {
            ...b.value,
            dataset: b.value.dataset.map((p, S) => a.id === `donut_${S}` ? {
              ...p,
              value: t,
              VALUE: t
            } : p)
          }, $t.value = requestAnimationFrame(u));
        };
        k.value = k.value.filter((p) => p !== a.id), u();
      } else if (r.length > 1) {
        let u = function() {
          t < l / 100 ? (Ve.value = !1, cancelAnimationFrame(zt.value), k.value.push(a.id), b.value = {
            ...b.value,
            dataset: b.value.dataset.map((p, S) => a.id === `donut_${S}` ? {
              ...p,
              value: 0,
              VALUE: 0
            } : p)
          }) : (t /= 1.1, b.value = {
            ...b.value,
            dataset: b.value.dataset.map((p, S) => a.id === `donut_${S}` ? {
              ...p,
              value: t,
              VALUE: t
            } : p)
          }, zt.value = requestAnimationFrame(u));
        };
        u();
      }
    }
    const ue = L(null);
    function ia(a) {
      ue.value = a;
    }
    const Tt = O(() => e.value.donutThicknessRatio < 0.01 ? 0.01 : e.value.donutThicknessRatio > 0.4 ? 0.4 : e.value.donutThicknessRatio), T = O(() => {
      if (_.value !== w.DONUT) return null;
      const a = b.value.dataset.map((h, z) => ({
        ...h,
        value: h.VALUE || h.DATA || h.SERIE || h.VALUES || h.NUM || 0,
        name: h.NAME || h.DESCRIPTION || h.TITLE || h.LABEL || `Serie ${z}`,
        id: `donut_${z}`
      })).map((h, z) => ({
        ...h,
        color: h.COLOR ? it(h.COLOR) : $e.value[z + e.value.paletteStartIndex] || ee[z + e.value.paletteStartIndex] || ee[(z + e.value.paletteStartIndex) % ee.length],
        immutableValue: h.value
      }));
      function r(h, z) {
        return J({
          v: isNaN(h.value / Ct(z)) ? 0 : h.value / Ct(z) * 100,
          s: "%",
          r: e.value.dataLabelRoundingPercentage
        });
      }
      function t(h) {
        return h.proportion * 100 > e.value.donutHideLabelUnderPercentage;
      }
      function l(h, z) {
        const c = Te.value.dataset.find((g, E) => `donut_${E}` === h).VALUE;
        return Math.abs(String(Number(c.toFixed(0))).length - String(Number(z.toFixed(0))).length);
      }
      function u({ datapoint: h, seriesIndex: z }) {
        Ce.value = { datapoint: h, seriesIndex: z, config: e.value, dataset: a }, oe.value = h.id;
        const c = e.value.tooltipCustomFormat;
        if (e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: h, seriesIndex: z }), je(c) && rt(() => c({
          datapoint: h,
          seriesIndex: z,
          series: a,
          config: e.value
        })))
          ge.value = c({
            datapoint: h,
            seriesIndex: z,
            series: a,
            config: e.value
          });
        else {
          let i = "";
          i += `<div style="width:100%;text-align:center;border-bottom:1px solid ${e.value.tooltipBorderColor};padding-bottom:6px;margin-bottom:3px;">${h.name}</div>`, i += `<div style="display:flex;flex-direction:row;gap:6px;align-items:center;"><svg viewBox="0 0 12 12" height="14" width="14"><circle cx="6" cy="6" r="6" stroke="none" fill="${h.color}"/></svg>`, i += `<b>${se(
            e.value.formatter,
            h.value,
            J({
              p: e.value.valuePrefix,
              v: h.value,
              s: e.value.valueSuffix,
              r: e.value.dataLabelRoundingValue
            }),
            { datapoint: h, seriesIndex: z }
          )}</b>`, i += `<span>(${J({ v: h.proportion * 100, s: "%", r: e.value.dataLabelRoundingPercentage })})</span></div>`, ge.value = `<div>${i}</div>`;
        }
        me.value = !0;
      }
      function p({ datapoint: h, seriesIndex: z }) {
        e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: h, seriesIndex: z }), me.value = !1, oe.value = null, ue.value = null;
      }
      function S({ datapoint: h, seriesIndex: z }) {
        e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: h, seriesIndex: z }), pe("selectDatapoint", h);
      }
      const P = {
        centerX: A.value.width / 2,
        centerY: A.value.height / 2
      }, F = a.filter((h) => !k.value.includes(h.id)).map((h) => h.value || 0).reduce((h, z) => h + z, 0), ae = a.map((h, z) => ({
        ...h,
        proportion: (h.value || 0) / F,
        value: h.value || 0,
        absoluteValue: Te.value.dataset.find((c, i) => `donut_${i}` === h.id).VALUE,
        shape: "circle"
      })), re = A.value.width / 2, Q = A.value.height / 2, W = A.value.height * e.value.donutRadiusRatio;
      return {
        dataset: ae.filter((h) => !k.value.includes(h.id)),
        legend: ae,
        drawingArea: P,
        displayArcPercentage: r,
        isArcBigEnough: t,
        useTooltip: u,
        killTooltip: p,
        selectDatapoint: S,
        getSpaces: l,
        total: F,
        cx: re,
        cy: Q,
        radius: W,
        chart: Oa(
          { series: a.filter((h) => !k.value.includes(h.id)) },
          re,
          Q,
          W,
          W,
          1.99999,
          2,
          1,
          360,
          105.25,
          A.value.height * Tt.value
        )
      };
    }), y = L({
      start: 0,
      end: b.value.maxSeriesLength
    });
    function ua() {
      It();
    }
    const Et = L(null);
    async function It() {
      await ot(), await ot();
      const { zoomStartIndex: a, zoomEndIndex: r } = e.value, t = Et.value;
      (a != null || r != null) && t ? (a != null && t.setStartValue(a), r != null && t.setEndValue(ra(r + 1))) : (y.value = {
        start: 0,
        end: b.value.maxSeriesLength
      }, Fe.value += 1);
    }
    function ra(a) {
      const r = b.value.maxSeriesLength;
      return a > r ? r : a < 0 || e.value.zoomStartIndex !== null && a < e.value.zoomStartIndex ? e.value.zoomStartIndex !== null ? e.value.zoomStartIndex + 1 : 1 : a;
    }
    const sa = O(() => {
      if (!e.value.zoomMinimap.show || _.value === w.DONUT) return [];
      let a = [];
      de(b.value.dataset) && (a = b.value.dataset), Ge(b.value.dataset) && (a = b.value.dataset.map((u, p) => ({
        values: u.VALUE || u.DATA || u.SERIE || u.SERIES || u.VALUES || u.NUM || 0,
        id: _.value === w.LINE ? `line_${p}` : `bar_${p}`
      })).filter((u) => !k.value.includes(u.id)));
      const r = de(a) ? a.length : Math.max(...a.map((u) => u.values.length));
      let t = [];
      if (de(a))
        t = a;
      else
        for (let u = 0; u < r; u += 1)
          t.push(a.map((p) => p.values[u] || 0).reduce((p, S) => (p || 0) + (S || 0), 0));
      const l = Math.min(...t);
      return t.map((u) => u + (l < 0 ? Math.abs(l) : 0));
    });
    function Pt() {
      let a = 0;
      return Be.value && (a = Array.from(Be.value.querySelectorAll("text")).reduce((l, u) => {
        const p = u.getComputedTextLength();
        return p > l ? p : l;
      }, 0)), a + 4;
    }
    const tt = L(0), va = jt((a) => {
      tt.value = a;
    }, 100);
    qt((a) => {
      const r = ie.value;
      if (!r) return;
      const t = new ResizeObserver((l) => {
        va(l[0].contentRect.height);
      });
      t.observe(r), a(() => t.disconnect());
    }), Vt(() => {
      tt.value = 0;
    });
    const qe = O(() => {
      let a = 0, r = 0;
      return ie.value && (r = tt.value), a + r;
    }), d = O(() => {
      if (_.value !== w.LINE) return null;
      const a = {
        height: A.value.height,
        width: A.value.width
      };
      let r = Pt();
      if (ie.value) {
        const c = ie.value.getBBox().x;
        c < 0 && (r += Math.abs(c));
      }
      const t = {
        left: r + e.value.xyPaddingLeft,
        top: e.value.xyPaddingTop,
        right: a.width - e.value.xyPaddingRight,
        bottom: a.height - e.value.xyPaddingBottom - qe.value,
        width: Math.max(10, a.width - e.value.xyPaddingLeft - e.value.xyPaddingRight - r),
        height: Math.max(10, a.height - e.value.xyPaddingTop - e.value.xyPaddingBottom - qe.value)
      };
      let l = [];
      de(b.value.dataset) && (l = [
        {
          values: b.value.dataset.slice(y.value.start, y.value.end),
          absoluteValues: b.value.dataset,
          absoluteIndices: b.value.dataset.map((c, i) => i).slice(y.value.start, y.value.end),
          name: e.value.title,
          color: $e.value[e.value.paletteStartIndex] || ee[e.value.paletteStartIndex],
          id: "line_0"
        }
      ]), Ge(b.value.dataset) && (l = b.value.dataset.map((c, i) => ({
        ...c,
        values: c.VALUE || c.DATA || c.SERIE || c.SERIES || c.VALUES || c.NUM || 0,
        name: c.NAME || c.DESCRIPTION || c.TITLE || c.LABEL || `Serie ${i}`,
        id: `line_${i}`
      })).map((c, i) => ({
        ...c,
        color: c.COLOR ? it(c.COLOR) : $e.value[i + e.value.paletteStartIndex] || ee[i + e.value.paletteStartIndex] || ee[(i + e.value.paletteStartIndex) % ee.length],
        values: c.values.slice(y.value.start, y.value.end),
        absoluteValues: c.values,
        absoluteIndices: c.values.map((g, E) => E).slice(y.value.start, y.value.end)
      })));
      const u = {
        max: Math.max(...l.filter((c) => !k.value.includes(c.id)).flatMap((c) => c.values)),
        min: Math.min(...l.filter((c) => !k.value.includes(c.id)).flatMap((c) => c.values)),
        maxSeries: Math.max(...l.map((c) => c.values.length))
      }, p = u.max === u.min ? Xe(u.min, u.min + 1, e.value.xyScaleSegments) : Xe(u.min < 0 ? u.min : 0, u.max < 0 ? 0 : u.max, e.value.xyScaleSegments), S = u.min < 0 ? Math.abs(u.min) : 0, P = u.max < 0 ? t.top : t.bottom - S / (p.max + S) * t.height, F = t.width / u.maxSeries, ae = p.ticks.map((c) => ({
        y: t.bottom - t.height * ((c + S) / (p.max + S)),
        x: t.left - 8,
        value: c
      })), re = l.map((c, i) => ({
        ...c,
        shape: "circle",
        coordinates: c.values.map((g, E) => ({
          x: t.left + F * (E + 1) - F / 2,
          y: t.bottom - (g + S) / (p.max + S) * t.height,
          value: g
        }))
      })).map((c) => {
        let i = [];
        return c.coordinates.forEach((g) => {
          i.push(`${g.x},${g.y} `);
        }), {
          ...c,
          linePath: i.join(" ")
        };
      });
      function Q(c) {
        return l.map((i) => ({
          ...i,
          value: i.values[c],
          absoluteIndex: i.absoluteIndices[c]
        })).filter((i) => !k.value.includes(i.id));
      }
      function W(c) {
        oe.value = c, ue.value = c;
        const i = Q(c);
        Ce.value = { datapoint: i, seriesIndex: c, config: e.value, dataset: l };
        const g = e.value.tooltipCustomFormat;
        if (e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: i, seriesIndex: c + y.value.start }), je(g) && rt(() => g({
          datapoint: i,
          seriesIndex: c,
          series: l,
          config: e.value
        })))
          ge.value = g({
            datapoint: i,
            seriesIndex: c,
            series: l,
            config: e.value
          });
        else {
          let E = "";
          j.value[i[0].absoluteIndex] && (E += `<div style="border-bottom:1px solid ${e.value.tooltipBorderColor};padding-bottom:6px;margin-bottom:3px;">${j.value[i[0].absoluteIndex].text}</div>`), i.forEach((B, q) => {
            E += `
                    <div style="display:flex; flex-wrap: wrap; align-items:center; gap:3px;">
                        <svg viewBox="0 0 12 12" height="14" width="12"><circle cx="6" cy="6" r="6" stroke="none" fill="${B.color}"/></svg>
                        <span>${B.name}:</span>
                        <b>${se(
              e.value.formatter,
              B.value,
              J({
                p: e.value.valuePrefix,
                v: B.value,
                s: e.value.valueSuffix,
                r: e.value.dataLabelRoundingValue
              }),
              { datapoint: B, seriesIndex: q }
            )}
                        </b>
                    </div>
                `;
          }), ge.value = E;
        }
        me.value = !0;
      }
      function h(c) {
        const i = Q(c);
        e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: i, seriesIndex: c + y.value.start }), oe.value = null, ue.value = null, me.value = !1;
      }
      function z(c) {
        const i = Q(c);
        e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: i, seriesIndex: c + y.value.start }), pe("selectDatapoint", i);
      }
      return {
        absoluteZero: P,
        dataset: re.filter((c) => !k.value.includes(c.id)),
        legend: re,
        drawingArea: t,
        extremes: u,
        slotSize: F,
        yLabels: ae,
        useTooltip: W,
        killTooltip: h,
        selectDatapoint: z
      };
    }), s = O(() => {
      if (_.value !== w.BAR) return null;
      const a = {
        height: A.value.height,
        width: A.value.width
      };
      let r = Pt();
      if (ie.value) {
        const i = ie.value.getBBox().x;
        i < 0 && (r += Math.abs(i));
      }
      const t = {
        left: r + e.value.xyPaddingLeft,
        top: e.value.xyPaddingTop,
        right: a.width - e.value.xyPaddingRight,
        bottom: a.height - e.value.xyPaddingBottom - qe.value,
        width: Math.max(10, a.width - e.value.xyPaddingLeft - e.value.xyPaddingRight - r),
        height: Math.max(10, a.height - e.value.xyPaddingTop - e.value.xyPaddingBottom - qe.value)
      };
      let l = [];
      de(b.value.dataset) && (l = [
        {
          values: b.value.dataset.slice(y.value.start, y.value.end),
          absoluteValues: b.value.dataset,
          absoluteIndices: b.value.dataset.map((i, g) => g).slice(y.value.start, y.value.end),
          name: e.value.title,
          color: $e.value[e.value.paletteStartIndex] || ee[e.value.paletteStartIndex],
          id: "bar_0"
        }
      ]), Ge(b.value.dataset) && (l = b.value.dataset.map((i, g) => ({
        ...i,
        values: i.VALUE || i.DATA || i.SERIE || i.SERIES || i.VALUES || i.NUM || 0,
        name: i.NAME || i.DESCRIPTION || i.TITLE || i.LABEL || `Serie ${g}`,
        id: `bar_${g}`
      })).map((i, g) => ({
        ...i,
        color: i.COLOR ? it(i.COLOR) : $e.value[g + e.value.paletteStartIndex] || ee[g + e.value.paletteStartIndex] || ee[(g + e.value.paletteStartIndex) % ee.length],
        values: i.values.slice(y.value.start, y.value.end),
        absoluteValues: i.values,
        absoluteIndices: i.values.map((E, B) => B).slice(y.value.start, y.value.end)
      })));
      const u = {
        max: Math.max(...l.filter((i) => !k.value.includes(i.id)).flatMap((i) => i.values)) < 0 ? 0 : Math.max(...l.filter((i) => !k.value.includes(i.id)).flatMap((i) => i.values)) ?? 1,
        min: Math.min(...l.filter((i) => !k.value.includes(i.id)).flatMap((i) => i.values)) ?? 0,
        maxSeries: Math.max(...l.filter((i) => !k.value.includes(i.id)).map((i) => i.values.length)) ?? 0
      }, p = u.min === u.max ? Xe(u.min, u.min + 1, e.value.xyScaleSegments) : Xe(u.min < 0 ? u.min : 0, u.max, e.value.xyScaleSegments), S = p.min < 0 ? Math.abs(p.min) : 0, P = t.bottom - S / (p.max + S) * t.height, F = t.width / u.maxSeries, ae = p.ticks.map((i) => ({
        y: t.bottom - t.height * ((i + S) / (p.max + S)),
        x: t.left - 8,
        value: i
      })), re = l.map((i, g) => ({
        ...i,
        shape: "square",
        coordinates: i.values.map((E, B) => {
          const q = (E + S) / (u.max + S) * t.height, le = Math.abs(E) / Math.abs(u.min) * (t.height - P), lt = S / (u.max + S) * t.height, Ie = F / l.filter((ye) => !k.value.includes(ye.id)).length - e.value.barGap / l.filter((ye) => !k.value.includes(ye.id)).length;
          return {
            x: t.left + F * B + Ie * g + e.value.barGap / 2,
            y: E > 0 ? t.bottom - q : P,
            height: E > 0 ? q - lt : le,
            value: E,
            width: Ie
          };
        })
      })), Q = l.filter((i) => !k.value.includes(i.id)).map((i, g) => ({
        ...i,
        coordinates: i.values.map((E, B) => {
          const q = (E + S) / (u.max + S) * t.height, le = Math.abs(E) / (u.max + S) * t.height, lt = S / (u.max + S) * t.height, Ie = F / l.filter((ye) => !k.value.includes(ye.id)).length - e.value.barGap / l.filter((ye) => !k.value.includes(ye.id)).length;
          return {
            x: t.left + F * B + Ie * g + e.value.barGap / 2,
            y: E > 0 ? t.bottom - q : P,
            height: E > 0 ? q - lt : le,
            value: E,
            width: Ie
          };
        })
      }));
      function W(i) {
        return l.map((g) => ({
          ...g,
          value: g.values[i],
          absoluteIndex: g.absoluteIndices[i]
        })).filter((g) => !k.value.includes(g.id));
      }
      function h(i) {
        oe.value = i, ue.value = i;
        const g = W(i);
        Ce.value = { datapoint: g, seriesIndex: i, config: e.value, dataset: l };
        const E = e.value.tooltipCustomFormat;
        if (e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: g, seriesIndex: i + y.value.start }), je(E) && rt(() => E({
          datapoint: g,
          seriesIndex: i,
          series: l,
          config: e.value
        })))
          ge.value = E({
            point: g,
            seriesIndex: i,
            series: l,
            config: e.value
          });
        else {
          let B = "";
          j.value[g[0].absoluteIndex] && (B += `<div style="border-bottom:1px solid ${e.value.tooltipBorderColor};padding-bottom:6px;margin-bottom:3px;">${j.value[g[0].absoluteIndex].text}</div>`), g.forEach((q, le) => {
            B += `
                    <div style="display:flex; flex-wrap: wrap; align-items:center; gap:3px;">
                        <svg viewBox="0 0 12 12" height="14" width="12"><rect x=0 y="0" width="12" height="12" rx="1" stroke="none" fill="${q.color}"/></svg>
                        <span>${q.name}:</span>
                        <b>${se(
              e.value.formatter,
              q.value,
              J({
                p: e.value.valuePrefix,
                v: q.value,
                s: e.value.valueSuffix,
                r: e.value.dataLabelRoundingValue
              }),
              { datapoint: q, seriesIndex: le }
            )}
                        </b>
                    </div>
                `;
          }), ge.value = B;
        }
        me.value = !0;
      }
      function z(i) {
        const g = W(i);
        e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: g, seriesIndex: i + y.value.start }), me.value = !1, oe.value = null, ue.value = null;
      }
      function c(i) {
        const g = W(i);
        e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: g, seriesIndex: i + y.value.start }), pe("selectDatapoint", g);
      }
      return {
        absoluteZero: P,
        dataset: Q.filter((i) => !k.value.includes(i.id)),
        absoluteDataset: Q,
        legend: re,
        drawingArea: t,
        extremes: u,
        slotSize: F,
        yLabels: ae,
        useTooltip: h,
        killTooltip: z,
        selectDatapoint: c
      };
    });
    function da(a) {
      if (!a) return;
      const r = a.getTotalLength();
      a.style.transition = "none", a.style.strokeDasharray = `${r}`, a.style.strokeDashoffset = `${r}`;
    }
    function Mt(a, { fromOpacity: r = "0", fromScale: t = "0.85" } = {}) {
      a.forEach((l) => {
        l.style.animation = "none", l.style.transition = "none", l.style.opacity = r, l.style.transform = `scale(${t})`, l.style.transformBox = "fill-box", l.style.transformOrigin = "50% 50%";
      });
    }
    function ca(a) {
      return a.cx?.baseVal?.value ?? parseFloat(a.getAttribute("cx"));
    }
    function ha(a) {
      const r = a.getAttribute("x");
      if (r != null) return parseFloat(r);
      const t = a.getCTM?.();
      return t ? t.e : 0;
    }
    function Ot(a, r) {
      if (!a.length) return [];
      const t = a.map((P) => ({ el: P, x: r(P) })).filter((P) => Number.isFinite(P.x));
      t.sort((P, F) => P.x - F.x);
      let l = 1 / 0;
      for (let P = 1; P < t.length; P++) {
        const F = t[P].x - t[P - 1].x;
        F > 0 && F < l && (l = F);
      }
      const u = (l === 1 / 0 ? 1 : l) / 2, p = [];
      let S = { x: t[0].x, items: [t[0].el] };
      for (let P = 1; P < t.length; P++) {
        const { x: F, el: ae } = t[P];
        Math.abs(F - S.x) <= u ? S.items.push(ae) : (p.push(S), S = { x: F, items: [ae] });
      }
      return p.push(S), p;
    }
    function fa({
      pathDuration: a,
      pathEasing: r = "ease-in-out",
      pointDuration: t,
      labelDuration: l,
      pointDelay: u = 0,
      labelDelay: p = 0,
      pointStep: S = 0,
      labelStep: P = 0,
      intraSeriesStep: F = 0
    } = {}) {
      const ae = Array.isArray(_e?.value) ? _e.value : [_e?.value].filter(Boolean), re = Array.isArray(ze?.value) ? ze.value : [ze?.value].filter(Boolean), Q = [...ae, ...re].filter(Boolean), W = ne?.value || null, h = W ? Array.from(W.querySelectorAll(".vue-ui-quick-chart-plot")) : [], z = W ? Array.from(W.querySelectorAll(".vue-ui-quick-chart-label")) : [];
      Q.forEach(da), Mt(h, { fromOpacity: "0", fromScale: "0.75" }), Mt(z, { fromOpacity: "0", fromScale: "0.98" }), h.forEach((g) => g.classList.remove("quick-animation")), z.forEach((g) => g.classList.remove("quick-animation")), W && W.offsetWidth;
      const c = h.length ? Ot(h, ca) : [], i = z.length ? Ot(z, ha) : [];
      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          Q.forEach((g) => {
            g.style.transition = `stroke-dashoffset ${a}ms ${r}`, g.style.strokeDashoffset = "0";
          }), c.forEach((g, E) => {
            g.items.forEach((B, q) => {
              const le = u + E * S + q * F;
              B.style.transition = `opacity ${t}ms ease-out ${le}ms, transform ${t}ms ease-out ${le}ms`, B.style.opacity = "1", B.style.transform = "scale(1)";
            });
          }), i.forEach((g, E) => {
            g.items.forEach((B, q) => {
              const le = p + E * P + q * F;
              B.style.transition = `opacity ${l}ms ease-out ${le}ms, transform ${l}ms ease-out ${le}ms`, B.style.opacity = "1", B.style.transform = "scale(1)";
            });
          });
        });
      });
    }
    const ma = O(() => {
      if (_.value === w.LINE)
        return d.value.legend.map((a) => (Math.min(...a.absoluteValues.map((r) => r ?? 0)), {
          ...a,
          isVisible: !k.value.includes(a.id),
          type: "line",
          series: a.absoluteValues
        }));
      if (_.value === w.BAR)
        return s.value.legend.map((a) => (Math.min(...a.absoluteValues.map((r) => r ?? 0)), {
          ...a,
          isVisible: !k.value.includes(a.id),
          type: "bar",
          series: a.absoluteValues
        }));
    }), j = L([]);
    let Rt = 0;
    qt(() => {
      const a = ++Rt;
      (async () => {
        const r = await Ha({
          values: e.value.xyPeriods,
          maxDatapoints: b.value.maxSeriesLength,
          formatter: e.value.datetimeFormatter,
          start: y.value.start,
          end: y.value.end
        });
        a === Rt && (j.value = r);
      })();
    });
    const Se = O(() => {
      const a = e.value.xyPeriodsModulo;
      return e.value.xyPeriods.length ? Math.min(a, [...new Set(j.value.map((r) => r.text))].length) : a;
    }), He = L(!1);
    function Ft(a) {
      He.value = a, gt.value += 1;
    }
    function Bt() {
      ke.value.showTooltip = !ke.value.showTooltip;
    }
    const We = L(!1);
    function at() {
      We.value = !We.value;
    }
    async function ga({ scale: a = 2 } = {}) {
      if (!ne.value) return;
      const { width: r, height: t } = ne.value.getBoundingClientRect(), l = r / t, { imageUri: u, base64: p } = await Ga({ domElement: ne.value, base64: !0, img: !0, scale: a });
      return {
        imageUri: u,
        base64: p,
        title: e.value.title,
        width: r,
        height: t,
        aspectRatio: l
      };
    }
    const pa = O(() => A.value.width), ya = O(() => A.value.height);
    ja({
      timeLabelsEls: ie,
      timeLabels: j,
      slicer: y,
      configRef: e,
      rotationPath: ["xyPeriodLabelsRotation"],
      autoRotatePath: ["xyPeriodLabelsAutoRotate", "enable"],
      isAutoSize: !1,
      rotation: e.value.xyPeriodLabelsAutoRotate.angle,
      height: ya.value,
      width: pa.value
    });
    const xa = O(() => e.value.backgroundColor), ka = O(() => _.value === w.DONUT ? T.value.legend : _.value === w.LINE ? d.value.legend : s.value.legend), ba = O(() => ({
      show: e.value.showLegend,
      bold: !1,
      backgroundColor: e.value.backgroundColor,
      color: e.value.color,
      fontSize: e.value.legendFontSize,
      position: e.value.legendPosition
    })), Sa = O(() => ({
      text: e.value.title,
      color: e.value.color,
      fontSize: e.value.titleFontSize,
      bold: e.value.titleBold,
      textAlign: e.value.titleTextAlign,
      subtitle: {
        text: ""
      }
    })), { exportSvg: wa, getSvg: Aa } = qa({
      svg: Ke,
      title: Sa,
      legend: ba,
      legendItems: ka,
      backgroundColor: xa
    });
    async function Nt({ isCb: a }) {
      if (a) {
        const { blob: r, url: t, text: l, dataUrl: u } = await Aa();
        e.value.userOptionsCallbacks.svg({ blob: r, url: t, text: l, dataUrl: u });
      } else
        wa();
    }
    async function Dt() {
      if (pe("copyAlt", {
        config: e.value,
        dataset: {
          line: d.value,
          bar: s.value,
          donut: T.value
        }
      }), !e.value.userOptionsCallbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptionsCallbacks.altCopy({
        config: e.value,
        dataset: {
          line: d.value,
          bar: s.value,
          donut: T.value
        }
      }));
    }
    return R({
      getImage: ga,
      generatePdf: wt,
      generateImage: At,
      generateSvg: Nt,
      toggleTooltip: Bt,
      toggleAnnotator: at,
      toggleFullscreen: Ft,
      copyAlt: Dt
    }), (a, r) => ta.value ? (o(), n("div", {
      key: 0,
      id: `${_.value}_${K.value}`,
      ref_key: "quickChart",
      ref: ne,
      class: fe({ "vue-data-ui-component": !0, "vue-ui-quick-chart": !0, "vue-data-ui-wrapper-fullscreen": He.value }),
      style: N(`background:${e.value.backgroundColor};color:${e.value.color};font-family:${e.value.fontFamily}; position: relative; ${e.value.responsive ? "height: 100%" : ""}`),
      onMouseenter: r[2] || (r[2] = () => St(!0)),
      onMouseleave: r[3] || (r[3] = () => St(!1))
    }, [
      e.value.userOptionsButtons.annotator ? (o(), we(v(ce), {
        key: 0,
        svgRef: v(Ke),
        backgroundColor: e.value.backgroundColor,
        color: e.value.color,
        active: We.value,
        isCursorPointer: xe.value,
        onClose: at
      }, {
        "annotator-action-close": H(() => [
          M(a.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": H(({ color: t }) => [
          M(a.$slots, "annotator-action-color", Z(Y({ color: t })), void 0, !0)
        ]),
        "annotator-action-draw": H(({ mode: t }) => [
          M(a.$slots, "annotator-action-draw", Z(Y({ mode: t })), void 0, !0)
        ]),
        "annotator-action-undo": H(({ disabled: t }) => [
          M(a.$slots, "annotator-action-undo", Z(Y({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-redo": H(({ disabled: t }) => [
          M(a.$slots, "annotator-action-redo", Z(Y({ disabled: t })), void 0, !0)
        ]),
        "annotator-action-delete": H(({ disabled: t }) => [
          M(a.$slots, "annotator-action-delete", Z(Y({ disabled: t })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : m("", !0),
      aa.value ? (o(), n("div", {
        key: 1,
        ref_key: "noTitle",
        ref: mt,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : m("", !0),
      e.value.showUserOptions && (bt.value || Ne.value) ? (o(), we(v(vt), {
        ref: "details",
        key: `user_option_${gt.value}`,
        backgroundColor: e.value.backgroundColor,
        color: e.value.color,
        isPrinting: v(De),
        isImaging: v(Ue),
        uid: K.value,
        hasTooltip: e.value.userOptionsButtons.tooltip && e.value.showTooltip,
        hasPdf: e.value.userOptionsButtons.pdf,
        hasImg: e.value.userOptionsButtons.img,
        hasSvg: e.value.userOptionsButtons.svg,
        hasFullscreen: e.value.userOptionsButtons.fullscreen,
        hasAltCopy: e.value.userOptionsButtons.altCopy,
        hasXls: !1,
        isTooltip: ke.value.showTooltip,
        isFullscreen: He.value,
        titles: { ...e.value.userOptionsButtonTitles },
        chartElement: ne.value,
        position: e.value.userOptionsPosition,
        hasAnnotator: e.value.userOptionsButtons.annotator,
        isAnnotation: We.value,
        callbacks: e.value.userOptionsCallbacks,
        printScale: e.value.userOptionsPrint.scale,
        isCursorPointer: xe.value,
        onToggleFullscreen: Ft,
        onGeneratePdf: v(wt),
        onGenerateImage: v(At),
        onGenerateSvg: Nt,
        onToggleTooltip: Bt,
        onToggleAnnotator: at,
        onCopyAlt: Dt,
        style: N({
          visibility: bt.value ? Ne.value ? "visible" : "hidden" : "visible"
        })
      }, _a({ _: 2 }, [
        a.$slots.menuIcon ? {
          name: "menuIcon",
          fn: H(({ isOpen: t, color: l }) => [
            M(a.$slots, "menuIcon", Z(Y({ isOpen: t, color: l })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        a.$slots.optionTooltip ? {
          name: "optionTooltip",
          fn: H(() => [
            M(a.$slots, "optionTooltip", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        a.$slots.optionPdf ? {
          name: "optionPdf",
          fn: H(() => [
            M(a.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        a.$slots.optionImg ? {
          name: "optionImg",
          fn: H(() => [
            M(a.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        a.$slots.optionSvg ? {
          name: "optionSvg",
          fn: H(() => [
            M(a.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        a.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: H(({ toggleFullscreen: t, isFullscreen: l }) => [
            M(a.$slots, "optionFullscreen", Z(Y({ toggleFullscreen: t, isFullscreen: l })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        a.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: H(({ toggleAnnotator: t, isAnnotator: l }) => [
            M(a.$slots, "optionAnnotator", Z(Y({ toggleAnnotator: t, isAnnotator: l })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        a.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: H(({ altCopy: t }) => [
            M(a.$slots, "optionAltCopy", Z(Y({ altCopy: t })), void 0, !0)
          ]),
          key: "7"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasTooltip", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "isTooltip", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "isCursorPointer", "onGeneratePdf", "onGenerateImage", "style"])) : m("", !0),
      e.value.title ? (o(), n("div", {
        key: 3,
        ref_key: "quickChartTitle",
        ref: dt,
        class: "vue-ui-quick-chart-title",
        style: N(`background:transparent;color:${e.value.color};font-size:${e.value.titleFontSize}px;font-weight:${e.value.titleBold ? "bold" : "normal"};text-align:${e.value.titleTextAlign}`)
      }, U(e.value.title), 5)) : m("", !0),
      $("div", {
        id: `legend-top-${K.value}`
      }, null, 8, il),
      _.value ? (o(), n("svg", {
        key: 4,
        ref_key: "svgRef",
        ref: Ke,
        xmlns: v(Ra),
        viewBox: la.value,
        style: N(`max-width:100%;overflow:visible;background:transparent;color:${e.value.color}`)
      }, [
        Ae(v(G)),
        a.$slots["chart-background"] && _.value === w.BAR ? (o(), n("foreignObject", {
          key: 0,
          x: s.value.drawingArea.left,
          y: s.value.drawingArea.top,
          width: s.value.drawingArea.width,
          height: s.value.drawingArea.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          M(a.$slots, "chart-background", {}, void 0, !0)
        ], 8, rl)) : m("", !0),
        a.$slots["chart-background"] && _.value === w.LINE ? (o(), n("foreignObject", {
          key: 1,
          x: d.value.drawingArea.left,
          y: d.value.drawingArea.top,
          width: d.value.drawingArea.width,
          height: d.value.drawingArea.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          M(a.$slots, "chart-background", {}, void 0, !0)
        ], 8, sl)) : m("", !0),
        a.$slots["chart-background"] && _.value === w.DONUT ? (o(), n("foreignObject", {
          key: 2,
          x: 0,
          y: 0,
          width: A.value.width,
          height: A.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          M(a.$slots, "chart-background", {}, void 0, !0)
        ], 8, vl)) : m("", !0),
        $("defs", null, [
          $("filter", {
            id: `blur_${K.value}`,
            x: "-50%",
            y: "-50%",
            width: "200%",
            height: "200%"
          }, [
            $("feGaussianBlur", {
              in: "SourceGraphic",
              stdDeviation: 2,
              id: `blur_std_${K.value}`
            }, null, 8, cl),
            r[4] || (r[4] = $("feColorMatrix", {
              type: "saturate",
              values: "0"
            }, null, -1))
          ], 8, dl),
          $("filter", {
            id: `shadow_${K.value}`,
            "color-interpolation-filters": "sRGB"
          }, [
            $("feDropShadow", {
              dx: "0",
              dy: "0",
              stdDeviation: "10",
              "flood-opacity": "0.5",
              "flood-color": e.value.donutShadowColor
            }, null, 8, fl)
          ], 8, hl)
        ]),
        _.value === w.DONUT ? (o(), n(x, { key: 3 }, [
          e.value.showDataLabels ? (o(), n("g", ml, [
            (o(!0), n(x, null, I(T.value.chart, (t, l) => (o(), n(x, null, [
              T.value.isArcBigEnough(t) ? (o(), n("path", {
                key: 0,
                d: v(Fa)(t, { x: A.value.width / 2, y: A.value.height / 2 }, 16, 16, !1, !1, A.value.height * Tt.value, 12, e.value.donutCurvedMarkers),
                stroke: t.color,
                "stroke-width": e.value.donutLabelMarkerStrokeWidth,
                "stroke-linecap": "round",
                "stroke-linejoin": "round",
                fill: "none",
                filter: Ee(t.id)
              }, null, 8, gl)) : m("", !0)
            ], 64))), 256))
          ])) : m("", !0),
          $("circle", {
            cx: T.value.cx,
            cy: T.value.cy,
            r: T.value.radius,
            fill: e.value.backgroundColor,
            filter: e.value.donutUseShadow ? `url(#shadow_${K.value})` : ""
          }, null, 8, pl),
          $("g", yl, [
            (o(!0), n(x, null, I(T.value.chart, (t, l) => (o(), n("path", {
              d: t.arcSlice,
              fill: t.color,
              stroke: e.value.donutStroke || e.value.backgroundColor,
              "stroke-width": e.value.donutStrokeWidth,
              filter: Ee(t.id)
            }, null, 8, xl))), 256)),
            (o(!0), n(x, null, I(T.value.chart, (t, l) => (o(), n("path", {
              d: t.arcSlice,
              fill: "transparent",
              onMouseenter: (u) => T.value.useTooltip({ datapoint: t, seriesIndex: l }),
              onMouseout: (u) => T.value.killTooltip({ datapoint: t, seriesIndex: l }),
              onClick: (u) => T.value.selectDatapoint({ datapoint: t, seriesIndex: l })
            }, null, 40, kl))), 256))
          ]),
          e.value.showDataLabels ? (o(), n("g", bl, [
            (o(!0), n(x, null, I(T.value.chart, (t, l) => (o(), n(x, null, [
              T.value.isArcBigEnough(t) ? (o(), n("circle", {
                key: 0,
                cx: v(Oe)(t).x,
                cy: v(ut)(t) - 3.7,
                fill: t.color,
                stroke: e.value.backgroundColor,
                "stroke-width": 1,
                r: 3,
                filter: Ee(t.id)
              }, null, 8, Sl)) : m("", !0),
              T.value.isArcBigEnough(t) ? (o(), n("text", {
                key: 1,
                "text-anchor": v(Oe)(t, !0, 20).anchor,
                x: v(Oe)(t, !0).x,
                y: v(ut)(t),
                fill: e.value.color,
                "font-size": e.value.dataLabelFontSize,
                filter: Ee(t.id)
              }, U(T.value.displayArcPercentage(t, T.value.chart)) + " (" + U(v(se)(
                e.value.formatter,
                t.value,
                v(J)({
                  p: e.value.valuePrefix,
                  v: t.value,
                  s: e.value.valueSuffix,
                  r: e.value.dataLabelRoundingValue
                }),
                { datapoint: t, seriesIndex: l }
              )) + ") ", 9, wl)) : m("", !0),
              T.value.isArcBigEnough(t, !0, 20) ? (o(), n("text", {
                key: 2,
                "text-anchor": v(Oe)(t).anchor,
                x: v(Oe)(t, !0).x,
                y: v(ut)(t) + e.value.dataLabelFontSize,
                fill: e.value.color,
                "font-size": e.value.dataLabelFontSize,
                filter: Ee(t.id)
              }, U(t.name), 9, Al)) : m("", !0)
            ], 64))), 256))
          ])) : m("", !0),
          e.value.donutShowTotal ? (o(), n("g", Ll, [
            $("text", {
              "text-anchor": "middle",
              x: T.value.drawingArea.centerX,
              y: T.value.drawingArea.centerY - e.value.donutTotalLabelFontSize / 2,
              "font-size": e.value.donutTotalLabelFontSize,
              fill: e.value.color
            }, U(e.value.donutTotalLabelText), 9, Cl),
            $("text", {
              "text-anchor": "middle",
              x: T.value.drawingArea.centerX,
              y: T.value.drawingArea.centerY + e.value.donutTotalLabelFontSize,
              "font-size": e.value.donutTotalLabelFontSize,
              fill: e.value.color
            }, U(v(J)({
              p: e.value.valuePrefix,
              v: T.value.total,
              s: e.value.valueSuffix,
              r: e.value.dataLabelRoundingValue
            })), 9, _l)
          ])) : m("", !0)
        ], 64)) : m("", !0),
        _.value === w.LINE ? (o(), n(x, { key: 4 }, [
          e.value.xyShowGrid ? (o(), n("g", zl, [
            (o(!0), n(x, null, I(d.value.yLabels, (t) => (o(), n(x, null, [
              t.y <= d.value.drawingArea.bottom ? (o(), n("line", {
                key: 0,
                x1: d.value.drawingArea.left,
                x2: d.value.drawingArea.right,
                y1: t.y,
                y2: t.y,
                stroke: e.value.xyGridStroke,
                "stroke-width": e.value.xyGridStrokeWidth,
                "stroke-linecap": "round"
              }, null, 8, $l)) : m("", !0)
            ], 64))), 256)),
            (o(!0), n(x, null, I(d.value.extremes.maxSeries + 1, (t, l) => (o(), n("line", {
              x1: d.value.drawingArea.left + d.value.slotSize * l,
              x2: d.value.drawingArea.left + d.value.slotSize * l,
              y1: d.value.drawingArea.top,
              y2: d.value.drawingArea.bottom,
              stroke: e.value.xyGridStroke,
              "stroke-width": e.value.xyGridStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, Tl))), 256))
          ])) : m("", !0),
          e.value.xyShowAxis ? (o(), n("g", El, [
            $("line", {
              x1: d.value.drawingArea.left,
              x2: d.value.drawingArea.left,
              y1: d.value.drawingArea.top,
              y2: d.value.drawingArea.bottom,
              stroke: e.value.xyAxisStroke,
              "stroke-width": e.value.xyAxisStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, Il),
            $("line", {
              x1: d.value.drawingArea.left,
              x2: d.value.drawingArea.right,
              y1: isNaN(d.value.absoluteZero) ? d.value.drawingArea.bottom : d.value.absoluteZero,
              y2: isNaN(d.value.absoluteZero) ? d.value.drawingArea.bottom : d.value.absoluteZero,
              stroke: e.value.xyAxisStroke,
              "stroke-width": e.value.xyAxisStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, Pl)
          ])) : m("", !0),
          e.value.xyShowScale ? (o(), n("g", {
            key: 2,
            class: "yLabels",
            ref_key: "scaleLabels",
            ref: Be
          }, [
            (o(!0), n(x, null, I(d.value.yLabels, (t, l) => (o(), n(x, null, [
              t.y <= d.value.drawingArea.bottom ? (o(), n("line", {
                key: 0,
                x1: t.x + 4,
                x2: d.value.drawingArea.left,
                y1: t.y,
                y2: t.y,
                stroke: e.value.xyAxisStroke,
                "stroke-width": e.value.xyAxisStrokeWidth,
                "stroke-linecap": "round"
              }, null, 8, Ml)) : m("", !0),
              t.y <= d.value.drawingArea.bottom ? (o(), n("text", {
                key: 1,
                x: t.x,
                y: t.y + e.value.xyLabelsYFontSize / 3,
                "text-anchor": "end",
                "font-size": e.value.xyLabelsYFontSize,
                fill: e.value.color
              }, U(v(se)(
                e.value.formatter,
                t.value,
                v(J)({
                  p: e.value.valuePrefix,
                  v: t.value,
                  s: e.value.valueSuffix,
                  r: e.value.dataLabelRoundingValue
                }),
                { datapoint: t, seriesIndex: l }
              )), 9, Ol)) : m("", !0)
            ], 64))), 256))
          ], 512)) : m("", !0),
          e.value.xyShowScale && e.value.xyPeriods.length ? (o(), n("g", Rl, [
            (o(!0), n(x, null, I(j.value.map((t) => t.text), (t, l) => (o(), n(x, null, [
              !e.value.xyPeriodsShowOnlyAtModulo || e.value.xyPeriodsShowOnlyAtModulo && l % Math.floor((y.value.end - y.value.start) / Se.value) === 0 || y.value.end - y.value.start <= Se.value ? (o(), n("line", {
                key: 0,
                x1: d.value.drawingArea.left + d.value.slotSize * (l + 1) - d.value.slotSize / 2,
                x2: d.value.drawingArea.left + d.value.slotSize * (l + 1) - d.value.slotSize / 2,
                y1: d.value.drawingArea.bottom,
                y2: d.value.drawingArea.bottom + 4,
                stroke: e.value.xyAxisStroke,
                "stroke-width": e.value.xyAxisStrokeWidth,
                "stroke-linecap": "round"
              }, null, 8, Fl)) : m("", !0)
            ], 64))), 256)),
            $("g", {
              ref_key: "timeLabelsEls",
              ref: ie
            }, [
              (o(!0), n(x, null, I(j.value.map((t) => t.text), (t, l) => (o(), n(x, null, [
                !e.value.xyPeriodsShowOnlyAtModulo || e.value.xyPeriodsShowOnlyAtModulo && l % Math.floor((y.value.end - y.value.start) / Se.value) === 0 || y.value.end - y.value.start <= Se.value ? (o(), n("g", Bl, [
                  String(t).includes(`
`) ? (o(), n("text", {
                    key: 1,
                    class: "vue-data-ui-time-label",
                    "font-size": e.value.xyLabelsXFontSize,
                    "text-anchor": e.value.xyPeriodLabelsRotation > 0 ? "start" : e.value.xyPeriodLabelsRotation < 0 ? "end" : "middle",
                    fill: e.value.color,
                    transform: `translate(${d.value.drawingArea.left + d.value.slotSize * (l + 1) - d.value.slotSize / 2}, ${d.value.drawingArea.bottom + e.value.xyLabelsXFontSize + 6}), rotate(${e.value.xyPeriodLabelsRotation})`,
                    innerHTML: v(Wt)({
                      content: String(t),
                      fontSize: e.value.xyLabelsXFontSize,
                      fill: e.value.color,
                      x: 0,
                      y: 0
                    })
                  }, null, 8, Dl)) : (o(), n("text", {
                    key: 0,
                    class: "vue-data-ui-time-label",
                    "font-size": e.value.xyLabelsXFontSize,
                    "text-anchor": e.value.xyPeriodLabelsRotation > 0 ? "start" : e.value.xyPeriodLabelsRotation < 0 ? "end" : "middle",
                    fill: e.value.color,
                    transform: `translate(${d.value.drawingArea.left + d.value.slotSize * (l + 1) - d.value.slotSize / 2}, ${d.value.drawingArea.bottom + e.value.xyLabelsXFontSize + 6}), rotate(${e.value.xyPeriodLabelsRotation})`
                  }, U(t), 9, Nl))
                ])) : m("", !0)
              ], 64))), 256))
            ], 512)
          ])) : m("", !0),
          $("g", Ul, [
            (o(!0), n(x, null, I(d.value.dataset, (t, l) => (o(), n("g", Vl, [
              e.value.lineSmooth ? (o(), n(x, { key: 0 }, [
                $("path", {
                  ref_for: !0,
                  ref_key: "pathWrapper",
                  ref: _e,
                  d: `M ${v(Xt)(t.coordinates)}`,
                  stroke: e.value.backgroundColor,
                  "stroke-width": e.value.lineStrokeWidth + 1,
                  "stroke-linecap": "round",
                  fill: "none",
                  class: fe({ "quick-animation": !v(D), "vue-data-ui-line-animated": e.value.lineAnimated && !v(D) }),
                  style: N({ transition: v(D) ? void 0 : "all 0.3s ease-in-out" })
                }, null, 14, ql),
                $("path", {
                  ref_for: !0,
                  ref_key: "pathTop",
                  ref: ze,
                  d: `M ${v(Xt)(t.coordinates)}`,
                  stroke: t.color,
                  "stroke-width": e.value.lineStrokeWidth,
                  "stroke-linecap": "round",
                  fill: "none",
                  class: fe({ "quick-animation": !v(D), "vue-data-ui-line-animated": e.value.lineAnimated && !v(D) }),
                  style: N({ transition: v(D) ? void 0 : "all 0.3s ease-in-out" })
                }, null, 14, Hl)
              ], 64)) : (o(), n(x, { key: 1 }, [
                $("path", {
                  ref_for: !0,
                  ref_key: "pathWrapper",
                  ref: _e,
                  d: `M ${t.linePath}`,
                  stroke: e.value.backgroundColor,
                  "stroke-width": e.value.lineStrokeWidth + 1,
                  "stroke-linecap": "round",
                  fill: "none",
                  class: fe({ "quick-animation": !v(D), "vue-data-ui-line-animated": e.value.lineAnimated && !v(D) }),
                  style: N({ transition: v(D) ? void 0 : "all 0.3s ease-in-out" })
                }, null, 14, Wl),
                $("path", {
                  ref_for: !0,
                  ref_key: "pathTop",
                  ref: ze,
                  d: `M ${t.linePath}`,
                  stroke: t.color,
                  "stroke-width": e.value.lineStrokeWidth,
                  "stroke-linecap": "round",
                  fill: "none",
                  class: fe({ "quick-animation": !v(D), "vue-data-ui-line-animated": e.value.lineAnimated && !v(D) }),
                  style: N({ transition: v(D) ? void 0 : "all 0.3s ease-in-out" })
                }, null, 14, Xl)
              ], 64)),
              (o(!0), n(x, null, I(t.coordinates, (u, p) => (o(), n("circle", {
                cx: u.x,
                cy: v(ve)(u.y),
                r: 3,
                fill: t.color,
                stroke: e.value.backgroundColor,
                "stroke-width": "0.5",
                class: fe({ "vue-ui-quick-chart-plot": !0, "quick-animation": !v(D) }),
                style: N({ transition: v(D) ? void 0 : "all 0.3s ease-in-out" })
              }, null, 14, jl))), 256))
            ]))), 256))
          ]),
          e.value.showDataLabels ? (o(), n("g", Gl, [
            (o(!0), n(x, null, I(d.value.dataset, (t, l) => (o(), n(x, null, [
              (o(!0), n(x, null, I(t.coordinates, (u, p) => (o(), n("text", {
                "text-anchor": "middle",
                "font-size": e.value.dataLabelFontSize,
                fill: t.color,
                x: u.x,
                y: v(ve)(u.y) - e.value.dataLabelFontSize / 2,
                class: fe({ "vue-ui-quick-chart-label": !0, "quick-animation": !v(D) }),
                style: N({ transition: v(D) ? void 0 : "all 0.3s ease-in-out" })
              }, U(v(se)(
                e.value.formatter,
                v(ve)(u.value),
                v(J)({
                  p: e.value.valuePrefix,
                  v: v(ve)(u.value),
                  s: e.value.valueSuffix,
                  r: e.value.dataLabelRoundingValue
                }),
                { datapoint: u, seriesIndex: p }
              )), 15, Zl))), 256))
            ], 64))), 256))
          ])) : m("", !0),
          Qe.value ? (o(), n("g", Yl, [
            (o(!0), n(x, null, I(d.value.extremes.maxSeries, (t, l) => (o(), n("rect", {
              x: d.value.drawingArea.left + l * d.value.slotSize,
              y: d.value.drawingArea.top,
              height: d.value.drawingArea.height <= 0 ? 1e-5 : d.value.drawingArea.height,
              width: d.value.slotSize <= 0 ? 1e-5 : d.value.slotSize,
              fill: [oe.value, ue.value].includes(l) ? e.value.xyHighlighterColor : "transparent",
              style: N(`opacity:${e.value.xyHighlighterOpacity}`),
              onMouseenter: (u) => d.value.useTooltip(l),
              onMouseleave: (u) => d.value.killTooltip(l),
              onClick: (u) => d.value.selectDatapoint(l)
            }, null, 44, Jl))), 256))
          ])) : m("", !0)
        ], 64)) : m("", !0),
        _.value === w.BAR ? (o(), n(x, { key: 5 }, [
          e.value.xyShowGrid ? (o(), n("g", Kl, [
            (o(!0), n(x, null, I(s.value.yLabels, (t) => (o(), n(x, null, [
              t.y <= s.value.drawingArea.bottom ? (o(), n("line", {
                key: 0,
                x1: s.value.drawingArea.left,
                x2: s.value.drawingArea.right,
                y1: t.y,
                y2: t.y,
                stroke: e.value.xyGridStroke,
                "stroke-width": e.value.xyGridStrokeWidth,
                "stroke-linecap": "round"
              }, null, 8, Ql)) : m("", !0)
            ], 64))), 256)),
            k.value.length < s.value.legend.length ? (o(!0), n(x, { key: 0 }, I(s.value.extremes.maxSeries + 1, (t, l) => (o(), n("line", {
              x1: s.value.drawingArea.left + s.value.slotSize * l,
              x2: s.value.drawingArea.left + s.value.slotSize * l,
              y1: s.value.drawingArea.top,
              y2: s.value.drawingArea.bottom,
              stroke: e.value.xyGridStroke,
              "stroke-width": e.value.xyGridStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, eo))), 256)) : m("", !0)
          ])) : m("", !0),
          e.value.xyShowAxis ? (o(), n("g", to, [
            $("line", {
              x1: s.value.drawingArea.left,
              x2: s.value.drawingArea.left,
              y1: s.value.drawingArea.top,
              y2: s.value.drawingArea.bottom,
              stroke: e.value.xyAxisStroke,
              "stroke-width": e.value.xyAxisStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, ao),
            $("line", {
              x1: s.value.drawingArea.left,
              x2: s.value.drawingArea.right,
              y1: isNaN(s.value.absoluteZero) ? s.value.drawingArea.bottom : s.value.absoluteZero,
              y2: isNaN(s.value.absoluteZero) ? s.value.drawingArea.bottom : s.value.absoluteZero,
              stroke: e.value.xyAxisStroke,
              "stroke-width": e.value.xyAxisStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, lo)
          ])) : m("", !0),
          e.value.xyShowScale ? (o(), n("g", {
            key: 2,
            class: "yLabels",
            ref_key: "scaleLabels",
            ref: Be
          }, [
            (o(!0), n(x, null, I(s.value.yLabels, (t, l) => (o(), n(x, null, [
              t.y <= s.value.drawingArea.bottom ? (o(), n("line", {
                key: 0,
                x1: t.x + 4,
                x2: s.value.drawingArea.left,
                y1: t.y,
                y2: t.y,
                stroke: e.value.xyAxisStroke,
                "stroke-width": e.value.xyAxisStrokeWidth,
                "stroke-linecap": "round"
              }, null, 8, oo)) : m("", !0),
              t.y <= s.value.drawingArea.bottom ? (o(), n("text", {
                key: 1,
                x: t.x,
                y: t.y + e.value.xyLabelsYFontSize / 3,
                "text-anchor": "end",
                "font-size": e.value.xyLabelsYFontSize,
                fill: e.value.color
              }, U(v(se)(
                e.value.formatter,
                t.value,
                v(J)({
                  p: e.value.valuePrefix,
                  v: t.value,
                  s: e.value.valueSuffix,
                  r: e.value.dataLabelRoundingValue
                }),
                { datapoint: t, seriesIndex: l }
              )), 9, no)) : m("", !0)
            ], 64))), 256))
          ], 512)) : m("", !0),
          e.value.xyShowScale && e.value.xyPeriods.length ? (o(), n("g", io, [
            (o(!0), n(x, null, I(e.value.xyPeriods.slice(y.value.start, y.value.end), (t, l) => (o(), n("line", {
              x1: s.value.drawingArea.left + s.value.slotSize * (l + 1) - s.value.slotSize / 2,
              x2: s.value.drawingArea.left + s.value.slotSize * (l + 1) - s.value.slotSize / 2,
              y1: s.value.drawingArea.bottom,
              y2: s.value.drawingArea.bottom + 4,
              stroke: e.value.xyAxisStroke,
              "stroke-width": e.value.xyAxisStrokeWidth,
              "stroke-linecap": "round"
            }, null, 8, uo))), 256)),
            $("g", {
              ref_key: "timeLabelsEls",
              ref: ie
            }, [
              (o(!0), n(x, null, I(j.value.map((t) => t.text), (t, l) => (o(), n(x, null, [
                !e.value.xyPeriodsShowOnlyAtModulo || e.value.xyPeriodsShowOnlyAtModulo && l % Math.floor((y.value.end - y.value.start) / Se.value) === 0 || y.value.end - y.value.start <= Se.value ? (o(), n("g", ro, [
                  String(t).includes(`
`) ? (o(), n("text", {
                    key: 1,
                    class: "vue-data-ui-time-label",
                    "font-size": e.value.xyLabelsXFontSize,
                    "text-anchor": e.value.xyPeriodLabelsRotation > 0 ? "start" : e.value.xyPeriodLabelsRotation < 0 ? "end" : "middle",
                    fill: e.value.color,
                    transform: `translate(${s.value.drawingArea.left + s.value.slotSize * (l + 1) - s.value.slotSize / 2}, ${s.value.drawingArea.bottom + e.value.xyLabelsXFontSize + 6}), rotate(${e.value.xyPeriodLabelsRotation})`,
                    innerHTML: v(Wt)({
                      content: String(t),
                      fontSize: e.value.xyLabelsXFontSize,
                      fill: e.value.color,
                      x: 0,
                      y: 0
                    })
                  }, null, 8, vo)) : (o(), n("text", {
                    key: 0,
                    class: "vue-data-ui-time-label",
                    "font-size": e.value.xyLabelsXFontSize,
                    "text-anchor": e.value.xyPeriodLabelsRotation > 0 ? "start" : e.value.xyPeriodLabelsRotation < 0 ? "end" : "middle",
                    fill: e.value.color,
                    transform: `translate(${s.value.drawingArea.left + s.value.slotSize * (l + 1) - s.value.slotSize / 2}, ${s.value.drawingArea.bottom + e.value.xyLabelsXFontSize + 6}), rotate(${e.value.xyPeriodLabelsRotation})`
                  }, U(t), 9, so))
                ])) : m("", !0)
              ], 64))), 256))
            ], 512)
          ])) : m("", !0),
          $("g", co, [
            (o(!0), n(x, null, I(s.value.dataset, (t, l) => (o(), n(x, null, [
              (o(!0), n(x, null, I(t.coordinates, (u, p) => (o(), n("rect", {
                x: u.x,
                width: u.width <= 0 ? 1e-5 : u.width,
                height: v(ve)(u.height <= 0 ? 1e-5 : u.height),
                y: v(ve)(u.y),
                fill: t.color,
                stroke: e.value.backgroundColor,
                "stroke-width": e.value.barStrokeWidth,
                "stroke-linecap": "round",
                class: fe({ "vue-data-ui-bar-animated": e.value.barAnimated && u.value < 0 && !v(D) })
              }, [
                e.value.barAnimated && u.value > 0 && !v(De) && !v(Ue) ? (o(), n("animate", {
                  key: 0,
                  attributeName: "height",
                  from: 0,
                  to: u.height,
                  dur: "0.5s"
                }, null, 8, fo)) : m("", !0),
                e.value.barAnimated && u.value > 0 && !v(De) && !v(Ue) ? (o(), n("animate", {
                  key: 1,
                  attributeName: "y",
                  from: s.value.absoluteZero,
                  to: s.value.absoluteZero - u.height,
                  dur: "0.5s"
                }, null, 8, mo)) : m("", !0)
              ], 10, ho))), 256))
            ], 64))), 256))
          ]),
          e.value.showDataLabels ? (o(), n("g", go, [
            (o(!0), n(x, null, I(s.value.dataset, (t, l) => (o(), n(x, null, [
              (o(!0), n(x, null, I(t.coordinates, (u, p) => (o(), n("text", {
                x: u.x + u.width / 2,
                y: v(ve)(u.y) - e.value.dataLabelFontSize / 2,
                "text-anchor": "middle",
                "font-size": e.value.dataLabelFontSize,
                fill: t.color,
                class: "quick-animation"
              }, U(v(se)(
                e.value.formatter,
                v(ve)(u.value),
                v(J)({
                  p: e.value.valuePrefix,
                  v: v(ve)(u.value),
                  s: e.value.valueSuffix,
                  r: e.value.dataLabelRoundingValue
                }),
                { datapoint: u, seriesIndex: p }
              )), 9, po))), 256))
            ], 64))), 256))
          ])) : m("", !0),
          Qe.value && k.value.length < s.value.legend.length ? (o(), n("g", yo, [
            (o(!0), n(x, null, I(s.value.extremes.maxSeries, (t, l) => (o(), n("rect", {
              x: s.value.drawingArea.left + l * s.value.slotSize,
              y: s.value.drawingArea.top,
              height: s.value.drawingArea.height <= 0 ? 1e-5 : s.value.drawingArea.height,
              width: s.value.slotSize <= 0 ? 1e-5 : s.value.slotSize,
              fill: [oe.value, ue.value].includes(l) ? e.value.xyHighlighterColor : "transparent",
              style: N(`opacity:${e.value.xyHighlighterOpacity}`),
              onMouseenter: (u) => s.value.useTooltip(l),
              onMouseleave: (u) => s.value.killTooltip(l),
              onClick: (u) => s.value.selectDatapoint(l)
            }, null, 44, xo))), 256))
          ])) : m("", !0)
        ], 64)) : m("", !0),
        [w.LINE, w.BAR].includes(_.value) ? (o(), n("g", ko, [
          e.value.xAxisLabel && _.value === w.LINE ? (o(), n("g", {
            key: 0,
            ref_key: "xAxisLabel",
            ref: yt
          }, [
            $("text", {
              "font-size": e.value.axisLabelsFontSize,
              fill: e.value.color,
              "text-anchor": "middle",
              x: d.value.drawingArea.left + d.value.drawingArea.width / 2,
              y: A.value.height - e.value.axisLabelsFontSize / 3
            }, U(e.value.xAxisLabel), 9, bo)
          ], 512)) : m("", !0),
          e.value.xAxisLabel && _.value === w.BAR ? (o(), n("g", {
            key: 1,
            ref_key: "xAxisLabel",
            ref: yt
          }, [
            $("text", {
              "font-size": e.value.axisLabelsFontSize,
              fill: e.value.color,
              "text-anchor": "middle",
              x: s.value.drawingArea.left + s.value.drawingArea.width / 2,
              y: A.value.height - e.value.axisLabelsFontSize / 3
            }, U(e.value.xAxisLabel), 9, So)
          ], 512)) : m("", !0),
          e.value.yAxisLabel && _.value === w.LINE ? (o(), n("g", {
            key: 2,
            ref_key: "yAxisLabel",
            ref: xt
          }, [
            $("text", {
              "font-size": e.value.axisLabelsFontSize,
              fill: e.value.color,
              transform: `translate(${e.value.axisLabelsFontSize}, ${d.value.drawingArea.top + d.value.drawingArea.height / 2}) rotate(-90)`,
              "text-anchor": "middle"
            }, U(e.value.yAxisLabel), 9, wo)
          ], 512)) : m("", !0),
          e.value.yAxisLabel && _.value === w.BAR ? (o(), n("g", {
            key: 3,
            ref_key: "yAxisLabel",
            ref: xt
          }, [
            $("text", {
              "font-size": e.value.axisLabelsFontSize,
              fill: e.value.color,
              transform: `translate(${e.value.axisLabelsFontSize}, ${s.value.drawingArea.top + s.value.drawingArea.height / 2}) rotate(-90)`,
              "text-anchor": "middle"
            }, U(e.value.yAxisLabel), 9, Ao)
          ], 512)) : m("", !0)
        ])) : m("", !0)
      ], 12, ul)) : m("", !0),
      a.$slots.watermark ? (o(), n("div", Lo, [
        M(a.$slots, "watermark", Z(Y({ isPrinting: v(De) || v(Ue) })), void 0, !0)
      ])) : m("", !0),
      [w.BAR, w.LINE].includes(_.value) && e.value.zoomXy && b.value.maxSeriesLength > 1 ? (o(), n("div", {
        key: `slicer_${Fe.value}`,
        ref_key: "quickChartSlicer",
        ref: ht
      }, [
        (o(), we(Za, {
          ref_key: "slicerComponent",
          ref: Et,
          key: `slicer_${Fe.value}`,
          background: e.value.zoomColor,
          borderColor: e.value.backgroundColor,
          fontSize: e.value.zoomFontSize,
          useResetSlot: e.value.zoomUseResetSlot,
          labelLeft: e.value.xyPeriods[y.value.start] && j.value[0] ? j.value[0].text : "",
          labelRight: e.value.xyPeriods[y.value.end - 1] && j.value?.length ? j.value.at(-1).text : "",
          textColor: e.value.color,
          inputColor: e.value.zoomColor,
          selectColor: e.value.zoomHighlightColor,
          max: b.value.maxSeriesLength,
          min: 0,
          valueStart: y.value.start,
          valueEnd: y.value.end,
          smoothMinimap: e.value.zoomMinimap.smooth,
          minimapSelectedColor: e.value.zoomMinimap.selectedColor,
          minimapSelectedColorOpacity: e.value.zoomMinimap.selectedColorOpacity,
          minimapSelectionRadius: e.value.zoomMinimap.selectionRadius,
          minimapLineColor: e.value.zoomMinimap.lineColor,
          minimap: sa.value,
          minimapIndicatorColor: e.value.zoomMinimap.indicatorColor,
          verticalHandles: e.value.zoomMinimap.verticalHandles,
          minimapSelectedIndex: ue.value,
          start: y.value.start,
          "onUpdate:start": r[0] || (r[0] = (t) => y.value.start = t),
          end: y.value.end,
          "onUpdate:end": r[1] || (r[1] = (t) => y.value.end = t),
          refreshStartPoint: e.value.zoomStartIndex !== null ? e.value.zoomStartIndex : 0,
          refreshEndPoint: e.value.zoomEndIndex !== null ? e.value.zoomEndIndex + 1 : b.value.maxSeriesLength,
          enableRangeHandles: e.value.zoomEnableRangeHandles,
          enableSelectionDrag: e.value.zoomEnableSelectionDrag,
          minimapCompact: e.value.zoomMinimap.compact,
          minimapMerged: e.value.zoomMinimap.merged,
          allMinimaps: ma.value,
          minimapFrameColor: e.value.zoomMinimap.frameColor,
          additionalMinimapHeight: e.value.zoomMinimap.additionalHeight,
          handleType: e.value.zoomMinimap.handleType,
          handleWidth: e.value.zoomMinimap.handleWidth,
          handleBorderWidth: e.value.zoomMinimap.handleBorderWidth,
          handleIconColor: e.value.zoomMinimap.handleIconColor,
          handleBorderColor: e.value.zoomMinimap.handleBorderColor,
          handleFill: e.value.zoomMinimap.handleFill,
          focusOnDrag: e.value.zoomFocusOnDrag,
          focusRangeRatio: e.value.zoomFocusRangeRatio,
          maxWidth: e.value.zoomMaxWidth,
          onReset: ua,
          onTrapMouse: ia
        }, {
          "reset-action": H(({ reset: t }) => [
            M(a.$slots, "reset-action", Z(Y({ reset: t })), void 0, !0)
          ]),
          _: 3
        }, 8, ["background", "borderColor", "fontSize", "useResetSlot", "labelLeft", "labelRight", "textColor", "inputColor", "selectColor", "max", "valueStart", "valueEnd", "smoothMinimap", "minimapSelectedColor", "minimapSelectedColorOpacity", "minimapSelectionRadius", "minimapLineColor", "minimap", "minimapIndicatorColor", "verticalHandles", "minimapSelectedIndex", "start", "end", "refreshStartPoint", "refreshEndPoint", "enableRangeHandles", "enableSelectionDrag", "minimapCompact", "minimapMerged", "allMinimaps", "minimapFrameColor", "additionalMinimapHeight", "handleType", "handleWidth", "handleBorderWidth", "handleIconColor", "handleBorderColor", "handleFill", "focusOnDrag", "focusRangeRatio", "maxWidth"]))
      ])) : m("", !0),
      $("div", {
        id: `legend-bottom-${K.value}`
      }, null, 8, Co),
      pt.value ? (o(), we(za, {
        key: 7,
        to: e.value.legendPosition === "top" ? `#legend-top-${K.value}` : `#legend-bottom-${K.value}`
      }, [
        e.value.showLegend ? (o(), n("div", {
          key: 0,
          ref_key: "quickChartLegend",
          ref: ct,
          class: "vue-ui-quick-chart-legend",
          style: N(`background:transparent;color:${e.value.color}`)
        }, [
          (T.value?.legend?.length > 2 || d.value?.legend?.length > 2 || s.value?.legend?.length > 2) && e.value.showLegendSelectAllToggle && !v(D) ? (o(), we(Ja, {
            key: 0,
            backgroundColor: e.value.legendSelectAllToggleBackgroundColor,
            color: e.value.legendSelectAllToggleColor,
            fontSize: e.value.legendFontSize,
            checked: k.value.length > 0,
            isCursorPointer: xe.value,
            onToggle: oa
          }, null, 8, ["backgroundColor", "color", "fontSize", "checked", "isCursorPointer"])) : m("", !0),
          _.value === w.DONUT ? (o(!0), n(x, { key: 1 }, I(T.value.legend, (t, l) => (o(), n("div", {
            class: "vue-ui-quick-chart-legend-item",
            onClick: (u) => {
              na(t, T.value.dataset), pe("selectLegend", t);
            },
            style: N(`cursor: ${T.value.legend.length > 1 && xe.value ? "pointer" : "default"}; opacity:${k.value.includes(t.id) ? "0.5" : "1"}`)
          }, [
            e.value.useCustomLegend ? M(a.$slots, "legend", nt({
              key: 0,
              ref_for: !0
            }, { legend: t }), void 0, !0) : (o(), n(x, { key: 1 }, [
              Ae(v(V), {
                name: e.value.legendIcon,
                stroke: t.color,
                size: e.value.legendIconSize
              }, null, 8, ["name", "stroke", "size"]),
              $("span", {
                style: N(`font-size:${e.value.legendFontSize}px`)
              }, U(t.name), 5),
              $("span", {
                style: N(`font-size:${e.value.legendFontSize}px;font-variant-numeric:tabular-nums`)
              }, U(k.value.includes(t.id) ? "-" : v(se)(
                e.value.formatter,
                t.absoluteValue,
                v(J)({
                  p: e.value.valuePrefix,
                  v: t.absoluteValue,
                  s: e.value.valueSuffix,
                  r: e.value.dataLabelRoundingValue
                }),
                { datapoint: t, seriesIndex: l }
              )), 5),
              k.value.includes(t.id) ? (o(), n("span", {
                key: 0,
                style: N(`font-size:${e.value.legendFontSize}px`)
              }, " ( - % ) ", 4)) : Ve.value ? (o(), n("span", {
                key: 1,
                style: N(`font-size:${e.value.legendFontSize}px; font-variant-numeric: tabular-nums;`)
              }, " ( - % ) ", 4)) : (o(), n("span", {
                key: 2,
                style: N(`font-size:${e.value.legendFontSize}px; font-variant-numeric: tabular-nums;`)
              }, " (" + U(v(J)({
                v: t.value / T.value.total * 100,
                s: "%",
                r: e.value.dataLabelRoundingPercentage
              })) + ") ", 5))
            ], 64))
          ], 12, _o))), 256)) : m("", !0),
          _.value === w.LINE ? (o(!0), n(x, { key: 2 }, I(d.value.legend, (t, l) => (o(), n("div", {
            class: "vue-ui-quick-chart-legend-item",
            onClick: (u) => {
              _t(t.id, d.value.legend.length - 1), pe("selectLegend", t);
            },
            style: N(`cursor: ${d.value.legend.length > 1 && xe.value ? "pointer" : "default"}; opacity:${k.value.includes(t.id) ? "0.5" : "1"}`)
          }, [
            e.value.useCustomLegend ? M(a.$slots, "legend", nt({
              key: 0,
              ref_for: !0
            }, { legend: t }), void 0, !0) : (o(), n(x, { key: 1 }, [
              Ae(v(V), {
                name: e.value.legendIcon,
                stroke: t.color,
                size: e.value.legendIconSize
              }, null, 8, ["name", "stroke", "size"]),
              $("span", {
                style: N(`font-size:${e.value.legendFontSize}px`)
              }, U(t.name), 5)
            ], 64))
          ], 12, zo))), 256)) : m("", !0),
          _.value === w.BAR ? (o(!0), n(x, { key: 3 }, I(s.value.legend, (t, l) => (o(), n("div", {
            class: "vue-ui-quick-chart-legend-item",
            onClick: (u) => {
              _t(t.id, s.value.legend.length - 1), pe("selectLegend", t);
            },
            style: N(`cursor: ${s.value.legend.length > 1 && xe.value ? "pointer" : "default"}; opacity:${k.value.includes(t.id) ? "0.5" : "1"}`)
          }, [
            e.value.useCustomLegend ? M(a.$slots, "legend", nt({
              key: 0,
              ref_for: !0
            }, { legend: t }), void 0, !0) : (o(), n(x, { key: 1 }, [
              Ae(v(V), {
                name: e.value.legendIcon,
                stroke: t.color,
                size: e.value.legendIconSize
              }, null, 8, ["name", "stroke", "size"]),
              $("span", {
                style: N(`font-size:${e.value.legendFontSize}px`)
              }, U(t.name), 5)
            ], 64))
          ], 12, $o))), 256)) : m("", !0)
        ], 4)) : m("", !0)
      ], 8, ["to"])) : m("", !0),
      a.$slots.source ? (o(), n("div", {
        key: 8,
        ref_key: "source",
        ref: ft,
        dir: "auto"
      }, [
        M(a.$slots, "source", {}, void 0, !0)
      ], 512)) : m("", !0),
      Ae(v(Le), {
        teleportTo: e.value.tooltipTeleportTo,
        show: ke.value.showTooltip && me.value,
        backgroundColor: e.value.backgroundColor,
        color: e.value.color,
        borderRadius: e.value.tooltipBorderRadius,
        borderColor: e.value.tooltipBorderColor,
        borderWidth: e.value.tooltipBorderWidth,
        fontSize: e.value.tooltipFontSize,
        backgroundOpacity: e.value.tooltipBackgroundOpacity,
        position: e.value.tooltipPosition,
        offsetY: e.value.tooltipOffsetY,
        parent: ne.value,
        content: ge.value,
        isFullscreen: He.value,
        isCustom: v(je)(e.value.tooltipCustomFormat),
        smooth: e.value.tooltipSmooth,
        smoothForce: e.value.tooltipSmoothForce,
        smoothSnapThreshold: e.value.tooltipSmoothSnapThreshold,
        backdropFilter: e.value.tooltipBackdropFilter
      }, {
        "tooltip-before": H(() => [
          M(a.$slots, "tooltip-before", Z(Y({ ...Ce.value })), void 0, !0)
        ]),
        "tooltip-after": H(() => [
          M(a.$slots, "tooltip-after", Z(Y({ ...Ce.value })), void 0, !0)
        ]),
        _: 3
      }, 8, ["teleportTo", "show", "backgroundColor", "color", "borderRadius", "borderColor", "borderWidth", "fontSize", "backgroundOpacity", "position", "offsetY", "parent", "content", "isFullscreen", "isCustom", "smooth", "smoothForce", "smoothSnapThreshold", "backdropFilter"]),
      M(a.$slots, "skeleton", {}, () => [
        v(D) ? (o(), we(Ua, { key: 0 })) : m("", !0)
      ], !0)
    ], 46, nl)) : (o(), n("div", To, [
      Ae(v(V), {
        name: "circleCancel",
        stroke: "red"
      }),
      r[5] || (r[5] = $("span", null, "Dataset is not processable", -1))
    ]));
  }
}, Zo = /* @__PURE__ */ Ka(Eo, [["__scopeId", "data-v-4baff879"]]);
export {
  Zo as default
};
