// vue-data-ui/dist/vue-ui-sparkgauge-DdRtHQDe.js
import { defineAsyncComponent as V, ref as b, computed as n, toRefs as X, onMounted as F, watch as B, openBlock as u, createElementBlock as v, normalizeStyle as y, toDisplayString as k, createCommentVNode as d, unref as i, createVNode as Y, renderSlot as C, createElementVNode as m, normalizeClass as H, createBlock as K } from "vue";
import { u as w } from "./useNestedProp-BBT0mbHh.js";
import { j as J, t as D, o as Q, q as E, r as W, an as Z, X as ee, k as te, K as x, l as ae } from "./lib-GwPnJSrY.js";
import { u as le } from "./useConfig-CIgyKZ6R.js";
import { u as oe, B as se } from "./BaseScanner-BhhkKVoj.js";
import { u as ne } from "./useThemeCheck-D1ZYXUwG.js";
import { u as ue } from "./useChartAccessibility-9icAAmYg.js";
import ie from "./vue_ui_sparkgauge-KnjASOZs.js";
const re = ["xmlns", "viewBox"], ve = ["width", "height"], ce = ["id"], fe = ["stop-color"], de = ["stop-color"], me = ["d", "stroke", "stroke-linecap"], ge = ["d", "stroke", "stroke-linecap", "stroke-dashoffset"], ye = ["x", "y", "width", "height"], he = ["x", "y", "font-size", "fill", "font-weight"], pe = {
  key: 2,
  ref: "source",
  dir: "auto"
}, Se = {
  __name: "vue-ui-sparkgauge",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  setup(T) {
    const j = V(() => import("./PackageVersion-DN9d0ME5.js")), { vue_ui_sparkgauge: I } = le(), { isThemeValid: O, warnInvalidTheme: R } = ne(), r = T, _ = b(J()), e = b(p()), G = n(() => D({
      defaultConfig: {
        style: {
          animation: { show: !1 },
          background: "#99999930",
          colors: {
            min: "#CACACA",
            max: "#6A6A6A"
          },
          track: {
            autoColor: !0
          },
          gutter: {
            color: "#6A6A6A80"
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: h, FINAL_DATASET: c } = oe({
      ...X(r),
      FINAL_CONFIG: e,
      prepareConfig: p,
      skeletonDataset: r.config?.skeletonDataset ?? {
        value: 0,
        min: -1,
        max: 1,
        title: ""
      },
      skeletonConfig: D({
        defaultConfig: e.value,
        userConfig: G.value
      })
    }), { svgRef: U } = ue({ config: { text: r.dataset?.title || "" } });
    function p() {
      const t = w({
        userConfig: r.config,
        defaultConfig: I
      }), o = t.theme;
      if (!o) return t;
      if (!O.value(t))
        return R(t), t;
      const f = w({
        userConfig: ie[o] || r.config,
        defaultConfig: t
      });
      return w({
        userConfig: r.config,
        defaultConfig: f
      });
    }
    F(() => {
      A();
    });
    const $ = n(() => e.value.debug);
    function A() {
      Q(r.dataset) ? E({
        componentName: "VueUiSparkgauge",
        type: "dataset",
        debug: $.value
      }) : W({
        datasetObject: r.dataset,
        requiredAttributes: ["value", "min", "max"]
      }).forEach((t) => {
        E({
          componentName: "VueUiSparkgauge",
          type: "datasetAttribute",
          property: t,
          debug: $.value
        });
      });
    }
    B(() => r.config, (t) => {
      e.value = p(), a.value = e.value.style.animation.show ? s.value.min : c.value.value, A();
    }, { deep: !0 });
    const l = n(() => ({
      height: e.value.style.height,
      width: 128,
      base: e.value.style.basePosition
    })), s = n(() => {
      const t = c.value.min ?? 0, o = c.value.max ?? 0, f = o - t;
      return {
        min: t,
        max: o,
        diff: f
      };
    }), a = b(e.value.style.animation.show ? s.value.min : c.value.value);
    B(() => c.value.value, () => {
      S(c.value.value || 0);
    });
    const q = n(() => a.value > s.value.max ? s.value.max : a.value < s.value.min ? s.value.min : a.value), L = n(() => s.value.diff / e.value.style.animation.speedMs);
    F(() => {
      S(c.value.value || 0);
    });
    function S(t) {
      function o() {
        a.value < t ? a.value = Math.min(a.value + L.value, t) : a.value > t && (a.value = Math.max(a.value - L.value, t)), a.value !== t && requestAnimationFrame(o);
      }
      o();
    }
    const g = n(() => c.value.title ?? ""), z = n(() => {
      const t = s.value.diff;
      if (!isFinite(t) || t === 0) return 0;
      const o = s.value.min, f = q.value;
      return f >= 0 ? (f - o) / t : (Math.abs(o) - Math.abs(f)) / t;
    }), N = n(() => Z(e.value.style.colors.min, e.value.style.colors.max, s.value.min, s.value.max, a.value)), M = n(() => e.value.style.dataLabel.autoColor ? N.value : e.value.style.dataLabel.color), P = n(() => e.value.style.track.autoColor ? N.value : e.value.style.track.color);
    return (t, o) => (u(), v("div", {
      class: "vue-data-ui-component vue-ui-sparkgauge",
      style: y(`font-family:${e.value.style.fontFamily};width: 100%; background:${e.value.style.background}`)
    }, [
      e.value.style.title.show && g.value && e.value.style.title.position === "top" ? (u(), v("div", {
        key: 0,
        class: "vue-data-ui-sparkgauge-label",
        style: y(`font-size:${e.value.style.title.fontSize}px;text-align:${e.value.style.title.textAlign};font-weight:${e.value.style.title.bold ? "bold" : "normal"};color:${e.value.style.title.color}`)
      }, k(g.value), 5)) : d("", !0),
      (u(), v("svg", {
        ref_key: "svgRef",
        ref: U,
        xmlns: i(ee),
        viewBox: `0 0 ${l.value.width} ${l.value.height}`,
        style: "overflow: visible; background:transparent; width:100%;"
      }, [
        Y(i(j)),
        t.$slots["chart-background"] && !i(h) ? (u(), v("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: l.value.width,
          height: l.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          C(t.$slots, "chart-background")
        ], 8, ve)) : d("", !0),
        m("defs", null, [
          m("linearGradient", {
            id: `gradient_${_.value}`,
            x1: "-10%",
            y1: "100%",
            x2: "110%",
            y2: "100%"
          }, [
            m("stop", {
              offset: "0%",
              "stop-color": e.value.style.colors.min
            }, null, 8, fe),
            m("stop", {
              offset: "100%",
              "stop-color": e.value.style.colors.max
            }, null, 8, de)
          ], 8, ce)
        ]),
        m("path", {
          d: `M10 ${l.value.base} A 1 1 0 1 1 118 ${l.value.base}`,
          stroke: e.value.style.gutter.color,
          "stroke-width": 8,
          "stroke-linecap": e.value.style.gutter.strokeLinecap,
          fill: "none"
        }, null, 8, me),
        z.value !== 0 ? (u(), v("path", {
          key: 1,
          d: `M10 ${l.value.base} A 1 1 0 1 1 118 ${l.value.base}`,
          stroke: e.value.style.colors.showGradient ? `url(#gradient_${_.value})` : P.value,
          "stroke-width": 8,
          "stroke-linecap": e.value.style.track.strokeLinecap,
          fill: "none",
          "stroke-dasharray": 169.5,
          "stroke-dashoffset": 169.5 - 169.5 * z.value,
          class: H({ "vue-ui-sparkgauge-track": e.value.style.animation.show }),
          style: y(e.value.style.animation.show ? `animation: vue-ui-sparkgauge-animation ${e.value.style.animation.speedMs}ms ease-in;` : "")
        }, null, 14, ge)) : d("", !0),
        i(h) ? (u(), v("rect", {
          key: 2,
          x: l.value.width / 2 - e.value.style.dataLabel.fontSize / 2,
          y: l.value.base + 6 + e.value.style.dataLabel.offsetY - e.value.style.dataLabel.fontSize,
          width: e.value.style.dataLabel.fontSize,
          height: e.value.style.dataLabel.fontSize,
          fill: "#6A6A6A50",
          rx: 3
        }, null, 8, ye)) : (u(), v("text", {
          key: 3,
          "text-anchor": "middle",
          x: l.value.width / 2,
          y: l.value.base + 6 + e.value.style.dataLabel.offsetY,
          "font-size": e.value.style.dataLabel.fontSize,
          fill: M.value,
          "font-weight": e.value.style.dataLabel.bold ? "bold" : "normal"
        }, k(i(te)(
          e.value.style.dataLabel.formatter,
          i(x)(a.value),
          i(ae)({
            p: e.value.style.dataLabel.prefix,
            v: i(x)(a.value),
            s: e.value.style.dataLabel.suffix,
            r: e.value.style.dataLabel.rounding
          }),
          { datapoint: i(x)(a.value), color: M.value }
        )), 9, he))
      ], 8, re)),
      e.value.style.title.show && g.value && e.value.style.title.position === "bottom" ? (u(), v("div", {
        key: 1,
        class: "vue-data-ui-sparkgauge-label",
        style: y(`font-size:${e.value.style.title.fontSize}px;text-align:${e.value.style.title.textAlign};font-weight:${e.value.style.title.bold ? "bold" : "normal"};font-weight:${e.value.style.title.bold ? "bold" : "normal"};color:${e.value.style.title.color}`)
      }, k(g.value), 5)) : d("", !0),
      t.$slots.source ? (u(), v("div", pe, [
        C(t.$slots, "source")
      ], 512)) : d("", !0),
      C(t.$slots, "skeleton", {}, () => [
        i(h) ? (u(), K(se, { key: 0 })) : d("", !0)
      ])
    ], 4));
  }
};
export {
  Se as default
};
