// vue-data-ui/dist/vue-ui-donut-evolution-C_ck_TOR.js
import { defineAsyncComponent as M, ref as i, shallowRef as St, computed as d, onMounted as wl, toRefs as Al, nextTick as he, watch as ye, watchEffect as _t, onBeforeUnmount as Ll, openBlock as s, createElementBlock as n, unref as c, normalizeStyle as Y, normalizeClass as N, createBlock as z, withCtx as g, renderSlot as m, normalizeProps as D, guardReactiveProps as I, createCommentVNode as v, createElementVNode as k, createSlots as $t, createVNode as zt, Fragment as _, renderList as $, toDisplayString as C, Teleport as Sl, resolveDynamicComponent as _l, mergeProps as $l, createTextVNode as Fe } from "vue";
import { j as Tt, o as Pt, q as ge, r as zl, t as Dt, m as Tl, p as fe, n as Pl, s as Ve, a7 as Dl, a8 as It, D as Il, a3 as Me, k as Be, l as be, X as Ol, u as le, a9 as Nl, Z as Rl, a4 as El, a5 as Fl, a6 as Ot, a0 as Vl, a1 as Ml, aa as Nt, ab as Bl } from "./lib-GwPnJSrY.js";
import { t as Rt, u as Ul } from "./useResponsive-DfdjqQps.js";
import { u as Hl } from "./useConfig-CIgyKZ6R.js";
import { u as Wl } from "./usePrinter-DiK5R6k-.js";
import { u as Xl, B as Gl } from "./BaseScanner-BhhkKVoj.js";
import { u as Yl } from "./useSvgExport-ByUukOZt.js";
import { u as jl } from "./useTimeLabels-C-A2uZKi.js";
import { u as Ue } from "./useNestedProp-BBT0mbHh.js";
import { u as ql } from "./useThemeCheck-D1ZYXUwG.js";
import { u as Zl } from "./useUserOptionState-BIvW1Kz7.js";
import { u as Jl } from "./useChartAccessibility-9icAAmYg.js";
import { u as Kl } from "./useTimeLabelCollider-CIsgDrl9.js";
import Ql from "./img-BYYzY_QU.js";
import ea from "./Title-BfyrP-jp.js";
import ta from "./vue_ui_donut_evolution-DFQU-rqE.js";
import la from "./Legend-1LMMsPxL.js";
import { S as aa } from "./Slicer-DZJA9IG-.js";
import { B as oa } from "./BaseLegendToggle-Y61xahI4.js";
import { _ as ua } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const sa = ["id"], na = ["id"], ra = ["xmlns", "viewBox"], ia = ["x", "y", "width", "height"], ca = ["id"], va = ["stop-color"], da = ["stop-color"], ha = ["id"], ya = ["stop-color"], ga = ["stop-color"], fa = ["stop-color"], ba = { key: 1 }, ma = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], pa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], xa = { key: 0 }, ka = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], Ca = ["font-size", "fill", "transform"], wa = ["x", "y", "font-size", "fill"], Aa = ["x1", "x2", "y1", "y2", "stroke", "stroke-width"], La = ["x", "y", "font-size", "fill", "font-weight"], Sa = { key: 0 }, _a = ["text-anchor", "font-size", "fill", "transform"], $a = ["text-anchor", "font-size", "fill", "transform", "innerHTML"], za = ["x1", "y1", "x2", "y2", "stroke", "stroke-width"], Ta = { key: 1 }, Pa = ["cx", "cy", "r", "fill"], Da = { key: 0 }, Ia = { key: 0 }, Oa = ["d", "stroke"], Na = ["text-anchor", "x", "y", "fill"], Ra = ["cx", "cy", "r", "fill"], Ea = { key: 0 }, Fa = ["cx", "cy", "fill"], Va = { key: 1 }, Ma = ["d", "fill", "stroke"], Ba = { key: 2 }, Ua = ["d", "fill", "stroke"], Ha = ["x", "y", "font-size", "fill"], Wa = ["x", "y", "width", "height", "fill"], Xa = ["x", "y", "width", "height", "onMouseenter", "onMouseleave", "onClick"], Ga = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Ya = ["id"], ja = ["onClick"], qa = { key: 0 }, Za = { key: 1 }, Ja = {
  __name: "vue-ui-donut-evolution",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Array,
      default() {
        return [];
      }
    }
  },
  emits: ["selectLegend", "copyAlt"],
  setup(Et, { expose: Ft, emit: Vt }) {
    const Mt = M(() => import("./vue-ui-accordion-BGHTqIBt.js")), Bt = M(() => import("./BaseIcon-CNx8Tzl5.js")), Ut = M(() => import("./DataTable-Bo1pmAqG.js")), Ht = M(() => import("./PackageVersion-DN9d0ME5.js")), Wt = M(() => import("./PenAndPaper-DWV1FMIi.js")), Xt = M(() => import("./UserOptions-CrLqEimB.js")), Gt = M(() => import("./vue-ui-donut-CzCgCVsx.js")), He = M(() => import("./BaseDraggableDialog-DSpKTn3R.js")), { vue_ui_donut_evolution: Yt } = Hl(), { isThemeValid: jt, warnInvalidTheme: qt } = ql(), S = Et, O = i(Tt()), p = i([]), b = i(null), ae = i(null), me = i(!1), B = i(null), R = i(null), Zt = i(null), We = i(0), Xe = i(0), Ge = i(0), Ye = i(0), je = i(0), pe = i(null), oe = i(null), xe = i(null), ke = i(null), qe = i(!1), j = i(null), Ce = i(null), ue = i(!1), se = i(!1), Ze = i(null), Je = i(null), Ke = i(null), Qe = i(null), et = i(null), we = i(!1), q = St(null), ne = St(null), tt = i(null), Ae = Vt, lt = d(() => !!S.dataset && S.dataset.length);
    wl(() => {
      qe.value = !0, at();
    });
    const re = d(() => !!e.value.debug);
    function at() {
      if (Pt(S.dataset) ? (ge({
        componentName: "VueUiDonutEvolution",
        type: "dataset",
        debug: re.value
      }), ie.value = !0) : S.dataset.length && S.dataset.forEach((t, a) => {
        zl({
          datasetObject: t,
          requiredAttributes: ["name", "values"]
        }).forEach((l) => {
          ge({
            componentName: "VueUiDonutEvolution",
            type: "datasetSerieAttribute",
            property: l,
            index: a,
            debug: re.value
          }), ie.value = !0;
        });
      }), Pt(S.dataset) || (ie.value = e.value.loading), setTimeout(() => {
        we.value = !0;
      }, 10), _e(), e.value.responsive) {
        const t = Rt(() => {
          we.value = !1;
          const { width: a, height: l } = Ul({
            chart: R.value,
            title: e.value.style.chart.title.text ? Ze.value : null,
            legend: e.value.style.chart.legend.show ? Je.value : null,
            slicer: e.value.style.chart.zoom.show && ot.value > 1 ? et.value : null,
            source: Ke.value
          });
          requestAnimationFrame(() => {
            X.value.width = a, X.value.height = l - 12, clearTimeout(tt.value), tt.value = setTimeout(() => {
              we.value = !0;
            }, 10);
          });
        });
        q.value && (ne.value && q.value.unobserve(ne.value), q.value.disconnect()), q.value = new ResizeObserver(t), ne.value = R.value.parentNode, q.value.observe(ne.value);
      }
    }
    const e = i(Te()), E = d(() => e.value.userOptions.useCursorPointer), Jt = d(() => Dt({
      defaultConfig: {
        useCssAnimation: !1,
        table: { show: !1 },
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            layout: {
              dataLabels: { show: !1 },
              grid: {
                stroke: "#6A6A6A",
                axis: {
                  yLabel: "",
                  xLabel: ""
                },
                yAxis: {
                  scaleMin: null,
                  scaleMax: null,
                  autoScale: !1,
                  dataLabels: { show: !1 }
                },
                xAxis: {
                  dataLabels: { show: !1 }
                }
              },
              line: {
                stroke: "#CACACA60"
              }
            },
            legend: {
              backgroundColor: "transparent",
              showValue: !1,
              showPercentage: !1
            },
            zoom: {
              show: !1,
              startIndex: null,
              endIndex: null
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: Le, FINAL_DATASET: Se, manualLoading: ie } = Xl({
      ...Al(S),
      FINAL_CONFIG: e,
      prepareConfig: Te,
      callback: () => {
        Promise.resolve().then(async () => {
          await _e();
        });
      },
      skeletonDataset: S.config?.skeletonDataset ?? [
        {
          name: "",
          values: [1, 2, 3, 5, 8, 13],
          color: "#AAAAAA"
        },
        {
          name: "",
          values: [1, 2, 3, 5, 8, 13],
          color: "#BABABA"
        },
        {
          name: "",
          values: [1, 2, 3, 5, 8, 13],
          color: "#CACACA"
        }
      ],
      skeletonConfig: Dt({
        defaultConfig: e.value,
        userConfig: Jt.value
      })
    }), X = i({
      width: e.value.style.chart.layout.width,
      height: e.value.style.chart.layout.height
    }), ot = d(() => Math.max(...Se.value.map((t) => t.values.length))), h = i({
      start: 0,
      end: ot.value
    });
    function ut() {
      _e();
    }
    async function _e() {
      await he(), await he();
      const { startIndex: t, endIndex: a } = e.value.style.chart.zoom, l = Qe.value;
      h.value = { start: 0, max: P.value }, (t != null || a != null) && l ? (t != null ? l.setStartValue(t) : (h.value.start = 0, l.setStartValue(0)), a != null ? l.setEndValue(Kt(a + 1)) : (h.value.end = P.value, l.setEndValue(P.value))) : (h.value = {
        start: 0,
        end: P.value
      }, Xe.value += 1);
    }
    function Kt(t) {
      const a = P.value;
      return t > a ? a : t < 0 || e.value.style.chart.zoom.startIndex !== null && t < e.value.style.chart.zoom.startIndex ? e.value.style.chart.zoom.startIndex !== null ? e.value.style.chart.zoom.startIndex + 1 : 1 : t;
    }
    const { userOptionsVisible: $e, setUserOptionsVisibility: st, keepUserOptionState: nt } = Zl({ config: e.value }), { svgRef: ze } = Jl({ config: e.value.style.chart.title });
    function Te() {
      const t = Ue({
        userConfig: S.config,
        defaultConfig: Yt
      });
      let a = {};
      const l = t.theme;
      if (l)
        if (!jt.value(t))
          qt(t), a = t;
        else {
          const u = Ue({
            userConfig: ta[l] || S.config,
            defaultConfig: t
          });
          a = {
            ...Ue({
              userConfig: S.config,
              defaultConfig: u
            }),
            customPalette: t.customPalette ? t.customPalette : Tl[l] || fe
          };
        }
      else
        a = t;
      return a;
    }
    ye(() => S.config, (t) => {
      Le.value || (e.value = Te()), $e.value = !e.value.userOptions.showOnChartHover, at(), Ge.value += 1, Ye.value += 1, je.value += 1, U.value.showTable = e.value.table.show;
    }, { deep: !0 }), ye(() => S.dataset, (t) => {
      Array.isArray(t) && t.length > 0 && (ie.value = !1), ut();
    }, { deep: !0 });
    const { isPrinting: Pe, isImaging: De, generatePdf: rt, generateImage: it } = Wl({
      elementId: O.value,
      fileName: e.value.style.chart.title.text || "vue-ui-donut-evolution",
      options: e.value.userOptions.print
    }), Qt = d(() => e.value.userOptions.show && !e.value.style.chart.title.text), el = d(() => Pl(e.value.customPalette)), U = i({
      showTable: e.value.table.show
    });
    ye(e, () => {
      U.value = {
        showTable: e.value.table.show
      };
    }, { immediate: !0 });
    const H = d(() => ({
      top: e.value.style.chart.layout.padding.top,
      right: e.value.style.chart.layout.padding.right,
      bottom: e.value.style.chart.layout.padding.bottom,
      left: e.value.style.chart.layout.padding.left
    }));
    function tl() {
      let t = 0;
      pe.value && (t = Array.from(pe.value.querySelectorAll("text")).reduce((o, f) => {
        const y = f.getComputedTextLength();
        return y > o ? y : o;
      }, 0));
      const a = ke.value ? ke.value.getBoundingClientRect().width + e.value.style.chart.layout.grid.axis.yLabelOffsetX + e.value.style.chart.layout.grid.axis.fontSize : 0;
      return t + a + 5;
    }
    const Ie = i(0), ll = Rt((t) => {
      Ie.value = t;
    }, 100);
    _t((t) => {
      const a = oe.value;
      if (!a) return;
      const l = new ResizeObserver((u) => {
        ll(u[0].contentRect.height);
      });
      l.observe(a), t(() => l.disconnect());
    }), Ll(() => {
      Ie.value = 0;
    });
    const ct = d(() => {
      let t = 0;
      xe.value && (t = xe.value.getBBox().height);
      let a = 0;
      return oe.value && (a = Ie.value), t + a + e.value.style.chart.layout.grid.axis.fontSize + e.value.style.chart.layout.grid.xAxis.dataLabels.offsetY;
    }), r = d(() => {
      const t = tl(), a = e.value.style.chart.layout.dataLabels.fontSize * 3, l = X.value.width, u = X.value.height, o = H.value.left + t, f = l - H.value.right, y = l - o - H.value.right, x = u - H.value.top - H.value.bottom - a - ct.value, V = H.value.top + a, ee = u - H.value.bottom - ct.value;
      return {
        top: V,
        left: o,
        right: f,
        bottom: ee,
        absoluteHeight: u,
        absoluteWidth: l,
        centerX: o + Math.max(10, y) / 2,
        centerY: V + Math.max(10, x) / 2,
        width: Math.max(10, y),
        height: Math.max(10, x)
      };
    }), T = d(() => (Se.value.forEach((t, a) => {
      [null, void 0].includes(t.name) && ge({
        componentName: "VueUiDonutEvolution",
        type: "datasetSerieAttribute",
        property: "name",
        index: a,
        debug: re.value
      }), [null, void 0].includes(t.values) && ge({
        componentName: "VueUiDonutEvolution",
        type: "datasetSerieAttribute",
        property: "values",
        index: a,
        debug: re.value
      });
    }), Se.value.map((t, a) => ({
      ...t,
      values: Dl(t.values),
      color: Ve(t.color) || el.value[a] || fe[a] || fe[a % fe.length],
      length: (t.values || []).length,
      uid: Tt()
    })))), ce = d(() => T.value.filter((t) => !p.value.includes(t.uid)).map((t) => ({
      ...t,
      values: t.values.filter((a, l) => l >= h.value.start && l <= h.value.end)
    }))), P = d(() => Math.max(...T.value.map((t) => t.length))), w = i([]);
    let vt = 0;
    _t(() => {
      const t = ++vt;
      (async () => {
        const a = await jl({
          values: e.value.style.chart.layout.grid.xAxis.dataLabels.values,
          maxDatapoints: P.value,
          formatter: e.value.style.chart.layout.grid.xAxis.dataLabels.datetimeFormatter,
          start: h.value.start,
          end: h.value.end
        });
        t === vt && (w.value = a);
      })();
    });
    const A = d(() => r.value.width / (h.value.end - h.value.start)), ve = d(() => {
      const t = [];
      for (let a = 0; a < h.value.end - h.value.start; a += 1) {
        const l = ce.value.map((x) => x.values[a] ?? null), u = l.filter((x) => [void 0, null].includes(x)).length === l.length, o = l.reduce((x, V) => x + V, 0), f = l.map((x) => x / o), y = r.value.left + A.value * a + A.value / 2;
        t.push({
          index: a,
          percentages: f,
          subtotal: u || o < 0 ? null : o,
          values: l,
          x: y
        });
      }
      return t;
    }), Oe = d(() => {
      const t = Math.max(...ve.value.map((o) => o.subtotal).filter((o) => It(o))) ?? 1, a = Math.min(...ve.value.map((o) => o.subtotal).filter((o) => It(o))) ?? 0;
      let l = Math.max(e.value.style.chart.layout.grid.yAxis.scaleMax ?? 0, t), u;
      return e.value.style.chart.layout.grid.yAxis.scaleMin != null ? u = Math.min(a, e.value.style.chart.layout.grid.yAxis.scaleMin) : u = 0, e.value.style.chart.layout.grid.yAxis.autoScale && (u = a, l = t), u === l && (u = l / 2, l *= 1.5), {
        max: l,
        min: u
      };
    }), F = d(() => {
      const t = ve.value.length === 1 ? Oe.value.max * 2 : Oe.value.max;
      return Il(Oe.value.min, t, e.value.style.chart.layout.grid.yAxis.dataLabels.steps);
    });
    function dt(t) {
      return (t - F.value.min) / (F.value.max - F.value.min);
    }
    const al = d(() => F.value.ticks.map((t) => ({
      y: r.value.bottom - r.value.height * dt(t),
      value: t
    }))), L = d(() => {
      const t = ve.value, a = Math.max(...t.map((l) => l.subtotal));
      return t.length === 1 && a * 2, t.map((l, u) => {
        const o = Math.min(r.value.width / 24, A.value / 2 * 0.7), f = o > r.value.width / 16 ? r.value.width / 16 : o, y = b.value === l.index ? r.value.width / 16 : f, x = t.length > 4 ? o * 2 : o * 2 > A.value / 2 * 0.7 ? A.value / 2 * 0.7 : o * 2, V = r.value.bottom - r.value.height * dt(l.subtotal), ee = ce.value.map((te) => ({
          color: te.color,
          name: te.name,
          value: te.values[u] ?? 0
        })).toSorted((te, Cl) => Cl.value - te.value);
        return {
          ...l,
          y: V,
          radius: f,
          activeRadius: y,
          hoverRadius: x,
          donut: Me({ series: ee }, l.x, V, f, f, 1.99999, 2, 1, 360, 105.25, f / 2),
          donutHover: Me({ series: ee }, l.x, V, x, x, 1.99999, 2, 1, 360, 105.25, x / 2),
          donutFocus: Me({ series: ee }, r.value.centerX, r.value.centerY, r.value.height / 3.6, r.value.height / 3.6, 1.99999, 2, 1, 360, 105.25, r.value.height / 6)
        };
      });
    });
    function ht(t, a, l) {
      return Be(
        e.value.style.chart.layout.dataLabels.formatter,
        t,
        be({
          p: e.value.style.chart.layout.dataLabels.prefix,
          v: t,
          s: e.value.style.chart.layout.dataLabels.suffix,
          r: e.value.style.chart.layout.dataLabels.rounding
        }),
        { datapoint: a, index: l }
      );
    }
    function ol(t, a) {
      return isNaN(t.value / Nt(a, "value")) ? 0 : (t.value / Nt(a, "value") * 100).toFixed(0) + "%";
    }
    function yt(t) {
      b.value = null, ae.value = null, e.value.events.datapointLeave && e.value.events.datapointLeave({ datapoint: t, seriesIndex: t.seriesIndex + h.value.start });
    }
    function ul(t) {
      e.value.events.datapointEnter && e.value.events.datapointEnter({ datapoint: t, seriesIndex: t.index + h.value.start }), b.value = t.index, ae.value = t;
    }
    const Z = i(null);
    function gt(t, a) {
      e.value.events.datapointClick && e.value.events.datapointClick({ datapoint: t, seriesIndex: t.index + h.value.start }), !(!t.subtotal || !e.value.style.chart.dialog.show) && (ae.value = null, b.value = null, me.value = !0, B.value = t, vl(t), [null, void 0].includes(a) || (Z.value = a));
    }
    const J = d(() => T.value.map((t, a) => ({
      name: t.name,
      value: t.values.slice(h.value.start, h.value.end).reduce((l, u) => l + u, 0),
      shape: "circle",
      uid: t.uid,
      color: t.color
    })).sort((t, a) => a.value - t.value).map((t, a) => ({
      ...t,
      opacity: p.value.includes(t.uid) ? 0.5 : 1,
      segregate: () => K(t.uid),
      isSegregated: p.value.includes(t.uid),
      display: `${t.name}${e.value.style.chart.legend.showPercentage || e.value.style.chart.legend.showValue ? ": " : ""}${e.value.style.chart.legend.showValue ? Be(
        e.value.style.chart.layout.dataLabels.formatter,
        t.value,
        be({
          p: e.value.style.chart.layout.dataLabels.prefix,
          v: t.value,
          s: e.value.style.chart.layout.dataLabels.suffix,
          r: e.value.style.chart.legend.roundingValue
        }),
        { datapoint: t, seriesIndex: a }
      ) : ""}${e.value.style.chart.legend.showPercentage ? p.value.includes(t.uid) ? `${e.value.style.chart.legend.showValue ? " (" : ""}- %${e.value.style.chart.legend.showValue ? ")" : ""}` : `${e.value.style.chart.legend.showValue ? " (" : ""}${isNaN(t.value / ft.value) ? "-" : be({
        v: t.value / ft.value * 100,
        s: "%",
        r: e.value.style.chart.legend.roundingPercentage
      })}${e.value.style.chart.legend.showValue ? ")" : ""}` : ""}`
    }))), ft = d(() => L.value.map((t) => t.subtotal).reduce((t, a) => t + a, 0)), sl = d(() => ({
      cy: "donut-div-legend",
      backgroundColor: e.value.style.chart.legend.backgroundColor,
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 12,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    }));
    function nl() {
      p.value.length ? p.value = [] : J.value.forEach((t) => {
        p.value.push(t.uid);
      });
    }
    function K(t) {
      if (p.value.includes(t))
        p.value = p.value.filter((a) => a !== t), Ae("selectLegend", ce.value);
      else {
        if (p.value.length === T.value.length - 1) return;
        p.value.push(t), Ae("selectLegend", ce.value);
      }
      B.value && gt(L.value.find((a, l) => l === Z.value));
    }
    function bt(t) {
      if (!T.value.length)
        return e.value.debug && console.warn("VueUiDonutEvolution - There are no series to show."), null;
      const a = T.value.find((l) => l.name === t);
      return a || (e.value.debug && console.warn(`VueUiDonutEvolution - Series name not found "${t}"`), null);
    }
    function rl(t) {
      const a = bt(t);
      a !== null && p.value.includes(a.uid) && K(a.uid);
    }
    function il(t) {
      const a = bt(t);
      a !== null && (p.value.includes(a.uid) || K(a.uid));
    }
    const G = d(() => {
      const t = [""].concat(T.value.filter((o) => !p.value.includes(o.uid)).map((o) => ({
        name: o.name,
        color: o.color
      }))).concat(["Σ"]);
      let a = [];
      for (let o = 0; o < P.value; o += 1) {
        const f = T.value.filter((y) => !p.value.includes(y.uid)).map((y) => y.values[o] ?? 0).reduce((y, x) => y + x, 0);
        a.push([w.value[o] ? w.value[o].text : "-"].concat(T.value.filter((y) => !p.value.includes(y.uid)).map((y) => ({
          value: y.values[o] ?? 0,
          percentage: y.values[o] ? y.values[o] / f * 100 : 0
        }))).concat([`${e.value.style.chart.layout.dataLabels.prefix}${Number(f.toFixed(e.value.table.td.roundingValue))}${e.value.style.chart.layout.dataLabels.suffix}`]));
      }
      const l = {
        th: {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          outline: e.value.table.th.outline
        },
        td: {
          backgroundColor: e.value.table.td.backgroundColor,
          color: e.value.table.td.color,
          outline: e.value.table.td.outline
        },
        breakpoint: e.value.table.responsiveBreakpoint
      }, u = [
        e.value.table.columnNames.period
      ].concat(T.value.filter((o) => !p.value.includes(o.uid)).map((o) => o.name)).concat(e.value.table.columnNames.total);
      return { head: t, body: a, config: l, colNames: u };
    });
    function cl() {
      return T.value;
    }
    function Ne(t = null) {
      he(() => {
        const a = [[e.value.style.chart.title.text], [e.value.style.chart.title.subtitle.text], [""]], l = [...G.value.head.map((y) => y.name ?? y)], u = [...G.value.body.map((y) => y.map((x) => x.value ?? x))], o = a.concat([l]).concat(u), f = Vl(o);
        t ? t(f) : Ml({ csvContent: f, title: e.value.style.chart.title.text || "vue-ui-donut-evolution" });
      });
    }
    const W = i(!1);
    function mt(t) {
      W.value = t, We.value += 1;
    }
    function pt() {
      U.value.showTable = !U.value.showTable;
    }
    const de = i(!1);
    function Re() {
      de.value = !de.value;
    }
    function xt(t) {
      return t.proportion * 100 > e.value.style.chart.donuts.hover.hideLabelsUnderValue;
    }
    const kt = i([]), Ct = i({}), Ee = i(null);
    function vl(t) {
      kt.value = t.donut.map((a) => ({
        name: a.name,
        values: [a.value],
        color: a.color
      })), Ct.value = Bl({
        ...e.value.style.chart.dialog.donutChart,
        responsive: !0,
        theme: e.value.theme
      }), Ee.value && Ee.value.open();
    }
    async function dl({ scale: t = 2 } = {}) {
      if (!R.value) return;
      const { width: a, height: l } = R.value.getBoundingClientRect(), u = a / l, { imageUri: o, base64: f } = await Ql({ domElement: R.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: o,
        base64: f,
        title: e.value.style.chart.title.text,
        width: a,
        height: l,
        aspectRatio: u
      };
    }
    const hl = d(() => X.value.width), yl = d(() => X.value.height);
    Kl({
      timeLabelsEls: oe,
      timeLabels: w,
      slicer: h,
      configRef: e,
      rotationPath: ["style", "chart", "layout", "grid", "xAxis", "dataLabels", "rotation"],
      autoRotatePath: ["style", "chart", "layout", "grid", "xAxis", "dataLabels", "autoRotate", "enable"],
      isAutoSize: !1,
      width: hl,
      height: yl,
      rotation: e.value.style.chart.layout.grid.xAxis.dataLabels.autoRotate.angle
    });
    const Q = d(() => {
      const t = e.value.table.useDialog && !e.value.table.show, a = U.value.showTable;
      return {
        component: t ? He : Mt,
        title: `${e.value.style.chart.title.text}${e.value.style.chart.title.subtitle.text ? `: ${e.value.style.chart.title.subtitle.text}` : ""}`,
        props: t ? {
          backgroundColor: e.value.table.th.backgroundColor,
          color: e.value.table.th.color,
          headerColor: e.value.table.th.color,
          headerBg: e.value.table.th.backgroundColor,
          isFullscreen: W.value,
          fullscreenParent: R.value,
          forcedWidth: Math.min(800, window.innerWidth * 0.8),
          isCursorPointer: E.value
        } : {
          hideDetails: !0,
          config: {
            open: a,
            maxHeight: 1e4,
            body: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            },
            head: {
              backgroundColor: e.value.style.chart.backgroundColor,
              color: e.value.style.chart.color
            }
          }
        }
      };
    });
    ye(() => U.value.showTable, (t) => {
      e.value.table.show || (t && e.value.table.useDialog && j.value ? j.value.open() : "close" in j.value && j.value.close());
    });
    function wt() {
      U.value.showTable = !1, Ce.value && Ce.value.setTableIconState(!1);
    }
    const gl = d(() => J.value.map((t) => ({
      ...t,
      name: t.display
    }))), fl = d(() => e.value.style.chart.backgroundColor), bl = d(() => e.value.style.chart.legend), ml = d(() => e.value.style.chart.title), { exportSvg: pl, getSvg: xl } = Yl({
      svg: ze,
      title: ml,
      legend: bl,
      legendItems: gl,
      backgroundColor: fl
    });
    async function At({ isCb: t }) {
      se.value = !0, await he();
      try {
        if (t) {
          const { blob: a, url: l, text: u, dataUrl: o } = await xl();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: a, url: l, text: u, dataUrl: o }));
        } else
          await Promise.resolve(pl());
      } finally {
        se.value = !1;
      }
    }
    function kl(t) {
      if (t?.stage === "start") {
        ue.value = !0;
        return;
      }
      if (t?.stage === "end") {
        ue.value = !1;
        return;
      }
      it();
    }
    async function Lt() {
      if (Ae("copyAlt", {
        config: e.value,
        dataset: L.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: L.value
      }));
    }
    return Ft({
      getData: cl,
      getImage: dl,
      generatePdf: rt,
      generateCsv: Ne,
      generateImage: it,
      generateSvg: At,
      hideSeries: il,
      showSeries: rl,
      toggleTable: pt,
      toggleAnnotator: Re,
      toggleFullscreen: mt,
      copyAlt: Lt
    }), (t, a) => (s(), n("div", {
      ref_key: "donutEvolutionChart",
      ref: R,
      class: N(`vue-data-ui-component vue-ui-donut-evolution ${W.value ? "vue-data-ui-wrapper-fullscreen" : ""} ${e.value.useCssAnimation ? "" : "vue-ui-dna"}`),
      style: Y(`font-family:${e.value.style.fontFamily};width:100%; text-align:center;background:${e.value.style.chart.backgroundColor}`),
      id: O.value,
      onMouseenter: a[5] || (a[5] = () => c(st)(!0)),
      onMouseleave: a[6] || (a[6] = () => c(st)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (s(), z(c(Wt), {
        key: 0,
        svgRef: c(ze),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: de.value,
        isCursorPointer: E.value,
        onClose: Re
      }, {
        "annotator-action-close": g(() => [
          m(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": g(({ color: l }) => [
          m(t.$slots, "annotator-action-color", D(I({ color: l })), void 0, !0)
        ]),
        "annotator-action-draw": g(({ mode: l }) => [
          m(t.$slots, "annotator-action-draw", D(I({ mode: l })), void 0, !0)
        ]),
        "annotator-action-undo": g(({ disabled: l }) => [
          m(t.$slots, "annotator-action-undo", D(I({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-redo": g(({ disabled: l }) => [
          m(t.$slots, "annotator-action-redo", D(I({ disabled: l })), void 0, !0)
        ]),
        "annotator-action-delete": g(({ disabled: l }) => [
          m(t.$slots, "annotator-action-delete", D(I({ disabled: l })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : v("", !0),
      Qt.value ? (s(), n("div", {
        key: 1,
        ref_key: "noTitle",
        ref: Zt,
        class: "vue-data-ui-no-title-space",
        style: "height:36px; width: 100%;background:transparent"
      }, null, 512)) : v("", !0),
      e.value.style.chart.title.text ? (s(), n("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Ze,
        style: "width:100%;background:transparent;padding-bottom:24px",
        onMouseleave: yt
      }, [
        (s(), z(ea, {
          key: `title_${Ge.value}`,
          config: {
            title: {
              cy: "donut-evolution-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "donut-evolution-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 544)) : v("", !0),
      k("div", {
        id: `legend-top-${O.value}`
      }, null, 8, na),
      e.value.userOptions.show && lt.value && (c(nt) || c($e)) ? (s(), z(c(Xt), {
        ref_key: "userOptionsRef",
        ref: Ce,
        key: `user_options_${We.value}`,
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: c(Pe),
        isImaging: c(De),
        uid: O.value,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: e.value.userOptions.buttons.csv,
        hasTable: e.value.userOptions.buttons.table,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: W.value,
        titles: { ...e.value.userOptions.buttonTitles },
        chartElement: R.value,
        position: e.value.userOptions.position,
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: de.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        tableDialog: e.value.table.useDialog,
        isCursorPointer: E.value,
        onToggleFullscreen: mt,
        onGeneratePdf: c(rt),
        onGenerateCsv: Ne,
        onGenerateImage: kl,
        onGenerateSvg: At,
        onToggleTable: pt,
        onToggleAnnotator: Re,
        onCopyAlt: Lt,
        style: Y({
          visibility: c(nt) ? c($e) ? "visible" : "hidden" : "visible"
        })
      }, $t({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: g(({ isOpen: l, color: u }) => [
            m(t.$slots, "menuIcon", D(I({ isOpen: l, color: u })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: g(() => [
            m(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionCsv ? {
          name: "optionCsv",
          fn: g(() => [
            m(t.$slots, "optionCsv", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: g(() => [
            m(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: g(() => [
            m(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionTable ? {
          name: "optionTable",
          fn: g(() => [
            m(t.$slots, "optionTable", {}, void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: g(({ toggleFullscreen: l, isFullscreen: u }) => [
            m(t.$slots, "optionFullscreen", D(I({ toggleFullscreen: l, isFullscreen: u })), void 0, !0)
          ]),
          key: "6"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: g(({ toggleAnnotator: l, isAnnotator: u }) => [
            m(t.$slots, "optionAnnotator", D(I({ toggleAnnotator: l, isAnnotator: u })), void 0, !0)
          ]),
          key: "7"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: g(({ altCopy: l }) => [
            m(t.$slots, "optionAltCopy", D(I({ altCopy: l })), void 0, !0)
          ]),
          key: "8"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasSvg", "hasXls", "hasTable", "hasFullscreen", "hasAltCopy", "isFullscreen", "titles", "chartElement", "position", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "tableDialog", "isCursorPointer", "onGeneratePdf", "style"])) : v("", !0),
      (s(), n("svg", {
        ref_key: "svgRef",
        ref: ze,
        xmlns: c(Ol),
        class: N({ "vue-data-ui-fullscreen--on": W.value, "vue-data-ui-fulscreen--off": !W.value }),
        viewBox: `0 0 ${r.value.absoluteWidth} ${r.value.absoluteHeight}`,
        style: Y(`max-width:100%; overflow: visible; background:transparent;color:${e.value.style.chart.color};`)
      }, [
        zt(c(Ht)),
        t.$slots["chart-background"] ? (s(), n("foreignObject", {
          key: 0,
          x: r.value.left,
          y: r.value.top,
          width: r.value.width,
          height: r.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          m(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, ia)) : v("", !0),
        k("defs", null, [
          k("linearGradient", {
            id: `hover_${O.value}`,
            x1: "0%",
            y1: "0%",
            x2: "0%",
            y2: "100%"
          }, [
            k("stop", {
              offset: "0%",
              "stop-color": c(le)(e.value.style.chart.backgroundColor, e.value.style.chart.layout.highlighter.opacity)
            }, null, 8, va),
            k("stop", {
              offset: "100%",
              "stop-color": c(le)(e.value.style.chart.layout.highlighter.color, e.value.style.chart.layout.highlighter.opacity)
            }, null, 8, da)
          ], 8, ca),
          k("radialGradient", {
            id: `focus_${O.value}`
          }, [
            k("stop", {
              offset: "0%",
              "stop-color": c(le)(c(Ve)(e.value.style.chart.backgroundColor), 0)
            }, null, 8, ya),
            k("stop", {
              offset: "77%",
              "stop-color": c(le)("#FFFFFF", 30)
            }, null, 8, ga),
            k("stop", {
              offset: "100%",
              "stop-color": c(le)(c(Ve)(e.value.style.chart.backgroundColor), 0)
            }, null, 8, fa)
          ], 8, ha)
        ]),
        e.value.style.chart.layout.grid.show ? (s(), n("g", ba, [
          k("line", {
            x1: r.value.left,
            x2: r.value.left,
            y1: r.value.top,
            y2: r.value.top + r.value.height,
            stroke: e.value.style.chart.layout.grid.stroke,
            "stroke-width": e.value.style.chart.layout.grid.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, ma),
          k("line", {
            x1: r.value.left,
            x2: r.value.right,
            y1: r.value.bottom,
            y2: r.value.bottom,
            stroke: e.value.style.chart.layout.grid.stroke,
            "stroke-width": e.value.style.chart.layout.grid.strokeWidth,
            "stroke-linecap": "round"
          }, null, 8, pa),
          e.value.style.chart.layout.grid.showVerticalLines ? (s(), n("g", xa, [
            (s(!0), n(_, null, $(h.value.end - h.value.start, (l, u) => (s(), n("line", {
              x1: r.value.left + (u + 1) * A.value,
              x2: r.value.left + (u + 1) * A.value,
              y1: r.value.top,
              y2: r.value.bottom,
              stroke: e.value.style.chart.layout.grid.stroke,
              "stroke-width": e.value.style.chart.layout.grid.strokeWidth,
              "stroke-linecap": "round"
            }, null, 8, ka))), 256))
          ])) : v("", !0)
        ])) : v("", !0),
        k("g", null, [
          e.value.style.chart.layout.grid.axis.yLabel ? (s(), n("text", {
            key: 0,
            ref_key: "yAxisLabel",
            ref: ke,
            "font-size": e.value.style.chart.layout.grid.axis.fontSize,
            fill: e.value.style.chart.layout.grid.axis.color,
            transform: `translate(${e.value.style.chart.layout.grid.axis.fontSize}, ${r.value.top + r.value.height / 2}) rotate(-90)`,
            "text-anchor": "middle",
            style: { transition: "none" }
          }, C(e.value.style.chart.layout.grid.axis.yLabel), 9, Ca)) : v("", !0),
          e.value.style.chart.layout.grid.axis.xLabel ? (s(), n("text", {
            key: 1,
            ref_key: "xAxisLabel",
            ref: xe,
            "text-anchor": "middle",
            x: r.value.absoluteWidth / 2,
            y: r.value.absoluteHeight - 3,
            "font-size": e.value.style.chart.layout.grid.axis.fontSize,
            fill: e.value.style.chart.layout.grid.axis.color
          }, C(e.value.style.chart.layout.grid.axis.xLabel), 9, wa)) : v("", !0)
        ]),
        e.value.style.chart.layout.grid.yAxis.dataLabels.show ? (s(), n("g", {
          key: 2,
          ref_key: "scaleLabels",
          ref: pe,
          class: N({ "donut-opacity": !0, "donut-behind": b.value !== null })
        }, [
          (s(!0), n(_, null, $(al.value, (l, u) => (s(), n("g", null, [
            l.value >= F.value.min && l.value <= F.value.max ? (s(), n("line", {
              key: 0,
              x1: r.value.left,
              x2: r.value.left - 5,
              y1: l.y,
              y2: l.y,
              stroke: e.value.style.chart.layout.grid.stroke,
              "stroke-width": e.value.style.chart.layout.grid.strokeWidth
            }, null, 8, Aa)) : v("", !0),
            l.value >= F.value.min && l.value <= F.value.max ? (s(), n("text", {
              key: 1,
              x: r.value.left + e.value.style.chart.layout.grid.yAxis.dataLabels.offsetX - 7,
              y: l.y + e.value.style.chart.layout.grid.yAxis.dataLabels.fontSize / 3,
              "font-size": e.value.style.chart.layout.grid.yAxis.dataLabels.fontSize,
              "text-anchor": "end",
              fill: e.value.style.chart.layout.grid.yAxis.dataLabels.color,
              "font-weight": e.value.style.chart.layout.grid.yAxis.dataLabels.bold ? "bold" : "normal"
            }, C(c(Nl)(l.value) ? c(Be)(
              e.value.style.chart.layout.dataLabels.formatter,
              l.value,
              c(be)({
                p: e.value.style.chart.layout.dataLabels.prefix,
                v: l.value,
                s: e.value.style.chart.layout.dataLabels.suffix,
                r: e.value.style.chart.layout.grid.yAxis.dataLabels.roundingValue
              }),
              { datapoint: l, seriesIndex: u }
            ) : ""), 9, La)) : v("", !0)
          ]))), 256))
        ], 2)) : v("", !0),
        e.value.style.chart.layout.grid.xAxis.dataLabels.show ? (s(), n("g", {
          key: 3,
          ref_key: "timeLabelsEls",
          ref: oe,
          class: N({ "donut-opacity": !0 })
        }, [
          (s(!0), n(_, null, $(h.value.end - h.value.start, (l, u) => (s(), n("g", null, [
            (e.value.style.chart.layout.grid.xAxis.dataLabels.showOnlyFirstAndLast && (u === 0 || u === P.value - 1) || !e.value.style.chart.layout.grid.xAxis.dataLabels.showOnlyFirstAndLast) && w.value[u] && w.value[u].text ? (s(), n("g", Sa, [
              String(w.value[u].text).includes(`
`) ? (s(), n("text", {
                key: 1,
                class: "vue-data-ui-time-label",
                "text-anchor": e.value.style.chart.layout.grid.xAxis.dataLabels.rotation > 0 ? "start" : e.value.style.chart.layout.grid.xAxis.dataLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.layout.grid.xAxis.dataLabels.fontSize,
                fill: e.value.style.chart.layout.grid.xAxis.dataLabels.color,
                transform: `translate(${r.value.left + A.value * u + A.value / 2}, ${r.value.bottom + e.value.style.chart.layout.grid.xAxis.dataLabels.fontSize + e.value.style.chart.layout.grid.xAxis.dataLabels.offsetY}), rotate(${e.value.style.chart.layout.grid.xAxis.dataLabels.rotation})`,
                innerHTML: c(Rl)({
                  content: String(w.value[u].text),
                  fontSize: e.value.style.chart.layout.grid.xAxis.dataLabels.fontSize,
                  fill: e.value.style.chart.layout.grid.xAxis.dataLabels.color,
                  x: 0,
                  y: 0
                })
              }, null, 8, $a)) : (s(), n("text", {
                key: 0,
                class: "vue-data-ui-time-label",
                "text-anchor": e.value.style.chart.layout.grid.xAxis.dataLabels.rotation > 0 ? "start" : e.value.style.chart.layout.grid.xAxis.dataLabels.rotation < 0 ? "end" : "middle",
                "font-size": e.value.style.chart.layout.grid.xAxis.dataLabels.fontSize,
                fill: e.value.style.chart.layout.grid.xAxis.dataLabels.color,
                transform: `translate(${r.value.left + A.value * u + A.value / 2}, ${r.value.bottom + e.value.style.chart.layout.grid.xAxis.dataLabels.fontSize + e.value.style.chart.layout.grid.xAxis.dataLabels.offsetY}), rotate(${e.value.style.chart.layout.grid.xAxis.dataLabels.rotation})`
              }, C(w.value[u].text || ""), 9, _a))
            ])) : v("", !0)
          ]))), 256))
        ], 512)) : v("", !0),
        (s(!0), n(_, null, $(L.value, (l, u) => (s(), n("g", null, [
          e.value.style.chart.layout.line.show && u < L.value.length - 1 && ![l.subtotal, L.value[u + 1].subtotal].includes(null) ? (s(), n("line", {
            key: 0,
            class: N({ "donut-opacity": !0, "donut-behind": b.value !== null }),
            x1: l.x,
            y1: l.y,
            x2: L.value[u + 1].x,
            y2: L.value[u + 1].y,
            stroke: e.value.style.chart.layout.line.stroke,
            "stroke-width": e.value.style.chart.layout.line.strokeWidth,
            "stroke-linecap": "round",
            "stroke-linejoin": "round"
          }, null, 10, za)) : v("", !0),
          l.subtotal !== null ? (s(), n("g", Ta, [
            l.subtotal ? (s(), n("circle", {
              key: 0,
              cx: l.x,
              cy: l.y,
              r: l.activeRadius,
              fill: e.value.style.chart.backgroundColor
            }, null, 8, Pa)) : v("", !0)
          ])) : v("", !0)
        ]))), 256)),
        (s(!0), n(_, null, $(L.value, (l, u) => (s(), n("g", {
          class: N({ "donut-opacity": !0, "donut-behind": u !== b.value && b.value !== null })
        }, [
          l.subtotal ? (s(), n("g", Da, [
            b.value !== null && b.value === u ? (s(), n("g", Ia, [
              (s(!0), n(_, null, $(l.donutHover, (o) => (s(), n("g", null, [
                xt(o) ? (s(), n("path", {
                  key: 0,
                  d: c(El)(o, { x: o.center.endX, y: o.center.endY }, 12, 12, { x: l.x, y: l.y }, !1, 20),
                  stroke: o.color,
                  "stroke-width": "1",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  fill: "none"
                }, null, 8, Oa)) : v("", !0)
              ]))), 256)),
              (s(!0), n(_, null, $(l.donutHover, (o, f) => (s(), n("g", null, [
                xt(o) ? (s(), n("text", {
                  key: 0,
                  "data-cy-hover-label": "",
                  "text-anchor": c(Ot)(o, !0, 0).anchor,
                  x: c(Ot)(o, !0, 9).x,
                  y: c(Fl)(o, 14, 10),
                  fill: e.value.style.chart.layout.grid.yAxis.dataLabels.color,
                  "font-size": 8,
                  "font-weight": "bold"
                }, C(o.name) + ": " + C(ol(o, l.donut)) + " (" + C(o.value === null ? "-" : ht(o.value, o, f)) + ") ", 9, Na)) : v("", !0)
              ]))), 256)),
              k("g", null, [
                k("circle", {
                  cx: l.x,
                  cy: l.y,
                  r: l.hoverRadius,
                  fill: e.value.style.chart.backgroundColor
                }, null, 8, Ra)
              ])
            ])) : v("", !0)
          ])) : v("", !0)
        ], 2))), 256)),
        (s(!0), n(_, null, $(L.value, (l, u) => (s(), n("g", {
          class: N({ "donut-opacity": !0, "donut-behind": u !== b.value && b.value !== null })
        }, [
          l.subtotal !== null ? (s(), n("g", Ea, [
            l.subtotal === 0 ? (s(), n("circle", {
              key: 0,
              cx: l.x,
              cy: l.y,
              r: 3,
              fill: e.value.style.chart.color
            }, null, 8, Fa)) : b.value !== null && b.value === u ? (s(), n("g", Va, [
              (s(!0), n(_, null, $(l.donutHover, (o, f) => (s(), n("path", {
                d: o.arcSlice,
                fill: `${o.color}`,
                "stroke-width": 1,
                stroke: e.value.style.chart.backgroundColor
              }, null, 8, Ma))), 256))
            ])) : (s(), n("g", Ba, [
              (s(!0), n(_, null, $(l.donut, (o, f) => (s(), n("path", {
                d: o.arcSlice,
                fill: `${o.color}`,
                "stroke-width": 0.5,
                stroke: e.value.style.chart.backgroundColor
              }, null, 8, Ua))), 256))
            ]))
          ])) : v("", !0)
        ], 2))), 256)),
        (s(!0), n(_, null, $(L.value, (l, u) => (s(), n("g", {
          class: N({ "donut-opacity": !0, "donut-behind": u !== b.value && b.value !== null || me.value && u !== B.value.index })
        }, [
          l.subtotal !== null && e.value.style.chart.layout.dataLabels.show ? (s(), n("text", {
            key: 0,
            "text-anchor": "middle",
            x: l.x,
            y: b.value === l.index && l.subtotal ? l.y + e.value.style.chart.layout.dataLabels.fontSize / 3 : l.y - l.radius - e.value.style.chart.layout.dataLabels.fontSize + e.value.style.chart.layout.dataLabels.offsetY,
            "font-size": e.value.style.chart.layout.dataLabels.fontSize,
            "font-weight": "bold",
            fill: e.value.style.chart.layout.dataLabels.color
          }, C(ht(l.subtotal, l, u)), 9, Ha)) : v("", !0)
        ], 2))), 256)),
        (s(!0), n(_, null, $(L.value, (l, u) => (s(), n("rect", {
          x: r.value.left + u * A.value,
          y: r.value.top,
          width: A.value,
          height: r.value.height,
          fill: [b.value, Z.value].includes(l.index) ? `url(#hover_${O.value})` : "transparent",
          class: N({ "donut-hover": E.value && l.subtotal && [b.value, Z.value].includes(l.index) }),
          style: {
            pointerEvents: "none"
          }
        }, null, 10, Wa))), 256)),
        (s(!0), n(_, null, $(L.value, (l, u) => (s(), n("rect", {
          "data-cy-trap": "",
          x: r.value.left + u * A.value,
          y: r.value.top,
          width: A.value,
          height: r.value.height,
          fill: "transparent",
          onMouseenter: (o) => ul(l),
          onMouseleave: (o) => yt(l),
          onClick: (o) => gt(l, u),
          class: N({ "donut-hover": E.value && b.value === l.index && l.subtotal })
        }, null, 42, Xa))), 256)),
        m(t.$slots, "svg", {
          svg: {
            ...r.value,
            isPrintingImg: c(Pe) | c(De) | ue.value,
            isPrintingSvg: se.value
          }
        }, void 0, !0)
      ], 14, ra)),
      t.$slots.watermark ? (s(), n("div", Ga, [
        m(t.$slots, "watermark", D(I({ isPrinting: c(Pe) || c(De) || ue.value || se.value })), void 0, !0)
      ])) : v("", !0),
      k("div", {
        ref_key: "chartSlicer",
        ref: et,
        style: Y(`width:100%;background:${e.value.style.chart.backgroundColor}`),
        "data-dom-to-png-ignore": ""
      }, [
        P.value > 1 && e.value.style.chart.zoom.show ? (s(), z(aa, {
          ref_key: "slicerComponent",
          ref: Qe,
          key: `slicer_${Xe.value}`,
          background: e.value.style.chart.zoom.color,
          borderColor: e.value.style.chart.backgroundColor,
          fontSize: e.value.style.chart.zoom.fontSize,
          useResetSlot: e.value.style.chart.zoom.useResetSlot,
          labelLeft: w.value[0] ? w.value[0].text : "",
          labelRight: w.value.at(-1) ? w.value.at(-1).text : "",
          textColor: e.value.style.chart.color,
          inputColor: e.value.style.chart.zoom.color,
          selectColor: e.value.style.chart.zoom.highlightColor,
          max: P.value,
          min: 0,
          valueStart: h.value.start,
          valueEnd: h.value.end,
          start: h.value.start,
          "onUpdate:start": a[0] || (a[0] = (l) => h.value.start = l),
          end: h.value.end,
          "onUpdate:end": a[1] || (a[1] = (l) => h.value.end = l),
          refreshStartPoint: e.value.style.chart.zoom.startIndex !== null ? e.value.style.chart.zoom.startIndex : 0,
          refreshEndPoint: e.value.style.chart.zoom.endIndex !== null ? e.value.style.chart.zoom.endIndex + 1 : P.value,
          enableRangeHandles: e.value.style.chart.zoom.enableRangeHandles,
          enableSelectionDrag: e.value.style.chart.zoom.enableSelectionDrag,
          focusOnDrag: e.value.style.chart.zoom.focusOnDrag,
          focusRangeRatio: e.value.style.chart.zoom.focusRangeRatio,
          isCursorPointer: E.value,
          maxWidth: e.value.style.chart.zoom.maxWidth,
          onReset: ut
        }, {
          "reset-action": g(({ reset: l }) => [
            m(t.$slots, "reset-action", D(I({ reset: l })), void 0, !0)
          ]),
          _: 3
        }, 8, ["background", "borderColor", "fontSize", "useResetSlot", "labelLeft", "labelRight", "textColor", "inputColor", "selectColor", "max", "valueStart", "valueEnd", "start", "end", "refreshStartPoint", "refreshEndPoint", "enableRangeHandles", "enableSelectionDrag", "focusOnDrag", "focusRangeRatio", "isCursorPointer", "maxWidth"])) : v("", !0)
      ], 4),
      k("div", {
        id: `legend-bottom-${O.value}`
      }, null, 8, Ya),
      qe.value ? (s(), z(Sl, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${O.value}` : `#legend-bottom-${O.value}`
      }, [
        k("div", {
          ref_key: "chartLegend",
          ref: Je
        }, [
          e.value.style.chart.legend.show ? (s(), z(la, {
            key: `legend_${je.value}`,
            legendSet: J.value,
            config: sl.value,
            onClickMarker: a[2] || (a[2] = ({ legend: l }) => K(l.uid))
          }, {
            item: g(({ legend: l, index: u }) => [
              k("div", {
                onClick: (o) => K(l.uid),
                style: Y(`opacity:${p.value.includes(l.uid) ? 0.5 : 1}`)
              }, C(l.display), 13, ja)
            ]),
            legendToggle: g(() => [
              J.value.length > 2 && e.value.style.chart.legend.selectAllToggle.show && !c(Le) ? (s(), z(oa, {
                key: 0,
                backgroundColor: e.value.style.chart.legend.selectAllToggle.backgroundColor,
                color: e.value.style.chart.legend.selectAllToggle.color,
                fontSize: e.value.style.chart.legend.fontSize,
                checked: p.value.length > 0,
                onToggle: nl
              }, null, 8, ["backgroundColor", "color", "fontSize", "checked"])) : v("", !0)
            ]),
            _: 1
          }, 8, ["legendSet", "config"])) : v("", !0),
          m(t.$slots, "legend", { legend: J.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : v("", !0),
      t.$slots.source ? (s(), n("div", {
        key: 6,
        ref_key: "source",
        ref: Ke,
        dir: "auto"
      }, [
        m(t.$slots, "source", {}, void 0, !0)
      ], 512)) : v("", !0),
      lt.value && e.value.userOptions.buttons.table ? (s(), z(_l(Q.value.component), $l({ key: 7 }, Q.value.props, {
        ref_key: "tableUnit",
        ref: j,
        onClose: wt
      }), $t({
        content: g(() => [
          (s(), z(c(Ut), {
            key: `table_${Ye.value}`,
            colNames: G.value.colNames,
            head: G.value.head,
            body: G.value.body,
            config: G.value.config,
            title: e.value.table.useDialog ? "" : Q.value.title,
            withCloseButton: !e.value.table.useDialog,
            isCursorPointer: E.value,
            onClose: wt
          }, {
            th: g(({ th: l }) => [
              Fe(C(l.name ?? l), 1)
            ]),
            td: g(({ td: l }) => [
              l.value === null ? (s(), n("span", qa, "-")) : (s(), n("b", Za, C(isNaN(l.value) ? "" : e.value.style.chart.layout.dataLabels.prefix) + C(!isNaN(l.value) && l.value !== null ? Number(l.value.toFixed(e.value.table.td.roundingValue)).toLocaleString() : l) + C(isNaN(l.value) ? "" : e.value.style.chart.layout.dataLabels.suffix), 1)),
              k("span", null, C(l.percentage && !isNaN(l.percentage) ? `(${Number(l.percentage.toFixed(e.value.table.td.roundingPercentage)).toLocaleString()}%)` : ""), 1)
            ]),
            _: 1
          }, 8, ["colNames", "head", "body", "config", "title", "withCloseButton", "isCursorPointer"]))
        ]),
        _: 2
      }, [
        e.value.table.useDialog ? {
          name: "title",
          fn: g(() => [
            Fe(C(Q.value.title), 1)
          ]),
          key: "0"
        } : void 0,
        e.value.table.useDialog ? {
          name: "actions",
          fn: g(() => [
            k("button", {
              tabindex: "0",
              class: "vue-ui-user-options-button",
              onClick: a[3] || (a[3] = (l) => Ne(e.value.userOptions.callbacks.csv)),
              style: Y({ cursor: E.value ? "cursor" : "default" })
            }, [
              zt(c(Bt), {
                name: "fileCsv",
                stroke: Q.value.props.color
              }, null, 8, ["stroke"])
            ], 4)
          ]),
          key: "1"
        } : void 0
      ]), 1040)) : v("", !0),
      e.value.style.chart.dialog.show ? (s(), z(c(He), {
        key: 8,
        ref_key: "dialog",
        ref: Ee,
        onClose: a[4] || (a[4] = (l) => {
          B.value = null, me.value = !1, ae.value = null, b.value = null, Z.value = null;
        }),
        backgroundColor: e.value.style.chart.dialog.backgroundColor,
        color: e.value.style.chart.dialog.color,
        headerBg: e.value.style.chart.dialog.header.backgroundColor,
        headerColor: e.value.style.chart.dialog.header.color,
        isFullscreen: W.value,
        fullscreenParent: R.value
      }, {
        title: g(() => [
          Fe(C(w.value[Number(B.value.index)] ? w.value[Number(B.value.index)].text : ""), 1)
        ]),
        content: g(() => [
          B.value ? (s(), z(c(Gt), {
            key: 0,
            config: Ct.value,
            dataset: kt.value
          }, null, 8, ["config", "dataset"])) : v("", !0)
        ]),
        _: 1
      }, 8, ["backgroundColor", "color", "headerBg", "headerColor", "isFullscreen", "fullscreenParent"])) : v("", !0),
      c(Le) ? (s(), z(Gl, { key: 9 })) : v("", !0)
    ], 46, sa));
  }
}, po = /* @__PURE__ */ ua(Ja, [["__scopeId", "data-v-30771125"]]);
export {
  po as default
};
