// vue-data-ui/dist/vue-ui-bullet-DPoGGtaG.js
import { defineAsyncComponent as X, ref as h, computed as i, onMounted as Je, toRefs as Ke, watch as me, onBeforeUnmount as Qe, openBlock as o, createElementBlock as u, unref as n, normalizeStyle as Y, normalizeClass as ye, createBlock as L, withCtx as f, renderSlot as c, normalizeProps as b, guardReactiveProps as k, createCommentVNode as d, createElementVNode as S, createSlots as Ze, createVNode as et, Fragment as J, renderList as K, toDisplayString as R, Teleport as tt, nextTick as at } from "vue";
import { u as lt } from "./useConfig-CIgyKZ6R.js";
import { j as st, t as pe, x as ot, D as nt, s as rt, k as z, l as U, X as ut, o as be, q as ke, r as it } from "./lib-GwPnJSrY.js";
import { t as ct, u as vt } from "./useResponsive-DfdjqQps.js";
import { u as ht } from "./usePrinter-DiK5R6k-.js";
import { u as dt, B as gt } from "./BaseScanner-BhhkKVoj.js";
import { u as ft } from "./useSvgExport-ByUukOZt.js";
import { u as Q } from "./useNestedProp-BBT0mbHh.js";
import { u as mt } from "./useThemeCheck-D1ZYXUwG.js";
import { u as yt } from "./useUserOptionState-BIvW1Kz7.js";
import { u as pt } from "./useChartAccessibility-9icAAmYg.js";
import bt from "./vue_ui_bullet-Ckqieba5.js";
import kt from "./Legend-1LMMsPxL.js";
import wt from "./Title-BfyrP-jp.js";
import Ct from "./img-BYYzY_QU.js";
import { _ as xt } from "./_plugin-vue_export-helper-CHgC5LLL.js";
const At = ["id"], $t = {
  key: 1,
  ref: "noTitle",
  class: "vue-data-ui-no-title-space",
  style: "height:36px; width: 100%;background:transparent"
}, Ot = ["id"], Lt = ["xmlns", "viewBox"], St = ["width", "height"], _t = { key: 1 }, Pt = ["x", "y", "height", "width", "fill", "stroke"], Bt = ["x", "y", "height", "width", "rx", "fill", "stroke", "stroke-width"], It = ["x", "y", "height", "width", "fill", "stroke", "stroke-width"], Tt = ["x", "y", "font-size", "font-weight", "fill"], Ft = ["x", "y", "height", "width", "rx", "fill", "stroke", "stroke-width"], Ht = { key: 3 }, Mt = ["x", "y", "fill", "font-size", "font-weight"], Rt = { key: 4 }, zt = ["x1", "x2", "y1", "y2", "stroke"], Ut = {
  key: 4,
  class: "vue-data-ui-watermark"
}, Nt = ["id"], Et = {
  key: 0,
  class: "vue-ui-bullet-legend-item",
  dir: "auto"
}, Dt = { style: { "margin-right": "2px" } }, Wt = {
  __name: "vue-ui-bullet",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Object,
      default() {
        return {};
      }
    }
  },
  emits: ["copyAlt"],
  setup(we, { expose: Ce, emit: xe }) {
    const Ae = X(() => import("./PackageVersion-DN9d0ME5.js")), $e = X(() => import("./PenAndPaper-DWV1FMIi.js")), Oe = X(() => import("./UserOptions-CrLqEimB.js")), { vue_ui_bullet: Le } = lt(), { isThemeValid: Se, warnInvalidTheme: _e } = mt(), x = we, Pe = xe, w = h(null), Z = h(null), ee = h(0), te = h(null), ae = h(null), Be = h(0), _ = h(null), I = h(null), le = h(!1), T = h(!1), F = h(!1), A = i({
      get: () => v.value.hasOwnProperty("value"),
      set: (t) => t
    }), P = i(() => v.value.segments ? Array.isArray(v.value.segments) ? v.value.segments.length ? !0 : (console.warn(`VueUiBullet: dataset segments is empty. Provide segments with this datastructure:

segments: [
    {
        name: string;
        from: number;
        to: number;
        color?: string;
    },
    {...}
]
        `), A.value = !1, !1) : (console.warn(`VueUiBullet: dataset segments must be an array of objects with this datastructure:

segments: [
    {
        name: string;
        from: number;
        to: number;
        color?: string;
    },
    {...}
] 
        `), A.value = !1, !1) : (console.warn(`VueUiBullet: dataset segments is empty. Provide segments with this datastructure:

segments: [
    {
        name: string;
        from: number;
        to: number;
        color?: string;
    },
    {...}
]
        `), A.value = !1, !1)), se = i(() => !!e.value.debug), Ie = i(() => {
      const { top: t, right: l, bottom: a, left: r } = e.value.style.chart.padding;
      return {
        top: t,
        right: l,
        bottom: a,
        left: r
      };
    });
    function oe() {
      if (be(v.value) ? (ke({
        componentName: "VueUiBullet",
        type: "dataset",
        debug: se.value
      }), H.value = !0) : P.value ? v.value.segments.forEach((t, l) => {
        it({
          datasetObject: t,
          requiredAttributes: ["name", "from", "to"]
        }).forEach((a) => {
          A.value = !1, ke({
            componentName: "VueUiBullet segment",
            type: "datasetSerieAttribute",
            property: a,
            index: l,
            debug: se.value
          });
        });
      }) : (A.value = !1, H.value = !0), be(v.value) || (H.value = e.value.loading), e.value.responsive) {
        const t = ct(() => {
          const { width: l, height: a } = vt({
            chart: w.value,
            title: e.value.style.chart.title.text ? Z.value : null,
            legend: e.value.style.chart.legend.show ? te.value : null,
            source: ae.value,
            padding: Ie.value
          }), p = (e.value.style.chart.legend.show ? 24 : 0) || 12;
          requestAnimationFrame(() => {
            O.value.width = l, O.value.height = a - p;
          });
        });
        _.value && (I.value && _.value.unobserve(I.value), _.value.disconnect()), _.value = new ResizeObserver(t), I.value = w.value.parentNode, _.value.observe(I.value);
      }
      e.value.style.chart.animation.show && !$.value && ce(v.value.value || 0);
    }
    Je(() => {
      le.value = !0, oe();
    });
    const C = h(st());
    function N() {
      const t = Q({
        userConfig: x.config,
        defaultConfig: Le
      }), l = t.theme;
      if (!l) return t;
      if (!Se.value(t))
        return _e(t), t;
      const a = Q({
        userConfig: bt[l] || x.config,
        defaultConfig: t
      });
      return Q({
        userConfig: x.config,
        defaultConfig: a
      });
    }
    const e = h(N()), ne = i(() => e.value.userOptions.useCursorPointer), Te = i(() => pe({
      defaultConfig: {
        userOptions: { show: !1 },
        style: {
          chart: {
            backgroundColor: "#99999930",
            segments: {
              dataLabels: { show: !1 },
              ticks: {
                stroke: "#8A8A8A"
              }
            },
            valueBar: {
              label: { show: !1 }
            }
          }
        }
      },
      userConfig: e.value.skeletonConfig ?? {}
    })), { loading: $, FINAL_DATASET: v, manualLoading: H } = dt({
      ...Ke(x),
      FINAL_CONFIG: e,
      prepareConfig: N,
      skeletonDataset: x.config?.skeletonDataset ?? {
        value: 100,
        target: 100,
        segments: [
          {
            name: "",
            from: 0,
            to: 33,
            color: "#AAAAAA"
          },
          {
            name: "",
            from: 33,
            to: 66,
            color: "#BABABA"
          },
          {
            name: "",
            from: 66,
            to: 100,
            color: "#CACACA"
          }
        ]
      },
      skeletonConfig: pe({
        defaultConfig: e.value,
        userConfig: Te.value
      })
    }), { userOptionsVisible: E, setUserOptionsVisibility: re, keepUserOptionState: ue } = yt({ config: e.value }), { svgRef: D } = pt({ config: e.value.style.chart.title }), O = h({
      width: e.value.style.chart.width,
      height: e.value.style.chart.height
    }), Fe = i(() => O.value.width), He = i(() => O.value.height);
    me(() => x.config, (t) => {
      $.value || (e.value = N()), E.value = !e.value.userOptions.showOnChartHover, O.value.width = e.value.style.chart.width, O.value.height = e.value.style.chart.height, oe(), ee.value += 1;
    }, { deep: !0 });
    const s = i(() => {
      const t = He.value, l = Fe.value, a = e.value.style.chart.padding.left, r = l - e.value.style.chart.padding.right, p = e.value.style.chart.padding.top, g = t - e.value.style.chart.padding.bottom;
      return {
        height: Math.max(1e-3, t),
        width: Math.max(1e-3, l),
        left: a,
        right: r,
        top: p,
        bottom: g,
        chartWidth: Math.max(1e-3, r - a),
        chartHeight: Math.max(1e-3, g - p)
      };
    }), Me = i(() => {
      if (!P.value) return [];
      const t = [];
      for (let l = 0; l < v.value.segments.length; l += 1)
        t.push(ot(e.value.style.chart.segments.baseColor, l / v.value.segments.length));
      return t;
    }), W = i(() => P.value ? {
      min: Math.min(...v.value.segments.map((t) => t.from)),
      max: Math.max(...v.value.segments.map((t) => t.to))
    } : { min: 0, max: 1 }), m = h(Re());
    me(() => v.value, (t) => {
      t.hasOwnProperty("value") && (H.value = !1), e.value.style.chart.animation.show && !$.value ? ce(t.value || 0) : m.value = t.value || 0;
    }, { deep: !0 });
    function Re() {
      return e.value.style.chart.animation.show && !$.value ? W.value.min : v.value.value || 0;
    }
    const ie = h(null);
    function ce(t) {
      const l = Math.abs(t - m.value) / e.value.style.chart.animation.animationFrames;
      function a() {
        m.value < t ? m.value = Math.min(m.value + l, t) : m.value > t && (m.value = Math.max(m.value - l, t)), m.value !== t && (ie.value = requestAnimationFrame(a));
      }
      a();
    }
    Qe(() => {
      cancelAnimationFrame(ie.value);
    });
    const y = i(() => {
      if (!P.value)
        return [];
      const t = nt(W.value.min, W.value.max, e.value.style.chart.segments.ticks.divisions), l = t.min >= 0 ? 0 : Math.abs(t.min), a = {
        x: s.value.left + (v.value.target + l) / (t.max + l) * s.value.chartWidth - e.value.style.chart.target.width / 2
      }, r = {
        width: (m.value + l) / (t.max + l) * s.value.chartWidth
      }, p = t.ticks.map((g) => ({
        value: g,
        y: s.value.bottom + e.value.style.chart.segments.dataLabels.fontSize + 3 + e.value.style.chart.segments.dataLabels.offsetY,
        x: s.value.left + (g + l) / (t.max + l) * s.value.chartWidth
      }));
      return {
        scale: t,
        target: a,
        value: r,
        ticks: p,
        chunks: v.value.segments.map((g, Ye) => ({
          ...g,
          color: g.color ? rt(g.color) : Me.value[Ye],
          x: s.value.left + s.value.chartWidth * ((g.from + l) / (t.max + l)),
          y: s.value.top,
          height: s.value.chartHeight,
          width: s.value.chartWidth * (Math.abs(g.to - g.from) / (t.max + l))
        }))
      };
    }), V = i(() => !y.value || !y.value.chunks || !y.value.chunks.length ? [] : y.value.chunks.map((t) => {
      const l = z(
        e.value.style.chart.segments.dataLabels.formatter,
        t.from,
        U({
          p: e.value.style.chart.segments.dataLabels.prefix,
          v: t.from,
          s: e.value.style.chart.segments.dataLabels.suffix,
          r: e.value.style.chart.segments.dataLabels.rounding
        })
      ), a = z(
        e.value.style.chart.segments.dataLabels.formatter,
        t.to,
        U({
          p: e.value.style.chart.segments.dataLabels.prefix,
          v: t.to,
          s: e.value.style.chart.segments.dataLabels.suffix,
          r: e.value.style.chart.segments.dataLabels.rounding
        })
      ), r = `${l} — ${a}`;
      return {
        ...t,
        shape: "square",
        value: r,
        display: `${t.name}: ${r}`
      };
    })), ze = i(() => ({
      cy: "bullet-div-legend",
      backgroundColor: "transparent",
      color: e.value.style.chart.legend.color,
      fontSize: e.value.style.chart.legend.fontSize,
      paddingBottom: 6,
      fontWeight: e.value.style.chart.legend.bold ? "bold" : ""
    })), { isPrinting: G, isImaging: j, generatePdf: ve, generateImage: he } = ht({
      elementId: `bullet_${C.value}`,
      fileName: e.value.style.chart.title.text || "vue-ui-bullet",
      options: e.value.userOptions.print
    }), Ue = i(() => e.value.userOptions.show && !e.value.style.chart.title.text), B = h(!1);
    function de(t) {
      B.value = t, Be.value += 1;
    }
    function Ne() {
      return y.value;
    }
    const M = h(!1);
    function q() {
      M.value = !M.value;
    }
    async function Ee({ scale: t = 2 } = {}) {
      if (!w.value) return;
      const { width: l, height: a } = w.value.getBoundingClientRect(), r = l / a, { imageUri: p, base64: g } = await Ct({ domElement: w.value, base64: !0, img: !0, scale: t });
      return {
        imageUri: p,
        base64: g,
        title: e.value.style.chart.title.text,
        width: l,
        height: a,
        aspectRatio: r
      };
    }
    const De = i(() => V.value.map((t) => ({
      ...t,
      name: t.display
    }))), We = i(() => e.value.style.chart.backgroundColor), Ve = i(() => e.value.style.chart.legend), Ge = i(() => e.value.style.chart.title), { exportSvg: je, getSvg: qe } = ft({
      svg: D,
      title: Ge,
      legend: Ve,
      legendItems: De,
      backgroundColor: We
    });
    async function ge({ isCb: t }) {
      F.value = !0, await at();
      try {
        if (t) {
          const { blob: l, url: a, text: r, dataUrl: p } = await qe();
          await Promise.resolve(e.value.userOptions.callbacks.svg({ blob: l, url: a, text: r, dataUrl: p }));
        } else
          await Promise.resolve(je());
      } finally {
        F.value = !1;
      }
    }
    function Xe(t) {
      if (t?.stage === "start") {
        T.value = !0;
        return;
      }
      if (t?.stage === "end") {
        T.value = !1;
        return;
      }
      he();
    }
    async function fe() {
      if (Pe("copyAlt", {
        config: e.value,
        dataset: v.value
      }), !e.value.userOptions.callbacks.altCopy) {
        console.warn("Vue Data UI - A callback must be set for `altCopy` in userOptions.");
        return;
      }
      await Promise.resolve(e.value.userOptions.callbacks.altCopy({
        config: e.value,
        dataset: v.value
      }));
    }
    return Ce({
      getData: Ne,
      getImage: Ee,
      generatePdf: ve,
      generateImage: he,
      generateSvg: ge,
      toggleAnnotator: q,
      toggleFullscreen: de,
      copyAlt: fe
    }), (t, l) => (o(), u("div", {
      ref_key: "bulletChart",
      ref: w,
      class: ye(`vue-data-ui-component vue-ui-bullet ${B.value ? "vue-data-ui-wrapper-fullscreen" : ""}`),
      style: Y(`font-family:${e.value.style.fontFamily};width:100%;background:${e.value.style.chart.backgroundColor};${e.value.responsive ? "height:100%" : ""}`),
      id: `bullet_${C.value}`,
      onMouseenter: l[0] || (l[0] = () => n(re)(!0)),
      onMouseleave: l[1] || (l[1] = () => n(re)(!1))
    }, [
      e.value.userOptions.buttons.annotator ? (o(), L(n($e), {
        key: 0,
        svgRef: n(D),
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        active: M.value,
        isCursorPointer: ne.value,
        onClose: q
      }, {
        "annotator-action-close": f(() => [
          c(t.$slots, "annotator-action-close", {}, void 0, !0)
        ]),
        "annotator-action-color": f(({ color: a }) => [
          c(t.$slots, "annotator-action-color", b(k({ color: a })), void 0, !0)
        ]),
        "annotator-action-draw": f(({ mode: a }) => [
          c(t.$slots, "annotator-action-draw", b(k({ mode: a })), void 0, !0)
        ]),
        "annotator-action-undo": f(({ disabled: a }) => [
          c(t.$slots, "annotator-action-undo", b(k({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-redo": f(({ disabled: a }) => [
          c(t.$slots, "annotator-action-redo", b(k({ disabled: a })), void 0, !0)
        ]),
        "annotator-action-delete": f(({ disabled: a }) => [
          c(t.$slots, "annotator-action-delete", b(k({ disabled: a })), void 0, !0)
        ]),
        _: 3
      }, 8, ["svgRef", "backgroundColor", "color", "active", "isCursorPointer"])) : d("", !0),
      Ue.value ? (o(), u("div", $t, null, 512)) : d("", !0),
      e.value.style.chart.title.text ? (o(), u("div", {
        key: 2,
        ref_key: "chartTitle",
        ref: Z,
        style: "width:100%;background:transparent;"
      }, [
        (o(), L(wt, {
          lineHeight: "1.3rem",
          key: `title_${ee.value}`,
          config: {
            title: {
              cy: "bullet-div-title",
              ...e.value.style.chart.title
            },
            subtitle: {
              cy: "bullet-div-subtitle",
              ...e.value.style.chart.title.subtitle
            }
          }
        }, null, 8, ["config"]))
      ], 512)) : d("", !0),
      S("div", {
        id: `legend-top-${C.value}`
      }, null, 8, Ot),
      e.value.userOptions.show && A.value && (n(ue) || n(E)) ? (o(), L(n(Oe), {
        key: 3,
        ref: "details",
        backgroundColor: e.value.style.chart.backgroundColor,
        color: e.value.style.chart.color,
        isPrinting: n(G),
        isImaging: n(j),
        uid: C.value,
        hasTooltip: !1,
        hasPdf: e.value.userOptions.buttons.pdf,
        hasImg: e.value.userOptions.buttons.img,
        hasSvg: e.value.userOptions.buttons.svg,
        hasXls: !1,
        hasTable: !1,
        hasLabel: !1,
        hasFullscreen: e.value.userOptions.buttons.fullscreen,
        hasAltCopy: e.value.userOptions.buttons.altCopy,
        isFullscreen: B.value,
        chartElement: w.value,
        position: e.value.userOptions.position,
        titles: { ...e.value.userOptions.buttonTitles },
        hasAnnotator: e.value.userOptions.buttons.annotator,
        isAnnotation: M.value,
        callbacks: e.value.userOptions.callbacks,
        printScale: e.value.userOptions.print.scale,
        isCursorPointer: ne.value,
        onToggleFullscreen: de,
        onGeneratePdf: n(ve),
        onGenerateImage: Xe,
        onGenerateSvg: ge,
        onToggleAnnotator: q,
        onCopyAlt: fe,
        style: Y({
          visibility: n(ue) ? n(E) ? "visible" : "hidden" : "visible"
        })
      }, Ze({ _: 2 }, [
        t.$slots.menuIcon ? {
          name: "menuIcon",
          fn: f(({ isOpen: a, color: r }) => [
            c(t.$slots, "menuIcon", b(k({ isOpen: a, color: r })), void 0, !0)
          ]),
          key: "0"
        } : void 0,
        t.$slots.optionPdf ? {
          name: "optionPdf",
          fn: f(() => [
            c(t.$slots, "optionPdf", {}, void 0, !0)
          ]),
          key: "1"
        } : void 0,
        t.$slots.optionImg ? {
          name: "optionImg",
          fn: f(() => [
            c(t.$slots, "optionImg", {}, void 0, !0)
          ]),
          key: "2"
        } : void 0,
        t.$slots.optionSvg ? {
          name: "optionSvg",
          fn: f(() => [
            c(t.$slots, "optionSvg", {}, void 0, !0)
          ]),
          key: "3"
        } : void 0,
        t.$slots.optionFullscreen ? {
          name: "optionFullscreen",
          fn: f(({ toggleFullscreen: a, isFullscreen: r }) => [
            c(t.$slots, "optionFullscreen", b(k({ toggleFullscreen: a, isFullscreen: r })), void 0, !0)
          ]),
          key: "4"
        } : void 0,
        t.$slots.optionAnnotator ? {
          name: "optionAnnotator",
          fn: f(({ toggleAnnotator: a, isAnnotator: r }) => [
            c(t.$slots, "optionAnnotator", b(k({ toggleAnnotator: a, isAnnotator: r })), void 0, !0)
          ]),
          key: "5"
        } : void 0,
        t.$slots.optionAltCopy ? {
          name: "optionAltCopy",
          fn: f(({ altCopy: a }) => [
            c(t.$slots, "optionAltCopy", b(k({ altCopy: a })), void 0, !0)
          ]),
          key: "6"
        } : void 0
      ]), 1032, ["backgroundColor", "color", "isPrinting", "isImaging", "uid", "hasPdf", "hasImg", "hasSvg", "hasFullscreen", "hasAltCopy", "isFullscreen", "chartElement", "position", "titles", "hasAnnotator", "isAnnotation", "callbacks", "printScale", "isCursorPointer", "onGeneratePdf", "style"])) : d("", !0),
      (o(), u("svg", {
        ref_key: "svgRef",
        ref: D,
        xmlns: n(ut),
        class: ye({ "vue-data-ui-fullscreen--on": B.value, "vue-data-ui-fulscreen--off": !B.value, "vue-ui-bullet-svg": !0 }),
        viewBox: `0 0 ${s.value.width} ${s.value.height}`,
        style: Y(`width: 100%; overflow: visible; background:transparent;color:${e.value.style.chart.color}`)
      }, [
        et(n(Ae)),
        t.$slots["chart-background"] ? (o(), u("foreignObject", {
          key: 0,
          x: 0,
          y: 0,
          width: s.value.width,
          height: s.value.height,
          style: {
            pointerEvents: "none"
          }
        }, [
          c(t.$slots, "chart-background", {}, void 0, !0)
        ], 8, St)) : d("", !0),
        P.value ? (o(), u("g", _t, [
          (o(!0), u(J, null, K(y.value.chunks, (a) => (o(), u("rect", {
            x: a.x,
            y: a.y,
            height: a.height,
            width: a.width,
            fill: a.color,
            "stroke-width": 1,
            stroke: e.value.style.chart.backgroundColor,
            style: { transition: "x 0.3s ease-in-out, width 0.3s ease-in-out" }
          }, null, 8, Pt))), 256)),
          !e.value.style.chart.target.onTop && e.value.style.chart.target.show ? (o(), u("rect", {
            key: 0,
            x: y.value.target.x,
            y: s.value.top + (s.value.chartHeight - s.value.chartHeight * e.value.style.chart.target.heightRatio) / 2,
            height: s.value.chartHeight * e.value.style.chart.target.heightRatio,
            width: e.value.style.chart.target.width,
            rx: e.value.style.chart.target.rounded ? e.value.style.chart.target.width / 2 : 0,
            fill: e.value.style.chart.target.color,
            stroke: e.value.style.chart.target.stroke,
            "stroke-width": e.value.style.chart.target.strokeWidth
          }, null, 8, Bt)) : d("", !0),
          S("rect", {
            x: s.value.left,
            y: s.value.top + (s.value.chartHeight - s.value.chartHeight * e.value.style.chart.valueBar.heightRatio) / 2,
            height: s.value.chartHeight * e.value.style.chart.valueBar.heightRatio,
            width: y.value.value.width,
            fill: e.value.style.chart.valueBar.color,
            stroke: e.value.style.chart.valueBar.stroke,
            "stroke-width": e.value.style.chart.valueBar.strokeWidth
          }, null, 8, It),
          e.value.style.chart.valueBar.label.show ? (o(), u("text", {
            key: 1,
            x: s.value.left + y.value.value.width,
            y: s.value.top - 6 + e.value.style.chart.valueBar.label.offsetY,
            "font-size": e.value.style.chart.valueBar.label.fontSize,
            "font-weight": e.value.style.chart.valueBar.label.bold ? "bold" : "normal",
            fill: e.value.style.chart.valueBar.label.color,
            "text-anchor": "middle"
          }, R(n(z)(
            e.value.style.chart.segments.dataLabels.formatter,
            m.value,
            n(U)({
              p: e.value.style.chart.segments.dataLabels.prefix,
              v: m.value,
              s: e.value.style.chart.segments.dataLabels.suffix,
              r: e.value.style.chart.segments.dataLabels.rounding
            })
          )), 9, Tt)) : d("", !0),
          e.value.style.chart.target.onTop && e.value.style.chart.target.show ? (o(), u("rect", {
            key: 2,
            x: y.value.target.x,
            y: s.value.top + (s.value.chartHeight - s.value.chartHeight * e.value.style.chart.target.heightRatio) / 2,
            height: s.value.chartHeight * e.value.style.chart.target.heightRatio,
            width: e.value.style.chart.target.width,
            rx: e.value.style.chart.target.rounded ? e.value.style.chart.target.width / 2 : 0,
            fill: e.value.style.chart.target.color,
            stroke: e.value.style.chart.target.stroke,
            "stroke-width": e.value.style.chart.target.strokeWidth,
            style: { transition: "x 0.3s ease-in-out" }
          }, null, 8, Ft)) : d("", !0),
          e.value.style.chart.segments.dataLabels.show ? (o(), u("g", Ht, [
            (o(!0), u(J, null, K(y.value.ticks, (a) => (o(), u("text", {
              x: a.x,
              y: a.y,
              "text-anchor": "middle",
              fill: e.value.style.chart.segments.dataLabels.color,
              "font-size": e.value.style.chart.segments.dataLabels.fontSize + "px",
              "font-weight": e.value.style.chart.segments.dataLabels.bold ? "bold" : "normal"
            }, R(n(z)(
              e.value.style.chart.segments.dataLabels.formatter,
              a.value,
              n(U)({
                p: e.value.style.chart.segments.dataLabels.prefix,
                v: a.value,
                s: e.value.style.chart.segments.dataLabels.suffix,
                r: e.value.style.chart.segments.dataLabels.rounding
              })
            )), 9, Mt))), 256))
          ])) : d("", !0),
          e.value.style.chart.segments.dataLabels.show && e.value.style.chart.segments.ticks.show ? (o(), u("g", Rt, [
            (o(!0), u(J, null, K(y.value.ticks, (a) => (o(), u("line", {
              x1: a.x,
              x2: a.x,
              y1: s.value.bottom,
              y2: s.value.bottom + 3,
              stroke: e.value.style.chart.segments.ticks.stroke,
              "stroke-width": 1,
              "stroke-linecap": "round"
            }, null, 8, zt))), 256))
          ])) : d("", !0)
        ])) : d("", !0),
        c(t.$slots, "svg", {
          svg: {
            ...s.value,
            isPrintingImg: n(G) | n(j) | T.value,
            isPrintingSvg: F.value
          }
        }, void 0, !0)
      ], 14, Lt)),
      t.$slots.watermark ? (o(), u("div", Ut, [
        c(t.$slots, "watermark", b(k({ isPrinting: n(G) || n(j) || T.value || F.value })), void 0, !0)
      ])) : d("", !0),
      S("div", {
        id: `legend-bottom-${C.value}`
      }, null, 8, Nt),
      le.value ? (o(), L(tt, {
        key: 5,
        to: e.value.style.chart.legend.position === "top" ? `#legend-top-${C.value}` : `#legend-bottom-${C.value}`
      }, [
        S("div", {
          ref_key: "chartLegend",
          ref: te
        }, [
          e.value.style.chart.legend.show ? (o(), L(kt, {
            key: 0,
            clickable: !1,
            legendSet: V.value,
            config: ze.value
          }, {
            item: f(({ legend: a }) => [
              n($) ? d("", !0) : (o(), u("div", Et, [
                S("span", Dt, R(a.name) + ":", 1),
                S("span", null, R(a.value), 1)
              ]))
            ]),
            _: 1
          }, 8, ["legendSet", "config"])) : d("", !0),
          c(t.$slots, "legend", { legend: V.value }, void 0, !0)
        ], 512)
      ], 8, ["to"])) : d("", !0),
      t.$slots.source ? (o(), u("div", {
        key: 6,
        ref_key: "source",
        ref: ae,
        dir: "auto"
      }, [
        c(t.$slots, "source", {}, void 0, !0)
      ], 512)) : d("", !0),
      c(t.$slots, "skeleton", {}, () => [
        n($) ? (o(), L(gt, { key: 0 })) : d("", !0)
      ], !0)
    ], 46, At));
  }
}, na = /* @__PURE__ */ xt(Wt, [["__scopeId", "data-v-e2e0cf53"]]);
export {
  na as default
};
