// vue-data-ui/dist/vue_ui_table_sparkline-yQdijcyJ.js
const o = { userOptions: { show: !1 } }, r = { userOptions: { show: !1 }, title: { backgroundColor: "#1A1A1A", color: "#CCCCCC", subtitle: { color: "#757575" } }, thead: { backgroundColor: "#1A1A1A", color: "#CCCCCC" }, tbody: { backgroundColor: "#1A1A1A", color: "#CCCCCC" } }, l = { title: { backgroundColor: "#1A1A1A", color: "#CCCCCC", subtitle: { color: "#757575" } }, thead: { backgroundColor: "#1A1A1A", color: "#CCCCCC" }, tbody: { backgroundColor: "#1A1A1A", color: "#CCCCCC" } }, c = { title: { backgroundColor: "#FFF8E1", color: "#424242", subtitle: { color: "#757575" } }, thead: { backgroundColor: "#FFF8E1", color: "#424242" }, tbody: { backgroundColor: "#FFF8E1", color: "#424242" } }, C = { title: { backgroundColor: "#1E1E1E", color: "#FFF8E1", subtitle: { color: "#BDBDBD" } }, thead: { backgroundColor: "#1E1E1E", color: "#BDBDBD" }, tbody: { backgroundColor: "#1E1E1E", color: "#BDBDBD" } }, t = { title: { backgroundColor: "#1A1A1A", color: "#66CC66", subtitle: { color: "#99AA99" } }, thead: { backgroundColor: "#1A1A1A", color: "#99AA99" }, tbody: { backgroundColor: "#1A1A1A", color: "#99AA99" } }, a = { title: { backgroundColor: "#fbfafa", color: "#8A9892", subtitle: { color: "#99AA99" } }, thead: { backgroundColor: "#fbfafa", color: "#8A9892" }, tbody: { backgroundColor: "#fbfafa", color: "#A0AC94" } }, b = { title: { backgroundColor: "#f6f6fb", color: "#50606C", subtitle: { color: "#718890" } }, thead: { backgroundColor: "#f6f6fb", color: "#50606C" }, tbody: { backgroundColor: "#f6f6fb", color: "#50606C" } }, A = {
  default: {},
  minimal: o,
  minimalDark: r,
  dark: l,
  celebration: c,
  celebrationNight: C,
  hack: t,
  zen: a,
  concrete: b
};
export {
  c as celebration,
  C as celebrationNight,
  b as concrete,
  l as dark,
  A as default,
  t as hack,
  o as minimal,
  r as minimalDark,
  a as zen
};
