// vue-data-ui/dist/vue_ui_relation_circle-CCroen7h.js
const o = { userOptions: { show: !1 } }, l = { userOptions: { show: !1 }, style: { backgroundColor: "#1A1A1A", color: "#CCCCCC", labels: { color: "#CCCCCC" }, circle: { stroke: "#5A5A5A" }, plot: { color: "#1A1A1A" }, title: { color: "#CCCCCC", subtitle: { color: "#757575" } } } }, r = { style: { backgroundColor: "#1A1A1A", color: "#CCCCCC", labels: { color: "#CCCCCC" }, circle: { stroke: "#5A5A5A" }, plot: { color: "#1A1A1A" }, title: { color: "#CCCCCC", subtitle: { color: "#757575" } } } }, c = { style: { backgroundColor: "#FFF8E1", color: "#424242", labels: { color: "#424242" }, circle: { stroke: "#5D403780" }, plot: { color: "#5D4037" }, title: { color: "#424242", subtitle: { color: "#757575" } } } }, t = { style: { backgroundColor: "#1E1E1E", color: "#BDBDBD", labels: { color: "#BDBDBD" }, circle: { stroke: "#5D403780" }, plot: { color: "#5D4037" }, title: { color: "#FFF8E1", subtitle: { color: "#BDBDBD" } } } }, e = { style: { backgroundColor: "#1A1A1A", color: "#99AA99", labels: { color: "#AACCAA" }, circle: { stroke: "#333333" }, plot: { color: "#66CC66" }, title: { color: "#66CC66", subtitle: { color: "#99AA99" } } } }, C = { style: { backgroundColor: "#fbfafa", color: "#8A9892", labels: { color: "#A0AC94" }, circle: { stroke: "transparent" }, plot: { color: "#C09E85" }, title: { color: "#8A9892", subtitle: { color: "#99AA99" } } } }, s = { style: { backgroundColor: "#f6f6fb", color: "#50606C", labels: { color: "#61747E" }, circle: { stroke: "#D2D7E0" }, plot: { color: "#97ACB7" }, title: { color: "#50606C", subtitle: { color: "#718890" } } } }, A = {
  default: {},
  minimal: o,
  minimalDark: l,
  dark: r,
  celebration: c,
  celebrationNight: t,
  hack: e,
  zen: C,
  concrete: s
};
export {
  c as celebration,
  t as celebrationNight,
  s as concrete,
  r as dark,
  A as default,
  e as hack,
  o as minimal,
  l as minimalDark,
  C as zen
};
