// vue-data-ui/dist/vue_ui_spark_trend-3BjrSprO.js
const o = { style: { arrow: { colors: { positive: "#7FA09B", negative: "#9D7D92" } } } }, e = { style: { backgroundColor: "#1A1A1A", dataLabel: { color: "#CCCCCC" }, trendLabel: { color: "#CCCCCC" }, arrow: { colors: { positive: "#7FA09B", negative: "#9D7D92" } } } }, l = { style: { backgroundColor: "#1A1A1A", dataLabel: { color: "#CCCCCC" }, trendLabel: { color: "#CCCCCC" } } }, a = { style: { backgroundColor: "#FFF8E1", line: { smooth: !1 }, dataLabel: { color: "#424242" }, trendLabel: { color: "#424242" } } }, r = { style: { backgroundColor: "#1E1E1E", line: { smooth: !1 }, dataLabel: { color: "#BDBDBD" }, trendLabel: { color: "#BDBDBD" } } }, t = { style: { backgroundColor: "#1A1A1A", line: { stroke: "#00FF00", smooth: !1 }, dataLabel: { color: "#AACCAA" }, trendLabel: { color: "#99CC99" }, arrow: { colors: { positive: "#66CC66", negative: "#196619" } } } }, C = { style: { backgroundColor: "#fbfafa", line: { stroke: "#B9B99D" }, dataLabel: { color: "#8A9892" }, trendLabel: { color: "#99CC99" }, arrow: { colors: { positive: "#B9B99D", negative: "#C09E85", neutral: "#A5AFC1" } } } }, n = { style: { backgroundColor: "#f6f6fb", line: { stroke: "#4A6A75" }, dataLabel: { color: "#61747E" }, trendLabel: { color: "#8A9892" }, arrow: { colors: { positive: "#4A6A75", negative: "#C09E85", neutral: "#ABB1BC" } } } }, c = {
  default: {},
  minimal: o,
  minimalDark: e,
  dark: l,
  celebration: a,
  celebrationNight: r,
  hack: t,
  zen: C,
  concrete: n
};
export {
  a as celebration,
  r as celebrationNight,
  n as concrete,
  l as dark,
  c as default,
  t as hack,
  o as minimal,
  e as minimalDark,
  C as zen
};
