// vue-data-ui/dist/vue_ui_thermometer-BTmjRBH8.js
const o = { userOptions: { show: !1 } }, t = { userOptions: { show: !1 }, style: { chart: { backgroundColor: "#1A1A1A", color: "#CCCCCC", graduations: { stroke: "#2A2A2A" }, label: { color: "#CCCCCC" } }, title: { color: "#CCCCCC", subtitle: { color: "#757575" } } } }, r = { style: { chart: { backgroundColor: "#1A1A1A", color: "#CCCCCC", graduations: { stroke: "#2A2A2A" }, label: { color: "#CCCCCC" } }, title: { color: "#CCCCCC", subtitle: { color: "#757575" } } } }, l = { style: { chart: { backgroundColor: "#FFF8E1", color: "#424242", graduations: { stroke: "#FFF8E1", strokeWidth: 1, gradient: { show: !0 } }, label: { color: "#424242" } }, title: { color: "#424242", subtitle: { color: "#757575" } } } }, e = { style: { chart: { backgroundColor: "#1E1E1E", color: "#BDBDBD", graduations: { stroke: "#5D4037", strokeWidth: 1, gradient: { show: !0 } }, label: { color: "#BDBDBD" } }, title: { color: "#FFF8E1", subtitle: { color: "#BDBDBD" } } } }, c = { style: { chart: { backgroundColor: "#1A1A1A", color: "#99AA99", graduations: { stroke: "#1A1A1A", strokeWidth: 1, gradient: { show: !0 } }, label: { color: "#66CC66" } }, title: { color: "#66CC66", subtitle: { color: "#99AA99" } } } }, s = { style: { chart: { backgroundColor: "#fbfafa", color: "#8A9892", graduations: { show: !1, gradient: { show: !0 } }, label: { color: "#B9B99D" } }, title: { color: "#8A9892", subtitle: { color: "#99AA99" } } } }, a = { style: { chart: { backgroundColor: "#f6f6fb", color: "#50606C", graduations: { stroke: "#1A1A1A", gradient: { show: !0 } }, label: { color: "#61747E" } }, title: { color: "#50606C", subtitle: { color: "#718890" } } } }, C = {
  default: {},
  minimal: o,
  minimalDark: t,
  dark: r,
  celebration: l,
  celebrationNight: e,
  hack: c,
  zen: s,
  concrete: a
};
export {
  l as celebration,
  e as celebrationNight,
  a as concrete,
  r as dark,
  C as default,
  c as hack,
  o as minimal,
  t as minimalDark,
  s as zen
};
