// vue-data-ui/dist/vue_ui_sparkgauge-KnjASOZs.js
const o = { style: { colors: { min: "#9D7D92", max: "#678681" } } }, t = { style: { background: "#1A1A1A", title: { color: "#CCCCCC" }, dataLabel: { color: "#CCCCCC" }, colors: { min: "#9D7D92", max: "#678681" }, gutter: { color: "#3A3A3A" } } }, l = { style: { background: "#1A1A1A", title: { color: "#CCCCCC" }, dataLabel: { color: "#CCCCCC" }, gutter: { color: "#3A3A3A" } } }, c = { style: { background: "#FFF8E1", title: { color: "#424242" }, dataLabel: { color: "#424242" }, colors: { min: "#F57C00", max: "#D32F2F" }, gutter: { color: "#5D403740" } } }, r = { style: { background: "#1E1E1E", title: { color: "#BDBDBD" }, dataLabel: { color: "#BDBDBD" }, colors: { min: "#F57C00", max: "#D32F2F" }, gutter: { color: "#5D403740" } } }, a = { style: { background: "#1A1A1A", title: { color: "#66CC66" }, dataLabel: { color: "#AACCAA" }, colors: { min: "#334C33", max: "#33CC33" }, track: { color: "#33CC33" }, gutter: { color: "#333333" } } }, e = { style: { background: "#fbfafa", title: { color: "#8A9892" }, dataLabel: { color: "#A0AC94" }, colors: { min: "#C09E85", max: "#B9B99D" }, track: { color: "#B9B99D" }, gutter: { color: "#f5f0eb" } } }, C = { style: { background: "#f6f6fb", title: { color: "#50606C" }, dataLabel: { color: "#61747E" }, colors: { min: "#92B0AB", max: "#4A6A75" }, track: { color: "#4A6A75" }, gutter: { color: "#e1e5e8" } } }, n = {
  default: {},
  minimal: o,
  minimalDark: t,
  dark: l,
  celebration: c,
  celebrationNight: r,
  hack: a,
  zen: e,
  concrete: C
};
export {
  c as celebration,
  r as celebrationNight,
  C as concrete,
  l as dark,
  n as default,
  a as hack,
  o as minimal,
  t as minimalDark,
  e as zen
};
